---
title: "使用leafcutter进行差异可变剪切分析"
date: 2025-02-12T00:30:01Z
draft: ["false"]
tags: [
  "fetched",
  "小明的数据分析笔记本"
]
categories: ["Acdemic"]
---
使用leafcutter进行差异可变剪切分析 by 小明的数据分析笔记本
------
<div><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com"><h2 data-tool="mdnice编辑器"><span></span><span>leafcutter对应的论文</span><span></span><span> </span></h2><p data-tool="mdnice编辑器">https://www.nature.com/articles/s41588-017-0004-9</p><blockquote data-tool="mdnice编辑器"><span></span><h1><span></span><span><span>Annotation-free quantification of RNA splicing using LeafCutter</span></span><span></span></h1></blockquote><h2 data-tool="mdnice编辑器"><span></span><span>对应的github主页</span><span></span><span> </span></h2><p data-tool="mdnice编辑器">https://github.com/davidaknowles/leafcutter</p><h2 data-tool="mdnice编辑器"><span></span><span>帮助文档主页</span><span></span><span> </span></h2><p data-tool="mdnice编辑器">https://davidaknowles.github.io/leafcutter/index.html</p><p data-tool="mdnice编辑器">文档里写python是2.7，R语言是3.6。版本比较老，所以安装过程比较麻烦</p><p data-tool="mdnice编辑器">文档里还提到了regtools</p><p data-tool="mdnice编辑器">对应的论文</p><blockquote data-tool="mdnice编辑器"><span></span><h1><span></span><span><span>Integrated analysis of genomic and transcriptomic data for the discovery of splice-associated variants in cancer</span></span><span></span></h1></blockquote><p data-tool="mdnice编辑器">https://www.nature.com/articles/s41467-023-37266-6</p><p data-tool="mdnice编辑器">这个工具对应的文档</p><p data-tool="mdnice编辑器">https://regtools.readthedocs.io/en/latest/</p><p data-tool="mdnice编辑器">安装需要自己编译，还好编译的过程没有报错</p><h1 data-tool="mdnice编辑器"><span></span><span>用拟南芥的转录组数据做测试</span><span></span></h1><p data-tool="mdnice编辑器">下载原始数据</p><pre data-tool="mdnice编辑器"><span></span><code>conda activate kingfisher<br>python kingfisher_download.py ERR.list<br></code></pre><h2 data-tool="mdnice编辑器"><span></span><span>第一步参考基因组构建索引</span><span></span><span> </span></h2><pre data-tool="mdnice编辑器"><span></span><code>mkdir atSTARindex<br>time ~/anaconda3/envs/rnaseq/bin/STAR --runMode genomeGenerate \<br>--genomeDir atSTARindex/ \<br>--genomeFastaFiles \<br>../../pan.genome/at.nc/version2.5.2019-10-09/minimap2.syri/00.ref/at.col0.chr.fna \<br>--sjdbGTFfile \<br>../../pan.genome/at.nc/version2.5.2019-10-09/minimap2.syri/00.ref/at.col0.chr.gtf \<br>--sjdbOverhang 100 --genomeSAindexNbases 12<br><br>###1个G的基因组用时大约20分钟<br>###拟南芥基因组3分钟<br></code></pre><h2 data-tool="mdnice编辑器"><span></span><span>第二步转录组数据与参考基因组比对</span><span></span><span> </span></h2><pre data-tool="mdnice编辑器"><span></span><code>mkdir 01.star.output.bam<br><br>~/anaconda3/envs/rnaseq/bin/STAR --genomeDir atSTARindex/ \<br>--twopassMode Basic --outSAMstrandField intronMotif \<br>--readFilesCommand zcat --outSAMtype BAM SortedByCoordinate \<br>--runThreadN 12 --outFileNamePrefix 01.star.output.bam/ERR1886187_ \<br>--readFilesIn 00.raw.data/ERR1886187_1.fastq.gz 00.raw.data/ERR1886187_2.fastq.gz<br><br>## 这个比对只要5分钟，但是之前用很小的数据不知道为啥比对时间比较长<br><br>~/anaconda3/envs/rnaseq/bin/STAR --genomeDir atSTARindex/ \<br>--twopassMode Basic --outSAMstrandField intronMotif \<br>--readFilesCommand zcat --outSAMtype BAM SortedByCoordinate \<br>--runThreadN 12 --outFileNamePrefix 01.star.output.bam/ERR1886193_ \<br>--readFilesIn 00.raw.data/ERR1886193_1.fastq.gz 00.raw.data/ERR1886193_2.fastq.gz<br><br>~/anaconda3/envs/rnaseq/bin/STAR --genomeDir atSTARindex/ --twopassMode Basic --outSAMstrandField intronMotif --readFilesCommand zcat --outSAMtype BAM SortedByCoordinate --runThreadN 12 --outFileNamePrefix 01.star.output.bam/ERR1886283_ --readFilesIn 00.raw.data/ERR1886283_1.fastq.gz 00.raw.data/ERR1886283_2.fastq.gz<br><br>~/anaconda3/envs/rnaseq/bin/STAR --genomeDir atSTARindex/ --twopassMode Basic --outSAMstrandField intronMotif --readFilesCommand zcat --outSAMtype BAM SortedByCoordinate --runThreadN 12 --outFileNamePrefix 01.star.output.bam/ERR1886289_ --readFilesIn 00.raw.data/ERR1886289_1.fastq.gz 00.raw.data/ERR1886289_2.fastq.gz<br><br>~/anaconda3/envs/rnaseq/bin/STAR --genomeDir atSTARindex/ --twopassMode Basic --outSAMstrandField intronMotif --readFilesCommand zcat --outSAMtype BAM SortedByCoordinate --runThreadN 12 --outFileNamePrefix 01.star.output.bam/ERR1886352_ --readFilesIn 00.raw.data/ERR1886352_1.fastq.gz 00.raw.data/ERR1886352_2.fastq.gz<br><br>~/anaconda3/envs/rnaseq/bin/STAR --genomeDir atSTARindex/ --twopassMode Basic --outSAMstrandField intronMotif --readFilesCommand zcat --outSAMtype BAM SortedByCoordinate --runThreadN 12 --outFileNamePrefix 01.star.output.bam/ERR1886358_ --readFilesIn 00.raw.data/ERR1886358_1.fastq.gz 00.raw.data/ERR1886358_2.fastq.gz<br><br><br><br>#~/anaconda3/envs/rnaseq/bin/STAR --genomeDir atchr5index/ --twopassMode Basic \<br>#--outSAMstrandField intronMotif \<br>#--readFilesCommand zcat --outSAMtype BAM Unsorted \<br>#--runThreadN 8 --outFileNamePrefix 02.star.output.bam/ERR1886187_ \<br>#--readFilesIn exampleDataNatureRNAseq/01.clean.reads/ERR1886187_1.fq.gz \<br>#exampleDataNatureRNAseq/01.clean.reads/ERR1886187_2.fq.gz<br><br>## 这个比对为什么这么慢呢？很小的数据<br><br>#~/anaconda3/envs/rnaseq/bin/STAR --genomeDir atchr5index/ --twopassMode Basic \<br>#--outSAMstrandField intronMotif \<br>#--readFilesCommand zcat --outSAMtype BAM Unsorted \<br>#--runThreadN 8 --outFileNamePrefix 02.star.output.bam/ERR1886193_ \<br>#--readFilesIn exampleDataNatureRNAseq/01.clean.reads/ERR1886193_1.fq.gz \<br>#exampleDataNatureRNAseq/01.clean.reads/ERR1886193_2.fq.gz<br></code></pre><h2 data-tool="mdnice编辑器"><span></span><span>第三步是 bam文件转换为junc<span></span></span><span></span><span> </span></h2><p data-tool="mdnice编辑器">这一步是在干啥暂时没有看明白，这一步需要用到 regtools</p><pre data-tool="mdnice编辑器"><span></span><code>#~/anaconda3/envs/syri/bin/samtools sort -@ 8 -O BAM 02.star.output.bam/ERR1886187_Aligned.out.bam -o ERR1886187.sorted.bam<br>## 这里还是需要把bam文件排序，要不然samtools index会报错，在STAR比对的时候可以直接设置bam文件排序<br><br>#~/anaconda3/envs/syri/bin/samtools sort -@ 8 -O BAM 02.star.output.bam/ERR1886193_Aligned.out.bam -o ERR1886193.sorted.bam<br>#~/anaconda3/envs/syri/bin/samtools index ERR1886193.sorted.bam<br><br>#~/biotools/regtools/build/regtools junctions extract -s RF -a 8 -m 50 -M 500000 ERR1886187.sorted.bam -o ERR1886187.junc<br><br>## 这里-s参数需要用哪个选项暂时不太确定<br>## -a参数的作用暂时不太明白<br>#~/biotools/regtools/build/regtools junctions extract -s RF -a 8 -m 50 -M 500000 ERR1886193.sorted.bam -o ERR1886193.junc<br><br><br>~/anaconda3/envs/syri/bin/samtools index 01.star.output.bam/ERR1886187_Aligned.sortedByCoord.out.bam<br>~/anaconda3/envs/syri/bin/samtools index 01.star.output.bam/ERR1886193_Aligned.sortedByCoord.out.bam<br>~/anaconda3/envs/syri/bin/samtools index 01.star.output.bam/ERR1886283_Aligned.sortedByCoord.out.bam<br>~/anaconda3/envs/syri/bin/samtools index 01.star.output.bam/ERR1886289_Aligned.sortedByCoord.out.bam<br>~/anaconda3/envs/syri/bin/samtools index 01.star.output.bam/ERR1886352_Aligned.sortedByCoord.out.bam<br>~/anaconda3/envs/syri/bin/samtools index 01.star.output.bam/ERR1886358_Aligned.sortedByCoord.out.bam<br><br><br>~/biotools/regtools/build/regtools junctions extract -s FR -a 8 -m 50 -M 500000 01.star.output.bam/ERR1886187_Aligned.sortedByCoord.out.bam -o 02.regtools.output.junc/ERR1886187.junc<br>~/biotools/regtools/build/regtools junctions extract -s FR -a 8 -m 50 -M 500000 01.star.output.bam/ERR1886193_Aligned.sortedByCoord.out.bam -o 02.regtools.output.junc/ERR1886193.junc<br>~/biotools/regtools/build/regtools junctions extract -s FR -a 8 -m 50 -M 500000 01.star.output.bam/ERR1886283_Aligned.sortedByCoord.out.bam -o 02.regtools.output.junc/ERR1886283.junc<br>~/biotools/regtools/build/regtools junctions extract -s FR -a 8 -m 50 -M 500000 01.star.output.bam/ERR1886289_Aligned.sortedByCoord.out.bam -o 02.regtools.output.junc/ERR1886289.junc<br>~/biotools/regtools/build/regtools junctions extract -s FR -a 8 -m 50 -M 500000 01.star.output.bam/ERR1886352_Aligned.sortedByCoord.out.bam -o 02.regtools.output.junc/ERR1886352.junc<br>~/biotools/regtools/build/regtools junctions extract -s FR -a 8 -m 50 -M 500000 01.star.output.bam/ERR1886358_Aligned.sortedByCoord.out.bam -o 02.regtools.output.junc/ERR1886358.junc<br><br><br><br>~/biotools/regtools-0.5.0/build/regtools junctions extract -s 0 -a 8 -i 50 -I 500000 01.star.output.bam/ERR1886358_Aligned.sortedByCoord.out.bam -o 02.regtools.output.junc/ERR1886358_01.junc<br>~/biotools/regtools-0.5.0/build/regtools junctions extract -s 0 -a 8 -i 50 -I 500000 01.star.output.bam/ERR1886187_Aligned.sortedByCoord.out.bam -o 02.regtools.output.junc/ERR1886187_01.junc<br></code></pre><p data-tool="mdnice编辑器">regtools 这个工具需要指定转录组文库的类型</p><p data-tool="mdnice编辑器">比如是 RF FR 链特异性文库之类的</p><p data-tool="mdnice编辑器">https://www.biostars.org/p/344264/</p><p data-tool="mdnice编辑器">参考一下这个链接</p><figure data-tool="mdnice编辑器"><img alt="" data-imgfileid="100014884" data-ratio="1.2878953107960742" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/t1wZDoUyFk6I6hjErxb4J4v5tb99HqGMjpn6sRCicibBXPaDRJcuGFLFSHib5O7yXSBEK3IElicqiaAVN1fMqMxPOaw/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="917" src="https://mmbiz.qpic.cn/sz_mmbiz_png/t1wZDoUyFk6I6hjErxb4J4v5tb99HqGMjpn6sRCicibBXPaDRJcuGFLFSHib5O7yXSBEK3IElicqiaAVN1fMqMxPOaw/640?wx_fmt=png&amp;from=appmsg"></figure><p data-tool="mdnice编辑器">这个我没有看明白</p><figure data-tool="mdnice编辑器"><img alt="" data-imgfileid="100014883" data-ratio="0.16843501326259946" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/t1wZDoUyFk6I6hjErxb4J4v5tb99HqGMObiadzTCDHruDjia3fsCwvMeWwibTnHpO0U1Y9dn2Hpf4lfbrR2p87hmA/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1508" src="https://mmbiz.qpic.cn/sz_mmbiz_png/t1wZDoUyFk6I6hjErxb4J4v5tb99HqGMObiadzTCDHruDjia3fsCwvMeWwibTnHpO0U1Y9dn2Hpf4lfbrR2p87hmA/640?wx_fmt=png&amp;from=appmsg"></figure><p data-tool="mdnice编辑器">这里有句话 ，常规的illumina paired-end 转录组文库都是FR类型？</p><p data-tool="mdnice编辑器">暂时没有想明白这个问题，如果有大佬懂这个问题欢迎留言解释一下这个问题</p><h2 data-tool="mdnice编辑器"><span></span><span>第四步是内含子聚类 intron clustering</span><span></span><span> </span></h2><pre data-tool="mdnice编辑器"><span></span><code>ls 02.regtools.output.junc/*.junc &gt; juncfiles.txt <br>conda activate leafcutter<br>python ~/biotools/leafcutter/clustering/leafcutter_cluster_regtools.py -j juncfiles.txt -m 5 -o testAT -l 500000<br><br></code></pre><p data-tool="mdnice编辑器">这里的输出一直是空的，在这里搞了好长时间没有搞明白</p><p data-tool="mdnice编辑器">原来对染色体名字是有要求的 只能是 chr1 chr2 chr3 这种形式 我的染色体是 Chr1 C是大写</p><figure data-tool="mdnice编辑器"><img alt="" data-imgfileid="100014880" data-ratio="0.24041297935103245" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/t1wZDoUyFk6I6hjErxb4J4v5tb99HqGMbDeFswhvAIt8TVyo3XLg8kb0bEj0bjzrD117iaiaLU8LpJjkvKVVxOfA/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="678" src="https://mmbiz.qpic.cn/sz_mmbiz_png/t1wZDoUyFk6I6hjErxb4J4v5tb99HqGMbDeFswhvAIt8TVyo3XLg8kb0bEj0bjzrD117iaiaLU8LpJjkvKVVxOfA/640?wx_fmt=png&amp;from=appmsg"></figure><p data-tool="mdnice编辑器">一直是0 clusters</p><p data-tool="mdnice编辑器">正常输出应该是</p><figure data-tool="mdnice编辑器"><img alt="" data-imgfileid="100014881" data-ratio="0.19951040391676866" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/t1wZDoUyFk6I6hjErxb4J4v5tb99HqGMS6iamPk1hTFKnhR2US4PMIrWBia4egNQetyL0XDOpDEAZ01M1iazVbxfg/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="817" src="https://mmbiz.qpic.cn/sz_mmbiz_png/t1wZDoUyFk6I6hjErxb4J4v5tb99HqGMS6iamPk1hTFKnhR2US4PMIrWBia4egNQetyL0XDOpDEAZ01M1iazVbxfg/640?wx_fmt=png&amp;from=appmsg"></figure><pre data-tool="mdnice编辑器"><span></span><code>sed -i "s/Chr/chr/g" 02.regtools.output.junc/ERR1886187.junc<br>sed -i "s/Chr/chr/g" 02.regtools.output.junc/ERR1886193.junc<br>sed -i "s/Chr/chr/g" 02.regtools.output.junc/ERR1886283.junc<br>sed -i "s/Chr/chr/g" 02.regtools.output.junc/ERR1886289.junc<br>sed -i "s/Chr/chr/g" 02.regtools.output.junc/ERR1886352.junc<br>sed -i "s/Chr/chr/g" 02.regtools.output.junc/ERR1886358.junc<br><br>ls 02.regtools.output.junc/*.junc &gt; juncfiles.txt<br><br>python ~/biotools/leafcutter/clustering/leafcutter_cluster_regtools.py -j juncfiles.txt -m 5 -o testAT -l 500000<br><br>less -S testAT_perind_numers.counts.gz<br></code></pre><figure data-tool="mdnice编辑器"><img alt="" data-imgfileid="100014882" data-ratio="0.5465288035450517" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/t1wZDoUyFk6I6hjErxb4J4v5tb99HqGMk6MXkM7lQE9sYKuD8gsNxibSj7T859OenBZXGMLBlrcqdnQO7RsEltw/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="677" src="https://mmbiz.qpic.cn/sz_mmbiz_png/t1wZDoUyFk6I6hjErxb4J4v5tb99HqGMk6MXkM7lQE9sYKuD8gsNxibSj7T859OenBZXGMLBlrcqdnQO7RsEltw/640?wx_fmt=png&amp;from=appmsg"></figure><p data-tool="mdnice编辑器">我安装的R语言版本是3.6.3，差异可变剪切分析的时候又遇到报错</p><pre data-tool="mdnice编辑器"><span></span><code>Loading required package: Rcpp<br>Error: package or namespace load failed for ‘leafcutter’:<br> package ‘leafcutter’ was installed by an R version with different internals; it needs to be reinstalled for use with this R version<br></code></pre><p data-tool="mdnice编辑器">这个R包没有安装成功，搜索了了一下，有人遇到同样的错误</p><p data-tool="mdnice编辑器">https://github.com/davidaknowles/leafcutter/issues/87</p><p data-tool="mdnice编辑器">这个里提到了一个方法，我试了也没有成功</p><p data-tool="mdnice编辑器">这个都是count值，差异分析是否可以直接用DESeq2，暂时没有搞明白，再仔细看看论文</p><p data-tool="mdnice编辑器">这个包里还提供了一个脚本可以把上面的输出直接转换成可以用于做 splicing QTL的格式</p><pre data-tool="mdnice编辑器"><span></span><code>python ~/biotools/leafcutter/scripts/prepare_phenotype_table.py testAT_perind_numers.counts.gz -p 10<br></code></pre><p data-tool="mdnice编辑器">报错</p><p data-tool="mdnice编辑器">Parsing chromosome names...</p><p data-tool="mdnice编辑器">Starting...</p><p data-tool="mdnice编辑器">Traceback (most recent call last):</p><p data-tool="mdnice编辑器">File "/home/myan/biotools/leafcutter/scripts/prepare_phenotype_table.py", line 197, in<module></module></p><pre data-tool="mdnice编辑器"><code>main(args[0], get_chromosomes(args[0]), get_blacklist_chromosomes(options.cbl), int(options.npcs) )<br></code></pre><p data-tool="mdnice编辑器">File "/home/myan/biotools/leafcutter/scripts/prepare_phenotype_table.py", line 86, in main</p><pre data-tool="mdnice编辑器"><code>chrom = dic['chrom']#.replace("chr",'')<br></code></pre><p data-tool="mdnice编辑器">KeyError: 'chrom'</p><p data-tool="mdnice编辑器">这个python脚本的报错应该能解决，今天不搞了，安装这个软件搞了好长时间，最后也没搞定</p><p data-tool="mdnice编辑器"><strong>欢迎大家关注我的公众号</strong></p><p data-tool="mdnice编辑器"><strong>小明的数据分析笔记本</strong></p><section><mp-common-profile data-pluginname="mpprofile" data-id="MzI3NzQ3MTcxMg==" data-headimg="http://mmbiz.qpic.cn/mmbiz_png/t1wZDoUyFk5t1sOnM0iabvBhnfIj5YpyqrMib0E1MGCd9ibcYxaOPZd0GWhQBDvK2BPEwsicQxd6y5MHLfphnwHnow/0?wx_fmt=png" data-nickname="小明的数据分析笔记本" data-alias="" data-signature="分享R语言和python在生物信息领域做数据分析和数据可视化的简单小例子；偶尔会分享一些组学数据处理相关的内容" data-from="0" data-is_biz_ban="0" data-service_type="1"></mp-common-profile></section><p data-tool="mdnice编辑器"><strong></strong></p><blockquote data-tool="mdnice编辑器"><span></span><p>小明的数据分析笔记本 公众号 主要分享：1、R语言和python做数据分析和数据可视化的简单小例子；2、园艺植物相关转录组学、基因组学、群体遗传学文献阅读笔记；3、生物信息学入门学习资料及自己的学习笔记！</p></blockquote><figure data-tool="mdnice编辑器"><img alt="" data-imgfileid="100014889" data-ratio="0.425531914893617" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/t1wZDoUyFk6I6hjErxb4J4v5tb99HqGMd2XCDTia9aw4RVTv1tBdzN9ibrMG38TXpgMY90ByUnPZbUQ5U6m8MKlQ/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1175" src="https://mmbiz.qpic.cn/sz_mmbiz_png/t1wZDoUyFk6I6hjErxb4J4v5tb99HqGMd2XCDTia9aw4RVTv1tBdzN9ibrMG38TXpgMY90ByUnPZbUQ5U6m8MKlQ/640?wx_fmt=png&amp;from=appmsg"></figure></section><p><br></p><p><mp-style-type data-value="3"></mp-style-type></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/vQFwwn1MA-F8aZk46c4cMw",target="_blank" rel="noopener noreferrer">原文链接</a>
