---
title: "单细胞实战之样本整理，细胞注释和部分图表绘制---从入门到进阶(初级篇1）"
date: 2025-02-09T01:36:23Z
draft: ["false"]
tags: [
  "fetched",
  "单细胞天地"
]
categories: ["Acdemic"]
---
单细胞实战之样本整理，细胞注释和部分图表绘制---从入门到进阶(初级篇1） by 单细胞天地
------
<div><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com"><p data-tool="mdnice编辑器">在完成了马拉松课程后，我们应该对单细胞分析有了基本了解。接下来，我们将开启新的篇章——<strong>单细胞实战：从入门到进阶</strong>。</p><p data-tool="mdnice编辑器">我们采用的示例数据集来自GSE188711，该数据集一共包含了6个单细胞测序样本(其中包含了3个左边结肠和3个右半结肠的样本)，并采用Illumina NovaSeq 6000平台完成测序。首先我们需要从GEO官网中下载该数据集。今后很长一段时间，我们将围绕这个数据集展开一系列的分析学习。<img alt="" data-imgfileid="100044156" data-ratio="1.0473856209150327" data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3iaxSsws3icXWsJZXBRDMNsY7JJw3utk87jc0Wqmd2lnSMicN0n0WEHzIQ/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="612" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3iaxSsws3icXWsJZXBRDMNsY7JJw3utk87jc0Wqmd2lnSMicN0n0WEHzIQ/640?wx_fmt=png&amp;from=appmsg">对应的文献是2022年1月发表在JCI insight上的Resolving the difference between left-sided and right-sided colorectal cancer by single-cell sequencing. PMID：34793335.<img alt="" data-imgfileid="100044154" data-ratio="0.4228094575799722" data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic33YAUSeAFZfwhHyUKRY83ib5KwiblSryOIV17sxRChLAI1roUwwiaHpG8A/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="719" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic33YAUSeAFZfwhHyUKRY83ib5KwiblSryOIV17sxRChLAI1roUwwiaHpG8A/640?wx_fmt=png&amp;from=appmsg">此外，可以通过向公众号发送关键词‘单细胞’，直接获取Seurat V5版本的完整代码。<img alt="" data-imgfileid="100044158" data-ratio="2.2203703703703703" data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3tNiaTUFLIwwF6z9gl9K9pBfIgLn56bq5iauD8lXlRQzlia3j3YVUuctsQ/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3tNiaTUFLIwwF6z9gl9K9pBfIgLn56bq5iauD8lXlRQzlia3j3YVUuctsQ/640?wx_fmt=png&amp;from=appmsg">本次推文中，我们将重点复习并学习降维聚类过程中的三个重要步骤：<strong>样本整理，细胞注释和部分图表绘制</strong>。</p><h4 data-tool="mdnice编辑器"><span></span>分析步骤<span></span></h4><h5 data-tool="mdnice编辑器"><span></span>1.整理数据<span></span></h5><p data-tool="mdnice编辑器">下载完数据之后，我们需要考虑怎么把数据进行归类整理。这种类型的数据需要把同一样本中的barcodes，features和matrix归类到一个文件夹中，并在最终归类的文件夹中的文件只需要保留barcodes，features和matrix名称。同时我们也需要考虑到为了更好区分不同样本的信息，默认的文件名称是GSM编号和样本信息，如果样本数量很少，比如这里一共就6个样本，那么手动分组其实也未尝不可。<img alt="" data-imgfileid="100044157" data-ratio="0.5583333333333333" data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3hpFOdpy3XzsDicZfbYxMg8dyhOMicRiaeh80xiau0snavOEr8Y91JhLpibQ/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3hpFOdpy3XzsDicZfbYxMg8dyhOMicRiaeh80xiau0snavOEr8Y91JhLpibQ/640?wx_fmt=png&amp;from=appmsg"></p><pre data-tool="mdnice编辑器"><span></span><code>fs=list.files(<span>'./'</span>,<span>'features'</span>)<br>fs<br>samples1 = fs <br>samples2=gsub(<span>'_features'</span>,<span>''</span>,fs)<br>samples2<br>samples2=gsub(<span>'.tsv.gz'</span>,<span>''</span>,samples2)<br>samples2 <br><br><span># 创建 outputs 文件夹</span><br>dir.create(<span>"outputs"</span>, showWarnings = <span>FALSE</span>)<br><br>lapply(<span>1</span>:length(samples2), <span>function</span>(i){<br>  <span>#i=1</span><br>  x=samples2[i]<br>  y=samples1[i]<br>  x;y<br>  <br>  <span># 创建子目录路径</span><br>  output_dir = file.path(<span>"outputs"</span>, x)<br>  dir.create(output_dir, recursive = <span>TRUE</span>)<br>  <br>  <span># 移动并重命名文件</span><br>  file.copy(from = y,<br>            to = file.path(output_dir, <span>'features.tsv.gz'</span>))<br>  file.copy(from = gsub(<span>'tsv'</span>, <span>'mtx'</span>, gsub(<span>'features'</span>, <span>'matrix'</span>, y)),<br>            to = file.path(output_dir, <span>'matrix.mtx.gz'</span>))<br>  file.copy(from = gsub(<span>'features'</span>, <span>'barcodes'</span>, y),<br>            to = file.path(output_dir, <span>'barcodes.tsv.gz'</span>))<br>  <br>})<br></code></pre><p data-tool="mdnice编辑器">这段代码我们首先提取了含有features的文件名称，做这一步就是为后面进行样本和文件夹重命名做准备，同理这里用matrix，barcodes提取数据也是可以的。<img alt="" data-imgfileid="100044155" data-ratio="0.15648148148148147" data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3EbuWKG5FxBcJdDyY7vRnFdZkPib6eIib5Hn4TE025kKx9iaL0pDZGQ4Tw/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3EbuWKG5FxBcJdDyY7vRnFdZkPib6eIib5Hn4TE025kKx9iaL0pDZGQ4Tw/640?wx_fmt=png&amp;from=appmsg">得到了fs名称之后，我们需要对其进行优化，分步去除“_features”和“.tsv.gz”，最终得到了含有GSM编号和样本特征的名称，这个也将最终变成文件夹名称。<img alt="" data-imgfileid="100044161" data-ratio="0.11018518518518519" data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3MxUgWGkcsYZvjBu9OBsRY6vCcWFzYP9F8cZicIhic1rx9QibjESNkhjnA/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3MxUgWGkcsYZvjBu9OBsRY6vCcWFzYP9F8cZicIhic1rx9QibjESNkhjnA/640?wx_fmt=png&amp;from=appmsg">接着我们创建outputs文件夹用于保存所有整合好的文件，同时我们也需要创建更新后的子目录。移动并重名文件夹的时候使用了gsub做了多次变化，并使用file.copy对文件进行了重命名。最终得到如下的文件夹+文件。<img alt="" data-imgfileid="100044159" data-ratio="0.2675925925925926" data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3LdGHfK0F0QjzdbMSKRoTz0dFljlOCZ16GtbMyM6PGZLBKDeGqvH87w/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3LdGHfK0F0QjzdbMSKRoTz0dFljlOCZ16GtbMyM6PGZLBKDeGqvH87w/640?wx_fmt=png&amp;from=appmsg"></p><h5 data-tool="mdnice编辑器"><span></span>2.降维聚类分群<span></span></h5><p data-tool="mdnice编辑器">使用Seurat V5</p><pre data-tool="mdnice编辑器"><span></span><code>rm(list=ls())<br>options(stringsAsFactors = <span>F</span>) <br><span>source</span>(<span>'scRNA_scripts/lib.R'</span>)<br>getwd()<br></code></pre><h6 data-tool="mdnice编辑器"><span></span>step1:导入数据<span></span></h6><pre data-tool="mdnice编辑器"><span></span><code><span>###### step1: 导入数据 ######   </span><br><span># 付费环节 800 元人民币</span><br><span># 参考：https://mp.weixin.qq.com/s/tw7lygmGDAbpzMTx57VvFw</span><br>dir=<span>'/Users/zaneflying/Desktop/practiceGSE188711/GSE188711_RAW/outputs/'</span><br>samples=list.files(dir)<br>samples<br><span>library</span>(data.table)<br>sceList = lapply(samples,<span>function</span>(pro){<br>sce = CreateSeuratObject(counts =  Read10X(file.path(dir,pro)) ,<br>                         project = pro,<br>                         min.cells = <span>5</span>,<br>                         min.features = <span>300</span> ) <br><span>return</span>(sce)<br>})<br><span># 把scrList的每个样本进行命名</span><br>names(sceList) =  samples<br><span># 合并样本</span><br>sce.all &lt;- merge(sceList[[<span>1</span>]], y= sceList[ -<span>1</span> ] ,<br>                 add.cell.ids =  samples) <br>dim(sce.all)<br><span># [1] 21308 32300</span><br><span># V5版本的需要joinLayers</span><br>LayerData(sce.all, assay = <span>"RNA"</span>, layer = <span>"counts"</span>)<br>sce.all &lt;- JoinLayers(sce.all)<br>dim(sce.all[[<span>"RNA"</span>]]$counts )<br>sp=<span>'human'</span><br></code></pre><h6 data-tool="mdnice编辑器"><span></span>step2:QC质控<span></span></h6><pre data-tool="mdnice编辑器"><span></span><code><span># 如果为了控制代码复杂度和行数 </span><br><span># 可以省略了质量控制环节</span><br><span>###### step2: QC质控 ######</span><br>dir.create(<span>"./1-QC"</span>)<br>setwd(<span>"./1-QC"</span>)<br><span># 如果过滤的太狠，就需要去修改这个过滤代码</span><br><span>source</span>(<span>'../scRNA_scripts/qc.R'</span>)<br>sce.all.filt = basic_qc(sce.all)<br>print(dim(sce.all))<br><span>#[1] 21308 32300</span><br>print(dim(sce.all.filt))<br><span>#[1] 21308 25621</span><br>setwd(<span>'../'</span>)<br>getwd()<br></code></pre><p data-tool="mdnice编辑器">QC中设定了筛选线粒体基因百分比小于 25% ，核糖体基因表达比例大于 3%和血红蛋白基因表达比例小于 1%的细胞，关于过滤标准需要按照不同组织进行划分，目前没有“确定”的标准。原文质控完之后的细胞数目为27927，我们这里的细胞数据为25621，数量上是有差别的，但这也是无法避免的，通常很难做到得到跟原作者一样的数据结果。<img alt="" data-imgfileid="100044160" data-ratio="0.2212962962962963" data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3HMMBfC7r0yelhMsn20ugFZQiadHXnibsUBicDDGxDopzwGaLIOfUd8lRg/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3HMMBfC7r0yelhMsn20ugFZQiadHXnibsUBicDDGxDopzwGaLIOfUd8lRg/640?wx_fmt=png&amp;from=appmsg"><img alt="" data-imgfileid="100044163" data-ratio="0.7111111111111111" data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3rCEOklvQhIp4bmCQOTaszlBItDuZ6micdBx5FKQt2adTQCTmhb1epEg/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3rCEOklvQhIp4bmCQOTaszlBItDuZ6micdBx5FKQt2adTQCTmhb1epEg/640?wx_fmt=png&amp;from=appmsg"><img alt="" data-imgfileid="100044162" data-ratio="0.6277777777777778" data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3h5TjZXX7NEDUYh8ibNZzJDxHGFdV0xlSqf4bwaZ0YwviaexqBrxeRCKg/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3h5TjZXX7NEDUYh8ibNZzJDxHGFdV0xlSqf4bwaZ0YwviaexqBrxeRCKg/640?wx_fmt=png&amp;from=appmsg"></p><h6 data-tool="mdnice编辑器"><span></span>step3：harmony整合多个单细胞样品<span></span></h6><pre data-tool="mdnice编辑器"><span></span><code><span>###### step3: harmony整合多个单细胞样品 ######</span><br><span>if</span>(<span>T</span>){<br>  dir.create(<span>"2-harmony"</span>)<br>  getwd()<br>  setwd(<span>"2-harmony"</span>)<br>  <span>source</span>(<span>'../scRNA_scripts/harmony.R'</span>)<br>  <span># 默认 ScaleData 没有添加"nCount_RNA", "nFeature_RNA"</span><br>  <span># 默认的</span><br>  sce.all.int = run_harmony(sce.all.filt)<br>  setwd(<span>'../'</span>)  <br>}<br></code></pre><p data-tool="mdnice编辑器">原文中分辨率选择了0.6，而根据进化树和分群情况情况，可以有两种策略，第一种是选择较低的分辨率(比如0.2)进行粗分，之后再进行细分，第二种是选择较高的分辨率(比如0.8)粗分之后再进行修整。两种方法都可以，也都需要经过粗分到细分过程。本次分辨率选择0.8进行分群。<img alt="" data-imgfileid="100044167" data-ratio="1.0185185185185186" data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3ylW4qkxoskkuaicrRez8YW28SvYGnwL19SLeEVfRVHCGCQkhKJsbYQA/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3ylW4qkxoskkuaicrRez8YW28SvYGnwL19SLeEVfRVHCGCQkhKJsbYQA/640?wx_fmt=png&amp;from=appmsg"></p><h6 data-tool="mdnice编辑器"><span></span>step4：看标记基因库<span></span></h6><pre data-tool="mdnice编辑器"><span></span><code><span>###### step4:  看标记基因库 ######</span><br>table(Idents(sce.all.int))<br>table(sce.all.int$seurat_clusters)<br>table(sce.all.int$RNA_snn_res.0.1) <br>table(sce.all.int$RNA_snn_res.0.8) <br><br>getwd()<br>dir.create(<span>'check-by-0.1'</span>)<br>setwd(<span>'check-by-0.1'</span>)<br>sel.clust = <span>"RNA_snn_res.0.1"</span><br>sce.all.int &lt;- SetIdent(sce.all.int, value = sel.clust)<br>table(sce.all.int@active.ident) <br><br><span>source</span>(<span>'../scRNA_scripts/check-all-markers.R'</span>)<br>setwd(<span>'../'</span>) <br>getwd()<br><br>dir.create(<span>'check-by-0.5'</span>)<br>setwd(<span>'check-by-0.5'</span>)<br>sel.clust = <span>"RNA_snn_res.0.5"</span><br>sce.all.int &lt;- SetIdent(sce.all.int, value = sel.clust)<br>table(sce.all.int@active.ident) <br><span>source</span>(<span>'../scRNA_scripts/check-all-markers.R'</span>)<br>setwd(<span>'../'</span>) <br>getwd()<br><br>dir.create(<span>'check-by-0.8'</span>)<br>setwd(<span>'check-by-0.8'</span>)<br>sel.clust = <span>"RNA_snn_res.0.8"</span><br>sce.all.int &lt;- SetIdent(sce.all.int, value = sel.clust)<br>table(sce.all.int@active.ident) <br><span>source</span>(<span>'../scRNA_scripts/check-all-markers.R'</span>)<br>setwd(<span>'../'</span>) <br>getwd()<br><br>last_markers_to_check<br></code></pre><p data-tool="mdnice编辑器">曾老师的代码中包含了多种细胞和肿瘤的标记库，此外还存在COSG函数分析得到了每个细胞群中的差异基因。接下来重点查看0.8分辨率的结果，笔者查看标记结果的方法主要为以下12个字：<strong>由浅入深，重点关注，反复对比</strong>。</p><p data-tool="mdnice编辑器">查看结果的具体顺序为<strong>1. Last_markers_and_umap结果；2.查看需要特别关注或者last_markers_and_umap结果中不明确或者特定组织特定标记群结果；3.查看对比cosg/findallmarkers等差异基因分析工具得到的结果</strong>，时刻提醒自己要结合生物学知识。<img alt="" data-imgfileid="100044165" data-ratio="0.6824074074074075" data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3oicMzzm1lVrhquHs7eAbNgCMQkMPVSPmOIUBsFe4cKiavkMJQPcO10dw/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3oicMzzm1lVrhquHs7eAbNgCMQkMPVSPmOIUBsFe4cKiavkMJQPcO10dw/640?wx_fmt=png&amp;from=appmsg"></p><ol data-tool="mdnice编辑器"><li><section>Last_markers_and_umap的结果是气泡图结合UMAP图：这里的标记主要是目前我们所已知公认的标记，每个人都可以根据自己主攻方向设定标记集。除此之外，我们也需要根据UMAP图中的不同细胞群之间的位置和细胞群的信息来反推我们的注释是否存在问题。</section></li><li><section>查看需要特别关注或者last_markers_and_umap结果中不明确或者特定组织特定标记群结果。比如在后续分析中需要特别关注T细胞，那么可以尝试先观察一下T细胞标记集的结果；或者比如在last_markers_and_umap结果中发现某一群细胞无法被清晰的注释，那么就可以考虑通过更加细节的标记进一步对比</section></li><li><section>查看对比cosg/findallmarkers等差异基因分析工具得到的结果，时刻提醒自己要结合生物学知识。根据差异分析基因工具得到不同细胞群中的差异基因，然后根据这些差异基因去验证前面两步注释的结果。</section></li></ol><p data-tool="mdnice编辑器">接下来打算正式对细胞进行注释分群，首先回溯一下文章中的分群情况以及每个细胞群所对应的标志物。研究者将细胞十分细致的分成了19个亚群，特别有意思的是研究者还分出了一群Cancer stem cells细胞群，这群细胞的高表达标记物为TFF3, SPKK4, AGR2。不过按照目前的研究来看，可能还需要更多的标志物才能更好的定义干细胞。<img alt="" data-imgfileid="100044169" data-ratio="0.6592592592592592" data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3SibZsOYOTicznRk8r3fJ7xXg45b9wHqyUiaiazRo8bib3OVBePOGzR1LiavQ/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3SibZsOYOTicznRk8r3fJ7xXg45b9wHqyUiaiazRo8bib3OVBePOGzR1LiavQ/640?wx_fmt=png&amp;from=appmsg"><img alt="" data-imgfileid="100044166" data-ratio="0.32037037037037036" data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3icCoYiblPIr29mse1B2lcuSax9RIL6cStbTcsVgIxjrNHSwUpxg83cfA/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3icCoYiblPIr29mse1B2lcuSax9RIL6cStbTcsVgIxjrNHSwUpxg83cfA/640?wx_fmt=png&amp;from=appmsg">作为一个示例数据，我们并不计划严格按照研究者提供的细胞分群信息进行细胞注释，而是结合以下信息综合进行细胞注释。</p><ol data-tool="mdnice编辑器"><li><section>细胞亚群的认定流程：<strong>第一：判断标志物是否是细胞群所特有的，比如T细胞的CD3, B细胞的MS4A1(CD20)，NK细胞的KLRB1，髓系细胞的C1QA/C1QB，肝细胞的APOC3。第二：需要考虑标志物的平均表达量高低，一般会倾向关注表达量很高的标志物。第三：若两种分属不同细胞亚群的标志物在同一个细胞簇中出现时，需要考虑是否是特殊的细胞群</strong>。</section></li><li><section>如果细胞群出现特殊情况，举个不恰当的例子，比如在下面这个图中有那么多个簇都属于T细胞的时候，需要考虑是否是存在问题(要考虑T细胞的数量和百分比是否符合常理)，这个时候需要回溯多个图表进行综合考量。比如需要回溯nFeature，nCount图看看这个簇的值是否异常(包括线粒体，核糖体和血红蛋白的值)，也需要看看COSG/FindAllmaker差异分析结果判断一下是否符合T细胞的特征，同时一定要查阅一下文献看看是否有高分文献进行过类型细胞的报道。如果确认了上述情况都没有问题，那咱们也不要犹豫，这就是真实的结果~ 后续就可以继续研究为什么这批样本中T细胞的比例如此之高。<img alt="" data-imgfileid="100044168" data-ratio="0.9879629629629629" data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3LhIYE5ia8htqicmbcuk6V95t8O5bcicLH8WaIssteYAxkXUbicVrZDsZ0w/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3LhIYE5ia8htqicmbcuk6V95t8O5bcicLH8WaIssteYAxkXUbicVrZDsZ0w/640?wx_fmt=png&amp;from=appmsg">cosg结果与前面的结果对比验证<img alt="" data-imgfileid="100044170" data-ratio="1.1629392971246006" data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3fiaum1r3HJTqqs2t4icUblicBOiak0IXnzNvmqUPtEfias01RwS7ibI0hIvA/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="626" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3fiaum1r3HJTqqs2t4icUblicBOiak0IXnzNvmqUPtEfias01RwS7ibI0hIvA/640?wx_fmt=png&amp;from=appmsg"></section></li></ol><h6 data-tool="mdnice编辑器"><span></span>step5: 确定单细胞亚群生物学名字<span></span></h6><pre data-tool="mdnice编辑器"><span></span><code><span>###### step5: 确定单细胞亚群生物学名字 ######</span><br><span>source</span>(<span>'scRNA_scripts/lib.R'</span>)<br>sce.all.int = readRDS(<span>'2-harmony/sce.all_int.rds'</span>)<br>sp=<span>'human'</span><br>tmp=sce.all.int@meta.data<br>colnames(sce.all.int@meta.data) <br>table(sce.all.int$RNA_snn_res.0.8)<br><span># pbmc_small &lt;- BuildClusterTree(object = sce.all.int)</span><br><span># plot(Tool(object = pbmc_small, slot = 'BuildClusterTree'))</span><br><span># plot(pbmc_small@tools$BuildClusterTree)</span><br><br><span># 付费环节 800 元人民币</span><br><span># 如果是手动给各个单细胞亚群命名</span><br><span>if</span>(<span>T</span>){<br>  sce.all.int<br>  celltype=data.frame(ClusterID=<span>0</span>:<span>18</span> ,<br>                      celltype= <span>0</span>:<span>18</span>  ) <br>  <span>#定义细胞亚群        </span><br>  celltype[celltype$ClusterID %<span>in</span>% c( <span>0</span>,<span>2</span>,<span>3</span>  ),<span>2</span>]=<span>'T/NK cells'</span>  <br>  celltype[celltype$ClusterID %<span>in</span>% c( <span>1</span>,<span>15</span>  ),<span>2</span>]=<span>'B cells'</span>  <br>  celltype[celltype$ClusterID %<span>in</span>% c( <span>4</span>,<span>16</span>  ),<span>2</span>]=<span>'plasma'</span>  <br>  celltype[celltype$ClusterID %<span>in</span>% c( <span>11</span>  ),<span>2</span>]=<span>'proliferative cells'</span>  <br>  celltype[celltype$ClusterID %<span>in</span>% c( <span>10</span>,<span>18</span>  ),<span>2</span>]=<span>'macrophages'</span>  <br>  celltype[celltype$ClusterID %<span>in</span>% c( <span>7</span>  ),<span>2</span>]=<span>'neutrophils'</span>    <br>  celltype[celltype$ClusterID %<span>in</span>% c( <span>13</span>  ),<span>2</span>]=<span>'endothelial cells'</span> <br>  celltype[celltype$ClusterID %<span>in</span>% c( <span>8</span> ),<span>2</span>]=<span>'fibroblasts'</span>        <br>  celltype[celltype$ClusterID %<span>in</span>% c( <span>5</span>,<span>9</span>,<span>17</span> ),<span>2</span>]=<span>'epithelial/cancer cells'</span>         <br>  celltype[celltype$ClusterID %<span>in</span>% c( <span>12</span> ),<span>2</span>]=<span>'mast cells'</span>   <br>  celltype[celltype$ClusterID %<span>in</span>% c( <span>14</span> ),<span>2</span>]=<span>'VSMCs'</span>  <br>  celltype[celltype$ClusterID %<span>in</span>% c( <span>6</span> ),<span>2</span>]=<span>'dendritic cells'</span><br>  sce.all.int<br>  head(celltype)<br>  celltype<br>  table(celltype$celltype)<br>  sce.all.int@meta.data$celltype = <span>"NA"</span><br>  <br>  <span>for</span>(i <span>in</span> <span>1</span>:nrow(celltype)){<br>    sce.all.int@meta.data[which(sce.all.int@meta.data$RNA_snn_res.0.8 == celltype$ClusterID[i]),<span>'celltype'</span>] &lt;- celltype$celltype[i]}<br>  Idents(sce.all.int)=sce.all.int$celltype<br>  <br>  <br>}<br><br>table(sce.all.int$celltype)<br><span># B cells         dendritic cells       endothelial cells epithelial/cancer cells </span><br><span># 3804                    1244                     549                    2421 </span><br><span># fibroblasts             macrophages              mast cells             neutrophils </span><br><span># 1040                    1014                     596                    1198 </span><br><span># plasma     proliferative cells              T/NK cells                   VSMCs </span><br><span># 2096                     803                   10445                     411</span><br></code></pre><p data-tool="mdnice编辑器">在这一次分群中，一共分出了12种细胞，分别为B cells，dendritic cells，endothelial cells，epithelial/cancer cells, fibroblasts，macrophages，mast cells，neutrophils，plasma，proliferative cells，T/NK cells，VSMCs。</p><p data-tool="mdnice编辑器">接下来我们可以先大概粗看一下每个细胞亚群的细胞数量，其中dendritic cells, macrophages和neutrophils这几个群数据过于相近！而根据我们的先验知识，通常树突细胞在免疫细胞中占比是最少的，中性粒细胞比例会高一些，巨噬细胞在这三种细胞中占比是最高的。因此在看到三者的细胞数量过于相似时需要警惕这种注释方式是否正确。</p><p data-tool="mdnice编辑器">那么回溯的注释是否正确的顺序可以从以下几点考虑：<strong>1. 首先是回看标志物的气泡图和UMAP图；2.如果确定无误，那么就要从样本的生物学角度考虑是否符合“逻辑”；3.如果都是正确的那就大胆的命名，如果拿不准那就退一步往这些细胞群的上一级命名。</strong> 那么根据这几点回溯了数据之后发现暂时不能确定这就是准确的细胞分群，因此我们决定将这几群细胞统一命名为myeloid cells.</p><pre data-tool="mdnice编辑器"><span></span><code><span># 重新注释，不需要的细胞亚群用#注释掉了</span><br><span>if</span>(<span>T</span>){<br>  sce.all.int<br>  celltype=data.frame(ClusterID=<span>0</span>:<span>18</span> ,<br>                      celltype= <span>0</span>:<span>18</span>  ) <br>  <span>#定义细胞亚群        </span><br>  celltype[celltype$ClusterID %<span>in</span>% c( <span>0</span>,<span>2</span>,<span>3</span>  ),<span>2</span>]=<span>'T/NK cells'</span>  <br>  celltype[celltype$ClusterID %<span>in</span>% c( <span>1</span>,<span>15</span>  ),<span>2</span>]=<span>'B cells'</span>  <br>  celltype[celltype$ClusterID %<span>in</span>% c( <span>4</span>,<span>16</span>  ),<span>2</span>]=<span>'plasma'</span>  <br>  celltype[celltype$ClusterID %<span>in</span>% c( <span>11</span>  ),<span>2</span>]=<span>'proliferative cells'</span>  <br>  <span>#celltype[celltype$ClusterID %in% c( 10,18  ),2]='macrophages'  </span><br>  <span>#celltype[celltype$ClusterID %in% c( 7  ),2]='neutrophils'    </span><br>  celltype[celltype$ClusterID %<span>in</span>% c( <span>13</span>  ),<span>2</span>]=<span>'endothelial cells'</span> <br>  celltype[celltype$ClusterID %<span>in</span>% c( <span>8</span> ),<span>2</span>]=<span>'fibroblasts'</span>        <br>  celltype[celltype$ClusterID %<span>in</span>% c( <span>5</span>,<span>9</span>,<span>17</span> ),<span>2</span>]=<span>'epithelial/cancer cells'</span>         <br>  celltype[celltype$ClusterID %<span>in</span>% c( <span>12</span> ),<span>2</span>]=<span>'mast cells'</span>   <br>  celltype[celltype$ClusterID %<span>in</span>% c( <span>14</span> ),<span>2</span>]=<span>'VSMCs'</span>  <br>  <span>#celltype[celltype$ClusterID %in% c( 6 ),2]='dendritic cells'</span><br>  celltype[celltype$ClusterID %<span>in</span>% c( <span>6</span>,<span>7</span>,<span>10</span>,<span>18</span> ),<span>2</span>]=<span>'myeloid cells'</span><br>  sce.all.int<br>  head(celltype)<br>  celltype<br>  table(celltype$celltype)<br>  sce.all.int@meta.data$celltype = <span>"NA"</span><br>  <br>  <span>for</span>(i <span>in</span> <span>1</span>:nrow(celltype)){<br>    sce.all.int@meta.data[which(sce.all.int@meta.data$RNA_snn_res.0.8 == celltype$ClusterID[i]),<span>'celltype'</span>] &lt;- celltype$celltype[i]}<br>  Idents(sce.all.int)=sce.all.int$celltype<br>  <br>  <br>}<br><br>table(sce.all.int$celltype)<br><span># B cells       endothelial cells epithelial/cancer cells             fibroblasts </span><br><span># 3804                     549                    2421                    1040 </span><br><span># mast cells           myeloid cells             plasma        proliferative cells </span><br><span># 596                    3456                    2096                     803 </span><br><span># T/NK cells               VSMCs </span><br><span># 10445                     411 </span><br></code></pre><p data-tool="mdnice编辑器">我们可以看到此时三种细胞已经整合成myeloid cells了</p><pre data-tool="mdnice编辑器"><span></span><code><span># 过滤无法命名的 </span><br>sce.all.int = sce.all.int[,!sce.all.int$celltype %<span>in</span>% <span>0</span>:<span>100</span>]<br>table(sce.all.int$celltype) <br>table(sce.all.int$orig.ident) <br><span># GSM5688706_WGC     GSM5688707_JCA GSM5688708_LS-CRC3 GSM5688709_RS-CRC1  GSM5688710_R_CRC3 </span><br><span># 3640               2894               5796               5354               3831 </span><br><span># GSM5688711_R_CRC4 </span><br><span># 4106 </span><br>sce.all.int<br><span># An object of class Seurat </span><br><span># 21308 features across 25621 samples within 1 assay </span><br><span># Active assay: RNA (21308 features, 2000 variable features)</span><br><span># 3 layers present: counts, data, scale.data</span><br><span># 3 dimensional reductions calculated: pca, harmony, umap</span><br>head(colnames(sce.all.int))<br><span># [1] "GSM5688706_WGC_AAACCCAAGGCATGGT-1" "GSM5688706_WGC_AAACCCATCCGTGGGT-1"</span><br><span># [3] "GSM5688706_WGC_AAACGAAAGGGCTGAT-1" "GSM5688706_WGC_AAACGAAAGGTACATA-1"</span><br><span># [5] "GSM5688706_WGC_AAACGAAAGTCACGCC-1" "GSM5688706_WGC_AAACGAACAACATCGT-1"</span><br></code></pre><p data-tool="mdnice编辑器">接下来把左右半结肠的临床信息添加进去, 请至GEO官网进行查看。</p><pre data-tool="mdnice编辑器"><span></span><code>a &lt;- sce.all.int@meta.data<br>a$location &lt;- ifelse(a$orig.ident %<span>in</span>% c(<span>"GSM5688706_WGC"</span>,<span>"GSM5688707_JCA"</span>,<span>"GSM5688708_LS-CRC3"</span>),<span>"left"</span>,<span>"right"</span>)<br>table(a$location)<br><span># left right </span><br><span># 12330 13291 </span><br>sce.all.int@meta.data &lt;- a<br>colnames(sce.all.int@meta.data)<br><span># [1] "orig.ident"       "nCount_RNA"       "nFeature_RNA"     "percent_mito"     "percent_ribo"     "percent_hb"      </span><br><span># [7] "RNA_snn_res.0.01" "seurat_clusters"  "RNA_snn_res.0.05" "RNA_snn_res.0.1"  "RNA_snn_res.0.2"  "RNA_snn_res.0.3" </span><br><span># [13] "RNA_snn_res.0.5"  "RNA_snn_res.0.8"  "RNA_snn_res.1"    "celltype"         "location"</span><br><br>gplots::balloonplot(table(sce.all.int$celltype ,<br>                          sce.all.int$location))<br><br>gplots::balloonplot(table(sce.all.int$celltype ,<br>                          sce.all.int$orig.ident))<br></code></pre><p data-tool="mdnice编辑器">并通过气泡图查看一下左右半结肠中各个细胞亚群的细胞数量是否有差异。粗看下来可能mast cell在数量上是最显著的细胞群，同时对比原文中的结果发现mast细胞的比例也确实在右边结肠中比例高。<img alt="" data-imgfileid="100044172" data-ratio="0.5268518518518519" data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3HhRgBByziaDib8Xm2vicgj0VibxYgqyn4hVexWxaREtyB1sy8Hz1M33Rsw/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3HhRgBByziaDib8Xm2vicgj0VibxYgqyn4hVexWxaREtyB1sy8Hz1M33Rsw/640?wx_fmt=png&amp;from=appmsg"><img alt="" data-imgfileid="100044171" data-ratio="0.46944444444444444" data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3fS6E0zicwZI31EUxCqC6FbuBAu5ibyzkZeyrkxjSRicVFKj4G8TaokhDA/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3fS6E0zicwZI31EUxCqC6FbuBAu5ibyzkZeyrkxjSRicVFKj4G8TaokhDA/640?wx_fmt=png&amp;from=appmsg"><img alt="" data-imgfileid="100044173" data-ratio="0.42407407407407405" data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3Fnkn1hrhMGqbGMCEfu3q9JbA2pcmicjNTF9aia5CF0WibByOrQfzJRQAg/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3Fnkn1hrhMGqbGMCEfu3q9JbA2pcmicjNTF9aia5CF0WibByOrQfzJRQAg/640?wx_fmt=png&amp;from=appmsg"></p><pre data-tool="mdnice编辑器"><span></span><code><span># 如果前面成功的给各个细胞亚群命名了</span><br><span># 就可以运行下面的代码 </span><br><span>if</span>(<span>"celltype"</span> %<span>in</span>% colnames(sce.all.int@meta.data ) ){<br>  <br>  sel.clust = <span>"celltype"</span><br>  sce.all.int &lt;- SetIdent(sce.all.int, value = sel.clust)<br>  table(sce.all.int@active.ident) <br>  phe=sce.all.int@meta.data<br>  save(phe,file = <span>'phe.Rdata'</span>)<br>  pdf(<span>'celltype-vs-orig.ident.pdf'</span>,width = <span>10</span>)<br>  gplots::balloonplot(table(sce.all.int$celltype,sce.all.int$orig.ident))<br>  dev.off()<br>  <br>  dir.create(<span>'check-by-celltype'</span>)<br>  setwd(<span>'check-by-celltype'</span>)<br>  <span>source</span>(<span>'../scRNA_scripts/check-all-markers.R'</span>)<br>  setwd(<span>'../'</span>) <br>  getwd()<br>}<br></code></pre><figure data-tool="mdnice编辑器"><img alt="" data-imgfileid="100044174" data-ratio="0.7851851851851852" data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3tlcO8gZpLAujtATocibM55j6CcKapnlTt9xFMVlia5hT64mcR7welahg/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3tlcO8gZpLAujtATocibM55j6CcKapnlTt9xFMVlia5hT64mcR7welahg/640?wx_fmt=png&amp;from=appmsg"></figure><h6 data-tool="mdnice编辑器"><span></span>step6. 载入并且理解数据<span></span></h6><pre data-tool="mdnice编辑器"><span></span><code>rm(list=ls())<br>options(stringsAsFactors = <span>F</span>)<br>V5_path = <span>"/Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/seurat5/"</span><br>.libPaths(V5_path)<br>.libPaths()<br><br><span>library</span>(Seurat)<br><span>library</span>(ggplot2)<br><span>library</span>(clustree)<br><span>library</span>(cowplot)<br><span>library</span>(dplyr)<br><span>library</span>(stringr)<br><span>library</span>(ggsci) <br><span>library</span>(patchwork) <br><span>library</span>(ggsci)<br><span>library</span>(ggpubr)<br><span>library</span>(RColorBrewer) <br><br>getwd()<br>dir.create(<span>'paper-figures/'</span>)<br>setwd(<span>'paper-figures/'</span>) <br><br><span>## 1. 载入，并且理解数据 ---- </span><br>sce.all = readRDS(<span>'../2-harmony/sce.all_int.rds'</span>)<br>sp=<span>'human'</span> <br>load(<span>'../phe.Rdata'</span>) <br>identical(rownames(phe) , colnames(sce.all))<br>sce.all=sce.all[,colnames(sce.all) %<span>in</span>% rownames(phe) ]<br>sce.all@meta.data = phe[match(colnames(sce.all),rownames(phe) ),]<br>sel.clust = <span>"celltype"</span><br>sce.all &lt;- SetIdent(sce.all, value = sel.clust)<br>table(sce.all@active.ident) <br>DimPlot(sce.all) <br>colnames(sce.all@meta.data) <br><br>sce.all.int=sce.all <br><br>seuratObj &lt;- RunUMAP(sce.all.int,  <br>                     reduction = <span>"pca"</span>,<br>                     dims = <span>1</span>:<span>15</span> )<br>colnames(sce.all.int@meta.data)<br>p3=DimPlot(sce.all.int, reduction = <span>"umap"</span>, <br>        group.by = <span>"celltype"</span> ) <span>#+ NoLegend()</span><br>p4=DimPlot(seuratObj, reduction = <span>"umap"</span> , <br>          group.by = <span>"celltype"</span>)+ NoLegend()<br><br>p1=DimPlot(sce.all.int, reduction = <span>"umap"</span>, <br>           group.by = <span>"orig.ident"</span> )+ NoLegend()<br>p2=DimPlot(seuratObj, reduction = <span>"umap"</span> , <br>           group.by = <span>"orig.ident"</span>)+ NoLegend()<br><br>(p1+p2)/(p3+p4)<br>ggsave(<span>'harmony-or-not.pdf'</span>,width = <span>12</span>,height = <span>12</span>)<br></code></pre><p data-tool="mdnice编辑器">下图左边部分是harmony之后的结果，右边部分是pca的结果。我们可以发现相比于右边的结果，左边结果中细胞之间会融合更好。<img alt="" data-imgfileid="100044178" data-ratio="1.000925925925926" data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3FKicA2TnbqqFPmZNW0SYiaEibtbAib7XlNPiccPSWicV6fbsq7cbOzbBpXMA/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3FKicA2TnbqqFPmZNW0SYiaEibtbAib7XlNPiccPSWicV6fbsq7cbOzbBpXMA/640?wx_fmt=png&amp;from=appmsg"></p><pre data-tool="mdnice编辑器"><span></span><code>plot1 &lt;- DimPlot(sce, reduction = <span>"umap"</span>, group.by = <span>"RNA_snn_res.0.8"</span>,label = <span>T</span>,repel = <span>T</span>)+  <br>  theme(legend.key.size = unit(<span>0.02</span>, <span>'cm'</span>)) + theme(plot.title = element_text(size=<span>12</span>),<br>                                                    axis.title = element_text(size=<span>10</span>),<br>                                                    axis.text = element_text(size=<span>10</span>),<br>                                                    legend.text = element_text(size=<span>10</span>))<br>plot2 &lt;- DimPlot(sce, reduction = <span>"umap"</span>, group.by = <span>"celltype"</span>,label = <span>F</span>,cols = color,repel = <span>T</span>) +<br>  theme(legend.key.size = unit(<span>0.02</span>, <span>'cm'</span>)) + theme(plot.title = element_text(size=<span>12</span>),<br>                                                    axis.title = element_text(size=<span>10</span>),<br>                                                    axis.text = element_text(size=<span>10</span>),<br>                                                    legend.text = element_text(size=<span>10</span>))<br>plot3 &lt;- DimPlot(sce, reduction = <span>"umap"</span>, group.by = <span>"location"</span>,label = <span>T</span>,cols = sample(color,<span>6</span>),repel = <span>T</span>) +<br>  theme(plot.title = element_text(size=<span>12</span>),<br>        axis.title = element_text(size=<span>10</span>),<br>        axis.text = element_text(size=<span>10</span>),<br>        legend.text = element_text(size=<span>10</span>))<br> <br>fig1 &lt;- plot1|plot2|plot3<br>fig1<br>ggsave(fig1,filename = <span>"fig1.pdf"</span>,units = <span>"cm"</span>,width = <span>32</span>,height = <span>1</span><br></code></pre><figure data-tool="mdnice编辑器"><img alt="" data-imgfileid="100044176" data-ratio="0.36666666666666664" data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3JP0oBcG47KYbNKkMiaibS9pL8XPTD90lkW34ehqj5m4o3crvykPM4sHw/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3JP0oBcG47KYbNKkMiaibS9pL8XPTD90lkW34ehqj5m4o3crvykPM4sHw/640?wx_fmt=png&amp;from=appmsg"></figure><pre data-tool="mdnice编辑器"><span></span><code><span># 排序为后续做图做准备</span><br>ord = c(  <span>"T/NK cells"</span>,<span>"plasma"</span>,<span>"myeloid cells"</span>,<span>"B cells"</span>,<br>          <span>"epithelial/cancer cells"</span>, <span>"mast cells"</span> ,<span>"proliferative cells"</span>,<br>          <span>"endothelial cells"</span>,<span>"VSMCs"</span>,<span>"fibroblasts"</span>  )<br>ord = unique(sce.all$celltype) <br>all(ord %<span>in</span>% unique(sce.all$celltype) )<br>sce.all$celltype = factor(sce.all$celltype ,levels = ord)<br>table(sce.all$celltype)<br>DimPlot(sce.all) <br><br>table(Idents(sce))<br>table(Idents(sce.all))<br><br><span>if</span>(<span>T</span>){<br>  sce= sce.all<br>  pro = <span>'cosg_celltype_'</span><br>  <span>library</span>(COSG)<br>  marker_cosg &lt;- cosg(<br>    sce,<br>    groups=<span>'all'</span>,<br>    assay=<span>'RNA'</span>,<br>    slot=<span>'data'</span>,<br>    mu=<span>1</span>,<br>    n_genes_user=<span>100</span>)<br>  <br>  save(marker_cosg,file = paste0(pro,<span>'_marker_cosg.Rdata'</span>))<br>  <br>  <br>  <span>## Top10 genes</span><br>  <span>library</span>(dplyr)  <br>  top_10 &lt;- unique(as.character(apply(marker_cosg$names,<span>2</span>,head,<span>10</span>)))<br>  <span># width &lt;-0.006*dim(sce)[2];width</span><br>  <span># height &lt;- 0.25*length(top_10)+4.5;height</span><br>  <br>  width &lt;- <span>15</span>+<span>0.5</span>*length(unique(Idents(sce)));width<br>  height &lt;- <span>8</span>+<span>0.1</span>*length(top_10);height<br>  <br>  <span># levels(  Idents(sce) )  = 0:length(levels(  Idents(sce) ))</span><br>  DoHeatmap( subset(sce,downsample=<span>100</span>), top_10 , <br>             size=<span>3</span>)<br>  <br>  ggsave(filename=paste0(pro,<span>'DoHeatmap_check_top10_markers_by_clusters.pdf'</span>) ,<br>         <span># limitsize = FALSE,</span><br>         units = <span>"cm"</span>,width=width,height=height)<br>  width &lt;- <span>8</span>+<span>0.6</span>*length(unique(Idents(sce)));width<br>  height &lt;- <span>8</span>+<span>0.2</span>*length(top_10);height<br>  DotPlot(sce, features = top_10 ,<br>          assay=<span>'RNA'</span>  )  + coord_flip() +FontSize(y.text = <span>4</span>)<br>  ggsave(paste0(pro,<span>'DotPlot_check_top10_markers_by_clusters.pdf'</span>),<br>         units = <span>"cm"</span>,width=width,height=height)<br>  <br>  <br>  <span>## Top3 genes</span><br>  top_3 &lt;- unique(as.character(apply(marker_cosg$names,<span>2</span>,head,<span>3</span>)))<br>  <br>  width &lt;- <span>15</span>+<span>0.2</span>*length(unique(Idents(sce)));width<br>  height &lt;- <span>8</span>+<span>0.1</span>*length(top_3);height<br>  <br>  DoHeatmap( subset(sce,downsample=<span>100</span>), top_3 ,<br>             size=<span>3</span>)<br>  ggsave(filename=paste0(pro,<span>'DoHeatmap_check_top3_markers_by_clusters.pdf'</span>) ,<br>         units = <span>"cm"</span>,width=width,height=height)<br>  <br>  width &lt;- <span>8</span>+<span>0.2</span>*length(unique(Idents(sce)));width<br>  height &lt;- <span>8</span>+<span>0.1</span>*length(top_3);height<br>  DotPlot(sce, features = top_3 ,<br>          assay=<span>'RNA'</span>  )  + coord_flip()<br>  ggsave(paste0(pro,<span>'DotPlot_check_top3_markers_by_clusters.pdf'</span>),width=width,height=height)<br>  <br>  <br>}<br> <br><br>getwd()<br>pro = <span>'cosg_celltype_'</span><br>load(file = paste0(pro,<span>'_marker_cosg.Rdata'</span>))<br><br>top_10 &lt;- unique(as.character(apply(marker_cosg$names,<span>2</span>,head,<span>10</span>))) <br>sce.Scale &lt;- ScaleData(subset(sce.all,downsample=<span>100</span>),features =  top_10  )  <br>table(sce.Scale$celltype)<br><span>library</span>(paletteer) <br>color &lt;- c(paletteer_d(<span>"awtools::bpalette"</span>),<br>           paletteer_d(<span>"awtools::a_palette"</span>),<br>           paletteer_d(<span>"awtools::mpalette"</span>),<br>           paletteer_d(<span>"awtools::spalette"</span>))<br> <br>table(Idents(sce.Scale)) <br><br>ll =  as.list(as.data.frame(apply(marker_cosg$names,<span>2</span>,head,<span>10</span>)))<br>ll = ll[ord]<br>rmg=names(table(unlist(ll))[table(unlist(ll))&gt;<span>1</span>])<br>ll = lapply(ll, <span>function</span>(x) x[!x %<span>in</span>% rmg])<br>ll<br>DoHeatmap(sce.Scale,<br>          features = unlist(ll),<br>          group.by = <span>"celltype"</span>,<br>          raster = <span>F</span>,<br>          assay = <span>'RNA'</span>, label = <span>T</span>)+<br>  scale_fill_gradientn(colors = c(<span>"white"</span>,<span>"grey"</span>,<span>"firebrick3"</span>))<br><br>ggsave(filename = <span>"fig4-top10-marker_pheatmap.pdf"</span>,units = <span>"cm"</span>,<br>       width = <span>25</span>,height = <span>30</span>)<br></code></pre><figure data-tool="mdnice编辑器"><img alt="" data-imgfileid="100044179" data-ratio="1.1592592592592592" data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3w8mHfZR5GHROsC62WukmzIZMSxUZGMGss1Rjzkp8cg1g2ic2jPmJ6Dg/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3w8mHfZR5GHROsC62WukmzIZMSxUZGMGss1Rjzkp8cg1g2ic2jPmJ6Dg/640?wx_fmt=png&amp;from=appmsg"></figure><pre data-tool="mdnice编辑器"><span></span><code>load(<span>'../phe.Rdata'</span>)<br>phe$group = sce$group = ifelse(sce$orig.ident %<span>in</span>% c(<span>"GSM5688706_WGC"</span>,<span>"GSM5688707_JCA"</span>,<span>"GSM5688708_LS-CRC3"</span>),<span>"left"</span>,<span>"right"</span>)<br>head(phe) <br>gplots::balloonplot(<br>  table(phe$celltype,phe$orig.ident)<br>)<br><br>x = phe$orig.ident<br>y = phe$celltype<br>tbl =  table(x,y)<br>df = dcast(as.data.frame(tbl),x~y)<br>ptb = round(<span>100</span>*tbl/rowSums(df[,-<span>1</span>]),<span>2</span>)<br>df = as.data.frame(ptb)<br>colnames(df)=c(<span>'orig.ident'</span>,<span>'celltype'</span>,<span>'Percentage'</span>)<br>head(df)<br>gpinfo=unique(phe[,c(<span>'orig.ident'</span>,<span>'group'</span>)]);gpinfo<br>df=merge(df,gpinfo,by=<span>'orig.ident'</span>)<br>head(df)<br></code></pre><figure data-tool="mdnice编辑器"><img alt="" data-imgfileid="100044177" data-ratio="0.49537037037037035" data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3ibWiarSTZicmibf0DT8BWFaALuVOSk2Qap5Uh9AR1lxBZPSbUahp43FhYA/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjR35GfwTfKcUa12sZxeiaeic3ibWiarSTZicmibf0DT8BWFaALuVOSk2Qap5Uh9AR1lxBZPSbUahp43FhYA/640?wx_fmt=png&amp;from=appmsg"></figure><h4 data-tool="mdnice编辑器"><span></span>参考资料：<span></span></h4><ol data-tool="mdnice编辑器"><li><section>Resolving the difference between left-sided and right-sided colorectal cancer by single-cell sequencing.JCI Insight. 2022 Jan 11;7(1):e152616.</section></li></ol><p data-tool="mdnice编辑器"><strong>致谢</strong>：感谢曾老师以及生信技能树团队全体成员。</p><p data-tool="mdnice编辑器"><strong>注</strong>：若对内容有疑惑或者有发现明确错误的朋友，请联系后台(欢迎交流)。</p><span>- END -</span></section><p><mp-style-type data-value="3"></mp-style-type></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/wivGnWxfKfQROZvMxBlk-g",target="_blank" rel="noopener noreferrer">原文链接</a>
