---
title: "单细胞实战之Scissor和BayesPrism映射——入门到进阶(初级篇4）"
date: 2025-02-15T15:54:46Z
draft: ["false"]
tags: [
  "fetched",
  "单细胞天地"
]
categories: ["Acdemic"]
---
单细胞实战之Scissor和BayesPrism映射——入门到进阶(初级篇4） by 单细胞天地
------
<div><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com"><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com"><blockquote data-tool="mdnice编辑器"><span></span><p>我们在上一讲内容中学习了<a target="_blank" href="https://mp.weixin.qq.com/s?__biz=MzI1Njk4ODE0MQ==&amp;mid=2247528022&amp;idx=1&amp;sn=232a80f8e5d72bde16fe496211604318&amp;scene=21#wechat_redirect" textvalue="pseudobulks和GSVA富集分析" linktype="text" imgurl="" imgdata="null" data-itemshowtype="0" tab="innerlink" data-linktype="2">pseudobulks和GSVA富集分析</a>~ 接下来我们我们学习一下<strong>两个“映射”工具——Scissor和BayesPrism</strong>。同时本次分析还需要TCGA-COAD的数据，可以自己清洗数据，也可以直接下载网盘内容分享的文件：TCGA-COAD_sur_model.Rdata <span>链接: https://pan.baidu.com/s/18ZawCeYKkUMvwC9QVQ9yhA 提取码: xhab</span><span></span></p></blockquote></section><h4 data-tool="mdnice编辑器"><span></span>映射工具<span></span></h4><p data-tool="mdnice编辑器">在做基础实验的时候，研究者都希望能够改变各种条件来进行对比分析，从而探索自己所感兴趣的方向。在做数据分析的时候也是一样的，我们希望有一个数据集能够附加了很多临床信息/表型，然后二次分析者们就可以进一步挖掘。然而现实情况总是数据集质量非常不错，但是附加的临床信息/表型却十分有限，这种状况在单细胞数据分析时代中更加常见。因此如何将大量的含有临床信息/表型的bulkRNA测序数据和单细胞数据构成联系，这也是算法开发者们所重点关注的方向之一。</p><p data-tool="mdnice编辑器">目前关于映射的工具有很多，从bulkRNA到单细胞上我们常用Scissor(“剪刀”法)，从单细胞上映射到bulk层面最常用的就是BayesPrism(贝叶斯棱镜法)了，这次内容我们就来学习一下Scissor法和BayesPrism。</p><h5 data-tool="mdnice编辑器"><span></span>1.Scissor<span></span></h5><p data-tool="mdnice编辑器">图片的内容展示了Scissor法的整一个分析流程，通过结合bulk RNA-seq表型数据和单细胞转录组数据，用于识别与特定表型（如生存率、治疗反应）相关的细胞亚群。首先，计算bulk样本与单细胞之间的相关性矩阵，并结合细胞相似性网络进行正则化回归，筛选出与表型显著相关的Scissor⁺ (正相关) 和Scissor⁻ (负相关)细胞。随后，通过UMAP可视化、统计检验和通路富集分析，进一步解析这些细胞的特征及其在生物学功能中的作用。<img alt="" data-imgfileid="100044345" data-ratio="1.189127972819932" data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjQSNb81YJ5yTpdtwlu7GOvNARwvgNd5wMiaySgfwZiaKBe2WQp5RUnoic0cZnVx1xpdwx4Vf6jokFnQQ/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="883" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjQSNb81YJ5yTpdtwlu7GOvNARwvgNd5wMiaySgfwZiaKBe2WQp5RUnoic0cZnVx1xpdwx4Vf6jokFnQQ/640?wx_fmt=png&amp;from=appmsg">从文献中对Scissor的概述来看，Scissor的核心计算步骤采用的是<strong>皮尔逊相关性分析</strong>，那么<strong>皮尔逊相关性分析的主要优势是计算高效、适用于大规模数据，且能够很好地衡量bulk数据和单细胞数据的线性关系，局限是仅适用于线性相关性，对非线性关系的处理较弱</strong>。</p><p data-tool="mdnice编辑器">但测序数据就是高维数据，笔者从目前的知识储备来理解，线性相关性方法是在大样本情况下更加稳健，高维数据往往包含大量噪声，非线性方法（如Spearman秩相关性等方法）在小样本、非线性情况下可能表现良好，但在高维数据中容易受数据稀疏性影响，导致计算不稳定或过拟合。还有就是Scissor的核心任务是将 bulk RNA-seq数据与单细胞数据进行匹配，即找到与bulk数据模式最接近的单细胞群体。由于 bulk RNA-seq 反映的是大量细胞的加权平均表达，其主要变化趋势通常表现就是为线性模式。</p><h5 data-tool="mdnice编辑器"><span></span>2.BayesPrism<span></span></h5><p data-tool="mdnice编辑器">那么从单细胞数据映射到bulkRN数据我们通常采用的是BayesPrism，其作为一种工具可以根据scRNA数据(作为先验模型)去推断bulkRNA数据中肿瘤微环境组成(不同免疫细胞组分/不同细胞群)和基因表达情况。开发者展示的图片就很形象了，左边图展示了把标注了不同细胞类型的单细胞数据作为先验信息(prior info)的基因信息和bulkRNA测序数据输入进贝叶斯棱镜算法，右图就是通过了棱镜算法之后把bulkRNA测序数据中每个样本里所占的标注细胞类型的细胞百分比展示出来。<img alt="" data-imgfileid="100044344" data-ratio="0.7196356275303644" data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjQSNb81YJ5yTpdtwlu7GOvN1TzTTPS58z6dgEgFpfo4AibA91DprscXFEbWI3nXniavfo6e4GogOicIA/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="988" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjQSNb81YJ5yTpdtwlu7GOvN1TzTTPS58z6dgEgFpfo4AibA91DprscXFEbWI3nXniavfo6e4GogOicIA/640?wx_fmt=png&amp;from=appmsg">从文献中对BayesPrism的概述来看，其通过贝叶斯建模，利用单细胞RNA-seq作为先验信息，联合推断bulk RNA-seq数据中的细胞类型比例和细胞类型特异性基因表达。该模型<strong>显式建模</strong>(显式建模是指在数学模型中<strong>明确地表达某些变量之间的关系</strong>，而不是隐含地或通过经验性调整来描述它们。 在 BayesPrism中，显式建模的核心思想是直接在模型中纳入单细胞RNA-seq 和 bulk RNA-seq 之间的基因表达差异，而不是假设它们完全一致)并边缘化(边缘化是概率论中的一个概念，指的是通过对某些未观测变量进行积分或求和，使其影响被消除，从而得到目标变量的分布。<strong>简单来说，就是通过整合掉某些变量，使得模型对关键变量的推断更加稳健</strong>)单细胞与bulk之间的基因表达差异，以减少技术和生物学偏差，从而提高推断准确性。在模型优化和推断过程中，BayesPrism采用后验分布推断(贝叶斯统计中，后验分布指的是在观测到数据之后，进而调整参数分布。<strong>它是基于先验知识和观测数据综合得出的结果</strong>)进行细胞类型比例估计，并结合单细胞参考数据预测bulk 样本中各细胞类型的基因表达水平，从而克服传统去卷积方法的局限性。</p><h4 data-tool="mdnice编辑器"><span></span>分析步骤——Scissor<span></span></h4><h5 data-tool="mdnice编辑器"><span></span>1.导入<span></span></h5><pre data-tool="mdnice编辑器"><span></span><code>rm(list=ls())<br><span>source</span>(<span>'scRNA_scripts/lib.R'</span>)<br><span>source</span>(<span>'scRNA_scripts/mycolors.R'</span>)<br><span>library</span>(BiocParallel)<br><span>library</span>(Scissor)<br>register(MulticoreParam(workers = <span>8</span>, progressbar = <span>TRUE</span>))  <span># 单线程模式</span><br>load(<span>"TCGA-COAD_sur_model.Rdata"</span>)<br>scRNA &lt;- qread(<span>"./4-Corrective-data/sce.all.qs"</span>)<br><br>dir.create(<span>"./7-Scissors"</span>)<br>setwd(<span>"./7-Scissors"</span>)<br></code></pre><h5 data-tool="mdnice编辑器"><span></span>2.预处理数据<span></span></h5><pre data-tool="mdnice编辑器"><span></span><code><span># Check TCGA-COAD</span><br>head(meta)[<span>1</span>:<span>5</span>,<span>1</span>:<span>5</span>]<br><span>#                                ID  OS.time OS                     histology gender</span><br><span># TCGA-AA-3818-01A TCGA-AA-3818-01A 1.000000  1          Colon Adenocarcinoma FEMALE</span><br><span># TCGA-AA-3543-01A TCGA-AA-3543-01A 1.000000  0 Colon Mucinous Adenocarcinoma   MALE</span><br><span># TCGA-AA-3856-01A TCGA-AA-3856-01A 1.000000  0          Colon Adenocarcinoma   MALE</span><br><span># TCGA-AA-A00R-01A TCGA-AA-A00R-01A 1.000000  0          Colon Adenocarcinoma FEMALE</span><br><span># TCGA-AA-A01V-01A TCGA-AA-A01V-01A 1.033333  0          Colon Adenocarcinoma   MALE</span><br>head(exprSet)[<span>1</span>:<span>5</span>,<span>1</span>:<span>5</span>]<br><span>#            TCGA-AA-3818-01A TCGA-AA-3543-01A TCGA-AA-3856-01A TCGA-AA-A00R-01A TCGA-AA-A01V-01A</span><br><span># WASH7P          0.003626661        0.6383137      0.001720891       0.06671424        1.1729266</span><br><span># AL627309.6      0.919089866        1.8961174      0.148456647       2.29267751        0.6987401</span><br><span># WASH9P          1.927197805        2.1416464      1.727985983       2.11144083        2.3731305</span><br><span># MTND1P23        1.792300493        1.4819065      2.676033573       3.46413905        5.0904344</span><br><span># MTND2P28        8.726735194        7.2472009      8.578934482       8.10135516        8.8684236</span><br><br>table(scRNA$celltype)<br><span>#       B cells              T/NK cells epithelial/cancer cells                  plasma </span><br><span>#          3686                   10160                    2268                    1540 </span><br><span># myeloid cells                   VSMCs     proliferative cells             fibroblasts </span><br><span>#          3198                     395                     765                     984 </span><br><span>#    mast cells       endothelial cells </span><br><span>#           568                     516 </span><br>sel.clust = <span>"celltype"</span><br>scRNA &lt;- SetIdent(scRNA, value = sel.clust)<br>table(scRNA@active.ident) <br></code></pre><h5 data-tool="mdnice编辑器"><span></span>2.由于Scissor不支持V5，因此需要修改代码<span></span></h5><p data-tool="mdnice编辑器">修改了数据读取方式：sc_exprs &lt;- as.matrix(sc_dataset@assay@layers$data)</p><pre data-tool="mdnice编辑器"><span></span><code>RUNScissor &lt;- <span>function</span> (bulk_dataset, sc_dataset, phenotype, tag = <span>NULL</span>, alpha = <span>NULL</span>, <br>                        cutoff = <span>0.2</span>, family = c(<span>"gaussian"</span>, <span>"binomial"</span>, <span>"cox"</span>), <br>                        Save_file = <span>"Scissor_inputs.RData"</span>, Load_file = <span>NULL</span>) <br>{<br>  <span>library</span>(Seurat)<br>  <span>library</span>(Matrix)<br>  <span>library</span>(preprocessCore)<br>  <br>  <span>if</span> (is.null(Load_file)) {<br>    common &lt;- intersect(rownames(bulk_dataset), rownames(sc_dataset))<br>    <span>if</span> (length(common) == <span>0</span>) {<br>      <span>stop</span>(<span>"There is no common genes between the given single-cell and bulk samples."</span>)<br>    }<br>    <br>    <span>if</span> (class(sc_dataset) == <span>"Seurat"</span>) {<br>      sc_exprs &lt;- as.matrix(sc_dataset@assays$RNA@layers$data)<br>      rownames(sc_exprs) &lt;- rownames(sc_dataset)<br>      colnames(sc_exprs) &lt;- colnames(sc_dataset)<br>      network &lt;- as.matrix(sc_dataset@graphs$RNA_snn)<br>    } <span>else</span> {<br>      sc_exprs &lt;- as.matrix(sc_dataset)<br>      Seurat_tmp &lt;- CreateSeuratObject(sc_dataset)<br>      Seurat_tmp &lt;- FindVariableFeatures(Seurat_tmp, selection.method = <span>"vst"</span>, verbose = <span>FALSE</span>)<br>      Seurat_tmp &lt;- ScaleData(Seurat_tmp, verbose = <span>FALSE</span>)<br>      Seurat_tmp &lt;- RunPCA(Seurat_tmp, features = VariableFeatures(Seurat_tmp), verbose = <span>FALSE</span>)<br>      Seurat_tmp &lt;- FindNeighbors(Seurat_tmp, dims = <span>1</span>:<span>10</span>, verbose = <span>FALSE</span>)<br>      network &lt;- as.matrix(Seurat_tmp@graphs$RNA_snn)<br>    }<br>    <br>    diag(network) &lt;- <span>0</span><br>    network[which(network != <span>0</span>)] &lt;- <span>1</span><br>    dataset0 &lt;- cbind(bulk_dataset[common, ], sc_exprs[common, ])<br>    dataset0 &lt;- as.matrix(dataset0)<br>    dataset1 &lt;- normalize.quantiles(dataset0)<br>    rownames(dataset1) &lt;- rownames(dataset0)<br>    colnames(dataset1) &lt;- colnames(dataset0)<br>    Expression_bulk &lt;- dataset1[, <span>1</span>:ncol(bulk_dataset)]<br>    Expression_cell &lt;- dataset1[, (ncol(bulk_dataset) + <span>1</span>):ncol(dataset1)]<br>    X &lt;- cor(Expression_bulk, Expression_cell)<br>    <br>    quality_check &lt;- quantile(X)<br>    print(<span>"|**************************************************|"</span>)<br>    print(<span>"Performing quality-check for the correlations"</span>)<br>    print(<span>"The five-number summary of correlations:"</span>)<br>    print(quality_check)<br>    print(<span>"|**************************************************|"</span>)<br>    <br>    <span>if</span> (quality_check[<span>3</span>] &lt; <span>0.01</span>) {<br>      <span>warning</span>(<span>"The median correlation between the single-cell and bulk samples is relatively low."</span>)<br>    }<br>    <br>    <span>if</span> (family == <span>"binomial"</span>) {<br>      Y &lt;- as.numeric(phenotype)<br>      z &lt;- table(Y)<br>      <span>if</span> (length(z) != length(tag)) {<br>        <span>stop</span>(<span>"The length differs between tags and phenotypes. Please check Scissor inputs and selected regression type."</span>)<br>      } <span>else</span> {<br>        print(sprintf(<span>"Current phenotype contains %d %s and %d %s samples."</span>, z[<span>1</span>], tag[<span>1</span>], z[<span>2</span>], tag[<span>2</span>]))<br>        print(<span>"Perform logistic regression on the given phenotypes:"</span>)<br>      }<br>    }<br>    <br>    <span>if</span> (family == <span>"gaussian"</span>) {<br>      Y &lt;- as.numeric(phenotype)<br>      z &lt;- table(Y)<br>      <span>if</span> (length(z) != length(tag)) {<br>        <span>stop</span>(<span>"The length differs between tags and phenotypes. Please check Scissor inputs and selected regression type."</span>)<br>      } <span>else</span> {<br>        tmp &lt;- paste(z, tag)<br>        print(paste0(<span>"Current phenotype contains "</span>, paste(tmp[<span>1</span>:(length(z) - <span>1</span>)], collapse = <span>", "</span>), <span>", and "</span>, tmp[length(z)], <span>" samples."</span>))<br>        print(<span>"Perform linear regression on the given phenotypes:"</span>)<br>      }<br>    }<br>    <br>    <span>if</span> (family == <span>"cox"</span>) {<br>      Y &lt;- as.matrix(phenotype)<br>      <span>if</span> (ncol(Y) != <span>2</span>) {<br>        <span>stop</span>(<span>"The size of survival data is wrong. Please check Scissor inputs and selected regression type."</span>)<br>      } <span>else</span> {<br>        print(<span>"Perform cox regression on the given clinical outcomes:"</span>)<br>      }<br>    }<br>    <br>    save(X, Y, network, Expression_bulk, Expression_cell, file = Save_file)<br>  } <span>else</span> {<br>    load(Load_file)<br>  }<br>  <br>  <span>if</span> (is.null(alpha)) {<br>    alpha &lt;- c(<span>0.005</span>, <span>0.01</span>, <span>0.05</span>, <span>0.1</span>, <span>0.2</span>, <span>0.3</span>, <span>0.4</span>, <span>0.5</span>, <span>0.6</span>, <span>0.7</span>, <span>0.8</span>, <span>0.9</span>)<br>  }<br>  <br>  <span>for</span> (i <span>in</span> <span>1</span>:length(alpha)) {<br>    set.seed(<span>123</span>)<br>    fit0 &lt;- APML1(X, Y, family = family, penalty = <span>"Net"</span>, alpha = alpha[i], Omega = network, nlambda = <span>100</span>, nfolds = min(<span>10</span>, nrow(X)))<br>    fit1 &lt;- APML1(X, Y, family = family, penalty = <span>"Net"</span>, alpha = alpha[i], Omega = network, lambda = fit0$lambda.min)<br>    <br>    <span>if</span> (family == <span>"binomial"</span>) {<br>      Coefs &lt;- as.numeric(fit1$Beta[<span>2</span>:(ncol(X) + <span>1</span>)])<br>    } <span>else</span> {<br>      Coefs &lt;- as.numeric(fit1$Beta)<br>    }<br>    <br>    Cell1 &lt;- colnames(X)[which(Coefs &gt; <span>0</span>)]<br>    Cell2 &lt;- colnames(X)[which(Coefs &lt; <span>0</span>)]<br>    percentage &lt;- (length(Cell1) + length(Cell2)) / ncol(X)<br>    print(sprintf(<span>"alpha = %s"</span>, alpha[i]))<br>    print(sprintf(<span>"Scissor identified %d Scissor+ cells and %d Scissor- cells."</span>, length(Cell1), length(Cell2)))<br>    print(sprintf(<span>"The percentage of selected cell is: %s%%"</span>, formatC(percentage * <span>100</span>, format = <span>"f"</span>, digits = <span>3</span>)))<br>    <br>    <span>if</span> (percentage &lt; cutoff) {<br>      <span>break</span><br>    }<br>    cat(<span>"\n"</span>)<br>  }<br>  <br>  print(<span>"|**************************************************|"</span>)<br>  <span>return</span>(list(para = list(alpha = alpha[i], lambda = fit0$lambda.min, family = family), Coefs = Coefs, Scissor_pos = Cell1, Scissor_neg = Cell2))<br>}<br></code></pre><h6 data-tool="mdnice编辑器"><span></span>2.1 二分类数据<span></span></h6><ol data-tool="mdnice编辑器"><li><section>请注意单细胞数据集的细胞量，细胞量稍一大就会报错；</section></li><li><section>似乎对bulkRNA和phenotype的数据格式没有特殊要求，或者可以考虑全部改成matrix；</section></li><li><section>tag中的顺序需要与phenotype中顺序一致；</section></li></ol><pre data-tool="mdnice编辑器"><span></span><code><span># 单细胞数据</span><br><span># 按比例取样</span><br>set.seed(<span>123</span>)<br>sc_dataset = scRNA[, sample(<span>1</span>:ncol(scRNA),round(ncol(scRNA)/<span>10</span>)) ]<br>sc_dataset &lt;- FindNeighbors(sc_dataset, dims = <span>1</span>:<span>20</span>) <br>table(sc_dataset$celltype)<br><br><span># 单细胞数据 </span><br><span># sc_dataset &lt;- scRNA</span><br><br><span># bulk数据##############</span><br><span># 二分类数据，后续要使用binomial</span><br><span># 分析时数据中不能存在na数据，去除或者改成0(需要提前确认！)</span><br>class(exprSet)<br>bulk_dataset &lt;- as.matrix(exprSet)<br><span>#bulk_dataset &lt;- na.omit(bulk_dataset)</span><br><span>#bulk_dataset[is.na(bulk_dataset)] &lt;- 0</span><br>class(bulk_dataset)<br><br><span># 提取phenotype</span><br>identical(rownames(meta),colnames(exprSet))<br>meta$cluster &lt;- ifelse(meta$OS == <span>"1"</span>, <span>"1"</span>, <span>"0"</span>)  <span># 先将 OS 变量转换为二分类</span><br>phenotype &lt;- as.numeric(meta[,<span>"cluster"</span>])<br>class(phenotype)<br>phenotype<br><br><span># tag</span><br>table(phenotype)<br><span># phenotype</span><br><span>#   0   1 </span><br><span># 332  92 </span><br>tag &lt;- c(<span>"Live"</span>,<span>"Dead"</span>) <br><br><span># 保存数据</span><br>save(bulk_dataset,phenotype,tag,file = <span>"bulkData.Rdata"</span>)<br>qsave(sc_dataset,<span>"sc_dataset.qs"</span>)<br></code></pre><p data-tool="mdnice编辑器"><strong>开始运行+可视化</strong></p><ol data-tool="mdnice编辑器"><li><section>alpha默认值是0.05，越大惩罚力度越大；alpha需要多调整</section></li><li><section>cutoff默认20%，不要超过单细胞总数的50%</section></li><li><section>family = binomial</section></li></ol><pre data-tool="mdnice编辑器"><span></span><code><span># V5</span><br>infos1 &lt;- RUNScissor(bulk_dataset, <br>                     sc_dataset, <br>                     phenotype, <br>                     tag = tag,<br>                     alpha = <span>0.001</span>, <span># 默认0.05</span><br>                     cutoff = <span>0.2</span>, <span># Scissor选择的细胞不能超过单细胞总数的50%，默认20%</span><br>                     family = <span>"binomial"</span>,  <span># 生存数据需要使用cox</span><br>                     Save_file = <span>'./result.RData'</span>)<br><span># [1] "|**************************************************|"</span><br><span># [1] "Performing quality-check for the correlations"</span><br><span># [1] "The five-number summary of correlations:"</span><br><span>#        0%       25%       50%       75%      100% </span><br><span># 0.1395088 0.3866064 0.4255620 0.4629632 0.7986783 </span><br><span># [1] "|**************************************************|"</span><br><span># [1] "Current phenotype contains 332 Live and 92 Dead samples."</span><br><span># [1] "Perform logistic regression on the given phenotypes:"</span><br><span># [1] "alpha = 0.001"</span><br><span># [1] "Scissor identified 2281 Scissor+ cells and 2368 Scissor- cells."</span><br><span># [1] "The percentage of selected cell is: 96.532%"</span><br><span># </span><br><span># [1] "|**************************************************|"</span><br>infos1 &lt;- qread(<span>"infos1.qs"</span>)<br><br><span># 可视化</span><br>Scissor_select &lt;- rep(<span>0</span>, ncol(sc_dataset))<br>names(Scissor_select) &lt;- colnames(sc_dataset)<br>Scissor_select[infos1$Scissor_pos] &lt;- <span>"Scissor+"</span><br>Scissor_select[infos1$Scissor_neg] &lt;- <span>"Scissor-"</span><br>sc_dataset &lt;- AddMetaData(sc_dataset, <br>                          metadata = Scissor_select, <br>                          col.name = <span>"scissor"</span>)<br>UMAP_scissor &lt;- DimPlot(sc_dataset, reduction = <span>'umap'</span>, <br>                        group.by = <span>'scissor'</span>,<br>                        cols = c(<span>'grey'</span>,<span>'#0d1b46'</span>,<span>'tomato'</span>), <br>                        pt.size = <span>0.001</span>, order = c(<span>"Scissor+"</span>,<span>"Scissor-"</span>))<br>UMAP_scissor+DimPlot(sc_dataset,group.by = <span>"celltype"</span>)<br>ggsave(<span>"scissor.pdf"</span>,width = <span>18</span>,height = <span>7</span>)<br></code></pre><p data-tool="mdnice编辑器">我们粗略的看一下这张图(肉眼看比例哈，因为这里的alpha值是随便设定的)，scissor+细胞主要集中于mast，endothelial，VSMCs，T/NK细胞还有部分的myeloid细胞，这个结果其实让我有点犹豫和困惑，比如T细胞几乎都是阳性细胞，并且癌症细胞中没有阳性细胞。接下来我们把时间也加上去重新分析看看。<img alt="" data-imgfileid="100044347" data-ratio="0.5694444444444444" data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjQSNb81YJ5yTpdtwlu7GOvNZ4aj4kQXQ2tMkvoP3mEsVNEgQCx8ibYPtvR4NnvKpwT9BAol9DNbjsg/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjQSNb81YJ5yTpdtwlu7GOvNZ4aj4kQXQ2tMkvoP3mEsVNEgQCx8ibYPtvR4NnvKpwT9BAol9DNbjsg/640?wx_fmt=png&amp;from=appmsg"></p><h5 data-tool="mdnice编辑器"><span></span>2.1 生存数据<span></span></h5><ol data-tool="mdnice编辑器"><li><section>不需要tag数据</section></li><li><section>需要提取时间和状态数据，重命名为time和state</section></li><li><section>官方中time数据类型为num，state数据类型为dbl，因此phenotype应当需要为数值型</section></li></ol><pre data-tool="mdnice编辑器"><span></span><code><span># 按比例取样</span><br>set.seed(<span>123</span>)<br>sc_dataset = scRNA[, sample(<span>1</span>:ncol(scRNA),round(ncol(scRNA)/<span>5</span>)) ]<br>sc_dataset &lt;- FindNeighbors(sc_dataset, dims = <span>1</span>:<span>30</span>) <br>table(sc_dataset$celltype)<br><br><span># 单细胞数据(如果数据量很小的话) </span><br><span># sc_dataset &lt;- scRNA</span><br><br><span># bulk数据##############</span><br><span># 分析时数据中不能存在na数据，去除或者改成0(需要提前确认！)</span><br>class(exprSet)<br>bulk_dataset &lt;- as.matrix(exprSet)<br><span>#bulk_dataset &lt;- na.omit(bulk_dataset)</span><br><span>#bulk_dataset[is.na(bulk_dataset)] &lt;- 0</span><br><br><span># 提取phenotype</span><br>identical(rownames(meta),colnames(exprSet))<br>phenotype &lt;- meta[,c(<span>"OS.time"</span>,<span>"OS"</span>)]<br>colnames(phenotype) &lt;- c(<span>"time"</span>, <span>"status"</span>)<br>str(phenotype)<br><span># 'data.frame': 424 obs. of  2 variables:</span><br><span>#  $ time  : num  1 1 1 1 1.03 ...</span><br><span>#  $ status: int  1 0 0 0 0 0 0 0 0 0 ...</span><br>head(phenotype)<br><span>#                      time status</span><br><span># TCGA-AA-3818-01A 1.000000      1</span><br><span># TCGA-AA-3543-01A 1.000000      0</span><br><span># TCGA-AA-3856-01A 1.000000      0</span><br><span># TCGA-AA-A00R-01A 1.000000      0</span><br><span># TCGA-AA-A01V-01A 1.033333      0</span><br><span># TCGA-AA-3496-01A 1.033333      0</span><br><br><span># 保存数据</span><br>save(bulk_dataset,phenotype,tag,file = <span>"bulkData.Rdata"</span>)<br>qsave(sc_dataset,<span>"sc_dataset.qs"</span>)<br></code></pre><p data-tool="mdnice编辑器"><strong>开始运行+可视化</strong></p><ol data-tool="mdnice编辑器"><li><section>alpha默认值是0.05，越大惩罚力度越大；alpha需要多调整</section></li><li><section>cutoff默认20%，不要超过单细胞总数的50%</section></li><li><section>family = cox</section></li></ol><pre data-tool="mdnice编辑器"><span></span><code><span># V5</span><br>infos1 &lt;- RUNScissor(bulk_dataset, <br>                     sc_dataset, <br>                     phenotype, <br>                     <span>#tag = tag,</span><br>                     alpha = <span>0.00025</span>, <span># 默认0.05</span><br>                     cutoff = <span>0.2</span>, <span># Scissor选择的细胞不能超过单细胞总数的50%，默认20%</span><br>                     family = <span>"cox"</span>,  <span># 生存数据需要使用cox</span><br>                     Save_file = <span>'./result.RData'</span>)<br><span># [1] "|**************************************************|"</span><br><span># [1] "Performing quality-check for the correlations"</span><br><span># [1] "The five-number summary of correlations:"</span><br><span>#        0%       25%       50%       75%      100% </span><br><span># 0.1395088 0.3866064 0.4255620 0.4629632 0.7986783 </span><br><span># [1] "|**************************************************|"</span><br><span># [1] "Perform cox regression on the given clinical outcomes:"</span><br><span># [1] "alpha = 0.00025"</span><br><span># [1] "Scissor identified 1554 Scissor+ cells and 398 Scissor- cells."</span><br><span># [1] "The percentage of selected cell is: 40.532%"</span><br><span># </span><br><span># [1] "|**************************************************|"</span><br>infos1 &lt;- qread(<span>"infos1.qs"</span>)<br></code></pre><p data-tool="mdnice编辑器">我们粗略的看一下这张图，scissor+细胞主要集中于mast，endothelial，VSMCs，fibroblasts,部分T/NK和B细胞还有部分的myeloid细胞。除了增加了fibroblasts进入了阳性细胞行列，其他细胞的情况是类似的。</p><p data-tool="mdnice编辑器">那么其他类似的细胞系暂且不谈，笔者思考了一下<strong>为什么成纤维细胞在这两种不同的分析方式中存在不同的情况。有一种可能性是在二分类分析中，成纤维细胞被归为 Scissor-，说明其在存活患者中更富集，或在死亡患者中表达较低，这是一个整体表达的情况。然而，在 Scissor结合生存分析时，成纤维细胞变为Scissor+，表明其与较短生存时间相关，可能在长期预后上促进肿瘤进展或免疫抑制</strong>。因此其实这个结果挺有趣的，<strong>但到底成纤维细胞和疾病进展(我认为癌症中的死亡很大程度上是因为疾病进展)是怎么样的因果关系还是值得推敲的(比如是成纤维细胞加速了疾病进展？还是疾病本身进展的同时导致了呈现成纤维细胞的异常？当然最有可能的是这两种情况都存在)</strong> 。<img alt="" data-imgfileid="100044346" data-ratio="0.524074074074074" data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjQSNb81YJ5yTpdtwlu7GOvNdzxUnkyB90mc310icLmhnO2gib7yO3VOgJ4Zf3QgK817b5WkKyE3mh8g/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjQSNb81YJ5yTpdtwlu7GOvNdzxUnkyB90mc310icLmhnO2gib7yO3VOgJ4Zf3QgK817b5WkKyE3mh8g/640?wx_fmt=png&amp;from=appmsg"></p><h4 data-tool="mdnice编辑器"><span></span>分析步骤——BayesPrism<span></span></h4><h5 data-tool="mdnice编辑器"><span></span>1.导入<span></span></h5><pre data-tool="mdnice编辑器"><span></span><code>rm(list = ls())<br><span>library</span>(BayesPrism)<br><span>library</span>(TCGAbiolinks)<br><span>library</span>(stringr)<br><span>library</span>(SummarizedExperiment)<br><span>library</span>(Seurat)<br><span>library</span>(qs)<br><span>library</span>(BiocParallel)<br>BiocParallel::register(MulticoreParam(workers = <span>8</span>, progressbar = <span>TRUE</span>))<br>load(<span>"TCGA-COAD_sur_model.Rdata"</span>)<br>scRNA &lt;- qread(<span>"./4-Corrective-data/sce.all.qs"</span>)<br><br>dir.create(<span>"./8-BayesPrism/"</span>)<br>setwd(<span>"./8-BayesPrism"</span>)<br></code></pre><h5 data-tool="mdnice编辑器"><span></span>2.TCGA预处理<span></span></h5><pre data-tool="mdnice编辑器"><span></span><code><span>#请注意我们要做BayesPrism分析</span><br><span>#请确保使用一致的基因注释（TCGA RNA-seq 使用 GENCODE v22）。</span><br><span>#建议使用未规范化和未转换（ unnormalized and untransformed）的count data。当原始计数不可用时，线性归一化（例如 TPM、RPM、RPKM、FPKM）也是可以接受的，因为 BayesPrism 对reference和mixture之间的线性乘法差异具有鲁棒性（换个单词 即Robust）。理想情况下，如果使用规范化数据，最好提供参考数据和通过相同方法转换的批量数据。应避免log转换。因此建议使用小洁老师旧的探针注释方式</span><br>proj = <span>"TCGA-COAD"</span><br>exp &lt;- <span>2</span>^exprSet - <span>1</span><br>exp_pre &lt;- t(exp)<br></code></pre><h5 data-tool="mdnice编辑器"><span></span>3.scRNA数据预处理<span></span></h5><pre data-tool="mdnice编辑器"><span></span><code><span># 单细胞数据</span><br>sc_dataset &lt;- scRNA<br><span># 按比例抽样</span><br>set.seed(<span>123</span>)<br>sc_dataset = sc_dataset[, sample(<span>1</span>:ncol(sc_dataset),round(ncol(sc_dataset)/<span>5</span>)) ]<br>table(sc_dataset$celltype)<br><br><span>#sc_dat &lt;- as.data.frame(sc_dataset@assays$RNA@counts) # V4</span><br>sc_dat &lt;- GetAssayData(sc_dataset,layer = <span>"counts"</span>) <span># V5</span><br>sc_dat_pre &lt;- t(sc_dat)<br>table(sc_dataset$celltype)<br><br><span># 保存一下</span><br>qsave(sc_dat_pre,<span>"sc_dat_pre.qs"</span>)<br>qsave(sc_dataset,<span>"sc_dataset.qs"</span>)<br>qsave(exp_pre,<span>"exp_pre.qs"</span>)<br></code></pre><h5 data-tool="mdnice编辑器"><span></span>4.对状态和细胞类型进行质量控制<span></span></h5><pre data-tool="mdnice编辑器"><span></span><code><span># 对细胞状态进行质量控制</span><br><span># plot.cor.phi(input=sc_dat_pre,</span><br><span>#               input.labels=sc_dataset$status,</span><br><span>#               title="cell state correlation",</span><br><span>#               #specify pdf.prefix if need to output to pdf</span><br><span>#               #pdf.prefix="gbm.cor.cs",</span><br><span>#               cexRow=0.2, cexCol=0.2,</span><br><span>#               margins=c(2,2))</span><br><span># 对细胞类型进行质量控制</span><br>plot.cor.phi (input=sc_dat_pre, <br>              input.labels=sc_dataset$celltype, <br>              title=<span>"cell type correlation"</span>,<br>              <span>#specify pdf.prefix if need to output to pdf</span><br>              <span>#pdf.prefix="gbm.cor.ct",</span><br>              cexRow=<span>0.5</span>, cexCol=<span>0.5</span>,<br>)<br></code></pre><p data-tool="mdnice编辑器">笔者认为细胞状态参数有助于我们对肿瘤细胞或髓系细胞在不同样本中的状态进行细分，从而更好地揭示细胞群体的异质性。例如，整合转移、复发和原发样本的数据，并对细胞进行标注，有助于更细致地探索各细胞亚群的情况。<img alt="" data-imgfileid="100044343" data-ratio="0.8432835820895522" data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjQSNb81YJ5yTpdtwlu7GOvNicb5lMxoteTd76HvibMfM5PECiaF12Tics1ibYlDmxkOR0voTBCjzLDicmVA/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="804" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjQSNb81YJ5yTpdtwlu7GOvNicb5lMxoteTd76HvibMfM5PECiaF12Tics1ibYlDmxkOR0voTBCjzLDicmVA/640?wx_fmt=png&amp;from=appmsg"><img alt="" data-imgfileid="100044348" data-ratio="0.37922077922077924" data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjQSNb81YJ5yTpdtwlu7GOvNydr9ayfQOqpzNeHT4oEZZCoINLuib1vxMsiaOc2M7mpYwFzGFq0lxxSw/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="770" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjQSNb81YJ5yTpdtwlu7GOvNydr9ayfQOqpzNeHT4oEZZCoINLuib1vxMsiaOc2M7mpYwFzGFq0lxxSw/640?wx_fmt=png&amp;from=appmsg"></p><ol data-tool="mdnice编辑器"><li><section>不同的细胞状态要保留至少大于20或者50个细胞</section></li><li><section>纳入分析的数据建议 1）不同细胞类型是具有足够多显著差异表达基因的细胞群体，例如大于50个或100个。<strong>对于转录组相似度过高的细胞群体，建议将其视为细胞状态</strong>，在最终的Gibbs采样之前将其聚合。<strong>因此，细胞状态通常适用于那些在表型空间中形成连续体的细胞，而非形成离散簇的细胞</strong>。2）为具有显著异质性的细胞类型定义多个细胞状态，例如恶性细胞等，以便解析其转录组。</section></li></ol><p data-tool="mdnice编辑器">可以看出每个细胞群分的还算清晰<img alt="" data-imgfileid="100044349" data-ratio="0.9287037037037037" data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjQSNb81YJ5yTpdtwlu7GOvNe4uLyic02Oibl5icfNicW2ASafbMtTSKlwuRBn1f6EEyu5IZfKZk3nYclA/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjQSNb81YJ5yTpdtwlu7GOvNe4uLyic02Oibl5icfNicW2ASafbMtTSKlwuRBn1f6EEyu5IZfKZk3nYclA/640?wx_fmt=png&amp;from=appmsg"></p><h5 data-tool="mdnice编辑器"><span></span>5.过滤离群基因-单细胞和bulk<span></span></h5><p data-tool="mdnice编辑器">表达量较高的基因，如核糖体蛋白基因和线粒体基因，可能主导数据的分布并偏倚推断结果。这些基因通常在区分细胞类型方面信息较少，并且可能引入较大的虚假方差。因此，它们可能对去卷积分析产生不利影响。因此需要去除这些基因。</p><pre data-tool="mdnice编辑器"><span></span><code><span>#查看单细胞数据的离群基因</span><br><span>#sc.stat 显示每个基因的归一化平均表达(x 轴)和最大特异性(y 轴)的对数，以及每个基因是否属于一个潜在的异常值类别。Other _ Rb 是由核糖体假基因组成的一组精选基因。</span><br>sc.stat &lt;- plot.scRNA.outlier(<br>  input=sc_dat_pre, <span>#make sure the colnames are gene symbol or ENSMEBL ID </span><br>  cell.type.labels=sc_dataset$celltype,<br>  species=<span>"hs"</span>, <span>#currently only human(hs) and mouse(mm) annotations are supported</span><br>  return.raw=<span>TRUE</span> <span>#return the data used for plotting. </span><br>  <span>#pdf.prefix="gbm.sc.stat" specify pdf.prefix if need to output to pdf</span><br>)<br>head(sc.stat)  <br><br><span>#查看bulk细胞数据的离群基因</span><br>bk.stat &lt;- plot.bulk.outlier(<br>  bulk.input=exp_pre,<span>#make sure the colnames are gene symbol or ENSMEBL ID </span><br>  sc.input=sc_dat_pre, <span>#make sure the colnames are gene symbol or ENSMEBL ID </span><br>  cell.type.labels=sc_dataset$celltype,<br>  species=<span>"hs"</span>, <span>#currently only human(hs) and mouse(mm) annotations are supported</span><br>  return.raw=<span>TRUE</span><br>  <span>#pdf.prefix="gbm.bk.stat" specify pdf.prefix if need to output to pdf</span><br>)<br>head(bk.stat)<br><br><span>#过滤基因</span><br>sc.dat.filtered &lt;- cleanup.genes (input=sc_dat_pre,<br>                                   input.type=<span>"count.matrix"</span>,<br>                                   species=<span>"hs"</span>, <br>                                   gene.group=c( <span>"Rb"</span>,<span>"Mrp"</span>,<span>"other_Rb"</span>,<br>                                                 <span>"chrM"</span>,<span>"MALAT1"</span>,<span>"chrX"</span>,<span>"chrY"</span>) ,<br>                                   exp.cells=<span>5</span>)<br>plot.bulk.vs.sc (sc.input = sc.dat.filtered,<br>                 bulk.input = exp_pre<br>                 <span>#pdf.prefix="gbm.bk.vs.sc" specify pdf.prefix if need to output to pdf</span><br>)<br></code></pre><p data-tool="mdnice编辑器"><img alt="" data-imgfileid="100044351" data-ratio="0.9287037037037037" data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjQSNb81YJ5yTpdtwlu7GOvNjghqTjUgDy7PNB0kc4ibsF6e44Nbz6cwmPcSI4InVBg9YWSCzuEGZdQ/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjQSNb81YJ5yTpdtwlu7GOvNjghqTjUgDy7PNB0kc4ibsF6e44Nbz6cwmPcSI4InVBg9YWSCzuEGZdQ/640?wx_fmt=png&amp;from=appmsg"><img alt="" data-imgfileid="100044352" data-ratio="0.9287037037037037" data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjQSNb81YJ5yTpdtwlu7GOvNqzpIcqVjJvnlWUNibRmCtMtQpLIMQ4tT1mmJKX4UZ3Vh8HVBLuMwCZw/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjQSNb81YJ5yTpdtwlu7GOvNqzpIcqVjJvnlWUNibRmCtMtQpLIMQ4tT1mmJKX4UZ3Vh8HVBLuMwCZw/640?wx_fmt=png&amp;from=appmsg"></p><h5 data-tool="mdnice编辑器"><span></span>6.根据 “蛋白编码基因”确定子集。<span></span></h5><p data-tool="mdnice编辑器">接下来，我们检查不同类型基因(scRNA和bulkRNA)的表达一致性。</p><pre data-tool="mdnice编辑器"><span></span><code>sc.dat.filtered.pc &lt;-  select.gene.type(sc.dat.filtered,<br>                                        gene.type = <span>"protein_coding"</span>)<br><span>#根据‘signature gene’确定子集</span><br><span>#或者，在细胞类型被定义为其中一些显示非常相似的转录或存在严重的批次效应的情况下，例如，reference 和mixture 来自不同的测定手段，选择signature gene是可行的。这是因为特征基因的选择可以丰富用于反卷积的基因信息，同时减少技术批量效应造成的噪声影响。</span><br></code></pre><p data-tool="mdnice编辑器">基本上蛋白基因的表达一致性分数还是挺高的。<img alt="" data-imgfileid="100044350" data-ratio="0.9287037037037037" data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjQSNb81YJ5yTpdtwlu7GOvNtC9FTxDwMH6YHMricay3937UyaDswG0Af15yCOExEjxzWNfASNpricTA/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjQSNb81YJ5yTpdtwlu7GOvNtC9FTxDwMH6YHMricay3937UyaDswG0Af15yCOExEjxzWNfASNpricTA/640?wx_fmt=png&amp;from=appmsg"></p><h5 data-tool="mdnice编辑器"><span></span>7.构建BayesPrism模型<span></span></h5><pre data-tool="mdnice编辑器"><span></span><code>myPrism &lt;- new.prism(<br>   reference=sc.dat.filtered.pc, <br>   mixture=exp_pre,<br>   input.type=<span>"count.matrix"</span>, <br>   cell.type.labels = sc_dataset$celltype, <br>   cell.state.labels = <span>NULL</span>, <span># 不同的细胞状态不能分配到同一细胞类型中</span><br>   key= <span>"epithelial/cancer cells"</span>,   <span>#"Malignant cells",#key是 cell.type.label 中对应于恶性细胞类型的字符向量。如无恶性细胞，即设置为 NULL，此时，所有类型的细胞将被 treated equally。</span><br>   outlier.cut=<span>0.01</span>,<br>   outlier.fraction=<span>0.1</span>,<br> )<br></code></pre><h5 data-tool="mdnice编辑器"><span></span>8.运行BayesPrism<span></span></h5><pre data-tool="mdnice编辑器"><span></span><code><span>#后续即运行BayesPrism控制 Gibbs 采样和优化的参数可以使用 Gibbs.control 和 opt.control 指定。?run.prism获取details。这里建议使用默认参数</span><br>bp.res &lt;- run.prism(prism = myPrism, n.cores=<span>20</span>)<br><br>qsave(bp.res,file = <span>"bp.res.qs"</span>)<br></code></pre><h5 data-tool="mdnice编辑器"><span></span>9. 提取细胞型分数 θ 的后验平均值/变异分数(CV)/Z分数<span></span></h5><pre data-tool="mdnice编辑器"><span></span><code>bp.res &lt;- qread(<span>"bp.res.qs"</span>)<br><span># 提取细胞型分数θ的后验平均值</span><br>theta &lt;- get.fraction (bp=bp.res,<br>            which.theta=<span>"final"</span>,<br>            state.or.type=<span>"type"</span>)<br><br>head(theta)<br><span>#                  epithelial/cancer cells       plasma      B cells   T/NK cells myeloid cells proliferative cells</span><br><span># TCGA-AA-3818-01A               0.9271370 6.325108e-05 1.375325e-04 0.0004325820    0.01680760        0.0004305814</span><br><span># TCGA-AA-3543-01A               0.8562443 1.974167e-04 3.342300e-05 0.0085692794    0.05991585        0.0042596279</span><br><span># TCGA-AA-3856-01A               0.8191022 9.401481e-04 2.925654e-02 0.0004518154    0.07363352        0.0119511269</span><br><span># TCGA-AA-A00R-01A               0.6588630 5.125138e-05 3.889203e-05 0.0106335478    0.08606630        0.1664692259</span><br><span># TCGA-AA-A01V-01A               0.9290664 6.600326e-05 3.966345e-03 0.0001245812    0.01926844        0.0005808367</span><br><span># TCGA-AA-3496-01A               0.7466722 6.637973e-04 4.654608e-03 0.0028542579    0.06975685        0.0024060684</span><br><span>#                  fibroblasts   mast cells endothelial cells        VSMCs</span><br><span># TCGA-AA-3818-01A  0.02857603 6.929316e-05        0.02026500 0.0060810801</span><br><span># TCGA-AA-3543-01A  0.05132464 7.272148e-04        0.01832902 0.0003992496</span><br><span># TCGA-AA-3856-01A  0.03616049 4.760719e-04        0.02599443 0.0020336560</span><br><span># TCGA-AA-A00R-01A  0.04469131 2.575208e-05        0.03237835 0.0007823437</span><br><span># TCGA-AA-A01V-01A  0.02405473 1.266463e-04        0.02253263 0.0002133477</span><br><span># TCGA-AA-3496-01A  0.11691740 2.093585e-04        0.05484618 0.0010192445</span><br>write.csv(theta,<span>"theta.csv"</span>)<br></code></pre><p data-tool="mdnice编辑器">得到了每个样本的分析就很好办了~ 比如就可以按照不同样本中不同细胞类型的中位值将样本重新分组，然后进行表达量或者生存分析~</p><h4 data-tool="mdnice编辑器"><span></span>参考资料：<span></span></h4><ol data-tool="mdnice编辑器"><li><section>Identifying phenotype-associated subpopulations by integrating bulk and single-cell sequencing data.  Nat Biotechnol. 2022 Apr;40(4):527-538. Nucleic Acids Res. 2022 Nov 28;50(21):12112-12130.</section></li><li><section>scAB detects multiresolution cell states with clinical significance by integrating single-cell genomics and bulk sequencing data.</section></li><li><section>Cell type and gene expression deconvolution with BayesPrism enables Bayesian integrative analysis across bulk and single-cell RNA sequencing in oncology. Nat Cancer. 2022 Apr;3(4):505-517.</section></li><li><section>Scissor：https://github.com/sunduanchen/Scissor</section></li><li><section>scAB：https://github.com/jinworks/scAB</section></li><li><section>BayesPrism：https://github.com/Danko-Lab/BayesPrism/tree/main</section></li><li><section>生信技能树：https://mp.weixin.qq.com/s/swPDu2PoGTBEutzaR7YYSg https://mp.weixin.qq.com/s/3dZQxDdY6M1WwMMcus5Gkg</section></li></ol><p data-tool="mdnice编辑器"><strong>致谢</strong>：感谢曾老师以及生信技能树团队全体成员。更多精彩内容可关注公众号：<strong>生信技能树，单细胞天地，生信菜鸟团</strong>等公众号。</p><p data-tool="mdnice编辑器"><strong>注</strong>：若对内容有疑惑或者有发现明确错误的朋友，请联系后台(欢迎交流)。</p><span>- END -</span></section><p><br></p><p><mp-style-type data-value="3"></mp-style-type></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/NTb0gHcKEMOkD2wJ-DHd5Q",target="_blank" rel="noopener noreferrer">原文链接</a>
