---
title: "《单细胞最佳实践》——数据整合部分详细解读"
date: 2025-09-15T00:15:36Z
draft: ["false"]
tags: [
  "fetched",
  "生信菜鸟团"
]
categories: ["Acdemic"]
---
《单细胞最佳实践》——数据整合部分详细解读 by 生信菜鸟团
------
<div><section><span leaf=""><br></span></section><section><span leaf="">前期内容：</span></section><section><span leaf=""><a target="_blank" href="https://mp.weixin.qq.com/s?__biz=MzUzMTEwODk0Ng==&amp;mid=2247536042&amp;idx=1&amp;sn=afc12844aae04c8a576db26cc707c89b&amp;scene=21#wechat_redirect" textvalue="《单细胞最佳实践》—重新回顾质控部分的新理解" data-itemshowtype="0" linktype="text" data-linktype="2">《单细胞最佳实践》—重新回顾质控部分的新理解</a></span></section><section><span leaf=""><a target="_blank" href="https://mp.weixin.qq.com/s?__biz=MzUzMTEwODk0Ng==&amp;mid=2247536159&amp;idx=1&amp;sn=7e954ba72f26c62ea7da0239c1da7f63&amp;scene=21#wechat_redirect" textvalue="《python单细胞最佳实践》——归一化部分详细解读" data-itemshowtype="0" linktype="text" data-linktype="2">《python单细胞最佳实践》——归一化部分详细解读</a></span></section><section><span leaf=""><a target="_blank" href="https://mp.weixin.qq.com/s?__biz=MzUzMTEwODk0Ng==&amp;mid=2247536298&amp;idx=1&amp;sn=e54eb5d86df6faf5ef2c133bd0a68086&amp;scene=21#wechat_redirect" textvalue="《单细胞最佳实践》——特征选择、降维及聚类部分详细解读" data-itemshowtype="0" linktype="text" data-linktype="2">《单细胞最佳实践》——特征选择、降维及聚类部分详细解读</a></span></section><section><span leaf=""><a target="_blank" href="https://mp.weixin.qq.com/s?__biz=MzUzMTEwODk0Ng==&amp;mid=2247536417&amp;idx=1&amp;sn=2d748cc0e4805482ed53236dc524fb05&amp;scene=21#wechat_redirect" textvalue="《单细胞最佳实践》——注释部分详细解读" data-itemshowtype="0" linktype="text" data-linktype="2">《单细胞最佳实践》——注释部分详细解读</a></span></section><section><section><span leaf=""><a target="_blank" href="https://mp.weixin.qq.com/s?__biz=MzUzMTEwODk0Ng==&amp;mid=2247536586&amp;idx=1&amp;sn=7b0e82078ac7623d3d4a7c6b55904ba4&amp;scene=21#wechat_redirect" textvalue="《单细胞最佳实践》——自动注释部分详细解读 · 续" data-itemshowtype="0" linktype="text" data-linktype="2">《单细胞最佳实践》——自动注释部分详细解读 · 续</a></span></section><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com" data-pm-slice="0 0 []"><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">7.1 Motivation</span></span><span></span><span></span></h2><p data-tool="mdnice编辑器"><span leaf="">大多数 scRNA-seq 数据分析的一个核心挑战就是批次效应（batch effects）。批次效应是指由于在不同组别或“批次”中处理细胞而导致的测量表达水平的变化。通常，批次效应的来源多样且难以确定。一些批次效应来源可能是技术性的，例如样本处理、实验方案或测序深度的差异，但生物学效应，比如供体之间的差异、组织或采样位置，也经常被解释为批次效应。生物学因素是否应被视为批次效应，可能取决于实验设计和所提出的问题。去批次对于在联合分析中寻找跨批次数据的共同结构非常重要，通常只有在去除批次效应后，才能识别出先前被批次间差异所掩盖的稀有细胞群体。同时，去批次也帮助注释中 “query-to-reference mapping” 的方法，将多个数据集整合，从而回答那些任何一个单独数据集都无法解答的、更宏观、更复杂的生物学问题。</span></p><p data-tool="mdnice编辑器"><span leaf="">去除批次效应时必须做出两个关键选择：（1）方法和参数；（2）批次协变量。因为批次效应可以在数据生成的不同层级上产生（如不同的样本、供体、数据集等），批次协变量的选择表明了哪种级别的变异会被保留而哪种级别的变异会被去除。批次分辨率越细，去除的批次效应就越多，然而这种细的批次效应也可能与有意义的生物学信号混淆。因此，批次协变量的选择将要取决于你整合任务的目标。</span></p><h3 data-tool="mdnice编辑器"><span></span><span><span leaf="">7.1.1 整合模型的类别</span></span><span></span></h3><p data-tool="mdnice编辑器"><span leaf="">去除 scRNA-seq 中批次效应的方法通常（最多）由三个步骤组成：</span></p><ol><li><section><span leaf="">降维</span></section></li><li><section><span leaf="">建模并去除批次效应</span></section></li><li><section><span leaf="">投影回高维空间</span></section></li></ol><p data-tool="mdnice编辑器"><span leaf="">虽然步骤 2 是任何去批次方法的核心部分，但许多方法首先将数据投影到低维空间以提高信噪比，并在低维空间中进行批次校正以获得更好的性能，然后再将数据投影回高维特征空间，输出经过批次校正的基因表达矩阵。</span></p><p data-tool="mdnice编辑器"><span leaf="">批次效应去除方法可能在这三个步骤中每一个都有所不同，可能使用各种线性或非线性的降维方法、批次效应模型，并且可能输出不同格式的批次校正数据。总体而言，可以将去批次的方法分为 4 类，按发展顺序分别是全局模型、线性嵌入模型、基于图的方法和深度学习方法。</span></p><p data-tool="mdnice编辑器"><strong><span leaf="">全局模型（Global Model）：</span></strong><span leaf="">起源于 bulk RNA-seq，对所有细胞进行一致的缩放和/或平移（相加和/或相乘），常见例子是 </span><code><span leaf="">ComBat</span></code><span leaf="">。</span></p><p data-tool="mdnice编辑器"><strong><span leaf="">线性嵌入模型（Linear Embedding Model）：</span></strong><span leaf="">第一个单细胞特异的批次去除方法。这些方法通常使用奇异值分解（SVD）的变体（应该指的是 PCA）来嵌入数据到低维空间，然后在嵌入中寻找批次之间相似细胞的局部邻域，并以局部自适应（非线性）的方式来校正批次效应。主要例子包括开创性的互近邻对（MNN）、</span><code><span leaf="">Seurat</span></code><span leaf="">、</span><code><span leaf="">Scanorama </span></code><span leaf="">、</span><code><span leaf="">FastMNN</span></code><span leaf="">、</span><code><span leaf="">Harmony</span></code><span leaf="">。</span></p><p data-tool="mdnice编辑器"><strong><span leaf="">基于图的方法（Graph-based methods）：</span></strong><span leaf="">通常是运行最快的方法。这些方法使用最近邻图来表示每个批次的数据。强制在不同批次的细胞间建立连接，然后通过修剪强制连接的边来考虑细胞类型组成的差异，从而校正批次效应。最突出的例子是 </span><code><span leaf="">bbknn</span></code><span leaf="">。</span></p><p data-tool="mdnice编辑器"><strong><span leaf="">深度学习方法（Deep Learning approaches）：</span></strong><span leaf="">最新、最复杂，通常需要最多的数据才能获得良好的性能。大多数深度学习方法基于自编码器网络，要么在条件变分自编码器（CVAE）中根据批次协变量调节降维，要么在嵌入空间中拟合局部线性校正。例子有 </span><code><span leaf="">scVI</span></code><span leaf="">、</span><code><span leaf="">scANVI</span></code><span leaf=""> 和 </span><code><span leaf="">scGen</span></code><span leaf="">。</span></p><p data-tool="mdnice编辑器"><span leaf="">还有一些方法可以使用细胞身份标签，给哪些生物学变异不能作为批次效应被去除提供一些参考。由于批次效应去除通常是一项预处理任务，所以这类方法可能在很多场合下并不适用。</span></p><figure data-tool="mdnice编辑器"><span leaf=""><img alt="不同整合方法总览" data-imgfileid="100053190" data-ratio="0.4685185185185185" data-src="https://mmbiz.qpic.cn/mmbiz_jpg/iaRJcrq2Los8kIEum763iaPGg2MSjt4QTUY5yIrZ7gMZtQ12iaLgRfKGuuIE4FtgjfufL6Cl1PF8s79wEUfOQOR6w/640?wx_fmt=jpeg&amp;from=appmsg" data-type="jpeg" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_jpg/iaRJcrq2Los8kIEum763iaPGg2MSjt4QTUY5yIrZ7gMZtQ12iaLgRfKGuuIE4FtgjfufL6Cl1PF8s79wEUfOQOR6w/640?wx_fmt=jpeg&amp;from=appmsg"></span><figcaption><span leaf="">不同整合方法总览</span></figcaption></figure><h3 data-tool="mdnice编辑器"><span></span><span><span leaf="">7.1.2 批次去除的复杂性</span></span><span></span></h3><p data-tool="mdnice编辑器"><span leaf="">scRNA-seq 中的批次效应去除被分为两个子任务：批次校正（batch correction）和数据整合（data integration）。批次校正处理同一实验中样本间的批次效应，这个效应通常是拟线性的（quasi-linear）。数据整合处理复杂的、通常是嵌套的批次效应，这些数据集可能使用不同的实验方案生成，并且细胞类型标签在不同数据集之间并不一致。虽然这么区分，但实际使用中这些术语经常互换使用。鉴于差异的复杂性，不同的方法在不同的基准测试中被证明对这两个子任务是最优的，这并不奇怪。</span></p><h3 data-tool="mdnice编辑器"><span></span><span><span leaf="">7.1.3 数据整合方法的比较</span></span><span></span></h3><p data-tool="mdnice编辑器"><span leaf="">一些基准测试先前已经评估了批次校正和数据整合方法的性能。在去除批次效应时，方法可能会过度校正去除掉有意义的生物学变异。所以需要同时考虑批次效应去除和生物变异保护来评估整合性能。</span></p><p data-tool="mdnice编辑器"><span leaf="">kBET 是第一个量化批次校正的指标。基准测试表明，全局模型中 </span><code><span leaf="">ComBat</span></code><span leaf=""> 表现最好 [1]；2020 和 2021 年的基准测试 [2] [3] 表明 </span><code><span leaf="">Seurat</span></code><span leaf=""> 和 </span><code><span leaf="">Harmony</span></code><span leaf=""> 在简单的批次校正任务中表现良好。一项对于复杂整合任务的基准研究 [4] 使用 14 个指标测试了 16 种方法，每个任务中表现最好的方法有所不同，但使用细胞类型标签的方法在所有任务中表现最好。深度学习方法 </span><code><span leaf="">scANVI</span></code><span leaf="">（带标签）、</span><code><span leaf="">scVI</span></code><span leaf=""> 和 </span><code><span leaf="">scGen</span></code><span leaf="">（带标签）以及线性嵌入模型 </span><code><span leaf="">Scanorama</span></code><span leaf=""> 表现最佳，尤其是在复杂任务上，而 </span><code><span leaf="">Harmony</span></code><span leaf=""> 在不太复杂的任务上表现良好。</span></p><h3 data-tool="mdnice编辑器"><span></span><span><span leaf="">7.1.4 选择整合方法</span></span><span></span></h3><p data-tool="mdnice编辑器"><span leaf="">虽然数据整合方法已经经过广泛的基准测试，但不存在适用于所有场景的最佳方法。像 </span><code><span leaf="">scIB</span></code><span leaf=""> 和 </span><code><span leaf="">batchbench</span></code><span leaf=""> 这种集成性能度量标准和评估流程的软件包，可以用在你自己的数据上来评估整合的性能。然而许多指标（特别是衡量生物学变异保留的指标）需要细胞类型标签的金标准。参数优化可以调整许多方法以适用于特定任务，但</span><strong><span leaf="">总的来说，</span><code><span leaf="">Harmony</span></code><span leaf=""> 和 </span><code><span leaf="">Seurat</span></code><span leaf=""> 在简单的批次校正任务中一贯表现良好，</span><code><span leaf="">scVI</span></code><span leaf="">、</span><code><span leaf="">scGen</span></code><span leaf="">、</span><code><span leaf="">scANVI</span></code><span leaf=""> 和 </span><code><span leaf="">Scanorama</span></code><span leaf=""> 在更加复杂的数据整合任务中表现良好。</span></strong><span leaf="">关于数据整合方法选择的详细指南可以在 [4] 中找到。</span></p><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">7.2 数据集</span></span><span></span><span></span></h2><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># Python packages</span></span><span leaf=""><br></span><span><span leaf="">import</span></span><span leaf=""> anndata2ri</span><span leaf=""><br></span><span><span leaf="">import</span></span><span leaf=""> bbknn</span><span leaf=""><br></span><span><span leaf="">import</span></span><span leaf=""> matplotlib.pyplot </span><span><span leaf="">as</span></span><span leaf=""> plt</span><span leaf=""><br></span><span><span leaf="">import</span></span><span leaf=""> numpy </span><span><span leaf="">as</span></span><span leaf=""> np</span><span leaf=""><br></span><span><span leaf="">import</span></span><span leaf=""> pandas </span><span><span leaf="">as</span></span><span leaf=""> pd</span><span leaf=""><br></span><span><span leaf="">import</span></span><span leaf=""> scanpy </span><span><span leaf="">as</span></span><span leaf=""> sc</span><span leaf=""><br></span><span><span leaf="">import</span></span><span leaf=""> scib</span><span leaf=""><br></span><span><span leaf="">import</span></span><span leaf=""> scvi</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># R interface</span></span><span leaf=""><br></span><span><span leaf="">from</span></span><span leaf=""> rpy2.robjects </span><span><span leaf="">import</span></span><span leaf=""> pandas2ri</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># pandas2ri.activate()</span></span><span leaf=""><br></span><span><span leaf=""># anndata2ri.activate()</span></span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">%load_ext rpy2.ipython</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">%%R</span><span leaf=""><br></span><span><span leaf=""># R packages</span></span><span leaf=""><br></span><span leaf="">library(Seurat)</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">每次安装环境都挺费劲的，这里推荐安装的时候新建一个虚拟环境，避免 </span><code><span leaf="">scvi</span></code><span leaf=""> 与其他包的冲突与兼容问题，</span><code><span leaf="">scvi</span></code><span leaf=""> 目前支持 python 3.10 - 3.13。</span></p><p data-tool="mdnice编辑器"><span leaf="">数据集使用的是 10x Multiome 技术，在同一细胞中可以同时测量 scRNA-seq 表达值和染色质可及性（scATAC-seq）。从 figshare 中下载数据集：</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">curl -L -o openproblems_bmmc_multiome_genes_filtered.h5ad \</span><span leaf=""><br></span><span leaf="">     -H </span><span><span leaf="">"User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36"</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">     https://figshare.com/ndownloader/files/45452260</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">完整的数据集包含 69,249 个细胞和 129,921 个特征，表达矩阵有 </span><code><span leaf="">counts</span></code><span leaf=""> 原始计数和 </span><code><span leaf="">logcounts</span></code><span leaf=""> 标准化后的对数计数值并且 </span><code><span leaf="">logcounts</span></code><span leaf=""> layer 也存储在 </span><code><span leaf="">.X</span></code><span leaf=""> 中。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">adata_raw = sc.read(</span><span><span leaf="">"openproblems_bmmc_multiome_genes_filtered.h5ad"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">adata_raw.layers[</span><span><span leaf="">"logcounts"</span></span><span leaf="">] = adata_raw.X</span><span leaf=""><br></span><span leaf="">adata_raw</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">AnnData object </span><span><span leaf="">with</span></span><span leaf=""> n_obs × n_vars = </span><span><span leaf="">69249</span></span><span leaf=""> × </span><span><span leaf="">129921</span></span><span leaf=""><br></span><span leaf="">    obs: </span><span><span leaf="">'GEX_pct_counts_mt'</span></span><span leaf="">, </span><span><span leaf="">'GEX_n_counts'</span></span><span leaf="">, </span><span><span leaf="">'GEX_n_genes'</span></span><span leaf="">, </span><span><span leaf="">'GEX_size_factors'</span></span><span leaf="">, </span><span><span leaf="">'GEX_phase'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_nCount_peaks'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_atac_fragments'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_reads_in_peaks_frac'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_blacklist_fraction'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_nucleosome_signal'</span></span><span leaf="">, </span><span><span leaf="">'cell_type'</span></span><span leaf="">, </span><span><span leaf="">'batch'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_pseudotime_order'</span></span><span leaf="">, </span><span><span leaf="">'GEX_pseudotime_order'</span></span><span leaf="">, </span><span><span leaf="">'Samplename'</span></span><span leaf="">, </span><span><span leaf="">'Site'</span></span><span leaf="">, </span><span><span leaf="">'DonorNumber'</span></span><span leaf="">, </span><span><span leaf="">'Modality'</span></span><span leaf="">, </span><span><span leaf="">'VendorLot'</span></span><span leaf="">, </span><span><span leaf="">'DonorID'</span></span><span leaf="">, </span><span><span leaf="">'DonorAge'</span></span><span leaf="">, </span><span><span leaf="">'DonorBMI'</span></span><span leaf="">, </span><span><span leaf="">'DonorBloodType'</span></span><span leaf="">, </span><span><span leaf="">'DonorRace'</span></span><span leaf="">, </span><span><span leaf="">'Ethnicity'</span></span><span leaf="">, </span><span><span leaf="">'DonorGender'</span></span><span leaf="">, </span><span><span leaf="">'QCMeds'</span></span><span leaf="">, </span><span><span leaf="">'DonorSmoker'</span></span><span leaf=""><br></span><span leaf="">    var: </span><span><span leaf="">'feature_types'</span></span><span leaf="">, </span><span><span leaf="">'gene_id'</span></span><span leaf=""><br></span><span leaf="">    uns: </span><span><span leaf="">'ATAC_gene_activity_var_names'</span></span><span leaf="">, </span><span><span leaf="">'dataset_id'</span></span><span leaf="">, </span><span><span leaf="">'genome'</span></span><span leaf="">, </span><span><span leaf="">'organism'</span></span><span leaf=""><br></span><span leaf="">    obsm: </span><span><span leaf="">'ATAC_gene_activity'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_lsi_full'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_lsi_red'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_umap'</span></span><span leaf="">, </span><span><span leaf="">'GEX_X_pca'</span></span><span leaf="">, </span><span><span leaf="">'GEX_X_umap'</span></span><span leaf=""><br></span><span leaf="">    layers: </span><span><span leaf="">'counts'</span></span><span leaf="">, </span><span><span leaf="">'logcounts'</span></span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><code><span leaf="">.obs</span></code><span leaf=""> 中包含了一部分在预处理部分计算的变量用于质控，另一部分包含样本的 metadata。还有 </span><code><span leaf="">cell_type</span></code><span leaf=""> 和 </span><code><span leaf="">batch</span></code><span leaf=""> 分别表示每个细胞的注释标签和测序批次。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">label_key = </span><span><span leaf="">"cell_type"</span></span><span leaf=""><br></span><span leaf="">batch_key = </span><span><span leaf="">"batch"</span></span><span leaf=""><br></span></code></pre><blockquote><span></span><p><strong><span leaf="">使用什么作为批次标签？</span></strong></p><p><span leaf="">决定使用什么作为数据整合的“批次”是整合数据时的核心选择之一。最常见的办法是将每一个样本定义为一个批次，像这个数据集示例一样，这通常会产生最强的批次校正效果。然而，样本间的批次效应通常会与想要保留的生物学差异混淆。比如说从同一个组织的两个位置获取样本，使用样本作为批次会倾向于消除它们之间的差异，同时也会消除位置之间的差异。在这种情况下使用供体作为批次消除个体之间的差异而不是位置之间的差异可能更合适。而当整合多个数据集想要捕获个体间的差异的时候，数据集就可以作为一个有用的批次协变量。</span></p></blockquote><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">adata_raw.obs[batch_key].value_counts()</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">batch</span><span leaf=""><br></span><span leaf="">s4d8     </span><span><span leaf="">9876</span></span><span leaf=""><br></span><span leaf="">s4d1     </span><span><span leaf="">8023</span></span><span leaf=""><br></span><span leaf="">s3d10    </span><span><span leaf="">6781</span></span><span leaf=""><br></span><span leaf="">s1d2     </span><span><span leaf="">6740</span></span><span leaf=""><br></span><span leaf="">s1d1     </span><span><span leaf="">6224</span></span><span leaf=""><br></span><span leaf="">s2d4     </span><span><span leaf="">6111</span></span><span leaf=""><br></span><span leaf="">s2d5     </span><span><span leaf="">4895</span></span><span leaf=""><br></span><span leaf="">s3d3     </span><span><span leaf="">4325</span></span><span leaf=""><br></span><span leaf="">s4d9     </span><span><span leaf="">4325</span></span><span leaf=""><br></span><span leaf="">s1d3     </span><span><span leaf="">4279</span></span><span leaf=""><br></span><span leaf="">s2d1     </span><span><span leaf="">4220</span></span><span leaf=""><br></span><span leaf="">s3d7     </span><span><span leaf="">1771</span></span><span leaf=""><br></span><span leaf="">s3d6     </span><span><span leaf="">1679</span></span><span leaf=""><br></span><span leaf="">Name: count, dtype: int64</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">数据集中总共有 13 个不同的批次，这里的 </span><code><span leaf="">s4d8</span></code><span leaf="">，</span><code><span leaf="">s</span></code><span leaf=""> 表示样本 sample，</span><code><span leaf="">d</span></code><span leaf=""> 表示供体 donor。为了简单起见并减少计算时间，使用 3 个样本进行后续的计算。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">keep_batches = [</span><span><span leaf="">"s1d3"</span></span><span leaf="">, </span><span><span leaf="">"s2d1"</span></span><span leaf="">, </span><span><span leaf="">"s3d7"</span></span><span leaf="">]</span><span leaf=""><br></span><span leaf="">adata = adata_raw[adata_raw.obs[batch_key].isin(keep_batches)].copy()</span><span leaf=""><br></span><span leaf="">adata</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">AnnData object </span><span><span leaf="">with</span></span><span leaf=""> n_obs × n_vars = </span><span><span leaf="">10270</span></span><span leaf=""> × </span><span><span leaf="">129921</span></span><span leaf=""><br></span><span leaf="">    obs: </span><span><span leaf="">'GEX_pct_counts_mt'</span></span><span leaf="">, </span><span><span leaf="">'GEX_n_counts'</span></span><span leaf="">, </span><span><span leaf="">'GEX_n_genes'</span></span><span leaf="">, </span><span><span leaf="">'GEX_size_factors'</span></span><span leaf="">, </span><span><span leaf="">'GEX_phase'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_nCount_peaks'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_atac_fragments'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_reads_in_peaks_frac'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_blacklist_fraction'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_nucleosome_signal'</span></span><span leaf="">, </span><span><span leaf="">'cell_type'</span></span><span leaf="">, </span><span><span leaf="">'batch'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_pseudotime_order'</span></span><span leaf="">, </span><span><span leaf="">'GEX_pseudotime_order'</span></span><span leaf="">, </span><span><span leaf="">'Samplename'</span></span><span leaf="">, </span><span><span leaf="">'Site'</span></span><span leaf="">, </span><span><span leaf="">'DonorNumber'</span></span><span leaf="">, </span><span><span leaf="">'Modality'</span></span><span leaf="">, </span><span><span leaf="">'VendorLot'</span></span><span leaf="">, </span><span><span leaf="">'DonorID'</span></span><span leaf="">, </span><span><span leaf="">'DonorAge'</span></span><span leaf="">, </span><span><span leaf="">'DonorBMI'</span></span><span leaf="">, </span><span><span leaf="">'DonorBloodType'</span></span><span leaf="">, </span><span><span leaf="">'DonorRace'</span></span><span leaf="">, </span><span><span leaf="">'Ethnicity'</span></span><span leaf="">, </span><span><span leaf="">'DonorGender'</span></span><span leaf="">, </span><span><span leaf="">'QCMeds'</span></span><span leaf="">, </span><span><span leaf="">'DonorSmoker'</span></span><span leaf=""><br></span><span leaf="">    var: </span><span><span leaf="">'feature_types'</span></span><span leaf="">, </span><span><span leaf="">'gene_id'</span></span><span leaf=""><br></span><span leaf="">    uns: </span><span><span leaf="">'ATAC_gene_activity_var_names'</span></span><span leaf="">, </span><span><span leaf="">'dataset_id'</span></span><span leaf="">, </span><span><span leaf="">'genome'</span></span><span leaf="">, </span><span><span leaf="">'organism'</span></span><span leaf=""><br></span><span leaf="">    obsm: </span><span><span leaf="">'ATAC_gene_activity'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_lsi_full'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_lsi_red'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_umap'</span></span><span leaf="">, </span><span><span leaf="">'GEX_X_pca'</span></span><span leaf="">, </span><span><span leaf="">'GEX_X_umap'</span></span><span leaf=""><br></span><span leaf="">    layers: </span><span><span leaf="">'counts'</span></span><span leaf="">, </span><span><span leaf="">'logcounts'</span></span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">剩下 10,270 个细胞。</span><code><span leaf="">.var</span></code><span leaf=""> 存储了两列特征的注释，</span><code><span leaf="">feature_types</span></code><span leaf=""> 表示特征的类型（RNA 或 ATAC），</span><code><span leaf="">gene_id</span></code><span leaf=""> 为与每个特征关联的基因。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">adata.var[</span><span><span leaf="">"feature_types"</span></span><span leaf="">].value_counts()</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">feature_types</span><span leaf=""><br></span><span leaf="">ATAC    </span><span><span leaf="">116490</span></span><span leaf=""><br></span><span leaf="">GEX      </span><span><span leaf="">13431</span></span><span leaf=""><br></span><span leaf="">Name: count, dtype: int64</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">这里面有超过 10 万个 ATAC 特征以及大约只有 13,000 个基因表达特征（GEX）。这是一个多模态数据，在数据整合这一节暂时用不到 ATAC 数据，所以先把它去掉，只保留基因表达数据。同时还进行一个简单的过滤，确保特征中没有全零值。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">adata = adata[:, adata.var[</span><span><span leaf="">"feature_types"</span></span><span leaf="">] == </span><span><span leaf="">"GEX"</span></span><span leaf="">].copy()</span><span leaf=""><br></span><span leaf="">sc.pp.filter_genes(adata, min_cells=</span><span><span leaf="">1</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">adata</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">AnnData object </span><span><span leaf="">with</span></span><span leaf=""> n_obs × n_vars = </span><span><span leaf="">10270</span></span><span leaf=""> × </span><span><span leaf="">13431</span></span><span leaf=""><br></span><span leaf="">    obs: </span><span><span leaf="">'GEX_pct_counts_mt'</span></span><span leaf="">, </span><span><span leaf="">'GEX_n_counts'</span></span><span leaf="">, </span><span><span leaf="">'GEX_n_genes'</span></span><span leaf="">, </span><span><span leaf="">'GEX_size_factors'</span></span><span leaf="">, </span><span><span leaf="">'GEX_phase'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_nCount_peaks'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_atac_fragments'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_reads_in_peaks_frac'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_blacklist_fraction'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_nucleosome_signal'</span></span><span leaf="">, </span><span><span leaf="">'cell_type'</span></span><span leaf="">, </span><span><span leaf="">'batch'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_pseudotime_order'</span></span><span leaf="">, </span><span><span leaf="">'GEX_pseudotime_order'</span></span><span leaf="">, </span><span><span leaf="">'Samplename'</span></span><span leaf="">, </span><span><span leaf="">'Site'</span></span><span leaf="">, </span><span><span leaf="">'DonorNumber'</span></span><span leaf="">, </span><span><span leaf="">'Modality'</span></span><span leaf="">, </span><span><span leaf="">'VendorLot'</span></span><span leaf="">, </span><span><span leaf="">'DonorID'</span></span><span leaf="">, </span><span><span leaf="">'DonorAge'</span></span><span leaf="">, </span><span><span leaf="">'DonorBMI'</span></span><span leaf="">, </span><span><span leaf="">'DonorBloodType'</span></span><span leaf="">, </span><span><span leaf="">'DonorRace'</span></span><span leaf="">, </span><span><span leaf="">'Ethnicity'</span></span><span leaf="">, </span><span><span leaf="">'DonorGender'</span></span><span leaf="">, </span><span><span leaf="">'QCMeds'</span></span><span leaf="">, </span><span><span leaf="">'DonorSmoker'</span></span><span leaf=""><br></span><span leaf="">    var: </span><span><span leaf="">'feature_types'</span></span><span leaf="">, </span><span><span leaf="">'gene_id'</span></span><span leaf="">, </span><span><span leaf="">'n_cells'</span></span><span leaf=""><br></span><span leaf="">    uns: </span><span><span leaf="">'ATAC_gene_activity_var_names'</span></span><span leaf="">, </span><span><span leaf="">'dataset_id'</span></span><span leaf="">, </span><span><span leaf="">'genome'</span></span><span leaf="">, </span><span><span leaf="">'organism'</span></span><span leaf=""><br></span><span leaf="">    obsm: </span><span><span leaf="">'ATAC_gene_activity'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_lsi_full'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_lsi_red'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_umap'</span></span><span leaf="">, </span><span><span leaf="">'GEX_X_pca'</span></span><span leaf="">, </span><span><span leaf="">'GEX_X_umap'</span></span><span leaf=""><br></span><span leaf="">    layers: </span><span><span leaf="">'counts'</span></span><span leaf="">, </span><span><span leaf="">'logcounts'</span></span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">因为对原先的数据集取了子集，所以需要重新标准化数据，然后用这个数据进行后续的数据整合。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">adata.X = adata.layers[</span><span><span leaf="">"counts"</span></span><span leaf="">].copy()</span><span leaf=""><br></span><span leaf="">sc.pp.normalize_total(adata)</span><span leaf=""><br></span><span leaf="">sc.pp.log1p(adata)</span><span leaf=""><br></span><span leaf="">adata.layers[</span><span><span leaf="">"logcounts"</span></span><span leaf="">] = adata.X.copy()</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">大多数的整合方法需要一个包含了所有样本和批次变量的对象，如果是每个样本有一个对象，可以使用 anndata 的 </span><code><span leaf="">concat()</span></code><span leaf=""> 函数进行连接，详细信息可见  https://anndata.readthedocs.io/en/stable/concatenation.html。</span></p><blockquote><span></span><p><strong><span leaf="">整合 UMI 和全长数据：</span></strong></p><p><span leaf="">全长转录本测序会受到基因长度偏差的影响，在 UMI 相同的情况下，更长的基因将有更多的 reads。所以建议在尝试进行数据整合之前先将全长样本的 counts 数据转换成对基因长度进行校正后的数据（比如 TPM）。如果要整合的所有样本都是全长数据，就没有必要这样做。</span></p></blockquote><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">7.3 未整合的数据</span></span><span></span><span></span></h2><p data-tool="mdnice编辑器"><span leaf="">建议在实施任何整合之前先看一下原始数据的结果，这可以帮人大致了解批次效应有多大以及可能是由什么原因造成的（从而决定考虑哪些变量作为批次标签）。如果样本之间有重叠的话甚至可以不进行整合，这在同一个实验室中的小鼠和细胞系研究中并不少见，因为可以控制大部分导致批次效应的变量。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">sc.pp.highly_variable_genes(adata)</span><span leaf=""><br></span><span leaf="">sc.tl.pca(adata)</span><span leaf=""><br></span><span leaf="">sc.pp.neighbors(adata)</span><span leaf=""><br></span><span leaf="">sc.tl.umap(adata)</span><span leaf=""><br></span><span leaf="">adata</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">AnnData object </span><span><span leaf="">with</span></span><span leaf=""> n_obs × n_vars = </span><span><span leaf="">10270</span></span><span leaf=""> × </span><span><span leaf="">13431</span></span><span leaf=""><br></span><span leaf="">    obs: </span><span><span leaf="">'GEX_pct_counts_mt'</span></span><span leaf="">, </span><span><span leaf="">'GEX_n_counts'</span></span><span leaf="">, </span><span><span leaf="">'GEX_n_genes'</span></span><span leaf="">, </span><span><span leaf="">'GEX_size_factors'</span></span><span leaf="">, </span><span><span leaf="">'GEX_phase'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_nCount_peaks'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_atac_fragments'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_reads_in_peaks_frac'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_blacklist_fraction'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_nucleosome_signal'</span></span><span leaf="">, </span><span><span leaf="">'cell_type'</span></span><span leaf="">, </span><span><span leaf="">'batch'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_pseudotime_order'</span></span><span leaf="">, </span><span><span leaf="">'GEX_pseudotime_order'</span></span><span leaf="">, </span><span><span leaf="">'Samplename'</span></span><span leaf="">, </span><span><span leaf="">'Site'</span></span><span leaf="">, </span><span><span leaf="">'DonorNumber'</span></span><span leaf="">, </span><span><span leaf="">'Modality'</span></span><span leaf="">, </span><span><span leaf="">'VendorLot'</span></span><span leaf="">, </span><span><span leaf="">'DonorID'</span></span><span leaf="">, </span><span><span leaf="">'DonorAge'</span></span><span leaf="">, </span><span><span leaf="">'DonorBMI'</span></span><span leaf="">, </span><span><span leaf="">'DonorBloodType'</span></span><span leaf="">, </span><span><span leaf="">'DonorRace'</span></span><span leaf="">, </span><span><span leaf="">'Ethnicity'</span></span><span leaf="">, </span><span><span leaf="">'DonorGender'</span></span><span leaf="">, </span><span><span leaf="">'QCMeds'</span></span><span leaf="">, </span><span><span leaf="">'DonorSmoker'</span></span><span leaf=""><br></span><span leaf="">    var: </span><span><span leaf="">'feature_types'</span></span><span leaf="">, </span><span><span leaf="">'gene_id'</span></span><span leaf="">, </span><span><span leaf="">'n_cells'</span></span><span leaf="">, </span><span><span leaf="">'highly_variable'</span></span><span leaf="">, </span><span><span leaf="">'means'</span></span><span leaf="">, </span><span><span leaf="">'dispersions'</span></span><span leaf="">, </span><span><span leaf="">'dispersions_norm'</span></span><span leaf=""><br></span><span leaf="">    uns: </span><span><span leaf="">'ATAC_gene_activity_var_names'</span></span><span leaf="">, </span><span><span leaf="">'dataset_id'</span></span><span leaf="">, </span><span><span leaf="">'genome'</span></span><span leaf="">, </span><span><span leaf="">'organism'</span></span><span leaf="">, </span><span><span leaf="">'log1p'</span></span><span leaf="">, </span><span><span leaf="">'hvg'</span></span><span leaf="">, </span><span><span leaf="">'pca'</span></span><span leaf="">, </span><span><span leaf="">'neighbors'</span></span><span leaf="">, </span><span><span leaf="">'umap'</span></span><span leaf=""><br></span><span leaf="">    obsm: </span><span><span leaf="">'ATAC_gene_activity'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_lsi_full'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_lsi_red'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_umap'</span></span><span leaf="">, </span><span><span leaf="">'GEX_X_pca'</span></span><span leaf="">, </span><span><span leaf="">'GEX_X_umap'</span></span><span leaf="">, </span><span><span leaf="">'X_pca'</span></span><span leaf="">, </span><span><span leaf="">'X_umap'</span></span><span leaf=""><br></span><span leaf="">    varm: </span><span><span leaf="">'PCs'</span></span><span leaf=""><br></span><span leaf="">    layers: </span><span><span leaf="">'counts'</span></span><span leaf="">, </span><span><span leaf="">'logcounts'</span></span><span leaf=""><br></span><span leaf="">    obsp: </span><span><span leaf="">'distances'</span></span><span leaf="">, </span><span><span leaf="">'connectivities'</span></span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">绘制 UMAP 图，点的着色分别是细胞类型（数据集自带）以及批次标签：</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">adata.uns[batch_key + </span><span><span leaf="">"_colors"</span></span><span leaf="">] = [</span><span leaf=""><br></span><span leaf="">    </span><span><span leaf="">"<a topic-id="mfgg65bf-mkjutm" data-topic="1">#1b9e77</a>"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">    </span><span><span leaf="">"<a topic-id="mfgg65bg-e2a8jy" data-topic="1">#d95f02</a>"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">    </span><span><span leaf="">"<a topic-id="mfgg65bg-hp0m39" data-topic="1">#7570b3</a>"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">]  </span><span><span leaf=""># Set custom colours for batches</span></span><span leaf=""><br></span><span leaf="">sc.pl.umap(adata, color=[label_key, batch_key], wspace=</span><span><span leaf="">1</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-imgfileid="100053186" data-ratio="0.20555555555555555" data-src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2Los8kIEum763iaPGg2MSjt4QTUb8BkZxwhJVjpWlBAOmlDDae6WicpibK4Ox7MhrTddPe76lnxoaatFQiaw/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2Los8kIEum763iaPGg2MSjt4QTUb8BkZxwhJVjpWlBAOmlDDae6WicpibK4Ox7MhrTddPe76lnxoaatFQiaw/640?wx_fmt=png&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">这种图可以看到批次之间的明显分离。这个例子里，虽然相同标签的细胞通常彼此靠近，但批次之间存在偏移。如果我们使用原始数据进行聚类分析，最终可能会得到一些包含单个批次的簇，这在注释阶段难以解释。也可能忽略那些在任何单个样本中都不足以产生自己簇的稀有细胞类型。</span></p><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">7.4 已知批次的特征选择</span></span><span></span><span></span></h2><p data-tool="mdnice编辑器"><span leaf="">在有多个样本的情况下，执行特征选择的时候就需要设置好批次标签，因为对整个数据集进行基因选择，很可能捕获到的是批次效应而不是感兴趣的生物学信号。它对选择稀有细胞类型相关的基因也有一定帮助。</span></p><p data-tool="mdnice编辑器"><span leaf="">设置 scanpy 中的 </span><code><span leaf="">highly_variable_genes()</span></code><span leaf=""> 函数中的 </span><code><span leaf="">batch_key</span></code><span leaf=""> 参数，scanpy 可以单独计算每个批次中的 HVGs（高变基因）然后选择那些在更多批次中出现高变基因。使用这个函数是因为它已经有内置的参数了，其他方法可能需要单独在每个批次中进行计算然后手动整合结果。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">sc.pp.highly_variable_genes(</span><span leaf=""><br></span><span leaf="">    adata, n_top_genes=</span><span><span leaf="">2000</span></span><span leaf="">, flavor=</span><span><span leaf="">"cell_ranger"</span></span><span leaf="">, batch_key=batch_key</span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><code><span leaf="">.var</span></code><span leaf=""> 中新加了几列，包括 </span><code><span leaf="">highly_variable_nbatches</span></code><span leaf=""> 表示基因在几个批次中被视为 HVG，</span><code><span leaf="">highly_variable_intersection</span></code><span leaf=""> 表示基因是否在每个批次中都是 HVG，</span><code><span leaf="">highly_variable</span></code><span leaf=""> 表示结合每个批次的结果后最终确定是否为 HVG。可以看一下每个批次中 HVG 的数目：</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">n_batches = adata.var[</span><span><span leaf="">"highly_variable_nbatches"</span></span><span leaf="">].value_counts()</span><span leaf=""><br></span><span leaf="">ax = n_batches.plot(kind=</span><span><span leaf="">"bar"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">n_batches</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">highly_variable_nbatches</span><span leaf=""><br></span><span><span leaf="">0</span></span><span leaf="">    </span><span><span leaf="">9931</span></span><span leaf=""><br></span><span><span leaf="">1</span></span><span leaf="">    </span><span><span leaf="">1824</span></span><span leaf=""><br></span><span><span leaf="">2</span></span><span leaf="">     </span><span><span leaf="">852</span></span><span leaf=""><br></span><span><span leaf="">3</span></span><span leaf="">     </span><span><span leaf="">824</span></span><span leaf=""><br></span><span leaf="">Name: count, dtype: int64</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-imgfileid="100053187" data-ratio="0.7471074380165289" data-src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2Los8kIEum763iaPGg2MSjt4QTUMMDtLoDtRhmYo5djmWWiaY0uHvicic9KIVJdxEEITR29ZTW1adpyQ27Fw/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="605" src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2Los8kIEum763iaPGg2MSjt4QTUMMDtLoDtRhmYo5djmWWiaY0uHvicic9KIVJdxEEITR29ZTW1adpyQ27Fw/640?wx_fmt=png&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">大多数的基因都不是  HVG，这是普遍情况，HVG 的具体数量取决于要整合的样本之间的差异程度。top2000 的 HVG 在这里包含了所有 在两到三个批次中同时存在的高变基因和大部分只在一个批次中存在的高变基因。</span></p><blockquote><span></span><p><strong><span leaf="">选择多少基因作为 HVG？</span></strong></p><p><span leaf="">这个问题没有一个明确的答案。</span><code><span leaf="">scvi-tools</span></code><span leaf=""> 的作者建议 1000 - 10000 个基因，但具体多少取决于数据集的复杂性、批次的数目等。也有研究表明人们通常在标准分析中使用 1000 - 6000 个 HVG。虽然选择较少的基因有助于消除批次效应（因为变异性大的基因通常只描述显性生物学变异），但还是建议选择稍微多一点的基因。不过更多的基因也会增加运行数据整合方法的时间。</span></p></blockquote><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">adata_hvg = adata[:, adata.var[</span><span><span leaf="">"highly_variable"</span></span><span leaf="">]].copy()</span><span leaf=""><br></span><span leaf="">adata_hvg</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">AnnData object </span><span><span leaf="">with</span></span><span leaf=""> n_obs × n_vars = </span><span><span leaf="">10270</span></span><span leaf=""> × </span><span><span leaf="">2000</span></span><span leaf=""><br></span><span leaf="">    obs: </span><span><span leaf="">'GEX_pct_counts_mt'</span></span><span leaf="">, </span><span><span leaf="">'GEX_n_counts'</span></span><span leaf="">, </span><span><span leaf="">'GEX_n_genes'</span></span><span leaf="">, </span><span><span leaf="">'GEX_size_factors'</span></span><span leaf="">, </span><span><span leaf="">'GEX_phase'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_nCount_peaks'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_atac_fragments'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_reads_in_peaks_frac'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_blacklist_fraction'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_nucleosome_signal'</span></span><span leaf="">, </span><span><span leaf="">'cell_type'</span></span><span leaf="">, </span><span><span leaf="">'batch'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_pseudotime_order'</span></span><span leaf="">, </span><span><span leaf="">'GEX_pseudotime_order'</span></span><span leaf="">, </span><span><span leaf="">'Samplename'</span></span><span leaf="">, </span><span><span leaf="">'Site'</span></span><span leaf="">, </span><span><span leaf="">'DonorNumber'</span></span><span leaf="">, </span><span><span leaf="">'Modality'</span></span><span leaf="">, </span><span><span leaf="">'VendorLot'</span></span><span leaf="">, </span><span><span leaf="">'DonorID'</span></span><span leaf="">, </span><span><span leaf="">'DonorAge'</span></span><span leaf="">, </span><span><span leaf="">'DonorBMI'</span></span><span leaf="">, </span><span><span leaf="">'DonorBloodType'</span></span><span leaf="">, </span><span><span leaf="">'DonorRace'</span></span><span leaf="">, </span><span><span leaf="">'Ethnicity'</span></span><span leaf="">, </span><span><span leaf="">'DonorGender'</span></span><span leaf="">, </span><span><span leaf="">'QCMeds'</span></span><span leaf="">, </span><span><span leaf="">'DonorSmoker'</span></span><span leaf=""><br></span><span leaf="">    var: </span><span><span leaf="">'feature_types'</span></span><span leaf="">, </span><span><span leaf="">'gene_id'</span></span><span leaf="">, </span><span><span leaf="">'n_cells'</span></span><span leaf="">, </span><span><span leaf="">'highly_variable'</span></span><span leaf="">, </span><span><span leaf="">'means'</span></span><span leaf="">, </span><span><span leaf="">'dispersions'</span></span><span leaf="">, </span><span><span leaf="">'dispersions_norm'</span></span><span leaf="">, </span><span><span leaf="">'highly_variable_nbatches'</span></span><span leaf="">, </span><span><span leaf="">'highly_variable_intersection'</span></span><span leaf=""><br></span><span leaf="">    uns: </span><span><span leaf="">'ATAC_gene_activity_var_names'</span></span><span leaf="">, </span><span><span leaf="">'dataset_id'</span></span><span leaf="">, </span><span><span leaf="">'genome'</span></span><span leaf="">, </span><span><span leaf="">'organism'</span></span><span leaf="">, </span><span><span leaf="">'log1p'</span></span><span leaf="">, </span><span><span leaf="">'hvg'</span></span><span leaf="">, </span><span><span leaf="">'pca'</span></span><span leaf="">, </span><span><span leaf="">'neighbors'</span></span><span leaf="">, </span><span><span leaf="">'umap'</span></span><span leaf="">, </span><span><span leaf="">'batch_colors'</span></span><span leaf="">, </span><span><span leaf="">'cell_type_colors'</span></span><span leaf=""><br></span><span leaf="">    obsm: </span><span><span leaf="">'ATAC_gene_activity'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_lsi_full'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_lsi_red'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_umap'</span></span><span leaf="">, </span><span><span leaf="">'GEX_X_pca'</span></span><span leaf="">, </span><span><span leaf="">'GEX_X_umap'</span></span><span leaf="">, </span><span><span leaf="">'X_pca'</span></span><span leaf="">, </span><span><span leaf="">'X_umap'</span></span><span leaf=""><br></span><span leaf="">    varm: </span><span><span leaf="">'PCs'</span></span><span leaf=""><br></span><span leaf="">    layers: </span><span><span leaf="">'counts'</span></span><span leaf="">, </span><span><span leaf="">'logcounts'</span></span><span leaf=""><br></span><span leaf="">    obsp: </span><span><span leaf="">'distances'</span></span><span leaf="">, </span><span><span leaf="">'connectivities'</span></span><span leaf=""><br></span></code></pre><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">7.5 基于变分自编码器（VAE）的整合</span></span><span></span><span></span></h2><p data-tool="mdnice编辑器"><code><span leaf="">scVI</span></code><span leaf="">（single-cell Variational Inference）是一种基于条件变分自编码器的方法，可以在 </span><code><span leaf="">scvi-tools</span></code><span leaf=""> 包中获取。变分自编码器是一种试图降低数据集维度的人工神经网络，“条件”部分指的是将这种降维过程基于特定协变量（批次）进行调节，使得这个协变量不会影响低维表征。在基准研究中，</span><code><span leaf="">scVI</span></code><span leaf=""> 被证明在多个数据集上表现良好。</span><code><span leaf="">scVI</span></code><span leaf=""> 直接对原始计数数据建模，所以需要注意提供计数矩阵而不是标准化后的表达矩阵。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">adata_scvi = adata_hvg.copy()</span><span leaf=""><br></span></code></pre><h3 data-tool="mdnice编辑器"><span></span><span><span leaf="">7.5.1 数据准备</span></span><span></span></h3><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">scvi.model.SCVI.setup_anndata(adata_scvi, layer=</span><span><span leaf="">"counts"</span></span><span leaf="">, batch_key=batch_key)</span><span leaf=""><br></span><span leaf="">adata_scvi</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">AnnData object </span><span><span leaf="">with</span></span><span leaf=""> n_obs × n_vars = </span><span><span leaf="">10270</span></span><span leaf=""> × </span><span><span leaf="">2000</span></span><span leaf=""><br></span><span leaf="">    obs: </span><span><span leaf="">'GEX_pct_counts_mt'</span></span><span leaf="">, </span><span><span leaf="">'GEX_n_counts'</span></span><span leaf="">, </span><span><span leaf="">'GEX_n_genes'</span></span><span leaf="">, </span><span><span leaf="">'GEX_size_factors'</span></span><span leaf="">, </span><span><span leaf="">'GEX_phase'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_nCount_peaks'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_atac_fragments'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_reads_in_peaks_frac'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_blacklist_fraction'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_nucleosome_signal'</span></span><span leaf="">, </span><span><span leaf="">'cell_type'</span></span><span leaf="">, </span><span><span leaf="">'batch'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_pseudotime_order'</span></span><span leaf="">, </span><span><span leaf="">'GEX_pseudotime_order'</span></span><span leaf="">, </span><span><span leaf="">'Samplename'</span></span><span leaf="">, </span><span><span leaf="">'Site'</span></span><span leaf="">, </span><span><span leaf="">'DonorNumber'</span></span><span leaf="">, </span><span><span leaf="">'Modality'</span></span><span leaf="">, </span><span><span leaf="">'VendorLot'</span></span><span leaf="">, </span><span><span leaf="">'DonorID'</span></span><span leaf="">, </span><span><span leaf="">'DonorAge'</span></span><span leaf="">, </span><span><span leaf="">'DonorBMI'</span></span><span leaf="">, </span><span><span leaf="">'DonorBloodType'</span></span><span leaf="">, </span><span><span leaf="">'DonorRace'</span></span><span leaf="">, </span><span><span leaf="">'Ethnicity'</span></span><span leaf="">, </span><span><span leaf="">'DonorGender'</span></span><span leaf="">, </span><span><span leaf="">'QCMeds'</span></span><span leaf="">, </span><span><span leaf="">'DonorSmoker'</span></span><span leaf="">, </span><span><span leaf="">'_scvi_batch'</span></span><span leaf="">, </span><span><span leaf="">'_scvi_labels'</span></span><span leaf=""><br></span><span leaf="">    var: </span><span><span leaf="">'feature_types'</span></span><span leaf="">, </span><span><span leaf="">'gene_id'</span></span><span leaf="">, </span><span><span leaf="">'n_cells'</span></span><span leaf="">, </span><span><span leaf="">'highly_variable'</span></span><span leaf="">, </span><span><span leaf="">'means'</span></span><span leaf="">, </span><span><span leaf="">'dispersions'</span></span><span leaf="">, </span><span><span leaf="">'dispersions_norm'</span></span><span leaf="">, </span><span><span leaf="">'highly_variable_nbatches'</span></span><span leaf="">, </span><span><span leaf="">'highly_variable_intersection'</span></span><span leaf=""><br></span><span leaf="">    uns: </span><span><span leaf="">'ATAC_gene_activity_var_names'</span></span><span leaf="">, </span><span><span leaf="">'batch_colors'</span></span><span leaf="">, </span><span><span leaf="">'cell_type_colors'</span></span><span leaf="">, </span><span><span leaf="">'dataset_id'</span></span><span leaf="">, </span><span><span leaf="">'genome'</span></span><span leaf="">, </span><span><span leaf="">'hvg'</span></span><span leaf="">, </span><span><span leaf="">'log1p'</span></span><span leaf="">, </span><span><span leaf="">'neighbors'</span></span><span leaf="">, </span><span><span leaf="">'organism'</span></span><span leaf="">, </span><span><span leaf="">'pca'</span></span><span leaf="">, </span><span><span leaf="">'umap'</span></span><span leaf="">, </span><span><span leaf="">'_scvi_uuid'</span></span><span leaf="">, </span><span><span leaf="">'_scvi_manager_uuid'</span></span><span leaf=""><br></span><span leaf="">    obsm: </span><span><span leaf="">'ATAC_gene_activity'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_lsi_full'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_lsi_red'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_umap'</span></span><span leaf="">, </span><span><span leaf="">'GEX_X_pca'</span></span><span leaf="">, </span><span><span leaf="">'GEX_X_umap'</span></span><span leaf="">, </span><span><span leaf="">'X_pca'</span></span><span leaf="">, </span><span><span leaf="">'X_umap'</span></span><span leaf=""><br></span><span leaf="">    varm: </span><span><span leaf="">'PCs'</span></span><span leaf=""><br></span><span leaf="">    layers: </span><span><span leaf="">'counts'</span></span><span leaf="">, </span><span><span leaf="">'logcounts'</span></span><span leaf=""><br></span><span leaf="">    obsp: </span><span><span leaf="">'connectivities'</span></span><span leaf="">, </span><span><span leaf="">'distances'</span></span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><code><span leaf="">scVI</span></code><span leaf=""> 创建的内容会以 </span><code><span leaf="">_scvi</span></code><span leaf=""> 作为前缀，这个没办法手动修改。这个包的作者建议在模型训练完毕之前不要对 </span><code><span leaf="">adata</span></code><span leaf=""> 进行任何更改。在其他数据集上，有可能会看到关于输入表达矩阵的警告，通常需要检查一下 </span><code><span leaf="">.X</span></code><span leaf=""> 的数据是否是 count 值，但如果使用的全长数据进行了基因长度校正，或者使用其他不产生整数 count 的定量方法，也有可能会出现这种情况。</span></p><h3 data-tool="mdnice编辑器"><span></span><span><span leaf="">7.5.2 构建模型</span></span><span></span></h3><p data-tool="mdnice编辑器"><span leaf="">现在可以构造一个 </span><code><span leaf="">scVI</span></code><span leaf=""> 模型对象。除了 </span><code><span leaf="">scVI</span></code><span leaf=""> 以外，</span><code><span leaf="">scvi-tools</span></code><span leaf=""> 还包含了很多其他模型。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">model_scvi = scvi.model.SCVI(adata_scvi)</span><span leaf=""><br></span><span leaf="">model_scvi</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-imgfileid="100053188" data-ratio="0.10833333333333334" data-src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2Los8kIEum763iaPGg2MSjt4QTUOS10E7TMMYiapMDGLtHZ84D7TcBGLJpoHw3C1CxA1s0O2FNLBGiaytrg/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2Los8kIEum763iaPGg2MSjt4QTUOS10E7TMMYiapMDGLtHZ84D7TcBGLJpoHw3C1CxA1s0O2FNLBGiaytrg/640?wx_fmt=png&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><code><span leaf="">scVI</span></code><span leaf=""> 模型对象包含所提供的 anndata 对象以及模型本身的神经网络。现在的模型是没有经过训练的。如果想要修改网络结构，可以向模型构建函数提供额外的参数，这里只使用了默认设置。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">model_scvi.view_anndata_setup()</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-imgfileid="100053189" data-ratio="0.7476635514018691" data-src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2Los8kIEum763iaPGg2MSjt4QTUGDicXo58DIDLaRbppk7r9B29beSp8qCbNhjzTYXg9wePa97SYW0YjEQ/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="749" src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2Los8kIEum763iaPGg2MSjt4QTUGDicXo58DIDLaRbppk7r9B29beSp8qCbNhjzTYXg9wePa97SYW0YjEQ/640?wx_fmt=png&amp;from=appmsg"></span></figure><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2Los8kIEum763iaPGg2MSjt4QTUshr9Og3AzqPRKSbhoP0vPCH0tcG5VyuJN2EfcEPIRgYOdMiaugwp7Zg/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.9177304964539007" data-type="png" data-w="705" data-imgfileid="100053195" src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2Los8kIEum763iaPGg2MSjt4QTUshr9Og3AzqPRKSbhoP0vPCH0tcG5VyuJN2EfcEPIRgYOdMiaugwp7Zg/640?wx_fmt=png&amp;from=appmsg"></span></figure><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2Los8kIEum763iaPGg2MSjt4QTUHUQlHmFBgMBa5enibkV5GkrQWuL19OaNvF8yP6klEjOia3DIaj5DHAkw/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.5" data-type="png" data-w="1006" data-imgfileid="100053194" src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2Los8kIEum763iaPGg2MSjt4QTUHUQlHmFBgMBa5enibkV5GkrQWuL19OaNvF8yP6klEjOia3DIaj5DHAkw/640?wx_fmt=png&amp;from=appmsg"></span></figure><h3 data-tool="mdnice编辑器"><span></span><span><span leaf="">7.5.3 训练模型</span></span><span></span></h3><p data-tool="mdnice编辑器"><span leaf="">模型将针对设定的训练轮次（epoch）进行训练，每个训练轮次中所有细胞数据都会通过神经网络传递一次。默认情况下，</span><code><span leaf="">scVI</span></code><span leaf=""> 采用启发式算法来设定训练轮次：对于细胞数量少于 2 万个的数据集，使用 400 个训练轮次；当细胞数量超过 2 万时，训练轮次数会随细胞数量增加而逐步减少。依据在于当网络在每个训练轮次中处理更多细胞数据时，学习到的信息量等同于在细胞数量较少情况下经过更多轮次训练所获得的信息量。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">model_scvi.train()</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">GPU available: </span><span><span leaf="">False</span></span><span leaf="">, used: </span><span><span leaf="">False</span></span><span leaf=""><br></span><span leaf="">TPU available: </span><span><span leaf="">False</span></span><span leaf="">, using: </span><span><span leaf="">0</span></span><span leaf=""> TPU cores</span><span leaf=""><br></span><span leaf="">HPU available: </span><span><span leaf="">False</span></span><span leaf="">, using: </span><span><span leaf="">0</span></span><span leaf=""> HPUs</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2Los8kIEum763iaPGg2MSjt4QTUkfSbVBbZEedJA3hLVYk6eMlVjtOl6Wymujsa4XcCvPEiaGfzQvVqcGA/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.03611111111111111" data-type="png" data-w="1080" data-imgfileid="100053191" src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2Los8kIEum763iaPGg2MSjt4QTUkfSbVBbZEedJA3hLVYk6eMlVjtOl6Wymujsa4XcCvPEiaGfzQvVqcGA/640?wx_fmt=png&amp;from=appmsg"></span></figure><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">`Trainer.fit` stopped: `max_epochs=</span><span><span leaf="">400</span></span><span leaf="">` reached.</span><span leaf=""><br></span></code></pre><blockquote><span></span><p><strong><span leaf="">提前停止：</span></strong></p><p><span leaf="">除了设定目标训练轮次外，还可在训练函数中设置 </span><code><span leaf="">early_stopping=True</span></code><span leaf="">。可以允许 </span><code><span leaf="">scVI</span></code><span leaf=""> 根据模型收敛情况自主决定提前终止训练。具体的停止条件可通过其他参数进行控制。</span></p></blockquote><h3 data-tool="mdnice编辑器"><span></span><span><span leaf="">7.5.4 提取嵌入</span></span><span></span></h3><p data-tool="mdnice编辑器"><span leaf="">从训练好的模型中提取每个细胞的潜在表征，也就是一个经过批次校正的多维嵌入空间，将它存储在 </span><code><span leaf="">.obsm</span></code><span leaf=""> 中，对应的键为 </span><code><span leaf="">X_scvi</span></code><span leaf="">。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">adata_scvi.obsm[</span><span><span leaf="">"X_scVI"</span></span><span leaf="">] = model_scvi.get_latent_representation()</span><span leaf=""><br></span></code></pre><h3 data-tool="mdnice编辑器"><span></span><span><span leaf="">7.5.5 计算批次校正后的 UMAP</span></span><span></span></h3><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">sc.pp.neighbors(adata_scvi, use_rep=</span><span><span leaf="">"X_scVI"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">sc.tl.umap(adata_scvi)</span><span leaf=""><br></span><span leaf="">adata_scvi</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">AnnData object </span><span><span leaf="">with</span></span><span leaf=""> n_obs × n_vars = </span><span><span leaf="">10270</span></span><span leaf=""> × </span><span><span leaf="">2000</span></span><span leaf=""><br></span><span leaf="">    obs: </span><span><span leaf="">'GEX_pct_counts_mt'</span></span><span leaf="">, </span><span><span leaf="">'GEX_n_counts'</span></span><span leaf="">, </span><span><span leaf="">'GEX_n_genes'</span></span><span leaf="">, </span><span><span leaf="">'GEX_size_factors'</span></span><span leaf="">, </span><span><span leaf="">'GEX_phase'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_nCount_peaks'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_atac_fragments'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_reads_in_peaks_frac'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_blacklist_fraction'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_nucleosome_signal'</span></span><span leaf="">, </span><span><span leaf="">'cell_type'</span></span><span leaf="">, </span><span><span leaf="">'batch'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_pseudotime_order'</span></span><span leaf="">, </span><span><span leaf="">'GEX_pseudotime_order'</span></span><span leaf="">, </span><span><span leaf="">'Samplename'</span></span><span leaf="">, </span><span><span leaf="">'Site'</span></span><span leaf="">, </span><span><span leaf="">'DonorNumber'</span></span><span leaf="">, </span><span><span leaf="">'Modality'</span></span><span leaf="">, </span><span><span leaf="">'VendorLot'</span></span><span leaf="">, </span><span><span leaf="">'DonorID'</span></span><span leaf="">, </span><span><span leaf="">'DonorAge'</span></span><span leaf="">, </span><span><span leaf="">'DonorBMI'</span></span><span leaf="">, </span><span><span leaf="">'DonorBloodType'</span></span><span leaf="">, </span><span><span leaf="">'DonorRace'</span></span><span leaf="">, </span><span><span leaf="">'Ethnicity'</span></span><span leaf="">, </span><span><span leaf="">'DonorGender'</span></span><span leaf="">, </span><span><span leaf="">'QCMeds'</span></span><span leaf="">, </span><span><span leaf="">'DonorSmoker'</span></span><span leaf="">, </span><span><span leaf="">'_scvi_batch'</span></span><span leaf="">, </span><span><span leaf="">'_scvi_labels'</span></span><span leaf=""><br></span><span leaf="">    var: </span><span><span leaf="">'feature_types'</span></span><span leaf="">, </span><span><span leaf="">'gene_id'</span></span><span leaf="">, </span><span><span leaf="">'n_cells'</span></span><span leaf="">, </span><span><span leaf="">'highly_variable'</span></span><span leaf="">, </span><span><span leaf="">'means'</span></span><span leaf="">, </span><span><span leaf="">'dispersions'</span></span><span leaf="">, </span><span><span leaf="">'dispersions_norm'</span></span><span leaf="">, </span><span><span leaf="">'highly_variable_nbatches'</span></span><span leaf="">, </span><span><span leaf="">'highly_variable_intersection'</span></span><span leaf=""><br></span><span leaf="">    uns: </span><span><span leaf="">'ATAC_gene_activity_var_names'</span></span><span leaf="">, </span><span><span leaf="">'batch_colors'</span></span><span leaf="">, </span><span><span leaf="">'cell_type_colors'</span></span><span leaf="">, </span><span><span leaf="">'dataset_id'</span></span><span leaf="">, </span><span><span leaf="">'genome'</span></span><span leaf="">, </span><span><span leaf="">'hvg'</span></span><span leaf="">, </span><span><span leaf="">'log1p'</span></span><span leaf="">, </span><span><span leaf="">'neighbors'</span></span><span leaf="">, </span><span><span leaf="">'organism'</span></span><span leaf="">, </span><span><span leaf="">'pca'</span></span><span leaf="">, </span><span><span leaf="">'umap'</span></span><span leaf="">, </span><span><span leaf="">'_scvi_uuid'</span></span><span leaf="">, </span><span><span leaf="">'_scvi_manager_uuid'</span></span><span leaf=""><br></span><span leaf="">    obsm: </span><span><span leaf="">'ATAC_gene_activity'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_lsi_full'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_lsi_red'</span></span><span leaf="">, </span><span><span leaf="">'ATAC_umap'</span></span><span leaf="">, </span><span><span leaf="">'GEX_X_pca'</span></span><span leaf="">, </span><span><span leaf="">'GEX_X_umap'</span></span><span leaf="">, </span><span><span leaf="">'X_pca'</span></span><span leaf="">, </span><span><span leaf="">'X_umap'</span></span><span leaf="">, </span><span><span leaf="">'X_scVI'</span></span><span leaf=""><br></span><span leaf="">    varm: </span><span><span leaf="">'PCs'</span></span><span leaf=""><br></span><span leaf="">    layers: </span><span><span leaf="">'counts'</span></span><span leaf="">, </span><span><span leaf="">'logcounts'</span></span><span leaf=""><br></span><span leaf="">    obsp: </span><span><span leaf="">'connectivities'</span></span><span leaf="">, </span><span><span leaf="">'distances'</span></span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">sc.pl.umap(adata_scvi, color=[label_key, batch_key], wspace=</span><span><span leaf="">1</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2Los8kIEum763iaPGg2MSjt4QTUvOokicibmqfziaZ7Jib7SyWCswkvkQKrQNXCrEkCcxtAbrFUEDnH7QGZLA/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.20555555555555555" data-type="png" data-w="1080" data-imgfileid="100053192" src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2Los8kIEum763iaPGg2MSjt4QTUvOokicibmqfziaZ7Jib7SyWCswkvkQKrQNXCrEkCcxtAbrFUEDnH7QGZLA/640?wx_fmt=png&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">之前批次之间是分开的，现在批次重叠更多。</span></p><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">7.6 使用细胞类型标签进行 VAE 整合</span></span><span></span><span></span></h2><p data-tool="mdnice编辑器"><span leaf="">用 </span><code><span leaf="">scVI</span></code><span leaf=""> 的时候没有用到细胞类型标签，当至少拥有部分细胞类型标签时，就可以使用 </span><code><span leaf="">scANVI</span></code><span leaf="">（single-cell ANnotation using Variational Inference），这是 </span><code><span leaf="">scVI</span></code><span leaf=""> 模型的扩展版本，能够同时整合细胞类型标签和批次信息。由于拥有这些额外的信息，</span><code><span leaf="">scANVI</span></code><span leaf=""> 可以在消除批次效应的同时尽力保持细胞标签之间的差异。基准测试表明，与 </span><code><span leaf="">scVI</span></code><span leaf=""> 相比，</span><code><span leaf="">scANVI</span></code><span leaf=""> 往往能更好地保留生物学信号，但有时在消除批次效应方面效果稍逊。</span></p><blockquote><span></span><p><strong><span leaf="">标签统一：</span></strong></p><p><span leaf="">如果使用 </span><code><span leaf="">scANVI</span></code><span leaf=""> 整合多个已有标签的数据集，需要首先进行标签的统一，指的是确保待整合数据集中标签的一致性。比如，某个细胞在一个数据集中可能被标注为 "T 细胞"，而相同类型的细胞在另一个数据集中可能被标记为 "CD8+ T 细胞"。如何最优地协调标签仍是一个开放性问题，通常需要领域专家的参与。</span></p></blockquote><p data-tool="mdnice编辑器"><code><span leaf="">scANVI</span></code><span leaf=""> 是在已训练的 </span><code><span leaf="">scVI</span></code><span leaf=""> 模型基础上进行优化，如果尚未训练 </span><code><span leaf="">scVI</span></code><span leaf=""> 模型，则需要先完成这一步，同时需要指定 </span><code><span leaf="">adata.obs</span></code><span leaf=""> 中包含细胞标签的列。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">model_scanvi = scvi.model.SCANVI.from_scvi_model(</span><span leaf=""><br></span><span leaf="">    model_scvi, labels_key=label_key, unlabeled_category=</span><span><span leaf="">"unlabelled"</span></span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">print(model_scanvi)</span><span leaf=""><br></span><span leaf="">model_scanvi.view_anndata_setup()</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2Los8kIEum763iaPGg2MSjt4QTUoTcEgIaK984MC5vfuq3hhRK0uG3kEGYGjiaAfze6y12gibXBpcPk8bAQ/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.09537037037037037" data-type="png" data-w="1080" data-imgfileid="100053193" src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2Los8kIEum763iaPGg2MSjt4QTUoTcEgIaK984MC5vfuq3hhRK0uG3kEGYGjiaAfze6y12gibXBpcPk8bAQ/640?wx_fmt=png&amp;from=appmsg"></span></figure><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2Los8kIEum763iaPGg2MSjt4QTU5vPoWOudafGeGl10ubeXqKuOl0LqibKaJRrzvCyT04viaWHIvYYzXKibw/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.8058124174372523" data-type="png" data-w="757" data-imgfileid="100053198" src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2Los8kIEum763iaPGg2MSjt4QTU5vPoWOudafGeGl10ubeXqKuOl0LqibKaJRrzvCyT04viaWHIvYYzXKibw/640?wx_fmt=png&amp;from=appmsg"></span></figure><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2Los8kIEum763iaPGg2MSjt4QTUFGKAHHfOibI87kMC4fDG22OuvahXjTFia7jMBayTbicc4U75K2GO4ia26A/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.9080459770114943" data-type="png" data-w="696" data-imgfileid="100053199" src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2Los8kIEum763iaPGg2MSjt4QTUFGKAHHfOibI87kMC4fDG22OuvahXjTFia7jMBayTbicc4U75K2GO4ia26A/640?wx_fmt=png&amp;from=appmsg"></span></figure><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2Los8kIEum763iaPGg2MSjt4QTUrwicxlK6UYicx99FRWzfYiaibBSHtwRwMiagiaPFQCQ73WHicKs72QfOc3AuQ/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.2898052691867125" data-type="png" data-w="873" data-imgfileid="100053197" src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2Los8kIEum763iaPGg2MSjt4QTUrwicxlK6UYicx99FRWzfYiaibBSHtwRwMiagiaPFQCQ73WHicKs72QfOc3AuQ/640?wx_fmt=png&amp;from=appmsg"></span></figure><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2Los8kIEum763iaPGg2MSjt4QTU5N3PnyVQSaribhAeOia5Xvmya4mZX1jl7yVBDSAibKVtm9ZPv35FXAPWA/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.760932944606414" data-type="png" data-w="1029" data-imgfileid="100053200" src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2Los8kIEum763iaPGg2MSjt4QTU5N3PnyVQSaribhAeOia5Xvmya4mZX1jl7yVBDSAibKVtm9ZPv35FXAPWA/640?wx_fmt=png&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><code><span leaf="">scANVI</span></code><span leaf=""> 模型和之前的 </span><code><span leaf="">scVI</span></code><span leaf=""> 有点类似。我们也可以更改模型的网络结构但这里使用的就是默认参数。同样的，使用启发式方法选择 epoch 的数量，这里比 </span><code><span leaf="">scVI</span></code><span leaf=""> 的少很多，因为这只是在完善 </span><code><span leaf="">scVI</span></code><span leaf=""> 模型而不是从头开始训练整个网络。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">max_epochs_scanvi = int(np.min([</span><span><span leaf="">10</span></span><span leaf="">, np.max([</span><span><span leaf="">2</span></span><span leaf="">, round(max_epochs_scvi / </span><span><span leaf="">3.0</span></span><span leaf="">)])]))</span><span leaf=""><br></span><span leaf="">model_scanvi.train(max_epochs=max_epochs_scanvi)</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">INFO     Training </span><span><span leaf="">for</span></span><span leaf=""> </span><span><span leaf="">10</span></span><span leaf=""> epochs.                                                                                   </span><span leaf=""><br></span><span leaf="">GPU available: </span><span><span leaf="">False</span></span><span leaf="">, used: </span><span><span leaf="">False</span></span><span leaf=""><br></span><span leaf="">TPU available: </span><span><span leaf="">False</span></span><span leaf="">, using: </span><span><span leaf="">0</span></span><span leaf=""> TPU cores</span><span leaf=""><br></span><span leaf="">HPU available: </span><span><span leaf="">False</span></span><span leaf="">, using: </span><span><span leaf="">0</span></span><span leaf=""> HPUs</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2Los8kIEum763iaPGg2MSjt4QTU1HBOc3yrTn5clMbibCb8APff1FDvLj7ia5tMwtGAYvt0nVGLIjzGD7iaA/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.028703703703703703" data-type="png" data-w="1080" data-imgfileid="100053196" src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2Los8kIEum763iaPGg2MSjt4QTU1HBOc3yrTn5clMbibCb8APff1FDvLj7ia5tMwtGAYvt0nVGLIjzGD7iaA/640?wx_fmt=png&amp;from=appmsg"></span></figure><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">`Trainer.fit` stopped: `max_epochs=</span><span><span leaf="">10</span></span><span leaf="">` reached.</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">从模型中提取新的潜在表示，并创建一个新的 UMAP 嵌入：</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">adata_scanvi = adata_scvi.copy()</span><span leaf=""><br></span><span leaf="">adata_scanvi.obsm[</span><span><span leaf="">"X_scANVI"</span></span><span leaf="">] = model_scanvi.get_latent_representation()</span><span leaf=""><br></span><span leaf="">sc.pp.neighbors(adata_scanvi, use_rep=</span><span><span leaf="">"X_scANVI"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">sc.tl.umap(adata_scanvi)</span><span leaf=""><br></span><span leaf="">sc.pl.umap(adata_scanvi, color=[label_key, batch_key], wspace=</span><span><span leaf="">1</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2Los8kIEum763iaPGg2MSjt4QTUDtNIibWdE3faUmuX7RbmcN9GXF90wI2UBbzD2ILgtfgyQEib4R0P9GxA/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.20555555555555555" data-type="png" data-w="1080" data-imgfileid="100053205" src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2Los8kIEum763iaPGg2MSjt4QTUDtNIibWdE3faUmuX7RbmcN9GXF90wI2UBbzD2ILgtfgyQEib4R0P9GxA/640?wx_fmt=png&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">光看 UMAP 很难看出 </span><code><span leaf="">scANVI</span></code><span leaf=""> 和 </span><code><span leaf="">scVI</span></code><span leaf=""> 之间的差别，但后面可以用具体的指标来量化它们的效能。</span></p><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">参考文献</span></span><span></span><span></span></h2><ol><li><section><span leaf="">Maren Büttner, Zhichao Miao, F Alexander Wolf, Sarah A Teichmann, and Fabian J Theis. A test metric for assessing single-cell RNA-seq batch correction. </span><em><span leaf="">Nature methods</span></em><span leaf="">, 16(1):43–49, January 2019. doi:10.1038/s41592-018-0254-1                                    .</span></section></li><li><section><span leaf="">Hoa Thi Nhu Tran, Kok Siong Ang, Marion Chevrier, Xiaomeng Zhang, Nicole Yee Shin Lee, Michelle Goh, and Jinmiao Chen. A benchmark of batch-effect correction methods for single-cell RNA sequencing data. </span><em><span leaf="">Genome biology</span></em><span leaf="">, 21(1):12, January 2020. doi:10.1186/s13059-019-1850-9                                    .</span></section></li><li><section><span leaf="">Ruben Chazarra-Gil, Stijn van Dongen, Vladimir Yu Kiselev, and Martin Hemberg. Flexible comparison of batch correction methods for single-cell RNA-seq using BatchBench. </span><em><span leaf="">Nucleic acids research</span></em><span leaf="">, 49(7):e42, April 2021. doi:10.1093/nar/gkab004                                    .</span></section></li><li><section><span leaf="">Malte D Luecken, M Büttner, K Chaichoompu, A Danese, M Interlandi, M F Mueller, D C Strobl, L Zappia, M Dugas, M Colomé-Tatché, and Fabian J Theis. Benchmarking atlas-level data integration in single-cell genomics. </span><em><span leaf="">Nature methods</span></em><span leaf="">, December 2021. doi:10.1038/s41592-021-01336-8                                    .</span></section></li></ol></section><p data-tool="mdnice编辑器"><span leaf=""><br></span></p></section><p><mp-style-type data-value="10000"></mp-style-type></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/E05v1MjHYc6DEhwl8SCd5g",target="_blank" rel="noopener noreferrer">原文链接</a>
