---
title: "gsMap软件代码教程"
date: 2025-09-03T03:25:44Z
draft: ["false"]
tags: [
  "fetched",
  "生信宝库"
]
categories: ["Acdemic"]
---
gsMap软件代码教程 by 生信宝库
------
<div><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com" data-pm-slice="0 0 []"><h1 data-tool="mdnice编辑器"><span></span><span><span leaf="">前言</span></span><span></span></h1><p data-tool="mdnice编辑器"><span leaf="">gsMap是一个整合空间转录组和GWAS数据用于推断与复杂性状关联的细胞类型和空间区域的软件包。在之前的推文：<a target="_blank" href="https://mp.weixin.qq.com/s?__biz=MzI4MjY5ODI1Nw==&amp;mid=2247495882&amp;idx=1&amp;sn=fd13fd8f2a48fc5fc4d970dd38fcabd2&amp;scene=21#wechat_redirect" textvalue="Nature重磅：gsMap工具实现基因组和空转联合分析" data-itemshowtype="0" linktype="text" data-linktype="2">Nature重磅：gsMap工具实现基因组和空转联合分析</a>中，我们首先介绍了gsMap的功能框架；并且在后续推文：<a target="_blank" href="https://mp.weixin.qq.com/s?__biz=MzI4MjY5ODI1Nw==&amp;mid=2247495883&amp;idx=1&amp;sn=f54cb5a437ef2d8e3c6a854375e0400f&amp;scene=21#wechat_redirect" textvalue="gsMap算法原理介绍" data-itemshowtype="0" linktype="text" data-linktype="2">gsMap算法原理介绍</a>中，介绍了gsMap的基本工作原理；本文将继续通过代码实操，介绍gsMap的具体使用方法。</span></p><h1 data-tool="mdnice编辑器"><span></span><span><span leaf="">安装</span></span><span></span></h1><p data-tool="mdnice编辑器"><span leaf="">首先创建Python 3.10的Conda环境</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">conda create -n gsMap python=3.10</span><span leaf=""><br></span><span leaf="">conda activate gsMap</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">gsMap有3种安装方法，小编采用的是使用pip安装</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">pip install gsMap</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">也能使用conda安装</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">conda install bioconda::gsmap</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">当pip或者conda自动安装失败时，也能本地源码安装</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">git </span><span><span leaf="">clone</span></span><span leaf=""> https://github.com/JianYang-Lab/gsMap</span><span leaf=""><br></span><span><span leaf="">cd</span></span><span leaf=""> gsMap</span><span leaf=""><br></span><span leaf="">pip install -e .</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">安装好后，使用帮助命令检验是否安装成功</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">gsmap --</span><span><span leaf="">help</span></span><span leaf=""><br></span></code></pre><h1 data-tool="mdnice编辑器"><span></span><span><span leaf="">数据格式</span></span><span></span></h1><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">下载参考信息</span></span><span></span></h2><p data-tool="mdnice编辑器"><span leaf="">gsMap首先需要一系列参考信息，这些可通过如下命令下载（有条件建议开🪜更快）</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">wget https://yanglab.westlake.edu.cn/data/gsMap/gsMap_resource.tar.gz</span><span leaf=""><br></span><span leaf="">tar -xvzf gsMap_resource.tar.gz</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">检查一下下载的文件</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">tree -L 2 gsMap_resource</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">gsMap_resource</span><span leaf=""><br></span><span leaf="">├── genome_annotation</span><span leaf=""><br></span><span leaf="">│   ├── enhancer</span><span leaf=""><br></span><span leaf="">│   └── gtf</span><span leaf=""><br></span><span leaf="">├── homologs</span><span leaf=""><br></span><span leaf="">│   ├── human_macaque_marmoset_mouse_homologs.txt.gz</span><span leaf=""><br></span><span leaf="">│   ├── macaque_human_homologs.txt</span><span leaf=""><br></span><span leaf="">│   └── mouse_human_homologs.txt</span><span leaf=""><br></span><span leaf="">├── LD_Reference_Panel</span><span leaf=""><br></span><span leaf="">│   └── 1000G_EUR_Phase3_plink</span><span leaf=""><br></span><span leaf="">├── LDSC_resource</span><span leaf=""><br></span><span leaf="">│   ├── hapmap3_snps</span><span leaf=""><br></span><span leaf="">│   └── weights_hm3_no_hla</span><span leaf=""><br></span><span leaf="">└── quick_mode</span><span leaf=""><br></span><span leaf="">    ├── baseline</span><span leaf=""><br></span><span leaf="">    ├── SNP_gene_pair</span><span leaf=""><br></span><span leaf="">    └── snp_gene_weight_matrix.h5ad</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">可以发现，参考数据主要包含参考基因组注释信息，各组织的enhancer调控信息，跨物种同源基因信息、LD参考数据、SNP权重信息以及一些缓存好的人的SNP和基因对应信息，便于快速读取。这些参考数据在每次运行gsMap时直接使用即可。</span></p><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">下载示例数据</span></span><span></span></h2><p data-tool="mdnice编辑器"><span leaf="">作者提供了一套示例输入数据供软件测试和学习，使用以下命令下载</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">wget https://yanglab.westlake.edu.cn/data/gsMap/gsMap_example_data.tar.gz</span><span leaf=""><br></span><span leaf="">tar -xvzf gsMap_example_data.tar.gz</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">检查示例数据的目录</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">gsMap_example_data</span><span leaf=""><br></span><span leaf="">├── GWAS</span><span leaf=""><br></span><span leaf="">│   ├── BCX2_MCHC_EA_GWAMA.sumstats.gz</span><span leaf=""><br></span><span leaf="">│   ├── GIANT_EUR_Height_2022_Nature.sumstats.gz</span><span leaf=""><br></span><span leaf="">│   ├── gwas_config.yaml</span><span leaf=""><br></span><span leaf="">│   └── IQ_NG_2018.sumstats.gz</span><span leaf=""><br></span><span leaf="">└── ST</span><span leaf=""><br></span><span leaf="">    ├── E16.5_E1S1.MOSTA.h5ad</span><span leaf=""><br></span><span leaf="">    └── E16.5_E2S11.MOSTA.h5ad</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">示例数据包含输入的空转（ST）和GWAS数据。空转数据是大家非常熟悉的h5ad格式，能直接被Scanpy读取为AnnData对象</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">import</span></span><span leaf=""> scanpy </span><span><span leaf="">as</span></span><span leaf=""> sc</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">adata = sc.read_h5ad(</span><span><span leaf="">"gsMap_example_data/ST/E16.5_E1S1.MOSTA.h5ad"</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">AnnData对象中必须包含原始count矩阵、空间坐标信息。细胞/组织注释信息为可选项。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">print(adata.layers[</span><span><span leaf="">'count'</span></span><span leaf="">].shape)</span><span leaf=""><br></span><span leaf="">print(adata.obsm[</span><span><span leaf="">"spatial"</span></span><span leaf="">].shape)</span><span leaf=""><br></span><span leaf="">print(adata.obs[</span><span><span leaf="">"annotation"</span></span><span leaf="">].value_counts().head())</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">(121767, 28204)</span><span leaf=""><br></span><span leaf="">(121767, 2)</span><span leaf=""><br></span><span leaf="">annotation</span><span leaf=""><br></span><span leaf="">Brain                17374</span><span leaf=""><br></span><span leaf="">Liver                14167</span><span leaf=""><br></span><span leaf="">Cavity               11287</span><span leaf=""><br></span><span leaf="">Muscle                9853</span><span leaf=""><br></span><span leaf="">Connective tissue     8803</span><span leaf=""><br></span><span leaf="">Name: count, dtype: int64</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><em><span leaf="">这示例数据的细胞量比我做的正式课题还多，没点钞能力发不了nature啊。</span></em><span leaf=""><br></span><span leaf="">GWAS数据需整理成以下格式，包含SNP名称、变异前后碱基、Z检验统计量和样本量。论文中</span><span data-formula="\chi^2=Z^2"><span data-formula="\chi^2=Z^2"><svg xmlns="http://www.w3.org/2000/svg" role="img" focusable="false" viewbox="0 -833.9 3547.7 1037.9" aria-hidden="true"><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="matrix(1 0 0 -1 0 0)"><g data-mml-node="math"><g data-mml-node="msup"><g data-mml-node="mi"><path data-c="3C7" d="M576 -125Q576 -147 547 -175T487 -204H476Q394 -204 363 -157Q334 -114 293 26L284 59Q283 58 248 19T170 -66T92 -151T53 -191Q49 -194 43 -194Q36 -194 31 -189T25 -177T38 -154T151 -30L272 102L265 131Q189 405 135 405Q104 405 87 358Q86 351 68 351Q48 351 48 361Q48 369 56 386T89 423T148 442Q224 442 258 400Q276 375 297 320T330 222L341 180Q344 180 455 303T573 429Q579 431 582 431Q600 431 600 414Q600 407 587 392T477 270Q356 138 353 134L362 102Q392 -10 428 -89T490 -168Q504 -168 517 -156T536 -126Q539 -116 543 -115T557 -114T571 -115Q576 -118 576 -125Z"></path></g><g data-mml-node="mn" transform="translate(626, 363) scale(0.707)"><path data-c="32" d="M109 429Q82 429 66 447T50 491Q50 562 103 614T235 666Q326 666 387 610T449 465Q449 422 429 383T381 315T301 241Q265 210 201 149L142 93L218 92Q375 92 385 97Q392 99 409 186V189H449V186Q448 183 436 95T421 3V0H50V19V31Q50 38 56 46T86 81Q115 113 136 137Q145 147 170 174T204 211T233 244T261 278T284 308T305 340T320 369T333 401T340 431T343 464Q343 527 309 573T212 619Q179 619 154 602T119 569T109 550Q109 549 114 549Q132 549 151 535T170 489Q170 464 154 447T109 429Z"></path></g></g><g data-mml-node="mo" transform="translate(1307.3, 0)"><path data-c="3D" d="M56 347Q56 360 70 367H707Q722 359 722 347Q722 336 708 328L390 327H72Q56 332 56 347ZM56 153Q56 168 72 173H708Q722 163 722 153Q722 140 707 133H70Q56 140 56 153Z"></path></g><g data-mml-node="msup" transform="translate(2363.1, 0)"><g data-mml-node="mi"><path data-c="5A" d="M58 8Q58 23 64 35Q64 36 329 334T596 635L586 637Q575 637 512 637H500H476Q442 637 420 635T365 624T311 598T266 548T228 469Q227 466 226 463T224 458T223 453T222 450L221 448Q218 443 202 443Q185 443 182 453L214 561Q228 606 241 651Q249 679 253 681Q256 683 487 683H718Q723 678 723 675Q723 673 717 649Q189 54 188 52L185 49H274Q369 50 377 51Q452 60 500 100T579 247Q587 272 590 277T603 282H607Q628 282 628 271Q547 5 541 2Q538 0 300 0H124Q58 0 58 8Z"></path></g><g data-mml-node="mn" transform="translate(781, 363) scale(0.707)"><path data-c="32" d="M109 429Q82 429 66 447T50 491Q50 562 103 614T235 666Q326 666 387 610T449 465Q449 422 429 383T381 315T301 241Q265 210 201 149L142 93L218 92Q375 92 385 97Q392 99 409 186V189H449V186Q448 183 436 95T421 3V0H50V19V31Q50 38 56 46T86 81Q115 113 136 137Q145 147 170 174T204 211T233 244T261 278T284 308T305 340T320 369T333 401T340 431T343 464Q343 527 309 573T212 619Q179 619 154 602T119 569T109 550Q109 549 114 549Q132 549 151 535T170 489Q170 464 154 447T109 429Z"></path></g></g></g></g><g></g></svg></span></span><span leaf="">。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">zcat IQ_NG_2018.sumstats.gz | head -n 5</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">SNP A1 A2 Z N</span><span leaf=""><br></span><span leaf="">rs12184267 T C 0.916 225955</span><span leaf=""><br></span><span leaf="">rs12184277 G A 0.656 226215</span><span leaf=""><br></span><span leaf="">rs12184279 A C 1.050 226224</span><span leaf=""><br></span><span leaf="">rs116801199 T G 0.300 226626</span><span leaf=""><br></span></code></pre><h1 data-tool="mdnice编辑器"><span></span><span><span leaf="">快速入门</span></span><span></span></h1><p data-tool="mdnice编辑器"><span leaf="">整理好输入数据以后，运行gsMap就很方便了，只需一行命令：</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">gsmap quick_mode \</span><span leaf=""><br></span><span leaf="">   --workdir </span><span><span leaf="">'.'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">   --homolog_file </span><span><span leaf="">'../gsMap_resource/homologs/mouse_human_homologs.txt'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">   --sample_name </span><span><span leaf="">'E16.5_E1S1.MOSTA'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">   --gsMap_resource_dir </span><span><span leaf="">'../gsMap_resource'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">   --hdf5_path </span><span><span leaf="">'../gsMap_example_data/ST/E16.5_E1S1.MOSTA.h5ad'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">   --annotation </span><span><span leaf="">'annotation'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">   --data_layer </span><span><span leaf="">'count'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">   --sumstats_file </span><span><span leaf="">'../gsMap_example_data/GWAS/IQ_NG_2018.sumstats.gz'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">   --trait_name </span><span><span leaf="">'IQ'</span></span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">这里的参数也非常显而易见，相信不需要过多解释小伙伴们也能看懂。就是注意</span><code><span leaf="">workdir</span></code><span leaf="">参数表示的是输出文件夹的位置，而输出文件夹则按</span><code><span leaf="">sample_name</span></code><span leaf="">参数进行命名。</span><span leaf=""><br></span><span leaf="">一套全新的数据运行时间会很长，这里大概运行了8个多小时，得到了以下输出文件夹</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">tree -L 2 E16.5_E1S1.MOSTA/</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">E16.5_E1S1.MOSTA/</span><span leaf=""><br></span><span leaf="">├── cauchy_combination</span><span leaf=""><br></span><span leaf="">│   └── E16.5_E1S1.MOSTA_IQ.Cauchy.csv.gz</span><span leaf=""><br></span><span leaf="">├── find_latent_representations</span><span leaf=""><br></span><span leaf="">│   └── E16.5_E1S1.MOSTA_add_latent.h5ad</span><span leaf=""><br></span><span leaf="">├── generate_ldscore</span><span leaf=""><br></span><span leaf="">│   ├── baseline -&gt; /data/h01006/software_test/gsMap/gsMap_resource/quick_mode/baseline</span><span leaf=""><br></span><span leaf="">│   ├── E16.5_E1S1.MOSTA_generate_ldscore.done</span><span leaf=""><br></span><span leaf="">│   └── SNP_gene_pair -&gt; /data/h01006/software_test/gsMap/gsMap_resource/quick_mode/SNP_gene_pair</span><span leaf=""><br></span><span leaf="">├── gsMap_pipeline_E16.5_E1S1.MOSTA_20250701_171043.log</span><span leaf=""><br></span><span leaf="">├── latent_to_gene</span><span leaf=""><br></span><span leaf="">│   └── E16.5_E1S1.MOSTA_gene_marker_score.feather</span><span leaf=""><br></span><span leaf="">├── report</span><span leaf=""><br></span><span leaf="">│   └── IQ</span><span leaf=""><br></span><span leaf="">└── spatial_ldsc</span><span leaf=""><br></span><span leaf="">    └── E16.5_E1S1.MOSTA_IQ.csv.gz</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">其他的输出文件我们后面再介绍，现在主要关注report文件夹。里面按照性状名称储存结果，每个结果都有一个html格式的报告文件，可以用浏览器打开。</span><span leaf=""><br></span><span leaf="">第一个结果是性状与spot的关联性p值的组织空间分布图，展示了空间组织中每个spot的-log10(p)值。这个图是交互式的，将光标移上去可以看到对应spot的注释、空间位置和p值，双击右边的组织类型还能单独观测对应的组织。<img data-src="https://mmbiz.qpic.cn/mmbiz_png/GL6g5Y3aR7calxDyI9mY9DEr36eCTibCYM5Tia6AHuAKpLaEvK3ycVphB9OvAFQq9bodquqyxn8p3Vw6YrXoBThA/640?wx_fmt=png&amp;from=appmsg" data-ratio="1.099074074074074" data-type="png" data-w="1080" data-imgfileid="100012394" src="https://mmbiz.qpic.cn/mmbiz_png/GL6g5Y3aR7calxDyI9mY9DEr36eCTibCYM5Tia6AHuAKpLaEvK3ycVphB9OvAFQq9bodquqyxn8p3Vw6YrXoBThA/640?wx_fmt=png&amp;from=appmsg">然后是各组织柯西合并检验后的结果，可以发现柯西合并检验的p值要明显小于单纯的区域内p值的中位数。<img data-src="https://mmbiz.qpic.cn/mmbiz_png/GL6g5Y3aR7calxDyI9mY9DEr36eCTibCYomRyDW3RttUQ53hZ6l10O0C7UJgLY57GGGKdR0jXicDBMhpBK73UFtA/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.2175925925925926" data-type="png" data-w="1080" data-imgfileid="100012390" src="https://mmbiz.qpic.cn/mmbiz_png/GL6g5Y3aR7calxDyI9mY9DEr36eCTibCYomRyDW3RttUQ53hZ6l10O0C7UJgLY57GGGKdR0jXicDBMhpBK73UFtA/640?wx_fmt=png&amp;from=appmsg">曼哈顿图整体呈现的是GWAS的结果，展示了SNP与性状的关联性p值。这张图也是交互式的，光标指向SNP可以看到它的基因组位置、SNP名称、对应的基因、注释的细胞或组织以及注释LD分数与GWAS卡方统计量的Pearson相关系数（PCC）。<img data-src="https://mmbiz.qpic.cn/mmbiz_png/GL6g5Y3aR7calxDyI9mY9DEr36eCTibCYpuuzAw1nMA62XVxPID0ImEEE6CEBYeTic7LrdQbiaEtD6R9BR3j0EQcg/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.23981481481481481" data-type="png" data-w="1080" data-imgfileid="100012392" src="https://mmbiz.qpic.cn/mmbiz_png/GL6g5Y3aR7calxDyI9mY9DEr36eCTibCYpuuzAw1nMA62XVxPID0ImEEE6CEBYeTic7LrdQbiaEtD6R9BR3j0EQcg/640?wx_fmt=png&amp;from=appmsg">然后是基因表达和GSS分数的分布，可以选择任何基因查看。可以看出，GSS比基因表达量差异更为明显。<img data-src="https://mmbiz.qpic.cn/mmbiz_png/GL6g5Y3aR7calxDyI9mY9DEr36eCTibCYJdAkvOhzL3NibddmOX15bd62ic1wLzmr1tUTXNL4hz9Hp9lpJAbAVJTQ/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.9351851851851852" data-type="png" data-w="1080" data-imgfileid="100012391" src="https://mmbiz.qpic.cn/mmbiz_png/GL6g5Y3aR7calxDyI9mY9DEr36eCTibCYJdAkvOhzL3NibddmOX15bd62ic1wLzmr1tUTXNL4hz9Hp9lpJAbAVJTQ/640?wx_fmt=png&amp;from=appmsg"><img data-src="https://mmbiz.qpic.cn/mmbiz_png/GL6g5Y3aR7calxDyI9mY9DEr36eCTibCYryIQ6Lje5fqNeZLSBFd2urkAnUllXdibcibPiaEBickXHf9HlEgBUx5e4w/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.7435185185185185" data-type="png" data-w="1080" data-imgfileid="100012393" src="https://mmbiz.qpic.cn/mmbiz_png/GL6g5Y3aR7calxDyI9mY9DEr36eCTibCYryIQ6Lje5fqNeZLSBFd2urkAnUllXdibcibPiaEBickXHf9HlEgBUx5e4w/640?wx_fmt=png&amp;from=appmsg">然后是50个表达最特异的基因和组织区域/细胞类型对及其GSS和PCC的展示。<img data-src="https://mmbiz.qpic.cn/mmbiz_png/GL6g5Y3aR7calxDyI9mY9DEr36eCTibCY2heMtXsgClnsAwWHlg6xwjA7bqQUoT2LLyiaEfZPayicuU5Y3WrcKz3Q/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.6537037037037037" data-type="png" data-w="1080" data-imgfileid="100012396" src="https://mmbiz.qpic.cn/mmbiz_png/GL6g5Y3aR7calxDyI9mY9DEr36eCTibCY2heMtXsgClnsAwWHlg6xwjA7bqQUoT2LLyiaEfZPayicuU5Y3WrcKz3Q/640?wx_fmt=png&amp;from=appmsg">最后就是一些运行相关信息了<img data-src="https://mmbiz.qpic.cn/mmbiz_png/GL6g5Y3aR7calxDyI9mY9DEr36eCTibCYnd3EgmoqCh7JJwP8kicqdXD7UibLc5AGKg5icWibiaXdW9OERoUKEjT7O5A/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.5611111111111111" data-type="png" data-w="1080" data-imgfileid="100012395" src="https://mmbiz.qpic.cn/mmbiz_png/GL6g5Y3aR7calxDyI9mY9DEr36eCTibCYnd3EgmoqCh7JJwP8kicqdXD7UibLc5AGKg5icWibiaXdW9OERoUKEjT7O5A/640?wx_fmt=png&amp;from=appmsg"></span></p><h1 data-tool="mdnice编辑器"><span></span><span><span leaf="">分步运行</span></span><span></span></h1><p data-tool="mdnice编辑器"><span leaf="">为了帮助用户更好的理解gsMap算法的底层逻辑以及让数据处理过程更灵活，作者设置了gsMap的分布运行功能，下面是gsMap每个步骤分开运行的教程。</span></p><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">寻找潜在表征（可选）</span></span><span></span></h2><p data-tool="mdnice编辑器"><span leaf="">gsMap的第一步是将空间转录组的表达矩阵、空间坐标及细胞类型输入GAT网络中学习每个spot的潜在嵌入表征。它由下面一行代码完成：</span><span leaf=""><br></span><strong><span leaf="">这一步12万个细胞大约需要60G内存</span></strong></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">gsmap run_find_latent_representations \</span><span leaf=""><br></span><span leaf="">    --workdir </span><span><span leaf="">'.'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">    --sample_name </span><span><span leaf="">'E16.5_E1S1.MOSTA'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">    --input_hdf5_path </span><span><span leaf="">'../gsMap_example_data/ST/E16.5_E1S1.MOSTA.h5ad'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">    --annotation </span><span><span leaf="">'annotation'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">    --data_layer </span><span><span leaf="">'count'</span></span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">运行完成后，在输出目录中生成了目录</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">find_latent_representations/</span><span leaf=""><br></span><span leaf="">└── E16.5_E1S1.MOSTA_add_latent.h5ad</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">使用scanpy读取这个h5ad文件，里面存储的是一个adata对象。它的</span><code><span leaf="">obsm</span></code><span leaf="">中多出了</span><code><span leaf="">latent_GVAE</span></code><span leaf="">，是一个二维数组，每一行存储着每个spot的30维嵌入。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">import</span></span><span leaf=""> scanpy </span><span><span leaf="">as</span></span><span leaf=""> sc</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">adata = sc.read_h5ad(</span><span><span leaf="">"step/E16.5_E1S1.MOSTA/find_latent_representations/E16.5_E1S1.MOSTA_add_latent.h5ad"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">adata.obsm[</span><span><span leaf="">'latent_GVAE'</span></span><span leaf="">].shape</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">array([[ 0.00952553, -0.08907691, -0.06804881, ..., -0.03316928, -0.04227457,  0.01298398],</span><span leaf=""><br></span><span leaf="">       [ 0.01311077, -0.09554639, -0.06435487, ..., -0.03155136, -0.04299155,  0.0202301 ],</span><span leaf=""><br></span><span leaf="">       [ 0.01806523, -0.09366009, -0.06114935, ..., -0.02755435, -0.0463993 ,  0.02227518],</span><span leaf=""><br></span><span leaf="">       ...,</span><span leaf=""><br></span><span leaf="">       [-0.03978971, -0.07202866, -0.02012973, ..., -0.01995548, -0.06187748, -0.00288487],</span><span leaf=""><br></span><span leaf="">       [-0.03788416, -0.07053255, -0.01871764, ..., -0.02421809, -0.06079815,  0.00054346],</span><span leaf=""><br></span><span leaf="">       [-0.04278402, -0.06833802, -0.01483879, ..., -0.02659683, -0.06472453,  0.00366877]], dtype=float32)</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">当然这一步是可选的，如果小伙伴们不喜欢使用GVAE获得潜在嵌入的话也可以通过其他降维方法如PCA等来得到spot的表征。</span></p><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">计算GSS</span></span><span></span></h2><p data-tool="mdnice编辑器"><span leaf="">在获得了spot的表征后，gsMap通过表征间的余弦相似度确定空间域，然后计算每个spot为中心的空间域的GSS分数。这个过程用下面的代码实现，有两个很重要的参数：</span></p><ul><li><section><code><span leaf="">num_neighbour_spatial</span></code><span leaf="">：表示与focal spot空间欧氏距离最近的201个spot，这是初步划定的搜索范围。</span></section></li><li><section><code><span leaf="">num_neighbour</span></code><span leaf="">：表示在上述搜索范围内选择与focal spot嵌入的余弦相似度最高的51个spot作为微空间域。</span><code><span leaf="">mouse_human_homologs.txt</span></code><span leaf="">是同源基因转换的参考文件，它要求左边列给的是空转对应物种的基因名，右边列给的是GWAS对应物种的基因名。</span></section></li></ul><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">head mouse_human_homologs.txt</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">MOUSE_GENE_SYM HUMAN_GENE_SYM</span><span leaf=""><br></span><span leaf="">Hsfy2 HSFY1</span><span leaf=""><br></span><span leaf="">Gm5788 EIF1AY</span><span leaf=""><br></span><span leaf="">Gm21677 RBMY1B</span><span leaf=""><br></span><span leaf="">Tbl1x TBL1Y</span><span leaf=""><br></span><span leaf="">Pcdh11x PCDH11Y</span><span leaf=""><br></span><span leaf="">Uty UTY</span><span leaf=""><br></span><span leaf="">Dazl DAZ1</span><span leaf=""><br></span><span leaf="">Tgif2lx2 TGIF2LY</span><span leaf=""><br></span><span leaf="">Prkx PRKY</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">运行这一步的代码：</span><span leaf=""><br></span><strong><span leaf="">这一步12万个细胞大约需要45G内存</span></strong></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">gsmap run_latent_to_gene \</span><span leaf=""><br></span><span leaf="">    --workdir </span><span><span leaf="">'.'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">    --sample_name </span><span><span leaf="">'E16.5_E1S1.MOSTA'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">    --annotation </span><span><span leaf="">'annotation'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">    --latent_representation </span><span><span leaf="">'latent_GVAE'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">    --num_neighbour 51 \</span><span leaf=""><br></span><span leaf="">    --num_neighbour_spatial 201 \</span><span leaf=""><br></span><span leaf="">    --homolog_file </span><span><span leaf="">'../gsMap_resource/homologs/mouse_human_homologs.txt'</span></span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">运行完以后得到以下结果</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">latent_to_gene/</span><span leaf=""><br></span><span leaf="">└── E16.5_E1S1.MOSTA_gene_marker_score.feather</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">feather文件是一种二进制文件用于更高效地储存数据框，可以使用pyarrow包进行读取。这是一个行为基因，列为spot的GSS打分矩阵。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">import</span></span><span leaf=""> pyarrow.feather </span><span><span leaf="">as</span></span><span leaf=""> feather</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">gss = feather.read_feather(</span><span><span leaf="">"step/E16.5_E1S1.MOSTA/latent_to_gene/E16.5_E1S1.MOSTA_gene_marker_score.feather"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">gss.iloc[</span><span><span leaf="">0</span></span><span leaf="">:</span><span><span leaf="">5</span></span><span leaf="">, </span><span><span leaf="">0</span></span><span leaf="">:</span><span><span leaf="">5</span></span><span leaf="">]</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">cell_name HUMAN_GENE_SYM 190_194 190_195 190_196 190_197</span><span leaf=""><br></span><span leaf="">0 C11orf58 0.000000 0.00000 0.00000 0.000000</span><span leaf=""><br></span><span leaf="">1 C18orf21 0.000000 0.00000 0.00000 0.000000</span><span leaf=""><br></span><span leaf="">2 C2orf76 1.933594 1.84082 1.84082 1.834961</span><span leaf=""><br></span><span leaf="">3 KIAA0930 0.000000 0.00000 0.00000 0.000000</span><span leaf=""><br></span><span leaf="">4 C2orf49 0.000000 0.00000 0.00000 0.000000</span><span leaf=""><br></span></code></pre><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">生成LD分数</span></span><span></span></h2><p data-tool="mdnice编辑器"><span leaf="">得到基因的GSS后，下一步就是建立基因和SNP的映射。有以下3种映射方式：</span></p><ol><li><section><span leaf="">通过基因TSS位置，需要指定参数：</span></section></li></ol><ul><li><section><code><span leaf="">gene_window_size</span></code><span leaf="">：基因body上下游的距离。在这个窗口内的SNP都会直接跟相应的基因映射。如果一个SNP落入多个基因的窗口中，那么将会映射到距离TSS最近的基因。</span></section></li></ul><ol start="2"><li><section><span leaf="">通过enhancer和基因的调控关系，落入基因body和enhancer的SNP都会映射到对应基因，需要指定以下参数：</span></section></li></ol><ul><li><section><code><span leaf="">enhancer_annotation_file</span></code><span leaf="">：指定enhancer-基因对应关系的注释文件。</span></section></li><li><section><code><span leaf="">snp_multienhancer_strategy</span></code><span leaf="">：当SNP落入多个基因的enhancer区域时采取的策略。'max_mkscore'表示映射到GSS最大的基因；'nearest_TSS'表示映射到TSS最近的基因。</span></section></li><li><section><code><span leaf="">gene_window_enhancer_priority</span></code><span leaf="">：当SNP同时落入基因窗口和enhancer时采取的策略。在仅考虑enhancer时需设为'enhancer_only'。在下一种情况下，'gene_window_first'和'enhancer_first'分别表示优先映射基因窗口和enhancer。</span><span leaf=""><br></span><span leaf="">enhancer注释文件为这种形式：</span></section></li></ul><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">head ABC_roadmap_merged.bed</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">Chromosome Start End symbol</span><span leaf=""><br></span><span leaf="">chr1 32803048 32803704 A3GALT2</span><span leaf=""><br></span><span leaf="">chr1 32856711 32857379 A3GALT2</span><span leaf=""><br></span><span leaf="">chr1 33169042 33169769 A3GALT2</span><span leaf=""><br></span><span leaf="">chr1 33225064 33226048 A3GALT2</span><span leaf=""><br></span><span leaf="">chr1 33226459 33227433 A3GALT2</span><span leaf=""><br></span><span leaf="">chr1 33446521 33447063 A3GALT2</span><span leaf=""><br></span><span leaf="">chr1 33454352 33455149 A3GALT2</span><span leaf=""><br></span><span leaf="">chr1 33457470 33459019 A3GALT2</span><span leaf=""><br></span><span leaf="">chr1 33515824 33516453 A3GALT2</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><strong><span leaf="">gsMap的参考文件中还提供了人不同组织器官中的enhancer注释，可以根据自己分析的样本类型灵活选择，而快速模式则没有这个参数可选。</span></strong><span leaf=""><br></span><span leaf="">3. 同时考虑基因body上下游窗口和enhancer，需要指定上述所有参数，小编这次用这种情况进行演示，这步使用以下代码运行。</span><span leaf=""><br></span><strong><span leaf="">这一步大约需要40G内存</span></strong></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">for</span></span><span leaf=""> CHROM </span><span><span leaf="">in</span></span><span leaf=""> {1..22}</span><span leaf=""><br></span><span><span leaf="">do</span></span><span leaf=""><br></span><span leaf="">    gsmap run_generate_ldscore \</span><span leaf=""><br></span><span leaf="">        --workdir </span><span><span leaf="">'.'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">        --sample_name </span><span><span leaf="">'E16.5_E1S1.MOSTA'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">        --chrom </span><span><span leaf="">$CHROM</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">        --bfile_root </span><span><span leaf="">'../gsMap_resource/LD_Reference_Panel/1000G_EUR_Phase3_plink/1000G.EUR.QC'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">        --keep_snp_root </span><span><span leaf="">'../gsMap_resource/LDSC_resource/hapmap3_snps/hm'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">        --gtf_annotation_file </span><span><span leaf="">'../gsMap_resource/genome_annotation/gtf/gencode.v46lift37.basic.annotation.gtf'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">        --gene_window_size 50000 \</span><span leaf=""><br></span><span leaf="">        --enhancer_annotation_file </span><span><span leaf="">'../gsMap_resource/genome_annotation/enhancer/by_tissue/ALL/ABC_roadmap_merged.bed'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">        --snp_multiple_enhancer_strategy </span><span><span leaf="">'max_mkscore'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">        --gene_window_enhancer_priority </span><span><span leaf="">'gene_window_first'</span></span><span leaf=""><br></span><span><span leaf="">done</span></span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">这一步会产生很多结果，咱不管那些乱七八糟的，重点看这两个目录：</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">SNP_gene_pair/</span><span leaf=""><br></span><span leaf="">├── SNP_gene_pair_chr10.feather</span><span leaf=""><br></span><span leaf="">├── ...</span><span leaf=""><br></span><span leaf="">└── SNP_gene_pair_chr9.feather</span><span leaf=""><br></span><span leaf="">w_ld/</span><span leaf=""><br></span><span leaf="">├── weights.10.l2.ldscore.gz</span><span leaf=""><br></span><span leaf="">├── ...</span><span leaf=""><br></span><span leaf="">└── weights.9.l2.ldscore.gz</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><code><span leaf="">SNP_gene_pair</span></code><span leaf="">储存的就是SNP和基因的映射信息。我们先读取chr1那个文件看看</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">import</span></span><span leaf=""> pyarrow.feather </span><span><span leaf="">as</span></span><span leaf=""> feather</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">snp_gene = feather.read_feather(</span><span><span leaf="">"step/E16.5_E1S1.MOSTA/generate_ldscore/SNP_gene_pair/SNP_gene_pair_chr1.feather"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">snp_gene.iloc[</span><span><span leaf="">860</span></span><span leaf="">:</span><span><span leaf="">865</span></span><span leaf="">, :]</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf=""> SNP gene_name CHR BP CM</span><span leaf=""><br></span><span leaf="">860 rs145881258 SAMD11 1 876067 0.620827</span><span leaf=""><br></span><span leaf="">861 rs71628924 SAMD11 1 876200 0.620827</span><span leaf=""><br></span><span leaf="">862 rs4372192 SAMD11 1 876499 0.620827</span><span leaf=""><br></span><span leaf="">863 rs72902600 SAMD11 1 876949 0.620827</span><span leaf=""><br></span><span leaf="">864 rs114982608 NOC2L 1 877147 0.620827</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">在这个表格中，如果SNP没有映射的基因，则gene_name为None；BP表示SNP在染色体上的物理位置，单位为bp；CM表示SNP在染色体上的遗传位置，单位为厘摩。</span><span leaf=""><br></span><span leaf="">这一步除了建立SNP和基因的映射外，还会计算SNP的连锁不平衡（LD）分数，它的形式为：</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">zcat weights.1.l2.ldscore.gz | head</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">CHR SNP BP CM L2</span><span leaf=""><br></span><span leaf="">1 rs3131972 752721 0.4888678 4.282165</span><span leaf=""><br></span><span leaf="">1 rs3131969 754182 0.48973366 4.4342103</span><span leaf=""><br></span><span leaf="">1 rs3131967 754334 0.48979417 4.4342103</span><span leaf=""><br></span><span leaf="">1 rs1048488 760912 0.49250725 4.3271613</span><span leaf=""><br></span><span leaf="">1 rs12562034 768448 0.49571378 1.2600546</span><span leaf=""><br></span><span leaf="">1 rs12124819 776546 0.49944228 2.0366306</span><span leaf=""><br></span><span leaf="">1 rs4040617 779322 0.50070773 4.124312</span><span leaf=""><br></span><span leaf="">1 rs2905036 792480 0.53674447 2.9040823</span><span leaf=""><br></span><span leaf="">1 rs4970383 838555 0.62082653 3.3426797</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">其中L2就是SNP的LD分数，它越高说明SNP跟临近SNP的连锁不平衡程度越高。然而gsMap的参考数据中提供了人的各SNP的LD分数，所以后续分析一般不用这里算出来的LD分数。</span></p><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">空间LDSC分析</span></span><span></span></h2><p data-tool="mdnice编辑器"><span leaf="">得到了基因和SNP的映射关系后，下一步就是整合GWAS数据进行LDSC分析了。这里没啥参数特别需要解释的，就是</span><code><span leaf="">w_file</span></code><span leaf="">后面接的是LD分数文件的前缀。</span><span leaf=""><br></span><strong><span leaf="">这一步大约需要40G内存</span></strong></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">gsmap run_spatial_ldsc \</span><span leaf=""><br></span><span leaf="">        --workdir </span><span><span leaf="">"."</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">        --sample_name </span><span><span leaf="">'E16.5_E1S1.MOSTA'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">        --trait_name </span><span><span leaf="">'IQ'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">        --sumstats_file </span><span><span leaf="">'../gsMap_example_data/GWAS/IQ_NG_2018.sumstats.gz'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">        --w_file </span><span><span leaf="">'../gsMap_resource/LDSC_resource/weights_hm3_no_hla/weights.'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">        --num_process 10</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">运行结果就是一个csv表格，存储着每个spot的LDSC回归的</span><span data-formula="\beta"><span data-formula="\beta"><svg xmlns="http://www.w3.org/2000/svg" role="img" focusable="false" viewbox="0 -705 566 899" aria-hidden="true"><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="matrix(1 0 0 -1 0 0)"><g data-mml-node="math"><g data-mml-node="mi"><path data-c="3B2" d="M29 -194Q23 -188 23 -186Q23 -183 102 134T186 465Q208 533 243 584T309 658Q365 705 429 705H431Q493 705 533 667T573 570Q573 465 469 396L482 383Q533 332 533 252Q533 139 448 65T257 -10Q227 -10 203 -2T165 17T143 40T131 59T126 65L62 -188Q60 -194 42 -194H29ZM353 431Q392 431 427 419L432 422Q436 426 439 429T449 439T461 453T472 471T484 495T493 524T501 560Q503 569 503 593Q503 611 502 616Q487 667 426 667Q384 667 347 643T286 582T247 514T224 455Q219 439 186 308T152 168Q151 163 151 147Q151 99 173 68Q204 26 260 26Q302 26 349 51T425 137Q441 171 449 214T457 279Q457 337 422 372Q380 358 347 358H337Q258 358 258 389Q258 396 261 403Q275 431 353 431Z"></path></g></g></g><g></g></svg></span></span><span leaf="">值以及block-jackknife方法计算的标准误、z统计量和p值。这个表格可以进一步导入Seurat/AnnData对象中进行分析和可视化。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">zcat E16.5_E1S1.MOSTA_IQ.csv.gz | head</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">spot,beta,se,z,p</span><span leaf=""><br></span><span leaf="">190_194,-8.760567138111535e-09,2.3904128827941287e-09,-3.664876139669813,0.9998762708438821</span><span leaf=""><br></span><span leaf="">190_195,-1.017986287894897e-08,1.8104173514659798e-09,-5.622937092768061,0.999999990613111</span><span leaf=""><br></span><span leaf="">190_196,-9.676589019076809e-09,1.7393880753927583e-09,-5.563214532726867,0.999999986757502</span><span leaf=""><br></span><span leaf="">190_197,-8.654820922724944e-09,1.7101907418698534e-09,-5.060734285850543,0.9999997911775429</span><span leaf=""><br></span><span leaf="">190_198,-9.285564032846918e-09,1.6392959203110825e-09,-5.664361094173183,0.9999999926213325</span><span leaf=""><br></span><span leaf="">190_199,-8.68369809825401e-09,1.706017320378527e-09,-5.090040994617389,0.9999998210069435</span><span leaf=""><br></span><span leaf="">190_200,-7.79527285731608e-09,1.6845775390074085e-09,-4.627434877179493,0.9999981488870429</span><span leaf=""><br></span><span leaf="">190_201,-8.744531510222193e-09,1.7063147137160712e-09,-5.124805781682589,0.9999998510776975</span><span leaf=""><br></span><span leaf="">190_202,-7.892295642543661e-09,1.7265129666357858e-09,-4.57123450275747,0.9999975757034143</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">至此，我们已经得到了每个spot与表型关联的显著性p值，gsMap分析的主体部分也已经完成了。后面两部用户可以根据自己的需求选择性运行。</span></p><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">柯西合并检验（可选）</span></span><span></span></h2><p data-tool="mdnice编辑器"><span leaf="">柯西合并检验就是把spot的p值按组织或细胞类型进行合并。使用下面代码完成。</span><span leaf=""><br></span><strong><span leaf="">这一步大约需要12G的内存</span></strong></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">gsmap run_cauchy_combination \</span><span leaf=""><br></span><span leaf="">        --workdir </span><span><span leaf="">'.'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">        --sample_name </span><span><span leaf="">'E16.5_E1S1.MOSTA'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">        --trait_name </span><span><span leaf="">'IQ'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">        --annotation </span><span><span leaf="">'annotation'</span></span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">运行结果是一个csv表格，记录着柯西合并检验的p值以及简单取中位数的p值。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">zcat E16.5_E1S1.MOSTA_IQ.Cauchy.csv.gz | head</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">annotation,p_cauchy,p_median</span><span leaf=""><br></span><span leaf="">Brain,1.7822683407008737e-26,9.443479488838352e-16</span><span leaf=""><br></span><span leaf="">Spinal cord,1.2455551008640988e-20,5.945566346850132e-13</span><span leaf=""><br></span><span leaf="">Dorsal root ganglion,6.3162808316974406e-12,3.984616094128103e-09</span><span leaf=""><br></span><span leaf="">Choroid plexus,1.9943233731112286e-09,4.902085774857428e-06</span><span leaf=""><br></span><span leaf="">Sympathetic nerve,8.755048019892797e-09,6.910435592108675e-07</span><span leaf=""><br></span><span leaf="">Mucosal epithelium,5.986670119750848e-06,0.0067753931043148</span><span leaf=""><br></span><span leaf="">Meninges,9.19767922591408e-06,0.1482752689335825</span><span leaf=""><br></span><span leaf="">Muscle,1.2536003082930236e-05,0.0059370242200209</span><span leaf=""><br></span><span leaf="">Cartilage primordium,2.195056642739779e-05,0.0325189969669502</span><span leaf=""><br></span></code></pre><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">报告生成（可选）</span></span><span></span></h2><p data-tool="mdnice编辑器"><span leaf="">在完成上述过程后，如果用户想生成一份像快速模式那种报告，可以用下面代码生成：</span><span leaf=""><br></span><strong><span leaf="">这一步需要约60G内存</span></strong></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">gsmap run_report \</span><span leaf=""><br></span><span leaf="">        --workdir </span><span><span leaf="">'.'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">        --sample_name </span><span><span leaf="">'E16.5_E1S1.MOSTA'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">        --trait_name </span><span><span leaf="">'IQ'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">        --annotation </span><span><span leaf="">'annotation'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">        --sumstats_file </span><span><span leaf="">'../gsMap_example_data/GWAS/IQ_NG_2018.sumstats.gz'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">        --top_corr_genes 50</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><code><span leaf="">top_corr_genes</span></code><span leaf="">控制的是报告中展示的GSS与-log10(p_value)相关性最高的50个基因；如果要展示感兴趣的基因，可以用</span><code><span leaf="">selected_genes</span></code><span leaf="">指定。</span></p><h1 data-tool="mdnice编辑器"><span></span><span><span leaf="">高级用法</span></span><span></span></h1><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">自定义潜在嵌入</span></span><span></span></h2><p data-tool="mdnice编辑器"><span leaf="">小编在前面也提到过，潜在嵌入不一定需要GVAE学习出来的，可以通过任何降维算法计算得到。这种情况下，只需要把嵌入矩阵作为AnnData对象的obsm域，其中每一行为单个spot的嵌入向量就行。然后在</span><code><span leaf="">gsmap run_latent_to_gene</span></code><span leaf="">的</span><code><span leaf="">latent_representation</span></code><span leaf="">参数中指定嵌入矩阵的名称就好：</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># This could be any key in the obsm field of the AnnData object.</span></span><span leaf=""><br></span><span leaf="">latent_customized = </span><span><span leaf="">'latent_customized'</span></span><span leaf=""><br></span><span leaf="">gsmap run_latent_to_gene \</span><span leaf=""><br></span><span leaf="">    --input_hdf5_path </span><span><span leaf="">'sample1.h5ad'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">    --workdir </span><span><span leaf="">'./workdir'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">    --sample_name </span><span><span leaf="">'sample1'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">    --annotation </span><span><span leaf="">'annotation'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">    --latent_representation </span><span><span leaf="">$latent_customized</span></span><span leaf=""><br></span></code></pre><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">条件分析</span></span><span></span></h2><p data-tool="mdnice编辑器"><span leaf="">在生成LD分数步骤中，也可以加入一些基线指标用作LDSC分析的对照。人的基线参考数据可以下载为</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">wget https://yanglab.westlake.edu.cn/data/gsMap/gsMap_additional_annotation.tar.gz</span><span leaf=""><br></span><span leaf="">tar -xvzf gsMap_additional_annotation.tar.gz</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">在gsMap_additional_annotation目录中存放着每条染色体的基线注释文件，以1号染色体为例</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">zcat baseline.1.annot.gz | head | cut -f 1-7</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">CHR BP SNP CM Coding_UCSC Coding_UCSC.extend.500 Conserved_LindbladToh</span><span leaf=""><br></span><span leaf="">1 11008 rs575272151 0.0 0 0 0</span><span leaf=""><br></span><span leaf="">1 11012 rs544419019 0.0 0 0 0</span><span leaf=""><br></span><span leaf="">1 13110 rs540538026 0.0 0 0 0</span><span leaf=""><br></span><span leaf="">1 13116 rs62635286 0.0 0 0 0</span><span leaf=""><br></span><span leaf="">1 13118 rs200579949 0.0 0 0 0</span><span leaf=""><br></span><span leaf="">1 13273 rs531730856 0.0 0 0 0</span><span leaf=""><br></span><span leaf="">1 13550 rs554008981 0.0 0 0 0</span><span leaf=""><br></span><span leaf="">1 14464 rs546169444 0.0 0 0 0</span><span leaf=""><br></span><span leaf="">1 14599 rs531646671 0.0 0 0 0</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">从第5列开始，后面每一列都记录着SNP所在位置的属性，如是否是编码区、promoter、intron、5'UTR等。这些属性往往是跟表型无关的，所以可以作为LDSC分析基线参考变量。</span><span leaf=""><br></span><span leaf="">要加上这些基线变量，只需在生成LD分数步骤将</span><code><span leaf="">additional_baseline_annotation</span></code><span leaf="">参数设为基线注释文件夹名称即可。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">for</span></span><span leaf=""> CHROM </span><span><span leaf="">in</span></span><span leaf=""> {1..22}</span><span leaf=""><br></span><span><span leaf="">do</span></span><span leaf=""><br></span><span leaf="">        gsmap run_generate_ldscore \</span><span leaf=""><br></span><span leaf="">                --workdir </span><span><span leaf="">'.'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">                --sample_name </span><span><span leaf="">'E16.5_E1S1.MOSTA'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">                --chrom </span><span><span leaf="">$CHROM</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">                --bfile_root </span><span><span leaf="">'../gsMap_resource/LD_Reference_Panel/1000G_EUR_Phase3_plink/1000G.EUR.QC'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">                --keep_snp_root </span><span><span leaf="">'../gsMap_resource/LDSC_resource/hapmap3_snps/hm'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">                --gtf_annotation_file </span><span><span leaf="">'../gsMap_resource/genome_annotation/gtf/gencode.v46lift37.basic.annotation.gtf'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">                --gene_window_size 50000 \</span><span leaf=""><br></span><span leaf="">                --additional_baseline_annotation </span><span><span leaf="">'gsMap_additional_annotation'</span></span><span leaf=""><br></span><span><span leaf="">done</span></span><span leaf=""><br></span></code></pre><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">生物学重复</span></span><span></span></h2><p data-tool="mdnice编辑器"><span leaf="">在某些情况下，我们会做多个空转切片作为生物学重复。gsMap对生物学重复的整合主要在以下几个方面。</span></p><h3 data-tool="mdnice编辑器"><span></span><span><span leaf="">计算切片均值</span></span><span></span></h3><p data-tool="mdnice编辑器"><span leaf="">gsMap可以计算每个基因在所有切片中的基因表达排序的均值。这些均值又可作为参照用于每个样本的GSS的计算，从而可以增加样本间数据的一致性和可比性。通过以下代码进行计算：</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">gsmap create_slice_mean \</span><span leaf=""><br></span><span leaf="">        --sample_name_list </span><span><span leaf="">'E16.5_E1S1.MOSTA'</span></span><span leaf=""> </span><span><span leaf="">'E16.5_E2S11.MOSTA'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">        --h5ad_list </span><span><span leaf="">'../gsMap_example_data/ST/E16.5_E1S1.MOSTA.h5ad'</span></span><span leaf=""> </span><span><span leaf="">'../gsMap_example_data/ST/E16.5_E2S11.MOSTA.h5ad'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">        --slice_mean_output_file </span><span><span leaf="">'sample_slice_mean.parquet'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">        --data_layer </span><span><span leaf="">'count'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">        --homolog_file </span><span><span leaf="">'../gsMap_resource/homologs/mouse_human_homologs.txt'</span></span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">计算得到了一个parquet文件，它也是一种存储数据框的文件格式，可以通过pandas读取</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">import</span></span><span leaf=""> pandas </span><span><span leaf="">as</span></span><span leaf=""> pd</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">df = pd.read_parquet(</span><span><span leaf="">"replicate/sample_slice_mean.parquet"</span></span><span leaf="">, engine=</span><span><span leaf="">"pyarrow"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">df.head()</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf=""> G_Mean frac</span><span leaf=""><br></span><span leaf="">gene  </span><span leaf=""><br></span><span leaf="">0610010F05Rik 8014.482320 0.095213</span><span leaf=""><br></span><span leaf="">0610010K14Rik 8760.099930 0.220422</span><span leaf=""><br></span><span leaf="">0610012G03Rik 9218.004314 0.286876</span><span leaf=""><br></span><span leaf="">0610030E20Rik 7728.556558 0.050133</span><span leaf=""><br></span><span leaf="">0610040J01Rik 7518.777238 0.016676</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">可以发现，这个数据框存储着每个基因的平均表达排名和表达比例。</span></p><h3 data-tool="mdnice编辑器"><span></span><span><span leaf="">切片均值用于快速模式</span></span><span></span></h3><p data-tool="mdnice编辑器"><span leaf="">使用多样本均值修正快速模式运行，只需在</span><code><span leaf="">gM_slices</span></code><span leaf="">参数指定parquet文件即可。注意这里只以其中一个样本为例，另一个样本也是相同的方式进行修正。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">gsmap quick_mode \</span><span leaf=""><br></span><span leaf="">        --workdir </span><span><span leaf="">'.'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">        --homolog_file </span><span><span leaf="">'../gsMap_resource/homologs/mouse_human_homologs.txt'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">        --sample_name </span><span><span leaf="">'E16.5_E1S1.MOSTA'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">        --gsMap_resource_dir </span><span><span leaf="">'../gsMap_resource'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">        --hdf5_path </span><span><span leaf="">'../gsMap_example_data/ST/E16.5_E1S1.MOSTA.h5ad'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">        --annotation </span><span><span leaf="">'annotation'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">        --data_layer </span><span><span leaf="">'count'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">        --sumstats_file </span><span><span leaf="">'../gsMap_example_data/GWAS/IQ_NG_2018.sumstats.gz'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">        --trait_name </span><span><span leaf="">'IQ'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">        --gM_slices </span><span><span leaf="">'sample_slice_mean.parquet'</span></span><span leaf=""><br></span></code></pre><h3 data-tool="mdnice编辑器"><span></span><span><span leaf="">切片均值用于分步模式</span></span><span></span></h3><p data-tool="mdnice编辑器"><span leaf="">在分步模式中，</span><code><span leaf="">gM_slice</span></code><span leaf="">参数在计算GSS那一步指定。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">gsmap run_latent_to_gene \</span><span leaf=""><br></span><span leaf="">    --workdir </span><span><span leaf="">'.'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">    --sample_name </span><span><span leaf="">'E16.5_E1S1.MOSTA'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">    --annotation </span><span><span leaf="">'annotation'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">    --latent_representation </span><span><span leaf="">'latent_GVAE'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">    --num_neighbour 51 \</span><span leaf=""><br></span><span leaf="">    --num_neighbour_spatial 201 \</span><span leaf=""><br></span><span leaf="">    --homolog_file </span><span><span leaf="">'../gsMap_resource/homologs/mouse_human_homologs.txt'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">    --gM_slices </span><span><span leaf="">'sample_slice_mean.parquet'</span></span><span leaf=""><br></span></code></pre><h3 data-tool="mdnice编辑器"><span></span><span><span leaf="">多样本柯西合并检验</span></span><span></span></h3><p data-tool="mdnice编辑器"><span leaf="">gsMap还能对多样本中的相同细胞或组织类型进行柯西合并检验，只需在</span><code><span leaf="">sample_name_list</span></code><span leaf="">参数指定多个样本的输出目录，并且在</span><code><span leaf="">output_file</span></code><span leaf="">参数额外指定输出文件路径即可。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">gsmap run_cauchy_combination \</span><span leaf=""><br></span><span leaf="">    --workdir </span><span><span leaf="">'.'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">    --sample_name_list </span><span><span leaf="">'E16.5_E1S1.MOSTA'</span></span><span leaf=""> </span><span><span leaf="">'E16.5_E2S11.MOSTA'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">    --trait_name </span><span><span leaf="">'IQ'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">    --annotation </span><span><span leaf="">'annotation'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">    --output_file </span><span><span leaf="">'combined_IQ_cauchy_combination.csv.gz'</span></span><span leaf=""><br></span></code></pre><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">10X Visium数据实例</span></span><span></span></h2><p data-tool="mdnice编辑器"><span leaf="">前面的教程都是用Stereo-seq的示例数据，然而除了Stereo-seq，10X Visium也是一种广泛使用的空间组学技术。虽然Visium没有单细胞分辨率，但是也能使用gsMap在组织或者空间亚群层面进行分析。</span><span leaf=""><br></span><span leaf="">首先下载Visium的示例数据</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">wget https://yanglab.westlake.edu.cn/data/gsMap/gsMap_resource.tar.gz</span><span leaf=""><br></span><span leaf="">tar -xvzf gsMap_resource.tar.gz</span><span leaf=""><br></span><span leaf="">tree Visium_example_data</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">Visium_example_data</span><span leaf=""><br></span><span leaf="">├── GWAS</span><span leaf=""><br></span><span leaf="">│   ├── IQ_NG_2018.sumstats.gz</span><span leaf=""><br></span><span leaf="">│   └── Serum_creatinine.sumstats.gz</span><span leaf=""><br></span><span leaf="">└── ST</span><span leaf=""><br></span><span leaf="">    ├── V1_Adult_Mouse_Brain_Coronal_Section.h5ad</span><span leaf=""><br></span><span leaf="">    ├── V1_Mouse_Brain_Sagittal_Posterior_Section.h5ad</span><span leaf=""><br></span><span leaf="">    └── V1_Mouse_Kidney.h5ad</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">可以发现，示例数据中GWAS数据包含有IQ和血清肌酐这两种表型。我们重点关注空转数据，包含成年小鼠冠状切片、成年小鼠矢状切片以及成年小鼠肾脏。这里我们读取成年小鼠冠状切片的样本看看</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">import</span></span><span leaf=""> scanpy </span><span><span leaf="">as</span></span><span leaf=""> sc</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">corocal_brain = sc.read_h5ad(</span><span><span leaf="">"visium/Visium_example_data/ST/V1_Adult_Mouse_Brain_Coronal_Section.h5ad"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">corocal_brain</span><span leaf=""><br></span><span leaf="">corocal_brain.obs.head()</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">AnnData object with n_obs × n_vars = 2902 × 32285</span><span leaf=""><br></span><span leaf="">    obs: </span><span><span leaf="">'in_tissue'</span></span><span leaf="">, </span><span><span leaf="">'array_row'</span></span><span leaf="">, </span><span><span leaf="">'array_col'</span></span><span leaf="">, </span><span><span leaf="">'domain'</span></span><span leaf=""><br></span><span leaf="">    var: </span><span><span leaf="">'gene_ids'</span></span><span leaf="">, </span><span><span leaf="">'feature_types'</span></span><span leaf="">, </span><span><span leaf="">'genome'</span></span><span leaf=""><br></span><span leaf="">    uns: </span><span><span leaf="">'spatial'</span></span><span leaf=""><br></span><span leaf="">    obsm: </span><span><span leaf="">'spatial'</span></span><span leaf=""><br></span><span leaf="">    layers: </span><span><span leaf="">'count'</span></span><span leaf=""><br></span><span leaf=""> in_tissue array_row array_col domain</span><span leaf=""><br></span><span leaf="">AAACAAGTATCTCCCA-1 1 50 102 4</span><span leaf=""><br></span><span leaf="">AAACAGAGCGACTCCT-1 1 14 94 2</span><span leaf=""><br></span><span leaf="">AAACAGTGTTCCTGGG-1 1 73 43 1</span><span leaf=""><br></span><span leaf="">AAACATTTCCCGGATT-1 1 61 97 5</span><span leaf=""><br></span><span leaf="">AAACCGGGTAGGTACC-1 1 42 28 3</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">可以发现，10X Visium的AnnData格式与Stereo-seq的基本类似。区别就在于10X Visium无法注释spot为具体的细胞类型，只有通过聚类划分的domain。</span><span leaf=""><br></span><span leaf="">Visium数据的gsMap分析方法跟Stereo-seq是完全一样的。为了省事，小编直接把3个样本的gsMap分析写成了一个脚本统一运行。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">gsmap quick_mode \</span><span leaf=""><br></span><span leaf="">   --workdir </span><span><span leaf="">'.'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">   --homolog_file </span><span><span leaf="">'../gsMap_resource/homologs/mouse_human_homologs.txt'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">   --sample_name </span><span><span leaf="">'V1_Adult_Mouse_Brain_Coronal_Section'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">   --gsMap_resource_dir </span><span><span leaf="">'../gsMap_resource'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">   --hdf5_path </span><span><span leaf="">'Visium_example_data/ST/V1_Adult_Mouse_Brain_Coronal_Section.h5ad'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">   --annotation </span><span><span leaf="">'domain'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">   --data_layer </span><span><span leaf="">'count'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">   --sumstats_file </span><span><span leaf="">'Visium_example_data/GWAS/IQ_NG_2018.sumstats.gz'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">   --trait_name </span><span><span leaf="">'IQ'</span></span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">gsmap quick_mode \</span><span leaf=""><br></span><span leaf="">   --workdir </span><span><span leaf="">'.'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">   --homolog_file </span><span><span leaf="">'../gsMap_resource/homologs/mouse_human_homologs.txt'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">   --sample_name </span><span><span leaf="">'V1_Mouse_Brain_Sagittal_Posterior_Section'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">   --gsMap_resource_dir </span><span><span leaf="">'../gsMap_resource'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">   --hdf5_path </span><span><span leaf="">'Visium_example_data/ST/V1_Mouse_Brain_Sagittal_Posterior_Section.h5ad'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">   --annotation </span><span><span leaf="">'domain'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">   --data_layer </span><span><span leaf="">'count'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">   --sumstats_file </span><span><span leaf="">'Visium_example_data/GWAS/IQ_NG_2018.sumstats.gz'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">   --trait_name </span><span><span leaf="">'IQ'</span></span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">gsmap quick_mode \</span><span leaf=""><br></span><span leaf="">   --workdir </span><span><span leaf="">'.'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">   --homolog_file </span><span><span leaf="">'../gsMap_resource/homologs/mouse_human_homologs.txt'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">   --sample_name </span><span><span leaf="">'V1_Mouse_Kidney'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">   --gsMap_resource_dir </span><span><span leaf="">'../gsMap_resource'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">   --hdf5_path </span><span><span leaf="">'Visium_example_data/ST/V1_Mouse_Kidney.h5ad'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">   --annotation </span><span><span leaf="">'domain'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">   --data_layer </span><span><span leaf="">'count'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">   --sumstats_file </span><span><span leaf="">'Visium_example_data/GWAS/Serum_creatinine.sumstats.gz'</span></span><span leaf=""> \</span><span leaf=""><br></span><span leaf="">   --trait_name </span><span><span leaf="">'Serum_creatinine'</span></span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">跑完以后，我们看看成年小鼠冠状切片的分析结果报告。<img data-src="https://mmbiz.qpic.cn/mmbiz_png/GL6g5Y3aR7calxDyI9mY9DEr36eCTibCY5kibj6KA1XI5My6eLnKIibj4JjkrLUn6n6hOX8MMjBuqCfGPBEJ5qIyA/640?wx_fmt=png&amp;from=appmsg" data-ratio="1.124074074074074" data-type="png" data-w="1080" data-imgfileid="100012398" src="https://mmbiz.qpic.cn/mmbiz_png/GL6g5Y3aR7calxDyI9mY9DEr36eCTibCY5kibj6KA1XI5My6eLnKIibj4JjkrLUn6n6hOX8MMjBuqCfGPBEJ5qIyA/640?wx_fmt=png&amp;from=appmsg">可以发现，Visium数据报告的可视化效果不太好，spot过大，并且也没显示H&amp;E染色的切片，所以最好还是将分析结果导入Scanpy中进行可视化。我们首先可视化一下p值。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">import</span></span><span leaf=""> pandas </span><span><span leaf="">as</span></span><span leaf=""> pd</span><span leaf=""><br></span><span><span leaf="">import</span></span><span leaf=""> numpy </span><span><span leaf="">as</span></span><span leaf=""> np</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">p_value_res = pd.read_csv(</span><span><span leaf="">"visium/V1_Adult_Mouse_Brain_Coronal_Section/spatial_ldsc/V1_Adult_Mouse_Brain_Coronal_Section_IQ.csv.gz"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">p_value_res.index = p_value_res[</span><span><span leaf="">'spot'</span></span><span leaf="">].values</span><span leaf=""><br></span><span leaf="">corocal_brain.obs = corocal_brain.obs.merge(</span><span leaf=""><br></span><span leaf="">    right=p_value_res,</span><span leaf=""><br></span><span leaf="">    how=</span><span><span leaf="">'left'</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">    left_index=</span><span><span leaf="">True</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">    right_index=</span><span><span leaf="">True</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">corocal_brain.obs[</span><span><span leaf="">'-log10(p)'</span></span><span leaf="">] = </span><span><span leaf="">-1</span></span><span leaf=""> * corocal_brain.obs[</span><span><span leaf="">'p'</span></span><span leaf="">].apply(</span><span><span leaf="">lambda</span></span><span leaf=""> x: </span><span><span leaf="">0</span></span><span leaf=""> </span><span><span leaf="">if</span></span><span leaf=""> x == </span><span><span leaf="">0</span></span><span leaf=""> </span><span><span leaf="">else</span></span><span leaf=""> np.log10(x))</span><span leaf=""><br></span><span leaf="">sc.pl.spatial(corocal_brain, color=</span><span><span leaf="">'-log10(p)'</span></span><span leaf="">, size=</span><span><span leaf="">1</span></span><span leaf="">, frameon=</span><span><span leaf="">False</span></span><span leaf="">, img_key=</span><span><span leaf="">'hires'</span></span><span leaf="">, cmap=</span><span><span leaf="">'magma'</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/GL6g5Y3aR7calxDyI9mY9DEr36eCTibCYR8Lr9c3c95mpTbia1iakyfg5EEX6YmaaKNLOdQ8JhbWGKpfvyFl3Y6uA/640?wx_fmt=png&amp;from=appmsg" data-ratio="1.312849162011173" data-type="png" data-w="716" data-imgfileid="100012397" src="https://mmbiz.qpic.cn/mmbiz_png/GL6g5Y3aR7calxDyI9mY9DEr36eCTibCYR8Lr9c3c95mpTbia1iakyfg5EEX6YmaaKNLOdQ8JhbWGKpfvyFl3Y6uA/640?wx_fmt=png&amp;from=appmsg">可以发现这个样本是以荧光成像为底片做的空转，可视化清楚多了。</span><span leaf=""><br></span><span leaf="">报告里还有针对基因的GSS的可视化展示，那我们也在Scanpy里实现一下叭。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">from</span></span><span leaf=""> pyarrow </span><span><span leaf="">import</span></span><span leaf=""> feather</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">gss = feather.read_feather(</span><span><span leaf="">"visium/V1_Adult_Mouse_Brain_Coronal_Section/latent_to_gene/V1_Adult_Mouse_Brain_Coronal_Section_gene_marker_score.feather"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">gss.index = gss[</span><span><span leaf="">'HUMAN_GENE_SYM'</span></span><span leaf="">].values</span><span leaf=""><br></span><span leaf="">gss = gss.drop(columns=[</span><span><span leaf="">'HUMAN_GENE_SYM'</span></span><span leaf="">])</span><span leaf=""><br></span><span leaf="">gss = gss.T</span><span leaf=""><br></span><span leaf="">corocal_brain.obs = corocal_brain.obs.merge(</span><span leaf=""><br></span><span leaf="">    right=gss,</span><span leaf=""><br></span><span leaf="">    how=</span><span><span leaf="">'left'</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">    left_index=</span><span><span leaf="">True</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">    right_index=</span><span><span leaf="">True</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">sc.pl.spatial(corocal_brain, color=</span><span><span leaf="">'EPHX4'</span></span><span leaf="">, size=</span><span><span leaf="">1</span></span><span leaf="">, frameon=</span><span><span leaf="">False</span></span><span leaf="">, img_key=</span><span><span leaf="">'hires'</span></span><span leaf="">, cmap=</span><span><span leaf="">'magma'</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/GL6g5Y3aR7calxDyI9mY9DEr36eCTibCYOxa7QFQPNPHic07KPjUSJvJ1xMdV5OjvodYzVM8IVZQWZZpSMyd1BibA/640?wx_fmt=png&amp;from=appmsg" data-ratio="1.3183730715287518" data-type="png" data-w="713" data-imgfileid="100012399" src="https://mmbiz.qpic.cn/mmbiz_png/GL6g5Y3aR7calxDyI9mY9DEr36eCTibCYOxa7QFQPNPHic07KPjUSJvJ1xMdV5OjvodYzVM8IVZQWZZpSMyd1BibA/640?wx_fmt=png&amp;from=appmsg">注意这里我把</span><code><span leaf="">gss</span></code><span leaf="">矩阵放在</span><code><span leaf="">obs</span></code><span leaf="">中是因为空转是鼠的数据，但是计算出的</span><code><span leaf="">gss</span></code><span leaf="">矩阵是人的基因。如果空转数据也是人的话需要把</span><code><span leaf="">gss</span></code><span leaf="">矩阵放在</span><code><span leaf="">layer</span></code><span leaf="">中，需要可视化的时候直接将</span><code><span leaf="">gss</span></code><span leaf="">矩阵作为主矩阵</span><code><span leaf="">X</span></code><span leaf="">。</span><span leaf=""><br></span><span leaf="">gsMap的</span><span leaf="">软件教程就</span><span leaf="">到这里啦，小伙伴们下次见～</span></p><p data-tool="mdnice编辑器"><span leaf=""><br></span></p><h1 data-tool="mdnice编辑器"><span></span><span leaf="">GitHub链接：https://github.com/JianYang-Lab/gsMap</span></h1></section><p data-pm-slice="2 1 []"><span leaf=""><br></span></p><section nodeleaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_jpg/GL6g5Y3aR7cwqaibulibXfhkQdkqSW0H9M5riaAVfSBb3ibibCZHnqViahn3dyukTiaJU0sKzC2xXD8U2Pa39ra5dz9xg/640?wx_fmt=jpeg&amp;from=appmsg" data-ratio="1" data-s="300,640" data-type="jpeg" data-w="430" data-croporisrc="https://mmbiz.qpic.cn/mmbiz_jpg/GL6g5Y3aR7cwqaibulibXfhkQdkqSW0H9M5riaAVfSBb3ibibCZHnqViahn3dyukTiaJU0sKzC2xXD8U2Pa39ra5dz9xg/0?wx_fmt=jpeg&amp;from=appmsg" data-cropselx2="203" data-cropsely2="203" data-imgfileid="100010845" src="https://mmbiz.qpic.cn/mmbiz_jpg/GL6g5Y3aR7cwqaibulibXfhkQdkqSW0H9M5riaAVfSBb3ibibCZHnqViahn3dyukTiaJU0sKzC2xXD8U2Pa39ra5dz9xg/640?wx_fmt=jpeg&amp;from=appmsg"></section><h3 data-tool="mdnice编辑器"><span><span leaf="">关注公众号，下回更新不迷路</span></span></h3><section><span leaf=""><br></span></section><section nodeleaf=""><mp-common-profile data-pluginname="mpprofile" data-nickname="生信宝库" data-alias="sxbk2020" data-from="2" data-headimg="http://mmbiz.qpic.cn/mmbiz_png/GL6g5Y3aR7f0yanILMQZCnw1duTMROQRvgDqVjlYcrljTRy1E4ZLppLG6zicdd3h0IwjLpxnum1V5KsowibJM1sw/0?wx_fmt=png" data-signature="本公众号只用于生信知识的收集与传播，以及生信人之间互相交流和学习，不会涉及任何商业利益。本公众号各小编平时忙于科研，更新文章较其它同类型公众号较慢，但保持宁缺毋滥的本心，只更新对大家有用的推文。" data-id="MzI4MjY5ODI1Nw==" data-is_biz_ban="0"></mp-common-profile></section><section><span leaf=""><br></span></section><p><mp-style-type data-value="3"></mp-style-type></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/LwFv0cy9LbqQdSuLevnZ8w",target="_blank" rel="noopener noreferrer">原文链接</a>
