---
title: "NC高分杂志cellchat细胞通讯结果个性化绘制之弦图"
date: 2025-09-06T13:15:38Z
draft: ["false"]
tags: [
  "fetched",
  "生信技能树"
]
categories: ["Acdemic"]
---
NC高分杂志cellchat细胞通讯结果个性化绘制之弦图 by 生信技能树
------
<div><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com" data-pm-slice="0 0 []"><p data-tool="mdnice编辑器"><span leaf="">前面，我们尝试了对单细胞或者空间转录组细胞通讯结果进行可视化，相关的准备的帖子有：</span></p><ul><li><section><span leaf=""><a target="_blank" href="https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247542658&amp;idx=1&amp;sn=adabcd404adc4cd33d3991909888ec76&amp;scene=21#wechat_redirect" textvalue="" linktype="text" data-linktype="2">cellchat细胞通讯绘制弦图函数的参数这么难搞定吗？</a></span></section></li><li><section><span leaf=""><a target="_blank" href="https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247544490&amp;idx=1&amp;sn=9a3e5918921f89126470514d7516094e&amp;scene=21#wechat_redirect" textvalue="" linktype="text" data-linktype="2">基本功修炼：Chord diagram 和弦图的基础函数</a></span></section></li></ul><p data-tool="mdnice编辑器"><span leaf="">今天就开干！学习的弦图个性化图来自文献<span textstyle="">《Microglia coordinate cellular interactions during spinal cord repair in mice》</span>，于<span textstyle="">2022年7月14号发表在 Nature Communications 杂志上</span>，弦图如下。</span></p><p data-tool="mdnice编辑器"><span leaf="">含义：</span><strong><span leaf="">弦图显示了假手术（sham）、7天（7 dpi）和28天（28 dpi）脊髓中，小胶质细胞（棕色）、MDMs（蓝色）和星形胶质细胞（橙色）之间的相互作用，这些相互作用来自载体（vehicle）处理的小鼠（a–c）和小胶质细胞耗竭的小鼠（d–f）。</span></strong></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxm6hAcT089C68fBxHacF0w39IDpmYNUUQgmj1jJ4BIcPMjwDJnQjbrdJjmOB061y1VWdnVTAqWicg/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.4425925925925926" data-type="png" data-w="1080" data-imgfileid="100061252" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxm6hAcT089C68fBxHacF0w39IDpmYNUUQgmj1jJ4BIcPMjwDJnQjbrdJjmOB061y1VWdnVTAqWicg/640?wx_fmt=png&amp;from=appmsg"></span></figure><h2 data-cacheurl="" data-remoteid="" data-tool="mdnice编辑器"><span></span><span><span leaf="">数据背景</span></span></h2><p data-tool="mdnice编辑器"><span leaf="">创伤性脊髓损伤（SCI）会引发一种以驻留组织的小胶质细胞microglia和单核细胞衍生的巨噬细胞（MDMs）为主导的神经炎症反应。<span textstyle="">由于激活的microglia和MDMs在体内具有相同的形态特征，并且表达相似的表型标记，因此历史上一直难以识别由microglia特异性协调的损伤反应。</span>在此，作者通过药物耗竭小microglia，并利用解剖学、组织病理学、神经纤维示踪、bulk RNA-seq和scRNA-seq揭示了由microglia控制的SCI的细胞和分子反应。</span></p><p data-tool="mdnice编辑器"><span leaf="">结果发现，microglia 对于SCI的恢复至关重要，并在中枢神经系统驻留的胶质细胞和浸润性白细胞中协调损伤反应。耗竭microglia会加剧组织损伤并恶化功能恢复。相反，通过测序数据识别出的microglia依赖性信号轴在microglia耗竭的小鼠中得以恢复，可以防止继发性损伤并促进恢复。进一步的生物信息学分析揭示，通过利用microglia、astrocytes和MDMs之间的关键配体-受体相互作用，可能会实现SCI后的最佳修复。</span></p><h3 data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><span></span><span><span leaf="">实验设计</span></span><span></span></h3><p data-tool="mdnice编辑器"><span leaf="">小鼠在手术前2周开始接受载体（vehicle）或PLX5622处理，持续至术后7天或28天。在终点时，通过解剖一段1厘米的新鲜组织（T4–T13），包括硬膜，以椎板切除术为中心，对新鲜（未灌注）小鼠脊髓进行单细胞RNA测序。</span></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxm6hAcT089C68fBxHacF0wnT1Q7POc8jicesYTYwIxs4dC5AKZYmyY7SRSu8hAxrELnZpBSia9Gcicw/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.3638888888888889" data-type="png" data-w="1080" data-imgfileid="100061253" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxm6hAcT089C68fBxHacF0wnT1Q7POc8jicesYTYwIxs4dC5AKZYmyY7SRSu8hAxrELnZpBSia9Gcicw/640?wx_fmt=png&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">top3亚群差异基因：</span></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxm6hAcT089C68fBxHacF0w5dFToTlO5yasNFXTxl7JwjzsBJ4y0erjzNEvaCYt2ID6oqRyy2KBgQ/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.32222222222222224" data-type="png" data-w="1080" data-imgfileid="100061250" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxm6hAcT089C68fBxHacF0w5dFToTlO5yasNFXTxl7JwjzsBJ4y0erjzNEvaCYt2ID6oqRyy2KBgQ/640?wx_fmt=png&amp;from=appmsg"></span></figure><h3 data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><span></span><span><span leaf="">样本如下：</span></span><span></span></h3><p data-tool="mdnice编辑器"><span leaf="">Bulk RNAseq of spinal cords：sham Vehicle (n=4), sham PLX5622 (n=4), 7dpi Vehicle (n=3), and 7dpi PLX5622 mice (n=3)</span></p><p data-tool="mdnice编辑器"><span leaf="">Single cell RNAseq of spinal cords：sham, 7 dpi and 28 dpi Vehicle mice and sham, 7 dpi and 28 dpi PLX5622 mice</span></p><p data-tool="mdnice编辑器"><span leaf="">GSE196928：https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE196928</span></p><pre data-tool="mdnice编辑器"><code><span leaf="">GSM5904811 A.1_S23</span><span leaf=""><br></span><span leaf="">GSM5904812 A.25_S2</span><span leaf=""><br></span><span leaf="">GSM5904813 A.27_S3</span><span leaf=""><br></span><span leaf="">GSM5904814 A.3_S1</span><span leaf=""><br></span><span leaf="">GSM5904815 B.2_S24</span><span leaf=""><br></span><span leaf="">GSM5904816 B.26_S25</span><span leaf=""><br></span><span leaf="">GSM5904817 B.4_S4</span><span leaf=""><br></span><span leaf="">GSM5904818 C.29_S6</span><span leaf=""><br></span><span leaf="">GSM5904819 C.31_S27</span><span leaf=""><br></span><span leaf="">GSM5904820 C.5_S26</span><span leaf=""><br></span><span leaf="">GSM5904821 C.7_S5</span><span leaf=""><br></span><span leaf="">GSM5904822 D.30_S8</span><span leaf=""><br></span><span leaf="">GSM5904823 D.32_S29</span><span leaf=""><br></span><span leaf="">GSM5904824 D.6_S7</span><span leaf=""><br></span><span leaf="">GSM5904825 B1</span><span leaf=""><br></span><span leaf="">GSM5904826 B2</span><span leaf=""><br></span><span leaf="">GSM5904827 B3</span><span leaf=""><br></span><span leaf="">GSM5904828 B4</span><span leaf=""><br></span><span leaf="">GSM5904829 B7</span><span leaf=""><br></span><span leaf="">GSM5904830 B8</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">R代码放了两个地方：</span></p><ul><li><section><p><span leaf="">https://github.com/OSU-BMBL/Spinal-cord-scRNAseq</span></p></section></li><li><section><p><span leaf="">https://doi.org/10.5281/zenodo.6590552</span></p></section></li></ul><h2 data-cacheurl="" data-remoteid="" data-tool="mdnice编辑器"><span></span><span><span leaf="">单细胞注释</span></span></h2><p data-tool="mdnice编辑器"><span leaf="">先把数据下载下来，进行预处理：</span></p><pre data-tool="mdnice编辑器"><code><span><span leaf="">###</span></span><span leaf=""><br></span><span><span leaf="">### Create: Jianming Zeng</span></span><span leaf=""><br></span><span><span leaf="">### Date:  2023-12-31  </span></span><span leaf=""><br></span><span><span leaf="">### Email: jmzeng1314@163.com</span></span><span leaf=""><br></span><span><span leaf="">### Blog: http://www.bio-info-trainee.com/</span></span><span leaf=""><br></span><span><span leaf="">### Forum:  http://www.biotrainee.com/thread-1376-1-1.html</span></span><span leaf=""><br></span><span><span leaf="">### CAFS/SUSTC/Eli Lilly/University of Macau</span></span><span leaf=""><br></span><span><span leaf="">### Update Log: 2023-12-31   First version </span></span><span leaf=""><br></span><span><span leaf="">### Update Log: 2024-12-09   by juan zhang (492482942@qq.com)</span></span><span leaf=""><br></span><span><span leaf="">### </span></span><span leaf=""><br></span><span leaf="">rm(list=ls())</span><span leaf=""><br></span><span leaf="">options(stringsAsFactors = F)</span><span leaf=""><br></span><span leaf="">library(ggsci)</span><span leaf=""><br></span><span leaf="">library(dplyr) </span><span leaf=""><br></span><span leaf="">library(future)</span><span leaf=""><br></span><span leaf="">library(Seurat)</span><span leaf=""><br></span><span leaf="">library(clustree)</span><span leaf=""><br></span><span leaf="">library(cowplot)</span><span leaf=""><br></span><span leaf="">library(data.table)</span><span leaf=""><br></span><span leaf="">library(ggplot2)</span><span leaf=""><br></span><span leaf="">library(patchwork)</span><span leaf=""><br></span><span leaf="">library(stringr)</span><span leaf=""><br></span><span leaf="">library(qs)</span><span leaf=""><br></span><span leaf="">library(Matrix)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 创建目录</span></span><span leaf=""><br></span><span leaf="">getwd()</span><span leaf=""><br></span><span leaf="">gse &lt;- </span><span><span leaf="">"GSE196928"</span></span><span leaf=""><br></span><span leaf="">dir.create(gse)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf="">###### step1: 导入数据 ######   </span></span><span leaf=""><br></span><span leaf="">ct &lt;- data.table::fread(</span><span><span leaf="">"GSE196928/GSE196928_counts.csv.gz"</span></span><span leaf="">,data.table = F)</span><span leaf=""><br></span><span leaf="">ct[1:5, 1:5]</span><span leaf=""><br></span><span leaf="">dim(ct)</span><span leaf=""><br></span><span leaf="">rownames(ct) &lt;- ct[,1]</span><span leaf=""><br></span><span leaf="">ct &lt;- ct[,-1]</span><span leaf=""><br></span><span leaf="">ct[1:5, 1:5]</span><span leaf=""><br></span><span leaf="">head(colnames(ct))</span><span leaf=""><br></span><span><span leaf=""># 去掉前面的X</span></span><span leaf=""><br></span><span leaf="">colnames(ct) &lt;- substring(colnames(ct), 2)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">phe &lt;- data.table::fread(</span><span><span leaf="">'GSE196928/GSE196928_metadata.csv.gz'</span></span><span leaf="">,data.table = F)</span><span leaf=""><br></span><span leaf="">head(phe)</span><span leaf=""><br></span><span leaf="">tail(phe)</span><span leaf=""><br></span><span leaf="">table(phe</span><span><span leaf="">$orig</span></span><span leaf="">.ident)</span><span leaf=""><br></span><span leaf="">rownames(phe) &lt;- phe[,1]</span><span leaf=""><br></span><span leaf="">phe &lt;- phe[,-1]</span><span leaf=""><br></span><span leaf="">rownames(phe) &lt;- gsub(</span><span><span leaf="">" "</span></span><span leaf="">, </span><span><span leaf="">"."</span></span><span leaf="">, rownames(phe))</span><span leaf=""><br></span><span leaf="">rownames(phe) &lt;- gsub(</span><span><span leaf="">"-"</span></span><span leaf="">, </span><span><span leaf="">"."</span></span><span leaf="">, rownames(phe))</span><span leaf=""><br></span><span leaf="">identical(rownames(phe),colnames(ct))</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 创建对象</span></span><span leaf=""><br></span><span leaf="">sce.all &lt;- CreateSeuratObject(counts = ct, meta.data = phe, min.cells = 3)</span><span leaf=""><br></span><span leaf="">sce.all</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 查看特征</span></span><span leaf=""><br></span><span leaf="">as.data.frame(sce.all@assays</span><span><span leaf="">$RNA</span></span><span><span leaf="">$counts</span></span><span leaf="">[1:10, 1:2])</span><span leaf=""><br></span><span leaf="">head(sce.all@meta.data, 10)</span><span leaf=""><br></span><span leaf="">table(sce.all</span><span><span leaf="">$orig</span></span><span leaf="">.ident) </span><span leaf=""><br></span><span leaf="">table(sce.all</span><span><span leaf="">$integrated_snn_res</span></span><span leaf="">.0.15) </span><span leaf=""><br></span><span leaf="">table(sce.all</span><span><span leaf="">$seurat_clusters</span></span><span leaf="">) </span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">library(qs)</span><span leaf=""><br></span><span leaf="">qsave(sce.all, file=</span><span><span leaf="">"GSE196928/sce.all.qs"</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">作者这里有直接分好的类别，</span></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxm6hAcT089C68fBxHacF0w1rHroXUFeicBjAZWZY9MaucPc7WOYovoTeHVHIOCsT6m4jZiaT6pMf0Q/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.11405529953917051" data-type="png" data-w="868" data-imgfileid="100061249" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxm6hAcT089C68fBxHacF0w1rHroXUFeicBjAZWZY9MaucPc7WOYovoTeHVHIOCsT6m4jZiaT6pMf0Q/640?wx_fmt=png&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">接下来经过数据标准化、降维聚类，harmony去批次，然后根据文献的代码注释：</span></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxm6hAcT089C68fBxHacF0waafoibZm558wMUsJwhEyicBvl66bawg44H1nk5J79N3VSicthtGIBUaQQ/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.6200750469043153" data-type="png" data-w="1066" data-imgfileid="100061251" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxm6hAcT089C68fBxHacF0waafoibZm558wMUsJwhEyicBvl66bawg44H1nk5J79N3VSicthtGIBUaQQ/640?wx_fmt=png&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">最终注释结果如下，于文献中一致：</span></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxm6hAcT089C68fBxHacF0wPacu7bMYF63VFcsrxfsshqZcJjFLHrUy7yW4npXHQgM3O49CgUqcmA/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.7981481481481482" data-type="png" data-w="1080" data-imgfileid="100061258" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxm6hAcT089C68fBxHacF0wPacu7bMYF63VFcsrxfsshqZcJjFLHrUy7yW4npXHQgM3O49CgUqcmA/640?wx_fmt=png&amp;from=appmsg"></span></figure><h2 data-cacheurl="" data-remoteid="" data-tool="mdnice编辑器"><span></span><span><span leaf="">单细胞cellchat分析</span></span></h2><p data-tool="mdnice编辑器"><span leaf="">这里作者按照样本分组进行分细胞通讯分析，我们只分析其中一组来学习上面的弦图绘制：选择 net.ctrl.00d 组别</span></p><pre data-tool="mdnice编辑器"><code><span leaf="">library(CellChat)</span><span leaf=""><br></span><span leaf="">packageVersion(</span><span><span leaf="">"CellChat"</span></span><span leaf="">)</span><span leaf=""><br></span><span><span leaf=""># [1] ‘2.2.0’</span></span><span leaf=""><br></span><span leaf="">library(patchwork)</span><span leaf=""><br></span><span leaf="">library(Hmisc)</span><span leaf=""><br></span><span><span leaf=""># library(hash)</span></span><span leaf=""><br></span><span leaf="">library(tidyverse)</span><span leaf=""><br></span><span leaf="">library(circlize)</span><span leaf=""><br></span><span leaf="">library(ggsci)</span><span leaf=""><br></span><span leaf="">library(igraph)</span><span leaf=""><br></span><span leaf="">library(gtools)</span><span leaf=""><br></span><span leaf="">library(ComplexHeatmap)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># counts : the gene expression matrix</span></span><span leaf=""><br></span><span><span leaf=""># meta   : the meta data matched with the gene expression matrix</span></span><span leaf=""><br></span><span leaf="">load(file = </span><span><span leaf="">"2-harmony/sce.all.filt.RData"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">sce.all.filt</span><span leaf=""><br></span><span leaf="">head(sce.all.filt@meta.data)</span><span leaf=""><br></span><span leaf="">table(sce.all.filt</span><span><span leaf="">$orig</span></span><span leaf="">.ident)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">sce.all.filt &lt;- subset(sce.all.filt, orig.ident==</span><span><span leaf="">"00 d Control"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">sce.all.filt</span><span leaf=""><br></span></code></pre><h3 data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><span></span><span><span leaf="">细胞通讯：</span></span><span></span></h3><p data-tool="mdnice编辑器"><span leaf="">前面也做了细胞通讯结果可视化：<a target="_blank" href="https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247544734&amp;idx=1&amp;sn=afb7ead473e91445b695efea54807327&amp;scene=21#wechat_redirect" textvalue="" linktype="text" data-linktype="2">高分杂志同款cellchat细胞通讯结果气泡图绘制（IF=25.083）</a></span></p><pre data-tool="mdnice编辑器"><code><span leaf="">counts &lt;- sce.all.filt@assays</span><span><span leaf="">$RNA</span></span><span><span leaf="">$counts</span></span><span leaf=""><br></span><span leaf="">counts &lt;- as.matrix(counts) </span><span><span leaf=""># transform data frame into matrix</span></span><span leaf=""><br></span><span leaf="">counts[1:4,1:4]</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">meta &lt;- sce.all.filt@meta.data</span><span leaf=""><br></span><span leaf="">meta</span><span><span leaf="">$samples</span></span><span leaf=""> &lt;- meta</span><span><span leaf="">$orig</span></span><span leaf="">.ident</span><span leaf=""><br></span><span leaf="">head(meta)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 创建对象</span></span><span leaf=""><br></span><span leaf="">cellchat &lt;- createCellChat(object = counts, meta = meta, group.by = </span><span><span leaf="">"celltype"</span></span><span leaf="">) </span><span><span leaf=""># create a CellChat object</span></span><span leaf=""><br></span><span leaf="">levels(cellchat@idents) </span><span><span leaf=""># check the cell types</span></span><span leaf=""><br></span><span leaf="">CellChatDB &lt;- CellChatDB.mouse </span><span><span leaf=""># set the database according to the organism</span></span><span leaf=""><br></span><span leaf="">dplyr::glimpse(CellChatDB</span><span><span leaf="">$interaction</span></span><span leaf="">) </span><span><span leaf=""># show the structure of the database</span></span><span leaf=""><br></span><span leaf="">unique(CellChatDB</span><span><span leaf="">$interaction</span></span><span><span leaf="">$annotation</span></span><span leaf="">) </span><span><span leaf=""># check the annotation of interactions</span></span><span leaf=""><br></span><span leaf="">CellChatDB.use &lt;- CellChatDB </span><span><span leaf=""># use all databases</span></span><span leaf=""><br></span><span leaf="">cellchat@DB &lt;- CellChatDB.use </span><span><span leaf=""># set the used database in the object</span></span><span leaf=""><br></span><span leaf="">cellchat &lt;- subsetData(cellchat) </span><span><span leaf=""># subset the expression data of signaling genes for saving computation cost</span></span><span leaf=""><br></span><span leaf="">cellchat &lt;- identifyOverExpressedGenes(cellchat)</span><span leaf=""><br></span><span leaf="">cellchat &lt;- identifyOverExpressedInteractions(cellchat)</span><span leaf=""><br></span><span><span leaf=""># cellchat &lt;- projectData(cellchat, PPI.mouse)</span></span><span leaf=""><br></span><span leaf="">cellchat &lt;- computeCommunProb(cellchat, raw.use = T, nboot = 1, population.size = T)</span><span leaf=""><br></span><span leaf="">cellchat &lt;- filterCommunication(cellchat, min.cells = 10)</span><span leaf=""><br></span><span leaf="">save(cellchat, file = </span><span><span leaf="">"cellchat.RData"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf="">#load("cellchat.RData")</span></span><span leaf=""><br></span><span leaf="">df.net &lt;- subsetCommunication(cellchat) </span><span><span leaf=""># get the pairs of cell types or ligand receptors in the interactions</span></span><span leaf=""><br></span><span leaf="">head(df.net)</span><span leaf=""><br></span><span leaf="">cell_type_pair &lt;- paste0(df.net</span><span><span leaf="">$source</span></span><span leaf="">, </span><span><span leaf="">'_'</span></span><span leaf="">, df.net</span><span><span leaf="">$target</span></span><span leaf="">) </span><span><span leaf=""># build the pairs of interaction cell types</span></span><span leaf=""><br></span><span leaf="">cell_type_pair</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">mean_exp &lt;- sapply(1:nrow(df.net), </span><span><span leaf="">function</span></span><span leaf="">(x) {</span><span leaf=""><br></span><span><span leaf="">#x &lt;- 1</span></span><span leaf=""><br></span><span><span leaf="">print</span></span><span leaf="">(x)</span><span leaf=""><br></span><span leaf="">  ct1 &lt;- df.net</span><span><span leaf="">$source</span></span><span leaf="">[x] </span><span><span leaf=""># source cell type</span></span><span leaf=""><br></span><span leaf="">  ct2 &lt;- df.net</span><span><span leaf="">$target</span></span><span leaf="">[x] </span><span><span leaf=""># target cell type</span></span><span leaf=""><br></span><span leaf="">  cells1 &lt;- rownames(meta[meta</span><span><span leaf="">$celltype</span></span><span leaf=""> == ct1,]) </span><span><span leaf=""># cells belonging to the source cell type</span></span><span leaf=""><br></span><span leaf="">  cells2 &lt;- rownames(meta[meta</span><span><span leaf="">$celltype</span></span><span leaf=""> == ct2,])</span><span><span leaf=""># cells belonging to the target cell type</span></span><span leaf=""><br></span><span leaf="">  gene_pair1 &lt;- df.net</span><span><span leaf="">$ligand</span></span><span leaf="">[x] </span><span><span leaf=""># genes corresponding to the ligand</span></span><span leaf=""><br></span><span leaf="">  genes1 &lt;- capitalize(tolower(strsplit(gene_pair1, split = </span><span><span leaf="">'_'</span></span><span leaf="">)[[1]]))</span><span leaf=""><br></span><span leaf="">  gene_pair2 &lt;- df.net</span><span><span leaf="">$receptor</span></span><span leaf="">[x] </span><span><span leaf=""># genes corresponding to the receptor</span></span><span leaf=""><br></span><span leaf="">  genes2 &lt;- capitalize(tolower(strsplit(gene_pair2, split = </span><span><span leaf="">'_'</span></span><span leaf="">)[[1]]))</span><span leaf=""><br></span><span><span leaf="">if</span></span><span leaf=""> (length(genes2) &gt; 1 &amp; genes2[2] == </span><span><span leaf="">'R2'</span></span><span leaf="">) {</span><span leaf=""><br></span><span leaf="">    genes2[2] &lt;- gsub(pattern = </span><span><span leaf="">'1'</span></span><span leaf="">, replacement = </span><span><span leaf="">'2'</span></span><span leaf="">, genes2[1])</span><span leaf=""><br></span><span leaf="">  }</span><span leaf=""><br></span><span><span leaf="">#cat("Ligand genes: ", genes1, "\n")</span></span><span leaf=""><br></span><span><span leaf="">#cat("Receptor genes: ", genes2, "\n\n")</span></span><span leaf=""><br></span><span><span leaf="">#mean1 &lt;- mean(counts[genes1, cells1]) # the mean expression value of the ligand within the source cell type</span></span><span leaf=""><br></span><span><span leaf="">#mean2 &lt;- mean(counts[genes2, cells2]) # the mean expression value of the receptor within the source cell type</span></span><span leaf=""><br></span><span leaf="">  mean1 &lt;- mean(counts[intersect(rownames(sce.all.filt), genes1), cells1 ]) </span><span><span leaf=""># the mean expression value of the ligand within the source cell type</span></span><span leaf=""><br></span><span leaf="">  mean2 &lt;- mean(counts[intersect(rownames(sce.all.filt), genes2), cells2 ]) </span><span><span leaf=""># the mean expression value of the receptor within the source cell type</span></span><span leaf=""><br></span><span leaf="">  avg &lt;- (mean1 + mean2) / 2 </span><span><span leaf=""># the mean expression value of the ligands and receptors</span></span><span leaf=""><br></span><span><span leaf="">return</span></span><span leaf="">(avg)</span><span leaf=""><br></span><span leaf="">})</span><span leaf=""><br></span><span leaf="">df.net &lt;- cbind(cell_type_pair, mean_exp, df.net) </span><span><span leaf=""># add the mean expression and cell type pairs</span></span><span leaf=""><br></span><span leaf="">head(df.net)</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxm6hAcT089C68fBxHacF0wVzyvPyibXhiaBbOaEYZzJiceXKydIb0leErWKYNmBaUC0ficFnNPxMTNibw/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.21944444444444444" data-type="png" data-w="1080" data-imgfileid="100061257" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxm6hAcT089C68fBxHacF0wVzyvPyibXhiaBbOaEYZzJiceXKydIb0leErWKYNmBaUC0ficFnNPxMTNibw/640?wx_fmt=png&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">细胞通讯的结果就做好啦！</span></p><p data-tool="mdnice编辑器"><span leaf="">后面使用 基础函数绘制弦图的代码篇幅比较长，我们下次分享（待续~）</span></p></section><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com" data-pm-slice="0 0 []"><h4 data-tool="mdnice编辑器"><span leaf="">文末友情宣传</span></h4><p data-tool="mdnice编辑器"><span leaf="">强烈建议你推荐给身边的</span><strong><span leaf="">博士后以及年轻生物学PI</span></strong><span leaf="">，多一点数据认知，让他们的科研上一个台阶：</span></p><ul></ul></section><ul><li><section><span leaf=""><a target="_blank" href="https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247544311&amp;idx=1&amp;sn=d41b5838e799f52280e78703135bb603&amp;scene=21#wechat_redirect" textvalue="生信入门&amp;数据挖掘线上直播课8月班" data-itemshowtype="0" linktype="text" data-linktype="2">生信入门&amp;数据挖掘线上直播课8月班</a><span textstyle="">，你的生物信息学入门课</span></span></section></li><li><section><span leaf=""><a target="_blank" href="https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247525079&amp;idx=1&amp;sn=0b997af16a58195b4192691373048fd5&amp;scene=21#wechat_redirect" textvalue="时隔5年，我们的生信技能树VIP学徒继续招生啦" data-itemshowtype="0" linktype="text" data-linktype="2"><span textstyle="">时隔5年，我们的生信技能树VIP学徒继续招生啦</span></a></span></section></li><li><section><span leaf=""><a target="_blank" href="https://mp.weixin.qq.com/s?__biz=MzUzMTEwODk0Ng==&amp;mid=2247530048&amp;idx=1&amp;sn=28aa7bbd5e00521f79e074496a5f5d66&amp;scene=21#wechat_redirect" textvalue="满足你生信分析计算需求的低价解决方案" data-itemshowtype="0" linktype="text" data-linktype="2">满足你生信分析计算需求的低价解决方案</a></span></section></li><li><section><span leaf=""><a target="_blank" href="https://mp.weixin.qq.com/mp/appmsgalbum?__biz=MzAxMDkxODM1Ng==&amp;action=getalbum&amp;album_id=1679199708449144836&amp;scene=173&amp;subscene=207&amp;sessionid=1745492310&amp;enterid=1745492314&amp;from_msgid=2247541298&amp;from_itemidx=1&amp;count=3&amp;nolastread=1#wechat_redirect" textvalue="生信故事会" data-itemshowtype="0" linktype="text" data-linktype="2">生信故事会</a>，来看看他们的生信入门故事</span></section></li><li><section><span leaf=""><a target="_blank" href="https://mp.weixin.qq.com/mp/appmsgalbum?__biz=MzAxMDkxODM1Ng==&amp;action=getalbum&amp;album_id=3690970204957147140&amp;scene=173&amp;subscene=207&amp;sessionid=1745066271&amp;enterid=1745066274&amp;from_msgid=2247540702&amp;from_itemidx=1&amp;count=3&amp;nolastread=1#wechat_redirect" textvalue="生信马拉松答疑专辑" data-itemshowtype="0" linktype="text" data-linktype="2">生信马拉松答疑专辑</a>，获取你的生信专属答疑</span></section></li></ul><section><span leaf=""><br></span></section><p><mp-style-type data-value="3"></mp-style-type></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/m40ulfefUrjIHbFmxQ59Yg",target="_blank" rel="noopener noreferrer">原文链接</a>
