---
title: "Science杂志高颜值GSEA打分排序图"
date: 2025-09-19T01:16:08Z
draft: ["false"]
tags: [
  "fetched",
  "生信技能树"
]
categories: ["Acdemic"]
---
Science杂志高颜值GSEA打分排序图 by 生信技能树
------
<div><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com"><blockquote data-tool="mdnice编辑器"><p>前面给大家介绍过一个高颜值富集分析结果展示图：<a href="https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247537055&amp;idx=1&amp;sn=26544d5687fbe6001391e869ea84e692&amp;scene=21#wechat_redirect" data-linktype="2">一种很新的功能富集结果展示方法</a>。今天来还是来复现这篇 2024 年 6 月份发表在 顶刊 science 杂志上的文献《<strong>Defining the KRAS- and ERK-dependent transcriptome in KRAS-mutant cancers</strong>》中另一个富集分析GSEA结果图，图片以及含义如下：</p></blockquote><h4 data-tool="mdnice编辑器"><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com"><p data-tool="mdnice编辑器">含义：<strong>作者对 KRAS versus NS siRNA 处理的 PDAC 细胞系差异基因进行 GSEA 分析，基因集为 50 个 Hallmark gene sets</strong>。然后取每条通路的 NES 打分从高到低进行排序，并绘制散点图。<strong>点的大小</strong>为 Detection：应该为每个通路中基因在所有样本中表达的 count 大于 5 的占比（我们这里并没有表达矩阵，就用基因集大小替代好了）。<strong>点的颜色</strong>为通路显著性：-log10 pvalue。</p></section><span></span></h4><figure data-tool="mdnice编辑器"><img data-imgfileid="100054251" data-ratio="0.7162162162162162" data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwUJ2WlCfwngHZR7Z6ZyCFqOkNpkaY1C7tKCK23ssJO9koMwkK9fDJEZXLhCPqib3Kj3bmolGffePw/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="518" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwUJ2WlCfwngHZR7Z6ZyCFqOkNpkaY1C7tKCK23ssJO9koMwkK9fDJEZXLhCPqib3Kj3bmolGffePw/640?wx_fmt=png&amp;from=appmsg"></figure><p data-tool="mdnice编辑器">图注：</p><blockquote data-tool="mdnice编辑器"><span></span><p>Fig. 1. Establishment and evaluation of a KRAS dependent gene expression program in KRAS-mutant PDAC. (C) GSEA for 50 Hallmark gene sets within DEGs after KRAS siRNA treatment. Detection is based on the presence of a gene in all eight PDAC cell lines at &gt;5 reads.</p></blockquote><p data-tool="mdnice编辑器">关于可不可以用差异基因进行GSEA分析，我们前面讨论过：<a href="https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247537475&amp;idx=1&amp;sn=0ba7600556f909d01727f59010d7a9f9&amp;scene=21#wechat_redirect" data-linktype="2">IF10+杂志文章只用统计学显著的差异基因做GSEA就合理吗？</a></p><h2 data-tool="mdnice编辑器"><span></span><span>数据背景介绍</span><span></span></h2><h3 data-tool="mdnice编辑器"><span></span><span></span><span>1、PDAC细胞系KRAS siRNA RNA测序</span><span></span></h3><p data-tool="mdnice编辑器">为了确定 KRAS 依赖性转录组，作者对一组八个人类 KRAS 突变型 胰腺导管腺癌PDAC的细胞系在经过 24 小时 KRAS 小干扰 RNA（siRNA）处理后的基因转录变化（图 S1A 和 B）进行了 RNA 测序（RNA-seq）。数据可在这里下载：<strong>PRJNA980201</strong> https://www.ebi.ac.uk/ena/browser/view/PRJNA980201?show=related-records。</p><p data-tool="mdnice编辑器">Note：</p><ul data-tool="mdnice编辑器"><li><section><p>NS缩写：control non-specific ("NS") siRNA。</p></section></li><li><section><p>KRAS siRNA是一种小干扰RNA（siRNA），通过特异性靶向KRAS基因的mRNA，从而抑制其表达，达到治疗KRAS突变肿瘤的目的。</p></section></li></ul><figure data-tool="mdnice编辑器"><img data-imgfileid="100054250" data-ratio="0.1969532100108814" data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwUJ2WlCfwngHZR7Z6ZyCFqofI03HYCvE1woibvSXdysAV4VQJTKfiaamENsy54hNmAng8UkWRbxx1w/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="919" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwUJ2WlCfwngHZR7Z6ZyCFqofI03HYCvE1woibvSXdysAV4VQJTKfiaamENsy54hNmAng8UkWRbxx1w/640?wx_fmt=png&amp;from=appmsg"></figure><p data-tool="mdnice编辑器">差异分析结果如下：The 677 KRAS-dependent (UP) and 1051 KRAS-inhibited (DN) genes (<strong>log2FC &gt; 0.5, adj. p &lt; 0.05</strong>) are indicated by the light blue– and light red–shaded rectangles underlaid on plot, respectively。</p><figure data-tool="mdnice编辑器"><img data-imgfileid="100054252" data-ratio="0.7696629213483146" data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwUJ2WlCfwngHZR7Z6ZyCFqBJMsUKI1Fic0uH2Pia7sU3suBNncH2t98I4Ey2uyj19YcOgTSuShbIyQ/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="534" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwUJ2WlCfwngHZR7Z6ZyCFqBJMsUKI1Fic0uH2Pia7sU3suBNncH2t98I4Ey2uyj19YcOgTSuShbIyQ/640?wx_fmt=png&amp;from=appmsg"></figure><p data-tool="mdnice编辑器">差异结果在表格S1：<strong>Data S1.</strong> Differential expression statistics for genes in KRAS versus NS siRNA treated PDAC cells, using RNA-sequencing。</p><h3 data-tool="mdnice编辑器"><span></span><span></span><span>2、50 Hallmark gene sets</span><span></span></h3><p data-tool="mdnice编辑器">Hallmark 通路可以在GSEA 的 MSigDB 数据库去下载 gmt 格式：<span>https://www.gsea-msigdb.org/gsea/msigdb/human/collections.jsp</span></p><figure data-tool="mdnice编辑器"><img data-imgfileid="100054249" data-ratio="0.43866171003717475" data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwUJ2WlCfwngHZR7Z6ZyCFqMo55d45P5eTlY7IDhn8TcSbnLKM1uytacWh99zMib7D4Z26FuttOCXw/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="807" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwUJ2WlCfwngHZR7Z6ZyCFqMo55d45P5eTlY7IDhn8TcSbnLKM1uytacWh99zMib7D4Z26FuttOCXw/640?wx_fmt=png&amp;from=appmsg"></figure><h2 data-tool="mdnice编辑器"><span></span><span>GSEA分析</span><span></span></h2><p data-tool="mdnice编辑器">GSEA的输入数据为差异基因排序列表，排序指标为log2FC值：</p><h3 data-tool="mdnice编辑器"><span></span><span></span><span>1、读取数据</span><span></span></h3><pre data-tool="mdnice编辑器"><span></span><code>rm(list=ls())<br><span># 加载R包</span><br>library(ggplot2)<br>library(tibble)<br>library(ggrepel)<br>library(tidyverse)<br>library(dplyr)<br>library(patchwork)<br><br><span>##### 01、加载数据</span><br><br><span># 加载：KRAS vs NS siRNA（log2FC）</span><br>group1 &lt;- read.csv(<span>"./data/science.adk0775_data_s1.csv"</span> )<br>head(group1)<br><br><span># 提取FDR值和FC值</span><br>group1 &lt;- group1[,c(<span>"external_gene_name"</span>,<span>"logFC"</span>,<span>"FDR"</span> )]<br>group1 &lt;- group1[group1<span>$external_gene_name</span>!=<span>""</span>, ]<br>group1 &lt;- distinct(group1,external_gene_name,.keep_all = T)<br>rownames(group1) &lt;- group1<span>$external_gene_name</span><br>head(group1)<br><br><span># 增加一列上下调：log2FC &gt; 0.5, adj. p &lt; 0.05</span><br>group1<span>$g1</span> &lt;- <span>"normal"</span><br>group1<span>$g1</span>[group1<span>$logFC</span> &gt;0.5 &amp; group1<span>$FDR</span> &lt; 0.05 ] &lt;- <span>"up"</span><br>group1<span>$g1</span>[group1<span>$logFC</span> &lt; -0.5 &amp; group1<span>$FDR</span> &lt; 0.05 ] &lt;- <span>"down"</span><br>table(group1<span>$g1</span>)<br><span># down normal     up </span><br><span># 675  12876   1043 </span><br><br><span># 获得差异基因</span><br>DEGs &lt;- group1[group1<span>$g1</span>!=<span>"normal"</span>,]<br>head(DEGs)<br><br><span># 得到排序列表</span><br>DEGs &lt;- DEGs[order(DEGs<span>$logFC</span>, decreasing = T),]<br>genelist &lt;- DEGs<span>$logFC</span><br>names(genelist) &lt;- DEGs<span>$external_gene_name</span><br>head(genelist)<br><span># NPPC   SMTNL2    KCNK3     CHPF PCDHGA12  EXOC3L1 </span><br><span># 2.860472 2.712068 2.551034 2.536558 2.499657 2.446140</span><br><br>tail(genelist)<br></code></pre><p data-tool="mdnice编辑器">得到 1718个差异基因：</p><figure data-tool="mdnice编辑器"><img data-imgfileid="100054248" data-ratio="0.2505091649694501" data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwUJ2WlCfwngHZR7Z6ZyCFqicOicMH2G7m3dlABTCAmRdMWLGTqxY2GlPdhbpTvmu7y8Xmia6psE2riag/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="491" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwUJ2WlCfwngHZR7Z6ZyCFqicOicMH2G7m3dlABTCAmRdMWLGTqxY2GlPdhbpTvmu7y8Xmia6psE2riag/640?wx_fmt=png&amp;from=appmsg"></figure><h3 data-tool="mdnice编辑器"><span></span><span></span><span>2、读取 50 Hallmark gene sets 通路并富集：</span><span></span></h3><pre data-tool="mdnice编辑器"><span></span><code><span>## === HALLMARK通路富集</span><br>geneset &lt;- read.gmt(<span>"data/h.all.v2024.1.Hs.symbols.gmt"</span>)<br>table(geneset<span>$term</span>)<br>geneset<span>$term</span> &lt;- gsub(pattern = <span>"HALLMARK_"</span>,<span>""</span>, geneset<span>$term</span>)<br><br><span># 运行,输出全部结果</span><br>egmt &lt;- GSEA(genelist, TERM2GENE=geneset, pvalueCutoff = 1, minGSSize = 1, maxGSSize = 500000)<br>colnames(egmt@result)<br>head(egmt[, 1:6])<br></code></pre><h3 data-tool="mdnice编辑器"><span></span><span></span><span>3、使用 ggplot2 定制化绘图</span><span></span></h3><p data-tool="mdnice编辑器">取出绘图需要的列，并进行相关设置：</p><pre data-tool="mdnice编辑器"><span></span><code><span># 绘图</span><br>data &lt;- egmt[,c(<span>"ID"</span>, <span>"NES"</span>,<span>"setSize"</span>,<span>"pvalue"</span>)]<br>data<span>$setSize_1</span> &lt;- data<span>$setSize</span>/10<br>head(data)<br><br><span># 给Y轴的通路名设置为因子，排序</span><br>data &lt;- data[order(data<span>$NES</span>, decreasing = T),]<br>data<span>$ID</span> &lt;- factor(data<span>$ID</span>, levels = data<span>$ID</span>)<br>data<span>$xlab</span> &lt;- 1:49<br>head(data)<br>summary(data<span>$NES</span>)<br>summary(data<span>$setSize_1</span>)<br><br><span># 图中标出的通路名字</span><br>label &lt;- c( <span>"KRAS_SIGNALING_DN"</span>, <span>"INTERFERON_ALPHA_RESPONSE"</span>,<br>  <span>"UV_RESPONSE_DN"</span>, <span>"EMT"</span>, <span>"TNFA_SIGNALING_VIA_NFKB"</span>, <span>"MYC_TARGETS_V2"</span>,<br>  <span>"KRAS_SIGNALING_UP"</span>, <span>"MYC_TARGETS_V1"</span>, <span>"G2M_CHECKPOINT"</span>, <span>"E2F_TARGETS"</span> )<br><br><span># 没有EMT通路</span><br>data_label &lt;- data[data<span>$ID</span> %<span>in</span>% label,]<br>data_label<br><span># 图中通路的颜色</span><br>data_label<span>$col</span> &lt;- c(<span>"black"</span>, <span>"grey80"</span>,<span>"grey80"</span>,<span>"grey80"</span>,<span>"grey80"</span>,<span>"#ffb882"</span>,<span>"grey80"</span>,<span>"#ff5eff"</span>,<span>"#ff5eff"</span>)<br></code></pre><p data-tool="mdnice编辑器">使用ggplot2绘图：</p><pre data-tool="mdnice编辑器"><span></span><code>p &lt;- ggplot(data = data, aes(x = xlab, y = NES)) +<br>  geom_point(aes(size = setSize_1, alpha = -log10(pvalue)), shape = 21, stroke = 0.7,fill = <span>"#0000ff"</span>, colour = <span>"black"</span>) + <span># stroke：设置点的边框宽度。</span><br>  scale_size_continuous(range = c(0.2, 8)) + <br>  xlab(label = <span>"Hallmark gene sets"</span>) + <br>  ylab(label = <span>"Normalized enrichment score (NES)"</span>) + <br>  theme_classic(base_size = 15) + <br>  scale_x_continuous(breaks = seq(0, 50, by = 10), labels = seq(0, 50, by = 10) ) + <span># 设置x轴的刻度线和刻度标签</span><br>  scale_y_continuous( breaks = seq(-4, 2.3, by = 1), labels = seq(-4, 2.3, by = 1) ) + <br>  guides(size = guide_legend(title = <span>"Detection\n(proportion)"</span>),<br>         alpha = guide_legend(title = <span>"Significance\n(-log10 p-val.)"</span>) ) +  <span># 修改图例标题</span><br>  theme(<br>    axis.line = element_line(color = <span>"black"</span>, size = 0.6), <span># 加粗x轴和y轴的线条</span><br>    axis.text = element_text(face = <span>"bold"</span>), <span># 加粗x轴和y轴的标签</span><br>    axis.title = element_text( size = 13)    <span># 加粗x轴和y轴的标题</span><br>  )<br><br>p<br></code></pre><figure data-tool="mdnice编辑器"><img data-imgfileid="100054256" data-ratio="0.7892074198988196" data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwUJ2WlCfwngHZR7Z6ZyCFqcibmbDVg5C24XJV4FoEic4QALbeJGBcvoGSFFGhHWwGCC9AqibJVwcQ3g/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="593" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwUJ2WlCfwngHZR7Z6ZyCFqcibmbDVg5C24XJV4FoEic4QALbeJGBcvoGSFFGhHWwGCC9AqibJVwcQ3g/640?wx_fmt=png&amp;from=appmsg"></figure><p data-tool="mdnice编辑器">添加通路标签：</p><pre data-tool="mdnice编辑器"><span></span><code><span># 添加label：vjust（垂直调整）或hjust（水平调整）</span><br>p3 &lt;- p + <br>  geom_text_repel(data= data_label, aes(x = xlab, y = NES,label = ID), size = 3, color = data_label<span>$col</span>, <br>                  force=20,                <span># 将重叠的文本标签相互推开的强度。force 参数的值越大，标签之间的排斥力度也越大，这会导致标签在图中更分散地排列</span><br>                  point.padding = 0.5,     <span># 设置文本标签与对应点之间的最小距离</span><br>                  min.segment.length = 0,  <span># 长度大于0就可以添加引线</span><br>                  hjust = 1.2,             <span># 文本标签的右侧与指定位置对齐</span><br>                  segment.color=<span>"grey20"</span>,<br>                  segment.size=0.3,        <span># 设置引导线的粗细</span><br>                  segment.alpha=0.8,       <span># 文本标签中连接线段的透明度</span><br>                  nudge_y=-0.1             <span># 在y轴方向上微调标签位置</span><br>                  ) <br>p3<br><br><span># 保存</span><br>ggsave(filename = <span>"Figure1C.pdf"</span>, plot = p3, width = 6.2, height = 5)<br></code></pre><p data-tool="mdnice编辑器">结果如下：</p><figure data-tool="mdnice编辑器"><img data-imgfileid="100054257" data-ratio="0.8226837060702875" data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwUJ2WlCfwngHZR7Z6ZyCFqJRkmmKSx2TqmMmFehAggpsgdMedXiaEeaAhen9baZibqw88icWjdA7w3A/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="626" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwUJ2WlCfwngHZR7Z6ZyCFqJRkmmKSx2TqmMmFehAggpsgdMedXiaEeaAhen9baZibqw88icWjdA7w3A/640?wx_fmt=png&amp;from=appmsg"></figure><h3 data-tool="mdnice编辑器"><span></span><span></span><span>友情宣传：</span><span></span></h3><p data-tool="mdnice编辑器"><a href="https://mp.weixin.qq.com/s?__biz=MzI1Njk4ODE0MQ==&amp;mid=2247527230&amp;idx=1&amp;sn=7156afcd5ab734c7d391b9048695747a&amp;scene=21#wechat_redirect" data-linktype="2">生信入门&amp;数据挖掘线上直播课2025年1月班</a></p><p data-tool="mdnice编辑器"><a href="http://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247524148&amp;idx=1&amp;sn=7806da6feb41a36493c519c1cfc1d3ac&amp;chksm=9b4bdf8fac3c569960369602f1ef26639cb366b250f233b2297d1f059471c0458335bfc0b829&amp;scene=21#wechat_redirect" data-linktype="2">时隔5年，我们的生信技能树VIP学徒继续招生啦</a></p><p data-tool="mdnice编辑器"><a href="https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247535760&amp;idx=2&amp;sn=1e02a2e982a046ecf6389231e6768d5b&amp;scene=21#wechat_redirect" data-linktype="2">满足你生信分析计算需求的低价解决方案</a></p></section><p><mp-style-type data-value="3"></mp-style-type></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/udzFxcwSpqJACkPM26oHlw",target="_blank" rel="noopener noreferrer">原文链接</a>
