---
title: "几行代码实现 8 种单细胞批次整合方法！？SeuratIntegrate 包"
date: 2025-09-06T14:28:23Z
draft: ["false"]
tags: [
  "fetched",
  "生信碱移"
]
categories: ["Acdemic"]
---
几行代码实现 8 种单细胞批次整合方法！？SeuratIntegrate 包 by 生信碱移
------
<div><section data-tool="markdown编辑器" data-website="https://markdown.com.cn/editor" data-pm-slice="0 0 []"><section data-mpa-powered-by="yiban.io" data-style='white-space: normal; max-width: 100%; letter-spacing: 0.544px; text-size-adjust: auto; background-color: rgb(255, 255, 255); font-family: "Helvetica Neue", Helvetica, "Hiragino Sans GB", "Microsoft YaHei", Arial, sans-serif; box-sizing: border-box !important; overflow-wrap: break-word !important;' data-pm-slice="9 4 []"><section><section><section><section data-id="85660" data-custom="rgb(117, 117, 118)" data-color="rgb(117, 117, 118)"><section data-style="margin-top: 2em; padding-top: 0.5em; padding-bottom: 0.5em; max-width: 100%; border-style: solid none; text-decoration: inherit; border-top-color: rgb(204, 204, 204); border-bottom-color: rgb(204, 204, 204); border-top-width: 1px; border-bottom-width: 1px; box-sizing: border-box !important; overflow-wrap: break-word !important;"><p><span><span leaf="">生信碱移</span></span></p><section><strong><span leaf="">单细胞批次整合</span></strong></section></section></section></section></section></section></section><blockquote><section><span><span leaf="">SeuratIntegrate 包集成了八种主流单细胞数据批次整合方法，通过几行代码就可以实现全流程调用，以及整合性能评估。</span></span></section></blockquote><p data-tool="markdown.com.cn编辑器"><span leaf="">单细胞数据集的快速增长使得研究来自不同生物学和技术背景的多个样本成为可能。一些研究将类似的样本整合构建了大规模的细胞图谱，可以用于<span textstyle="">细胞映射</span>、<span textstyle="">疾病相关亚群</span>鉴定等分析。<span textstyle="">尽管如此，这些图谱汇编了来自不同实验、试剂批次或测序技术的数据，将如此大规模的数据需要克服可能掩盖真实生物学差异的混杂效应。</span></span></p><section nodeleaf=""><img data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/LvUIqvYKCeVMhL3qmYYHdLKliahnx99TgrduRMkvicYCPC2r1NduEK2D09n91Nz4EicBmmTjaAoDnibNwAzcvgAmpA/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.578125" data-s="300,640" data-type="png" data-w="640" type="block" data-imgfileid="100012798" src="https://mmbiz.qpic.cn/sz_mmbiz_png/LvUIqvYKCeVMhL3qmYYHdLKliahnx99TgrduRMkvicYCPC2r1NduEK2D09n91Nz4EicBmmTjaAoDnibNwAzcvgAmpA/640?wx_fmt=png&amp;from=appmsg"></section><p data-tool="markdown.com.cn编辑器"><span leaf=""><span textstyle="">图：68种批次整合方法的排名</span><span textstyle="">。小编先前介绍了这篇单细胞批次整合方法的测评文章，感兴趣的老铁可以</span><a target="_blank" href="https://mp.weixin.qq.com/s?__biz=MzkyNTIzMzYyMA==&amp;mid=2247495480&amp;idx=1&amp;sn=349f589ee2f8fc982eb2b5d7617b5f08&amp;scene=21#wechat_redirect" textvalue="点击此处阅读" data-itemshowtype="0" linktype="text" data-linktype="2"><span textstyle="">点击此处阅读</span></a><span textstyle="">。</span></span></p><p data-tool="markdown.com.cn编辑器"><span leaf=""><span textstyle="">Seurat</span> 和 <span textstyle="">Scanpy</span> 分别是最广泛使用的基于 R 和基于 Python 的单细胞数据分析工具。<span textstyle="">两者都提供了全面的功能，包括前面提到的批次效应校正和数据整合。</span>Seurat 一开始采用了一种基于“锚点”的整合策略，通过分配最近邻来批次效应校正。其 V5 版本添加了对 Harmony 和基于 Python 的 scVI 的原生支持，使得受支持的整合方法总数增加到五种<span textstyle="">(scVI也是目前最强的批次整合工具之一)</span>。<span textstyle="">多项研究表明，不同的数据集可能需要不同的批次整合方法。</span></span></p><p data-tool="markdown.com.cn编辑器"><span leaf="">为此，来自波尔多大学的研究者开发了 <span textstyle="">SeuratIntegrate</span>，一个灵活且功能全面的 R 包，作为 Seurat 的扩展。<span textstyle="">SeuratIntegrate 能够在 R 环境中完整运行基于 R 和基于 Python 的几种批次整合方法并评估其性能。</span></span></p><section nodeleaf=""><img data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/LvUIqvYKCeVMhL3qmYYHdLKliahnx99Tgzaeqv556pGDdpBmxziaaJmLs4y2EBkZUOicibxORzc6uict2fJe0vc27uA/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.5248" data-s="300,640" data-type="png" data-w="625" type="block" data-imgfileid="100012799" src="https://mmbiz.qpic.cn/sz_mmbiz_png/LvUIqvYKCeVMhL3qmYYHdLKliahnx99Tgzaeqv556pGDdpBmxziaaJmLs4y2EBkZUOicibxORzc6uict2fJe0vc27uA/640?wx_fmt=png&amp;from=appmsg"></section><p data-tool="markdown.com.cn编辑器"><span leaf=""><span textstyle="">DOI: </span></span><span leaf=""><span textstyle="">10.1093/bioinformatics/btaf358</span></span></p><p data-tool="markdown.com.cn编辑器"><span leaf="">具体来说，<span textstyle="">SeuratIntegrate</span> 集成了<span textstyle="">八种</span>主流单细胞数据整合方法，<span textstyle="">其中 R 语言方法包括</span> <span textstyle="">MNN</span><span textstyle="">（互为最近邻）、</span><span textstyle="">Harmony</span><span textstyle="">（迭代嵌入校正）</span>和 <span textstyle="">ComBat</span><span textstyle="">（经验贝叶斯调整）</span><span textstyle="">；</span><span textstyle="">Python 方法则涵盖</span><span textstyle="">基于深度学习的</span><span textstyle=""> </span><span textstyle="">scVI</span><span textstyle="">（变分自编码器）、</span><span textstyle="">scANVI</span><span textstyle="">（半监督变分自编码器）和</span><span textstyle=""> </span><span textstyle="">trVAE</span><span textstyle="">（条件变分自编码器），以及基于邻域与流形的</span><span textstyle=""> </span><span textstyle="">BBKNN</span><span textstyle="">（批次均衡近邻）和 </span><span textstyle="">Scanorama</span><span textstyle="">（流形对齐）。</span></span></p><section nodeleaf=""><img data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/LvUIqvYKCeVMhL3qmYYHdLKliahnx99TgXoabiaM3fDTfVNKXSIZlicTFOTKRWPoLzlXfguzxRFic1Vfwc0hBiaqIsg/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.44244105409153955" data-s="300,640" data-type="png" data-w="721" type="block" data-imgfileid="100012800" src="https://mmbiz.qpic.cn/sz_mmbiz_png/LvUIqvYKCeVMhL3qmYYHdLKliahnx99TgXoabiaM3fDTfVNKXSIZlicTFOTKRWPoLzlXfguzxRFic1Vfwc0hBiaqIsg/640?wx_fmt=png&amp;from=appmsg"></section><p data-tool="markdown.com.cn编辑器"><span leaf=""><span textstyle="">图：SeuratIntegrate 流程概览</span><span textstyle="">。可以看到它不但支持8种批次整合方法，也提供了用于比较整合结果的多种评估指标。</span></span></p><p data-tool="markdown.com.cn编辑器"><span leaf=""><span textstyle="">本文简要介绍SeuratIntegrate的使用，感兴趣的同学可以进一步阅读原文或下方链接</span>：</span></p><ul><li><section><span leaf="">https://cbib.github.io/Seurat-Integrate/articles/SeuratIntegrate.html</span></section></li></ul><h3 data-tool="markdown.com.cn编辑器"><span></span><span><span leaf="">00.R包安装</span></span><span></span></h3><p data-tool="markdown.com.cn编辑器"><span leaf=""><span textstyle="">可以使用以下代码安装该包</span>：</span></p><pre data-tool="markdown.com.cn编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">if</span></span><span leaf=""> (!</span><span><span leaf="">require</span></span><span leaf="">(</span><span><span leaf="">"BiocManager"</span></span><span leaf="">, quietly = </span><span><span leaf="">TRUE</span></span><span leaf="">))</span><span leaf=""><br></span><span leaf="">  install.packages(</span><span><span leaf="">"BiocManager"</span></span><span leaf="">)</span><span leaf=""><br></span><span><span leaf="">if</span></span><span leaf=""> (!</span><span><span leaf="">require</span></span><span leaf="">(</span><span><span leaf="">"remotes"</span></span><span leaf="">, quietly = </span><span><span leaf="">TRUE</span></span><span leaf="">))</span><span leaf=""><br></span><span leaf="">  install.packages(</span><span><span leaf="">"remotes"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">remotes::install_github(</span><span><span leaf="">"cbib/Seurat-Integrate"</span></span><span leaf="">, dependencies = </span><span><span leaf="">NA</span></span><span leaf="">, repos = BiocManager::repositories()) </span><span leaf=""><br></span></code></pre><p data-tool="markdown.com.cn编辑器"><span leaf=""><span textstyle="">安装以后调用所需的R包</span>：</span></p><pre data-tool="markdown.com.cn编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">library</span></span><span leaf="">(Seurat)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(SeuratIntegrate)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(dplyr)</span><span leaf=""><br></span></code></pre><h3 data-tool="markdown.com.cn编辑器"><span></span><span><span leaf="">01.使用示例</span></span><span></span></h3><p data-tool="markdown.com.cn编辑器"><span leaf=""><span textstyle="">作者提供了一个相对小的</span><span textstyle="">示例数据</span><span textstyle="">（200 个肝脏免疫细胞，~6.5K 基因）展示运行过程</span>：</span></p><pre data-tool="markdown.com.cn编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># 小型示例数据（200 个肝脏免疫细胞，~6.5K 基因）</span></span><span leaf=""><br></span><span leaf="">data(</span><span><span leaf="">"liver_small"</span></span><span leaf="">, package = </span><span><span leaf="">"SeuratIntegrate"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">dim(liver_small)  </span><span><span leaf=""># genes x cells</span></span><span leaf=""><br></span><span><span leaf=""># 6534 x 200</span></span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 数据包含 4 个 样本，来自 2 个研究（批次效应）；除此之外，还包含细胞类型标签等</span></span><span leaf=""><br></span><span leaf="">liver_small[[]] |&gt; dplyr::select(First_author, ID_sample, Tissue_type) |&gt; head()</span><span leaf=""><br></span><span leaf="">liver_small[[]][, c(</span><span><span leaf="">"predicted.celltype.l1"</span></span><span leaf="">,</span><span><span leaf="">"predicted.celltype.l2"</span></span><span leaf="">)] |&gt; head()</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 使用手工注释作为金标准进行性能评估，ID列作为批次变量</span></span><span leaf=""><br></span><span leaf="">cell.var  &lt;- </span><span><span leaf="">"manual_cell_type_short"</span></span><span leaf=""><br></span><span leaf="">batch.var &lt;- </span><span><span leaf="">"ID_sample"</span></span><span leaf=""><br></span></code></pre><p data-tool="markdown.com.cn编辑器"><span leaf=""><span textstyle="">① 环境配置</span>。使用</span><code><span leaf="">getCache()</span></code><span leaf="">函数可以查看当前配置好的环境：</span></p><pre data-tool="markdown.com.cn编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">getCache()</span><span leaf=""><br></span></code></pre><section nodeleaf=""><img data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/LvUIqvYKCeVMhL3qmYYHdLKliahnx99TgxNkkkjBoJGE35H8bRtt8pECsbXJNiakmj5DlmV4ajJM8FyibqWXmaWSA/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.334384858044164" data-s="300,640" data-type="png" data-w="951" type="block" data-imgfileid="100012801" src="https://mmbiz.qpic.cn/sz_mmbiz_png/LvUIqvYKCeVMhL3qmYYHdLKliahnx99TgxNkkkjBoJGE35H8bRtt8pECsbXJNiakmj5DlmV4ajJM8FyibqWXmaWSA/640?wx_fmt=png&amp;from=appmsg"></section><p data-tool="markdown.com.cn编辑器"><span leaf="">从上图可以看到，基于R的combat、harmony、mnn环境已经配置成功，但是基于Python的 BBKNN、scVI、scANVI、Scanorama、trVAE环境尚未配置。<span textstyle="">不过，SeuratIntegrate 提供了环境快速配置工具，具体代码如下</span>：</span></p><pre data-tool="markdown.com.cn编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># 仅需 BBKNN 时，下面代码自动准备 bbknn 的 conda 环境</span></span><span leaf=""><br></span><span leaf="">UpdateEnvCache(</span><span><span leaf="">"bbknn"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 其它 Python 方法（按需）：</span></span><span leaf=""><br></span><span><span leaf=""># UpdateEnvCache("scvi")</span></span><span leaf=""><br></span><span><span leaf=""># UpdateEnvCache("scanorama")</span></span><span leaf=""><br></span><span><span leaf=""># UpdateEnvCache("trVAE")</span></span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 创建指定 UMAP 版本的conda环境</span></span><span leaf=""><br></span><span leaf="">reticulate::conda_create(</span><span><span leaf="">"umap_0.5.4"</span></span><span leaf="">, packages = </span><span><span leaf="">"umap-learn=0.5.4"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 再次查看一下</span></span><span leaf=""><br></span><span leaf="">getCache()</span><span leaf=""><br></span></code></pre><section nodeleaf=""><img data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/LvUIqvYKCeVMhL3qmYYHdLKliahnx99TgibxLIjkgjvBhTSCIcfSmwYmPO13UbVAuibW54Tc5dYbNHK4jQUj9ZORg/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.34649122807017546" data-s="300,640" data-type="png" data-w="912" type="block" data-imgfileid="100012802" src="https://mmbiz.qpic.cn/sz_mmbiz_png/LvUIqvYKCeVMhL3qmYYHdLKliahnx99TgibxLIjkgjvBhTSCIcfSmwYmPO13UbVAuibW54Tc5dYbNHK4jQUj9ZORg/640?wx_fmt=png&amp;from=appmsg"></section><p data-tool="markdown.com.cn编辑器"><span leaf=""><span textstyle="">② 环境配置好以后，可以对数据进行预处理</span>。将seurat对象按批次拆分，以便每个批次独立归一化：</span></p><pre data-tool="markdown.com.cn编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">batch.var &lt;- </span><span><span leaf="">"ID_sample"</span></span><span leaf="">  </span><span><span leaf=""># save for later</span></span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">liver_small &lt;- split(liver_small, f = liver_small$ID_sample)</span><span leaf=""><br></span><span leaf="">liver_small &lt;- SCTransform(liver_small)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">liver_small &lt;- RunPCA(liver_small)</span><span leaf=""><br></span><span leaf="">liver_small &lt;- FindNeighbors(liver_small, dims = </span><span><span leaf="">1</span></span><span leaf="">:</span><span><span leaf="">20</span></span><span leaf="">, k.param = </span><span><span leaf="">15L</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">liver_small &lt;- RunUMAP(liver_small, dims = </span><span><span leaf="">1</span></span><span leaf="">:</span><span><span leaf="">20</span></span><span leaf="">, n.neighbors = </span><span><span leaf="">15</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 可视化一下批次效应</span></span><span leaf=""><br></span><span leaf="">DimPlot(liver_small, group.by = c(</span><span><span leaf="">"First_author"</span></span><span leaf="">, </span><span><span leaf="">"ID_sample"</span></span><span leaf="">))</span><span leaf=""><br></span></code></pre><section nodeleaf=""><img data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/LvUIqvYKCeVMhL3qmYYHdLKliahnx99Tg1X3oHGichFtEpPT0icTGwKa6ePSr9dsPwQzTkYHw3qWrI8Lu2Nmgy4bQ/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.4465909090909091" data-s="300,640" data-type="png" data-w="880" type="block" data-imgfileid="100012803" src="https://mmbiz.qpic.cn/sz_mmbiz_png/LvUIqvYKCeVMhL3qmYYHdLKliahnx99Tg1X3oHGichFtEpPT0icTGwKa6ePSr9dsPwQzTkYHw3qWrI8Lu2Nmgy4bQ/640?wx_fmt=png&amp;from=appmsg"></section><p data-tool="markdown.com.cn编辑器"><span leaf=""><span textstyle="">可视化手工注释结果（真实标签）</span>：</span></p><pre data-tool="markdown.com.cn编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">DimPlot(liver_small, group.by = cell.var)</span><span leaf=""><br></span></code></pre><section nodeleaf=""><img data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/LvUIqvYKCeVMhL3qmYYHdLKliahnx99TgDIXT9FEE5PweCASF5PYIFGWUH6zTUhMEicCoSpuXoG6hJXO9mZj2Ltg/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.7834394904458599" data-s="300,640" data-type="png" data-w="471" type="block" data-imgfileid="100012804" src="https://mmbiz.qpic.cn/sz_mmbiz_png/LvUIqvYKCeVMhL3qmYYHdLKliahnx99TgDIXT9FEE5PweCASF5PYIFGWUH6zTUhMEicCoSpuXoG6hJXO9mZj2Ltg/640?wx_fmt=png&amp;from=appmsg"></section><p data-tool="markdown.com.cn编辑器"><span leaf="">可以看到存在一定的批次效应，如果这样去走常规流程聚类，可能混淆真实的细胞标签以及批次信息。</span></p><p data-tool="markdown.com.cn编辑器"><span leaf=""><span textstyle="">③ 执行批次整合</span>。这里演示的方法分别是 ComBat（对counts去批次）、BBKNN（对 KNN 图去批次）、Harmony（对PCA嵌入去批次）。<span textstyle="">不同工具针对的去批次对象有所不同（对后续的结果会有影响）</span></span><span leaf="" data-pm-slice='1 1 ["para",{"tagName":"section","attributes":{"data-tool":"markdown编辑器","data-website":"https://markdown.com.cn/editor","style":"font-size: 16px;word-spacing: 0px;letter-spacing: 0px;word-break: break-word;word-wrap: break-word;text-align: justify;margin-top: -10px;line-height: 1.75;color: #3f3f3f;font-family: Optima-Regular, Optima, PingFangTC-Light, PingFangSC-light, PingFangTC-light;background-image: linear-gradient(90deg, rgba(50, 0, 0, 0.05) 3%, rgba(0, 0, 0, 0) 3%), linear-gradient(360deg, rgba(50, 0, 0, 0.05) 3%, rgba(0, 0, 0, 0) 3%);background-size: 20px 20px;background-position: center center;padding: 25px;","data-pm-slice":"0 0 []"},"namespaceURI":"http://www.w3.org/1999/xhtml"},"para",{"tagName":"p","attributes":{"data-tool":"markdown.com.cn编辑器","style":"padding-top: 8px;padding-bottom: 8px;margin: 10px 0px;letter-spacing: 1.5px;margin-top: 0px;margin-bottom: 0px;font-size: 16px;line-height: 2;color: #3f3f3f !important;"},"namespaceURI":"http://www.w3.org/1999/xhtml"}]'><span textstyle="">，还是建议大家搞懂原理，不然容易混淆</span></span><span leaf="">：</span></p><pre data-tool="markdown.com.cn编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">liver_small &lt;- DoIntegrate(</span><span leaf=""><br></span><span leaf="">  object = liver_small,</span><span leaf=""><br></span><span><span leaf=""># 选择要跑的方法（可按需增减）</span></span><span leaf=""><br></span><span leaf="">  CombatIntegration(), </span><span><span leaf=""># combat</span></span><span leaf=""><br></span><span leaf="">  bbknnIntegration(orig = </span><span><span leaf="">"pca"</span></span><span leaf="">, ndims.use = </span><span><span leaf="">20</span></span><span leaf="">), </span><span><span leaf=""># bbknn</span></span><span leaf=""><br></span><span leaf="">  SeuratIntegrate::HarmonyIntegration(orig = </span><span><span leaf="">"pca"</span></span><span leaf="">, dims = </span><span><span leaf="">1</span></span><span leaf="">:</span><span><span leaf="">20</span></span><span leaf="">), </span><span><span leaf=""># harmony</span></span><span leaf=""><br></span><span><span leaf=""># 其它配置</span></span><span leaf=""><br></span><span leaf="">  use.hvg    = </span><span><span leaf="">TRUE</span></span><span leaf="">,                 </span><span><span leaf=""># 使用高变基因</span></span><span leaf=""><br></span><span leaf="">  use.future = c(</span><span><span leaf="">FALSE</span></span><span leaf="">, </span><span><span leaf="">TRUE</span></span><span leaf="">, </span><span><span leaf="">FALSE</span></span><span leaf="">) </span><span><span leaf=""># 是否并行（与上面方法一一对应哈）</span></span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span><span leaf="">## Integration 1 in 3: integrating using 'CombatIntegration'</span></span><span leaf=""><br></span><span><span leaf="">## Integration 2 in 3: integrating using 'bbknnIntegration'</span></span><span leaf=""><br></span><span><span leaf="">## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span><span leaf=""><br></span><span><span leaf="">## </span></span><span leaf=""><br></span><span><span leaf="">## Number of nodes: 200</span></span><span leaf=""><br></span><span><span leaf="">## Number of edges: 1788</span></span><span leaf=""><br></span><span><span leaf="">## </span></span><span leaf=""><br></span><span><span leaf="">## Running Louvain algorithm...</span></span><span leaf=""><br></span><span><span leaf="">## Maximum modularity in 10 random starts: 0.6217</span></span><span leaf=""><br></span><span><span leaf="">## Number of communities: 5</span></span><span leaf=""><br></span><span><span leaf="">## Elapsed time: 0 seconds</span></span><span leaf=""><br></span><span><span leaf="">## Integration 3 in 3: integrating using 'SeuratIntegrate::HarmonyIntegration'</span></span><span leaf=""><br></span></code></pre><p data-tool="markdown.com.cn编辑器"><span leaf=""><span textstyle="">看看运行后的对象：</span></span></p><pre data-tool="markdown.com.cn编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">## An object of class Seurat </span></span><span leaf=""><br></span><span><span leaf="">## 22602 features across 200 samples within 4 assays </span></span><span leaf=""><br></span><span><span leaf="">## Active assay: SCT (6534 features, 3000 variable features)</span></span><span leaf=""><br></span><span><span leaf="">##  3 layers present: counts, data, scale.data</span></span><span leaf=""><br></span><span><span leaf="">##  3 other assays present: RNA, combat.reconstructed, bbknn.ridge</span></span><span leaf=""><br></span><span><span leaf="">##  4 dimensional reductions calculated: pca, umap, pca.bbknn, harmony</span></span><span leaf=""><br></span><span><span leaf="">## </span></span><span leaf=""><br></span><span><span leaf="">## #####</span></span><span leaf=""><br></span><span><span leaf="">## </span></span><span leaf=""><br></span><span><span leaf="">## [1] "SCT_nn"                              </span></span><span leaf=""><br></span><span><span leaf="">## [2] "SCT_snn"                             </span></span><span leaf=""><br></span><span><span leaf="">## [3] "bbknn_scale.data_connectivities"     </span></span><span leaf=""><br></span><span><span leaf="">## [4] "bbknn_scale.data_distances"          </span></span><span leaf=""><br></span><span><span leaf="">## [5] "bbknn_ridge.residuals_connectivities"</span></span><span leaf=""><br></span><span><span leaf="">## [6] "bbknn_ridge.residuals_distances"</span></span><span leaf=""><br></span></code></pre><p data-tool="markdown.com.cn编辑器"><span leaf=""><span textstyle="">④ 有了批次整合结果就可以做后续的聚类鉴定了，当然这里演示的是批次整合结果的评估</span>（已经有了手工注释的金标准）。需要注意，不同批次整合的方法需要分别处理，以便统一拿到 PCA 或者 KNN graph 等用于后续分析，大家可以参考如下处理方法：</span></p><pre data-tool="markdown.com.cn编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># 对于ComBat： -&gt; PCA -&gt; KNN -&gt; 聚类</span></span><span leaf=""><br></span><span leaf="">DefaultAssay(liver_small) &lt;- </span><span><span leaf="">"combat.reconstructed"</span></span><span leaf=""><br></span><span leaf="">VariableFeatures(liver_small) &lt;- VariableFeatures(liver_small[[</span><span><span leaf="">"SCT"</span></span><span leaf="">]])  </span><span><span leaf=""># 继承 HVGs</span></span><span leaf=""><br></span><span leaf="">liver_small &lt;- ScaleData(liver_small)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">liver_small &lt;- RunPCA(liver_small, npcs = </span><span><span leaf="">50L</span></span><span leaf="">, reduction.name = </span><span><span leaf="">"pca.combat"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">liver_small &lt;- FindNeighbors(</span><span leaf=""><br></span><span leaf="">  liver_small, reduction = </span><span><span leaf="">"pca.combat"</span></span><span leaf="">, dims = </span><span><span leaf="">1</span></span><span leaf="">:</span><span><span leaf="">20</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  return.neighbor = </span><span><span leaf="">TRUE</span></span><span leaf="">, graph.name = </span><span><span leaf="">"knn.combat"</span></span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">liver_small &lt;- SymmetrizeKnn(liver_small, graph.name = </span><span><span leaf="">"knn.combat"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">liver_small &lt;- FindOptimalClusters(</span><span leaf=""><br></span><span leaf="">  liver_small, graph.name = </span><span><span leaf="">"knn.combat_symmetric"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  cluster.name = glue::glue(</span><span><span leaf="">"combat_{cell.var}_{{metric}}"</span></span><span leaf="">),</span><span leaf=""><br></span><span leaf="">  cell.var = cell.var, optimisation.metric = c(</span><span><span leaf="">"nmi"</span></span><span leaf="">,</span><span><span leaf="">"ari"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">对于Harmony：-&gt; KNN -&gt; 聚类</span><span leaf=""><br></span><span leaf="">DefaultAssay(liver_small) &lt;- </span><span><span leaf="">"SCT"</span></span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 未整合基线</span></span><span leaf=""><br></span><span leaf="">liver_small &lt;- FindNeighbors(</span><span leaf=""><br></span><span leaf="">  liver_small, reduction = </span><span><span leaf="">"pca"</span></span><span leaf="">, dims = </span><span><span leaf="">1</span></span><span leaf="">:</span><span><span leaf="">20</span></span><span leaf="">, k.param = </span><span><span leaf="">20L</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  return.neighbor = </span><span><span leaf="">TRUE</span></span><span leaf="">, graph.name = </span><span><span leaf="">"knn.unintegrated"</span></span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">liver_small &lt;- SymmetrizeKnn(liver_small, graph.name = </span><span><span leaf="">"knn.unintegrated"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">liver_small &lt;- FindOptimalClusters(</span><span leaf=""><br></span><span leaf="">  liver_small, graph.name = </span><span><span leaf="">"knn.unintegrated_symmetric"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  cluster.name = glue::glue(</span><span><span leaf="">"unintegrated_{cell.var}_{{metric}}"</span></span><span leaf="">),</span><span leaf=""><br></span><span leaf="">  cell.var = cell.var</span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># Harmony 嵌入</span></span><span leaf=""><br></span><span leaf="">liver_small &lt;- FindNeighbors(</span><span leaf=""><br></span><span leaf="">  liver_small, reduction = </span><span><span leaf="">"harmony"</span></span><span leaf="">, dims = </span><span><span leaf="">1</span></span><span leaf="">:</span><span><span leaf="">20</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  return.neighbor = </span><span><span leaf="">TRUE</span></span><span leaf="">, graph.name = </span><span><span leaf="">"knn.harmony"</span></span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">liver_small &lt;- SymmetrizeKnn(liver_small, graph.name = </span><span><span leaf="">"knn.harmony"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">liver_small &lt;- FindOptimalClusters(</span><span leaf=""><br></span><span leaf="">  liver_small, graph.name = </span><span><span leaf="">"knn.harmony_symmetric"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  cluster.name = glue::glue(</span><span><span leaf="">"harmony_{cell.var}_{{metric}}"</span></span><span leaf="">),</span><span leaf=""><br></span><span leaf="">  cell.var = cell.var</span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 对于 graph 输出（BBKNN）： -&gt; 聚类</span></span><span leaf=""><br></span><span leaf="">liver_small &lt;- SymmetrizeKnn(liver_small, graph.name = </span><span><span leaf="">"bbknn_ridge.residuals_distances"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">liver_small &lt;- FindOptimalClusters(</span><span leaf=""><br></span><span leaf="">  liver_small, graph.name = </span><span><span leaf="">"bbknn_ridge.residuals_distances_symmetric"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  cell.var = cell.var, cluster.name = glue::glue(</span><span><span leaf="">"bbknn_{cell.var}_{{metric}}"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span></code></pre><p data-tool="markdown.com.cn编辑器"><span leaf=""><span textstyle="">三种批次整合方式的流程是不一样的，仔细看小编的注释（搞不明白的建议仔细搞懂单细胞分析的流程）。</span></span></p><p data-tool="markdown.com.cn编辑器"><span leaf=""><span textstyle="">⑤ 批次整合性能评估，有好几种评分方式。</span></span></p><p data-tool="markdown.com.cn编辑器"><span leaf=""><span textstyle="">无需 cell-type 变量</span><span textstyle="">（embedding 上）：</span></span></p><pre data-tool="markdown.com.cn编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># PCA 回归 / 密度（不支持 graph 输出）</span></span><span leaf=""><br></span><span leaf="">liver_small &lt;- AddScoreRegressPC(liver_small, </span><span><span leaf="">"unintegrated"</span></span><span leaf="">, batch.var, reduction = </span><span><span leaf="">"pca"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">liver_small &lt;- AddScoreRegressPC(liver_small, </span><span><span leaf="">"combat"</span></span><span leaf="">,       batch.var, reduction = </span><span><span leaf="">"pca.combat"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">liver_small &lt;- AddScoreRegressPC(liver_small, </span><span><span leaf="">"harmony"</span></span><span leaf="">,      batch.var, reduction = </span><span><span leaf="">"harmony"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">liver_small &lt;- AddScoreDensityPC(liver_small, </span><span><span leaf="">"unintegrated"</span></span><span leaf="">, batch.var, reduction = </span><span><span leaf="">"pca"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">liver_small &lt;- AddScoreDensityPC(liver_small, </span><span><span leaf="">"combat"</span></span><span leaf="">,       batch.var, reduction = </span><span><span leaf="">"pca.combat"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">liver_small &lt;- AddScoreDensityPC(liver_small, </span><span><span leaf="">"harmony"</span></span><span leaf="">,      batch.var, reduction = </span><span><span leaf="">"harmony"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 细胞周期回归（基于 PC）</span></span><span leaf=""><br></span><span leaf="">liver_small &lt;- CellCycleScoringPerBatch(liver_small, batch.var = batch.var,</span><span leaf=""><br></span><span leaf="">                                        s.features = cc.genes$s.genes,</span><span leaf=""><br></span><span leaf="">                                        g2m.features = cc.genes$g2m.genes)</span><span leaf=""><br></span><span leaf="">liver_small &lt;- AddScoreRegressPC.CellCycle(liver_small, </span><span><span leaf="">"unintegrated"</span></span><span leaf="">, batch.var, what = </span><span><span leaf="">"pca"</span></span><span leaf="">,        compute.cc = </span><span><span leaf="">FALSE</span></span><span leaf="">, dims.use = </span><span><span leaf="">1</span></span><span leaf="">:</span><span><span leaf="">20</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">liver_small &lt;- AddScoreRegressPC.CellCycle(liver_small, </span><span><span leaf="">"combat"</span></span><span leaf="">,       batch.var, what = </span><span><span leaf="">"pca.combat"</span></span><span leaf="">, compute.cc = </span><span><span leaf="">FALSE</span></span><span leaf="">, dims.use = </span><span><span leaf="">1</span></span><span leaf="">:</span><span><span leaf="">20</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">liver_small &lt;- AddScoreRegressPC.CellCycle(liver_small, </span><span><span leaf="">"harmony"</span></span><span leaf="">,      batch.var, what = </span><span><span leaf="">"harmony"</span></span><span leaf="">,    compute.cc = </span><span><span leaf="">FALSE</span></span><span leaf="">, dims.use = </span><span><span leaf="">1</span></span><span leaf="">:</span><span><span leaf="">20</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><p data-tool="markdown.com.cn编辑器"><span leaf=""><span textstyle="">基于 ASW</span>（需要 cell-type；不支持 graph）:</span></p><pre data-tool="markdown.com.cn编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">liver_small &lt;- AddScoreASW(liver_small, </span><span><span leaf="">"unintegrated"</span></span><span leaf="">, cell.var = cell.var, what = </span><span><span leaf="">"pca"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">liver_small &lt;- AddScoreASW(liver_small, </span><span><span leaf="">"combat"</span></span><span leaf="">,       cell.var = cell.var, what = </span><span><span leaf="">"pca.combat"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">liver_small &lt;- AddScoreASW(liver_small, </span><span><span leaf="">"harmony"</span></span><span leaf="">,      cell.var = cell.var, what = </span><span><span leaf="">"harmony"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># ASW.batch（可选：per.cell.var = FALSE 可不依赖 cell-type）</span></span><span leaf=""><br></span><span leaf="">liver_small &lt;- AddScoreASWBatch(liver_small, </span><span><span leaf="">"unintegrated"</span></span><span leaf="">, batch.var, cell.var, what = </span><span><span leaf="">"pca"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">liver_small &lt;- AddScoreASWBatch(liver_small, </span><span><span leaf="">"combat"</span></span><span leaf="">,       batch.var, cell.var, what = </span><span><span leaf="">"pca.combat"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">liver_small &lt;- AddScoreASWBatch(liver_small, </span><span><span leaf="">"harmony"</span></span><span leaf="">,      batch.var, cell.var, what = </span><span><span leaf="">"harmony"</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><p data-tool="markdown.com.cn编辑器"><span leaf=""><span textstyle="">基于图的指标</span>（需要 cell-type）</span></p><pre data-tool="markdown.com.cn编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># 图连通性</span></span><span leaf=""><br></span><span leaf="">liver_small &lt;- AddScoreConnectivity(liver_small, </span><span><span leaf="">"unintegrated"</span></span><span leaf="">, graph.name = </span><span><span leaf="">"knn.unintegrated_symmetric"</span></span><span leaf="">, cell.var = cell.var)</span><span leaf=""><br></span><span leaf="">liver_small &lt;- AddScoreConnectivity(liver_small, </span><span><span leaf="">"combat"</span></span><span leaf="">,       graph.name = </span><span><span leaf="">"knn.combat_symmetric"</span></span><span leaf="">,       cell.var = cell.var)</span><span leaf=""><br></span><span leaf="">liver_small &lt;- AddScoreConnectivity(liver_small, </span><span><span leaf="">"harmony"</span></span><span leaf="">,      graph.name = </span><span><span leaf="">"knn.harmony_symmetric"</span></span><span leaf="">,      cell.var = cell.var)</span><span leaf=""><br></span><span leaf="">liver_small &lt;- AddScoreConnectivity(liver_small, </span><span><span leaf="">"bbknn"</span></span><span leaf="">,        graph.name = </span><span><span leaf="">"bbknn_ridge.residuals_distances_symmetric"</span></span><span leaf="">, cell.var = cell.var)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># LISI（cLISI/iLISI；embedding 或 graph）</span></span><span leaf=""><br></span><span leaf="">liver_small &lt;- AddScoreLISI(liver_small, </span><span><span leaf="">"unintegrated"</span></span><span leaf="">, batch.var, cell.var, reduction = </span><span><span leaf="">"pca"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">liver_small &lt;- AddScoreLISI(liver_small, </span><span><span leaf="">"combat"</span></span><span leaf="">,       batch.var, cell.var, reduction = </span><span><span leaf="">"pca.combat"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">liver_small &lt;- AddScoreLISI(liver_small, </span><span><span leaf="">"harmony"</span></span><span leaf="">,      batch.var, cell.var, reduction = </span><span><span leaf="">"harmony"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">liver_small &lt;- AddScoreLISI(liver_small, </span><span><span leaf="">"bbknn"</span></span><span leaf="">,        batch.var, cell.var,</span><span leaf=""><br></span><span leaf="">                            reduction = </span><span><span leaf="">NULL</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">                            graph.name = </span><span><span leaf="">"bbknn_ridge.residuals_distances_symmetric"</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><p data-tool="markdown.com.cn编辑器"><span leaf=""><span textstyle="">细胞类型的分隔程度</span>（ARI / NMI；只需 cell-type + 聚类结果）</span></p><pre data-tool="markdown.com.cn编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># ARI</span></span><span leaf=""><br></span><span leaf="">liver_small &lt;- AddScoreARI(liver_small, </span><span><span leaf="">"unintegrated"</span></span><span leaf="">, cell.var, clust.var = paste(</span><span><span leaf="">"unintegrated"</span></span><span leaf="">, cell.var, </span><span><span leaf="">"ari"</span></span><span leaf="">, sep = </span><span><span leaf="">"_"</span></span><span leaf="">))</span><span leaf=""><br></span><span leaf="">liver_small &lt;- AddScoreARI(liver_small, </span><span><span leaf="">"combat"</span></span><span leaf="">,       cell.var, clust.var = paste(</span><span><span leaf="">"combat"</span></span><span leaf="">,       cell.var, </span><span><span leaf="">"ari"</span></span><span leaf="">, sep = </span><span><span leaf="">"_"</span></span><span leaf="">))</span><span leaf=""><br></span><span leaf="">liver_small &lt;- AddScoreARI(liver_small, </span><span><span leaf="">"harmony"</span></span><span leaf="">,      cell.var, clust.var = paste(</span><span><span leaf="">"harmony"</span></span><span leaf="">,      cell.var, </span><span><span leaf="">"ari"</span></span><span leaf="">, sep = </span><span><span leaf="">"_"</span></span><span leaf="">))</span><span leaf=""><br></span><span leaf="">liver_small &lt;- AddScoreARI(liver_small, </span><span><span leaf="">"bbknn"</span></span><span leaf="">,        cell.var, clust.var = paste(</span><span><span leaf="">"bbknn"</span></span><span leaf="">,        cell.var, </span><span><span leaf="">"ari"</span></span><span leaf="">, sep = </span><span><span leaf="">"_"</span></span><span leaf="">))</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># NMI</span></span><span leaf=""><br></span><span leaf="">liver_small &lt;- AddScoreNMI(liver_small, </span><span><span leaf="">"unintegrated"</span></span><span leaf="">, cell.var, clust.var = paste(</span><span><span leaf="">"unintegrated"</span></span><span leaf="">, cell.var, </span><span><span leaf="">"nmi"</span></span><span leaf="">, sep = </span><span><span leaf="">"_"</span></span><span leaf="">))</span><span leaf=""><br></span><span leaf="">liver_small &lt;- AddScoreNMI(liver_small, </span><span><span leaf="">"combat"</span></span><span leaf="">,       cell.var, clust.var = paste(</span><span><span leaf="">"combat"</span></span><span leaf="">,       cell.var, </span><span><span leaf="">"nmi"</span></span><span leaf="">, sep = </span><span><span leaf="">"_"</span></span><span leaf="">))</span><span leaf=""><br></span><span leaf="">liver_small &lt;- AddScoreNMI(liver_small, </span><span><span leaf="">"harmony"</span></span><span leaf="">,      cell.var, clust.var = paste(</span><span><span leaf="">"harmony"</span></span><span leaf="">,      cell.var, </span><span><span leaf="">"nmi"</span></span><span leaf="">, sep = </span><span><span leaf="">"_"</span></span><span leaf="">))</span><span leaf=""><br></span><span leaf="">liver_small &lt;- AddScoreNMI(liver_small, </span><span><span leaf="">"bbknn"</span></span><span leaf="">,        cell.var, clust.var = paste(</span><span><span leaf="">"bbknn"</span></span><span leaf="">,</span><span leaf=""><br></span></code></pre><p data-tool="markdown.com.cn编辑器"><span leaf=""><span textstyle="">⑥ 可视化结果，首先看看方法的性能：</span></span></p><pre data-tool="markdown.com.cn编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">liver_small &lt;- ScaleScores(liver_small)</span><span leaf=""><br></span><span leaf="">PlotScores(liver_small)</span><span leaf=""><br></span></code></pre><section nodeleaf=""><img data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/LvUIqvYKCeVMhL3qmYYHdLKliahnx99TgVZfltMH1jkkPbtoyAZpM0EwMtUtt9gyGGadKJzau6dFvVenU6Qn21w/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.682033096926714" data-s="300,640" data-type="png" data-w="846" type="block" data-imgfileid="100012807" src="https://mmbiz.qpic.cn/sz_mmbiz_png/LvUIqvYKCeVMhL3qmYYHdLKliahnx99TgVZfltMH1jkkPbtoyAZpM0EwMtUtt9gyGGadKJzau6dFvVenU6Qn21w/640?wx_fmt=png&amp;from=appmsg"></section><p data-tool="markdown.com.cn编辑器"><span leaf=""><span textstyle="">然后可以看看降维结果</span>：</span></p><pre data-tool="markdown.com.cn编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">liver_small &lt;- RunUMAP(liver_small, dims = </span><span><span leaf="">1</span></span><span leaf="">:</span><span><span leaf="">20</span></span><span leaf="">, reduction = </span><span><span leaf="">"pca.combat"</span></span><span leaf="">, reduction.name = </span><span><span leaf="">"umap.combat"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">liver_small &lt;- RunUMAP(liver_small, dims = </span><span><span leaf="">1</span></span><span leaf="">:</span><span><span leaf="">20</span></span><span leaf="">, reduction = </span><span><span leaf="">"harmony"</span></span><span leaf="">,    reduction.name = </span><span><span leaf="">"umap.harmony"</span></span><span leaf="">)</span><span leaf=""><br></span><span><span leaf=""># 对 BBKNN 使用其图直接跑 umap-learn</span></span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(future)</span><span leaf=""><br></span><span leaf="">plan(multisession)</span><span leaf=""><br></span><span leaf="">liver_small %&lt;-% {</span><span leaf=""><br></span><span leaf="">  reticulate::use_condaenv(</span><span><span leaf="">"umap_0.5.4"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">  RunUMAP(</span><span leaf=""><br></span><span leaf="">    liver_small,</span><span leaf=""><br></span><span leaf="">    graph = </span><span><span leaf="">"bbknn_ridge.residuals_connectivities"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">    umap.method = </span><span><span leaf="">"umap-learn"</span></span><span leaf="">, n.epochs = </span><span><span leaf="">200L</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">    reduction.name = </span><span><span leaf="">"umap.bbknn"</span></span><span leaf=""><br></span><span leaf="">  )</span><span leaf=""><br></span><span leaf="">}</span><span leaf=""><br></span><span leaf="">plan(sequential)</span><span leaf=""><br></span><span><span leaf=""># 看一下bbknn的结果</span></span><span leaf=""><br></span><span leaf="">DimPlot(liver_small, group.by = c(batch.var, cell.var), reduction = </span><span><span leaf="">"umap.bbknn"</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><section nodeleaf=""><img data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/LvUIqvYKCeVMhL3qmYYHdLKliahnx99TgKBjAPQDAK45xrIaNpicu6445gias7bIpSBjNL0l1oibKibRZmA6iaoq2ibkA/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.42857142857142855" data-s="300,640" data-type="png" data-w="889" type="block" data-imgfileid="100012808" src="https://mmbiz.qpic.cn/sz_mmbiz_png/LvUIqvYKCeVMhL3qmYYHdLKliahnx99TgKBjAPQDAK45xrIaNpicu6445gias7bIpSBjNL0l1oibKibRZmA6iaoq2ibkA/640?wx_fmt=png&amp;from=appmsg"></section><p data-pm-slice="8 3 []"><span><strong><span><span><strong><span><span><strong><span leaf=""><span textstyle="">可以</span></span></strong></span><span leaf=""><span textstyle="">测试一下</span></span></span></strong></span></span></strong></span></p><p><strong><span><span leaf=""><span textstyle="">至少支持的方法还可以</span></span><span leaf=""><br></span></span></strong></p><p><strong><span><strong><span><span><strong><span><span leaf=""><span textstyle="">简单分享到这 </span></span></span></strong><strong><span><span leaf=""><img data-src="https://res.wx.qq.com/t/wx_fed/we-emoji/res/v1.3.10/assets/newemoji/Yellowdog.png" data-ratio="1" data-w="128" src="https://res.wx.qq.com/t/wx_fed/we-emoji/res/v1.3.10/assets/newemoji/Yellowdog.png"></span></span></strong></span></span></strong></span></strong></p><p><strong><span><span leaf=""><span textstyle="">欢迎各位佬哥佬姐关注</span></span></span></strong></p><section nodeleaf=""><mp-common-profile data-pluginname="mpprofile" data-nickname="生信碱移" data-alias="liudoufu307" data-from="2" data-headimg="http://mmbiz.qpic.cn/mmbiz_png/LvUIqvYKCeXYZNMxRMnjiaicO2a27jDZ2FgQga8TdeQcsGRJRIn2IInkKtfcbbMXOBSViaPXpTOBulUlNzd11pzow/0?wx_fmt=png" data-signature="春来秋至，分享我的所见与所识" data-id="MzkyNTIzMzYyMA=="></mp-common-profile></section><section><span leaf=""><br></span></section><section><strong><span><span leaf="">END~</span></span></strong><span leaf=""><br></span></section><p><span><strong><span leaf="">仅供粉丝老铁们参考</span></strong></span></p><p><strong><span leaf="">如有侵权或错误，请联系删除改正～</span></strong></p><p><strong><span leaf=""><br></span></strong></p><section data-class="_mbEditor" data-id="32689"><section><section><section data-pm-slice='5 4 ["para",{"tagName":"section","attributes":{"data-class":"_mbEditor","data-id":"32689","style":"-webkit-tap-highlight-color: transparent;margin-bottom: 0px;outline: 0px;font-size: 16px;color: rgb(63, 63, 63);font-family: Optima-Regular, Optima, PingFangTC-Light, PingFangSC-light, PingFangTC-light;letter-spacing: 2px;background-color: rgb(255, 255, 255);"},"namespaceURI":"http://www.w3.org/1999/xhtml"},"para",{"tagName":"section","attributes":{"style":"-webkit-tap-highlight-color: transparent;outline: 0px;border-width: 0px;border-style: none;border-color: initial;"},"namespaceURI":"http://www.w3.org/1999/xhtml"},"para",{"tagName":"section","attributes":{"style":"-webkit-tap-highlight-color: transparent;outline: 0px;border-style: solid;-webkit-border-image: url(\"https://mmbiz.qpic.cn/mmbiz_png/b96CibCt70iaY62dey05FUCzFxmD5kYLSzMkskfzNG4hzY0V1ZcwXYEZxM9BiaYUTyd2xhLv2B8qrS54uy2mIyfxQ/640\") 94 170 140 fill;border-width: 20px 32px 40px 24px;font-size: 14px;line-height: 30px;"},"namespaceURI":"http://www.w3.org/1999/xhtml"}]'><p><span><strong mpa-from-tpl="t"><span leaf="">推荐好文，点击查看:</span></strong></span></p><p><span leaf=""><a target="_blank" href="https://mp.weixin.qq.com/s?__biz=MzkyNTIzMzYyMA==&amp;mid=2247496423&amp;idx=1&amp;sn=ed2aa25cd04563ce5bfb85674bc4fa9e&amp;scene=21#wechat_redirect" textvalue="【教程】还在做空间反卷积吗？抗收缩隐马尔可夫+伪细胞采样，这篇 NC 的算法直接实现单个细胞级别的空间转录组学分析！！" data-itemshowtype="0" linktype="text" data-linktype="2"><span textstyle="">【教程】无需注释即可筛选疾病相关细胞？！这篇纯生信 Science Advances 算法可以马上安装到 R 语言里了</span></a><span textstyle="">！</span></span></p><p><span><span leaf=""><a target="_blank" href="https://mp.weixin.qq.com/s?__biz=MzkyNTIzMzYyMA==&amp;mid=2247495672&amp;idx=1&amp;sn=5eb99652b90c34a044608f4e07109203&amp;scene=21#wechat_redirect" textvalue="【工具】Nature级别算法gsMap：空转+GWAS，将空间信息映射到人类复杂性状及疾病" data-itemshowtype="0" linktype="text" data-linktype="2"><span textstyle="">【工具】Nature级别算法gsMap: 空转+GWAS！将空间信息映射到人类复杂性状及疾病</span></a></span></span></p><p><span><span leaf=""><a target="_blank" href="https://mp.weixin.qq.com/s?__biz=MzkyNTIzMzYyMA==&amp;mid=2247496398&amp;idx=1&amp;sn=fc65c65645f05d76181ffcac96132b85&amp;scene=21#wechat_redirect" textvalue="【文献】这篇顶刊纯生信领先常规生信分析一个版本！" data-itemshowtype="0" linktype="text" data-linktype="2"><span textstyle="">【文献】发表一篇纯生信算法文章真的很难吗？！这篇 Communications Biology 的思路得赶紧学起来了</span></a><span textstyle="">！</span></span></span></p><p><span><span leaf=""><a target="_blank" href="http://mp.weixin.qq.com/s?__biz=MzkyNTIzMzYyMA==&amp;mid=2247494231&amp;idx=2&amp;sn=b3fb91ed7fa5d5775e16422cd502b0e8&amp;chksm=c1cb12ecf6bc9bfab457e75d1deb37045c9a86e9609e87f3707c38d9dd3083eefcf47689699b&amp;scene=21#wechat_redirect" textvalue="【期刊】无版面费低分 SCI 期刊，近期发表多篇纯网药‍研究！" data-itemshowtype="0" linktype="text" data-linktype="2"><span textstyle="">【期刊】无版面费低分 SCI 期刊，近期发表多篇纯网药研究！</span></a></span></span></p></section></section></section></section></section><p><mp-style-type data-value="3"></mp-style-type></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/Bj5gijBo3qdAZx6uRYYZ0A",target="_blank" rel="noopener noreferrer">原文链接</a>
