---
title: "基于Seurat的空转单样本数据分析流程学习(二)-SPOTlight"
date: 2025-09-07T01:22:51Z
draft: ["false"]
tags: [
  "fetched",
  "单细胞天地"
]
categories: ["Acdemic"]
---
基于Seurat的空转单样本数据分析流程学习(二)-SPOTlight by 单细胞天地
------
<div><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com" data-pm-slice="0 0 []"><p data-tool="mdnice编辑器"><span leaf="">上一篇推文采用了RCTD工具，<a target="_blank" href="https://mp.weixin.qq.com/s?__biz=MzI1Njk4ODE0MQ==&amp;mid=2247532214&amp;idx=1&amp;sn=e29ac93b2ba661de579ab2f5bf4b1fbf&amp;scene=21#wechat_redirect" textvalue="基于Seurat的空转单样本数据分析流程学习(一)" data-itemshowtype="0" linktype="text" data-linktype="2">基于Seurat的空转单样本数据分析流程学习(一)</a></span></p><p data-tool="mdnice编辑器"><span leaf="">本次我们使用SPOTlight映射，SPOTlight 是一种用于解析（去卷积）细胞类型及其在每个捕获位置中所占比例的工具，这些位置通常由多种细胞混合组成。该方法最初是为 10X Visium 空间转录组学技术开发的，但它同样适用于所有能够返回细胞混合信号的技术。SPOTlight 的核心原理是：通过 NMFreg 模型学习每种细胞类型的主题特征（topic profile signatures），然后寻找能够最佳拟合目标点（spot）的细胞类型组合。<img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjSuaNwoBo7cEMMyyZZG8Uhor2rpeIeicGKS26q3s5LpibklzX9MmTHBVicp98d3Ar02MUOfmMqpq46cQ/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.34448160535117056" data-type="png" data-w="598" data-imgfileid="100048627" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjSuaNwoBo7cEMMyyZZG8Uhor2rpeIeicGKS26q3s5LpibklzX9MmTHBVicp98d3Ar02MUOfmMqpq46cQ/640?wx_fmt=png&amp;from=appmsg"></span></p><h4 data-tool="mdnice编辑器"><span></span><span><span leaf="">分析流程</span></span><span></span></h4><p data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjSuaNwoBo7cEMMyyZZG8UhoX1qj0jyKGdDBwODmpuA3lPNiao5sY2zYlrmeuIqrvbfgIXbwJS7MzWA/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.18981481481481483" data-type="png" data-w="1080" data-imgfileid="100048629" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjSuaNwoBo7cEMMyyZZG8UhoX1qj0jyKGdDBwODmpuA3lPNiao5sY2zYlrmeuIqrvbfgIXbwJS7MzWA/640?wx_fmt=png&amp;from=appmsg"><img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjSuaNwoBo7cEMMyyZZG8UhoEl6HaF6yRYUScPEx8VfCAwhxBqnkjh8ecuKkLpTlfkI0oV0TX77MuA/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.275" data-type="png" data-w="1080" data-imgfileid="100048630" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjSuaNwoBo7cEMMyyZZG8UhoEl6HaF6yRYUScPEx8VfCAwhxBqnkjh8ecuKkLpTlfkI0oV0TX77MuA/640?wx_fmt=png&amp;from=appmsg"></span></p><ol><li><section><span leaf="">aligned_fiducials.jpg：带有芯片定位点（fiducial markers）的图像。这些定位点是Visium芯片上的参照标记，用来帮助显微镜图像与芯片的spot array对齐；在图像配准过程中保证组织切片位置与空间条码坐标准确匹配。</span></section></li><li><section><span leaf="">detected_tissue_image.jpg：组织识别后的二值图像。用于标记哪些区域包含组织（组织蒙版 mask），帮助区分组织区域和背景区域；常用于筛选有效的 spots。</span></section></li><li><section><span leaf="">filtered_feature_bc_matrix.h5：核心的基因表达矩阵文件（HDF5 格式）。行表示基因，列表示空间位置（spot/barcode），数值为UMI counts；该矩阵已去除低质量基因和低质量spot，是Seurat、Scanpy等空间分析的主要输入。</span></section></li><li><section><span leaf="">scalefactors_json.json：存放图像与空间坐标之间的缩放因子。主要字段包括高/低分辨率图像的缩放比例和spot 在原始图像上的直径；用于将spot坐标准确映射到组织切片图像上，保证可视化时位置匹配。</span></section></li><li><section><span leaf="">tissue_hires_image.png：高分辨率 H&amp;E 染色组织切片图像。用于在空间可视化中叠加基因表达信息，生成高质量的展示图像。</span></section></li><li><section><span leaf="">tissue_lowres_image.png：低分辨率 H&amp;E 染色组织切片图像。常用于快速加载和浏览，在交互式分析中默认使用，节省内存和计算资源。</span></section></li><li><section><span leaf="">tissue_positions_list.csv：每个 spot 的空间位置信息文件。包括 spot 的条码、是否位于组织区域（1=组织内，0=背景）、芯片阵列上的行列位置以及在图像中的像素坐标；用于确定每个 spot 在组织图像上的具体位置。</span></section></li></ol><p data-tool="mdnice编辑器"><span leaf="">在空间转录组分析中，必要的文件包括 filtered_feature_bc_matrix.h5（基因表达矩阵）、tissue_positions_list.csv（spot的空间位置信息）、scalefactors_json.json（坐标与图像的缩放比例）以及 tissue_lowres_image.png 或 tissue_hires_image.png（组织切片图像，用于可视化）。</span></p><h5 data-tool="mdnice编辑器"><span></span><span><span leaf="">1.导入</span></span><span></span></h5><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">rm(list = ls())</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(Seurat)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(future)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(ggplot2)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(hdf5r)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(SingleCellExperiment)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(SpatialExperiment)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(scater)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(scran)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">dir.create(</span><span><span leaf="">"0-GSM8633891-2"</span></span><span leaf="">)   </span><span leaf=""><br></span><span leaf="">setwd(</span><span><span leaf="">"0-GSM8633891-2"</span></span><span leaf="">)    </span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">把GSM8633891数据集整理成了这样的一个文件结构，并且把名称进行修改。由于空转数据基本上都是一个个处理，所以手动解压、改名、移动可能更快。<img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjSuaNwoBo7cEMMyyZZG8UhozaicB8tNRfcXNpplj3VCNicc2Rz0ozj0mTFOYeRhyW3I55NleDYMRv0g/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.5974842767295597" data-type="png" data-w="636" data-imgfileid="100048628" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjSuaNwoBo7cEMMyyZZG8UhozaicB8tNRfcXNpplj3VCNicc2Rz0ozj0mTFOYeRhyW3I55NleDYMRv0g/640?wx_fmt=png&amp;from=appmsg"></span></p><h5 data-tool="mdnice编辑器"><span></span><span><span leaf="">2.单样本读取数据</span></span><span></span></h5><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">data &lt;- Read10X_h5(</span><span><span leaf="">"./GSM8633891/filtered_feature_bc_matrix.h5"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">dim(data)</span><span leaf=""><br></span><span leaf="">object &lt;- CreateSeuratObject(counts = data, </span><span leaf=""><br></span><span leaf="">                             assay = </span><span><span leaf="">"Spatial"</span></span><span leaf="">, </span><span leaf=""><br></span><span leaf="">                             min.cells = </span><span><span leaf="">3</span></span><span leaf="">, </span><span><span leaf=""><a topic-id="mf7pcxfl-uajk67" data-topic="1">#过滤在少于3个细胞中表达的基因</a>，以减少低表达基因的干扰</span></span><span leaf=""><br></span><span leaf="">                             project = </span><span><span leaf="">"GSM8633891"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">object</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 再读取</span></span><span leaf=""><br></span><span leaf="">image &lt;- Read10X_Image(image.dir = </span><span><span leaf="">"GSM8633891/"</span></span><span leaf="">, </span><span leaf=""><br></span><span leaf="">                       image.name = </span><span><span leaf="">"tissue_hires_image.png"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">                       filter.matrix = </span><span><span leaf="">TRUE</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">image</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">dim(image)</span><span leaf=""><br></span><span leaf="">image &lt;- image[Cells(x = object)]</span><span><span leaf=""># 筛选图像对象中包含的SPOT</span></span><span leaf=""><br></span><span leaf="">DefaultAssay(object = image) &lt;- </span><span><span leaf="">"Spatial"</span></span><span><span leaf=""># 设置图像对象的默认assay为"Spatial"</span></span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 添加图片到object中</span></span><span leaf=""><br></span><span leaf="">object[[</span><span><span leaf="">"GSM8633891"</span></span><span leaf="">]] &lt;- image</span><span leaf=""><br></span><span leaf="">object@images[[</span><span><span leaf="">"GSM8633891"</span></span><span leaf="">]]@scale.factors$lowres &lt;- object@images[[</span><span><span leaf="">"GSM8633891"</span></span><span leaf="">]]@scale.factors$hires</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">sce.all &lt;- object</span><span leaf=""><br></span><span><span leaf="">## 简单探索一下数据结构</span></span><span leaf=""><br></span><span leaf="">as.data.frame(sce.all[[</span><span><span leaf="">"Spatial"</span></span><span leaf="">]]$counts[</span><span><span leaf="">1</span></span><span leaf="">:</span><span><span leaf="">4</span></span><span leaf="">,</span><span><span leaf="">1</span></span><span leaf="">:</span><span><span leaf="">4</span></span><span leaf="">])</span><span leaf=""><br></span><span leaf="">as.data.frame(LayerData(sce.all, assay = </span><span><span leaf="">"Spatial"</span></span><span leaf="">, layer = </span><span><span leaf="">"counts"</span></span><span leaf="">)[</span><span><span leaf="">1</span></span><span leaf="">:</span><span><span leaf="">5</span></span><span leaf="">,</span><span><span leaf="">1</span></span><span leaf="">:</span><span><span leaf="">5</span></span><span leaf="">])</span><span leaf=""><br></span><span leaf="">head(sce.all@meta.data)</span><span leaf=""><br></span><span leaf="">table(sce.all$orig.ident)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">Layers(sce.all)</span><span leaf=""><br></span><span leaf="">Assays(sce.all)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(qs)</span><span leaf=""><br></span><span leaf="">qsave(sce.all, file=</span><span><span leaf="">"GSM8633891/sce.all.qs"</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><h5 data-tool="mdnice编辑器"><span></span><span><span leaf="">3.数据预处理</span></span><span></span></h5><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">plot1 &lt;- VlnPlot(sce.all, </span><span leaf=""><br></span><span leaf="">                 features = c(</span><span><span leaf="">"nCount_Spatial"</span></span><span leaf="">,</span><span><span leaf="">"nFeature_Spatial"</span></span><span leaf="">), </span><span leaf=""><br></span><span leaf="">                 pt.size = </span><span><span leaf="">0.1</span></span><span leaf="">) + </span><span leaf=""><br></span><span leaf="">  NoLegend()</span><span leaf=""><br></span><span leaf="">plot1</span><span leaf=""><br></span><span leaf="">ggsave(filename = </span><span><span leaf="">"1-QC/VlnPlot.pdf"</span></span><span leaf="">, </span><span leaf=""><br></span><span leaf="">       width = </span><span><span leaf="">12</span></span><span leaf="">,height = </span><span><span leaf="">6</span></span><span leaf="">, plot = plot1)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf="">######</span></span><span leaf=""><br></span><span leaf="">plot2 &lt;- SpatialFeaturePlot(sce.all, </span><span leaf=""><br></span><span leaf="">                            features = </span><span><span leaf="">"nCount_Spatial"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">                            pt.size.factor = </span><span><span leaf="">2</span></span><span leaf="">) </span><span leaf=""><br></span><span leaf="">plot2</span><span leaf=""><br></span><span leaf="">ggsave(filename = </span><span><span leaf="">"1-QC/FeaturePlot_nCount.pdf"</span></span><span leaf="">, </span><span leaf=""><br></span><span leaf="">       width = </span><span><span leaf="">18</span></span><span leaf="">,height = </span><span><span leaf="">6</span></span><span leaf="">, plot = plot2)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf="">########</span></span><span leaf=""><br></span><span leaf="">plot3 &lt;- SpatialFeaturePlot(sce.all, </span><span leaf=""><br></span><span leaf="">                            features =</span><span><span leaf="">"nFeature_Spatial"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">                            pt.size.factor = </span><span><span leaf="">2</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">plot3</span><span leaf=""><br></span><span leaf="">ggsave(filename =</span><span><span leaf="">"1-QC/FeaturePlot_nFeature.pdf"</span></span><span leaf="">, </span><span leaf=""><br></span><span leaf="">       width = </span><span><span leaf="">18</span></span><span leaf="">,height = </span><span><span leaf="">6</span></span><span leaf="">, plot = plot3)</span><span leaf=""><br></span></code></pre><h5 data-tool="mdnice编辑器"><span></span><span><span leaf="">4.过滤SPOT</span></span><span></span></h5><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">## 过滤spot</span></span><span leaf=""><br></span><span leaf="">sce.all &lt;- subset(sce.all, subset = nFeature_Spatial &gt; </span><span><span leaf="">200</span></span><span leaf=""> )</span><span leaf=""><br></span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf="">### 计算线粒体/核糖体/特定基因集的比例</span></span><span leaf=""><br></span><span><span leaf=""><a topic-id="mf7pcxfl-besl8q" data-topic="1">#mito</a>_genes &lt;- grep("^MT-", rownames(sce.all),ignore.case = T, value = T)</span></span><span leaf=""><br></span><span><span leaf="">## 可能是13个线粒体基因</span></span><span leaf=""><br></span><span><span leaf=""><a topic-id="mf7pcxfl-fbli9o" data-topic="1">#print</a>(mito_genes)</span></span><span leaf=""><br></span><span><span leaf=""><a topic-id="mf7pcxfl-skw4jy" data-topic="1">#sce</a>.all &lt;- PercentageFeatureSet(sce.all, </span></span><span leaf=""><br></span><span><span leaf="">#                                features = mito_genes, </span></span><span leaf=""><br></span><span><span leaf="">#                                col.name = "percent_mito")</span></span><span leaf=""><br></span></code></pre><h5 data-tool="mdnice编辑器"><span></span><span><span leaf="">5.标准化</span></span><span></span></h5><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">sce.all &lt;- SCTransform(sce.all, assay =</span><span><span leaf="">"Spatial"</span></span><span leaf="">, verbose = </span><span><span leaf="">T</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">sce.all &lt;- RunPCA(sce.all, assay =</span><span><span leaf="">"SCT"</span></span><span leaf="">, verbose = </span><span><span leaf="">FALSE</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">sce.all &lt;- FindNeighbors(sce.all, reduction =</span><span><span leaf="">"pca"</span></span><span leaf="">, dims = </span><span><span leaf="">1</span></span><span leaf="">:</span><span><span leaf="">30</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">sce.all &lt;- FindClusters(sce.all, verbose = </span><span><span leaf="">FALSE</span></span><span leaf="">,resolution = </span><span><span leaf="">0.8</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">sce.all &lt;- RunUMAP(sce.all, reduction =</span><span><span leaf="">"pca"</span></span><span leaf="">, dims = </span><span><span leaf="">1</span></span><span leaf="">:</span><span><span leaf="">30</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">p1 &lt;- DimPlot(sce.all, </span><span leaf=""><br></span><span leaf="">              reduction =</span><span><span leaf="">"umap"</span></span><span leaf="">, </span><span leaf=""><br></span><span leaf="">              label = </span><span><span leaf="">TRUE</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">              label.size = </span><span><span leaf="">7</span></span><span leaf="">);p1</span><span leaf=""><br></span><span leaf="">ggsave(filename =</span><span><span leaf="">"1-QC/DimPlot.pdf"</span></span><span leaf="">, width = </span><span><span leaf="">9</span></span><span leaf="">,height = </span><span><span leaf="">6</span></span><span leaf="">, plot = p1)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">p2 &lt;- SpatialDimPlot(sce.all, label = </span><span><span leaf="">TRUE</span></span><span leaf="">, </span><span leaf=""><br></span><span leaf="">                     label.size = </span><span><span leaf="">3</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">                     pt.size.factor = </span><span><span leaf="">2</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">                     image.alpha = </span><span><span leaf="">0.6</span></span><span leaf="">);p2</span><span leaf=""><br></span><span leaf="">ggsave(filename =</span><span><span leaf="">"1-QC/SpatialDimPlot.pdf"</span></span><span leaf="">, width = </span><span><span leaf="">18</span></span><span leaf="">,height = </span><span><span leaf="">6</span></span><span leaf="">, plot = p2)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">p1|p2</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_jpg/siaia0BDGJdjSuaNwoBo7cEMMyyZZG8UhoS7zveJAQOUpNicSTrQwxDEdXK0sjibSsS3URx4aChsJWb4w06d4piaP2w/640?wx_fmt=jpeg&amp;from=appmsg" data-ratio="0.5046296296296297" data-type="jpeg" data-w="1080" data-imgfileid="100048631" src="https://mmbiz.qpic.cn/mmbiz_jpg/siaia0BDGJdjSuaNwoBo7cEMMyyZZG8UhoS7zveJAQOUpNicSTrQwxDEdXK0sjibSsS3URx4aChsJWb4w06d4piaP2w/640?wx_fmt=jpeg&amp;from=appmsg"></span></figure><h6 data-tool="mdnice编辑器"><span></span><span><span leaf="">5.1 修改空转数据的数据样式，从Seurat变成SpatialExperiment</span></span><span></span></h6><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># 修改空转数据的数据样式，从Seurat变成SpatialExperiment</span></span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(SpatialExperiment)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(SingleCellExperiment)</span><span leaf=""><br></span><span leaf="">sce &lt;- as.SingleCellExperiment(sce.all)</span><span leaf=""><br></span><span><span leaf=""># 检查 assay</span></span><span leaf=""><br></span><span leaf="">assayNames(sce)       </span><span><span leaf=""># 通常包含 counts 或 Spatial_counts</span></span><span leaf=""><br></span><span leaf="">assay(sce, </span><span><span leaf="">"counts"</span></span><span leaf="">)  </span><span><span leaf=""># 确认 counts 是否存在</span></span><span leaf=""><br></span><span><span leaf=""># 添加空间坐标（来自 Seurat object）</span></span><span leaf=""><br></span><span leaf="">spatial_coords &lt;- GetTissueCoordinates(sce.all)</span><span leaf=""><br></span><span><span leaf=""># 保留数值列</span></span><span leaf=""><br></span><span leaf="">spatial_coords_mat &lt;- as.matrix(spatial_coords[, c(</span><span><span leaf="">"x"</span></span><span leaf="">,</span><span><span leaf="">"y"</span></span><span leaf="">)])</span><span leaf=""><br></span><span><span leaf=""># 转换为SpatialExperiment</span></span><span leaf=""><br></span><span leaf="">spe &lt;- SpatialExperiment(</span><span leaf=""><br></span><span leaf="">  assays = assays(sce),           </span><span><span leaf=""># 保留 counts/logcounts</span></span><span leaf=""><br></span><span leaf="">  colData = colData(sce),         </span><span><span leaf=""># 保留 meta data</span></span><span leaf=""><br></span><span leaf="">  spatialCoords = as.matrix(spatial_coords_mat)  </span><span><span leaf=""># 添加空间坐标</span></span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">class(spe)              </span><span leaf=""><br></span></code></pre><h5 data-tool="mdnice编辑器"><span></span><span><span leaf="">6.基于现有单细胞数据集去映射</span></span><span></span></h5><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># 读取参考数据集</span></span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(qs)</span><span leaf=""><br></span><span leaf="">ref &lt;- qread(</span><span><span leaf="">".././0-reference/sce_celltype.qs"</span></span><span leaf="">)</span><span leaf=""><br></span><span><span leaf=""># 建议大家都不要用特殊符号</span></span><span leaf=""><br></span><span><span leaf=""># R语言不喜欢这些特殊符号</span></span><span leaf=""><br></span><span leaf="">a &lt;- ref@meta.data</span><span leaf=""><br></span><span leaf="">a$celltype &lt;- sub(</span><span><span leaf="">"/"</span></span><span leaf="">,</span><span><span leaf="">"_"</span></span><span leaf="">,a$celltype)</span><span leaf=""><br></span><span leaf="">ref@meta.data &lt;- a</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 该注释存储在对象元数据的“celltype”列中。</span></span><span leaf=""><br></span><span leaf="">DimPlot(ref, group.by = </span><span><span leaf="">"celltype"</span></span><span leaf="">, label = </span><span><span leaf="">TRUE</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 修改单细胞数据的数据格式</span></span><span leaf=""><br></span><span leaf="">sce &lt;- as.SingleCellExperiment(ref)</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjSuaNwoBo7cEMMyyZZG8UhofpiaROYO5qdZjQAOyia0MQ6roguSUVTd4F1O9TOqynvSP7T6uh5TOFeQ/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.8895833333333333" data-type="png" data-w="480" data-imgfileid="100048634" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjSuaNwoBo7cEMMyyZZG8UhofpiaROYO5qdZjQAOyia0MQ6roguSUVTd4F1O9TOqynvSP7T6uh5TOFeQ/640?wx_fmt=png&amp;from=appmsg"></span></figure><h5 data-tool="mdnice编辑器"><span></span><span><span leaf="">7.SPOTling分析前数据处理</span></span><span></span></h5><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># 研究者认为每个细胞身份的细胞数下采样至约100个并不会影响模型性能。</span></span><span leaf=""><br></span><span><span leaf=""># 每个细胞身份包含超过100个细胞时，仅能带来边际性的提升，但会显著增加计算时间和资源消耗。</span></span><span leaf=""><br></span><span><span leaf=""># 将基因集限制为每种细胞类型的标记基因，并结合最多3,000个高变基因，可以进一步优化性能和计算资源。</span></span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""><a topic-id="mf7pcxfl-7mnszs" data-topic="1">#特征选择</a></span></span><span leaf=""><br></span><span><span leaf=""><a topic-id="mf7pcxfl-46j008" data-topic="1">#第一步是获取每种细胞类型的标记基因</a>。</span></span><span leaf=""><br></span><span><span leaf=""><a topic-id="mf7pcxfl-xw4sh2" data-topic="1">#遵循OSCA中描述的标准化流程</a>。</span></span><span leaf=""><br></span><span><span leaf=""><a topic-id="mf7pcxfl-xwfudi" data-topic="1">#首先执行文库大小标准化</a>，以校正细胞特异性的偏差：</span></span><span leaf=""><br></span><span leaf="">sce &lt;- logNormCounts(sce)</span><span leaf=""><br></span></code></pre><h5 data-tool="mdnice编辑器"><span></span><span><span leaf="">8.获取重点基因</span></span><span></span></h5><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># 获取一个向量，用于指示哪些基因 既不是核糖体基因，也不是线粒体基因。</span></span><span leaf=""><br></span><span leaf="">genes &lt;- !grepl(pattern = </span><span><span leaf="">"^Rp[l|s]|Mt"</span></span><span leaf="">, x = rownames(sce))</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">dec &lt;- modelGeneVar(sce, subset.row = genes)</span><span leaf=""><br></span><span leaf="">plot(dec$mean, dec$total, xlab = </span><span><span leaf="">"Mean log-expression"</span></span><span leaf="">, ylab = </span><span><span leaf="">"Variance"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">curve(metadata(dec)$trend(x), col = </span><span><span leaf="">"blue"</span></span><span leaf="">, add = </span><span><span leaf="">TRUE</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 获取排名前3000的基因</span></span><span leaf=""><br></span><span leaf="">hvg &lt;- getTopHVGs(dec, n = </span><span><span leaf="">3000</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><h5 data-tool="mdnice编辑器"><span></span><span><span leaf="">9.获取每个细胞身份的标志基因</span></span><span></span></h5><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># 可以使用任何方法，只要它能返回一个权重，表示该基因对于该细胞类型的重要性。</span></span><span leaf=""><br></span><span><span leaf=""># 常见的指标包括 avgLogFC、AUC、pct.expressed、p-value 等。</span></span><span leaf=""><br></span><span leaf="">colLabels(sce) &lt;- colData(sce)$celltype </span><span><span leaf=""># 是否使用celltype要根据自己的数据集进行修改</span></span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 计算标记基因</span></span><span leaf=""><br></span><span leaf="">mgs &lt;- scoreMarkers(sce, subset.row = genes)</span><span leaf=""><br></span></code></pre><h5 data-tool="mdnice编辑器"><span></span><span><span leaf="">10.保留那些与每个细胞身份相关的基因</span></span><span></span></h5><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># 接下来，想保留那些与每个细胞身份相关的基因。</span></span><span leaf=""><br></span><span leaf="">mgs_fil &lt;- lapply(names(mgs), </span><span><span leaf="">function</span></span><span leaf="">(i) {</span><span leaf=""><br></span><span leaf="">    x &lt;- mgs[[i]]</span><span leaf=""><br></span><span leaf="">    </span><span><span leaf=""># 筛选并保留相关的标志基因（marker genes），即 AUC &gt; 0.8 的基因。</span></span><span leaf=""><br></span><span leaf="">    x &lt;- x[x$mean.AUC &gt; </span><span><span leaf="">0.8</span></span><span leaf="">, ]</span><span leaf=""><br></span><span leaf="">    </span><span><span leaf=""># 将基因按权重从高到低排序。</span></span><span leaf=""><br></span><span leaf="">    x &lt;- x[order(x$mean.AUC, decreasing = </span><span><span leaf="">TRUE</span></span><span leaf="">), ]</span><span leaf=""><br></span><span leaf="">    </span><span><span leaf=""># 向数据框中添加基因和簇（cluster）ID。</span></span><span leaf=""><br></span><span leaf="">    x$gene &lt;- rownames(x)</span><span leaf=""><br></span><span leaf="">    x$cluster &lt;- i</span><span leaf=""><br></span><span leaf="">    data.frame(x)</span><span leaf=""><br></span><span leaf="">})</span><span leaf=""><br></span><span leaf="">mgs_df &lt;- do.call(rbind, mgs_fil)</span><span leaf=""><br></span></code></pre><h5 data-tool="mdnice编辑器"><span></span><span><span leaf="">11.提取细胞</span></span><span></span></h5><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""><a topic-id="mf7pcxfl-almnwb" data-topic="1">#随机从每个细胞身份中选择最多100个细胞</a>。如果某个细胞类型少于 100 个细胞，则会使用该类型的全部细胞。对于生物学差异非常大的细胞身份，可以使用更少的细胞，因为它们的转录谱差异很大。而当细胞身份在转录水平上较为相似时，需要增加细胞数量以捕捉它们之间的生物学异质性。</span></span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 按细胞身份划分细胞索引。</span></span><span leaf=""><br></span><span leaf="">idx &lt;- split(seq(ncol(sce)), ref$celltype)</span><span leaf=""><br></span><span><span leaf=""># 对每个细胞身份最多采样100个细胞并进行子集选择。</span></span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 分析</span></span><span leaf=""><br></span><span leaf="">n_cells &lt;- </span><span><span leaf="">100</span></span><span leaf=""><br></span><span leaf="">cs_keep &lt;- lapply(idx, </span><span><span leaf="">function</span></span><span leaf="">(i) {</span><span leaf=""><br></span><span leaf="">    n &lt;- length(i)</span><span leaf=""><br></span><span leaf="">    </span><span><span leaf="">if</span></span><span leaf=""> (n &lt; n_cells)</span><span leaf=""><br></span><span leaf="">        n_cells &lt;- n</span><span leaf=""><br></span><span leaf="">    sample(i, n_cells)</span><span leaf=""><br></span><span leaf="">})</span><span leaf=""><br></span><span leaf="">sce &lt;- sce[, unlist(cs_keep)]</span><span leaf=""><br></span></code></pre><h5 data-tool="mdnice编辑器"><span></span><span><span leaf="">12.SPOTlight解卷积</span></span><span></span></h5><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">res &lt;- SPOTlight(</span><span leaf=""><br></span><span leaf="">    x = sce,</span><span leaf=""><br></span><span leaf="">    y = spe,</span><span leaf=""><br></span><span leaf="">    groups = as.character(sce$celltype),</span><span leaf=""><br></span><span leaf="">    mgs = mgs_df,</span><span leaf=""><br></span><span leaf="">    hvg = hvg,</span><span leaf=""><br></span><span leaf="">    weight_id = </span><span><span leaf="">"mean.AUC"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">    group_id = </span><span><span leaf="">"cluster"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">    gene_id = </span><span><span leaf="">"gene"</span></span><span leaf="">)  </span><span leaf=""><br></span></code></pre><h5 data-tool="mdnice编辑器"><span></span><span><span leaf="">13 可视化</span></span><span></span></h5><h6 data-tool="mdnice编辑器"><span></span><span><span leaf="">13.1 查看Topic profiles</span></span><span></span></h6><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""><a topic-id="mf7pcxfl-ajqk9f" data-topic="1">#查看Topic</a> profiles。通过设置 facet = FALSE，希望评估每个topic 签名对每个细胞身份的特异性。理想情况下，每个细胞身份都会对应一个独特的topic profile</span></span><span leaf=""><br></span><span leaf="">plotTopicProfiles(</span><span leaf=""><br></span><span leaf="">    x = mod,</span><span leaf=""><br></span><span leaf="">    y = sce$celltype,</span><span leaf=""><br></span><span leaf="">    facet = </span><span><span leaf="">FALSE</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">    min_prop = </span><span><span leaf="">0.01</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">    ncol = </span><span><span leaf="">1</span></span><span leaf="">) +</span><span leaf=""><br></span><span leaf="">    theme(aspect.ratio = </span><span><span leaf="">1</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjSuaNwoBo7cEMMyyZZG8Uho3Vklaka4DRvgX8ZRH4bcR5h7ExIfYhSDDSHGAzJpdboESJUepSBoLg/640?wx_fmt=png&amp;from=appmsg" data-ratio="1.3674418604651162" data-type="png" data-w="860" data-imgfileid="100048633" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjSuaNwoBo7cEMMyyZZG8Uho3Vklaka4DRvgX8ZRH4bcR5h7ExIfYhSDDSHGAzJpdboESJUepSBoLg/640?wx_fmt=png&amp;from=appmsg"></span></figure><h6 data-tool="mdnice编辑器"><span></span><span><span leaf="">13.2 Topic profiles分面权重值</span></span><span></span></h6><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">plotTopicProfiles(</span><span leaf=""><br></span><span leaf="">    x = mod,</span><span leaf=""><br></span><span leaf="">    y = sce$celltype,</span><span leaf=""><br></span><span leaf="">    facet = </span><span><span leaf="">TRUE</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">    min_prop = </span><span><span leaf="">0.01</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">    ncol = </span><span><span leaf="">3</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjSuaNwoBo7cEMMyyZZG8Uho3D5HFRPNs9Ran51ibNic5pOC5Tc6dVC9UKgkcyhI3d7clbWdA0GW30vA/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.9342592592592592" data-type="png" data-w="1080" data-imgfileid="100048636" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjSuaNwoBo7cEMMyyZZG8Uho3D5HFRPNs9Ran51ibNic5pOC5Tc6dVC9UKgkcyhI3d7clbWdA0GW30vA/640?wx_fmt=png&amp;from=appmsg"></span></figure><h6 data-tool="mdnice编辑器"><span></span><span><span leaf="">13.3 相关性分析</span></span><span></span></h6><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">plotCorrelationMatrix(mat)</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjSuaNwoBo7cEMMyyZZG8Uhog517al932VJltDakNrtwxBuCINBNiaqShfX1NibAG7nXWMS94rc9Swsw/640?wx_fmt=png&amp;from=appmsg" data-ratio="1.3674418604651162" data-type="png" data-w="860" data-imgfileid="100048635" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjSuaNwoBo7cEMMyyZZG8Uhog517al932VJltDakNrtwxBuCINBNiaqShfX1NibAG7nXWMS94rc9Swsw/640?wx_fmt=png&amp;from=appmsg"></span></figure><h6 data-tool="mdnice编辑器"><span></span><span><span leaf="">13.4 共定位</span></span><span></span></h6><p data-tool="mdnice编辑器"><span leaf="">现在已经知道每个spot中存在的细胞类型，就可以绘制一个表示空间相互作用的图。在图中，细胞类型之间的权越大，表示它们在同一个spot中出现的频率越高。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># 共定位（Co-localization）</span></span><span leaf=""><br></span><span leaf="">plotInteractions(mat, which = </span><span><span leaf="">"heatmap"</span></span><span leaf="">, metric = </span><span><span leaf="">"prop"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">plotInteractions(mat, which = </span><span><span leaf="">"heatmap"</span></span><span leaf="">, metric = </span><span><span leaf="">"jaccard"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">plotInteractions(mat, which = </span><span><span leaf="">"network"</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjSuaNwoBo7cEMMyyZZG8UholEKynvK9265LTpOyHdF2dxfvUPl2Ih9cnwsiaeUl9DbJtqe1iaTaGTuQ/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.3" data-type="png" data-w="1080" data-imgfileid="100048632" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjSuaNwoBo7cEMMyyZZG8UholEKynvK9265LTpOyHdF2dxfvUPl2Ih9cnwsiaeUl9DbJtqe1iaTaGTuQ/640?wx_fmt=png&amp;from=appmsg"></span></figure><h6 data-tool="mdnice编辑器"><span></span><span><span leaf="">13.5  散点饼图</span></span><span></span></h6><p data-tool="mdnice编辑器"><span leaf="">绘制每个SPOT中不同细胞的百分比饼图</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""><a topic-id="mf7pcxfl-dsbys2" data-topic="1">#还可以将每个</a> spot 中的细胞类型比例可视化为饼图的不同扇区。</span></span><span leaf=""><br></span><span leaf="">ct &lt;- colnames(mat)</span><span leaf=""><br></span><span leaf="">mat[mat &lt; </span><span><span leaf="">0.1</span></span><span leaf="">] &lt;- </span><span><span leaf="">0</span></span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 定义多种颜色</span></span><span leaf=""><br></span><span leaf="">paletteMartin &lt;- c(</span><span leaf=""><br></span><span leaf="">    </span><span><span leaf="">"<a topic-id="mf7pcxfl-lwfn7t" data-topic="1">#000000</a>"</span></span><span leaf="">, </span><span><span leaf="">"<a topic-id="mf7pcxfl-usc03y" data-topic="1">#004949</a>"</span></span><span leaf="">, </span><span><span leaf="">"<a topic-id="mf7pcxfl-dranm3" data-topic="1">#009292</a>"</span></span><span leaf="">, </span><span><span leaf="">"<a topic-id="mf7pcxfl-6zmewu" data-topic="1">#ff6db6</a>"</span></span><span leaf="">, </span><span><span leaf="">"<a topic-id="mf7pcxfl-1qasqj" data-topic="1">#ffb6db</a>"</span></span><span leaf="">, </span><span leaf=""><br></span><span leaf="">    </span><span><span leaf="">"<a topic-id="mf7pcxfl-bi5k9k" data-topic="1">#490092</a>"</span></span><span leaf="">, </span><span><span leaf="">"<a topic-id="mf7pcxfl-e9t138" data-topic="1">#006ddb</a>"</span></span><span leaf="">, </span><span><span leaf="">"<a topic-id="mf7pcxfl-7s1e56" data-topic="1">#b66dff</a>"</span></span><span leaf="">, </span><span><span leaf="">"<a topic-id="mf7pcxfl-vlb66f" data-topic="1">#6db6ff</a>"</span></span><span leaf="">, </span><span><span leaf="">"<a topic-id="mf7pcxfl-ic6x36" data-topic="1">#b6dbff</a>"</span></span><span leaf="">, </span><span leaf=""><br></span><span leaf="">    </span><span><span leaf="">"<a topic-id="mf7pcxfl-ivhpet" data-topic="1">#920000</a>"</span></span><span leaf="">, </span><span><span leaf="">"<a topic-id="mf7pcxfl-xb2jp3" data-topic="1">#924900</a>"</span></span><span leaf="">, </span><span><span leaf="">"<a topic-id="mf7pcxfl-q2ajp0" data-topic="1">#db6d00</a>"</span></span><span leaf="">, </span><span><span leaf="">"<a topic-id="mf7pcxfl-zwcnee" data-topic="1">#24ff24</a>"</span></span><span leaf="">, </span><span><span leaf="">"<a topic-id="mf7pcxfl-eg4hb0" data-topic="1">#ffff6d</a>"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">pal &lt;- colorRampPalette(paletteMartin)(length(ct))</span><span leaf=""><br></span><span leaf="">names(pal) &lt;- ct</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">plotSpatialScatterpie(</span><span leaf=""><br></span><span leaf="">    x = spatialCoords(spe),</span><span leaf=""><br></span><span leaf="">    y = mat,</span><span leaf=""><br></span><span leaf="">    cell_types = colnames(mat),</span><span leaf=""><br></span><span leaf="">    img = </span><span><span leaf="">F</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">    scatterpie_alpha = </span><span><span leaf="">1</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">    pie_scale = </span><span><span leaf="">0.4</span></span><span leaf="">) +</span><span leaf=""><br></span><span leaf="">    scale_fill_manual(</span><span leaf=""><br></span><span leaf="">        values = pal,</span><span leaf=""><br></span><span leaf="">        breaks = names(pal))</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">对比了一下RCTD的结果，稍有差别。<img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjSuaNwoBo7cEMMyyZZG8UhosQ0fexCbnQIgyqFlf38MtcdN5RqoKp7yojnYTJtKug9E4yBU1bho7A/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.9342592592592592" data-type="png" data-w="1080" data-imgfileid="100048642" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjSuaNwoBo7cEMMyyZZG8UhosQ0fexCbnQIgyqFlf38MtcdN5RqoKp7yojnYTJtKug9E4yBU1bho7A/640?wx_fmt=png&amp;from=appmsg"></span></p><h6 data-tool="mdnice编辑器"><span></span><span><span leaf="">13.6 残差法评估准确度</span></span><span></span></h6><p data-tool="mdnice编辑器"><span leaf="">还可以查看模型对每个spot预测的比例效果如何。通过观察每个spot平方和残差来评估，数值越高说明不准确度越高。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">spe$res_ss &lt;- res[[</span><span><span leaf="">2</span></span><span leaf="">]][colnames(spe)]</span><span leaf=""><br></span><span leaf="">xy &lt;- spatialCoords(spe)</span><span leaf=""><br></span><span leaf="">spe$x &lt;- xy[, </span><span><span leaf="">1</span></span><span leaf="">]</span><span leaf=""><br></span><span leaf="">spe$y &lt;- xy[, </span><span><span leaf="">2</span></span><span leaf="">]</span><span leaf=""><br></span><span leaf="">ggcells(spe, aes(y, x, color = res_ss)) +</span><span leaf=""><br></span><span leaf="">    geom_point() +</span><span leaf=""><br></span><span leaf="">    scale_color_viridis_c() +</span><span leaf=""><br></span><span leaf="">    coord_fixed() +</span><span leaf=""><br></span><span leaf="">    theme_bw()</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjSuaNwoBo7cEMMyyZZG8UhoxUMWwHfNzEMDCc5u4uCA4WHQ5AHQEOgXlsRgu84ycUYqJSxibYEDkNQ/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.9342592592592592" data-type="png" data-w="1080" data-imgfileid="100048641" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjSuaNwoBo7cEMMyyZZG8UhoxUMWwHfNzEMDCc5u4uCA4WHQ5AHQEOgXlsRgu84ycUYqJSxibYEDkNQ/640?wx_fmt=png&amp;from=appmsg"></span></figure><h4 data-tool="mdnice编辑器"><span></span><span><span leaf="">参考资料：</span></span><span></span></h4><ol><li><section><span leaf="">SPOTlight: seeded NMF regression to deconvolute spatial transcriptomics spots with single-cell transcriptomes. Nucleic Acids Res 49(9):e50.</span></section></li><li><section><span leaf="">SPOTight: https://github.com/MarcElosua/SPOTlight</span></section></li></ol></section><section><span leaf=""><br></span></section><p><mp-style-type data-value="3"></mp-style-type></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/GrZbAjyxXUzZXbK-lO8Fmw",target="_blank" rel="noopener noreferrer">原文链接</a>
