---
title: "给我那刚入门R的朋友写一个Visium HD数据分析教程"
date: 2025-09-25T23:46:30Z
draft: ["false"]
tags: [
  "fetched",
  "生信技能树"
]
categories: ["Acdemic"]
---
给我那刚入门R的朋友写一个Visium HD数据分析教程 by 生信技能树
------
<div><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com" data-pm-slice="0 0 []"><blockquote><p><span leaf="">我这个朋友其实很厉害，<span textstyle="">bash编程六的一匹</span>，但是呢他就是不学R。劝他学R的时候，我python还没学呢；劝他一起学python的时候，我R还没学呢！</span><strong><span leaf="">「出来混，迟早要还的，这就来了！」</span></strong><span leaf="">我python都学完了！现在他的工作需要他做一个空转Visium HD的数据分析，这下好了，python生态的SpatialData和R生态的seurat这下子都不会了！</span></p></blockquote><p data-tool="mdnice编辑器"><span leaf="">来写一个详细一点的教程吧~</span></p><p data-tool="mdnice编辑器"><span leaf="">Seurat之前支持分析10x Visium和SLIDE-seq的数据，现在已经可以支持 Visium HD了，该技术比之前的版本在空间分辨率上有了显著的提升。Visium HD数据是由在2微米×2微米的网格中标记的空间模式化的寡核苷酸生成的。然而，由于这种分辨率的数据较为稀疏，因此将相邻的网格合并在一起，以创建8微米和16微米的分辨率。</span><strong><span leaf="">「10x建议使用8微米网格的数据进行分析，但Seurat支持同时加载多种网格数据——并将它们存储在单个对象中作为多个检测。」</span></strong></p><p data-tool="mdnice编辑器"><span leaf="">在本教程中，作者提供了Seurat支持的用于分析Visium HD数据的一些空间工作流程的概述，特别是：</span></p><ul><li><section><span leaf="">无监督聚类</span></section></li><li><section><span leaf="">空间组织域的识别</span></section></li><li><section><span leaf="">子集空间区域</span></section></li><li><section><span leaf="">与scRNA-seq数据的整合</span></section></li><li><section><span leaf="">比较不同细胞类型的空间定位</span></section></li></ul><blockquote><p><span leaf="">注意：Visium HD是一种新的数据类型，我们预计会随着对空间数据分析方法的进一步测试而更新本教程。我们强烈鼓励用户探索不同的参数设置如何影响他们的结果，迭代地分析数据（并与生物学专家合作），并验证意外或令人惊讶的生物学发现。</span></p></blockquote><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">示例数据</span></span></h2><p data-tool="mdnice编辑器"><span leaf="">先安装一下各种包：</span></p><pre data-tool="mdnice编辑器"><code><span><span leaf=""># 需要安装两个包</span></span><span leaf=""><br></span><span><span leaf="">## 使用西湖大学的 Bioconductor镜像</span></span><span leaf=""><br></span><span leaf="">options(BioC_mirror=</span><span><span leaf="">"https://mirrors.westlake.edu.cn/bioconductor"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">options(</span><span><span leaf="">"repos"</span></span><span leaf="">=c(CRAN=</span><span><span leaf="">"https://mirrors.westlake.edu.cn/CRAN/"</span></span><span leaf="">))</span><span leaf=""><br></span><span><span leaf=""># 读取 Visium HD数据需要的包</span></span><span leaf=""><br></span><span leaf="">install.packages(</span><span><span leaf="">"hdf5r"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">install.packages(</span><span><span leaf="">"arrow"</span></span><span leaf="">) </span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">rm(list=ls())</span><span leaf=""><br></span><span leaf="">library(Seurat)</span><span leaf=""><br></span><span leaf="">library(ggplot2)</span><span leaf=""><br></span><span leaf="">library(patchwork)</span><span leaf=""><br></span><span leaf="">library(dplyr)</span><span leaf=""><br></span></code></pre><h4 data-tool="mdnice编辑器"><span></span><span leaf="">加载Visium HD数据</span><span></span></h4><p data-tool="mdnice编辑器"><span leaf="">本次的数据为小鼠大脑的Visium HD数据集：</span></p><ul><li><section><span leaf="">Visium HD小鼠大脑数据集可在此处下载 https://www.10xgenomics.com/datasets/visium-hd-cytassist-gene-expression-libraries-of-mouse-brain-he</span></section></li><li><section><span leaf="">Seurat可以将多个binning/分辨率存储在不同的检测中</span></section></li><li><section><span leaf="">bin.size参数指定要加载的分辨率（默认加载8和16微米的分辨率）</span></section></li><li><section><span leaf="">用户可以通过更改 assay 来切换分辨率</span></section></li></ul><pre data-tool="mdnice编辑器"><code><span><span leaf="">## 1.加载数据</span></span><span leaf=""><br></span><span leaf="">localdir &lt;- </span><span><span leaf="">"../data/Visium_HD_Mouse_Brain/"</span></span><span leaf=""><br></span><span leaf="">object &lt;- Load10X_Spatial(data.dir = localdir, bin.size = c(8, 16))</span><span leaf=""><br></span><span leaf="">object</span><span leaf=""><br></span><span leaf="">head(object@meta.data)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># Setting default assay changes between 8um and 16um binning</span></span><span leaf=""><br></span><span leaf="">Assays(object)</span><span leaf=""><br></span><span leaf="">DefaultAssay(object) &lt;- </span><span><span leaf="">"Spatial.008um"</span></span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">vln.plot &lt;- VlnPlot(object, features = </span><span><span leaf="">"nCount_Spatial.008um"</span></span><span leaf="">, pt.size = 0) + theme(axis.text = element_text(size = 4)) + NoLegend()</span><span leaf=""><br></span><span leaf="">count.plot &lt;- SpatialFeaturePlot(object, features = </span><span><span leaf="">"nCount_Spatial.008um"</span></span><span leaf="">) + theme(legend.position = </span><span><span leaf="">"right"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># note that many spots have very few counts, in-part</span></span><span leaf=""><br></span><span><span leaf=""># due to low cellular density in certain tissue regions</span></span><span leaf=""><br></span><span leaf="">vln.plot | count.plot</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 查看每个bin的表型</span></span><span leaf=""><br></span><span leaf="">head(object@meta.data)</span><span leaf=""><br></span><span><span leaf=""># 查看样本数</span></span><span leaf=""><br></span><span leaf="">table(object</span><span><span leaf="">$orig</span></span><span leaf="">.ident)</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">Visium_HD_Mouse_Brain的目录结构：</span></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxtgwQibOGbib7wECfB2SUQib3kPyyV3iactDd5BDNVXD87w4xRlm2cNzlYW2xdUzwo2Tt228rewGYBmA/640?wx_fmt=png&amp;from=appmsg" data-ratio="1.5122655122655122" data-type="png" data-w="693" data-imgfileid="100062257" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxtgwQibOGbib7wECfB2SUQib3kPyyV3iactDd5BDNVXD87w4xRlm2cNzlYW2xdUzwo2Tt228rewGYBmA/640?wx_fmt=png&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">现在出来的第一个图就跟官网教程中不一样，我那朋友就有点困惑：</span></p><blockquote><p><span leaf="">不一样的点在官网这里的小提琴是很多个，但是自己跑出来的只有一个。</span></p></blockquote><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_jpg/cZNhZQ6j4wxtgwQibOGbib7wECfB2SUQib36ib0YPIxE1GepK8SshTkGMOslGWYKs1trEZLsSorIicoBkU5RO4sxAlw/640?wx_fmt=jpeg&amp;from=appmsg" data-ratio="0.5185185185185185" data-type="jpeg" data-w="1080" data-imgfileid="100062259" src="https://mmbiz.qpic.cn/mmbiz_jpg/cZNhZQ6j4wxtgwQibOGbib7wECfB2SUQib36ib0YPIxE1GepK8SshTkGMOslGWYKs1trEZLsSorIicoBkU5RO4sxAlw/640?wx_fmt=jpeg&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">我的解释：</span></p><blockquote><p><span leaf="">这里的小提琴横坐标是样本数，官网那个是细胞注释后的亚群，我们才刚刚开始读取数据进来，什么都没做肯定没有细胞亚群结果的。这里展示的是样本，通过 table(object@meta.data$orig.ident) 可以看到只有一个样本s，所有就只有一个小提琴。</span></p></blockquote><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxtgwQibOGbib7wECfB2SUQib3zr35nrdibIo374ibs0Hrz5DmJibKqAUptscibHlPqjXypkBtnyyY8I36lQ/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.25462962962962965" data-type="png" data-w="1080" data-imgfileid="100062256" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxtgwQibOGbib7wECfB2SUQib3zr35nrdibIo374ibs0Hrz5DmJibKqAUptscibHlPqjXypkBtnyyY8I36lQ/640?wx_fmt=png&amp;from=appmsg"></span></figure><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">数据集标准化</span></span></h2><p data-tool="mdnice编辑器"><span leaf="">在本教程中，使用标准的对数标准化方法处理空间数据。针对空间数据的最佳标准化方法仍在开发和评估之中，可以阅读Phipson/Davis和Fan实验室的研究论文，以了解更多关于空间标准化的潜在注意事项。</span></p><ul><li><section><span leaf="">Phipson/Davis：https://genomebiology.biomedcentral.com/articles/10.1186/s13059-024-03241-7</span></section></li><li><section><span leaf="">Fan实验室：https://genomebiology.biomedcentral.com/articles/10.1186/s13059-024-03303-w</span></section></li></ul><pre data-tool="mdnice编辑器"><code><span><span leaf="">## 2.数据标准化</span></span><span leaf=""><br></span><span><span leaf=""># normalize both 8um and 16um bins</span></span><span leaf=""><br></span><span leaf="">DefaultAssay(object) &lt;- </span><span><span leaf="">"Spatial.008um"</span></span><span leaf=""><br></span><span leaf="">object &lt;- NormalizeData(object)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">DefaultAssay(object) &lt;- </span><span><span leaf="">"Spatial.016um"</span></span><span leaf=""><br></span><span leaf="">object &lt;- NormalizeData(object)</span><span leaf=""><br></span></code></pre><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">可视化基因表达</span></span></h2><p data-tool="mdnice编辑器"><span leaf="">调整</span><code><span leaf="">pt.size.factor</span></code><span leaf="">参数（默认设置为1.2）有助于在该高清数据集中可视化分子和组织学信息。</span></p><p data-tool="mdnice编辑器"><span leaf="">还可以调整形状（</span><code><span leaf="">shape</span></code><span leaf="">）和描边（</span><code><span leaf="">stroke</span></code><span leaf="">，轮廓）参数以优化可视化效果。</span></p><pre data-tool="mdnice编辑器"><code><span><span leaf="">## 3.基因表达可视化</span></span><span leaf=""><br></span><span><span leaf=""># switch spatial resolution to 16um from 8um</span></span><span leaf=""><br></span><span leaf="">DefaultAssay(object) &lt;- </span><span><span leaf="">"Spatial.016um"</span></span><span leaf=""><br></span><span leaf="">p1 &lt;- SpatialFeaturePlot(object, features = </span><span><span leaf="">"Rorb"</span></span><span leaf="">) + ggtitle(</span><span><span leaf="">"Rorb expression (16um)"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># switch back to 8um</span></span><span leaf=""><br></span><span leaf="">DefaultAssay(object) &lt;- </span><span><span leaf="">"Spatial.008um"</span></span><span leaf=""><br></span><span leaf="">p2 &lt;- SpatialFeaturePlot(object, features = </span><span><span leaf="">"Hpca"</span></span><span leaf="">) + ggtitle(</span><span><span leaf="">"Hpca expression (8um)"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">p1 | p2</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_jpg/cZNhZQ6j4wxtgwQibOGbib7wECfB2SUQib3bUNTjAicghPYLuqU0lvCAPpaBpLxxJ7dJKpOub2uykq06t8L8wF7Dww/640?wx_fmt=jpeg&amp;from=appmsg" data-ratio="0.5981481481481481" data-type="jpeg" data-w="1080" data-imgfileid="100062260" src="https://mmbiz.qpic.cn/mmbiz_jpg/cZNhZQ6j4wxtgwQibOGbib7wECfB2SUQib3bUNTjAicghPYLuqU0lvCAPpaBpLxxJ7dJKpOub2uykq06t8L8wF7Dww/640?wx_fmt=jpeg&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">这两个基因非常具有代表性，查一下功能：</span></p><p data-tool="mdnice编辑器"><strong><span leaf="">「Rorb」</span></strong><span leaf="">：Rorb在</span><strong><span leaf="">「大脑皮层」</span></strong><span leaf="">中高表达，特别是在皮层第四层神经元中，作为保守的皮质第四层神经元标志物</span></p><p data-tool="mdnice编辑器"><strong><span leaf="">「Hpca」</span></strong><span leaf="">：HPCA在中枢神经系统神经元中发挥重要作用，尤其是在</span><strong><span leaf="">「海马体」</span></strong><span leaf="">中。它可能参与调节神经元的兴奋性和突触传递</span></p><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">无监督聚类</span></span></h2><p data-tool="mdnice编辑器"><span leaf="">虽然标准的单细胞RNA测序（scRNA-seq）聚类工作流程也可以应用于空间数据集，但我们在处理Visium HD数据集时发现，Seurat v5的https://satijalab.org/seurat/articles/seurat5_sketch_analysis 工作流程表现出了更好的性能，尤其是在识别稀有且空间受限的细胞群体方面。</span></p><p data-tool="mdnice编辑器"><span leaf="">正如研究 https://www.nature.com/articles/s41587-023-01767-y 与 https://www.sciencedirect.com/science/article/pii/S2405471219301528 中所描述的那样，基于 seurat5_sketch_analysis 的分析旨在以一种保留稀有群体的方式“抽样”大型数据集。在这里，我们对Visium HD数据集进行 seurat5_sketch_analysis 处理，在抽样的细胞上进行聚类，然后将聚类标签投影回整个数据集。</span></p><pre data-tool="mdnice编辑器"><code><span><span leaf="">## 4.聚类</span></span><span leaf=""><br></span><span><span leaf=""># note that data is already normalized</span></span><span leaf=""><br></span><span leaf="">DefaultAssay(object) &lt;- </span><span><span leaf="">"Spatial.008um"</span></span><span leaf=""><br></span><span leaf="">object &lt;- FindVariableFeatures(object)</span><span leaf=""><br></span><span leaf="">object &lt;- ScaleData(object)</span><span leaf=""><br></span><span><span leaf=""># we select 50,0000 cells and create a new 'sketch' assay</span></span><span leaf=""><br></span><span leaf="">object &lt;- SketchData( object = object, ncells = 50000, method = </span><span><span leaf="">"LeverageScore"</span></span><span leaf="">, sketched.assay = </span><span><span leaf="">"sketch"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># switch analysis to sketched cells</span></span><span leaf=""><br></span><span leaf="">Assays(object)</span><span leaf=""><br></span><span leaf="">DefaultAssay(object) &lt;- </span><span><span leaf="">"sketch"</span></span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># perform clustering workflow</span></span><span leaf=""><br></span><span leaf="">object &lt;- FindVariableFeatures(object)</span><span leaf=""><br></span><span leaf="">object &lt;- ScaleData(object)</span><span leaf=""><br></span><span leaf="">object &lt;- RunPCA(object, assay = </span><span><span leaf="">"sketch"</span></span><span leaf="">, reduction.name = </span><span><span leaf="">"pca.sketch"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">object &lt;- FindNeighbors(object, assay = </span><span><span leaf="">"sketch"</span></span><span leaf="">, reduction = </span><span><span leaf="">"pca.sketch"</span></span><span leaf="">, dims = 1:50)</span><span leaf=""><br></span><span leaf="">object &lt;- FindClusters(object, cluster.name = </span><span><span leaf="">"seurat_cluster.sketched"</span></span><span leaf="">, resolution = 3)</span><span leaf=""><br></span><span leaf="">object &lt;- RunUMAP(object, reduction = </span><span><span leaf="">"pca.sketch"</span></span><span leaf="">, reduction.name = </span><span><span leaf="">"umap.sketch"</span></span><span leaf="">, return.model = T, dims = 1:50)</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">现在我们可以使用ProjectData函数，将从50,000个草图细胞中学习到的聚类标签以及降维结果（PCA和UMAP）投射到整个数据集。</span></p><p data-tool="mdnice编辑器"><span leaf="">在得到的结果对象中，对于所有细胞：</span></p><ul><li><section><span leaf="">聚类标签将存储在object$seurat_cluster.projected中</span></section></li><li><section><span leaf="">投射的PCA嵌入将存储在object[["pca.008um"]]中</span></section></li><li><section><span leaf="">投射的UMAP嵌入将存储在object[["umap.sketch"]]中</span></section></li></ul><pre data-tool="mdnice编辑器"><code><span leaf="">object &lt;- ProjectData(</span><span leaf=""><br></span><span leaf="">  object = object,</span><span leaf=""><br></span><span leaf="">  assay = </span><span><span leaf="">"Spatial.008um"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  full.reduction = </span><span><span leaf="">"full.pca.sketch"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  sketched.assay = </span><span><span leaf="">"sketch"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  sketched.reduction = </span><span><span leaf="">"pca.sketch"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  umap.model = </span><span><span leaf="">"umap.sketch"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  dims = 1:50,</span><span leaf=""><br></span><span leaf="">  refdata = list(seurat_cluster.projected = </span><span><span leaf="">"seurat_cluster.sketched"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">head(object@meta.data)</span><span leaf=""><br></span><span leaf="">table(object@meta.data</span><span><span leaf="">$seurat_cluster</span></span><span leaf="">.projected)</span><span leaf=""><br></span><span><span leaf=""># 聚类标签将存储在object$seurat_cluster.projected中</span></span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxtgwQibOGbib7wECfB2SUQib3S461iaIgabQAwJggriaXTVwyD0cgIlLMMD0pT27gstwzj8oWEK0pumYQ/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.2490740740740741" data-type="png" data-w="1080" data-imgfileid="100062258" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxtgwQibOGbib7wECfB2SUQib3S461iaIgabQAwJggriaXTVwyD0cgIlLMMD0pT27gstwzj8oWEK0pumYQ/640?wx_fmt=png&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">meta.data：是一个数据框，每一行代表一个bin区间，其余列都是这个bin区间的各种特征，比如nCount是bin区间中表达umi count数，nFeature是bin区间中表达的基因数等。</span></p><p data-tool="mdnice编辑器"><span leaf="">接着可视化草图细胞的聚类结果，以及整个数据集的投射聚类结果：</span></p><pre data-tool="mdnice编辑器"><code><span><span leaf=""># 可视化草图细胞的聚类结果，以及整个数据集的投射聚类结果</span></span><span leaf=""><br></span><span leaf="">DefaultAssay(object) &lt;- </span><span><span leaf="">"sketch"</span></span><span leaf=""><br></span><span leaf="">Idents(object) &lt;- </span><span><span leaf="">"seurat_cluster.sketched"</span></span><span leaf=""><br></span><span leaf="">p1 &lt;- DimPlot(object, reduction = </span><span><span leaf="">"umap.sketch"</span></span><span leaf="">, label = F) + ggtitle(</span><span><span leaf="">"Sketched clustering (50,000 cells)"</span></span><span leaf="">) + theme(legend.position = </span><span><span leaf="">"bottom"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">p1</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># switch to full dataset</span></span><span leaf=""><br></span><span leaf="">DefaultAssay(object) &lt;- </span><span><span leaf="">"Spatial.008um"</span></span><span leaf=""><br></span><span leaf="">Idents(object) &lt;- </span><span><span leaf="">"seurat_cluster.projected"</span></span><span leaf=""><br></span><span leaf="">p2 &lt;- DimPlot(object, reduction = </span><span><span leaf="">"full.umap.sketch"</span></span><span leaf="">, label = F) + ggtitle(</span><span><span leaf="">"Projected clustering (full dataset)"</span></span><span leaf="">) + theme(legend.position = </span><span><span leaf="">"bottom"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">p2</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">p1 | p2</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxtgwQibOGbib7wECfB2SUQib3AAWB0UGk11qPxaHibMysOuQ1KJzJtKIibNeRB3r7trgKwdVhtRJOuFOQ/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.662962962962963" data-type="png" data-w="1080" data-imgfileid="100062265" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxtgwQibOGbib7wECfB2SUQib3AAWB0UGk11qPxaHibMysOuQ1KJzJtKIibNeRB3r7trgKwdVhtRJOuFOQ/640?wx_fmt=png&amp;from=appmsg"></span></figure><pre data-tool="mdnice编辑器"><code><span leaf="">SpatialDimPlot(object, label = T, repel = T, label.size = 4)</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_jpg/cZNhZQ6j4wxtgwQibOGbib7wECfB2SUQib3dzg16pg5D14SKaSNlicK3bjsFt7XK4EZyibbomkoFl2M1Yusf1LJr0Ow/640?wx_fmt=jpeg&amp;from=appmsg" data-ratio="0.7379629629629629" data-type="jpeg" data-w="1080" data-imgfileid="100062261" src="https://mmbiz.qpic.cn/mmbiz_jpg/cZNhZQ6j4wxtgwQibOGbib7wECfB2SUQib3dzg16pg5D14SKaSNlicK3bjsFt7XK4EZyibbomkoFl2M1Yusf1LJr0Ow/640?wx_fmt=jpeg&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">聚类结果中有一些亚群非常具有空间组织区域代表性，有一些区域则比较混杂。</span></p><p data-tool="mdnice编辑器"><strong><span leaf="">「可以单独突出显示了几个聚类的空间定位：」</span></strong><span leaf="">当存在许多不同的聚类（其中一些在空间上是受限的，而另一些则是混合的）时，绘制所有聚类的空间位置可能会难以解释。</span></p><pre data-tool="mdnice编辑器"><code><span leaf="">Idents(object) &lt;- </span><span><span leaf="">"seurat_cluster.projected"</span></span><span leaf=""><br></span><span leaf="">cells &lt;- CellsByIdentities(object, idents = c(0, 4, 32, 34, 35))</span><span leaf=""><br></span><span leaf="">p &lt;- SpatialDimPlot(object,</span><span leaf=""><br></span><span leaf="">  cells.highlight = cells[setdiff(names(cells), </span><span><span leaf="">"NA"</span></span><span leaf="">)],</span><span leaf=""><br></span><span leaf="">  cols.highlight = c(</span><span><span leaf="">"<a topic-id="mfxvl621-bmntru" data-topic="1">#FFFF00</a>"</span></span><span leaf="">, </span><span><span leaf="">"grey50"</span></span><span leaf="">), facet.highlight = T, combine = T</span><span leaf=""><br></span><span leaf="">) + NoLegend()</span><span leaf=""><br></span><span leaf="">p</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxtgwQibOGbib7wECfB2SUQib3ODCTFdseukmDPFXbJP8XNUy92cvITvJVIflwIqx9BEK2UU5j0YTeoA/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.7055555555555556" data-type="png" data-w="1080" data-imgfileid="100062264" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxtgwQibOGbib7wECfB2SUQib3ODCTFdseukmDPFXbJP8XNUy92cvITvJVIflwIqx9BEK2UU5j0YTeoA/640?wx_fmt=png&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">这里突出展示的每个cluster非常具有代表性，可以结合小脑矢状面切片的解剖组织区域注释图进行对比查看。</span></p><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">亚群高表达基因</span></span></h2><pre data-tool="mdnice编辑器"><code><span><span leaf="">## 5.亚群top基因 ----</span></span><span leaf=""><br></span><span><span leaf=""># Crete downsampled object to make visualization either</span></span><span leaf=""><br></span><span leaf="">DefaultAssay(object)</span><span leaf=""><br></span><span leaf="">Idents(object) </span><span leaf=""><br></span><span leaf="">DefaultAssay(object) &lt;- </span><span><span leaf="">"Spatial.008um"</span></span><span leaf=""><br></span><span leaf="">Idents(object) &lt;- </span><span><span leaf="">"seurat_cluster.projected"</span></span><span leaf=""><br></span><span><span leaf=""># 每个亚群降采样到1000个</span></span><span leaf=""><br></span><span leaf="">object_subset &lt;- subset(object, cells = Cells(object[[</span><span><span leaf="">"Spatial.008um"</span></span><span leaf="">]]), downsample = 1000)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># Order clusters by similarity</span></span><span leaf=""><br></span><span leaf="">DefaultAssay(object_subset) &lt;- </span><span><span leaf="">"Spatial.008um"</span></span><span leaf=""><br></span><span leaf="">Idents(object_subset) &lt;- </span><span><span leaf="">"seurat_cluster.projected"</span></span><span leaf=""><br></span><span leaf="">object_subset &lt;- BuildClusterTree(object_subset, assay = </span><span><span leaf="">"Spatial.008um"</span></span><span leaf="">, reduction = </span><span><span leaf="">"full.pca.sketch"</span></span><span leaf="">, reorder = T)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">markers &lt;- FindAllMarkers(object_subset, assay = </span><span><span leaf="">"Spatial.008um"</span></span><span leaf="">, only.pos = TRUE)</span><span leaf=""><br></span><span leaf="">top5 &lt;- markers %&gt;%</span><span leaf=""><br></span><span leaf="">  group_by(cluster) %&gt;%</span><span leaf=""><br></span><span leaf="">  dplyr::filter(avg_log2FC &gt; 1) %&gt;%</span><span leaf=""><br></span><span leaf="">  slice_head(n = 3) %&gt;%</span><span leaf=""><br></span><span leaf="">  ungroup()</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">object_subset &lt;- ScaleData(object_subset, assay = </span><span><span leaf="">"Spatial.008um"</span></span><span leaf="">, features = top5</span><span><span leaf="">$gene</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">p &lt;- DoHeatmap(object_subset, assay = </span><span><span leaf="">"Spatial.008um"</span></span><span leaf="">, features = top5</span><span><span leaf="">$gene</span></span><span leaf="">, size = 2.5) + theme(axis.text = element_text(size = 5.5)) + NoLegend()</span><span leaf=""><br></span><span leaf="">p</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxtgwQibOGbib7wECfB2SUQib3bG54Z1qFHhiceNEVvGlHdmU9XNzBmHaOVQSGGh56bLspyu4iaGK9Xx7Q/640?wx_fmt=png&amp;from=appmsg" data-ratio="1.2546296296296295" data-type="png" data-w="1080" data-imgfileid="100062263" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxtgwQibOGbib7wECfB2SUQib3bG54Z1qFHhiceNEVvGlHdmU9XNzBmHaOVQSGGh56bLspyu4iaGK9Xx7Q/640?wx_fmt=png&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">每一行为一个基因，每一列为一个bin区间，这里subsample每个亚群取了1000个bins进行展示。</span></p><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">识别空间定义的组织域</span></span></h2><p data-tool="mdnice编辑器"><span leaf="">虽然之前的分析是独立考虑每个单元格的，但空间数据使得细胞不仅能通过其邻域来定义，还能通过其更广泛的空间背景来定义。</span></p><p data-tool="mdnice编辑器"><span leaf="">在Singhal等人的研究中，作者介绍了BANKSY（Building Aggregates with a Neighborhood Kernel and Spatial Yardstick）。BANKSY执行多种任务，但我们发现它在识别和分割组织域方面特别有价值。在进行聚类时，BANKSY通过添加一个单元格更广泛邻域内基因表达水平的均值和梯度，来增强一个单元格的表达模式。</span></p><p data-tool="mdnice编辑器"><span leaf="">我们感谢作者们通过 SeuratWrappers 框架使 BANKSY 与 Seurat 兼容，这需要单独安装BANKSY包：</span></p><pre data-tool="mdnice编辑器"><code><span><span leaf="">if</span></span><span leaf=""> (!requireNamespace(</span><span><span leaf="">"Banksy"</span></span><span leaf="">, quietly = TRUE)) {</span><span leaf=""><br></span><span leaf="">  remotes::install_github(</span><span><span leaf="">"prabhakarlab/Banksy@devel"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">}</span><span leaf=""><br></span><span leaf="">library(SeuratWrappers)</span><span leaf=""><br></span><span leaf="">packageVersion(</span><span><span leaf="">"SeuratWrappers"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">library(Banksy)</span><span leaf=""><br></span><span leaf="">packageVersion(</span><span><span leaf="">"Banksy"</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><h4 data-tool="mdnice编辑器"><span></span><span leaf="">在运行BANKSY之前，用户需要考虑两个重要的模型参数：</span><span></span></h4><ul><li><section><strong><span leaf="">「k_geom」</span></strong><span leaf="">：局部邻域大小。较大的值会产生较大的域。</span></section></li><li><section><strong><span leaf="">「lambda」</span></strong><span leaf="">：邻域的影响。较大的值会产生更在空间上更连贯的域。</span></section></li></ul><p data-tool="mdnice编辑器"><strong><span leaf="">「RunBanksy」</span></strong><span leaf=""> 函数会创建一个新的BANKSY分析模块，该模块可用于降维和聚类：</span></p><pre data-tool="mdnice编辑器"><code><span leaf="">object &lt;- RunBanksy(object,</span><span leaf=""><br></span><span leaf="">                    lambda = 0.8, verbose = TRUE,</span><span leaf=""><br></span><span leaf="">                    assay = </span><span><span leaf="">"Spatial.008um"</span></span><span leaf="">, slot = </span><span><span leaf="">"data"</span></span><span leaf="">, features = </span><span><span leaf="">"variable"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">                    k_geom = 50</span><span leaf=""><br></span><span leaf="">                    )</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 就上面的结果降为聚类分群</span></span><span leaf=""><br></span><span leaf="">DefaultAssay(object) &lt;- </span><span><span leaf="">"BANKSY"</span></span><span leaf=""><br></span><span leaf="">object &lt;- RunPCA(object, assay = </span><span><span leaf="">"BANKSY"</span></span><span leaf="">, reduction.name = </span><span><span leaf="">"pca.banksy"</span></span><span leaf="">, features = rownames(object), npcs = 30)</span><span leaf=""><br></span><span leaf="">object &lt;- FindNeighbors(object, reduction = </span><span><span leaf="">"pca.banksy"</span></span><span leaf="">, dims = 1:30)</span><span leaf=""><br></span><span leaf="">object &lt;- FindClusters(object, cluster.name = </span><span><span leaf="">"banksy_cluster"</span></span><span leaf="">, resolution = 0.5)</span><span leaf=""><br></span><span leaf="">Idents(object) &lt;- </span><span><span leaf="">"banksy_cluster"</span></span><span leaf=""><br></span><span leaf="">p &lt;- SpatialDimPlot(object, group.by = </span><span><span leaf="">"banksy_cluster"</span></span><span leaf="">, label = T, repel = T, label.size = 4)</span><span leaf=""><br></span><span leaf="">p</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">运行 RunBanksy 报错：</span></p><blockquote><pre><code><span leaf="">Fetching data from slot data from assay Spatial.008um</span><span leaf=""><br></span><span leaf="">Subsetting by features</span><span leaf=""><br></span><span leaf="">Computing neighbor matrix</span><span leaf=""><br></span><span leaf="">错误于RunBanksy(object, lambda = 0.8, verbose = TRUE, assay = </span><span><span leaf="">"Spatial.008um"</span></span><span leaf="">, : </span><span leaf=""><br></span><span leaf="">  找不到对象</span><span><span leaf="">'compute.banksyMatrices'</span></span><span leaf=""><br></span></code></pre></blockquote><p data-tool="mdnice编辑器"><span leaf="">解决办法：https://github.com/prabhakarlab/Banksy/issues/26</span></p><p data-tool="mdnice编辑器"><span leaf="">解决好了重新运行上面的代码就可以了，空间域聚类结果如下：</span></p><p data-tool="mdnice编辑器"><span leaf="">也可以高亮某个domain：</span></p><pre data-tool="mdnice编辑器"><code><span leaf="">banksy_cells &lt;- CellsByIdentities(object)</span><span leaf=""><br></span><span leaf="">p &lt;- SpatialDimPlot(object, cells.highlight = banksy_cells[setdiff(names(banksy_cells), </span><span><span leaf="">"NA"</span></span><span leaf="">)], cols.highlight = c(</span><span><span leaf="">"<a topic-id="mfxvl621-6es201" data-topic="1">#FFFF00</a>"</span></span><span leaf="">, </span><span><span leaf="">"grey50"</span></span><span leaf="">), facet.highlight = T, combine = T) + NoLegend()</span><span leaf=""><br></span><span leaf="">p</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxtgwQibOGbib7wECfB2SUQib3T6NMZZlpLfJWnsPYGcmIibXrQwZia6CyaHQDwewo0MEO1mxzbESO4VNA/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.9324074074074075" data-type="png" data-w="1080" data-imgfileid="100062262" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxtgwQibOGbib7wECfB2SUQib3T6NMZZlpLfJWnsPYGcmIibXrQwZia6CyaHQDwewo0MEO1mxzbESO4VNA/640?wx_fmt=png&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">到这里基本分析就做完了，那后面的细胞类型鉴定是非常难的，这里先不放了，而且这里用到的都是低分辨率的那一套做法：使用单细胞数据运行RCTD去卷积，对于这种高分辨率的更推荐去试试细胞分割算法。</span></p><p data-tool="mdnice编辑器"><span leaf="">今天分享到这~</span></p></section><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com" data-pm-slice="0 0 []"><p data-tool="mdnice编辑器"><span leaf="">友情转发：</span></p></section><ul><li><section><span leaf=""><a target="_blank" href="https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247545889&amp;idx=1&amp;sn=b7b37a458eead4645137126753d58c34&amp;scene=21#wechat_redirect" textvalue="生信入门&amp;数据挖掘线上直播课10月班" data-itemshowtype="0" linktype="text" data-linktype="2"><span textstyle="">生信入门&amp;数据挖掘线上直播课10月班</span></a><span textstyle="">，你的生物信息学入门课</span></span></section></li><li><section><span leaf=""><a target="_blank" href="https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247525079&amp;idx=1&amp;sn=0b997af16a58195b4192691373048fd5&amp;scene=21#wechat_redirect" textvalue="时隔5年，我们的生信技能树VIP学徒继续招生啦" data-itemshowtype="0" linktype="text" data-linktype="2"><span textstyle="">时隔5年，我们的生信技能树VIP学徒继续招生啦</span></a></span></section></li><li><section><span leaf=""><a target="_blank" href="https://mp.weixin.qq.com/s?__biz=MzUzMTEwODk0Ng==&amp;mid=2247530048&amp;idx=1&amp;sn=28aa7bbd5e00521f79e074496a5f5d66&amp;scene=21#wechat_redirect" textvalue="满足你生信分析计算需求的低价解决方案" data-itemshowtype="0" linktype="text" data-linktype="2">满足你生信分析计算需求的低价解决方案</a></span></section></li><li><section><span leaf=""><a target="_blank" href="https://mp.weixin.qq.com/mp/appmsgalbum?__biz=MzAxMDkxODM1Ng==&amp;action=getalbum&amp;album_id=1679199708449144836&amp;scene=173&amp;subscene=207&amp;sessionid=1745492310&amp;enterid=1745492314&amp;from_msgid=2247541298&amp;from_itemidx=1&amp;count=3&amp;nolastread=1#wechat_redirect" textvalue="生信故事会" data-itemshowtype="0" linktype="text" data-linktype="2">生信故事会</a>，来看看他们的生信入门故事</span></section></li><li><section><span leaf=""><a target="_blank" href="https://mp.weixin.qq.com/mp/appmsgalbum?__biz=MzAxMDkxODM1Ng==&amp;action=getalbum&amp;album_id=3690970204957147140&amp;scene=173&amp;subscene=207&amp;sessionid=1745066271&amp;enterid=1745066274&amp;from_msgid=2247540702&amp;from_itemidx=1&amp;count=3&amp;nolastread=1#wechat_redirect" textvalue="生信马拉松答疑专辑" data-itemshowtype="0" linktype="text" data-linktype="2">生信马拉松答疑专辑</a>，获取你的生信专属答疑</span></section></li></ul><p><mp-style-type data-value="3"></mp-style-type></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/TDgVsP9w4LTF4mJLXR6MoQ",target="_blank" rel="noopener noreferrer">原文链接</a>
