---
title: "ç©ºè½¬æ•°æ®åˆ†æä¹‹ç”Ÿæ€ä½èšç±»ç®—æ³•Banksy"
date: 2024-05-05T01:09:01Z
draft: ["false"]
tags: [
  "fetched",
  "ç”Ÿä¿¡éšç¬”"
]
categories: ["Acdemic"]
---
ç©ºè½¬æ•°æ®åˆ†æä¹‹ç”Ÿæ€ä½èšç±»ç®—æ³•Banksy by ç”Ÿä¿¡éšç¬”
------
<div><section data-tool="mdniceç¼–è¾‘å™¨" data-website="https://www.mdnice.com"><p data-tool="mdniceç¼–è¾‘å™¨">ä¼ ç»Ÿçš„å•ç»†èƒèšç±»ç®—æ³•ï¼Œä¾‹å¦‚Seuratå’ŒScanpy Pipelineä¸­çš„Louvainå’ŒLeidenç­‰èšç±»ç®—æ³•ï¼Œé€šå¸¸åœ¨å¤„ç†ç©ºé—´æ•°æ®æ—¶å¿½ç•¥äº†ç©ºé—´ä¿¡æ¯ã€‚ç„¶è€Œï¼Œç”±äºç»†èƒçŠ¶æ€å—å…¶å‘¨å›´ç»†èƒçš„å½±å“ï¼Œå°†è½¬å½•ç»„æ•°æ®ä¸ç»†èƒçš„ç©ºé—´ä¿¡æ¯ç»“åˆèµ·æ¥è¿›è¡Œèšç±»åˆ†æï¼Œå°†æ›´æœ‰åŠ©äºæ­ç¤ºç»†èƒåœ¨ç»„ç»‡ä¸­çš„åˆ†å¸ƒå’Œç›¸äº’ä½œç”¨ã€‚</p><p data-tool="mdniceç¼–è¾‘å™¨">2024å¹´2æœˆï¼Œæ¥è‡ªShyam Prabhakarâ€‰å’ŒKok Hao Chenç ”ç©¶å›¢é˜Ÿçš„Vipul Singhalã€Nigel Chouå’ŒJoseph Leeç­‰äººåœ¨<em><strong>Nature Genetics</strong></em>ä¸Šå‘è¡¨äº†ä¸€ç¯‡é¢˜ä¸ºã€ŠBANKSY unifies cell typing and tissue domain segmentation for scalable spatial omics data analysisã€‹ï¼ˆhttps://www.nature.com/articles/s41588-024-01664-3ï¼‰çš„ç ”ç©¶æ–‡ç« ã€‚è¿™é¡¹å·¥ä½œå¼€å‘äº†ä¸€ç§åä¸º<strong>BANKSY</strong>çš„ç©ºé—´èšç±»ç®—æ³•ã€‚BANKSYå¯é€šè¿‡ç»“åˆç»†èƒ/Spotsçš„è½¬å½•ç»„ä¿¡æ¯å’Œå…¶å±€éƒ¨å¾®ç¯å¢ƒä¿¡æ¯ï¼Œæ˜¾è‘—æå‡ç©ºé—´ç»„å­¦æ•°æ®åˆ†æçš„å‡†ç¡®æ€§å’Œæ•ˆç‡ã€‚BANKSYç®—æ³•ä¸ä»…èƒ½å¤Ÿå¤„ç†å¤§è§„æ¨¡æ•°æ®é›†ï¼Œæé«˜ç»†èƒç±»å‹èšç±»å’Œç»„ç»‡åŸŸåˆ†å‰²çš„æ€§èƒ½ï¼Œè¿˜å±•ç°äº†åœ¨å¤šç§RNAå’Œè›‹ç™½è´¨æ•°æ®é›†ä¸Šçš„å¹¿æ³›é€‚ç”¨æ€§ã€‚</p><p data-tool="mdniceç¼–è¾‘å™¨">å…·ä½“æ¥è¯´ï¼ŒBANKSYæ˜¯ä¸€ç§èšç±»ç©ºé—´ç»„å­¦æ•°æ®çš„æ–¹æ³•ï¼Œé€šè¿‡å°†æ¯ä¸ªç»†èƒçš„ç‰¹å¾ä¸å…¶ç©ºé—´è¿‘é‚»ç‰¹å¾çš„å¹³å‡å€¼ä»¥åŠé‚»åŸŸç‰¹å¾æ¢¯åº¦ç›¸ç»“åˆæ¥å¢å¼ºã€‚é€šè¿‡å°†é‚»åŸŸä¿¡æ¯çº³å…¥èšç±»ä¸­ï¼ŒBANKSYèƒ½å¤Ÿï¼š</p><ul data-tool="mdniceç¼–è¾‘å™¨"><li><section>åœ¨å˜ˆæ‚çš„æ•°æ®ä¸­æ”¹å–„ç»†èƒ/Spotsç±»å‹åˆ†é…ï¼›</section></li><li><section>åŒºåˆ†å¾®ç¯å¢ƒåˆ†å±‚çš„å¾®å¦™ä¸åŒç»†èƒ/Spotsç±»å‹ï¼›</section></li><li><section>ç¡®å®šå…±äº«ç›¸åŒå¾®ç¯å¢ƒçš„ç©ºé—´åŸŸã€‚</section></li></ul><p data-tool="mdniceç¼–è¾‘å™¨">BANKSYå®˜æ–¹æ•™ç¨‹è¯¦è§ï¼š</p><ul data-tool="mdniceç¼–è¾‘å™¨"><li><section>githubï¼šhttps://github.com/prabhakarlab/Banksy/</section></li><li><section>Rè¯­è¨€ç‰ˆæœ¬ï¼šhttps://prabhakarlab.github.io/Banksy/</section></li><li><section>Pythonç‰ˆæœ¬ï¼šhttps://github.com/prabhakarlab/Banksy_py</section></li></ul><h3 data-tool="mdniceç¼–è¾‘å™¨"><span>ä¸€. BANKSYç®—æ³•çš„å·¥ä½œæµç¨‹</span></h3><figure data-tool="mdniceç¼–è¾‘å™¨"><img data-imgfileid="100004541" data-ratio="0.9462962962962963" data-src="https://mmbiz.qpic.cn/mmbiz_jpg/fTW9zRI3eqUffobfsLtm11jVN9MLsqTr38nyia7E2uV7f0GphVh5Ol75dL7S9YdfREEiaQ89kHMpJNgOLJ9CLMLw/640?wx_fmt=other&amp;from=appmsg" data-type="other" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_jpg/fTW9zRI3eqUffobfsLtm11jVN9MLsqTr38nyia7E2uV7f0GphVh5Ol75dL7S9YdfREEiaQ89kHMpJNgOLJ9CLMLw/640?wx_fmt=other&amp;from=appmsg"><figcaption>Fig. 1</figcaption></figure><p data-tool="mdniceç¼–è¾‘å™¨">å›¾ä¾‹ï¼š<strong>a</strong>, The original geneâ€“cell expression matrix (purple) was augmented with neighborhood-averaged expression matrices corresponding to the mean local expression (dark pink) and the AGF (light pink). Here, <em>Î»</em> is a mixing parameter that controls the importance of cellsâ€™ own expression and neighborhood expression effects, <em>G</em>(<em>r</em>) is a radially symmetric Gaussian kernel that decays from magnitude 1 at <em>r</em>â€‰=â€‰0, â€˜expressionâ€™ refers to each geneâ€™s expression level in each cell, the mean is taken over cells in the respective index cellâ€™s neighborhood and the <em>e<strong>i</strong>Ï•**G</em>(<em>r</em>) term confers gradient sensitivity to the AGF. <strong>b</strong>, Heatmap of the real and imaginary components of a gradient-sensitive AGF kernel. The plots show an unnormalized AGF kernel: the real part (cosâ¡(ğœ™)) senses the gradient along the <em>x</em> axis and the imaginary part (sinâ¡(ğœ™)) senses the gradient along the <em>y</em> axis. <strong>c</strong>, Simplified schematic of two distinct cell types in the neighbor-augmented space. The neighbor expression features, representing the local microenvironment, help to separate two clusters that would be difficult to separate based on the cellsâ€™ own expression alone. For simplicity, we show â€˜pureâ€™ microenvironments containing only a single cell type (cell type 1 in zone A and cell type 2 in zone B), although BANKSY is equally applicable to heterogeneous microenvironments containing mixtures of cell types (Extended Data Fig. 1).</p><p data-tool="mdniceç¼–è¾‘å™¨"><strong>ä¸€ç§ç©ºé—´èšç±»çš„ç­–ç•¥æ˜¯å°†ç»†èƒçš„ç©ºé—´åæ ‡é™„åŠ åˆ°å®ƒä»¬çš„åŸºå› è¡¨è¾¾å‘é‡ä¸­</strong>ã€‚ç„¶è€Œï¼Œè¿™ä¼šå¯¼è‡´ç©ºé—´ä¸Šç›¸è·è¾ƒè¿œçš„ç»†èƒåˆ†å…¥ä¸åŒçš„èšç±»ä¸­ï¼Œå³ä½¿å®ƒä»¬å…·æœ‰ç›¸åŒçš„è½¬å½•ç»„ã€‚<strong>ä¸€ç§æ›´ç›´è§‚çš„æ–¹æ³•æ˜¯é™„åŠ æŸç§è¡¨ç¤ºç»†èƒå¾®ç¯å¢ƒçš„è¡¨ç¤º</strong>ã€‚<strong>BANKSYç®—æ³•</strong>é‡‡ç”¨å¹³å‡é‚»åŸŸè¡¨è¾¾å’ŒAGF<strong>ï¼ˆå›¾1a,1bï¼‰</strong>æ¥ä»£è¡¨æ¯ä¸ªç»†èƒå‘¨å›´çš„è½¬å½•ç»„å¾®ç¯å¢ƒ<strong>ã€‚</strong>é‡è¦çš„æ˜¯ï¼ŒAGFï¼ˆå›¾1bï¼‰å¯ä»¥è¢«è§†ä¸ºè¡¡é‡æ¯ä¸ªç»†èƒé‚»åŸŸå†…åŸºå› è¡¨è¾¾æ¢¯åº¦ï¼Œå¯¹æ ·æœ¬æ—‹è½¬æ˜¯ä¸å˜çš„ã€‚æ¥ä¸‹æ¥ï¼Œè¿™äº›<strong>é¢å¤–çš„ç‰¹å¾</strong>è¢«ç”¨æ¥å°†ç»†èƒåµŒå…¥åˆ°ä¸€ä¸ªé‚»åŸŸå¢å¼ºçš„ä¹˜ç§¯ç©ºé—´ä¸­ï¼ˆå›¾1cï¼‰ã€‚ç»è¿‡é™ç»´å¤„ç†ï¼Œç„¶ååœ¨ç»“æœåµŒå…¥ç©ºé—´ä¸­æ„å»ºå›¾å½¢ï¼Œå¯ä»¥ä½¿ç”¨ä»»ä½•å›¾åˆ’åˆ†ç®—æ³•è¿›è¡Œèšç±»ã€‚é»˜è®¤æƒ…å†µä¸‹ï¼ŒBANKSYä½¿ç”¨Leidenç¤¾åŒºæ£€æµ‹ç®—æ³•ï¼Œå› ä¸ºå…¶é€Ÿåº¦å’Œå¯æ‰©å±•æ€§ï¼Œè™½ç„¶ä¹Ÿæä¾›äº†å…¶ä»–æ–¹æ³•ï¼ˆLouvainã€åŸºäºæ¨¡å‹çš„èšç±»å’Œk-meansï¼‰ä½œä¸ºé€‰é¡¹ã€‚</p><p data-tool="mdniceç¼–è¾‘å™¨"><strong>ä¸ºäº†æ§åˆ¶ç»†èƒè‡ªèº«å’Œé‚»åŸŸç‰¹å¾å¯¹åµŒå…¥ä¸­ç»†èƒ-ç»†èƒè·ç¦»çš„ç›¸å¯¹è´¡çŒ®</strong>ï¼ŒBANKSYä½¿ç”¨ä¸€ä¸ªæ··åˆå‚æ•°ï¼Œ<strong>Î»</strong>âˆˆâ€‰[0,â€‰1]ï¼Œæ¥æƒè¡¡ç»†èƒè½¬å½•ç»„çŸ©é˜µå’Œé‚»åŸŸè¡¨è¾¾çŸ©é˜µï¼ˆå¹³å‡å€¼å’ŒAGFï¼›å›¾1ï¼‰çš„è´¡çŒ®ã€‚è¾ƒå°çš„Î»è®¾ç½®å¼ºè°ƒç»†èƒè‡ªèº«çš„è½¬å½•ç»„ï¼Œå› æ­¤å¯¼è‡´ç»†èƒæ ¹æ®ç»†èƒç±»å‹è¿›è¡Œèšç±»ã€‚åœ¨æé™æƒ…å†µä¸‹ï¼Œå½“Î»â€‰=â€‰0æ—¶ï¼ŒBANKSYç®€åŒ–ä¸ºä¼ ç»Ÿçš„éç©ºé—´èšç±»ã€‚é€šè¿‡å¢åŠ Î»ï¼Œå¢åŠ é‚»åŸŸç­¾åçš„å½±å“ï¼Œå¯ä»¥å°†BANKSYä»ç»†èƒç±»å‹èšç±»æ¨¡å¼åˆ‡æ¢åˆ°ç»„ç»‡åŸŸåˆ†å‰²æ¨¡å¼ï¼Œå¯¼è‡´ç»†èƒæ ¹æ®ç»„ç»‡åŸŸè¿›è¡Œèšç±»ã€‚</p><blockquote data-tool="mdniceç¼–è¾‘å™¨"><p>å¼•ç”¨è‡ªã€<a href="https://mp.weixin.qq.com/s?__biz=MzkwOTQwOTE4OA==&amp;mid=2247486914&amp;idx=1&amp;sn=98e4409fb89ebc8642f27d63e97a0ca0&amp;scene=21#wechat_redirect" data-linktype="2">Nat.Genet. | BANKSYï¼šåŸºäºé‚»åŸŸæ ¸å’Œç©ºé—´ç»„å­¦çš„å¯æ‰©å±•åˆ†æç®—æ³•ç»Ÿä¸€ç»†èƒåˆ†å‹å’Œç»„ç»‡åŸŸåˆ†å‰²ï¼Œè§£æå±€éƒ¨å¾®ç¯å¢ƒ</a>ã€‘</p><p>è¯¦è§NGæ–‡ç« çš„ç»“æœå’Œæ–¹æ³•éƒ¨åˆ†ï¼šhttps://www.nature.com/articles/s41588-024-01664-3</p></blockquote><h3 data-tool="mdniceç¼–è¾‘å™¨"><span>äºŒ. BANKSYçš„Rä»£ç å®ç°</span></h3><h4 data-tool="mdniceç¼–è¾‘å™¨"><span>Step0. RåŒ…å®‰è£…</span></h4><p data-tool="mdniceç¼–è¾‘å™¨">BANKSYå¯¹Rè¯­è¨€ç‰ˆæœ¬è¦æ±‚æ¯”è¾ƒé«˜ï¼š</p><ul data-tool="mdniceç¼–è¾‘å™¨"><li><section>æ•™ç¨‹è¦æ±‚Rè¯­è¨€ç‰ˆæœ¬ <code>&gt;= 4.4.0</code></section></li></ul><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code><span>#BiocManager::install("SpatialExperiment")</span><br>BiocManager::install(<span>'Banksy'</span>)<br></code></pre><ul data-tool="mdniceç¼–è¾‘å™¨"><li><section>å¦‚æœRç‰ˆæœ¬æ˜¯<code>4.3.0</code>çš„å¯ä»¥é€šè¿‡githubå®‰è£…ï¼ˆhttps://github.com/prabhakarlab/Banksy/issues/13ï¼‰</section></li></ul><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code><span>#BiocManager::install("SpatialExperiment")</span><br>remotes::install_github(<span>"prabhakarlab/Banksy"</span>)<br></code></pre><p data-tool="mdniceç¼–è¾‘å™¨">å®˜æ–¹æ•™ç¨‹æä¾›äº†å¿«é€Ÿæµ‹è¯•çš„ä»£ç ï¼ˆhttps://prabhakarlab.github.io/Banksy/index.htmlï¼‰ï¼š</p><h4 data-tool="mdniceç¼–è¾‘å™¨"><span>Step1. åŠ è½½RåŒ…å’Œç¤ºä¾‹æ•°æ®</span></h4><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code><span>library</span>(Banksy)<br><br><span>library</span>(SummarizedExperiment)<br><span>library</span>(SpatialExperiment)<br><span>library</span>(scuttle)<br><br><span>library</span>(scater)<br><span>library</span>(cowplot)<br><span>library</span>(ggplot2)<br></code></pre><p data-tool="mdniceç¼–è¾‘å™¨">æµ‹è¯•æ•°æ®æ˜¯ç»å…¸çš„å°é¼ æµ·é©¬10X visiumç©ºè½¬æ•°æ®ã€‚</p><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code>data(hippocampus)<br>gcmÂ &lt;-Â hippocampus$expression<br>gcm[<span>1</span>:<span>10</span>,Â <span>1</span>:<span>10</span>]<br>locsÂ &lt;-Â as.matrix(hippocampus$locations)<br></code></pre><p data-tool="mdniceç¼–è¾‘å™¨">Banksyéœ€è¦ç”¨æˆ·å°†æ•°æ®è½¬åŒ–ä¸ºSpatialExperimentå¯¹è±¡ï¼š</p><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code>seÂ &lt;-Â SpatialExperiment(assayÂ =Â list(countsÂ =Â gcm),Â <br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â spatialCoordsÂ =Â locs)<br></code></pre><h4 data-tool="mdniceç¼–è¾‘å™¨"><span>Step2. è´¨æ§å’Œæ ‡å‡†åŒ–åˆ†æ</span></h4><blockquote data-tool="mdniceç¼–è¾‘å™¨"><p>è¿™é‡Œå¦‚æœå·²ç»èµ°è¿‡Seuratçš„è´¨æ§æµç¨‹ï¼Œå¯ä»¥è·³è¿‡QCéƒ¨åˆ†ã€‚</p></blockquote><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code><span>#Â QCÂ basedÂ onÂ totalÂ counts</span><br>qcstatsÂ &lt;-Â perCellQCMetrics(se)<br>thresÂ &lt;-Â quantile(qcstats$total,Â c(<span>0.05</span>,Â <span>0.98</span>))<br>keepÂ &lt;-Â (qcstats$totalÂ &gt;Â thres[<span>1</span>])Â &amp;Â (qcstats$totalÂ &lt;Â thres[<span>2</span>])<br>seÂ &lt;-Â se[,Â keep]<br><br><span>#Â NormalizationÂ toÂ meanÂ libraryÂ size</span><br>seÂ &lt;-Â computeLibraryFactors(se)<br>anameÂ &lt;-Â <span>"normcounts"</span><br>assay(se,Â aname)Â &lt;-Â normalizeCounts(se,Â logÂ =Â <span>FALSE</span>)<br></code></pre><h4 data-tool="mdniceç¼–è¾‘å™¨"><span>Step3. è®¡ç®—BANKSYçš„é‚»åŸŸçŸ©é˜µ</span></h4><p data-tool="mdniceç¼–è¾‘å™¨">è®¾ç½®<code>compute_agf=TRUE</code>è®¡ç®—åŠ æƒé‚»åŸŸå¹³å‡å€¼ï¼ˆâ„³ï¼‰å’Œæ–¹ä½Gaboræ»¤æ³¢å™¨ï¼ˆğ’¢ï¼‰ã€‚ç”¨äºè®¡ç®—â„³å’Œğ’¢çš„ç©ºé—´è¿‘é‚»ä¸ªæ•°åˆ†åˆ«ä¸º<code>k_geom[1]=15</code>å’Œ<code>k_geom[2]=30</code>ã€‚æˆ‘ä»¬åœ¨<code>lambda=0</code>å¤„è¿è¡ŒBANKSYï¼Œå¯¹åº”äºéç©ºé—´èšç±»;<code>lambda=0.2</code>å¯¹åº”äºBANKSYç©ºé—´ç»†èƒèšç±»ã€‚</p><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code>lambdaÂ &lt;-Â c(<span>0</span>,Â <span>0.2</span>)<br>k_geomÂ &lt;-Â c(<span>15</span>,Â <span>30</span>)<br><br>seÂ &lt;-Â Banksy::computeBanksy(se,Â assay_nameÂ =Â aname,Â compute_agfÂ =Â <span>TRUE</span>,Â k_geomÂ =Â k_geom)<br></code></pre><h4 data-tool="mdniceç¼–è¾‘å™¨"><span>Step4. PCAã€é™ç»´å’Œèšç±»</span></h4><p data-tool="mdniceç¼–è¾‘å™¨">æ¥ä¸‹æ¥ï¼Œåœ¨BANKSYçŸ©é˜µä¸Šè¿è¡ŒPCAå¹¶è¿è¡Œèšç±»ã€‚è®¾ç½®<code>use_agf=TRUE</code>ä½¿ç”¨â„³å’Œğ’¢æ„å»ºBANKSYçŸ©é˜µã€‚</p><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code>set.seed(<span>1000</span>)<br>seÂ &lt;-Â Banksy::runBanksyPCA(se,Â use_agfÂ =Â <span>TRUE</span>,Â lambdaÂ =Â lambda)<br>seÂ &lt;-Â Banksy::runBanksyUMAP(se,Â use_agfÂ =Â <span>TRUE</span>,Â lambdaÂ =Â lambda)<br>seÂ &lt;-Â Banksy::clusterBanksy(se,Â use_agfÂ =Â <span>TRUE</span>,Â lambdaÂ =Â lambda,Â resolutionÂ =Â <span>1.2</span>)<br></code></pre><p data-tool="mdniceç¼–è¾‘å™¨">å¯ä»¥ä½¿ç”¨<code>connectClusters</code>å‡½æ•°å°†ä¸åŒçš„èšç±»ç»“æœé‡æ–°æ ‡è®°ï¼Œä»¥æœ€å°åŒ–å®ƒä»¬ä¹‹é—´çš„å·®å¼‚ã€‚</p><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code>seÂ &lt;-Â Banksy::connectClusters(se)<br><span>#&gt;Â clust_M1_lam0.2_k50_res1.2Â --&gt;Â clust_M1_lam0_k50_res1.2</span><br></code></pre><h4 data-tool="mdniceç¼–è¾‘å™¨"><span>Step5. å¯è§†åŒ–</span></h4><p data-tool="mdniceç¼–è¾‘å™¨">å¯è§†åŒ–éç©ºé—´èšç±»<code>(lambda=0)</code>å’ŒBANKSYèšç±»<code>(lambda=0.2)</code>çš„èšç±»ç»“æœï¼š</p><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code>cnamesÂ &lt;-Â colnames(colData(se))<br>cnamesÂ &lt;-Â cnames[grep(<span>"^clust"</span>,Â cnames)]<br>colData(se)Â &lt;-Â cbind(colData(se),Â spatialCoords(se))<br><br>plot_nspÂ &lt;-Â plotColData(se,<br>Â Â Â Â xÂ =Â <span>"sdimx"</span>,Â yÂ =Â <span>"sdimy"</span>,<br>Â Â Â Â point_sizeÂ =Â <span>0.6</span>,Â colour_byÂ =Â cnames[<span>1</span>]<br>)<br>plot_bankÂ &lt;-Â plotColData(se,<br>Â Â Â Â xÂ =Â <span>"sdimx"</span>,Â yÂ =Â <span>"sdimy"</span>,<br>Â Â Â Â point_sizeÂ =Â <span>0.6</span>,Â colour_byÂ =Â cnames[<span>2</span>]<br>)<br><br><br>plot_grid(plot_nspÂ +Â coord_equal(),Â plot_bankÂ +Â coord_equal(),Â ncolÂ =Â <span>2</span>)<br></code></pre><figure data-tool="mdniceç¼–è¾‘å™¨"><img data-imgfileid="100004539" data-ratio="0.35714285714285715" data-src="https://mmbiz.qpic.cn/mmbiz_jpg/fTW9zRI3eqUffobfsLtm11jVN9MLsqTrJkmvwkjjWHPTRUcqK1VWIialnPlgpHQCOkECp64HASHVl1HEtUIcexA/640?wx_fmt=other&amp;from=appmsg" data-type="other" data-w="980" src="https://mmbiz.qpic.cn/mmbiz_jpg/fTW9zRI3eqUffobfsLtm11jVN9MLsqTrJkmvwkjjWHPTRUcqK1VWIialnPlgpHQCOkECp64HASHVl1HEtUIcexA/640?wx_fmt=other&amp;from=appmsg"></figure><p data-tool="mdniceç¼–è¾‘å™¨">æˆ‘ä»¬è¿˜å¯ä»¥åˆ†åˆ«å¯è§†åŒ–æ¯ä¸ªäºšç¾¤ï¼š</p><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code>plot_grid(<br>Â Â Â Â plot_nspÂ +Â facet_wrap(~colour_by),<br>Â Â Â Â plot_bankÂ +Â facet_wrap(~colour_by),<br>Â Â Â Â ncolÂ =Â <span>2</span><br>)<br></code></pre><figure data-tool="mdniceç¼–è¾‘å™¨"><img data-imgfileid="100004540" data-ratio="0.4444444444444444" data-src="https://mmbiz.qpic.cn/mmbiz_jpg/fTW9zRI3eqUffobfsLtm11jVN9MLsqTrQWSdbDohEbmXPmwKyXic0icMGviaT7DlpA7DqV56aaMiaHjhMicoeD4fxzg/640?wx_fmt=other&amp;from=appmsg" data-type="other" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_jpg/fTW9zRI3eqUffobfsLtm11jVN9MLsqTrQWSdbDohEbmXPmwKyXic0icMGviaT7DlpA7DqV56aaMiaHjhMicoeD4fxzg/640?wx_fmt=other&amp;from=appmsg"></figure><p data-tool="mdniceç¼–è¾‘å™¨">å¯è§†åŒ–æ¯”è¾ƒéç©ºé—´å’ŒBANKSYçš„umapç»“æœï¼š</p><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code>rdnamesÂ &lt;-Â reducedDimNames(se)<br><br>umap_nspÂ &lt;-Â plotReducedDim(se,<br>Â Â Â Â dimredÂ =Â grep(<span>"UMAP.*lam0$"</span>,Â rdnames,Â valueÂ =Â <span>TRUE</span>),<br>Â Â Â Â colour_byÂ =Â cnames[<span>1</span>]<br>)<br>umap_bankÂ &lt;-Â plotReducedDim(se,<br>Â Â Â Â dimredÂ =Â grep(<span>"UMAP.*lam0.2$"</span>,Â rdnames,Â valueÂ =Â <span>TRUE</span>),<br>Â Â Â Â colour_byÂ =Â cnames[<span>2</span>]<br>)<br>plot_grid(<br>Â Â Â Â umap_nsp,<br>Â Â Â Â umap_bank,<br>Â Â Â Â ncolÂ =Â <span>2</span><br>)<br></code></pre><figure data-tool="mdniceç¼–è¾‘å™¨"><img data-imgfileid="100004537" data-ratio="0.35714285714285715" data-src="https://mmbiz.qpic.cn/mmbiz_jpg/fTW9zRI3eqUffobfsLtm11jVN9MLsqTriaXNz1uxZtyVehKUfE9VU1pTTm2jqHFa9hDkjqFOQ7lJXUk2JWhD3cw/640?wx_fmt=other&amp;from=appmsg" data-type="other" data-w="980" src="https://mmbiz.qpic.cn/mmbiz_jpg/fTW9zRI3eqUffobfsLtm11jVN9MLsqTriaXNz1uxZtyVehKUfE9VU1pTTm2jqHFa9hDkjqFOQ7lJXUk2JWhD3cw/640?wx_fmt=other&amp;from=appmsg"><figcaption>img</figcaption></figure><h3 data-tool="mdniceç¼–è¾‘å™¨"><span>ä¸‰. BANKSYçš„å‡½æ•°å’Œå…³é”®å‚æ•°</span></h3><h4 data-tool="mdniceç¼–è¾‘å™¨"><span>3.1 RåŒ…BANKSYå°è£…çš„å‡½æ•°</span></h4><section data-tool="mdniceç¼–è¾‘å™¨"><table><thead><tr><th>Neighborhood matrix computation</th><th>Annotation</th></tr></thead><tbody><tr><td><code>computeBanksy()</code></td><td>Compute the component neighborhood matrices for the BANKSY matrix.</td></tr><tr><td><code>getBanksyMatrix()</code></td><td>Builds the BANKSY matrix from neighborhood matrices.</td></tr><tr><td><strong>Dimensionality reduction</strong></td><td><br></td></tr><tr><td><code>runBanksyPCA()</code></td><td>Run PCA on a BANKSY matrix.</td></tr><tr><td><code>runBanksyUMAP()</code></td><td>Run UMAP on a BANKSY embedding.</td></tr><tr><td><strong>Clustering</strong></td><td><br></td></tr><tr><td><code>clusterBanksy()</code></td><td>Perform clustering in BANKSY's neighborhood-augmented feature space.</td></tr><tr><td><code>connectClusters()</code></td><td>Relabel cluster labels across parameter runs to maximise their similarity.</td></tr><tr><td><code>compareClusters()</code></td><td>Compare cluster outputs based on various clustering comparison measures.</td></tr><tr><td><code>smoothLabels()</code></td><td>k-Nearest neighbor cluster label smoothing.</td></tr><tr><td><code>clusterNames()</code></td><td>Get names of clustering runs.</td></tr><tr><td><strong>Datasets</strong></td><td><br></td></tr><tr><td><code>hippocampus</code></td><td>Mouse Hippocampus VeraFISH data</td></tr><tr><td><code>rings</code></td><td>An unrealistic simulation of spatially-resolved omics data.</td></tr><tr><td><strong>Misc.</strong></td><td><br></td></tr><tr><td><code>simulateDataset()</code></td><td>Simulate an unrealistic spatial omics dataset.</td></tr></tbody></table></section><h4 data-tool="mdniceç¼–è¾‘å™¨"><span>3.2 ç›¸å…³å‡½æ•°çš„å…³é”®å‚æ•°</span></h4><p data-tool="mdniceç¼–è¾‘å™¨">BANKSYçš„æ ¸å¿ƒå‡½æ•°æœ‰ä¸€äº›å…³é”®å‚æ•°ï¼Œåœ¨è¿™é‡Œä½œè€…åšäº†è¯¦ç»†çš„ä»‹ç»ï¼ˆhttps://prabhakarlab.github.io/Banksy/articles/parameter-selection.htmlï¼‰ï¼š</p><h5 data-tool="mdniceç¼–è¾‘å™¨"><span>3.2.1 <code>computeBanksy()</code>:</span></h5><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code>k_geomÂ &lt;-Â c(<span>15</span>,Â <span>30</span>)<br>seÂ &lt;-Â computeBanksy(se,Â assay_nameÂ =Â aname,Â compute_agfÂ =Â <span>TRUE</span>,Â k_geomÂ =Â k_geom)<br></code></pre><ul data-tool="mdniceç¼–è¾‘å™¨"><li><section><p><strong>AGF usage</strong>: For characterising neighborhoods, <em>BANKSY</em> computes the weighted neighborhood mean (<code>H_0</code>) and the azimuthal Gabor filter (<code>H_1</code>), which estimates gene expression gradients. Setting <code>compute_agf=TRUE</code> computes both <code>H_0</code> and <code>H_1</code>.</p></section></li><li><section><p><strong>k-geometric</strong>: <code>k_geom</code> specifies the number of neighbors used to compute each <code>H_m</code> for <code>m=0,1</code>. If a single value is specified, the same <code>k_geom</code> will be used for each feature matrix. Alternatively, multiple values of <code>k_geom</code> can be provided for each feature matrix. Here, we use <code>k_geom[1]=15</code> and <code>k_geom[2]=30</code> for <code>H_0</code> and <code>H_1</code> respectively. More neighbors are used to compute gradients.</p></section></li></ul><p data-tool="mdniceç¼–è¾‘å™¨"><code>computeBanksy</code> populates the <code>assays</code> slot with <code>H_0</code> and <code>H_1</code> in this instance:</p><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code>se<br><span>#&gt;Â class:Â SpatialExperimentÂ </span><br><span>#&gt;Â dim:Â 120Â 10205Â </span><br><span>#&gt;Â metadata(1):Â BANKSY_params</span><br><span>#&gt;Â assays(4):Â countsÂ normcountsÂ H0Â H1</span><br><span>#&gt;Â rownames(120):Â Sparcl1Â Slc1a2Â ...Â Notch3Â Egfr</span><br><span>#&gt;Â rowDataÂ names(0):</span><br><span>#&gt;Â colnames(10205):Â cell_1276Â cell_691Â ...Â cell_11635Â cell_10849</span><br><span>#&gt;Â colDataÂ names(4):Â sample_idÂ sdimxÂ sdimyÂ sizeFactor</span><br><span>#&gt;Â reducedDimNames(0):</span><br><span>#&gt;Â mainExpName:Â NULL</span><br><span>#&gt;Â altExpNames(0):</span><br><span>#&gt;Â spatialCoordsÂ names(2)Â :Â sdimxÂ sdimy</span><br><span>#&gt;Â imgDataÂ names(1):Â sample_id</span><br></code></pre><h5 data-tool="mdniceç¼–è¾‘å™¨"><span>3.2.2 <code>runBanksyPCA()</code></span></h5><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code>lambdaÂ &lt;-Â c(<span>0</span>,Â <span>0.2</span>)<br>seÂ &lt;-Â runBanksyPCA(se,Â use_agfÂ =Â c(<span>FALSE</span>,Â <span>TRUE</span>),Â lambdaÂ =Â lambda,Â seedÂ =Â <span>1000</span>)<br><span>#&gt;Â UsingÂ seed=1000</span><br><span>#&gt;Â UsingÂ seed=1000</span><br><span>#&gt;Â UsingÂ seed=1000</span><br><span>#&gt;Â UsingÂ seed=1000</span><br></code></pre><p data-tool="mdniceç¼–è¾‘å™¨">The <code>lambda</code> parameter is a mixing parameter in <code>[0,1]</code> which determines how much spatial information is incorporated for downstream analysis. With smaller values of <code>lambda</code>, BANKY operates in <em>cell-typing</em> mode, while at higher levels of <code>lambda</code>, BANKSY operates in <em>domain-finding</em> mode. As a starting point, we recommend <code>lambda=0.2</code> for cell-typing and <code>lambda=0.8</code> for zone-finding. Here, we run <code>lambda=0</code> which corresponds to non-spatial clustering, and <code>lambda=0.2</code> for spatially-informed cell-typing. We compute PCs with and without the AGF (<code>H_1</code>).</p><p data-tool="mdniceç¼–è¾‘å™¨"><code>runBanksyPCA</code> populates the <code>reducedDims</code> slot, with each combination of <code>use_agf</code> and <code>lambda</code> provided.</p><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code>reducedDimNames(se)<br><span>#&gt;Â [1]Â "PCA_M0_lam0"Â Â Â "PCA_M0_lam0.2"Â "PCA_M1_lam0"Â Â Â "PCA_M1_lam0.2"</span><br></code></pre><h5 data-tool="mdniceç¼–è¾‘å™¨"><span>3.2.3 <code>clusterBanksy()</code>èšç±»å‚æ•°</span></h5><p data-tool="mdniceç¼–è¾‘å™¨">Next, we cluster the BANKSY embedding with Leiden graph-based clustering. This admits two parameters: <code>k_neighbors</code> and <code>resolution</code>. <code>k_neighbors</code> determines the number of k nearest neighbors used to construct the shared nearest neighbors graph. Leiden clustering is then performed on the resultant graph with resolution <code>resolution</code>. For reproducibiltiy we set a seed for each parameter combination.</p><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code>kÂ &lt;-Â <span>50</span><br>resÂ &lt;-Â <span>1</span><br>seÂ &lt;-Â clusterBanksy(se,Â use_agfÂ =Â c(<span>FALSE</span>,Â <span>TRUE</span>),Â lambdaÂ =Â lambda,Â k_neighborsÂ =Â k,Â resolutionÂ =Â res,Â seedÂ =Â <span>1000</span>)<br><span>#&gt;Â UsingÂ seed=1000</span><br><span>#&gt;Â UsingÂ seed=1000</span><br><span>#&gt;Â UsingÂ seed=1000</span><br><span>#&gt;Â UsingÂ seed=1000</span><br></code></pre><p data-tool="mdniceç¼–è¾‘å™¨"><code>clusterBanksy</code> populates <code>colData(se)</code> with cluster labels:</p><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code>colnames(colData(se))<br><span>#&gt;Â [1]Â "sample_id"Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â "sdimx"Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â </span><br><span>#&gt;Â [3]Â "sdimy"Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â "sizeFactor"Â Â Â Â Â Â Â Â Â Â Â Â Â Â </span><br><span>#&gt;Â [5]Â "clust_M0_lam0_k50_res1"Â Â Â "clust_M0_lam0.2_k50_res1"</span><br><span>#&gt;Â [7]Â "clust_M1_lam0_k50_res1"Â Â Â "clust_M1_lam0.2_k50_res1"</span><br></code></pre><p data-tool="mdniceç¼–è¾‘å™¨"><strong>Comparing cluster results</strong></p><p data-tool="mdniceç¼–è¾‘å™¨">To compare clustering runs visually, different runs can be relabeled to minimise their differences with <code>connectClusters</code>:</p><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code>seÂ &lt;-Â connectClusters(se)<br><span>#&gt;Â clust_M1_lam0_k50_res1Â --&gt;Â clust_M0_lam0_k50_res1</span><br><span>#&gt;Â clust_M0_lam0.2_k50_res1Â --&gt;Â clust_M1_lam0_k50_res1</span><br><span>#&gt;Â clust_M1_lam0.2_k50_res1Â --&gt;Â clust_M0_lam0.2_k50_res1</span><br></code></pre><p data-tool="mdniceç¼–è¾‘å™¨">Visualise spatial coordinates with cluster labels.</p><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code>cnamesÂ &lt;-Â colnames(colData(se))<br>cnamesÂ &lt;-Â cnames[grep(<span>"^clust"</span>,Â cnames)]<br>cplotsÂ &lt;-Â lapply(cnames,Â <span>function</span>(cnm)Â {<br>Â Â Â Â plotColData(se,Â xÂ =Â <span>"sdimx"</span>,Â yÂ =Â <span>"sdimy"</span>,Â point_sizeÂ =Â <span>0.1</span>,Â colour_byÂ =Â cnm)Â +<br>Â Â Â Â Â Â Â Â coord_equal()Â +<br>Â Â Â Â Â Â Â Â labs(titleÂ =Â cnm)Â +<br>Â Â Â Â Â Â Â Â theme(legend.titleÂ =Â element_blank())Â +<br>Â Â Â Â Â Â Â Â guides(colourÂ =Â guide_legend(override.aesÂ =Â list(sizeÂ =Â <span>2</span>)))<br>})<br><br>plot_grid(plotlistÂ =Â cplots,Â ncolÂ =Â <span>2</span>)<br></code></pre><figure data-tool="mdniceç¼–è¾‘å™¨"><img data-imgfileid="100004538" data-ratio="0.96" data-src="https://mmbiz.qpic.cn/mmbiz_jpg/fTW9zRI3eqUffobfsLtm11jVN9MLsqTrjLr46I1pORSGWMvzIqRu7KTsOtA3lRicBkLla8HiaJibibkLdaTsWSEqHg/640?wx_fmt=other&amp;from=appmsg" data-type="other" data-w="525" src="https://mmbiz.qpic.cn/mmbiz_jpg/fTW9zRI3eqUffobfsLtm11jVN9MLsqTrjLr46I1pORSGWMvzIqRu7KTsOtA3lRicBkLla8HiaJibibkLdaTsWSEqHg/640?wx_fmt=other&amp;from=appmsg"></figure><p data-tool="mdniceç¼–è¾‘å™¨">Compare all cluster outputs with <code>compareClusters</code>. This function computes pairwise cluster comparison metrics between the clusters in <code>colData(se)</code> based on adjusted Rand index (ARI):</p><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code>compareClusters(se,Â funcÂ =Â <span>"ARI"</span>)<br><span>#&gt;Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â clust_M0_lam0_k50_res1Â clust_M0_lam0.2_k50_res1</span><br><span>#&gt;Â clust_M0_lam0_k50_res1Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 1.000Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.67</span><br><span>#&gt;Â clust_M0_lam0.2_k50_res1Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.670Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 1.00</span><br><span>#&gt;Â clust_M1_lam0_k50_res1Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 1.000Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.67</span><br><span>#&gt;Â clust_M1_lam0.2_k50_res1Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.747Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.87</span><br><span>#&gt;Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â clust_M1_lam0_k50_res1Â clust_M1_lam0.2_k50_res1</span><br><span>#&gt;Â clust_M0_lam0_k50_res1Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 1.000Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.747</span><br><span>#&gt;Â clust_M0_lam0.2_k50_res1Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.670Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.870</span><br><span>#&gt;Â clust_M1_lam0_k50_res1Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 1.000Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.747</span><br><span>#&gt;Â clust_M1_lam0.2_k50_res1Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.747Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 1.000</span><br></code></pre><p data-tool="mdniceç¼–è¾‘å™¨">or normalized mutual information (NMI):</p><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code>compareClusters(se,Â funcÂ =Â <span>"NMI"</span>)<br><span>#&gt;Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â clust_M0_lam0_k50_res1Â clust_M0_lam0.2_k50_res1</span><br><span>#&gt;Â clust_M0_lam0_k50_res1Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 1.000Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.741</span><br><span>#&gt;Â clust_M0_lam0.2_k50_res1Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.741Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 1.000</span><br><span>#&gt;Â clust_M1_lam0_k50_res1Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 1.000Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.741</span><br><span>#&gt;Â clust_M1_lam0.2_k50_res1Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.782Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.915</span><br><span>#&gt;Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â clust_M1_lam0_k50_res1Â clust_M1_lam0.2_k50_res1</span><br><span>#&gt;Â clust_M0_lam0_k50_res1Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 1.000Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.782</span><br><span>#&gt;Â clust_M0_lam0.2_k50_res1Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.741Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.915</span><br><span>#&gt;Â clust_M1_lam0_k50_res1Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 1.000Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.782</span><br><span>#&gt;Â clust_M1_lam0.2_k50_res1Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.782Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 1.000</span><br></code></pre><p data-tool="mdniceç¼–è¾‘å™¨">See <code>?compareClusters</code> for the full list of comparison measures.</p><h3 data-tool="mdniceç¼–è¾‘å™¨"><span>å››. æ€»ç»“</span></h3><p data-tool="mdniceç¼–è¾‘å™¨">ç›¸è¾ƒäºä¼ ç»Ÿçš„å•ç»†èƒ/ç©ºè½¬èšç±»ç®—æ³•ï¼Œ<strong>BANKSY</strong>åœ¨ç»†èƒ/Spotsçš„è¡¨è¾¾çŸ©é˜µçš„åŸºç¡€ä¸Šï¼Œè¿›ä¸€æ­¥è”åˆäº†ç©ºé—´åæ ‡ä¿¡æ¯ï¼Œè¿™ç§ç­–ç•¥èƒ½å¤Ÿæœ‰æ•ˆæé«˜ç»†èƒ/Spotsåˆ†ç±»çš„å‡†ç¡®æ€§å’Œæ•ˆç‡ï¼Œè¿›è€Œæ­ç¤ºç»†èƒé—´çš„ç›¸äº’ä½œç”¨å’Œå¾®ç¯å¢ƒå½±å“ã€‚æ­¤å¤–ï¼ŒBANKSYåœ¨å¤„ç†å¤§è§„æ¨¡æ•°æ®é›†æ—¶è¿è¡Œé€Ÿåº¦éå¸¸å¿«ï¼Œä¸”å…¼å®¹å¤šæ ·æœ¬åˆ†æä»¥åŠä¼ ç»Ÿçš„æ•´åˆåˆ†æç®—æ³•å¦‚Harmonyã€‚æœ€åï¼ŒBANKSYé€‚<span>ç”¨äº</span>å¤šç§<span>ç©ºé—´</span>æ•°æ®ç±»å‹<span>ï¼ˆä¾‹å¦‚ 10x Visiumã€Slide-seqã€MERFISHã€CosMXã€CODEX</span><span>ï¼‰</span><span>ï¼Œå½“ç„¶</span>ä¹Ÿé€‚åˆæœ€æ–°çš„<a target="_blank" href="https://mp.weixin.qq.com/s?__biz=MzUzMTEwODk0Ng==&amp;mid=2247520976&amp;idx=1&amp;sn=8a03a261e67bb11fa6b22c6853e234d9&amp;scene=21#wechat_redirect" textvalue="10X Visium HD" linktype="text" imgurl="" imgdata="null" tab="innerlink" data-linktype="2">10X Visium HD</a>ç©ºè½¬æ•°æ®~</p><p data-tool="mdniceç¼–è¾‘å™¨">ä¸‹é¢æ˜¯æ¥è‡ªæ¨æ–‡ã€<a href="https://mp.weixin.qq.com/s?__biz=MzkwOTQwOTE4OA==&amp;mid=2247486914&amp;idx=1&amp;sn=98e4409fb89ebc8642f27d63e97a0ca0&amp;scene=21#wechat_redirect" data-linktype="2">Nat.Genet. | BANKSYï¼šåŸºäºé‚»åŸŸæ ¸å’Œç©ºé—´ç»„å­¦çš„å¯æ‰©å±•åˆ†æç®—æ³•ç»Ÿä¸€ç»†èƒåˆ†å‹å’Œç»„ç»‡åŸŸåˆ†å‰²ï¼Œè§£æå±€éƒ¨å¾®ç¯å¢ƒ</a>ã€‘çš„ç‚¹è¯„ï¼š</p><blockquote data-tool="mdniceç¼–è¾‘å™¨"><p>BANKSYç®—æ³•çš„æ ¸å¿ƒç†å¿µï¼Œå³ä¸€ä¸ªç»†èƒçš„çŠ¶æ€å’ŒåŠŸèƒ½å¯ä»¥é€šè¿‡å…¶æ‰€å¤„çš„å¾®ç¯å¢ƒæ¥æ›´å‡†ç¡®åœ°è¯†åˆ«ï¼Œå¯å‘äº†æœªæ¥ç®—æ³•å¼€å‘çš„æ–¹å‘ã€‚åœ¨å¾ˆå¤šç”Ÿç‰©å­¦ç ”ç©¶ä¸­ï¼Œç»†èƒçš„ç©ºé—´ä½ç½®å¯¹å…¶åŠŸèƒ½çš„å½±å“æ˜¯ä¸å¯å¿½è§†çš„ï¼Œè€ŒBANKSYé€šè¿‡å°†ç©ºé—´é‚»åŸŸæè¿°ç¬¦ä¸ç»†èƒè‡ªèº«çš„åŸºå› è¡¨è¾¾å‘é‡ç›¸ç»“åˆï¼Œæä¾›äº†ä¸€ç§å…¨æ–°çš„è§†è§’æ¥ç†è§£ç»†èƒåœ¨ç»„ç»‡ä¸­çš„è§’è‰²ã€‚æ­¤å¤–ï¼ŒBANKSYçš„å¯æ‰©å±•æ€§å’Œå¯¹å‚æ•°å˜åŒ–çš„é²æ£’æ€§ï¼Œä¹Ÿä¸ºå¤„ç†å¤§è§„æ¨¡ç©ºé—´ç»„å­¦æ•°æ®é›†æä¾›äº†æœ‰æ•ˆçš„è§£å†³æ–¹æ¡ˆã€‚è¿™ç§å°†ç”Ÿç‰©ä¿¡æ¯å­¦ä¸æœºå™¨å­¦ä¹ æŠ€æœ¯ç›¸ç»“åˆçš„æ–¹æ³•ï¼Œä¸ä»…ä¸ºç©ºé—´ç»„å­¦æ•°æ®åˆ†ææä¾›äº†æ–°çš„å·¥å…·ï¼Œä¹Ÿä¸ºæœªæ¥åœ¨å…¶ä»–ç»„å­¦æ•°æ®åˆ†æä¸­æ•´åˆä¸åŒç±»å‹ä¿¡æ¯æä¾›äº†å¯èƒ½çš„æ–¹å‘ã€‚è¿™ç§è·¨å­¦ç§‘çš„èåˆï¼Œç‰¹åˆ«æ˜¯åœ¨ç”Ÿç‰©å­¦å’Œè®¡ç®—ç§‘å­¦ä¹‹é—´çš„äº¤å‰ï¼Œå°†æ˜¯æ¨åŠ¨ç”Ÿå‘½ç§‘å­¦ç ”ç©¶è¿›æ­¥çš„å…³é”®ã€‚</p></blockquote></section><p><br></p><p><mp-style-type data-value="3"></mp-style-type></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/VhLW5J326GAHIv-gBWIZng",target="_blank" rel="noopener noreferrer">åŸæ–‡é“¾æ¥</a>
