---
title: "《单细胞最佳实践》—重新回顾质控部分的新理解"
date: 2025-08-02T20:13:16Z
draft: ["false"]
tags: [
  "fetched",
  "生信菜鸟团"
]
categories: ["Acdemic"]
---
《单细胞最佳实践》—重新回顾质控部分的新理解 by 生信菜鸟团
------
<div><section><span leaf=""><br></span></section><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com"><p><span leaf="">在我接触单细胞数据分析的时候，早就已经有一套非常完整的 seurat 的流程了，直接照着官网给的代码示例分析就好了，<span textstyle="">为什么要这么分析呢我也不清楚，只知道照着做，而且目前单细胞数据有那么多，每篇文章里各种参数都有一些细节上的差别，很多参数都是根据经验来确定的</span>，没有一个明确的标准。为了丰富这方面的知识以及加深理解，《Single Cell Best Practices》这本在线书籍可能有一定的帮助，链接是 https://www.sc-best-practices.org/preamble.html </span></p><section nodeleaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2Los8bEPEs9LdQB954UWZlvJcXBlIAKIow70ySAc29QO8KQCCiaiaxe7uK7t2U419jvtF0spn3N0Yt9o5Q/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.9" data-s="300,640" data-type="png" data-w="1080" type="block" data-imgfileid="100052391" src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2Los8bEPEs9LdQB954UWZlvJcXBlIAKIow70ySAc29QO8KQCCiaiaxe7uK7t2U419jvtF0spn3N0Yt9o5Q/640?wx_fmt=png&amp;from=appmsg"></section><p><span leaf=""><br></span></p><p><span leaf="">这本书全面概述了从预处理到可视化和统计评估等基本分析的步骤，而且还在更新中，示例代码以 python 为主（如果想要以 R 为主的可以看 Orchestrating Single-Cell Analysis with Bioconductor 这本书），这本书前面还介绍了一下 scRNA-seq 的原理和原始数据的处理，我就直接跳到质量控制部分。</span></p><p><span leaf=""><br></span></p><h2 data-tool="mdnice编辑器"><span><span leaf="">1.1 Motivation（为什么要进行质控）</span></span><span></span></h2><p data-tool="mdnice编辑器"><span leaf="">scRNA-seq 数据存在 drop-out 现象，这意味着数据是含有过多 0 的稀疏矩阵。如果在质量控制期间过滤掉过多细胞，可能会丢失稀有的细胞亚群，消除掉一些生物学效应；而如果筛选过于宽松，在预处理流程中没有排除低质量细胞，那么在后续注释细胞时可能会遇到困难。因此，选择合适的预处理方法至关重要。</span></p><h2 data-tool="mdnice编辑器"><span><span leaf="">1.2 环境设置与数据</span></span><span></span></h2><p data-tool="mdnice编辑器"><span leaf="">首先是加载数据集，这个数据是托管在 figshare 上的一个健康样本的人类骨髓单核细胞数据。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">import</span></span><span leaf=""> numpy </span><span><span leaf="">as</span></span><span leaf=""> np</span><span leaf=""><br></span><span><span leaf="">import</span></span><span leaf=""> scanpy </span><span><span leaf="">as</span></span><span leaf=""> sc</span><span leaf=""><br></span><span><span leaf="">import</span></span><span leaf=""> seaborn </span><span><span leaf="">as</span></span><span leaf=""> sns</span><span leaf=""><br></span><span><span leaf="">import</span></span><span leaf=""> matplotlib.pyplot </span><span><span leaf="">as</span></span><span leaf=""> plt</span><span leaf=""><br></span><span><span leaf="">from</span></span><span leaf=""> scipy.stats </span><span><span leaf="">import</span></span><span leaf=""> median_abs_deviation</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">sc.settings.verbosity = </span><span><span leaf="">0</span></span><span leaf=""><br></span><span leaf="">sc.settings.set_figure_params(</span><span leaf=""><br></span><span leaf="">    dpi=</span><span><span leaf="">80</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">    facecolor=</span><span><span leaf="">"white"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">    frameon=</span><span><span leaf="">False</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">adata = sc.read_10x_h5(</span><span leaf=""><br></span><span leaf="">    filename=</span><span><span leaf="">"filtered_feature_bc_matrix.h5"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">    backup_url=</span><span><span leaf="">"https://figshare.com/ndownloader/files/39546196"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">adata</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">UserWarning: Variable names are </span><span><span leaf="">not</span></span><span leaf=""> unique. To make them unique, call `.var_names_make_unique`</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">AnnData object </span><span><span leaf="">with</span></span><span leaf=""> n_obs × n_vars = </span><span><span leaf="">16934</span></span><span leaf=""> × </span><span><span leaf="">36601</span></span><span leaf=""><br></span><span leaf="">    var: </span><span><span leaf="">'gene_ids'</span></span><span leaf="">, </span><span><span leaf="">'feature_types'</span></span><span leaf="">, </span><span><span leaf="">'genome'</span></span><span leaf="">, </span><span><span leaf="">'interval'</span></span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">读取数据后，scanpy 显示警告，并非所有变量名都是唯一的。这表明一些变量（基因）出现了不止一次，这可能导致下游分析任务出现错误或意外行为。执行 </span><code><span leaf="">var_names_make_unique()</span></code><span leaf="">，通过将数字字符串附加到每个重复的索引元素（“1”、“2” 等）上来使变量名唯一。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">adata.var_names_make_unique()</span><span leaf=""><br></span><span leaf="">adata</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">AnnData object </span><span><span leaf="">with</span></span><span leaf=""> n_obs × n_vars = </span><span><span leaf="">16934</span></span><span leaf=""> × </span><span><span leaf="">36601</span></span><span leaf=""><br></span><span leaf="">    var: </span><span><span leaf="">'gene_ids'</span></span><span leaf="">, </span><span><span leaf="">'feature_types'</span></span><span leaf="">, </span><span><span leaf="">'genome'</span></span><span leaf="">, </span><span><span leaf="">'interval'</span></span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">这个数据集是 </span><code><span leaf="">n_obs × n_vars = 16934 × 36601</span></code><span leaf=""> ，其中 </span><code><span leaf="">n_obs</span></code><span leaf=""> 表示 </span><code><span leaf="">barcodes</span></code><span leaf="">，</span><code><span leaf="">n_vars</span></code><span leaf=""> 为 </span><code><span leaf="">number of transcripts</span></code><span leaf=""> 转录本的数目。这里的 </span><code><span leaf="">barcodes/observation</span></code><span leaf=""> 在大多数的下游分析中假设数据集中的每个观测值代表来自一个完整单细胞的测量结果，在某些情况下这种假设可能被低质量细胞、细胞外游离RNA或者双细胞所破坏。另外还有 </span><code><span leaf="">.var</span></code><span leaf=""> 中包括 Ensemble ID、特征类型、基因组和所在位置在内的更多信息，可以简单看一下：</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">adata.var.head()</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2Los8bEPEs9LdQB954UWZlvJcXoZaFborCuJnAQpfTq7LbC7WWdQ2uiajKzqkuB98Picmy48z3kN2kWlhA/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.21448467966573817" data-type="png" data-w="718" data-imgfileid="100052369" src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2Los8bEPEs9LdQB954UWZlvJcXoZaFborCuJnAQpfTq7LbC7WWdQ2uiajKzqkuB98Picmy48z3kN2kWlhA/640?wx_fmt=png&amp;from=appmsg"></span></figure><h2 data-tool="mdnice编辑器"><span><span leaf="">1.3 过滤低质量细胞</span></span><span></span></h2><p data-tool="mdnice编辑器"><span leaf="">质量控制的第一步是从数据集中移除低质量细胞。当一个细胞检测到的基因数量少、计数深度低且线粒体计数比例高时，它可能具有破裂的细胞膜，这可能表明细胞正在死亡。由于这些细胞通常不是分析的主要目标，并且可能扭曲下游分析，我们会在质量控制期间将其移除。为了识别它们，我们定义细胞质量控制（QC）阈值。细胞QC通常基于以下三个 QC 协变量进行：</span></p><ol><li><section><span leaf="">每个条形码的计数数量（计数深度）</span></section></li><li><section><span leaf="">每个条形码检测到的基因数量</span></section></li><li><section><span leaf="">每个条形码中线粒体基因计数的比例</span></section></li></ol><p data-tool="mdnice编辑器"><span leaf="">在细胞 QC 中，这些协变量通过阈值设定进行过滤。低计数深度、检测到的基因少以及高比例的线粒体 reads 可能反映了细胞膜破裂的细胞，其胞质mRNA已泄漏出去，因此只有线粒体中的mRNA仍然存在。然而，必须联合考虑这三个 QC 协变量，否则可能导致对细胞信号的错误解释。</span><em><span leaf="">例如，具有相对较高线粒体计数比例的细胞可能参与呼吸过程，不应被过滤掉。而具有低或高计数的细胞可能对应于静止细胞群或体积较大的细胞。</span></em><span leaf="">因此，设定阈值时最好考虑多个协变量。一般来说，建议排除更少的细胞，并尽可能宽松，以避免过滤掉有活力的细胞群或小的亚群。</span></p><p data-tool="mdnice编辑器"><span leaf="">对于少量或小型数据集的 QC 通常通过手动方式进行，即查看不同 QC 协变量的分布并识别离群值，随后进行过滤。然而，随着数据集规模增大，这项任务变得越来越耗时，可能值得考虑使用</span><code><span leaf="">MAD</span></code><span leaf="">（中位数绝对偏差）进行自动阈值设定。</span></p><section><span data-tool="mdnice编辑器"><section data-formula="MAD=median(|X_i-median(X)|)"><svg xmlns="http://www.w3.org/2000/svg" role="img" focusable="false" viewbox="0 -750 15947 1000" aria-hidden="true"><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="matrix(1 0 0 -1 0 0)"><g data-mml-node="math"><g data-mml-node="mi"><path data-c="4D" d="M289 629Q289 635 232 637Q208 637 201 638T194 648Q194 649 196 659Q197 662 198 666T199 671T201 676T203 679T207 681T212 683T220 683T232 684Q238 684 262 684T307 683Q386 683 398 683T414 678Q415 674 451 396L487 117L510 154Q534 190 574 254T662 394Q837 673 839 675Q840 676 842 678T846 681L852 683H948Q965 683 988 683T1017 684Q1051 684 1051 673Q1051 668 1048 656T1045 643Q1041 637 1008 637Q968 636 957 634T939 623Q936 618 867 340T797 59Q797 55 798 54T805 50T822 48T855 46H886Q892 37 892 35Q892 19 885 5Q880 0 869 0Q864 0 828 1T736 2Q675 2 644 2T609 1Q592 1 592 11Q592 13 594 25Q598 41 602 43T625 46Q652 46 685 49Q699 52 704 61Q706 65 742 207T813 490T848 631L654 322Q458 10 453 5Q451 4 449 3Q444 0 433 0Q418 0 415 7Q413 11 374 317L335 624L267 354Q200 88 200 79Q206 46 272 46H282Q288 41 289 37T286 19Q282 3 278 1Q274 0 267 0Q265 0 255 0T221 1T157 2Q127 2 95 1T58 0Q43 0 39 2T35 11Q35 13 38 25T43 40Q45 46 65 46Q135 46 154 86Q158 92 223 354T289 629Z"></path></g><g data-mml-node="mi" transform="translate(1051, 0)"><path data-c="41" d="M208 74Q208 50 254 46Q272 46 272 35Q272 34 270 22Q267 8 264 4T251 0Q249 0 239 0T205 1T141 2Q70 2 50 0H42Q35 7 35 11Q37 38 48 46H62Q132 49 164 96Q170 102 345 401T523 704Q530 716 547 716H555H572Q578 707 578 706L606 383Q634 60 636 57Q641 46 701 46Q726 46 726 36Q726 34 723 22Q720 7 718 4T704 0Q701 0 690 0T651 1T578 2Q484 2 455 0H443Q437 6 437 9T439 27Q443 40 445 43L449 46H469Q523 49 533 63L521 213H283L249 155Q208 86 208 74ZM516 260Q516 271 504 416T490 562L463 519Q447 492 400 412L310 260L413 259Q516 259 516 260Z"></path></g><g data-mml-node="mi" transform="translate(1801, 0)"><path data-c="44" d="M287 628Q287 635 230 637Q207 637 200 638T193 647Q193 655 197 667T204 682Q206 683 403 683Q570 682 590 682T630 676Q702 659 752 597T803 431Q803 275 696 151T444 3L430 1L236 0H125H72Q48 0 41 2T33 11Q33 13 36 25Q40 41 44 43T67 46Q94 46 127 49Q141 52 146 61Q149 65 218 339T287 628ZM703 469Q703 507 692 537T666 584T629 613T590 629T555 636Q553 636 541 636T512 636T479 637H436Q392 637 386 627Q384 623 313 339T242 52Q242 48 253 48T330 47Q335 47 349 47T373 46Q499 46 581 128Q617 164 640 212T683 339T703 469Z"></path></g><g data-mml-node="mo" transform="translate(2906.8, 0)"><path data-c="3D" d="M56 347Q56 360 70 367H707Q722 359 722 347Q722 336 708 328L390 327H72Q56 332 56 347ZM56 153Q56 168 72 173H708Q722 163 722 153Q722 140 707 133H70Q56 140 56 153Z"></path></g><g data-mml-node="mi" transform="translate(3962.6, 0)"><path data-c="6D" d="M21 287Q22 293 24 303T36 341T56 388T88 425T132 442T175 435T205 417T221 395T229 376L231 369Q231 367 232 367L243 378Q303 442 384 442Q401 442 415 440T441 433T460 423T475 411T485 398T493 385T497 373T500 364T502 357L510 367Q573 442 659 442Q713 442 746 415T780 336Q780 285 742 178T704 50Q705 36 709 31T724 26Q752 26 776 56T815 138Q818 149 821 151T837 153Q857 153 857 145Q857 144 853 130Q845 101 831 73T785 17T716 -10Q669 -10 648 17T627 73Q627 92 663 193T700 345Q700 404 656 404H651Q565 404 506 303L499 291L466 157Q433 26 428 16Q415 -11 385 -11Q372 -11 364 -4T353 8T350 18Q350 29 384 161L420 307Q423 322 423 345Q423 404 379 404H374Q288 404 229 303L222 291L189 157Q156 26 151 16Q138 -11 108 -11Q95 -11 87 -5T76 7T74 17Q74 30 112 181Q151 335 151 342Q154 357 154 369Q154 405 129 405Q107 405 92 377T69 316T57 280Q55 278 41 278H27Q21 284 21 287Z"></path></g><g data-mml-node="mi" transform="translate(4840.6, 0)"><path data-c="65" d="M39 168Q39 225 58 272T107 350T174 402T244 433T307 442H310Q355 442 388 420T421 355Q421 265 310 237Q261 224 176 223Q139 223 138 221Q138 219 132 186T125 128Q125 81 146 54T209 26T302 45T394 111Q403 121 406 121Q410 121 419 112T429 98T420 82T390 55T344 24T281 -1T205 -11Q126 -11 83 42T39 168ZM373 353Q367 405 305 405Q272 405 244 391T199 357T170 316T154 280T149 261Q149 260 169 260Q282 260 327 284T373 353Z"></path></g><g data-mml-node="mi" transform="translate(5306.6, 0)"><path data-c="64" d="M366 683Q367 683 438 688T511 694Q523 694 523 686Q523 679 450 384T375 83T374 68Q374 26 402 26Q411 27 422 35Q443 55 463 131Q469 151 473 152Q475 153 483 153H487H491Q506 153 506 145Q506 140 503 129Q490 79 473 48T445 8T417 -8Q409 -10 393 -10Q359 -10 336 5T306 36L300 51Q299 52 296 50Q294 48 292 46Q233 -10 172 -10Q117 -10 75 30T33 157Q33 205 53 255T101 341Q148 398 195 420T280 442Q336 442 364 400Q369 394 369 396Q370 400 396 505T424 616Q424 629 417 632T378 637H357Q351 643 351 645T353 664Q358 683 366 683ZM352 326Q329 405 277 405Q242 405 210 374T160 293Q131 214 119 129Q119 126 119 118T118 106Q118 61 136 44T179 26Q233 26 290 98L298 109L352 326Z"></path></g><g data-mml-node="mi" transform="translate(5826.6, 0)"><path data-c="69" d="M184 600Q184 624 203 642T247 661Q265 661 277 649T290 619Q290 596 270 577T226 557Q211 557 198 567T184 600ZM21 287Q21 295 30 318T54 369T98 420T158 442Q197 442 223 419T250 357Q250 340 236 301T196 196T154 83Q149 61 149 51Q149 26 166 26Q175 26 185 29T208 43T235 78T260 137Q263 149 265 151T282 153Q302 153 302 143Q302 135 293 112T268 61T223 11T161 -11Q129 -11 102 10T74 74Q74 91 79 106T122 220Q160 321 166 341T173 380Q173 404 156 404H154Q124 404 99 371T61 287Q60 286 59 284T58 281T56 279T53 278T49 278T41 278H27Q21 284 21 287Z"></path></g><g data-mml-node="mi" transform="translate(6171.6, 0)"><path data-c="61" d="M33 157Q33 258 109 349T280 441Q331 441 370 392Q386 422 416 422Q429 422 439 414T449 394Q449 381 412 234T374 68Q374 43 381 35T402 26Q411 27 422 35Q443 55 463 131Q469 151 473 152Q475 153 483 153H487Q506 153 506 144Q506 138 501 117T481 63T449 13Q436 0 417 -8Q409 -10 393 -10Q359 -10 336 5T306 36L300 51Q299 52 296 50Q294 48 292 46Q233 -10 172 -10Q117 -10 75 30T33 157ZM351 328Q351 334 346 350T323 385T277 405Q242 405 210 374T160 293Q131 214 119 129Q119 126 119 118T118 106Q118 61 136 44T179 26Q217 26 254 59T298 110Q300 114 325 217T351 328Z"></path></g><g data-mml-node="mi" transform="translate(6700.6, 0)"><path data-c="6E" d="M21 287Q22 293 24 303T36 341T56 388T89 425T135 442Q171 442 195 424T225 390T231 369Q231 367 232 367L243 378Q304 442 382 442Q436 442 469 415T503 336T465 179T427 52Q427 26 444 26Q450 26 453 27Q482 32 505 65T540 145Q542 153 560 153Q580 153 580 145Q580 144 576 130Q568 101 554 73T508 17T439 -10Q392 -10 371 17T350 73Q350 92 386 193T423 345Q423 404 379 404H374Q288 404 229 303L222 291L189 157Q156 26 151 16Q138 -11 108 -11Q95 -11 87 -5T76 7T74 17Q74 30 112 180T152 343Q153 348 153 366Q153 405 129 405Q91 405 66 305Q60 285 60 284Q58 278 41 278H27Q21 284 21 287Z"></path></g><g data-mml-node="mo" transform="translate(7300.6, 0)"><path data-c="28" d="M94 250Q94 319 104 381T127 488T164 576T202 643T244 695T277 729T302 750H315H319Q333 750 333 741Q333 738 316 720T275 667T226 581T184 443T167 250T184 58T225 -81T274 -167T316 -220T333 -241Q333 -250 318 -250H315H302L274 -226Q180 -141 137 -14T94 250Z"></path></g><g data-mml-node="mo" transform="translate(7689.6, 0)"><path data-c="7C" d="M139 -249H137Q125 -249 119 -235V251L120 737Q130 750 139 750Q152 750 159 735V-235Q151 -249 141 -249H139Z"></path></g><g data-mml-node="msub" transform="translate(7967.6, 0)"><g data-mml-node="mi"><path data-c="58" d="M42 0H40Q26 0 26 11Q26 15 29 27Q33 41 36 43T55 46Q141 49 190 98Q200 108 306 224T411 342Q302 620 297 625Q288 636 234 637H206Q200 643 200 645T202 664Q206 677 212 683H226Q260 681 347 681Q380 681 408 681T453 682T473 682Q490 682 490 671Q490 670 488 658Q484 643 481 640T465 637Q434 634 411 620L488 426L541 485Q646 598 646 610Q646 628 622 635Q617 635 609 637Q594 637 594 648Q594 650 596 664Q600 677 606 683H618Q619 683 643 683T697 681T738 680Q828 680 837 683H845Q852 676 852 672Q850 647 840 637H824Q790 636 763 628T722 611T698 593L687 584Q687 585 592 480L505 384Q505 383 536 304T601 142T638 56Q648 47 699 46Q734 46 734 37Q734 35 732 23Q728 7 725 4T711 1Q708 1 678 1T589 2Q528 2 496 2T461 1Q444 1 444 10Q444 11 446 25Q448 35 450 39T455 44T464 46T480 47T506 54Q523 62 523 64Q522 64 476 181L429 299Q241 95 236 84Q232 76 232 72Q232 53 261 47Q262 47 267 47T273 46Q276 46 277 46T280 45T283 42T284 35Q284 26 282 19Q279 6 276 4T261 1Q258 1 243 1T201 2T142 2Q64 2 42 0Z"></path></g><g data-mml-node="mi" transform="translate(828, -150) scale(0.707)"><path data-c="69" d="M184 600Q184 624 203 642T247 661Q265 661 277 649T290 619Q290 596 270 577T226 557Q211 557 198 567T184 600ZM21 287Q21 295 30 318T54 369T98 420T158 442Q197 442 223 419T250 357Q250 340 236 301T196 196T154 83Q149 61 149 51Q149 26 166 26Q175 26 185 29T208 43T235 78T260 137Q263 149 265 151T282 153Q302 153 302 143Q302 135 293 112T268 61T223 11T161 -11Q129 -11 102 10T74 74Q74 91 79 106T122 220Q160 321 166 341T173 380Q173 404 156 404H154Q124 404 99 371T61 287Q60 286 59 284T58 281T56 279T53 278T49 278T41 278H27Q21 284 21 287Z"></path></g></g><g data-mml-node="mo" transform="translate(9311.7, 0)"><path data-c="2212" d="M84 237T84 250T98 270H679Q694 262 694 250T679 230H98Q84 237 84 250Z"></path></g><g data-mml-node="mi" transform="translate(10312, 0)"><path data-c="6D" d="M21 287Q22 293 24 303T36 341T56 388T88 425T132 442T175 435T205 417T221 395T229 376L231 369Q231 367 232 367L243 378Q303 442 384 442Q401 442 415 440T441 433T460 423T475 411T485 398T493 385T497 373T500 364T502 357L510 367Q573 442 659 442Q713 442 746 415T780 336Q780 285 742 178T704 50Q705 36 709 31T724 26Q752 26 776 56T815 138Q818 149 821 151T837 153Q857 153 857 145Q857 144 853 130Q845 101 831 73T785 17T716 -10Q669 -10 648 17T627 73Q627 92 663 193T700 345Q700 404 656 404H651Q565 404 506 303L499 291L466 157Q433 26 428 16Q415 -11 385 -11Q372 -11 364 -4T353 8T350 18Q350 29 384 161L420 307Q423 322 423 345Q423 404 379 404H374Q288 404 229 303L222 291L189 157Q156 26 151 16Q138 -11 108 -11Q95 -11 87 -5T76 7T74 17Q74 30 112 181Q151 335 151 342Q154 357 154 369Q154 405 129 405Q107 405 92 377T69 316T57 280Q55 278 41 278H27Q21 284 21 287Z"></path></g><g data-mml-node="mi" transform="translate(11190, 0)"><path data-c="65" d="M39 168Q39 225 58 272T107 350T174 402T244 433T307 442H310Q355 442 388 420T421 355Q421 265 310 237Q261 224 176 223Q139 223 138 221Q138 219 132 186T125 128Q125 81 146 54T209 26T302 45T394 111Q403 121 406 121Q410 121 419 112T429 98T420 82T390 55T344 24T281 -1T205 -11Q126 -11 83 42T39 168ZM373 353Q367 405 305 405Q272 405 244 391T199 357T170 316T154 280T149 261Q149 260 169 260Q282 260 327 284T373 353Z"></path></g><g data-mml-node="mi" transform="translate(11656, 0)"><path data-c="64" d="M366 683Q367 683 438 688T511 694Q523 694 523 686Q523 679 450 384T375 83T374 68Q374 26 402 26Q411 27 422 35Q443 55 463 131Q469 151 473 152Q475 153 483 153H487H491Q506 153 506 145Q506 140 503 129Q490 79 473 48T445 8T417 -8Q409 -10 393 -10Q359 -10 336 5T306 36L300 51Q299 52 296 50Q294 48 292 46Q233 -10 172 -10Q117 -10 75 30T33 157Q33 205 53 255T101 341Q148 398 195 420T280 442Q336 442 364 400Q369 394 369 396Q370 400 396 505T424 616Q424 629 417 632T378 637H357Q351 643 351 645T353 664Q358 683 366 683ZM352 326Q329 405 277 405Q242 405 210 374T160 293Q131 214 119 129Q119 126 119 118T118 106Q118 61 136 44T179 26Q233 26 290 98L298 109L352 326Z"></path></g><g data-mml-node="mi" transform="translate(12176, 0)"><path data-c="69" d="M184 600Q184 624 203 642T247 661Q265 661 277 649T290 619Q290 596 270 577T226 557Q211 557 198 567T184 600ZM21 287Q21 295 30 318T54 369T98 420T158 442Q197 442 223 419T250 357Q250 340 236 301T196 196T154 83Q149 61 149 51Q149 26 166 26Q175 26 185 29T208 43T235 78T260 137Q263 149 265 151T282 153Q302 153 302 143Q302 135 293 112T268 61T223 11T161 -11Q129 -11 102 10T74 74Q74 91 79 106T122 220Q160 321 166 341T173 380Q173 404 156 404H154Q124 404 99 371T61 287Q60 286 59 284T58 281T56 279T53 278T49 278T41 278H27Q21 284 21 287Z"></path></g><g data-mml-node="mi" transform="translate(12521, 0)"><path data-c="61" d="M33 157Q33 258 109 349T280 441Q331 441 370 392Q386 422 416 422Q429 422 439 414T449 394Q449 381 412 234T374 68Q374 43 381 35T402 26Q411 27 422 35Q443 55 463 131Q469 151 473 152Q475 153 483 153H487Q506 153 506 144Q506 138 501 117T481 63T449 13Q436 0 417 -8Q409 -10 393 -10Q359 -10 336 5T306 36L300 51Q299 52 296 50Q294 48 292 46Q233 -10 172 -10Q117 -10 75 30T33 157ZM351 328Q351 334 346 350T323 385T277 405Q242 405 210 374T160 293Q131 214 119 129Q119 126 119 118T118 106Q118 61 136 44T179 26Q217 26 254 59T298 110Q300 114 325 217T351 328Z"></path></g><g data-mml-node="mi" transform="translate(13050, 0)"><path data-c="6E" d="M21 287Q22 293 24 303T36 341T56 388T89 425T135 442Q171 442 195 424T225 390T231 369Q231 367 232 367L243 378Q304 442 382 442Q436 442 469 415T503 336T465 179T427 52Q427 26 444 26Q450 26 453 27Q482 32 505 65T540 145Q542 153 560 153Q580 153 580 145Q580 144 576 130Q568 101 554 73T508 17T439 -10Q392 -10 371 17T350 73Q350 92 386 193T423 345Q423 404 379 404H374Q288 404 229 303L222 291L189 157Q156 26 151 16Q138 -11 108 -11Q95 -11 87 -5T76 7T74 17Q74 30 112 180T152 343Q153 348 153 366Q153 405 129 405Q91 405 66 305Q60 285 60 284Q58 278 41 278H27Q21 284 21 287Z"></path></g><g data-mml-node="mo" transform="translate(13650, 0)"><path data-c="28" d="M94 250Q94 319 104 381T127 488T164 576T202 643T244 695T277 729T302 750H315H319Q333 750 333 741Q333 738 316 720T275 667T226 581T184 443T167 250T184 58T225 -81T274 -167T316 -220T333 -241Q333 -250 318 -250H315H302L274 -226Q180 -141 137 -14T94 250Z"></path></g><g data-mml-node="mi" transform="translate(14039, 0)"><path data-c="58" d="M42 0H40Q26 0 26 11Q26 15 29 27Q33 41 36 43T55 46Q141 49 190 98Q200 108 306 224T411 342Q302 620 297 625Q288 636 234 637H206Q200 643 200 645T202 664Q206 677 212 683H226Q260 681 347 681Q380 681 408 681T453 682T473 682Q490 682 490 671Q490 670 488 658Q484 643 481 640T465 637Q434 634 411 620L488 426L541 485Q646 598 646 610Q646 628 622 635Q617 635 609 637Q594 637 594 648Q594 650 596 664Q600 677 606 683H618Q619 683 643 683T697 681T738 680Q828 680 837 683H845Q852 676 852 672Q850 647 840 637H824Q790 636 763 628T722 611T698 593L687 584Q687 585 592 480L505 384Q505 383 536 304T601 142T638 56Q648 47 699 46Q734 46 734 37Q734 35 732 23Q728 7 725 4T711 1Q708 1 678 1T589 2Q528 2 496 2T461 1Q444 1 444 10Q444 11 446 25Q448 35 450 39T455 44T464 46T480 47T506 54Q523 62 523 64Q522 64 476 181L429 299Q241 95 236 84Q232 76 232 72Q232 53 261 47Q262 47 267 47T273 46Q276 46 277 46T280 45T283 42T284 35Q284 26 282 19Q279 6 276 4T261 1Q258 1 243 1T201 2T142 2Q64 2 42 0Z"></path></g><g data-mml-node="mo" transform="translate(14891, 0)"><path data-c="29" d="M60 749L64 750Q69 750 74 750H86L114 726Q208 641 251 514T294 250Q294 182 284 119T261 12T224 -76T186 -143T145 -194T113 -227T90 -246Q87 -249 86 -250H74Q66 -250 63 -250T58 -247T55 -238Q56 -237 66 -225Q221 -64 221 250T66 725Q56 737 55 738Q55 746 60 749Z"></path></g><g data-mml-node="mo" transform="translate(15280, 0)"><path data-c="7C" d="M139 -249H137Q125 -249 119 -235V251L120 737Q130 750 139 750Q152 750 159 735V-235Q151 -249 141 -249H139Z"></path></g><g data-mml-node="mo" transform="translate(15558, 0)"><path data-c="29" d="M60 749L64 750Q69 750 74 750H86L114 726Q208 641 251 514T294 250Q294 182 284 119T261 12T224 -76T186 -143T145 -194T113 -227T90 -246Q87 -249 86 -250H74Q66 -250 63 -250T58 -247T55 -238Q56 -237 66 -225Q221 -64 221 250T66 725Q56 737 55 738Q55 746 60 749Z"></path></g></g></g><g></g></svg></section></span></section><p data-tool="mdnice编辑器"><span leaf="">其中 X</span><sub><span leaf="">i</span></sub><span leaf=""> 是观测值 i 对应的 QC 指标。如果细胞偏离中位数超过 5 个 </span><code><span leaf="">MAD</span></code><span leaf=""> ，我们将其标记为离群值，这是一种相对宽松的过滤策略。需要强调的是，在细胞注释后重新评估过滤策略可能是合理的。</span></p><p data-tool="mdnice编辑器"><span leaf="">计算 QC 协变量使用的是 scanpy 函数 </span><code><span leaf="">sc.pp.calculate_qc_metrics</span></code><span leaf="">，该函数还可以计算特定基因群计数的比例。因此，我们定义线粒体基因、核糖体基因和血红蛋白基因。需要注意的是，根据数据集中考虑的物种，线粒体计数要么以 "mt-" 为前缀，要么以 "MT-" 为前缀进行注释。这里使用的数据集是人类骨髓，因此以 "MT-" 为前缀，小鼠数据集前缀通常是小写 "mt-"。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># mitochondrial genes 线粒体基因</span></span><span leaf=""><br></span><span leaf="">adata.var[</span><span><span leaf="">"mt"</span></span><span leaf="">] = adata.var_names.str.startswith(</span><span><span leaf="">"MT-"</span></span><span leaf="">)</span><span leaf=""><br></span><span><span leaf=""># ribosomal genes 核糖体基因</span></span><span leaf=""><br></span><span leaf="">adata.var[</span><span><span leaf="">"ribo"</span></span><span leaf="">] = adata.var_names.str.startswith((</span><span><span leaf="">"RPS"</span></span><span leaf="">, </span><span><span leaf="">"RPL"</span></span><span leaf="">))</span><span leaf=""><br></span><span><span leaf=""># hemoglobin genes 血红蛋白基因</span></span><span leaf=""><br></span><span leaf="">adata.var[</span><span><span leaf="">"hb"</span></span><span leaf="">] = adata.var_names.str.contains(</span><span><span leaf="">"^HB[^(P)]"</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">sc.pp.calculate_qc_metrics(</span><span leaf=""><br></span><span leaf="">    adata, qc_vars=[</span><span><span leaf="">"mt"</span></span><span leaf="">, </span><span><span leaf="">"ribo"</span></span><span leaf="">, </span><span><span leaf="">"hb"</span></span><span leaf="">], inplace=</span><span><span leaf="">True</span></span><span leaf="">, percent_top=[</span><span><span leaf="">20</span></span><span leaf="">], log1p=</span><span><span leaf="">True</span></span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">adata</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">AnnData object </span><span><span leaf="">with</span></span><span leaf=""> n_obs × n_vars = </span><span><span leaf="">16934</span></span><span leaf=""> × </span><span><span leaf="">36601</span></span><span leaf=""><br></span><span leaf="">    obs: </span><span><span leaf="">'n_genes_by_counts'</span></span><span leaf="">, </span><span><span leaf="">'log1p_n_genes_by_counts'</span></span><span leaf="">, </span><span><span leaf="">'total_counts'</span></span><span leaf="">, </span><span><span leaf="">'log1p_total_counts'</span></span><span leaf="">, </span><span><span leaf="">'pct_counts_in_top_20_genes'</span></span><span leaf="">, </span><span><span leaf="">'total_counts_mt'</span></span><span leaf="">, </span><span><span leaf="">'log1p_total_counts_mt'</span></span><span leaf="">, </span><span><span leaf="">'pct_counts_mt'</span></span><span leaf="">, </span><span><span leaf="">'total_counts_ribo'</span></span><span leaf="">, </span><span><span leaf="">'log1p_total_counts_ribo'</span></span><span leaf="">, </span><span><span leaf="">'pct_counts_ribo'</span></span><span leaf="">, </span><span><span leaf="">'total_counts_hb'</span></span><span leaf="">, </span><span><span leaf="">'log1p_total_counts_hb'</span></span><span leaf="">, </span><span><span leaf="">'pct_counts_hb'</span></span><span leaf=""><br></span><span leaf="">    var: </span><span><span leaf="">'gene_ids'</span></span><span leaf="">, </span><span><span leaf="">'feature_types'</span></span><span leaf="">, </span><span><span leaf="">'genome'</span></span><span leaf="">, </span><span><span leaf="">'interval'</span></span><span leaf="">, </span><span><span leaf="">'mt'</span></span><span leaf="">, </span><span><span leaf="">'ribo'</span></span><span leaf="">, </span><span><span leaf="">'hb'</span></span><span leaf="">, </span><span><span leaf="">'n_cells_by_counts'</span></span><span leaf="">, </span><span><span leaf="">'mean_counts'</span></span><span leaf="">, </span><span><span leaf="">'log1p_mean_counts'</span></span><span leaf="">, </span><span><span leaf="">'pct_dropout_by_counts'</span></span><span leaf="">, </span><span><span leaf="">'total_counts'</span></span><span leaf="">, </span><span><span leaf="">'log1p_total_counts'</span></span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">可以看到这个函数向 </span><code><span leaf="">.obs</span></code><span leaf=""> 和 </span><code><span leaf="">.var</span></code><span leaf=""> 添加了一些额外的列。重点介绍其中的几个（更多关于不同指标的信息可以在 scanpy 文档中找到）：</span></p><ul><li><section><code><span leaf="">n_genes_by_counts</span></code><span leaf=""> 是一个细胞中有表达的基因数量。</span></section></li><li><section><code><span leaf="">total_counts</span></code><span leaf=""> 是一个细胞的总计数数量，这也可能被称为文库大小（library size）。</span></section></li><li><section><code><span leaf="">pct_counts_mt</span></code><span leaf=""> 是一个细胞中线粒体计数占总计数的比例。</span></section></li></ul><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">fig, axes = plt.subplots(</span><span><span leaf="">1</span></span><span leaf="">, </span><span><span leaf="">3</span></span><span leaf="">, figsize=(</span><span><span leaf="">18</span></span><span leaf="">, </span><span><span leaf="">5</span></span><span leaf="">))</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">sns.histplot(adata.obs[</span><span><span leaf="">"total_counts"</span></span><span leaf="">], bins=</span><span><span leaf="">100</span></span><span leaf="">, ax=axes[</span><span><span leaf="">0</span></span><span leaf="">])</span><span leaf=""><br></span><span leaf="">sc.pl.violin(adata, </span><span><span leaf="">'pct_counts_mt'</span></span><span leaf="">, ax=axes[</span><span><span leaf="">1</span></span><span leaf="">], show=</span><span><span leaf="">False</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">sc.pl.scatter(adata, </span><span><span leaf="">'total_counts'</span></span><span leaf="">, </span><span><span leaf="">'n_genes_by_counts'</span></span><span leaf="">, ax=axes[</span><span><span leaf="">2</span></span><span leaf="">], show=</span><span><span leaf="">False</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">plt.tight_layout()</span><span leaf=""><br></span><span leaf="">plt.show()</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2Los8bEPEs9LdQB954UWZlvJcXaQBCmSC3sPZSeDq2LibIldl2aicgYUcev7KPMwZuZgz5QXFGAef9qYzA/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.262037037037037" data-type="png" data-w="1080" data-imgfileid="100052371" src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2Los8bEPEs9LdQB954UWZlvJcXaQBCmSC3sPZSeDq2LibIldl2aicgYUcev7KPMwZuZgz5QXFGAef9qYzA/640?wx_fmt=png&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">这些图表明一些细胞具有相对较高的线粒体计数比例，这通常与细胞退化有关。但由于每个细胞的计数数量足够高，且大多数细胞的线粒体 reads 比例低于 20%，我们仍然可以处理这些数据。基于这些图，可以定义用于过滤细胞的手动阈值。接下来，我们将展示基于 </span><code><span leaf="">MAD</span></code><span leaf=""> 的自动阈值设定和过滤。</span></p><p data-tool="mdnice编辑器"><span leaf="">首先，我们定义一个函数 </span><code><span leaf="">is_outlier()</span></code><span leaf=""> ，它有两个参数 </span><code><span leaf="">metric</span></code><span leaf="">（</span><code><span leaf="">.obs</span></code><span leaf=""> 中的一列） 和 </span><code><span leaf="">nmads</span></code><span leaf="">（过滤允许的 </span><code><span leaf="">MAD</span></code><span leaf=""> 数量）：</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span><span leaf="">def</span></span><span leaf=""> </span><span><span leaf="">is_outlier</span></span><span><span leaf="">(adata, metric: str, nmads: int)</span></span><span leaf="">:</span></span><span leaf=""><br></span><span leaf="">    M = adata.obs[metric]</span><span leaf=""><br></span><span leaf="">    outlier = (M &lt; np.median(M) - nmads * median_abs_deviation(M)) | (</span><span leaf=""><br></span><span leaf="">        np.median(M) + nmads * median_abs_deviation(M) &lt; M</span><span leaf=""><br></span><span leaf="">    )</span><span leaf=""><br></span><span leaf="">    </span><span><span leaf="">return</span></span><span leaf=""> outlier</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">我们应用于 </span><code><span leaf="">log1p_total_counts</span></code><span leaf="">、</span><code><span leaf="">log1p_n_genes_by_counts</span></code><span leaf=""> 和 </span><code><span leaf="">pct_counts_in_top_20_genes</span></code><span leaf=""> 这三个 QC 协变量，每个都使用 5 个 MAD 的阈值。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">adata.obs[</span><span><span leaf="">"outlier"</span></span><span leaf="">] = (</span><span leaf=""><br></span><span leaf="">    is_outlier(adata, </span><span><span leaf="">"log1p_total_counts"</span></span><span leaf="">, </span><span><span leaf="">5</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">    | is_outlier(adata, </span><span><span leaf="">"log1p_n_genes_by_counts"</span></span><span leaf="">, </span><span><span leaf="">5</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">    | is_outlier(adata, </span><span><span leaf="">"pct_counts_in_top_20_genes"</span></span><span leaf="">, </span><span><span leaf="">5</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">adata.obs.outlier.value_counts()</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">outlier</span><span leaf=""><br></span><span><span leaf="">False</span></span><span leaf="">    </span><span><span leaf="">16065</span></span><span leaf=""><br></span><span><span leaf="">True</span></span><span leaf="">       </span><span><span leaf="">869</span></span><span leaf=""><br></span><span leaf="">Name: count, dtype: int64</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><code><span leaf="">pct_counts_mt</span></code><span leaf=""> 使用 3 个 MAD 进行过滤。此外，过滤掉线粒体计数比例超过 8% 的细胞。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">adata.obs[</span><span><span leaf="">"mt_outlier"</span></span><span leaf="">] = is_outlier(adata, </span><span><span leaf="">"pct_counts_mt"</span></span><span leaf="">, </span><span><span leaf="">3</span></span><span leaf="">) | (</span><span leaf=""><br></span><span leaf="">    adata.obs[</span><span><span leaf="">"pct_counts_mt"</span></span><span leaf="">] &gt; </span><span><span leaf="">8</span></span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">adata.obs.mt_outlier.value_counts()</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">mt_outlier</span><span leaf=""><br></span><span><span leaf="">False</span></span><span leaf="">    </span><span><span leaf="">15240</span></span><span leaf=""><br></span><span><span leaf="">True</span></span><span leaf="">      </span><span><span leaf="">1694</span></span><span leaf=""><br></span><span leaf="">Name: count, dtype: int64</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">adata = adata[(~adata.obs.outlier) &amp; (~adata.obs.mt_outlier)].copy()</span><span leaf=""><br></span><span leaf="">sc.pl.scatter(adata, </span><span><span leaf="">"total_counts"</span></span><span leaf="">, </span><span><span leaf="">"n_genes_by_counts"</span></span><span leaf="">, color=</span><span><span leaf="">"pct_counts_mt"</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2Los8bEPEs9LdQB954UWZlvJcX3HYWfhSSN36ToicMpevzy0ptu6d5ENmsMRn6UiaZVbCx3ib5CWwGjU40Q/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.8744186046511628" data-type="png" data-w="430" data-imgfileid="100052368" src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2Los8bEPEs9LdQB954UWZlvJcX3HYWfhSSN36ToicMpevzy0ptu6d5ENmsMRn6UiaZVbCx3ib5CWwGjU40Q/640?wx_fmt=png&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">过滤掉低质量细胞之后还剩 14814 个细胞。</span></p><h2 data-tool="mdnice编辑器"><span><span leaf="">1.4 环境RNA（cell-free RNA）校正</span></span><span></span></h2><p data-tool="mdnice编辑器"><span leaf="">对于基于液滴的单细胞 RNA 测序实验，稀释液中存在一定量的背景 mRNA，这些 mRNA 与细胞一起被分配到液滴中并进行测序。这样做的最终效果是产生背景污染，该污染代表的不是来自液滴内细胞的表达，而是来自包含细胞的溶液。游离 mRNA 分子代表稀释液中存在的背景 mRNA。输入溶液中游离 mRNA 的这种污染通常称为 “soup”，它是由细胞裂解产生的。</span></p><p data-tool="mdnice编辑器"><span leaf="">什么时候需要去除环境 RNA 呢？各种实验因素导致样本中死细胞比例过高，观察到明显的“汤效应”（Soup-like Profile），例如在初步分析中，发现许多细胞都表达一些通常只在特定稀有细胞类型中高表达或普遍低表达的基因，或者 QC 指标中每个细胞的线粒体基因比例异常低（因为环境RNA主要来自胞浆 mRNA），这强烈提示环境RNA污染严重。</span></p><p data-tool="mdnice编辑器"><span leaf="">去除环境 mRNA 的方法有 </span><code><span leaf="">SoupX</span></code><span leaf=""> 和 </span><code><span leaf="">DecontX</span></code><span leaf="">，都是基于 R 的，详细内容可以参考 https://www.sc-best-practices.org/preprocessing_visualization/quality_control.html。</span></p><h2 data-tool="mdnice编辑器"><span><span leaf="">1.5 Doublets 检测</span></span><span></span></h2><p data-tool="mdnice编辑器"><span leaf="">Doublets 被定义为在同一个细胞条形码下测序的两个细胞，它们被捕获在同一个液滴中。如果 doublets 由相同细胞类型（但来自不同个体）形成，就是同型（homotypic）的，否则就是异型（heterotypic）的。同型双细胞不一定可以从计数矩阵中识别，并且通常被认为是无害的，因为它们可以通过 cell hashing 或 SNP 来识别。异性双细胞的识别至关重要，因为它们很可能被错误分类，并可能导致下游分析步骤失真。</span></p><p data-tool="mdnice编辑器"><span leaf="">《Benchmarking Computational Doublet-Detection Methods for Single-Cell RNA Sequencing Data》这篇文章对九种不同的双细胞检测方法进行了基准测试，并评估了它们在计算效率和双峰检测精度方面的性能，链接：https://www.sciencedirect.com/science/article/pii/S2405471220304592</span></p><p data-tool="mdnice编辑器"><span leaf="">这本书里展示了用 </span><code><span leaf="">scDblFinder</span></code><span leaf=""> R 包实现双细胞检测的过程，详情可看 https://www.sc-best-practices.org/preprocessing_visualization/quality_control.html</span></p><p data-tool="mdnice编辑器"><span leaf="">也可以直接使用 scanpy 包中的 </span><code><span leaf="">scrublet</span></code><span leaf=""> 方法：</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">adata = sc.pp.scrublet(adata, random_state=</span><span><span leaf="">123</span></span><span leaf="">, expected_doublet_rate=</span><span><span leaf="">0.05</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">sc.external.pl.scrublet_score_distribution(adata)</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2Los8bEPEs9LdQB954UWZlvJcXgp3hdK16cttTKHfW6Y8SCocFjndsmYrBNuL6JMak6q85mNjT5pRYHw/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.3513157894736842" data-type="png" data-w="760" data-imgfileid="100052370" src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2Los8bEPEs9LdQB954UWZlvJcXgp3hdK16cttTKHfW6Y8SCocFjndsmYrBNuL6JMak6q85mNjT5pRYHw/640?wx_fmt=png&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">然后直接过滤掉就可以了</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">adata_clean = adata[adata.obs[</span><span><span leaf="">"predicted_doublet"</span></span><span leaf="">] == False, :].copy()</span><span leaf=""><br></span></code></pre><h2 data-tool="mdnice编辑器"><span><span leaf="">1.6 关键</span></span><span></span></h2><ol><li><section><span leaf="">对低质量细胞的过滤应该基于绝对中位偏差和宽松的阈值，以避免对较小亚群产生偏差。</span></section></li><li><section><span leaf="">可以用 </span><code><span leaf="">scDblFinder</span></code><span leaf=""> 等方法有效识别双细胞。</span></section></li><li><section><span leaf="">双细胞检测方法不应该在多个批次的整合 scRNA-seq 数据上运行。</span></section></li><li><section><span leaf="">Feature-based filtering does not show benefits for downstream tasks. 基于特征的过滤对下游任务没有显示出益处。“Feature-based filtering” 应该指的是 </span><code><span leaf="">sc.pp.filter_genes(adata, min_cells=3)</span></code><span leaf=""> 这种根据特定标准移除表达量过低或在太少细胞中表达的基因，这里的核心意思指的应该是，这种基于基因的过滤步骤未能被证明能显著改善后续分析任务（如下游任务）的质量或结果。</span></section></li></ol><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com" data-pm-slice="0 0 []"><h4 data-tool="mdnice编辑器"><span leaf="">文末友情宣传</span></h4><p data-tool="mdnice编辑器"><span leaf="">强烈建议你推荐给身边的</span><strong><span leaf="">博士后以及年轻生物学PI</span></strong><span leaf="">，多一点数据认知，让他们的科研上一个台阶：</span></p><ul></ul></section><ul><li><section><span leaf=""><a target="_blank" href="https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247544311&amp;idx=1&amp;sn=d41b5838e799f52280e78703135bb603&amp;scene=21#wechat_redirect" textvalue="生信入门&amp;数据挖掘线上直播课8月班" data-itemshowtype="0" linktype="text" data-linktype="2">生信入门&amp;数据挖掘线上直播课8月班</a><span textstyle="">，你的生物信息学入门课</span></span></section></li><li><section><span leaf=""><a target="_blank" href="https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247525079&amp;idx=1&amp;sn=0b997af16a58195b4192691373048fd5&amp;scene=21#wechat_redirect" textvalue="时隔5年，我们的生信技能树VIP学徒继续招生啦" data-itemshowtype="0" linktype="text" data-linktype="2"><span textstyle="">时隔5年，我们的生信技能树VIP学徒继续招生啦</span></a></span></section></li><li><section><span leaf=""><a target="_blank" href="https://mp.weixin.qq.com/s?__biz=MzUzMTEwODk0Ng==&amp;mid=2247530048&amp;idx=1&amp;sn=28aa7bbd5e00521f79e074496a5f5d66&amp;scene=21#wechat_redirect" textvalue="满足你生信分析计算需求的低价解决方案" data-itemshowtype="0" linktype="text" data-linktype="2">满足你生信分析计算需求的低价解决方案</a></span></section></li><li><section><span leaf=""><a target="_blank" href="https://mp.weixin.qq.com/mp/appmsgalbum?__biz=MzAxMDkxODM1Ng==&amp;action=getalbum&amp;album_id=1679199708449144836&amp;scene=173&amp;subscene=207&amp;sessionid=1745492310&amp;enterid=1745492314&amp;from_msgid=2247541298&amp;from_itemidx=1&amp;count=3&amp;nolastread=1#wechat_redirect" textvalue="生信故事会" data-itemshowtype="0" linktype="text" data-linktype="2">生信故事会</a>，来看看他们的生信入门故事</span></section></li><li><section><span leaf=""><a target="_blank" href="https://mp.weixin.qq.com/mp/appmsgalbum?__biz=MzAxMDkxODM1Ng==&amp;action=getalbum&amp;album_id=3690970204957147140&amp;scene=173&amp;subscene=207&amp;sessionid=1745066271&amp;enterid=1745066274&amp;from_msgid=2247540702&amp;from_itemidx=1&amp;count=3&amp;nolastread=1#wechat_redirect" textvalue="生信马拉松答疑专辑" data-itemshowtype="0" linktype="text" data-linktype="2">生信马拉松答疑专辑</a>，获取你的生信专属答疑</span></section></li></ul><section><span leaf=""><br></span></section></section><section><span leaf=""><br></span></section><p><mp-style-type data-value="10000"></mp-style-type></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/BJ5Clb_OmBogjZCZ2aTEVg",target="_blank" rel="noopener noreferrer">原文链接</a>
