---
title: "Science杂志高颜值GSEA打分排序图"
date: 2025-08-26T04:11:22Z
draft: ["false"]
tags: [
  "fetched",
  "珠海健明生物医药科技"
]
categories: ["Acdemic"]
---
Science杂志高颜值GSEA打分排序图 by 珠海健明生物医药科技
------
<div><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com"><blockquote><p><span leaf="">前面给大家介绍过一个高颜值富集分析结果展示图：<a target="_blank" href="https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247537055&amp;idx=1&amp;sn=26544d5687fbe6001391e869ea84e692&amp;scene=21#wechat_redirect" textvalue="" linktype="text" data-linktype="2">一种很新的功能富集结果展示方法</a>。今天来还是来复现这篇 2024 年 6 月份发表在 顶刊 science 杂志上的文献《</span><strong><span leaf="">Defining the KRAS- and ERK-dependent transcriptome in KRAS-mutant cancers</span></strong><span leaf="">》中另一个富集分析GSEA结果图，图片以及含义如下：</span></p></blockquote><h4 data-tool="mdnice编辑器"><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com"><p data-tool="mdnice编辑器"><span leaf="">含义：</span><strong><span leaf="">作者对 KRAS versus NS siRNA 处理的 PDAC 细胞系差异基因进行 GSEA 分析，基因集为 50 个 Hallmark gene sets</span></strong><span leaf="">。然后取每条通路的 NES 打分从高到低进行排序，并绘制散点图。</span><strong><span leaf="">点的大小</span></strong><span leaf="">为 Detection：应该为每个通路中基因在所有样本中表达的 count 大于 5 的占比（我们这里并没有表达矩阵，就用基因集大小替代好了）。</span><strong><span leaf="">点的颜色</span></strong><span leaf="">为通路显著性：-log10 pvalue。</span></p></section><span></span></h4><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwUJ2WlCfwngHZR7Z6ZyCFqOkNpkaY1C7tKCK23ssJO9koMwkK9fDJEZXLhCPqib3Kj3bmolGffePw/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.7162162162162162" data-type="png" data-w="518" data-imgfileid="100054251" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwUJ2WlCfwngHZR7Z6ZyCFqOkNpkaY1C7tKCK23ssJO9koMwkK9fDJEZXLhCPqib3Kj3bmolGffePw/640?wx_fmt=png&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">图注：</span></p><blockquote><span></span><p><span leaf="">Fig. 1. Establishment and evaluation of a KRAS dependent gene expression program in KRAS-mutant PDAC. (C) GSEA for 50 Hallmark gene sets within DEGs after KRAS siRNA treatment. Detection is based on the presence of a gene in all eight PDAC cell lines at &gt;5 reads.</span></p></blockquote><p data-tool="mdnice编辑器"><span leaf="">关于可不可以用差异基因进行GSEA分析，我们前面讨论过：<a target="_blank" href="https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247537475&amp;idx=1&amp;sn=0ba7600556f909d01727f59010d7a9f9&amp;scene=21#wechat_redirect" textvalue="" linktype="text" data-linktype="2">IF10+杂志文章只用统计学显著的差异基因做GSEA就合理吗？</a></span></p><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">数据背景介绍</span></span><span></span></h2><h3 data-tool="mdnice编辑器"><span></span><span></span><span><span leaf="">1、PDAC细胞系KRAS siRNA RNA测序</span></span><span></span></h3><p data-tool="mdnice编辑器"><span leaf="">为了确定 KRAS 依赖性转录组，作者对一组八个人类 KRAS 突变型 胰腺导管腺癌PDAC的细胞系在经过 24 小时 KRAS 小干扰 RNA（siRNA）处理后的基因转录变化（图 S1A 和 B）进行了 RNA 测序（RNA-seq）。数据可在这里下载：</span><strong><span leaf="">PRJNA980201</span></strong><span leaf=""> https://www.ebi.ac.uk/ena/browser/view/PRJNA980201?show=related-records。</span></p><p data-tool="mdnice编辑器"><span leaf="">Note：</span></p><ul><li><section><p><span leaf="">NS缩写：control non-specific ("NS") siRNA。</span></p></section></li><li><section><p><span leaf="">KRAS siRNA是一种小干扰RNA（siRNA），通过特异性靶向KRAS基因的mRNA，从而抑制其表达，达到治疗KRAS突变肿瘤的目的。</span></p></section></li></ul><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwUJ2WlCfwngHZR7Z6ZyCFqofI03HYCvE1woibvSXdysAV4VQJTKfiaamENsy54hNmAng8UkWRbxx1w/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.1969532100108814" data-type="png" data-w="919" data-imgfileid="100054250" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwUJ2WlCfwngHZR7Z6ZyCFqofI03HYCvE1woibvSXdysAV4VQJTKfiaamENsy54hNmAng8UkWRbxx1w/640?wx_fmt=png&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">差异分析结果如下：The 677 KRAS-dependent (UP) and 1051 KRAS-inhibited (DN) genes (</span><strong><span leaf="">log2FC &gt; 0.5, adj. p &lt; 0.05</span></strong><span leaf="">) are indicated by the light blue– and light red–shaded rectangles underlaid on plot, respectively。</span></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwUJ2WlCfwngHZR7Z6ZyCFqBJMsUKI1Fic0uH2Pia7sU3suBNncH2t98I4Ey2uyj19YcOgTSuShbIyQ/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.7696629213483146" data-type="png" data-w="534" data-imgfileid="100054252" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwUJ2WlCfwngHZR7Z6ZyCFqBJMsUKI1Fic0uH2Pia7sU3suBNncH2t98I4Ey2uyj19YcOgTSuShbIyQ/640?wx_fmt=png&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">差异结果在表格S1：</span><strong><span leaf="">Data S1.</span></strong><span leaf=""> Differential expression statistics for genes in KRAS versus NS siRNA treated PDAC cells, using RNA-sequencing。</span></p><h3 data-tool="mdnice编辑器"><span></span><span></span><span><span leaf="">2、50 Hallmark gene sets</span></span><span></span></h3><p data-tool="mdnice编辑器"><span leaf="">Hallmark 通路可以在GSEA 的 MSigDB 数据库去下载 gmt 格式：</span><span><span leaf="">https://www.gsea-msigdb.org/gsea/msigdb/human/collections.jsp</span></span></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwUJ2WlCfwngHZR7Z6ZyCFqMo55d45P5eTlY7IDhn8TcSbnLKM1uytacWh99zMib7D4Z26FuttOCXw/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.43866171003717475" data-type="png" data-w="807" data-imgfileid="100054249" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwUJ2WlCfwngHZR7Z6ZyCFqMo55d45P5eTlY7IDhn8TcSbnLKM1uytacWh99zMib7D4Z26FuttOCXw/640?wx_fmt=png&amp;from=appmsg"></span></figure><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">GSEA分析</span></span><span></span></h2><p data-tool="mdnice编辑器"><span leaf="">GSEA的输入数据为差异基因排序列表，排序指标为log2FC值：</span></p><h3 data-tool="mdnice编辑器"><span></span><span></span><span><span leaf="">1、读取数据</span></span><span></span></h3><pre data-tool="mdnice编辑器"><span></span><code><span leaf="">rm(list=ls())</span><span leaf=""><br></span><span><span leaf=""># 加载R包</span></span><span leaf=""><br></span><span leaf="">library(ggplot2)</span><span leaf=""><br></span><span leaf="">library(tibble)</span><span leaf=""><br></span><span leaf="">library(ggrepel)</span><span leaf=""><br></span><span leaf="">library(tidyverse)</span><span leaf=""><br></span><span leaf="">library(dplyr)</span><span leaf=""><br></span><span leaf="">library(patchwork)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf="">##### 01、加载数据</span></span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 加载：KRAS vs NS siRNA（log2FC）</span></span><span leaf=""><br></span><span leaf="">group1 &lt;- read.csv(</span><span><span leaf="">"./data/science.adk0775_data_s1.csv"</span></span><span leaf=""> )</span><span leaf=""><br></span><span leaf="">head(group1)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 提取FDR值和FC值</span></span><span leaf=""><br></span><span leaf="">group1 &lt;- group1[,c(</span><span><span leaf="">"external_gene_name"</span></span><span leaf="">,</span><span><span leaf="">"logFC"</span></span><span leaf="">,</span><span><span leaf="">"FDR"</span></span><span leaf=""> )]</span><span leaf=""><br></span><span leaf="">group1 &lt;- group1[group1</span><span><span leaf="">$external_gene_name</span></span><span leaf="">!=</span><span><span leaf="">""</span></span><span leaf="">, ]</span><span leaf=""><br></span><span leaf="">group1 &lt;- distinct(group1,external_gene_name,.keep_all = T)</span><span leaf=""><br></span><span leaf="">rownames(group1) &lt;- group1</span><span><span leaf="">$external_gene_name</span></span><span leaf=""><br></span><span leaf="">head(group1)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 增加一列上下调：log2FC &gt; 0.5, adj. p &lt; 0.05</span></span><span leaf=""><br></span><span leaf="">group1</span><span><span leaf="">$g1</span></span><span leaf=""> &lt;- </span><span><span leaf="">"normal"</span></span><span leaf=""><br></span><span leaf="">group1</span><span><span leaf="">$g1</span></span><span leaf="">[group1</span><span><span leaf="">$logFC</span></span><span leaf=""> &gt;0.5 &amp; group1</span><span><span leaf="">$FDR</span></span><span leaf=""> &lt; 0.05 ] &lt;- </span><span><span leaf="">"up"</span></span><span leaf=""><br></span><span leaf="">group1</span><span><span leaf="">$g1</span></span><span leaf="">[group1</span><span><span leaf="">$logFC</span></span><span leaf=""> &lt; -0.5 &amp; group1</span><span><span leaf="">$FDR</span></span><span leaf=""> &lt; 0.05 ] &lt;- </span><span><span leaf="">"down"</span></span><span leaf=""><br></span><span leaf="">table(group1</span><span><span leaf="">$g1</span></span><span leaf="">)</span><span leaf=""><br></span><span><span leaf=""># down normal     up </span></span><span leaf=""><br></span><span><span leaf=""># 675  12876   1043 </span></span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 获得差异基因</span></span><span leaf=""><br></span><span leaf="">DEGs &lt;- group1[group1</span><span><span leaf="">$g1</span></span><span leaf="">!=</span><span><span leaf="">"normal"</span></span><span leaf="">,]</span><span leaf=""><br></span><span leaf="">head(DEGs)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 得到排序列表</span></span><span leaf=""><br></span><span leaf="">DEGs &lt;- DEGs[order(DEGs</span><span><span leaf="">$logFC</span></span><span leaf="">, decreasing = T),]</span><span leaf=""><br></span><span leaf="">genelist &lt;- DEGs</span><span><span leaf="">$logFC</span></span><span leaf=""><br></span><span leaf="">names(genelist) &lt;- DEGs</span><span><span leaf="">$external_gene_name</span></span><span leaf=""><br></span><span leaf="">head(genelist)</span><span leaf=""><br></span><span><span leaf=""># NPPC   SMTNL2    KCNK3     CHPF PCDHGA12  EXOC3L1 </span></span><span leaf=""><br></span><span><span leaf=""># 2.860472 2.712068 2.551034 2.536558 2.499657 2.446140</span></span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">tail(genelist)</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">得到 1718个差异基因：</span></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwUJ2WlCfwngHZR7Z6ZyCFqicOicMH2G7m3dlABTCAmRdMWLGTqxY2GlPdhbpTvmu7y8Xmia6psE2riag/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.2505091649694501" data-type="png" data-w="491" data-imgfileid="100054248" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwUJ2WlCfwngHZR7Z6ZyCFqicOicMH2G7m3dlABTCAmRdMWLGTqxY2GlPdhbpTvmu7y8Xmia6psE2riag/640?wx_fmt=png&amp;from=appmsg"></span></figure><h3 data-tool="mdnice编辑器"><span></span><span></span><span><span leaf="">2、读取 50 Hallmark gene sets 通路并富集：</span></span><span></span></h3><pre data-tool="mdnice编辑器"><span></span><code><span><span leaf="">## === HALLMARK通路富集</span></span><span leaf=""><br></span><span leaf="">geneset &lt;- read.gmt(</span><span><span leaf="">"data/h.all.v2024.1.Hs.symbols.gmt"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">table(geneset</span><span><span leaf="">$term</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">geneset</span><span><span leaf="">$term</span></span><span leaf=""> &lt;- gsub(pattern = </span><span><span leaf="">"HALLMARK_"</span></span><span leaf="">,</span><span><span leaf="">""</span></span><span leaf="">, geneset</span><span><span leaf="">$term</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 运行,输出全部结果</span></span><span leaf=""><br></span><span leaf="">egmt &lt;- GSEA(genelist, TERM2GENE=geneset, pvalueCutoff = 1, minGSSize = 1, maxGSSize = 500000)</span><span leaf=""><br></span><span leaf="">colnames(egmt@result)</span><span leaf=""><br></span><span leaf="">head(egmt[, 1:6])</span><span leaf=""><br></span></code></pre><h3 data-tool="mdnice编辑器"><span></span><span></span><span><span leaf="">3、使用 ggplot2 定制化绘图</span></span><span></span></h3><p data-tool="mdnice编辑器"><span leaf="">取出绘图需要的列，并进行相关设置：</span></p><pre data-tool="mdnice编辑器"><span></span><code><span><span leaf=""># 绘图</span></span><span leaf=""><br></span><span leaf="">data &lt;- egmt[,c(</span><span><span leaf="">"ID"</span></span><span leaf="">, </span><span><span leaf="">"NES"</span></span><span leaf="">,</span><span><span leaf="">"setSize"</span></span><span leaf="">,</span><span><span leaf="">"pvalue"</span></span><span leaf="">)]</span><span leaf=""><br></span><span leaf="">data</span><span><span leaf="">$setSize_1</span></span><span leaf=""> &lt;- data</span><span><span leaf="">$setSize</span></span><span leaf="">/10</span><span leaf=""><br></span><span leaf="">head(data)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 给Y轴的通路名设置为因子，排序</span></span><span leaf=""><br></span><span leaf="">data &lt;- data[order(data</span><span><span leaf="">$NES</span></span><span leaf="">, decreasing = T),]</span><span leaf=""><br></span><span leaf="">data</span><span><span leaf="">$ID</span></span><span leaf=""> &lt;- factor(data</span><span><span leaf="">$ID</span></span><span leaf="">, levels = data</span><span><span leaf="">$ID</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">data</span><span><span leaf="">$xlab</span></span><span leaf=""> &lt;- 1:49</span><span leaf=""><br></span><span leaf="">head(data)</span><span leaf=""><br></span><span leaf="">summary(data</span><span><span leaf="">$NES</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">summary(data</span><span><span leaf="">$setSize_1</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 图中标出的通路名字</span></span><span leaf=""><br></span><span leaf="">label &lt;- c( </span><span><span leaf="">"KRAS_SIGNALING_DN"</span></span><span leaf="">, </span><span><span leaf="">"INTERFERON_ALPHA_RESPONSE"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  </span><span><span leaf="">"UV_RESPONSE_DN"</span></span><span leaf="">, </span><span><span leaf="">"EMT"</span></span><span leaf="">, </span><span><span leaf="">"TNFA_SIGNALING_VIA_NFKB"</span></span><span leaf="">, </span><span><span leaf="">"MYC_TARGETS_V2"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  </span><span><span leaf="">"KRAS_SIGNALING_UP"</span></span><span leaf="">, </span><span><span leaf="">"MYC_TARGETS_V1"</span></span><span leaf="">, </span><span><span leaf="">"G2M_CHECKPOINT"</span></span><span leaf="">, </span><span><span leaf="">"E2F_TARGETS"</span></span><span leaf=""> )</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 没有EMT通路</span></span><span leaf=""><br></span><span leaf="">data_label &lt;- data[data</span><span><span leaf="">$ID</span></span><span leaf=""> %</span><span><span leaf="">in</span></span><span leaf="">% label,]</span><span leaf=""><br></span><span leaf="">data_label</span><span leaf=""><br></span><span><span leaf=""># 图中通路的颜色</span></span><span leaf=""><br></span><span leaf="">data_label</span><span><span leaf="">$col</span></span><span leaf=""> &lt;- c(</span><span><span leaf="">"black"</span></span><span leaf="">, </span><span><span leaf="">"grey80"</span></span><span leaf="">,</span><span><span leaf="">"grey80"</span></span><span leaf="">,</span><span><span leaf="">"grey80"</span></span><span leaf="">,</span><span><span leaf="">"grey80"</span></span><span leaf="">,</span><span><span leaf="">"#ffb882"</span></span><span leaf="">,</span><span><span leaf="">"grey80"</span></span><span leaf="">,</span><span><span leaf="">"#ff5eff"</span></span><span leaf="">,</span><span><span leaf="">"#ff5eff"</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">使用ggplot2绘图：</span></p><pre data-tool="mdnice编辑器"><span></span><code><span leaf="">p &lt;- ggplot(data = data, aes(x = xlab, y = NES)) +</span><span leaf=""><br></span><span leaf="">  geom_point(aes(size = setSize_1, alpha = -log10(pvalue)), shape = 21, stroke = 0.7,fill = </span><span><span leaf="">"#0000ff"</span></span><span leaf="">, colour = </span><span><span leaf="">"black"</span></span><span leaf="">) + </span><span><span leaf=""># stroke：设置点的边框宽度。</span></span><span leaf=""><br></span><span leaf="">  scale_size_continuous(range = c(0.2, 8)) + </span><span leaf=""><br></span><span leaf="">  xlab(label = </span><span><span leaf="">"Hallmark gene sets"</span></span><span leaf="">) + </span><span leaf=""><br></span><span leaf="">  ylab(label = </span><span><span leaf="">"Normalized enrichment score (NES)"</span></span><span leaf="">) + </span><span leaf=""><br></span><span leaf="">  theme_classic(base_size = 15) + </span><span leaf=""><br></span><span leaf="">  scale_x_continuous(breaks = seq(0, 50, by = 10), labels = seq(0, 50, by = 10) ) + </span><span><span leaf=""># 设置x轴的刻度线和刻度标签</span></span><span leaf=""><br></span><span leaf="">  scale_y_continuous( breaks = seq(-4, 2.3, by = 1), labels = seq(-4, 2.3, by = 1) ) + </span><span leaf=""><br></span><span leaf="">  guides(size = guide_legend(title = </span><span><span leaf="">"Detection\n(proportion)"</span></span><span leaf="">),</span><span leaf=""><br></span><span leaf="">         alpha = guide_legend(title = </span><span><span leaf="">"Significance\n(-log10 p-val.)"</span></span><span leaf="">) ) +  </span><span><span leaf=""># 修改图例标题</span></span><span leaf=""><br></span><span leaf="">  theme(</span><span leaf=""><br></span><span leaf="">    axis.line = element_line(color = </span><span><span leaf="">"black"</span></span><span leaf="">, size = 0.6), </span><span><span leaf=""># 加粗x轴和y轴的线条</span></span><span leaf=""><br></span><span leaf="">    axis.text = element_text(face = </span><span><span leaf="">"bold"</span></span><span leaf="">), </span><span><span leaf=""># 加粗x轴和y轴的标签</span></span><span leaf=""><br></span><span leaf="">    axis.title = element_text( size = 13)    </span><span><span leaf=""># 加粗x轴和y轴的标题</span></span><span leaf=""><br></span><span leaf="">  )</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">p</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwUJ2WlCfwngHZR7Z6ZyCFqcibmbDVg5C24XJV4FoEic4QALbeJGBcvoGSFFGhHWwGCC9AqibJVwcQ3g/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.7892074198988196" data-type="png" data-w="593" data-imgfileid="100054256" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwUJ2WlCfwngHZR7Z6ZyCFqcibmbDVg5C24XJV4FoEic4QALbeJGBcvoGSFFGhHWwGCC9AqibJVwcQ3g/640?wx_fmt=png&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">添加通路标签：</span></p><pre data-tool="mdnice编辑器"><span></span><code><span><span leaf=""># 添加label：vjust（垂直调整）或hjust（水平调整）</span></span><span leaf=""><br></span><span leaf="">p3 &lt;- p + </span><span leaf=""><br></span><span leaf="">  geom_text_repel(data= data_label, aes(x = xlab, y = NES,label = ID), size = 3, color = data_label</span><span><span leaf="">$col</span></span><span leaf="">, </span><span leaf=""><br></span><span leaf="">                  force=20,                </span><span><span leaf=""># 将重叠的文本标签相互推开的强度。force 参数的值越大，标签之间的排斥力度也越大，这会导致标签在图中更分散地排列</span></span><span leaf=""><br></span><span leaf="">                  point.padding = 0.5,     </span><span><span leaf=""># 设置文本标签与对应点之间的最小距离</span></span><span leaf=""><br></span><span leaf="">                  min.segment.length = 0,  </span><span><span leaf=""># 长度大于0就可以添加引线</span></span><span leaf=""><br></span><span leaf="">                  hjust = 1.2,             </span><span><span leaf=""># 文本标签的右侧与指定位置对齐</span></span><span leaf=""><br></span><span leaf="">                  segment.color=</span><span><span leaf="">"grey20"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">                  segment.size=0.3,        </span><span><span leaf=""># 设置引导线的粗细</span></span><span leaf=""><br></span><span leaf="">                  segment.alpha=0.8,       </span><span><span leaf=""># 文本标签中连接线段的透明度</span></span><span leaf=""><br></span><span leaf="">                  nudge_y=-0.1             </span><span><span leaf=""># 在y轴方向上微调标签位置</span></span><span leaf=""><br></span><span leaf="">                  ) </span><span leaf=""><br></span><span leaf="">p3</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 保存</span></span><span leaf=""><br></span><span leaf="">ggsave(filename = </span><span><span leaf="">"Figure1C.pdf"</span></span><span leaf="">, plot = p3, width = 6.2, height = 5)</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">结果如下：</span></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwUJ2WlCfwngHZR7Z6ZyCFqJRkmmKSx2TqmMmFehAggpsgdMedXiaEeaAhen9baZibqw88icWjdA7w3A/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.8226837060702875" data-type="png" data-w="626" data-imgfileid="100054257" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwUJ2WlCfwngHZR7Z6ZyCFqJRkmmKSx2TqmMmFehAggpsgdMedXiaEeaAhen9baZibqw88icWjdA7w3A/640?wx_fmt=png&amp;from=appmsg"></span></figure><h3 data-tool="mdnice编辑器"><span></span><span></span><span><span leaf="">友情宣传：</span></span><span></span></h3><p data-tool="mdnice编辑器"><span leaf=""><a target="_blank" href="https://mp.weixin.qq.com/s?__biz=MzI1Njk4ODE0MQ==&amp;mid=2247527230&amp;idx=1&amp;sn=7156afcd5ab734c7d391b9048695747a&amp;scene=21#wechat_redirect" textvalue="" linktype="text" data-linktype="2">生信入门&amp;数据挖掘线上直播课2025年1月班</a></span></p><p data-tool="mdnice编辑器"><span leaf=""><a target="_blank" href="http://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247524148&amp;idx=1&amp;sn=7806da6feb41a36493c519c1cfc1d3ac&amp;chksm=9b4bdf8fac3c569960369602f1ef26639cb366b250f233b2297d1f059471c0458335bfc0b829&amp;scene=21#wechat_redirect" textvalue="" linktype="text" data-linktype="2">时隔5年，我们的生信技能树VIP学徒继续招生啦</a></span></p><p data-tool="mdnice编辑器"><span leaf=""><a target="_blank" href="https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247535760&amp;idx=2&amp;sn=1e02a2e982a046ecf6389231e6768d5b&amp;scene=21#wechat_redirect" textvalue="" linktype="text" data-linktype="2">满足你生信分析计算需求的低价解决方案</a></span></p></section><p><mp-style-type data-value="3"></mp-style-type></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/bsze8fvUBKvfTvyiXhUP1w",target="_blank" rel="noopener noreferrer">原文链接</a>
