---
title: "单细胞共表达基因相关性｜FigureYa301scCoExpr"
date: 2025-11-25T13:53:06Z
draft: ["false"]
tags: [
  "fetched",
  "小丫画图"
]
categories: ["Acdemic"]
---
单细胞共表达基因相关性｜FigureYa301scCoExpr by 小丫画图
------
<div><section><section powered-by="xiumi.us"><section><section><section><section><section><p><span><span leaf="">小丫碎碎念：</span></span></p><p><span><span leaf=""><br></span></span></p><p><span leaf="">提需求的小伙伴有一双发现美的眼睛，总能从文章中发现神奇的可视化方案。</span></p><p><span leaf=""><br></span></p><p><span leaf="">G 图右上角的 Merge 很有趣，像免疫荧光实验的 merge 效果：</span></p><ul><li><section><span leaf="">红色 = 只表达基因1</span></section></li><li><section><span leaf="">绿色 = 只表达基因2</span></section></li><li><section><span leaf="">黄色 = 共表达两个基因</span></section></li><li><section><span leaf="">灰色 = 两个基因都不表达</span></section></li></ul><p><span leaf=""><br></span></p></section></section></section><p><strong><span><span leaf="">需求描述</span></span></strong></p><section><span leaf="">用来展示单细胞的基因共表达。用2种颜色分别映射2个基因的表达量/连续变量，然后merge展示，有点像免疫荧光实验的merge效果，颜色不是固定值，映射到连续性表达量的也是渐变色。</span></section><section><span leaf=""><img data-imgfileid="100003848" data-ratio="0.8451612903225807" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/fTGTY7BwuzfJyJrTXlmsJs0ibMwCPHp5DsiaFEU8gvicntKWgaTvBq3v5VzyPfcxM6BfB4u3Itia43JxAcJKEvU2xQ/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="620" src="https://mmbiz.qpic.cn/sz_mmbiz_png/fTGTY7BwuzfJyJrTXlmsJs0ibMwCPHp5DsiaFEU8gvicntKWgaTvBq3v5VzyPfcxM6BfB4u3Itia43JxAcJKEvU2xQ/640?wx_fmt=png&amp;from=appmsg"></span></section><p data-pm-slice="0 0 []"><span leaf=""><span textstyle="">From https://www.thno.org/v12p3104.htm</span></span></p><p><span leaf=""><span textstyle="">Figure 5. scRNA-seq analysis reveals subsets of CAFs as the major source of THBS2 production.</span></span></p><p><span leaf=""><span textstyle="">G, UMAP plots showing the co-expression of THBS2 and FAP across individual single CAFs. Red/green dots represent CAFs expressing THBS2+ only/FAP+ only (upper left/middle panels), respectively, and yellow dots indicate CAFs co-expressing THBS2 and FAP only (upper right panel); lower left panel showing the </span></span><strong><span leaf=""><span textstyle="">co-expression matrix</span></span></strong><span leaf=""><span textstyle=""> across CAFs with different expression of FAP and THBS2; lower right panels showing the </span></span><strong><span leaf=""><span textstyle="">correlation (Pearson)</span></span></strong><span leaf=""><span textstyle=""> between THBS2 and FAP across THBS2+ CAFs. (1) Strong correlation: Pearson’s r ≥ 0.8; (2) Moderate: 0.5 ≤ Pearson’s r &lt; 0.8; (3) Weak: 0.3 ≤ Pearson’s r &lt; 0.5; (4) No correlation: Pearson’s r &lt; 0.3). The darker the color, the higher the expression level of the indicated markers. Of note, only a minority of CAFs co-express high THBS2 and FAP.</span></span></p><p><span leaf=""><br></span></p><h2><strong><span><span leaf="">应用场景</span></span></strong></h2><section><span leaf="">展示单细胞的基因共表达，以及共表达基因相关性。</span></section><p><span leaf=""><br></span></p></section></section></section></section><h2><strong><span><span leaf="">环境设置</span></span></strong></h2><section><ul><li><li><li><li><li><li><li></ul><pre data-lang="bash"><code><span leaf=""><span>source</span>(<span>"install_dependencies.R"</span>) <span>#自动识别缺哪些R包，自动安装</span></span></code><code><span leaf="">library(Seurat) <span># 单细胞数据处理 Single-cell data processing</span></span></code><code><span leaf="">library(SeuratData) <span># 单细胞数据集 SeuratData</span></span></code><code><span leaf="">library(ggplot2) <span># 绘图 Plotting</span></span></code><code><span leaf="">library(ggpubr) <span># 配色 Plotting</span></span></code><code><span leaf="">library(patchwork) <span># 拼图 Plotting</span></span></code><code><span leaf="">library(magrittr) <span># 数据处理 Data processing</span></span></code></pre></section><h2 data-heading="true"><strong><span><span leaf="">输入文件预处理</span></span></strong></h2><p data-pm-slice="0 0 []"><span><span leaf="">使用最常用的SeuratData中的pbmc3k进行演示。</span></span></p><section><ul><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li></ul><pre data-lang="apache"><code><span leaf=""><span># InstallData("pbmc3k") </span></span></code><code><span leaf=""><span># 加载数据集</span></span></code><code><span leaf=""><span># load dataset</span></span></code><code><span leaf=""><span>LoadData</span>(<span>"pbmc3k"</span>)</span></code><code><span leaf=""><span>pbmc3k</span> &lt;- UpdateSeuratObject(pbmc3k)</span></code><code><span leaf=""><span># 计算线粒体基因比例</span></span></code><code><span leaf=""><span># Calculate mitochondrial gene percentage</span></span></code><code><span leaf=""><span>pbmc3k</span>[[<span>"percent.mt"</span>]] &lt;- PercentageFeatureSet(pbmc3k, pattern = <span>"^MT-"</span>)</span></code><code><span leaf=""><span># 质控数据可视化</span></span></code><code><span leaf=""><span># Quality control visualization</span></span></code><code><span leaf=""><span>VlnPlot</span>(pbmc3k, features = c(<span>"nFeature_RNA"</span>, <span>"nCount_RNA"</span>, <span>"percent.mt"</span>), ncol = <span>3</span>)</span></code><code><span leaf=""><span>plot1</span> &lt;- FeatureScatter(pbmc3k, feature1 = <span>"nCount_RNA"</span>, feature2 = <span>"percent.mt"</span>)</span></code><code><span leaf=""><span>plot2</span> &lt;- FeatureScatter(pbmc3k, feature1 = <span>"nCount_RNA"</span>, feature2 = <span>"nFeature_RNA"</span>)</span></code><code><span leaf=""><span>plot1</span> + plot2</span></code><code><span leaf=""><span># 通过控制基因数和线粒体基因比例过滤掉低质量细胞，过滤完后剩余2638</span></span></code><code><span leaf=""><span># Filter low-quality cells by controlling gene counts and mitochondrial percentage, retaining 2,638 cells after filtering</span></span></code><code><span leaf=""><span>pbmc3k</span> &lt;- subset(pbmc3k, subset = nFeature_RNA &gt; <span>200</span> &amp; nFeature_RNA &lt; <span>2500</span> &amp; percent.mt &lt; <span>5</span>)</span></code><code><span leaf=""><span>pbmc3k</span></span></code><code><span leaf=""><span># 进行normalize并计算高变异基因</span></span></code><code><span leaf=""><span># Normalize data and identify highly variable genes</span></span></code><code><span leaf=""><span>pbmc3k</span> &lt;- NormalizeData(pbmc3k, normalization.method = <span>"LogNormalize"</span>, scale.factor = <span>10000</span>)</span></code><code><span leaf=""><span>pbmc3k</span> &lt;- FindVariableFeatures(pbmc3k, selection.method = <span>"vst"</span>, nfeatures = <span>2000</span>)</span></code><code><span leaf=""><span># Scale数据</span></span></code><code><span leaf=""><span># Scale data</span></span></code><code><span leaf=""><span>all</span>.genes &lt;- rownames(pbmc3k)</span></code><code><span leaf=""><span>pbmc3k</span> &lt;- ScaleData(pbmc3k, features = <span>all</span>.genes)</span></code><code><span leaf=""><span># 进行主成分分析</span></span></code><code><span leaf=""><span># PCA</span></span></code><code><span leaf=""><span>pbmc3k</span> &lt;- RunPCA(pbmc3k, features = VariableFeatures(object = pbmc3k))</span></code><code><span leaf=""><span>#DimHeatmap(pbmc3k, dims = 1:30, cells = 500, balanced = TRUE)</span></span></code><code><span leaf=""><span>pbmc3k</span> &lt;- JackStraw(pbmc3k, num.replicate = <span>100</span>)</span></code><code><span leaf=""><span>pbmc3k</span> &lt;- ScoreJackStraw(pbmc3k,dims = <span>1</span>:<span>20</span>)</span></code><code><span leaf=""><span>JackStrawPlot</span>(pbmc3k, dims = <span>1</span>:<span>20</span>)</span></code><code><span leaf=""><span>ElbowPlot</span>(pbmc3k)</span></code><code><span leaf=""><span># 选择前10个pc基因降维</span></span></code><code><span leaf=""><span># Dimensionality reduction using top 10 PCs</span></span></code><code><span leaf=""><span>npcs</span> = <span>1</span>:<span>10</span></span></code><code><span leaf=""><span>pbmc3k</span> &lt;- FindNeighbors(pbmc3k, dims = npcs)</span></code><code><span leaf=""><span>pbmc3k</span> &lt;- FindClusters(pbmc3k, resolution = <span>0</span>.<span>5</span>)</span></code><code><span leaf=""><span>pbmc3k</span> &lt;- RunUMAP(pbmc3k, dims = <span>1</span>:<span>10</span>)</span></code><code><span leaf=""><span>DimPlot</span>(pbmc3k, reduction = <span>"umap"</span>, label = T, pt.size = <span>1</span>,raster = F)+</span></code><code><span leaf="">  <span>theme</span>(aspect.ratio = <span>1</span>)</span></code><code><span leaf=""><span>DimPlot</span>(pbmc3k, reduction = <span>"umap"</span>, group.by = <span>"seurat_annotations"</span>,label = T, pt.size = <span>1</span>,raster = F)+</span></code><code><span leaf="">  <span>theme</span>(aspect.ratio = <span>1</span>)+</span></code><code><span leaf="">  <span>NoLegend</span>()</span></code><code><span leaf=""><span>#save(pbmc3k, file = "./pbmc3k_anno_CellType.rdata")</span></span></code></pre></section><h2 data-heading="true"><strong><span><span leaf="">开始画图</span></span></strong></h2><p data-pm-slice="0 0 []"><span><span leaf="">选择CD3G和CD247这两个T细胞marker基因进行基因共表达可视化</span></span></p><section><ul><li><li><li><li><li><li><li><li></ul><pre data-lang="apache"><code><span leaf=""><span># FeaturePlot绘制UMAP图上共表达基因展示，blend设置为TRUE</span></span></code><code><span leaf=""><span># FeaturePlot for co-expression visualization on UMAP,and blend is set to TRUE</span></span></code><code><span leaf=""><span>FeaturePlot</span>(pbmc3k, features = c(<span>"CD3G"</span>, <span>"CD247"</span>), </span></code><code><span leaf="">            <span>blend</span> = TRUE, </span></code><code><span leaf="">            <span>cols</span> = c(<span>"gray80"</span>,<span>"red"</span>, <span>"green"</span>), </span></code><code><span leaf="">            <span>pt</span>.size = <span>0</span>.<span>5</span>, raster = F) +  </span></code><code><span leaf="">  <span>theme</span>(aspect.ratio = <span>1</span>)</span></code><code><span leaf=""><span>ggsave</span>(<span>"UMAP_co-expression matrix.pdf"</span>, width = <span>12</span>, height = <span>4</span>)</span></code></pre></section><p data-pm-slice="0 0 []"><span><span leaf=""><img data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/fTGTY7BwuzfJyJrTXlmsJs0ibMwCPHp5DfZkhibN30wJicJf9jSrY9U5M0UlvFtZxG9ibZLNa4p42RjQicyDkiaibJK1g/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.3333333333333333" data-type="png" data-w="1080" data-imgfileid="100003851" src="https://mmbiz.qpic.cn/sz_mmbiz_png/fTGTY7BwuzfJyJrTXlmsJs0ibMwCPHp5DfZkhibN30wJicJf9jSrY9U5M0UlvFtZxG9ibZLNa4p42RjQicyDkiaibJK1g/640?wx_fmt=png&amp;from=appmsg"></span></span></p><p data-pm-slice="0 0 []"><span><span leaf="">自定义函数，用于计算共表达基因相关性并可视化</span></span></p><section><ul><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li></ul><pre data-lang="php"><code><span leaf=""><span># object: seurat对象</span></span></code><code><span leaf=""><span># gene1和gene2: 两个共表达基因名</span></span></code><code><span leaf=""><span># cor.method: 计算各相关性方法，可选spearman或pearson,</span></span></code><code><span leaf=""><span># jitter.num: 为了防止所有数据聚集在一起，jitter.num使数据进行轻微抖动，设置为0则为原始数据绘图，</span></span></code><code><span leaf=""><span># pos: 若为TRUE，怎只使用两个基因都表达的细胞计算相关性，可避免好多细胞中对其中一个细胞表达量为0</span></span></code><code><span leaf=""><span># object: Seurat object  </span></span></code><code><span leaf=""><span># gene1 &amp; gene2: Names of two co-expressed genes  </span></span></code><code><span leaf=""><span># cor.method: Correlation method ("spearman" or "pearson")  </span></span></code><code><span leaf=""><span># jitter.num: Adds slight jitter to prevent data clustering (set to 0 for raw data plotting)  </span></span></code><code><span leaf=""><span># pos: If TRUE, only uses cells expressing both genes to avoid zero-expression bias  </span></span></code><code><span leaf="">getScatterplot &lt;- <span><span>function</span></span><span>(</span><span><span><span>object</span></span></span><span><span>, gene1, gene2, cor.method = </span></span><span><span><span>"spearman"</span></span></span><span><span>,</span></span></span></code><code><span leaf="">                           jitter.num = <span>0.15</span>, pos = <span>TRUE</span>){</span></code><code><span leaf="">  <span>if</span><span> </span>(!gene1 %in% <span>rownames</span>(<span>object</span>)) {</span></code><code><span leaf="">    <span>print</span>(<span>"gene1 was not found"</span>)</span></code><code><span leaf="">    <span>if</span><span> </span>(!gene2 %in% <span>rownames</span>(<span>object</span>)) {</span></code><code><span leaf="">      <span>print</span>(<span>"gene2 was not found"</span>)</span></code><code><span leaf="">    }</span></code><code><span leaf="">  }<span>else</span>{</span></code><code><span leaf="">    exp.mat &lt;- <span>GetAssayData</span>(<span>object</span> = <span>object</span>, assay = <span>"RNA"</span>) %&gt;% .[<span>c</span>(gene1,gene2),] %&gt;% </span></code><code><span leaf="">      <span>as</span>.<span>matrix</span>() %&gt;% <span>t</span>() %&gt;% <span>as</span>.data.<span>frame</span>()</span></code><code><span leaf="">    <span>if</span><span> </span>(pos) {</span></code><code><span leaf="">      exp.mat &lt;- exp.mat[<span>which</span>(exp.mat[,<span>1</span>] &gt; <span>0</span> &amp; exp.mat[,<span>2</span>] &gt; <span>0</span>),]</span></code><code><span leaf="">    }</span></code><code><span leaf="">    plots &lt;- <span>ggplot</span>(data=exp.mat, mapping = <span>aes_string</span>(x = gene1, y = gene2)) + </span></code><code><span leaf="">      <span>geom_smooth</span>(method = <span>'lm'</span>, se = T, color=<span>'red'</span>, size=<span>1</span>)+</span></code><code><span leaf="">      <span>stat_cor</span>(method = cor.method)+ </span></code><code><span leaf="">      <span>geom_jitter</span>(width = jitter.num, height = jitter.num, color = <span>"black"</span>, size = <span>1</span>, alpha=<span>1</span>)+</span></code><code><span leaf="">      <span>theme_bw</span>()+</span></code><code><span leaf="">      <span>theme</span>(panel.grid=<span>element_blank</span>(),</span></code><code><span leaf="">            legend.text=<span>element_text</span>(colour= <span>'black'</span>,size=<span>10</span>),</span></code><code><span leaf="">            axis.text= <span>element_text</span>(colour= <span>'black'</span>,size=<span>10</span>),</span></code><code><span leaf="">            axis.line= <span>element_line</span>(colour= <span>'black'</span>),</span></code><code><span leaf="">            panel.border = <span>element_rect</span>(size = <span>1</span>, linetype = <span>"solid"</span>, colour = <span>"black"</span>),</span></code><code><span leaf="">            panel.background=<span>element_rect</span>(fill=<span>"white"</span>))</span></code><code><span leaf="">    <span>return</span>(plots)</span></code><code><span leaf="">  }</span></code><code><span leaf="">}</span></code></pre></section><p data-pm-slice="0 0 []"><span><span leaf="">计算共表达基因相关性，并可视化</span></span></p><section><ul><li><li><li><li><li></ul><pre data-lang="apache"><code><span leaf=""><span># 设置抖动参数，防止数据聚集在一起 Setting jitter parameter to prevent data clustering</span></span></code><code><span leaf=""><span>getScatterplot</span>(pbmc3k, gene1 = <span>"CD3G"</span>, gene2 = <span>"CD247"</span>, </span></code><code><span leaf="">               <span>jitter</span>.num = <span>0</span>.<span>15</span>, pos = TRUE) + </span></code><code><span leaf="">  <span>theme</span>(aspect.ratio = <span>1</span>)</span></code><code><span leaf=""><span>ggsave</span>(<span>"correlation_dotplot.pdf"</span>)</span></code></pre></section><p><span leaf=""><img data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/fTGTY7BwuzfJyJrTXlmsJs0ibMwCPHp5DV3xicCMxccjQ3iaOIsKo2cL1icVgdHJhQCgp7sqcPU3kPSyibKb08WuVnQ/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.7138888888888889" data-type="png" data-w="1080" data-imgfileid="100003852" src="https://mmbiz.qpic.cn/sz_mmbiz_png/fTGTY7BwuzfJyJrTXlmsJs0ibMwCPHp5DV3xicCMxccjQ3iaOIsKo2cL1icVgdHJhQCgp7sqcPU3kPSyibKb08WuVnQ/640?wx_fmt=png&amp;from=appmsg"></span></p><section><ul><li><li><li><li></ul><pre data-lang="apache"><code><span leaf=""><span># 不添加抖动，显示原始数据位置 No jitter, display original data position</span></span></code><code><span leaf=""><span>getScatterplot</span>(pbmc3k, gene1 = <span>"CD3G"</span>, gene2 = <span>"CD247"</span>, </span></code><code><span leaf="">               <span>jitter</span>.num = <span>0</span>, pos = FALSE) + </span></code><code><span leaf="">  <span>theme</span>(aspect.ratio = <span>1</span>)</span></code></pre></section><p><span leaf=""><img data-imgfileid="100003853" data-ratio="0.7138888888888889" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/fTGTY7BwuzfJyJrTXlmsJs0ibMwCPHp5D5T6oacO1eOOOqHOwTv8ViaRnoEu4QZXAeqmXkUEc5T9djY6nVGWaibUQ/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/sz_mmbiz_png/fTGTY7BwuzfJyJrTXlmsJs0ibMwCPHp5D5T6oacO1eOOOqHOwTv8ViaRnoEu4QZXAeqmXkUEc5T9djY6nVGWaibUQ/640?wx_fmt=png&amp;from=appmsg"></span></p><section data-pm-slice="5 3 []"><section><span data-pm-slice="0 0 []"><h1 tabindex="-1" dir="auto" data-pm-slice="0 0 []"><span><strong><span><span leaf="">代码、输入、输出文件</span></span></strong></span></h1></span></section><p><span><span leaf="">https://github.com/ying-ge/FigureYa/tree/main/</span><span leaf="">FigureYa301scCoExpr</span></span></p><p><span><span leaf=""><br></span></span></p><section data-pm-slice="3 3 []"><p><span data-pm-slice="0 0 []"><span><strong><span><span leaf="">下载某个FigureYa的压缩包</span></span></strong></span></span></p><p><span leaf="">https://github.com/ying-ge/FigureYa-compressed</span></p></section><p><span leaf=""><br></span></p></section><h1 tabindex="-1" dir="auto" data-pm-slice="0 0 []"><span><strong><span><span leaf="">Citation</span></span></strong></span></h1><section><span><span leaf="">Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005</span></span></section><p><mp-style-type data-value="3"></mp-style-type></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/4_-cJ4_ILUvbj5vGPv-zqw",target="_blank" rel="noopener noreferrer">原文链接</a>
