---
title: "基于Signac的monocle3轨迹分析学习"
date: 2025-11-23T23:22:27Z
draft: ["false"]
tags: [
  "fetched",
  "单细胞天地"
]
categories: ["Acdemic"]
---
基于Signac的monocle3轨迹分析学习 by 单细胞天地
------
<div><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com" data-pm-slice="0 0 []"><p data-tool="mdnice编辑器"><span leaf="">除了scRNA-seq数据可以使用 Monocle 进行轨迹分析之外，scATAC-seq数据同样可以进行轨迹分析。</span></p><h4 data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><span></span><span><span leaf="">分析流程</span></span><span></span></h4><h5 data-tool="mdnice编辑器"><span></span><span><span leaf="">1.导入</span></span><span></span></h5><p data-tool="mdnice编辑器"><span leaf="">为了方便在Seurat（被Signac使用）和 CellDataSet（被Monocle 3 使用）格式之间进行转换，将使用SeuratWrappers包中提供的转换函数</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">library</span></span><span leaf="">(Signac)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(Seurat)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(SeuratWrappers)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(monocle3)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(Matrix)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(ggplot2)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(patchwork)</span><span leaf=""><br></span></code></pre><h5 data-tool="mdnice编辑器"><span></span><span><span leaf="">2.数据读取</span></span><span></span></h5><p data-tool="mdnice编辑器"><span leaf="">使用的数据集是GSE129785，里面包含了很多数据。选择GSE129785_scATAC-Hematopoiesis-CD34开头的peaks/barcodes和GSM3722029_CD34_Progenitors_Rep1_fragments.tsv.gz数据作为练习。<img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTuNKAQMnHVZEjq85tepL3xNRKGdLAwezBPN0204Qy6hJevK5FOTKdThkCicjlVmZEwBS2ibWBdrwjQ/640?wx_fmt=png&amp;from=appmsg" data-ratio="1.2661290322580645" data-type="png" data-w="620" data-imgfileid="100050186" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTuNKAQMnHVZEjq85tepL3xNRKGdLAwezBPN0204Qy6hJevK5FOTKdThkCicjlVmZEwBS2ibWBdrwjQ/640?wx_fmt=png&amp;from=appmsg"></span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">filepath &lt;- </span><span><span leaf="">"GSE129785_scATAC-Hematopoiesis-CD34"</span></span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 读取barcode/peaks等信息</span></span><span leaf=""><br></span><span leaf="">peaks &lt;- read.table(paste0(filepath, </span><span><span leaf="">".peaks.txt.gz"</span></span><span leaf="">), header = </span><span><span leaf="">TRUE</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">cells &lt;- read.table(paste0(filepath, </span><span><span leaf="">".cell_barcodes.txt.gz"</span></span><span leaf="">), header = </span><span><span leaf="">TRUE</span></span><span leaf="">, stringsAsFactors = </span><span><span leaf="">FALSE</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">rownames(cells) &lt;- make.unique(cells$Barcodes)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 读取matrix矩阵</span></span><span leaf=""><br></span><span leaf="">mtx &lt;- readMM(file = paste0(filepath, </span><span><span leaf="">".mtx"</span></span><span leaf="">))</span><span leaf=""><br></span><span leaf="">mtx &lt;- as(object = mtx, Class = </span><span><span leaf="">"CsparseMatrix"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">colnames(mtx) &lt;- rownames(cells)</span><span leaf=""><br></span><span leaf="">rownames(mtx) &lt;- peaks$Feature</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># fragment 文件包含在 GSE129785_RAW.tar 压缩包中，但其中并未提供对应的索引文件。可以使用tabix自行为fragment文件建立索引</span></span><span leaf=""><br></span><span><span leaf=""># 在linux/终端中使用：tabix -p bed &lt;fragment_file&gt;，去建立索引，索引相当于目录可以让分析软件快速找到指定染色体、基因组区间和peak区域。</span></span><span leaf=""><br></span><span leaf="">bone_assay &lt;- CreateChromatinAssay(</span><span leaf=""><br></span><span leaf="">  counts = mtx,</span><span leaf=""><br></span><span leaf="">  min.cells = </span><span><span leaf="">5</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  fragments = </span><span><span leaf="">"GSM3722029_CD34_Progenitors_Rep1_fragments.tsv.gz"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  sep = c(</span><span><span leaf="">"_"</span></span><span leaf="">, </span><span><span leaf="">"_"</span></span><span leaf="">),</span><span leaf=""><br></span><span leaf="">  genome = </span><span><span leaf="">"hg19"</span></span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span><span leaf=""># Computing hash</span></span><span leaf=""><br></span><span><span leaf=""># Checking for 18489 cell barcodes</span></span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">bone &lt;- CreateSeuratObject(</span><span leaf=""><br></span><span leaf="">  counts = bone_assay,</span><span leaf=""><br></span><span leaf="">  meta.data = cells,</span><span leaf=""><br></span><span leaf="">  assay = </span><span><span leaf="">"ATAC"</span></span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 该数据集包含多种细胞类型</span></span><span leaf=""><br></span><span><span leaf=""># 对子集进行筛选，只保留一个重复的CD34⁺祖细胞样本</span></span><span leaf=""><br></span><span leaf="">bone &lt;- bone[, bone$Group_Barcode == </span><span><span leaf="">"CD34_Progenitors_Rep1"</span></span><span leaf="">]</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 添加细胞注释</span></span><span leaf=""><br></span><span leaf="">cluster_names &lt;- c(</span><span><span leaf="">"HSC"</span></span><span leaf="">,   </span><span><span leaf="">"MEP"</span></span><span leaf="">,  </span><span><span leaf="">"CMP-BMP"</span></span><span leaf="">,  </span><span><span leaf="">"LMPP"</span></span><span leaf="">, </span><span><span leaf="">"CLP"</span></span><span leaf="">,  </span><span><span leaf="">"Pro-B"</span></span><span leaf="">,    </span><span><span leaf="">"Pre-B"</span></span><span leaf="">,    </span><span><span leaf="">"GMP"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">                  </span><span><span leaf="">"MDP"</span></span><span leaf="">,    </span><span><span leaf="">"pDC"</span></span><span leaf="">,  </span><span><span leaf="">"cDC"</span></span><span leaf="">,  </span><span><span leaf="">"Monocyte-1"</span></span><span leaf="">,   </span><span><span leaf="">"Monocyte-2"</span></span><span leaf="">,   </span><span><span leaf="">"Naive-B"</span></span><span leaf="">,  </span><span><span leaf="">"Memory-B"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">                  </span><span><span leaf="">"Plasma-cell"</span></span><span leaf="">,    </span><span><span leaf="">"Basophil"</span></span><span leaf="">, </span><span><span leaf="">"Immature-NK"</span></span><span leaf="">,  </span><span><span leaf="">"Mature-NK1"</span></span><span leaf="">,   </span><span><span leaf="">"Mature-NK2"</span></span><span leaf="">,   </span><span><span leaf="">"Naive-CD4-T1"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">                  </span><span><span leaf="">"Naive-CD4-T2"</span></span><span leaf="">,   </span><span><span leaf="">"Naive-Treg"</span></span><span leaf="">,   </span><span><span leaf="">"Memory-CD4-T"</span></span><span leaf="">, </span><span><span leaf="">"Treg"</span></span><span leaf="">, </span><span><span leaf="">"Naive-CD8-T1"</span></span><span leaf="">, </span><span><span leaf="">"Naive-CD8-T2"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">                  </span><span><span leaf="">"Naive-CD8-T3"</span></span><span leaf="">,   </span><span><span leaf="">"Central-memory-CD8-T"</span></span><span leaf="">, </span><span><span leaf="">"Effector-memory-CD8-T"</span></span><span leaf="">,    </span><span><span leaf="">"Gamma delta T"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">num.labels &lt;- length(cluster_names)</span><span leaf=""><br></span><span leaf="">names(cluster_names) &lt;- paste0( rep(</span><span><span leaf="">"Cluster"</span></span><span leaf="">, num.labels), seq(num.labels) )</span><span leaf=""><br></span><span leaf="">new.md &lt;- cluster_names[as.character(bone$Clusters)]</span><span leaf=""><br></span><span leaf="">names(new.md) &lt;- Cells(bone)</span><span leaf=""><br></span><span leaf="">bone$celltype &lt;- new.md</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">bone[[</span><span><span leaf="">"ATAC"</span></span><span leaf="">]]</span><span leaf=""><br></span></code></pre><h5 data-tool="mdnice编辑器"><span></span><span><span leaf="">3.将hg19基因组的基因注释添加到对象中，有助于计算质量控制指标（TSS富集分数）和绘图</span></span><span></span></h5><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">library</span></span><span leaf="">(EnsDb.Hsapiens.v75)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 从EnsDb中提取基因注释</span></span><span leaf=""><br></span><span leaf="">annotations &lt;- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v75)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 由于数据是比对到 hg19 的，因此将基因注释转换为UCSC风格</span></span><span leaf=""><br></span><span leaf="">seqlevels(annotations) &lt;- paste0(</span><span><span leaf="">'chr'</span></span><span leaf="">, seqlevels(annotations))</span><span leaf=""><br></span><span leaf="">genome(annotations) &lt;- </span><span><span leaf="">"hg19"</span></span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 将基因信息添加到对象中</span></span><span leaf=""><br></span><span leaf="">Annotation(bone) &lt;- annotations</span><span leaf=""><br></span></code></pre><h5 data-tool="mdnice编辑器"><span></span><span><span leaf="">4.质量控制</span></span><span></span></h5><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># 计算每个细胞的TSS富集、核小体信号分数以及基因组黑名单区域中的计数百分比，并使用这些指标来帮助从数据集中移除低质量细胞。</span></span><span leaf=""><br></span><span leaf="">bone &lt;- TSSEnrichment(bone)</span><span leaf=""><br></span><span leaf="">bone &lt;- NucleosomeSignal(bone)</span><span leaf=""><br></span><span leaf="">bone$blacklist_fraction &lt;- FractionCountsInRegion(bone, regions = blacklist_hg19)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">VlnPlot(</span><span leaf=""><br></span><span leaf="">  object = bone,</span><span leaf=""><br></span><span leaf="">  features = c(</span><span><span leaf="">"nCount_ATAC"</span></span><span leaf="">, </span><span><span leaf="">"TSS.enrichment"</span></span><span leaf="">, </span><span><span leaf="">"nucleosome_signal"</span></span><span leaf="">, </span><span><span leaf="">"blacklist_fraction"</span></span><span leaf="">),</span><span leaf=""><br></span><span leaf="">  pt.size = </span><span><span leaf="">0.1</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  ncol = </span><span><span leaf="">4</span></span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 过滤</span></span><span leaf=""><br></span><span leaf="">bone &lt;- bone[, (bone$nCount_ATAC &lt; </span><span><span leaf="">50000</span></span><span leaf="">) &amp;</span><span leaf=""><br></span><span leaf="">               (bone$TSS.enrichment &gt; </span><span><span leaf="">2</span></span><span leaf="">) &amp; </span><span leaf=""><br></span><span leaf="">               (bone$nucleosome_signal &lt; </span><span><span leaf="">5</span></span><span leaf="">)]</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTuNKAQMnHVZEjq85tepL3xAXFuqXYgLEKO4QcXym5Ctje2zmuiaBicSPbYrxDySmUz4YkTpbt1qJSQ/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.5555555555555556" data-type="png" data-w="1080" data-imgfileid="100050189" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTuNKAQMnHVZEjq85tepL3xAXFuqXYgLEKO4QcXym5Ctje2zmuiaBicSPbYrxDySmUz4YkTpbt1qJSQ/640?wx_fmt=png&amp;from=appmsg"></span></figure><h5 data-tool="mdnice编辑器"><span></span><span><span leaf="">5.数据集预处理</span></span><span></span></h5><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">bone &lt;- FindTopFeatures(bone, min.cells = </span><span><span leaf="">10</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">bone &lt;- RunTFIDF(bone)</span><span leaf=""><br></span><span leaf="">bone &lt;- RunSVD(bone, n = </span><span><span leaf="">100</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">DepthCor(bone)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">bone &lt;- RunUMAP(</span><span leaf=""><br></span><span leaf="">  bone,</span><span leaf=""><br></span><span leaf="">  reduction = </span><span><span leaf="">"lsi"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  dims = </span><span><span leaf="">2</span></span><span leaf="">:</span><span><span leaf="">50</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  reduction.name = </span><span><span leaf="">"UMAP"</span></span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">bone &lt;- FindNeighbors(bone, dims = </span><span><span leaf="">2</span></span><span leaf="">:</span><span><span leaf="">50</span></span><span leaf="">, reduction = </span><span><span leaf="">"lsi"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">bone &lt;- FindClusters(bone, resolution = </span><span><span leaf="">0.8</span></span><span leaf="">, algorithm = </span><span><span leaf="">3</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">DimPlot(bone, label = </span><span><span leaf="">TRUE</span></span><span leaf="">) + NoLegend()</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTuNKAQMnHVZEjq85tepL3xy6Osz9frtibyM0sHT1My5uaIedm0Qsic6GVxfz0nbqcgX65UlOzSe92g/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.9251439539347409" data-type="png" data-w="1042" data-imgfileid="100050185" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTuNKAQMnHVZEjq85tepL3xy6Osz9frtibyM0sHT1My5uaIedm0Qsic6GVxfz0nbqcgX65UlOzSe92g/640?wx_fmt=png&amp;from=appmsg"><img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTuNKAQMnHVZEjq85tepL3x0Mh7Eia3ibDeOr5mxEMsULiavXic1DwrwBBEw4lznQziaQ2dxfwd2okl3nQ/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.9251439539347409" data-type="png" data-w="1042" data-imgfileid="100050188" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTuNKAQMnHVZEjq85tepL3x0Mh7Eia3ibDeOr5mxEMsULiavXic1DwrwBBEw4lznQziaQ2dxfwd2okl3nQ/640?wx_fmt=png&amp;from=appmsg"></span></p><h5 data-tool="mdnice编辑器"><span></span><span><span leaf="">6.细胞注释</span></span><span></span></h5><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">for</span></span><span leaf="">(i </span><span><span leaf="">in</span></span><span leaf=""> levels(bone)) {</span><span leaf=""><br></span><span leaf="">  cells_to_reid &lt;- WhichCells(bone, idents = i)</span><span leaf=""><br></span><span leaf="">  newid &lt;- names(sort(table(bone$celltype[cells_to_reid]),decreasing=</span><span><span leaf="">TRUE</span></span><span leaf="">))[</span><span><span leaf="">1</span></span><span leaf="">]</span><span leaf=""><br></span><span leaf="">  Idents(bone, cells = cells_to_reid) &lt;- newid</span><span leaf=""><br></span><span leaf="">}</span><span leaf=""><br></span><span leaf="">bone$assigned_celltype &lt;- Idents(bone)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">DimPlot(bone, label = </span><span><span leaf="">TRUE</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTuNKAQMnHVZEjq85tepL3xTAcGI5bqZnPr1c0oiaWuncZBj76SVKDdHo91pDn3M1tVKHrfwFv72RA/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.9251439539347409" data-type="png" data-w="1042" data-imgfileid="100050187" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTuNKAQMnHVZEjq85tepL3xTAcGI5bqZnPr1c0oiaWuncZBj76SVKDdHo91pDn3M1tVKHrfwFv72RA/640?wx_fmt=png&amp;from=appmsg"></span></figure><h5 data-tool="mdnice编辑器"><span></span><span><span leaf="">7.可以对子集进行不同谱系的筛选，并为每个谱系构建轨迹。另一种构建轨迹的方法是使用整个数据集，并针对Monocle3发现的不同细胞分区分别构建伪时间轨迹</span></span><span></span></h5><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">DefaultAssay(bone) &lt;- </span><span><span leaf="">"ATAC"</span></span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">erythroid &lt;- bone[,  bone$assigned_celltype %</span><span><span leaf="">in</span></span><span leaf="">% c(</span><span><span leaf="">"HSC"</span></span><span leaf="">, </span><span><span leaf="">"MEP"</span></span><span leaf="">, </span><span><span leaf="">"CMP-BMP"</span></span><span leaf="">)]</span><span leaf=""><br></span><span leaf="">lymphoid &lt;- bone[, bone$assigned_celltype %</span><span><span leaf="">in</span></span><span leaf="">% c(</span><span><span leaf="">"HSC"</span></span><span leaf="">, </span><span><span leaf="">"LMPP"</span></span><span leaf="">, </span><span><span leaf="">"GMP"</span></span><span leaf="">, </span><span><span leaf="">"CLP"</span></span><span leaf="">, </span><span><span leaf="">"Pro-B"</span></span><span leaf="">, </span><span><span leaf="">"pDC"</span></span><span leaf="">, </span><span><span leaf="">"MDP"</span></span><span leaf="">, </span><span><span leaf="">"GMP"</span></span><span leaf="">)]</span><span leaf=""><br></span></code></pre><h5 data-tool="mdnice编辑器"><span></span><span><span leaf="">8.使用Monocle3构建轨迹</span></span><span></span></h5><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">erythroid.cds &lt;- as.cell_data_set(erythroid)</span><span leaf=""><br></span><span leaf="">erythroid.cds &lt;- cluster_cells(cds = erythroid.cds, reduction_method = </span><span><span leaf="">"UMAP"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">erythroid.cds &lt;- learn_graph(erythroid.cds, use_partition = </span><span><span leaf="">TRUE</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">lymphoid.cds &lt;- as.cell_data_set(lymphoid)</span><span leaf=""><br></span><span leaf="">lymphoid.cds &lt;- cluster_cells(cds = lymphoid.cds, reduction_method = </span><span><span leaf="">"UMAP"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">lymphoid.cds &lt;- learn_graph(lymphoid.cds, use_partition = </span><span><span leaf="">TRUE</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><h5 data-tool="mdnice编辑器"><span></span><span><span leaf="">9.要计算每条轨迹的伪时间估计，需要确定每条轨迹的起点。在该例子中，要知道造血干细胞（HSCs）是轨迹中其他细胞类型的祖细胞，因此可以将这些细胞设置为轨迹的根节点。Monocle3提供了一个交互式函数，用于在图中选择细胞作为根节点。如果在调用order_cells时不指定root_cells参数，该函数会自动启动。研究团队已经预先选择了一些细胞作为根节点，并将其保存到文件中以保证可重复性。这个文件可以官网下载。</span></span><span></span></h5><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># 加载预先筛选的造血干细胞（HSCs）</span></span><span leaf=""><br></span><span><span leaf=""># 这个文件是官方教程提供的，如果是自己的数据可能要自己做准备</span></span><span leaf=""><br></span><span leaf="">hsc &lt;- readLines(</span><span><span leaf="">"./hsc_cells.txt"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 对细胞进行排序</span></span><span leaf=""><br></span><span leaf="">erythroid.cds &lt;- order_cells(erythroid.cds, reduction_method = </span><span><span leaf="">"UMAP"</span></span><span leaf="">, root_cells = hsc)</span><span leaf=""><br></span><span leaf="">lymphoid.cds &lt;- order_cells(lymphoid.cds, reduction_method = </span><span><span leaf="">"UMAP"</span></span><span leaf="">, root_cells = hsc)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 绘制轨迹图</span></span><span leaf=""><br></span><span leaf="">plot_cells(</span><span leaf=""><br></span><span leaf="">  cds = erythroid.cds,</span><span leaf=""><br></span><span leaf="">  color_cells_by = </span><span><span leaf="">"pseudotime"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  show_trajectory_graph = </span><span><span leaf="">TRUE</span></span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">plot_cells(</span><span leaf=""><br></span><span leaf="">  cds = lymphoid.cds,</span><span leaf=""><br></span><span leaf="">  color_cells_by = </span><span><span leaf="">"pseudotime"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  show_trajectory_graph = </span><span><span leaf="">TRUE</span></span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">红系细胞<img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTuNKAQMnHVZEjq85tepL3xCUxp02SPvic5fIS6zTNakicf8j7n1FQEwsCjSsm5eC6HKCjickkWZyCfQ/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.9251439539347409" data-type="png" data-w="1042" data-imgfileid="100050190" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTuNKAQMnHVZEjq85tepL3xCUxp02SPvic5fIS6zTNakicf8j7n1FQEwsCjSsm5eC6HKCjickkWZyCfQ/640?wx_fmt=png&amp;from=appmsg">淋巴系细胞<img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTuNKAQMnHVZEjq85tepL3xRIosQw3bb9H40YrDQIqvvIrzD4xDDubmJmu3zLibdQSNuIWBx0VLO2Q/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.9251439539347409" data-type="png" data-w="1042" data-imgfileid="100050191" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTuNKAQMnHVZEjq85tepL3xRIosQw3bb9H40YrDQIqvvIrzD4xDDubmJmu3zLibdQSNuIWBx0VLO2Q/640?wx_fmt=png&amp;from=appmsg"></span></p><h6 data-tool="mdnice编辑器"><span></span><span><span leaf="">9.1 提取伪时间值并将其添加到Seurat对象中</span></span><span></span></h6><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">bone &lt;- AddMetaData(</span><span leaf=""><br></span><span leaf="">  object = bone,</span><span leaf=""><br></span><span leaf="">  metadata = erythroid.cds@principal_graph_aux@listData$UMAP$pseudotime,</span><span leaf=""><br></span><span leaf="">  col.name = </span><span><span leaf="">"Erythroid"</span></span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">bone &lt;- AddMetaData(</span><span leaf=""><br></span><span leaf="">  object = bone,</span><span leaf=""><br></span><span leaf="">  metadata = lymphoid.cds@principal_graph_aux@listData$UMAP$pseudotime,</span><span leaf=""><br></span><span leaf="">  col.name = </span><span><span leaf="">"Lymphoid"</span></span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">FeaturePlot(bone, c(</span><span><span leaf="">"Erythroid"</span></span><span leaf="">, </span><span><span leaf="">"Lymphoid"</span></span><span leaf="">), pt.size = </span><span><span leaf="">0.1</span></span><span leaf="">) &amp; scale_color_viridis_c()</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTuNKAQMnHVZEjq85tepL3xXRznD7RibS0jkF2eLMULpliaxRbh10Mia7Zton1eUeqFQnhAym72mk8Tw/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.725" data-type="png" data-w="1080" data-imgfileid="100050192" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTuNKAQMnHVZEjq85tepL3xXRznD7RibS0jkF2eLMULpliaxRbh10Mia7Zton1eUeqFQnhAym72mk8Tw/640?wx_fmt=png&amp;from=appmsg"></span></figure><h4 data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><span></span><span><span leaf="">参考资料：</span></span><span></span></h4><ol><li><section><span leaf="">Signac monocle3：https://stuartlab.org/signac/articles/monocle</span></section></li><li><section><span leaf="">hsc_cells文件：https://www.dropbox.com/scl/fi/q2s8kxebrz48wasmxgzs4/hsc_cells.txt?rlkey=5ytjega4r7jais43ltkyvahy6&amp;e=1</span></section></li></ol></section><section><span leaf=""><br></span></section><p><mp-style-type data-value="3"></mp-style-type></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/BqUTMAM7ljODhMuSMim88w",target="_blank" rel="noopener noreferrer">原文链接</a>
