---
title: "bulk RNA-seq | 下游分析 | GSVA与ssGSEA"
date: 2024-12-03T00:52:38Z
draft: ["false"]
tags: [
  "fetched",
  "被炸熟的虾"
]
categories: ["Acdemic"]
---
bulk RNA-seq | 下游分析 | GSVA与ssGSEA by 被炸熟的虾
------
<div><p><span>bulk RNA-seq</span><span>下游分析部分，我们已经先后记录了</span></p><ul><li><section><a href="https://mp.weixin.qq.com/s?__biz=MzkwNDQwMDI5NQ==&amp;mid=2247488268&amp;idx=1&amp;sn=8b6acae06a129b028d95648193a44de0&amp;scene=21#wechat_redirect" data-linktype="2"><span><strong><span>bulk RNA-seq | 下游分析 | 功能富集分析 ORA - clusterProfiler</span></strong></span></a></section></li><li><section><a href="https://mp.weixin.qq.com/s?__biz=MzkwNDQwMDI5NQ==&amp;mid=2247488268&amp;idx=1&amp;sn=8b6acae06a129b028d95648193a44de0&amp;scene=21#wechat_redirect" data-linktype="2"><span><strong><span>功能富集分析 ORA | clusterProfiler 读取本地化KEGG数据</span></strong></span></a></section></li><li><section><a href="https://mp.weixin.qq.com/s?__biz=MzkwNDQwMDI5NQ==&amp;mid=2247488571&amp;idx=1&amp;sn=fdd0eb2b878da4b21bac1fd009ce6e67&amp;chksm=c086c8f3f7f141e55dd6a9389479c909fbd5559403d8106e510bc7c678e4e55fdde8949aad10&amp;scene=21&amp;cur_album_id=2550626183059685377#wechat_redirect" data-linktype="2"><span><strong><span>bulk RNA-seq | 下游分析 | 基因集富集分析 GSEA- clusterProfiler</span></strong></span></a></section></li></ul><p><span>下面接着记录</span><span>GSVA</span><span>与</span><span>ssGSEA</span><span>方法。这一部分之前也写过：</span><a href="https://mp.weixin.qq.com/s?__biz=MzkwNDQwMDI5NQ==&amp;mid=2247484182&amp;idx=1&amp;sn=fe3f83e5e67677413fb749bcfa6cc132&amp;payreadticket=HOim2eFVyWiPlEQbB5GBnb5JWm4ODyJxEQwVs-HENBdUNJ61GdWxT4cA8Pd1vV3I7zGYvMo&amp;scene=21#wechat_redirect" data-linktype="2"><span><strong><span>ssGSEA与GSVA使用方法</span></strong></span></a><span>，主体内容还是那些，一个更新原因是不知道为什么，</span><span>GSVA</span><span>包（支持</span><span>GSVA</span><span>、</span><span>ssGSEA</span><span>、</span><span>zscore</span><span>与</span><span>plage</span><span>等多种方法）的作者更新了函数！！</span></p><h3><strong><span>1、ssGSEA</span></strong></h3><p><span>单样本基因集富集分析（</span><span>single sample gene set enrichment analysis</span><span>, </span><span>ssGSEA</span><span>），原理上与</span><span>GSEA</span><span>类似，不同的是</span><span>GSEA</span><span> 主要用于检测不同实验组（如实验组和对照组）之间基因集富集差异，而</span><span>ssGSEA</span><span>将样本内基因表达谱进行归一化处理，然后计算每个基因集对应的</span><span>ssGSEA</span><span>得分。通过这种方式，</span><span>ssGSEA</span><span>将单个样本的基因表达谱转换为基因集富集得分矩阵。因此，如果使用的是免疫细胞</span><span>marker</span><span>组成的基因集，</span><span>ssGESA</span><span>也可以计算免疫细胞浸润评分。</span></p><h3><strong><span>2、GSVA</span></strong></h3><p><span>基因集变异分析（</span><span>Gene Set Variation Analysis</span><span>，</span><span>GSVA</span><span>）和</span><span>ssGSEA</span><span>非常相似，同样是对单个样本进行计算，最后输出的也是一个行为不同基因集、列为样本的矩阵。主要区别在于</span><span>GSVA</span><span>使用的是非参数核密度估计方法，而</span><span>ssGSEA</span><span>是基于样本排序的积分方法。</span></p><h4>更新前</h4><p><span>输入表达矩阵以及基因集，然后选择方法就可以了。</span></p><pre><code>packageVersion(<span>"GSVA"</span>)<span>#[1] ‘1.42.0’</span><br></code></pre><p><img data-imgfileid="100004947" data-ratio="0.5341726618705036" data-s="300,640" data-type="png" data-w="556" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3Jy9UPcZiaia8b6vEKCDWJ1XVibssE1azPY6crODlg5uWpdePVHGibPS82kKmI5Hq9CSDo4uPDbOCcKqQ/640?wx_fmt=png&amp;from=appmsg" src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3Jy9UPcZiaia8b6vEKCDWJ1XVibssE1azPY6crODlg5uWpdePVHGibPS82kKmI5Hq9CSDo4uPDbOCcKqQ/640?wx_fmt=png&amp;from=appmsg"></p><ul><li><section><span>expr</span><span>：基因的表达量需要的是矩阵格式（</span><span>matrix</span><span>，不是</span><span>data.frame</span><span>），行为基因，列为样本；</span></section></li><li><section><span>gset.idx.list</span><span>：基因集需要的是</span><span>list</span><span>格式；</span></section></li><li><section><span>kcdf</span><span>：默认情况下，参数</span><span>kcdf="Gaussian"</span><span>，适用于对数转换的</span><span>microarray</span><span>、</span><span>RNA-seq</span><span>的</span><span>log-CPMs</span><span>、</span><span>log-RPKMs</span><span>或</span><span>log-TPMs</span><span>。当输入的表达的矩阵是</span><span>RNA-seq</span><span>的</span><span>raw Count</span><span>时，这个参数应该设置为</span><span>kcdf=”Poisson”</span><span>。</span></section></li></ul><pre><code>kcdf = c(<span>"Gaussian"</span>,<span>"Poisson"</span>,<span>"none"</span>)<br></code></pre><ul><li><section><span>method</span><span>：参数</span><span>method</span><span>用于指定估计每个样本的基因集富集分数的方法，默认情况下，</span><span>method=”gsva”</span><span>。</span><span><strong><span></span></strong></span></section></li></ul><pre><code>method = c(<span>"gsva"</span>, <span>"ssgsea"</span>,<span>"zscore"</span>,<span>"plage"</span>)<br></code></pre><h4><span>更新后</span></h4><pre><code>packageVersion(<span>"GSVA"</span>)<br><span>#[1] ‘2.0.1’</span><br></code></pre><p><span>第一步：需要先定义</span><span>param</span><span>。这个</span><span>param</span><span>包括表达矩阵以及基因集等。不同方法使用不同</span><span>param</span><span>定义函数！！</span></p><section><span><img data-imgfileid="100004948" data-ratio="0.3990740740740741" data-type="png" data-w="1080" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3Jy9UPcZiaia8b6vEKCDWJ1XVusHKpUQAMAHPaj5VU9GNUSpJLDjwEd3pT7rJCH8HoyTO6kCeWCghFA/640?wx_fmt=png&amp;from=appmsg" src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3Jy9UPcZiaia8b6vEKCDWJ1XVusHKpUQAMAHPaj5VU9GNUSpJLDjwEd3pT7rJCH8HoyTO6kCeWCghFA/640?wx_fmt=png&amp;from=appmsg"></span></section><p><span>参数</span><span>kcdf</span><span>也被纳入了这一步。</span></p><p><img data-imgfileid="100004949" data-ratio="0.5275288092189501" data-s="300,640" data-type="png" data-w="781" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3Jy9UPcZiaia8b6vEKCDWJ1XVtz4Fm7hGBNcqY4Jniav1SSBDMsMXZMbhzESnSkBb7sLfHXLSrfSIpzw/640?wx_fmt=png&amp;from=appmsg" src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3Jy9UPcZiaia8b6vEKCDWJ1XVtz4Fm7hGBNcqY4Jniav1SSBDMsMXZMbhzESnSkBb7sLfHXLSrfSIpzw/640?wx_fmt=png&amp;from=appmsg"></p><h4><section>示例流程</section></h4><p><span>首先依旧是数据准备：</span></p><pre><code><span>library</span>(GSVA)<br>set.seed(<span>12345</span>)<br>p &lt;- <span>10000</span> <span>## number of genes</span><br>n &lt;- <span>30</span>    <span>## number of samples</span><br><span>## simulate expression values from a standard Gaussian distribution</span><br>X &lt;- matrix(rnorm(p*n), nrow = p,<br>            dimnames = list(paste0(<span>"g"</span>, <span>1</span>:p), paste0(<span>"s"</span>, <span>1</span>:n)))<br>X[<span>1</span>:<span>5</span>, <span>1</span>:<span>5</span>]<br><span>## sample gene set sizes</span><br>gs &lt;- as.list(sample(<span>10</span>:<span>100</span>, size = <span>100</span>, replace = <span>TRUE</span>))<br><span>## sample gene sets</span><br>gs &lt;- lapply(gs, <span>function</span>(n, p)<br>                   paste0(<span>"g"</span>, sample(<span>1</span>:p, size = n, replace = <span>FALSE</span>)), p)<br>names(gs) &lt;- paste0(<span>"gs"</span>, <span>1</span>:length(gs))<br>lapply(gs[<span>1</span>:<span>3</span>],head)<br><span>#$gs1</span><br><span>#[1] "g7824" "g7976" "g7005" "g5451" "g1199" "g227" </span><br><span>#$gs2</span><br><span>#[1] "g5066" "g6691" "g542"  "g7062" "g2891" "g8778"</span><br><span>#$gs3</span><br><span>#[1] "g6546" "g1957" "g2913" "g8503" "g5013" "g8340"</span><br></code></pre><p><span>更新前的分析方法只要一步：</span></p><pre><code><span>##We can calculate GSVA enrichment scores as follows:</span><br><span>library</span>(GSVA)<br>gsva.es &lt;- gsva(expr = X, <br>                gset.idx.list = gs,<br>                method = <span>'gsva'</span>,<br>                kcdf = <span>'Gaussian'</span>,<br>                verbose = <span>FALSE</span>)<br>dim(gsva.es)<span>#[1] 100  30</span><br>gsva.es[<span>1</span>:<span>5</span>, <span>1</span>:<span>5</span>]<br><span>#              s1          s2          s3           s4           s5</span><br><span>#gs1 -0.163169867 -0.18230994  0.36746012 -0.010434415 -0.026349618</span><br><span>#gs2  0.007954471 -0.05686215 -0.20594532 -0.009643796 -0.057696047</span><br><span>#gs3 -0.161205687  0.24624689  0.09306792 -0.033363937 -0.002946779</span><br><span>#gs4  0.126560368 -0.08559508 -0.26097093 -0.181676230 -0.307544679</span><br><span>#gs5  0.076199136 -0.01887931 -0.02707651 -0.069945789 -0.086908249</span><br></code></pre><p><span>更新后多了一步：</span></p><pre><code><span>##First we should build a parameter object for the desired methodology.R</span><br>gsvaPar &lt;- gsvaParam(exprData = X, <br>                     geneSets = gs,<br>                     kcdf = <span>"Gaussian"</span>)<br>gsvaPar<br><span>##Second, we call the gsva() function with the parameter object as first argument. </span><br>gsva.es &lt;- gsva(gsvaPar, verbose = <span>FALSE</span>)<br>dim(gsva.es)<span>#[1] 100  30</span><br>gsva.es[<span>1</span>:<span>5</span>, <span>1</span>:<span>5</span>]<br><span>#              s1          s2          s3           s4           s5</span><br><span>#gs1 -0.163169867 -0.18230994  0.36746012 -0.010434415 -0.026349618</span><br><span>#gs2  0.007954471 -0.05686215 -0.20594532 -0.009643796 -0.057696047</span><br><span>#gs3 -0.161205687  0.24624689  0.09306792 -0.033363937 -0.002946779</span><br><span>#gs4  0.126560368 -0.08559508 -0.26097093 -0.181676230 -0.307544679</span><br><span>#gs5  0.076199136 -0.01887931 -0.02707651 -0.069945789 -0.086908249</span><br></code></pre><p><span>无法理解更新目的🤔。接下来，再使用新版本是</span><span>GSVA</span><span>包跑一遍之前的笔记流程：</span></p><h2><span>1.ssGSEA方法评估样本免疫浸润水平</span></h2><p><span>重现结果来自：</span><span>Fig3 A-B</span><span>，评估肿瘤样本的免疫细胞浸润水平：</span></p><blockquote><p>Local mutational diversity drives intratumoral immune heterogeneity in non-small cell lung cancer. Nat Commun. 2018 Dec 18;9(1):5361. doi: 10.1038/s41467-018-07767-w. PMID: 30560866.</p></blockquote><p><span>免疫基因集出自文章</span><span>PMID: 28052254</span><span>,包括</span><span>28</span><span>种免疫细胞的基因集：</span></p><blockquote><p>Pan-cancer Immunogenomic Analyses Reveal Genotype-Immunophenotype Relationships and Predictors of Response to Checkpoint Blockade. Cell Rep. 2017 Jan 3;18(1):248-262. doi: 10.1016/j.celrep.2016.12.019. PMID: 28052254.</p></blockquote><h3><span>数据准备</span></h3><pre><code>rm(list = ls())<br>options(stringsAsFactors = <span>F</span>)<br>setwd(<span>"D:/bioinformatics/9.Others/7 Immune cell infiltration/data"</span>)<br>set.seed(<span>123456</span>)<br><span>###################################表达谱提取####################################</span><br>raw_data &lt;- read.table(<span>'GSE112996_merged_fpkm_table.txt.gz'</span>,header = <span>T</span>,row.names = <span>1</span>,check.names = <span>F</span>)<br><span>#rowSums(raw_data[,-1])</span><br>raw_data &lt;- raw_data[rowSums(raw_data[,-<span>1</span>])!=<span>0</span>,]<br><br><span>#####################################ID转换######################################</span><br>ids &lt;- data.frame(GeneSymbol = raw_data$GeneName,Ensembl = rownames(raw_data))<br>ids$median &lt;- apply(raw_data[,-<span>1</span>],<span>1</span>,median)<br>ids &lt;- ids[order(ids$GeneSymbol,ids$median,decreasing = <span>T</span>),]<br>ids &lt;- ids[!duplicated(ids$GeneSymbol),]<br>rownames(ids) &lt;- ids$GeneSymbol<br>clean_data &lt;- raw_data[ids$Ensembl,-<span>1</span>]<br>rownames(clean_data) &lt;- ids$GeneSymbol<br><br>uni_matrix &lt;- log2(clean_data+<span>1</span>)<br>boxplot(uni_matrix)<br><br><span>###################################表型信息提取###################################</span><br>pheno &lt;- read.csv(file = <span>'GSE112996_series_matrix.txt.gz'</span>)<br>pheno &lt;- data.frame(num1 = strsplit(as.character(pheno[<span>42</span>,]),split=<span>'\t'</span>)[[<span>1</span>]][-<span>1</span>],<br>                    num2 = gsub(<span>'patient: No.'</span>,<span>'P'</span>,strsplit(as.character(pheno[<span>51</span>,]),split=<span>'\t'</span>)[[<span>1</span>]][-<span>1</span>]))<br>pheno &lt;- pheno[order(pheno$num2),]<br>identical(colnames(uni_matrix),pheno$num1)<br>table(colnames(uni_matrix) %<span>in</span>% pheno$num1)<br>uni_matrix &lt;- uni_matrix[,pheno$num1]<br><br><span>###################################免疫基因集###################################</span><br>geneset &lt;- rio::import(<span>"ssGSEA-PMID28052254-Signature-28.xlsx"</span>,skip = <span>2</span>)<br>geneset &lt;- split(geneset$Metagene,geneset$`Cell type`)<br>lapply(geneset[<span>1</span>:<span>3</span>], head)<br><span>## $`Activated B cell`</span><br><span>## [1] "ADAM28" "CD180"  "CD79B"  "BLK"    "CD19"   "MS4A1" </span><br><span>## $`Activated CD4 T cell`</span><br><span>## [1] "AIM2"  "BIRC3" "BRIP1" "CCL20" "CCL4"  "CCL5" </span><br><span>## $`Activated CD8 T cell`</span><br><span>## [1] "ADRM1"     "AHSA1"     "C1GALT1C1" "CCT6B"     "CD37"      "CD3D"</span><br>names(geneset)<br></code></pre><h3><span>运行ssGSEA</span></h3><p><span>运行很简单，更新后多了一步：</span></p><pre><code><span>library</span>(GSVA)<br><span>##First we should build a parameter object for the desired methodology.R</span><br>gsvaPar &lt;- ssgseaParam(exprData = as.matrix(uni_matrix), <br>                       geneSets = geneset,<br>                       normalize = <span>TRUE</span>)<br><span>##Second, we call the gsva() function with the parameter object as first argument. </span><br>ssGSEA_matrix &lt;- gsva(gsvaPar, verbose = <span>FALSE</span>)<br>dim(ssGSEA_matrix)<span>#[1] 100  30</span><br>ssGSEA_matrix[<span>1</span>:<span>5</span>, <span>1</span>:<span>5</span>]<br><span>###normalize参数用最大值与最小值间的绝对差对ssGSEA分数进行标准化,默认为TRUE</span><br></code></pre><h3><section>结果可视化</section></h3><section><strong>1、样本免疫细胞浸润水平热图：</strong></section><pre><code>scale_gsva_matrix &lt;- t(scale(t(ssGSEA_matrix)))<br>scale_gsva_matrix[scale_gsva_matrix&lt; -<span>2</span>] &lt;- -<span>2</span><br>scale_gsva_matrix[scale_gsva_matrix&gt;<span>2</span>] &lt;- <span>2</span><br><span>##重新排序</span><br>anti_tumor &lt;- c(<span>'Activated CD4 T cell'</span>, <span>'Activated CD8 T cell'</span>, <span>'Central memory CD4 T cell'</span>, <br>                <span>'Central memory CD8 T cell'</span>, <span>'Effector memeory CD4 T cell'</span>, <span>'Effector memeory CD8 T cell'</span>,<br>                <span>'Type 1 T helper cell'</span>, <span>'Type 17 T helper cell'</span>, <span>'Activated dendritic cell'</span>,<br>                <span>'CD56bright natural killer cell'</span>, <span>'Natural killer cell'</span>, <span>'Natural killer T cell'</span>)<br>pro_tumor &lt;- c(<span>'Regulatory T cell'</span>, <span>'Type 2 T helper cell'</span>, <span>'CD56dim natural killer cell'</span>,<br>               <span>'Immature dendritic cell'</span>, <span>'Macrophage'</span>, <span>'MDSC'</span>, <span>'Neutrophil'</span>, <span>'Plasmacytoid dendritic cell'</span>)<br><br>anti &lt;- rownames(scale_gsva_matrix) %<span>in</span>% anti_tumor<br>pro &lt;- rownames(scale_gsva_matrix) %<span>in</span>% pro_tumor<br>non &lt;- !(anti|pro)<br>scale_gsva_matrix &lt;- rbind(scale_gsva_matrix[anti,],scale_gsva_matrix[pro,],scale_gsva_matrix[non,])<br><span>###数值归一化到[0，1]</span><br>normalization &lt;- <span>function</span>(x){<span>return</span>((x-min(x))/(max(x)-min(x)))}  <br>nor_gsva_matrix &lt;- normalization(scale_gsva_matrix)<br><br><span>library</span>(pheatmap)<br>annotation_col = data.frame(patient = pheno$num2)<br>rownames(annotation_col) &lt;- colnames(uni_matrix)<br>Heatmap &lt;- pheatmap(nor_gsva_matrix,<br>                    show_colnames = <span>F</span>,<br>                    cluster_rows = <span>F</span>,cluster_cols = <span>F</span>,<br>                    annotation_col = annotation_col,<br>                    cellwidth = <span>5</span>,cellheight = <span>5</span>,<br>                    fontsize = <span>5</span>,<br>                    gaps_row = c(<span>12</span>,<span>20</span>))<br><span>#save(gsva_matrix,gsva_matrix1,pheno,file = 'score.Rdata')</span><br></code></pre><figure><img data-imgfileid="100004942" data-ratio="0.47706422018348627" data-w="981" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3IhQ4at5dbGaI6tiaeS34X1p2HhtCvE02DLffR8mSVI9tS9qPtDjztON2TjW7huRkbgTnf66ulh06g/640?wx_fmt=png" src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3IhQ4at5dbGaI6tiaeS34X1p2HhtCvE02DLffR8mSVI9tS9qPtDjztON2TjW7huRkbgTnf66ulh06g/640?wx_fmt=png"><figcaption>img</figcaption></figure><section><strong>2、Anti-tumor immunity vs. Pro-tumor suppression 散点图：</strong></section><section><span>计算不同角色免疫细胞</span><span>score</span><span>加和后，使用</span><span>ggplot2</span><span>绘制</span></section><pre><code>anti &lt;- as.data.frame(scale_gsva_matrix[rownames(scale_gsva_matrix) %<span>in</span>% anti_tumor,])<br>pro &lt;- as.data.frame(scale_gsva_matrix[rownames(scale_gsva_matrix) %<span>in</span>% pro_tumor,])<br>anti_n &lt;- apply(anti,<span>2</span>,sum)<br>pro_n &lt;- apply(pro,<span>2</span>,sum)<br>patient &lt;- pheno$num2[match(colnames(scale_gsva_matrix),pheno$num1)]<br><span>library</span>(ggplot2)<br><span>library</span>(ggpubr)<br>data &lt;- data.frame(anti = anti_n,pro = pro_n,patient=patient)<br>anti_pro&lt;- cor.test(anti_n,pro_n,method = <span>'pearson'</span>)<br>gg &lt;- ggplot(data,aes(x = anti, y = pro)) + <br>      xlim(-<span>20</span>,<span>15</span>) + ylim(-<span>15</span>,<span>10</span>) +<br>      labs(x = <span>"Anti-tumor immunity"</span>, y = <span>"Pro-tumor suppression"</span>) +<br>      geom_point(aes(color = patient),size = <span>3</span>) +<br>      geom_smooth(method =<span>'lm'</span>) +<br>      stat_cor(method = <span>"spearman"</span>,size = <span>8</span>) +<br>      scale_colour_manual(values = RColorBrewer::brewer.pal(<span>12</span>, <span>"Set3"</span>)) +<br>      theme_bw() + <br>      theme(axis.text.x = element_text(size = <span>18</span>), <br>            axis.text.y = element_text(size = <span>18</span>), <br>            axis.title.x = element_text(size = <span>20</span>), <br>            axis.title.y = element_text(size = <span>20</span>),<br>            legend.title = element_blank(),<br>            legend.text = element_text(size = <span>20</span>))<br>gg<br></code></pre><p><img data-imgfileid="100004944" data-ratio="0.875" data-w="1080" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3IhQ4at5dbGaI6tiaeS34X1p2sDozGvdL0atx5RWreSS8BFVH8iaEFibibdXXXnjlwuzKrElWKsd3yNwQ/640?wx_fmt=png" src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3IhQ4at5dbGaI6tiaeS34X1p2sDozGvdL0atx5RWreSS8BFVH8iaEFibibdXXXnjlwuzKrElWKsd3yNwQ/640?wx_fmt=png"></p><section><span>计算基因表达与浸润得分相关性：</span></section><pre><code><span>library</span>(Hmisc)<br>identical(colnames(ssGSEA_matrix),colnames(uni_matrix))<br>gs = geneset[[<span>1</span>]]<br>gs &lt;- gs[gs %<span>in</span>% rownames(uni_matrix)]<br>nc = t(rbind(ssGSEA_matrix,uni_matrix[gs,]))<br>nc[<span>1</span>:<span>4</span>,<span>1</span>:<span>4</span>]<br>m = rcorr(nc)$r[<span>1</span>:nrow(ssGSEA_matrix),(ncol(nc)-length(gs)+<span>1</span>):ncol(nc)]<br>m[<span>1</span>:<span>4</span>,<span>1</span>:<span>4</span>]<br><span>##                            ADAM28     CD180     CD79B       BLK</span><br><span>##Activated B cell         0.4613288 0.2455269 0.8656336 0.7610076</span><br><span>##Activated CD4 T cell     0.1121755 0.1026576 0.4376069 0.1188357</span><br><span>##Activated CD8 T cell     0.4475305 0.2016188 0.6413575 0.3857065</span><br><span>##Activated dendritic cell 0.5078425 0.1712112 0.1222768 0.2304565</span><br>p = rcorr(nc)$P[<span>1</span>:nrow(ssGSEA_matrix),(ncol(nc)-length(gs)+<span>1</span>):ncol(nc)]<br>p[<span>1</span>:<span>4</span>,<span>1</span>:<span>4</span>]<br><span>##                               ADAM28     CD180        CD79B          BLK</span><br><span>##Activated B cell         0.0014247371 0.1040157 1.687539e-14 1.299134e-09</span><br><span>##Activated CD4 T cell     0.4631674845 0.5021913 2.645818e-03 4.368624e-01</span><br><span>##Activated CD8 T cell     0.0020531477 0.1841364 2.064353e-06 8.878395e-03</span><br><span>##Activated dendritic cell 0.0003693238 0.2607891 4.236035e-01 1.277514e-01</span><br><br><span>####################画图</span><br><span># 原图是行名在右边的，而pheatmap默认行名在右边且无法修改。网上有大佬把pheatmap函数修改了一下，让它无缝跑到左边去。</span><br><span># https://stackoverflow.com/questions/57729914/how-can-you-show-the-rownames-in-pheatmap-on-the-left-side-of-the-graph。</span><br><span># 存在了modified_pheatmap.R脚本里。</span><br><span>library</span>(dplyr)<br>tmp = matrix(case_when(as.vector(p) &lt; <span>0.001</span>~<span>"****"</span>,<br>                       as.vector(p) &lt; <span>0.001</span>~<span>"***"</span>,<br>                       as.vector(p) &lt; <span>0.01</span>~<span>"**"</span>,<br>                       as.vector(p) &lt; <span>0.05</span>~<span>"*"</span>,<br>                       <span>T</span>~ <span>""</span>),nrow = nrow(p))<br><span>source</span>(<span>"./modified_pheatmap.R"</span>)<br>pheatmap(m,<br>         display_numbers = tmp,<br>         angle_col = <span>45</span>,<br>         color = colorRampPalette(c(<span>"#92b7d1"</span>, <span>"white"</span>, <span>"#d71e22"</span>))(<span>100</span>),<br>         border_color = <span>"white"</span>,<br>         treeheight_col = <span>0</span>,<br>         treeheight_row = <span>0</span>)  <br></code></pre><h2><span>2.GSVA计算KEGG通路得分</span></h2><section><span>使用相同转录组数据，评估</span><span>HIF1A</span><span>、</span><span>HAO1</span><span>与缺氧通路相关性：</span></section><pre><code><span>library</span>(GSVA)<br>set.seed(<span>123456</span>)<br><span>library</span>(clusterProfiler)<br>Hallmark_df &lt;- read.gmt(<span>"D:/data/MSigdb/human/msigdb_v7.5.1.Hs_GMTs/h.all.v7.5.1.symbols.gmt"</span>)<br>Hallmark_list &lt;- split(Hallmark_df$gene,Hallmark_df$term)<br>lapply(Hallmark_list[<span>1</span>:<span>3</span>], head)<br><span>#$HALLMARK_TNFA_SIGNALING_VIA_NFKB</span><br><span>#[1] "JUNB"    "CXCL2"   "ATF3"    "NFKBIA"  "TNFAIP3" "PTGS2"  </span><br><span>#$HALLMARK_HYPOXIA</span><br><span>#[1] "PGK1"  "PDK1"  "GBE1"  "PFKL"  "ALDOA" "ENO2" </span><br><span>#$HALLMARK_CHOLESTEROL_HOMEOSTASIS</span><br><span>#[1] "FDPS"    "CYP51A1" "IDI1"    "FDFT1"   "DHCR7"   "SQLE"   </span><br><br>gsvaPar &lt;- gsvaParam(exprData = as.matrix(uni_matrix), <br>                     geneSets = Hallmark_list,<br>                     kcdf = <span>"Gaussian"</span>)<br>Hallmark_GSVA &lt;- gsva(gsvaPar, verbose = <span>FALSE</span>)<br>dim(Hallmark_GSVA)<span>#[1] 100  30</span><br>Hallmark_GSVA[<span>1</span>:<span>4</span>,<span>1</span>:<span>4</span>]<br><span>#                                 160003803T-1 160003803T-2 160003803T-3 160003803T-4</span><br><span>#HALLMARK_TNFA_SIGNALING_VIA_NFKB   -0.3994131  -0.54548251  -0.48628751 -0.348786092</span><br><span>#HALLMARK_HYPOXIA                   -0.1446272  -0.32933707  -0.18655741 -0.005094561</span><br><span>#HALLMARK_CHOLESTEROL_HOMEOSTASIS   -0.1535734  -0.19870746  -0.05003305  0.071075502</span><br><span>#HALLMARK_MITOTIC_SPINDLE            0.1046816  -0.05921412   0.10908452 -0.125620296</span><br></code></pre><section><span>计算</span><span>HIF1A</span><span>与缺氧通路相关性</span></section><pre><code>identical(colnames(Hallmark_GSVA),colnames(uni_matrix))<br><br><span>library</span>(ggplot2)<br><span>library</span>(ggpubr)<br>data &lt;- data.frame(HIF1A = as.numeric(uni_matrix[<span>"HIF1A"</span>,]),<br>                   HAO1 = as.numeric(uni_matrix[<span>"HAO1"</span>,]),<br>                   HYPOXIA = as.numeric(Hallmark_GSVA[<span>"HALLMARK_HYPOXIA"</span>,]))<br>cor.test(data$HIF1A,data$HYPOXIA,method = <span>'spearman'</span>)<br>cor.test(data$HAO1,data$HYPOXIA,method = <span>'spearman'</span>)<br>gg &lt;- ggplot(data,aes(x = HIF1A, y = HYPOXIA)) + <br>      <span>#xlim(-20,15) + ylim(-15,10) +</span><br>      labs(x = <span>"Gene Expression"</span>, y = <span>"HallMark Hypoxia Score"</span>) +<br>      geom_point(colour = <span>"#368ad9"</span>,size = <span>4</span>) +<br>      geom_smooth(method =<span>'lm'</span>) +<br>      stat_cor(method = <span>"spearman"</span>,size = <span>8</span>) +<br>      theme_bw() + <br>      theme(axis.text.x = element_text(size = <span>18</span>), <br>            axis.text.y = element_text(size = <span>18</span>), <br>            axis.title.x = element_text(size = <span>20</span>), <br>            axis.title.y = element_text(size = <span>20</span>),<br>            legend.title = element_blank(),<br>            legend.text=element_text(size=<span>19</span>))<br>gg<br></code></pre><p><img data-imgfileid="100004943" data-ratio="0.875" data-w="1080" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3IhQ4at5dbGaI6tiaeS34X1p3pL7eriaDkwePiaGVjl3TkpMfYd8kRZa3PcUjLzboAIiat7C2JgAVNyVw/640?wx_fmt=png" src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3IhQ4at5dbGaI6tiaeS34X1p3pL7eriaDkwePiaGVjl3TkpMfYd8kRZa3PcUjLzboAIiat7C2JgAVNyVw/640?wx_fmt=png"></p><h3><span>GSVA结果差异分析</span></h3><p><span>使用</span><span>limma</span><span>执行差异分析，这里的分组随机生成</span></p><pre><code>Group &lt;- ifelse(pheno$num2 %<span>in</span>% c(<span>"P002"</span>,<span>"P023"</span>,<span>"P024"</span>),<span>"Cold"</span>,<span>"Hot"</span>)<br>design &lt;- model.matrix(~Group)<br><span>library</span>(limma)<br>fit &lt;- lmFit(Hallmark_GSVA, design)<br>fit &lt;- eBayes(fit)<br>DEG &lt;- topTable(fit, coef = <span>2</span>, number = <span>Inf</span>)<br>head(DEG)<br><span>#                                               logFC     AveExpr        t      P.Value    adj.P.Val       B</span><br><span>#HALLMARK_KRAS_SIGNALING_UP                 0.3562078 -0.03475377 5.828477 4.020872e-07 2.010436e-05 6.24561</span><br><span>#HALLMARK_COMPLEMENT                        0.3828715 -0.02954553 5.496154 1.308133e-06 3.270334e-05 5.10157</span><br><span>#HALLMARK_INTERFERON_GAMMA_RESPONSE         0.4950833 -0.01321527 5.281168 2.786445e-06 3.410775e-05 4.36965</span><br><span>#HALLMARK_INFLAMMATORY_RESPONSE             0.3852465 -0.03193564 5.248098 3.128339e-06 3.410775e-05 4.25774</span><br><span>#HALLMARK_ALLOGRAFT_REJECTION               0.4472482 -0.01588245 5.212951 3.537149e-06 3.410775e-05 4.13902</span><br><span>#HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION 0.4670243 -0.03387425 5.171115 4.092930e-06 3.410775e-05 3.99799</span><br>DEG &lt;- DEG[abs(DEG$logFC) &gt; <span>0.5</span> &amp; DEG$adj.P.Val &lt; <span>0.05</span>,]<span></span></code></pre><section><img data-imgfileid="100003752" data-ratio="0.05278592375366569" data-type="png" data-w="341" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/1LTeQhNfr8sUH75oYsoDaqjPCTiaukEmS8tWricW7LnLKKfIE9jKBexibqamsrlibaaXmuc2nicaYibfDFBNCmqX5mBw/640?wx_fmt=other&amp;wxfrom=5&amp;wx_lazy=1&amp;wx_co=1&amp;tp=webp" src="https://mmbiz.qpic.cn/sz_mmbiz_png/1LTeQhNfr8sUH75oYsoDaqjPCTiaukEmS8tWricW7LnLKKfIE9jKBexibqamsrlibaaXmuc2nicaYibfDFBNCmqX5mBw/640?wx_fmt=other&amp;wxfrom=5&amp;wx_lazy=1&amp;wx_co=1&amp;tp=webp"></section><section><span></span></section><section data-tools="135编辑器" data-id="116886" draggable="true"><section><section><section><section><strong data-brushtype="text">被炸熟的虾</strong></section></section></section><section><section><section data-width="35%"><section><img data-cropselx1="0" data-cropselx2="115" data-cropsely1="0" data-cropsely2="115" data-imgfileid="100003750" data-ratio="1" data-type="jpeg" data-w="258" data-width="100%" data-src="https://mmbiz.qpic.cn/sz_mmbiz_jpg/dRYYdqiaan3JjDfj2H8p6gg3CB25AGthbwzrotao4ev5tIe0utthbZRK8yOoDOuTzOSoTSnPWn61IdDCnXsnaiag/640?wx_fmt=other&amp;wxfrom=5&amp;wx_lazy=1&amp;wx_co=1&amp;tp=webp" src="https://mmbiz.qpic.cn/sz_mmbiz_jpg/dRYYdqiaan3JjDfj2H8p6gg3CB25AGthbwzrotao4ev5tIe0utthbZRK8yOoDOuTzOSoTSnPWn61IdDCnXsnaiag/640?wx_fmt=other&amp;wxfrom=5&amp;wx_lazy=1&amp;wx_co=1&amp;tp=webp"></section></section><section data-width="63%"><section><section><strong>自己的摸索，发现问题麻烦告诉作者，光速回来改正</strong><strong><img data-imgfileid="100003751" data-ratio="1" data-type="png" data-w="64" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3KLQGNxibvI4SdtUfxjINkz2DO8cGEPTgbcMJ357RvXJ7IvWU2p7anljpePpakI9oKu3icJxobBDp5w/640?wx_fmt=other&amp;wxfrom=5&amp;wx_lazy=1&amp;wx_co=1&amp;tp=webp" src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3KLQGNxibvI4SdtUfxjINkz2DO8cGEPTgbcMJ357RvXJ7IvWU2p7anljpePpakI9oKu3icJxobBDp5w/640?wx_fmt=other&amp;wxfrom=5&amp;wx_lazy=1&amp;wx_co=1&amp;tp=webp"></strong></section></section></section></section></section></section></section><p><mp-style-type data-value="10000"></mp-style-type></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/w4xosiGw__9HyHpjTqk41w",target="_blank" rel="noopener noreferrer">原文链接</a>
