---
title: "bulk RNA-seq | ä¸‹æ¸¸åˆ†æ | GSVAä¸ssGSEA"
date: 2024-12-03T00:52:38Z
draft: ["false"]
tags: [
  "fetched",
  "è¢«ç‚¸ç†Ÿçš„è™¾"
]
categories: ["Acdemic"]
---
bulk RNA-seq | ä¸‹æ¸¸åˆ†æ | GSVAä¸ssGSEA by è¢«ç‚¸ç†Ÿçš„è™¾
------
<div><p><span>bulk RNA-seq</span><span>ä¸‹æ¸¸åˆ†æéƒ¨åˆ†ï¼Œæˆ‘ä»¬å·²ç»å…ˆåè®°å½•äº†</span></p><ul><li><section><a href="https://mp.weixin.qq.com/s?__biz=MzkwNDQwMDI5NQ==&amp;mid=2247488268&amp;idx=1&amp;sn=8b6acae06a129b028d95648193a44de0&amp;scene=21#wechat_redirect" data-linktype="2"><span><strong><span>bulk RNA-seq | ä¸‹æ¸¸åˆ†æ | åŠŸèƒ½å¯Œé›†åˆ†æ ORA - clusterProfiler</span></strong></span></a></section></li><li><section><a href="https://mp.weixin.qq.com/s?__biz=MzkwNDQwMDI5NQ==&amp;mid=2247488268&amp;idx=1&amp;sn=8b6acae06a129b028d95648193a44de0&amp;scene=21#wechat_redirect" data-linktype="2"><span><strong><span>åŠŸèƒ½å¯Œé›†åˆ†æ ORA | clusterProfiler è¯»å–æœ¬åœ°åŒ–KEGGæ•°æ®</span></strong></span></a></section></li><li><section><a href="https://mp.weixin.qq.com/s?__biz=MzkwNDQwMDI5NQ==&amp;mid=2247488571&amp;idx=1&amp;sn=fdd0eb2b878da4b21bac1fd009ce6e67&amp;chksm=c086c8f3f7f141e55dd6a9389479c909fbd5559403d8106e510bc7c678e4e55fdde8949aad10&amp;scene=21&amp;cur_album_id=2550626183059685377#wechat_redirect" data-linktype="2"><span><strong><span>bulk RNA-seq | ä¸‹æ¸¸åˆ†æ | åŸºå› é›†å¯Œé›†åˆ†æ GSEA- clusterProfiler</span></strong></span></a></section></li></ul><p><span>ä¸‹é¢æ¥ç€è®°å½•</span><span>GSVA</span><span>ä¸</span><span>ssGSEA</span><span>æ–¹æ³•ã€‚è¿™ä¸€éƒ¨åˆ†ä¹‹å‰ä¹Ÿå†™è¿‡ï¼š</span><a href="https://mp.weixin.qq.com/s?__biz=MzkwNDQwMDI5NQ==&amp;mid=2247484182&amp;idx=1&amp;sn=fe3f83e5e67677413fb749bcfa6cc132&amp;payreadticket=HOim2eFVyWiPlEQbB5GBnb5JWm4ODyJxEQwVs-HENBdUNJ61GdWxT4cA8Pd1vV3I7zGYvMo&amp;scene=21#wechat_redirect" data-linktype="2"><span><strong><span>ssGSEAä¸GSVAä½¿ç”¨æ–¹æ³•</span></strong></span></a><span>ï¼Œä¸»ä½“å†…å®¹è¿˜æ˜¯é‚£äº›ï¼Œä¸€ä¸ªæ›´æ–°åŸå› æ˜¯ä¸çŸ¥é“ä¸ºä»€ä¹ˆï¼Œ</span><span>GSVA</span><span>åŒ…ï¼ˆæ”¯æŒ</span><span>GSVA</span><span>ã€</span><span>ssGSEA</span><span>ã€</span><span>zscore</span><span>ä¸</span><span>plage</span><span>ç­‰å¤šç§æ–¹æ³•ï¼‰çš„ä½œè€…æ›´æ–°äº†å‡½æ•°ï¼ï¼</span></p><h3><strong><span>1ã€ssGSEA</span></strong></h3><p><span>å•æ ·æœ¬åŸºå› é›†å¯Œé›†åˆ†æï¼ˆ</span><span>single sample gene set enrichment analysis</span><span>, </span><span>ssGSEA</span><span>ï¼‰ï¼ŒåŸç†ä¸Šä¸</span><span>GSEA</span><span>ç±»ä¼¼ï¼Œä¸åŒçš„æ˜¯</span><span>GSEA</span><span> ä¸»è¦ç”¨äºæ£€æµ‹ä¸åŒå®éªŒç»„ï¼ˆå¦‚å®éªŒç»„å’Œå¯¹ç…§ç»„ï¼‰ä¹‹é—´åŸºå› é›†å¯Œé›†å·®å¼‚ï¼Œè€Œ</span><span>ssGSEA</span><span>å°†æ ·æœ¬å†…åŸºå› è¡¨è¾¾è°±è¿›è¡Œå½’ä¸€åŒ–å¤„ç†ï¼Œç„¶åè®¡ç®—æ¯ä¸ªåŸºå› é›†å¯¹åº”çš„</span><span>ssGSEA</span><span>å¾—åˆ†ã€‚é€šè¿‡è¿™ç§æ–¹å¼ï¼Œ</span><span>ssGSEA</span><span>å°†å•ä¸ªæ ·æœ¬çš„åŸºå› è¡¨è¾¾è°±è½¬æ¢ä¸ºåŸºå› é›†å¯Œé›†å¾—åˆ†çŸ©é˜µã€‚å› æ­¤ï¼Œå¦‚æœä½¿ç”¨çš„æ˜¯å…ç–«ç»†èƒ</span><span>marker</span><span>ç»„æˆçš„åŸºå› é›†ï¼Œ</span><span>ssGESA</span><span>ä¹Ÿå¯ä»¥è®¡ç®—å…ç–«ç»†èƒæµ¸æ¶¦è¯„åˆ†ã€‚</span></p><h3><strong><span>2ã€GSVA</span></strong></h3><p><span>åŸºå› é›†å˜å¼‚åˆ†æï¼ˆ</span><span>Gene Set Variation Analysis</span><span>ï¼Œ</span><span>GSVA</span><span>ï¼‰å’Œ</span><span>ssGSEA</span><span>éå¸¸ç›¸ä¼¼ï¼ŒåŒæ ·æ˜¯å¯¹å•ä¸ªæ ·æœ¬è¿›è¡Œè®¡ç®—ï¼Œæœ€åè¾“å‡ºçš„ä¹Ÿæ˜¯ä¸€ä¸ªè¡Œä¸ºä¸åŒåŸºå› é›†ã€åˆ—ä¸ºæ ·æœ¬çš„çŸ©é˜µã€‚ä¸»è¦åŒºåˆ«åœ¨äº</span><span>GSVA</span><span>ä½¿ç”¨çš„æ˜¯éå‚æ•°æ ¸å¯†åº¦ä¼°è®¡æ–¹æ³•ï¼Œè€Œ</span><span>ssGSEA</span><span>æ˜¯åŸºäºæ ·æœ¬æ’åºçš„ç§¯åˆ†æ–¹æ³•ã€‚</span></p><h4>æ›´æ–°å‰</h4><p><span>è¾“å…¥è¡¨è¾¾çŸ©é˜µä»¥åŠåŸºå› é›†ï¼Œç„¶åé€‰æ‹©æ–¹æ³•å°±å¯ä»¥äº†ã€‚</span></p><pre><code>packageVersion(<span>"GSVA"</span>)<span>#[1]Â â€˜1.42.0â€™</span><br></code></pre><p><img data-imgfileid="100004947" data-ratio="0.5341726618705036" data-s="300,640" data-type="png" data-w="556" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3Jy9UPcZiaia8b6vEKCDWJ1XVibssE1azPY6crODlg5uWpdePVHGibPS82kKmI5Hq9CSDo4uPDbOCcKqQ/640?wx_fmt=png&amp;from=appmsg" src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3Jy9UPcZiaia8b6vEKCDWJ1XVibssE1azPY6crODlg5uWpdePVHGibPS82kKmI5Hq9CSDo4uPDbOCcKqQ/640?wx_fmt=png&amp;from=appmsg"></p><ul><li><section><span>expr</span><span>ï¼šåŸºå› çš„è¡¨è¾¾é‡éœ€è¦çš„æ˜¯çŸ©é˜µæ ¼å¼ï¼ˆ</span><span>matrix</span><span>ï¼Œä¸æ˜¯</span><span>data.frame</span><span>ï¼‰ï¼Œè¡Œä¸ºåŸºå› ï¼Œåˆ—ä¸ºæ ·æœ¬ï¼›</span></section></li><li><section><span>gset.idx.list</span><span>ï¼šåŸºå› é›†éœ€è¦çš„æ˜¯</span><span>list</span><span>æ ¼å¼ï¼›</span></section></li><li><section><span>kcdf</span><span>ï¼šé»˜è®¤æƒ…å†µä¸‹ï¼Œå‚æ•°</span><span>kcdf="Gaussian"</span><span>ï¼Œé€‚ç”¨äºå¯¹æ•°è½¬æ¢çš„</span><span>microarray</span><span>ã€</span><span>RNA-seq</span><span>çš„</span><span>log-CPMs</span><span>ã€</span><span>log-RPKMs</span><span>æˆ–</span><span>log-TPMs</span><span>ã€‚å½“è¾“å…¥çš„è¡¨è¾¾çš„çŸ©é˜µæ˜¯</span><span>RNA-seq</span><span>çš„</span><span>raw Count</span><span>æ—¶ï¼Œè¿™ä¸ªå‚æ•°åº”è¯¥è®¾ç½®ä¸º</span><span>kcdf=â€Poissonâ€</span><span>ã€‚</span></section></li></ul><pre><code>kcdfÂ =Â c(<span>"Gaussian"</span>,<span>"Poisson"</span>,<span>"none"</span>)<br></code></pre><ul><li><section><span>method</span><span>ï¼šå‚æ•°</span><span>method</span><span>ç”¨äºæŒ‡å®šä¼°è®¡æ¯ä¸ªæ ·æœ¬çš„åŸºå› é›†å¯Œé›†åˆ†æ•°çš„æ–¹æ³•ï¼Œé»˜è®¤æƒ…å†µä¸‹ï¼Œ</span><span>method=â€gsvaâ€</span><span>ã€‚</span><span><strong><span></span></strong></span></section></li></ul><pre><code>methodÂ =Â c(<span>"gsva"</span>,Â <span>"ssgsea"</span>,<span>"zscore"</span>,<span>"plage"</span>)<br></code></pre><h4><span>æ›´æ–°å</span></h4><pre><code>packageVersion(<span>"GSVA"</span>)<br><span>#[1]Â â€˜2.0.1â€™</span><br></code></pre><p><span>ç¬¬ä¸€æ­¥ï¼šéœ€è¦å…ˆå®šä¹‰</span><span>param</span><span>ã€‚è¿™ä¸ª</span><span>param</span><span>åŒ…æ‹¬è¡¨è¾¾çŸ©é˜µä»¥åŠåŸºå› é›†ç­‰ã€‚ä¸åŒæ–¹æ³•ä½¿ç”¨ä¸åŒ</span><span>param</span><span>å®šä¹‰å‡½æ•°ï¼ï¼</span></p><section><span><img data-imgfileid="100004948" data-ratio="0.3990740740740741" data-type="png" data-w="1080" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3Jy9UPcZiaia8b6vEKCDWJ1XVusHKpUQAMAHPaj5VU9GNUSpJLDjwEd3pT7rJCH8HoyTO6kCeWCghFA/640?wx_fmt=png&amp;from=appmsg" src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3Jy9UPcZiaia8b6vEKCDWJ1XVusHKpUQAMAHPaj5VU9GNUSpJLDjwEd3pT7rJCH8HoyTO6kCeWCghFA/640?wx_fmt=png&amp;from=appmsg"></span></section><p><span>å‚æ•°</span><span>kcdf</span><span>ä¹Ÿè¢«çº³å…¥äº†è¿™ä¸€æ­¥ã€‚</span></p><p><img data-imgfileid="100004949" data-ratio="0.5275288092189501" data-s="300,640" data-type="png" data-w="781" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3Jy9UPcZiaia8b6vEKCDWJ1XVtz4Fm7hGBNcqY4Jniav1SSBDMsMXZMbhzESnSkBb7sLfHXLSrfSIpzw/640?wx_fmt=png&amp;from=appmsg" src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3Jy9UPcZiaia8b6vEKCDWJ1XVtz4Fm7hGBNcqY4Jniav1SSBDMsMXZMbhzESnSkBb7sLfHXLSrfSIpzw/640?wx_fmt=png&amp;from=appmsg"></p><h4><section>ç¤ºä¾‹æµç¨‹</section></h4><p><span>é¦–å…ˆä¾æ—§æ˜¯æ•°æ®å‡†å¤‡ï¼š</span></p><pre><code><span>library</span>(GSVA)<br>set.seed(<span>12345</span>)<br>pÂ &lt;-Â <span>10000</span>Â <span>##Â numberÂ ofÂ genes</span><br>nÂ &lt;-Â <span>30</span>Â Â Â Â <span>##Â numberÂ ofÂ samples</span><br><span>##Â simulateÂ expressionÂ valuesÂ fromÂ aÂ standardÂ GaussianÂ distribution</span><br>XÂ &lt;-Â matrix(rnorm(p*n),Â nrowÂ =Â p,<br>Â Â Â Â Â Â Â Â Â Â Â Â dimnamesÂ =Â list(paste0(<span>"g"</span>,Â <span>1</span>:p),Â paste0(<span>"s"</span>,Â <span>1</span>:n)))<br>X[<span>1</span>:<span>5</span>,Â <span>1</span>:<span>5</span>]<br><span>##Â sampleÂ geneÂ setÂ sizes</span><br>gsÂ &lt;-Â as.list(sample(<span>10</span>:<span>100</span>,Â sizeÂ =Â <span>100</span>,Â replaceÂ =Â <span>TRUE</span>))<br><span>##Â sampleÂ geneÂ sets</span><br>gsÂ &lt;-Â lapply(gs,Â <span>function</span>(n,Â p)<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â paste0(<span>"g"</span>,Â sample(<span>1</span>:p,Â sizeÂ =Â n,Â replaceÂ =Â <span>FALSE</span>)),Â p)<br>names(gs)Â &lt;-Â paste0(<span>"gs"</span>,Â <span>1</span>:length(gs))<br>lapply(gs[<span>1</span>:<span>3</span>],head)<br><span>#$gs1</span><br><span>#[1]Â "g7824"Â "g7976"Â "g7005"Â "g5451"Â "g1199"Â "g227"Â </span><br><span>#$gs2</span><br><span>#[1]Â "g5066"Â "g6691"Â "g542"Â Â "g7062"Â "g2891"Â "g8778"</span><br><span>#$gs3</span><br><span>#[1]Â "g6546"Â "g1957"Â "g2913"Â "g8503"Â "g5013"Â "g8340"</span><br></code></pre><p><span>æ›´æ–°å‰çš„åˆ†ææ–¹æ³•åªè¦ä¸€æ­¥ï¼š</span></p><pre><code><span>##WeÂ canÂ calculateÂ GSVAÂ enrichmentÂ scoresÂ asÂ follows:</span><br><span>library</span>(GSVA)<br>gsva.esÂ &lt;-Â gsva(exprÂ =Â X,Â <br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â gset.idx.listÂ =Â gs,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â methodÂ =Â <span>'gsva'</span>,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â kcdfÂ =Â <span>'Gaussian'</span>,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â verboseÂ =Â <span>FALSE</span>)<br>dim(gsva.es)<span>#[1]Â 100Â Â 30</span><br>gsva.es[<span>1</span>:<span>5</span>,Â <span>1</span>:<span>5</span>]<br><span>#Â Â Â Â Â Â Â Â Â Â Â Â Â Â s1Â Â Â Â Â Â Â Â Â Â s2Â Â Â Â Â Â Â Â Â Â s3Â Â Â Â Â Â Â Â Â Â Â s4Â Â Â Â Â Â Â Â Â Â Â s5</span><br><span>#gs1Â -0.163169867Â -0.18230994Â Â 0.36746012Â -0.010434415Â -0.026349618</span><br><span>#gs2Â Â 0.007954471Â -0.05686215Â -0.20594532Â -0.009643796Â -0.057696047</span><br><span>#gs3Â -0.161205687Â Â 0.24624689Â Â 0.09306792Â -0.033363937Â -0.002946779</span><br><span>#gs4Â Â 0.126560368Â -0.08559508Â -0.26097093Â -0.181676230Â -0.307544679</span><br><span>#gs5Â Â 0.076199136Â -0.01887931Â -0.02707651Â -0.069945789Â -0.086908249</span><br></code></pre><p><span>æ›´æ–°åå¤šäº†ä¸€æ­¥ï¼š</span></p><pre><code><span>##FirstÂ weÂ shouldÂ buildÂ aÂ parameterÂ objectÂ forÂ theÂ desiredÂ methodology.R</span><br>gsvaParÂ &lt;-Â gsvaParam(exprDataÂ =Â X,Â <br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â geneSetsÂ =Â gs,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â kcdfÂ =Â <span>"Gaussian"</span>)<br>gsvaPar<br><span>##Second,Â weÂ callÂ theÂ gsva()Â functionÂ withÂ theÂ parameterÂ objectÂ asÂ firstÂ argument.Â </span><br>gsva.esÂ &lt;-Â gsva(gsvaPar,Â verboseÂ =Â <span>FALSE</span>)<br>dim(gsva.es)<span>#[1]Â 100Â Â 30</span><br>gsva.es[<span>1</span>:<span>5</span>,Â <span>1</span>:<span>5</span>]<br><span>#Â Â Â Â Â Â Â Â Â Â Â Â Â Â s1Â Â Â Â Â Â Â Â Â Â s2Â Â Â Â Â Â Â Â Â Â s3Â Â Â Â Â Â Â Â Â Â Â s4Â Â Â Â Â Â Â Â Â Â Â s5</span><br><span>#gs1Â -0.163169867Â -0.18230994Â Â 0.36746012Â -0.010434415Â -0.026349618</span><br><span>#gs2Â Â 0.007954471Â -0.05686215Â -0.20594532Â -0.009643796Â -0.057696047</span><br><span>#gs3Â -0.161205687Â Â 0.24624689Â Â 0.09306792Â -0.033363937Â -0.002946779</span><br><span>#gs4Â Â 0.126560368Â -0.08559508Â -0.26097093Â -0.181676230Â -0.307544679</span><br><span>#gs5Â Â 0.076199136Â -0.01887931Â -0.02707651Â -0.069945789Â -0.086908249</span><br></code></pre><p><span>æ— æ³•ç†è§£æ›´æ–°ç›®çš„ğŸ¤”ã€‚æ¥ä¸‹æ¥ï¼Œå†ä½¿ç”¨æ–°ç‰ˆæœ¬æ˜¯</span><span>GSVA</span><span>åŒ…è·‘ä¸€éä¹‹å‰çš„ç¬”è®°æµç¨‹ï¼š</span></p><h2><span>1.ssGSEAæ–¹æ³•è¯„ä¼°æ ·æœ¬å…ç–«æµ¸æ¶¦æ°´å¹³</span></h2><p><span>é‡ç°ç»“æœæ¥è‡ªï¼š</span><span>Fig3 A-B</span><span>ï¼Œè¯„ä¼°è‚¿ç˜¤æ ·æœ¬çš„å…ç–«ç»†èƒæµ¸æ¶¦æ°´å¹³ï¼š</span></p><blockquote><p>Local mutational diversity drives intratumoral immune heterogeneity in non-small cell lung cancer. Nat Commun. 2018 Dec 18;9(1):5361. doi: 10.1038/s41467-018-07767-w. PMID: 30560866.</p></blockquote><p><span>å…ç–«åŸºå› é›†å‡ºè‡ªæ–‡ç« </span><span>PMID: 28052254</span><span>,åŒ…æ‹¬</span><span>28</span><span>ç§å…ç–«ç»†èƒçš„åŸºå› é›†ï¼š</span></p><blockquote><p>Pan-cancer Immunogenomic Analyses Reveal Genotype-Immunophenotype Relationships and Predictors of Response to Checkpoint Blockade. Cell Rep. 2017 Jan 3;18(1):248-262. doi: 10.1016/j.celrep.2016.12.019. PMID: 28052254.</p></blockquote><h3><span>æ•°æ®å‡†å¤‡</span></h3><pre><code>rm(listÂ =Â ls())<br>options(stringsAsFactorsÂ =Â <span>F</span>)<br>setwd(<span>"D:/bioinformatics/9.Others/7Â ImmuneÂ cellÂ infiltration/data"</span>)<br>set.seed(<span>123456</span>)<br><span>###################################è¡¨è¾¾è°±æå–####################################</span><br>raw_dataÂ &lt;-Â read.table(<span>'GSE112996_merged_fpkm_table.txt.gz'</span>,headerÂ =Â <span>T</span>,row.namesÂ =Â <span>1</span>,check.namesÂ =Â <span>F</span>)<br><span>#rowSums(raw_data[,-1])</span><br>raw_dataÂ &lt;-Â raw_data[rowSums(raw_data[,-<span>1</span>])!=<span>0</span>,]<br><br><span>#####################################IDè½¬æ¢######################################</span><br>idsÂ &lt;-Â data.frame(GeneSymbolÂ =Â raw_data$GeneName,EnsemblÂ =Â rownames(raw_data))<br>ids$medianÂ &lt;-Â apply(raw_data[,-<span>1</span>],<span>1</span>,median)<br>idsÂ &lt;-Â ids[order(ids$GeneSymbol,ids$median,decreasingÂ =Â <span>T</span>),]<br>idsÂ &lt;-Â ids[!duplicated(ids$GeneSymbol),]<br>rownames(ids)Â &lt;-Â ids$GeneSymbol<br>clean_dataÂ &lt;-Â raw_data[ids$Ensembl,-<span>1</span>]<br>rownames(clean_data)Â &lt;-Â ids$GeneSymbol<br><br>uni_matrixÂ &lt;-Â log2(clean_data+<span>1</span>)<br>boxplot(uni_matrix)<br><br><span>###################################è¡¨å‹ä¿¡æ¯æå–###################################</span><br>phenoÂ &lt;-Â read.csv(fileÂ =Â <span>'GSE112996_series_matrix.txt.gz'</span>)<br>phenoÂ &lt;-Â data.frame(num1Â =Â strsplit(as.character(pheno[<span>42</span>,]),split=<span>'\t'</span>)[[<span>1</span>]][-<span>1</span>],<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â num2Â =Â gsub(<span>'patient:Â No.'</span>,<span>'P'</span>,strsplit(as.character(pheno[<span>51</span>,]),split=<span>'\t'</span>)[[<span>1</span>]][-<span>1</span>]))<br>phenoÂ &lt;-Â pheno[order(pheno$num2),]<br>identical(colnames(uni_matrix),pheno$num1)<br>table(colnames(uni_matrix)Â %<span>in</span>%Â pheno$num1)<br>uni_matrixÂ &lt;-Â uni_matrix[,pheno$num1]<br><br><span>###################################å…ç–«åŸºå› é›†###################################</span><br>genesetÂ &lt;-Â rio::import(<span>"ssGSEA-PMID28052254-Signature-28.xlsx"</span>,skipÂ =Â <span>2</span>)<br>genesetÂ &lt;-Â split(geneset$Metagene,geneset$`CellÂ type`)<br>lapply(geneset[<span>1</span>:<span>3</span>],Â head)<br><span>##Â $`ActivatedÂ BÂ cell`</span><br><span>##Â [1]Â "ADAM28"Â "CD180"Â Â "CD79B"Â Â "BLK"Â Â Â Â "CD19"Â Â Â "MS4A1"Â </span><br><span>##Â $`ActivatedÂ CD4Â TÂ cell`</span><br><span>##Â [1]Â "AIM2"Â Â "BIRC3"Â "BRIP1"Â "CCL20"Â "CCL4"Â Â "CCL5"Â </span><br><span>##Â $`ActivatedÂ CD8Â TÂ cell`</span><br><span>##Â [1]Â "ADRM1"Â Â Â Â Â "AHSA1"Â Â Â Â Â "C1GALT1C1"Â "CCT6B"Â Â Â Â Â "CD37"Â Â Â Â Â Â "CD3D"</span><br>names(geneset)<br></code></pre><h3><span>è¿è¡ŒssGSEA</span></h3><p><span>è¿è¡Œå¾ˆç®€å•ï¼Œæ›´æ–°åå¤šäº†ä¸€æ­¥ï¼š</span></p><pre><code><span>library</span>(GSVA)<br><span>##FirstÂ weÂ shouldÂ buildÂ aÂ parameterÂ objectÂ forÂ theÂ desiredÂ methodology.R</span><br>gsvaParÂ &lt;-Â ssgseaParam(exprDataÂ =Â as.matrix(uni_matrix),Â <br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â geneSetsÂ =Â geneset,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â normalizeÂ =Â <span>TRUE</span>)<br><span>##Second,Â weÂ callÂ theÂ gsva()Â functionÂ withÂ theÂ parameterÂ objectÂ asÂ firstÂ argument.Â </span><br>ssGSEA_matrixÂ &lt;-Â gsva(gsvaPar,Â verboseÂ =Â <span>FALSE</span>)<br>dim(ssGSEA_matrix)<span>#[1]Â 100Â Â 30</span><br>ssGSEA_matrix[<span>1</span>:<span>5</span>,Â <span>1</span>:<span>5</span>]<br><span>###normalizeå‚æ•°ç”¨æœ€å¤§å€¼ä¸æœ€å°å€¼é—´çš„ç»å¯¹å·®å¯¹ssGSEAåˆ†æ•°è¿›è¡Œæ ‡å‡†åŒ–,é»˜è®¤ä¸ºTRUE</span><br></code></pre><h3><section>ç»“æœå¯è§†åŒ–</section></h3><section><strong>1ã€æ ·æœ¬å…ç–«ç»†èƒæµ¸æ¶¦æ°´å¹³çƒ­å›¾ï¼š</strong></section><pre><code>scale_gsva_matrixÂ &lt;-Â t(scale(t(ssGSEA_matrix)))<br>scale_gsva_matrix[scale_gsva_matrix&lt;Â -<span>2</span>]Â &lt;-Â -<span>2</span><br>scale_gsva_matrix[scale_gsva_matrix&gt;<span>2</span>]Â &lt;-Â <span>2</span><br><span>##é‡æ–°æ’åº</span><br>anti_tumorÂ &lt;-Â c(<span>'ActivatedÂ CD4Â TÂ cell'</span>,Â <span>'ActivatedÂ CD8Â TÂ cell'</span>,Â <span>'CentralÂ memoryÂ CD4Â TÂ cell'</span>,Â <br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â <span>'CentralÂ memoryÂ CD8Â TÂ cell'</span>,Â <span>'EffectorÂ memeoryÂ CD4Â TÂ cell'</span>,Â <span>'EffectorÂ memeoryÂ CD8Â TÂ cell'</span>,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â <span>'TypeÂ 1Â TÂ helperÂ cell'</span>,Â <span>'TypeÂ 17Â TÂ helperÂ cell'</span>,Â <span>'ActivatedÂ dendriticÂ cell'</span>,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â <span>'CD56brightÂ naturalÂ killerÂ cell'</span>,Â <span>'NaturalÂ killerÂ cell'</span>,Â <span>'NaturalÂ killerÂ TÂ cell'</span>)<br>pro_tumorÂ &lt;-Â c(<span>'RegulatoryÂ TÂ cell'</span>,Â <span>'TypeÂ 2Â TÂ helperÂ cell'</span>,Â <span>'CD56dimÂ naturalÂ killerÂ cell'</span>,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â <span>'ImmatureÂ dendriticÂ cell'</span>,Â <span>'Macrophage'</span>,Â <span>'MDSC'</span>,Â <span>'Neutrophil'</span>,Â <span>'PlasmacytoidÂ dendriticÂ cell'</span>)<br><br>antiÂ &lt;-Â rownames(scale_gsva_matrix)Â %<span>in</span>%Â anti_tumor<br>proÂ &lt;-Â rownames(scale_gsva_matrix)Â %<span>in</span>%Â pro_tumor<br>nonÂ &lt;-Â !(anti|pro)<br>scale_gsva_matrixÂ &lt;-Â rbind(scale_gsva_matrix[anti,],scale_gsva_matrix[pro,],scale_gsva_matrix[non,])<br><span>###æ•°å€¼å½’ä¸€åŒ–åˆ°[0ï¼Œ1]</span><br>normalizationÂ &lt;-Â <span>function</span>(x){<span>return</span>((x-min(x))/(max(x)-min(x)))}Â Â <br>nor_gsva_matrixÂ &lt;-Â normalization(scale_gsva_matrix)<br><br><span>library</span>(pheatmap)<br>annotation_colÂ =Â data.frame(patientÂ =Â pheno$num2)<br>rownames(annotation_col)Â &lt;-Â colnames(uni_matrix)<br>HeatmapÂ &lt;-Â pheatmap(nor_gsva_matrix,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â show_colnamesÂ =Â <span>F</span>,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â cluster_rowsÂ =Â <span>F</span>,cluster_colsÂ =Â <span>F</span>,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â annotation_colÂ =Â annotation_col,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â cellwidthÂ =Â <span>5</span>,cellheightÂ =Â <span>5</span>,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â fontsizeÂ =Â <span>5</span>,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â gaps_rowÂ =Â c(<span>12</span>,<span>20</span>))<br><span>#save(gsva_matrix,gsva_matrix1,pheno,fileÂ =Â 'score.Rdata')</span><br></code></pre><figure><img data-imgfileid="100004942" data-ratio="0.47706422018348627" data-w="981" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3IhQ4at5dbGaI6tiaeS34X1p2HhtCvE02DLffR8mSVI9tS9qPtDjztON2TjW7huRkbgTnf66ulh06g/640?wx_fmt=png" src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3IhQ4at5dbGaI6tiaeS34X1p2HhtCvE02DLffR8mSVI9tS9qPtDjztON2TjW7huRkbgTnf66ulh06g/640?wx_fmt=png"><figcaption>img</figcaption></figure><section><strong>2ã€Anti-tumor immunity vs. Pro-tumor suppression æ•£ç‚¹å›¾ï¼š</strong></section><section><span>è®¡ç®—ä¸åŒè§’è‰²å…ç–«ç»†èƒ</span><span>score</span><span>åŠ å’Œåï¼Œä½¿ç”¨</span><span>ggplot2</span><span>ç»˜åˆ¶</span></section><pre><code>antiÂ &lt;-Â as.data.frame(scale_gsva_matrix[rownames(scale_gsva_matrix)Â %<span>in</span>%Â anti_tumor,])<br>proÂ &lt;-Â as.data.frame(scale_gsva_matrix[rownames(scale_gsva_matrix)Â %<span>in</span>%Â pro_tumor,])<br>anti_nÂ &lt;-Â apply(anti,<span>2</span>,sum)<br>pro_nÂ &lt;-Â apply(pro,<span>2</span>,sum)<br>patientÂ &lt;-Â pheno$num2[match(colnames(scale_gsva_matrix),pheno$num1)]<br><span>library</span>(ggplot2)<br><span>library</span>(ggpubr)<br>dataÂ &lt;-Â data.frame(antiÂ =Â anti_n,proÂ =Â pro_n,patient=patient)<br>anti_pro&lt;-Â cor.test(anti_n,pro_n,methodÂ =Â <span>'pearson'</span>)<br>ggÂ &lt;-Â ggplot(data,aes(xÂ =Â anti,Â yÂ =Â pro))Â +Â <br>Â Â Â Â Â Â xlim(-<span>20</span>,<span>15</span>)Â +Â ylim(-<span>15</span>,<span>10</span>)Â +<br>Â Â Â Â Â Â labs(xÂ =Â <span>"Anti-tumorÂ immunity"</span>,Â yÂ =Â <span>"Pro-tumorÂ suppression"</span>)Â +<br>Â Â Â Â Â Â geom_point(aes(colorÂ =Â patient),sizeÂ =Â <span>3</span>)Â +<br>Â Â Â Â Â Â geom_smooth(methodÂ =<span>'lm'</span>)Â +<br>Â Â Â Â Â Â stat_cor(methodÂ =Â <span>"spearman"</span>,sizeÂ =Â <span>8</span>)Â +<br>Â Â Â Â Â Â scale_colour_manual(valuesÂ =Â RColorBrewer::brewer.pal(<span>12</span>,Â <span>"Set3"</span>))Â +<br>Â Â Â Â Â Â theme_bw()Â +Â <br>Â Â Â Â Â Â theme(axis.text.xÂ =Â element_text(sizeÂ =Â <span>18</span>),Â <br>Â Â Â Â Â Â Â Â Â Â Â Â axis.text.yÂ =Â element_text(sizeÂ =Â <span>18</span>),Â <br>Â Â Â Â Â Â Â Â Â Â Â Â axis.title.xÂ =Â element_text(sizeÂ =Â <span>20</span>),Â <br>Â Â Â Â Â Â Â Â Â Â Â Â axis.title.yÂ =Â element_text(sizeÂ =Â <span>20</span>),<br>Â Â Â Â Â Â Â Â Â Â Â Â legend.titleÂ =Â element_blank(),<br>Â Â Â Â Â Â Â Â Â Â Â Â legend.textÂ =Â element_text(sizeÂ =Â <span>20</span>))<br>gg<br></code></pre><p><img data-imgfileid="100004944" data-ratio="0.875" data-w="1080" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3IhQ4at5dbGaI6tiaeS34X1p2sDozGvdL0atx5RWreSS8BFVH8iaEFibibdXXXnjlwuzKrElWKsd3yNwQ/640?wx_fmt=png" src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3IhQ4at5dbGaI6tiaeS34X1p2sDozGvdL0atx5RWreSS8BFVH8iaEFibibdXXXnjlwuzKrElWKsd3yNwQ/640?wx_fmt=png"></p><section><span>è®¡ç®—åŸºå› è¡¨è¾¾ä¸æµ¸æ¶¦å¾—åˆ†ç›¸å…³æ€§ï¼š</span></section><pre><code><span>library</span>(Hmisc)<br>identical(colnames(ssGSEA_matrix),colnames(uni_matrix))<br>gsÂ =Â geneset[[<span>1</span>]]<br>gsÂ &lt;-Â gs[gsÂ %<span>in</span>%Â rownames(uni_matrix)]<br>ncÂ =Â t(rbind(ssGSEA_matrix,uni_matrix[gs,]))<br>nc[<span>1</span>:<span>4</span>,<span>1</span>:<span>4</span>]<br>mÂ =Â rcorr(nc)$r[<span>1</span>:nrow(ssGSEA_matrix),(ncol(nc)-length(gs)+<span>1</span>):ncol(nc)]<br>m[<span>1</span>:<span>4</span>,<span>1</span>:<span>4</span>]<br><span>##Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â ADAM28Â Â Â Â Â CD180Â Â Â Â Â CD79BÂ Â Â Â Â Â Â BLK</span><br><span>##ActivatedÂ BÂ cellÂ Â Â Â Â Â Â Â Â 0.4613288Â 0.2455269Â 0.8656336Â 0.7610076</span><br><span>##ActivatedÂ CD4Â TÂ cellÂ Â Â Â Â 0.1121755Â 0.1026576Â 0.4376069Â 0.1188357</span><br><span>##ActivatedÂ CD8Â TÂ cellÂ Â Â Â Â 0.4475305Â 0.2016188Â 0.6413575Â 0.3857065</span><br><span>##ActivatedÂ dendriticÂ cellÂ 0.5078425Â 0.1712112Â 0.1222768Â 0.2304565</span><br>pÂ =Â rcorr(nc)$P[<span>1</span>:nrow(ssGSEA_matrix),(ncol(nc)-length(gs)+<span>1</span>):ncol(nc)]<br>p[<span>1</span>:<span>4</span>,<span>1</span>:<span>4</span>]<br><span>##Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â ADAM28Â Â Â Â Â CD180Â Â Â Â Â Â Â Â CD79BÂ Â Â Â Â Â Â Â Â Â BLK</span><br><span>##ActivatedÂ BÂ cellÂ Â Â Â Â Â Â Â Â 0.0014247371Â 0.1040157Â 1.687539e-14Â 1.299134e-09</span><br><span>##ActivatedÂ CD4Â TÂ cellÂ Â Â Â Â 0.4631674845Â 0.5021913Â 2.645818e-03Â 4.368624e-01</span><br><span>##ActivatedÂ CD8Â TÂ cellÂ Â Â Â Â 0.0020531477Â 0.1841364Â 2.064353e-06Â 8.878395e-03</span><br><span>##ActivatedÂ dendriticÂ cellÂ 0.0003693238Â 0.2607891Â 4.236035e-01Â 1.277514e-01</span><br><br><span>####################ç”»å›¾</span><br><span>#Â åŸå›¾æ˜¯è¡Œååœ¨å³è¾¹çš„ï¼Œè€Œpheatmapé»˜è®¤è¡Œååœ¨å³è¾¹ä¸”æ— æ³•ä¿®æ”¹ã€‚ç½‘ä¸Šæœ‰å¤§ä½¬æŠŠpheatmapå‡½æ•°ä¿®æ”¹äº†ä¸€ä¸‹ï¼Œè®©å®ƒæ— ç¼è·‘åˆ°å·¦è¾¹å»ã€‚</span><br><span>#Â https://stackoverflow.com/questions/57729914/how-can-you-show-the-rownames-in-pheatmap-on-the-left-side-of-the-graphã€‚</span><br><span>#Â å­˜åœ¨äº†modified_pheatmap.Rè„šæœ¬é‡Œã€‚</span><br><span>library</span>(dplyr)<br>tmpÂ =Â matrix(case_when(as.vector(p)Â &lt;Â <span>0.001</span>~<span>"****"</span>,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â as.vector(p)Â &lt;Â <span>0.001</span>~<span>"***"</span>,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â as.vector(p)Â &lt;Â <span>0.01</span>~<span>"**"</span>,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â as.vector(p)Â &lt;Â <span>0.05</span>~<span>"*"</span>,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â <span>T</span>~Â <span>""</span>),nrowÂ =Â nrow(p))<br><span>source</span>(<span>"./modified_pheatmap.R"</span>)<br>pheatmap(m,<br>Â Â Â Â Â Â Â Â Â display_numbersÂ =Â tmp,<br>Â Â Â Â Â Â Â Â Â angle_colÂ =Â <span>45</span>,<br>Â Â Â Â Â Â Â Â Â colorÂ =Â colorRampPalette(c(<span>"#92b7d1"</span>,Â <span>"white"</span>,Â <span>"#d71e22"</span>))(<span>100</span>),<br>Â Â Â Â Â Â Â Â Â border_colorÂ =Â <span>"white"</span>,<br>Â Â Â Â Â Â Â Â Â treeheight_colÂ =Â <span>0</span>,<br>Â Â Â Â Â Â Â Â Â treeheight_rowÂ =Â <span>0</span>)Â Â <br></code></pre><h2><span>2.GSVAè®¡ç®—KEGGé€šè·¯å¾—åˆ†</span></h2><section><span>ä½¿ç”¨ç›¸åŒè½¬å½•ç»„æ•°æ®ï¼Œè¯„ä¼°</span><span>HIF1A</span><span>ã€</span><span>HAO1</span><span>ä¸ç¼ºæ°§é€šè·¯ç›¸å…³æ€§ï¼š</span></section><pre><code><span>library</span>(GSVA)<br>set.seed(<span>123456</span>)<br><span>library</span>(clusterProfiler)<br>Hallmark_dfÂ &lt;-Â read.gmt(<span>"D:/data/MSigdb/human/msigdb_v7.5.1.Hs_GMTs/h.all.v7.5.1.symbols.gmt"</span>)<br>Hallmark_listÂ &lt;-Â split(Hallmark_df$gene,Hallmark_df$term)<br>lapply(Hallmark_list[<span>1</span>:<span>3</span>],Â head)<br><span>#$HALLMARK_TNFA_SIGNALING_VIA_NFKB</span><br><span>#[1]Â "JUNB"Â Â Â Â "CXCL2"Â Â Â "ATF3"Â Â Â Â "NFKBIA"Â Â "TNFAIP3"Â "PTGS2"Â Â </span><br><span>#$HALLMARK_HYPOXIA</span><br><span>#[1]Â "PGK1"Â Â "PDK1"Â Â "GBE1"Â Â "PFKL"Â Â "ALDOA"Â "ENO2"Â </span><br><span>#$HALLMARK_CHOLESTEROL_HOMEOSTASIS</span><br><span>#[1]Â "FDPS"Â Â Â Â "CYP51A1"Â "IDI1"Â Â Â Â "FDFT1"Â Â Â "DHCR7"Â Â Â "SQLE"Â Â Â </span><br><br>gsvaParÂ &lt;-Â gsvaParam(exprDataÂ =Â as.matrix(uni_matrix),Â <br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â geneSetsÂ =Â Hallmark_list,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â kcdfÂ =Â <span>"Gaussian"</span>)<br>Hallmark_GSVAÂ &lt;-Â gsva(gsvaPar,Â verboseÂ =Â <span>FALSE</span>)<br>dim(Hallmark_GSVA)<span>#[1]Â 100Â Â 30</span><br>Hallmark_GSVA[<span>1</span>:<span>4</span>,<span>1</span>:<span>4</span>]<br><span>#Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 160003803T-1Â 160003803T-2Â 160003803T-3Â 160003803T-4</span><br><span>#HALLMARK_TNFA_SIGNALING_VIA_NFKBÂ Â Â -0.3994131Â Â -0.54548251Â Â -0.48628751Â -0.348786092</span><br><span>#HALLMARK_HYPOXIAÂ Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â -0.1446272Â Â -0.32933707Â Â -0.18655741Â -0.005094561</span><br><span>#HALLMARK_CHOLESTEROL_HOMEOSTASISÂ Â Â -0.1535734Â Â -0.19870746Â Â -0.05003305Â Â 0.071075502</span><br><span>#HALLMARK_MITOTIC_SPINDLEÂ Â Â Â Â Â Â Â Â Â Â Â 0.1046816Â Â -0.05921412Â Â Â 0.10908452Â -0.125620296</span><br></code></pre><section><span>è®¡ç®—</span><span>HIF1A</span><span>ä¸ç¼ºæ°§é€šè·¯ç›¸å…³æ€§</span></section><pre><code>identical(colnames(Hallmark_GSVA),colnames(uni_matrix))<br><br><span>library</span>(ggplot2)<br><span>library</span>(ggpubr)<br>dataÂ &lt;-Â data.frame(HIF1AÂ =Â as.numeric(uni_matrix[<span>"HIF1A"</span>,]),<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â HAO1Â =Â as.numeric(uni_matrix[<span>"HAO1"</span>,]),<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â HYPOXIAÂ =Â as.numeric(Hallmark_GSVA[<span>"HALLMARK_HYPOXIA"</span>,]))<br>cor.test(data$HIF1A,data$HYPOXIA,methodÂ =Â <span>'spearman'</span>)<br>cor.test(data$HAO1,data$HYPOXIA,methodÂ =Â <span>'spearman'</span>)<br>ggÂ &lt;-Â ggplot(data,aes(xÂ =Â HIF1A,Â yÂ =Â HYPOXIA))Â +Â <br>Â Â Â Â Â Â <span>#xlim(-20,15)Â +Â ylim(-15,10)Â +</span><br>Â Â Â Â Â Â labs(xÂ =Â <span>"GeneÂ Expression"</span>,Â yÂ =Â <span>"HallMarkÂ HypoxiaÂ Score"</span>)Â +<br>Â Â Â Â Â Â geom_point(colourÂ =Â <span>"#368ad9"</span>,sizeÂ =Â <span>4</span>)Â +<br>Â Â Â Â Â Â geom_smooth(methodÂ =<span>'lm'</span>)Â +<br>Â Â Â Â Â Â stat_cor(methodÂ =Â <span>"spearman"</span>,sizeÂ =Â <span>8</span>)Â +<br>Â Â Â Â Â Â theme_bw()Â +Â <br>Â Â Â Â Â Â theme(axis.text.xÂ =Â element_text(sizeÂ =Â <span>18</span>),Â <br>Â Â Â Â Â Â Â Â Â Â Â Â axis.text.yÂ =Â element_text(sizeÂ =Â <span>18</span>),Â <br>Â Â Â Â Â Â Â Â Â Â Â Â axis.title.xÂ =Â element_text(sizeÂ =Â <span>20</span>),Â <br>Â Â Â Â Â Â Â Â Â Â Â Â axis.title.yÂ =Â element_text(sizeÂ =Â <span>20</span>),<br>Â Â Â Â Â Â Â Â Â Â Â Â legend.titleÂ =Â element_blank(),<br>Â Â Â Â Â Â Â Â Â Â Â Â legend.text=element_text(size=<span>19</span>))<br>gg<br></code></pre><p><img data-imgfileid="100004943" data-ratio="0.875" data-w="1080" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3IhQ4at5dbGaI6tiaeS34X1p3pL7eriaDkwePiaGVjl3TkpMfYd8kRZa3PcUjLzboAIiat7C2JgAVNyVw/640?wx_fmt=png" src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3IhQ4at5dbGaI6tiaeS34X1p3pL7eriaDkwePiaGVjl3TkpMfYd8kRZa3PcUjLzboAIiat7C2JgAVNyVw/640?wx_fmt=png"></p><h3><span>GSVAç»“æœå·®å¼‚åˆ†æ</span></h3><p><span>ä½¿ç”¨</span><span>limma</span><span>æ‰§è¡Œå·®å¼‚åˆ†æï¼Œè¿™é‡Œçš„åˆ†ç»„éšæœºç”Ÿæˆ</span></p><pre><code>GroupÂ &lt;-Â ifelse(pheno$num2Â %<span>in</span>%Â c(<span>"P002"</span>,<span>"P023"</span>,<span>"P024"</span>),<span>"Cold"</span>,<span>"Hot"</span>)<br>designÂ &lt;-Â model.matrix(~Group)<br><span>library</span>(limma)<br>fitÂ &lt;-Â lmFit(Hallmark_GSVA,Â design)<br>fitÂ &lt;-Â eBayes(fit)<br>DEGÂ &lt;-Â topTable(fit,Â coefÂ =Â <span>2</span>,Â numberÂ =Â <span>Inf</span>)<br>head(DEG)<br><span>#Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â logFCÂ Â Â Â Â AveExprÂ Â Â Â Â Â Â Â tÂ Â Â Â Â Â P.ValueÂ Â Â Â adj.P.ValÂ Â Â Â Â Â Â B</span><br><span>#HALLMARK_KRAS_SIGNALING_UPÂ Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.3562078Â -0.03475377Â 5.828477Â 4.020872e-07Â 2.010436e-05Â 6.24561</span><br><span>#HALLMARK_COMPLEMENTÂ Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.3828715Â -0.02954553Â 5.496154Â 1.308133e-06Â 3.270334e-05Â 5.10157</span><br><span>#HALLMARK_INTERFERON_GAMMA_RESPONSEÂ Â Â Â Â Â Â Â Â 0.4950833Â -0.01321527Â 5.281168Â 2.786445e-06Â 3.410775e-05Â 4.36965</span><br><span>#HALLMARK_INFLAMMATORY_RESPONSEÂ Â Â Â Â Â Â Â Â Â Â Â Â 0.3852465Â -0.03193564Â 5.248098Â 3.128339e-06Â 3.410775e-05Â 4.25774</span><br><span>#HALLMARK_ALLOGRAFT_REJECTIONÂ Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.4472482Â -0.01588245Â 5.212951Â 3.537149e-06Â 3.410775e-05Â 4.13902</span><br><span>#HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITIONÂ 0.4670243Â -0.03387425Â 5.171115Â 4.092930e-06Â 3.410775e-05Â 3.99799</span><br>DEGÂ &lt;-Â DEG[abs(DEG$logFC)Â &gt;Â <span>0.5</span>Â &amp;Â DEG$adj.P.ValÂ &lt;Â <span>0.05</span>,]<span></span></code></pre><section><img data-imgfileid="100003752" data-ratio="0.05278592375366569" data-type="png" data-w="341" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/1LTeQhNfr8sUH75oYsoDaqjPCTiaukEmS8tWricW7LnLKKfIE9jKBexibqamsrlibaaXmuc2nicaYibfDFBNCmqX5mBw/640?wx_fmt=other&amp;wxfrom=5&amp;wx_lazy=1&amp;wx_co=1&amp;tp=webp" src="https://mmbiz.qpic.cn/sz_mmbiz_png/1LTeQhNfr8sUH75oYsoDaqjPCTiaukEmS8tWricW7LnLKKfIE9jKBexibqamsrlibaaXmuc2nicaYibfDFBNCmqX5mBw/640?wx_fmt=other&amp;wxfrom=5&amp;wx_lazy=1&amp;wx_co=1&amp;tp=webp"></section><section><span></span></section><section data-tools="135ç¼–è¾‘å™¨" data-id="116886" draggable="true"><section><section><section><section><strong data-brushtype="text">è¢«ç‚¸ç†Ÿçš„è™¾</strong></section></section></section><section><section><section data-width="35%"><section><img data-cropselx1="0" data-cropselx2="115" data-cropsely1="0" data-cropsely2="115" data-imgfileid="100003750" data-ratio="1" data-type="jpeg" data-w="258" data-width="100%" data-src="https://mmbiz.qpic.cn/sz_mmbiz_jpg/dRYYdqiaan3JjDfj2H8p6gg3CB25AGthbwzrotao4ev5tIe0utthbZRK8yOoDOuTzOSoTSnPWn61IdDCnXsnaiag/640?wx_fmt=other&amp;wxfrom=5&amp;wx_lazy=1&amp;wx_co=1&amp;tp=webp" src="https://mmbiz.qpic.cn/sz_mmbiz_jpg/dRYYdqiaan3JjDfj2H8p6gg3CB25AGthbwzrotao4ev5tIe0utthbZRK8yOoDOuTzOSoTSnPWn61IdDCnXsnaiag/640?wx_fmt=other&amp;wxfrom=5&amp;wx_lazy=1&amp;wx_co=1&amp;tp=webp"></section></section><section data-width="63%"><section><section><strong>è‡ªå·±çš„æ‘¸ç´¢ï¼Œå‘ç°é—®é¢˜éº»çƒ¦å‘Šè¯‰ä½œè€…ï¼Œå…‰é€Ÿå›æ¥æ”¹æ­£</strong><strong><img data-imgfileid="100003751" data-ratio="1" data-type="png" data-w="64" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3KLQGNxibvI4SdtUfxjINkz2DO8cGEPTgbcMJ357RvXJ7IvWU2p7anljpePpakI9oKu3icJxobBDp5w/640?wx_fmt=other&amp;wxfrom=5&amp;wx_lazy=1&amp;wx_co=1&amp;tp=webp" src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3KLQGNxibvI4SdtUfxjINkz2DO8cGEPTgbcMJ357RvXJ7IvWU2p7anljpePpakI9oKu3icJxobBDp5w/640?wx_fmt=other&amp;wxfrom=5&amp;wx_lazy=1&amp;wx_co=1&amp;tp=webp"></strong></section></section></section></section></section></section></section><p><mp-style-type data-value="10000"></mp-style-type></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/w4xosiGw__9HyHpjTqk41w",target="_blank" rel="noopener noreferrer">åŸæ–‡é“¾æ¥</a>
