---
title: "DNA 19. SCI 文章用于群体结构分析(admixture）"
date: 2024-12-07T00:39:02Z
draft: ["false"]
tags: [
  "fetched",
  "桓峰基因"
]
categories: ["Acdemic"]
---
DNA 19. SCI 文章用于群体结构分析(admixture） by 桓峰基因
------
<div><p><img data-galleryid="" data-imgfileid="100011142" data-ratio="0.42532146389713155" data-s="300,640" data-src="https://mmbiz.qpic.cn/mmbiz_png/5ia62S5YZaMy3YXZFI4ePX6LFx9DZYdhBhmic6xYib66qOibd1rnfkKCL2S8LwJRbPm0bZqm8Uqune50V2MREyqGYw/640?wx_fmt=other&amp;tp=webp&amp;wxfrom=5&amp;wx_lazy=1&amp;wx_co=1" data-type="png" data-w="1011" src="https://mmbiz.qpic.cn/mmbiz_png/5ia62S5YZaMy3YXZFI4ePX6LFx9DZYdhBhmic6xYib66qOibd1rnfkKCL2S8LwJRbPm0bZqm8Uqune50V2MREyqGYw/640?wx_fmt=other&amp;tp=webp&amp;wxfrom=5&amp;wx_lazy=1&amp;wx_co=1"></p><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com"><h2 data-tool="mdnice编辑器"><span></span><span>简    介</span></h2><p data-tool="mdnice编辑器">群体分层长期以来被认为是遗传关联研究中的一个混淆因素。根据多基因座基因型数据估算的祖先可用于对种群分层进行统计校正。一种流行的估算祖先的技术是基于模型的方法，它体现在广泛应用的程序结构中。在程序 EIGENSTRAT 中实现的另一种方法依赖于主成分分析，而不是基于模型的估计，并且不直接提供混合物分数。EIGENSTRAT 已获得的普及，部分原因是由于其显著的速度相比结构。我们提出了一种新的算法和程序，ADMIXTURE用于基于模型的估计无亲缘关系个体的祖先。ADMIXTURE 采用嵌入结构的似然模型。然而，ADMIXTURE 运行速度快得多，在几分钟内解决问题，需要结构小时。在我们的许多实验中，我们发现 ADMIXTURE 几乎和 EIGENSTRAT 一样快。ADMIXTURE 的运行时改进依赖于使用顺序二次规划进行块更新的快速块松弛方案，以及新颖的准牛顿收敛加速。我们的算法也运行得更快，并且比合并在程序FRAPPE中的期望最大化（EM）算法的实现具有更高的精度。我们的模拟表明，ADMIXTURE 对潜在外加剂系数和祖先等位基因频率的最大似然估计与结构的贝叶斯估计一样准确。在真实世界的数据集上，ADMIXTURE 的估计可以直接与 structure 和 EIGENSTRAT 的估计进行比较。综上所述，我们的结果表明 ADMIXTURE 的计算速度打开了在基于模型的祖先估计中使用更大标记集的可能性，并且其估计适合用于校正关联研究中的人口分层。</p><p><img data-imgfileid="100011153" data-ratio="0.3042973286875726" data-s="300,640" data-src="https://mmbiz.qpic.cn/mmbiz_png/5ia62S5YZaMzEJdsbjMiaic33xESI3D5ibgOhRQWXWC5NwN3iaL7O2sRicLkUbCsD8EEia03yRl3ibgJpZPAI8vicibZpfibw/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="861" src="https://mmbiz.qpic.cn/mmbiz_png/5ia62S5YZaMzEJdsbjMiaic33xESI3D5ibgOhRQWXWC5NwN3iaL7O2sRicLkUbCsD8EEia03yRl3ibgJpZPAI8vicibZpfibw/640?wx_fmt=png&amp;from=appmsg"></p><h2 data-tool="mdnice编辑器"><span></span><span>软件包安装</span></h2><p data-tool="mdnice编辑器">软件包下载链接：</p><p data-tool="mdnice编辑器">https://dalexander.github.io/admixture/binaries/admixture_linux-1.3.0.tar.gz</p><p data-tool="mdnice编辑器">测试数据下载链接：</p><p data-tool="mdnice编辑器">https://dalexander.github.io/admixture/hapmap3-files.tar.gz</p><p data-tool="mdnice编辑器">解压之后进入文件夹里面找到 admixture，运行即可。</p><pre data-tool="mdnice编辑器"><span></span><code>$tar -xvf hapmap3-files.tar.gz<br>$cd   dist/admixture_linux-<span>1.3</span><span>.0</span>/<br> ./admixture --help<br>****                   ADMIXTURE Version <span>1.3</span><span>.0</span>                  ****<br>****                    Copyright <span>2008</span>-<span>2015</span>                     ****<br>****           David Alexander, Suyash Shringarpure,            ****<br>****                John  Novembre, Ken Lange                   ****<br>****                                                            ****<br>****                 Please cite our paper!                     ****<br>****   Information at www.genetics.ucla.edu/software/admixture  ****<br><br>                                                                              <br>  ADMIXTURE basic usage:  (see manual <span>for</span> complete reference)                 <br>    % admixture [options] inputFile K                                         <br>                                                                              <br>  where:                                                                      <br>    K is the number of populations; and                                       <br>    inputFile may be:                                                         <br>      - a PLINK .bed file                                                     <br>      - a PLINK <span>"12"</span> coded .ped file                                        <br>                                                                              <br>  Output will be <span>in</span> files inputBasename.K.Q, inputBasename.K.P                <br>                                                                              <br>  General options:                                                            <br>    -jX          : do computation on X threads                                <br>    --seed=X     : use random seed X <span>for</span> initialization                       <br>                                                                              <br>  Algorithm options:                                                          <br>     -m=                                                                      <br>    --method=[em|block]     : set method.  block is default                   <br>                                                                              <br>     -a=                                                                      <br>    --acceleration=none   |                                                   <br>                   sqs&lt;X&gt; |                                                   <br>                   qn&lt;X&gt;      : set acceleration                              <br>                                                                              <br>  Convergence criteria:                                                       <br>    -C=X : set major convergence criterion (<span>for</span> point estimation)             <br>    -c=x : set minor convergence criterion (<span>for</span> bootstrap and CV reestimates) <br>                                                                              <br>  Bootstrap standard errors:                                                  <br>    -B[X]      : do bootstrapping [with X replicates]<br></code></pre><h2 data-tool="mdnice编辑器"><span></span><span>数据准备</span></h2><p data-tool="mdnice编辑器">解压测试数据集，可以看到里面包含了三个文件，这三个文件都是通过plink软件生产而来。</p><pre data-tool="mdnice编辑器"><span></span><code>$tar -xvf hapmap3-files.tar.gz<br>$ls<br>hapmap3.bed hapmap3.bim hapmap3.fam<br></code></pre><p data-tool="mdnice编辑器">可以通过PLINK生产 .ped 12"编码文件，该文件由以下命令生成</p><pre data-tool="mdnice编辑器"><span></span><code>plink --file hapmap --recode12 --out hapmap<br></code></pre><h2 data-tool="mdnice编辑器"><span></span><span>实例操作</span></h2><h3 data-tool="mdnice编辑器"><span></span>已知群体结构<span></span></h3><p data-tool="mdnice编辑器">你相信样本中的个体的祖先来自三个祖先人群，然后像这样进行混合：</p><pre data-tool="mdnice编辑器"><span></span><code>./admixture hapmap3.bed <span>3</span><br></code></pre><p data-tool="mdnice编辑器">若是 .ped 执行下面的命令，生成两个文件.Q和.P：</p><pre data-tool="mdnice编辑器"><span></span><code>$./admixture hapmap3.ped <span>3</span><br><span>## 产生文件</span><br>hapmap3.3.Q hapmap3.3.P<br></code></pre><p data-tool="mdnice编辑器">如果还需要标准错误，那么应该使用命令，可以看到结果文件多了 .Q_se。</p><pre data-tool="mdnice编辑器"><span></span><code>./admixture -B hapmap3.ped <span>3</span><br><span>## 生成文件</span><br>hapmap3.3.Q hapmap3.3.P hapmap3.3.Q_se<br></code></pre><p data-tool="mdnice编辑器">如果您确实需要分析所有标记，或者由于其他原因（非常大的K，广泛的交叉验证或引导），您可以考虑在多线程模式下运行 ADMIXTURE。例如，如果您的计算机有四个处理器，您可以使用这样的命令：</p><pre data-tool="mdnice编辑器"><span></span><code>./admixture hapmap3.bed <span>3</span> -j4<br></code></pre><h3 data-tool="mdnice编辑器"><span></span>未知结构如何选择K的正确值?<span></span></h3><p data-tool="mdnice编辑器">使用 ADMIXTURE 的交叉验证程序。与其他K值相比，良好的K值将表现出较低的交叉验证误差。只需在ADMIXTURE 命令行中添加——cv标志即可启用交叉验证。在这个默认设置中，交叉验证过程将执行5倍CV，例如，使用——cv =10，可以得到10倍CV。在输出中报告交叉验证错误。</p><pre data-tool="mdnice编辑器"><span></span><code>$<span>for</span> K <span>in</span> <span>1</span> <span>2</span> <span>3</span> <span>4</span> <span>5</span> <span>6</span> <span>7</span> <span>8</span> <span>9</span> <span>10</span> ; do ./dist/admixture_linux-<span>1.3</span><span>.0</span>/admixture --cv hapmap3.bed <span>10</span> | tee log10.out; done <br><br>$grep -h CV log*.out &gt; CVerror.txt<br>CV error (K=<span>10</span>): <span>0.53864</span><br>CV error (K=<span>1</span>): <span>0.55248</span><br>CV error (K=<span>2</span>): <span>0.48190</span><br>CV error (K=<span>3</span>): <span>0.47835</span><br>CV error (K=<span>4</span>): <span>0.48236</span><br>CV error (K=<span>5</span>): <span>0.49001</span><br>CV error (K=<span>6</span>): <span>0.50069</span><br>CV error (K=<span>7</span>): <span>0.51484</span><br>CV error (K=<span>8</span>): <span>0.51761</span><br>CV error (K=<span>9</span>): <span>0.52387</span><br></code></pre><h3 data-tool="mdnice编辑器"><span></span>群体结构可视化<span></span></h3><p data-tool="mdnice编辑器">将CV结果复制粘贴至文本中中，ggplot2 绘制折线图。图中可看出最佳分群数为K=3。</p><pre data-tool="mdnice编辑器"><span></span><code>CVerror = read.table(<span>"CVerror.txt"</span>)<br>CVerror[, <span>3</span>]<br><span>##  [1] "(K=10):" "(K=1):"  "(K=2):"  "(K=3):"  "(K=4):"  "(K=5):"  "(K=6):" </span><br><span>##  [8] "(K=7):"  "(K=8):"  "(K=9):"</span><br><span>library</span>(stringr)<br>data = data.frame(K = str_extract(CVerror[, <span>3</span>], <span>"([:digit:]{1,2})"</span>), Error = CVerror[,<br>    <span>4</span>])<br>data$K = as.integer(data$K)<br><span>library</span>(ggplot2)<br>ggplot(data, aes(x = K, y = Error)) + geom_line() + geom_point() + ylab(<span>"Cross−validation error"</span>) +<br>    scale_x_continuous(breaks = seq(<span>1</span>, <span>10</span>, <span>1</span>))<br></code></pre><p><img data-imgfileid="100011155" data-ratio="0.7667560321715817" data-s="300,640" data-src="https://mmbiz.qpic.cn/mmbiz_png/5ia62S5YZaMzEJdsbjMiaic33xESI3D5ibgOY2TvlZPpUwlMcwxMEexneoeAQB85ibeu6QR8tEGhVZicqGIDkoAkUG1g/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="746" src="https://mmbiz.qpic.cn/mmbiz_png/5ia62S5YZaMzEJdsbjMiaic33xESI3D5ibgOY2TvlZPpUwlMcwxMEexneoeAQB85ibeu6QR8tEGhVZicqGIDkoAkUG1g/640?wx_fmt=png&amp;from=appmsg"></p><h3 data-tool="mdnice编辑器"><span></span>如何画出Q的估计值呢？<span></span></h3><p data-tool="mdnice编辑器">Q估计值作为一个简单的矩阵输出，要制作堆叠的条形图，使用barplot命令。</p><pre data-tool="mdnice编辑器"><span></span><code>tbl = read.table(<span>"hapmap3.3.Q"</span>)<br>colnames(tbl) = c(<span>"Q1"</span>, <span>"Q2"</span>, <span>"Q3"</span>)<br>barplot(t(as.matrix(tbl)), col = rainbow(<span>3</span>), xlab = <span>"Individual #"</span>, ylab = <span>"Ancestry"</span>,<br>    border = <span>NA</span>)<br>legend(x = <span>5</span>, y = -<span>0.01</span>, pch = <span>15</span>, colnames(tbl), col = rainbow(<span>3</span>), bty = <span>"n"</span>, horiz = <span>TRUE</span>,<br>    xpd = <span>TRUE</span>)<br></code></pre><p><img data-imgfileid="100011156" data-ratio="0.38055555555555554" data-s="300,640" data-src="https://mmbiz.qpic.cn/mmbiz_png/5ia62S5YZaMzEJdsbjMiaic33xESI3D5ibgO8eXYPvIb3JJ74k7ibtTGEbm7ISSQVBNMmn7BiaSulvJibVWOtViaIIaiarw/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/5ia62S5YZaMzEJdsbjMiaic33xESI3D5ibgO8eXYPvIb3JJ74k7ibtTGEbm7ISSQVBNMmn7BiaSulvJibVWOtViaIIaiarw/640?wx_fmt=png&amp;from=appmsg"></p><h4 data-tool="mdnice编辑器"><span></span><span>Reference</span><span></span></h4><p data-tool="mdnice编辑器">Alexander DH, Novembre J, Lange K. Fast model-based estimation of ancestry in unrelated individuals. Genome Res. 2009 Sep;19(9):1655-64.</p><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com"><h2 data-tool="mdnice编辑器"><span>基因组生信分析教程</span></h2><p data-tool="mdnice编辑器"><a href="https://mp.weixin.qq.com/s?__biz=MzU0NDY2NTUwOA==&amp;mid=2247486323&amp;idx=1&amp;sn=ce8841a6ad47b58b8b57dc6d8f0eacc4&amp;scene=21#wechat_redirect" data-linktype="2"><span>DNA 1. Germline Mutation Vs. Somatic Mutation 傻傻分不清楚</span></a></p><p data-tool="mdnice编辑器"><a href="https://mp.weixin.qq.com/s?__biz=MzU0NDY2NTUwOA==&amp;mid=2247486359&amp;idx=1&amp;sn=67577469c7bfff0d644411704b5060f5&amp;scene=21#wechat_redirect" data-linktype="2"><span>DNA 2. SCI 文章中基因组变异分析神器之 maftools</span></a></p><p data-tool="mdnice编辑器"><a href="https://mp.weixin.qq.com/s?__biz=MzU0NDY2NTUwOA==&amp;mid=2247486473&amp;idx=1&amp;sn=ae1820d8cf2e2967f1364b44be155d43&amp;scene=21#wechat_redirect" data-linktype="2"><span>DNA 3. SCI 文章中基因组变异分析神器之 maftools</span></a></p><p data-tool="mdnice编辑器"><a href="https://mp.weixin.qq.com/s?__biz=MzU0NDY2NTUwOA==&amp;mid=2247486617&amp;idx=1&amp;sn=058b375db9c39275f37ab1aee2ce0fc5&amp;scene=21#wechat_redirect" data-linktype="2"><span>DNA 4. SCI 文章中基因组的突变信号（maftools）</span></a></p><p data-tool="mdnice编辑器"><a href="https://mp.weixin.qq.com/s?__biz=MzU0NDY2NTUwOA==&amp;mid=2247487200&amp;idx=1&amp;sn=57baa9dc823c2da7c0ede981ddf782f7&amp;scene=21#wechat_redirect" data-linktype="2"><span>DNA 5. 基因组变异文件VCF格式详解</span></a></p><p data-tool="mdnice编辑器"><a href="https://mp.weixin.qq.com/s?__biz=MzU0NDY2NTUwOA==&amp;mid=2247487672&amp;idx=1&amp;sn=9ac1c81307e9015a0183e783dd95722f&amp;scene=21#wechat_redirect" data-linktype="2"><span>DNA 6. 基因组变异之绘制精美瀑布图（ComplexHeatmap）</span></a></p><p data-tool="mdnice编辑器"><a href="https://mp.weixin.qq.com/s?__biz=MzU0NDY2NTUwOA==&amp;mid=2247487727&amp;idx=1&amp;sn=2d61c35afd5aa1ff935d2dc1a0cd25d5&amp;scene=21#wechat_redirect" data-linktype="2"><span>DNA 7. 基因组拷贝数变异分析及可视化 (GISTIC2.0)</span></a></p><p data-tool="mdnice编辑器"><a href="https://mp.weixin.qq.com/s?__biz=MzU0NDY2NTUwOA==&amp;mid=2247487743&amp;idx=1&amp;sn=d1d16293d22fef847fbd8e3c591c44c7&amp;scene=21#wechat_redirect" data-linktype="2"><span>DNA 8. 癌症的突变异质性及寻找新的癌症驱动基因(MutSigCV)</span></a></p><p data-tool="mdnice编辑器"><a href="https://mp.weixin.qq.com/s?__biz=MzU0NDY2NTUwOA==&amp;mid=2247487783&amp;idx=1&amp;sn=8b16e34567f9ca14497fec6b8589c94b&amp;scene=21#wechat_redirect" data-linktype="2"><span>DNA 9. 揭秘肿瘤异质性与TMB, MSI之间的相关性</span></a></p><p data-tool="mdnice编辑器"><a href="https://mp.weixin.qq.com/s?__biz=MzU0NDY2NTUwOA==&amp;mid=2247487869&amp;idx=1&amp;sn=a2e1ca0c1ba8bf66e4d1d313c76f2758&amp;scene=21#wechat_redirect" data-linktype="2"><span>DNA 10. 识别癌症驱动基因 (OncodriveCLUST)</span></a></p><p data-tool="mdnice编辑器"><a href="https://mp.weixin.qq.com/s?__biz=MzU0NDY2NTUwOA==&amp;mid=2247487906&amp;idx=1&amp;sn=ce88e0b9f3891e7a4ea6df32b0d3231d&amp;scene=21#wechat_redirect" data-linktype="2"><span>DNA 11. 识别肿瘤蛋白质三维结构上突变热点(HotSpot3D)</span></a></p><p data-tool="mdnice编辑器"><a href="https://mp.weixin.qq.com/s?__biz=MzU0NDY2NTUwOA==&amp;mid=2247488195&amp;idx=1&amp;sn=6e162e15007e445c329528b1a74eebed&amp;scene=21#wechat_redirect" data-linktype="2"><span>DNA 12. SCI 文章绘图之全基因组关联分析可视化(GWAS)</span></a></p><p data-tool="mdnice编辑器"><a href="https://mp.weixin.qq.com/s?__biz=MzU0NDY2NTUwOA==&amp;mid=2247490396&amp;idx=1&amp;sn=e9c17ff4d7142531c52962a4193fe8f5&amp;scene=21#wechat_redirect" data-linktype="2"><span>DNA 13. SCI 文章肿瘤突变负荷计算方法（TMB）</span></a></p><p data-tool="mdnice编辑器"><a href="https://mp.weixin.qq.com/s?__biz=MzU0NDY2NTUwOA==&amp;mid=2247490420&amp;idx=1&amp;sn=92b8a82c636627fd9c29128c44cf7720&amp;scene=21#wechat_redirect" data-linktype="2"><span>DNA 14. SCI 文章肿瘤微卫星不稳定性计算方法（MSI）</span></a></p><p data-tool="mdnice编辑器"><a href="https://mp.weixin.qq.com/s?__biz=MzU0NDY2NTUwOA==&amp;mid=2247490437&amp;idx=1&amp;sn=c7f3735a1b23f650a4c2445f8e33d54c&amp;scene=21#wechat_redirect" data-linktype="2"><span>DNA 15. SCI 文章肿瘤微卫星不稳定性之 MSIsensor 系列软件</span></a></p><p data-tool="mdnice编辑器"><a href="https://mp.weixin.qq.com/s?__biz=MzU0NDY2NTUwOA==&amp;mid=2247490525&amp;idx=1&amp;sn=ac397eba03287d230a69283c50df5c2e&amp;scene=21#wechat_redirect" data-linktype="2"><span>DNA 16. SCI 文章研究表型与基因型之间的关系工具（TASSEL）</span></a></p><p data-tool="mdnice编辑器"><a href="https://mp.weixin.qq.com/s?__biz=MzU0NDY2NTUwOA==&amp;mid=2247490640&amp;idx=1&amp;sn=f0f48c752bb3d9c150af6e9f5c194e1b&amp;scene=21#wechat_redirect" data-linktype="2"><span>DNA 17. SCI 文章用于分析基因组多样性工具 (rTASSEL)</span></a></p><p data-tool="mdnice编辑器"><a href="https://mp.weixin.qq.com/s?__biz=MzU0NDY2NTUwOA==&amp;mid=2247492457&amp;idx=1&amp;sn=64bcbfdd7b3fbf7c15db79ea9b8e23de&amp;scene=21#wechat_redirect" data-linktype="2"><span>DNA 18. SCI 文章用于识别肿瘤驱动基因(driveR)</span></a></p></section><section><mp-common-videosnap data-pluginname="mpvideosnap" data-url="https://findermp.video.qq.com/251/20304/stodownload?encfilekey=rjD5jyTuFrIpZ2ibE8T7YmwgiahniaXswqzCBEK5NcB3ianSmVFbhnuib65QVXxFSsCFYic6Xibibl8992xXDXPIyw75afwiarQrRqnYHXvtVxQmZEI7VLmnoFhq4Eg&amp;token=cztXnd9GyrHtDgPsibUJDcj4ak1wriaJcYEKPq9YsEMQiclIia8jD4dUng1WIZ5xWicuN9zjGyKnb01qQUE0SFj533lZcHmcTVNib6u4eEvbAVNfLQ2ME5PKIic3K57pEMnnqAt&amp;idx=1&amp;hy=SH&amp;m=&amp;scene=2&amp;uzid=2" data-headimgurl="http://wx.qlogo.cn/finderhead/Q3auHgzwzM6NFe6qfZ98pccnStLOBzJIic5564gE1oW2m34RtSicQHIQ/0" data-username="v2_060000231003b20faec8c6eb891cc2ddca0ce43cb07767727967ddf958ff98912bf15d898b0d@finder" data-nickname="桓峰基因" data-desc="DNA 19. SCI 文章用于群体结构分析(admixture）#基因组分析#DNA#群体分析#GWAS@桓峰基因" data-nonceid="1487655130172002924" data-type="video" data-mediatype="undefined" data-authiconurl="" data-from="new" data-width="1920" data-height="1080" data-id="export/UzFfAgtgekIEAQAAAAAAl8AQOwzoOwAAAAstQy6ubaLX4KHWvLEZgBPEt6JERzRRFoOKzNPgMIs_ZT4SI647UGk4_SM3y672" data-isdisabled="0" data-errortips=""></mp-common-videosnap></section><p data-tool="mdnice编辑器"><br></p></section><p><mp-style-type data-value="3"></mp-style-type></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/Z49Q0viKeQS3U7petAw6nw",target="_blank" rel="noopener noreferrer">原文链接</a>
