---
title: "scATAC 教程一文详解"
date: 2024-12-08T09:59:53Z
draft: ["false"]
tags: [
  "fetched",
  "老俊俊的生信笔记"
]
categories: ["Acdemic"]
---
scATAC 教程一文详解 by 老俊俊的生信笔记
------
<div><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com" data-mpa-powered-by="yiban.io"><section><mp-common-profile data-pluginname="mpprofile" data-id="MzkyMTI1MTYxNA==" data-headimg="http://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usezgsqIGqjITSMggCTSoViaYeoKe2xoZr1IIvNJoztibQxibYHLDDoiabwAc6Ggws3Tvdo8EPss2nLgaVQ/0?wx_fmt=png" data-nickname="老俊俊的生信笔记" data-alias="JunJunLab" data-signature="老俊俊的生信技能和知识分享,我不是巨人,但你可以站在我的肩膀上更进一步!" data-from="0" data-is_biz_ban="0" data-service_type="1"></mp-common-profile></section></section><section><mp-common-clmusic data-pluginname="insertaudio" type="1" music_name="%E5%85%B6%E5%AE%9E%E9%83%BD%E6%B2%A1%E6%9C%89" albumurl="http://wx.y.gtimg.cn/music/photo_new/T002R500x500M000000KL9de0YaLIR_1.jpg" singer="杨宗纬" duration="231000" username="" music_source="1" is_vip="1" listenid="78221633842093184"></mp-common-clmusic></section><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com"><h2 data-tool="mdnice编辑器"><span></span><span>引言</span><span></span></h2><blockquote data-tool="mdnice编辑器"><span>❝</span><p><strong>scATAC-seq</strong> 结合了 <strong>染色质可接近性分析</strong> 和 <strong>单细胞分辨率</strong> 的优势，可用于 <strong>研究细胞异质性和调控机制</strong>，尤其在肿瘤、免疫学和发育生物学领域显示出巨大潜力。如果有 <strong>bulk-ATAC-seq</strong> 的基础，那么理解 <strong>scATAC-seq</strong> 会有很大帮助的。目前争对单细胞级别的的表观测序有 <strong>signac</strong>，<strong>ArchR</strong>，<strong>snapATAC/snapATAC2</strong>，<strong>epiScanpy</strong> 等等工具。今天我们跟着 <strong>signac</strong> 来具体了解一下数据的类型，分析流程。</p></blockquote><blockquote data-tool="mdnice编辑器"><span>❝</span><p><strong>signac</strong> 工具是 <strong>seurat</strong> 团队开发的，能够很好的对接 <strong>seurat</strong> 软件。</p></blockquote><p data-tool="mdnice编辑器"><strong>安装：</strong></p><pre data-tool="mdnice编辑器"><span></span><code>setRepositories(ind=<span>1</span>:<span>3</span>)<br>install.packages(<span>"Signac"</span>)<br></code></pre><blockquote data-tool="mdnice编辑器"><span>❝</span><p>我们知道单细胞转录组输入文件 <strong>一般有三个</strong>（<strong>每个细胞的 id：矩阵的列名</strong>，<strong>细胞里的基因名：行名</strong>，<strong>以及每个基因每个细胞的表达量：矩阵的内容</strong>），最终组合成一个完整的矩阵而已。<strong>scATAC-seq</strong> 也是一样，（<strong>每个细胞的 id：矩阵的列名</strong>，<strong>细胞里的基因区域/染色质开放区域：行名</strong>，<strong>以及每个基因每个区域的定量：矩阵的内容</strong>），最终也是组合成一个完整的矩阵而已。是不是基本差不多，只不过基因名称换成了检测到的基因组的区间位置而已，<strong>还有就是需要所有细胞不同区间位置的 fragments 这样一个文件</strong>。</p></blockquote><p data-tool="mdnice编辑器"><strong>下面是一个官网示例（https://support.10xgenomics.com/single-cell-atac/software/pipelines/latest/output/matrices）：</strong></p><pre data-tool="mdnice编辑器"><span></span><code><span>require</span>(magrittr)<br><span>require</span>(readr)<br><span>require</span>(Matrix)<br><span>require</span>(tidyr)<br><span>require</span>(dplyr)<br><br><span># peak-bc matrix</span><br>mex_dir_path &lt;- <span>"/opt/sample345/outs/filtered_peak_bc_matrix"</span><br><br>mtx_path &lt;- paste(mex_dir_path, <span>"matrix.mtx"</span>, sep = <span>'/'</span>)<br>feature_path &lt;- paste(mex_dir_path, <span>"peaks.bed"</span>, sep = <span>'/'</span>)<br>barcode_path &lt;- paste(mex_dir_path, <span>"barcodes.tsv"</span>, sep = <span>'/'</span>)<br><br>features &lt;- readr::read_tsv(feature_path, col_names = <span>F</span>) %&gt;% tidyr::unite(feature)<br>barcodes &lt;- readr::read_tsv(barcode_path, col_names = <span>F</span>) %&gt;% tidyr::unite(barcode)<br><br>mtx &lt;- Matrix::readMM(mtx_path) %&gt;%<br>  magrittr::set_rownames(features$feature) %&gt;%<br>  magrittr::set_colnames(barcodes$barcode)<br><br><span># tf-bc matrix</span><br>mex_dir_path &lt;- <span>"/opt/sample345/outs/filtered_tf_bc_matrix"</span><br><br>mtx_path &lt;- paste(mex_dir_path, <span>"matrix.mtx"</span>, sep = <span>'/'</span>)<br>feature_path &lt;- paste(mex_dir_path, <span>"motifs.tsv"</span>, sep = <span>'/'</span>)<br>barcode_path &lt;- paste(mex_dir_path, <span>"barcodes.tsv"</span>, sep = <span>'/'</span>)<br><br>features &lt;- readr::read_tsv(feature_path, col_names = c(<span>'feature'</span>, <span>'common_name'</span>))<br>barcodes &lt;- readr::read_tsv(barcode_path, col_names = <span>F</span>) %&gt;% tidyr::unite(barcode)<br><br>mtx &lt;- Matrix::readMM(mtx_path) %&gt;%<br>  magrittr::set_rownames(features$feature) %&gt;%<br>  magrittr::set_colnames(barcodes$barcode)<br></code></pre><p data-tool="mdnice编辑器"><strong>最终都是准备成 mtx 这样一个表达矩阵。</strong></p><p data-tool="mdnice编辑器"><strong>cellranger-atac 输出下面三个文件：</strong></p><pre data-tool="mdnice编辑器"><span></span><code>$ cd /home/jdoe/runs/sample345/outs<br>$ tree filtered_peak_bc_matrix<br>filtered_peak_bc_matrix<br>├── barcodes.tsv<br>├── peaks.bed<br>└── matrix.mtx<br><span>1</span> directories, <span>3</span> files<br></code></pre><h2 data-tool="mdnice编辑器"><span></span><span>10xgenomics scATAC-seq 建库原理</span><span></span></h2><blockquote data-tool="mdnice编辑器"><span>❝</span><p>和 <strong>scRNA-seq</strong> 类似但又不同，<strong>scATAC-seq</strong> 捕获的 DNA 的序列。<strong>核心是利用 Tn5 转座酶的原理切割 DNA 上开放的区域，并连接好测序接头。</strong></p></blockquote><figure data-tool="mdnice编辑器"><img data-imgfileid="100032160" data-ratio="0.8252911813643927" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBVEopxFSiaJUSW49UIdUIhlF4lCe19Gu1VhCguP1pzickCb22bhVKibhGrw/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="601" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBVEopxFSiaJUSW49UIdUIhlF4lCe19Gu1VhCguP1pzickCb22bhVKibhGrw/640?wx_fmt=png&amp;from=appmsg"></figure><p data-tool="mdnice编辑器"><strong>第一步： 将细胞核和转座酶混合物孵育进行 DNA 开放区域的片段切割和添加接头。（Nuclei suspensions are incubated in a Transposition Mix that includes a Transposase. The Transposase enters the nuclei and preferentially fragments the DNA in open regions of the chromatin. Simultaneously, adapter sequences are added to the ends of the DNA fragments）</strong></p><p data-tool="mdnice编辑器"><strong>第二步：通过微液滴技术生成油包水的 GEM。(GEMs are generated by combining barcoded Gel Beads, transposed nuclei, a Master Mix, and Partitioning Oil on a Chromium Next GEM Chip H. To achieve single nuclei resolution, the nuclei are delivered at a limiting dilution, such that the majority (~90-99%) of generated GEMs contains no nuclei, while the remainder largely contain a single nucleus)</strong></p><figure data-tool="mdnice编辑器"><img data-imgfileid="100032162" data-ratio="0.8985507246376812" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBVRRq0icG8407rL7CsmtMz08pTvacGKx8oPGPDTDCnydhN7De7j511Ylw/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="552" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBVRRq0icG8407rL7CsmtMz08pTvacGKx8oPGPDTDCnydhN7De7j511Ylw/640?wx_fmt=png&amp;from=appmsg"></figure><p data-tool="mdnice编辑器"><strong>磁珠上的接头与 tn5 切割下来的 DNA 片段进行互补配对，进行序列延伸。</strong></p><figure data-tool="mdnice编辑器"><img data-imgfileid="100032161" data-ratio="0.9758064516129032" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBVRjAVYaRjibUmqKK2IWa7rzBd83tmMOLMZM3s7AJLSuhKYW0mUvnvQ6A/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="496" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBVRjAVYaRjibUmqKK2IWa7rzBd83tmMOLMZM3s7AJLSuhKYW0mUvnvQ6A/640?wx_fmt=png&amp;from=appmsg"></figure><p data-tool="mdnice编辑器"><strong>第三步：对 GEM (Gel Beads-in-Emulsion, 即胶珠乳液) 反应后的混合产物进行必要的清理。这一步旨在去除多余的试剂、残留的条形码（barcodes），以及其他潜在杂质，以保证后续文库构建和测序的高质量。（Silane magnetic beads are used to remove leftover biochemical reagents from the post GEM reaction mixture. Solid Phase Reversible Immobilization (SPRI) beads are used to eliminate unused barcodes from the sample.）</strong></p><p data-tool="mdnice编辑器"><strong>第四步：破碎油包水液滴合并所有磁珠，添加 P7 接头序列，用于后续桥式扩增。（ P7 and a sample index are added during library construction via PCR. The final libraries contain the P5 and P7 sequences used in Illumina® bridge amplification）</strong></p><figure data-tool="mdnice编辑器"><img data-imgfileid="100032159" data-ratio="0.7418244406196214" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBV2b0BFrHMU7TRzSObmeKe0oma0KMnjgPzOrLic4ibrbywmRcS4WZydf1A/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="581" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBV2b0BFrHMU7TRzSObmeKe0oma0KMnjgPzOrLic4ibrbywmRcS4WZydf1A/640?wx_fmt=png&amp;from=appmsg"></figure><p data-tool="mdnice编辑器"><strong>第五步：扩增序列，文库构建，进行测序。（ The Chromium Next GEM Single Cell ATAC Reagent Kits v2 protocol produces Illumina® ready sequencing libraries. Illumina® sequencer compatibility, sample indices, sequencing depth &amp; run parameters, library loading and pooling are summarized.）</strong></p><figure data-tool="mdnice编辑器"><img data-imgfileid="100032158" data-ratio="0.29314159292035397" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBVibDXHWu09f3fdwA0URicJ0GADGL6MXf8nXacBCT0OWMbs5GxHenwXPicA/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="904" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBVibDXHWu09f3fdwA0URicJ0GADGL6MXf8nXacBCT0OWMbs5GxHenwXPicA/640?wx_fmt=png&amp;from=appmsg"></figure><blockquote data-tool="mdnice编辑器"><span>❝</span><p>测序数据使用 <strong>Cell Ranger ATAC</strong> 进行分析，也是有 <strong>cellranger-atac mkfastq</strong> 拆分 fastq 文件和 <strong>cellranger-atac count</strong> 定量：</p></blockquote><figure data-tool="mdnice编辑器"><img data-imgfileid="100032166" data-ratio="0.500330906684315" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBV6NibA6cMAU0KvSS9S934pGcP02dtrDS1SNicwA85FnibNtTZEibZpZ1bqA/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1511" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBV6NibA6cMAU0KvSS9S934pGcP02dtrDS1SNicwA85FnibNtTZEibZpZ1bqA/640?wx_fmt=png&amp;from=appmsg"></figure><h2 data-tool="mdnice编辑器"><span></span><span>signac 分析流程</span><span></span></h2><blockquote data-tool="mdnice编辑器"><span>❝</span><p><strong>signac 功能：</strong></p><ul><li><section><strong>Calling peaks</strong></section></li><li><section><strong>Quantifying per-cell counts in different genomic regions</strong></section></li><li><section><strong>Calculating single-cell QC metrics</strong></section></li><li><section><strong>Dimensional reduction, visualization, and clustering</strong></section></li><li><section><strong>Identifying cell-type-specific peaks</strong></section></li><li><section><strong>Visualizing ‘pseudo-bulk’ coverage tracks</strong></section></li><li><section><strong>Integration of multiple single-cell datasets</strong></section></li><li><section><strong>Integration with single-cell RNA-seq datasets</strong></section></li><li><section><strong>Sequence motif enrichment analysis</strong></section></li><li><section><strong>Transcription factor footprinting analysis</strong></section></li><li><section><strong>Linking peaks to correlated genes</strong></section></li><li><section><strong>Parallelization through the future package</strong></section></li><li><section><strong>Seamless interface with Seurat, SeuratWrappers, SeuratDisk, and SeuratData functionality</strong></section></li><li><section><strong>Interoperability with Bioconductor tools</strong></section></li></ul></blockquote><h3 data-tool="mdnice编辑器"><span></span><span>数据下载</span><span></span></h3><pre data-tool="mdnice编辑器"><span></span><code>wget https://cf.10xgenomics.com/samples/cell-atac/2.1.0/10k_pbmc_ATACv2_nextgem_Chromium_Controller/10k_pbmc_ATACv2_nextgem_Chromium_Controller_filtered_peak_bc_matrix.h5<br>wget https://cf.10xgenomics.com/samples/cell-atac/2.1.0/10k_pbmc_ATACv2_nextgem_Chromium_Controller/10k_pbmc_ATACv2_nextgem_Chromium_Controller_singlecell.csv<br>wget https://cf.10xgenomics.com/samples/cell-atac/2.1.0/10k_pbmc_ATACv2_nextgem_Chromium_Controller/10k_pbmc_ATACv2_nextgem_Chromium_Controller_fragments.tsv.gz<br>wget https://cf.10xgenomics.com/samples/cell-atac/2.1.0/10k_pbmc_ATACv2_nextgem_Chromium_Controller/10k_pbmc_ATACv2_nextgem_Chromium_Controller_fragments.tsv.gz.tbi<br></code></pre><p data-tool="mdnice编辑器"><strong>10k_pbmc_ATACv2_nextgem_Chromium_Controller_singlecell.csv</strong> 是细胞的 metadata 信息：</p><figure data-tool="mdnice编辑器"><img data-imgfileid="100032167" data-ratio="0.5321805955811719" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBVwwicYB457SobvtSmR7rM7ibcfcibAsK1eJbAyI2lNdRvVH2bcVYbPxY3A/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1041" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBVwwicYB457SobvtSmR7rM7ibcfcibAsK1eJbAyI2lNdRvVH2bcVYbPxY3A/640?wx_fmt=png&amp;from=appmsg"></figure><p data-tool="mdnice编辑器"><strong>10k_pbmc_ATACv2_nextgem_Chromium_Controller_fragments.tsv.gz</strong> 所有细胞每个位置区间的 read 数：</p><figure data-tool="mdnice编辑器"><img data-imgfileid="100032165" data-ratio="0.5092682926829268" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBVOFaxapuTWWriaYOd58aY3lgct3FUBfzZwSdqnoskJicT1yeRhSxN5iaRQ/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1025" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBVOFaxapuTWWriaYOd58aY3lgct3FUBfzZwSdqnoskJicT1yeRhSxN5iaRQ/640?wx_fmt=png&amp;from=appmsg"></figure><figure data-tool="mdnice编辑器"><img data-imgfileid="100032163" data-ratio="0.41935483870967744" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBVfhUzxA6SPDKAfJ1ru2rWWS2kId8neEH1D1YnWBUvqTlDaz36szQDZQ/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="527" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBVfhUzxA6SPDKAfJ1ru2rWWS2kId8neEH1D1YnWBUvqTlDaz36szQDZQ/640?wx_fmt=png&amp;from=appmsg"></figure><p data-tool="mdnice编辑器"><strong>10k_pbmc_ATACv2_nextgem_Chromium_Controller_filtered_peak_bc_matrix.h5</strong> 为每个细胞每个开放染色质区间的矩阵，类似单细胞表达矩阵，行不再是基因名，而是位置区间，值代表 Tn5 的结合次数。一般是过滤后的在多个细胞中保守的可及性矩阵。</p><pre data-tool="mdnice编辑器"><span></span><code>counts &lt;- Read10X_h5(filename = <span>"10k_pbmc_ATACv2_nextgem_Chromium_Controller_filtered_peak_bc_matrix.h5"</span>)<br><br>str(counts)<br><span># Formal class 'dgCMatrix' [package "Matrix"] with 6 slots</span><br><span># ..@ i       : int [1:115807238] 14 37 40 86 98 99 110 113 115 122 ...</span><br><span># ..@ p       : int [1:10247] 0 12605 17997 31303 45352 52467 66981 91458 102075 112905 ...</span><br><span># ..@ Dim     : int [1:2] 165434 10246</span><br><span># ..@ Dimnames:List of 2</span><br><span># .. ..$ : chr [1:165434] "chr1:9772-10660" "chr1:180712-181178" "chr1:181200-181607" "chr1:191183-192084" ...</span><br><span># .. ..$ : chr [1:10246] "AAACGAAAGAGAGGTA-1" "AAACGAAAGCAGGAGG-1" "AAACGAAAGGAAGAAC-1" "AAACGAAAGTCGACCC-1" ...</span><br><span># ..@ x       : num [1:115807238] 2 2 2 2 2 2 2 2 2 2 ...</span><br><span># ..@ factors : list()</span><br></code></pre><h3 data-tool="mdnice编辑器"><span></span><span>创建 seurat 对象</span><span></span></h3><blockquote data-tool="mdnice编辑器"><span>❝</span><p>需要三个文件：<strong>peak 矩阵</strong>，<strong>metadata</strong> 和 <strong>fragments 文件</strong>。</p></blockquote><pre data-tool="mdnice编辑器"><span></span><code><span># setRepositories(ind=1:3)</span><br><span># install.packages("Signac")</span><br><br><span>library</span>(Signac)<br><span>library</span>(Seurat)<br><span>library</span>(GenomicRanges)<br><span>library</span>(ggplot2)<br><span>library</span>(patchwork)<br><br>counts &lt;- Read10X_h5(filename = <span>"10k_pbmc_ATACv2_nextgem_Chromium_Controller_filtered_peak_bc_matrix.h5"</span>)<br><br>str(counts)<br><span># Formal class 'dgCMatrix' [package "Matrix"] with 6 slots</span><br><span># ..@ i       : int [1:115807238] 14 37 40 86 98 99 110 113 115 122 ...</span><br><span># ..@ p       : int [1:10247] 0 12605 17997 31303 45352 52467 66981 91458 102075 112905 ...</span><br><span># ..@ Dim     : int [1:2] 165434 10246</span><br><span># ..@ Dimnames:List of 2</span><br><span># .. ..$ : chr [1:165434] "chr1:9772-10660" "chr1:180712-181178" "chr1:181200-181607" "chr1:191183-192084" ...</span><br><span># .. ..$ : chr [1:10246] "AAACGAAAGAGAGGTA-1" "AAACGAAAGCAGGAGG-1" "AAACGAAAGGAAGAAC-1" "AAACGAAAGTCGACCC-1" ...</span><br><span># ..@ x       : num [1:115807238] 2 2 2 2 2 2 2 2 2 2 ...</span><br><span># ..@ factors : list()</span><br><br>class(counts)<br><br>metadata &lt;- read.csv(file = <span>"10k_pbmc_ATACv2_nextgem_Chromium_Controller_singlecell.csv"</span>,<br>                     header = <span>TRUE</span>,<br>                     row.names = <span>1</span>)<br><br>chrom_assay &lt;- CreateChromatinAssay(counts = counts,<br>                                    sep = c(<span>":"</span>, <span>"-"</span>),<br>                                    fragments = <span>"10k_pbmc_ATACv2_nextgem_Chromium_Controller_fragments.tsv.gz"</span>,<br>                                    min.cells = <span>10</span>,<br>                                    min.features = <span>200</span>)<br>chrom_assay<br><span># ChromatinAssay data with 165434 features for 10246 cells</span><br><span># Variable features: 0</span><br><span># Genome:</span><br><span>#   Annotation present: FALSE</span><br><span># Motifs present: FALSE</span><br><span># Fragment files: 1</span><br><br>pbmc &lt;- CreateSeuratObject(counts = chrom_assay,<br>                           assay = <span>"peaks"</span>,<br>                           meta.data = metadata)<br>pbmc<br><span># An object of class Seurat</span><br><span># 165434 features across 10246 samples within 1 assay</span><br><span># Active assay: peaks (165434 features, 0 variable features)</span><br><span># 2 layers present: counts, data</span><br></code></pre><p data-tool="mdnice编辑器"><strong>添加基因注释信息：</strong></p><pre data-tool="mdnice编辑器"><span></span><code><span>library</span>(AnnotationHub)<br>ah &lt;- AnnotationHub()<br><br><span># Search for the Ensembl 98 EnsDb for Homo sapiens on AnnotationHub</span><br>query(ah, <span>"EnsDb.Hsapiens.v98"</span>)<br><span># AnnotationHub with 1 record</span><br><span># # snapshotDate(): 2024-04-30</span><br><span># # names(): AH75011</span><br><span># # $dataprovider: Ensembl</span><br><span># # $species: Homo sapiens</span><br><span># # $rdataclass: EnsDb</span><br><span># # $rdatadateadded: 2019-05-02</span><br><span># # $title: Ensembl 98 EnsDb for Homo sapiens</span><br><span># # $description: Gene and protein annotations for Homo sapiens bas...</span><br><span># # $taxonomyid: 9606</span><br><span># # $genome: GRCh38</span><br><span># # $sourcetype: ensembl</span><br><span># # $sourceurl: http://www.ensembl.org</span><br><span># # $sourcesize: NA</span><br><span># # $tags: c("98", "AHEnsDbs", "Annotation", "EnsDb",</span><br><span># #   "Ensembl", "Gene", "Protein", "Transcript")</span><br><span># # retrieve record with 'object[["AH75011"]]'</span><br><br>ensdb_v98 &lt;- ah[[<span>"AH75011"</span>]]<br><br><span># extract gene annotations from EnsDb</span><br>annotations &lt;- GetGRangesFromEnsDb(ensdb = ensdb_v98)<br><br><span># change to UCSC style since the data was mapped to hg38</span><br>seqlevels(annotations) &lt;- paste0(<span>'chr'</span>, seqlevels(annotations))<br>genome(annotations) &lt;- <span>"hg38"</span><br><br><span># add the gene information to the object</span><br>Annotation(pbmc) &lt;- annotations<br></code></pre><h3 data-tool="mdnice编辑器"><span></span><span>质量控制</span><span></span></h3><blockquote data-tool="mdnice编辑器"><span>❝</span><p><strong>质量控制可以参考 bulk 的 ATAC-seq 的指标。</strong></p></blockquote><ul data-tool="mdnice编辑器"><li><section><strong>核小体结合模式：测序检测的片段长度分布类似于 bulk ATAC 的，在单个核小体，两个核小体，三个核小体等长度位置处有有峰。然后小于单个核小体的长度（147bp）应该是最多的（无核小体片段）。</strong></section></li></ul><figure data-tool="mdnice编辑器"><img data-imgfileid="100032164" data-ratio="0.4753694581280788" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBV7NejHsSLJicwjOKonzL36ziaWqPK9CrprCp5pHn0aRQfPWIgOLqFwPibA/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="812" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBV7NejHsSLJicwjOKonzL36ziaWqPK9CrprCp5pHn0aRQfPWIgOLqFwPibA/640?wx_fmt=png&amp;from=appmsg"></figure><ul data-tool="mdnice编辑器"><li><section><p><strong>转录起始位点（TSS）富集分数：转录起始位点 (TSS) 是基因表达中重要的功能区域，往往具有开放的染色质状态。开放的 TSS 会有较多的 ATAC-seq 片段覆盖。 低 TSS 富集分数表示实验富集能力较差，常见于样本污染、有死细胞标记的情况。</strong></p></section></li><li><section><p><strong>峰值片段总数：峰值片段总数表示一个细胞内落在已识别开放染色质区域（即 ATAC-seq 峰）的片段数量。这是测序深度和样本复杂性的一个直观指标。如果读取数太少，代表测序深度不足或样本质量较差，这些细胞可能需要被移除（假阳性较高）。极高峰值片段数量：可能代表双细胞 (doublet)、多核簇 (nuclei clump)、或者其他实验偏差，需要特别注意。</strong></p></section></li><li><section><p><strong>峰值片段比例 ：表示来自测序的所有片段中，有多少比例落在峰区域（染色质开放区）。该指标反映出染色质信号有效性。高比例：说明测序捕获了更多的真实起作用的开放染色质区域。低比例：可能代表测序数据中有较多背景噪声或技术偏差。</strong></p></section></li><li><section><p><strong>黑名单区域内的 reads 比例：某些基因组区域（称为黑名单区域）往往因测序技术导致的伪信号（如重复序列、结构变化区域等）特别高。ENCODE 提供的黑名单列出此类区域，帮助过滤虚假信号。高黑名单比例（异常）：表示该细胞有较多技术性噪声片段，可能是实验污染或系统偏差的表现。Blacklist 区域可以通过 ENCODE 提供的参考组，如人类 (hg19, hg38) 和小鼠 (mm9, mm10) 等。</strong></p></section></li></ul><figure data-tool="mdnice编辑器"><img data-imgfileid="100032169" data-ratio="0.7436241610738255" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBVNodib3zhsGyjNt6z1SSso1dLTfbialCG6DevB8ods7Kw9ziaB4QwE1BRQ/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="745" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBVNodib3zhsGyjNt6z1SSso1dLTfbialCG6DevB8ods7Kw9ziaB4QwE1BRQ/640?wx_fmt=png&amp;from=appmsg"></figure><p data-tool="mdnice编辑器">计算质控相关的指标：</p><pre data-tool="mdnice编辑器"><span></span><code><span># ===============================================================</span><br><span># qc</span><br><span># ===============================================================</span><br><br><span># compute nucleosome signal score per cell</span><br>pbmc &lt;- NucleosomeSignal(object = pbmc)<br><br><span># compute TSS enrichment score per cell</span><br>pbmc &lt;- TSSEnrichment(object = pbmc)<br><br><span># add fraction of reads in peaks</span><br>pbmc$pct_reads_in_peaks &lt;- pbmc$peak_region_fragments / pbmc$passed_filters * <span>100</span><br><br><span># add blacklist ratio</span><br>pbmc$blacklist_ratio &lt;- FractionCountsInRegion(<br>  object = pbmc,<br>  assay = <span>'peaks'</span>,<br>  regions = blacklist_hg38_unified<br>)<br><br><span># plot</span><br>DensityScatter(pbmc,<br>               x = <span>'nCount_peaks'</span>,<br>               y = <span>'TSS.enrichment'</span>,<br>               log_x = <span>TRUE</span>, quantiles = <span>TRUE</span>)<br></code></pre><figure data-tool="mdnice编辑器"><img data-imgfileid="100032171" data-ratio="0.6178571428571429" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBVH23HAedFIQbCpZkBSUlkc8k3YbQ10h4VvGhscEeSLXYjMDqP25epKg/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1400" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBVH23HAedFIQbCpZkBSUlkc8k3YbQ10h4VvGhscEeSLXYjMDqP25epKg/640?wx_fmt=png&amp;from=appmsg"></figure><p data-tool="mdnice编辑器">片段长度分布：</p><pre data-tool="mdnice编辑器"><span></span><code><span># fragment length</span><br>pbmc$nucleosome_group &lt;- ifelse(pbmc$nucleosome_signal &gt; <span>4</span>, <span>'NS &gt; 4'</span>, <span>'NS &lt; 4'</span>)<br>FragmentHistogram(object = pbmc, group.by = <span>'nucleosome_group'</span>)<br></code></pre><figure data-tool="mdnice编辑器"><img data-imgfileid="100032168" data-ratio="0.43504171632896305" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBV8LwcxzMibvibCZiaIeODpUcib0MtDg8ibZuV1rc5CWcEicnhGn5HjwUSeJnQ/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="839" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBV8LwcxzMibvibCZiaIeODpUcib0MtDg8ibZuV1rc5CWcEicnhGn5HjwUSeJnQ/640?wx_fmt=png&amp;from=appmsg"></figure><p data-tool="mdnice编辑器">其他指标：</p><pre data-tool="mdnice编辑器"><span></span><code><span># plot the distribution of each QC metric separately</span><br>VlnPlot(object = pbmc,<br>        features = c(<span>'nCount_peaks'</span>, <span>'TSS.enrichment'</span>, <span>'blacklist_ratio'</span>, <span>'nucleosome_signal'</span>, <span>'pct_reads_in_peaks'</span>),<br>        pt.size = <span>0.1</span>,<br>        ncol = <span>5</span>)<br></code></pre><figure data-tool="mdnice编辑器"><img data-imgfileid="100032173" data-ratio="0.431144683323649" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBVL1sYz5GdPQO2ic96rHcM8m1amDyiaugGbwg02gJXucPXCGKoibAEpNdlg/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1721" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBVL1sYz5GdPQO2ic96rHcM8m1amDyiaugGbwg02gJXucPXCGKoibAEpNdlg/640?wx_fmt=png&amp;from=appmsg"></figure><p data-tool="mdnice编辑器"><strong>过滤，设置阈值进行过滤细胞：</strong></p><pre data-tool="mdnice编辑器"><span></span><code><span># remove cells that are outliers for these QC metrics</span><br>pbmc &lt;- subset(x = pbmc,<br>               subset = nCount_peaks &gt; <span>9000</span> &amp;<br>                 nCount_peaks &lt; <span>100000</span> &amp;<br>                 pct_reads_in_peaks &gt; <span>40</span> &amp;<br>                 blacklist_ratio &lt; <span>0.01</span> &amp;<br>                 nucleosome_signal &lt; <span>4</span> &amp;<br>                 TSS.enrichment &gt; <span>4</span>)<br><br>pbmc<br><span># An object of class Seurat</span><br><span># 165434 features across 9651 samples within 1 assay</span><br><span># Active assay: peaks (165434 features, 0 variable features)</span><br><span># 2 layers present: counts, data</span><br></code></pre><h3 data-tool="mdnice编辑器"><span></span><span>归一化，特征选择和线性降维</span><span></span></h3><blockquote data-tool="mdnice编辑器"><span>❝</span><p><strong>类似与 scRNA-seq 里面的步骤，不过所用的函数不太一样：</strong></p></blockquote><hr data-tool="mdnice编辑器"><blockquote data-tool="mdnice编辑器"><span>❝</span><p><strong>1.TF-IDF 归一化：TF-IDF 是一种从信息检索中借来的方法，用于处理单细胞 ATAC-seq 数据的稀疏性问题。它对数据进行两步归一化，目的是调整测序深度的差异，并赋予罕见区域（peaks）较高的权重。</strong></p><ul><li><section><strong>细胞级归一化：每个细胞里的峰（peak）计数会按总片段数归一化，校正不同细胞之间测序深度的差异。这样可以确保峰的计数在不同的细胞中具有一定的可比性。</strong></section></li><li><section><strong>峰级归一化：通过对峰的稀有性进行加权，稀有峰（仅在较少细胞中检测到的峰）获得更高的值，而常见峰（几乎在每个细胞中都存在）值较低。</strong></section></li></ul></blockquote><blockquote data-tool="mdnice编辑器"><span>❝</span><p><strong>2.特征选择：在 scRNA-seq 中，我们通常使用基因表达的变异性（Variable Gene Selection）来筛选对细胞异质性贡献最大的基因作为“高变基因”。但在 scATAC-seq 中，由于信号稀疏（低动态范围），很难沿用这种选取方法，因此我们需要其他策略。</strong></p><ul><li><section><strong>使用一定比例的峰：可以选择顶端百分比的峰，比如 "前 25% 峰" （对应于 min.cutoff="q75"），即特征值最高的那部分峰。这是因为，在 ATAC-seq 中，峰（chromatin accessibility features）最相关或最独特的一小部分就足够用来捕获细胞间的主要差异。这种方法通常可以显著减少计算量，加快降维计算。</strong></section></li><li><section><strong>剔除低共同存在的峰：即去掉只存在于很少数量细胞中的峰。可使用 FindTopFeatures() 函数，在分析中剔除低质量或非常稀有的峰。</strong></section></li><li><section><strong>全部峰：也可以使用所有的峰（不进行任何过滤），虽然运行时间会更慢，但结果通常相似。</strong></section></li></ul></blockquote><blockquote data-tool="mdnice编辑器"><span>❝</span><p><strong>3.线性降维：SVD（奇异值分解）是一种经典的线性降维方法，它可以用来从高维稀疏数据中提取主要信号（本例中的峰占据的维度很多）。简单来说，SVD 会将输入的矩阵（这里是 TF-IDF 矩阵）分解成几个主要成分（components），每个成分表示主要的变化方向。</strong></p></blockquote><pre data-tool="mdnice编辑器"><span></span><code><span># ====================================================================</span><br><span># Normalization and linear dimensional reduction</span><br><span># ====================================================================</span><br><br>pbmc &lt;- RunTFIDF(pbmc)<br>pbmc &lt;- FindTopFeatures(pbmc, min.cutoff = <span>'q0'</span>)<br>pbmc &lt;- RunSVD(pbmc)<br></code></pre><ul data-tool="mdnice编辑器"><li><section><strong>LSI（Latent Semantic Indexing）：是基于奇异值分解（SVD）的降维方法。在降维后的低维度数据中，每一个 LSI 成分（component）都是数据某种变化模式的表征。</strong></section></li></ul><p data-tool="mdnice编辑器"><strong><em>Signac 提供的 DepthCor() 函数，这个函数计算每个 LSI 成分与每个细胞的测序深度（Depth，细胞中总片段数）之间的相关性。用来衡量两者之间相互影响的程度。由于 LSI 是从整个矩阵中提取主要的变化模式，第一主成分（LSI 第一个成分）常常捕捉到的是测序深度的差异，而非生物学信号。如果 LSI 的第一成分（或其他成分）主要由测序深度引起，这些技术噪声会对下游分析（如聚类、分类或可视化）造成干扰。如果 LSI 的某些成分（例如第 1 成分）与测序深度的相关系数较高，则说明这些成分主要由测序深度驱动。这些成分应该被移除，因为它们反映的是技术变量而不是生物学变量。</em></strong></p><pre data-tool="mdnice编辑器"><span></span><code>DepthCor(pbmc)<br></code></pre><figure data-tool="mdnice编辑器"><img data-imgfileid="100032170" data-ratio="0.508869179600887" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBV82xWyiaOufcHs3nofPq5sEicMb5DvZ8d3dVS9UJ5A2s1RqU1jrLoMGGQ/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="902" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBV82xWyiaOufcHs3nofPq5sEicMb5DvZ8d3dVS9UJ5A2s1RqU1jrLoMGGQ/640?wx_fmt=png&amp;from=appmsg"></figure><h3 data-tool="mdnice编辑器"><span></span><span>非线性降维、聚类</span><span></span></h3><p data-tool="mdnice编辑器"><strong>这里选择主成分不使用第一个：</strong></p><pre data-tool="mdnice编辑器"><span></span><code>pbmc &lt;- FindNeighbors(object = pbmc, reduction = <span>'lsi'</span>, dims = <span>2</span>:<span>30</span>)<br>pbmc &lt;- FindClusters(object = pbmc, verbose = <span>FALSE</span>, algorithm = <span>3</span>)<br>pbmc &lt;- RunUMAP(object = pbmc, reduction = <span>'lsi'</span>, dims = <span>2</span>:<span>30</span>)<br><br><span># plot</span><br>DimPlot(object = pbmc, label = <span>TRUE</span>) + NoLegend()<br></code></pre><figure data-tool="mdnice编辑器"><img data-imgfileid="100032175" data-ratio="0.8974820143884892" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBVI6WgcfORqHaM2eebLLbZf5icVW2EB1ZDVU9icvjxYvCp2tkria0xTNViaA/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="556" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBVI6WgcfORqHaM2eebLLbZf5icVW2EB1ZDVU9icvjxYvCp2tkria0xTNViaA/640?wx_fmt=png&amp;from=appmsg"></figure><h3 data-tool="mdnice编辑器"><span></span><span>创建基因活性矩阵</span><span></span></h3><blockquote data-tool="mdnice编辑器"><span>❝</span><p><strong>scATAC-seq 数据主要测量的是染色质的开放性，而非编码区域的功能远没有编码区域（基因）研究得清楚。这使得直接根据染色质的开放性来注释细胞群体更具挑战性。然而，某些染色质开放性的区域（如基因启动子区域的可及性）与基因表达强度有关。因此，我们可以间接通过开放性评估基因活性，进而实现基因层面的细胞群体注释。</strong></p></blockquote><p data-tool="mdnice编辑器"><strong>步骤：</strong></p><ul data-tool="mdnice编辑器"><li><section><p>获取基因坐标：将基因编码区域的坐标从基因组特征（annotation file, 如 GTF 文件）中提取出来。基因区域通常覆盖了基因体（gene body），并包含了上游 2kb 的启动子区域（promoter，这是基因转录调控的重要位置）。</p></section></li><li><section><p>统计与基因相关的片段数量：统计在每个基因区域（包括启动子区域）内有多少测序片段（fragments）。每个片段的计数代表该基因区域在相应的细胞中的“活性”，因为染色质开放性与基因活性相关。Fragment counting 可以通过 <strong>FeatureMatrix()</strong> 函数实现。</p></section></li><li><section><p>生成基因活性矩阵：对于每个基因 G 和每个细胞 C，计算矩阵中 <strong>G×C</strong> 的值，即：<strong>在基因 G 的基因区域及其启动子区域内的片段数量</strong>。输出类似 RNA-seq 基因表达矩阵的稀疏矩阵。</p></section></li></ul><blockquote data-tool="mdnice编辑器"><span>❝</span><p>上述所有步骤都可以通过 <strong>GeneActivity()</strong> 函数轻松实现：GeneActivity() 函数会：<strong>自动提取基因坐标</strong>。<strong>自动延伸基因坐标范围（包括延伸至上游 2kb 的启动子区域）</strong>。<strong>使用 FeatureMatrix() 函数统计片段，并生成基因活性矩阵</strong>。输出：<strong>返回一个以基因为行、以细胞为列的基因活性矩阵（Gene Activity Matrix）</strong>。该矩阵可以作为另一个“Assay”添加到 Seurat 对象中，用于进一步分析。</p></blockquote><pre data-tool="mdnice编辑器"><span></span><code><span># ====================================================================</span><br><span># Create a gene activity matrix</span><br><span># ====================================================================</span><br><br>gene.activities &lt;- GeneActivity(pbmc)<br><br>str(gene.activities)<br><span># Formal class 'dgCMatrix' [package "Matrix"] with 6 slots</span><br><span># ..@ i       : int [1:85029853] 0 2 4 6 8 9 10 12 13 14 ...</span><br><span># ..@ p       : int [1:9652] 0 8524 13848 23402 33423 39897 50602 62725 71537 79867 ...</span><br><span># ..@ Dim     : int [1:2] 19615 9651</span><br><span># ..@ Dimnames:List of 2</span><br><span># .. ..$ : chr [1:19615] "PLCXD1" "GTPBP6" "PPP2R3B" "SHOX" ...</span><br><span># .. ..$ : chr [1:9651] "AAACGAAAGAGAGGTA-1" "AAACGAAAGCAGGAGG-1" "AAACGAAAGGAAGAAC-1" "AAACGAAAGTCGACCC-1" ...</span><br><span># ..@ x       : num [1:85029853] 1 1 1 1 1 2 1 4 1 1 ...</span><br><span># ..@ factors : list()</span><br><br><span># add the gene activity matrix to the Seurat object as a new assay and normalize it</span><br>pbmc[[<span>'RNA'</span>]] &lt;- CreateAssayObject(counts = gene.activities)<br><br>pbmc &lt;- NormalizeData(<br>  object = pbmc,<br>  assay = <span>'RNA'</span>,<br>  normalization.method = <span>'LogNormalize'</span>,<br>  scale.factor = median(pbmc$nCount_RNA)<br>)<br><br><br>DefaultAssay(pbmc) &lt;- <span>'RNA'</span><br></code></pre><p data-tool="mdnice编辑器">可视化经典基因的 marker：</p><pre data-tool="mdnice编辑器"><span></span><code><span># plot</span><br>FeaturePlot(<br>  object = pbmc,<br>  features = c(<span>'MS4A1'</span>, <span>'CD3D'</span>, <span>'LEF1'</span>, <span>'NKG7'</span>, <span>'TREM1'</span>, <span>'LYZ'</span>),<br>  pt.size = <span>0.1</span>,<br>  max.cutoff = <span>'q95'</span>,<br>  ncol = <span>3</span><br>)<br></code></pre><figure data-tool="mdnice编辑器"><img data-imgfileid="100032177" data-ratio="0.5484351713859911" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBVbHYJt7YW0Dz7vr2Jmia5BpQ9RuCZldxibcCiasADmI7EP2trIQg3UJfbA/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1342" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBVbHYJt7YW0Dz7vr2Jmia5BpQ9RuCZldxibcCiasADmI7EP2trIQg3UJfbA/640?wx_fmt=png&amp;from=appmsg"></figure><h2 data-tool="mdnice编辑器"><span></span><span>与 scRNA-seq 数据整合分析</span><span></span></h2><blockquote data-tool="mdnice编辑器"><span>❝</span><p><strong>单细胞 ATAC-seq 数据主要反映染色质开放性，但其信号稀疏和复杂，使得直接注释细胞群体（尤其是细胞亚型）具有难度</strong>。染色质开放性与基因表达具有一定的相关性，但它依赖于假设，且存在噪声。仅基于 scATAC-seq 的基因活性矩阵可能不足以提供准确的生物学解释。<strong>借助与 scRNA-seq 数据的整合，我们可以使用 RNA 测序数据中更一致和清晰的基因表达信号，来辅助解释 ATAC-seq 数据中的细胞群体。</strong></p></blockquote><p data-tool="mdnice编辑器">这里加载已经处理好的数据：</p><pre data-tool="mdnice编辑器"><span></span><code><span># ====================================================================</span><br><span># Integrating with scRNA-seq data</span><br><span># ====================================================================</span><br><br><span># wget https://signac-objects.s3.amazonaws.com/pbmc_10k_v3.rds</span><br><br><span># Load the pre-processed scRNA-seq data for PBMCs</span><br>pbmc_rna &lt;- readRDS(<span>"pbmc_10k_v3.rds"</span>)<br>pbmc_rna &lt;- UpdateSeuratObject(pbmc_rna)<br>pbmc_rna<br><span># An object of class Seurat</span><br><span># 19089 features across 9432 samples within 1 assay</span><br><span># Active assay: RNA (19089 features, 3000 variable features)</span><br><span># 3 layers present: counts, data, scale.data</span><br><span># 3 dimensional reductions calculated: pca, tsne, umap</span><br></code></pre><p data-tool="mdnice编辑器"><strong>FindTransferAnchors</strong> 函数通过寻找共享特征或模式，在两个数据集之间构建对齐关系（<strong>“锚点”，anchors</strong>）：</p><pre data-tool="mdnice编辑器"><span></span><code>transfer.anchors &lt;- FindTransferAnchors(<br>  reference = pbmc_rna,<br>  query = pbmc,<br>  reduction = <span>'cca'</span><br>)<br><span># Running CCA</span><br><span># Merging objects</span><br><span># Finding neighborhoods</span><br><span># Finding anchors</span><br><span># Found 19099 anchors</span><br></code></pre><p data-tool="mdnice编辑器"><strong>TransferData()</strong> 函数为每个细胞添加预测的细胞类型及得分，然后把结果添加到 <strong>metadata</strong> 里面：</p><pre data-tool="mdnice编辑器"><span></span><code>predicted.labels &lt;- TransferData(<br>  anchorset = transfer.anchors,<br>  refdata = pbmc_rna$celltype,<br>  weight.reduction = pbmc[[<span>'lsi'</span>]],<br>  dims = <span>2</span>:<span>30</span><br>)<br><span># Finding integration vectors</span><br><span># Finding integration vector weights</span><br><span># 0%   10   20   30   40   50   60   70   80   90   100%</span><br><span># [----|----|----|----|----|----|----|----|----|----|</span><br><span># **************************************************|</span><br><span># Predicting cell labels</span><br><br>pbmc &lt;- AddMetaData(object = pbmc, metadata = predicted.labels)<br></code></pre><figure data-tool="mdnice编辑器"><img data-imgfileid="100032174" data-ratio="0.3609845031905196" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBVSogziaFKAfH55crIqY3DU4mqxibBKLn1EaFunms8SrfKDzGNprfj2nTA/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1097" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBVSogziaFKAfH55crIqY3DU4mqxibBKLn1EaFunms8SrfKDzGNprfj2nTA/640?wx_fmt=png&amp;from=appmsg"></figure><p data-tool="mdnice编辑器">可视化对比一下：</p><pre data-tool="mdnice编辑器"><span></span><code><span># plot</span><br>plot1 &lt;- DimPlot(object = pbmc_rna,<br>                 group.by = <span>'celltype'</span>,<br>                 label = <span>TRUE</span>,<br>                 repel = <span>TRUE</span>) +<br>  NoLegend() + ggtitle(<span>'scRNA-seq'</span>)<br><br>plot2 &lt;- DimPlot(object = pbmc,<br>                 group.by = <span>'predicted.id'</span>,<br>                 label = <span>TRUE</span>,<br>                 repel = <span>TRUE</span>) +<br>  NoLegend() + ggtitle(<span>'scATAC-seq'</span>)<br><br>plot1 + plot2<br></code></pre><figure data-tool="mdnice编辑器"><img data-imgfileid="100032176" data-ratio="0.5489815712900097" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBVicq1eicp31LhgCseRglb1yBK0mIYhOb6Q3IRJkQZ0iau3cSys3VeSRokA/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1031" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBVicq1eicp31LhgCseRglb1yBK0mIYhOb6Q3IRJkQZ0iau3cSys3VeSRokA/640?wx_fmt=png&amp;from=appmsg"></figure><hr data-tool="mdnice编辑器"><p data-tool="mdnice编辑器"><strong>上面在 scATAC</strong> 中可以看到有一群细胞被注释为 <strong>platelets（血小板）</strong>，scATAC-seq 的目标是检测细胞核中染色质的可及性（chromatin accessibility），<strong>而血小板没有细胞核</strong>。因此这里认为预测为血小板的细胞可能实际是 <strong>巨核细胞（megakaryocytes）</strong>，<strong>巨核细胞是血小板的前体细胞，存在于骨髓之中，负责产生血小板。</strong></p><p data-tool="mdnice编辑器">这里我们可以可视化一下每个细胞类型的预测分数，可以看到 <strong>platelets（血小板）</strong> 预测分数是比较低的：</p><pre data-tool="mdnice编辑器"><span></span><code>VlnPlot(pbmc, <span>'prediction.score.max'</span>, group.by = <span>'predicted.id'</span>)<br></code></pre><figure data-tool="mdnice编辑器"><img data-imgfileid="100032178" data-ratio="0.4077669902912621" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBVDQ3H5e7D7ew5zbu9ibkJI7BryLJJK8Q3pgXWMJ9ibpcRjrnTeibP3GhmQ/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1545" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBVDQ3H5e7D7ew5zbu9ibkJI7BryLJJK8Q3pgXWMJ9ibpcRjrnTeibP3GhmQ/640?wx_fmt=png&amp;from=appmsg"></figure><p data-tool="mdnice编辑器"><strong>platelets（血小板）</strong> 细胞数量也比较少，我们可以过滤一下这些细胞：</p><pre data-tool="mdnice编辑器"><span></span><code><span># filter</span><br>predicted_id_counts &lt;- table(pbmc$predicted.id)<br>predicted_id_counts<br><span># B cell progenitor        CD14+ Monocytes        CD16+ Monocytes             CD4 Memory              CD4 Naive           CD8 effector</span><br><span>#               294                   2542                    241                   2597                   1122                    971</span><br><span>#         CD8 Naive         Dendritic cell Double negative T cell              NK bright                 NK dim                    pDC</span><br><span>#               625                     67                    172                     44                    690                     31</span><br><span>#          Platelet             pre-B cell</span><br><span>#                10                    245</span><br><br><span># Identify the predicted.id values that have more than 20 cells</span><br>major_predicted_ids &lt;- names(predicted_id_counts[predicted_id_counts &gt; <span>20</span>])<br>pbmc &lt;- pbmc[, pbmc$predicted.id %<span>in</span>% major_predicted_ids]<br><br><span># change cell identities to the per-cell predicted labels</span><br>Idents(pbmc) &lt;- pbmc$predicted.id<br></code></pre><h2 data-tool="mdnice编辑器"><span></span><span>寻找染色质差异开放区域</span><span></span></h2><p data-tool="mdnice编辑器">基于 <strong>Wilcoxon test</strong> 方法差异分析比较慢，安装 <strong>presto</strong> 加速运行：</p><pre data-tool="mdnice编辑器"><span></span><code><span>if</span> (!requireNamespace(<span>"remotes"</span>, quietly = <span>TRUE</span>))<br>  install.packages(<span>'remotes'</span>)<br>remotes::install_github(<span>'immunogenomics/presto'</span>)<br></code></pre><p data-tool="mdnice编辑器">比如寻找 <strong>CD4 Naive</strong> 和 <strong>CD14+ Monocytes</strong> 细胞间的差异区域：</p><pre data-tool="mdnice编辑器"><span></span><code><span># ====================================================================</span><br><span># Find differentially accessible peaks between cell types</span><br><span># ====================================================================</span><br><br><span># change back to working with peaks instead of gene activities</span><br>DefaultAssay(pbmc) &lt;- <span>'peaks'</span><br><br><span># wilcox is the default option for test.use</span><br>da_peaks &lt;- FindMarkers(object = pbmc,<br>                        ident.1 = <span>"CD4 Naive"</span>,<br>                        ident.2 = <span>"CD14+ Monocytes"</span>,<br>                        test.use = <span>'wilcox'</span>,<br>                        min.pct = <span>0.1</span>)<br><br>head(da_peaks)<br><span>#                           p_val avg_log2FC pct.1 pct.2 p_val_adj</span><br><span># chr6-13302533-13303459        0  -5.320377 0.025 0.772         0</span><br><span># chr19-54207815-54208728       0  -4.314469 0.052 0.793         0</span><br><span># chr17-78198651-78199583       0  -5.571622 0.023 0.760         0</span><br><span># chr12-119988511-119989430     0   4.146181 0.783 0.090         0</span><br><span># chr7-142808530-142809435      0   3.654013 0.757 0.088         0</span><br><span># chr17-82126458-82127377       0   4.998374 0.704 0.043         0</span><br></code></pre><p data-tool="mdnice编辑器">可视化其中一个区域：</p><pre data-tool="mdnice编辑器"><span></span><code><span># plot</span><br>plot1 &lt;- VlnPlot(<br>  object = pbmc,<br>  features = rownames(da_peaks)[<span>1</span>],<br>  pt.size = <span>0.1</span>,<br>  idents = c(<span>"CD4 Naive"</span>,<span>"CD14+ Monocytes"</span>)<br>)<br><br>plot2 &lt;- FeaturePlot(<br>  object = pbmc,<br>  features = rownames(da_peaks)[<span>1</span>],<br>  pt.size = <span>0.1</span><br>)<br><br>plot1 | plot2<br></code></pre><figure data-tool="mdnice编辑器"><img data-imgfileid="100032181" data-ratio="0.5061124694376528" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBVqcSQvPrwnraic9Xt2FRoQBkOiaQnUvDrOjcmQuxib7l0sU7THTC1LuaGg/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1227" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBVqcSQvPrwnraic9Xt2FRoQBkOiaQnUvDrOjcmQuxib7l0sU7THTC1LuaGg/640?wx_fmt=png&amp;from=appmsg"></figure><p data-tool="mdnice编辑器">对 peak 区域进行注释，便于知道是哪个基因名称：</p><pre data-tool="mdnice编辑器"><span></span><code><span># annotate peaks</span><br>open_cd4naive &lt;- rownames(da_peaks[da_peaks$avg_log2FC &gt; <span>3</span>, ])<br>open_cd14mono &lt;- rownames(da_peaks[da_peaks$avg_log2FC &lt; -<span>3</span>, ])<br><br>closest_genes_cd4naive &lt;- ClosestFeature(pbmc, regions = open_cd4naive)<br>closest_genes_cd14mono &lt;- ClosestFeature(pbmc, regions = open_cd14mono)<br><br>head(closest_genes_cd4naive)<br><span>#                           tx_id gene_name         gene_id   gene_biotype type            closest_region              query_region distance</span><br><span># ENSE00002206071 ENST00000397558    BICDL1 ENSG00000135127 protein_coding exon chr12-119989869-119990297 chr12-119988511-119989430      438</span><br><span># ENST00000632998 ENST00000632998     PRSS2 ENSG00000275896 protein_coding  utr  chr7-142774509-142774564  chr7-142808530-142809435    33965</span><br><span># ENST00000665763 ENST00000665763    CCDC57 ENSG00000176155 protein_coding  gap   chr17-82101867-82127691   chr17-82126458-82127377        0</span><br><span># ENST00000645515 ENST00000645515  ATP6V0A4 ENSG00000105929 protein_coding  cds  chr7-138752625-138752837  chr7-138752283-138753197        0</span><br><span># ENST00000545320 ENST00000545320       CD6 ENSG00000013725 protein_coding  gap   chr11-60971915-60987907   chr11-60985909-60986801        0</span><br><span># ENST00000603548 ENST00000603548     FBXW7 ENSG00000109670 protein_coding  utr  chr4-152320544-152322880  chr4-152100248-152101142   219401</span><br></code></pre><p data-tool="mdnice编辑器">针对差异区域进行功能富集分析：</p><pre data-tool="mdnice编辑器"><span></span><code><span># GO enrichment analysis with clusterProfiler</span><br><span>library</span>(clusterProfiler)<br><span>library</span>(org.Hs.eg.db)<br><span>library</span>(enrichplot)<br><br>cd4naive_ego &lt;- enrichGO(gene = closest_genes_cd4naive$gene_id,<br>                         keyType = <span>"ENSEMBL"</span>,<br>                         OrgDb = org.Hs.eg.db,<br>                         ont = <span>"BP"</span>,<br>                         pAdjustMethod = <span>"BH"</span>,<br>                         pvalueCutoff = <span>0.05</span>,<br>                         qvalueCutoff = <span>0.05</span>,<br>                         readable = <span>TRUE</span>)<br><br>barplot(cd4naive_ego,showCategory = <span>10</span>)<br></code></pre><figure data-tool="mdnice编辑器"><img data-imgfileid="100032180" data-ratio="0.8813056379821959" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBVGjhFBkBTyEjQ970pYAhL3cs0ADAARPSDibMfayibIJc37TQKZN7A67qQ/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="674" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBVGjhFBkBTyEjQ970pYAhL3cs0ADAARPSDibMfayibIJc37TQKZN7A67qQ/640?wx_fmt=png&amp;from=appmsg"></figure><h2 data-tool="mdnice编辑器"><span></span><span>峰图展示差异可及性基因组区域</span><span></span></h2><p data-tool="mdnice编辑器"><strong>CoveragePlot() 可视化基因组区域的染色质可及性，生成类似于 bulk 的信号轨迹，并支持与基因注释、Peaks 和基因组连接等信息的整合分析：</strong></p><pre data-tool="mdnice编辑器"><span></span><code><span># ====================================================================</span><br><span># Plotting genomic regions</span><br><span># ====================================================================</span><br><br>pbmc &lt;- SortIdents(pbmc)<br><span># Creating pseudobulk profiles for 13 variables across 165434 features</span><br><span># Computing euclidean distance between pseudobulk profiles</span><br><span># Clustering distance matrix</span><br><br><span># find DA peaks overlapping gene of interest</span><br>regions_highlight &lt;- subsetByOverlaps(StringToGRanges(open_cd4naive), LookupGeneCoords(pbmc, <span>"CD4"</span>))<br><span># GRanges object with 1 range and 0 metadata columns:</span><br><span>#       seqnames          ranges strand</span><br><span>#          &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt;</span><br><span>#   [1]    chr12 6786415-6787326      *</span><br><span>#   -------</span><br><span>#   seqinfo: 24 sequences from an unspecified genome; no seqlengths</span><br><br>CoveragePlot(<br>  object = pbmc,<br>  region = <span>"CD4"</span>,<br>  region.highlight = regions_highlight,<br>  extend.upstream = <span>1000</span>,<br>  extend.downstream = <span>1000</span><br>) &amp; theme(axis.ticks.y = element_blank(),<br>          strip.text.y.left = element_text(hjust = <span>1</span>))<br></code></pre><figure data-tool="mdnice编辑器"><img data-imgfileid="100032179" data-ratio="1.1857142857142857" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBVsCOxicswPlYIm7uU3mibrlTAxwLDLwe7Rtjv0fFEjIibIbQMca3tapstQ/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="560" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBVsCOxicswPlYIm7uU3mibrlTAxwLDLwe7Rtjv0fFEjIibIbQMca3tapstQ/640?wx_fmt=png&amp;from=appmsg"></figure><h2 data-tool="mdnice编辑器"><span></span><span>motif（基序）分析</span><span></span></h2><blockquote data-tool="mdnice编辑器"><span>❝</span><p><strong>DNA motif</strong> 是一种短 DNA 序列的模式，通常代表转录因子的结合位点或其他调控序列元件。在单细胞 ATAC-seq 数据中，研究这些 motif 的位置和发生情况有助于理解哪些转录因子可能活跃，以及它们在调控基因表达中的可能作用。<strong>signac</strong> 有 <strong>两种 motif 分析</strong>：</p><ul><li><section><strong>差异开放性区域的 motif 富集分析：</strong> 判断某些 motif 是否显著富集在特定组别的染色质开放区域中，这可以帮助识别哪些转录因子在调控这些区域的基因表达。</section></li><li><section><strong>不同细胞群体间的 motif 活性分析：</strong> 在不同细胞群体之间比较 motif 的活跃程度，探索不同细胞类型或状态下的转录因子调控差异。</section></li></ul></blockquote><p data-tool="mdnice编辑器"><strong>首先添加 motif 信息到对象里：</strong></p><p data-tool="mdnice编辑器">Signac 提供了 <strong>AddMotifs()</strong> 函数来将 DNA 序列 motif 信息添加到一个 Seurat 对象中。包括 <strong>位置频率矩阵(PFM,Position Frequency Matrix)</strong> 或者 <strong>位置权重矩阵(PWM, Position Weight Matrix)</strong> ：</p><pre data-tool="mdnice编辑器"><span></span><code><span># ====================================================================</span><br><span># Motif analysis with Signac</span><br><span># ====================================================================</span><br><span># BiocManager::install("JASPAR2020")</span><br><span># BiocManager::install("TFBSTools")</span><br><span># install.packages("motifmatchr")</span><br><span># BiocManager::install("BSgenome.Hsapiens.UCSC.hg38")</span><br><span>library</span>(JASPAR2020)<br><span>library</span>(TFBSTools)<br><span>library</span>(BSgenome.Hsapiens.UCSC.hg38)<br><br><span># Adding motif information to the Seurat object</span><br><span># Get a list of motif position frequency matrices from the JASPAR database</span><br>pfm &lt;- getMatrixSet(x = JASPAR2020,<br>                    opts = list(collection = <span>"CORE"</span>,<br>                                tax_group = <span>'vertebrates'</span>,<br>                                all_versions = <span>FALSE</span>))<br><br><span># add motif information</span><br>pbmc &lt;- AddMotifs(object = pbmc,<br>                  genome = BSgenome.Hsapiens.UCSC.hg38,<br>                  pfm = pfm)<br></code></pre><h3 data-tool="mdnice编辑器"><span></span><span>寻找差异染色质开放区域富集的 motif</span><span></span></h3><blockquote data-tool="mdnice编辑器"><span>❝</span><p>使用 <strong>FindMarkers()</strong> 函数对单细胞 ATAC-seq 数据进行初步分析，以找到在两组细胞之间差异显著的 peaks（即染色质开放性不同的区域）。对于 ATAC-seq 数据，由于稀疏性，<strong>min.pct</strong> 参数这个阈值<strong>通常需要降低</strong>（如 0.01 或更低），以捕获更多差异性 peaks。<strong>FindMotifs()</strong> 函数来对差异的 peaks 来寻找富集的 motif：</p></blockquote><pre data-tool="mdnice编辑器"><span></span><code><span># ====================================================================</span><br><span># Finding overrepresented motifs</span><br><span># ====================================================================</span><br><br>da_peaks &lt;- FindMarkers(<br>  object = pbmc,<br>  ident.1 = <span>"CD4 Naive"</span>,<br>  ident.2 = <span>"CD14+ Monocytes"</span>,<br>  only.pos = <span>TRUE</span>,<br>  test.use = <span>'LR'</span>,<br>  min.pct = <span>0.05</span>,<br>  latent.vars = <span>'nCount_peaks'</span><br>)<br><br><span># get top differentially accessible peaks</span><br>top.da.peak &lt;- rownames(da_peaks[da_peaks$p_val &lt; <span>0.005</span> &amp; da_peaks$pct.1 &gt; <span>0.2</span>, ])<br><br><span># test enrichment</span><br>enriched.motifs &lt;- FindMotifs(object = pbmc,<br>                              features = top.da.peak)<br></code></pre><p data-tool="mdnice编辑器">可视化前几个 motif（<strong>这里因为运行速度很慢，我就粘贴一下官网的示例图</strong>）：</p><pre data-tool="mdnice编辑器"><span></span><code>MotifPlot(object = mouse_brain,<br>          motifs = head(rownames(enriched.motifs)))<br></code></pre><figure data-tool="mdnice编辑器"><img data-imgfileid="100032182" data-ratio="0.6178571428571429" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBVMXj8Iic17BzH0g3JxvfBQE7Q89hczHfhKdeUwrZ2TLQmtuSlCsWjF7w/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1400" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBVMXj8Iic17BzH0g3JxvfBQE7Q89hczHfhKdeUwrZ2TLQmtuSlCsWjF7w/640?wx_fmt=png&amp;from=appmsg"></figure><h3 data-tool="mdnice编辑器"><span></span><span>计算 motif 活性来寻找差异的 motif</span><span></span></h3><blockquote data-tool="mdnice编辑器"><span>❝</span><p>利用 <strong>chromVAR</strong> 算法计算每个细胞上的 <strong>motif 活性得分（motif activity score）</strong>。</p></blockquote><p data-tool="mdnice编辑器">这里记得去除对象里在 <strong>BSgenome.Hsapiens.UCSC.hg38</strong> 里没有的染色体，否则会报错：</p><pre data-tool="mdnice编辑器"><span></span><code><span># ====================================================================</span><br><span># Computing motif activities</span><br><span># ====================================================================</span><br><span># devtools::install_github("GreenleafLab/chromVAR")</span><br><span>library</span>(chromVAR)<br><br><span># remove these sequences that are not present in the BSgenome</span><br>features.keep &lt;- as.character(seqnames(granges(pbmc))) %<span>in</span>% standardChromosomes(granges(pbmc))<br><br>pbmc &lt;- pbmc[features.keep, ]<br><br><span>library</span>(BiocParallel)<br><span># accelarate running</span><br>register(SerialParam())<br><br>pbmc &lt;- RunChromVAR(object = pbmc,<br>                    genome = BSgenome.Hsapiens.UCSC.hg38)<br><span># Computing GC bias per region</span><br><span># Selecting background regions</span><br><span># Computing deviations from background</span><br><span># ...</span><br><br>DefaultAssay(pbmc) &lt;- <span>'chromvar'</span><br><br><span># look at the activity of xx</span><br>FeaturePlot(object = pbmc,<br>            features = <span>"MA0497.1"</span>,<br>            min.cutoff = <span>'q10'</span>,<br>            max.cutoff = <span>'q90'</span>,<br>            pt.size = <span>0.1</span>)<br></code></pre><p data-tool="mdnice编辑器"><strong>同样速度也很慢，就不展示了。</strong></p><h2 data-tool="mdnice编辑器"><span></span><span>Footprinting 分析</span><span></span></h2><blockquote data-tool="mdnice编辑器"><span>❝</span><p>scATAC-seq（单细胞染色质可及性测序）的<strong>Footprinting</strong>分析是一种重要分析方法，用来深入研究 DNA 结合蛋白（如转录因子）的结合位点特征及其调控作用。<strong>在开放染色质区域（peaks），部分 DNA 片段因被转录因子或其他结合蛋白保护，不容易被 Tn5 转座酶切割。这种保护会在染色质可及性测序中表现为特定区域的可及性减少，因此会形成一个突然下降的凹槽，称为"footprint"（足迹）。</strong></p></blockquote><p data-tool="mdnice编辑器">我们可以使用 <strong>Footprint()</strong> 函数来进行分析，使用 <strong>PlotFootprint()</strong> 函数来进行可视化，注意在进行 <strong>Footprinting</strong>分析前需要添加对应物种的 motif 矩阵到对象里，即 <strong>getMatrixSet</strong> 和 <strong>AddMotifs</strong> 处理后：</p><pre data-tool="mdnice编辑器"><span></span><code><span># ===================================================================</span><br><span># Motif footprinting</span><br><span># ===================================================================</span><br><span># extract position frequency matrices for the motifs</span><br>pwm &lt;- getMatrixSet(x = JASPAR2020,<br>                    opts = list(species = <span>9606</span>, all_versions = <span>FALSE</span>))<br><br><span># add motif information</span><br>pbmc &lt;- AddMotifs(pbmc, genome = BSgenome.Hsapiens.UCSC.hg38, pfm = pwm)<br></code></pre><p data-tool="mdnice编辑器">随便看几个，可视化：</p><pre data-tool="mdnice编辑器"><span></span><code><span># check names</span><br>mtif_name &lt;- pbmc@assays$peaks@motifs@motif.names<br><br>motif_p &lt;- names(mtif_name[match(c(<span>"GATA2"</span>, <span>"CEBPA"</span>, <span>"EBF1"</span>,<span>"KLF4"</span>),mtif_name)])<br>motif_p<br><span># [1] "MA0036.3" "MA0102.4" "MA0154.4" "MA0039.4"</span><br><br><span># gather the footprinting information for sets of motifs</span><br>pbmc &lt;- Footprint(object = pbmc,<br>                  motif.name = motif_p,<br>                  genome = BSgenome.Hsapiens.UCSC.hg38)<br><br><br><span># plot the footprint data for each group of cells</span><br>PlotFootprint(pbmc, features = motif_p) +<br>  patchwork::plot_layout(ncol = <span>2</span>)<br></code></pre><figure data-tool="mdnice编辑器"><img data-imgfileid="100032183" data-ratio="0.5697211155378487" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBVndfnDKO7dVYGDkeoIFibFpuaiapz1bSQvTggxvSkHM2e59pHFrSF841g/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1255" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBVndfnDKO7dVYGDkeoIFibFpuaiapz1bSQvTggxvSkHM2e59pHFrSF841g/640?wx_fmt=png&amp;from=appmsg"></figure><h2 data-tool="mdnice编辑器"><span></span><span>结尾</span><span></span></h2><blockquote data-tool="mdnice编辑器"><span>❝</span><p><strong>路漫漫其修远兮,吾将上下而求索。</strong></p></blockquote><hr data-tool="mdnice编辑器"><p data-tool="mdnice编辑器">欢迎加入生信交流群。加我微信我也拉你进 <strong>微信群聊</strong> <strong>老俊俊生信交流群</strong> <strong>(微信交流群需收取 20 元入群费用,一旦交费,拒不退还!(防止骗子和便于管理))</strong> 。</p><figure data-tool="mdnice编辑器"><img data-imgfileid="100032188" data-ratio="0.6083707025411061" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBVEeM5xvhNKicUjpjDzDu7Mvic0Pafjc2yCpz6RQLxPyyylKYtuicvbWHfw/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="669" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usez7Of7VQBJk3Poz4tElibRBVEeM5xvhNKicUjpjDzDu7Mvic0Pafjc2yCpz6RQLxPyyylKYtuicvbWHfw/640?wx_fmt=png&amp;from=appmsg"></figure><hr data-tool="mdnice编辑器"><p data-tool="mdnice编辑器"><strong></strong></p><center data-tool="mdnice编辑器"><strong> 往期回顾目录 </strong></center><p data-tool="mdnice编辑器"><br></p><blockquote data-tool="mdnice编辑器"><span>❝</span><p><strong></strong></p><center><strong><a href="https://mp.weixin.qq.com/s?__biz=MzkyMTI1MTYxNA==&amp;mid=2247515726&amp;idx=1&amp;sn=0fb48c324f7207d95fb1e5c774b4fd1a&amp;token=2079312128&amp;lang=zh_CN&amp;scene=21#wechat_redirect" data-linktype="2">单细胞数据类型读取一文解决！</a></strong></center><strong><center><a href="https://mp.weixin.qq.com/s?__biz=MzkyMTI1MTYxNA==&amp;mid=2247515672&amp;idx=1&amp;sn=3878299038763a15e2503340942d9660&amp;token=902807353&amp;lang=zh_CN&amp;scene=21#wechat_redirect" data-linktype="2">从 gtf 文件构建 orgdb 和 txdb 数据库</a></center></strong><strong><center><a href="https://mp.weixin.qq.com/s?__biz=MzkyMTI1MTYxNA==&amp;mid=2247515655&amp;idx=1&amp;sn=03c1bf4a081fb561468a4f3c40331615&amp;token=1970442110&amp;lang=zh_CN&amp;scene=21#wechat_redirect" data-linktype="2">easybio 基于 CellMarker2.0 数据库进行亚群注释</a></center></strong><strong><center><a href="https://mp.weixin.qq.com/s?__biz=MzkyMTI1MTYxNA==&amp;mid=2247515609&amp;idx=1&amp;sn=2f81a1d58af368bc0a261a59f4026652&amp;token=1970442110&amp;lang=zh_CN&amp;scene=21#wechat_redirect" data-linktype="2">写一个 R 版本的 ORFfinder</a></center></strong><strong><center><a href="https://mp.weixin.qq.com/s?__biz=MzkyMTI1MTYxNA==&amp;mid=2247515569&amp;idx=1&amp;sn=8838873a67cd108744a6206f1c9e3972&amp;chksm=c18487c0f6f30ed66faed2ec61160fb8fe8466fc650e32560a16051f032264864f577b04fb29&amp;token=837342039&amp;lang=zh_CN&amp;scene=21#wechat_redirect" data-linktype="2">TCGA+GTEX 数据联合分析</a></center></strong><strong><center><a href="https://mp.weixin.qq.com/s?__biz=MzkyMTI1MTYxNA==&amp;mid=2247515543&amp;idx=1&amp;sn=49d6991c5d610fbe3fda761b7421934d&amp;chksm=c18487e6f6f30ef002f9e7d88bb19c2fed2de330786d1ad69d36c713b444f19b64aca79e27c2&amp;token=837342039&amp;lang=zh_CN&amp;scene=21#wechat_redirect" data-linktype="2">SNAF 新型计算工具鉴定广谱癌症新抗原</a></center></strong><strong><center><a href="https://mp.weixin.qq.com/s?__biz=MzkyMTI1MTYxNA==&amp;mid=2247515529&amp;idx=1&amp;sn=b047eb05ccbce0a9ad6ee734b4188416&amp;chksm=c18487f8f6f30eee3c027ac7c5f9b0babcc2fe737dfc2a6c5236e906ec6853b1f508e184af5f&amp;token=2137498509&amp;lang=zh_CN&amp;scene=21#wechat_redirect" data-linktype="2">听说你想给 visCluster 添加行注释？</a></center></strong><strong><center><a href="https://mp.weixin.qq.com/s?__biz=MzkyMTI1MTYxNA==&amp;mid=2247515507&amp;idx=1&amp;sn=bfd0f92a6dbad64dd41d8242e4868132&amp;chksm=c1848702f6f30e140bbe092108d9eeadf4e9e7b9ca13fa33a5575bfc778a7d2af01c982a622f&amp;token=2137498509&amp;lang=zh_CN&amp;scene=21#wechat_redirect" data-linktype="2">Julia 中也可以使用 ggplot？</a></center></strong><strong><center><a href="https://mp.weixin.qq.com/s?__biz=MzkyMTI1MTYxNA==&amp;mid=2247515491&amp;idx=1&amp;sn=d2cfe70986882a8cbc4aecb23fbc1dad&amp;chksm=c1848712f6f30e0413b5808b4cb284b7a98b4e374812d71b8e7e1f3278a3308d6ea1e8a7600d&amp;token=697561698&amp;lang=zh_CN&amp;scene=21#wechat_redirect" data-linktype="2">trackc：HiC 数据可视化神器！</a></center></strong><strong><center><a href="https://mp.weixin.qq.com/s?__biz=MzkyMTI1MTYxNA==&amp;mid=2247515447&amp;idx=1&amp;sn=d8e01ea8fe0b37ee430142ea058a0f9d&amp;chksm=c1848746f6f30e50d75249f9a198946d211e08317960ce7388f2dde95db27d3f9351195aa211&amp;token=697561698&amp;lang=zh_CN&amp;scene=21#wechat_redirect" data-linktype="2">scplotter：单细胞可视化神包！</a></center></strong><p><br></p></blockquote></section><p><br></p><p><mp-style-type data-value="3"></mp-style-type></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/Y0E7FOy9en0SeMt3tlQRlw",target="_blank" rel="noopener noreferrer">原文链接</a>
