---
title: "scATAC 多样本整合"
date: 2024-12-18T09:06:22Z
draft: ["false"]
tags: [
  "fetched",
  "老俊俊的生信笔记"
]
categories: ["Acdemic"]
---
scATAC 多样本整合 by 老俊俊的生信笔记
------
<div><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com" data-mpa-powered-by="yiban.io"><section><mp-common-profile data-pluginname="mpprofile" data-id="MzkyMTI1MTYxNA==" data-headimg="http://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usezgsqIGqjITSMggCTSoViaYeoKe2xoZr1IIvNJoztibQxibYHLDDoiabwAc6Ggws3Tvdo8EPss2nLgaVQ/0?wx_fmt=png" data-nickname="老俊俊的生信笔记" data-alias="JunJunLab" data-signature="老俊俊的生信技能和知识分享,我不是巨人,但你可以站在我的肩膀上更进一步!" data-from="0" data-is_biz_ban="0" data-service_type="1"></mp-common-profile></section><h2 data-tool="mdnice编辑器"><span></span><span>引言</span><span></span></h2><blockquote data-tool="mdnice编辑器"><span>❝</span><p>跟着 <strong>signac</strong> 官网教程顺一下。<strong>如何合并多个样本的 scATAC 数据</strong>。</p></blockquote><h2 data-tool="mdnice编辑器"><span></span><span>数据下载</span><span></span></h2><p data-tool="mdnice编辑器"><strong>包含不同细胞数量的 PBMC 的 scATAC 数据：</strong></p><figure data-tool="mdnice编辑器"><img data-imgfileid="100032487" data-ratio="0.1880597014925373" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usezscUFu1RjQZ3nY9bS02aoKsA0V0OOUuwe23DAh74tRR8NUmzlFBlibXF2etwnjh1u3QbG5s1BicNsw/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="670" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usezscUFu1RjQZ3nY9bS02aoKsA0V0OOUuwe23DAh74tRR8NUmzlFBlibXF2etwnjh1u3QbG5s1BicNsw/640?wx_fmt=png&amp;from=appmsg"></figure><pre data-tool="mdnice编辑器"><span></span><code><span># 500 cell</span><br>wget https://cf.10xgenomics.com/samples/cell-atac/<span>1.1</span><span>.0</span>/atac_pbmc_500_nextgem/atac_pbmc_500_nextgem_fragments.tsv.gz<br>wget https://cf.10xgenomics.com/samples/cell-atac/<span>1.1</span><span>.0</span>/atac_pbmc_500_nextgem/atac_pbmc_500_nextgem_fragments.tsv.gz.tbi<br>wget https://cf.10xgenomics.com/samples/cell-atac/<span>1.1</span><span>.0</span>/atac_pbmc_500_nextgem/atac_pbmc_500_nextgem_peaks.bed<br>wget https://cf.10xgenomics.com/samples/cell-atac/<span>1.1</span><span>.0</span>/atac_pbmc_500_nextgem/atac_pbmc_500_nextgem_singlecell.csv<br>wget https://cf.10xgenomics.com/samples/cell-atac/<span>1.1</span><span>.0</span>/atac_pbmc_500_nextgem/atac_pbmc_500_nextgem_filtered_peak_bc_matrix.h5<br><br><span># 1k cell</span><br>wget https://cf.10xgenomics.com/samples/cell-atac/<span>1.1</span><span>.0</span>/atac_pbmc_1k_nextgem/atac_pbmc_1k_nextgem_fragments.tsv.gz<br>wget https://cf.10xgenomics.com/samples/cell-atac/<span>1.1</span><span>.0</span>/atac_pbmc_1k_nextgem/atac_pbmc_1k_nextgem_fragments.tsv.gz.tbi<br>wget https://cf.10xgenomics.com/samples/cell-atac/<span>1.1</span><span>.0</span>/atac_pbmc_1k_nextgem/atac_pbmc_1k_nextgem_peaks.bed<br>wget https://cf.10xgenomics.com/samples/cell-atac/<span>1.1</span><span>.0</span>/atac_pbmc_1k_nextgem/atac_pbmc_1k_nextgem_singlecell.csv<br>wget https://cf.10xgenomics.com/samples/cell-atac/<span>1.1</span><span>.0</span>/atac_pbmc_1k_nextgem/atac_pbmc_1k_nextgem_filtered_peak_bc_matrix.h5<br><br><span># 5k cell</span><br>wget https://cf.10xgenomics.com/samples/cell-atac/<span>1.1</span><span>.0</span>/atac_pbmc_5k_nextgem/atac_pbmc_5k_nextgem_fragments.tsv.gz<br>wget https://cf.10xgenomics.com/samples/cell-atac/<span>1.1</span><span>.0</span>/atac_pbmc_5k_nextgem/atac_pbmc_5k_nextgem_fragments.tsv.gz.tbi<br>wget https://cf.10xgenomics.com/samples/cell-atac/<span>1.1</span><span>.0</span>/atac_pbmc_5k_nextgem/atac_pbmc_5k_nextgem_peaks.bed<br>wget https://cf.10xgenomics.com/samples/cell-atac/<span>1.1</span><span>.0</span>/atac_pbmc_5k_nextgem/atac_pbmc_5k_nextgem_singlecell.csv<br>wget https://cf.10xgenomics.com/samples/cell-atac/<span>1.1</span><span>.0</span>/atac_pbmc_5k_nextgem/atac_pbmc_5k_nextgem_filtered_peak_bc_matrix.h5<br><br><span># 10k cell</span><br>wget https://cf.10xgenomics.com/samples/cell-atac/<span>1.1</span><span>.0</span>/atac_pbmc_10k_nextgem/atac_pbmc_10k_nextgem_fragments.tsv.gz<br>wget https://cf.10xgenomics.com/samples/cell-atac/<span>1.1</span><span>.0</span>/atac_pbmc_10k_nextgem/atac_pbmc_10k_nextgem_fragments.tsv.gz.tbi<br>wget https://cf.10xgenomics.com/samples/cell-atac/<span>1.1</span><span>.0</span>/atac_pbmc_10k_nextgem/atac_pbmc_10k_nextgem_peaks.bed<br>wget https://cf.10xgenomics.com/samples/cell-atac/<span>1.1</span><span>.0</span>/atac_pbmc_10k_nextgem/atac_pbmc_10k_nextgem_singlecell.csv<br>wget https://cf.10xgenomics.com/samples/cell-atac/<span>1.1</span><span>.0</span>/atac_pbmc_10k_nextgem/atac_pbmc_10k_nextgem_filtered_peak_bc_matrix.h5<br></code></pre><h2 data-tool="mdnice编辑器"><span></span><span>创建共有的 peaks 集合</span><span></span></h2><blockquote data-tool="mdnice编辑器"><span>❝</span><p>因为每个样本检测到的开放区域可能不同，因此在合并前需要创建一个共有的 peak 集合，可以使用 <strong>GenomicRanges</strong> 包的 <strong>reduce</strong> 或者 <strong>disjoin</strong> 函数来处理 peaks：</p></blockquote><pre data-tool="mdnice编辑器"><span></span><code>gr &lt;- GRanges(seqnames = <span>"chr1"</span>, ranges = IRanges(start = c(<span>20</span>, <span>70</span>, <span>300</span>), end = c(<span>120</span>, <span>200</span>, <span>400</span>)))<br>gr<br><span>## GRanges object with 3 ranges and 0 metadata columns:</span><br><span>##       seqnames    ranges strand</span><br><span>##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span><br><span>##   [1]     chr1    20-120      *</span><br><span>##   [2]     chr1    70-200      *</span><br><span>##   [3]     chr1   300-400      *</span><br><span>##   -------</span><br><span>##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span><br></code></pre><figure data-tool="mdnice编辑器"><img data-imgfileid="100032488" data-ratio="0.16666666666666666" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usezscUFu1RjQZ3nY9bS02aoKALlgnXgjtzRDicwokHQtdaPoIL3bVjEVkQSeaYZ84ttcURTeiac3E8ibQ/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1152" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usezscUFu1RjQZ3nY9bS02aoKALlgnXgjtzRDicwokHQtdaPoIL3bVjEVkQSeaYZ84ttcURTeiac3E8ibQ/640?wx_fmt=png&amp;from=appmsg"></figure><figure data-tool="mdnice编辑器"><img data-imgfileid="100032485" data-ratio="0.16666666666666666" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usezscUFu1RjQZ3nY9bS02aoKhHhxmZl2MqCibM8p7niag6RicWWsKnaXjXGPHKOPIlC6So6zrolnUPzxA/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1152" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usezscUFu1RjQZ3nY9bS02aoKhHhxmZl2MqCibM8p7niag6RicWWsKnaXjXGPHKOPIlC6So6zrolnUPzxA/640?wx_fmt=png&amp;from=appmsg"></figure><figure data-tool="mdnice编辑器"><img data-imgfileid="100032486" data-ratio="0.16666666666666666" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usezscUFu1RjQZ3nY9bS02aoK9lwZtyjf3LicGWEdOcqVpkibT2tvJQZ0jExFjj7fjzyy5qmS7tq1m1Vw/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1152" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usezscUFu1RjQZ3nY9bS02aoK9lwZtyjf3LicGWEdOcqVpkibT2tvJQZ0jExFjj7fjzyy5qmS7tq1m1Vw/640?wx_fmt=png&amp;from=appmsg"></figure><p data-tool="mdnice编辑器">读取 peaks 文件，然后使用 <strong>reduce</strong> 取并集：</p><pre data-tool="mdnice编辑器"><span></span><code><span>library</span>(Signac)<br><span>library</span>(Seurat)<br><span>library</span>(GenomicRanges)<br><span>library</span>(future)<br><br>plan(<span>"multicore"</span>, workers = <span>4</span>)<br>options(future.globals.maxSize = <span>50000</span> * <span>1024</span>^<span>2</span>) <span># for 50 Gb RAM</span><br><br><span># read in peak sets</span><br>peaks.500 &lt;- read.table(<br>  file = <span>"pbmc500/atac_pbmc_500_nextgem_peaks.bed"</span>,<br>  col.names = c(<span>"chr"</span>, <span>"start"</span>, <span>"end"</span>)<br>)<br>peaks.1k &lt;- read.table(<br>  file = <span>"pbmc1k/atac_pbmc_1k_nextgem_peaks.bed"</span>,<br>  col.names = c(<span>"chr"</span>, <span>"start"</span>, <span>"end"</span>)<br>)<br>peaks.5k &lt;- read.table(<br>  file = <span>"pbmc5k/atac_pbmc_5k_nextgem_peaks.bed"</span>,<br>  col.names = c(<span>"chr"</span>, <span>"start"</span>, <span>"end"</span>)<br>)<br>peaks.10k &lt;- read.table(<br>  file = <span>"pbmc10k/atac_pbmc_10k_nextgem_peaks.bed"</span>,<br>  col.names = c(<span>"chr"</span>, <span>"start"</span>, <span>"end"</span>)<br>)<br><br><span># convert to genomic ranges</span><br>gr.500 &lt;- makeGRangesFromDataFrame(peaks.500)<br>gr.1k &lt;- makeGRangesFromDataFrame(peaks.1k)<br>gr.5k &lt;- makeGRangesFromDataFrame(peaks.5k)<br>gr.10k &lt;- makeGRangesFromDataFrame(peaks.10k)<br><br><span># Create a unified set of peaks to quantify in each dataset</span><br>combined.peaks &lt;- reduce(x = c(gr.500, gr.1k, gr.5k, gr.10k))<br><br><span># Filter out bad peaks based on length</span><br>peakwidths &lt;- width(combined.peaks)<br>combined.peaks &lt;- combined.peaks[peakwidths  &lt; <span>10000</span> &amp; peakwidths &gt; <span>20</span>]<br>combined.peaks<br><span>## GRanges object with 89951 ranges and 0 metadata columns:</span><br><span>##           seqnames            ranges strand</span><br><span>##              &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt;</span><br><span>##       [1]     chr1     565153-565499      *</span><br><span>##       [2]     chr1     569185-569620      *</span><br><span>##       [3]     chr1     713551-714783      *</span><br><span>##       [4]     chr1     752418-753020      *</span><br><span>##       [5]     chr1     762249-763345      *</span><br><span>##       ...      ...               ...    ...</span><br><span>##   [89947]     chrY 23422151-23422632      *</span><br><span>##   [89948]     chrY 23583994-23584463      *</span><br><span>##   [89949]     chrY 23602466-23602779      *</span><br><span>##   [89950]     chrY 28816593-28817710      *</span><br><span>##   [89951]     chrY 58855911-58856251      *</span><br><span>##   -------</span><br><span>##   seqinfo: 24 sequences from an unspecified genome; no seqlengths</span><br></code></pre><h2 data-tool="mdnice编辑器"><span></span><span>创建 Fragment 对象</span><span></span></h2><blockquote data-tool="mdnice编辑器"><span>❝</span><p>使用 <strong>CreateFragmentObject</strong> 函数为每个样本创建 Fragment 对象：</p></blockquote><p data-tool="mdnice编辑器">首先读取 metadata 做一些过滤：</p><pre data-tool="mdnice编辑器"><span></span><code><span># load metadata</span><br>md.500 &lt;- read.table(<br>  file = <span>"pbmc500/atac_pbmc_500_nextgem_singlecell.csv"</span>,<br>  stringsAsFactors = <span>FALSE</span>,<br>  sep = <span>","</span>,<br>  header = <span>TRUE</span>,<br>  row.names = <span>1</span><br>)[-<span>1</span>, ] <span># remove the first row</span><br><br>md.1k &lt;- read.table(<br>  file = <span>"pbmc1k/atac_pbmc_1k_nextgem_singlecell.csv"</span>,<br>  stringsAsFactors = <span>FALSE</span>,<br>  sep = <span>","</span>,<br>  header = <span>TRUE</span>,<br>  row.names = <span>1</span><br>)[-<span>1</span>, ]<br><br>md.5k &lt;- read.table(<br>  file = <span>"pbmc5k/atac_pbmc_5k_nextgem_singlecell.csv"</span>,<br>  stringsAsFactors = <span>FALSE</span>,<br>  sep = <span>","</span>,<br>  header = <span>TRUE</span>,<br>  row.names = <span>1</span><br>)[-<span>1</span>, ]<br><br>md.10k &lt;- read.table(<br>  file = <span>"pbmc10k/atac_pbmc_10k_nextgem_singlecell.csv"</span>,<br>  stringsAsFactors = <span>FALSE</span>,<br>  sep = <span>","</span>,<br>  header = <span>TRUE</span>,<br>  row.names = <span>1</span><br>)[-<span>1</span>, ]<br><br><span># perform an initial filtering of low count cells</span><br>md.500 &lt;- md.500[md.500$passed_filters &gt; <span>500</span>, ]<br>md.1k &lt;- md.1k[md.1k$passed_filters &gt; <span>500</span>, ]<br>md.5k &lt;- md.5k[md.5k$passed_filters &gt; <span>500</span>, ]<br>md.10k &lt;- md.10k[md.10k$passed_filters &gt; <span>1000</span>, ] <span># sequenced deeper so set higher cutoff</span><br></code></pre><p data-tool="mdnice编辑器">开始创建：</p><pre data-tool="mdnice编辑器"><span></span><code><span># create fragment objects</span><br>frags.500 &lt;- CreateFragmentObject(<br>  path = <span>"pbmc500/atac_pbmc_500_nextgem_fragments.tsv.gz"</span>,<br>  cells = rownames(md.500)<br>)<br><br>frags.1k &lt;- CreateFragmentObject(<br>  path = <span>"pbmc1k/atac_pbmc_1k_nextgem_fragments.tsv.gz"</span>,<br>  cells = rownames(md.1k)<br>)<br><br>frags.5k &lt;- CreateFragmentObject(<br>  path = <span>"pbmc5k/atac_pbmc_5k_nextgem_fragments.tsv.gz"</span>,<br>  cells = rownames(md.5k)<br>)<br><br>frags.10k &lt;- CreateFragmentObject(<br>  path = <span>"pbmc10k/atac_pbmc_10k_nextgem_fragments.tsv.gz"</span>,<br>  cells = rownames(md.10k)<br>)<br></code></pre><h2 data-tool="mdnice编辑器"><span></span><span>创建 peaks x cell 矩阵</span><span></span></h2><blockquote data-tool="mdnice编辑器"><span>❝</span><p>使用 <strong>FeatureMatrix</strong> 对前面得到的共有 peak 进行定量，得到 <strong>peaks x cell</strong> 矩阵：</p></blockquote><pre data-tool="mdnice编辑器"><span></span><code>pbmc500.counts &lt;- FeatureMatrix(<br>  fragments = frags.500,<br>  features = combined.peaks,<br>  cells = rownames(md.500)<br>)<br><br>pbmc1k.counts &lt;- FeatureMatrix(<br>  fragments = frags.1k,<br>  features = combined.peaks,<br>  cells = rownames(md.1k)<br>)<br><br>pbmc5k.counts &lt;- FeatureMatrix(<br>  fragments = frags.5k,<br>  features = combined.peaks,<br>  cells = rownames(md.5k)<br>)<br><br>pbmc10k.counts &lt;- FeatureMatrix(<br>  fragments = frags.10k,<br>  features = combined.peaks,<br>  cells = rownames(md.10k)<br>)<br></code></pre><h2 data-tool="mdnice编辑器"><span></span><span>创建 seurat 对象</span><span></span></h2><blockquote data-tool="mdnice编辑器"><span>❝</span><p>有了上面数据，就可以创建<strong>seurat</strong>对象了：</p></blockquote><pre data-tool="mdnice编辑器"><span></span><code>pbmc500_assay &lt;- CreateChromatinAssay(pbmc500.counts, fragments = frags.500)<br>pbmc500 &lt;- CreateSeuratObject(pbmc500_assay, assay = <span>"ATAC"</span>, meta.data=md.500)<br><br>pbmc1k_assay &lt;- CreateChromatinAssay(pbmc1k.counts, fragments = frags.1k)<br>pbmc1k &lt;- CreateSeuratObject(pbmc1k_assay, assay = <span>"ATAC"</span>, meta.data=md.1k)<br><br>pbmc5k_assay &lt;- CreateChromatinAssay(pbmc5k.counts, fragments = frags.5k)<br>pbmc5k &lt;- CreateSeuratObject(pbmc5k_assay, assay = <span>"ATAC"</span>, meta.data=md.5k)<br><br>pbmc10k_assay &lt;- CreateChromatinAssay(pbmc10k.counts, fragments = frags.10k)<br>pbmc10k &lt;- CreateSeuratObject(pbmc10k_assay, assay = <span>"ATAC"</span>, meta.data=md.10k)<br></code></pre><h2 data-tool="mdnice编辑器"><span></span><span>合并样本走标准流程</span><span></span></h2><blockquote data-tool="mdnice编辑器"><span>❝</span><p>使用 <strong>merge</strong> 函数合并多样本，然后走标准分析流程：</p></blockquote><pre data-tool="mdnice编辑器"><span></span><code><span># add information to identify dataset of origin</span><br>pbmc500$dataset &lt;- <span>'pbmc500'</span><br>pbmc1k$dataset &lt;- <span>'pbmc1k'</span><br>pbmc5k$dataset &lt;- <span>'pbmc5k'</span><br>pbmc10k$dataset &lt;- <span>'pbmc10k'</span><br><br><span># merge all datasets, adding a cell ID to make sure cell names are unique</span><br>combined &lt;- merge(<br>  x = pbmc500,<br>  y = list(pbmc1k, pbmc5k, pbmc10k),<br>  add.cell.ids = c(<span>"500"</span>, <span>"1k"</span>, <span>"5k"</span>, <span>"10k"</span>)<br>)<br>combined[[<span>"ATAC"</span>]]<br><span>## ChromatinAssay data with 89951 features for 21688 cells</span><br><span>## Variable features: 0</span><br><span>## Genome:</span><br><span>## Annotation present: FALSE</span><br><span>## Motifs present: FALSE</span><br><span>## Fragment files: 4</span><br></code></pre><p data-tool="mdnice编辑器">下游流程：</p><pre data-tool="mdnice编辑器"><span></span><code>combined &lt;- RunTFIDF(combined)<br>combined &lt;- FindTopFeatures(combined, min.cutoff = <span>20</span>)<br>combined &lt;- RunSVD(combined)<br>combined &lt;- RunUMAP(combined, dims = <span>2</span>:<span>50</span>, reduction = <span>'lsi'</span>)<br>DimPlot(combined, group.by = <span>'dataset'</span>, pt.size = <span>0.1</span>)<br></code></pre><figure data-tool="mdnice编辑器"><img data-imgfileid="100032489" data-ratio="0.6178571428571429" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usezscUFu1RjQZ3nY9bS02aoKLQgHQaAViac9pDicibUSon20U0XUmcsklyN6GrxzOWr3SbzLddSPiaGhtw/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1400" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usezscUFu1RjQZ3nY9bS02aoKLQgHQaAViac9pDicibUSon20U0XUmcsklyN6GrxzOWr3SbzLddSPiaGhtw/640?wx_fmt=png&amp;from=appmsg"></figure><pre data-tool="mdnice编辑器"><span></span><code>CoveragePlot(<br>  object = combined,<br>  group.by = <span>'dataset'</span>,<br>  region = <span>"chr14-99700000-99760000"</span><br>)<br></code></pre><figure data-tool="mdnice编辑器"><img data-imgfileid="100032492" data-ratio="0.5485714285714286" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usezscUFu1RjQZ3nY9bS02aoKtc4ogxICqqibZ6iabSNDloUr0I0iapBibM1InZMbaG2jArAjJjLhMzPiaOw/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1400" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usezscUFu1RjQZ3nY9bS02aoKtc4ogxICqqibZ6iabSNDloUr0I0iapBibM1InZMbaG2jArAjJjLhMzPiaOw/640?wx_fmt=png&amp;from=appmsg"></figure><h2 data-tool="mdnice编辑器"><span></span><span>没有 fragment 文件情况</span><span></span></h2><blockquote data-tool="mdnice编辑器"><span>❝</span><p>有时候如果你没有 <strong>fragment</strong> 文件，也可以直接合并多个样本，<strong>merge 函数内部会对有重叠的区域自动进行拓展成等宽的 peak，但是这种操作可能会带来一些不准确的问题</strong>。因此还是建议始终优先基于<strong>片段文件</strong>进行<strong>统一特征定义与重新量化</strong>。</p></blockquote><pre data-tool="mdnice编辑器"><span></span><code><span># load the count matrix for each object that was generated by cellranger</span><br>counts.500 &lt;- Read10X_h5(<span>"pbmc500/atac_pbmc_500_nextgem_filtered_peak_bc_matrix.h5"</span>)<br>counts.1k &lt;- Read10X_h5(<span>"pbmc1k/atac_pbmc_1k_nextgem_filtered_peak_bc_matrix.h5"</span>)<br>counts.5k &lt;- Read10X_h5(<span>"pbmc5k/atac_pbmc_5k_nextgem_filtered_peak_bc_matrix.h5"</span>)<br>counts.10k &lt;- Read10X_h5(<span>"pbmc10k/atac_pbmc_10k_nextgem_filtered_peak_bc_matrix.h5"</span>)<br><br><span># create objects</span><br>pbmc500_assay &lt;- CreateChromatinAssay(counts = counts.500, sep = c(<span>":"</span>, <span>"-"</span>), min.features = <span>500</span>)<br>pbmc500 &lt;- CreateSeuratObject(pbmc500_assay, assay = <span>"peaks"</span>)<br>pbmc1k_assay &lt;- CreateChromatinAssay(counts = counts.1k, sep = c(<span>":"</span>, <span>"-"</span>), min.features = <span>500</span>)<br>pbmc1k &lt;- CreateSeuratObject(pbmc1k_assay, assay = <span>"peaks"</span>)<br>pbmc5k_assay &lt;- CreateChromatinAssay(counts = counts.5k, sep = c(<span>":"</span>, <span>"-"</span>), min.features = <span>500</span>)<br>pbmc5k &lt;- CreateSeuratObject(pbmc5k_assay, assay = <span>"peaks"</span>)<br>pbmc10k_assay &lt;- CreateChromatinAssay(counts = counts.10k, sep = c(<span>":"</span>, <span>"-"</span>), min.features = <span>1000</span>)<br>pbmc10k &lt;- CreateSeuratObject(pbmc10k_assay, assay = <span>"peaks"</span>)<br><br><span># add information to identify dataset of origin</span><br>pbmc500$dataset &lt;- <span>'pbmc500'</span><br>pbmc1k$dataset &lt;- <span>'pbmc1k'</span><br>pbmc5k$dataset &lt;- <span>'pbmc5k'</span><br>pbmc10k$dataset &lt;- <span>'pbmc10k'</span><br><br><span># merge</span><br>combined &lt;- merge(<br>  x = pbmc500,<br>  y = list(pbmc1k, pbmc5k, pbmc10k),<br>  add.cell.ids = c(<span>"500"</span>, <span>"1k"</span>, <span>"5k"</span>, <span>"10k"</span>)<br>)<br><br><span># process</span><br>combined &lt;- RunTFIDF(combined)<br>combined &lt;- FindTopFeatures(combined, min.cutoff = <span>20</span>)<br>combined &lt;- RunSVD(combined)<br>combined &lt;- RunUMAP(combined, dims = <span>2</span>:<span>50</span>, reduction = <span>'lsi'</span>)<br><br>DimPlot(combined, group.by = <span>'dataset'</span>, pt.size = <span>0.1</span>)<br></code></pre><figure data-tool="mdnice编辑器"><img data-imgfileid="100032494" data-ratio="0.6178571428571429" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usezscUFu1RjQZ3nY9bS02aoK0WhuRA0GNTpbJZxtkurmsjdlCsC24SHb0ZgIQMfHro5tHGQWgmxSgQ/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1400" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usezscUFu1RjQZ3nY9bS02aoK0WhuRA0GNTpbJZxtkurmsjdlCsC24SHb0ZgIQMfHro5tHGQWgmxSgQ/640?wx_fmt=png&amp;from=appmsg"></figure><h2 data-tool="mdnice编辑器"><span></span><span>结尾</span><span></span></h2><blockquote data-tool="mdnice编辑器"><span>❝</span><p><strong>路漫漫其修远兮,吾将上下而求索。</strong></p></blockquote><hr data-tool="mdnice编辑器"><p data-tool="mdnice编辑器">欢迎加入生信交流群。加我微信我也拉你进 <strong>微信群聊</strong> <strong>老俊俊生信交流群</strong> <strong>(微信交流群需收取 20 元入群费用,一旦交费,拒不退还!(防止骗子和便于管理))</strong> 。</p><figure data-tool="mdnice编辑器"><img data-imgfileid="100032493" data-ratio="0.6083707025411061" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usezscUFu1RjQZ3nY9bS02aoKyBQ8HWZJK1ke13mhde4SicJ2Cgm2MF5pV5pk4zViba0QIIrwAdp96Seg/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="669" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usezscUFu1RjQZ3nY9bS02aoKyBQ8HWZJK1ke13mhde4SicJ2Cgm2MF5pV5pk4zViba0QIIrwAdp96Seg/640?wx_fmt=png&amp;from=appmsg"></figure><hr data-tool="mdnice编辑器"><p data-tool="mdnice编辑器"><strong></strong></p><center data-tool="mdnice编辑器"><strong> 往期回顾目录</strong></center><blockquote data-tool="mdnice编辑器"><span>❝</span><center><strong><a href="https://mp.weixin.qq.com/s?__biz=MzkyMTI1MTYxNA==&amp;mid=2247516132&amp;idx=1&amp;sn=ce65838712aa2e5e4c3c5b7dcfdb71b4&amp;token=1001473693&amp;lang=zh_CN&amp;scene=21#wechat_redirect" data-linktype="2">单细胞亚群注释相关工具</a></strong></center><strong><center><a href="https://mp.weixin.qq.com/s?__biz=MzkyMTI1MTYxNA==&amp;mid=2247516072&amp;idx=1&amp;sn=f82a27c6936f38c3a078ea259178216d&amp;token=1001473693&amp;lang=zh_CN&amp;scene=21#wechat_redirect" data-linktype="2">听说你想美化 signac 的 track 图？</a></center></strong><strong><center><a href="https://mp.weixin.qq.com/s?__biz=MzkyMTI1MTYxNA==&amp;mid=2247516054&amp;idx=1&amp;sn=4315e02f23804d9578464dac4c9d5646&amp;token=291087746&amp;lang=zh_CN&amp;scene=21#wechat_redirect" data-linktype="2">scanpy 处理单细胞数据</a></center></strong><strong><center><a href="https://mp.weixin.qq.com/s?__biz=MzkyMTI1MTYxNA==&amp;mid=2247516004&amp;idx=1&amp;sn=82611ea509826412aaf3b606e865d61a&amp;token=350842983&amp;lang=zh_CN&amp;scene=21#wechat_redirect" data-linktype="2">snapATAC2 分析教程</a></center></strong><strong><center><a href="https://mp.weixin.qq.com/s?__biz=MzkyMTI1MTYxNA==&amp;mid=2247515887&amp;idx=1&amp;sn=f12b573b99aa6bf20c5aae995f5a57e1&amp;token=350842983&amp;lang=zh_CN&amp;scene=21#wechat_redirect" data-linktype="2">scATAC 教程一文详解</a></center></strong><strong><center><a href="https://mp.weixin.qq.com/s?__biz=MzkyMTI1MTYxNA==&amp;mid=2247515726&amp;idx=1&amp;sn=0fb48c324f7207d95fb1e5c774b4fd1a&amp;token=2079312128&amp;lang=zh_CN&amp;scene=21#wechat_redirect" data-linktype="2">单细胞数据类型读取一文解决！</a></center></strong><strong><center><a href="https://mp.weixin.qq.com/s?__biz=MzkyMTI1MTYxNA==&amp;mid=2247515672&amp;idx=1&amp;sn=3878299038763a15e2503340942d9660&amp;token=902807353&amp;lang=zh_CN&amp;scene=21#wechat_redirect" data-linktype="2">从 gtf 文件构建 orgdb 和 txdb 数据库</a></center></strong><strong><center><a href="https://mp.weixin.qq.com/s?__biz=MzkyMTI1MTYxNA==&amp;mid=2247515655&amp;idx=1&amp;sn=03c1bf4a081fb561468a4f3c40331615&amp;token=1970442110&amp;lang=zh_CN&amp;scene=21#wechat_redirect" data-linktype="2">easybio 基于 CellMarker2.0 数据库进行亚群注释</a></center></strong><strong><center><a href="https://mp.weixin.qq.com/s?__biz=MzkyMTI1MTYxNA==&amp;mid=2247515609&amp;idx=1&amp;sn=2f81a1d58af368bc0a261a59f4026652&amp;token=1970442110&amp;lang=zh_CN&amp;scene=21#wechat_redirect" data-linktype="2">写一个 R 版本的 ORFfinder</a></center></strong><strong><center><a href="https://mp.weixin.qq.com/s?__biz=MzkyMTI1MTYxNA==&amp;mid=2247515569&amp;idx=1&amp;sn=8838873a67cd108744a6206f1c9e3972&amp;chksm=c18487c0f6f30ed66faed2ec61160fb8fe8466fc650e32560a16051f032264864f577b04fb29&amp;token=837342039&amp;lang=zh_CN&amp;scene=21#wechat_redirect" data-linktype="2">TCGA+GTEX 数据联合分析</a></center></strong></blockquote></section><p><mp-style-type data-value="3"></mp-style-type></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/hcyqkN6hzt0kOc2gdjUDHA",target="_blank" rel="noopener noreferrer">原文链接</a>
