---
title: "使用decoupleR一次性实现11种基因集的活性打分（R与Python我都要）"
date: 2024-12-04T13:58:24Z
draft: ["false"]
tags: [
  "fetched",
  "生信技能树"
]
categories: ["Acdemic"]
---
使用decoupleR一次性实现11种基因集的活性打分（R与Python我都要） by 生信技能树
------
<div><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com"><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com"><blockquote data-tool="mdnice编辑器"><span></span><p>关于<span><strong>Python和R编程语言的优先级</strong></span>的讨论实在是太多了， 小朋友才做选择，成年人就是全都学！我们有：<a target="_blank" href="https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247535452&amp;idx=1&amp;sn=2da733a67e6511e9210753c65e8a73b1&amp;scene=21#wechat_redirect" textvalue="掌握Python，解锁单细胞数据的无限可能" linktype="text" imgurl="" imgdata="null" data-itemshowtype="0" tab="innerlink" data-linktype="2">掌握Python，解锁单细胞数据的无限可能</a>，也有：<a target="_blank" href="https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247535420&amp;idx=2&amp;sn=7f5bb012c70f342460c44bdc47c28218&amp;scene=21#wechat_redirect" textvalue="生信入门&amp;数据挖掘线上直播课12月班" linktype="text" imgurl="" imgdata="null" data-itemshowtype="0" tab="innerlink" data-linktype="2">生信入门&amp;数据挖掘线上直播课12月班</a>，掌握多种编程语言更方便应对各种生物信息学数据分析场合！</p></blockquote></section><h3 data-tool="mdnice编辑器"><span></span></h3><h3 data-tool="mdnice编辑器"><span>就是因为考虑到绝大部分小伙伴是Python和R编程语言的二选一，</span><span>所以为了自己的工具使用更广泛，很多开发者会特意分发不同版本的软件</span><span>。其中对基因集等进行活性打分推断的包实在是太多了，但是同时开发了R代码和python代码的应该不多见，且集合了11种打分方法，我们推荐decoupleR：</span></h3><p data-tool="mdnice编辑器"><strong><span>官网github链接：https://github.com/saezlab/decoupleR</span></strong></p><p data-tool="mdnice编辑器">文献信息：</p><blockquote data-tool="mdnice编辑器"><p>Badia-i-Mompel P.,。。。。<strong>2022. decoupleR: ensemble of computational methods to infer biological activities from omics data. Bioinformatics Advance</strong>s. https://doi.org/10.1093/bioadv/vbac016</p></blockquote><figure data-tool="mdnice编辑器"><img data-imgfileid="100051830" data-ratio="0.48703703703703705" data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwVXxKHmPFTITLe10IUeiaHOSpoW72XPSYo3K5VibhSvRccgY1Bj6vPyopNqDaw2iahhSJc68Uz4PDlQ/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwVXxKHmPFTITLe10IUeiaHOSpoW72XPSYo3K5VibhSvRccgY1Bj6vPyopNqDaw2iahhSJc68Uz4PDlQ/640?wx_fmt=png&amp;from=appmsg"><figcaption> </figcaption></figure><p data-tool="mdnice编辑器">11个被囊括的包如下：</p><figure data-tool="mdnice编辑器"><img data-imgfileid="100051827" data-ratio="0.7670588235294118" data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwVXxKHmPFTITLe10IUeiaHOy7ladTk9oH3evVSsib3SNQfWhAicJ5jaR1MwhciaAseYAWIgvvRms7sEA/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="850" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwVXxKHmPFTITLe10IUeiaHOy7ladTk9oH3evVSsib3SNQfWhAicJ5jaR1MwhciaAseYAWIgvvRms7sEA/640?wx_fmt=png&amp;from=appmsg"><figcaption> image-20241202163856344</figcaption></figure><h3 data-tool="mdnice编辑器">decoupleR包的功能：</h3><ul data-tool="mdnice编辑器"><li><section>bulk RNA-seq 中通路活性推断</section></li><li><section>scRNA-seq 中通路活性推断</section></li><li><section>bulk RNA-seq 中转录因子活性推断</section></li><li><section>scRNA-seq 中转录因子活性推断</section></li><li><section>激酶和转录因子活性估计</section></li></ul><h3 data-tool="mdnice编辑器">安装此包：</h3><pre data-tool="mdnice编辑器"><code>options(BioC_mirror=<span>"https://mirrors.westlake.edu.cn/bioconductor"</span>)<br>options(<span>"repos"</span>=c(CRAN=<span>"https://mirrors.westlake.edu.cn/CRAN/"</span>))<br><br>install.packages(<span>'remotes'</span>)<br>remotes::install_github(<span>'saezlab/decoupleR'</span>)<br></code></pre><p data-tool="mdnice编辑器">下面来简单看一下其中一个功能：<span>Pathway activity inference in bulk RNA-seq</span></p><p data-tool="mdnice编辑器"><span>https://saezlab.github.io/decoupleR/articles/pw_bk.html</span></p><h3 data-tool="mdnice编辑器">小试牛刀：bulk RNA-seq 中通路活性推断</h3><p data-tool="mdnice编辑器">在这个笔记本中，我们展示了如何使用decoupleR对一个bulk RNA测序数据集进行通路活性推断，该数据集中胰腺癌细胞系中的转录因子FOXA2被敲除。</p><p data-tool="mdnice编辑器">数据包括<strong>3个野生型（WT）样本和3个敲除型（KO）样本，GEO编号为：https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE119931，是一个表观调控的项目，其中转录组数据部分样品信息如下所示：</strong></p><p><img data-galleryid="" data-imgfileid="100051845" data-ratio="0.9052823315118397" data-s="300,640" data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwVXxKHmPFTITLe10IUeiaHOO66zSAZokdMdLldVoGldStZicrSIBdyWLjRaibrJibmziaQjd9icYkHTdkQ/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1098" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwVXxKHmPFTITLe10IUeiaHOO66zSAZokdMdLldVoGldStZicrSIBdyWLjRaibrJibmziaQjd9icYkHTdkQ/640?wx_fmt=png&amp;from=appmsg"></p><h4 data-tool="mdnice编辑器">1、首先加载示例数据</h4><p data-tool="mdnice编辑器">作者已经对此数据集进行了预处理，并放到了包中：</p><pre data-tool="mdnice编辑器"><code><span># 加载</span><br>inputs_dir &lt;- system.file(<span>"extdata"</span>, package = <span>"decoupleR"</span>)<br>data &lt;- readRDS(file.path(inputs_dir, <span>"bk_data.rds"</span>))<br><br><span># 简单探索</span><br>class(data)<br>str(data)<br></code></pre><p data-tool="mdnice编辑器">可以看到数据为一个list对象，包含三个数据：counts值（为the normalized log-transformed counts）、design分组矩阵、limma_ttop差异基因结果</p><pre data-tool="mdnice编辑器"><code>List of 3<br> $ counts    : spc_tbl_ [11,093 × 7] (S3: spec_tbl_df/tbl_df/tbl/data.frame)<br>  ..$ gene              : chr [1:11093] <span>"NOC2L"</span> <span>"PLEKHN1"</span> <span>"PERM1"</span> <span>"ISG15"</span> ...<br>  ..$ PANC1.WT.Rep1     : num [1:11093] 10.05 7.54 6.28 10.94 6.96 ...<br>  ..$ PANC1.WT.Rep2     : num [1:11093] 11.95 8.13 6.42 11.47 7.2 ...<br>  ..$ PANC1.WT.Rep3     : num [1:11093] 12.06 8.71 6.59 11.43 7.52 ...<br>  ..$ PANC1.FOXA2KO.Rep1: num [1:11093] 12.31 8.05 6.29 11.55 7.06 ...<br>  ..$ PANC1.FOXA2KO.Rep2: num [1:11093] 12.14 8.29 6.49 11.37 7.49 ...<br>  ..$ PANC1.FOXA2KO.Rep3: num [1:11093] 11.49 8.62 6.78 11.18 7.07 ...<br>  ..- attr(*, <span>"spec"</span>)=<br>  .. .. cols(<br>  .. ..   gene = col_character(),<br>  .. ..   PANC1.WT.Rep1 = col_double(),<br>  .. ..   PANC1.WT.Rep2 = col_double(),<br>  .. ..   PANC1.WT.Rep3 = col_double(),<br>  .. ..   PANC1.FOXA2KO.Rep1 = col_double(),<br>  .. ..   PANC1.FOXA2KO.Rep2 = col_double(),<br>  .. ..   PANC1.FOXA2KO.Rep3 = col_double()<br>  .. .. )<br> $ design    : spc_tbl_ [6 × 2] (S3: spec_tbl_df/tbl_df/tbl/data.frame)<br>  ..$ sample   : chr [1:6] <span>"PANC1.WT.Rep1"</span> <span>"PANC1.WT.Rep2"</span> <span>"PANC1.WT.Rep3"</span> <span>"PANC1.FOXA2KO.Rep1"</span> ...<br>  ..$ condition: chr [1:6] <span>"PANC1.WT"</span> <span>"PANC1.WT"</span> <span>"PANC1.WT"</span> <span>"PANC1.FOXA2KO"</span> ...<br>  ..- attr(*, <span>"spec"</span>)=<br>  .. .. cols(<br>  .. ..   sample = col_character(),<br>  .. ..   condition = col_character()<br>  .. .. )<br> $ limma_ttop: spc_tbl_ [11,093 × 7] (S3: spec_tbl_df/tbl_df/tbl/data.frame)<br>  ..$ ID       : chr [1:11093] <span>"RHBDL2"</span> <span>"PLEKHH2"</span> <span>"HEG1"</span> <span>"CLU"</span> ...<br>  ..$ logFC    : num [1:11093] -1.82 -1.57 -1.73 -1.79 2.09 ...<br>  ..$ AveExpr  : num [1:11093] 8.6 7.7 8.64 12.22 9.61 ...<br>  ..$ t        : num [1:11093] -12.81 -10.79 -9.79 -9.76 8.95 ...<br>  ..$ P.Value  : num [1:11093] 0.00000303 0.00000993 0.0000194 0.00001976 0.00003552 ...<br>  ..$ adj.P.Val: num [1:11093] 0.0336 0.0548 0.0548 0.0548 0.0788 ...<br>  ..$ B        : num [1:11093] 3.95 3.27 2.84 2.83 2.43 ...<br>  ..- attr(*, <span>"spec"</span>)=<br>  .. .. cols(<br>  .. ..   ID = col_character(),<br>  .. ..   logFC = col_double(),<br>  .. ..   AveExpr = col_double(),<br>  .. ..   t = col_double(),<br>  .. ..   P.Value = col_double(),<br>  .. ..   adj.P.Val = col_double(),<br>  .. ..   B = col_double()<br>  .. .. )<br></code></pre><p data-tool="mdnice编辑器">提取count：</p><pre data-tool="mdnice编辑器"><code><span># Remove NAs and set row names</span><br>counts &lt;- data<span>$counts</span> %&gt;%<br>  dplyr::mutate_if(~ any(is.na(.x)), <br>                   ~ dplyr::if_else(is.na(.x), 0, .x)) %&gt;% <br>  tibble::column_to_rownames(var = <span>"gene"</span>) %&gt;% <br>  as.matrix()<br><br>head(counts)<br></code></pre><figure data-tool="mdnice编辑器"><img data-imgfileid="100051829" data-ratio="0.1446360153256705" data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwVXxKHmPFTITLe10IUeiaHODTUShCbGhO6GSagStl5ey6w4m75UqPrLZ5ycy2wUxOmzPwedfYEXZA/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1044" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwVXxKHmPFTITLe10IUeiaHODTUShCbGhO6GSagStl5ey6w4m75UqPrLZ5ycy2wUxOmzPwedfYEXZA/640?wx_fmt=png&amp;from=appmsg"><figcaption> </figcaption></figure><p data-tool="mdnice编辑器">提取The design meta-data:</p><pre data-tool="mdnice编辑器"><code>design &lt;- data<span>$design</span><br>design<br></code></pre><figure data-tool="mdnice编辑器"><img data-imgfileid="100051828" data-ratio="0.26598837209302323" data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwVXxKHmPFTITLe10IUeiaHOmkSmrF6LmPwS8qo4f0yc9tCoK5icEgUAV37lG7yIll3CWbfoRSDuQfg/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="688" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwVXxKHmPFTITLe10IUeiaHOmkSmrF6LmPwS8qo4f0yc9tCoK5icEgUAV37lG7yIll3CWbfoRSDuQfg/640?wx_fmt=png&amp;from=appmsg"><figcaption> </figcaption></figure><p data-tool="mdnice编辑器">提取limma结果：会用到结果中的t值</p><pre data-tool="mdnice编辑器"><code><span># Extract t-values per gene</span><br>deg &lt;- data<span>$limma_ttop</span> %&gt;%<br>  dplyr::select(ID, t) %&gt;% <br>  dplyr::filter(!is.na(t)) %&gt;% <br>  tibble::column_to_rownames(var = <span>"ID"</span>) %&gt;%<br>  as.matrix()<br><br>head(deg)<br></code></pre><figure data-tool="mdnice编辑器"><img data-imgfileid="100051826" data-ratio="0.27238805970149255" data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwVXxKHmPFTITLe10IUeiaHODqiaTMcsenVOOZVEXaBU9ZJvxj6ibY8OGblf2X9HKicnzPzug74rsgP7A/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="536" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwVXxKHmPFTITLe10IUeiaHODqiaTMcsenVOOZVEXaBU9ZJvxj6ibY8OGblf2X9HKicnzPzug74rsgP7A/640?wx_fmt=png&amp;from=appmsg"><figcaption> </figcaption></figure><h4 data-tool="mdnice编辑器">2、评分使用基因集PROGENy</h4><p data-tool="mdnice编辑器">访问网址：https://saezlab.github.io/progeny/</p><p data-tool="mdnice编辑器">PROGENy是一个包含了经过整理的通路及其靶基因的集合，并且为每个相互作用提供了权重。在这个例子中，<strong>我们将使用人类权重（也提供了其他生物体的权重），并且我们将使用按p值排名的前500个responsive genes。以下是每个通路的简要描述：</strong></p><ul data-tool="mdnice编辑器"><li><section>雄激素（Androgen）：参与男性生殖器官的生长和发育。</section></li><li><section>表皮生长因子受体（EGFR）：在哺乳动物细胞中调节生长、存活、迁移、凋亡、增殖和分化。</section></li><li><section>雌激素（Estrogen）：促进女性生殖器官的生长和发育。</section></li><li><section>缺氧（Hypoxia）：在氧气水平低时促进血管生成和代谢重编程。</section></li><li><section>JAK-STAT 信号通路：涉及免疫、细胞分裂、细胞死亡和肿瘤形成。</section></li><li><section>丝裂原活化蛋白激酶（MAPK）：整合外部信号并促进细胞生长和增殖。</section></li><li><section>核因子κB（NFkB）：调节免疫反应、细胞因子产生和细胞存活。</section></li><li><section>肿瘤蛋白p53（p53）：调节细胞周期、凋亡、DNA修复和肿瘤抑制。</section></li><li><section>磷脂酰肌醇3激酶（PI3K）：促进生长和增殖。</section></li><li><section>转化生长因子β（TGFb）：涉及大多数组织的发展、稳态和修复。</section></li><li><section>肿瘤坏死因子α（TNFa）：介导造血、免疫监视、肿瘤退化和防止感染。</section></li><li><section>TRAIL：诱导凋亡。</section></li><li><section>血管内皮生长因子（VEGF）：介导血管生成、血管通透性和细胞迁移。</section></li><li><section>WNT 信号通路：在发育过程中调节器官形态发生和组织修复。</section></li></ul><p data-tool="mdnice编辑器">获取：</p><pre data-tool="mdnice编辑器"><code><span># 这里需要安装一下OmnipathR包</span><br><span># BiocManager::install("OmnipathR")</span><br><br>net &lt;- decoupleR::get_progeny(organism = <span>'human'</span>, top = 500)<br>net<br>head(net)<br>table(net<span>$source</span>)<br><br><span># 保存一下，这里不是很好下载，后面看看是不是有什么地方可以加速这个下载</span><br>saveRDS(net,file = <span>"net.rds"</span>)<br></code></pre><figure data-tool="mdnice编辑器"><img data-imgfileid="100051834" data-ratio="0.22407407407407406" data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwVXxKHmPFTITLe10IUeiaHOG0iaHicsjn3cnGFY0kWibtAFc0h2Bl3ugBA475mEx5UR9Jmeia3ic3Vse2g/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwVXxKHmPFTITLe10IUeiaHOG0iaHicsjn3cnGFY0kWibtAFc0h2Bl3ugBA475mEx5UR9Jmeia3ic3Vse2g/640?wx_fmt=png&amp;from=appmsg"><figcaption> </figcaption></figure><h4 data-tool="mdnice编辑器">3、使用decoupleR评估PROGENy基因集在此数据中的活性</h4><p data-tool="mdnice编辑器">活性推断采用的模型为：Multivariate Linear Model (MLM)，<strong>MLM在文章中方法的可靠性排前三</strong>。推断的打分如果为positive，表示 the pathway is active；如果是negative，则inactive。</p><figure data-tool="mdnice编辑器"><img data-imgfileid="100051833" data-ratio="0.6848934198331789" data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwVXxKHmPFTITLe10IUeiaHOicXNP9u4L2Eyibk1vcvkjPickg5UKt5USEog1wUB1XOhQC3Gweu1ViaZzA/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1079" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwVXxKHmPFTITLe10IUeiaHOicXNP9u4L2Eyibk1vcvkjPickg5UKt5USEog1wUB1XOhQC3Gweu1ViaZzA/640?wx_fmt=png&amp;from=appmsg"><figcaption> </figcaption></figure><pre data-tool="mdnice编辑器"><code><span># Run mlm</span><br>sample_acts &lt;- decoupleR::run_mlm(mat = counts, <br>                                  net = net, <br>                                  .<span>source</span> = <span>'source'</span>, <br>                                  .target = <span>'target'</span>,<br>                                  .mor = <span>'weight'</span>, <br>                                  minsize = 5)<br>sample_acts<br></code></pre><p data-tool="mdnice编辑器">结果如下：</p><figure data-tool="mdnice编辑器"><img data-imgfileid="100051832" data-ratio="0.39295039164490864" data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwVXxKHmPFTITLe10IUeiaHOyiaibdwAVzpHpiclONribHvX0iaemjovYAJ6U6kRzGmXJl7SdUfDmJpdMZA/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="766" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwVXxKHmPFTITLe10IUeiaHOyiaibdwAVzpHpiclONribHvX0iaemjovYAJ6U6kRzGmXJl7SdUfDmJpdMZA/640?wx_fmt=png&amp;from=appmsg"><figcaption> </figcaption></figure><h4 data-tool="mdnice编辑器">4、打分结果可视化</h4><p data-tool="mdnice编辑器">首先提取每个样本中每条通路的score：</p><pre data-tool="mdnice编辑器"><code><span># Transform to wide matrix</span><br>sample_acts_mat &lt;- sample_acts %&gt;%<br>                   tidyr::pivot_wider(id_cols = <span>'condition'</span>, <br>                                      names_from = <span>'source'</span>,<br>                                      values_from = <span>'score'</span>) %&gt;%<br>                   tibble::column_to_rownames(<span>'condition'</span>) %&gt;%<br>                   as.matrix()<br><br>sample_acts_mat<br><br><span># Scale per feature</span><br>sample_acts_mat &lt;- scale(sample_acts_mat)<br></code></pre><p data-tool="mdnice编辑器">打分矩阵：</p><figure data-tool="mdnice编辑器"><img data-imgfileid="100051835" data-ratio="0.2324074074074074" data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwVXxKHmPFTITLe10IUeiaHOSzqNUyiagZOxQaYcOB1IjNLOHFhLtRnuOWYvSPp5oKrbib2snPPnASFQ/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwVXxKHmPFTITLe10IUeiaHOSzqNUyiagZOxQaYcOB1IjNLOHFhLtRnuOWYvSPp5oKrbib2snPPnASFQ/640?wx_fmt=png&amp;from=appmsg"><figcaption> </figcaption></figure><p data-tool="mdnice编辑器">绘图：</p><pre data-tool="mdnice编辑器"><code><span># Color scale</span><br>colors &lt;- rev(RColorBrewer::brewer.pal(n = 11, name = <span>"RdBu"</span>))<br>colors.use &lt;- grDevices::colorRampPalette(colors = colors)(100)<br><br>my_breaks &lt;- c(seq(-2, 0, length.out = ceiling(100 / 2) + 1),<br>               seq(0.05,2, length.out = floor(100 / 2)))<br><br><span># Plot</span><br>pheatmap::pheatmap(mat = sample_acts_mat,<br>                   color = colors.use,<br>                   border_color = <span>"white"</span>,<br>                   breaks = my_breaks,<br>                   cellwidth = 20,<br>                   cellheight = 20,<br>                   treeheight_row = 20,<br>                   treeheight_col = 20)<br></code></pre><p data-tool="mdnice编辑器">打分结果：</p><figure data-tool="mdnice编辑器"><img data-imgfileid="100051831" data-ratio="0.45838218053927315" data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwVXxKHmPFTITLe10IUeiaHOKRYLCN4C6Lg7rlZMth1jjbhFw0icIwBPwTzUibibibTEZwd8ENsBgK7UmQ/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="853" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwVXxKHmPFTITLe10IUeiaHOKRYLCN4C6Lg7rlZMth1jjbhFw0icIwBPwTzUibibibTEZwd8ENsBgK7UmQ/640?wx_fmt=png&amp;from=appmsg"><figcaption> </figcaption></figure><h4 data-tool="mdnice编辑器">5、也可以从差异表达基因（DEGs）的t值推断通路活性</h4><pre data-tool="mdnice编辑器"><code><span># Run mlm</span><br>contrast_acts &lt;- decoupleR::run_mlm(mat  =deg, <br>                                    net = net, <br>                                    .<span>source</span> = <span>'source'</span>, <br>                                    .target = <span>'target'</span>,<br>                                    .mor = <span>'weight'</span>, <br>                                    minsize = 5)<br><br>contrast_acts<br><br><span># 可视化</span><br>colors &lt;- rev(RColorBrewer::brewer.pal(n = 11, name = <span>"RdBu"</span>)[c(2, 10)])<br><br>p &lt;- ggplot2::ggplot(data = contrast_acts, <br>                     mapping = ggplot2::aes(x = stats::reorder(<span>source</span>, score), <br>                                            y = score)) + <br>     ggplot2::geom_bar(mapping = ggplot2::aes(fill = score),<br>                       color = <span>"black"</span>,<br>                       <span>stat</span> = <span>"identity"</span>) +<br>     ggplot2::scale_fill_gradient2(low = colors[1], <br>                                   mid = <span>"whitesmoke"</span>, <br>                                   high = colors[2], <br>                                   midpoint = 0) + <br>     ggplot2::theme_minimal() +<br>     ggplot2::theme(axis.title = element_text(face = <span>"bold"</span>, size = 12),<br>              axis.text.x = ggplot2::element_text(angle = 45, <br>                                                  hjust = 1, <br>                                                  size = 10, <br>                                                  face = <span>"bold"</span>),<br>              axis.text.y = ggplot2::element_text(size = 10, <br>                                                  face = <span>"bold"</span>),<br>              panel.grid.major = element_blank(), <br>              panel.grid.minor = element_blank()) +<br>     ggplot2::xlab(<span>"Pathways"</span>)<br><br>p<br></code></pre><p data-tool="mdnice编辑器">结果：与野生型（WT）相比，敲除型（KO）中的p53和TRAIL通路被抑制，而MAPKK和JAK-STAT通路似乎被激活，<strong>与上面的每个样本中活性打分热图基本一致</strong>。</p><figure data-tool="mdnice编辑器"><img data-imgfileid="100051837" data-ratio="0.6254635352286774" data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwVXxKHmPFTITLe10IUeiaHOcEf0Rdufp3eYagxqHdzticIq6PSGFQTTqMOPoCmTBwibInHcpV9kyoTg/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="809" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwVXxKHmPFTITLe10IUeiaHOcEf0Rdufp3eYagxqHdzticIq6PSGFQTTqMOPoCmTBwibInHcpV9kyoTg/640?wx_fmt=png&amp;from=appmsg"><figcaption> </figcaption></figure><p data-tool="mdnice编辑器">我们可以进一步沿着它们的t值可视化每个通路中最有反应的基因，以解释结果。例如，让我们来看看属于MAPK通路的基因：</p><pre data-tool="mdnice编辑器"><code>pathway &lt;- <span>'MAPK'</span><br><br>df &lt;- net %&gt;%<br>      dplyr::filter(<span>source</span> == pathway) %&gt;%<br>      dplyr::arrange(target) %&gt;%<br>      dplyr::mutate(ID = target, <br>                    color = <span>"3"</span>) %&gt;%<br>      tibble::column_to_rownames(<span>'target'</span>)<br><br>inter &lt;- sort(dplyr::intersect(rownames(deg), rownames(df)))<br><br>df &lt;- df[inter, ]<br><br>df[<span>'t_value'</span>] &lt;- deg[inter, ]<br><br>df &lt;- df %&gt;%<br>      dplyr::mutate(color = dplyr::if_else(weight &gt; 0 &amp; t_value &gt; 0, <span>'1'</span>, color)) %&gt;%<br>      dplyr::mutate(color = dplyr::if_else(weight &gt; 0 &amp; t_value &lt; 0, <span>'2'</span>, color)) %&gt;%<br>      dplyr::mutate(color = dplyr::if_else(weight &lt; 0 &amp; t_value &gt; 0, <span>'2'</span>, color)) %&gt;%<br>      dplyr::mutate(color = dplyr::if_else(weight &lt; 0 &amp; t_value &lt; 0, <span>'1'</span>, color))<br><br>colors &lt;- rev(RColorBrewer::brewer.pal(n = 11, name = <span>"RdBu"</span>)[c(2, 10)])<br><br>p &lt;- ggplot2::ggplot(data = df, <br>                     mapping = ggplot2::aes(x = weight, <br>                                            y = t_value, <br>                                            color = color)) + <br>     ggplot2::geom_point(size = 2.5, <br>                         color = <span>"black"</span>) + <br>     ggplot2::geom_point(size = 1.5) +<br>     ggplot2::scale_colour_manual(values = c(colors[2], colors[1], <span>"grey"</span>)) +<br>     ggrepel::geom_label_repel(mapping = ggplot2::aes(label = ID)) + <br>     ggplot2::theme_minimal() +<br>     ggplot2::theme(legend.position = <span>"none"</span>) +<br>     ggplot2::geom_vline(xintercept = 0, linetype = <span>'dotted'</span>) +<br>     ggplot2::geom_hline(yintercept = 0, linetype = <span>'dotted'</span>) +<br>     ggplot2::ggtitle(pathway)<br><br>p<br></code></pre><p data-tool="mdnice编辑器">结果如下：</p><figure data-tool="mdnice编辑器"><img data-imgfileid="100051836" data-ratio="0.6359959555106168" data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwVXxKHmPFTITLe10IUeiaHOzlpUNGggZpOEtqSojht0VDtvibniczpTSgcicIeulQzJiaxRBlqdswOUzQ/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="989" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwVXxKHmPFTITLe10IUeiaHOzlpUNGggZpOEtqSojht0VDtvibniczpTSgcicIeulQzJiaxRBlqdswOUzQ/640?wx_fmt=png&amp;from=appmsg"><figcaption> </figcaption></figure><p data-tool="mdnice编辑器">可以去试试看~</p><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com"><h3 data-tool="mdnice编辑器"><span></span><span>学徒作业</span><span></span></h3><p data-tool="mdnice编辑器">前面我们介绍的GSE119931数据集里面有两<span><strong>个不同的细胞系实验，都是两分组，就是FOXA2-KO和WT的差异分析</strong></span>，大家试试看读取 https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE119931 里面的表达量矩阵：</p><pre data-tool="mdnice编辑器"><span></span><code>GSM3387462 PANC1.WT.Rep1<br>GSM3387463 PANC1.WT.Rep2<br>GSM3387464 PANC1.WT.Rep3<br>GSM3387465 PANC1.FOXA2-KO.Rep1<br>GSM3387466 PANC1.FOXA2-KO.Rep2<br>GSM3387467 PANC1.FOXA2-KO.Rep3<br>GSM3387468 CFPAC1.WT.Rep1<br>GSM3387469 CFPAC1.WT.Rep2<br>GSM3387470 CFPAC1.WT.Rep3 <br>GSM3387474 CFPAC1.FOXA2-KD.Rep1<br>GSM3387475 CFPAC1.FOXA2-KD.Rep2<br>GSM3387476 CFPAC1.FOXA2-KD.Rep3<br></code></pre><p data-tool="mdnice编辑器">找到了GSE119931_raw_counts_GRCh38.p13_NCBI.tsv.gz，这个2022-12-15的749.7 Kb的文件，然后进行两次差异分析，然后做交集看看，韦恩图或者更多丰富的方法， 也可以去看看<strong>文献：FOXA2 controls the cis-regulatory networks of pancreatic cancer cells in a differentiation grade-specific manner. EMBO J 2019 Oct 做更丰富的生物学解读！</strong></p></section></section><p><mp-style-type data-value="3"></mp-style-type></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/Fo7EZFeO0c5v0XoFw4FB_w",target="_blank" rel="noopener noreferrer">原文链接</a>
