---
title: "单细胞上游处理 cellranger 定量"
date: 2024-12-19T11:45:30Z
draft: ["false"]
tags: [
  "fetched",
  "老俊俊的生信笔记"
]
categories: ["Acdemic"]
---
单细胞上游处理 cellranger 定量 by 老俊俊的生信笔记
------
<div><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com" data-mpa-powered-by="yiban.io"><section><mp-common-profile data-pluginname="mpprofile" data-id="MzkyMTI1MTYxNA==" data-headimg="http://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usezgsqIGqjITSMggCTSoViaYeoKe2xoZr1IIvNJoztibQxibYHLDDoiabwAc6Ggws3Tvdo8EPss2nLgaVQ/0?wx_fmt=png" data-nickname="老俊俊的生信笔记" data-alias="JunJunLab" data-signature="老俊俊的生信技能和知识分享,我不是巨人,但你可以站在我的肩膀上更进一步!" data-from="0" data-is_biz_ban="0" data-service_type="1"></mp-common-profile></section></section><section><mp-common-clmusic data-pluginname="insertaudio" type="1" music_name="%E7%99%BE%E5%B9%B4%E5%AD%A4%E5%AF%82" albumurl="http://wx.y.gtimg.cn/music/photo_new/T002R500x500M000003j0loN16WKM8_3.jpg" singer="王菲" duration="317000" username="" music_source="1" is_vip="1" listenid="78221626641369408"></mp-common-clmusic></section><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com"><h2 data-tool="mdnice编辑器"><span></span><span>引言</span><span></span></h2><blockquote data-tool="mdnice编辑器"><span>❝</span><p><strong>cellranger</strong> 是由 <strong>10X Genomics</strong> 提供的软件工具包，用于处理其单细胞测序数据（如单细胞转录组测序 scRNA-seq、单细胞免疫测序等）。它是一个专门为 10X Genomics 平台生成的数据（如 Chromium 系统）设计并优化的分析工具，主要用于原始测序数据处理和后续分析。cellranger 包含以下核心命令，每个命令用于不同的功能：</p><ul><li><section><strong>cellranger mkfastq:</strong> 通常用于数据的前期处理，把测序机器输出的 BCL 文件转换为标准的 FASTQ 文件格式。</section></li><li><section><strong>cellranger count:</strong> 对单个样本进行表达定量分析，生成细胞的<strong>基因表达矩阵</strong>。输入为 <strong>FASTQ</strong> 文件，输出为基因表达数据。</section></li><li><section><strong>cellranger aggr：</strong> 它允许将不同样本整合成为一个组合的细胞数据集，同时对测序深度进行标准化。</section></li><li><section><strong>cellranger reanalyze:</strong> 对现有的单细胞数据进行重新分析。允许用户在 <strong>filtered_feature_bc_matrix</strong> 基础上调整聚类参数、可视化和降维分析。</section></li><li><section><strong>cellranger vdj:</strong> 针对单细胞 <strong>V(D)J 免疫测序（免疫组库测序）</strong> 数据进行分析。用于分析 TCR（T 细胞受体）或 BCR（B 细胞受体）信息，例如克隆扩增、序列重排等。</section></li><li><section><strong>cellranger multi:</strong> 支持多种功能（如抗体标记数据整合），特别是在多模态实验数据中的分析。</section></li></ul></blockquote><blockquote data-tool="mdnice编辑器"><span>❝</span><p>我们主要测试一下这个软件的<strong>下载安装</strong>，对<strong>文献里下载 fastq 文件进行定量输出表达矩阵</strong>这些过程。</p></blockquote><p data-tool="mdnice编辑器"><strong>文库组成：</strong></p><p data-tool="mdnice编辑器"><strong>样本拆分后一般有两个 fastq（barcode+UMI 序列(26/28bp,取决于文库版本)和 cDNA 序列）或者三个 fastq 文件（sample index 序列，barcode+UMI 序列和 cDNA 序列）</strong>，需要自己手动检查一下每个文件是什么内容。</p><figure data-tool="mdnice编辑器"><img data-imgfileid="100032520" data-ratio="0.1905" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useworkLahshuoWrQe0ickusLYxojdFdVcpYyqtPib0mEHIP5sD6jFichtduCOdcopIVw0icNGJWzKMV8gQ/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="2000" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useworkLahshuoWrQe0ickusLYxojdFdVcpYyqtPib0mEHIP5sD6jFichtduCOdcopIVw0icNGJWzKMV8gQ/640?wx_fmt=png&amp;from=appmsg"></figure><h2 data-tool="mdnice编辑器"><span></span><span>cellranger 下载安装</span><span></span></h2><p data-tool="mdnice编辑器"><strong>官网：https://www.10xgenomics.com/company</strong></p><figure data-tool="mdnice编辑器"><img data-imgfileid="100032521" data-ratio="0.3662420382165605" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useworkLahshuoWrQe0ickusLYflQOy5FZJmUia3MrqEu8osrlouO1AFhm1poQpjKUnqIicQT0Xx1A1O0Q/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1570" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useworkLahshuoWrQe0ickusLYflQOy5FZJmUia3MrqEu8osrlouO1AFhm1poQpjKUnqIicQT0Xx1A1O0Q/640?wx_fmt=png&amp;from=appmsg"></figure><p data-tool="mdnice编辑器">找到 cellranger，然后需要注册一下邮箱：</p><figure data-tool="mdnice编辑器"><img data-imgfileid="100032519" data-ratio="0.35526315789473684" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useworkLahshuoWrQe0ickusLYUUJLpJCLZ9TlK7HCNUwOTRCFqge4ib9m4PhNhSoyK7VrCfdRvStFOdw/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1596" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useworkLahshuoWrQe0ickusLYUUJLpJCLZ9TlK7HCNUwOTRCFqge4ib9m4PhNhSoyK7VrCfdRvStFOdw/640?wx_fmt=png&amp;from=appmsg"></figure><p data-tool="mdnice编辑器">有三种方式下载，<strong>直接下载到本地，或者 curl/wget 命令下载，开箱即用型：</strong></p><figure data-tool="mdnice编辑器"><img data-imgfileid="100032522" data-ratio="0.5829351535836177" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useworkLahshuoWrQe0ickusLY0NcdsqibbtibwEZFTwZFyphE69UaqlEWY71pnlgLVukOo7CAaBgRmo7w/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1465" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useworkLahshuoWrQe0ickusLY0NcdsqibbtibwEZFTwZFyphE69UaqlEWY71pnlgLVukOo7CAaBgRmo7w/640?wx_fmt=png&amp;from=appmsg"></figure><pre data-tool="mdnice编辑器"><span></span><code>curl -o cellranger-9.0.0.tar.gz "https://cf.10xgenomics.com/releases/cell-exp/cellranger-9.0.0.tar.gz?Expires=1734643244&amp;Key-Pair-Id=APKAI7S6A5RYOXBWRPDA&amp;Signature=fYYn9M0ttG56GYYZ2f2B~guVJzi7urOwPYCNfeuvMuz5RuIL7UrnsH1ekXFstD0BWdG7f2KwwhKxuW59jutCTQUaI3jegf8~IyzjkaoUMi99NcQsuIaCMp15PSFZIdjD6lkbJsteCguFzmLZSUUHtYIGjbDISVlfGVttxx~yauVoAtrES3wgZLwY-69FPVNIoo8ccWHNOe2WtDdzIUvX4brICmCvTWda83AanZQ7HezPd7q1cQo2Z2ndkWovCrQmy4ZLhwf2nX90SJS0OuCn3O78Yq3EU1lHZczKk9iHzdIJYBCnes3rdXmkhsikzlfyAQPtLe3mC9TtUOocncy~Fg__"<br><br>wget -O cellranger-9.0.0.tar.gz "https://cf.10xgenomics.com/releases/cell-exp/cellranger-9.0.0.tar.gz?Expires=1734643244&amp;Key-Pair-Id=APKAI7S6A5RYOXBWRPDA&amp;Signature=fYYn9M0ttG56GYYZ2f2B~guVJzi7urOwPYCNfeuvMuz5RuIL7UrnsH1ekXFstD0BWdG7f2KwwhKxuW59jutCTQUaI3jegf8~IyzjkaoUMi99NcQsuIaCMp15PSFZIdjD6lkbJsteCguFzmLZSUUHtYIGjbDISVlfGVttxx~yauVoAtrES3wgZLwY-69FPVNIoo8ccWHNOe2WtDdzIUvX4brICmCvTWda83AanZQ7HezPd7q1cQo2Z2ndkWovCrQmy4ZLhwf2nX90SJS0OuCn3O78Yq3EU1lHZczKk9iHzdIJYBCnes3rdXmkhsikzlfyAQPtLe3mC9TtUOocncy~Fg__"<br></code></pre><h3 data-tool="mdnice编辑器"><span></span><span>准备参考基因组</span><span></span></h3><p data-tool="mdnice编辑器">网站有已经构建好的 <strong>人/小鼠</strong> 的：</p><figure data-tool="mdnice编辑器"><img data-imgfileid="100032518" data-ratio="0.22967479674796748" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useworkLahshuoWrQe0ickusLYEeGQBvuH4cywTtyhDMIOuUXYXicQ8FxgSnYXSeKdNw4xCTLSWqVia4QQ/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="984" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useworkLahshuoWrQe0ickusLYEeGQBvuH4cywTtyhDMIOuUXYXicQ8FxgSnYXSeKdNw4xCTLSWqVia4QQ/640?wx_fmt=png&amp;from=appmsg"></figure><figure data-tool="mdnice编辑器"><img data-imgfileid="100032527" data-ratio="0.25729166666666664" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useworkLahshuoWrQe0ickusLY50TEN5lvIpRz8C2Ovw6KeY8AH2zwvfPuMHmicP8Wktbc7ic1akzhazibQ/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="960" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useworkLahshuoWrQe0ickusLY50TEN5lvIpRz8C2Ovw6KeY8AH2zwvfPuMHmicP8Wktbc7ic1akzhazibQ/640?wx_fmt=png&amp;from=appmsg"></figure><pre data-tool="mdnice编辑器"><span></span><code>wget "https://cf.10xgenomics.com/supp/cell-exp/refdata-gex-GRCh38-2024-A.tar.gz"<br>tar -zxvf refdata-gex-GRCh38-2024-A.tar.gz<br><br>wget "https://cf.10xgenomics.com/supp/cell-exp/refdata-gex-GRCm39-2024-A.tar.gz"<br>tar -zxvf refdata-gex-GRCm39-2024-A.tar.gz<br></code></pre><hr data-tool="mdnice编辑器"><p data-tool="mdnice编辑器">你也可以自己构建索引文件，首先使用 <strong>cellranger mkgtf</strong> 准备好 <strong>cellranger</strong> 需要的 <strong>gtf</strong> 的格式：</p><pre data-tool="mdnice编辑器"><span></span><code>cellranger mkgtf your.gtf out.gtf<br></code></pre><p data-tool="mdnice编辑器">然后使用 <strong>cellranger mkref</strong> 构建索引文件：</p><pre data-tool="mdnice编辑器"><span></span><code>cellranger mkref \<br>    --genome=my_genome \<br>    --fasta=mygenome.dna.primary_assembly.fa \<br>    --genes=out.gtf<br></code></pre><h2 data-tool="mdnice编辑器"><span></span><span>测试</span><span></span></h2><blockquote data-tool="mdnice编辑器"><span>❝</span><p>我们使用 <strong>GEO</strong> 的数据下载 fastq 来测试一下。</p></blockquote><p data-tool="mdnice编辑器"><strong>GSE182790</strong></p><figure data-tool="mdnice编辑器"><img data-imgfileid="100032523" data-ratio="0.6677667766776678" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useworkLahshuoWrQe0ickusLYlqne1pY0V4Zry7CzfEGT280XncrjbAHNoKHQaib3xJPaGibDYsXIBUYQ/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="909" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useworkLahshuoWrQe0ickusLYlqne1pY0V4Zry7CzfEGT280XncrjbAHNoKHQaib3xJPaGibDYsXIBUYQ/640?wx_fmt=png&amp;from=appmsg"></figure><figure data-tool="mdnice编辑器"><img data-imgfileid="100032526" data-ratio="0.3511491181186531" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useworkLahshuoWrQe0ickusLYFSatPvAm0XBwsxeLAUE51ZSicCyzRPvZJMkw5TDoxrNXSggEHOItd5A/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1871" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useworkLahshuoWrQe0ickusLYFSatPvAm0XBwsxeLAUE51ZSicCyzRPvZJMkw5TDoxrNXSggEHOItd5A/640?wx_fmt=png&amp;from=appmsg"></figure><p data-tool="mdnice编辑器"><strong>我们看到有些一个样本有四个文件，推测可能是多个 runs/lanes 的数据，我们点击 SRA 进去检查可以看到：</strong></p><figure data-tool="mdnice编辑器"><img data-imgfileid="100032524" data-ratio="0.7289272030651341" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useworkLahshuoWrQe0ickusLYEcW3dyvFHdXxLzuMLnKYxCTpxY4LP5sRDGDl3dLLpMUP8jiagqLqnFQ/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1044" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useworkLahshuoWrQe0ickusLYEcW3dyvFHdXxLzuMLnKYxCTpxY4LP5sRDGDl3dLLpMUP8jiagqLqnFQ/640?wx_fmt=png&amp;from=appmsg"></figure><p data-tool="mdnice编辑器">我们选择一个样本，然后去 sra-exploer 拿到下载地址，对应有 <strong>8(2x4)</strong> 个文件：</p><pre data-tool="mdnice编辑器"><span></span><code><span>#</span><span>!/usr/bin/env bash</span><br>ascp -QT -l 300m -P33001 -i $HOME/.aspera/connect/etc/asperaweb_id_dsa.openssh era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR156/039/SRR15614139/SRR15614139_1.fastq.gz . &amp;&amp; mv SRR15614139_1.fastq.gz SRR15614139_GSM5537050_H9_t0_Homo_sapiens_RNA-Seq_1.fastq.gz<br>ascp -QT -l 300m -P33001 -i $HOME/.aspera/connect/etc/asperaweb_id_dsa.openssh era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR156/039/SRR15614139/SRR15614139_2.fastq.gz . &amp;&amp; mv SRR15614139_2.fastq.gz SRR15614139_GSM5537050_H9_t0_Homo_sapiens_RNA-Seq_2.fastq.gz<br>ascp -QT -l 300m -P33001 -i $HOME/.aspera/connect/etc/asperaweb_id_dsa.openssh era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR156/040/SRR15614140/SRR15614140_1.fastq.gz . &amp;&amp; mv SRR15614140_1.fastq.gz SRR15614140_GSM5537050_H9_t0_Homo_sapiens_RNA-Seq_1.fastq.gz<br>ascp -QT -l 300m -P33001 -i $HOME/.aspera/connect/etc/asperaweb_id_dsa.openssh era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR156/040/SRR15614140/SRR15614140_2.fastq.gz . &amp;&amp; mv SRR15614140_2.fastq.gz SRR15614140_GSM5537050_H9_t0_Homo_sapiens_RNA-Seq_2.fastq.gz<br>ascp -QT -l 300m -P33001 -i $HOME/.aspera/connect/etc/asperaweb_id_dsa.openssh era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR156/041/SRR15614141/SRR15614141_1.fastq.gz . &amp;&amp; mv SRR15614141_1.fastq.gz SRR15614141_GSM5537050_H9_t0_Homo_sapiens_RNA-Seq_1.fastq.gz<br>ascp -QT -l 300m -P33001 -i $HOME/.aspera/connect/etc/asperaweb_id_dsa.openssh era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR156/041/SRR15614141/SRR15614141_2.fastq.gz . &amp;&amp; mv SRR15614141_2.fastq.gz SRR15614141_GSM5537050_H9_t0_Homo_sapiens_RNA-Seq_2.fastq.gz<br>ascp -QT -l 300m -P33001 -i $HOME/.aspera/connect/etc/asperaweb_id_dsa.openssh era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR156/042/SRR15614142/SRR15614142_1.fastq.gz . &amp;&amp; mv SRR15614142_1.fastq.gz SRR15614142_GSM5537050_H9_t0_Homo_sapiens_RNA-Seq_1.fastq.gz<br>ascp -QT -l 300m -P33001 -i $HOME/.aspera/connect/etc/asperaweb_id_dsa.openssh era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR156/042/SRR15614142/SRR15614142_2.fastq.gz . &amp;&amp; mv SRR15614142_2.fastq.gz SRR15614142_GSM5537050_H9_t0_Homo_sapiens_RNA-Seq_2.fastq.gz<br></code></pre><p data-tool="mdnice编辑器">下载两个 runs 的测试，改一下名称，命名格式 <strong>Sample_S1_L00X_R1_001.fastq.gz，Sample_S1_L00X_R2_001.fastq.gz</strong>：</p><pre data-tool="mdnice编辑器"><span></span><code>(base) ➜  1.raw-data ls -lh<br>total 5.3G<br>-rwxr-xr-x 1 root root 2.1K Dec 19 15:22 sra_explorer_fastq_aspera_download.sh<br>-rw-r--r-- 1 root root 622M Dec 19 15:24 test_S1_L001_R1_001.fastq.gz<br>-rw-r--r-- 1 root root 2.2G Dec 19 15:34 test_S1_L001_R2_001.fastq.gz<br>-rw-r--r-- 1 root root 563M Dec 19 15:37 test_S1_L002_R1_001.fastq.gz<br>-rw-r--r-- 1 root root 2.0G Dec 19 15:47 test_S1_L002_R2_001.fastq.gz<br>(base) ➜  1.raw-data<br></code></pre><p data-tool="mdnice编辑器">看看帮助文档：</p><pre data-tool="mdnice编辑器"><span></span><code>./cellranger-9.0.0/cellranger count -h<br><span>#</span><span> Count gene expression and/or feature barcode reads from a single sample and GEM well</span><br><span>#</span><br><span>#</span><span> Usage: cellranger count [OPTIONS] --id &lt;ID&gt; --create-bam &lt;<span>true</span>|<span>false</span>&gt;</span><br><span>#</span><br><span>#</span><span>   Options:</span><br><span>#</span><span>   --id &lt;ID&gt;                                   A unique run id and output folder name [a-zA-Z0-9_-]+</span><br><span>#</span><span>   --description &lt;TEXT&gt;                        Sample description to embed <span>in</span> output files [default: ]</span><br><span>#</span><span> --transcriptome &lt;PATH&gt;                      Path of folder containing 10x-compatible transcriptome reference</span><br><span>#</span><span> --fastqs &lt;PATH&gt;                             Path to input FASTQ data</span><br><span>#</span><span> --project &lt;TEXT&gt;                            Name of the project folder within a mkfastq or bcl2fastq-generated folder from <span>which</span> to pick</span><br><span>#</span><span> FASTQs</span><br><span>#</span><span> --sample &lt;PREFIX&gt;                           Prefix of the filenames of FASTQs to select</span><br><span>#</span><span> --lanes &lt;NUMS&gt;                              Only use FASTQs from selected lanes</span><br><span>#</span><span> --libraries &lt;CSV&gt;                           CSV file declaring input library data sources</span><br><span>#</span><span> --feature-ref &lt;CSV&gt;                         Feature reference CSV file, declaring Feature Barcode constructs and associated barcodes</span><br><span>#</span><span> --expect-cells &lt;NUM&gt;                        Expected number of recovered cells, used as input to cell calling algorithm</span><br><span>#</span><span> --force-cells &lt;NUM&gt;                         Force pipeline to use this number of cells, bypassing cell calling algorithm. [MINIMUM: 10]</span><br><span>#</span><span> --create-bam &lt;<span>true</span>|<span>false</span>&gt;                   Enable or <span>disable</span> BAM file generation. Setting --create-bam=<span>false</span> reduces the total computation</span><br><span>#</span><span> time and the size of the output directory (BAM file not generated). We recommend setting</span><br><span>#</span><span> --create-bam=<span>true</span> <span>if</span> unsure. See https://10xgen.com/create-bam <span>for</span> additional guidance</span><br><span>#</span><span> [possible values: <span>true</span>, <span>false</span>]</span><br><span>#</span><span> --nosecondary                               Disable secondary analysis, e.g. clustering. Optional</span><br><span>#</span><span> --r1-length &lt;NUM&gt;                           Hard trim the input Read 1 to this length before analysis</span><br><span>#</span><span> --r2-length &lt;NUM&gt;                           Hard trim the input Read 2 to this length before analysis</span><br><span>#</span><span> --include-introns &lt;<span>true</span>|<span>false</span>&gt;              Include intronic reads <span>in</span> count [default: <span>true</span>] [possible values: <span>true</span>, <span>false</span>]</span><br><span>#</span><span> --chemistry &lt;CHEM&gt;                          Assay configuration. NOTE: by default the assay configuration is detected automatically, <span>which</span></span><br><span>#</span><span> is the recommended mode. You usually will not need to specify a chemistry. Options are: <span>'auto'</span></span><br><span>#</span><span> <span>for</span> autodetection, <span>'threeprime'</span> <span>for</span> Single Cell 3<span>', '</span>fiveprime<span>' for  Single Cell 5'</span>, <span>'SC3Pv1'</span></span><br><span>#</span><span> or <span>'SC3Pv2'</span> or <span>'SC3Pv3'</span> or <span>'SC3Pv4'</span> <span>for</span> Single Cell 3<span>' v1/v2/v3/v4, '</span>SC3Pv3LT<span>' for Single Cell</span></span><br><span>#</span><span><span>                                                   3'</span> v3 LT, <span>'SC3Pv3HT'</span> <span>for</span> Single Cell 3<span>' v3 HT, '</span>SC5P-PE<span>' or '</span>SC5P-PE-v3<span>' or '</span>SC5P-R2<span>' or</span></span><br><span>#</span><span><span>                                                   '</span>SC5P-R2-v3<span>' for Single Cell 5'</span>, paired-end/R2-only, <span>'SC-FB'</span> <span>for</span> Single Cell Antibody-only 3<span>'</span></span><br><span>#</span><span><span>                                                   v2 or 5'</span>. To analyze the GEX portion of multiome data, chemistry must be <span>set</span> to <span>'ARC-v1'</span></span><br><span>#</span><span> [default: auto]</span><br><span>#</span><span> --no-libraries                              Proceed with processing using a --feature-ref but no Feature Barcode libraries specified with</span><br><span>#</span><span> the <span>'libraries'</span> flag</span><br><span>#</span><span> --check-library-compatibility &lt;<span>true</span>|<span>false</span>&gt;  Whether to check <span>for</span> barcode compatibility between libraries. [default: <span>true</span>] [possible values:</span><br><span>#</span><span>                                                                                                                              <span>true</span>, <span>false</span>]</span><br><span>#</span><span> --tenx-cloud-token-path &lt;PATH&gt;              The path to the 10x Cloud Analysis user token used to <span>enable</span> cell annotation. If not provided,</span><br><span>#</span><span> will default to the location stored through cellranger cloud auth setup</span><br><span>#</span><span> --cell-annotation-model &lt;MODEL&gt;             Cell annotation model to use. Valid model names can be viewed by running `cellranger cloud</span><br><span>#</span><span> annotation models` or on the 10x Genomics Support site (https://www.10xgenomics.com/support).</span><br><span>#</span><span> If <span>"auto"</span>, uses the default model <span>for</span> the species. If not provided, does not run cell</span><br><span>#</span><span> annotation</span><br><span>#</span><span> --min-crispr-umi &lt;NUM&gt;                      Minimum CRISPR UMI threshold [default: 3]</span><br><span>#</span><span> --dry                                       Do not execute the pipeline. Generate a pipeline invocation (.mro) file and stop</span><br><span>#</span><span> --jobmode &lt;MODE&gt;                            Job manager to use. Valid options: <span>local</span> (default), sge, lsf, slurm or path to a .template</span><br><span>#</span><span> file. Search <span>for</span> <span>help</span> on <span>"Cluster Mode"</span> at support.10xgenomics.com <span>for</span> more details on</span><br><span>#</span><span> configuring the pipeline to use a compute cluster</span><br><span>#</span><span> --localcores &lt;NUM&gt;                          Set max cores the pipeline may request at one time. Only applies to <span>local</span> <span>jobs</span></span><br><span>#</span><span> --localmem &lt;NUM&gt;                            Set max GB the pipeline may request at one time. Only applies to <span>local</span> <span>jobs</span></span><br><span>#</span><span> --localvmem &lt;NUM&gt;                           Set max virtual address space <span>in</span> GB <span>for</span> the pipeline. Only applies to <span>local</span> <span>jobs</span></span><br><span>#</span><span> --mempercore &lt;NUM&gt;                          Reserve enough threads <span>for</span> each job to ensure enough memory will be available, assuming each</span><br><span>#</span><span> core on your cluster has at least this much memory available. Only applies to cluster jobmodes</span><br><span>#</span><span> --maxjobs &lt;NUM&gt;                             Set max <span>jobs</span> submitted to cluster at one time. Only applies to cluster jobmodes</span><br><span>#</span><span> --jobinterval &lt;NUM&gt;                         Set delay between submitting <span>jobs</span> to cluster, <span>in</span> ms. Only applies to cluster jobmodes</span><br><span>#</span><span> --overrides &lt;PATH&gt;                          The path to a JSON file that specifies stage-level overrides <span>for</span> cores and memory.</span><br><span>#</span><span> Finer-grained than --localcores, --mempercore and --localmem. Consult</span><br><span>#</span><span> https://10xgen.com/resource-override <span>for</span> an example override file</span><br><span>#</span><span> --output-dir &lt;PATH&gt;                         Output the results to this directory</span><br><span>#</span><span> --uiport &lt;PORT&gt;                             Serve web UI at http://localhost:PORT</span><br><span>#</span><span> --<span>disable</span>-ui                                Do not serve the web UI</span><br><span>#</span><span> --noexit                                    Keep web UI running after pipestance completes or fails</span><br><span>#</span><span> --nopreflight                               Skip preflight checks</span><br><span>#</span><span> -h, --<span>help</span>                                      Print <span>help</span></span><br></code></pre><p data-tool="mdnice编辑器">定量<strong>（--nosecondary 不使用 cellranger 进行后续的降维聚类等分析步骤）</strong>：</p><pre data-tool="mdnice编辑器"><span></span><code>./cellranger-9.0.0/cellranger count --id test_sc \<br>    --fastqs ./1.raw-data \<br>    --sample test \<br>    --transcriptome /mnt/data/home/tycloud/junjun_proj/index-data/refdata-gex-GRCh38-2024-A \<br>    --nosecondary \<br>    --localcores 15 \<br>    --output-dir cellranger_output \<br>    --create-bam false<br></code></pre><p data-tool="mdnice编辑器">查看输出文件，主要用到 <strong>filtered_feature_bc_matrix.h5</strong> 和 <strong>filtered_feature_bc_matrix</strong> 里面的三个标准数据，接下来就可以在 R 里面操作走<strong>seurat</strong>流程了：</p><pre data-tool="mdnice编辑器"><span></span><code>(base) ➜  1.raw-data cd ../cellranger_output/outs<br>(base) ➜  outs ls -lh<br>total 196M<br>drwxr-xr-x 2 root root 4.0K Dec 19 17:20 filtered_feature_bc_matrix<br>-rw-r--r-- 1 root root  18M Dec 19 17:20 filtered_feature_bc_matrix.h5<br>-rw-r--r-- 1 root root  668 Dec 19 17:23 metrics_summary.csv<br>-rw-r--r-- 1 root root 146M Dec 19 17:20 molecule_info.h5<br>drwxr-xr-x 2 root root 4.0K Dec 19 17:16 raw_feature_bc_matrix<br>-rw-r--r-- 1 root root  29M Dec 19 17:16 raw_feature_bc_matrix.h5<br>-rw-r--r-- 1 root root 4.7M Dec 19 17:23 web_summary.html<br><br>(base) ➜  outs ls -l filtered_feature_bc_matrix/<br>total 47808<br>-rw-r--r-- 1 root root    26555 Dec 19 17:20 barcodes.tsv.gz<br>-rw-r--r-- 1 root root   341963 Dec 19 17:20 features.tsv.gz<br>-rw-r--r-- 1 root root 48582326 Dec 19 17:20 matrix.mtx.gz<br></code></pre><p data-tool="mdnice编辑器"><strong>web_summary.html</strong> 提供了关于分析结果的总结信息，主要用于快速了解测序和分析结果的质量、流程成功与否以及相关的关键指标：</p><figure data-tool="mdnice编辑器"><img data-imgfileid="100032525" data-ratio="0.6427457098283932" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useworkLahshuoWrQe0ickusLYzIH2pEJw7viaUrNnicD5CKicFmqJKj8IAibJiajy0n70mHqficoMj2sIPmjQ/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1282" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useworkLahshuoWrQe0ickusLYzIH2pEJw7viaUrNnicD5CKicFmqJKj8IAibJiajy0n70mHqficoMj2sIPmjQ/640?wx_fmt=png&amp;from=appmsg"></figure><p data-tool="mdnice编辑器">比对情况：</p><figure data-tool="mdnice编辑器"><img data-imgfileid="100032531" data-ratio="0.6515625" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useworkLahshuoWrQe0ickusLYcG2gBsn7OlvwNPkP8CIuv8XRABQgcOTOyTwnYZKJKIJK91wFIEJvibw/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="640" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useworkLahshuoWrQe0ickusLYcG2gBsn7OlvwNPkP8CIuv8XRABQgcOTOyTwnYZKJKIJK91wFIEJvibw/640?wx_fmt=png&amp;from=appmsg"></figure><p data-tool="mdnice编辑器">测序饱和度：</p><figure data-tool="mdnice编辑器"><img data-imgfileid="100032532" data-ratio="0.5003855050115652" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useworkLahshuoWrQe0ickusLYt4SO5LorRNsiaolGzEIEd0aFyzlwXqqeSaCEmpnWzNRsqvqVcjdhbOA/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1297" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useworkLahshuoWrQe0ickusLYt4SO5LorRNsiaolGzEIEd0aFyzlwXqqeSaCEmpnWzNRsqvqVcjdhbOA/640?wx_fmt=png&amp;from=appmsg"></figure><h2 data-tool="mdnice编辑器"><span></span><span>结尾</span><span></span></h2><blockquote data-tool="mdnice编辑器"><span>❝</span><p><strong>路漫漫其修远兮,吾将上下而求索。</strong></p></blockquote><hr data-tool="mdnice编辑器"><p data-tool="mdnice编辑器">欢迎加入生信交流群。加我微信我也拉你进 <strong>微信群聊</strong> <strong>老俊俊生信交流群</strong> <strong>(微信交流群需收取 20 元入群费用,一旦交费,拒不退还!(防止骗子和便于管理))</strong> 。</p><figure data-tool="mdnice编辑器"><img data-imgfileid="100032530" data-ratio="0.6083707025411061" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useworkLahshuoWrQe0ickusLYTgTp0w7ULoesBicuQMfo4asGROdA1aDHWLu1ibksV8UeQw2oiaFnRiaxXg/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="669" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useworkLahshuoWrQe0ickusLYTgTp0w7ULoesBicuQMfo4asGROdA1aDHWLu1ibksV8UeQw2oiaFnRiaxXg/640?wx_fmt=png&amp;from=appmsg"></figure><hr data-tool="mdnice编辑器"><p data-tool="mdnice编辑器"><strong></strong></p><center data-tool="mdnice编辑器"><strong> 往期回顾目录 </strong></center><p data-tool="mdnice编辑器"><br></p><blockquote data-tool="mdnice编辑器"><span>❝</span><center><strong><a href="https://mp.weixin.qq.com/s?__biz=MzkyMTI1MTYxNA==&amp;mid=2247516165&amp;idx=1&amp;sn=338f0ac0dc903a1c8d5be3ae6227ebb3&amp;token=941065474&amp;lang=zh_CN&amp;scene=21#wechat_redirect" data-linktype="2">scATAC 多样本整合</a></strong></center><strong><center><a href="https://mp.weixin.qq.com/s?__biz=MzkyMTI1MTYxNA==&amp;mid=2247516132&amp;idx=1&amp;sn=ce65838712aa2e5e4c3c5b7dcfdb71b4&amp;token=1001473693&amp;lang=zh_CN&amp;scene=21#wechat_redirect" data-linktype="2">单细胞亚群注释相关工具</a></center></strong><strong><center><a href="https://mp.weixin.qq.com/s?__biz=MzkyMTI1MTYxNA==&amp;mid=2247516072&amp;idx=1&amp;sn=f82a27c6936f38c3a078ea259178216d&amp;token=1001473693&amp;lang=zh_CN&amp;scene=21#wechat_redirect" data-linktype="2">听说你想美化 signac 的 track 图？</a></center></strong><strong><center><a href="https://mp.weixin.qq.com/s?__biz=MzkyMTI1MTYxNA==&amp;mid=2247516054&amp;idx=1&amp;sn=4315e02f23804d9578464dac4c9d5646&amp;token=291087746&amp;lang=zh_CN&amp;scene=21#wechat_redirect" data-linktype="2">scanpy 处理单细胞数据</a></center></strong><strong><center><a href="https://mp.weixin.qq.com/s?__biz=MzkyMTI1MTYxNA==&amp;mid=2247516004&amp;idx=1&amp;sn=82611ea509826412aaf3b606e865d61a&amp;token=350842983&amp;lang=zh_CN&amp;scene=21#wechat_redirect" data-linktype="2">snapATAC2 分析教程</a></center></strong><strong><center><a href="https://mp.weixin.qq.com/s?__biz=MzkyMTI1MTYxNA==&amp;mid=2247515887&amp;idx=1&amp;sn=f12b573b99aa6bf20c5aae995f5a57e1&amp;token=350842983&amp;lang=zh_CN&amp;scene=21#wechat_redirect" data-linktype="2">scATAC 教程一文详解</a></center></strong><strong><center><a href="https://mp.weixin.qq.com/s?__biz=MzkyMTI1MTYxNA==&amp;mid=2247515726&amp;idx=1&amp;sn=0fb48c324f7207d95fb1e5c774b4fd1a&amp;token=2079312128&amp;lang=zh_CN&amp;scene=21#wechat_redirect" data-linktype="2">单细胞数据类型读取一文解决！</a></center></strong><strong><center><a href="https://mp.weixin.qq.com/s?__biz=MzkyMTI1MTYxNA==&amp;mid=2247515672&amp;idx=1&amp;sn=3878299038763a15e2503340942d9660&amp;token=902807353&amp;lang=zh_CN&amp;scene=21#wechat_redirect" data-linktype="2">从 gtf 文件构建 orgdb 和 txdb 数据库</a></center></strong><strong><center><a href="https://mp.weixin.qq.com/s?__biz=MzkyMTI1MTYxNA==&amp;mid=2247515655&amp;idx=1&amp;sn=03c1bf4a081fb561468a4f3c40331615&amp;token=1970442110&amp;lang=zh_CN&amp;scene=21#wechat_redirect" data-linktype="2">easybio 基于 CellMarker2.0 数据库进行亚群注释</a></center></strong><strong><center><a href="https://mp.weixin.qq.com/s?__biz=MzkyMTI1MTYxNA==&amp;mid=2247515609&amp;idx=1&amp;sn=2f81a1d58af368bc0a261a59f4026652&amp;token=1970442110&amp;lang=zh_CN&amp;scene=21#wechat_redirect" data-linktype="2">写一个 R 版本的 ORFfinder</a></center></strong></blockquote></section><p><br></p><p><mp-style-type data-value="3"></mp-style-type></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/1bmR6FC_PD3pzo8d3TrYdA",target="_blank" rel="noopener noreferrer">原文链接</a>
