---
title: "本实验室可以做的分析：scATAC-seq分析-signac"
date: 2024-12-22T00:21:08Z
draft: ["false"]
tags: [
  "fetched",
  "东林的扯淡小屋"
]
categories: ["Acdemic"]
---
本实验室可以做的分析：scATAC-seq分析-signac by 东林的扯淡小屋
------
<div><p>首先下载相关文件：<br></p><p><img data-galleryid="" data-ratio="0.5638888888888889" data-s="300,640" data-type="png" data-w="1080" data-src="https://mmbiz.qpic.cn/mmbiz_png/kZ1wdgAscBr9WKRw7ibjDQ9yUBhqhSJxjqypd05JL6Oic71MPQPx4Ixd8I3ffvdSH4N6yGduInrtdVGLpl0LYbIg/640?wx_fmt=png" src="https://mmbiz.qpic.cn/mmbiz_png/kZ1wdgAscBr9WKRw7ibjDQ9yUBhqhSJxjqypd05JL6Oic71MPQPx4Ixd8I3ffvdSH4N6yGduInrtdVGLpl0LYbIg/640?wx_fmt=png"></p><p><img data-galleryid="" data-ratio="0.28854314002828857" data-s="300,640" data-type="png" data-w="707" data-src="https://mmbiz.qpic.cn/mmbiz_png/kZ1wdgAscBr9WKRw7ibjDQ9yUBhqhSJxjDMiaQibVDbibGud8S2KKzIFtItxJyJDznaT3ibwe9GIWK6iaAOdwOtn4HQg/640?wx_fmt=png" src="https://mmbiz.qpic.cn/mmbiz_png/kZ1wdgAscBr9WKRw7ibjDQ9yUBhqhSJxjDMiaQibVDbibGud8S2KKzIFtItxJyJDznaT3ibwe9GIWK6iaAOdwOtn4HQg/640?wx_fmt=png"></p><p><br></p><p>全套代码：<br></p><section><section><ul><li><li><li><li><li></ul><pre data-lang="sql"><code><span>Sys.setenv(LANGUAGE = "en") <span>#显示英文报错信息</span></span></code><code><span>gc()</span></code><code><span>memory.limit(9999999999)</span></code><code><span>set.seed(<span>123</span>)</span></code><code><span>rm(<span>list</span> = ls())  </span></code><code><span>options(stringsAsFactors = F)</span></code><code><span><span>library</span>(Seurat)</span></code><code><span><span>library</span>(tidyverse)</span></code><code><span><span>library</span>(patchwork)</span></code><code><span><span>library</span>(SCENIC)</span></code><code><span><span>library</span>(plyr)</span></code><code><span><span>library</span>(permute)</span></code><code><span><span>library</span>(data.table)</span></code><code><span><span>library</span>(SCopeLoomR)</span></code><code><span><span>library</span>(Signac)</span></code><code><span><span>library</span>(Seurat)</span></code><code><span><span>library</span>(EnsDb.Hsapiens.v75)</span></code><code><span><span>library</span>(ggplot2)</span></code><code><span><span>library</span>(patchwork)</span></code><code><span>setwd(<span>"/data1/yudonglin/scATAC"</span>)</span></code><code><span><br></span></code><code><span><br></span></code><code><span><br></span></code><code><span><br></span></code><code><span>counts &lt;- Read10X_h5(filename = <span>"atac_v1_pbmc_10k_filtered_peak_bc_matrix.h5"</span>)</span></code><code><span>metadata &lt;- read.csv(</span></code><code><span>  <span>file</span> = <span>"atac_v1_pbmc_10k_singlecell.csv"</span>,</span></code><code><span>  header = <span>TRUE</span>,</span></code><code><span>  row.names = <span>1</span></span></code><code><span>)</span></code><code><span><br></span></code><code><span>chrom_assay &lt;- CreateChromatinAssay(</span></code><code><span>  counts = counts,</span></code><code><span>  sep = c(<span>":"</span>, <span>"-"</span>),</span></code><code><span>  fragments = <span>'atac_v1_pbmc_10k_fragments.tsv.gz'</span>,</span></code><code><span>  min.cells = <span>10</span>,</span></code><code><span>  min.features = <span>200</span></span></code><code><span>)</span></code><code><span><br></span></code><code><span>pbmc &lt;- CreateSeuratObject(</span></code><code><span>  counts = chrom_assay,</span></code><code><span>  assay = <span>"peaks"</span>,</span></code><code><span>  meta.data = metadata</span></code><code><span>)</span></code><code><span><br></span></code><code><span><span># extract gene annotations from EnsDb</span></span></code><code><span>annotations &lt;- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v75)</span></code><code><span><br></span></code><code><span><span># change to UCSC style since the data was mapped to hg19</span></span></code><code><span>seqlevels(annotations) &lt;- paste0(<span>'chr'</span>, seqlevels(annotations))</span></code><code><span>genome(annotations) &lt;- <span>"hg19"</span></span></code><code><span><span># add the gene information to the object</span></span></code><code><span>Annotation(pbmc) &lt;- annotations</span></code><code><span><br></span></code><code><span><span># compute nucleosome signal score per cell</span></span></code><code><span>pbmc &lt;- NucleosomeSignal(<span>object</span> = pbmc)</span></code><code><span><br></span></code><code><span><span># compute TSS enrichment score per cell</span></span></code><code><span>pbmc &lt;- TSSEnrichment(<span>object</span> = pbmc, <span>fast</span> = <span>FALSE</span>)</span></code><code><span><br></span></code><code><span><span># add blacklist ratio and fraction of reads in peaks</span></span></code><code><span>pbmc$pct_reads_in_peaks &lt;- pbmc$peak_region_fragments / pbmc$passed_filters * <span>100</span></span></code><code><span>pbmc$blacklist_ratio &lt;- pbmc$blacklist_region_fragments / pbmc$peak_region_fragments</span></code><code><span><br></span></code><code><span>DensityScatter(pbmc, x = <span>'nCount_peaks'</span>, y = <span>'TSS.enrichment'</span>, log_x = <span>TRUE</span>, quantiles = <span>TRUE</span>)</span></code><code><span>pbmc$high.tss &lt;- ifelse(pbmc$TSS.enrichment &gt; <span>3</span>, <span>'High'</span>, <span>'Low'</span>)</span></code><code><span>TSSPlot(pbmc, group.by = <span>'high.tss'</span>) + NoLegend()</span></code><code><span>pbmc$nucleosome_group &lt;- ifelse(pbmc$nucleosome_signal &gt; <span>4</span>, <span>'NS &gt; 4'</span>, <span>'NS &lt; 4'</span>)</span></code><code><span>FragmentHistogram(<span>object</span> = pbmc, group.by = <span>'nucleosome_group'</span>)</span></code><code><span>VlnPlot(</span></code><code><span>  <span>object</span> = pbmc,</span></code><code><span>  features = c(<span>'nCount_peaks'</span>, <span>'TSS.enrichment'</span>, <span>'blacklist_ratio'</span>, <span>'nucleosome_signal'</span>, <span>'pct_reads_in_peaks'</span>),</span></code><code><span>  pt.size = <span>0.1</span>,</span></code><code><span>  ncol = <span>5</span></span></code><code><span>)</span></code><code><span>pbmc &lt;- subset(</span></code><code><span>  x = pbmc,</span></code><code><span>  subset = nCount_peaks &gt; <span>3000</span> &amp;</span></code><code><span>    nCount_peaks &lt; <span>30000</span> &amp;</span></code><code><span>    pct_reads_in_peaks &gt; <span>15</span> &amp;</span></code><code><span>    blacklist_ratio &lt; <span>0.05</span> &amp;</span></code><code><span>    nucleosome_signal &lt; <span>4</span> &amp;</span></code><code><span>    TSS.enrichment &gt; <span>3</span></span></code><code><span>)</span></code><code><span>pbmc</span></code><code><span>pbmc &lt;- RunTFIDF(pbmc)</span></code><code><span>pbmc &lt;- FindTopFeatures(pbmc, min.cutoff = <span>'q0'</span>)</span></code><code><span>pbmc &lt;- RunSVD(pbmc)</span></code><code><span>DepthCor(pbmc)</span></code><code><span>pbmc &lt;- RunUMAP(<span>object</span> = pbmc, reduction = <span>'lsi'</span>, dims = <span>2</span>:<span>30</span>)</span></code><code><span>pbmc &lt;- FindNeighbors(<span>object</span> = pbmc, reduction = <span>'lsi'</span>, dims = <span>2</span>:<span>30</span>)</span></code><code><span>pbmc &lt;- FindClusters(<span>object</span> = pbmc, verbose = <span>FALSE</span>, algorithm = <span>3</span>)</span></code><code><span>DimPlot(<span>object</span> = pbmc, label = <span>TRUE</span>) + NoLegend()</span></code><code><span>gene.activities &lt;- GeneActivity(pbmc)</span></code><code><span><span># add the gene activity matrix to the Seurat object as a new assay and normalize it</span></span></code><code><span>pbmc[[<span>'RNA'</span>]] &lt;- CreateAssayObject(counts = gene.activities)</span></code><code><span>pbmc &lt;- NormalizeData(</span></code><code><span>  <span>object</span> = pbmc,</span></code><code><span>  assay = <span>'RNA'</span>,</span></code><code><span>  normalization.method = <span>'LogNormalize'</span>,</span></code><code><span>  scale.factor = <span>median</span>(pbmc$nCount_RNA)</span></code><code><span>)</span></code><code><span><br></span></code><code><span>DefaultAssay(pbmc) &lt;- <span>'RNA'</span></span></code><code><span><br></span></code><code><span>FeaturePlot(</span></code><code><span>  <span>object</span> = pbmc,</span></code><code><span>  features = c(<span>'MS4A1'</span>, <span>'CD3D'</span>, <span>'LEF1'</span>, <span>'NKG7'</span>, <span>'TREM1'</span>, <span>'LYZ'</span>),</span></code><code><span>  pt.size = <span>0.1</span>,</span></code><code><span>  max.cutoff = <span>'q95'</span>,</span></code><code><span>  ncol = <span>3</span></span></code><code><span>)</span></code><code><span><span># Load the pre-processed scRNA-seq data for PBMCs</span></span></code><code><span>pbmc_rna &lt;- readRDS(<span>"pbmc_10k_v3.rds"</span>)</span></code><code><span><br></span></code><code><span>transfer.anchors &lt;- FindTransferAnchors(</span></code><code><span>  <span>reference</span> = pbmc_rna,</span></code><code><span>  <span>query</span> = pbmc,</span></code><code><span>  reduction = <span>'cca'</span></span></code><code><span>)</span></code><code><span><br></span></code><code><span>predicted.labels &lt;- TransferData(</span></code><code><span>  anchorset = transfer.anchors,</span></code><code><span>  refdata = pbmc_rna$celltype,</span></code><code><span>  weight.reduction = pbmc[[<span>'lsi'</span>]],</span></code><code><span>  dims = <span>2</span>:<span>30</span></span></code><code><span>)</span></code><code><span><br></span></code><code><span>pbmc &lt;- AddMetaData(<span>object</span> = pbmc, metadata = predicted.labels)</span></code><code><span>plot1 &lt;- DimPlot(</span></code><code><span>  <span>object</span> = pbmc_rna,</span></code><code><span>  group.by = <span>'celltype'</span>,</span></code><code><span>  label = <span>TRUE</span>,</span></code><code><span>  repel = <span>TRUE</span>) + NoLegend() + ggtitle(<span>'scRNA-seq'</span>)</span></code><code><span><br></span></code><code><span>plot2 &lt;- DimPlot(</span></code><code><span>  <span>object</span> = pbmc,</span></code><code><span>  group.by = <span>'predicted.id'</span>,</span></code><code><span>  label = <span>TRUE</span>,</span></code><code><span>  repel = <span>TRUE</span>) + NoLegend() + ggtitle(<span>'scATAC-seq'</span>)</span></code><code><span><br></span></code><code><span>plot1 + plot2</span></code><code><span><br></span></code><code><span><span># replace each label with its most likely prediction</span></span></code><code><span><span>for</span>(i <span>in</span> <span>levels</span>(pbmc)) {</span></code><code><span>  cells_to_reid &lt;- WhichCells(pbmc, idents = i)</span></code><code><span>  newid &lt;- <span>names</span>(which.max(<span>table</span>(pbmc$predicted.id[cells_to_reid])))</span></code><code><span>  Idents(pbmc, cells = cells_to_reid) &lt;- newid</span></code><code><span>}</span></code><code><span><span># change back to working with peaks instead of gene activities</span></span></code><code><span>DefaultAssay(pbmc) &lt;- <span>'peaks'</span></span></code><code><span><br></span></code><code><span>da_peaks &lt;- FindMarkers(</span></code><code><span>  <span>object</span> = pbmc,</span></code><code><span>  ident<span>.1</span> = <span>"CD4 Naive"</span>,</span></code><code><span>  ident<span>.2</span> = <span>"CD14+ Monocytes"</span>,</span></code><code><span>  test.use = <span>'LR'</span>,</span></code><code><span>  latent.vars = <span>'nCount_peaks'</span></span></code><code><span>)</span></code><code><span><br></span></code><code><span><span>head</span>(da_peaks)</span></code><code><span>plot1 &lt;- VlnPlot(</span></code><code><span>  <span>object</span> = pbmc,</span></code><code><span>  features = rownames(da_peaks)[<span>1</span>],</span></code><code><span>  pt.size = <span>0.1</span>,</span></code><code><span>  idents = c(<span>"CD4 Naive"</span>,<span>"CD14+ Monocytes"</span>)</span></code><code><span>)</span></code><code><span>plot2 &lt;- FeaturePlot(</span></code><code><span>  <span>object</span> = pbmc,</span></code><code><span>  features = rownames(da_peaks)[<span>1</span>],</span></code><code><span>  pt.size = <span>0.1</span></span></code><code><span>)</span></code><code><span><br></span></code><code><span>plot1 | plot2</span></code><code><span>fc &lt;- FoldChange(pbmc, ident<span>.1</span> = <span>"CD4 Naive"</span>, ident<span>.2</span> = <span>"CD14+ Monocytes"</span>)</span></code><code><span><span># order by fold change</span></span></code><code><span>fc &lt;- fc[<span>order</span>(fc$avg_log2FC, decreasing = <span>TRUE</span>), ]</span></code><code><span><span>head</span>(fc)</span></code><code><span>open_cd4naive &lt;- rownames(da_peaks[da_peaks$avg_log2FC &gt; <span>3</span>, ])</span></code><code><span>open_cd14mono &lt;- rownames(da_peaks[da_peaks$avg_log2FC &lt; <span>-3</span>, ])</span></code><code><span><br></span></code><code><span>closest_genes_cd4naive &lt;- ClosestFeature(pbmc, regions = open_cd4naive)</span></code><code><span>closest_genes_cd14mono &lt;- ClosestFeature(pbmc, regions = open_cd14mono)</span></code><code><span><span>head</span>(closest_genes_cd4naive)</span></code><code><span><span>head</span>(closest_genes_cd14mono)</span></code><code><span><br></span></code><code><span><span># set plotting order</span></span></code><code><span><span>levels</span>(pbmc) &lt;- c(<span>"CD4 Naive"</span>,<span>"CD4 Memory"</span>,<span>"CD8 Naive"</span>,<span>"CD8 effector"</span>,<span>"Double negative T cell"</span>,<span>"NK dim"</span>,<span>"pre-B cell"</span>,<span>'B cell progenitor'</span>,<span>"pDC"</span>,<span>"Dendritic cell"</span>,<span>"CD14+ Monocytes"</span>,<span>'CD16+ Monocytes'</span>)</span></code><code><span><br></span></code><code><span>CoveragePlot(</span></code><code><span>  <span>object</span> = pbmc,</span></code><code><span>  region = rownames(da_peaks)[<span>1</span>],</span></code><code><span>  extend.upstream = <span>40000</span>,</span></code><code><span>  extend.downstream = <span>20000</span></span></code><code><span>)</span></code><code><span><br></span></code><code><span><br></span></code><code><span><span>#参考资料：https://stuartlab.org/signac/articles/pbmc_vignette</span></span></code><code><span><br></span></code></pre></section></section><p><img data-galleryid="" data-ratio="1.0629629629629629" data-s="300,640" data-type="jpeg" data-w="1080" data-src="https://mmbiz.qpic.cn/mmbiz_png/kZ1wdgAscBr9WKRw7ibjDQ9yUBhqhSJxj6ia5iaxicW6I3gD75SeIJQbsRTScgbyToMXDeRBc19ThbgesxIvEanIpA/640?wx_fmt=png" src="https://mmbiz.qpic.cn/mmbiz_png/kZ1wdgAscBr9WKRw7ibjDQ9yUBhqhSJxj6ia5iaxicW6I3gD75SeIJQbsRTScgbyToMXDeRBc19ThbgesxIvEanIpA/640?wx_fmt=png"></p><p><img data-galleryid="" data-ratio="1.0629629629629629" data-s="300,640" data-type="jpeg" data-w="1080" data-src="https://mmbiz.qpic.cn/mmbiz_png/kZ1wdgAscBr9WKRw7ibjDQ9yUBhqhSJxjGdLzgHYO5e3b2bBRILFendmtia7noFIxhTEL9mmicBGugAXgMWicLE8SA/640?wx_fmt=png" src="https://mmbiz.qpic.cn/mmbiz_png/kZ1wdgAscBr9WKRw7ibjDQ9yUBhqhSJxjGdLzgHYO5e3b2bBRILFendmtia7noFIxhTEL9mmicBGugAXgMWicLE8SA/640?wx_fmt=png"></p><p><img data-galleryid="" data-ratio="1.0629629629629629" data-s="300,640" data-type="jpeg" data-w="1080" data-src="https://mmbiz.qpic.cn/mmbiz_png/kZ1wdgAscBr9WKRw7ibjDQ9yUBhqhSJxjTb7JQIyRuib5XLsCicvRRicWemiaJjQyEANXvvNB1rUb24Hl0kNMCNvFSw/640?wx_fmt=png" src="https://mmbiz.qpic.cn/mmbiz_png/kZ1wdgAscBr9WKRw7ibjDQ9yUBhqhSJxjTb7JQIyRuib5XLsCicvRRicWemiaJjQyEANXvvNB1rUb24Hl0kNMCNvFSw/640?wx_fmt=png"></p><p><img data-galleryid="" data-ratio="0.5990740740740741" data-s="300,640" data-type="jpeg" data-w="1080" data-src="https://mmbiz.qpic.cn/mmbiz_png/kZ1wdgAscBr9WKRw7ibjDQ9yUBhqhSJxjodEEfzXhe9jjkOk9tObqBf9C8ia1EbphE9rbYVM5VEUrhicNnb8pIcrw/640?wx_fmt=png" src="https://mmbiz.qpic.cn/mmbiz_png/kZ1wdgAscBr9WKRw7ibjDQ9yUBhqhSJxjodEEfzXhe9jjkOk9tObqBf9C8ia1EbphE9rbYVM5VEUrhicNnb8pIcrw/640?wx_fmt=png"></p><p><img data-galleryid="" data-ratio="1.0629629629629629" data-s="300,640" data-type="jpeg" data-w="1080" data-src="https://mmbiz.qpic.cn/mmbiz_png/kZ1wdgAscBr9WKRw7ibjDQ9yUBhqhSJxj1icQQ7bs4zVgtKkLJnaSiaTArnrcuxicAINv0TPPInRReWF3BsZc1dzwg/640?wx_fmt=png" src="https://mmbiz.qpic.cn/mmbiz_png/kZ1wdgAscBr9WKRw7ibjDQ9yUBhqhSJxj1icQQ7bs4zVgtKkLJnaSiaTArnrcuxicAINv0TPPInRReWF3BsZc1dzwg/640?wx_fmt=png"></p><p><img data-galleryid="" data-ratio="1.0629629629629629" data-s="300,640" data-type="jpeg" data-w="1080" data-src="https://mmbiz.qpic.cn/mmbiz_png/kZ1wdgAscBr9WKRw7ibjDQ9yUBhqhSJxjfqA8152b8x8OnPBYsnr8pqCzqkuQs0S6kkN30gibnFDTdfoteNHzrKA/640?wx_fmt=png" src="https://mmbiz.qpic.cn/mmbiz_png/kZ1wdgAscBr9WKRw7ibjDQ9yUBhqhSJxjfqA8152b8x8OnPBYsnr8pqCzqkuQs0S6kkN30gibnFDTdfoteNHzrKA/640?wx_fmt=png"></p><p><img data-galleryid="" data-ratio="1.0629629629629629" data-s="300,640" data-type="jpeg" data-w="1080" data-src="https://mmbiz.qpic.cn/mmbiz_png/kZ1wdgAscBr9WKRw7ibjDQ9yUBhqhSJxjrRPcYhEqfDKjuUyWjOXN6k3Hepy1fficauibVtK8l5eczyicYaHWicRoYw/640?wx_fmt=png" src="https://mmbiz.qpic.cn/mmbiz_png/kZ1wdgAscBr9WKRw7ibjDQ9yUBhqhSJxjrRPcYhEqfDKjuUyWjOXN6k3Hepy1fficauibVtK8l5eczyicYaHWicRoYw/640?wx_fmt=png"></p><p><img data-galleryid="" data-ratio="1.0629629629629629" data-s="300,640" data-type="jpeg" data-w="1080" data-src="https://mmbiz.qpic.cn/mmbiz_png/kZ1wdgAscBr9WKRw7ibjDQ9yUBhqhSJxjiaqjdUAxDdwjibRK58h3j6LPiaL2rIsWfibRjvNcf8dRJyX3XWPyfhicxkw/640?wx_fmt=png" src="https://mmbiz.qpic.cn/mmbiz_png/kZ1wdgAscBr9WKRw7ibjDQ9yUBhqhSJxjiaqjdUAxDdwjibRK58h3j6LPiaL2rIsWfibRjvNcf8dRJyX3XWPyfhicxkw/640?wx_fmt=png"></p><p><img data-galleryid="" data-ratio="1.0629629629629629" data-s="300,640" data-type="jpeg" data-w="1080" data-src="https://mmbiz.qpic.cn/mmbiz_png/kZ1wdgAscBr9WKRw7ibjDQ9yUBhqhSJxjhnPCy8YgCSgLibjUWYgic3XYNoJn8cK0ibjH3iaNRmQzWdpm6v1iclMOBEg/640?wx_fmt=png" src="https://mmbiz.qpic.cn/mmbiz_png/kZ1wdgAscBr9WKRw7ibjDQ9yUBhqhSJxjhnPCy8YgCSgLibjUWYgic3XYNoJn8cK0ibjH3iaNRmQzWdpm6v1iclMOBEg/640?wx_fmt=png"></p><p><img data-galleryid="" data-ratio="1.0629629629629629" data-s="300,640" data-type="jpeg" data-w="1080" data-src="https://mmbiz.qpic.cn/mmbiz_png/kZ1wdgAscBr9WKRw7ibjDQ9yUBhqhSJxjKGkuCIa4lm4pmkXxy8kCcjtOMCtpRSibNjPST08dkPmheCmj3xqkNQA/640?wx_fmt=png" src="https://mmbiz.qpic.cn/mmbiz_png/kZ1wdgAscBr9WKRw7ibjDQ9yUBhqhSJxjKGkuCIa4lm4pmkXxy8kCcjtOMCtpRSibNjPST08dkPmheCmj3xqkNQA/640?wx_fmt=png"><span></span></p><p><mp-style-type data-value="3"></mp-style-type></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/qqi6gAa7w-4bFIed9duOAw",target="_blank" rel="noopener noreferrer">原文链接</a>
