---
title: "脚本封装学习----以单细胞基础分析为例（R代码）"
date: 2024-12-05T23:58:00Z
draft: ["false"]
tags: [
  "fetched",
  "单细胞空间交响乐"
]
categories: ["Acdemic"]
---
脚本封装学习----以单细胞基础分析为例（R代码） by 单细胞空间交响乐
------
<div><h4 data-line="0">作者，Evil Genius</h4><h4 data-line="2">今天呢，教大家一个简单的内容，关于脚本封装，其中我们封装脚本需要实现的目标是单细胞分析脚本，包括质控，去除双细胞，数据标准化处理，多样本整合，去除批次，降维聚类 ，差异富集。</h4><p><img data-imgfileid="100010252" data-ratio="1.101010101010101" data-type="other" data-w="693" data-src="https://mmbiz.qpic.cn/mmbiz_jpg/srXAibe95Mmma1wBnhmEbyRWia3nQOwAJhNpP8ia20rrWhxW1wWXiblZBGcL3l8q0dZUSNzYWqibe9K6tBQ10Mdib6iag/640?wx_fmt=jpeg" src="https://mmbiz.qpic.cn/mmbiz_jpg/srXAibe95Mmma1wBnhmEbyRWia3nQOwAJhNpP8ia20rrWhxW1wWXiblZBGcL3l8q0dZUSNzYWqibe9K6tBQ10Mdib6iag/640?wx_fmt=jpeg"><br></p><h4 data-line="5">封装脚本学会之后都是一通百通的，是在公司常用的方法，也是基本功，大家送给公司的分析都是这么做出来的。</h4><h4 data-line="7"><span>R一般用argparse包进行传参，并且脚本一般需要定义大量的函数。</span></h4><h4 data-line="9">其中我们需要封装的方法包括Seurat、clusterProfiler、Harmony、DoubletFinder。</h4><h4 data-line="11">下面是代码示例,封装脚本都是一气呵成的（上课给大家讲过，有条件的多多学习），<span>一条龙出所有结果。</span></h4><pre><code>library(argparse)<br>library(Seurat)            <span># 用于单细胞RNA-seq分析</span><br>library(dplyr)             <span># 数据操作</span><br>library(ggplot2)           <span># 可视化</span><br>library(clusterProfiler)   <span># 富集分析</span><br>library(org.Hs.eg.db)      <span># 基因注释数据库</span><br>library(Harmony)           <span># 用于批次效应去除</span><br>library(DoubletFinder)     <span># 双细胞预测</span><br><br><span># 创建命令行解析器</span><br>parser &lt;- ArgumentParser(description = <span>'Single-cell RNA-seq analysis pipeline'</span>)<br><br><span># 添加命令行参数</span><br>parser$add_argument(<span>'--sample_paths'</span>, type = <span>'character'</span>, nargs = <span>'+'</span>, required = <span>TRUE</span>, help = <span>'Paths to the sample directories'</span>)<br>parser$add_argument(<span>'--output_dir'</span>, type = <span>'character'</span>, <span>default</span> = <span>'output'</span>, help = <span>'Directory to save the results'</span>)<br>parser$add_argument(<span>'--resolution'</span>, type = <span>'numeric'</span>, <span>default</span> = <span>0.5</span>, help = <span>'Clustering resolution (default: 0.5)'</span>)<br>parser$add_argument(<span>'--dims'</span>, type = <span>'integer'</span>, <span>default</span> = <span>1</span>:<span>20</span>, help = <span>'PCA dimensions for clustering (default: 1:20)'</span>)<br>parser$add_argument(<span>'--use_doublet_finder'</span>, type = <span>'logical'</span>, <span>default</span> = <span>TRUE</span>, help = <span>'Whether to use DoubletFinder for doublet detection (default: TRUE)'</span>)<br><br>args &lt;- parser$parse_args()<br><br><span># 1. 数据加载与质控函数</span><br>load_and_qc &lt;- <span><span>function</span><span>(sample_paths, min_cells = <span>3</span>, min_features = <span>200</span>, max_mt_percent = <span>5</span>, use_doublet_finder = TRUE)</span> </span>{<br>  samples &lt;- lapply(sample_paths, <span><span>function</span><span>(path)</span> </span>{<br>    data &lt;- Read10X(data.dir = path)<br>    seurat_obj &lt;- CreateSeuratObject(counts = data, project = <span>"SingleCell"</span>, min.cells = min_cells, min.features = min_features)<br>    seurat_obj[[<span>"percent.mt"</span>]] &lt;- PercentageFeatureSet(seurat_obj, pattern = <span>"^MT-"</span>)<br>    <br>    <span># 去除低质量细胞和双细胞</span><br>    seurat_obj &lt;- subset(seurat_obj, subset = nFeature_RNA &gt; <span>200</span> &amp; nFeature_RNA &lt; <span>2500</span> &amp; percent.mt &lt; max_mt_percent)<br>    <br>    <span># 如果使用DoubletFinder</span><br>    <span>if</span> (use_doublet_finder) {<br>      seurat_obj &lt;- doublet_removal(seurat_obj)<br>    }<br>    <br>    <span>return</span>(seurat_obj)<br>  })<br>  <br>  <span>return</span>(samples)<br>}<br><br><span># 2. 基于指标（nFeature_RNA 和 nCount_RNA）去除双细胞的函数</span><br>doublet_removal &lt;- <span><span>function</span><span>(seurat_obj, min_features = <span>200</span>, max_features = <span>2500</span>, min_count = <span>500</span>, max_count = <span>10000</span>)</span> </span>{<br>  <span># 通过检查基因数和转录本数去除双细胞</span><br>  seurat_obj &lt;- subset(seurat_obj, subset = nFeature_RNA &gt; min_features &amp; nFeature_RNA &lt; max_features)<br>  seurat_obj &lt;- subset(seurat_obj, subset = nCount_RNA &gt; min_count &amp; nCount_RNA &lt; max_count)<br>  <br>  <span># 可选：使用 DoubletFinder 去除预测的双细胞</span><br>  seurat_obj &lt;- DoubletFinder::doubletFinder_v3(seurat_obj, PCs = <span>1</span>:<span>20</span>, pN = <span>0.25</span>, pK = <span>0.05</span>, nExp = round(<span>0.05</span> * ncol(seurat_obj)))<br>  <br>  <span># 根据DoubletFinder的预测，过滤掉双细胞</span><br>  seurat_obj &lt;- subset(seurat_obj, subset = DF.classifications != <span>"Doublet"</span>)<br>  <br>  <span>return</span>(seurat_obj)<br>}<br><br><span># 3. 数据标准化和高变基因选择函数</span><br>normalize_and_find_variable_genes &lt;- <span><span>function</span><span>(seurat_objs)</span> </span>{<br>  <span>for</span> (i in <span>1</span>:length(seurat_objs)) {<br>    seurat_objs[[i]] &lt;- NormalizeData(seurat_objs[[i]])<br>    seurat_objs[[i]] &lt;- FindVariableFeatures(seurat_objs[[i]], selection.method = <span>"vst"</span>, nfeatures = <span>2000</span>)<br>  }<br>  <span>return</span>(seurat_objs)<br>}<br><br><span># 4. 多样本整合函数</span><br>integrate_samples &lt;- <span><span>function</span><span>(seurat_objs)</span> </span>{<br>  <span># 合并样本</span><br>  merged_samples &lt;- merge(seurat_objs[[<span>1</span>]], y = seurat_objs[<span>2</span>:length(seurat_objs)], add.cell.ids = paste0(<span>"Sample"</span>, <span>1</span>:length(seurat_objs)))<br>  merged_samples &lt;- NormalizeData(merged_samples)<br>  merged_samples &lt;- FindVariableFeatures(merged_samples, selection.method = <span>"vst"</span>, nfeatures = <span>2000</span>)<br>  <br>  <span>return</span>(merged_samples)<br>}<br><br><span># 5. 批次效应去除函数</span><br>remove_batch_effects &lt;- <span><span>function</span><span>(merged_samples)</span> </span>{<br>  merged_samples &lt;- RunPCA(merged_samples, features = VariableFeatures(object = merged_samples))<br>  merged_samples &lt;- RunHarmony(merged_samples, group.by.vars = <span>"orig.ident"</span>)  <span># "orig.ident" 是批次信息</span><br>  <br>  <span>return</span>(merged_samples)<br>}<br><br><span># 6. 降维和聚类函数</span><br>perform_clustering &lt;- <span><span>function</span><span>(merged_samples, dims = <span>1</span>:<span>20</span>, resolution = <span>0.5</span>)</span> </span>{<br>  merged_samples &lt;- FindNeighbors(merged_samples, dims = dims)<br>  merged_samples &lt;- FindClusters(merged_samples, resolution = resolution)<br>  <br>  <span># 运行UMAP降维</span><br>  merged_samples &lt;- RunUMAP(merged_samples, dims = dims)<br>  <br>  <span>return</span>(merged_samples)<br>}<br><br><span># 7. 差异表达分析函数</span><br>find_differential_expression &lt;- <span><span>function</span><span>(merged_samples)</span> </span>{<br>  <span># 对所有聚类进行差异表达分析</span><br>  cluster_ids &lt;- unique(merged_samples$seurat_clusters)<br>  all_results &lt;- <span>list</span>()<br>  <br>  <span>for</span> (i in <span>1</span>:(length(cluster_ids) - <span>1</span>)) {<br>    <span>for</span> (j in (i + <span>1</span>):length(cluster_ids)) {<br>      cluster_1 &lt;- cluster_ids[i]<br>      cluster_2 &lt;- cluster_ids[j]<br>      <br>      cluster_markers &lt;- FindMarkers(merged_samples, ident<span>.1</span> = cluster_1, ident<span>.2</span> = cluster_2)<br>      all_results[[paste0(<span>"Cluster_"</span>, cluster_1, <span>"_vs_Cluster_"</span>, cluster_2)]] &lt;- cluster_markers<br>    }<br>  }<br>  <br>  <span>return</span>(all_results)<br>}<br><br><span># 8. 富集分析函数</span><br>perform_go_enrichment &lt;- <span><span>function</span><span>(diff_genes)</span> </span>{<br>  <span># 选取显著差异基因（p值&lt;0.05）</span><br>  gene_list &lt;- rownames(diff_genes)[which(diff_genes$p_val_adj &lt; <span>0.05</span>)]<br>  <br>  <span># 转换基因ID为Entrez ID</span><br>  gene_symbols &lt;- bitr(gene_list, fromType = <span>"SYMBOL"</span>, toType = <span>"ENTREZID"</span>, OrgDb = org.Hs.eg.db)<br>  <br>  <span># 进行GO富集分析</span><br>  go_results &lt;- enrichGO(gene = gene_symbols$ENTREZID, OrgDb = org.Hs.eg.db, keyType = <span>"ENTREZID"</span>, ont = <span>"BP"</span>, pAdjustMethod = <span>"BH"</span>, qvalueCutoff = <span>0.05</span>)<br>  <br>  <span>return</span>(go_results)<br>}<br><br><span># 主分析函数：执行所有分析步骤</span><br>single_cell_analysis &lt;- <span><span>function</span><span>(sample_paths, resolution = <span>0.5</span>, dims = <span>1</span>:<span>20</span>, output_dir = <span>"output"</span>, use_doublet_finder = TRUE)</span> </span>{<br>  <span># 1. 数据加载与质控</span><br>  samples &lt;- load_and_qc(sample_paths, use_doublet_finder = use_doublet_finder)<br>  <br>  <span># 2. 数据标准化与高变基因选择</span><br>  samples &lt;- normalize_and_find_variable_genes(samples)<br>  <br>  <span># 3. 多样本整合</span><br>  merged_samples &lt;- integrate_samples(samples)<br>  <br>  <span># 4. 批次效应去除</span><br>  merged_samples &lt;- remove_batch_effects(merged_samples)<br>  <br>  <span># 5. 降维与聚类</span><br>  merged_samples &lt;- perform_clustering(merged_samples, dims = dims, resolution = resolution)<br>  <br>  <span># 6. 差异表达分析</span><br>  cluster_markers_list &lt;- find_differential_expression(merged_samples)<br>  <br>  <span># 7. 富集分析（对每个差异基因集进行富集分析）</span><br>  go_results_list &lt;- <span>list</span>()<br>  <span>for</span> (key in names(cluster_markers_list)) {<br>    cluster_markers &lt;- cluster_markers_list[[key]]<br>    go_results &lt;- perform_go_enrichment(cluster_markers)<br>    go_results_list[[key]] &lt;- go_results<br>  }<br>  <br>  <span># 保存结果</span><br>  <span>if</span> (!dir.exists(output_dir)) {<br>    dir.create(output_dir)<br>  }<br>  <br>  <span># 保存 Seurat 对象、差异分析结果和富集分析结果</span><br>  saveRDS(merged_samples, file.path(output_dir, <span>"merged_samples.rds"</span>))<br>  saveRDS(cluster_markers_list, file.path(output_dir, <span>"cluster_markers_list.rds"</span>))<br>  saveRDS(go_results_list, file.path(output_dir, <span>"go_results_list.rds"</span>))<br>  <br>  <span># 可视化结果</span><br>  DimPlot(merged_samples, reduction = <span>"umap"</span>, group.by = <span>"seurat_clusters"</span>) + <br>    ggsave(file.path(output_dir, <span>"umap_plot.png"</span>))<br>  <br>  <span># GO 富集分析可视化</span><br>  <span>for</span> (key in names(go_results_list)) {<br>    dotplot(go_results_list[[key]]) + <br>      ggsave(file.path(output_dir, paste0(key, <span>"_go_enrichment_plot.png"</span>)))<br>  }<br>}<br><br><span># 执行主分析函数</span><br>single_cell_analysis(sample_paths = args$sample_paths, <br>                     resolution = args$resolution, <br>                     dims = args$dims, <br>                     output_dir = args$output_dir,<br>                     use_doublet_finder = args$use_doublet_finder)<br></code></pre><h4 data-line="198">在命令行的运行示例</h4><pre><code>Rscript single_cell_analysis.R --sample_paths /path/to/sample1 /path/to/sample2 --output_dir /path/to/output --resolution <span>0.6</span> --dims <span>1</span>:<span>20</span> --use_doublet_finder <span>TRUE</span><br></code></pre><h4 data-line="202">qsub、slurm等投递到后台，等待结果即可。</h4><h4 data-line="204">其中参数的意义</h4><ul><li><p><span>--sample_paths</span>：指定你要分析的样本目录的路径。你可以传递多个样本目录路径，这样脚本会合并多个样本进行分析。例如：/path/to/sample1 和 /path/to/sample2 是样本目录的路径。每个目录下应包含单细胞数据（例如 matrix.mtx, features.tsv, barcodes.tsv 文件）。</p></li><li><p><span>--output_dir</span>：指定分析结果的输出目录。例如：/path/to/output 是保存分析结果和图像的文件夹。如果该文件夹不存在，脚本会自动创建。</p></li><li><p><span>--resolution</span>：指定聚类的分辨率参数。resolution 控制聚类的数量，较高的值会导致更多的聚类。默认值是 0.5，但你可以根据数据的特点调整。例如：--resolution 0.6。</p></li><li><p><span>--dims</span>：指定用于降维和聚类的 PCA 维度范围。例如，使用前 30 个主成分进行聚类：--dims 1:30。</p></li><li><p><span>--use_doublet_finder</span>：是否使用 DoubletFinder 进行双细胞预测和去除。可以传入 TRUE 或 FALSE。如果为 TRUE，脚本会使用 DoubletFinder 来预测双细胞并将其去除。默认值为 TRUE。如果不希望使用 DoubletFinder，可以将其设置为 FALSE：--use_doublet_finder TRUE。</p></li></ul><h4 data-line="211">这样就会一键式分析得到单细胞基础分析的所有结果。</h4><h4 data-line="213">生活很好，有你更好</h4><section><mp-common-cpsad data-pluginname="mpcps" data-templateid="list" data-traceid="72d7b963-57d4-4fc0-ad23-89a970d58883" data-goodssouce="1" data-pid="104_3381840471" data-appuin="3860655629" data-cpsversion="v116"></mp-common-cpsad></section><p><br></p><p><mp-style-type data-value="3"></mp-style-type></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/eq8nk601ICRes4jS4GMXYQ",target="_blank" rel="noopener noreferrer">原文链接</a>
