---
title: "bbknnR 2.0发布——Seurat也能做BBKNN批次矫正"
date: 2025-06-05T03:45:32Z
draft: ["false"]
tags: [
  "fetched",
  "Eula的生信小金鱼"
]
categories: ["Acdemic"]
---
bbknnR 2.0发布——Seurat也能做BBKNN批次矫正 by Eula的生信小金鱼
------
<div><section data-tool="markdown编辑器" data-website="https://markdown.com.cn/editor" data-pm-slice="0 0 []"><h1 data-tool="markdown.com.cn编辑器"><span></span><span><span leaf=""><mp-common-clmusic data-pluginname="insertaudio" type="1" music_name="crossing field (交叉领域)" albumurl="http://wx.y.gtimg.cn/music/photo_new/T002R500x500M000002U1a1A15HIyr_3.jpg" singer="LiSA" count="0" is_vip="1" duration="249000" music_source="1" listenid="78248336074600730"></mp-common-clmusic></span></span></h1><blockquote><span><span leaf="">“</span></span><p><span leaf="">柴火的往日，烈焰的今时，皆为永不回头的生命。所行一切，火必将审问。</span></p><span></span></blockquote><p data-tool="markdown.com.cn编辑器"><span leaf="">在单细胞转录组分析中，批次效应一直是影响数据整合和下游分析的难题。BBKNN（Batch Balanced KNN）作为一种高效的批次效应校正方法，最初在 Python 的 Scanpy 框架中实现[1]。因其在构建 nearest neighbor graph 时考虑批次平衡而受到scanpy用户的广泛欢迎。然而，对于 R 语言用户而言，过去想要使用 BBKNN 需要依赖 Python 环境和 reticulate 包，增加了配置和维护的复杂性。</span></p><p data-tool="markdown.com.cn编辑器"><span leaf="">大约三年前，我写了</span><code><span leaf="">bbknnR</span></code><span leaf="">包，借助</span><code><span leaf="">RcppAnnoy</span></code><span leaf="">在R中实现了BBKNN的算法。同时提供了</span><code><span leaf="">reticulate + bbknn</span></code><span leaf="">作为使用原生python实现的可选接口。最近，得益于</span><code><span leaf="">rnndescent</span></code><span leaf="">包的诞生，我把</span><code><span leaf="">bbknnR</span></code><span leaf="">更新到了2.0，不再依赖任何python的东西。换句话说，</span><strong><span leaf="">Seurat玩家也可以无缝使用BBKNN算法来做批次矫正啦！</span></strong></p><p data-tool="markdown.com.cn编辑器"><code><span leaf="">bbknnR</span></code><span leaf="">包已经发布到CRAN，可以直接安装：</span></p><pre data-tool="markdown.com.cn编辑器"><code><span leaf="">install.packages("bbknnR")</span><span leaf=""><br></span></code></pre><h2 data-tool="markdown.com.cn编辑器"><span></span><span><span leaf="">测试准备</span></span></h2><ul><li><section><span leaf="">自组装台式机：CPU Intel i9-9820X 3.30GHz、128Gb内存、三星980Pro固态硬盘</span></section></li><li><section><span leaf="">数据集：64个人脐带血单个核细胞样本，来自https://explore.data.humancellatlas.org/</span></section></li><li><section><span leaf="">软件环境：windows11系统，R-4.4.2</span></section></li><li><section><span leaf="">R包：BPCells + Seurat 5.3.0</span></section></li></ul><p data-tool="markdown.com.cn编辑器"><span leaf="">每个关键步骤都将使用</span><code><span leaf="">Sys.time()</span></code><span leaf="">和</span><code><span leaf="">gc()</span></code><span leaf="">记录运行使用的时间和累计使用的内存。</span></p><pre data-tool="markdown.com.cn编辑器"><code><span><span leaf="">library</span></span><span leaf="">(Seurat)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(BPCells)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(dplyr)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(ggplot2)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(bbknnR)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># needs to be set for large dataset analysis</span></span><span leaf=""><br></span><span leaf="">options(future.globals.maxSize = </span><span><span leaf="">1e9</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><h2 data-tool="markdown.com.cn编辑器"><span></span><span><span leaf="">数据加载</span></span></h2><p data-tool="markdown.com.cn编辑器"><span leaf="">获取所有样本（</span><strong><span leaf="">批次</span></strong><span leaf="">）信息：</span></p><pre data-tool="markdown.com.cn编辑器"><code><span leaf="">samples &lt;- gsub(</span><span><span leaf="">"_raw.*"</span></span><span leaf="">, </span><span><span leaf="">""</span></span><span leaf="">, list.files(</span><span><span leaf="">"CordBlood/"</span></span><span leaf="">))</span><span leaf=""><br></span><span leaf="">samples</span><span leaf=""><br></span></code></pre><pre data-tool="markdown.com.cn编辑器"><code><span leaf="">  [1] "MantonCB1_HiSeq_1" "MantonCB1_HiSeq_2" "MantonCB1_HiSeq_3"</span><span leaf=""><br></span><span leaf="">  [4] "MantonCB1_HiSeq_4" "MantonCB1_HiSeq_5" "MantonCB1_HiSeq_6"</span><span leaf=""><br></span><span leaf="">  [7] "MantonCB1_HiSeq_7" "MantonCB1_HiSeq_8" "MantonCB2_HiSeq_1"</span><span leaf=""><br></span><span leaf=""> [10] "MantonCB2_HiSeq_2" "MantonCB2_HiSeq_3" "MantonCB2_HiSeq_4"</span><span leaf=""><br></span><span leaf=""> [13] "MantonCB2_HiSeq_5" "MantonCB2_HiSeq_6" "MantonCB2_HiSeq_7"</span><span leaf=""><br></span><span leaf=""> [16] "MantonCB2_HiSeq_8" "MantonCB3_HiSeq_1" "MantonCB3_HiSeq_2"</span><span leaf=""><br></span><span leaf=""> [19] "MantonCB3_HiSeq_3" "MantonCB3_HiSeq_4" "MantonCB3_HiSeq_5"</span><span leaf=""><br></span><span leaf=""> [22] "MantonCB3_HiSeq_6" "MantonCB3_HiSeq_7" "MantonCB3_HiSeq_8"</span><span leaf=""><br></span><span leaf=""> [25] "MantonCB4_HiSeq_1" "MantonCB4_HiSeq_2" "MantonCB4_HiSeq_3"</span><span leaf=""><br></span><span leaf=""> [28] "MantonCB4_HiSeq_4" "MantonCB4_HiSeq_5" "MantonCB4_HiSeq_6"</span><span leaf=""><br></span><span leaf=""> [31] "MantonCB4_HiSeq_7" "MantonCB4_HiSeq_8" "MantonCB5_HiSeq_1"</span><span leaf=""><br></span><span leaf=""> [34] "MantonCB5_HiSeq_2" "MantonCB5_HiSeq_3" "MantonCB5_HiSeq_4"</span><span leaf=""><br></span><span leaf=""> [37] "MantonCB5_HiSeq_5" "MantonCB5_HiSeq_6" "MantonCB5_HiSeq_7"</span><span leaf=""><br></span><span leaf=""> [40] "MantonCB5_HiSeq_8" "MantonCB6_HiSeq_1" "MantonCB6_HiSeq_2"</span><span leaf=""><br></span><span leaf=""> [43] "MantonCB6_HiSeq_3" "MantonCB6_HiSeq_4" "MantonCB6_HiSeq_5"</span><span leaf=""><br></span><span leaf=""> [46] "MantonCB6_HiSeq_6" "MantonCB6_HiSeq_7" "MantonCB6_HiSeq_8"</span><span leaf=""><br></span><span leaf=""> [49] "MantonCB7_HiSeq_1" "MantonCB7_HiSeq_2" "MantonCB7_HiSeq_3"</span><span leaf=""><br></span><span leaf=""> [52] "MantonCB7_HiSeq_4" "MantonCB7_HiSeq_5" "MantonCB7_HiSeq_6"</span><span leaf=""><br></span><span leaf=""> [55] "MantonCB7_HiSeq_7" "MantonCB7_HiSeq_8" "MantonCB8_HiSeq_1"</span><span leaf=""><br></span><span leaf=""> [58] "MantonCB8_HiSeq_2" "MantonCB8_HiSeq_3" "MantonCB8_HiSeq_4"</span><span leaf=""><br></span><span leaf=""> [61] "MantonCB8_HiSeq_5" "MantonCB8_HiSeq_6" "MantonCB8_HiSeq_7"</span><span leaf=""><br></span><span leaf=""> [64] "MantonCB8_HiSeq_8"</span><span leaf=""><br></span></code></pre><p data-tool="markdown.com.cn编辑器"><span leaf="">用</span><code><span leaf="">open_matrix_10x_hdf5</span></code><span leaf="">来读取每个样本的</span><code><span leaf="">raw_feature_bc_matrix.h5</span></code><span leaf="">。把每个样本的基因ID替换成symbol，给barcode加上样本前缀，并过滤细胞。最后，用</span><code><span leaf="">cbind</span></code><span leaf="">把所有矩阵合并成一个。</span></p><pre data-tool="markdown.com.cn编辑器"><code><span leaf="">genes &lt;- hdf5r.Extra::h5Read(</span><span leaf=""><br></span><span leaf="">  list.files(</span><span><span leaf="">"CordBlood/"</span></span><span leaf="">, full.names = </span><span><span leaf="">TRUE</span></span><span leaf="">)[</span><span><span leaf="">1</span></span><span leaf="">], </span><span leaf=""><br></span><span><span leaf="">"matrix/features/name"</span></span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">cells &lt;- readLines(</span><span><span leaf="">"cells.list"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">mat &lt;- list()</span><span leaf=""><br></span><span><span leaf="">for</span></span><span leaf=""> (i </span><span><span leaf="">in</span></span><span leaf=""> samples) {</span><span leaf=""><br></span><span leaf="">  tmp.file &lt;- paste0(i, </span><span><span leaf="">"_raw_feature_bc_matrix.h5"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">  tmp.mat &lt;- open_matrix_10x_hdf5(file.path(</span><span><span leaf="">"CordBlood"</span></span><span leaf="">, tmp.file))</span><span leaf=""><br></span><span leaf="">  rownames(tmp.mat) &lt;- make.unique(genes)</span><span leaf=""><br></span><span leaf="">  colnames(tmp.mat) &lt;- paste0(i, </span><span><span leaf="">"-"</span></span><span leaf="">, colnames(tmp.mat))</span><span leaf=""><br></span><span leaf="">  tmp.mat &lt;- tmp.mat[, colnames(tmp.mat) %</span><span><span leaf="">in</span></span><span leaf="">% cells]</span><span leaf=""><br></span><span leaf="">  colnames(tmp.mat) &lt;- gsub(</span><span><span leaf="">"_HiSeq_"</span></span><span leaf="">, </span><span><span leaf="">"HiSeq"</span></span><span leaf="">, colnames(tmp.mat))</span><span leaf=""><br></span><span leaf="">  colnames(tmp.mat) &lt;- sub(</span><span><span leaf="">"\\-"</span></span><span leaf="">, </span><span><span leaf="">"_"</span></span><span leaf="">, colnames(tmp.mat))</span><span leaf=""><br></span><span leaf="">  mat[[i]] &lt;- tmp.mat</span><span leaf=""><br></span><span leaf="">}</span><span leaf=""><br></span><span leaf="">mat &lt;- do.call(cbind, mat)</span><span leaf=""><br></span><span leaf="">mat &lt;- write_matrix_dir(mat, </span><span><span leaf="">"CordBlood_BPCells"</span></span><span leaf="">, overwrite = </span><span><span leaf="">TRUE</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">mat</span><span leaf=""><br></span></code></pre><pre data-tool="markdown.com.cn编辑器"><code><span leaf=""> 33538 x 239544 IterableMatrix object with class MatrixDir</span><span leaf=""><br></span><span leaf=""><br></span><span leaf=""> Row names: MIR1302-2HG, FAM138A ... FAM231C</span><span leaf=""><br></span><span leaf=""> Col names: MantonCB1HiSeq1_AAACCTGAGGAGTTGC-1, MantonCB1HiSeq1_AAACCTGCACAGACAG-1 ... MantonCB8HiSeq8_TTTGTCATCCAGATCA-1</span><span leaf=""><br></span><span leaf=""><br></span><span leaf=""> Data type: uint32_t</span><span leaf=""><br></span><span leaf=""> Storage order: column major</span><span leaf=""><br></span><span leaf=""><br></span><span leaf=""> Queued Operations:</span><span leaf=""><br></span><span leaf=""> 1. Load compressed matrix from directory E:\BPCells\CordBlood_BPCells</span><span leaf=""><br></span></code></pre><pre data-tool="markdown.com.cn编辑器"><code><span leaf="">Time: 406.32 seconds. Memory max: 1445.6 MB</span><span leaf=""><br></span></code></pre><p data-tool="markdown.com.cn编辑器"><span leaf="">将合并后的矩阵数据写入到硬盘中的某个目录下，以后</span><code><span leaf="">BPCells</span></code><span leaf="">读取时就可以跳过合并矩阵的迭代步骤，实现更高效的读写。接下来就是熟悉的</span><code><span leaf="">Seurat</span></code><span leaf="">标准流程了。</span></p><h2 data-tool="markdown.com.cn编辑器"><span></span><span><span leaf="">Seurat流程</span></span></h2><pre data-tool="markdown.com.cn编辑器"><code><span leaf="">obj &lt;- CreateSeuratObject(mat)</span><span leaf=""><br></span><span leaf="">obj &lt;- NormalizeData(obj, verbose = </span><span><span leaf="">FALSE</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">obj &lt;- FindVariableFeatures(obj, verbose = </span><span><span leaf="">FALSE</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">obj &lt;- ScaleData(obj, verbose = </span><span><span leaf="">FALSE</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">obj &lt;- RunPCA(obj, verbose = </span><span><span leaf="">FALSE</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">obj</span><span leaf=""><br></span></code></pre><pre data-tool="markdown.com.cn编辑器"><code><span leaf=""> An object of class Seurat </span><span leaf=""><br></span><span leaf=""> 33538 features across 239544 samples within 1 assay </span><span leaf=""><br></span><span leaf=""> Active assay: RNA (33538 features, 2000 variable features)</span><span leaf=""><br></span><span leaf="">  3 layers present: counts, data, scale.data</span><span leaf=""><br></span><span leaf="">  1 dimensional reduction calculated: pca</span><span leaf=""><br></span></code></pre><pre data-tool="markdown.com.cn编辑器"><code><span leaf="">Time: 298.75 seconds. Memory max: 1495.7 MB</span><span leaf=""><br></span></code></pre><h2 data-tool="markdown.com.cn编辑器"><span></span><span><span leaf="">BBKNN</span></span></h2><p data-tool="markdown.com.cn编辑器"><span leaf="">接下来就是用</span><code><span leaf="">RunBBKNN</span></code><span leaf="">矫正批次。默认使用的KNN搜索算法是</span><code><span leaf="">annoy</span></code><span leaf="">，这里为了和python的版本一致，我们选择</span><code><span leaf="">nndescent</span></code><span leaf="">。默认情况下，当作用于Seurat对象时，会在完成批次矫正后，</span><strong><span leaf="">自动计算tSNE和UMAP</span></strong><span leaf="">：</span></p><pre data-tool="markdown.com.cn编辑器"><code><span leaf="">obj &lt;- RunBBKNN(obj, </span><span><span leaf="">"orig.ident"</span></span><span leaf="">, method = </span><span><span leaf="">"nndescent"</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><pre data-tool="markdown.com.cn编辑器"><code><span leaf="">Time: 3145.14 seconds. Memory max: 7894.7 MB</span><span leaf=""><br></span></code></pre><p data-tool="markdown.com.cn编辑器"><span leaf="">可以来查看一下批次矫正的效果：</span></p><pre data-tool="markdown.com.cn编辑器"><code><span leaf="">DimPlot(obj, group.by = </span><span><span leaf="">"orig.ident"</span></span><span leaf="">) + NoLegend()</span><span leaf=""><br></span></code></pre><figure data-tool="markdown.com.cn编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/weQiaZ04F2Nszmz2AyARqiaYXtTMKclUjmAUUmpQLARIYDlIl5lGXQma8za1yHphYhqicg8HUTq2XKvuibzic4miaTFA/640?wx_fmt=png&amp;from=appmsg" alt="bbknn1" data-ratio="0.8888888888888888" data-type="png" data-w="432" data-imgfileid="100000339" src="https://mmbiz.qpic.cn/mmbiz_png/weQiaZ04F2Nszmz2AyARqiaYXtTMKclUjmAUUmpQLARIYDlIl5lGXQma8za1yHphYhqicg8HUTq2XKvuibzic4miaTFA/640?wx_fmt=png&amp;from=appmsg"></span><figcaption><span leaf="">bbknn1</span></figcaption></figure><p data-tool="markdown.com.cn编辑器"><span leaf="">借助</span><code><span leaf="">Rtsne</span></code><span leaf="">包的</span><code><span leaf="">Rtsne_neighbor()</span></code><span leaf="">，BBKNN的结果也可以用于计算tSNE降维，这在</span><strong><span leaf="">原python实现中是没有的</span></strong><span leaf="">！</span></p><pre data-tool="markdown.com.cn编辑器"><code><span leaf="">DimPlot(obj, reduction = </span><span><span leaf="">"tsne"</span></span><span leaf="">, group.by = </span><span><span leaf="">"orig.ident"</span></span><span leaf="">) + NoLegend()</span><span leaf=""><br></span></code></pre><figure data-tool="markdown.com.cn编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/weQiaZ04F2Nszmz2AyARqiaYXtTMKclUjmTXxOMoibFciaQCrJCkUmv2qpgiboib8ica1vTPhciaqU4CQJztbwFc7MVeug/640?wx_fmt=png&amp;from=appmsg" alt="bbknn2" data-ratio="0.8888888888888888" data-type="png" data-w="432" data-imgfileid="100000337" src="https://mmbiz.qpic.cn/mmbiz_png/weQiaZ04F2Nszmz2AyARqiaYXtTMKclUjmTXxOMoibFciaQCrJCkUmv2qpgiboib8ica1vTPhciaqU4CQJztbwFc7MVeug/640?wx_fmt=png&amp;from=appmsg"></span><figcaption><span leaf="">bbknn2</span></figcaption></figure><h2 data-tool="markdown.com.cn编辑器"><span></span><span><span leaf="">聚类分群</span></span></h2><p data-tool="markdown.com.cn编辑器"><span leaf="">要使用BBKNN的结果做分群，只需要指定</span><code><span leaf="">FindClusters(graph.name = "RNA_bbknn")</span></code><span leaf="">：</span></p><pre data-tool="markdown.com.cn编辑器"><code><span leaf="">obj &lt;- FindClusters(obj, graph.name = </span><span><span leaf="">"RNA_bbknn"</span></span><span leaf="">, verbose = </span><span><span leaf="">FALSE</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><pre data-tool="markdown.com.cn编辑器"><code><span leaf="">Time: 354.6 seconds. Memory max: 7894.7 MB</span><span leaf=""><br></span></code></pre><pre data-tool="markdown.com.cn编辑器"><code><span leaf="">DimPlot(obj) + NoLegend()</span><span leaf=""><br></span></code></pre><figure data-tool="markdown.com.cn编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/weQiaZ04F2Nszmz2AyARqiaYXtTMKclUjmHpAC4AcI5eqkRJOlQBibBZrGb1P0EEZphR29s14ty6ibUV6U53uGnJ3Q/640?wx_fmt=png&amp;from=appmsg" alt="bbknn3" data-ratio="0.8888888888888888" data-type="png" data-w="432" data-imgfileid="100000338" src="https://mmbiz.qpic.cn/mmbiz_png/weQiaZ04F2Nszmz2AyARqiaYXtTMKclUjmHpAC4AcI5eqkRJOlQBibBZrGb1P0EEZphR29s14ty6ibUV6U53uGnJ3Q/640?wx_fmt=png&amp;from=appmsg"></span><figcaption><span leaf="">bbknn3</span></figcaption></figure><p data-tool="markdown.com.cn编辑器"><span leaf="">查看marker基因分布：</span></p><pre data-tool="markdown.com.cn编辑器"><code><span leaf="">FeaturePlot(obj, features = c(</span><span><span leaf="">"CD3E"</span></span><span leaf="">, </span><span><span leaf="">"CD19"</span></span><span leaf="">, </span><span><span leaf="">"CD14"</span></span><span leaf="">, </span><span><span leaf="">"CD8A"</span></span><span leaf="">), ncol = </span><span><span leaf="">2</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><figure data-tool="markdown.com.cn编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/weQiaZ04F2Nszmz2AyARqiaYXtTMKclUjmicNSuEORH2kjd6cUiaLod4k0ITyd4BB0NJchqaHkQzMP4tuAgGkUx6oA/640?wx_fmt=png&amp;from=appmsg" alt="bbknn4" data-ratio="0.7916666666666666" data-type="png" data-w="576" data-imgfileid="100000336" src="https://mmbiz.qpic.cn/mmbiz_png/weQiaZ04F2Nszmz2AyARqiaYXtTMKclUjmicNSuEORH2kjd6cUiaLod4k0ITyd4BB0NJchqaHkQzMP4tuAgGkUx6oA/640?wx_fmt=png&amp;from=appmsg"></span><figcaption><span leaf="">bbknn4</span></figcaption></figure><h2 data-tool="markdown.com.cn编辑器"><span></span><span><span leaf="">小结</span></span></h2><p data-tool="markdown.com.cn编辑器"><code><span leaf="">bbknnR</span></code><span leaf="">是我入坑生信开发的第一个R包，时隔三年它能变成完全体，离不开整个开源社区的贡献。过去BBKNN算法几乎是Scanpy生态的专属，如今也是可以无缝衔接Seurat生态了。如果您正在寻找一种R环境可用的批次效应校正方法，</span><code><span leaf="">bbknnR</span></code><span leaf="">将是一个值得尝试的选择。欢迎大家在实际项目中应用，并通过 GitHub 提交反馈和建议：https://github.com/ycli1995/bbknnR</span></p><h2 data-tool="markdown.com.cn编辑器"><span></span><span><span leaf="">参考文献</span></span></h2><p data-tool="markdown.com.cn编辑器"><span leaf="">[1] Polański, Krzysztof et al. “BBKNN: fast batch alignment of single cell transcriptomes.” Bioinformatics (Oxford, England) vol. 36,3 (2020): 964-965.</span></p></section><section><span leaf=""><br></span></section><p><mp-style-type data-value="3"></mp-style-type></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/N_JQDseIfF5vYZrnMeTPcg",target="_blank" rel="noopener noreferrer">原文链接</a>
