---
title: "RNA 50.轻松搞定TCGA-LUAD基因筛选"
date: 2025-06-30T01:00:54Z
draft: ["false"]
tags: [
  "fetched",
  "桓峰基因"
]
categories: ["Acdemic"]
---
RNA 50.轻松搞定TCGA-LUAD基因筛选 by 桓峰基因
------
<div><section><span leaf=""><img alt="图片" data-ratio="0.4260230849947534" data-src="https://mmbiz.qpic.cn/mmbiz_png/5ia62S5YZaMwdlWerL4CKlDUMsIkLJRhhG5h5z4GIu47zBB6uDicHPerk6n5j84jZnbhFP8CVmgI5ibniaab750LiaA/640?wx_fmt=other&amp;wxfrom=5&amp;wx_lazy=1&amp;wx_co=1&amp;tp=webp" data-w="953" src="https://mmbiz.qpic.cn/mmbiz_png/5ia62S5YZaMwdlWerL4CKlDUMsIkLJRhhG5h5z4GIu47zBB6uDicHPerk6n5j84jZnbhFP8CVmgI5ibniaab750LiaA/640?wx_fmt=other&amp;wxfrom=5&amp;wx_lazy=1&amp;wx_co=1&amp;tp=webp"></span></section><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com" data-pm-slice="0 0 []"><hr><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">安装必要包</span></span><span></span></h2><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">if</span></span><span leaf=""> (!</span><span><span leaf="">require</span></span><span leaf="">(</span><span><span leaf="">"BiocManager"</span></span><span leaf="">)) install.packages(</span><span><span leaf="">"BiocManager"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">BiocManager::install(c(</span><span><span leaf="">"TCGAbiolinks"</span></span><span leaf="">, </span><span><span leaf="">"SummarizedExperiment"</span></span><span leaf="">, </span><span><span leaf="">"DESeq2"</span></span><span leaf="">, </span><span><span leaf="">"apeglm"</span></span><span leaf="">, </span><span leaf=""><br></span><span><span leaf="">"IHW"</span></span><span leaf="">, </span><span><span leaf="">"glmnet"</span></span><span leaf="">))</span><span leaf=""><br></span><span leaf="">install.packages(c(</span><span><span leaf="">"caret"</span></span><span leaf="">, </span><span><span leaf="">"dplyr"</span></span><span leaf="">, </span><span><span leaf="">"ggplot2"</span></span><span leaf="">, </span><span><span leaf="">"pheatmap"</span></span><span leaf="">))</span><span leaf=""><br></span></code></pre><h1 data-tool="mdnice编辑器"><span></span><span><span leaf="">加载包</span></span><span></span></h1><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">library</span></span><span leaf="">(TCGAbiolinks)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(SummarizedExperiment)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(DESeq2)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(survival)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(survminer)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(caret)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(dplyr)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(ggplot2)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(pheatmap)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(glmnet)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(timeROC)</span><span leaf=""><br></span></code></pre><h1 data-tool="mdnice编辑器"><span></span><span><span leaf="">数据下载与预处理</span></span><span></span></h1><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">1.下载TCGA-LUAD表达数据</span></span><span></span></h2><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">query &lt;- GDCquery(</span><span leaf=""><br></span><span leaf="">  project = </span><span><span leaf="">"TCGA-LUAD"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  data.category = </span><span><span leaf="">"Transcriptome Profiling"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  data.type = </span><span><span leaf="">"Gene Expression Quantification"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  workflow.type = </span><span><span leaf="">"STAR - Counts"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  experimental.strategy = </span><span><span leaf="">"RNA-Seq"</span></span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">GDCdownload(query, method = </span><span><span leaf="">"api"</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">2. 提取基因表达矩阵</span></span><span></span></h2><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">exp_data &lt;- GDCprepare(query)</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">count_matrix &lt;- assays(exp_data)$unstranded</span><span leaf=""><br></span><span leaf="">rownames(count_matrix) &lt;- rowData(exp_data)$gene_name</span><span leaf=""><br></span></code></pre><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">3. 获取样本分组信息</span></span><span></span></h2><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">sample_info &lt;- colData(exp_data) %&gt;%</span><span leaf=""><br></span><span leaf="">  as.data.frame() %&gt;%</span><span leaf=""><br></span><span leaf="">  select(barcode = barcode, patient = patient, sample_type = definition) %&gt;%</span><span leaf=""><br></span><span leaf="">  mutate(</span><span leaf=""><br></span><span leaf="">    group = ifelse(grepl(</span><span><span leaf="">"Normal"</span></span><span leaf="">, sample_type), </span><span><span leaf="">"Normal"</span></span><span leaf="">, </span><span><span leaf="">"Tumor"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">  )</span><span leaf=""><br></span></code></pre><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">4. 下载临床数据</span></span><span></span></h2><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">clinical_query &lt;- GDCquery(</span><span leaf=""><br></span><span leaf="">  project = </span><span><span leaf="">"TCGA-LUAD"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  data.category = </span><span><span leaf="">"Clinical"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  data.type = </span><span><span leaf="">"Clinical Supplement"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  data.format = </span><span><span leaf="">"BCR Biotab"</span></span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">GDCdownload(clinical_query)</span><span leaf=""><br></span></code></pre><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">5.提取生存数据</span></span><span></span></h2><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">clinical_data &lt;- GDCprepare(clinical_query)</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">colnames(clinical_data$clinical_patient_luad)</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">##   [1] "bcr_patient_uuid"                           </span><span leaf=""><br></span><span leaf="">##   [2] "bcr_patient_barcode"                        </span><span leaf=""><br></span><span leaf="">##   [3] "form_completion_date"                       </span><span leaf=""><br></span><span leaf="">##   [4] "histologic_diagnosis...4"                   </span><span leaf=""><br></span><span leaf="">##   [5] "race"                                       </span><span leaf=""><br></span><span leaf="">##   [6] "ethnicity"                                  </span><span leaf=""><br></span><span leaf="">##   [7] "location_lung_parenchyma"                   </span><span leaf=""><br></span><span leaf="">##   [8] "height_cm_at_diagnosis"                     </span><span leaf=""><br></span><span leaf="">##   [9] "prospective_collection"                     </span><span leaf=""><br></span><span leaf="">##  [10] "weight_kg_at_diagnosis"                     </span><span leaf=""><br></span><span leaf="">##  [11] "retrospective_collection"                   </span><span leaf=""><br></span><span leaf="">##  [12] "gender"                                     </span><span leaf=""><br></span><span leaf="">##  [13] "birth_days_to"                              </span><span leaf=""><br></span><span leaf="">##  [14] "history_other_malignancy"                   </span><span leaf=""><br></span><span leaf="">##  [15] "history_neoadjuvant_treatment"              </span><span leaf=""><br></span><span leaf="">##  [16] "cause_of_death"                             </span><span leaf=""><br></span><span leaf="">##  [17] "cause_of_death_other"                       </span><span leaf=""><br></span><span leaf="">##  [18] "residual_tumor"                             </span><span leaf=""><br></span><span leaf="">##  [19] "tumor_status"                               </span><span leaf=""><br></span><span leaf="">##  [20] "ajcc_staging_edition"                       </span><span leaf=""><br></span><span leaf="">##  [21] "ecog_score"                                 </span><span leaf=""><br></span><span leaf="">##  [22] "ajcc_nodes_pathologic_pn"                   </span><span leaf=""><br></span><span leaf="">##  [23] "karnofsky_score"                            </span><span leaf=""><br></span><span leaf="">##  [24] "ajcc_metastasis_pathologic_pm"              </span><span leaf=""><br></span><span leaf="">##  [25] "radiation_treatment_adjuvant"               </span><span leaf=""><br></span><span leaf="">##  [26] "vital_status"                               </span><span leaf=""><br></span><span leaf="">##  [27] "pharmaceutical_tx_adjuvant"                 </span><span leaf=""><br></span><span leaf="">##  [28] "tobacco_smoking_history_indicator"          </span><span leaf=""><br></span><span leaf="">##  [29] "tobacco_smoking_age_started"                </span><span leaf=""><br></span><span leaf="">##  [30] "last_contact_days_to"                       </span><span leaf=""><br></span><span leaf="">##  [31] "tobacco_smoking_pack_years_smoked"          </span><span leaf=""><br></span><span leaf="">##  [32] "hiv_status"                                 </span><span leaf=""><br></span><span leaf="">##  [33] "death_days_to"                              </span><span leaf=""><br></span><span leaf="">##  [34] "nadir_cd4_counts"                           </span><span leaf=""><br></span><span leaf="">##  [35] "cd4_counts_at_diagnosis"                    </span><span leaf=""><br></span><span leaf="">##  [36] "pulmonary_function_test_indicator"          </span><span leaf=""><br></span><span leaf="">##  [37] "hiv_rna_load_at_diagnosis"                  </span><span leaf=""><br></span><span leaf="">##  [38] "fev1_percent_ref_prebroncholiator"          </span><span leaf=""><br></span><span leaf="">##  [39] "prior_aids_conditions"                      </span><span leaf=""><br></span><span leaf="">##  [40] "fev1_percent_ref_postbroncholiator"         </span><span leaf=""><br></span><span leaf="">##  [41] "hbv_test_results"                           </span><span leaf=""><br></span><span leaf="">##  [42] "fev1_fvc_ratio_prebroncholiator"            </span><span leaf=""><br></span><span leaf="">##  [43] "hcv_test_results"                           </span><span leaf=""><br></span><span leaf="">##  [44] "fev1_fvc_ratio_postbroncholiator"           </span><span leaf=""><br></span><span leaf="">##  [45] "hpv_test_results"                           </span><span leaf=""><br></span><span leaf="">##  [46] "carbon_monoxide_diffusion_dlco"             </span><span leaf=""><br></span><span leaf="">##  [47] "kshv_hhv8_test_results"                     </span><span leaf=""><br></span><span leaf="">##  [48] "kras_gene_analysis_indicator"               </span><span leaf=""><br></span><span leaf="">##  [49] "haart_therapy_prior_to_dx"                  </span><span leaf=""><br></span><span leaf="">##  [50] "kras_mutation_found"                        </span><span leaf=""><br></span><span leaf="">##  [51] "haart_therapy_at_dx"                        </span><span leaf=""><br></span><span leaf="">##  [52] "kras_mutation_identified_type"              </span><span leaf=""><br></span><span leaf="">##  [53] "cdc_hiv_risk_group"                         </span><span leaf=""><br></span><span leaf="">##  [54] "egfr_mutation_status"                       </span><span leaf=""><br></span><span leaf="">##  [55] "egfr_mutation_identified_type"              </span><span leaf=""><br></span><span leaf="">##  [56] "prior_mailgnancy_type"                      </span><span leaf=""><br></span><span leaf="">##  [57] "eml4_alk_translocation_status"              </span><span leaf=""><br></span><span leaf="">##  [58] "history_immunological_disease"              </span><span leaf=""><br></span><span leaf="">##  [59] "eml4_alk_translocation_variant"             </span><span leaf=""><br></span><span leaf="">##  [60] "history_immunological_disease_other"        </span><span leaf=""><br></span><span leaf="">##  [61] "eml4_alk_analysis_type"                     </span><span leaf=""><br></span><span leaf="">##  [62] "history_immunosuppresive_rx"                </span><span leaf=""><br></span><span leaf="">##  [63] "history_immunosuppressive_rx_other"         </span><span leaf=""><br></span><span leaf="">##  [64] "tobacco_smoking_year_started"               </span><span leaf=""><br></span><span leaf="">##  [65] "history_relevant_infectious_dx"             </span><span leaf=""><br></span><span leaf="">##  [66] "tobacco_smoking_year_stopped"               </span><span leaf=""><br></span><span leaf="">##  [67] "history_immunosuppresive_dx_other"          </span><span leaf=""><br></span><span leaf="">##  [68] "histologic_diagnosis...68"                  </span><span leaf=""><br></span><span leaf="">##  [69] "submitted_tumor_site"                       </span><span leaf=""><br></span><span leaf="">##  [70] "performance_status_timing"                  </span><span leaf=""><br></span><span leaf="">##  [71] "laterality"                                 </span><span leaf=""><br></span><span leaf="">##  [72] "anatomic_organ_subdivision"                 </span><span leaf=""><br></span><span leaf="">##  [73] "site_of_primary_tumor_other"                </span><span leaf=""><br></span><span leaf="">##  [74] "treatment_outcome_first_course"             </span><span leaf=""><br></span><span leaf="">##  [75] "new_tumor_event_dx_indicator"               </span><span leaf=""><br></span><span leaf="">##  [76] "initial_pathologic_dx_year"                 </span><span leaf=""><br></span><span leaf="">##  [77] "method_initial_path_dx"                     </span><span leaf=""><br></span><span leaf="">##  [78] "method_initial_path_dx_other"               </span><span leaf=""><br></span><span leaf="">##  [79] "lymph_nodes_examined"                       </span><span leaf=""><br></span><span leaf="">##  [80] "lymph_nodes_examined_count"                 </span><span leaf=""><br></span><span leaf="">##  [81] "lymph_nodes_examined_he_count"              </span><span leaf=""><br></span><span leaf="">##  [82] "lymph_nodes_examined_ihc_count"             </span><span leaf=""><br></span><span leaf="">##  [83] "pos_lymph_node_location"                    </span><span leaf=""><br></span><span leaf="">##  [84] "other_pos_node_location"                    </span><span leaf=""><br></span><span leaf="">##  [85] "ajcc_tumor_clinical_ct"                     </span><span leaf=""><br></span><span leaf="">##  [86] "ajcc_tumor_pathologic_pt"                   </span><span leaf=""><br></span><span leaf="">##  [87] "ajcc_nodes_clinical_cn"                     </span><span leaf=""><br></span><span leaf="">##  [88] "ajcc_metastasis_clinical_cm"                </span><span leaf=""><br></span><span leaf="">##  [89] "ajcc_clinical_tumor_stage"                  </span><span leaf=""><br></span><span leaf="">##  [90] "ajcc_pathologic_tumor_stage"                </span><span leaf=""><br></span><span leaf="">##  [91] "followup_lost_to"                           </span><span leaf=""><br></span><span leaf="">##  [92] "age_at_initial_pathologic_diagnosis"        </span><span leaf=""><br></span><span leaf="">##  [93] "cancer_diagnosis_cancer_type_icd9_text_name"</span><span leaf=""><br></span><span leaf="">##  [94] "days_to_form_completion"                    </span><span leaf=""><br></span><span leaf="">##  [95] "days_to_hiv_diagnosis"                      </span><span leaf=""><br></span><span leaf="">##  [96] "days_to_initial_pathologic_diagnosis"       </span><span leaf=""><br></span><span leaf="">##  [97] "days_to_patient_progression_free"           </span><span leaf=""><br></span><span leaf="">##  [98] "days_to_sample_procurement"                 </span><span leaf=""><br></span><span leaf="">##  [99] "days_to_tumor_progression"                  </span><span leaf=""><br></span><span leaf="">## [100] "disease_code"                               </span><span leaf=""><br></span><span leaf="">## [101] "egfr_mutation_identified"                   </span><span leaf=""><br></span><span leaf="">## [102] "eml4_alk_translocation_identified"          </span><span leaf=""><br></span><span leaf="">## [103] "extranodal_involvement"                     </span><span leaf=""><br></span><span leaf="">## [104] "family_member_relationship_type"            </span><span leaf=""><br></span><span leaf="">## [105] "icd_10"                                     </span><span leaf=""><br></span><span leaf="">## [106] "icd_o_3_histology"                          </span><span leaf=""><br></span><span leaf="">## [107] "icd_o_3_site"                               </span><span leaf=""><br></span><span leaf="">## [108] "informed_consent_verified"                  </span><span leaf=""><br></span><span leaf="">## [109] "margin_status"                              </span><span leaf=""><br></span><span leaf="">## [110] "measure_of_response"                        </span><span leaf=""><br></span><span leaf="">## [111] "metastatic_site_at_diagnosis"               </span><span leaf=""><br></span><span leaf="">## [112] "metastatic_site_at_diagnosis_other"         </span><span leaf=""><br></span><span leaf="">## [113] "number_cycles"                              </span><span leaf=""><br></span><span leaf="">## [114] "patient_id"                                 </span><span leaf=""><br></span><span leaf="">## [115] "pharm_regimen"                              </span><span leaf=""><br></span><span leaf="">## [116] "pharm_regimen_other"                        </span><span leaf=""><br></span><span leaf="">## [117] "prior_systemic_therapy_type"                </span><span leaf=""><br></span><span leaf="">## [118] "project_code"                               </span><span leaf=""><br></span><span leaf="">## [119] "regimen_indication"                         </span><span leaf=""><br></span><span leaf="">## [120] "relative_family_cancer_history"             </span><span leaf=""><br></span><span leaf="">## [121] "stage_other"                                </span><span leaf=""><br></span><span leaf="">## [122] "stem_cell_transplantation"                  </span><span leaf=""><br></span><span leaf="">## [123] "tissue_source_site"</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">clinical_df &lt;- clinical_data$clinical_patient_luad %&gt;% </span><span leaf=""><br></span><span leaf="">  select(</span><span leaf=""><br></span><span leaf="">    bcr_patient_barcode, </span><span leaf=""><br></span><span leaf="">    OS_time = death_days_to,         </span><span><span leaf=""># 使用死亡时间作为主要生存时间</span></span><span leaf=""><br></span><span leaf="">    last_followup = last_contact_days_to,  </span><span><span leaf=""># 最后随访时间</span></span><span leaf=""><br></span><span leaf="">    vital_status</span><span leaf=""><br></span><span leaf="">  ) %&gt;% </span><span leaf=""><br></span><span leaf="">  mutate(</span><span leaf=""><br></span><span><span leaf=""># 转换时间为数值型</span></span><span leaf=""><br></span><span leaf="">    OS_time = as.numeric(as.character(OS_time)),</span><span leaf=""><br></span><span leaf="">    last_followup = as.numeric(as.character(last_followup)),</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 处理生存时间：如果死亡时间为空，则使用最后随访时间</span></span><span leaf=""><br></span><span leaf="">    OS_time = ifelse(is.na(OS_time), last_followup, OS_time),</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 处理生存状态</span></span><span leaf=""><br></span><span leaf="">    OS_status = case_when(</span><span leaf=""><br></span><span leaf="">      vital_status == </span><span><span leaf="">"Dead"</span></span><span leaf=""> ~ </span><span><span leaf="">1</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">      vital_status == </span><span><span leaf="">"Alive"</span></span><span leaf=""> ~ </span><span><span leaf="">0</span></span><span leaf="">,</span><span leaf=""><br></span><span><span leaf="">TRUE</span></span><span leaf=""> ~ </span><span><span leaf="">NA_real_</span></span><span leaf=""><br></span><span leaf="">    )</span><span leaf=""><br></span><span leaf="">  ) %&gt;% </span><span leaf=""><br></span><span leaf="">  filter(!is.na(OS_time) &amp; OS_time &gt; </span><span><span leaf="">0</span></span><span leaf=""> &amp; !is.na(OS_status)) %&gt;%</span><span leaf=""><br></span><span leaf="">  select(bcr_patient_barcode, OS_time, OS_status)</span><span leaf=""><br></span></code></pre><h1 data-tool="mdnice编辑器"><span></span><span><span leaf="">差异表达分析</span></span><span></span></h1><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">1.创建DESeq2对象</span></span><span></span></h2><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">dds &lt;- DESeqDataSetFromMatrix(</span><span leaf=""><br></span><span leaf="">  countData = count_matrix,</span><span leaf=""><br></span><span leaf="">  colData = sample_info,</span><span leaf=""><br></span><span leaf="">  design = ~ group</span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span></code></pre><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">2.预处理：过滤低表达基因</span></span><span></span></h2><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">keep &lt;- rowSums(counts(dds) &gt;= </span><span><span leaf="">10</span></span><span leaf="">) &gt;= </span><span><span leaf="">5</span></span><span leaf=""><br></span><span leaf="">dds &lt;- dds[keep,]</span><span leaf=""><br></span></code></pre><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">3.差异表达分析</span></span><span></span></h2><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">dds &lt;- DESeq(dds, test = </span><span><span leaf="">"Wald"</span></span><span leaf="">, sfType = </span><span><span leaf="">"poscounts"</span></span><span leaf="">, useT = </span><span><span leaf="">TRUE</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">4.获取结果</span></span><span></span></h2><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">res &lt;- results(dds, </span><span leaf=""><br></span><span leaf="">               contrast = c(</span><span><span leaf="">"group"</span></span><span leaf="">, </span><span><span leaf="">"Tumor"</span></span><span leaf="">, </span><span><span leaf="">"Normal"</span></span><span leaf="">),</span><span leaf=""><br></span><span leaf="">               alpha = </span><span><span leaf="">0.05</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">5.结果收缩</span></span><span></span></h2><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">resLFC &lt;- lfcShrink(dds, </span><span leaf=""><br></span><span leaf="">                    coef = </span><span><span leaf="">"group_Tumor_vs_Normal"</span></span><span leaf="">, </span><span leaf=""><br></span><span leaf="">                    type = </span><span><span leaf="">"apeglm"</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">6.筛选显著差异基因</span></span><span></span></h2><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">deg_threshold &lt;- </span><span><span leaf="">0.01</span></span><span><span leaf=""># FDR阈值</span></span><span leaf=""><br></span><span leaf="">lfc_threshold &lt;- </span><span><span leaf="">2</span></span><span><span leaf=""># log2FC阈值</span></span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">sig_genes &lt;- resLFC %&gt;%</span><span leaf=""><br></span><span leaf="">  as.data.frame() %&gt;%</span><span leaf=""><br></span><span leaf="">  tibble::rownames_to_column(</span><span><span leaf="">"gene_name"</span></span><span leaf="">) %&gt;%</span><span leaf=""><br></span><span leaf="">  filter(padj &lt; deg_threshold &amp; abs(log2FoldChange) &gt; lfc_threshold) %&gt;%</span><span leaf=""><br></span><span leaf="">  arrange(padj, desc(abs(log2FoldChange)))</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">head(sig_genes)</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">##   gene_name baseMean log2FoldChange      lfcSE        pvalue          padj</span><span leaf=""><br></span><span leaf="">## 1    FAM83A 6577.665       6.857424 0.24218801 1.294714e-113 5.051716e-109</span><span leaf=""><br></span><span leaf="">## 2     MMP11 6458.548       5.244130 0.25964061 5.251530e-103  1.024521e-98</span><span leaf=""><br></span><span leaf="">## 3     PYCR1 3723.099       3.706301 0.14094837 1.116987e-102  1.452754e-98</span><span leaf=""><br></span><span leaf="">## 4     OTUD1 1101.777      -2.095951 0.08233153  1.304822e-97  1.272789e-93</span><span leaf=""><br></span><span leaf="">## 5     SYT12 1360.843       5.470375 0.28116973  1.170384e-93  9.133207e-90</span><span leaf=""><br></span><span leaf="">## 6 AFAP1-AS1 2831.703       6.306635 0.25975021  1.963892e-92  1.277119e-88</span><span leaf=""><br></span></code></pre><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">7.保存差异基因结果</span></span><span></span></h2><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">write.csv(sig_genes, </span><span><span leaf="">"TCGA_LUAD_DEGs.csv"</span></span><span leaf="">, row.names = </span><span><span leaf="">FALSE</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">8.差异表达可视化</span></span><span></span></h2><h3 data-tool="mdnice编辑器"><span></span><span><span leaf="">火山图</span></span><span></span></h3><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">EnhancedVolcano::EnhancedVolcano(</span><span leaf=""><br></span><span leaf="">  resLFC,</span><span leaf=""><br></span><span leaf="">  lab = rownames(resLFC),</span><span leaf=""><br></span><span leaf="">  x = </span><span><span leaf="">'log2FoldChange'</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  y = </span><span><span leaf="">'padj'</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  pCutoff = deg_threshold,</span><span leaf=""><br></span><span leaf="">  FCcutoff = lfc_threshold,</span><span leaf=""><br></span><span leaf="">  pointSize = </span><span><span leaf="">2.0</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  labSize = </span><span><span leaf="">4.0</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  title = </span><span><span leaf="">"LUAD差异表达基因 (Tumor vs Normal)"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  subtitle = </span><span><span leaf="">"TCGA-LUAD"</span></span><span leaf=""><br></span><span leaf="">) </span><span leaf=""><br></span></code></pre><h3 data-tool="mdnice编辑器"><span></span><span><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/5ia62S5YZaMwKQXmiay65Px9riaMFm3iccCqHg7h3YXg12bm12uyCyswk3g4um53DBnwyWF5BaRGwIqh7RB7hwK4VQ/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.5625841184387618" data-s="300,640" data-type="png" data-w="743" type="block" data-imgfileid="100012213" src="https://mmbiz.qpic.cn/mmbiz_png/5ia62S5YZaMwKQXmiay65Px9riaMFm3iccCqHg7h3YXg12bm12uyCyswk3g4um53DBnwyWF5BaRGwIqh7RB7hwK4VQ/640?wx_fmt=png&amp;from=appmsg"></span></span></h3><h3 data-tool="mdnice编辑器"><span><span leaf="">热图</span></span><span></span></h3><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">top_genes &lt;- head(sig_genes$gene_name, </span><span><span leaf="">50</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">norm_counts &lt;- counts(dds, normalized = </span><span><span leaf="">TRUE</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">heatmap_data &lt;- norm_counts[top_genes, ] </span><span leaf=""><br></span><span leaf="">colnames(heatmap_data) &lt;- substr(colnames(heatmap_data), </span><span><span leaf="">1</span></span><span leaf="">, </span><span><span leaf="">12</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">heatmap_data = heatmap_data[!duplicated(heatmap_data),]</span><span leaf=""><br></span><span leaf="">ann=data.frame(sample_info[,c(</span><span><span leaf="">"group"</span></span><span leaf="">,</span><span><span leaf="">"patient"</span></span><span leaf="">)])</span><span leaf=""><br></span><span leaf="">ann=ann[!duplicated(ann$patient),]</span><span leaf=""><br></span><span leaf="">pheatmap(</span><span leaf=""><br></span><span leaf="">  log2(heatmap_data + </span><span><span leaf="">1</span></span><span leaf="">),</span><span leaf=""><br></span><span leaf="">  scale = </span><span><span leaf="">"row"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  clustering_distance_rows = </span><span><span leaf="">"euclidean"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  clustering_distance_cols = </span><span><span leaf="">"euclidean"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  clustering_method = </span><span><span leaf="">"ward.D2"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  show_colnames = </span><span><span leaf="">FALSE</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  annotation_col = ann,</span><span leaf=""><br></span><span leaf="">  main = </span><span><span leaf="">"Top 50 DEGs in LUAD"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  filename = </span><span><span leaf="">"DEG_heatmap_LUAD.png"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  width = </span><span><span leaf="">10</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  height = </span><span><span leaf="">8</span></span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span></code></pre><h1 data-tool="mdnice编辑器"><span></span><span><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_jpg/5ia62S5YZaMwKQXmiay65Px9riaMFm3iccCqKyib2bAZtx07V2INFO87G48OwaJbTnAyicwn5rmJ4hLOBUzwyD1MMTpg/640?wx_fmt=jpeg" data-ratio="0.9203125" data-s="300,640" data-type="png" data-w="1280" type="block" data-croporisrc="https://mmbiz.qpic.cn/mmbiz_png/5ia62S5YZaMwKQXmiay65Px9riaMFm3iccCq4wqib42GMmg1VrH5pxJML40sPaSN4WZybGFIMcDTviccKicUk3VustGFg/0?wx_fmt=png&amp;from=appmsg" data-cropx2="2607.52688172043" data-cropy2="2397.8494623655915" data-imgfileid="100012214" src="https://mmbiz.qpic.cn/mmbiz_jpg/5ia62S5YZaMwKQXmiay65Px9riaMFm3iccCqKyib2bAZtx07V2INFO87G48OwaJbTnAyicwn5rmJ4hLOBUzwyD1MMTpg/640?wx_fmt=jpeg"></span></span></h1><h1 data-tool="mdnice编辑器"><span><span leaf="">Lasso回归筛选关键基因</span></span><span></span></h1><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">1. 准备Lasso回归数据</span></span><span></span></h2><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">top_genes_matrix = count_matrix[top_genes, ]</span><span leaf=""><br></span><span leaf="">colnames(top_genes_matrix) = substr(colnames(top_genes_matrix), </span><span><span leaf="">1</span></span><span leaf="">, </span><span><span leaf="">12</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 关键修复：确保只保留有交集的样本</span></span><span leaf=""><br></span><span leaf="">common_samples &lt;- intersect(colnames(top_genes_matrix), </span><span leaf=""><br></span><span leaf="">                            substr(clinical_df$bcr_patient_barcode, </span><span><span leaf="">1</span></span><span leaf="">, </span><span><span leaf="">12</span></span><span leaf="">))</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 检查是否有共同样本</span></span><span leaf=""><br></span><span><span leaf="">if</span></span><span leaf=""> (length(common_samples) == </span><span><span leaf="">0</span></span><span leaf="">) {</span><span leaf=""><br></span><span><span leaf="">stop</span></span><span leaf="">(</span><span><span leaf="">"错误：矩阵和临床数据无共同样本"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">}</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 提取共同样本的子集</span></span><span leaf=""><br></span><span leaf="">top_genes_matrix = top_genes_matrix[, common_samples, drop = </span><span><span leaf="">FALSE</span></span><span leaf="">]</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 匹配临床数据顺序</span></span><span leaf=""><br></span><span leaf="">matched_clinical &lt;- clinical_df[match(common_samples, </span><span leaf=""><br></span><span leaf="">                                      substr(clinical_df$bcr_patient_barcode, </span><span><span leaf="">1</span></span><span leaf="">, </span><span><span leaf="">12</span></span><span leaf="">)), ]</span><span leaf=""><br></span></code></pre><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">2.构建训练数据</span></span><span></span></h2><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">x_train &lt;- t(top_genes_matrix)</span><span leaf=""><br></span><span leaf="">y_train &lt;- cbind(</span><span leaf=""><br></span><span leaf="">  time = matched_clinical$OS_time,</span><span leaf=""><br></span><span leaf="">  status = matched_clinical$OS_status</span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span></code></pre><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">3.创建交叉验证的生存对象</span></span><span></span></h2><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">cvfit &lt;- cv.glmnet(</span><span leaf=""><br></span><span leaf="">  x = x_train,</span><span leaf=""><br></span><span leaf="">  y = y_train,</span><span leaf=""><br></span><span leaf="">  family = </span><span><span leaf="">"cox"</span></span><span leaf="">,  </span><span><span leaf=""># Cox比例风险模型</span></span><span leaf=""><br></span><span leaf="">  type.measure = </span><span><span leaf="">"C"</span></span><span leaf="">,  </span><span><span leaf=""># 使用C-index作为评估指标</span></span><span leaf=""><br></span><span leaf="">  nfolds = </span><span><span leaf="">10</span></span><span leaf="">,     </span><span><span leaf=""># 10折交叉验证</span></span><span leaf=""><br></span><span leaf="">  alpha = </span><span><span leaf="">1</span></span><span leaf="">,       </span><span><span leaf=""># Lasso回归 (alpha=1)</span></span><span leaf=""><br></span><span leaf="">  standardize = </span><span><span leaf="">FALSE</span></span><span leaf="">,  </span><span><span leaf=""># 数据已标准化</span></span><span leaf=""><br></span><span leaf="">  parallel = </span><span><span leaf="">TRUE</span></span><span><span leaf=""># 并行计算加速</span></span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span></code></pre><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">4.可视化交叉验证结果</span></span><span></span></h2><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">plot(cvfit)</span><span leaf=""><br></span></code></pre><h2 data-tool="mdnice编辑器"><span></span><span><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/5ia62S5YZaMwKQXmiay65Px9riaMFm3iccCqVMA3ybkP1xwD07mB7Fecdq74eDicHOGfsCGOwWAyiaveDuLQpOj6M7fw/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.654357459379616" data-s="300,640" data-type="png" data-w="677" type="block" data-imgfileid="100012216" src="https://mmbiz.qpic.cn/mmbiz_png/5ia62S5YZaMwKQXmiay65Px9riaMFm3iccCqVMA3ybkP1xwD07mB7Fecdq74eDicHOGfsCGOwWAyiaveDuLQpOj6M7fw/640?wx_fmt=png&amp;from=appmsg"></span></span></h2><h2 data-tool="mdnice编辑器"><span><span leaf="">5.选择最优lambda值</span></span><span></span></h2><p data-tool="mdnice编辑器"><span leaf="">lambda.1se 更严格</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">lambda_opt &lt;- cvfit$lambda.1se</span><span leaf=""><br></span></code></pre><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">6.可视化Lasso路径</span></span><span></span></h2><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">plot(cvfit$glmnet.fit, xvar = </span><span><span leaf="">"lambda"</span></span><span leaf="">, label = </span><span><span leaf="">TRUE</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">abline(v = log(lambda_opt), lty = </span><span><span leaf="">2</span></span><span leaf="">, col = </span><span><span leaf="">"red"</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><h2 data-tool="mdnice编辑器"><span></span><span><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/5ia62S5YZaMwKQXmiay65Px9riaMFm3iccCqkg93dL6LSS9s5jACJzgUgpZEuRyGm7PGj05D0tObJJ9icmaZDiak8o9Q/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.6460176991150443" data-s="300,640" data-type="png" data-w="678" type="block" data-imgfileid="100012217" src="https://mmbiz.qpic.cn/mmbiz_png/5ia62S5YZaMwKQXmiay65Px9riaMFm3iccCqkg93dL6LSS9s5jACJzgUgpZEuRyGm7PGj05D0tObJJ9icmaZDiak8o9Q/640?wx_fmt=png&amp;from=appmsg"></span></span></h2><h2 data-tool="mdnice编辑器"><span><span leaf="">7.提取非零系数基因</span></span><span></span></h2><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">lasso_coef &lt;- coef(cvfit, s = lambda_opt)</span><span leaf=""><br></span><span leaf="">selected_genes &lt;- rownames(lasso_coef)[which(lasso_coef != </span><span><span leaf="">0</span></span><span leaf="">)]</span><span leaf=""><br></span></code></pre><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">8.保存结果</span></span><span></span></h2><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">lasso_results &lt;- data.frame(</span><span leaf=""><br></span><span leaf="">  Gene = selected_genes,</span><span leaf=""><br></span><span leaf="">  Coefficient = as.vector(lasso_coef[lasso_coef != </span><span><span leaf="">0</span></span><span leaf="">])</span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">lasso_results </span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">##     Gene  Coefficient</span><span leaf=""><br></span><span leaf="">## 1 FAM83A 5.194384e-06</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">write.csv(lasso_results, </span><span><span leaf="">"lasso_selected_genes.csv"</span></span><span leaf="">, row.names = </span><span><span leaf="">FALSE</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><section nodeleaf=""><mp-common-videosnap data-pluginname="mpvideosnap" data-url="https://findermp.video.qq.com/251/20304/stodownload?encfilekey=rjD5jyTuFrIpZ2ibE8T7Ym3K77SEULgkiarlyZ9iaM6c7THHN1pKqEN1FDjo4781QOzTPXaOtCB2XicP9EuFmmTy1Rtv8n52ofdUnnVjL7E19rJUqov3VDYzyQ&amp;token=2lt8WBSnjTnCKu2HRiaMAaoAia1TetQquTzxFGX76WIbCryvweG7hWEpQzzxzgd2r01HTIpERjcp5BpxiaEuroQCJ0V6ic6AgfGgAKjFObpKw70oOtOWlniaIvncBAlXp1PtEDvlhSzNqSp6BxphhmzLIqI4l8rylGVFve9iaUorzZMvQ&amp;idx=1&amp;hy=SZ&amp;m=&amp;scene=2&amp;uzid=1" data-headimgurl="http://wx.qlogo.cn/finderhead/Q3auHgzwzM6NFe6qfZ98pccnStLOBzJIic5564gE1oW2m34RtSicQHIQ/0" data-username="v2_060000231003b20faec8c6eb891cc2ddca0ce43cb07767727967ddf958ff98912bf15d898b0d@finder" data-nickname="桓峰基因" data-desc="RNA 50.轻松搞定TCGA-LUAD数据下载到基因筛选#RNA#转录组#机器学习" data-nonceid="12924456554895691218" data-width="1920" data-height="1080" data-type="video" data-id="export/UzFfAgtgekIEAQAAAAAAvjgZenw2GAAAAAstQy6ubaLX4KHWvLEZgBPEjqIIa0J5YoyIzNPgMIueDUmomekbBgTWS5KeIqoF"></mp-common-videosnap></section><pre data-tool="mdnice编辑器"><code><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/5ia62S5YZaMzx8oc6vWnArVHF528DC3lZfpu2QacVGXIOnwibajbkXsabNk2Zia22icpPf4iaiby472Rk5Au5CxwZ6Zg/640?wx_fmt=png&amp;from=appmsg" data-ratio="1.7777777777777777" data-s="300,640" data-type="png" data-w="1080" type="block" data-imgfileid="100011907" src="https://mmbiz.qpic.cn/mmbiz_png/5ia62S5YZaMzx8oc6vWnArVHF528DC3lZfpu2QacVGXIOnwibajbkXsabNk2Zia22icpPf4iaiby472Rk5Au5CxwZ6Zg/640?wx_fmt=png&amp;from=appmsg"></span></code></pre></section><p><mp-style-type data-value="3"></mp-style-type></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/ZoamxYZZ0TSDu9Gyfv9owQ",target="_blank" rel="noopener noreferrer">原文链接</a>
