---
title: "使用R语言进行CRISPRscreen的定量"
date: 2025-06-17T12:56:34Z
draft: ["false"]
tags: [
  "fetched",
  "东林的扯淡小屋"
]
categories: ["Acdemic"]
---
使用R语言进行CRISPRscreen的定量 by 东林的扯淡小屋
------
<div><section><ul><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li></ul><pre data-lang="bash"><code><span leaf="">library(ShortRead)</span></code><code><span leaf="">library(parallel)</span></code><code><span leaf="">library(stringr)</span></code><code><span leaf="">library(dplyr)</span></code><code><span leaf=""><br></span></code><code><span leaf=""><span># 记录开始时间</span></span></code><code><span leaf="">start_time &lt;- Sys.time()</span></code><code><span leaf=""><br></span></code><code><span leaf=""><span># 读取gRNA库</span></span></code><code><span leaf="">gRNA_library &lt;- read.csv(<span>"HIV_library_withoutG_YT.csv"</span>, stringsAsFactors = FALSE)</span></code><code><span leaf="">target_sequences &lt;- unique(gRNA_library<span>$gRNA</span>.sequence)</span></code><code><span leaf=""><br></span></code><code><span leaf=""><span># 获取所有 fastq 文件名</span></span></code><code><span leaf="">fastq_files &lt;- list.files(pattern = <span>"_R1\\.fq\\.gz$"</span>)</span></code><code><span leaf=""><br></span></code><code><span leaf=""><br></span></code><code><span leaf=""><span># 并行核心数</span></span></code><code><span leaf="">num_cores &lt;- 50  </span></code><code><span leaf=""><br></span></code><code><span leaf=""><span># 初始化最终结果表，先放 id, gRNA.sequence, Gene</span></span></code><code><span leaf="">final_result &lt;- gRNA_library %&gt;%</span></code><code><span leaf="">  <span>select</span>(<span>id</span>, gRNA.sequence, Gene)</span></code><code><span leaf=""><br></span></code><code><span leaf=""><span># 循环处理每个 fastq 文件</span></span></code><code><span leaf=""><span>for</span> (fastq_file <span>in</span> fastq_files) {</span></code><code><span leaf="">  message(<span>"开始处理: "</span>, fastq_file)</span></code><code><span leaf=""><br></span></code><code><span leaf="">  <span># 读取 fastq</span></span></code><code><span leaf="">  fq &lt;- readFastq(fastq_file)</span></code><code><span leaf="">  reads &lt;- as.character(sread(fq))</span></code><code><span leaf=""><br></span></code><code><span leaf="">  <span># 建立并行集群</span></span></code><code><span leaf="">  cl &lt;- makeCluster(num_cores, <span>type</span> = <span>"PSOCK"</span>)</span></code><code><span leaf=""><br></span></code><code><span leaf="">  <span># 导出变量和函数到各线程</span></span></code><code><span leaf="">  clusterExport(cl, varlist = c(<span>"reads"</span>, <span>"str_detect"</span>, <span>"fixed"</span>))</span></code><code><span leaf=""><br></span></code><code><span leaf="">  <span># 并行计算 count</span></span></code><code><span leaf="">  count_list &lt;- parLapply(cl, target_sequences, <span>function</span>(<span>seq</span>) {</span></code><code><span leaf="">    <span>sum</span>(str_detect(reads, fixed(<span>seq</span>)))</span></code><code><span leaf="">  })</span></code><code><span leaf=""><br></span></code><code><span leaf="">  <span># 关闭并行集群</span></span></code><code><span leaf="">  stopCluster(cl)</span></code><code><span leaf=""><br></span></code><code><span leaf="">  <span># 汇总结果</span></span></code><code><span leaf="">  result &lt;- data.frame(</span></code><code><span leaf="">    gRNA.sequence = target_sequences,</span></code><code><span leaf="">    count = unlist(count_list)</span></code><code><span leaf="">  )</span></code><code><span leaf=""><br></span></code><code><span leaf="">  <span># 合并到最终表</span></span></code><code><span leaf="">  col_name &lt;- sub(<span>"_R1\\.fq\\.gz$"</span>, <span>""</span>, fastq_file)</span></code><code><span leaf="">  final_result &lt;- final_result %&gt;%</span></code><code><span leaf="">    left_join(result, by = <span>"gRNA.sequence"</span>) %&gt;%</span></code><code><span leaf="">    rename(!!col_name := count)</span></code><code><span leaf=""><br></span></code><code><span leaf="">  <span># 保存单个结果</span></span></code><code><span leaf="">  single_result &lt;- gRNA_library %&gt;%</span></code><code><span leaf="">    left_join(result, by = <span>"gRNA.sequence"</span>) %&gt;%</span></code><code><span leaf="">    mutate(count = ifelse(is.na(count), 0, count)) %&gt;%</span></code><code><span leaf="">    <span>select</span>(<span>id</span>, gRNA.sequence, Gene, count) %&gt;%</span></code><code><span leaf="">    rename(!!col_name := count)</span></code><code><span leaf=""><br></span></code><code><span leaf="">  write.csv(single_result, paste0(<span>"gRNA_count_"</span>, col_name, <span>".csv"</span>), row.names = FALSE)</span></code><code><span leaf="">}</span></code><code><span leaf=""><br></span></code><code><span leaf=""><span># 将NA补0</span></span></code><code><span leaf="">final_result[is.na(final_result)] &lt;- 0</span></code><code><span leaf=""><br></span></code><code><span leaf=""><span># 保存最终汇总表</span></span></code><code><span leaf="">write.csv(final_result, <span>"gRNA_count_merged_all_samples.csv"</span>, row.names = FALSE)</span></code><code><span leaf=""><br></span></code><code><span leaf=""><span># 打印总运行时间</span></span></code><code><span leaf="">end_time &lt;- Sys.time()</span></code><code><span leaf="">runtime &lt;- end_time - start_time</span></code><code><span leaf=""><span>print</span>(<span>paste</span>(<span>"总运行时间:"</span>, round(runtime, 2), <span>"秒"</span>))</span></code><code><span leaf=""><br></span></code></pre></section><p><span leaf="">在R读入所有的fastq文件，然后比对所有的sgrna序列并进行定量，得到具体的count矩阵。这个得到的具体count值比mageck的还完整，因为不考虑数据过滤，唯一的问题就是速度很慢，一个fastq文件需要处理好几个小时。这个是因为人家mageck是用C语言写的，有很多性能优化，R在内存处理等等方面还是差了很多。</span></p><p><span leaf="">所以现在才知道写mageck的大佬有多牛逼。。。</span></p><p><mp-style-type data-value="3"></mp-style-type></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/blQpCw34gvjH6BvCBXPWAw",target="_blank" rel="noopener noreferrer">原文链接</a>
