---
title: "å•ç»†èƒTPMæ•°æ®é‡ä¸ŠSeurat v5"
date: 2024-04-06T23:06:23Z
draft: ["false"]
tags: [
  "fetched",
  "ç”Ÿä¿¡æ˜Ÿçƒ"
]
categories: ["Acdemic"]
---
å•ç»†èƒTPMæ•°æ®é‡ä¸ŠSeurat v5 by ç”Ÿä¿¡æ˜Ÿçƒ
------
<div><section data-mpa-powered-by="yiban.io"><br></section><section><span>â€</span></section><section><img data-ratio="1" data-src="https://mmbiz.qpic.cn/mmbiz_png/8oKPbJgbBHrDic8XGmJ0b7oibVJajb0emLBHSvuibGG49ooBgtaAibE3TNJ00iaHviaMtdIKQJfCwtUfuHicDImtSfIxg/640?wx_fmt=png" data-type="png" data-w="64" width="20px" data-imgfileid="100010583" src="https://mmbiz.qpic.cn/mmbiz_png/8oKPbJgbBHrDic8XGmJ0b7oibVJajb0emLBHSvuibGG49ooBgtaAibE3TNJ00iaHviaMtdIKQJfCwtUfuHicDImtSfIxg/640?wx_fmt=png"><img data-imgfileid="100010584" data-ratio="1" data-src="https://mmbiz.qpic.cn/mmbiz_png/8oKPbJgbBHrDic8XGmJ0b7oibVJajb0emLPukRHCbicy4pNKeEv9qd7aWSfsx7roib2od3xPrRPicw3a0kbn0uQ6JmQ/640?wx_fmt=png" data-type="png" data-w="64" width="20px" src="https://mmbiz.qpic.cn/mmbiz_png/8oKPbJgbBHrDic8XGmJ0b7oibVJajb0emLPukRHCbicy4pNKeEv9qd7aWSfsx7roib2od3xPrRPicw3a0kbn0uQ6JmQ/640?wx_fmt=png"><span>Â ä»Šå¤©æ˜¯ç”Ÿä¿¡æ˜Ÿçƒé™ªä½ çš„<span><strong>ç¬¬937å¤©</strong></span></span><img data-imgfileid="100010585" data-ratio="1" data-src="https://mmbiz.qpic.cn/mmbiz_png/8oKPbJgbBHrDic8XGmJ0b7oibVJajb0emLBHSvuibGG49ooBgtaAibE3TNJ00iaHviaMtdIKQJfCwtUfuHicDImtSfIxg/640?wx_fmt=png" data-type="png" data-w="64" width="20px" src="https://mmbiz.qpic.cn/mmbiz_png/8oKPbJgbBHrDic8XGmJ0b7oibVJajb0emLBHSvuibGG49ooBgtaAibE3TNJ00iaHviaMtdIKQJfCwtUfuHicDImtSfIxg/640?wx_fmt=png"></section><hr><section><span><span>Â  Â </span><span>å¤§ç¥ä¸€å¥è¯ï¼Œèœé¸Ÿè·‘åŠå¹´ã€‚æˆ‘ä¸æ˜¯å¤§ç¥ï¼Œä½†æˆ‘å¯ä»¥ç¼©çŸ­ä½ èµ°å¼¯è·¯çš„åŠå¹´~</span></span></section><section><span>Â  Â å°±åƒæ­Œå„¿å”±çš„é‚£æ ·ï¼Œå¦‚æœä½ ä¸çŸ¥é“è¯¥å¾€å“ªå„¿èµ°ï¼Œå°±ç•™åœ¨è¿™å­¦ç‚¹ç”Ÿä¿¡å¥½ä¸å¥½~</span></section><p><span>Â  Â è¿™é‡Œæœ‰è±†è±†å’ŒèŠ±èŠ±çš„å­¦ä¹ å†ç¨‹ï¼Œä»æ–°æ‰‹åˆ°è¿›é˜¶ï¼Œç”Ÿä¿¡è·¯ä¸Šæœ‰ä½ æœ‰æˆ‘ï¼</span></p><section><h4><span><span> </span>0.å•ç»†èƒçš„TPMæ•°æ®</span></h4><p>ä¼—æ‰€å‘¨çŸ¥ï¼Œæ­£å¸¸çš„å•ç»†èƒæ•°æ®åº”è¯¥ç»™æˆ‘ä»¬æä¾›countï¼Œä½†ååæœ‰ä¸€äº›æ•°æ®æ˜¯ç‰¹æ®Šçš„ï¼Œç»™çš„æ˜¯tpmæ•°æ®ã€‚å‡†ç¡®æ¥è¯´åº”è¯¥æ˜¯log2(tpm/10+1)</p><blockquote><p>ä¸ºå•¥é™¤ä»¥10ï¼Œè¿™ä¹ˆè§£é‡Šçš„ï¼š<br>TPM values were divided by 10 since we estimate the complexity of our single cell libraries to be on the order of 100,000 transcripts and would like to avoid counting each transcript ~10 times, as would be the case with TPM, which may inflate the difference between the expression level of a gene in cells in which the gene is detected and those in which it is not detected.</p></blockquote><p>æŸ¥é˜…seuratçš„githubï¼Œå¯¹tpmæ•°æ®çš„è¯´æ˜æ˜¯</p><blockquote><p>If you have TPM data, you can simply manually log transform the gene expression matrix in the object@data slot before scaling the data. You have to replace your object@data slot with the desired gene expression matrix as follows: sce.all@data = log(x = norm + 1))</p></blockquote><p>åœ¨Seurat v5 ç‰ˆæœ¬ä¹‹å‰ï¼Œç›´æ¥è·³è¿‡NormalizeData è¿™ä¸ªå‡½æ•°å³å¯ï¼š</p><blockquote><p>Yes, run CreateSeuratObject() and skip the data normalization step since youâ€™re starting with normalized data</p></blockquote><p>ä½†æ˜¯å‘¢ï¼Œå½“æˆ‘ä»¬ä½¿ç”¨Seurat v5 ï¼Œç²—æš´çš„è·³è¿‡NormalizeData æ˜¯ä¼šæŠ¥é”™çš„ã€‚ä¾‹å¦‚ï¼š</p><h4><span><span> </span>1.è¯»å–æ•°æ®</span></h4><pre><code>datÂ =Â data.table::fread("GSE72056_melanoma_single_cell_revised_v2.txt.gz",data.tableÂ =Â F)<br>dat[1:4,1:4]<br><span><br>#</span><span>#Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Cell</span><br><span>#</span><span>#Â 1Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â tumor</span><br><span>#</span><span>#Â 2Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â malignant(1=no,2=yes,0=unresolved)</span><br><span>#</span><span>#Â 3Â non-malignantÂ cellÂ typeÂ (1=T,2=B,3=Macro.4=Endo.,5=CAF;6=NK)</span><br><span>#</span><span>#Â 4Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â C9orf152</span><br><span>#</span><span>#Â Â Â Cy72_CD45_H02_S758_combÂ CY58_1_CD45_B02_S974_combÂ Cy71_CD45_D08_S524_comb</span><br><span>#</span><span>#Â 1Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 72Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 58Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 71</span><br><span>#</span><span>#Â 2Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 1Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 1Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 2</span><br><span>#</span><span>#Â 3Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 2Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 1Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0</span><br><span>#</span><span>#Â 4Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0</span><br><br>datÂ =Â dat[-(1:3),]<br>library(tidyverse)<br>datÂ =Â distinct(dat,Cell,.keep_allÂ =Â T)<br>library(tibble)<br>datÂ =Â column_to_rownames(dat,"Cell")<br>dat[1:4,1:4]<br><span><br>#</span><span>#Â Â Â Â Â Â Â Â Â Â Cy72_CD45_H02_S758_combÂ CY58_1_CD45_B02_S974_comb</span><br><span>#</span><span>#Â C9orf152Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.0000Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.0000</span><br><span>#</span><span>#Â RPS11Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 9.2172Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 8.3745</span><br><span>#</span><span>#Â ELMO2Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.0000Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.0000</span><br><span>#</span><span>#Â CREB3L1Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.0000Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.0000</span><br><span>#</span><span>#Â Â Â Â Â Â Â Â Â Â Cy71_CD45_D08_S524_combÂ Cy81_FNA_CD45_B01_S301_comb</span><br><span>#</span><span>#Â C9orf152Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.0000Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.0000</span><br><span>#</span><span>#Â RPS11Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 9.3130Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 7.8876</span><br><span>#</span><span>#Â ELMO2Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 2.1263Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.0000</span><br><span>#</span><span>#Â CREB3L1Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.0000Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.0000</span><br></code></pre><h4><span><span> </span>2.æ­£å¸¸åˆ›å»ºå¯¹è±¡</span></h4><pre><code>library(Seurat)<br>sce.allÂ <span>&lt;<span>-</span>Â <span>CreateSeuratObject</span>(<span>counts</span>Â =Â <span>dat,</span>Â <br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â <span>project</span>Â =Â <span>"a"</span>,Â <br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â <span>min.cells</span>Â =Â <span>3,</span>Â <br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â <span>min.features</span>Â =Â <span>200)</span><br></span></code></pre><h4><span><span> </span>3.è´¨æ§</span></h4><pre><code>sce.all[[<span>"percent.mt"</span>]]Â &lt;-Â PercentageFeatureSet(sce.all,Â patternÂ =Â <span>"^MT-"</span>)<br>head(sce.all@meta.data,Â <span>3</span>)<br><br><span>##Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â orig.identÂ nCount_RNAÂ nFeature_RNAÂ percent.mt</span><br><span>##Â Cy72_CD45_H02_S758_combÂ Â Â Â Â Â Â Â Â Â Â Â aÂ Â Â 7143.363Â Â Â Â Â Â Â Â Â 3365Â Â Â Â Â Â Â Â Â Â 0</span><br><span>##Â CY58_1_CD45_B02_S974_combÂ Â Â Â Â Â Â Â Â Â aÂ Â Â 8915.833Â Â Â Â Â Â Â Â Â 3637Â Â Â Â Â Â Â Â Â Â 0</span><br><span>##Â Cy71_CD45_D08_S524_combÂ Â Â Â Â Â Â Â Â Â Â Â aÂ Â 12578.256Â Â Â Â Â Â Â Â Â 4660Â Â Â Â Â Â Â Â Â Â 0</span><br><br>VlnPlot(sce.all,Â <br>Â Â Â Â Â Â Â Â featuresÂ =Â c(<span>"nFeature_RNA"</span>,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â <span>"nCount_RNA"</span>,Â <br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â <span>"percent.mt"</span>),Â <br>Â Â Â Â Â Â Â Â ncolÂ =Â <span>3</span>,pt.sizeÂ =Â <span>0.5</span>)<br></code></pre><figure><img data-imgfileid="100010600" data-ratio="0.7138888888888889" data-src="https://mmbiz.qpic.cn/mmbiz_jpg/8oKPbJgbBHrBiciadYeIDMTzRywRO3OOpDGTlVIM90Bc8RANI5PGf75uEwrFAQWjsibrJkSdF71sVCpUEDjdkKo2A/640?wx_fmt=other&amp;from=appmsg" data-type="other" data-w="1080" title="image.png" src="https://mmbiz.qpic.cn/mmbiz_jpg/8oKPbJgbBHrBiciadYeIDMTzRywRO3OOpDGTlVIM90Bc8RANI5PGf75uEwrFAQWjsibrJkSdF71sVCpUEDjdkKo2A/640?wx_fmt=other&amp;from=appmsg"><figcaption>image.png</figcaption></figure><p>çº¿ç²’ä½“åŸºå› æ¯”ä¾‹ä¸º0ï¼Œå¦ä¸¤ä¸ªå›¾ä¹Ÿæ²¡æœ‰å¾ˆåˆ©ç¾¤çš„å€¼ï¼Œè¿™ä¸ªæ•°æ®å¯ä»¥ä¸è¿‡æ»¤ã€‚</p><h4><span><span> </span>4.ä¼å›¾è·³è¿‡NormalizeData</span></h4><p>å½“æˆ‘ä»¬ä¼å›¾è·³è¿‡NormalizeDataå°±ä¼šæŠ¥é”™</p><pre><code><span>sce</span>.<span>all</span>Â =Â sce.<span>all</span>Â %&gt;%Â <br>Â Â <span>#NormalizeData()Â %&gt;%Â Â </span><br>Â Â <span>FindVariableFeatures</span>()Â %&gt;%Â Â <br>Â Â <span>ScaleData</span>()<br><br><span>##Â ErrorÂ inÂ `ScaleData()`:</span><br><span>##Â !Â NoÂ layerÂ matchingÂ patternÂ 'data'Â found.Â PleaseÂ runÂ NormalizeDataÂ andÂ retry</span><br></code></pre><p>æ˜æ˜4ç‰ˆæœ¬è¿™æ ·åšæ˜¯æœ¨æœ‰é—®é¢˜çš„å˜ã€‚</p><h4><span><span> </span>5.æŸ¥å‘€ï¼</span></h4><p>å„ç§ç»†ææœ«èŠ‚éƒ½å¯ä»¥åœ¨seurat githubé¡µé¢ä¸Šæ‰¾åˆ°ç­”æ¡ˆï¼š</p><blockquote><p>In the previous versions, â€œdataâ€ was automatically populated with the raw counts prior to running NormalizeData, which is no longer the case in v5. Therefore, if you run ScaleData and the â€œdataâ€ layer does not exist, this will not work.</p><p>If you want to run ScaleData on the raw counts (which I wouldnâ€™t recommend doing in most cases), you could do the following: sce.all_small &lt;- ScaleData(sce.all_small, layer = â€œcountsâ€) https://github.com/satijalab/seurat/issues/8003</p></blockquote><p>æ‰€ä»¥è§£å†³åŠæ³•å°±æ˜¯åŠ ä¸Šä¸ªå‚æ•°layer = â€œcountsâ€ã€‚è™½ç„¶å«countsä½†æˆ‘ä»¬åˆ›å»ºå¯¹è±¡çš„æ—¶å€™æ”¾äº†ä¸ªä»€ä¹ˆä¸œè¥¿è¿›å»è‡ªå·±å¿ƒé‡Œæ²¡ç‚¹æ•°å˜›ï¼Ÿé‚£å…¶å®æ˜¯tpmğŸ˜„ã€‚</p><pre><code>sce.allÂ =Â sce.allÂ %&gt;%Â <br><span>Â Â #</span><span>NormalizeData()Â %&gt;%Â Â </span><br>Â Â FindVariableFeatures()Â %&gt;%Â Â <br>Â Â ScaleData(layerÂ =Â "counts")Â %&gt;%<br>Â Â RunPCA(pc.genesÂ =Â pbmc@var.genes)Â Â %&gt;%<br>Â Â FindNeighbors(dimsÂ =Â 1:15)Â %&gt;%Â <br>Â Â FindClusters(resolutionÂ =Â 0.5)Â %&gt;%Â <br>Â Â RunUMAP(dimsÂ =Â 1:15)Â <br><span><br>#</span><span>#Â ModularityÂ OptimizerÂ versionÂ 1.3.0Â byÂ LudoÂ WaltmanÂ andÂ NeesÂ JanÂ vanÂ Eck</span><br><span>#</span><span>#Â </span><br><span>#</span><span>#Â NumberÂ ofÂ nodes:Â 4645</span><br><span>#</span><span>#Â NumberÂ ofÂ edges:Â 150415</span><br><span>#</span><span>#Â </span><br><span>#</span><span>#Â RunningÂ LouvainÂ algorithm...</span><br><span>#</span><span>#Â MaximumÂ modularityÂ inÂ 10Â randomÂ starts:Â 0.9164</span><br><span>#</span><span>#Â NumberÂ ofÂ communities:Â 16</span><br><span>#</span><span>#Â ElapsedÂ time:Â 0Â seconds</span><br></code></pre><p>åé¢å°±å¯ä»¥æ­£å¸¸è¿›è¡Œå•¦ï¼Œé¡ºä¾¿æŠŠè¿™ä¸ªæ•°æ®åšå®Œ</p><h4><span><span> </span>6.è¿˜èƒ½åšä¸ªsingleR</span></h4><pre><code>#Â æ³¨é‡Š<br><span>library</span>(celldex)<br><span>library</span>(SingleR)<br>#refÂ &lt;-Â celldex::HumanPrimaryCellAtlasData()<br>#å› ä¸ºä¸‹è½½å¤ªæ…¢ï¼Œä½¿ç”¨äº†æå‰å­˜å¥½çš„æœ¬åœ°æ•°æ®<br>refÂ &lt;-Â <span>get</span>(load(<span>"ref_Human_all.RData"</span>))<br><span>library</span>(BiocParallel)<br>pred.scRNAÂ &lt;-Â SingleR(testÂ =Â sce.all<span>@assays</span>$RNA$counts,Â <br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â refÂ =Â ref,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â labelsÂ =Â ref$label.main,Â <br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â clustersÂ =Â sce.all<span>@active</span>.ident)<br>pred.scRNA$pruned.labels<br><br>##Â Â [<span>1</span>]Â <span>"T_cells"</span>Â Â Â Â Â Â Â Â Â Â Â <span>"T_cells"</span>Â Â Â Â Â Â Â Â Â Â Â <span>"B_cell"</span>Â Â Â Â Â Â Â Â Â Â Â <br>##Â Â [<span>4</span>]Â <span>"Neurons"</span>Â Â Â Â Â Â Â Â Â Â Â <span>"Neurons"</span>Â Â Â Â Â Â Â Â Â Â Â <span>"T_cells"</span>Â Â Â Â Â Â Â Â Â Â <br>##Â Â [<span>7</span>]Â <span>"Monocyte"</span>Â Â Â Â Â Â Â Â Â Â <span>"T_cells"</span>Â Â Â Â Â Â Â Â Â Â Â <span>"Tissue_stem_cells"</span><br>##Â [<span>10</span>]Â <span>"Neurons"</span>Â Â Â Â Â Â Â Â Â Â Â <span>"Neurons"</span>Â Â Â Â Â Â Â Â Â Â Â <span>"Endothelial_cells"</span><br>##Â [<span>13</span>]Â <span>"T_cells"</span>Â Â Â Â Â Â Â Â Â Â Â <span>"Fibroblasts"</span>Â Â Â Â Â Â Â <span>"T_cells"</span>Â Â Â Â Â Â Â Â Â Â <br>##Â [<span>16</span>]Â <span>"Neurons"</span><br><br>plotScoreHeatmap(pred.scRNA,Â clusters=pred.scRNA<span>@rownames</span>,Â fontsize.rowÂ =Â <span>9</span>,show_colnamesÂ =Â T)<br></code></pre><figure><img data-imgfileid="100010601" data-ratio="0.7138888888888889" data-src="https://mmbiz.qpic.cn/mmbiz_jpg/8oKPbJgbBHrBiciadYeIDMTzRywRO3OOpDsoicyvEISiaUf9pMBFia0ZmU844OC0YCIfWEnBaXKor9WRJZbe5eXdS8Q/640?wx_fmt=other&amp;from=appmsg" data-type="other" data-w="1080" title="" src="https://mmbiz.qpic.cn/mmbiz_jpg/8oKPbJgbBHrBiciadYeIDMTzRywRO3OOpDsoicyvEISiaUf9pMBFia0ZmU844OC0YCIfWEnBaXKor9WRJZbe5eXdS8Q/640?wx_fmt=other&amp;from=appmsg"><figcaption></figcaption></figure><pre><code>new.cluster.idsÂ &lt;-Â pred.scRNA$pruned.labels<br>names(new.cluster.ids)Â &lt;-Â levels(sce.all)<br>levels(sce.all)<br><br><span>##Â Â [1]Â <span>"0"</span>Â Â <span>"1"</span>Â Â <span>"2"</span>Â Â <span>"3"</span>Â Â <span>"4"</span>Â Â <span>"5"</span>Â Â <span>"6"</span>Â Â <span>"7"</span>Â Â <span>"8"</span>Â Â <span>"9"</span>Â Â <span>"10"</span>Â <span>"11"</span>Â <span>"12"</span>Â <span>"13"</span>Â <span>"14"</span></span><br><span>##Â [16]Â <span>"15"</span></span><br><br>sce.allÂ &lt;-Â RenameIdents(sce.all,new.cluster.ids)<br>levels(sce.all)<br><br><span>##Â [1]Â <span>"T_cells"</span>Â Â Â Â Â Â Â Â Â Â Â <span>"B_cell"</span>Â Â Â Â Â Â Â Â Â Â Â Â <span>"Neurons"</span>Â Â Â Â Â Â Â Â Â Â </span><br><span>##Â [4]Â <span>"Monocyte"</span>Â Â Â Â Â Â Â Â Â Â <span>"Tissue_stem_cells"</span>Â <span>"Endothelial_cells"</span></span><br><span>##Â [7]Â <span>"Fibroblasts"</span></span><br><br>pÂ =Â UMAPPlot(objectÂ =Â sce.all,Â pt.sizeÂ =Â <span>0.5</span>,Â labelÂ =Â <span>TRUE</span>)<br>p<br></code></pre><figure><img data-imgfileid="100010599" data-ratio="0.7138888888888889" data-src="https://mmbiz.qpic.cn/mmbiz_jpg/8oKPbJgbBHrBiciadYeIDMTzRywRO3OOpDNASnaeBot3NFr0iauK6nLsQo668rBTBwzfz0aVNUEMr9hicKKxpcZLcg/640?wx_fmt=other&amp;from=appmsg" data-type="other" data-w="1080" title="" src="https://mmbiz.qpic.cn/mmbiz_jpg/8oKPbJgbBHrBiciadYeIDMTzRywRO3OOpDNASnaeBot3NFr0iauK6nLsQo668rBTBwzfz0aVNUEMr9hicKKxpcZLcg/640?wx_fmt=other&amp;from=appmsg"><figcaption></figcaption></figure></section><section><blockquote><section><span>å¦‚æœå› ä¸ºä»£ç çœ‹ä¸æ‡‚ï¼Œè€Œè·Ÿä¸ä¸Šæ­£æ–‡çš„èŠ‚å¥ï¼Œå¯ä»¥æ¥æ‰¾æˆ‘ï¼Œç›¸å½“äºæ¥ä¸€ä¸ªæ–°æ‰‹ä¿æŠ¤æœŸã€‚æˆ‘çš„è¯¾ç¨‹éƒ½æ˜¯å¾ªç¯å¼€è¯¾ã€‚ä¸‹ä¸€æœŸçš„æ—¶é—´ï¼Œç‚¹è¿›å»å’¨è¯¢å¾®ä¿¡å’¯</span><br></section><section><a target="_blank" href="http://mp.weixin.qq.com/s?__biz=MzU4NjU4ODQ2MQ==&amp;mid=2247494223&amp;idx=1&amp;sn=956c0412cf4235ad73269a1c219cd90c&amp;chksm=fdfba20dca8c2b1b8cfafefb05d6fa736f2f126c8883086f8963e9aafeb2af39791e2e56e254&amp;scene=21#wechat_redirect" textvalue="ç”Ÿä¿¡åˆ†æâ€ç›´æ’­è¯¾ç¨‹" linktype="text" imgurl="" imgdata="null" data-itemshowtype="11" tab="innerlink" data-linktype="2">ç”Ÿä¿¡åˆ†æç›´æ’­è¯¾ç¨‹</a></section><section><a target="_blank" href="http://mp.weixin.qq.com/s?__biz=MzU4NjU4ODQ2MQ==&amp;mid=2247494009&amp;idx=1&amp;sn=035d253f973f65c0d1069e515eec2ce8&amp;chksm=fdfba13bca8c282de5a9222a2ab8bc0caa2ddf60cab543c2181e6d4a90b553f4dcde5f154adf&amp;scene=21#wechat_redirect" textvalue="ç”Ÿä¿¡æ–°æ‰‹ä¿æŠ¤å­¦ä¹ å°ç»„â€" linktype="text" imgurl="" imgdata="null" data-itemshowtype="11" tab="innerlink" data-linktype="2">ç”Ÿä¿¡æ–°æ‰‹ä¿æŠ¤å­¦ä¹ å°ç»„</a><br></section></blockquote></section><section><section><br></section></section><section><br></section><p><mp-style-type data-value="3"></mp-style-type></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/uFLgc62O-ER28vkNHJJWig",target="_blank" rel="noopener noreferrer">åŸæ–‡é“¾æ¥</a>
