---
title: "python版本的单细胞数据整合分析：scVI整合"
date: 2026-02-01T01:50:51Z
draft: ["false"]
tags: [
  "fetched",
  "生信技能树"
]
categories: ["Acdemic"]
---
python版本的单细胞数据整合分析：scVI整合 by 生信技能树
------
<div><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com" data-pm-slice="0 0 []"><blockquote><p><span leaf="">单细胞数据整合在分析中一直是个比较难得点，目前也没有一个很统一的能适用于所有情况的方法，今天先来看看 scVI 整合算法，该算法于2018年发表在Nat Methods杂志上，标题为《Deep Generative Modeling for Single-cell Transcriptomics》。<span textstyle="">在基准测试研究中，scVI 在一系列数据集上表现良好，能够在校正批次效应的同时保留生物学变异性。</span>参考：https://www.sc-best-practices.org/preamble.html</span></p></blockquote><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">批次</span></span></h2><p data-tool="mdnice编辑器"><span leaf="">在从多组学数据中去除批次效应时，必须做出两个核心选择：（1）方法和参数化，以及（2）批次协变量。由于批次效应可能在细胞的不同分组水平之间产生（例如样本、供体、数据集等），批次协变量的选择表明应保留哪些水平的变化，以及应去除哪些水平的变化。批次分辨率越高，去除的效应就越多。</span><strong><span leaf="">「批次协变量的选择将取决于你的整合任务的目标。你是想观察个体之间的差异，还是专注于共同细胞类型的变异？」</span></strong></p><p data-tool="mdnice编辑器"><span leaf="">下面是批次整合方法的三大类别：</span></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxNpbjicrBK9bTicNDBzllcLmiciaA1sxlbF0MZ1fL0icibibTmcUVVcbPht96q6B8w9fWibDm4v9icYgqCc4g/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.4759259259259259" data-type="png" data-w="1080" data-imgfileid="100062474" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxNpbjicrBK9bTicNDBzllcLmiciaA1sxlbF0MZ1fL0icibibTmcUVVcbPht96q6B8w9fWibDm4v9icYgqCc4g/640?wx_fmt=png&amp;from=appmsg"></span></figure><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">不同的方法比较</span></span></h2><p data-tool="mdnice编辑器"><span leaf="">这里有一些列各种基准测试的文章：</span></p><ul><li><section><p><span leaf="">benchmark：文献标题《A benchmark of batch-effect correction methods for single-cell RNA sequencing data》，杂志 Genome biology，</span></p></section></li><li><section><p><span leaf="">文献标题《Flexible comparison of batch correction methods for single-cell RNA-seq using BatchBench》，Nucleic acids research</span></p></section></li><li><section><p><span leaf="">Benchmarking atlas-level data integration in single-cell genomics. Nature methods, December 2021</span></p></section></li><li><section><p><span leaf="">Building the mega single-cell transcriptome ocular meta-atlas. GigaScience, October 2021</span></p></section></li></ul><p data-tool="mdnice编辑器"><span leaf="">总体而言，可以说<span textstyle="">Harmony和Seurat在简单的批效应校正任务中表现一致地良好，而scVI、scGen、scANVI和Scanorama在更复杂的数据整合任务中表现良好。</span>在选择方法时，我们建议首先考虑这些选项。</span></p><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">环境配置</span></span></h2><p data-tool="mdnice编辑器"><span leaf="">我这里还是老办法，使用的sc小环境，缺的模块安装一下：</span></p><pre data-tool="mdnice编辑器"><code><span><span leaf=""># bash代码</span></span><span leaf=""><br></span><span leaf="">conda activate sc</span><span leaf=""><br></span><span leaf="">conda install -y anndata2ri</span><span leaf=""><br></span><span leaf="">conda install -y bbknn</span><span leaf=""><br></span><span leaf="">pip install scib</span><span leaf=""><br></span><span leaf="">pip install -U scvi-tools[cuda]</span><span leaf=""><br></span></code></pre><blockquote><p><span leaf="">安装过程太折腾了，不是这个包版本不对就是那个不对！借助kimi轻松解决~</span></p></blockquote><h4 data-tool="mdnice编辑器"><span></span><span leaf="">导入模块：</span><span></span></h4><pre data-tool="mdnice编辑器"><code><span><span leaf=""># Python packages</span></span><span leaf=""><br></span><span leaf="">import anndata2ri</span><span leaf=""><br></span><span leaf="">import bbknn</span><span leaf=""><br></span><span leaf="">import matplotlib.pyplot as plt</span><span leaf=""><br></span><span leaf="">import numpy as np</span><span leaf=""><br></span><span leaf="">import pandas as pd</span><span leaf=""><br></span><span leaf="">import scanpy as sc</span><span leaf=""><br></span><span leaf="">import scib</span><span leaf=""><br></span><span leaf="">import scvi</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># # R interface</span></span><span leaf=""><br></span><span><span leaf=""># from rpy2.robjects import pandas2ri</span></span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># pandas2ri.activate()</span></span><span leaf=""><br></span><span><span leaf=""># anndata2ri.activate()</span></span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># %load_ext rpy2.ipython</span></span><span leaf=""><br></span><span><span leaf="">##===================</span></span><span leaf=""><br></span><span><span leaf="">## 上面的代码修改成了下面的</span></span><span leaf=""><br></span><span><span leaf=""># R interface</span></span><span leaf=""><br></span><span leaf="">from rpy2.robjects import conversion</span><span leaf=""><br></span><span leaf="">import pandas as pd</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 定义转换器</span></span><span leaf=""><br></span><span leaf="">pandas2ri_converter = conversion.Converter(</span><span><span leaf="">'pandas2ri'</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 激活转换上下文</span></span><span leaf=""><br></span><span leaf="">with pandas2ri_converter.context():</span><span leaf=""><br></span><span leaf="">    </span><span><span leaf=""># 在这个上下文中，pandas数据结构将自动转换为R对象</span></span><span leaf=""><br></span><span leaf="">    </span><span><span leaf=""># 你可以在这里调用R代码</span></span><span leaf=""><br></span><span leaf="">    pass</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 如果需要，可以在这里调用anndata2ri.activate()</span></span><span leaf=""><br></span><span leaf="">import anndata2ri</span><span leaf=""><br></span><span leaf="">anndata2ri.activate()</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 加载R魔法命令</span></span><span leaf=""><br></span><span leaf="">get_ipython().run_line_magic(</span><span><span leaf="">'load_ext'</span></span><span leaf="">, </span><span><span leaf="">'rpy2.ipython'</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">示例数据</span></span></h2><p data-tool="mdnice编辑器"><span leaf="">用来演示数据整合的数据集包含了几个骨髓单核细胞样本。来自这里 [Multimodal single cell data integration challenge: Results and lessons learned；A sandbox for prediction and integration of DNA, RNA, and proteins in single cells]。<span textstyle="">这里使用的数据版本已经经过预处理，以去除低质量的细胞。</span></span></p><p data-tool="mdnice编辑器"><span leaf="">可以使用scanpy读取这个数据集，以获得一个AnnData对象。</span></p><p data-tool="mdnice编辑器"><span leaf="">数据下载：如果下载有困难加我（微信 Biotree123）</span></p><pre data-tool="mdnice编辑器"><code><span><span leaf=""># bash命令</span></span><span leaf=""><br></span><span leaf="">wget https://figshare.com/ndownloader/files/45452260 -O openproblems_bmmc_multiome_genes_filtered.h5ad</span><span leaf=""><br></span></code></pre><h4 data-tool="mdnice编辑器"><span></span><span leaf="">读取数据：</span><span></span></h4><pre data-tool="mdnice编辑器"><code><span leaf="">adata_raw = sc.read_h5ad(</span><span><span leaf="">"openproblems_bmmc_multiome_genes_filtered.h5ad"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">adata_raw.layers[</span><span><span leaf="">"logcounts"</span></span><span leaf="">] = adata_raw.X</span><span leaf=""><br></span><span leaf="">adata_raw</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxNpbjicrBK9bTicNDBzllcLm8AhkR5Uno6sQIW6ZMwkemLdqJI4k5IQP9dCBL8iaIAick8oZQxBZibn0g/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.11203703703703703" data-type="png" data-w="1080" data-imgfileid="100062473" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxNpbjicrBK9bTicNDBzllcLm8AhkR5Uno6sQIW6ZMwkemLdqJI4k5IQP9dCBL8iaIAick8oZQxBZibn0g/640?wx_fmt=png&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">整个数据集包含69,249个细胞和129,921个特征的测量值。表达矩阵有两个版本，一个是</span><code><span leaf="">counts</span></code><span leaf="">，包含原始计数值；另一个是</span><code><span leaf="">logcounts</span></code><span leaf="">，包含标准化后的对数计数（这些值也存储在</span><code><span leaf="">adata.X</span></code><span leaf="">中）。</span></p><p data-tool="mdnice编辑器"><code><span leaf="">obs</span></code><span leaf="">细胞表型信息包含多个变量，其中一些是在预处理（用于质量控制）过程中计算的，还有一些包含有关样本的数据。这里感兴趣的是：</span></p><ul><li><section><code><span leaf="">cell_type</span></code><span leaf="">：每个细胞的注释标签</span></section></li><li><section><code><span leaf="">batch</span></code><span leaf="">：每个细胞的测序批次</span></section></li></ul><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">用什么作为批次标签？</span></span></h2><p data-tool="mdnice编辑器"><span leaf="">决定在数据整合中使用什么作为“批次”是整合数据时的核心选择之一。最常见的方法是将每个样本定义为一个批次（正如我们这里所做的那样），这通常会产生最强的批次校正效果。然而，样本通常与你可能想要保留的生物因素相混淆。</span></p><blockquote><p><span leaf="">例如，想象一个从组织的两个位置取样的实验。如果将样本视为批次，那么数据整合方法将试图消除它们之间的差异，因此也会消除位置之间的差异。在这种情况下，使用供体作为批次来消除个体之间的差异而不是位置之间的差异可能更为合适。</span></p><p><span leaf="">在整合许多数据集并希望捕捉个体之间差异的情况下，数据集可能是一个有用的批次协变量。</span></p></blockquote><p data-tool="mdnice编辑器"><span leaf="">在本例子中，为这两个位置使用一致的细胞类型标签，然后检测它们之间的差异，可能比为每个位置分别设置独立的簇，再分别进行注释并进行匹配要更好。</span></p><p data-tool="mdnice编辑器"><span leaf="">让我们看看不同的批次以及每个批次有多少个细胞：</span></p><pre data-tool="mdnice编辑器"><code><span leaf="">label_key = </span><span><span leaf="">"cell_type"</span></span><span leaf=""><br></span><span leaf="">batch_key = </span><span><span leaf="">"batch"</span></span><span leaf=""><br></span><span leaf="">adata_raw.obs[batch_key].value_counts()</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 为了简化操作并减少计算时间，我们将选择三个样本进行使用。</span></span><span leaf=""><br></span><span><span leaf=""># 选择三个样本进行分析</span></span><span leaf=""><br></span><span leaf="">keep_batches = [</span><span><span leaf="">"s1d3"</span></span><span leaf="">, </span><span><span leaf="">"s2d1"</span></span><span leaf="">, </span><span><span leaf="">"s3d7"</span></span><span leaf="">]</span><span leaf=""><br></span><span leaf="">adata = adata_raw[adata_raw.obs[batch_key].isin(keep_batches)].copy()</span><span leaf=""><br></span><span leaf="">adata</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><strong><span leaf="">「在这个数据集中有13个不同的批次。」</span></strong><span leaf="">在本次实验中，从一组供体中多次取样，并在不同的测序设施中进行测序，因此这里的名称是样本编号（例如“s1”）和供体（例如“d2”）的组合。</span></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxNpbjicrBK9bTicNDBzllcLmGZFX2ibaICwFlNsvibT2cl14z5kL2sxoQnprbIfq5DflhjLBzkVNZIDQ/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.7142857142857143" data-type="png" data-w="651" data-imgfileid="100062470" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxNpbjicrBK9bTicNDBzllcLmGZFX2ibaICwFlNsvibT2cl14z5kL2sxoQnprbIfq5DflhjLBzkVNZIDQ/640?wx_fmt=png&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">选择的三个样本，剩下10,270个细胞。</span></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxNpbjicrBK9bTicNDBzllcLmRdJboM8teAmZ707oGHiasXxwpuXG8UyWPic11iaiaibJ8NREO5BAAk7eJ3g/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.15925925925925927" data-type="png" data-w="1080" data-imgfileid="100062472" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxNpbjicrBK9bTicNDBzllcLmRdJboM8teAmZ707oGHiasXxwpuXG8UyWPic11iaiaibJ8NREO5BAAk7eJ3g/640?wx_fmt=png&amp;from=appmsg"></span></figure><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">数据预处理</span></span></h2><p data-tool="mdnice编辑器"><span leaf="">在</span><code><span leaf="">var</span></code><span leaf="">中存储了两种特征的注释：</span></p><ul><li><section><p><code><span leaf="">feature_types</span></code><span leaf="">——每种特征的类型（RNA或ATAC）</span></p></section></li><li><section><p><code><span leaf="">gene_id</span></code><span leaf="">——与每种特征相关的基因</span></p></section></li></ul><p data-tool="mdnice编辑器"><span leaf="">有超过10万个ATAC特征，但只有大约1.3万个基因表达（“GEX”）特征。将仅筛选出基因表达特征。还进行了简单的过滤，以确保没有特征的计数为零。还需要重新对数据进行标准化：</span></p><pre data-tool="mdnice编辑器"><code><span><span leaf=""># 筛选基因并简单过滤在全部细胞中表达为0 的基因</span></span><span leaf=""><br></span><span leaf="">adata = adata[:, adata.var[</span><span><span leaf="">"feature_types"</span></span><span leaf="">] == </span><span><span leaf="">"GEX"</span></span><span leaf="">].copy()</span><span leaf=""><br></span><span leaf="">sc.pp.filter_genes(adata, min_cells=</span><span><span leaf="">1</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">adata</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 数据标准化</span></span><span leaf=""><br></span><span leaf="">adata.X = adata.layers[</span><span><span leaf="">"counts"</span></span><span leaf="">].copy()</span><span leaf=""><br></span><span leaf="">sc.pp.normalize_total(adata)</span><span leaf=""><br></span><span leaf="">sc.pp.log1p(adata)</span><span leaf=""><br></span><span leaf="">adata.layers[</span><span><span leaf="">"logcounts"</span></span><span leaf="">] = adata.X.copy()</span><span leaf=""><br></span><span leaf="">adata</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxNpbjicrBK9bTicNDBzllcLmNtZkjBX8c9PlXemIRyyLaicXB8sEZfIKic3dYqEMQibg8WHWpMVAo915w/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.2740740740740741" data-type="png" data-w="1080" data-imgfileid="100062471" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxNpbjicrBK9bTicNDBzllcLmNtZkjBX8c9PlXemIRyyLaicXB8sEZfIKic3dYqEMQibg8WHWpMVAo915w/640?wx_fmt=png&amp;from=appmsg"></span></figure><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">整合之前</span></span></h2><p data-tool="mdnice编辑器"><strong><span leaf="">「在进行任何整合之前，始终建议先查看原始数据。」</span></strong><span leaf="">这可以提供一些关于批次效应有多大以及可能由什么引起的指示（因此哪些变量应被视为批次标签）。对于某些实验，如果样本已经重叠，甚至可能表明不需要进行整合。</span></p><h4 data-tool="mdnice编辑器"><span></span><span leaf="">进行高变基因（HVG）选择、PCA和UMAP降维：</span><span></span></h4><pre data-tool="mdnice编辑器"><code><span><span leaf=""># 整合之前</span></span><span leaf=""><br></span><span leaf="">sc.pp.highly_variable_genes(adata)</span><span leaf=""><br></span><span leaf="">sc.tl.pca(adata)</span><span leaf=""><br></span><span leaf="">sc.pp.neighbors(adata)</span><span leaf=""><br></span><span leaf="">sc.tl.umap(adata)</span><span leaf=""><br></span><span leaf="">adata.uns[batch_key + </span><span><span leaf="">"_colors"</span></span><span leaf="">] = [</span><span leaf=""><br></span><span leaf="">    </span><span><span leaf="">"<a topic-id="mgmbjce4-mrvjft" data-topic="1">#1b9e77</a>"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">    </span><span><span leaf="">"<a topic-id="mgmbjce4-saj70q" data-topic="1">#d95f02</a>"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">    </span><span><span leaf="">"<a topic-id="mgmbjce5-6k89k3" data-topic="1">#7570b3</a>"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">]  </span><span><span leaf=""># Set custom colours for batches</span></span><span leaf=""><br></span><span><span leaf="">import</span></span><span leaf=""> matplotlib.pyplot </span><span><span leaf="">as</span></span><span leaf=""> plt</span><span leaf=""><br></span><span><span leaf=""># 设置图片的宽度和高度</span></span><span leaf=""><br></span><span leaf="">plt.rcParams[</span><span><span leaf="">'figure.figsize'</span></span><span leaf="">] = (</span><span><span leaf="">6</span></span><span leaf="">, </span><span><span leaf="">7</span></span><span leaf="">)  </span><span><span leaf=""># 宽度为10英寸，高度为5英寸</span></span><span leaf=""><br></span><span leaf="">sc.pl.umap(adata, color=[label_key, batch_key], wspace=</span><span><span leaf="">1</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">绘制UMAP图，根据细胞身份和批次标签对点进行着色：</span></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxNpbjicrBK9bTicNDBzllcLmz7Dp1w1U1nGpicehiaEibazVKoFrPjjic053IBK6VxoYARFgibq8lZmao4A/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.3" data-type="png" data-w="1080" data-imgfileid="100062478" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxNpbjicrBK9bTicNDBzllcLmz7Dp1w1U1nGpicehiaEibazVKoFrPjjic053IBK6VxoYARFgibq8lZmao4A/640?wx_fmt=png&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">通常在查看umap时，会看到批次之间有明显的分离。在这里我们看到的情况更为微妙：<span textstyle="">尽管具有相同标签的细胞彼此靠近，但批次之间存在一定的偏移。</span></span></p><blockquote><p><span leaf="">如果批次能够完美重叠，或者我们可以在不进行校正的情况下发现有意义的细胞聚类，那么就没有必要进行整合。</span></p></blockquote><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">特征基因选择</span></span></h2><p data-tool="mdnice编辑器"><span leaf="">可以通过在 </span><code><span leaf="">scanpy</span></code><span leaf=""> 的 </span><code><span leaf="">highly_variable_genes()</span></code><span leaf=""> 函数中设置 </span><code><span leaf="">batch_key</span></code><span leaf=""> 参数来执行考虑批次的高变基因选择。</span><code><span leaf="">scanpy</span></code><span leaf=""> 将分别计算每个批次的高变基因（HVGs），并通过选择在最多个批次中高度可变的基因来合并结果。</span></p><pre data-tool="mdnice编辑器"><code><span><span leaf="">## 特征基因选择</span></span><span leaf=""><br></span><span leaf="">sc.pp.highly_variable_genes(</span><span leaf=""><br></span><span leaf="">    adata, n_top_genes=2000, flavor=</span><span><span leaf="">"cell_ranger"</span></span><span leaf="">, batch_key=batch_key</span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">adata.var</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxNpbjicrBK9bTicNDBzllcLm4bv1iaa8Mbz6PgWcngYafSwCzbgru8fhiaeeta8oSUQGN3VAVqLmAe5g/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.2962962962962963" data-type="png" data-w="1080" data-imgfileid="100062479" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxNpbjicrBK9bTicNDBzllcLm4bv1iaa8Mbz6PgWcngYafSwCzbgru8fhiaeeta8oSUQGN3VAVqLmAe5g/640?wx_fmt=png&amp;from=appmsg"></span></figure><ul><li><section><code><span leaf="">highly_variable_nbatches</span></code><span leaf="">：每个基因在多少个批次中被发现是高度可变的。</span></section></li><li><section><code><span leaf="">highly_variable_intersection</span></code><span leaf="">：每个基因是否在每个批次中都是高度可变的。</span></section></li><li><section><code><span leaf="">highly_variable</span></code><span leaf="">：在综合每个批次的结果后，每个基因是否被选为高度可变的。</span></section></li></ul><pre data-tool="mdnice编辑器"><code><span leaf="">n_batches = adata.var[</span><span><span leaf="">"highly_variable_nbatches"</span></span><span leaf="">].value_counts()</span><span leaf=""><br></span><span leaf="">ax = n_batches.plot(kind=</span><span><span leaf="">"bar"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">n_batches</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 选择只包含选定基因的对象，用于后续的整合分析</span></span><span leaf=""><br></span><span leaf="">adata_hvg = adata[:, adata.var[</span><span><span leaf="">"highly_variable"</span></span><span leaf="">]].copy()</span><span leaf=""><br></span><span leaf="">adata_hvg</span><span leaf=""><br></span><span><span leaf=""># AnnData object with n_obs × n_vars = 10270 × 2000</span></span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxNpbjicrBK9bTicNDBzllcLmz9uCCJ7bricUQukzrefPfXd5WqXoCzqHtNPxSzzMUOic5Vm2qAR87UAg/640?wx_fmt=png&amp;from=appmsg" data-ratio="1.08" data-type="png" data-w="575" data-imgfileid="100062475" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxNpbjicrBK9bTicNDBzllcLmz9uCCJ7bricUQukzrefPfXd5WqXoCzqHtNPxSzzMUOic5Vm2qAR87UAg/640?wx_fmt=png&amp;from=appmsg"></span></figure><h4 data-tool="mdnice编辑器"><span></span><span leaf="">使用多少基因？</span><span></span></h4><blockquote><p><strong><span leaf="">「这是一个没有明确答案的问题。」</span></strong><span leaf="">在下面使用的 </span><code><span leaf="">scvi-tools</span></code><span leaf=""> 包的作者推荐使用 1000 到 10000 个基因，但具体数量取决于上下文，包括数据集的复杂性和批次的数量。尽管选择较少的基因有助于去除批次效应最具变性的基因通常只描述主要的生物学变异），我们建议选择略多的基因，而不是选择太少，从而避免移除对稀有细胞类型或感兴趣通路重要的基因。</span></p></blockquote><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">scVI整合</span></span></h2><p data-tool="mdnice编辑器"><span leaf="">在基准测试研究中，scVI 在一系列数据集上表现良好，能够在校正批次效应的同时保留生物学变异性。</span><strong><span leaf="">「scVI 直接对原始计数进行建模」</span></strong><span leaf="">，因此提供原始计数矩阵而不是归一化的表达矩阵非常重要。</span></p><h3 data-tool="mdnice编辑器"><span></span><span><span leaf="">step1: 数据准备</span></span><span></span></h3><pre data-tool="mdnice编辑器"><code><span><span leaf=""># step1: 数据准备</span></span><span leaf=""><br></span><span leaf="">adata_scvi = adata_hvg.copy()</span><span leaf=""><br></span><span leaf="">scvi.model.SCVI.setup_anndata(adata_scvi, layer=</span><span><span leaf="">"counts"</span></span><span leaf="">, batch_key=batch_key)</span><span leaf=""><br></span><span leaf="">adata_scvi</span><span leaf=""><br></span></code></pre><h3 data-tool="mdnice编辑器"><span></span><span><span leaf="">step2: 构建 scVI 模型</span></span><span></span></h3><p data-tool="mdnice编辑器"><span leaf="">构建一个 scVI 模型对象：</span></p><pre data-tool="mdnice编辑器"><code><span><span leaf=""># step2: 构建 scVI 模型</span></span><span leaf=""><br></span><span leaf="">model_scvi = scvi.model.SCVI(adata_scvi)</span><span leaf=""><br></span><span leaf="">model_scvi</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># scVI 模型对象的结构</span></span><span leaf=""><br></span><span><span leaf=""># 在 scVI 模型中，你可以查看模型分配的各种信息，包括每个批次是如何被编码的</span></span><span leaf=""><br></span><span leaf="">model_scvi.view_anndata_setup()</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxNpbjicrBK9bTicNDBzllcLmEyGPvOJYytyibtIPia8BK7uRavwUyLTemyD8RRlbpibUjDmyZTlwOSU8w/640?wx_fmt=png&amp;from=appmsg" data-ratio="1.804661487236404" data-type="png" data-w="901" data-imgfileid="100062477" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxNpbjicrBK9bTicNDBzllcLmEyGPvOJYytyibtIPia8BK7uRavwUyLTemyD8RRlbpibUjDmyZTlwOSU8w/640?wx_fmt=png&amp;from=appmsg"></span></figure><h3 data-tool="mdnice编辑器"><span></span><span><span leaf="">step3：训练模型</span></span><span></span></h3><p data-tool="mdnice编辑器"><span leaf="">模型将被训练一定数量的周期（epochs），每个周期是一个训练迭代过程，其中每个细胞都会通过网络。</span></p><pre data-tool="mdnice编辑器"><code><span><span leaf=""># step3：训练模型</span></span><span leaf=""><br></span><span leaf="">max_epochs_scvi = np.min([round((</span><span><span leaf="">20000</span></span><span leaf=""> / adata.n_obs) * </span><span><span leaf="">400</span></span><span leaf="">), </span><span><span leaf="">400</span></span><span leaf="">])</span><span leaf=""><br></span><span leaf="">max_epochs_scvi </span><span><span leaf=""># 400</span></span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># train the model</span></span><span leaf=""><br></span><span leaf="">model_scvi.train() </span><span><span leaf=""># will take ~20-40 minutes</span></span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxNpbjicrBK9bTicNDBzllcLmcZzcdlA7aXxzOdY5M4kZNaZRew99xA4EEqbliagf6aXibXI966660Vhw/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.25833333333333336" data-type="png" data-w="1080" data-imgfileid="100062476" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxNpbjicrBK9bTicNDBzllcLmcZzcdlA7aXxzOdY5M4kZNaZRew99xA4EEqbliagf6aXibXI966660Vhw/640?wx_fmt=png&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">我这里服务器上有GPU，这个代码自动检测然后用了，</span><strong><span leaf="">「时间缩短到了4分钟！」</span></strong></p><h3 data-tool="mdnice编辑器"><span></span><span><span leaf="">step4：提取the embedding</span></span><span></span></h3><p data-tool="mdnice编辑器"><span leaf="">从训练好的模型中主要想提取的结果是每个细胞的潜在表示（latent representation）。这是一种多维嵌入，其中批次效应已被去除，可以像我们在分析单个数据集时使用PCA维度那样使用它。</span><strong><span leaf="">「这个结果存储在 </span><code><span leaf="">obsm</span></code><span leaf=""> 中，键为 </span><code><span leaf="">X_scvi</span></code><span leaf="">。」</span></strong></p><pre data-tool="mdnice编辑器"><code><span><span leaf=""># step4：提取the embedding</span></span><span leaf=""><br></span><span leaf="">adata_scvi.obsm[</span><span><span leaf="">"X_scVI"</span></span><span leaf="">] = model_scvi.get_latent_representation()</span><span leaf=""><br></span></code></pre><h3 data-tool="mdnice编辑器"><span></span><span><span leaf="">step5：UMAP降维</span></span><span></span></h3><p data-tool="mdnice编辑器"><span leaf="">像整合之前一样对数据进行可视化。这里会计算一个新的UMAP嵌入，但不是在PCA空间中寻找最近邻，而是从 scVI 给出的校正后的表示开始。</span></p><pre data-tool="mdnice编辑器"><code><span><span leaf=""># step5：UMAP降维</span></span><span leaf=""><br></span><span leaf="">sc.pp.neighbors(adata_scvi, use_rep=</span><span><span leaf="">"X_scVI"</span></span><span leaf="">) </span><span><span leaf=""># X_scVI</span></span><span leaf=""><br></span><span leaf="">sc.tl.umap(adata_scvi)</span><span leaf=""><br></span><span leaf="">sc.pl.umap(adata_scvi, color=[label_key, batch_key], wspace=</span><span><span leaf="">1</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxNpbjicrBK9bTicNDBzllcLmNZwYPECia8znWO5SJJ9jESgsiaTc0lUQ1cHpeHicRmOENyQT3tlFibkBbw/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.3" data-type="png" data-w="1080" data-imgfileid="100062480" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wxNpbjicrBK9bTicNDBzllcLmNZwYPECia8znWO5SJJ9jESgsiaTc0lUQ1cHpeHicRmOENyQT3tlFibkBbw/640?wx_fmt=png&amp;from=appmsg"></span></figure><h4 data-tool="mdnice编辑器"><span></span><span leaf="">看起来好多了！之前，不同的批次彼此分离。现在，批次之间有更多的重叠，并且每个细胞身份标签都有一个单独的团块。</span><span></span></h4><h4 data-tool="mdnice编辑器"><span></span><span leaf="">今天分享到这~</span><span></span></h4></section><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com" data-pm-slice="0 0 []"><p data-tool="mdnice编辑器"><span leaf=""><span textstyle="">友情转发：</span></span></p></section><ul><li><section><span leaf=""><a target="_blank" href="https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247545889&amp;idx=1&amp;sn=b7b37a458eead4645137126753d58c34&amp;scene=21#wechat_redirect" textvalue="生信入门&amp;数据挖掘线上直播课10月班" data-itemshowtype="0" linktype="text" data-linktype="2"><span textstyle="">生信入门&amp;数据挖掘线上直播课10月班</span></a><span textstyle="">，你的生物信息学入门课</span></span></section></li><li><section><span leaf=""><a target="_blank" href="https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247525079&amp;idx=1&amp;sn=0b997af16a58195b4192691373048fd5&amp;scene=21#wechat_redirect" textvalue="时隔5年，我们的生信技能树VIP学徒继续招生啦" data-itemshowtype="0" linktype="text" data-linktype="2"><span textstyle="">时隔5年，我们的生信技能树VIP学徒继续招生啦</span></a></span></section></li><li><section><span leaf=""><a target="_blank" href="https://mp.weixin.qq.com/s?__biz=MzUzMTEwODk0Ng==&amp;mid=2247530048&amp;idx=1&amp;sn=28aa7bbd5e00521f79e074496a5f5d66&amp;scene=21#wechat_redirect" textvalue="满足你生信分析计算需求的低价解决方案" data-itemshowtype="0" linktype="text" data-linktype="2">满足你生信分析计算需求的低价解决方案</a></span></section></li><li><section><span leaf=""><a target="_blank" href="https://mp.weixin.qq.com/mp/appmsgalbum?__biz=MzAxMDkxODM1Ng==&amp;action=getalbum&amp;album_id=1679199708449144836&amp;scene=173&amp;subscene=207&amp;sessionid=1745492310&amp;enterid=1745492314&amp;from_msgid=2247541298&amp;from_itemidx=1&amp;count=3&amp;nolastread=1#wechat_redirect" textvalue="生信故事会" data-itemshowtype="0" linktype="text" data-linktype="2">生信故事会</a>，来看看他们的生信入门故事</span></section></li><li><section><span leaf=""><a target="_blank" href="https://mp.weixin.qq.com/mp/appmsgalbum?__biz=MzAxMDkxODM1Ng==&amp;action=getalbum&amp;album_id=3690970204957147140&amp;scene=173&amp;subscene=207&amp;sessionid=1745066271&amp;enterid=1745066274&amp;from_msgid=2247540702&amp;from_itemidx=1&amp;count=3&amp;nolastread=1#wechat_redirect" textvalue="生信马拉松答疑专辑" data-itemshowtype="0" linktype="text" data-linktype="2">生信马拉松答疑专辑</a>，获取你的生信专属答疑</span></section></li></ul><section><span leaf=""><br></span></section><p><mp-style-type data-value="3"></mp-style-type></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/_sd1EGMOo9SNBBiVwvQtfg",target="_blank" rel="noopener noreferrer">原文链接</a>
