---
title: "bulk RNA-seq | 下游分析 | 功能富集分析 ORA - clusterProfiler"
date: 2024-09-14T08:43:35Z
draft: ["false"]
tags: [
  "fetched",
  "被炸熟的虾"
]
categories: ["Acdemic"]
---
bulk RNA-seq | 下游分析 | 功能富集分析 ORA - clusterProfiler by 被炸熟的虾
------
<div><p><span>在完成了</span><span>bulk RNA-seq</span><span>数据质控与差异分析后，通过设置不同的阈值，我们获得的差异基因数量可能从几个到上千个不等。</span><br></p><section><span></span></section><ul><li><section><a href="https://mp.weixin.qq.com/s?__biz=MzkwNDQwMDI5NQ==&amp;mid=2247488049&amp;idx=1&amp;sn=5f3892ab818deeece6d21f32e924e6b8&amp;scene=21#wechat_redirect" data-linktype="2"><span><strong><span>bulk RNA-seq | 下游分析 | 分析前评估</span></strong></span></a></section></li><li><section><a href="https://mp.weixin.qq.com/s?__biz=MzkwNDQwMDI5NQ==&amp;mid=2247488149&amp;idx=1&amp;sn=b402db9268122ddfc8c8f43fec7f1f3b&amp;scene=21#wechat_redirect" data-linktype="2"><span><strong><span>bulk RNA-seq | 下游分析 | 差异分析 DESeq2</span></strong></span></a></section></li><li><section><a href="https://mp.weixin.qq.com/s?__biz=MzkwNDQwMDI5NQ==&amp;mid=2247488192&amp;idx=1&amp;sn=5f1c50e176aa9495922e6e5ea69014c7&amp;scene=21#wechat_redirect" data-linktype="2"><span><strong><span>bulk RNA-seq | 下游分析 | 差异分析 DESeq2-补1-统计结果缺失值</span></strong></span></a></section></li></ul><section><span>一方面，大家不是计算机，确实很难记住所有基因的功能，另一方面生物功能的执行往往涉及多个基因或蛋白的相互作用，直接注释会发现大量功能交叠现象，导致分析结果冗余，不利于进一步的精细分析。这时最常听到的语句就是——你去做一下富集（</span><span>enrichment</span><span>）分析。</span></section><p><span>富集指的是使用某种方法使目的物质占比增大。如使用磁珠富集</span><span>T</span><span>细胞，最终的结果是使</span><span>T</span><span>细胞的占比增大。功能富集分析就是在目标基因集中寻找占比高的功能，从而推断目标基因集的关键功能。例如对差异基因集进行功能富集分析，寻找差异基因集“富集”的关键功能，进而推出样本间重要的功能差异。</span></p><p><span>严谨来说，基于数据来源和算法分类，富集分析方法主要有过度代表性分析（</span><span>Over Representation Analysis, ORA</span><span>）、功能分类打分（</span><span>Functional Class Scoring，FCS</span><span>）、通路拓扑结构（</span><span>Pathway Topology, PT</span><span>）和网络拓扑结构（</span><span>Network Topology, NT</span><span>）四种。</span></p><p><img data-galleryid="" data-imgfileid="100004569" data-ratio="0.675" data-s="300,640" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3Kw75B9aZmbs6Bmkib0icJ54XVcdt6I8QBibnoy2ra7kfPmd5oAsVfG8FykKuRZCSfareFfyUDiaLkWJQ/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3Kw75B9aZmbs6Bmkib0icJ54XVcdt6I8QBibnoy2ra7kfPmd5oAsVfG8FykKuRZCSfareFfyUDiaLkWJQ/640?wx_fmt=png&amp;from=appmsg"></p><section><span>图片来源：</span><span>Progress in Gene Functional Enrichment Analysis</span></section><p><span>我们常听说的</span><strong><span>GO</span><span>富集分析，</span><span>KEGG</span><span>富集分析，</span><span>WikiPathways</span><span>富集分析等直接以基因功能注释数据库取名的富集分析采用的都是</span><span>ORA</span><span>算法</span></strong><span>；而</span><strong><span>基因集富集分析（</span><span>GSEA</span><span>）、基因集变异分析（</span><span>GSVA</span><span>）、单样本基因集富集分析（</span><span>ssGSEA</span><span>）等，采用的是</span><span>FCS</span><span>算法</span></strong><span>。</span></p><p><span>很久前写过一篇笔记：</span><a href="https://mp.weixin.qq.com/s?__biz=MzkwNDQwMDI5NQ==&amp;mid=2247485584&amp;idx=1&amp;sn=fe72ec1999b3bb50a1f89112f8e109a8&amp;scene=21#wechat_redirect" data-linktype="2"></a><a href="https://mp.weixin.qq.com/s?__biz=MzkwNDQwMDI5NQ==&amp;mid=2247485584&amp;idx=1&amp;sn=fe72ec1999b3bb50a1f89112f8e109a8&amp;scene=21#wechat_redirect" data-linktype="2">记录 8 种常见的基因功能富集在线工具</a><span>（</span><span>Metascape</span><span>、</span><span>Enrichtool</span><span>、</span><span>DAVID</span><span>、</span><span>GeneTrail</span><span>、</span><span>G:profiler</span><span>、</span><span>WEBgestalt</span><span>、</span><span>KOBAS</span><span>与</span><span>modEnrichr</span><span>）。曾有大佬比较不同工具的分析效能，强调数据集的更新时间、内置数据集种类是选择富集分析工具时的主要参考指标，尤其是前者，如果不及时更新，很容易遗漏重要信息（</span><span>PMID: 27575621</span><span>）。</span></p><p><span><span>clusterProfiler</span>包可以抓取</span><span>GO</span><span>、</span><span>KEGG</span><span>等数据库的最新信息，因此</span><span>可以说是</span><span>R</span><span>语言富集分析（</span><span>ORA</span><span>）的不二选择。今天以我们以</span><code><span>DESeq2</span></code><span>差异分析结果为例，记录一下基础流程。</span></p><pre><code><span>library</span>(clusterProfiler)<br>packageVersion(<span>"clusterProfiler"</span>)<span>#[1]‘4.11.1’</span><br></code></pre><p><span>提取差异表达基因：</span></p><pre><code>setwd(<span>"D:/bioinformatics/2.NGS/1.bulk RNA-seq"</span>)<br>load(<span>"./data/Merge2.Rdata"</span>)<br>load(<span>"./data/DEG_DESeq2.Rdata"</span>)<br><span># 设定阈值，筛选显著上下调差异基因</span><br>logFC &lt;- <span>2</span><br>Pvalue &lt;- <span>0.01</span><br>DEG_DESeq2$Group &lt;- ifelse(DEG_DESeq2$log2FoldChange &gt; logFC &amp; DEG_DESeq2$padj &lt; Pvalue,<span>"Up"</span>,<br>                           ifelse(DEG_DESeq2$log2FoldChange &lt; -logFC &amp; DEG_DESeq2$padj &lt; Pvalue,<br>                                  <span>"Down"</span>,<span>"Stable"</span>))<br>table(DEG_DESeq2$Group)<br><span>#Down Stable     Up </span><br><span>#1824  18306    506</span><br></code></pre><h2><span>一、ORA 原理</span></h2><p><span>（想了一下还是记录了一点）</span></p><p><span>ORA</span><span>算法的主要思想是统计评估目标基因集（例如差异基因集）中属于特定功能条目（</span><span>term</span><span>）的基因的比例。如果这种比例出现的可能性（期望）很低，远低于随机事件发生的概率，那么我们就认为我们观察到的富集情况是“显著”的。</span></p><p><span>有点拗口，我们套用一下经典的“摸球”问题：</span></p><p><span>蓝色方框中的球是本次测序获得的所有基因【共</span><span>N</span><span>个】，某个特定通路</span><span>P</span><span>包括基因共</span><span>M</span><span>个（红色小球）；绿色圆圈是一次摸球事件，用来表示做了一次差异分析得到的差异基因【共</span><span>n</span><span>个】，这些基因中，有属于通路</span><span>P</span><span>的（红色球）【共</span><span>k</span><span>个】，有不属于的（黑色球）。问这次分析结果中，通路</span><span>P</span><span>的基因占比是否显著过高？</span></p><p><img data-galleryid="" data-imgfileid="100004570" data-ratio="0.6035911602209945" data-s="300,640" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3Kw75B9aZmbs6Bmkib0icJ54XWahmG2K2hQbsoR9o82BZyrlcQBEkk9kzibaOCTmiafF7mzl30AtNTYyg/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="724" src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3Kw75B9aZmbs6Bmkib0icJ54XWahmG2K2hQbsoR9o82BZyrlcQBEkk9kzibaOCTmiafF7mzl30AtNTYyg/640?wx_fmt=png&amp;from=appmsg"></p><p><span>无放回的抽样问题，至少有</span><span>k</span><span>个小球属于通路</span><span>P</span><span>的概率可以用超几何检验来计算。若</span><span>p ≤0.05</span><span>，说明在随机状态下，当前目标基因集中含有</span><span>k</span><span>个以上的</span><span>P</span><span>通路基因是极不可能的，但是当下却发生了，该事件就是一个“有意义”的事件，因此</span><span>P</span><span>通路显著富集。反之，富集不显著。</span></p><p><span>推荐看一下</span><span>Y</span><span>叔的讲解：</span><a href="https://mp.weixin.qq.com/s?__biz=MzI5NjUyNzkxMg==&amp;mid=2247489254&amp;idx=1&amp;sn=20a22e1844565fc54fcd11bfaeb4c265&amp;chksm=ec43a7a1db342eb76d747d5d050e3a08fae605011b0cfdcd325b86e003b0e94018437beb9f0d&amp;scene=21&amp;cur_album_id=1300601631955910656#wechat_redirect" data-linktype="2"><span></span></a><a href="https://mp.weixin.qq.com/s?__biz=MzI5NjUyNzkxMg==&amp;mid=2247489254&amp;idx=1&amp;sn=20a22e1844565fc54fcd11bfaeb4c265&amp;chksm=ec43a7a1db342eb76d747d5d050e3a08fae605011b0cfdcd325b86e003b0e94018437beb9f0d&amp;scene=21&amp;cur_album_id=1300601631955910656#wechat_redirect" data-linktype="2">富集分析的p值是怎么算出来的？</a></p><h2><span>二、GO &amp; KEGG</span></h2><p><span>Gene Ontology (GO)</span><span>是一个用于描述基因和蛋白质功能的标准化系统，</span><span>GO</span><span>数据库将基因/蛋白的属性分为</span><span>3</span><span>大类：</span></p><ul><li><section><span>生物过程（</span><span>Biological Process</span><span>,</span><span>BP</span><span>）：描述基因产物参与的生物学过程，例如细胞周期、信号转导等。</span></section></li><li><section><span>分子功能（</span><span>Molecular Function</span><span>,</span><span>MF</span><span>）：描述基因产物的具体分子活动，例如酶活性、结合活动等。</span></section></li><li><section><span>细胞组分（</span><span>Cellular Component</span><span>,</span><span>CC</span><span>）：描述基因产物的细胞位置或组成，例如细胞膜、细胞核等。</span></section></li></ul><p><span>KEGG</span><span> (</span><span>Kyoto Encyclopedia of Genes and Genomes</span><span>)数据库的侧重点是描述生物体内的代谢通路，以及这些通路在不同生物体系中的变化和调控。</span></p><p><span>两者通常会结合使用，以全面理解基因和蛋白质的功能及其在生物体内的作用。</span><span>对于这些最常用的注释数据库，</span><span>clusterprofiler</span><span>包分别提供了专用函数，函数语法很简单：</span></p><pre><code>enrichKEGG(gene,                <span>#待富集的基因列表，只接受entrez gene id</span><br>           organism = <span>"hsa"</span>,    <span>#指定物种</span><br>           keyType = <span>"kegg"</span>,<br>           pvalueCutoff = <span>0.05</span>, <span>#指定 p 值阈值（可指定 1 以输出全部）</span><br>           qvalueCutoff = <span>0.2</span>,  <span>#指定 q 值阈值（可指定 1 以输出全部）</span><br>           pAdjustMethod = <span>"BH"</span>,<span>#指定 p 值校正方法，比如holm,hochberg,hommel,bonferroni,BH,BY,fdr,none</span><br>           universe,<br>           minGSSize = <span>10</span>,      <span>#通路最少基因数量</span><br>           maxGSSize = <span>500</span>,     <span>#通路最大基因数量</span><br>           use_internal_data = <span>FALSE</span>) <span>#使用本地KEGG.db还是在线抓取</span><br>enrichGO(gene,                  <span>#待富集的基因列表</span><br>         OrgDb,                 <span>#指定物种的基因数据库</span><br>         keyType = <span>"ENTREZID"</span>,  <span>#指定给定的基因名称类型</span><br>         ont = <span>"MF"</span>,            <span>#GO Ontology，可选 BP、MF、CC，也可以指定ALL同时计算3者</span><br>         pvalueCutoff = <span>0.05</span>,   <span>#指定 p 值阈值（可指定 1 以输出全部）</span><br>         qvalueCutoff = <span>0.2</span>,    <span>#指定 q 值阈值（可指定 1 以输出全部）</span><br>         pAdjustMethod = <span>"BH"</span>,  <span>#指定 p 值校正方法，比如holm,hochberg,hommel,bonferroni,BH,BY,fdr,none</span><br>         universe,<br>         minGSSize = <span>10</span>,        <span>#Ontology最少基因数量</span><br>         maxGSSize = <span>500</span>,       <span>#Ontology最大基因数量</span><br>         readable = <span>FALSE</span>,      <span>#显示结果是否将gene ID转为gene Name</span><br>         pool = <span>FALSE</span>)</code></pre><h3><section><strong><span><strong><span><strong><span>KEGG富集分析</span></strong></span></strong></span></strong></section></h3><section><span>enrichKEGG</span><span>函数只接受</span><span>ENTREZ ID</span><span>，因此第一步需要利用</span><span>bitr</span><span>函数完成基因</span><span>ID</span><span>转换。</span><span>bitr</span><span>函数需要依赖</span><span>Orgdb</span><span>注释包提取不同</span><span>ID的</span><span>对应关系，不同物种的</span><span>Orgdb</span><span>注释包目录如下：</span></section><section><span>https://bioconductor.org/packages/release/BiocViews.html#___OrgDb</span></section><pre><code><span>#BiocManager::install("org.Hs.eg.db")</span><br><span>#library(org.Hs.eg.db)</span><br><span>#packageVersion("org.Hs.eg.db")#[1]‘3.15.0’</span><br><br><span>#BiocManager::install("org.Mm.eg.db")</span><br><span>library</span>(org.Mm.eg.db)<br>packageVersion(<span>"org.Mm.eg.db"</span>)<span>#[1]‘3.15.0’</span><br><br>ID_entrez &lt;- bitr(rownames(DEG_DESeq2[DEG_DESeq2$Group == <span>"Up"</span>,]),<span>#以上调基因为例</span><br>                  fromType = <span>"SYMBOL"</span>,   <span>#fromType是指你的数据ID类型是属于哪一类的</span><br>                  toType = <span>"ENTREZID"</span>,   <span>#toType是指你要转换成哪种ID类型，可以写多种，也可以只写一种</span><br>                  OrgDb = <span>"org.Mm.eg.db"</span>)<span>#Orgdb是指对应的注释包是哪个</span><br><span>#'select()' returned 1:1 mapping between keys and columns</span><br><span>#Warning message:</span><br><span>#In bitr(rownames(DEG_DESeq2[DEG_DESeq2$Group == "Up", ]), fromType = "SYMBOL",  :</span><br><span>#  4.35% of input gene IDs are fail to map...</span><br>head(ID_entrez)<br><span>#   SYMBOL ENTREZID</span><br><span>#1     Ltf    17002</span><br><span>#2    Lcn2    16819</span><br><span>#3   Farsb    23874</span><br><span>#4    Oas2   246728</span><br><span>#5 Clca3a2    80797</span><br><span>#6    Aqp5    11830</span><br></code></pre><p><span>个人习惯，如果有</span><span>ENSEMBL</span><span>，可以也尝试一下，看看转换率会不会高一丢丢，本例就没有。</span></p><pre><code>head(ids)<br>ID_ENSEMBL &lt;- ids[ids$SYMBOL %<span>in</span>% rownames(DEG_DESeq2[DEG_DESeq2$Group == <span>"Up"</span>,]),<span>"ENSEMBL"</span>]<br>ID_ENSEMBL &lt;- sapply(strsplit(ID_ENSEMBL,<span>"[.]"</span>),<span>"["</span>,<span>1</span>)<br><span>library</span>(org.Mm.eg.db)<br>ID_entrez2 &lt;- bitr(ID_ENSEMBL,<br>                   fromType = <span>"ENSEMBL"</span>,   <span>#fromType是指你的数据ID类型是属于哪一类的</span><br>                   toType = <span>"ENTREZID"</span>,   <span>#toType是指你要转换成哪种ID类型，可以写多种，也可以只写一种</span><br>                   OrgDb = <span>"org.Mm.eg.db"</span>)<span>#Orgdb是指对应的注释包是哪个</span><br><span>#'select()' returned 1:1 mapping between keys and columns</span><br><span>#Warning message:</span><br><span>#In bitr(rownames(DEG_DESeq2[DEG_DESeq2$Group == "Up", ]), fromType = "SYMBOL",  :</span><br><span>#  6.32% of input gene IDs are fail to map...</span><br>head(ID_entrez2)<br><span>#             ENSEMBL ENTREZID</span><br><span>#1 ENSMUSG00000023902   332221</span><br><span>#2 ENSMUSG00000027514    58203</span><br><span>#3 ENSMUSG00000018849   211652</span><br><span>#4 ENSMUSG00000035112    69847</span><br><span>#5 ENSMUSG00000040154   209232</span><br><span>#6 ENSMUSG00000017723    67701</span><br></code></pre><section><span>一步执行富集分析：</span></section><pre><code>packageVersion(<span>"clusterProfiler"</span>)<span>#[1]‘4.11.1’</span><br>Enrichment_KEGG &lt;- enrichKEGG(gene = unique(ID_entrez$ENTREZID),<br>                              organism = <span>"mmu"</span>, <span>#Mus musculus (house mouse)</span><br>                              pvalueCutoff = <span>0.5</span>,<br>                              qvalueCutoff = <span>0.5</span>,<br>                              pAdjustMethod = <span>"BH"</span>,<br>                              minGSSize = <span>5</span>,<br>                              maxGSSize = <span>10000</span>)<br>head(Enrichment_KEGG)<br><span>library</span>(openxlsx)<br>write.xlsx(Enrichment_KEGG@result,file = <span>"./Enrichment_KEGG.xlsx"</span>,colNames = <span>TRUE</span>,rowNames = <span>TRUE</span>)<br></code></pre><p><img data-imgfileid="100004571" data-ratio="0.17037037037037037" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3Kw75B9aZmbs6Bmkib0icJ54XXa1a3yJyXTibiclg1LJEK1gmcchPtaWc1dgcZF9AEurfQxycNRtU42sw/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3Kw75B9aZmbs6Bmkib0icJ54XXa1a3yJyXTibiclg1LJEK1gmcchPtaWc1dgcZF9AEurfQxycNRtU42sw/640?wx_fmt=png&amp;from=appmsg"></p><ul><li><section><span>ID</span><span>：</span><span>pathway</span><span>的</span>ID<span>名；</span></section></li><li><section><span>GeneRatio</span><span>：差异基因中富集到该</span><span>pathway</span><span>的基因数目/富集到所有</span><span>pathway</span><span>的总差异基因数目；</span></section></li><li><section>BgRatio<span>：该</span>pathway<span>的基因数目/所有</span>pathway<span><span>的</span>总背景基因数目；</span></section></li><li><section>pvalue<span>：显著性</span>p<span>值；</span></section></li><li><section>p.adjust<span>：多重检验矫正后的</span>p<span>值；</span></section></li><li><section>qvalue<span>：应用</span>FDR<span>校正的</span>q<span>值；</span></section></li><li><section>Count<span>：富集到该</span>pathway<span>的基因数目；</span></section></li><li><section>geneID<span>：富集到这个</span>pathway<span>的差异基因（</span>ENTREZ ID<span>）；</span></section></li></ul><p><span>我们可以根据</span>GeneRatio<span><span>与</span></span>BgRatio<span><span>信息手动计算出<span>显著性</span></span></span>p<span><span><span>值：</span></span></span></p><p><img data-galleryid="" data-imgfileid="100004583" data-ratio="0.49907407407407406" data-s="300,640" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3Kw75B9aZmbs6Bmkib0icJ54X9AXPib644GDUv2BgAxWL5sxINJ4AFcpsYkLYPyLOnG699ibGtBoFWB5g/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3Kw75B9aZmbs6Bmkib0icJ54X9AXPib644GDUv2BgAxWL5sxINJ4AFcpsYkLYPyLOnG699ibGtBoFWB5g/640?wx_fmt=png&amp;from=appmsg"></p><p><span>如果分析物种是人或大鼠，我们修改参数</span><code><span>organism</span></code><span>为</span>hsa（Homo sapiens，human）<span>或</span>rno（Rattus norvegicus，rat）<span>即可。</span></p><p><code><span>enrichKEGG</span></code><span>函数适配的物种目录如下：</span></p><section><span>https://www.genome.jp/kegg/tables/br08606.html<span>。</span></span></section><p><span>新的</span><code><span>clusterprofiler</span></code><span>版本返回结果新增了</span>category<span>信息，更新后，我们就可以很容易的按照通路所属类别展示结果了：</span></p><pre><code><span>library</span>(ggplot2)<br>p1 &lt;- dotplot(Enrichment_KEGG,x = <span>"GeneRatio"</span>,color = <span>"p.adjust"</span>,showCategory = <span>10</span>) +<br>      facet_grid(rows = vars(category),scales = <span>'free_y'</span>,space = <span>'free_y'</span>) +<br>      scale_color_gradientn(colors = c(<span>'#BF1E27'</span>,<span>'#FEB466'</span>,<span>'#F9FCCB'</span>,<span>'#6296C5'</span>,<span>'#38489D'</span>)) +<br>      theme_bw(base_size = <span>18</span>)<br>p2 &lt;- barplot(Enrichment_KEGG,x = <span>"GeneRatio"</span>,color = <span>"p.adjust"</span>,showCategory = <span>10</span>) +<br>      facet_grid(rows = vars(category),scales = <span>'free_y'</span>,space = <span>'free_y'</span>) +<br>      scale_fill_gradientn(colors = c(<span>'#BF1E27'</span>,<span>'#FEB466'</span>,<span>'#F9FCCB'</span>,<span>'#6296C5'</span>,<span>'#38489D'</span>)) + <br>      theme_bw(base_size = <span>18</span>)<br></code></pre><p><img data-galleryid="" data-imgfileid="100004578" data-ratio="0.4583333333333333" data-s="300,640" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3Kw75B9aZmbs6Bmkib0icJ54XyhbMYAETY7Uu7hkKRgk0JwTXGtGMhDU3UtsUSdxL6huZ1movXbpLiag/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3Kw75B9aZmbs6Bmkib0icJ54XyhbMYAETY7Uu7hkKRgk0JwTXGtGMhDU3UtsUSdxL6huZ1movXbpLiag/640?wx_fmt=png&amp;from=appmsg"></p><section><span></span></section><section><span>输出结果</span>geneID<span>列默认保留</span>ENTREZ ID<span>，不易阅读，可以使用</span><code><span>setReadable</span></code><span>函数将</span>geneID<span>列的</span>ENTREZ ID<span>重转为</span>Symbol<span>：</span></section><pre><code>Enrichment_KEGG2 &lt;- setReadable(Enrichment_KEGG,OrgDb = org.Mm.eg.db,keyType = <span>"ENTREZID"</span>)<br>head(Enrichment_KEGG2)<br><span>library</span>(openxlsx)<br>write.xlsx(Enrichment_KEGG2@result,file = <span>"./Enrichment_KEGG2.xlsx"</span>,colNames = <span>TRUE</span>,rowNames = <span>TRUE</span>)<br></code></pre><section><img data-imgfileid="100004577" data-ratio="0.17407407407407408" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3Kw75B9aZmbs6Bmkib0icJ54XWmeia9hxMdRSuUZTv60Jibbetsic0dXcjQMwM0wjGXsicwLJoLPYib80AeA/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3Kw75B9aZmbs6Bmkib0icJ54XWmeia9hxMdRSuUZTv60Jibbetsic0dXcjQMwM0wjGXsicwLJoLPYib80AeA/640?wx_fmt=png&amp;from=appmsg"><span></span></section><section><span>默认输出结果中是没有图中展示的</span><strong><span>富集倍数（</span>Fold Enrichment<span>）</span></strong><span>及</span><strong><span>富集因子（</span>Rich Factor<span>）信息</span></strong><span>的，如果有需要，可自行计算添加：</span></section><ul><li><section><span>富集因子：差异基因中富集到该通路的基因数（</span>count<span>）/背景基因中富集到该通路的基因数（</span>BgRatio<span>的分子）；</span></section></li><li><section><span>富集倍数：</span>GeneRatio/BgRatio<span>；</span></section></li></ul><pre><code><span>library</span>(dplyr)<br><span>#计算Rich Factor（富集因子）：</span><br>Enrichment_KEGG2 &lt;- mutate(Enrichment_KEGG,<br>                           RichFactor = Count / as.numeric(sub(<span>"/\\d+"</span>, <span>""</span>, BgRatio)))<br><span>#计算Fold Enrichment（富集倍数）：</span><br>Enrichment_KEGG2 &lt;- mutate(Enrichment_KEGG2, <br>                           FoldEnrichment = parse_ratio(GeneRatio) / parse_ratio(BgRatio))<br>head(Enrichment_KEGG2@result)<span></span></code></pre><h3><span><strong><span><strong><span><strong><span><strong><span>GO富集分析</span></strong></span></strong></span></strong></span></strong></span></h3><section><code><span>enrichGO</span></code><span>函数</span><strong><span>可以接受多种</span>ID</strong><span>，包括</span>SYMBOL<span>与</span>ENTREZ ID<span>：</span></section><pre><code><span>#####SYMBOL</span><br>Enrichment_GOBP &lt;- enrichGO(gene = rownames(DEG_DESeq2[DEG_DESeq2$Group == <span>"Up"</span>,]),<br>                            OrgDb = org.Mm.eg.db, <span>#记得修改对应org包</span><br>                            ont = <span>"BP"</span>, <span>#"BP"，"MF","CC"三选一,或"ALL"合并</span><br>                            keyType = <span>"SYMBOL"</span>,<br>                            pvalueCutoff = <span>0.1</span>,<br>                            qvalueCutoff = <span>0.1</span>,<br>                            pAdjustMethod = <span>"BH"</span>,<br>                            minGSSize = <span>5</span>,<br>                            maxGSSize = <span>10000</span>,<br>                            readable = <span>FALSE</span>)<br>head(Enrichment_GOBP@result)<br><span>#####ENTREZ ID</span><br>Enrichment_GOBP2 &lt;- enrichGO(gene = unique(ID_entrez$ENTREZID),<br>                            OrgDb = org.Mm.eg.db, <span>#记得修改对应org包</span><br>                            ont = <span>"BP"</span>, <span>#"BP"，"MF","CC"三选一,或"ALL"合并</span><br>                            keyType = <span>"ENTREZID"</span>,<br>                            pvalueCutoff = <span>0.1</span>,<br>                            qvalueCutoff = <span>0.1</span>,<br>                            pAdjustMethod = <span>"BH"</span>,<br>                            minGSSize = <span>5</span>,<br>                            maxGSSize = <span>10000</span>,<br>                            readable = <span>TRUE</span>)<br>head(Enrichment_GOBP2@result)<br><span>library</span>(openxlsx)<br>write.xlsx(Enrichment_GOBP@result,file = <span>"./Enrichment_GOBP.xlsx"</span>,colNames = <span>TRUE</span>,rowNames = <span>TRUE</span>)<br></code></pre><p><span><img data-imgfileid="100004579" data-ratio="0.18425925925925926" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3Kw75B9aZmbs6Bmkib0icJ54XI6vqlJ7yDJnzQhJbM9bjNvpYamRqYR0d8fnsia2Nlib8a8iaQfUqIl6NA/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3Kw75B9aZmbs6Bmkib0icJ54XI6vqlJ7yDJnzQhJbM9bjNvpYamRqYR0d8fnsia2Nlib8a8iaQfUqIl6NA/640?wx_fmt=png&amp;from=appmsg"></span></p><section><span>和</span>KEGG<span>富集结果表基本相同，共</span>9<span>列：</span></section><ul><li><section>ID<span>：</span>GO term ID<span>号；</span></section></li><li><section>Description<span>：</span>GO term<span>的功能描述；</span></section></li><li><section>GeneRatio<span>：差异基因中富集到该</span>GO term<span>的基因数/差异基因总数；</span></section></li><li><section>BgRatio<span>：背景基因中富集到该</span>GO term<span>的基因数/背景基因总数；</span></section></li><li><section>pvalue<span>：显著性</span>p<span>值；</span></section></li><li><section>p.adjust<span>：多重检验矫正后的</span>p<span>值；</span></section></li><li><section>qvalue<span>：应用</span>FDR<span>校正的</span>q<span>值；</span></section></li><li><section>geneID<span>：富集到这个</span>GO term<span>的差异基因（已转换为</span>gene name<span>）；</span></section></li><li><section>Count<span>：富集到该</span>GO term<span>的基因数量。</span></section></li></ul><section><span>因为只有被对应物种</span><code><span>org</span></code><span>包记录且被注释数据库收录的基因才会被纳入分析（</span>GeneRatio<span>列的总基因数目），因此，这两种方法的分析结果是完全相同的。</span></section><section><span><strong><span>注意</span></strong></span><span>：相比</span><code><span>enrichKEGG</span></code><span>函数，</span>enrichGO<span>函数可以直接通过设置参数</span><code><span>readable = FALSE</span></code><span>将</span>geneID<span>列的</span>ENTREZ ID<span>重转为</span><span>SYMBOL</span><span>（如果输入基因列表就是</span>SYMBOL<span>无需转换输出即是</span>SYMBOL<span>）。</span></section><h4><span><strong><span><strong><span><strong><span>移除冗余term</span></strong></span></strong></span></strong></span></h4><section><span>由</span><span>于</span>GO<span>的注释体系是一个有向无环图，</span>MF<span>、</span>CC<span>、</span>BP<span>为一级</span>term<span>，节点之间保持严格的“父子”关系，根据包含关系往下分为二级、三级以及更低层级的分类，层级越低对功能描述越详细。</span></section><pre><code><span>#BiocManager::install("topGO")</span><br><span>library</span>(topGO)<br>plotGOgraph(Enrichment_GOBP)<br><span>#goplot(Enrichment_GOBP) #来自enrichplot</span><br></code></pre><p><img data-galleryid="" data-imgfileid="100004580" data-ratio="1" data-s="300,640" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3Kw75B9aZmbs6Bmkib0icJ54Xlg3mWvJkK7UkSfoCibjiap8c8IJjEvZ6LKlvYic7MZPWBiaCD63YYIDeFw/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3Kw75B9aZmbs6Bmkib0icJ54Xlg3mWvJkK7UkSfoCibjiap8c8IJjEvZ6LKlvYic7MZPWBiaCD63YYIDeFw/640?wx_fmt=png&amp;from=appmsg"></p><p><span>详细讲解可以阅读：</span><a href="https://mp.weixin.qq.com/s?__biz=MzA5NzQzOTgzMw==&amp;mid=2650826394&amp;idx=1&amp;sn=040ca718d3c3e162d8263da35b215d7e&amp;chksm=8b541facbc2396ba48945b3ef231267b3bb39437643e01db692095973b3d5a4efe09057504bb&amp;scene=21#wechat_redirect" data-linktype="2"><span></span></a><a href="https://mp.weixin.qq.com/s?__biz=MzA5NzQzOTgzMw==&amp;mid=2650826394&amp;idx=1&amp;sn=040ca718d3c3e162d8263da35b215d7e&amp;chksm=8b541facbc2396ba48945b3ef231267b3bb39437643e01db692095973b3d5a4efe09057504bb&amp;scene=21#wechat_redirect" data-linktype="2">GO数据库的分类层级说明</a></p><section><span>父</span>term<span>可能仅因为包含“</span>over-represented<span>”的子</span>term<span>的所有基因而被显著富集。因此，富集的</span>GO<span>富集分析结果通常太长并且包含冗余术语。</span><code><span>clusterProfiler</span></code><span>提供了一个</span><code><span>simplify</span></code><span>函数来消除这种冗余的</span>GO term<span>。</span></section><section><span>小心</span><span>simplify</span><code><span></span></code><span>同名函数很多，建议指定<span>clusterProfiler</span>调用！！</span></section><pre><code>Enrichment_GOBP2 &lt;- clusterProfiler::simplify(Enrichment_GOBP, cutoff = <span>0.7</span>, <br>                                              by = <span>"p.adjust"</span>,select_fun = min)<br>dim(Enrichment_GOBP@result) <span>#[1] 4929    9</span><br>dim(Enrichment_GOBP2@result) <span>#[1] 228   9</span><br></code></pre><section><span>4929</span><span>个仅保留</span><span>228</span><span>，删减的非常多。</span></section><section><span><strong><span>注意</span></strong></span><span>：分析类别设置为</span><span>MF</span><span>、</span><span>CC</span><span>、</span><span>BP</span><span>都可以执行去冗余，但是</span>"ALL"<span>不可以。</span></section><h2><span>三、其他通路数据库</span></h2><section><span>除了</span><span>GO</span><span>与</span><span>KEGG</span><span>，</span><span></span><span><span>clusterProfiler</span>包也提供了一些其他数据库的富集分析接口，例如</span><span>WikiPathways</span><span>与</span><span>Pathway Commons</span><span>：</span></section><ul><li><section><p><strong><span>WikiPathways</span></strong><span>是一个由社区驱动的生物通路数据库，涵盖了不同物种的生物学通路。它是一个开放平台，用户可以提交和编辑通路信息。由于其社区驱动的特点，它可以包含更广泛的或新兴的通路信息。</span></p></section></li><li><section><p><strong><span>Pathway Commons</span></strong><span>是一个汇集了来自众多资源的多个物种的公开获取通路数据，整合了多种通路数据库。包括生化反应，复合物组装、转运和催化事件，以及涉及蛋白质复合物、</span><span>DNA</span><span>、</span><span>RNA</span><span>、小分子化合物和配合物在内的物理相互作用。</span></p></section></li></ul><pre><code><span>##ORA analysis for WikiPathways</span><br>enrichWP(gene, organism, <span>...</span>)<br><span>#enrichWP支持物种</span><br>get_wp_organisms()<br><span># [1] "Anopheles gambiae"        "Arabidopsis thaliana"     "Bos taurus"              </span><br><span># [4] "Caenorhabditis elegans"   "Canis familiaris"         "Danio rerio"             </span><br><span># [7] "Drosophila melanogaster"  "Equus caballus"           "Gallus gallus"           </span><br><span>#[10] "Homo sapiens"             "Mus musculus"             "Pan troglodytes"         </span><br><span>#[13] "Populus trichocarpa"      "Rattus norvegicus"        "Saccharomyces cerevisiae"</span><br><span>#[16] "Solanum lycopersicum"     "Sus scrofa"               "Zea mays"         </span><br><br><span>##ORA analysis for Pathway Commons</span><br>enrichPC(gene, <span>source</span>, keyType = <span>"hgnc"</span>, <span>...</span>)<br></code></pre><section><span>个人测试结果</span><span>enrichWP</span><span>函数只支持</span><span>ENTREZ ID</span><span>，似乎没有</span><span>keyType</span><span>参数选项的函数都只能老老实实使用</span><span>ENTREZ ID</span><span>。</span></section><pre><code>Enrichment_WP &lt;- enrichWP(gene = unique(ID_entrez$ENTREZID),<br>                           <span>#OrgDb = org.Mm.eg.db, #记得修改对应org包</span><br>                           organism = <span>"Mus musculus"</span>,<br>                           pvalueCutoff = <span>0.1</span>,<br>                           qvalueCutoff = <span>0.1</span>,<br>                           pAdjustMethod = <span>"BH"</span>,<br>                           minGSSize = <span>5</span>,<br>                           maxGSSize = <span>10000</span>)<br>head(Enrichment_WP@result)<br><br>Enrichment_WP2 &lt;- setReadable(Enrichment_WP,OrgDb = org.Mm.eg.db,keyType = <span>"ENTREZID"</span>)<br>head(Enrichment_WP2@result)<br><br>save(Enrichment_KEGG,Enrichment_KEGG2,<br>     Enrichment_GOBP,Enrichment_GOBP2,<br>     Enrichment_WP,Enrichment_WP2,<br>     file = <span>"./ORA.Rdata"</span>)<br></code></pre><h2><span>四、自定义注释基因集</span></h2><section><span><span>clusterProfiler</span><span>包提供了</span><span>enricher</span><span>函数作为</span><strong><span>通用</span><span>ORA</span><span>接口</span></strong><span>，可以输入其他来源的注释基因集。</span></span></section><section><span><span><span>enricher</span><span>函数</span>要求的</span><span>term</span><span>格式为</span><span>2</span><span>列</span><span>data.frame</span><span>，分别为</span><span>term</span><span>与</span><span>gene</span><span>。例如我们测试</span><span>Reactome</span><span>通路，刚好从</span><span>MSigDB</span><span>数据库下载，使用</span><span>read.gmt</span><span>函数读入：</span></span></section><ul><li><section><a href="https://mp.weixin.qq.com/s?__biz=MzkwNDQwMDI5NQ==&amp;mid=2247487045&amp;idx=1&amp;sn=43dae7bc4ae133c23091ab4c18d2af67&amp;scene=21#wechat_redirect" data-linktype="2"><span><strong><span>从MSigDB查找感兴趣的基因集</span></strong></span></a></section></li><li><section><a href="https://mp.weixin.qq.com/s?__biz=MzkwNDQwMDI5NQ==&amp;mid=2247487702&amp;idx=1&amp;sn=c8147cf4180cf62b60111bce5178c621&amp;scene=21#wechat_redirect" data-linktype="2"><span><strong><span>MSigDB | 基因集简述</span></strong></span></a></section></li></ul><pre><code>GeneSet &lt;- read.gmt(<span>"D:/data/MSigdb/mouse/msigdb_v2023.2.Mm_GMTs/m2.cp.reactome.v2023.2.Mm.entrez.gmt"</span>)<br>head(GeneSet)<br><span>#                    term  gene</span><br><span>#1 REACTOME_ABACAVIR_ADME 75894</span><br><span>#2 REACTOME_ABACAVIR_ADME 11522</span><br><span>#3 REACTOME_ABACAVIR_ADME 76952</span><br><span>#4 REACTOME_ABACAVIR_ADME 20517</span><br><span>#5 REACTOME_ABACAVIR_ADME 20518</span><br><span>#6 REACTOME_ABACAVIR_ADME 20519</span><br></code></pre><p><span>基本参数依旧是类似的：</span></p><pre><code>Enrichment_Reactome &lt;- enricher(unique(ID_entrez$ENTREZID), <span>#待富集的基因列表</span><br>                                pvalueCutoff = <span>0.5</span>,   <span>#指定 p 值阈值（可指定 1 以输出全部）</span><br>                                qvalueCutoff = <span>0.5</span>,   <span>#指定 q 值阈值（可指定 1 以输出全部）</span><br>                                pAdjustMethod = <span>"BH"</span>, <span>#指定 p 值校正方法</span><br>                                <span># universe = rownames(DEG_DESeq2), #背景基因</span><br>                                minGSSize = <span>5</span>,        <span>#term最少基因数量</span><br>                                maxGSSize = <span>10000</span>,    <span>#term最大基因数量</span><br>                                TERM2GENE = GeneSet)  <span>#annotation of term TO gene mapping</span><br></code></pre><p><strong><span>这里有两个个人认为需要注意的点：</span></strong></p><section><span>1、</span><span>输入基因列表需要与</span><span>注释基因集的</span><span></span><span>基因类型一致；</span></section><section><span>2、</span><span>如果<span>注释基因集</span>只是几个</span><span>涵盖基因数目很少的</span><span><span>term</span></span><span>，注意设置背景基因，默认使用</span><span>term</span><span>全部基因作为背景基因，例如只输入</span><span>REACTOME DEVELOPMENTAL BIOLOGY</span><span>：</span></section><pre><code>Enrichment_Reactome2 &lt;- enricher(unique(ID_entrez$ENTREZID),<br>                                 pvalueCutoff = <span>1</span>,<br>                                 qvalueCutoff = <span>1</span>,<br>                                 pAdjustMethod = <span>"BH"</span>,<br>                                 minGSSize = <span>5</span>,<br>                                 maxGSSize = <span>10000</span>,<br>                                 TERM2GENE = GeneSet[GeneSet$term == <span>"REACTOME_DEVELOPMENTAL_BIOLOGY"</span>,])<br><br>length(unique(GeneSet$gene)) <span>#[1] 9082</span><br>length(unique(GeneSet[GeneSet$term == <span>"REACTOME_DEVELOPMENTAL_BIOLOGY"</span>,<span>"gene"</span>])) <span>#[1] 484</span><br>Enrichment_Reactome@result[<span>"REACTOME_DEVELOPMENTAL_BIOLOGY"</span>,<span>3</span>:<span>7</span>]<br><span>#                               GeneRatio  BgRatio       pvalue    p.adjust      qvalue</span><br><span>#REACTOME_DEVELOPMENTAL_BIOLOGY    28/221 484/9082 1.629723e-05 0.002536935 0.002373105</span><br>Enrichment_Reactome2@result[<span>"REACTOME_DEVELOPMENTAL_BIOLOGY"</span>,<span>3</span>:<span>7</span>]<br><span>#                               GeneRatio BgRatio pvalue p.adjust qvalue</span><br><span>#REACTOME_DEVELOPMENTAL_BIOLOGY     28/28 484/484      1        1     NA</span><br></code></pre><section><span>但背景基因纳入标准又是啥呢？</span></section><h2><span>五、结果可视化</span></h2><section><span>富集分析结果可视化之前写过一些笔记，应该可以满足基本需求<img data-src="https://res.wx.qq.com/t/wx_fed/we-emoji/res/v1.3.10/assets/newemoji/2_12.png" data-ratio="1" data-w="128" src="https://res.wx.qq.com/t/wx_fed/we-emoji/res/v1.3.10/assets/newemoji/2_12.png">，有其他需要可以在评论区留言<img data-ratio="1" data-src="https://res.wx.qq.com/t/wx_fed/we-emoji/res/v1.3.10/assets/newemoji/Yellowdog.png" data-w="128" src="https://res.wx.qq.com/t/wx_fed/we-emoji/res/v1.3.10/assets/newemoji/Yellowdog.png"></span><br></section><ul><li><section><p><a href="https://mp.weixin.qq.com/s?__biz=MzkwNDQwMDI5NQ==&amp;mid=2247485520&amp;idx=1&amp;sn=c41cec901b0050f9a774e278a4c1662e&amp;chksm=c086d498f7f15d8ed2009697f9680b82f1396e2b35d20b3b44c772f52773f6da5f20c89f8953&amp;scene=21&amp;cur_album_id=3053206297459900416#wechat_redirect" data-linktype="2"><span><strong><span>富集分析可视化 | 第一节 条形图</span></strong></span></a></p></section></li><li><section><p><a href="https://mp.weixin.qq.com/s?__biz=MzkwNDQwMDI5NQ==&amp;mid=2247485533&amp;idx=2&amp;sn=121351b4921b4c086e4721492935d6e4&amp;chksm=c086d495f7f15d830097f4fa02fd2647ba59de4dc074eda523fe64407eca272f60e9494cc3d7&amp;scene=21&amp;cur_album_id=3053206297459900416#wechat_redirect" data-linktype="2"><span><strong><span>富集分析可视化 | 第二节 气泡图</span></strong></span></a></p></section></li><li><section><p><a href="https://mp.weixin.qq.com/s?__biz=MzkwNDQwMDI5NQ==&amp;mid=2247485533&amp;idx=1&amp;sn=65a2a6331ea45306e9982a8f5c896d3e&amp;chksm=c086d495f7f15d83b3044acf2c4f9711d1f85ba1a3bd11fea40af3beb2facf1bd04a0a323c84&amp;scene=21&amp;cur_album_id=3053206297459900416#wechat_redirect" data-linktype="2"><span><strong><span>富集分析可视化 | 第三节 棒棒糖图</span></strong></span></a></p></section></li><li><section><p><a href="http://mp.weixin.qq.com/s?__biz=MzkwNDQwMDI5NQ==&amp;mid=2247487131&amp;idx=1&amp;sn=2c9d8d00d8123c24ae37a90b518c9190&amp;chksm=c086d253f7f15b457ddaf3a8e5c73af9519561bcbf281562a91b9ac68e22b879d75a371f7ca9&amp;scene=21#wechat_redirect" data-linktype="2"><span><strong><span>富集分析可视化 | KEGG通路分组绘图</span></strong></span></a></p></section></li></ul><section><span></span></section><section><span>最近后台经常受到一些“没头没脑”的留言，询问也没回复<img data-ratio="1" data-src="https://res.wx.qq.com/t/wx_fed/we-emoji/res/v1.3.10/assets/newemoji/2_05.png" data-w="128" src="https://res.wx.qq.com/t/wx_fed/we-emoji/res/v1.3.10/assets/newemoji/2_05.png">。</span></section><section><span><br></span></section><h3><span><span><span><strong>参考文章：</strong></span></span></span></h3><p><span><strong><span>1、生信技工：</span></strong></span><span><strong><span>富集分析：（一）概述</span></strong></span></p><p><span><strong><span>2、联川生物：</span></strong></span><a href="https://mp.weixin.qq.com/s?__biz=MjM5MzIwNTI4OA==&amp;mid=2650330996&amp;idx=1&amp;sn=a93d1964bb6d9d08d06968fef03a9bff&amp;scene=21#wechat_redirect" data-linktype="2"><span><strong><span>【联川沧海】富集分析ORA（通用富集分析）、FCS（GSEA、ssGSEA、GSVA）一网打尽</span></strong></span></a></p><p><span><strong><span>3、生信技工：</span></strong></span><span><strong><span>富集分析：（一）概述</span></strong></span></p><p><span><strong><span>4、观科研：</span></strong></span><a href="https://mp.weixin.qq.com/s?__biz=MzkzNzQxMjQ4MA==&amp;mid=2247489057&amp;idx=2&amp;sn=7c4b3382137c597c07a100affbaaea89&amp;scene=21#wechat_redirect" data-linktype="2"><span><strong><span>Bulk RNA-seq | 第7期. GO富集，我收藏了</span></strong></span></a></p><p><span><strong><span>5、基迪奥生物：</span></strong></span><a href="https://mp.weixin.qq.com/s?__biz=MzA5NzQzOTgzMw==&amp;mid=2650931485&amp;idx=1&amp;sn=6f3d5f842b19252869e6a2b5955bccc0&amp;chksm=8b56782bbc21f13ddf2a9f94c82efef9b2cebdd64b65c248667c46b0523764a3dce6619bbfd0&amp;scene=21#wechat_redirect" data-linktype="2"><span><strong><span>KEGG富集分析及可视化，一把子拿捏！</span></strong></span></a></p><p><span><strong><span>6、基迪奥生物：</span></strong></span><a href="https://mp.weixin.qq.com/s?__biz=MzA5NzQzOTgzMw==&amp;mid=2650932365&amp;idx=1&amp;sn=f857135cbdda7c6d9decf4824da2ef68&amp;scene=21#wechat_redirect" data-linktype="2"><span><strong><span>GO富集分析及可视化，一把子拿捏！</span></strong></span></a></p><section><img data-imgfileid="100003752" data-ratio="0.05278592375366569" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/1LTeQhNfr8sUH75oYsoDaqjPCTiaukEmS8tWricW7LnLKKfIE9jKBexibqamsrlibaaXmuc2nicaYibfDFBNCmqX5mBw/640?wx_fmt=other&amp;wxfrom=5&amp;wx_lazy=1&amp;wx_co=1&amp;tp=webp" data-type="png" data-w="341" src="https://mmbiz.qpic.cn/sz_mmbiz_png/1LTeQhNfr8sUH75oYsoDaqjPCTiaukEmS8tWricW7LnLKKfIE9jKBexibqamsrlibaaXmuc2nicaYibfDFBNCmqX5mBw/640?wx_fmt=other&amp;wxfrom=5&amp;wx_lazy=1&amp;wx_co=1&amp;tp=webp"></section><section><br></section><section><span>最后罗嗦几句：</span><span>关于</span><span>clusterprofiler</span><span>包的用法，</span><span>Y</span><span>叔专门写了一本书：</span><span>https://yulab-smu.top/biomedical-knowledge-mining-book/</span></section><section><span></span></section><section><span>除此之外，在</span><span>Y</span><span>叔的公众号也有大量关于</span><span>clusterprofiler</span><span>包的推文：</span><a href="https://mp.weixin.qq.com/mp/appmsgalbum?__biz=MzI5NjUyNzkxMg==&amp;action=getalbum&amp;album_id=1300601631955910656&amp;sessionid=1686559868&amp;subscene=169&amp;ascene=0&amp;realreporttime=1686559874842&amp;clicktime=1686559874&amp;enterid=1686559874&amp;forceh5=1" data-linktype="2"><span></span></a><a href="https://mp.weixin.qq.com/mp/appmsgalbum?__biz=MzI5NjUyNzkxMg==&amp;action=getalbum&amp;album_id=1300601631955910656&amp;sessionid=1686559868&amp;subscene=169&amp;ascene=0&amp;realreporttime=1686559874842&amp;clicktime=1686559874&amp;enterid=1686559874&amp;forceh5=1" data-linktype="2">clusterprofiler合集</a><span>，其中很多推文是对使用者问题的回答，很多推文中说过问题</span><span>/bug</span><span>都被修复了。</span></section><section data-tools="135编辑器" data-id="116886" draggable="true"><section><section><section><section><strong data-brushtype="text">被炸熟的虾</strong></section></section></section><section><section><section data-width="35%"><section><img data-cropselx1="0" data-cropselx2="115" data-cropsely1="0" data-cropsely2="115" data-imgfileid="100003750" data-ratio="1" data-src="https://mmbiz.qpic.cn/sz_mmbiz_jpg/dRYYdqiaan3JjDfj2H8p6gg3CB25AGthbwzrotao4ev5tIe0utthbZRK8yOoDOuTzOSoTSnPWn61IdDCnXsnaiag/640?wx_fmt=other&amp;wxfrom=5&amp;wx_lazy=1&amp;wx_co=1&amp;tp=webp" data-type="jpeg" data-w="258" data-width="100%" src="https://mmbiz.qpic.cn/sz_mmbiz_jpg/dRYYdqiaan3JjDfj2H8p6gg3CB25AGthbwzrotao4ev5tIe0utthbZRK8yOoDOuTzOSoTSnPWn61IdDCnXsnaiag/640?wx_fmt=other&amp;wxfrom=5&amp;wx_lazy=1&amp;wx_co=1&amp;tp=webp"></section></section><section data-width="63%"><section><section><strong>自己的摸索，发现问题麻烦告诉作者，光速回来改正</strong><strong><img data-imgfileid="100003751" data-ratio="1" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3KLQGNxibvI4SdtUfxjINkz2DO8cGEPTgbcMJ357RvXJ7IvWU2p7anljpePpakI9oKu3icJxobBDp5w/640?wx_fmt=other&amp;wxfrom=5&amp;wx_lazy=1&amp;wx_co=1&amp;tp=webp" data-type="png" data-w="64" src="https://mmbiz.qpic.cn/sz_mmbiz_png/dRYYdqiaan3KLQGNxibvI4SdtUfxjINkz2DO8cGEPTgbcMJ357RvXJ7IvWU2p7anljpePpakI9oKu3icJxobBDp5w/640?wx_fmt=other&amp;wxfrom=5&amp;wx_lazy=1&amp;wx_co=1&amp;tp=webp"></strong></section></section></section></section></section></section></section><p><mp-style-type data-value="10000"></mp-style-type></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/in1O4Wte1jz8bqGwuAaAdw",target="_blank" rel="noopener noreferrer">原文链接</a>
