---
title: "单细胞亚群绝对数量和相对比例的探索"
date: 2024-09-14T15:39:45Z
draft: ["false"]
tags: [
  "fetched",
  "生信技能树"
]
categories: ["Acdemic"]
---
单细胞亚群绝对数量和相对比例的探索 by 生信技能树
------
<div><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com"><p data-tool="mdnice编辑器">前面我在：<a href="https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247533053&amp;idx=1&amp;sn=bd155638b280e875dcae14622198b65b&amp;scene=21#wechat_redirect" data-linktype="2">单细胞转录组降维聚类分群过滤基因和过滤细胞的区别</a> 介绍了文献，题目为“Revealing the transcriptional heterogeneity of organ-specific metastasis in human gastric cancer using single-cell RNA Sequencing”。通讯作者是浙江大学的范骁辉教授，于2022年发表在Clin Transl Med杂志（IF=10.6），这个胃癌单细胞数据集GSE163558的单细胞转录组在降维聚类分群后，就可以看细胞比例的变化情况。</p><p data-tool="mdnice编辑器">首先就需要查看文献里面的实验设计，一般来说会在文献的附件：</p><figure data-tool="mdnice编辑器"><img data-imgfileid="100049445" data-ratio="0.5601851851851852" data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwSXsEnIfoqrDlYuBCaTicbevpAWGyEia1ImKLJpC50I55USVpQ0MKXlLOKAZLzEiacJ0G82B5VicibPhA/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwSXsEnIfoqrDlYuBCaTicbevpAWGyEia1ImKLJpC50I55USVpQ0MKXlLOKAZLzEiacJ0G82B5VicibPhA/640?wx_fmt=png&amp;from=appmsg"></figure><p data-tool="mdnice编辑器">实验共收集了来自六名患者的样本，包括原发肿瘤、非肿瘤邻近组织和转移性肿瘤样本，合起来是10个样品的10x技术的单细胞转录组，可以分成6组，如下所示：</p><ol data-tool="mdnice编辑器"><li><section><strong>原发肿瘤样本（Primary Tumour, PT）</strong>：</section></li></ol><ul><li><section>收集了三份原发肿瘤样本，分别命名为PT1、PT2和PT3。这些样本代表了肿瘤最初发生的位置。</section></li></ul><li><section><strong>非肿瘤邻近组织样本（Adjacent Non-tumoural, NT）</strong>：</section></li><ul><li><section>收集了一份非肿瘤邻近组织样本，命名为NT1。此样本来自肿瘤附近的组织，但本身不是肿瘤，用于对照分析。</section></li></ul><li><section><strong>转移性肿瘤样本（Metastatic, M）</strong>：</section></li><ul><li><section>卵巢转移样本一份，命名为O1。此样本来自卵巢中的转移肿瘤。</section></li><li><section>腹膜转移样本一份，命名为P1。此样本来自腹膜中的转移肿瘤。</section></li><li><section>淋巴结转移样本两份，命名为LN1和LN2。这些样本来自淋巴结中的转移肿瘤。</section></li><li><section>肝脏转移样本两份，命名为Li1和Li2。这些样本来自肝脏中的转移肿瘤。</section></li><li><section><strong>肝脏转移样本（Liver Metastasis, Li）</strong>：</section></li><li><section><strong>淋巴结转移样本（Lymph Nodes Metastasis, LN）</strong>：</section></li><li><section><strong>腹膜转移样本（Peritoneal Metastasis, P）</strong>：</section></li><li><section><strong>卵巢转移样本（Ovary Metastasis, O）</strong>：</section></li></ul><p data-tool="mdnice编辑器">这样的话，仅仅是（Primary Tumour, PT）组，以及 （Lymph Nodes Metastasis, LN）里面是有多个样品的，其它的每个组都是单个样品。</p><h3 data-tool="mdnice编辑器"><span></span><span>整理临床信息</span><span></span></h3><p data-tool="mdnice编辑器">这个时候其实都不需要前面的降维聚类分群后的单细胞转录组表达量矩阵文件了，仅仅是每个细胞的身份信息即可：</p><pre data-tool="mdnice编辑器"><span></span><code>load(<span>'phe.Rdata'</span>)<br>phe$group = ifelse(grepl(<span>'Control'</span>,phe$orig.ident),<br>                   <span>'control'</span>,<span>'treat'</span>)<br>phe$group = gsub(<span>'[0-9]'</span>,<span>''</span>,phe$orig.ident)<br>head(phe) <br>gplots::balloonplot(<br>  table(phe$group,phe$orig.ident)<br>)<br><br>&gt;  head(phe[,c(<span>1</span>,<span>16</span>,<span>17</span>)])<br>                                  orig.ident    celltype group<br>GSM5004180_PT1_AAACCCACAACAACAA-<span>1</span>        PT1 neutrophils    PT<br>GSM5004180_PT1_AAACCCAGTCAAAGTA-<span>1</span>        PT1      Tcells    PT<br>GSM5004180_PT1_AAACCCAGTGGATACG-<span>1</span>        PT1 neutrophils    PT<br>GSM5004180_PT1_AAACCCATCGTTCCCA-<span>1</span>        PT1 neutrophils    PT<br>GSM5004180_PT1_AAACGAAAGCTGTTCA-<span>1</span>        PT1 neutrophils    PT<br>GSM5004180_PT1_AAACGAACAGGCACAA-<span>1</span>        PT1 neutrophils    PT<br></code></pre><p data-tool="mdnice编辑器">可以看到：</p><pre data-tool="mdnice编辑器"><span></span><code>      Li1   Li2   LN1   LN2   NT1    O1    P1   PT1   PT2   PT3<br>  Li  1293 10321     0     0     0     0     0     0     0     0<br>  LN     0     0  3898  7501     0     0     0     0     0     0<br>  NT     0     0     0     0  2254     0     0     0     0     0<br>  O      0     0     0     0     0  1439     0     0     0     0<br>  P      0     0     0     0     0     0  4038     0     0     0<br>  PT     0     0     0     0     0     0     0  2339  6528  2976<br></code></pre><h3 data-tool="mdnice编辑器"><span></span><span>批量组合不同变量（分组，样品）看细胞类型比例</span><span></span></h3><p data-tool="mdnice编辑器">我写了一个函数来针对我们的phe数据框里面的任意列的组合去做单细胞亚群细胞数量的统计：</p><pre data-tool="mdnice编辑器"><span></span><code>cal_table = <span>function</span>(x,y,prefix ){<br>  <span># x = phe$orig.ident</span><br>  <span># y = phe$celltype</span><br>  <span>library</span>(sur)<br>  <span>library</span>(reshape2)<br>  tbl =  table(x,y)<br>  pdf(paste0(prefix,<span>'-table.pdf'</span>),width = <span>10</span>,height = <span>10</span>)<br>  gplots::balloonplot( tbl )<br>  dev.off() <br>  df = dcast(as.data.frame(tbl),x~y)<br>  head(df)<br>  write.csv(  df ,paste0(prefix,<span>'-table.csv'</span>))<br>  <br>  <span># ptb = round(sur::percent.table(x,y),2)</span><br>  ptb = round(<span>100</span>*tbl/rowSums(df[,-<span>1</span>]),<span>2</span>)<br>  <br>  pdf(paste0(prefix,<span>'-percent-table.pdf'</span>),width = <span>10</span>,height = <span>10</span>)<br>  gplots::balloonplot( ptb )<br>  dev.off()<br>  write.csv(  dcast(as.data.frame(ptb),x~y) ,paste0(prefix,<span>'-percent-table.csv'</span>)) <br>  <br>}<br><br>cal_table(phe$orig.ident,phe$celltype,prefix = <span>'celltype-vs-orig.ident'</span>)<br>cal_table(phe$group,phe$celltype,prefix = <span>'celltype-vs-group'</span>)<br>cal_table(phe$group,phe$seurat_clusters,prefix = <span>'seurat_clusters-vs-group'</span>)<br>cal_table(phe$group,phe$RNA_snn_res.0.8,prefix = <span>'RNA_snn_res.0.8-vs-group'</span>)<br></code></pre><p data-tool="mdnice编辑器">首先是可以看绝对数量，可以看到的是t细胞占比非常夸张，都过半了，而中性粒细胞和b细胞呢，有两个组的特异性：</p><figure data-tool="mdnice编辑器"><img data-imgfileid="100049446" data-ratio="1.0592592592592593" data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwSXsEnIfoqrDlYuBCaTicbexSiaaZicrKJ5AgNuygiapdwFAoghNCVzGnduiaBZaCib86TBCGvPQwl6gfA/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwSXsEnIfoqrDlYuBCaTicbexSiaaZicrKJ5AgNuygiapdwFAoghNCVzGnduiaBZaCib86TBCGvPQwl6gfA/640?wx_fmt=png&amp;from=appmsg"></figure><p data-tool="mdnice编辑器">然后可以看到相对比例：</p><figure data-tool="mdnice编辑器"><img data-imgfileid="100049447" data-ratio="1.0583333333333333" data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwSXsEnIfoqrDlYuBCaTicbessgia1rF1iasxNKr4icdKf33sicuj5aZ1rMxkqEgqxQLUgBDF2ZCcgChoA/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwSXsEnIfoqrDlYuBCaTicbessgia1rF1iasxNKr4icdKf33sicuj5aZ1rMxkqEgqxQLUgBDF2ZCcgChoA/640?wx_fmt=png&amp;from=appmsg"></figure><p data-tool="mdnice编辑器">值得注意的是，上面的相对比例应该是在具体的某个分组内部，或者在某个样品内部计算不同单细胞亚群的比例。而不是在某个单细胞亚群内部去计算它来源于不同组或者不同样品的比例，如果计算错误会得到“南辕北辙”的结论。</p><h3 data-tool="mdnice编辑器"><span></span><span>上面的单细胞亚群表达量比例的表格也可以进行多种多样的可视化：</span><span></span></h3><p data-tool="mdnice编辑器">比如折线图：</p><pre data-tool="mdnice编辑器"><span></span><code>x=<span>'celltype'</span>;y=<span>'group'</span> <br>plot_data &lt;- data.frame(table(phe[, y ],<br>                              phe[, x ]))<br>head(plot_data)<br>plot_data<span>$Total</span> &lt;- apply(plot_data,1,<span>function</span>(x)sum(plot_data[plot_data<span>$Var1</span> == x[1],3]))<br>plot_data &lt;- plot_data %&gt;% mutate(Percentage = round(Freq/Total,3) * 100)<br>colnames(plot_data) &lt;- c(<span>"group"</span>,<span>"cluster"</span>,<span>"Freq"</span>,<span>"Total"</span>,<span>"Percentage"</span>) <br>head(plot_data)<br>ggplot(plot_data,aes(group,Percentage,group= cluster,color =cluster))+geom_point(size=4)+<br>  geom_line(position = position_dodge(0.1),cex=2)+theme_bw()+theme_test(base_size = 30) <br><br>ggsave(<span>"celltype-vs-group-percent-plot.pdf"</span>,width = 9,height = 7)<br><br></code></pre><p data-tool="mdnice编辑器">如下所示：</p><figure data-tool="mdnice编辑器"><img data-imgfileid="100049444" data-ratio="0.7891156462585034" data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwSXsEnIfoqrDlYuBCaTicbeIF3lYFiahfU7fkPCNUyARzWroSlKKCLpSxhoZq5sfXMQwON3oEcZlDg/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="882" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wwSXsEnIfoqrDlYuBCaTicbeIF3lYFiahfU7fkPCNUyARzWroSlKKCLpSxhoZq5sfXMQwON3oEcZlDg/640?wx_fmt=png&amp;from=appmsg"></figure><p data-tool="mdnice编辑器">也可以是箱线图：</p><pre data-tool="mdnice编辑器"><span></span><code>x = phe$orig.ident<br>y = phe$celltype<br>tbl =  table(x,y)<br>df = dcast(as.data.frame(tbl),x~y)<br>ptb = round(<span>100</span>*tbl/rowSums(df[,-<span>1</span>]),<span>2</span>)<br>df = as.data.frame(ptb)<br>colnames(df)=c(<span>'orig.ident'</span>,<span>'celltype'</span>,<span>'Percentage'</span>)<br>head(df)<br>gpinfo=unique(phe[,c(<span>'orig.ident'</span>,<span>'group'</span>)]);gpinfo<br>df=merge(df,gpinfo,by=<span>'orig.ident'</span>)<br>head(df)<br><br>p &lt;- ggplot(df,aes(x=group,y=Percentage))+<br>  geom_boxplot(outlier.alpha=<span>0</span>, aes(fill=celltype))+facet_grid(~celltype,scales = <span>"free"</span>)+<br>  geom_jitter(aes(x = group, y = Percentage,color=celltype))+ <br>  scale_color_npg() +<br>  scale_fill_npg() +<br>  theme_bw()+<br>  theme(axis.text.x = element_text(face = <span>"bold"</span>,angle = <span>45</span>, hjust=<span>1</span>, vjust=<span>0.5</span>,size = <span>14</span>), <br>        legend.position= <span>"none"</span>,<br>        strip.background = element_rect(color=<span>"black"</span>,fill = <span>"steelblue3"</span>, linetype = <span>"solid"</span>),<br>        strip.text = element_text(face = <span>"bold"</span>, color = <span>"black"</span>,hjust = <span>0.5</span>, size = <span>12</span>,),<br>        plot.title=element_text(face=<span>"bold.italic"</span>,size=<span>"20"</span>, color=<span>"brown"</span>,hjust = <span>0.5</span>),<br>        axis.text.y = element_text(face = <span>"bold"</span>,size = <span>14</span>) ,<br>        axis.title = element_text(size = <span>14</span>),<br>        panel.grid.major = element_blank(),<br>        panel.grid.minor = element_blank()) <br>p   <br>pro=<span>'Percentage-of-each--celltype-in-orig.ident-by-group'</span><br>ggsave(paste0(pro,<span>"_boxplot.pdf"</span>),width = <span>12</span>,height = <span>6</span>)<br><br></code></pre><p data-tool="mdnice编辑器">当然了，不同的数据集有不同的比例情况，是需要自己慢慢的折腾的。</p><h1>如何计算各单细胞亚群比例并且显示在图例</h1><blockquote data-tool="mdnice编辑器"><p>交流群有小伙伴问自己参考我们的降维聚类分群后，虽然很容易拿到可视化的umap图，但是看很多文献里面的图例就很丰富，问是不是有什么参数可以调整，得到如下所示的图例样式：</p></blockquote><p><img data-galleryid="" data-imgfileid="100049482" data-ratio="0.2675925925925926" data-s="300,640" data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wx6fdrrHBtP25NdiazhYrlndhkzkOxzwEA8JK1byvicH7SeeKHLP3ExCVnzMibaWeo3QbFhRKxKffd6Q/640?wx_fmt=png&amp;wxfrom=13&amp;tp=wxpic" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wx6fdrrHBtP25NdiazhYrlndhkzkOxzwEA8JK1byvicH7SeeKHLP3ExCVnzMibaWeo3QbFhRKxKffd6Q/640?wx_fmt=png&amp;wxfrom=13&amp;tp=wxpic"></p><figure data-tool="mdnice编辑器"><figcaption>亚群比例并且显示在图例</figcaption></figure><p data-tool="mdnice编辑器">这个目前我没有看到什么样的参数可以做到，不过实现起来并不可能，只需要自己计算好各个单细胞亚群的比例即可，以我们熟知的pbmc3k数据集为例，大家先安装这个数据集对应的包 <code>SeuratData</code>，就可以使用它啦：</p><pre data-tool="mdnice编辑器"><span data-lazy-bgimg="https://mmbiz.qpic.cn/mmbiz_svg/GqIlejFTbNg6nxSV9Rt0mJZSmPGMjvrgCCheQDmlHp96qfWMP7hR0wqPfPo4l3b7KKicT5owVGQccPQkPUKPYc9ppvh4PPQJib/640?wx_fmt=svg" data-fail="0"></span><code><span># install.packages('devtools')</span><br><span># devtools::install_github('satijalab/seurat-data')</span><br><span>library</span>(SeuratData) <span>#加载seurat数据集  </span><br><span>library</span>(Seurat)<br>getOption(<span>'timeout'</span>)<br>options(timeout=<span>10000</span>)<br><span>#InstallData("pbmc3k")  </span><br>data(<span>"pbmc3k"</span>)  <br>sce &lt;- pbmc3k.final  <br>colnames(sce@meta.data)<br>DimPlot(sce,reduction = <span>'umap'</span>,  <br>        group.by = <span>'seurat_clusters'</span>,<br>        label.box = <span>T</span>,  label = <span>T</span>,repel = <span>T</span>)+<br>  DimPlot(sce,reduction = <span>'umap'</span>,  <br>          group.by = <span>'seurat_annotations'</span>,<br>          label.box = <span>T</span>,  label = <span>T</span>,repel = <span>T</span>)<br>gplots::balloonplot(table(sce$seurat_annotations,sce$seurat_clusters))<br><br></code></pre><p data-tool="mdnice编辑器">很明显的可以看到这个pbmc3k数据集里面自带顺序编号的亚群以及作者给出来了的担心亚群生物学名字：</p><p><img data-galleryid="" data-imgfileid="100049483" data-ratio="0.45092592592592595" data-s="300,640" data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wx6fdrrHBtP25NdiazhYrlndvyo9LHwLFs76lnWzOkiaIw99iaV4DJGushibUG2iaGLSyhib4g2ZeSdribBg/640?wx_fmt=png&amp;tp=wxpic&amp;wxfrom=5&amp;wx_lazy=1&amp;wx_co=1" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wx6fdrrHBtP25NdiazhYrlndvyo9LHwLFs76lnWzOkiaIw99iaV4DJGushibUG2iaGLSyhib4g2ZeSdribBg/640?wx_fmt=png&amp;tp=wxpic&amp;wxfrom=5&amp;wx_lazy=1&amp;wx_co=1"></p><figure data-tool="mdnice编辑器"><figcaption>两个注释</figcaption></figure><p data-tool="mdnice编辑器">我们只需要简单的计算单细胞亚群比例并且显示在图例：</p><pre data-tool="mdnice编辑器"><span data-lazy-bgimg="https://mmbiz.qpic.cn/mmbiz_svg/GqIlejFTbNg6nxSV9Rt0mJZSmPGMjvrgCCheQDmlHp96qfWMP7hR0wqPfPo4l3b7KKicT5owVGQccPQkPUKPYc9ppvh4PPQJib/640?wx_fmt=svg" data-fail="0"></span><code><br>df = data.frame(clu=names(table(sce$seurat_clusters)),<br>                per=sprintf(<span>"%1.2f%%"</span>, <span>100</span>*table(sce$seurat_clusters)/length(sce$seurat_clusters)))<br>sce$per = df[match(sce$seurat_clusters,df$clu),<span>2</span>]<br>sce$new = paste0(sce$seurat_clusters,<span>":"</span>,sce$seurat_annotations,<span>"("</span>,sce$per,<span>")"</span>)<br>table(sce$new )<br>DimPlot(sce,reduction = <span>'umap'</span>,  <br>        group.by = <span>'new'</span>,<br>        label.box = <span>T</span>,  label = <span>F</span>,repel = <span>T</span>)<br><br></code></pre><p data-tool="mdnice编辑器">如下所示：</p><p><img data-galleryid="" data-imgfileid="100049484" data-ratio="0.5509259259259259" data-s="300,640" data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wx6fdrrHBtP25NdiazhYrlndlYwcIMyCfNRe57qzHjvuxpKvWtmyWEiaNyU1DZdU4B5viap6X5NyBKoA/640?wx_fmt=png&amp;tp=wxpic&amp;wxfrom=5&amp;wx_lazy=1&amp;wx_co=1" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wx6fdrrHBtP25NdiazhYrlndlYwcIMyCfNRe57qzHjvuxpKvWtmyWEiaNyU1DZdU4B5viap6X5NyBKoA/640?wx_fmt=png&amp;tp=wxpic&amp;wxfrom=5&amp;wx_lazy=1&amp;wx_co=1"></p><figure data-tool="mdnice编辑器"><figcaption>显示在图例</figcaption></figure><p data-tool="mdnice编辑器">是不是很简单啊！</p><h1>美化你的单细胞亚群比例图</h1><blockquote data-tool="mdnice编辑器"><p>单细胞数据分析里面最基础的就是降维聚类分群，参考前面的例子：<a href="https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247497956&amp;idx=1&amp;sn=5d4deb7cf7b7848b3e2273cbd663bb6a&amp;scene=21#wechat_redirect" data-linktype="2">人人都能学会的单细胞聚类分群注释</a>  ，这个大家基本上问题不大了，使用seurat标准流程即可，不过它默认出图并不好看，详见以前我们做的投票：<a href="https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247500816&amp;idx=1&amp;sn=25568540a54d63225def0804c05025f8&amp;scene=21#wechat_redirect" data-linktype="2">可视化单细胞亚群的标记基因的5个方法</a>，下面的5个基础函数相信大家都是已经烂熟于心了：</p></blockquote><ul data-tool="mdnice编辑器"><li><section>VlnPlot(pbmc, features = c("MS4A1", "CD79A"))</section></li><li><section>FeaturePlot(pbmc, features = c("MS4A1", "CD79A"))</section></li><li><section>RidgePlot(pbmc, features = c("MS4A1", "CD79A"), ncol = 1)</section></li><li><section>DotPlot(pbmc, features = unique(features)) + RotatedAxis()</section></li><li><section>DoHeatmap(subset(pbmc, downsample = 100), features = features, size = 3)</section></li></ul><p data-tool="mdnice编辑器">最近，<strong>郑州大学第一附属医院的史阳</strong>同学无私的分享了他对这些基础函数的改造，颜值说不上巅峰，但打败基础函数是没有问题的， 同时也算是抛砖引玉吧，希望广大生信技能树粉丝们都投稿分享自己的创作，投稿请发邮件到 jmzeng1314@163.com</p><h3 data-tool="mdnice编辑器"><span></span>仍然是以大家熟知的pbmc3k数据集为例<span></span></h3><p data-tool="mdnice编辑器">大家先安装这个数据集对应的包，并且对它进行降维聚类分群，参考前面的例子：<a href="https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247497956&amp;idx=1&amp;sn=5d4deb7cf7b7848b3e2273cbd663bb6a&amp;scene=21#wechat_redirect" data-linktype="2">人人都能学会的单细胞聚类分群注释</a>  ，而且每个亚群找高表达量基因，都存储为Rdata文件。</p><pre data-tool="mdnice编辑器"><span data-lazy-bgimg="https://mmbiz.qpic.cn/mmbiz_svg/GqIlejFTbNjwekMlGTsw0WMZDJZiaUUo6xZIGOcPKW2dw163M9ZlvLNFcNkene8KjxEtuIKQUZEd0nyreQQ7BurzPJ1UT9Miaw/640?wx_fmt=svg" data-fail="0"></span><code><span># 0.安装R包 ----</span><br><span># devtools::install_github('caleblareau/BuenColors')</span><br><span># utils::install.packages(pkgs = "ggstatsplot")</span><br><span># InstallData("pbmc3k") </span><br><br><span># 1.加载R包和测试数据 ----</span><br>rm(list = ls())<br><span>library</span>(SeuratData) <span>#加载seurat数据集  </span><br>getOption(<span>'timeout'</span>)<br>options(timeout = <span>10000</span>)<br>data(<span>"pbmc3k"</span>)  <br>sce &lt;- pbmc3k.final  <br><span>library</span>(Seurat)<br><br>DimPlot(sce,reduction = <span>"umap"</span>,label = <span>TRUE</span>) <br>unique(Idents(sce))<br>sce$celltype = Idents(sce)<br><br><span># 2.单细胞分析基本流程 ----</span><br><span># 主要是拿到 tSNE和Umap的坐标，因为默认的 pbmc3k.final 是没有的</span><br>sce &lt;- NormalizeData(sce, normalization.method =  <span>"LogNormalize"</span>,  <br>                     scale.factor = <span>1e4</span>) <br>sce &lt;- FindVariableFeatures(sce, <br>                            selection.method = <span>"vst"</span>, nfeatures = <span>2000</span>)  <br>sce &lt;- ScaleData(sce) <br>sce &lt;- RunPCA(object = sce, pc.genes = VariableFeatures(sce))  <br>sce &lt;- FindNeighbors(sce, dims = <span>1</span>:<span>15</span>)<br>sce &lt;- FindClusters(sce, resolution = <span>0.8</span>) <br>set.seed(<span>123</span>)<br>sce &lt;- RunTSNE(object = sce, dims = <span>1</span>:<span>15</span>, do.fast = <span>TRUE</span>) <br>sce &lt;- RunUMAP(object = sce, dims = <span>1</span>:<span>15</span>, do.fast = <span>TRUE</span>)<br>DimPlot(object = sce, reduction = <span>"umap"</span>,label = <span>TRUE</span>)   <br><span># 3.定义分组信息 ---- </span><br>Idents(sce) = sce$celltype<br><br><span># 4.寻找各个单细胞亚群的特征性高表达量基因----</span><br>sce.markers &lt;- FindAllMarkers(object = sce, <br>                              only.pos = <span>TRUE</span>,<br>                              min.pct = <span>0.25</span>, <br>                              thresh.use = <span>0.25</span>) <br><br>save(sce, sce.markers, file = <span>'tmp.Rdata'</span>)<br></code></pre><h3 data-tool="mdnice编辑器"><span></span>接下来改造 单细胞亚群比例图<span></span></h3><p data-tool="mdnice编辑器">先放一下改造效果图：</p><p><img data-galleryid="" data-imgfileid="100049491" data-ratio="0.5981481481481481" data-s="300,640" data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wyqjgXMibhrPibjbIbgZ9Ay9qwFoIAdhzJu3w2GubEJ2s5XCU2azEqJc01OrL2BIcr5ibUNdyhZMU8JQ/640?wx_fmt=png&amp;tp=wxpic&amp;wxfrom=5&amp;wx_lazy=1&amp;wx_co=1" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wyqjgXMibhrPibjbIbgZ9Ay9qwFoIAdhzJu3w2GubEJ2s5XCU2azEqJc01OrL2BIcr5ibUNdyhZMU8JQ/640?wx_fmt=png&amp;tp=wxpic&amp;wxfrom=5&amp;wx_lazy=1&amp;wx_co=1"></p><figure data-tool="mdnice编辑器"><figcaption> </figcaption></figure><p data-tool="mdnice编辑器">代码如下所示：</p><pre data-tool="mdnice编辑器"><span data-lazy-bgimg="https://mmbiz.qpic.cn/mmbiz_svg/GqIlejFTbNjwekMlGTsw0WMZDJZiaUUo6xZIGOcPKW2dw163M9ZlvLNFcNkene8KjxEtuIKQUZEd0nyreQQ7BurzPJ1UT9Miaw/640?wx_fmt=svg" data-fail="0"></span><code><span>source</span>(<span>'PropPlot.R'</span>)<br>table(sce$seurat_clusters, sce$celltype)<br>PropPlot(object = sce, groupBy = <span>'celltype'</span>) + <br>  PropPlot(object = sce, groupBy = <span>'seurat_clusters'</span>) <br></code></pre><p data-tool="mdnice编辑器">所有的细节都在<strong> PropPlot.R  文件里面的 PropPlot 函数</strong>， 是自定义的，内容如下所示：</p><pre data-tool="mdnice编辑器"><span data-lazy-bgimg="https://mmbiz.qpic.cn/mmbiz_svg/GqIlejFTbNjwekMlGTsw0WMZDJZiaUUo6xZIGOcPKW2dw163M9ZlvLNFcNkene8KjxEtuIKQUZEd0nyreQQ7BurzPJ1UT9Miaw/640?wx_fmt=svg" data-fail="0"></span><code>PropPlot &lt;- <span>function</span>(object, groupBy){<br>  <span># (1)获取绘图数据</span><br>  plot_data = object@meta.data %&gt;% <br>    dplyr::select(orig.ident, {{groupBy}}) %&gt;% <br>    dplyr::rename(group = as.name(groupBy))<br><br>  <span># (2)绘图</span><br>  figure = ggbarstats(data = plot_data, <br>                      x = group, y = orig.ident,<br>                      package = <span>'ggsci'</span>,<br>                      palette = <span>'category20c_d3'</span>,<br>                      results.subtitle = <span>FALSE</span>,<br>                      bf.message = <span>FALSE</span>,<br>                      proportion.test = <span>FALSE</span>,<br>                      label.args = list(size = <span>2</span>, <br>                                        fill = <span>'white'</span>, <br>                                        alpha = <span>0.85</span>,<br>                                        family = <span>'Arial'</span>,<br>                                        fontface = <span>'bold'</span>),<br>                      perc.k = <span>2</span>,<br>                      title = <span>''</span>,<br>                      xlab = <span>''</span>,<br>                      legend.title = <span>'Seurat Cluster'</span>,<br>                      ggtheme = ggpubr::theme_pubclean()) +<br>    theme(axis.ticks.x = element_blank(),<br>          axis.ticks.y = element_line(color = <span>'black'</span>, lineend = <span>'round'</span>),<br>          legend.position = <span>'right'</span>,<br>          axis.text.x = element_text(size = <span>15</span>, color = <span>'black'</span>, family = <span>'Arial'</span>),<br>          axis.text.y = element_text(size = <span>15</span>, color = <span>'black'</span>, family = <span>'Arial'</span>),<br>          legend.text = element_text(family = <span>'Arial'</span>, size = <span>10</span>, color = <span>'black'</span>),<br>          legend.title = element_text(family = <span>'Arial'</span>, size = <span>13</span>, color = <span>'black'</span>)) <br>  <br>  <span># (3)去除柱子下面的样本量标识：</span><br>  gginnards::delete_layers(x = figure, match_type = <span>'GeomText'</span>)<br>}<br><br></code></pre><p data-tool="mdnice编辑器">当然了，这个<strong> PropPlot.R  文件里面的 PropPlot 函数是可以自己重新定制化修改的哦！</strong></p><p data-tool="mdnice编辑器">可以看到两个命名体系差不多：</p><pre data-tool="mdnice编辑器"><span data-lazy-bgimg="https://mmbiz.qpic.cn/mmbiz_svg/GqIlejFTbNjwekMlGTsw0WMZDJZiaUUo6xZIGOcPKW2dw163M9ZlvLNFcNkene8KjxEtuIKQUZEd0nyreQQ7BurzPJ1UT9Miaw/640?wx_fmt=svg" data-fail="0"></span><code>&gt; table(sce<span>$celltype</span>,sce<span>$seurat_clusters</span>)<br>              <br>                 0   1   2   3   4   5   6   7   8<br>  Naive CD4 T  163 505   0   0  29   0   0   0   0<br>  Memory CD4 T 444  20   0   0  19   0   0   0   0<br>  CD14+ Mono     0   0 466   0   0  14   0   0   0<br>  B              1   0   0 342   1   0   0   0   0<br>  CD8 T          1   0   0   0 264   0   6   0   0<br>  FCGR3A+ Mono   0   0   0   0   0 162   0   0   0<br>  NK             0   0   0   0   2   0 153   0   0<br>  DC             0   0   0   0   0   0   0  32   0<br>  Platelet       0   0   1   0   0   0   0   0  13<br></code></pre><p data-tool="mdnice编辑器">所以，使用两种情况看比例，也差不多。</p><p data-tool="mdnice编辑器">这样的基础认知，也可以看基础10讲：</p><ul data-tool="mdnice编辑器"><li><section><a href="https://mp.weixin.qq.com/s?__biz=MzI1Njk4ODE0MQ==&amp;mid=2247486076&amp;idx=1&amp;sn=52bb851d7dc23461233a2cf458736151&amp;scene=21#wechat_redirect" data-linktype="2">01. 上游分析流程</a></section></li><li><section><a href="https://mp.weixin.qq.com/s?__biz=MzI1Njk4ODE0MQ==&amp;mid=2247486082&amp;idx=1&amp;sn=03cadceffb2c14ba95d97fe5caf38d94&amp;scene=21#wechat_redirect" data-linktype="2">02.课题多少个样品，测序数据量如何</a></section></li><li><section><a href="https://mp.weixin.qq.com/s?__biz=MzI1Njk4ODE0MQ==&amp;mid=2247486088&amp;idx=1&amp;sn=3a115338ee4937d20caab78627237553&amp;scene=21#wechat_redirect" data-linktype="2">03. 过滤不合格细胞和基因（数据质控很重要）</a></section></li><li><section><a href="https://mp.weixin.qq.com/s?__biz=MzI1Njk4ODE0MQ==&amp;mid=2247486096&amp;idx=1&amp;sn=1a99c4c5800b7e0287db3e8ef369fab8&amp;scene=21#wechat_redirect" data-linktype="2">04. 过滤线粒体核糖体基因</a></section></li><li><section><a href="https://mp.weixin.qq.com/s?__biz=MzI1Njk4ODE0MQ==&amp;mid=2247486098&amp;idx=1&amp;sn=bf9a71df848d74fe665ce7d5e283d5ff&amp;scene=21#wechat_redirect" data-linktype="2">05. 去除细胞效应和基因效应</a></section></li><li><section><a href="https://mp.weixin.qq.com/s?__biz=MzI1Njk4ODE0MQ==&amp;mid=2247486260&amp;idx=1&amp;sn=c6abf658de73594d1d77d8e1ffa7d153&amp;scene=21#wechat_redirect" data-linktype="2">06.单细胞转录组数据的降维聚类分群</a></section></li><li><section><a href="https://mp.weixin.qq.com/s?__biz=MzI1Njk4ODE0MQ==&amp;mid=2247486271&amp;idx=1&amp;sn=638b434b6deee63206af1c0eeda175ab&amp;scene=21#wechat_redirect" data-linktype="2">07.单细胞转录组数据处理之细胞亚群注释</a></section></li><li><section><a href="https://mp.weixin.qq.com/s?__biz=MzI1Njk4ODE0MQ==&amp;mid=2247486278&amp;idx=1&amp;sn=91250ef733833ff00371818b215dc124&amp;scene=21#wechat_redirect" data-linktype="2">08.把拿到的亚群进行更细致的分群</a></section></li><li><section><a href="https://mp.weixin.qq.com/s?__biz=MzI1Njk4ODE0MQ==&amp;mid=2247486287&amp;idx=1&amp;sn=49627c638ff9c04418282c53518aa7c7&amp;scene=21#wechat_redirect" data-linktype="2">09.单细胞转录组数据处理之细胞亚群比例比较</a></section></li></ul><p data-tool="mdnice编辑器">最基础的往往是降维聚类分群，参考前面的例子：<a href="https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247497956&amp;idx=1&amp;sn=5d4deb7cf7b7848b3e2273cbd663bb6a&amp;scene=21#wechat_redirect" data-linktype="2">人人都能学会的单细胞聚类分群注释</a></p><p data-tool="mdnice编辑器"><br></p></section><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com"><h3 data-tool="mdnice编辑器"><span>写在文末</span></h3></section><p>如果你也想做单细胞转录组数据分析，<span>最好是有自己的计算机资源哦，比如我们的</span><a href="https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247528363&amp;idx=1&amp;sn=5e02f3e9b2e148191e23ebc2c0d780e7&amp;scene=21#wechat_redirect" data-linktype="2">2024的共享服务器交个朋友福利价仍然是800</a><span>，而且还需要有基本的生物信息学基础，也可以看看我们的</span><a target="_blank" href="http://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247531929&amp;idx=1&amp;sn=f6f16b7bf6b907360d6d0052e3d10cf6&amp;chksm=9b4b3d22ac3cb434b6aa7753a4cf0f266578147ccf10b49cc834e46af578ee6de99be0accb30&amp;scene=21#wechat_redirect" textvalue="生物信息学马拉‍松授课（买一得五）" linktype="text" imgurl="" imgdata="null" data-itemshowtype="0" tab="innerlink" data-linktype="2" hasload="1">生物信息学马拉松授课（买一得五）</a><span> ，你的生物信息学入门课。 </span></p><p><mp-style-type data-value="3"></mp-style-type></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/OflFEujZ8x_qVKAoW3xVsQ",target="_blank" rel="noopener noreferrer">原文链接</a>
