---
title: "CIBERSORT用腻了吗？重磅神器DeSide用深度学习反卷积肿瘤微环境，一文详解教学（上）！"
date: 2025-03-22T08:35:55Z
draft: ["false"]
tags: [
  "fetched",
  "生信作曲家"
]
categories: ["Acdemic"]
---
CIBERSORT用腻了吗？重磅神器DeSide用深度学习反卷积肿瘤微环境，一文详解教学（上）！ by 生信作曲家
------
<div><section><section><span leaf=""><br></span></section><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/mo60jlFOtaA3M2Vga9UHKNuDoYcQb3VWFyH2ZBshkIm9x0icryP7PkAC6OMSr4Fy5vwLic6ECQVSSEe6zgHPqAuA/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.398640543782487" data-s="300,640" data-type="png" data-w="2501" type="block" data-imgfileid="100007210" src="https://mmbiz.qpic.cn/mmbiz_png/mo60jlFOtaA3M2Vga9UHKNuDoYcQb3VWFyH2ZBshkIm9x0icryP7PkAC6OMSr4Fy5vwLic6ECQVSSEe6zgHPqAuA/640?wx_fmt=png&amp;from=appmsg"></span></section><section><section><p><span><strong><span><span leaf=""><br></span></span></strong></span></p><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com" data-pm-slice="0 0 []"><h3 data-tool="mdnice编辑器"><span></span><span><span leaf="">前言</span></span></h3><h3 data-tool="mdnice编辑器"><span leaf="">关于Bulk反卷积的方法想必各位同学都比较熟悉了，主要有CIBERSORT、MCPcounter、Timer等等。</span></h3><h3 data-tool="mdnice编辑器"><span leaf="">今天，我们的主角DeSide，<span textstyle="">利用深度学习的方法来反卷积免疫细胞、肿瘤细胞比例，其具备很好的创新、高端和真实的优势。</span></span><span></span></h3><p data-tool="mdnice编辑器"><span leaf="">DeSide 是一种基于深度学习和单细胞解析的bulk肿瘤去卷积方法，可用于从bulk RNA-seq数据推测不同细胞类型的比例。</span></p><p data-tool="mdnice编辑器"><span leaf="" data-pm-slice="0 0 []">技术于2024年发表在PNAS上。<img data-src="https://mmbiz.qpic.cn/mmbiz_png/mo60jlFOtaCl6Kx7iaODqEB8Wd0O6fnem6YQqTG7fNKkJxKibnmRqsx0A9YFXMSkicJloGpibEf4Ls09v7ArVuCq1g/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.4020066889632107" data-type="png" data-w="1495" data-imgfileid="100007296" src="https://mmbiz.qpic.cn/mmbiz_png/mo60jlFOtaCl6Kx7iaODqEB8Wd0O6fnem6YQqTG7fNKkJxKibnmRqsx0A9YFXMSkicJloGpibEf4Ls09v7ArVuCq1g/640?wx_fmt=png&amp;from=appmsg"></span></p><p data-tool="mdnice编辑器"><span leaf=""><img data-imgfileid="100007295" data-ratio="0.4038573933372297" data-src="https://mmbiz.qpic.cn/mmbiz_png/mo60jlFOtaCl6Kx7iaODqEB8Wd0O6fnemAXaicqpQpriaNquHXo6ckQBc0diccAaSh7asumKrJEfulTcj5rVSmUofA/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1711" src="https://mmbiz.qpic.cn/mmbiz_png/mo60jlFOtaCl6Kx7iaODqEB8Wd0O6fnemAXaicqpQpriaNquHXo6ckQBc0diccAaSh7asumKrJEfulTcj5rVSmUofA/640?wx_fmt=png&amp;from=appmsg"></span></p><ul><li><section><span leaf=""><span textstyle="">技术亮点：</span></span></section></li></ul><p data-tool="mdnice编辑器"><span leaf="">DeSide 的深度神经网络（DNN）架构中，创新性地融合了两个全连接网络：pathway 网络和 GEP 网络。</span><strong><span leaf=""><span textstyle="">Pathway 网络从生物信号通路中提取特征</span></span></strong><span leaf="">（这是很多现有算法没有的），而 GEP 网络则关注基因表达谱信息。特别是，</span><strong><span leaf=""><span textstyle="">pathway（KEGG/Reactome）网络通过引入更宏观的特征维度</span></span></strong><span leaf="">，增强了输入数据的多样性，从而提升模型的泛化能力。</span></p><p data-tool="mdnice编辑器"><span leaf="">考虑到不同癌症类型的肿瘤细胞基因表达存在较大差异，DeSide 采用 sigmoid 作为 DNN 输出层的激活函数，使预测值的总和限定在 [0,1] 之间。</span><strong><span leaf="">模型首先估算免疫细胞等非癌细胞的比例</span></strong><span leaf="">，</span><strong><span leaf="">再通过 1 减去这些细胞的占比来推算肿瘤细胞的比例</span></strong><span leaf="">。这种间接计算方法</span><strong><span leaf="">减少了直接预测肿瘤细胞比例可能带来的误差</span></strong><span leaf="">。</span></p><p data-tool="mdnice编辑器"><span leaf="">小编值得一提的是，在利用深度学习解析细胞组成的研究中，</span><strong><span leaf="">DeSide 是首个采用这一策略的算法。</span></strong></p><p data-tool="mdnice编辑器"><span leaf=""><span textstyle="">简言之，DeSide对于肿瘤细胞、免疫细胞的比例推断都很理想，不仅考虑基因表达，还考虑通路富集！</span></span></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/mo60jlFOtaCl6Kx7iaODqEB8Wd0O6fnemqsZu1mH89aQQIk6Ec6OkEhpQXADnSfhiaSMXF4Oq2KB4QorNWuXSywg/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.6578171091445427" data-type="png" data-w="1695" data-imgfileid="100007298" src="https://mmbiz.qpic.cn/mmbiz_png/mo60jlFOtaCl6Kx7iaODqEB8Wd0O6fnemqsZu1mH89aQQIk6Ec6OkEhpQXADnSfhiaSMXF4Oq2KB4QorNWuXSywg/640?wx_fmt=png&amp;from=appmsg"></span></figure></section><section nodeleaf=""><img data-imgfileid="100007316" data-ratio="0.6770120988953182" data-s="300,640" data-src="https://mmbiz.qpic.cn/mmbiz_png/mo60jlFOtaCl6Kx7iaODqEB8Wd0O6fnemaxtrFrFLKoMJvdrwZv78jTEXgaCoLoDX85icAJBpHjwO9RqD8bbQc5A/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1901" type="block" src="https://mmbiz.qpic.cn/mmbiz_png/mo60jlFOtaCl6Kx7iaODqEB8Wd0O6fnemaxtrFrFLKoMJvdrwZv78jTEXgaCoLoDX85icAJBpHjwO9RqD8bbQc5A/640?wx_fmt=png&amp;from=appmsg"></section><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com" data-pm-slice="0 0 []"><figure data-tool="mdnice编辑器"><span leaf="">以上是DeSide 的细胞类型</span></figure><h3 data-tool="mdnice编辑器"><span></span><span><span leaf="">代码实操</span></span><span></span></h3><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">#</span></span><span><span><span leaf="">## 先在终端配置好环境</span></span></span><span leaf=""><br></span><span leaf="">conda create -n deside python=3.8</span><span leaf=""><br></span><span leaf="">conda activate deside</span><span leaf=""><br></span><span leaf="">pip install deside -i https://pypi.tuna.tsinghua.edu.cn/simple</span><span leaf=""><br></span><span leaf="">pip install ipykernel -i https://pypi.tuna.tsinghua.edu.cn/simple</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/mo60jlFOtaCl6Kx7iaODqEB8Wd0O6fnem0mCvPaKTjyTzsoOTq1ia4qA7J6gicOniaiag9ZKVGUc44mtb0nWekmMLpg/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.3066966797974114" data-type="png" data-w="1777" data-imgfileid="100007297" src="https://mmbiz.qpic.cn/mmbiz_png/mo60jlFOtaCl6Kx7iaODqEB8Wd0O6fnem0mCvPaKTjyTzsoOTq1ia4qA7J6gicOniaiag9ZKVGUc44mtb0nWekmMLpg/640?wx_fmt=png&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">接下来我们打开python来操作了，此处用jupyterlab演示，pycharm，vscode等均同理。只不过注意python虚拟环境切换即可。</span></p><h4 data-tool="mdnice编辑器"><span></span><span><span leaf="">注意事项</span></span><span></span></h4><p data-tool="mdnice编辑器"><span leaf="">有两种方法可以使用 DeSide。1.可以使用提供的</span><strong><span leaf=""><span textstyle="">预训练模型</span></span></strong><span leaf="">直接预测细胞比例，无需自己训练模型。2.使用自己的模型从单细胞数据开始训练。需要执行一系列步骤，包括数据预处理、训练模型以及使用训练好的模型来预测细胞比例。</span></p><p data-tool="mdnice编辑器"><span leaf=""><span textstyle="">本期我们主要教学预训练模型</span></span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">import</span></span><span leaf=""> os</span><span leaf=""><br></span><span><span leaf="">import</span></span><span leaf=""> pandas </span><span><span leaf="">as</span></span><span leaf=""> pd</span><span leaf=""><br></span><span><span leaf="">from</span></span><span leaf=""> IPython.display </span><span><span leaf="">import</span></span><span leaf=""> Image</span><span leaf=""><br></span><span><span leaf="">from</span></span><span leaf=""> deside.utility </span><span><span leaf="">import</span></span><span leaf=""> check_dir</span><span leaf=""><br></span><span><span leaf="">from</span></span><span leaf=""> deside.decon_cf </span><span><span leaf="">import</span></span><span leaf=""> DeSide</span><span leaf=""><br></span><span><span leaf="">from</span></span><span leaf=""> deside.plot </span><span><span leaf="">import</span></span><span leaf=""> plot_predicted_result</span><span leaf=""><br></span><span><span leaf="">from</span></span><span leaf=""> deside.utility.read_file </span><span><span leaf="">import</span></span><span leaf=""> read_gene_set</span><span leaf=""><br></span><span><span leaf="">import</span></span><span leaf=""> deside</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/mo60jlFOtaCl6Kx7iaODqEB8Wd0O6fnemlvGa4MibKKP8nTy1YGiaVw1E2CnzmtS6oy1OicIOH42gPfEUDvnacBgow/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.571076417419885" data-type="png" data-w="1217" data-imgfileid="100007299" src="https://mmbiz.qpic.cn/mmbiz_png/mo60jlFOtaCl6Kx7iaODqEB8Wd0O6fnemlvGa4MibKKP8nTy1YGiaVw1E2CnzmtS6oy1OicIOH42gPfEUDvnacBgow/640?wx_fmt=png&amp;from=appmsg"></span></figure><ul><li><section><span leaf="">准备数据 创建datasets文件夹：</span></section></li></ul><figure data-tool="mdnice编辑器"><span leaf=""><img data-imgfileid="100007301" data-ratio="0.6449912126537786" data-src="https://mmbiz.qpic.cn/mmbiz_png/mo60jlFOtaCl6Kx7iaODqEB8Wd0O6fnemqvIxY7QYTcicVqVymYfp0PDCibS5afVLXqkXfOCdiaRcv2UB7tkuWly3A/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="569" src="https://mmbiz.qpic.cn/mmbiz_png/mo60jlFOtaCl6Kx7iaODqEB8Wd0O6fnemqvIxY7QYTcicVqVymYfp0PDCibS5afVLXqkXfOCdiaRcv2UB7tkuWly3A/640?wx_fmt=png&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">准备好bulk数据表，csv格式如下，第一列gene_name，后面为样本<img data-src="https://mmbiz.qpic.cn/mmbiz_png/mo60jlFOtaCl6Kx7iaODqEB8Wd0O6fnemGlFyFIPo7J915n2ib8dPlXm7XYKSFb3wvtay2Uvbm1uKibrqJhjhpZ2g/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.5752490823282643" data-type="png" data-w="1907" data-imgfileid="100007304" src="https://mmbiz.qpic.cn/mmbiz_png/mo60jlFOtaCl6Kx7iaODqEB8Wd0O6fnemGlFyFIPo7J915n2ib8dPlXm7XYKSFb3wvtay2Uvbm1uKibrqJhjhpZ2g/640?wx_fmt=png&amp;from=appmsg"></span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">dataset_dir = </span><span><span leaf="">'./datasets/'</span></span><span leaf=""><br></span><span><span leaf=""># bulk gene expression profiles (GEPs) in TPM formart</span></span><span leaf=""><br></span><span leaf="">bulk_tpm_file_path = os.path.join(dataset_dir, </span><span><span leaf="">'LUAD_TPM.csv'</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">bulk_tpm = pd.read_csv(bulk_tpm_file_path, index_col=</span><span><span leaf="">0</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">print(bulk_tpm.shape)</span><span leaf=""><br></span><span leaf="">bulk_tpm.head(</span><span><span leaf="">2</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/mo60jlFOtaCl6Kx7iaODqEB8Wd0O6fnemsV9WE20dfopYrvn6y6AL4DU71zXdwvlo7y6geljJVOqiavyxttcIZ7w/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.2364502680166766" data-type="png" data-w="1679" data-imgfileid="100007303" src="https://mmbiz.qpic.cn/mmbiz_png/mo60jlFOtaCl6Kx7iaODqEB8Wd0O6fnemsV9WE20dfopYrvn6y6AL4DU71zXdwvlo7y6geljJVOqiavyxttcIZ7w/640?wx_fmt=png&amp;from=appmsg"></span></figure><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># create output directory创建输出目录</span></span><span leaf=""><br></span><span leaf="">result_dir = </span><span><span leaf="">'./results'</span></span><span leaf=""><br></span><span leaf="">y_pred_file_path = os.path.join(result_dir, </span><span><span leaf="">'y_pred.csv'</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">check_dir(result_dir)</span><span leaf=""><br></span><span leaf=""><br></span></code></pre><ul><li><section><span leaf="">继续数据准备：kegg,reactome通路表 准备好后则运行下方，也可以找管理员要我们准备好的。</span></section></li></ul><figure data-tool="mdnice编辑器"><span leaf=""><img data-imgfileid="100007302" data-ratio="0.47110675808031344" data-src="https://mmbiz.qpic.cn/mmbiz_png/mo60jlFOtaCl6Kx7iaODqEB8Wd0O6fnemdrRl7SrSbkA8SvFtFaBRVoKibtqsLDdPyMbhk4S1aKIyd2nDVk7rsuQ/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1021" src="https://mmbiz.qpic.cn/mmbiz_png/mo60jlFOtaCl6Kx7iaODqEB8Wd0O6fnemdrRl7SrSbkA8SvFtFaBRVoKibtqsLDdPyMbhk4S1aKIyd2nDVk7rsuQ/640?wx_fmt=png&amp;from=appmsg"></span></figure><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf=""><br></span><span><span leaf="">## 设定深度学习运行参数</span></span><span leaf=""><br></span><span><span leaf=""># hyper-parameters of the DNN model</span></span><span leaf=""><br></span><span leaf="">deside_parameters = {</span><span leaf=""><br></span><span leaf="">    </span><span><span leaf="">'architecture'</span></span><span leaf="">: ([</span><span><span leaf="">200</span></span><span leaf="">, </span><span><span leaf="">2000</span></span><span leaf="">, </span><span><span leaf="">2000</span></span><span leaf="">, </span><span><span leaf="">2000</span></span><span leaf="">, </span><span><span leaf="">50</span></span><span leaf="">], [</span><span><span leaf="">0.05</span></span><span leaf="">, </span><span><span leaf="">0.05</span></span><span leaf="">, </span><span><span leaf="">0.05</span></span><span leaf="">, </span><span><span leaf="">0.2</span></span><span leaf="">, </span><span><span leaf="">0</span></span><span leaf="">]),</span><span leaf=""><br></span><span leaf="">    </span><span><span leaf="">'architecture_for_pathway_network'</span></span><span leaf="">: ([</span><span><span leaf="">50</span></span><span leaf="">, </span><span><span leaf="">500</span></span><span leaf="">, </span><span><span leaf="">500</span></span><span leaf="">, </span><span><span leaf="">500</span></span><span leaf="">, </span><span><span leaf="">50</span></span><span leaf="">], [</span><span><span leaf="">0</span></span><span leaf="">, </span><span><span leaf="">0</span></span><span leaf="">, </span><span><span leaf="">0</span></span><span leaf="">, </span><span><span leaf="">0</span></span><span leaf="">, </span><span><span leaf="">0</span></span><span leaf="">]),</span><span leaf=""><br></span><span leaf="">    </span><span><span leaf="">'loss_function_alpha'</span></span><span leaf="">: </span><span><span leaf="">0.5</span></span><span leaf="">,  </span><span><span leaf=""># alpha*mae + (1-alpha)*rmse, mae means mean absolute error</span></span><span leaf=""><br></span><span leaf="">    </span><span><span leaf="">'normalization'</span></span><span leaf="">: </span><span><span leaf="">'layer_normalization'</span></span><span leaf="">,  </span><span><span leaf=""># batch_normalization / layer_normalization / None</span></span><span leaf=""><br></span><span leaf="">     </span><span><span leaf=""># 1 means to add a normalization layer, input | the first hidden layer | ... | output</span></span><span leaf=""><br></span><span leaf="">    </span><span><span leaf="">'normalization_layer'</span></span><span leaf="">: [</span><span><span leaf="">0</span></span><span leaf="">, </span><span><span leaf="">0</span></span><span leaf="">, </span><span><span leaf="">1</span></span><span leaf="">, </span><span><span leaf="">1</span></span><span leaf="">, </span><span><span leaf="">1</span></span><span leaf="">, </span><span><span leaf="">1</span></span><span leaf="">],  </span><span><span leaf=""># 1 more parameter than the number of hidden layers</span></span><span leaf=""><br></span><span leaf="">    </span><span><span leaf="">'pathway_network'</span></span><span leaf="">: </span><span><span leaf="">True</span></span><span leaf="">,  </span><span><span leaf=""># using an independent pathway network</span></span><span leaf=""><br></span><span leaf="">    </span><span><span leaf="">'last_layer_activation'</span></span><span leaf="">: </span><span><span leaf="">'sigmoid'</span></span><span leaf="">,  </span><span><span leaf=""># sigmoid / softmax</span></span><span leaf=""><br></span><span leaf="">    </span><span><span leaf="">'learning_rate'</span></span><span leaf="">: </span><span><span leaf="">1e-4</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">    </span><span><span leaf="">'batch_size'</span></span><span leaf="">: </span><span><span leaf="">128</span></span><span leaf="">}</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># read two gene sets as pathway mask</span></span><span leaf=""><br></span><span leaf="">gene_set_file_path1 = os.path.join(dataset_dir, </span><span><span leaf="">'gene_set'</span></span><span leaf="">, </span><span><span leaf="">'c2.cp.kegg.v2022.1.Hs.symbols.gmt'</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">gene_set_file_path2 = os.path.join(dataset_dir, </span><span><span leaf="">'gene_set'</span></span><span leaf="">, </span><span><span leaf="">'c2.cp.reactome.v2023.1.Hs.symbols.gmt'</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">all_pathway_files = [gene_set_file_path1, gene_set_file_path2]</span><span leaf=""><br></span><span leaf="">pathway_mask = read_gene_set(all_pathway_files)  </span><span><span leaf=""># genes by pathways</span></span><span leaf=""><br></span></code></pre><ul><li><section><span leaf="">关键的一步：开始预测</span></section></li></ul><section data-pm-slice="0 0 []"><span leaf=""><br></span><span leaf="" data-pm-slice="0 0 []">运行结果参考：<span textstyle="">不仅有肿瘤细胞的推断，还有其他热门的CD8(GZMB high)、区分CAF和肌成纤维等！ </span>当然结果也很合理，肿瘤细胞占据60%左右的比例，这个和真实情况也很接近！<span textstyle="">这方法的优势立刻凸显</span></span></section><section><span leaf=""><br></span></section><section><span leaf=""><img data-imgfileid="100007308" data-ratio="0.2723239046034387" data-src="https://mmbiz.qpic.cn/mmbiz_png/mo60jlFOtaCl6Kx7iaODqEB8Wd0O6fnem06GvEzViba7loW1C68xibne5hiauuq3J02WUpmiauX6Ue3w3yB3QOotf5w/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1803" src="https://mmbiz.qpic.cn/mmbiz_png/mo60jlFOtaCl6Kx7iaODqEB8Wd0O6fnem06GvEzViba7loW1C68xibne5hiauuq3J02WUpmiauX6Ue3w3yB3QOotf5w/640?wx_fmt=png&amp;from=appmsg"></span></section><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/mo60jlFOtaCl6Kx7iaODqEB8Wd0O6fnemKqiazLdRzTsibQ8QltibR7fhMsRqcjicSiadGZ96xhxzwiau1LXPkn4uytJQ/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.7069684335914235" data-type="png" data-w="1679" data-imgfileid="100007309" src="https://mmbiz.qpic.cn/mmbiz_png/mo60jlFOtaCl6Kx7iaODqEB8Wd0O6fnemKqiazLdRzTsibQ8QltibR7fhMsRqcjicSiadGZ96xhxzwiau1LXPkn4uytJQ/640?wx_fmt=png&amp;from=appmsg"></span></figure></section><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com" data-pm-slice="0 0 []"><figure data-tool="mdnice编辑器"><span leaf=""><br></span></figure></section><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com" data-pm-slice="0 0 []"><figure data-tool="mdnice编辑器"><span leaf="">购买后添加管理员获取示例数据、必备文件</span></figure><section nodeleaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_jpg/mo60jlFOtaBGpGicRlf1uIRz5z4icMGickywJW7jQIkU4lafb2SfXxTwtD6ia0Twl7EePYmgFGpYULxQ1Ym9GVbKibw/640?wx_fmt=jpeg" data-ratio="1" data-s="300,640" data-type="jpeg" data-w="512" type="block" data-imgfileid="100001321" src="https://mmbiz.qpic.cn/mmbiz_jpg/mo60jlFOtaBGpGicRlf1uIRz5z4icMGickywJW7jQIkU4lafb2SfXxTwtD6ia0Twl7EePYmgFGpYULxQ1Ym9GVbKibw/640?wx_fmt=jpeg"></section><p data-tool="mdnice编辑器"><span leaf=""><br></span></p><section><span leaf=""><br></span></section><p><mp-pay-preview-filter data-offset="45"></mp-pay-preview-filter></p></section></section></section></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/1z1FS27oJqgzTsef5xHTew",target="_blank" rel="noopener noreferrer">原文链接</a>
