---
title: "sclet：新手入门R单细胞分析的正确打开方式？"
date: 2025-03-24T00:52:41Z
draft: ["false"]
tags: [
  "fetched",
  "YuLabSMU"
]
categories: ["Acdemic"]
---
sclet：新手入门R单细胞分析的正确打开方式？ by YuLabSMU
------
<div><section><a href="https://mp.weixin.qq.com/s?__biz=MzI5NjUyNzkxMg==&amp;mid=2247495781&amp;idx=1&amp;sn=cba2f100c93deb1e935abd8f560407b2&amp;scene=21#wechat_redirect" imgurl="https://mmbiz.qpic.cn/mmbiz_png/MPBFtnFrw4mlkpbzzuuiaXoRnKMqTYpANKJMr6LVSYqKZicP7u0HD7EjIicYkXfGOVUxwhRWetQMDI9fVvvAfv0HA/640?wx_fmt=png&amp;from=appmsg" linktype="image" tab="innerlink" data-itemshowtype="0" target="_blank" data-linktype="1"><span><img data-imgfileid="100013255" data-ratio="0.8087591240875912" data-s="300,640" data-type="png" data-w="685" type="block" data-src="https://mmbiz.qpic.cn/mmbiz_png/MPBFtnFrw4mlkpbzzuuiaXoRnKMqTYpANKJMr6LVSYqKZicP7u0HD7EjIicYkXfGOVUxwhRWetQMDI9fVvvAfv0HA/640?wx_fmt=png" src="https://mmbiz.qpic.cn/mmbiz_png/MPBFtnFrw4mlkpbzzuuiaXoRnKMqTYpANKJMr6LVSYqKZicP7u0HD7EjIicYkXfGOVUxwhRWetQMDI9fVvvAfv0HA/640?wx_fmt=png"></span></a></section><section><span leaf="">当个老师不容易，因为我要讲课，于是我发出了这样的灵魂拷问。</span><p data-tool="mdnice编辑器"><span leaf="">评论区非常热闹，然后我推荐大家可以阅读一下评论区。比如底下截图的这个，就很好地说明了现状：</span></p><p data-tool="mdnice编辑器"><span leaf=""><img data-imgfileid="100013257" data-ratio="0.3972222222222222" data-s="300,640" data-type="png" data-w="1080" type="block" data-src="https://mmbiz.qpic.cn/mmbiz_png/MPBFtnFrw4mlkpbzzuuiaXoRnKMqTYpANnNaJtGKnGibFAnW5LSghnj3FLPIBZlnWWADwkZgoeXm9Od1uW5fBlcg/640?wx_fmt=png" src="https://mmbiz.qpic.cn/mmbiz_png/MPBFtnFrw4mlkpbzzuuiaXoRnKMqTYpANnNaJtGKnGibFAnW5LSghnj3FLPIBZlnWWADwkZgoeXm9Od1uW5fBlcg/640?wx_fmt=png"></span></p><p data-tool="mdnice编辑器"><span leaf="">Seurat大家喜欢，是因为一个没搞过单细胞的新手，他捣鼓一下午，就觉得自己能搞单细胞了，他完全可以把seurat的那个pbmc的例子跑下来，比如一个本科生，他的简历就可以写自己跑过单细胞流程了（此处似乎指导了马上研究生复试的学生们）。门槛很重要，seurat胜出不是没道理。</span></p><p data-tool="mdnice编辑器"><span leaf="">然而为什么不要seurat，上一个评论也是说得很在理。python生态就是抄R的呀，你们还天天吵什么到底是R还是python，数据结构非常重要，而python的AnnData与R的SingleCellExperiment是一模一样的，你要是用SingleCellExperiment，要切换成python其实是无缝的。</span></p><p data-tool="mdnice编辑器"><span leaf="">抛开两语言切换的问题，就Seurat和SingleCellExperiment比，为什么要SingleCellExperiment？因为Seurat的文档虽然傻瓜，但他从来就没有告诉过你数据结构是怎么考虑的？是怎样定义的？然后版本升级，还会改数据结构，然后他也不告诉你。这就是问题，反观SingleCellExperiment，定义得清清楚楚，简简单单。</span></p><p data-tool="mdnice编辑器"><span leaf="">而为什么这个很重要？你可能会说:我反正是当个调包侠，跑完拍拍屁股收工。<span>作为一个调包侠，这样子是当不好调包侠的！</span><span data-change-font="1">（类比一下，想着60分万岁，总考50多分很闹心。想着混毕业，你懂的）</span>你总有需要倒腾点数据的时候，完全活在别人圈死的小圈子里，就是要简单地倒腾点数据，也很容易流下没有技术的眼泪。</span></p><p data-tool="mdnice编辑器"><span leaf="">特别是我们作为生物信息学专业，莫说倒腾点数据，自己写点小功能总是要的吧？一个白盒子的SingleCellExperiement，稳定（不会老变）、清楚（有完善的文档描述）还设计良好的数据结构太重要了，Seurat虽然代码是公开的，但会变又没文档，只能算个灰盒子。基于SingleCellExperiement来做事情，显然是事半功倍，舒服多了。于是我跟学生们说，还是换到Bioconductor生态吧，诸多好处，就是以后换python了，也好切换。</span></p><p data-tool="mdnice编辑器"><span leaf="">然后我知道，大家都会路径依赖！这事很难！</span></p><p data-tool="mdnice编辑器"><span leaf="">再然后，我就在想，要不我基于SingleCellExperiment把Seurat那个pbmc的例子，完全复刻一下，不就是用Seurat半天就以为自己会单细胞了吗？那我让你同样的代码，用SingleCellExperiment来做，你也是半天学会，这不就完事了嘛。</span></p><p data-tool="mdnice编辑器"><span leaf=""><img data-imgfileid="100013258" data-ratio="1.0018518518518518" data-s="300,640" data-type="png" data-w="1080" type="block" data-src="https://mmbiz.qpic.cn/mmbiz_png/MPBFtnFrw4mlkpbzzuuiaXoRnKMqTYpAN5MS9f8zKO1iaaibnuLjiaPggjYjqqSWRsaHHoWgIrNwQewPxMiaA9t9MMg/640?wx_fmt=png" src="https://mmbiz.qpic.cn/mmbiz_png/MPBFtnFrw4mlkpbzzuuiaXoRnKMqTYpAN5MS9f8zKO1iaaibnuLjiaPggjYjqqSWRsaHHoWgIrNwQewPxMiaA9t9MMg/640?wx_fmt=png"></span></p><p data-tool="mdnice编辑器"><span leaf="">弄好了这层皮，我在2024年的10月1号把它推送到github，包的名字是sclet，sc当然是单细胞，let是小的意思嘛。意指这是个麻雀包，后面我搜一下，发现还是一个酒。</span></p><p data-tool="mdnice编辑器"><span leaf=""><img data-imgfileid="100013259" data-ratio="0.9704579025110783" data-s="300,640" data-type="png" data-w="677" type="block" data-src="https://mmbiz.qpic.cn/mmbiz_png/MPBFtnFrw4mlkpbzzuuiaXoRnKMqTYpAN7bxyXoFrm2o9fdlKVyH4D7edDnLCrwXF06SbV12fQ08iaH9bkkxVd8g/640?wx_fmt=png" src="https://mmbiz.qpic.cn/mmbiz_png/MPBFtnFrw4mlkpbzzuuiaXoRnKMqTYpAN7bxyXoFrm2o9fdlKVyH4D7edDnLCrwXF06SbV12fQ08iaH9bkkxVd8g/640?wx_fmt=png"></span></p><p data-tool="mdnice编辑器"><span leaf="">然后Description里我直接就写着，帮助学生从Seurat迁移到SingleCellExperiment，然后我就在假期之后的组会上演示了一遍，当老师真是操碎了心！</span></p><p data-tool="mdnice编辑器"><span leaf="">请欣赏余老师缝的虎皮裙子（</span><a href="https://mp.weixin.qq.com/s?__biz=MzI5NjUyNzkxMg==&amp;mid=2247489820&amp;idx=1&amp;sn=f5c9c793cc057093981453f01540176d&amp;scene=21#wechat_redirect"><span leaf="">梗点这里</span></a><span leaf="">）</span></p><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">读数据</span></span><span></span></h2><p data-tool="mdnice编辑器"><span leaf="">出来当然是个</span><code><span leaf="">SingleCellExperiment</span></code><span leaf="">的对象。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">library</span></span><span leaf="">(sclet)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">dir &lt;- </span><span><span leaf="">"filtered_gene_bc_matrices/hg19"</span></span><span leaf=""><br></span><span leaf="">pbmc &lt;- Read10X(dir)</span><span leaf=""><br></span></code></pre><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">质控</span></span><span></span></h2><p data-tool="mdnice编辑器"><span leaf="">注释的写法是</span><code><span leaf="">SingleCellExperiment</span></code><span leaf="">的传统写法，但我们可以用Seurat的写法了。函数名我都给你搞成一模一样的，就是要做到真假Seurat，无从分辨。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># colData(pbmc)$percent.mt &lt;- PercentageFeatureSet(pbmc, "^MT-")</span></span><span leaf=""><br></span><span leaf="">pbmc[[</span><span><span leaf="">"percent.mt"</span></span><span leaf="">]] &lt;- PercentageFeatureSet(pbmc, </span><span><span leaf="">"^MT-"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">VlnPlot(pbmc, features = c(</span><span><span leaf="">"nFeature_RNA"</span></span><span leaf="">, </span><span><span leaf="">"nCount_RNA"</span></span><span leaf="">, </span><span><span leaf="">"percent.mt"</span></span><span leaf="">))</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf=""><img data-imgfileid="100013303" data-ratio="0.75" data-type="png" data-w="1080" data-src="https://mmbiz.qpic.cn/mmbiz_png/MPBFtnFrw4mlkpbzzuuiaXoRnKMqTYpANc0qBOW1Kwl5vsuibFr82VFibv4ZFh2PmnTZRvOE0nk9Yxriaic8jBZCCuQ/640?wx_fmt=png" src="https://mmbiz.qpic.cn/mmbiz_png/MPBFtnFrw4mlkpbzzuuiaXoRnKMqTYpANc0qBOW1Kwl5vsuibFr82VFibv4ZFh2PmnTZRvOE0nk9Yxriaic8jBZCCuQ/640?wx_fmt=png"></span></p><p data-tool="mdnice编辑器"><span leaf="">过滤细胞可以先看一眼。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">subset_feature(pbmc, </span><span><span leaf="">10</span></span><span leaf="">)</span><span leaf=""><br></span><span><span leaf="">## ************************************************** &gt;=0 | 32738 (100%)</span></span><span leaf=""><br></span><span><span leaf="">## ************************* &gt;=1 | 16634 (51%)</span></span><span leaf=""><br></span><span><span leaf="">## ********************** &gt;=2 | 14702 (45%)</span></span><span leaf=""><br></span><span><span leaf="">## ********************* &gt;=3 | 13714 (42%)</span></span><span leaf=""><br></span><span><span leaf="">## ******************** &gt;=4 | 13034 (40%)</span></span><span leaf=""><br></span><span><span leaf="">## ******************* &gt;=5 | 12572 (38%)</span></span><span leaf=""><br></span><span><span leaf="">## ******************* &gt;=6 | 12198 (37%)</span></span><span leaf=""><br></span><span><span leaf="">## ****************** &gt;=7 | 11876 (36%)</span></span><span leaf=""><br></span><span><span leaf="">## ****************** &gt;=8 | 11627 (36%)</span></span><span leaf=""><br></span><span><span leaf="">## ***************** &gt;=9 | 11379 (35%)</span></span><span leaf=""><br></span><span><span leaf="">## ***************** &gt;=10 | 11139 (34%)</span></span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">pbmc &lt;- subset_feature(pbmc, mincell = </span><span><span leaf="">3</span></span><span leaf="">, peek=</span><span><span leaf="">FALSE</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">pbmc &lt;- subset(pbmc, subset = nFeature_RNA &gt; </span><span><span leaf="">200</span></span><span leaf=""> &amp; nFeature_RNA &lt; </span><span><span leaf="">2500</span></span><span leaf=""> &amp; percent.mt &lt; </span><span><span leaf="">5</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">pbmc</span><span leaf=""><br></span><span><span leaf="">## class: SingleCellExperiment </span></span><span leaf=""><br></span><span><span leaf="">## dim: 13714 2638 </span></span><span leaf=""><br></span><span><span leaf="">## metadata(1): Samples</span></span><span leaf=""><br></span><span><span leaf="">## assays(1): counts</span></span><span leaf=""><br></span><span><span leaf="">## rownames(13714): AL627309.1 AP006222.2 ...</span></span><span leaf=""><br></span><span><span leaf="">##   PNRC2_ENSG00000215700 SRSF10_ENSG00000215699</span></span><span leaf=""><br></span><span><span leaf="">## rowData names(2): ID Symbol</span></span><span leaf=""><br></span><span><span leaf="">## colnames: NULL</span></span><span leaf=""><br></span><span><span leaf="">## colData names(5): Sample Barcode nFeature_RNA</span></span><span leaf=""><br></span><span><span leaf="">##   nCount_RNA percent.mt</span></span><span leaf=""><br></span><span><span leaf="">## reducedDimNames(0):</span></span><span leaf=""><br></span><span><span leaf="">## mainExpName: NULL</span></span><span leaf=""><br></span><span><span leaf="">## altExpNames(0):</span></span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">然后画图，也给你画个一样的：</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">plot1 &lt;- FeatureScatter(pbmc, feature1 = </span><span><span leaf="">"nCount_RNA"</span></span><span leaf="">, feature2 = </span><span><span leaf="">"percent.mt"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">plot2 &lt;- FeatureScatter(pbmc, feature1 = </span><span><span leaf="">"nCount_RNA"</span></span><span leaf="">, feature2 = </span><span><span leaf="">"nFeature_RNA"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(aplot)</span><span leaf=""><br></span><span leaf="">plot_list(plot1, plot2)</span><span leaf=""><br></span></code></pre><h2 data-tool="mdnice编辑器"><span></span><span><span leaf=""><img data-ratio="0.75" data-type="png" data-w="1080" data-imgfileid="100013304" data-src="https://mmbiz.qpic.cn/mmbiz_png/MPBFtnFrw4mlkpbzzuuiaXoRnKMqTYpANwV2fkGrhUJYLM8TtkE5fe3d8cn4zHD0O4azFIhxxYYtPkicyZuibwedg/640?wx_fmt=png" src="https://mmbiz.qpic.cn/mmbiz_png/MPBFtnFrw4mlkpbzzuuiaXoRnKMqTYpANwV2fkGrhUJYLM8TtkE5fe3d8cn4zHD0O4azFIhxxYYtPkicyZuibwedg/640?wx_fmt=png"></span></span></h2><h2 data-tool="mdnice编辑器"><span><span leaf="">高可变基因</span></span><span></span></h2><p data-tool="mdnice编辑器"><code><span leaf="">SingleCellExperiment</span></code><span leaf="">生态鉴定高可变基因的方法与Seurat不一样，这会影响后面的分析，然后我要复刻嘛，于是我把Seurat的方法也给调用一下。</span></p><p data-tool="mdnice编辑器"><span leaf="">这里</span><code><span leaf="">FindVariableFeatures(method = "scran")</span></code><span leaf="">就是用</span><code><span leaf="">SingleCellExperiment</span></code><span leaf="">生态的</span><code><span leaf="">scran::modelGeneVar()</span></code><span leaf="">，默认</span><code><span leaf="">method = "seurat"</span></code><span leaf="">用了和seurat一样的方法。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">pbmc &lt;- NormalizeData(pbmc)</span><span leaf=""><br></span><span leaf="">pbmc &lt;- FindVariableFeatures(pbmc)</span><span leaf=""><br></span><span leaf="">top10 &lt;- head(VariableFeatures(pbmc), </span><span><span leaf="">10</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">top10</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">##  [1] "PPBP"   "LYZ"    "S100A9" "IGLL5"  "GNLY"   "FTL"   </span></span><span leaf=""><br></span><span><span leaf="">##  [7] "PF4"    "FTH1"   "GNG11"  "S100A8"</span></span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">VariableFeaturePlot(pbmc, label = top10)</span><span leaf=""><br></span></code></pre><h2 data-tool="mdnice编辑器"><span></span><span><span leaf=""><img data-ratio="0.75" data-type="png" data-w="1080" data-imgfileid="100013305" data-src="https://mmbiz.qpic.cn/mmbiz_png/MPBFtnFrw4mlkpbzzuuiaXoRnKMqTYpANpUViaIE0iakb2XTKwlxUfI5zfJIVXCOPuGzNVM3JAnMjykmjMxflgTzg/640?wx_fmt=png" src="https://mmbiz.qpic.cn/mmbiz_png/MPBFtnFrw4mlkpbzzuuiaXoRnKMqTYpANpUViaIE0iakb2XTKwlxUfI5zfJIVXCOPuGzNVM3JAnMjykmjMxflgTzg/640?wx_fmt=png"></span></span></h2><h2 data-tool="mdnice编辑器"><span><span leaf="">降维</span></span><span></span></h2><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">pbmc &lt;- ScaleData(pbmc)</span><span leaf=""><br></span><span leaf="">pbmc &lt;- runPCA(pbmc, subset_row = VariableFeatures(pbmc), exprs_values = </span><span><span leaf="">"scaled"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">ElbowPlot(pbmc)</span><span leaf=""><br></span></code></pre><h2 data-tool="mdnice编辑器"><span></span><span><span leaf=""><img data-ratio="0.7" data-type="png" data-w="1080" data-imgfileid="100013306" data-src="https://mmbiz.qpic.cn/mmbiz_png/MPBFtnFrw4mlkpbzzuuiaXoRnKMqTYpANtEOI5fBezl1GibQVWyqAVgy1Z1jwJmfZTkiaqIib9Y1bfxCz6yTsj6Jag/640?wx_fmt=png" src="https://mmbiz.qpic.cn/mmbiz_png/MPBFtnFrw4mlkpbzzuuiaXoRnKMqTYpANtEOI5fBezl1GibQVWyqAVgy1Z1jwJmfZTkiaqIib9Y1bfxCz6yTsj6Jag/640?wx_fmt=png"></span></span></h2><h2 data-tool="mdnice编辑器"><span><span leaf="">聚类</span></span><span></span></h2><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">set.seed(</span><span><span leaf="">2024</span></span><span leaf="">-</span><span><span leaf="">10</span></span><span leaf="">-</span><span><span leaf="">01</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">pbmc &lt;- FindNeighbors(pbmc, dims = </span><span><span leaf="">1</span></span><span leaf="">:</span><span><span leaf="">10</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">pbmc &lt;- FindClusters(pbmc)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">head(Idents(pbmc), </span><span><span leaf="">5</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">## AAACATACAACCAC-1 AAACATTGAGCTAC-1 AAACATTGATCAGC-1 </span></span><span leaf=""><br></span><span><span leaf="">##                1                2                3 </span></span><span leaf=""><br></span><span><span leaf="">## AAACCGTGCTTCCG-1 AAACCGTGTATGCG-1 </span></span><span leaf=""><br></span><span><span leaf="">##                4                5 </span></span><span leaf=""><br></span><span><span leaf="">## Levels: 1 2 3 4 5 6 7 8</span></span><span leaf=""><br></span></code></pre><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">UMAP</span></span><span></span></h2><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">pbmc &lt;- RunUMAP(pbmc, </span><span><span leaf="">1</span></span><span leaf="">:</span><span><span leaf="">10</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(ggsc)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(aplot)</span><span leaf=""><br></span><span leaf="">p1 &lt;- sc_dim(pbmc, reduction=</span><span><span leaf="">"PCA"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">p2 &lt;- sc_dim(pbmc, reduction=</span><span><span leaf="">"UMAP"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">plot_list(gglist = list(PCA=p1, UMAP=p2))</span><span leaf=""><br></span></code></pre><h2 data-tool="mdnice编辑器"><span></span><span><span leaf=""><img data-ratio="0.5768518518518518" data-type="png" data-w="1080" data-imgfileid="100013307" data-src="https://mmbiz.qpic.cn/mmbiz_png/MPBFtnFrw4mlkpbzzuuiaXoRnKMqTYpANK2DFeFVXHFywyycPHibCtk3iba6UpZHosVe7J5qYIVn4VkltfWOJ3eKQ/640?wx_fmt=png" src="https://mmbiz.qpic.cn/mmbiz_png/MPBFtnFrw4mlkpbzzuuiaXoRnKMqTYpANK2DFeFVXHFywyycPHibCtk3iba6UpZHosVe7J5qYIVn4VkltfWOJ3eKQ/640?wx_fmt=png"></span></span></h2><h2 data-tool="mdnice编辑器"><span><span leaf="">Marker基因</span></span><span></span></h2><p data-tool="mdnice编辑器"><span leaf="">没错呀，Seurat找Marker基因的方法，我们也给搞一遍。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># find all markers of cluster 2</span></span><span leaf=""><br></span><span leaf="">cluster2.markers &lt;- FindMarkers(pbmc, ident.1 = </span><span><span leaf="">2</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">head(cluster2.markers, n = </span><span><span leaf="">5</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">##                   pval          padj avg_log2FC pct.1 pct.2</span></span><span leaf=""><br></span><span><span leaf="">## CD79A     0.000000e+00  0.000000e+00   6.888332 0.923 0.041</span></span><span leaf=""><br></span><span><span leaf="">## MS4A1     0.000000e+00  0.000000e+00   5.689795 0.840 0.053</span></span><span leaf=""><br></span><span><span leaf="">## TCL1A    6.735525e-271 3.079033e-267   6.827811 0.617 0.021</span></span><span leaf=""><br></span><span><span leaf="">## CD79B    6.080891e-269 2.084833e-265   4.609790 0.903 0.142</span></span><span leaf=""><br></span><span><span leaf="">## HLA-DQA1 4.615369e-267 1.265904e-263   3.960900 0.886 0.116</span></span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># find all markers distinguishing cluster 5 from clusters 0 and 3</span></span><span leaf=""><br></span><span leaf="">cluster5.markers &lt;- FindMarkers(pbmc, ident.1 = </span><span><span leaf="">5</span></span><span leaf="">, ident.2 = c(</span><span><span leaf="">0</span></span><span leaf="">, </span><span><span leaf="">3</span></span><span leaf="">))</span><span leaf=""><br></span><span leaf="">head(cluster5.markers, n = </span><span><span leaf="">5</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">##                 pval          padj avg_log2FC pct.1 pct.2</span></span><span leaf=""><br></span><span><span leaf="">## GZMB   1.469274e-130 2.014962e-126   8.447240 0.955 0.033</span></span><span leaf=""><br></span><span><span leaf="">## NKG7   1.545205e-114 1.059547e-110   7.562537 1.000 0.142</span></span><span leaf=""><br></span><span><span leaf="">## GNLY   4.489308e-113 2.052212e-109   7.791358 0.975 0.120</span></span><span leaf=""><br></span><span><span leaf="">## FGFBP2 2.694631e-112 9.238541e-109   7.771659 0.860 0.035</span></span><span leaf=""><br></span><span><span leaf="">## PRF1   1.983747e-111 5.441020e-108   6.136492 0.936 0.087</span></span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># find markers for every cluster compared to all remaining cells, </span></span><span leaf=""><br></span><span><span leaf=""># report only the positive ones</span></span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(dplyr)</span><span leaf=""><br></span><span leaf="">pbmc.markers &lt;- FindAllMarkers(pbmc, only.pos = </span><span><span leaf="">TRUE</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">topN_marker &lt;- </span><span><span leaf="">function</span></span><span leaf="">(markers, n) {</span><span leaf=""><br></span><span leaf="">    markers %&gt;%</span><span leaf=""><br></span><span leaf="">        group_by(cluster) %&gt;%</span><span leaf=""><br></span><span leaf="">        dplyr::filter(avg_log2FC &gt; </span><span><span leaf="">1</span></span><span leaf="">) %&gt;%</span><span leaf=""><br></span><span leaf="">        slice_head(n = n) %&gt;%</span><span leaf=""><br></span><span leaf="">        ungroup()</span><span leaf=""><br></span><span leaf="">}</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">top10 &lt;- topN_marker(pbmc.markers, </span><span><span leaf="">10</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">split(top10$gene, top10$cluster)</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">## $`1`</span></span><span leaf=""><br></span><span><span leaf="">##  [1] "CCL5"  "GZMK"  "NKG7"  "GZMA"  "CST7"  "CD8A"  "CTSW" </span></span><span leaf=""><br></span><span><span leaf="">##  [8] "IL32"  "LYAR"  "KLRG1"</span></span><span leaf=""><br></span><span><span leaf="">## </span></span><span leaf=""><br></span><span><span leaf="">## $`2`</span></span><span leaf=""><br></span><span><span leaf="">##  [1] "CD79A"     "MS4A1"     "TCL1A"     "CD79B"    </span></span><span leaf=""><br></span><span><span leaf="">##  [5] "HLA-DQA1"  "LINC00926" "VPREB3"    "HLA-DQB1" </span></span><span leaf=""><br></span><span><span leaf="">##  [9] "CD74"      "HLA-DRA"  </span></span><span leaf=""><br></span><span><span leaf="">## </span></span><span leaf=""><br></span><span><span leaf="">## $`3`</span></span><span leaf=""><br></span><span><span leaf="">##  [1] "LTB"     "IL7R"    "IL32"    "LDHB"    "CD3D"   </span></span><span leaf=""><br></span><span><span leaf="">##  [6] "TNFRSF4" "AQP3"    "CD2"     "CD3E"    "CD40LG" </span></span><span leaf=""><br></span><span><span leaf="">## </span></span><span leaf=""><br></span><span><span leaf="">## $`4`</span></span><span leaf=""><br></span><span><span leaf="">##  [1] "LGALS2" "S100A8" "FCN1"   "S100A9" "MS4A6A" "CST3"  </span></span><span leaf=""><br></span><span><span leaf="">##  [7] "TYROBP" "CD14"   "LYZ"    "TYMP"  </span></span><span leaf=""><br></span><span><span leaf="">## </span></span><span leaf=""><br></span><span><span leaf="">## $`5`</span></span><span leaf=""><br></span><span><span leaf="">##  [1] "GZMB"   "FGFBP2" "SPON2"  "PRF1"   "GNLY"   "XCL2"  </span></span><span leaf=""><br></span><span><span leaf="">##  [7] "AKR1C3" "CLIC3"  "KLRD1"  "CST7"  </span></span><span leaf=""><br></span><span><span leaf="">## </span></span><span leaf=""><br></span><span><span leaf="">## $`6`</span></span><span leaf=""><br></span><span><span leaf="">##  [1] "CDKN1C"        "HES4"          "RP11-290F20.3"</span></span><span leaf=""><br></span><span><span leaf="">##  [4] "MS4A7"         "FCGR3A"        "CKB"          </span></span><span leaf=""><br></span><span><span leaf="">##  [7] "LILRA3"        "IFITM3"        "LRRC25"       </span></span><span leaf=""><br></span><span><span leaf="">## [10] "LST1"         </span></span><span leaf=""><br></span><span><span leaf="">## </span></span><span leaf=""><br></span><span><span leaf="">## $`7`</span></span><span leaf=""><br></span><span><span leaf="">##  [1] "CCR7"      "LDHB"      "LEF1"      "FHIT"     </span></span><span leaf=""><br></span><span><span leaf="">##  [5] "PRKCQ-AS1" "CD3E"      "NOSIP"     "PIK3IP1"  </span></span><span leaf=""><br></span><span><span leaf="">##  [9] "CD27"      "LINC00176"</span></span><span leaf=""><br></span><span><span leaf="">## </span></span><span leaf=""><br></span><span><span leaf="">## $`8`</span></span><span leaf=""><br></span><span><span leaf="">##  [1] "GP9"                   "ITGA2B"               </span></span><span leaf=""><br></span><span><span leaf="">##  [3] "AP001189.4"            "LY6G6F"               </span></span><span leaf=""><br></span><span><span leaf="">##  [5] "TMEM40"                "SEPT5"                </span></span><span leaf=""><br></span><span><span leaf="">##  [7] "HGD"                   "TREML1"               </span></span><span leaf=""><br></span><span><span leaf="">##  [9] "ITGB3_ENSG00000259207" "SPARC"</span></span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">当然还额外赠送了</span><code><span leaf="">fanyi</span></code><span leaf="">包的功能：</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">top1 &lt;- topN_marker(pbmc.markers, </span><span><span leaf="">1</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">gene_table &lt;- gene_summary_table(top1, gene_col = </span><span><span leaf="">'gene'</span></span><span leaf="">, </span><span leaf=""><br></span><span leaf="">    cluster_col = </span><span><span leaf="">'cluster'</span></span><span leaf="">, keyType=</span><span><span leaf="">'SYMBOL'</span></span><span leaf="">, output = </span><span><span leaf="">"data.frame"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">gene_table</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">##   ENTREZID   gene          pval          padj avg_log2FC</span></span><span leaf=""><br></span><span><span leaf="">## 1     1028 CDKN1C 9.252218e-233 1.268849e-228   5.835202</span></span><span leaf=""><br></span><span><span leaf="">## 2     1236   CCR7 7.265363e-105 1.423388e-101   2.579844</span></span><span leaf=""><br></span><span><span leaf="">## 3     2815    GP9  0.000000e+00  0.000000e+00  11.419366</span></span><span leaf=""><br></span><span><span leaf="">## 4     3002   GZMB 7.268536e-269 9.968070e-265   6.001673</span></span><span leaf=""><br></span><span><span leaf="">## 5     3957 LGALS2  0.000000e+00  0.000000e+00   6.351912</span></span><span leaf=""><br></span><span><span leaf="">## 6     4050    LTB 4.712630e-115 6.462900e-111   1.484305</span></span><span leaf=""><br></span><span><span leaf="">## 7     6352   CCL5 1.016869e-271 1.394534e-267   3.820950</span></span><span leaf=""><br></span><span><span leaf="">## 8      973  CD79A  0.000000e+00  0.000000e+00   6.888332</span></span><span leaf=""><br></span><span><span leaf="">##   pct.1 pct.2 cluster   name</span></span><span leaf=""><br></span><span><span leaf="">## 1 0.531 0.008       6 CDKN1C</span></span><span leaf=""><br></span><span><span leaf="">## 2 0.526 0.116       7   CCR7</span></span><span leaf=""><br></span><span><span leaf="">## 3 0.917 0.002       8    GP9</span></span><span leaf=""><br></span><span><span leaf="">## 4 0.955 0.068       5   GZMB</span></span><span leaf=""><br></span><span><span leaf="">## 5 0.906 0.047       4 LGALS2</span></span><span leaf=""><br></span><span><span leaf="">## 6 0.987 0.631       3    LTB</span></span><span leaf=""><br></span><span><span leaf="">## 7 0.982 0.201       1   CCL5</span></span><span leaf=""><br></span><span><span leaf="">## 8 0.923 0.041       2  CD79A</span></span><span leaf=""><br></span><span><span leaf="">##                            description</span></span><span leaf=""><br></span><span><span leaf="">## 1 cyclin dependent kinase inhibitor 1C</span></span><span leaf=""><br></span><span><span leaf="">## 2       C-C motif chemokine receptor 7</span></span><span leaf=""><br></span><span><span leaf="">## 3             glycoprotein IX platelet</span></span><span leaf=""><br></span><span><span leaf="">## 4                           granzyme B</span></span><span leaf=""><br></span><span><span leaf="">## 5                           galectin 2</span></span><span leaf=""><br></span><span><span leaf="">## 6                     lymphotoxin beta</span></span><span leaf=""><br></span><span><span leaf="">## 7         C-C motif chemokine ligand 5</span></span><span leaf=""><br></span><span><span leaf="">## 8                       CD79a molecule</span></span><span leaf=""><br></span><span><span leaf="">##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       summary</span></span><span leaf=""><br></span><span><span leaf="">## 1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     This gene is imprinted, with preferential expression of the maternal allele. The encoded protein is a tight-binding, strong inhibitor of several G1 cyclin/Cdk complexes and a negative regulator of cell proliferation. Mutations in this gene are implicated in sporadic cancers and Beckwith-Wiedemann syndorome, suggesting that this gene is a tumor suppressor candidate. Three transcript variants encoding two different isoforms have been found for this gene. [provided by RefSeq, Oct 2010]</span></span><span leaf=""><br></span><span><span leaf="">## 2                                                                                                      The protein encoded by this gene is a member of the G protein-coupled receptor family. This receptor was identified as a gene induced by the Epstein-Barr virus (EBV), and is thought to be a mediator of EBV effects on B lymphocytes. This receptor is expressed in various lymphoid tissues and activates B and T lymphocytes. It has been shown to control the migration of memory T cells to inflamed tissues, as well as stimulate dendritic cell maturation. The chemokine (C-C motif) ligand 19 (CCL19/ECL) has been reported to be a specific ligand of this receptor. Signals mediated by this receptor regulate T cell homeostasis in lymph nodes, and may also function in the activation and polarization of T cells, and in chronic inflammation pathogenesis. Alternative splicing of this gene results in multiple transcript variants. [provided by RefSeq, Sep 2014]</span></span><span leaf=""><br></span><span><span leaf="">## 3                                                                                                                                                                                                                                                                                                                                      This gene encodes a small membrane glycoprotein found on the surface of human platelets. It forms a 1-to-1 noncovalent complex with glycoprotein Ib, a platelet surface membrane glycoprotein complex that functions as a receptor for von Willebrand factor. The complete receptor complex includes noncovalent association of the alpha and beta subunits with the protein encoded by this gene and platelet glycoprotein V. Defects in this gene are a cause of Bernard-Soulier syndrome, also known as giant platelet disease. These patients have unusually large platelets and have a clinical bleeding tendency. [provided by RefSeq, Oct 2008]</span></span><span leaf=""><br></span><span><span leaf="">## 4                                                                                                                                                                                                                                                                                                                                                                                This gene encodes a member of the granzyme subfamily of proteins, part of the peptidase S1 family of serine proteases. The encoded preproprotein is secreted by natural killer (NK) cells and cytotoxic T lymphocytes (CTLs) and proteolytically processed to generate the active protease, which induces target cell apoptosis. This protein also processes cytokines and degrades extracellular matrix proteins, and these roles are implicated in chronic inflammation and wound healing. Expression of this gene may be elevated in human patients with cardiac fibrosis. [provided by RefSeq, Sep 2016]</span></span><span leaf=""><br></span><span><span leaf="">## 5                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       The protein encoded by this gene is a soluble beta-galactoside binding lectin. The encoded protein is found as a homodimer and can bind to lymphotoxin-alpha. A single nucleotide polymorphism in an intron of this gene can alter the transcriptional level of the protein, with a resultant increased risk of myocardial infarction. [provided by RefSeq, Jul 2008]</span></span><span leaf=""><br></span><span><span leaf="">## 6                                                                                                                                                                         Lymphotoxin beta is a type II membrane protein of the TNF family. It anchors lymphotoxin-alpha to the cell surface through heterotrimer formation. The predominant form on the lymphocyte surface is the lymphotoxin-alpha 1/beta 2 complex (e.g. 1 molecule alpha/2 molecules beta) and this complex is the primary ligand for the lymphotoxin-beta receptor. The minor complex is lymphotoxin-alpha 2/beta 1. LTB is an inducer of the inflammatory response system and involved in normal development of lymphoid tissue. Lymphotoxin-beta isoform b is unable to complex with lymphotoxin-alpha suggesting a function for lymphotoxin-beta which is independent of lympyhotoxin-alpha. Alternative splicing results in multiple transcript variants encoding different isoforms. [provided by RefSeq, Jul 2008]</span></span><span leaf=""><br></span><span><span leaf="">## 7 This gene is one of several chemokine genes clustered on the q-arm of chromosome 17. Chemokines form a superfamily of secreted proteins involved in immunoregulatory and inflammatory processes. The superfamily is divided into four subfamilies based on the arrangement of the N-terminal cysteine residues of the mature peptide. This chemokine, a member of the CC subfamily, functions as a chemoattractant for blood monocytes, memory T helper cells and eosinophils. It causes the release of histamine from basophils and activates eosinophils. This cytokine is one of the major HIV-suppressive factors produced by CD8+ cells. It functions as one of the natural ligands for the chemokine receptor chemokine (C-C motif) receptor 5 (CCR5), and it suppresses in vitro replication of the R5 strains of HIV-1, which use CCR5 as a coreceptor. Alternative splicing results in multiple transcript variants that encode different isoforms. [provided by RefSeq, Jul 2013]</span></span><span leaf=""><br></span><span><span leaf="">## 8                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  The B lymphocyte antigen receptor is a multimeric complex that includes the antigen-specific component, surface immunoglobulin (Ig). Surface Ig non-covalently associates with two other proteins, Ig-alpha and Ig-beta, which are necessary for expression and function of the B-cell antigen receptor. This gene encodes the Ig-alpha protein of the B-cell antigen component. Alternatively spliced transcript variants encoding different isoforms have been described. [provided by RefSeq, Jul 2008]</span></span><span leaf=""><br></span></code></pre><h2 data-tool="mdnice编辑器"><span></span><span><a href="https://mp.weixin.qq.com/s?__biz=MzI5NjUyNzkxMg==&amp;mid=2247496428&amp;idx=1&amp;sn=6cbb4dab24253752f9f96bde90b70b74&amp;scene=21#wechat_redirect" imgurl="https://mmbiz.qpic.cn/mmbiz_jpg/MPBFtnFrw4mlkpbzzuuiaXoRnKMqTYpANKwbibV0vLuTFQy90cWHcPRezDTFkdD54FZwpeEyzAOrkAeP9LykFOOw/640?wx_fmt=jpeg&amp;from=appmsg" linktype="image" tab="innerlink" data-itemshowtype="0" target="_blank" data-linktype="1"><span><img data-ratio="0.11203703703703703" data-type="jpeg" data-w="1080" data-imgfileid="100013308" data-src="https://mmbiz.qpic.cn/mmbiz_jpg/MPBFtnFrw4mlkpbzzuuiaXoRnKMqTYpANKwbibV0vLuTFQy90cWHcPRezDTFkdD54FZwpeEyzAOrkAeP9LykFOOw/640?wx_fmt=jpeg" src="https://mmbiz.qpic.cn/mmbiz_jpg/MPBFtnFrw4mlkpbzzuuiaXoRnKMqTYpANKwbibV0vLuTFQy90cWHcPRezDTFkdD54FZwpeEyzAOrkAeP9LykFOOw/640?wx_fmt=jpeg"></span></a></span></h2><h2 data-tool="mdnice编辑器"><span><span leaf="">细胞注释</span></span><span></span></h2><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">new_labels &lt;- c(</span><span><span leaf="">"CD8+ T"</span></span><span leaf="">,</span><span><span leaf="">"B"</span></span><span leaf="">,</span><span><span leaf="">"Memory CD4+"</span></span><span leaf="">,</span><span><span leaf="">"CD14+ Mono"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">    </span><span><span leaf="">"NK"</span></span><span leaf="">,</span><span><span leaf="">"FCGR3A+ Mono"</span></span><span leaf="">,</span><span><span leaf="">"Native CD4+ T"</span></span><span leaf="">,</span><span><span leaf="">"Platelet"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">pbmc &lt;- RenameIdents(pbmc, new_labels)</span><span leaf=""><br></span><span leaf="">sc_dim(pbmc, reduction=</span><span><span leaf="">"UMAP"</span></span><span leaf="">) + sc_dim_geom_label()</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf=""><img data-ratio="0.7138888888888889" data-type="png" data-w="1080" data-imgfileid="100013309" data-src="https://mmbiz.qpic.cn/mmbiz_png/MPBFtnFrw4mlkpbzzuuiaXoRnKMqTYpANGdVrSZ2IHfXBqFcNen6eTWLbiaa3WZqO3XbEqgUwnZUlOGgzdibicQK7w/640?wx_fmt=png" src="https://mmbiz.qpic.cn/mmbiz_png/MPBFtnFrw4mlkpbzzuuiaXoRnKMqTYpANGdVrSZ2IHfXBqFcNen6eTWLbiaa3WZqO3XbEqgUwnZUlOGgzdibicQK7w/640?wx_fmt=png"></span></p><p data-tool="mdnice编辑器"><span leaf="">基本上是一样的，最后出的图，虽然有一点不一样，但我们知道这是一些参数设置的影响（乃至于只是随机数设定上的差别），出来这个图的时候，我已经懒得去看seurat的代码，让umap图也一样了。</span></p><p data-tool="mdnice编辑器"><span leaf="">我当时跟学生讲</span><code><span leaf="">sclet</span></code><span leaf="">这个包的时候，就强调了，虽然我演示的复刻seurat功能，这张皮就像我弄的一个小玩具，但它并不是玩玩而已，因为你要用的并不是单纯</span><code><span leaf="">sclet</span></code><span leaf="">这一个包，而是</span><code><span leaf="">SingleCellExperiment</span></code><span leaf="">整个生态，这个包是劝大家切换的，是入门的。它的背后是一个完整的生态，有你要的各种功能。</span></p><h1 data-tool="mdnice编辑器"><span></span><span><span leaf="">后续</span></span><span></span></h1><p data-tool="mdnice编辑器"><span leaf="">当然虽说一张皮，我也希望能造更多的皮，让大家分析对接得更舒服一些。所以这个</span><code><span leaf="">sclet</span></code><span leaf="">包，也陆陆续续地对接了一些包的使用，并且都有文档。</span></p><section nodeleaf=""><img data-ratio="0.7759259259259259" data-s="300,640" data-type="png" data-w="1080" type="block" data-imgfileid="100013302" data-src="https://mmbiz.qpic.cn/mmbiz_png/MPBFtnFrw4mlkpbzzuuiaXoRnKMqTYpAN6Foj1mZFGcbAsHibNQcmzeeS16GD8QRcRAjhkyEk3w1lldvIARic4iaag/640?wx_fmt=png" src="https://mmbiz.qpic.cn/mmbiz_png/MPBFtnFrw4mlkpbzzuuiaXoRnKMqTYpAN6Foj1mZFGcbAsHibNQcmzeeS16GD8QRcRAjhkyEk3w1lldvIARic4iaag/640?wx_fmt=png"></section><section><span leaf=""><br></span></section><p data-tool="mdnice编辑器"><span leaf="">然后呢，如果你有兴趣：</span></p><ol><li><section><span leaf="">可以提需求，我们或者可以帮助对接一下，使其更易用，或者出更漂亮的图。</span></section></li><li><section><span leaf="">可以一起来缝虎皮裙子，造福社会。</span></section></li></ol></section><p><mp-style-type data-value="3"></mp-style-type></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/51qmaNG4RthL3xDta8F2Ew",target="_blank" rel="noopener noreferrer">原文链接</a>
