---
title: "一文完成单细胞pseudobulk分组以及多重分组比较分析"
date: 2025-03-18T03:18:16Z
draft: ["false"]
tags: [
  "fetched",
  "瓶子里有颗星星"
]
categories: ["Acdemic"]
---
一文完成单细胞pseudobulk分组以及多重分组比较分析 by 瓶子里有颗星星
------
<div><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com"><h1 data-tool="mdnice编辑器"><span leaf=""><br></span><span><span leaf=""><br></span></span><span leaf=""><br></span></h1><blockquote><section><span leaf=""><br></span><p><span leaf="">不只是单细胞，也不只是pseudobulk</span></p></section></blockquote><p data-tool="mdnice编辑器"><span leaf="">在学习过程中，通常是以单样本，不设置分组进行学习，这方面的官方教程和相应的科普文都很多，比如Seurat官网非常经典的PBMC3k示例教程https://satijalab.org/seurat/articles/pbmc3k_tutorial</span></p><p data-tool="mdnice编辑器"><span leaf="">但是在我们实际研究过程中，通常是需要进行分组进行分析和比较，比如治疗前后的比较，对照组和实验组之间的比较，还有像发在Nature Medicine 的这篇文章：Combined PD-1, BRAF and MEK inhibition inBRAFV600E colorectal cancer: a phase 2 trial， 更是设置了双重分组进行比较（是否响应+治疗前后）</span></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes05j0l1Sf5ncvuGtMqQGoxFaynP335zdVOW4IA0hwk8BfY9GUX11QsTg/640?wx_fmt=png&amp;from=appmsg" alt="image-20250217153206358" data-ratio="0.9166666666666666" data-type="png" data-w="1080" data-imgfileid="100002053" src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes05j0l1Sf5ncvuGtMqQGoxFaynP335zdVOW4IA0hwk8BfY9GUX11QsTg/640?wx_fmt=png&amp;from=appmsg"></span><figcaption><span leaf="">image-20250217153206358</span></figcaption></figure><p data-tool="mdnice编辑器"><span leaf="">而且，像单细胞测序样本爆炸式增长，单细胞自身数量以及所需计算资源也是远超Bulk RNA-seq的样本量的存在，尤其存在生物学重复的时候（比如实验组和对照组都进行了数例单细胞测序），这时候就需要引入pseudobulks。而且，关于pseudobulks分析，是有助于减少计算偏倚的，且加速运算，减少计算资源，且减少假阳性，详见发在NC的《Confronting false discoveries in single-cell differential expression》</span></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes06NkNvwQcSpx4FhCtSt1U6icsRCzHTl6hjWsdYJry6xKNgiaicTg61Cvjg/640?wx_fmt=png&amp;from=appmsg" alt="image-20250217154556309" data-ratio="0.7916666666666666" data-type="png" data-w="1080" data-imgfileid="100002054" src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes06NkNvwQcSpx4FhCtSt1U6icsRCzHTl6hjWsdYJry6xKNgiaicTg61Cvjg/640?wx_fmt=png&amp;from=appmsg"></span><figcaption><span leaf="">image-20250217154556309</span></figcaption></figure><p data-tool="mdnice编辑器"><span leaf="">所以本文我们以pseudobulks为例，介绍如何实现单个分组（只比较一种条件，如是否用药，用药前后等等）或双重分组（比如上述Nature Medicine文章，用药前后+是否响应）进行DEseq2分析，并包装成函数方便使用，一行函数得到结果和完成绘图。</span></p><p data-tool="mdnice编辑器"><span leaf="">函数内容包括</span></p><blockquote><section><span leaf=""><br></span><ol><li><section><p><span leaf="">对单细胞进行pseudobulk处理并进行DESeq2差异分析；</span></p></section></li><li><section><p><span leaf="">基于DESeq2进行绘图，</span></p></section><section><p><span leaf="">针对差异基因的两款Volcano plot；</span></p></section><section><p><span leaf="">针对差异基因的富集分析，图包括Barplot + 两款Lollipop，该分析又包括三种</span></p></section></li><ol><li><section><span leaf="">分析所有显著差异基因；</span></section></li><li><section><span leaf="">只分析上/下调的显著差异基因；</span></section></li></ol></ol><section><p><span leaf="">3. 针对Broad所有pathway activity进行分析并绘图，包括</span></p></section><ol><li><section><p><span leaf="">单个GSEA图（两种图形可选）</span></p></section></li><li><section><p><span leaf="">汇总显著的pathway 并进行上下调柱状图</span></p></section></li><li><section><p><span leaf="">自定义pathway 然后进行分析并绘图，同样也是包括单个GSEA图（两种图形可选）以及汇总显著的pathway 并进行上下调柱状图</span></p></section></li><li><section><p><span leaf="">查看目的兴趣基因在不同分组的情况（柱状图）</span></p></section></li></ol></section></blockquote><p data-tool="mdnice编辑器"><span leaf="">脚本特点</span></p><blockquote><section><span leaf=""><br></span><ol><li><section><span leaf="">人鼠通用（默认是人）；</span></section></li><li><section><span leaf="">psuedobulk分析可选择对全部细胞或者只处理指定细胞类型</span></section></li><li><section><span leaf="">可选单分组或双分组进行分析；</span></section></li><li><section><span leaf="">增加输出目录，结果可选组是否自动输出到本地文件夹；</span></section></li><li><section><span leaf="">使用Broad pathway时，默认是只使用Hallmark, 也可以选择其它，也可以选择所有（包含KEGG, GO, immunesigdb, biocarta等等）</span></section></li><li><section><span leaf="">富集分析的时候除了选择显著差异基因是否同时进行上下调富集分析，也可选择是否同时进行KEGG和GO分析，或者分别单独进行</span></section></li><li><section><span leaf="">画图细节处理，比如基因选择是否斜体，是否增加阴影等等。</span></section></li></ol></section></blockquote><h3 data-tool="mdnice编辑器"><span leaf=""><br></span><span><span leaf="">1. 加载R包和对单细胞矩阵进行初步处理</span></span><span leaf=""><br></span></h3><pre data-tool="mdnice编辑器"><span leaf=""><br></span><code><span leaf="">rm(list=ls())</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(Seurat)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(SeuratData)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(ggplot2)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(RColorBrewer)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(tidyverse)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(magrittr)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(ggpubr)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(ggh4x) </span><span leaf=""><br></span><span leaf="">ifnb.data = LoadData(</span><span><span leaf="">"ifnb"</span></span><span leaf="">) </span><span leaf=""><br></span><span leaf="">table(ifnb.data@meta.data$stim)</span><span leaf=""><br></span><span><span leaf="">#CTRL STIM </span></span><span leaf=""><br></span><span><span leaf="">#6548 7451 </span></span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">现在只有stim这个分组（CTRL和STIM ）以及细胞类型，没有生物学重复也没有其他的分组</span></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes0UicQOeEvh2GhiahmDOhUNeYuR4vVKKd39loQxDSAB5lVW31wFYJCwxCQ/640?wx_fmt=png&amp;from=appmsg" alt="image-20250217220420275" data-ratio="0.32037037037037036" data-type="png" data-w="1080" data-imgfileid="100002052" src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes0UicQOeEvh2GhiahmDOhUNeYuR4vVKKd39loQxDSAB5lVW31wFYJCwxCQ/640?wx_fmt=png&amp;from=appmsg"></span><figcaption><span leaf="">image-20250217220420275</span></figcaption></figure><p data-tool="mdnice编辑器"><span leaf="">进行简单处理</span></p><pre data-tool="mdnice编辑器"><span leaf=""><br></span><code><span leaf="">DefaultAssay(ifnb.data) &lt;- </span><span><span leaf="">'RNA'</span></span><span leaf=""><br></span><span leaf="">ifnb.data &lt;- ifnb.data %&gt;% </span><span leaf=""><br></span><span leaf="">     NormalizeData() %&gt;%</span><span leaf=""><br></span><span leaf="">         FindVariableFeatures() %&gt;% </span><span leaf=""><br></span><span leaf="">         ScaleData() %&gt;%</span><span leaf=""><br></span><span leaf="">         RunPCA()</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">增加生物学重复（STIM和CTRL每组各五个生物学重复），以及对这十个样本增加一列分组（对药物是否响应），这里我们假设stim组四个个响应，一个不响应，ctrl组则反过来。</span></p><pre data-tool="mdnice编辑器"><span leaf=""><br></span><code><span><span leaf=""># 按 stim 分组，平均分配生物学重复（Rep1, Rep2, Rep3）</span></span><span leaf=""><br></span><span leaf="">ifnb.data@meta.data  &lt;- ifnb.data@meta.data  %&gt;%</span><span leaf=""><br></span><span leaf="">  group_by(stim) %&gt;%</span><span leaf=""><br></span><span leaf="">  mutate(biological_replicate = paste0(stim, </span><span><span leaf="">"_"</span></span><span leaf="">, rep(c(</span><span><span leaf="">"Rep1"</span></span><span leaf="">, </span><span><span leaf="">"Rep2"</span></span><span leaf="">, </span><span><span leaf="">"Rep3"</span></span><span leaf="">, </span><span><span leaf="">"Rep4"</span></span><span leaf="">, </span><span><span leaf="">"Rep5"</span></span><span leaf="">), length.out = n())))</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">检查</span></p><pre data-tool="mdnice编辑器"><span leaf=""><br></span><code><span leaf="">table(ifnb.data$biological_replicate)</span><span leaf=""><br></span><span><span leaf=""># CTRL_Rep1 CTRL_Rep2 CTRL_Rep3 CTRL_Rep4 CTRL_Rep5 </span></span><span leaf=""><br></span><span><span leaf="">#    1310      1310      1310      1309      1309 </span></span><span leaf=""><br></span><span><span leaf=""># STIM_Rep1 STIM_Rep2 STIM_Rep3 STIM_Rep4 STIM_Rep5 </span></span><span leaf=""><br></span><span><span leaf="">#     1491      1490      1490      1490      1490</span></span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span leaf=""><br></span><code><span><span leaf=""># 分配药物响应情况：</span></span><span leaf=""><br></span><span leaf="">ifnb.data@meta.data &lt;- ifnb.data@meta.data %&gt;%</span><span leaf=""><br></span><span leaf="">  mutate(response = case_when(</span><span leaf=""><br></span><span leaf="">    stim == </span><span><span leaf="">"CTRL"</span></span><span leaf=""> &amp; biological_replicate %</span><span><span leaf="">in</span></span><span leaf="">% c(</span><span><span leaf="">"CTRL_Rep1"</span></span><span leaf="">, </span><span><span leaf="">"CTRL_Rep2"</span></span><span leaf="">, </span><span><span leaf="">"CTRL_Rep3"</span></span><span leaf="">,</span><span><span leaf="">"CTRL_Rep4"</span></span><span leaf="">) ~ </span><span><span leaf="">"No Response"</span></span><span leaf="">,  </span><span><span leaf=""># CTRL 组前四份为 No Response</span></span><span leaf=""><br></span><span leaf="">    stim == </span><span><span leaf="">"CTRL"</span></span><span leaf=""> &amp; biological_replicate == </span><span><span leaf="">"CTRL_Rep5"</span></span><span leaf=""> ~ </span><span><span leaf="">"Response"</span></span><span leaf="">,  </span><span><span leaf=""># CTRL  组最后一份为 Response</span></span><span leaf=""><br></span><span leaf="">    stim == </span><span><span leaf="">"STIM"</span></span><span leaf=""> &amp; biological_replicate %</span><span><span leaf="">in</span></span><span leaf="">% c(</span><span><span leaf="">"STIM_Rep1"</span></span><span leaf="">, </span><span><span leaf="">"STIM_Rep2"</span></span><span leaf="">, </span><span><span leaf="">"STIM_Rep3"</span></span><span leaf="">,</span><span><span leaf="">"STIM_Rep4"</span></span><span leaf="">) ~ </span><span><span leaf="">"Response"</span></span><span leaf="">,  </span><span><span leaf=""># STIM 组前四份为 Response</span></span><span leaf=""><br></span><span leaf="">    stim == </span><span><span leaf="">"STIM"</span></span><span leaf=""> &amp; biological_replicate == </span><span><span leaf="">"STIM_Rep5"</span></span><span leaf=""> ~ </span><span><span leaf="">"No Response"</span></span><span leaf="">  </span><span><span leaf=""># STIM 组最后一份为 No Response</span></span><span leaf=""><br></span><span leaf="">  ))</span><span leaf=""><br></span><span leaf="">rownames(ifnb.data@meta.data) &lt;- colnames(ifnb.data)</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">检查</span></p><pre data-tool="mdnice编辑器"><span leaf=""><br></span><code><span leaf="">table(ifnb.data$biological_replicate, ifnb.data$response)</span><span leaf=""><br></span><span><span leaf="">#           No Response Response</span></span><span leaf=""><br></span><span><span leaf="">#  CTRL_Rep1        1310        0</span></span><span leaf=""><br></span><span><span leaf="">#  CTRL_Rep2        1310        0</span></span><span leaf=""><br></span><span><span leaf="">#  CTRL_Rep3        1310        0</span></span><span leaf=""><br></span><span><span leaf="">#  CTRL_Rep4        1309        0</span></span><span leaf=""><br></span><span><span leaf="">#  CTRL_Rep5           0     1309</span></span><span leaf=""><br></span><span><span leaf="">#  STIM_Rep1           0     1491</span></span><span leaf=""><br></span><span><span leaf="">#  STIM_Rep2           0     1490</span></span><span leaf=""><br></span><span><span leaf="">#  STIM_Rep3           0     1490</span></span><span leaf=""><br></span><span><span leaf="">#  STIM_Rep4           0     1490</span></span><span leaf=""><br></span><span><span leaf="">#  STIM_Rep5        1490        0</span></span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">准备就绪，加载函数并创建输出文件夹</span></p><pre data-tool="mdnice编辑器"><span leaf=""><br></span><code><span><span leaf="">source</span></span><span leaf="">(</span><span><span leaf="">'Pairs2_pseudo_bulk_DESeq2.R'</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">output_dir = </span><span><span leaf="">'../../Pairs2_pseudo_bulk_DESeq2'</span></span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">DEseq2_output_dir = file.path(output_dir, </span><span><span leaf="">'01.DEseq2_pairs'</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">Volcano_output_dir = file.path(output_dir, </span><span><span leaf="">'02.Volcano'</span></span><span leaf="">) </span><span leaf=""><br></span><span leaf="">KEGG_GO_ORA_output_dir = file.path(output_dir, </span><span><span leaf="">'03.KEGG_GO_ORA'</span></span><span leaf="">)   </span><span leaf=""><br></span><span leaf="">Broad_GSEA_output_dir = file.path(output_dir, </span><span><span leaf="">'04.Broad_GSEA'</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">Col_GSEA_output_dir = file.path(output_dir, </span><span><span leaf="">'05.Col_GSEA'</span></span><span leaf="">)     </span><span leaf=""><br></span><span leaf="">Genes_output_dir = file.path(output_dir, </span><span><span leaf="">'06.Genes'</span></span><span leaf="">)   </span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf="">#dir.create(DEseq2_output_dir,recursive = TRUE)</span></span><span leaf=""><br></span><span><span leaf="">#dir.create(Volcano_output_dir,recursive = TRUE)     </span></span><span leaf=""><br></span><span><span leaf="">#dir.create(KEGG_GO_ORA_output_dir,recursive = TRUE)</span></span><span leaf=""><br></span><span><span leaf="">#dir.create(Col_GSEA_output_dir,recursive = TRUE)</span></span><span leaf=""><br></span><span><span leaf="">#dir.create(Broad_GSEA_output_dir,recursive = TRUE)</span></span><span leaf=""><br></span><span><span leaf="">#dir.create(Genes_output_dir,recursive = TRUE)</span></span><span leaf=""><br></span></code></pre><h3 data-tool="mdnice编辑器"><span leaf=""><br></span><span><span leaf="">2. 进行pseudobulk的DEseq2分析</span></span><span leaf=""><br></span></h3><p data-tool="mdnice编辑器"><span leaf="">先进行以STIM为例，纳入全部细胞的单分组示范</span></p><pre data-tool="mdnice编辑器"><span leaf=""><br></span><code><span leaf="">DESeq2_obj &lt;- Pairs2_pseudo_bulk_DESeq2(seurat_obj = ifnb.data, </span><span><span leaf="">#单细胞对象</span></span><span leaf=""><br></span><span leaf="">                                  sample_col = </span><span><span leaf="">'biological_replicate'</span></span><span leaf="">, </span><span><span leaf="">#生物学重复样本，也就是sampleID</span></span><span leaf=""><br></span><span leaf="">                                  group_col = </span><span><span leaf="">'stim'</span></span><span leaf="">, </span><span><span leaf="">#分组</span></span><span leaf=""><br></span><span leaf="">                                  group_level = c(</span><span><span leaf="">'CTRL'</span></span><span leaf="">, </span><span><span leaf="">'STIM'</span></span><span leaf="">), </span><span><span leaf="">#对照组在前，干预组在后（特地设置，防止方向结果相反）</span></span><span leaf=""><br></span><span leaf="">                                  min_cells = </span><span><span leaf="">10</span></span><span leaf="">,  </span><span><span leaf=""># 默认每个样本最低10个细胞，否则会被过滤</span></span><span leaf=""><br></span><span leaf="">                                  output_dir = DEseq2_output_dir </span><span><span leaf="">#可选，有这地址，则会将DEseq2的运行结果直接输出到本地</span></span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">head(DESeq2_obj) </span><span><span leaf="">#输出看下结果</span></span><span leaf=""><br></span><span><span leaf="">#         baseMean log2FoldChange      lfcSE      stat pvalue padj</span></span><span leaf=""><br></span><span><span leaf=""># ISG15  41601.8914       6.266955 0.05610425 111.70197      0    0</span></span><span leaf=""><br></span><span><span leaf=""># IFI6    4516.2598       4.508051 0.03983272 113.17457      0    0</span></span><span leaf=""><br></span><span><span leaf=""># IFI44L  1014.9313       4.734912 0.08243783  57.43616      0    0</span></span><span leaf=""><br></span><span><span leaf=""># IFI44    568.8719       3.820258 0.08538517  44.74146      0    0</span></span><span leaf=""><br></span><span><span leaf=""># GBP1    2083.3210       2.871889 0.04665976  61.54960      0    0</span></span><span leaf=""><br></span><span><span leaf=""># GBP4     813.5021       3.069299 0.06171407  49.73418      0    0</span></span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">如果提供本地地址，本地将会产生如下文件</span></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes0mT11OOSsGkQYrFwVmVS5ylBXo7txOY3hHEuFic2jLCKSoXrU5QObDpg/640?wx_fmt=png&amp;from=appmsg" alt="image-20250218095515204" data-ratio="0.6175925925925926" data-type="png" data-w="1080" data-imgfileid="100002050" src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes0mT11OOSsGkQYrFwVmVS5ylBXo7txOY3hHEuFic2jLCKSoXrU5QObDpg/640?wx_fmt=png&amp;from=appmsg"></span><figcaption><span leaf="">image-20250218095515204</span></figcaption></figure><p data-tool="mdnice编辑器"><span leaf="">如果需要选择指定细胞类型，则为</span></p><pre data-tool="mdnice编辑器"><span leaf=""><br></span><code><span leaf="">DESeq2_obj_CD14_Mono &lt;- Pairs2_pseudo_bulk_DESeq2(seurat_obj = ifnb.data, </span><span><span leaf="">#单细胞对象</span></span><span leaf=""><br></span><span leaf="">                                  sample_col = </span><span><span leaf="">'biological_replicate'</span></span><span leaf="">, </span><span><span leaf="">#生物学重复样本，也就是sampleID</span></span><span leaf=""><br></span><span leaf="">                                  group_col = </span><span><span leaf="">'stim'</span></span><span leaf="">, </span><span><span leaf="">#分组</span></span><span leaf=""><br></span><span leaf="">                                  group_level = c(</span><span><span leaf="">'CTRL'</span></span><span leaf="">, </span><span><span leaf="">'STIM'</span></span><span leaf="">), </span><span><span leaf="">#对照组在前，干预组在后（特地设置，防止方向结果相反）</span></span><span leaf=""><br></span><span leaf="">       celltype_col = </span><span><span leaf="">'seurat_annotations'</span></span><span leaf="">, </span><span><span leaf="">#细胞类型指定列</span></span><span leaf=""><br></span><span leaf="">                                  selected_celltype = </span><span><span leaf="">'CD14 Mono'</span></span><span leaf="">, </span><span><span leaf="">#所选择的细胞类型</span></span><span leaf=""><br></span><span leaf="">                                  min_cells = </span><span><span leaf="">10</span></span><span leaf="">,  </span><span><span leaf=""># 默认每个样本最低10个细胞，否则会被过滤</span></span><span leaf=""><br></span><span leaf="">                                  output_dir = </span><span><span leaf="">NULL</span></span><span><span leaf="">#这次不输出到本地</span></span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">head(DESeq2_obj_CD14_Mono) </span><span><span leaf="">#输出看下结果</span></span><span leaf=""><br></span><span><span leaf="">#         baseMean log2FoldChange      lfcSE      stat pvalue padj</span></span><span leaf=""><br></span><span><span leaf=""># ISG15  31252.9530       7.134495 0.04987722 143.04115      0    0</span></span><span leaf=""><br></span><span><span leaf=""># IFI6    2111.8723       4.333395 0.05508712  78.66440      0    0</span></span><span leaf=""><br></span><span><span leaf=""># GBP1    1066.7836       3.268803 0.06295213  51.92521      0    0</span></span><span leaf=""><br></span><span><span leaf=""># S100A8  1942.5120      -1.819645 0.04285002 -42.46545      0    0</span></span><span leaf=""><br></span><span><span leaf=""># MNDA     403.3226       3.791513 0.09683171  39.15570      0    0</span></span><span leaf=""><br></span><span><span leaf=""># RSAD2   2927.2526       7.348386 0.10696400  68.69961      0    0</span></span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">现在进行双重分组，代码则为</span></p><pre data-tool="mdnice编辑器"><span leaf=""><br></span><code><span leaf="">DESeq2_obj_pairs &lt;- Pairs2_pseudo_bulk_DESeq2(seurat_obj = ifnb.data, </span><span><span leaf="">#单细胞对象</span></span><span leaf=""><br></span><span leaf="">                                  sample_col = </span><span><span leaf="">'biological_replicate'</span></span><span leaf="">, </span><span><span leaf="">#生物学重复样本，也就是sampleID</span></span><span leaf=""><br></span><span leaf="">                                  group_col = </span><span><span leaf="">'stim'</span></span><span leaf="">, </span><span><span leaf="">#分组</span></span><span leaf=""><br></span><span leaf="">                                  group_level = c(</span><span><span leaf="">'CTRL'</span></span><span leaf="">, </span><span><span leaf="">'STIM'</span></span><span leaf="">), </span><span><span leaf="">#对照组在前，干预组在后（特地设置，防止方向结果相反）</span></span><span leaf=""><br></span><span leaf="">                                  condition_col = </span><span><span leaf="">'response'</span></span><span leaf="">, </span><span><span leaf="">#第二个分组设置为samples是否响应</span></span><span leaf=""><br></span><span leaf="">                                  condition_level = c(</span><span><span leaf="">'No Response'</span></span><span leaf="">, </span><span><span leaf="">'Response'</span></span><span leaf="">), </span><span><span leaf="">#同样设置因子水平，二次定义防止出错</span></span><span leaf=""><br></span><span leaf="">       celltype_col = </span><span><span leaf="">'seurat_annotations'</span></span><span leaf="">, </span><span><span leaf="">#细胞类型指定列</span></span><span leaf=""><br></span><span leaf="">                                  selected_celltype = </span><span><span leaf="">'CD14 Mono'</span></span><span leaf="">, </span><span><span leaf="">#所选择的细胞类型</span></span><span leaf=""><br></span><span leaf="">                                  min_cells = </span><span><span leaf="">10</span></span><span leaf="">,  </span><span><span leaf=""># 默认每个样本最低10个细胞，否则会被过滤</span></span><span leaf=""><br></span><span leaf="">                                  output_dir = </span><span><span leaf="">NULL</span></span><span><span leaf="">#不输出到本地</span></span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">head(DESeq2_obj_pairs)</span><span leaf=""><br></span><span><span leaf="">#            baseMean log2FoldChange     lfcSE        stat       pvalue         padj</span></span><span leaf=""><br></span><span><span leaf=""># CXCL10  17642.891549    -1.36717586 0.1617739 -8.45115128 2.884442e-17 1.734992e-13</span></span><span leaf=""><br></span><span><span leaf=""># NMB        38.442374    -2.58501092 0.6005312 -4.30454042 1.673327e-05 5.032532e-02</span></span><span leaf=""><br></span><span><span leaf=""># NOC2L      11.369199    -0.02175176 1.0442203 -0.02083063 9.833808e-01 9.997521e-01</span></span><span leaf=""><br></span><span><span leaf=""># HES4      293.039045     0.99224341 1.0972957  0.90426254 3.658562e-01 9.997521e-01</span></span><span leaf=""><br></span><span><span leaf=""># ISG15   31252.953015     0.07304462 0.1292091  0.56532101 5.718554e-01 9.997521e-01</span></span><span leaf=""><br></span><span><span leaf=""># TNFRSF4     6.594153    -0.10963990 1.4437204 -0.07594262 9.394648e-01 9.997521e-01</span></span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">三种情况都正常运行了，但注意⚠️ ==双重分组注意不能存在共线性问题，否则会报错==，直白意思是比如示例中，CTRL 组 </span><strong><span leaf="">全是</span></strong><span leaf=""> "No Response"，没有 "Response"，这会导致 Response 变量不能在 CTRL 组中估计。</span></p><p data-tool="mdnice编辑器"><span leaf="">接下来我们主要以第一个结果DESeq2_obj 为例进行图形绘制</span></p><h3 data-tool="mdnice编辑器"><span leaf=""><br></span><span><span leaf="">3. 火山图绘制显著差异基因</span></span><span leaf=""><br></span></h3><p data-tool="mdnice编辑器"><span leaf="">第一种</span></p><pre data-tool="mdnice编辑器"><span leaf=""><br></span><code><span leaf="">volcanoPlot_1 &lt;- get_volcanoPlot_1(DESeq2_obj, </span><span leaf=""><br></span><span leaf="">                        group_col = </span><span><span leaf="">'stim'</span></span><span leaf="">, </span><span><span leaf="">#设置和前面一样</span></span><span leaf=""><br></span><span leaf="">                        group_level = c(</span><span><span leaf="">'CTRL'</span></span><span leaf="">, </span><span><span leaf="">'STIM'</span></span><span leaf="">), </span><span><span leaf="">#设置和前面一样            </span></span><span leaf=""><br></span><span leaf="">                        gene_Italic = </span><span><span leaf="">T</span></span><span leaf="">, </span><span><span leaf="">#默认是不斜体，斜体渲染时间会稍微多一些</span></span><span leaf=""><br></span><span leaf="">                        logFC_threshold = </span><span><span leaf="">1.5</span></span><span leaf="">, </span><span><span leaf=""># 默认Foldchange 倍数为 1.5</span></span><span leaf=""><br></span><span leaf="">                        pval_threshold = </span><span><span leaf="">0.05</span></span><span leaf="">,  </span><span><span leaf=""># 默认P值cutoff为 0.05</span></span><span leaf=""><br></span><span leaf="">                        output_dir = Volcano_output_dir </span><span><span leaf="">#输出文件夹</span></span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">输出图片为</span></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes0SpUSaADg2AEgvjfv2JRtxDC3poLEuiclLfsicRmszJJb9DND8JiaeT04g/640?wx_fmt=png&amp;from=appmsg" alt="image-20250218115232101" data-ratio="0.6574074074074074" data-type="png" data-w="1080" data-imgfileid="100002051" src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes0SpUSaADg2AEgvjfv2JRtxDC3poLEuiclLfsicRmszJJb9DND8JiaeT04g/640?wx_fmt=png&amp;from=appmsg"></span><figcaption><span leaf="">image-20250218115232101</span></figcaption></figure><pre data-tool="mdnice编辑器"><span leaf=""><br></span><code><span><span leaf=""># 如果是双重分组（使用上面的 'CD14 Mono' 分析object）</span></span><span leaf=""><br></span><span leaf="">volcanoPlot_1_2 &lt;- get_volcanoPlot_1(DESeq2_obj_pairs, </span><span leaf=""><br></span><span leaf="">                        group_col = </span><span><span leaf="">'stim'</span></span><span leaf="">, </span><span><span leaf="">#设置和前面一样</span></span><span leaf=""><br></span><span leaf="">                        group_level = c(</span><span><span leaf="">'CTRL'</span></span><span leaf="">, </span><span><span leaf="">'STIM'</span></span><span leaf="">), </span><span><span leaf="">#设置和前面一样 </span></span><span leaf=""><br></span><span leaf="">     condition_col = </span><span><span leaf="">'response'</span></span><span leaf="">, </span><span><span leaf="">#设置和前面一样 </span></span><span leaf=""><br></span><span leaf="">                        condition_level = c(</span><span><span leaf="">'No Response'</span></span><span leaf="">, </span><span><span leaf="">'Response'</span></span><span leaf="">), </span><span><span leaf="">#设置和前面一样 </span></span><span leaf=""><br></span><span leaf="">                        gene_Italic = </span><span><span leaf="">F</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">                        label_gene =  c(</span><span><span leaf="">'CXCL10'</span></span><span leaf="">), </span><span><span leaf=""># 标注感兴趣的基因,可以设置多个（版面要足够大）</span></span><span leaf=""><br></span><span leaf="">                        output_dir = </span><span><span leaf="">NULL</span></span><span leaf=""> </span><span><span leaf="">#这次不输出</span></span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes0tKwQvw5D8T3N9iblm7GM1fVZsVqLzHqvM6H0gXElebDe8kFjEiapOU9Q/640?wx_fmt=png&amp;from=appmsg" alt="image-20250218152116320" data-ratio="0.5453703703703704" data-type="png" data-w="1080" data-imgfileid="100002058" src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes0tKwQvw5D8T3N9iblm7GM1fVZsVqLzHqvM6H0gXElebDe8kFjEiapOU9Q/640?wx_fmt=png&amp;from=appmsg"></span><figcaption><span leaf="">image-20250218152116320</span></figcaption></figure><p data-tool="mdnice编辑器"><span leaf="">第二种火山图</span></p><pre data-tool="mdnice编辑器"><span leaf=""><br></span><code><span leaf="">volcanoPlot_2 &lt;- get_volcanoPlot_2(DESeq2_obj, </span><span leaf=""><br></span><span leaf="">                        group_level = c(</span><span><span leaf="">'CTRL'</span></span><span leaf="">, </span><span><span leaf="">'STIM'</span></span><span leaf="">), </span><span><span leaf="">#设置和前面一样            </span></span><span leaf=""><br></span><span leaf="">                        output_dir = Volcano_output_dir </span><span><span leaf="">#输出文件夹</span></span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes0lUz39ej1B9OM7NuQ1CTRRCNGyGfu4PqC9vUCaSicKmicssUJr2bBvTdA/640?wx_fmt=png&amp;from=appmsg" alt="image-20250218152630644" data-ratio="0.7981481481481482" data-type="png" data-w="1080" data-imgfileid="100002057" src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes0lUz39ej1B9OM7NuQ1CTRRCNGyGfu4PqC9vUCaSicKmicssUJr2bBvTdA/640?wx_fmt=png&amp;from=appmsg"></span><figcaption><span leaf="">image-20250218152630644</span></figcaption></figure><h3 data-tool="mdnice编辑器"><span leaf=""><br></span><span><span leaf="">4. 差异基因富集分析</span></span><span leaf=""><br></span></h3><pre data-tool="mdnice编辑器"><span leaf=""><br></span><code><span leaf="">enrichplot_1 &lt;- enrich_DESeq(DESeq2_obj, </span><span leaf=""><br></span><span leaf="">                         species = </span><span><span leaf="">"human"</span></span><span leaf="">, </span><span><span leaf="">#默认是human, 鼠类可改为mouse</span></span><span leaf=""><br></span><span leaf="">                         logFC_threshold = </span><span><span leaf="">1.5</span></span><span leaf="">,  </span><span><span leaf="">#默认显著差异基因倍数是1.5</span></span><span leaf=""><br></span><span leaf="">                         pval_threshold = </span><span><span leaf="">0.05</span></span><span leaf="">, </span><span><span leaf="">#默认显著差异基因pvalue阈值是0.05</span></span><span leaf=""><br></span><span leaf="">                         analysis_type = </span><span><span leaf="">"both"</span></span><span leaf="">,  </span><span><span leaf="">#both是上下调基因都进行分析，"up" 是只进行上调基因分析, "down"是只下调</span></span><span leaf=""><br></span><span leaf="">                         enrich_type = </span><span><span leaf="">"both"</span></span><span leaf="">,   </span><span><span leaf="">#both是KEGG和GO都进行分析，"KEGG" 是只进行KEGG分析, "GO"是只进行GO分析, </span></span><span leaf=""><br></span><span leaf="">                         num_pathways = </span><span><span leaf="">5</span></span><span leaf="">, </span><span><span leaf="">#默认每项只选择最显著的5条pathway, 不满5条则全选</span></span><span leaf=""><br></span><span leaf="">                         mycol = </span><span><span leaf="">NULL</span></span><span leaf="">, </span><span><span leaf="">#默认无须自定义颜色</span></span><span leaf=""><br></span><span leaf="">                         shadow = </span><span><span leaf="">T</span></span><span leaf="">, </span><span><span leaf=""># 添加阴影，默认是不添加阴影</span></span><span leaf=""><br></span><span leaf="">                         output_dir = KEGG_GO_ORA_output_dir)</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">提供了输出文件夹，会产生如下结果</span></p><blockquote><section><span leaf=""><br></span><p><span leaf="">KEGG.pathway.csv # KEGG的所有富集pathway完整信息</span></p><p><span leaf="">GO.pathway.csv # GO的所有富集pathway完整信息</span></p><p><span leaf="">三款富集分析图</span></p></section></blockquote><p data-tool="mdnice编辑器"><code><span leaf="">Barplot.pdf 富集分析柱状图</span></code></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes0nPLfLezsGKtb6YCHibzKtvmjG1yzJaP1ahHy8lIq1iacluxwVzmRUfNg/640?wx_fmt=png&amp;from=appmsg" alt="image-20250218162017390" data-ratio="0.5509259259259259" data-type="png" data-w="1080" data-imgfileid="100002059" src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes0nPLfLezsGKtb6YCHibzKtvmjG1yzJaP1ahHy8lIq1iacluxwVzmRUfNg/640?wx_fmt=png&amp;from=appmsg"></span><figcaption><span leaf="">image-20250218162017390</span></figcaption></figure><p data-tool="mdnice编辑器"><code><span leaf="">Lollipop_1.pdf 富集分析棒棒糖图1</span></code><span leaf="">   → 手机样式外框</span></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes07JQVyb10xtClvhOHdxSxZQoF932kJhXxT17Uw9aXSQc1WExc25pfEQ/640?wx_fmt=png&amp;from=appmsg" alt="image-20250218155249488" data-ratio="0.687962962962963" data-type="png" data-w="1080" data-imgfileid="100002055" src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes07JQVyb10xtClvhOHdxSxZQoF932kJhXxT17Uw9aXSQc1WExc25pfEQ/640?wx_fmt=png&amp;from=appmsg"></span><figcaption><span leaf="">image-20250218155249488</span></figcaption></figure><p data-tool="mdnice编辑器"><code><span leaf="">Lollipop_2.pdf 富集分析棒棒糖图2</span></code></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes0g8oyfVAHUicmNyCWeD4wvic4QP00asdXJBzOQOoYJVvtmRujQViaAAU9g/640?wx_fmt=png&amp;from=appmsg" alt="image-20250218162049851" data-ratio="0.7083333333333334" data-type="png" data-w="1080" data-imgfileid="100002056" src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes0g8oyfVAHUicmNyCWeD4wvic4QP00asdXJBzOQOoYJVvtmRujQViaAAU9g/640?wx_fmt=png&amp;from=appmsg"></span><figcaption><span leaf="">image-20250218162049851</span></figcaption></figure><pre data-tool="mdnice编辑器"><span leaf=""><br></span><code><span><span leaf=""># 看下只进行GO和上调的情况, 并且富集pathway设置为8条，不添加阴影，不输出到目录</span></span><span leaf=""><br></span><span leaf="">enrichplot_2 &lt;- enrich_DESeq(DESeq2_obj, </span><span leaf=""><br></span><span leaf="">                         analysis_type = </span><span><span leaf="">"up"</span></span><span leaf="">,  </span><span><span leaf="">#只进行上调基因分析</span></span><span leaf=""><br></span><span leaf="">                         enrich_type = </span><span><span leaf="">"GO"</span></span><span leaf="">,   </span><span><span leaf="">#只进行GO分析, </span></span><span leaf=""><br></span><span leaf="">                         num_pathways = </span><span><span leaf="">8</span></span><span leaf="">, </span><span><span leaf="">#富集pathway设置为8条</span></span><span leaf=""><br></span><span leaf="">                         shadow = </span><span><span leaf="">F</span></span><span><span leaf=""># 默认就是不添加阴影，且除非enrich_type为both选项，否则会自动关闭阴影选项</span></span><span leaf=""><br></span><span leaf="">                         )</span><span leaf=""><br></span></code></pre><pre data-tool="mdnice编辑器"><span leaf=""><br></span><code><span leaf="">enrichplot_2$fig$Barplot</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes07BtdcvuWyxYmO4jJ6qR0wJeYCue1Wu1EmjENz9unMvKuZbMf7ibl8jg/640?wx_fmt=png&amp;from=appmsg" alt="image-20250218162347762" data-ratio="0.6342592592592593" data-type="png" data-w="1080" data-imgfileid="100002062" src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes07BtdcvuWyxYmO4jJ6qR0wJeYCue1Wu1EmjENz9unMvKuZbMf7ibl8jg/640?wx_fmt=png&amp;from=appmsg"></span><figcaption><span leaf="">image-20250218162347762</span></figcaption></figure><pre data-tool="mdnice编辑器"><span leaf=""><br></span><code><span leaf="">enrichplot_2$fig$Lollipop_1</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes0MlqYGhUYoPdVxtg0XAWnxyP8E5PvdpXYOKviaX38QujGVbTEXLdVjDA/640?wx_fmt=png&amp;from=appmsg" alt="image-20250218162448229" data-ratio="0.6129629629629629" data-type="png" data-w="1080" data-imgfileid="100002063" src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes0MlqYGhUYoPdVxtg0XAWnxyP8E5PvdpXYOKviaX38QujGVbTEXLdVjDA/640?wx_fmt=png&amp;from=appmsg"></span><figcaption><span leaf="">image-20250218162448229</span></figcaption></figure><pre data-tool="mdnice编辑器"><span leaf=""><br></span><code><span leaf=""> enrichplot_2$fig$Lollipop_2</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes0bY6FZkicEl0e4glbOXQgeiatuUsd8ibGjPg9k1NcETe2jgs1UnibNQuA0g/640?wx_fmt=png&amp;from=appmsg" alt="image-20250218162538209" data-ratio="0.7824074074074074" data-type="png" data-w="1080" data-imgfileid="100002064" src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes0bY6FZkicEl0e4glbOXQgeiatuUsd8ibGjPg9k1NcETe2jgs1UnibNQuA0g/640?wx_fmt=png&amp;from=appmsg"></span><figcaption><span leaf="">image-20250218162538209</span></figcaption></figure><h3 data-tool="mdnice编辑器"><span leaf=""><br></span><span><span leaf="">5. 对Broad pathway activity进行分析并绘图</span></span><span leaf=""><br></span></h3><p data-tool="mdnice编辑器"><span leaf="">这里我们从msigdbr R package获取Broad 相关pathway, 默认是使用human数据且只使用Hallmark（gene_set为NULL则使用全部pathway，慎重，因为分析时间会很长）, GSEA绘图函数是来自GseaVis R package函数，进行修改使其更适合该流程分析</span></p><pre data-tool="mdnice编辑器"><span leaf=""><br></span><code><span leaf="">GSEA_1 &lt;- GSEA_plot(DESeq2_obj, </span><span leaf=""><br></span><span leaf="">                      species = </span><span><span leaf="">"human"</span></span><span leaf="">, </span><span leaf=""><br></span><span leaf="">                      gene_set = </span><span><span leaf="">"H"</span></span><span leaf="">, </span><span leaf=""><br></span><span leaf="">                      NewGseaPlot = </span><span><span leaf="">T</span></span><span leaf="">,  </span><span><span leaf="">#第一款GSEA图</span></span><span leaf=""><br></span><span leaf="">                      output_dir = Broad_GSEA_output_dir)</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">若提供输出目录地址，输出包含</span></p><blockquote><section><span leaf=""><br></span><p><span leaf="">01.All.pathway.csv # 所有pathway的详细情况</span></p><p><span leaf="">02.Significant.pathway.pdf #显著pathway (pvalue &lt; 0.05)的结果</span></p><p><span leaf="">03.每一条pathway的详细情况</span></p></section></blockquote><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes060S3JIvYZXU3IPaZ9R2DOicsicib3Ho5NNQVTiasvPLlH12YQypDwNYjEQ/640?wx_fmt=png&amp;from=appmsg" alt="image-20250218181215487" data-ratio="0.4601851851851852" data-type="png" data-w="1080" data-imgfileid="100002060" src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes060S3JIvYZXU3IPaZ9R2DOicsicib3Ho5NNQVTiasvPLlH12YQypDwNYjEQ/640?wx_fmt=png&amp;from=appmsg"></span><figcaption><span leaf="">image-20250218181215487</span></figcaption></figure><p data-tool="mdnice编辑器"><code><span leaf="">02.Significant.pathway.pdf</span></code></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes0pstrdNTaiaCqkw028xA0azADFFxZxFibJUD4nJtiaWTibn7Z5EFHoOBdiaw/640?wx_fmt=png&amp;from=appmsg" alt="image-20250218180942112" data-ratio="0.4675925925925926" data-type="png" data-w="1080" data-imgfileid="100002061" src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes0pstrdNTaiaCqkw028xA0azADFFxZxFibJUD4nJtiaWTibn7Z5EFHoOBdiaw/640?wx_fmt=png&amp;from=appmsg"></span><figcaption><span leaf="">image-20250218180942112</span></figcaption></figure><p data-tool="mdnice编辑器"><span leaf="">单独的pathway</span></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes0lIkTwUbQwFvJicsBJyVfhTibibhSVULAaiavkkDj4ib0N0CCJk1URaibsDpw/640?wx_fmt=png&amp;from=appmsg" alt="image-20250218182333211" data-ratio="0.7861111111111111" data-type="png" data-w="1080" data-imgfileid="100002065" src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes0lIkTwUbQwFvJicsBJyVfhTibibhSVULAaiavkkDj4ib0N0CCJk1URaibsDpw/640?wx_fmt=png&amp;from=appmsg"></span><figcaption><span leaf="">image-20250218182333211</span></figcaption></figure><pre data-tool="mdnice编辑器"><span leaf=""><br></span><code><span><span leaf=""># 这里介绍如何使用其它category的pathway, 以C6 onocogene为例   查询地址为https://www.gsea-msigdb.org/gsea/msigdb/collections.jsp</span></span><span leaf=""><br></span><span leaf="">GSEA_2 &lt;- GSEA_plot(DESeq2_obj, , </span><span leaf=""><br></span><span leaf="">                      gene_set = </span><span><span leaf="">"C6"</span></span><span leaf="">,  </span><span><span leaf="">#  onocogenic siganture</span></span><span leaf=""><br></span><span leaf="">                      NewGseaPlot = </span><span><span leaf="">F</span></span><span leaf="">) </span><span><span leaf="">#使用另一种绘图风格</span></span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">只富集到上调的通路</span></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes0vq700LER9juMVjEQMTkUwrTGriaQrhD0AflY49ROXbicBeJEYOAYRxsw/640?wx_fmt=png&amp;from=appmsg" alt="image-20250218182459444" data-ratio="0.8555555555555555" data-type="png" data-w="1080" data-imgfileid="100002069" src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes0vq700LER9juMVjEQMTkUwrTGriaQrhD0AflY49ROXbicBeJEYOAYRxsw/640?wx_fmt=png&amp;from=appmsg"></span><figcaption><span leaf="">image-20250218182459444</span></figcaption></figure><p data-tool="mdnice编辑器"><span leaf="">取单独的pathway，用另外一种GSEA绘图风格</span></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes09qDoTQT7B9rCH1CvuicK2vd2S2ouXbib5f4dKjWCOAqF13oSRZ7iarnKw/640?wx_fmt=png&amp;from=appmsg" alt="image-20250218182630795" data-ratio="0.8361111111111111" data-type="png" data-w="1080" data-imgfileid="100002068" src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes09qDoTQT7B9rCH1CvuicK2vd2S2ouXbib5f4dKjWCOAqF13oSRZ7iarnKw/640?wx_fmt=png&amp;from=appmsg"></span><figcaption><span leaf="">image-20250218182630795</span></figcaption></figure><h3 data-tool="mdnice编辑器"><span leaf=""><br></span><span><span leaf="">6.对自定义的pathway进行分析并绘图</span></span><span leaf=""><br></span></h3><p data-tool="mdnice编辑器"><span leaf="">先自定义一些pathway (⚠️pathway里的基因只是随便列举, 不要使用)</span></p><pre data-tool="mdnice编辑器"><span leaf=""><br></span><code><span leaf="">paper_pathways=  </span><span><span leaf="">NULL</span></span><span leaf=""><br></span><span leaf="">paper_pathways$MAPK = c(c(</span><span><span leaf="">"MAPK1"</span></span><span leaf="">, </span><span><span leaf="">"MAPK3"</span></span><span leaf="">, </span><span><span leaf="">"MAPK8"</span></span><span leaf="">, </span><span><span leaf="">"MAPK9"</span></span><span leaf="">, </span><span><span leaf="">"MAPK14"</span></span><span leaf="">, </span><span leaf=""><br></span><span leaf="">                </span><span><span leaf="">"RAF1"</span></span><span leaf="">, </span><span><span leaf="">"BRAF"</span></span><span leaf="">, </span><span><span leaf="">"MEK1"</span></span><span leaf="">, </span><span><span leaf="">"MEK2"</span></span><span leaf="">, </span><span><span leaf="">"ERK1"</span></span><span leaf="">, </span><span><span leaf="">"ERK2"</span></span><span leaf="">, </span><span leaf=""><br></span><span leaf="">                </span><span><span leaf="">"JNK1"</span></span><span leaf="">, </span><span><span leaf="">"JNK2"</span></span><span leaf="">, </span><span><span leaf="">"JNK3"</span></span><span leaf="">, </span><span><span leaf="">"p38"</span></span><span leaf="">, </span><span><span leaf="">"RAS"</span></span><span leaf="">, </span><span><span leaf="">"EGFR"</span></span><span leaf="">, </span><span><span leaf="">"ISG15"</span></span><span leaf="">, </span><span><span leaf="">"IFI6"</span></span><span leaf="">, </span><span><span leaf="">"IFI44L"</span></span><span leaf="">,</span><span><span leaf="">"IFI44"</span></span><span leaf="">,</span><span><span leaf="">"GBP1"</span></span><span leaf="">, </span><span><span leaf="">"GBP4"</span></span><span leaf="">,</span><span><span leaf="">"HLA-E"</span></span><span leaf="">,    </span><span><span leaf="">"HLA-B"</span></span><span leaf=""> ,   </span><span><span leaf="">"PSMB9"</span></span><span leaf="">   ))</span><span leaf=""><br></span><span leaf="">paper_pathways$BRAF_mut = c(</span><span><span leaf="">"BRAF"</span></span><span leaf="">, </span><span><span leaf="">"KRAS"</span></span><span leaf="">, </span><span><span leaf="">"NRAS"</span></span><span leaf="">, </span><span><span leaf="">"HRAS"</span></span><span leaf="">, </span><span><span leaf="">"EGFR"</span></span><span leaf="">, </span><span><span leaf="">"GRB2"</span></span><span leaf="">, </span><span><span leaf="">"SOS1"</span></span><span leaf="">, </span><span leaf=""><br></span><span leaf="">                         </span><span><span leaf="">"MAP2K1"</span></span><span leaf="">, </span><span><span leaf="">"MAP2K2"</span></span><span leaf="">, </span><span><span leaf="">"MAPK1"</span></span><span leaf="">, </span><span><span leaf="">"MAPK3"</span></span><span leaf="">, </span><span><span leaf="">"DUSP6"</span></span><span leaf="">, </span><span><span leaf="">"SPRY2"</span></span><span leaf="">, </span><span><span leaf="">"PTEN"</span></span><span leaf="">, </span><span><span leaf="">"S100A8"</span></span><span leaf="">,</span><span><span leaf="">"MNDA"</span></span><span leaf="">,</span><span><span leaf="">"IFI16"</span></span><span leaf="">,                              </span><span><span leaf="">"EIF2AK2"</span></span><span leaf="">,  </span><span><span leaf="">"CMPK2"</span></span><span leaf="">,</span><span><span leaf="">"LAP3"</span></span><span leaf="">,     </span><span><span leaf="">"IL8"</span></span><span leaf="">,  </span><span><span leaf="">"CXCL10"</span></span><span leaf="">  )</span><span leaf=""><br></span><span leaf="">paper_pathways$PI3K  = c(</span><span><span leaf="">"PIK3CA"</span></span><span leaf="">, </span><span><span leaf="">"PIK3CB"</span></span><span leaf="">, </span><span><span leaf="">"PIK3CD"</span></span><span leaf="">, </span><span><span leaf="">"PIK3CG"</span></span><span leaf="">, </span><span leaf=""><br></span><span leaf="">                         </span><span><span leaf="">"PIK3R1"</span></span><span leaf="">, </span><span><span leaf="">"PIK3R2"</span></span><span leaf="">, </span><span><span leaf="">"AKT1"</span></span><span leaf="">, </span><span><span leaf="">"AKT2"</span></span><span leaf="">, </span><span><span leaf="">"AKT3"</span></span><span leaf="">, </span><span leaf=""><br></span><span leaf="">                         </span><span><span leaf="">"PTEN"</span></span><span leaf="">, </span><span><span leaf="">"PDK1"</span></span><span leaf="">, </span><span><span leaf="">"MTOR"</span></span><span leaf="">, </span><span><span leaf="">"RPTOR"</span></span><span leaf="">, </span><span><span leaf="">"RICTOR"</span></span><span leaf="">, </span><span leaf=""><br></span><span leaf="">                         </span><span><span leaf="">"TSC1"</span></span><span leaf="">, </span><span><span leaf="">"TSC2"</span></span><span leaf="">, </span><span><span leaf="">"FOXO1"</span></span><span leaf="">, </span><span><span leaf="">"FOXO3"</span></span><span leaf="">, </span><span><span leaf="">"BAD"</span></span><span leaf="">, </span><span><span leaf="">"BCL2"</span></span><span leaf="">, </span><span leaf=""><br></span><span leaf="">                         </span><span><span leaf="">"MCL1"</span></span><span leaf="">, </span><span><span leaf="">"GSK3B"</span></span><span leaf="">, </span><span><span leaf="">"RABGAP1L"</span></span><span leaf="">, </span><span><span leaf="">"TMSB10"</span></span><span leaf=""> ,  </span><span><span leaf="">"VAMP5"</span></span><span leaf="">,    </span><span><span leaf="">"IL1RN"</span></span><span leaf="">     )</span><span leaf=""><br></span><span leaf="">paper_pathways$RAS  = c(</span><span><span leaf="">"KRAS"</span></span><span leaf="">, </span><span><span leaf="">"HRAS"</span></span><span leaf="">, </span><span><span leaf="">"NRAS"</span></span><span leaf="">, </span><span><span leaf="">"EGFR"</span></span><span leaf="">, </span><span><span leaf="">"HER2"</span></span><span leaf="">, </span><span><span leaf="">"FGFR"</span></span><span leaf="">, </span><span leaf=""><br></span><span leaf="">                        </span><span><span leaf="">"GRB2"</span></span><span leaf="">, </span><span><span leaf="">"SOS1"</span></span><span leaf="">, </span><span><span leaf="">"BRAF"</span></span><span leaf="">, </span><span><span leaf="">"RAF1"</span></span><span leaf="">, </span><span><span leaf="">"MAP2K1"</span></span><span leaf="">, </span><span><span leaf="">"MAP2K2"</span></span><span leaf="">, </span><span leaf=""><br></span><span leaf="">                        </span><span><span leaf="">"MAPK1"</span></span><span leaf="">, </span><span><span leaf="">"MAPK3"</span></span><span leaf="">, </span><span><span leaf="">"PIK3CA"</span></span><span leaf="">, </span><span><span leaf="">"AKT1"</span></span><span leaf="">, </span><span><span leaf="">"MTOR"</span></span><span leaf="">, </span><span><span leaf="">"NMI"</span></span><span leaf="">  ,  </span><span><span leaf="">"IFIH1"</span></span><span leaf="">  ,  </span><span><span leaf="">"SSB"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">paper_pathways$MHCI  = c(</span><span><span leaf="">"HLA-A"</span></span><span leaf="">, </span><span><span leaf="">"HLA-B"</span></span><span leaf="">, </span><span><span leaf="">"HLA-C"</span></span><span leaf="">, </span><span><span leaf="">"PSMB8"</span></span><span leaf="">, </span><span><span leaf="">"PSMB9"</span></span><span leaf="">, </span><span><span leaf="">"PSMB10"</span></span><span leaf="">, </span><span leaf=""><br></span><span leaf="">                         </span><span><span leaf="">"TAP1"</span></span><span leaf="">, </span><span><span leaf="">"TAP2"</span></span><span leaf="">, </span><span><span leaf="">"TAPBP"</span></span><span leaf="">, </span><span><span leaf="">"ERAP1"</span></span><span leaf="">, </span><span><span leaf="">"ERAP2"</span></span><span leaf="">, </span><span><span leaf="">"B2M"</span></span><span leaf="">, </span><span leaf=""><br></span><span leaf="">                         </span><span><span leaf="">"CALR"</span></span><span leaf="">, </span><span><span leaf="">"PDIA3"</span></span><span leaf="">, </span><span><span leaf="">"CD8A"</span></span><span leaf="">, </span><span><span leaf="">"CD8B"</span></span><span leaf="">, </span><span><span leaf="">"LMP2"</span></span><span leaf="">, </span><span><span leaf="">"LMP7"</span></span><span leaf="">,</span><span><span leaf="">"SP110"</span></span><span leaf="">   , </span><span><span leaf="">"PLSCR1"</span></span><span leaf=""> ,  </span><span><span leaf="">"TNFSF10"</span></span><span leaf=""> ,</span><span><span leaf="">"CXCL11"</span></span><span leaf=""> ,  </span><span><span leaf="">"PPM1K"</span></span><span leaf="">  ,  </span><span><span leaf="">"HERC5"</span></span><span leaf="">   )</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">和上面的Broad pathway 用法稍有区别</span></p><pre data-tool="mdnice编辑器"><span leaf=""><br></span><code><span leaf="">Ind_GSEA &lt;- DIY_pathway_GSEA_plot(DESeq2_obj, </span><span leaf=""><br></span><span leaf="">             pathway_list = paper_pathways,</span><span leaf=""><br></span><span leaf="">                      NewGseaPlot = </span><span><span leaf="">T</span></span><span leaf="">,  </span><span><span leaf="">#第一款GSEA图</span></span><span leaf=""><br></span><span leaf="">                      output_dir = Col_GSEA_output_dir)</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes08V3b71SP5e15SeU7oMk4OSWjia6c3ReUj5djXMu5poN6BicO5NTM7nibA/640?wx_fmt=png&amp;from=appmsg" alt="image-20250218185146513" data-ratio="0.8564814814814815" data-type="png" data-w="1080" data-imgfileid="100002066" src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes08V3b71SP5e15SeU7oMk4OSWjia6c3ReUj5djXMu5poN6BicO5NTM7nibA/640?wx_fmt=png&amp;from=appmsg"></span><figcaption><span leaf="">image-20250218185146513</span></figcaption></figure><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes06C3Yia75yJicfucEe1NWvTGJSmNoenRXqAtXCWRTSYW4fAujyp670HrQ/640?wx_fmt=png&amp;from=appmsg" alt="image-20250218185218582" data-ratio="0.41944444444444445" data-type="png" data-w="1080" data-imgfileid="100002067" src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes06C3Yia75yJicfucEe1NWvTGJSmNoenRXqAtXCWRTSYW4fAujyp670HrQ/640?wx_fmt=png&amp;from=appmsg"></span><figcaption><span leaf="">image-20250218185218582</span></figcaption></figure><h3 data-tool="mdnice编辑器"><span leaf=""><br></span><span><span leaf="">7. 查看目的基因在分组的情况 (这个作用暂时并不大，确认方向可以看看)</span></span><span leaf=""><br></span></h3><pre data-tool="mdnice编辑器"><span leaf=""><br></span><code><span leaf="">tar_genes_plot &lt;-  Var_gene_plot(ifnb.data, </span><span><span leaf="">#需要单细胞对象，而非DEseq2后的结果</span></span><span leaf=""><br></span><span leaf="">        sample_col = </span><span><span leaf="">'biological_replicate'</span></span><span leaf="">,  </span><span><span leaf=""># 和前面设置一样</span></span><span leaf=""><br></span><span leaf="">        group_col = </span><span><span leaf="">'stim'</span></span><span leaf="">, </span><span><span leaf=""># 和前面设置一样</span></span><span leaf=""><br></span><span leaf="">        group_level = c(</span><span><span leaf="">'CTRL'</span></span><span leaf="">, </span><span><span leaf="">'STIM'</span></span><span leaf="">), </span><span><span leaf=""># 和前面设置一样</span></span><span leaf=""><br></span><span leaf="">        min_cells = </span><span><span leaf="">10</span></span><span leaf="">, </span><span><span leaf=""># 和前面设置一样</span></span><span leaf=""><br></span><span leaf="">        tar_genes = </span><span><span leaf="">NULL</span></span><span leaf="">, </span><span><span leaf="">#如果设定为NULL则取前三十个基因</span></span><span leaf=""><br></span><span leaf="">         output_dir =  Genes_output_dir                        </span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes0sKIFJcktk2M406YtGxQ5XWxJuJibaTy7DickAOu90Y5c6cgLDLMH9HaQ/640?wx_fmt=png&amp;from=appmsg" alt="image-20250218192233111" data-ratio="0.4564814814814815" data-type="png" data-w="1080" data-imgfileid="100002074" src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes0sKIFJcktk2M406YtGxQ5XWxJuJibaTy7DickAOu90Y5c6cgLDLMH9HaQ/640?wx_fmt=png&amp;from=appmsg"></span><figcaption><span leaf="">image-20250218192233111</span></figcaption></figure><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes0lxiaYFSXOOejQnsOysqzD637EHWjahmJrLmBJrygxhwULpiacXNYdH7w/640?wx_fmt=png&amp;from=appmsg" alt="image-20250218192316424" data-ratio="1.0019685039370079" data-type="png" data-w="1016" data-imgfileid="100002073" src="https://mmbiz.qpic.cn/sz_mmbiz_png/9LgsctS2IfFpKia4trdVr5wH3jImQKes0lxiaYFSXOOejQnsOysqzD637EHWjahmJrLmBJrygxhwULpiacXNYdH7w/640?wx_fmt=png&amp;from=appmsg"></span><figcaption><span leaf="">image-20250218192316424</span></figcaption></figure><p data-tool="mdnice编辑器"><span leaf="">确认方向，符合文件夹</span><code><span leaf="">01.DEseq2_pairs</span></code><span leaf=""> 的结果</span></p><p data-tool="mdnice编辑器"><span leaf="">以下是函数链接 （付费前请确认具有一定的R语言基础），</span></p><p data-tool="mdnice编辑器"><span leaf="">后续更新功能会通过留言板或者链接更新</span></p><p><mp-pay-preview-filter data-offset="94"></mp-pay-preview-filter></p></section></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/sgXJZBZ5Mtbgr2FHo8-ulg",target="_blank" rel="noopener noreferrer">原文链接</a>
