---
title: "å¯»æ‰¾motifï¼ŒRç‰ˆæœ¬çš„MEMEæ¥äº†ï¼Œä¸è¯•è¯•ï¼Ÿ"
date: 2023-01-06T03:15:17Z
draft: ["false"]
tags: [
  "fetched",
  "ç”Ÿä¿¡æ˜Ÿçƒ"
]
categories: ["Acdemic"]
---
å¯»æ‰¾motifï¼ŒRç‰ˆæœ¬çš„MEMEæ¥äº†ï¼Œä¸è¯•è¯•ï¼Ÿ by ç”Ÿä¿¡æ˜Ÿçƒ
------
<div><p><img data-ratio="1" data-src="https://mmbiz.qpic.cn/mmbiz_png/8oKPbJgbBHrLfDiavqTPmkMHiaIAkPgmZR9LERh1l5td8g2LicMYNcCHjh1ibIxqJys9neR8ibDH1IAzcxia9jrLP1tw/640?wx_fmt=png" data-type="png" data-w="20" width="20px" src="https://mmbiz.qpic.cn/mmbiz_png/8oKPbJgbBHrLfDiavqTPmkMHiaIAkPgmZR9LERh1l5td8g2LicMYNcCHjh1ibIxqJys9neR8ibDH1IAzcxia9jrLP1tw/640?wx_fmt=png"><img data-ratio="1" data-src="https://mmbiz.qpic.cn/mmbiz_png/8oKPbJgbBHrLfDiavqTPmkMHiaIAkPgmZRXmgtHtmB3L1qaPWfQYBIXPXWsv4xSThfiaveuJxFcribXCVia6S1Q5OWw/640?wx_fmt=png" data-type="png" data-w="20" width="20px" src="https://mmbiz.qpic.cn/mmbiz_png/8oKPbJgbBHrLfDiavqTPmkMHiaIAkPgmZRXmgtHtmB3L1qaPWfQYBIXPXWsv4xSThfiaveuJxFcribXCVia6S1Q5OWw/640?wx_fmt=png"><span>Â ä»Šå¤©æ˜¯ç”Ÿä¿¡æ˜Ÿçƒé™ªä½ çš„<span><strong>ç¬¬809å¤©</strong></span></span><img data-ratio="1" data-src="https://mmbiz.qpic.cn/mmbiz_png/8oKPbJgbBHrLfDiavqTPmkMHiaIAkPgmZR9LERh1l5td8g2LicMYNcCHjh1ibIxqJys9neR8ibDH1IAzcxia9jrLP1tw/640?wx_fmt=png" data-type="png" data-w="20" width="20px" src="https://mmbiz.qpic.cn/mmbiz_png/8oKPbJgbBHrLfDiavqTPmkMHiaIAkPgmZR9LERh1l5td8g2LicMYNcCHjh1ibIxqJys9neR8ibDH1IAzcxia9jrLP1tw/640?wx_fmt=png"><br></p><hr><p><span><span>Â  Â </span><span>å¤§ç¥ä¸€å¥è¯ï¼Œèœé¸Ÿè·‘åŠå¹´ã€‚æˆ‘ä¸æ˜¯å¤§ç¥ï¼Œä½†æˆ‘å¯ä»¥ç¼©çŸ­ä½ èµ°å¼¯è·¯çš„åŠå¹´~</span></span></p><p><span><span>Â  Â å°±åƒæ­Œå„¿å”±çš„é‚£æ ·ï¼Œå¦‚æœä½ ä¸çŸ¥é“è¯¥å¾€å“ªå„¿èµ°ï¼Œå°±ç•™åœ¨è¿™å­¦ç‚¹ç”Ÿä¿¡å¥½ä¸å¥½~</span></span></p><p><span>Â  Â è¿™é‡Œæœ‰è±†è±†å’ŒèŠ±èŠ±çš„å­¦ä¹ å†ç¨‹ï¼Œä»æ–°æ‰‹åˆ°è¿›é˜¶ï¼Œç”Ÿä¿¡è·¯ä¸Šæœ‰ä½ æœ‰æˆ‘ï¼</span></p><section><blockquote><p>è±†è±†å†™äº2021.4.28<br>ä¼šä»‹ç»MEMEçš„ç”¨æ³•ï¼Œä»¥åŠæ–°çš„Rç‰ˆæœ¬ä½¿ç”¨ï¼Œæœ€åçœ‹çœ‹å…¶ä»–ç±»ä¼¼å·¥å…·</p></blockquote><h3><span>ä»€ä¹ˆæ˜¯MEMEï¼Ÿ</span></h3><p>çœ‹ä»–çš„å®˜ç½‘ï¼šhttp://meme-suite.org/index.htmlï¼Œå¾ˆæ¸…æ¥šå†™ç€åŸºäºmotifè¿›è¡Œçš„ä¸€ç³»åˆ—åˆ†æã€‚</p><blockquote><p>åœ¨å®˜ç½‘ä¸­æ‰¾åˆ°çš„motifå®šä¹‰ï¼š<br>A motif is an approximate sequence pattern that occurs repeatedly in a group of related sequences.</p><p>å…³é”®è¯ï¼šå¤§é‡é‡å¤å‡ºç°ï¼›çŸ­åºåˆ—ï¼Œåªæ˜¯ä¸€æ®µpattern</p></blockquote><p>å½“ç„¶ï¼Œè¿™æ®µåºåˆ—ä¸å±€é™äºDNAï¼Œè¿˜å¯ä»¥æ˜¯RNAæˆ–è€…æ°¨åŸºé…¸åºåˆ—ï¼›å¹¶ä¸”å®ƒç»å¸¸æ˜¯å…·æœ‰åºåˆ—ç‰¹å¼‚æ€§çš„ç»“åˆä½ç‚¹ï¼ˆæ¯”å¦‚è½¬å½•å› å­ç»“åˆä½ç‚¹æˆ–è€…RNAå‰ªåˆ‡ä½ç‚¹ç­‰ï¼‰</p><p><img data-galleryid="" data-ratio="1.0331825037707392" data-s="300,640" data-src="https://mmbiz.qpic.cn/mmbiz_png/8oKPbJgbBHpp5Au8nialxzG7HZ9DBuzErTX0UzPzdEva18r0BPHd6zj9TlCKjy9ywBXLOFQ9UrzXU9FI6kRfVkA/640?wx_fmt=png" data-type="png" data-w="1326" src="https://mmbiz.qpic.cn/mmbiz_png/8oKPbJgbBHpp5Au8nialxzG7HZ9DBuzErTX0UzPzdEva18r0BPHd6zj9TlCKjy9ywBXLOFQ9UrzXU9FI6kRfVkA/640?wx_fmt=png"></p><p>å·¦ä¾§åˆ—è¡¨å°±å†™å‡ºäº†å®ƒçš„<strong>äº”å¤§åŠŸèƒ½</strong>ï¼š<br></p><ul><li><p><strong>Motif Discovery</strong>ï¼šé¢„æµ‹åºåˆ—ä¸Šçš„motifä¿¡æ¯ï¼Œå¯ä»¥æ˜¯DNAï¼ŒRNAï¼Œæ°¨åŸºé…¸åºåˆ—ï¼Œæ–¹æ³•æ˜¯ä»å¤´é¢„æµ‹ï¼ˆde novo discoveryï¼‰ã€‚æ¯”å¦‚æˆ‘ä»¬ChIPseqæ•°æ®æ‹¿åˆ°äº†ä¸€äº›peaksçš„åŒºé—´ï¼Œé‚£ä¹ˆå°±èƒ½æ‹¿åˆ°åºåˆ—ï¼Œç„¶åå°±èƒ½æ‹¿åˆ°åºåˆ—ä¸­çš„motif</p></li><ul><li><p><strong>MEME</strong>: Multiple EM for Motif Elicitation, åˆ©ç”¨de novoé¢„æµ‹å¾—åˆ°motifç»“æ„å’Œåœ¨è¾“å…¥åºåˆ—ä¸Šçš„ä½ç½®</p></li><li><p><strong>DREME</strong>: Discriminative Regular Expression Motif Elicitation, ä¸»è¦é’ˆå¯¹é•¿åº¦åœ¨100bpå·¦å³çš„åºåˆ—</p></li><li><p><strong>STREME</strong>: Sensitive, Thorough, Rapid, Enriched Motif Elicitationï¼Œå®ƒæ˜¯DREMEçš„æ›¿ä»£ç‰ˆï¼Œç‰¹ç‚¹æ˜¯åŸºäºçš„åºåˆ—æ•°é‡å¤šï¼ˆå¤šäº50ä¸ªï¼‰</p></li><li><p><span>MEME-ChIP</span></p></li><li><p><span>GLAM2</span></p></li><li><p><span>MoMo</span></p></li></ul><li><p><strong>Motif Enrichment</strong>ï¼šåˆ†æå·²çŸ¥çš„motifåœ¨æˆ‘ä»¬çš„åºåˆ—ä¸Šæœ‰æ²¡æœ‰ï¼Œæœ‰å¤šå°‘ï¼Ÿ</p></li><ul><li><p><span>CentriMo</span></p></li><li><p><span>AME</span></p></li><li><p><span>AME</span></p></li><li><p><span>SpaMo</span></p></li></ul><li><p><strong>Motif Scanning</strong>ï¼šåˆ†ææˆ‘ä»¬åºåˆ—ä¸Šå¯èƒ½å‡ºç°motifçš„ä½ç½®æ˜¯ï¼Ÿ</p></li><ul><li><p><span>FIMO</span></p></li><li><p><span>MAST</span></p></li><li><p><span>MCAST</span></p></li></ul><li><p><strong>Motif Comparison</strong>ï¼šä¸åŒmotifä¹‹é—´çš„ç›¸ä¼¼æ€§æœ‰å¤šå°‘ï¼Ÿ</p></li><ul><li><p><span>Tomtom</span></p></li></ul><li><p><strong>Gene Regulation</strong>ï¼šé¢„æµ‹åŸºå› å’Œè°ƒæ§å…ƒä»¶ï¼ˆç”¨æŸ“è‰²ä½“åŒºåŸŸè¡¨ç¤ºï¼‰ä¹‹é—´çš„è°ƒæ§å…³ç³»</p></li></ul><p><img data-galleryid="" data-ratio="0.5110782865583456" data-s="300,640" data-src="https://mmbiz.qpic.cn/mmbiz_png/8oKPbJgbBHpp5Au8nialxzG7HZ9DBuzEr3UkVQclFXsBQb8WnIqFEa7OrBHD40oIkn93W2icqUlkBAk6GVnfZTJA/640?wx_fmt=png" data-type="png" data-w="1354" src="https://mmbiz.qpic.cn/mmbiz_png/8oKPbJgbBHpp5Au8nialxzG7HZ9DBuzEr3UkVQclFXsBQb8WnIqFEa7OrBHD40oIkn93W2icqUlkBAk6GVnfZTJA/640?wx_fmt=png"></p><p>ä½†å®ƒçš„ç¼ºç‚¹å°±æ˜¯ï¼šå®ƒä½œä¸ºä¸€ä¸ªç½‘é¡µå·¥å…·ï¼Œä½†ç•¥æ˜¾ç¹çï¼Œé™åˆ¶äº†ä½¿ç”¨äººç¾¤</p><h3><span>æ–°çš„Rç‰ˆæœ¬æ€ä¹ˆç”¨ï¼Ÿ</span></h3><blockquote><p>å®˜æ–¹æ–‡æ¡£åœ¨ï¼šhttps://snystrom.github.io/memes-manual/</p></blockquote><p>ä»Šå¤©çœ‹åˆ°ä¸€ä¸ªæ¨æ–‡ï¼Œå‘ç°æœ‰äº†Rç‰ˆæœ¬çš„æ”¯æŒï¼Œä½œè€…è¿˜å‘äº†ç¯‡æ–‡ç« ï¼š<br>Memes: an R interface to the MEME Suite<br>ï¼ˆhttps://www.biorxiv.org/content/10.1101/2021.04.23.441089v1ï¼‰</p><blockquote><p>In addition to interfacing with popular MEME Suite tools, <strong>memes leverages existing R/Bioconductor data structures</strong> to store the complex, multidimensional data returned by MEME Suite tools for rapid data access and manipulation.</p></blockquote><figure><img data-ratio="1.154103852596315" data-src="https://mmbiz.qpic.cn/mmbiz_png/8oKPbJgbBHpp5Au8nialxzG7HZ9DBuzErNgS08kRkwog48PPRC6mKYoKRuwaFAzUwibKmBNm3z74LoIIwdMJ1eLw/640?wx_fmt=png" data-type="png" data-w="1194" title="" src="https://mmbiz.qpic.cn/mmbiz_png/8oKPbJgbBHpp5Au8nialxzG7HZ9DBuzErNgS08kRkwog48PPRC6mKYoKRuwaFAzUwibKmBNm3z74LoIIwdMJ1eLw/640?wx_fmt=png"></figure><h4><span>é¦–å…ˆå®‰è£…RåŒ…</span></h4><p>ç°åœ¨è¿™ä¸ªRåŒ…åº”è¯¥è¿˜æ˜¯åœ¨å¼€å‘ç‰ˆï¼Œéœ€è¦R 4.1ä»¥ä¸Šæ‰å¯ä»¥ã€‚æˆ‘äºæ˜¯ä¸‹è½½äº†R 4.2å’Œ4.1å®‰è£…è¯•ä¸€ä¸‹ï¼Œ<strong>ç»“æœå¤±è´¥ğŸ˜</strong></p><p>ä¸è¿‡æ²¡å…³ç³»ï¼Œå…ˆæŠŠå®‰è£…ä»£ç æ”¾åœ¨è¿™é‡Œï¼Œæ¯•ç«Ÿè¿‡äº›æ—¥å­å‡ºäº†æ­£å¼ç‰ˆ4.1å°±å¯ä»¥äº† <span>ï¼ˆR 4.1 scheduled for Tuesday 2021-05-18ï¼‰</span></p><pre><code><span>#Â éœ€è¦ï¼šBioconductor version: Development (3.13)</span><br><span>#Â BioconductorÂ versionÂ '3.13'Â requiresÂ RÂ versionÂ '4.1'</span><br><br><span>#Â ä½¿ç”¨RÂ 4.1Â åŠä»¥ä¸Šç‰ˆæœ¬</span><br><span>if</span>Â (!requireNamespace(<span>"remotes"</span>,Â quietly=<span>TRUE</span>))<br>Â Â Â Â install.packages(<span>"remotes"</span>)<br>remotes::install_github(<span>"snystrom/memes"</span>)<br></code></pre><h4><span>RåŒ…éœ€è¦æœ¬åœ°MEMEå¥—ä»¶ï¼ˆfor Linux, OS X and Cygwin systems)</span></h4><blockquote><p>ä¸å¾—ä¸è¯´ï¼Œå®‰è£…è¿‡ç¨‹éƒ½è¿™ä¹ˆç¹çï¼Œç¡®å®å½±å“äº†ä½¿ç”¨è€…å¿ƒæƒ…</p></blockquote><pre><code><span>#</span><span>Â ä¸‹è½½ï¼ŒÂ åœ¨Â https://meme-suite.org/doc/download.html/Â æ‰¾ç‰ˆæœ¬</span><br><span>#</span><span>Â å®‰è£…</span><br>tarÂ zxfÂ meme-5.3.3.tar.gz<br>cdÂ meme-5.3.3<br>./configureÂ --prefix=$HOME/memeÂ --enable-build-libxml2Â --enable-build-libxslt<br>make<br>makeÂ test<br>makeÂ install<br><span>#</span><span>Â æœ€åæ·»åŠ ç¯å¢ƒå˜é‡</span><br>exportÂ PATH=$HOME/meme/bin:$HOME/meme/libexec/meme-5.3.3:$PATH<br><span><br>#</span><span>Â è¿™ä¸ªå€’æ˜¯å®‰è£…æˆåŠŸäº†ï¼Œè€—æ—¶çº¦10min</span><br></code></pre><p>å¦‚æœä¹‹åå®‰è£…æˆåŠŸäº†ï¼Œé‚£ä¹ˆå°±ç»§ç»­è¿è¡Œï¼š</p><h4><span>å…ˆæ¥ä¸€ä¸ªæ£€æµ‹æ­¥éª¤</span></h4><pre><code><span>library</span>(memes)<br>check_meme_install()<br><span>#&gt;Â checkingÂ mainÂ install</span><br><span>#&gt;Â âœ“Â /opt/meme/bin</span><br><span>#&gt;Â checkingÂ utilÂ installs</span><br><span>#&gt;Â âœ“Â /opt/meme/bin/dreme</span><br><span>#&gt;Â âœ“Â /opt/meme/bin/ame</span><br><span>#&gt;Â âœ“Â /opt/meme/bin/fimo</span><br><span>#&gt;Â âœ“Â /opt/meme/bin/tomtom</span><br><span>#&gt;Â âœ“Â /opt/meme/bin/meme</span><br><span>#Â æˆ–è€…å¯ä»¥æŒ‡å®šè·¯å¾„</span><br>check_meme_install(meme_pathÂ =Â <span>'bad/path'</span>)<br></code></pre><p>æ ¸å¿ƒçš„å‡½æ•°æœ‰ï¼š</p><figure><img data-ratio="0.3624413145539906" data-src="https://mmbiz.qpic.cn/mmbiz_png/8oKPbJgbBHpp5Au8nialxzG7HZ9DBuzErDwDA3icquiaSY0qdY4mlFFbSSgFabDM4FHXe9ScRCN5OzH2qJjQicQ15w/640?wx_fmt=png" data-type="png" data-w="2130" title="" src="https://mmbiz.qpic.cn/mmbiz_png/8oKPbJgbBHpp5Au8nialxzG7HZ9DBuzErDwDA3icquiaSY0qdY4mlFFbSSgFabDM4FHXe9ScRCN5OzH2qJjQicQ15w/640?wx_fmt=png"></figure><h4><span>DREMEå¯»æ‰¾motif</span></h4><pre><code>suppressPackageStartupMessages(<span>library</span>(magrittr))<br>suppressPackageStartupMessages(<span>library</span>(GenomicRanges))<br><br><span>#Â æ‹¿åˆ°peaks</span><br>data(<span>"example_peaks"</span>,Â packageÂ =Â <span>"memes"</span>)<br><span>#Â æ‹¿åˆ°åŸºå› ç»„</span><br>dm.genomeÂ &lt;-Â BSgenome.Dmelanogaster.UCSC.dm6::BSgenome.Dmelanogaster.UCSC.dm6<br><br><span>#Â GenerateÂ sequencesÂ fromÂ 200bpÂ aboutÂ theÂ centerÂ ofÂ myÂ peaksÂ ofÂ interest</span><br>sequencesÂ &lt;-Â example_peaksÂ %&gt;%Â <br>Â Â resize(<span>200</span>,Â <span>"center"</span>)Â %&gt;%Â <br>Â Â get_sequence(dm.genome)<br><br><span>#Â æœ€åè¿è¡Œdreme</span><br><span>#Â runDreme()Â acceptsÂ XStringSetÂ orÂ aÂ pathÂ toÂ aÂ fastaÂ fileÂ asÂ input.</span><br><span>#Â InÂ aÂ realÂ analysis,Â eÂ shouldÂ typicallyÂ beÂ &lt;Â 1</span><br>dreme_resultsÂ &lt;-Â runDreme(sequences,Â controlÂ =Â <span>"shuffle"</span>,Â eÂ =Â <span>50</span>)<br><br><span>#Â å±•ç¤ºç»“æœ</span><br><span>library</span>(universalmotif)<br>dreme_resultsÂ %&gt;%Â <br>Â Â to_list()Â %&gt;%Â <br>Â Â view_motifs()<br></code></pre><figure><img data-ratio="0.7142857142857143" data-src="https://mmbiz.qpic.cn/mmbiz_png/8oKPbJgbBHpp5Au8nialxzG7HZ9DBuzErbHrAIOBNJt9u4tnUNXLqxMe6TCx0FdgP2zhSr45cURfBEyv9iatT3pw/640?wx_fmt=png" data-type="png" data-w="672" title="" src="https://mmbiz.qpic.cn/mmbiz_png/8oKPbJgbBHpp5Au8nialxzG7HZ9DBuzErbHrAIOBNJt9u4tnUNXLqxMe6TCx0FdgP2zhSr45cURfBEyv9iatT3pw/640?wx_fmt=png"></figure><h4><span>TOMTOMæ¯”è¾ƒmotif</span></h4><pre><code><span>#Â DiscoveredÂ motifsÂ canÂ beÂ matchedÂ toÂ knownÂ TFÂ motifsÂ usingÂ runTomTom()</span><br><br><span>#Â TomTomÂ usesÂ aÂ databaseÂ ofÂ knownÂ motifsÂ whichÂ canÂ beÂ passedÂ toÂ theÂ databaseÂ parameterÂ asÂ aÂ pathÂ toÂ aÂ .memeÂ formatÂ file,Â orÂ aÂ universalmotifÂ object.</span><br><span>#Â å¯ä»¥æŒ‡å®šæ•°æ®åº“ï¼š</span><br><span>#Â options(meme_dbÂ =Â system.file("extdata/flyFactorSurvey_cleaned.meme",Â packageÂ =Â "memes"))</span><br><br>full_resultsÂ &lt;-Â dreme_resultsÂ %&gt;%Â <br>Â Â runTomTom()<br><span>#Â å¯è§†åŒ–ç»“æœ</span><br>full_resultsÂ %&gt;%Â <br>Â Â view_tomtom_hits(top_nÂ =Â <span>1</span>)<br></code></pre><figure><img data-ratio="0.7142857142857143" data-src="https://mmbiz.qpic.cn/mmbiz_png/8oKPbJgbBHpp5Au8nialxzG7HZ9DBuzErdRSib3GzibZuWscBYGTfibo3ALA58gDTUYrGwXXt2UKX0vzqjadEdsSOA/640?wx_fmt=png" data-type="png" data-w="672" title="" src="https://mmbiz.qpic.cn/mmbiz_png/8oKPbJgbBHpp5Au8nialxzG7HZ9DBuzErdRSib3GzibZuWscBYGTfibo3ALA58gDTUYrGwXXt2UKX0vzqjadEdsSOA/640?wx_fmt=png"></figure><h4><span>AMEå¯Œé›†motif</span></h4><pre><code><span>#Â AMEÂ isÂ usedÂ toÂ testÂ forÂ enrichmentÂ ofÂ knownÂ motifsÂ inÂ targetÂ sequences</span><br><span>#Â åŒæ ·ä¹Ÿæ˜¯éœ€è¦æ•°æ®åº“çš„</span><br><span>#Â options(meme_dbÂ =Â "path/to/database.meme")</span><br><br>ame_resultsÂ &lt;-Â runAme(sequences,Â controlÂ =Â <span>"shuffle"</span>,Â evalue_report_thresholdÂ =Â <span>30</span>)<br>ame_results<br><span>#&gt;Â #Â AÂ tibble:Â 2Â xÂ 17</span><br><span>#&gt;Â Â Â Â rankÂ motif_dbÂ motif_idÂ motif_alt_idÂ consensusÂ Â pvalueÂ adj.pvalueÂ evalueÂ tests</span><br><span>#&gt;Â Â Â &lt;int&gt;Â &lt;chr&gt;Â Â Â Â &lt;chr&gt;Â Â Â Â &lt;chr&gt;Â Â Â Â Â Â Â Â &lt;chr&gt;Â Â Â Â Â Â Â &lt;dbl&gt;Â Â Â Â Â Â &lt;dbl&gt;Â Â &lt;dbl&gt;Â &lt;int&gt;</span><br><span>#&gt;Â 1Â Â Â Â Â 1Â /usr/loâ€¦Â Eip93F_â€¦Â Eip93FÂ Â Â Â Â Â Â ACWSCCRAâ€¦Â 5.14e-4Â Â Â Â Â 0.0339Â Â Â 18.8Â Â Â Â 67</span><br><span>#&gt;Â 2Â Â Â Â Â 2Â /usr/loâ€¦Â Cf2-PB_â€¦Â Cf2Â Â Â Â Â Â Â Â Â Â CSSHNKDTâ€¦Â 1.57e-3Â Â Â Â Â 0.0415Â Â Â 23.1Â Â Â Â 27</span><br><span>#&gt;Â #Â â€¦Â withÂ 8Â moreÂ variables:Â fasta_maxÂ &lt;dbl&gt;,Â posÂ &lt;int&gt;,Â negÂ &lt;int&gt;,</span><br><span>#&gt;Â #Â Â Â pwm_minÂ &lt;dbl&gt;,Â tpÂ &lt;int&gt;,Â tp_percentÂ &lt;dbl&gt;,Â fpÂ &lt;int&gt;,Â fp_percentÂ &lt;dbl&gt;</span><br><br><span>#Â å¯è§†åŒ–ç»“æœ</span><br>ame_resultsÂ %&gt;%Â <br>Â Â plot_ame_heatmap()<br></code></pre><figure><img data-ratio="0.6" data-src="https://mmbiz.qpic.cn/mmbiz_png/8oKPbJgbBHpp5Au8nialxzG7HZ9DBuzErfcgJAViaQbVKZgTgzavULR7QlGibLDFlu02KJxpYhLzibeTiaRFxic2InZg/640?wx_fmt=png" data-type="png" data-w="480" title="" src="https://mmbiz.qpic.cn/mmbiz_png/8oKPbJgbBHpp5Au8nialxzG7HZ9DBuzErfcgJAViaQbVKZgTgzavULR7QlGibLDFlu02KJxpYhLzibeTiaRFxic2InZg/640?wx_fmt=png"></figure><h4><span>FIMOåˆ†æmotifå‡ºç°çš„ä½ç½®</span></h4><p><code>MotifDb</code>çš„ä½¿ç”¨æ–¹æ³•:<br>https://bioconductor.org/packages/release/bioc/vignettes/MotifDb/inst/doc/MotifDb.html ä»¥åŠ https://rdrr.io/bioc/MotifDb/man/MotifDb.html</p><pre><code><span>#</span><span>Â MotifDb</span><br><span><br>#</span><span>Â TheÂ matricesÂ areÂ obtainedÂ fromÂ sixÂ publicÂ sources:</span><br>FlyFactorSurvey:Â Â Â Â 614<br>hPDI:Â Â Â Â 437<br>JASPAR_CORE:Â Â Â Â 459<br>jolma2013:Â Â Â Â 843<br>ScerTF:Â Â Â Â 196<br>stamlab:Â Â Â Â 683<br>UniPROBE:Â Â Â Â 380<br>cisbpÂ 1.02Â Â Â Â 874<br><span><br>#</span><span>Â RepresentingÂ primarilyÂ fiveÂ organsismsÂ (andÂ 49Â total)</span><br>Hsapiens:Â Â Â Â 2328<br>Dmelanogaster:Â Â Â Â 1008<br>Scerevisiae:Â Â Â Â 701<br>Mmusculus:Â Â Â Â 660<br>Athaliana:Â Â Â Â 160<br>Celegans:Â Â Â Â 44<br>other:Â Â Â Â 177<br></code></pre><p>ç»§ç»­çœ‹FIMOçš„ä½¿ç”¨ï¼š</p><pre><code><span>#Â QueryÂ MotifDbÂ forÂ aÂ motif</span><br>e93_motifÂ &lt;-Â MotifDb::query(MotifDb::MotifDb,Â <span>"Eip93F"</span>)Â %&gt;%Â <br>Â Â universalmotif::convert_motifs()<br><br><span>#Â ScanÂ forÂ theÂ E93Â motifÂ withinÂ givenÂ sequences</span><br>fimo_resultsÂ &lt;-Â runFimo(sequences,Â e93_motif,Â threshÂ =Â <span>1e-3</span>)<br><br><span>#Â å¯è§†åŒ–</span><br>plot_sequence_heatmap(fimo_results$matched_sequence)Â Â <br></code></pre><figure><img data-ratio="1.3333333333333333" data-src="https://mmbiz.qpic.cn/mmbiz_png/8oKPbJgbBHpp5Au8nialxzG7HZ9DBuzEr2789d6iaoJTumJp4v320QicerXmyXby8cG7RPicBBApzR2kSlfeQ8wQSQ/640?wx_fmt=png" data-type="png" data-w="288" title="" src="https://mmbiz.qpic.cn/mmbiz_png/8oKPbJgbBHpp5Au8nialxzG7HZ9DBuzEr2789d6iaoJTumJp4v320QicerXmyXby8cG7RPicBBApzR2kSlfeQ8wQSQ/640?wx_fmt=png"></figure><h4><span>è¿˜æ”¯æŒå¯¼å…¥MEMEç½‘é¡µå·¥å…·çš„ç»“æœ</span></h4><figure><img data-ratio="0.594758064516129" data-src="https://mmbiz.qpic.cn/mmbiz_png/8oKPbJgbBHpp5Au8nialxzG7HZ9DBuzErWqiaNZDKlBuoaz7mvOu4ficrBGTNHBAbibannb2YHR2zaTia8J9KJDVHjw/640?wx_fmt=png" data-type="png" data-w="992" title="" src="https://mmbiz.qpic.cn/mmbiz_png/8oKPbJgbBHpp5Au8nialxzG7HZ9DBuzErWqiaNZDKlBuoaz7mvOu4ficrBGTNHBAbibannb2YHR2zaTia8J9KJDVHjw/640?wx_fmt=png"></figure><h3><span>æœ€åï¼Œå·¥å…·ç»å¯¹ä¸æ­¢ä¸€ç§</span></h3><p>æ¯”å¦‚motifåˆ†æï¼Œhttp://veda.cs.uiuc.edu/MET/links.htmlè¿™é‡Œåˆ—å‡ºäº†ä¸€äº›ï¼š</p><figure><img data-ratio="0.3595608417200366" data-src="https://mmbiz.qpic.cn/mmbiz_png/8oKPbJgbBHpp5Au8nialxzG7HZ9DBuzErnfYX1VOkpaGSKiaZ2aJGEDFvR1HKXhyt6jKuT9TjE96jF9z6gar1pHQ/640?wx_fmt=png" data-type="png" data-w="2186" title="" src="https://mmbiz.qpic.cn/mmbiz_png/8oKPbJgbBHpp5Au8nialxzG7HZ9DBuzErnfYX1VOkpaGSKiaZ2aJGEDFvR1HKXhyt6jKuT9TjE96jF9z6gar1pHQ/640?wx_fmt=png"></figure><p>å¦å¤–ï¼Œ<strong>å¸¸ç”¨çš„homerä¹Ÿæ”¯æŒ</strong>ï¼Œä¾‹å¦‚ä¹‹å‰æŠ€èƒ½æ ‘å‘å¸ƒçš„ï¼šHomerè½¯ä»¶çš„ä»‹ç»-æœ€å…¨é¢è€Œè¯¦ç»†çš„æ‰¾motifæ•™ç¨‹</p><pre><code><span>#</span><span>Â å®‰è£…æ–¹æ³•1ï¼š</span><br>condaÂ installÂ -cÂ biocondaÂ homer<br><span>#</span><span>Â è¿™ä¸ªè·¯å¾„éœ€è¦è‡ªå·±ä¿®æ”¹</span><br>perlÂ ~/miniconda3/pkgs/homer-4.10-pl526hc9558a2_0/share/homer-4.10-0/configureHomer.plÂ -installÂ hg19<br><span>#</span><span>Â æœ€åæ•°æ®ä¸‹è½½åˆ°</span><br>lsÂ ~/miniconda3/pkgs/homer-4.10-pl526hc9558a2_0/share/homer-4.10-0/data/<br><span><br>#</span><span><span>#Â å®‰è£…æ–¹æ³•2ï¼š</span></span><br>cdÂ ~/biosoft<br>mkdirÂ homerÂ &amp;&amp;Â Â cdÂ homer<br>wgetÂ http://homer.salk.edu/homer/configureHomer.plÂ <br>perlÂ configureHomer.plÂ -installÂ hg19<br><span><br>#</span><span>Â å‡†å¤‡æ•°æ®:Â ç¬¬ä¸€åˆ—:Â æŸ“è‰²ä½“;Â ç¬¬äºŒåˆ—:Â èµ·å§‹ä½ç½®;Â ç¬¬ä¸‰åˆ—:Â ç»ˆæ­¢ä½ç½®;Â ç¬¬å››åˆ—:Â é“¾çš„æ–¹å‘</span><br>awkÂ '{printÂ $4"Â Â Â Â "$1"Â Â Â "$2"Â Â Â "$3"Â Â Â +"}'Â macs_peaks.bedÂ &gt;sample_homer.bed<br><span><br>#</span><span>Â è¿è¡Œ(åœ¨åŸºå› ç»„åŒºåŸŸä¸­å¯»æ‰¾å¯Œé›†Motifs)</span><br>findMotifsGenome.plÂ sample_homer.bedÂ hg19Â sample_homer_motifDirÂ -lenÂ 8,10,12<br><span>#</span><span>Â å‚æ•°è§£é‡Š</span><br><span>#</span><span>Â -è¾“å…¥æ–‡ä»¶ï¼šawkå¤„ç†å¥½çš„Homer Peak/Positions file</span><br><span>#</span><span>Â -å‚è€ƒåŸºå› ç»„ï¼šè¿™é‡Œæ˜¯hg19</span><br><span>#</span><span>Â -è¾“å‡ºæ–‡ä»¶ï¼šç»™ä¸€ä¸ªè·¯å¾„å’Œè¾“å‡ºæ–‡ä»¶çš„åå­—</span><br><span>#</span><span>Â -lenï¼šmotifå¤§å°è®¾ç½®ï¼Œé»˜è®¤8,10,12ï¼›è¶Šå¤§éœ€è¦çš„è®¡ç®—èµ„æºè¶Šå¤š</span><br></code></pre><p><strong>å“ˆä½›çš„è¯¾ç¨‹</strong>é€‰æ‹©äº†MEMEçš„<code>DREME</code>å·¥å…·ï¼šhttps://hbctraining.github.io/Intro-to-ChIPseq/lessons/12_functional_analysis.html</p><blockquote><p>DREME is a motif discovery algorithm designed to find short, core DNA-binding motifs of eukaryotic transcription factors and is optimized to handle large ChIP-seq data sets.<br>DREME is tailored to eukaryotic data by focusing on <strong>short motifs (4 to 8 nucleotides)</strong> encompassing the DNA-binding region of most eukaryotic monomeric transcription factors.</p></blockquote><p>ç°åœ¨MEMEæ›´æ–°äº†ï¼Œç”¨<code>STREME</code> å–ä»£äº†ä¹‹å‰çš„<code>DREME</code></p><p>è¿˜æœ‰ä¸€ä¸ª<code>MEME-ChIP</code> ï¼š</p><blockquote><p><strong>MEME-ChIP performs DREME and Tomtom analysis</strong> in addition to using tools to assess which motifs are most centrally enriched (motifs should be centered in the peaks) and to combine <strong>related motifs into similarity clusters.</strong> <br>It is able to identify <strong>longer motifs &lt; 30bp</strong>, but takes much longer to run.</p></blockquote><p><strong>å»å¹´è¿˜å‘å¸ƒäº†ä¸€ä¸ªä¸åŸºäºpeaksçš„å·¥å…·</strong>ï¼šNoPeak: k-mer-based motif discovery in ChIP-Seq data without peak calling ï¼ˆBioinformatics, btaa845, https://doi.org/10.1093/bioinformatics/btaa845ï¼Œ 29 September 2020ï¼‰</p><blockquote><p>Here, we present a novel approach to reliably identify transcription factor-binding motifs from ChIP-Seq data <strong>without peak detection</strong>. By evaluating the <strong>distributions of sequencing reads around the different k-mers in the genome,</strong> we are able to identify binding motifs in ChIP-Seq data that yield no results in traditional pipelines.</p></blockquote><p>è¿˜æœ‰å…¶ä»–<strong>åŸºäºRçš„å·¥å…·</strong>ï¼š</p><p>rGADEM (motif discovery):<br>Â http://bioconductor.org/packages/devel/bioc/html/rGADEM.html</p><p>MotIV (motif validation):<br>Â http://bioconductor.org/packages/devel/bioc/html/MotIV.html</p></section><section><hr></section><p><strong><span>ç‚¹å‡»åº•éƒ¨çš„â€œé˜…è¯»åŸæ–‡â€ï¼Œè·å¾—æ›´å¥½çš„é˜…è¯»ä½“éªŒå“¦</span></strong><span>ğŸ˜»</span></p><p><span><strong><span>åˆå­¦ç”Ÿä¿¡ï¼Œå¾ˆè£å¹¸å¸¦ä½ è¿ˆå‡ºç¬¬ä¸€æ­¥</span></strong></span><br></p><p><span><strong>ğŸ¤“</strong></span><strong><span>ç”Ÿä¿¡æ˜ŸçƒÂ </span></strong><span><strong>ğŸŒ</strong></span><span><span><strong>~Â </strong></span><strong><span>ä¸€ä¸ªä¸æ‹½æœ¯è¯­ã€é€šä¿—æ˜“æ‡‚çš„ç”Ÿä¿¡çŸ¥è¯†å¹³å°</span></strong></span></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/BqpzGZQsquwM7T73HEXDOg",target="_blank" rel="noopener noreferrer">åŸæ–‡é“¾æ¥</a>
