---
title: "è‚¿ç˜¤å¼‚è´¨æ€§å¼ºï¼Ÿè¯•è¯•è¿™ä¸ªæ ·æœ¬æ•´åˆåŒ…ï¼"
date: 2023-01-02T08:22:44Z
draft: ["false"]
tags: [
  "fetched",
  "ç”Ÿä¿¡èœé¸Ÿå›¢"
]
categories: ["Acdemic"]
---
è‚¿ç˜¤å¼‚è´¨æ€§å¼ºï¼Ÿè¯•è¯•è¿™ä¸ªæ ·æœ¬æ•´åˆåŒ…ï¼ by ç”Ÿä¿¡èœé¸Ÿå›¢
------
<div><section data-tool="mdniceç¼–è¾‘å™¨" data-website="https://www.mdnice.com"><p data-tool="mdniceç¼–è¾‘å™¨">Poisson scRNA Integration of Mixed Unknown Signalsï¼ˆPRIMUSï¼‰ï¼šæ··æ‚æœªçŸ¥ä¿¡å·çš„æ³Šæ¾scRNAæ•°æ®æ•´åˆ</p><h1 data-tool="mdniceç¼–è¾‘å™¨">èƒŒæ™¯ä»‹ç»</h1><p data-tool="mdniceç¼–è¾‘å™¨">PRIMUS æ¥æºäºä¸€ç¯‡Sci Advçš„æ–‡çŒ®Longitudinal single-cell RNA-seq analysis reveals stress-promoted chemoresistance in metastatic ovarian cancer[1]ã€‚ä½œè€…æœŸæœ›é€šè¿‡ç ”ç©¶åŒ–ç–—å‰ååµå·¢ç™Œæ ·æœ¬çš„å•ç»†èƒè½¬å½•ç»„å±‚é¢çš„å˜åŒ–ï¼Œç„¶è€Œç”±äºåµå·¢ç™Œè‚¿ç˜¤ç‰¹æœ‰çš„å¼‚è´¨æ€§ï¼Œä¸åŒè‚¿ç˜¤ä¹‹é—´ç»†èƒæˆåˆ†å·®å¼‚å¾ˆå¤§ï¼Œå¤šæ ·æœ¬ä¹‹é—´æ•´åˆå­˜åœ¨å›°éš¾ã€‚è€ŒåŸºäºä»¥å¾€çš„Seurat v3, Harmony, LIGER, mnnCorrect, and fastMNN ä»¥åŠä¸‰ç§Bulk RNAseqçš„æ•´åˆæ–¹æ³•ï¼ˆComBatã€ComBat-seqï¼Œand limmaï¼‰æ•´åˆä»ç„¶ä¸èƒ½å¾ˆå¥½çš„è§£å†³è¿™ä¸ªé—®é¢˜ã€‚åŸºäºæ­¤ï¼Œä½œè€…å›¢é˜Ÿå¼€å‘äº†PRIMUSã€‚</p><p data-tool="mdniceç¼–è¾‘å™¨">PRIMUSæ˜¯ä¸€ç§æ•´ä½“èšç±»æ–¹æ³•ï¼Œèƒ½å¤Ÿä» scRNA-seq æ•°æ®ä¸­è¯†åˆ«è¡¨å‹ç»†èƒç¾¤ï¼ŒåŒæ—¶è€ƒè™‘æ—¥æœŸæ¥æº(ä¾‹å¦‚æ‚£è€…ï¼Œæ ·æœ¬ï¼Œæ•°æ®é›†)ç‰¹å®šçš„ç»„åˆ†ä»¥åŠæŠ€æœ¯å™ªå£°ã€‚PRIMUS é‡‡ç”¨åŒçº¿æ€§æ³Šæ¾å›å½’æ¨¡å‹ï¼Œå°†è¡¨è¾¾æ•°æ®åŒæ—¶åˆ†è§£ä¸ºæ˜ç¡®çš„å¹²æ‰°å› ç´ ï¼ˆdefined nuisance factorsï¼‰ã€æœªå®šä¹‰çš„ç»†èƒè¡¨å‹åŠå…¶ç›¸åº”çš„è½¬å½•ç»„å­¦ç‰¹å¾ã€‚</p><blockquote data-tool="mdniceç¼–è¾‘å™¨"><p>[1] Â Zhang K.et al.Longitudinal single-cell RNA-seq analysis reveals stress-promoted chemoresistance in metastatic ovarian cancer. Sci Adv. 2022 Feb 25;8(8):eabm1831. doi: 10.1126/sciadv.abm1831. Epub 2022 Feb 23. PMID: 35196078</p></blockquote><figure data-tool="mdniceç¼–è¾‘å™¨"><img data-ratio="0.7454931071049841" data-src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2LosicLWxG3A6KyWDUKMATWIIicL6PPibcy0x9q5f7anBQXSjPesPGSUwcnLwPeMQuCxsx0reVChqDoFkxQ/640?wx_fmt=png" data-type="png" data-w="943" title="å›¾1 æ–‡çŒ®é™„å›¾S2 A-FÂ  æè¿°PRIMUS çš„æ•´åˆæ•ˆæœ" src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2LosicLWxG3A6KyWDUKMATWIIicL6PPibcy0x9q5f7anBQXSjPesPGSUwcnLwPeMQuCxsx0reVChqDoFkxQ/640?wx_fmt=png"><figcaption><span>å›¾1 æ–‡çŒ®é™„å›¾S2 A-FÂ  æè¿°PRIMUS çš„æ•´åˆæ•ˆæœ</span></figcaption></figure><h1 data-tool="mdniceç¼–è¾‘å™¨">PRIMUS å®‰è£…</h1><p data-tool="mdniceç¼–è¾‘å™¨">PRIMUS å­˜å‚¨åœ¨Githubï¼Œå› æ­¤ï¼Œå¯ä»¥ä½¿ç”¨devtools å®‰è£…ã€‚</p><pre data-tool="mdniceç¼–è¾‘å™¨"><code>devtools::install_github(<span>"KaiyangZ/PRIMUS"</span>)<br></code></pre><p data-tool="mdniceç¼–è¾‘å™¨">Vignette æ–‡æ¡£é“¾æ¥ï¼šhttps://htmlpreview.github.io/?https://github.com/KaiyangZ/PRIMUS/blob/master/vignettes/quickstart.html</p><h1 data-tool="mdniceç¼–è¾‘å™¨">PRIMUS æµç¨‹ä½¿ç”¨</h1><h2 data-tool="mdniceç¼–è¾‘å™¨"><span>æ•°æ®å‡†å¤‡</span><span> </span></h2><p data-tool="mdniceç¼–è¾‘å™¨">åœ¨ä¸ŠæœŸæˆ‘ä»¬è¯´è¿‡ï¼Œä¸€èˆ¬æƒ…å†µä¸‹ä½œè€…éƒ½ä¼šæä¾›åŒ…çš„ä½¿ç”¨æ–¹æ³•åŠç¤ºä¾‹æ•°æ®ï¼Œä½†äº‹æƒ…æ€»æœ‰ä¾‹å¤–ã€‚PRIMUS Â RåŒ…çš„ç¤ºä¾‹æ•°æ®å°±å¦‚æ­¤ã€‚ä½œè€…æä¾›çš„Counts æ–‡ä»¶å’ŒMeta æ–‡ä»¶éƒ½ä¸æ˜åŸå› ä¸¢å¤±äº†ï¼Œè¿™ä¸ªæ—¶å€™å°±éœ€è¦è‡ªå·±ä¸‹è½½å…¶ä»–çš„ç¤ºä¾‹æ–‡ä»¶ã€‚åœ¨è¿™é‡Œï¼Œæˆ‘é€‰æ‹©ä½œè€…æ–‡ç« ä¸­ç”¨åˆ°çš„èƒ°è…ºç™Œæ•°æ®ï¼ˆhttps://github.com/JinmiaoChenLab/Batch-effect-removal-benchmarking/tree/master/Data/dataset4ï¼‰ä½œä¸ºç¤ºä¾‹æ•°æ®è¿›è¡Œæ¼”ç¤ºã€‚</p><p data-tool="mdniceç¼–è¾‘å™¨"><img data-ratio="0.45353982300884954" data-src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2LosicLWxG3A6KyWDUKMATWIIicLeZuE38iaoqTERZROSgPiafjWURs3JnTLibXJ1r4fFjTuJn3CHHlBiciagibQ/640?wx_fmt=png" data-type="png" data-w="452" title="å›¾2Â  Â Methods æˆªå›¾" src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2LosicLWxG3A6KyWDUKMATWIIicLeZuE38iaoqTERZROSgPiafjWURs3JnTLibXJ1r4fFjTuJn3CHHlBiciagibQ/640?wx_fmt=png"><span>å›¾2Â  Â Methods æˆªå›¾</span></p><p data-tool="mdniceç¼–è¾‘å™¨">è€ŒGithub é¡µé¢æ˜¾ç¤ºï¼Œè¯¥æ•°æ®å­˜æ”¾åœ¨Dockerï¼Œ</p><p data-tool="mdniceç¼–è¾‘å™¨"><img data-ratio="0.35853131749460043" data-src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2LosicLWxG3A6KyWDUKMATWIIicL5iaZBELcCMMolwpBebFsibHFA6LMGkuKSFEQplxaSIfYgwprE8vlxphQ/640?wx_fmt=png" data-type="png" data-w="926" title="å›¾3 Github é¡µé¢æˆªå›¾" src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2LosicLWxG3A6KyWDUKMATWIIicL5iaZBELcCMMolwpBebFsibHFA6LMGkuKSFEQplxaSIfYgwprE8vlxphQ/640?wx_fmt=png"><span>å›¾3 Github é¡µé¢æˆªå›¾</span></p><p data-tool="mdniceç¼–è¾‘å™¨">ç‚¹å¼€é“¾æ¥ï¼Œå‘ç°è¿™é‡Œç”¨åˆ°äº†sudo å‘½ä»¤ï¼Œä¸€èˆ¬æƒ…å†µä¸‹ï¼Œæˆ‘ä»¬ä½¿ç”¨çš„linux æœåŠ¡å™¨å¾ˆéš¾æœ‰sudo æƒé™ã€‚åœ¨æˆ‘ç”¨çš„æœåŠ¡å™¨ä¸­ï¼Œsingularity æ›¿ä»£äº†docker çš„ç”¨æ³•ã€‚</p><ul data-tool="mdniceç¼–è¾‘å™¨"><li><section>å› æ­¤ï¼Œè¿™é‡Œå¯ä»¥ä½¿ç”¨singularity è·å–èƒ°è…ºç™Œæ•°æ®ã€‚singularity å­¦ä¹ åœ°å€ï¼ˆhttp://www.xtaohub.com/Container-Tech/Singularity-in-nutshell.htmlï¼‰</section></li></ul><pre data-tool="mdniceç¼–è¾‘å™¨"><code>singularity build --sandbox ./batcheffect  docker://jinmiaochenlab/batch-effect-removal-benchmarking<br>cp  ./batcheffect/batch_effect/dataset4 ./dataset4<br></code></pre><h2 data-tool="mdniceç¼–è¾‘å™¨"><span>scRNA å‰æœŸå¤„ç†</span><span> </span></h2><h3 data-tool="mdniceç¼–è¾‘å™¨">æ•°æ®è¯»å…¥</h3><p data-tool="mdniceç¼–è¾‘å™¨">è¯¥èƒ°è…ºç™Œæ•°æ®æ¥æºäºA benchmark of batch-effect correction methods for single-cell RNA sequencing data" (https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1850-9ï¼Œæ„Ÿå…´è¶£å¯ä»¥é˜…è¯»åŸæ–‡çŒ®ã€‚åœ¨è¿™é‡Œï¼Œæˆ‘ä»¬å…ˆè¯»å…¥æ•°æ®ï¼Œå¹¶è¿›è¡Œè¿›è¡Œå‰æœŸçš„ç®€å•å¤„ç†ï¼Œ</p><pre data-tool="mdniceç¼–è¾‘å™¨"><code><span>library</span>(SingleCellExperiment)<br><span>library</span>(scuttle)<br><span>library</span>(data.table)<br><span>library</span>(dplyr)<br>rm(list = ls())<br><span># è¯»å…¥æ•°æ®</span><br>counts = fread(<span>'./dataset4/myData_pancreatic_5batches.txt.gz'</span>) %&gt;% as.data.frame()<br>rownames(counts) &lt;- counts$V1; counts &lt;- counts[,<span>2</span>:ncol(counts)]<br>dim(counts)<br><span>#&gt; [1] 15558 14767</span><br>head(counts)[<span>1</span>:<span>5</span>,<span>1</span>:<span>2</span>]<br><span>#&gt;        human1_lib1.final_cell_0001 human1_lib1.final_cell_0002</span><br><span>#&gt;  A1BG                             0                           0</span><br><span>#&gt;  A1CF                             4                           0</span><br><span>#&gt;  A2M                              0                           0</span><br><span>#&gt;  A2ML1                            0                           0</span><br><span>#&gt;  A4GALT                           0                           0</span><br>meta = fread(<span>'./dataset4/mySample_pancreatic_5batches.txt.gz'</span>,) %&gt;% as.data.frame()<br>rownames(meta) &lt;- meta$V1; meta &lt;- meta[,<span>2</span>:ncol(meta)]<br>dim(meta)<br><span>#&gt; [1] 14767     5</span><br>colnames(meta)<br><span>#&gt;  [1] "batch"         "batchlb"       "celltype_orig" "cellname"      "celltype"</span><br>head(meta)[<span>1</span>:<span>5</span>,<span>1</span>:<span>3</span>]<br><span>#&gt;                            batch  batchlb celltype_orig</span><br><span>#&gt; human1_lib1.final_cell_0001     1 Baron_b1        acinar</span><br><span>#&gt; human1_lib1.final_cell_0002     1 Baron_b1        acinar</span><br><span>#&gt; human1_lib1.final_cell_0003     1 Baron_b1        acinar</span><br><span>#&gt; human1_lib1.final_cell_0004     1 Baron_b1        acinar</span><br><span>#&gt; human1_lib1.final_cell_0005     1 Baron_b1        acinar</span><br></code></pre><h3 data-tool="mdniceç¼–è¾‘å™¨">ç»†èƒæŠ½æ ·</h3><p data-tool="mdniceç¼–è¾‘å™¨">è¯»å…¥æ•°æ®åç®€å•æŸ¥çœ‹äº†ä¸‹æ•°æ®ï¼Œä¸éš¾çœ‹å‡ºï¼Œè¯¥èƒ°è…ºç™Œæ•°æ®å«æœ‰14767ä¸ªç»†èƒï¼Œ5ä¸ªæ ·æœ¬ï¼Œä¸ºäº†ç²¾ç®€æµç¨‹ï¼Œå‡å°‘è¿è¡Œæ—¶é—´ï¼Œåç»­å¯¹æ ·æœ¬è¿›è¡ŒéšæœºæŠ½æ ·ï¼Œä»¥å‡å°‘æ•°æ®é‡ã€‚</p><pre data-tool="mdniceç¼–è¾‘å™¨"><code><span>### æŠ½æ ·å‰æŸ¥çœ‹å„ç»†èƒåˆ†å¸ƒ ###</span><br>table(meta$celltype, meta$batch) <br><span>#&gt;                  1    2    3    4    5</span><br><span>#&gt;   acinar        958  219  185    6    0</span><br><span>#&gt;   alpha        2326  812  886  190  886</span><br><span>#&gt;   beta         2525  448  270  111  472</span><br><span>#&gt;   delta         601  193  114    9   49</span><br><span>#&gt;   ductal       1077  245  386   96    0</span><br><span>#&gt;   endothelial   252   21   16    0    0</span><br><span>#&gt;   epsilon        18    3    7    0    0</span><br><span>#&gt;   gamma         255  101  197   18   85</span><br><span>#&gt;   macrophage     55    0    0    0    0</span><br><span>#&gt;   mast           25    0    7    0    0</span><br><span>#&gt;   mesenchymal     0   80    0   27    0</span><br><span>#&gt;   MHC class II    0    0    5    0    0</span><br><span>#&gt;   schwann        13    0    0    0    0</span><br><span>#&gt;   stellate      457    0   54    0    0</span><br><span>#&gt;   t_cell          7    0    0    0    0</span><br><br><span>### ç”±äºéƒ¨åˆ†ç»†èƒæ•°ç¨€å°‘ï¼Œå› æ­¤è¿™é‡Œè€ƒè™‘æŒ‰æ¯æ ·æœ¬æŠ½300ç»†èƒä½œä¸ºç¤ºä¾‹æ–‡ä»¶ ###ï¼Œ</span><br>allCells=rownames(meta)<br>allbatch = levels(factor(meta$batch))<br>choose_Cells = unlist(lapply(allbatch, <span>function</span>(x){<br>  cgCells = allCells[meta$batch == x]<br>  cg=sample(cgCells,<span>300</span>,replace = <span>F</span>)  <br>  cg<br>}))<br>Ct &lt;- counts[,choose_Cells]<br>mt &lt;- meta[choose_Cells,]<br><span>### æŸ¥çœ‹æŠ½æ ·ç»“æœ ##</span><br>table(mt$celltype, mt$batch) <br><span>#&gt;                  1   2   3   4   5</span><br><span>#&gt;   acinar        21  28  32   5   0</span><br><span>#&gt;   alpha         93 116 120 122 178</span><br><span>#&gt;   beta          96  61  42  78  98</span><br><span>#&gt;   delta         18  25  16   5   7</span><br><span>#&gt;   ductal        34  38  49  62   0</span><br><span>#&gt;   endothelial   13   4   3   0   0</span><br><span>#&gt;   epsilon        2   0   1   0   0</span><br><span>#&gt;   gamma          7  17  29  10  17</span><br><span>#&gt;   macrophage     3   0   0   0   0</span><br><span>#&gt;   mast           1   0   2   0   0</span><br><span>#&gt;   mesenchymal    0  11   0  18   0</span><br><span>#&gt;   MHC class II   0   0   1   0   0</span><br><span>#&gt;   schwann        1   0   0   0   0</span><br><span>#&gt;   stellate      11   0   5   0   0</span><br></code></pre><p data-tool="mdniceç¼–è¾‘å™¨">T ç»†èƒå› ä¸ºæ•°é‡æ˜¯åœ¨æ˜¯å¤ªå°‘äº†ï¼Œä¸€ä¸ªéƒ½æ²¡æœ‰æŠ½åˆ°ï¼Œä¸è¿‡ä¸å½±å“æˆ‘ä»¬åç»­æ¼”ç¤ºã€‚</p><h3 data-tool="mdniceç¼–è¾‘å™¨">åˆ›å»ºSingleCellExperiment å¯¹è±¡</h3><p data-tool="mdniceç¼–è¾‘å™¨">PRIMUS å…¨ç¨‹ä½¿ç”¨çš„æ˜¯ SingleCellExperiment å¯¹è±¡ï¼Œè¿™é‡Œä¹Ÿä½¿ç”¨è¯¥å¯¹è±¡ã€‚</p><pre data-tool="mdniceç¼–è¾‘å™¨"><code>simData = logNormCounts(SingleCellExperiment(assays = list(counts = Ct )))<br>simData<br><span>#&gt;  class: SingleCellExperiment </span><br><span>#&gt;  dim: 15558 1500 </span><br><span>#&gt;  metadata(0):</span><br><span>#&gt;  assays(2): counts logcounts</span><br><span>#&gt;  rownames(15558): A1BG A1CF ... ZZEF1 ZZZ3</span><br><span>#&gt;  rowData names(0):</span><br><span>#&gt;  colnames(1500): human3_lib3.final_cell_0151 human3_lib4.final_cell_0497 ... Sample_396 Sample_698</span><br><span>#&gt;  colData names(1): sizeFactor</span><br><span>#&gt;  reducedDimNames(0):</span><br><span>#&gt;  mainExpName: NULL</span><br><span>#&gt;  altExpNames(0):</span><br><span># è¿™é‡Œå°†sizeFactor å­˜æ”¾åˆ°mtæ•°æ®ä¸­ï¼Œåç»­éœ€è¦è¯¥æ•°æ®</span><br>mt$sizeFactor  &lt;- colData(simData)$sizeFactor<br></code></pre><p data-tool="mdniceç¼–è¾‘å™¨">ä½œè€…åœ¨åŸæ–‡ä¸­æåŠçš„æ˜¯ï¼Œèƒ°è…ºç™Œæ•°æ®ä½¿ç”¨PRISM åŒ…ï¼ˆhttps://bitbucket.org/anthakki/prism/src/master/ï¼‰è®¡ç®— <code>scaling factor</code> ,æ¨¡æ‹Ÿæ•°æ®ä½¿ç”¨ <code>logNormCounts</code> è®¡ç®—çš„ç»“æœã€‚è¿™é‡Œå¦å¤–è®¡ç®—å®åœ¨æ˜¯æœ‰ç‚¹çƒ¦ï¼Œæˆ‘ç›´æ¥ç”¨äº†<code>logNormCounts</code>çš„ç»“æœã€‚</p><h3 data-tool="mdniceç¼–è¾‘å™¨">umap å¯è§†åŒ–</h3><p data-tool="mdniceç¼–è¾‘å™¨">å±•ç¤ºæœªè¿›è¡Œ PRIMUS çš„umapç»“æœã€‚</p><pre data-tool="mdniceç¼–è¾‘å™¨"><code><span>library</span>(uwot)<br><span>library</span>(ggplot2)<br><span>library</span>(cowplot)<br><br>umap.coor.raw &lt;- uwot::umap(t(simData@assays@data$logcounts),<br>                            n_components = <span>2</span>, metric = <span>"cosine"</span>,<br>                            min_dist = <span>0.3</span>, n_neighbors = <span>30</span>)<br><br>mt$UMAP_raw_1 = umap.coor.raw[, <span>1</span>]<br>mt$UMAP_raw_2 = umap.coor.raw[, <span>2</span>]<br><br>umap.raw.g &lt;- ggplot(mt, aes(x= UMAP_raw_1, y=UMAP_raw_2, color = celltype) ) +<br>  geom_point(size = <span>0.5</span>) +<br>  theme_classic() +<br>  xlab(<span>"UMAP_1"</span>) + <br>  ylab(<span>"UMAP_2"</span>) +<br>  ggtitle(<span>"colored by phenotypic cell group"</span>) +<br>  guides(colour = guide_legend(override.aes = list(size=<span>3</span>)), shape = guide_legend(override.aes = list(size=<span>3</span>)))<br><br>umap.raw.s &lt;- ggplot(mt, aes(x= UMAP_raw_1, y=UMAP_raw_2, color = batch) ) +<br>  geom_point(size = <span>0.5</span>) +<br>  theme_classic() +<br>  xlab(<span>"UMAP_1"</span>) + <br>  ylab(<span>"UMAP_2"</span>) +<br>  ggtitle(<span>"colored by sample"</span>) +<br>  guides(colour = guide_legend(override.aes = list(size=<span>3</span>)), shape = guide_legend(override.aes = list(size=<span>3</span>)))<br><br>cowplot::plot_grid(umap.raw.s, umap.raw.g, nrow = <span>1</span>)<br></code></pre><p><span>å›¾4 åˆå§‹ umap å±•ç¤ºï¼ˆsample+celltypeï¼‰</span></p><figure data-tool="mdniceç¼–è¾‘å™¨"><p><br></p><figcaption><br></figcaption></figure><h2 data-tool="mdniceç¼–è¾‘å™¨"><span>PRIMUS è¿ç®—</span><span> </span></h2><p data-tool="mdniceç¼–è¾‘å™¨">runPrimuså‡½æ•°ä»scRNA-seqæ•°æ®ä¸­è¯†åˆ«è¡¨å‹ç»†èƒç¾¤ï¼ŒåŒæ—¶è€ƒè™‘åˆ°å¹²æ‰°å› ç´ ï¼ˆè¿™é‡Œæ˜¯æŒ‡æ ·æœ¬çš„batch effectï¼‰ã€‚ä½œä¸ºè¾“å…¥ï¼ŒrunPrimuséœ€è¦åŸå§‹è®¡æ•°çŸ©é˜µï¼ˆmatrixï¼‰ã€æŒ‡å®šæ ·æœ¬æ ‡ç­¾çš„é‚»æ¥çŸ©é˜µå’Œæ¯ä¸ªç»†èƒçš„å¤§å°ç³»æ•°ã€‚runPrimusè¿”å›ä¸€ä¸ªåŒ…å«è¯†åˆ«çš„ç»†èƒClusterã€æ¯ä¸ªç¾¤çš„å»å™ªè½®å»“ä¸­å¿ƒç‚¹ï¼ˆthe denoised profile centroidsï¼‰ã€æ ·æœ¬ç‰¹å®šæ•ˆåº”å’Œæ¨¡å‹æ‹Ÿåˆæˆæœ¬çš„åˆ—è¡¨ï¼ˆListï¼‰ã€‚æˆ‘ä»¬å¯ä»¥é€šè¿‡å¤šæ¬¡éšæœºé‡å¯è¿è¡ŒPRIMUSï¼Œå¹¶é€‰æ‹©æœ€ä½³æ‹Ÿåˆï¼ˆæˆæœ¬æœ€ä½çš„é‚£ä¸ªï¼‰ã€‚</p><h3 data-tool="mdniceç¼–è¾‘å™¨">è¿ç®—</h3><pre data-tool="mdniceç¼–è¾‘å™¨"><code><span>library</span>(PRIMUS)<br><br><span># å°†æ ·æœ¬æ ‡ç­¾è½¬æ¢ä¸ºé‚»è¿‘çŸ©é˜µ</span><br>labels2weights &lt;- <span>function</span>(L, levels = sort(unique(L)))<br>  <span>return</span> (outer(L, levels, `==`) + <span>0L</span>)<br><br>D = t(labels2weights(mt$batch))<br><br><span># éšæœºè¿è¡ŒPRIMUS 5æ¬¡</span><br><span># kæ˜¯èšç±»çš„æ•°é‡ã€‚ç”¨æˆ·å¯ä»¥è®¾ç½®ä¸åŒçš„å€¼ï¼Œå¹¶æ ¹æ®BICé€‰æ‹©æœ€ä½³çš„kã€‚è¿™é‡Œå…ˆè®¾ç½®k =5 </span><br>fits &lt;- list()<br><br>Ct &lt;- as.matrix(Ct)<br><span>for</span> (i <span>in</span> seq.int(<span>5</span>) ){<br>  set.seed(i * <span>1234</span>)<br>    fits[[i]] = runPrimus(Y = Ct, D = D, g = mt$sizeFactor, k = <span>5</span>, max.iter = <span>100</span>)<br>}<br><br>fit = fits[[which.min(sapply(fits, <span>function</span>(x) x$cost))]]<br><br><span>library</span>(mclust)<br><br>adjustedRandIndex(fit$L, mt$celltype)<br><span>#&gt;  [1] 0.8846919    ## æ¨¡æ‹Ÿæ•°æ®çš„ç»“æœæ˜¯1ï¼Œ0.8 çš„ç»“æœ ä¹Ÿå¾ˆé«˜äº†   </span><br>Z = sapply(seq.int(ncol(Ct)), <span>function</span>(i) primus_centroid(A = fit$X %*% D[, i, drop = <span>F</span>],<br>                                                          B = Ct[, i, drop = <span>F</span>], g = mt$sizeFactor[i]))<br><br><span># take square root for variance-stabilizing</span><br>Z.sqrt = sqrt(Z)<br></code></pre><h3 data-tool="mdniceç¼–è¾‘å™¨">å¯è§†åŒ–</h3><pre data-tool="mdniceç¼–è¾‘å™¨"><code><span>#å¯è§†åŒ–</span><br><br><br>umap.coor.Z &lt;- uwot::umap(t(Z.sqrt), n_components = <span>2</span>,<br>                          metric = <span>"cosine"</span>,  min_dist = <span>0.3</span>,<br>                          n_neighbors = <span>30</span>, <br>                          y = as.factor(fit$L),<br>                          target_weight = <span>0.05</span>)<br><br>mt$Z_UMAP_1 = umap.coor.Z[, <span>1</span>]<br>mt$Z_UMAP_2 = umap.coor.Z[, <span>2</span>]<br>mt$cluster = as.factor(fit$L)<br>mt$celltype<br>umap.Z.g &lt;- ggplot(mt, aes(x= Z_UMAP_1, y=Z_UMAP_2, color = celltype) ) +<br>  geom_point(size = <span>0.5</span>) +<br>  theme_classic() +<br>  ggtitle(<span>"colored by phenotypic cell group"</span>) + <br>  guides(colour = guide_legend(override.aes = list(size=<span>3</span>)), shape = guide_legend(override.aes = list(size=<span>3</span>)))<br><br>umap.Z.s &lt;- ggplot(mt, aes(x= Z_UMAP_1, y=Z_UMAP_2, color = batch) ) +<br>  geom_point(size = <span>0.5</span>) +<br>  theme_classic() +<br>  ggtitle(<span>"colored by sample"</span>) + <br>  guides(colour = guide_legend(override.aes = list(size=<span>3</span>)), shape = guide_legend(override.aes = list(size=<span>3</span>)))<br><br>umap.Z.c &lt;- ggplot(mt, aes(x= Z_UMAP_1, y=Z_UMAP_2, color = cluster) ) +<br>  geom_point(size = <span>0.5</span>) +<br>  theme_classic() +<br>  ggtitle(<span>"colored by PRIMUS identified clusters"</span>) +<br>  guides(colour = guide_legend(override.aes = list(size=<span>3</span>)), shape = guide_legend(override.aes = list(size=<span>3</span>)) )<br><br>cowplot::plot_grid(umap.Z.s, umap.Z.g, umap.Z.c, nrow = <span>1</span>)<br><br></code></pre><p data-tool="mdniceç¼–è¾‘å™¨"><img data-ratio="0.26917613636363635" data-src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2LosicLWxG3A6KyWDUKMATWIIicLgxtjiaxB8EhsgUjP46whd5ZPuIlIsjbMZLcw1ISSYpYnPGM6gf0m7Ww/640?wx_fmt=png" data-type="png" data-w="1408" title="å›¾5Â  k=5 æ—¶ umapå›¾" src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2LosicLWxG3A6KyWDUKMATWIIicLgxtjiaxB8EhsgUjP46whd5ZPuIlIsjbMZLcw1ISSYpYnPGM6gf0m7Ww/640?wx_fmt=png"><span>å›¾5Â  k=5 æ—¶ umapå›¾</span></p><p data-tool="mdniceç¼–è¾‘å™¨">k =5 æ—¶ï¼Œå‡ ä¸ªå¤§ç±»çš„ç»†èƒèƒ½å¤Ÿå¾ˆå¥½çš„åŒºåˆ†å¼€ï¼Œè€Œå°‘éƒ¨åˆ†æ•°é‡å¾ˆå°‘çš„ç»†èƒåˆ™æˆæ•£ç‚¹æ ·åœ¨å¤§ç»†èƒç¾¤å‘¨å›´åˆ†å¸ƒã€‚è‹¥è®¾ç½®k =8 åˆ™ï¼š</p><pre data-tool="mdniceç¼–è¾‘å™¨"><code>adjustedRandIndex(fit$L, mt$celltype)<br><span>#&gt;  [1] 0.6460315</span><br></code></pre><p data-tool="mdniceç¼–è¾‘å™¨">éšåå¯è§†åŒ–ï¼ˆk=8ï¼‰</p><p data-tool="mdniceç¼–è¾‘å™¨"><img data-ratio="0.27586206896551724" data-src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2LosicLWxG3A6KyWDUKMATWIIicLbTTnhuz1N5tYx9LoBTW86Alre51xs5FEZ5vyC0pnibMfZMQfh15FVJw/640?wx_fmt=png" data-type="png" data-w="1276" title="å›¾6Â  k=8 æ—¶umap å›¾" src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2LosicLWxG3A6KyWDUKMATWIIicLbTTnhuz1N5tYx9LoBTW86Alre51xs5FEZ5vyC0pnibMfZMQfh15FVJw/640?wx_fmt=png"><span>å›¾6Â  k=8 æ—¶umap å›¾</span></p><p data-tool="mdniceç¼–è¾‘å™¨">k=8 æ—¶ï¼Œæ˜æ˜¾èšç±»çš„æ•ˆæœå‚å·®ï¼Œä½†æ˜¾ç„¶ï¼Œå¤§ç»†èƒç¾¤è¿˜æ˜¯å¾ˆå¥½çš„åŒºåˆ†å¼€äº†ï¼Œä½†å°éƒ¨åˆ†ç»†èƒç¾¤ä»æœ‰èšé›†ã€‚</p><h1 data-tool="mdniceç¼–è¾‘å™¨">ç»“æŸè¯­</h1><p data-tool="mdniceç¼–è¾‘å™¨">æœ¬ç¯‡ç®€è¦ä»‹ç»äº†æ–‡çŒ®ï¼šLongitudinal single-cell RNA-seq analysis reveals stress-promoted chemoresistance in metastatic ovarian cancerä¸­å¼€å‘çš„æ–°å¤šæ ·æœ¬æ•´åˆRåŒ…PRIMUSã€‚ä»å‡ºå‘ç‚¹æ¥è¯´ï¼Œè¯¥åŒ…è§£å†³äº†å¤šæ ·æœ¬å¼‚è´¨æ€§çš„é—®é¢˜ï¼ŒæŸäº›ç»†èƒç¾¤åªåœ¨å•ä¸ªæ ·æœ¬ä¸­æœ‰ï¼Œè€Œåœ¨å¦ä¸€ä¸ªæ ·æœ¬ä¸­æ— ï¼Œè¿™ç§æƒ…å†µä¸‹çš„æ ·æœ¬æ•´åˆã€‚æ ¹æ®åŸæ–‡çŒ®è¡¥å……æ•°æ®ä¸Šçš„æè¿°ï¼ŒSeurat v3 ä»¥åŠHarmonyåœ¨è¿™ç§æƒ…å†µä¸‹å¤„ç†æ•ˆæœä¸€èˆ¬ï¼Œç”±æ­¤è¯´æ˜PRIMUS æ•´åˆçš„ä¼˜åŠ¿ã€‚</p><p data-tool="mdniceç¼–è¾‘å™¨"><img data-ratio="0.20309278350515464" data-src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2LosicLWxG3A6KyWDUKMATWIIicLgjkOic5YmVyrJTMOGiaTJrvurB3Kw56YSOAuBNjmCSvibWUbugg9uzZ0Q/640?wx_fmt=png" data-type="png" data-w="970" title="å›¾7 æ–‡çŒ®é™„å›¾S2 I" src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2LosicLWxG3A6KyWDUKMATWIIicLgjkOic5YmVyrJTMOGiaTJrvurB3Kw56YSOAuBNjmCSvibWUbugg9uzZ0Q/640?wx_fmt=png"><span>å›¾7 æ–‡çŒ®é™„å›¾S2 I</span></p><p data-tool="mdniceç¼–è¾‘å™¨">Kå€¼çš„é€‰æ‹©åœ¨PRIMUSçš„æµç¨‹ä¸­è‡³å…³é‡è¦ï¼Œä½œè€…æåŠäº†BICæ–¹æ³•é€‰æ‹©Kå€¼ï¼Œä½†å…·ä½“å¦‚ä½•æ“ä½œï¼Œç›®å‰è¿˜æœªèƒ½æ‰¾åˆ°è¸ªè¿¹ï¼Œåç»­å­¦ä¹ è¡¥å……ä¹‹ã€‚</p><blockquote data-tool="mdniceç¼–è¾‘å™¨"><p>To select the optimal k, we fitted PRIMUS for k = 1,2, â€¦,25 with 10 different random initial parameter sets for each k, and k = 12 was selected on the basis of BIC (fig. S2A).<strong>BIC ï¼š</strong> Bayesian information criterion</p></blockquote><figure data-tool="mdniceç¼–è¾‘å™¨"><img data-ratio="0.8688946015424165" data-src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2LosicLWxG3A6KyWDUKMATWIIicLkg7YwicVseMtviaqWYOLdXIhh546x9LRak63oGrTJzJtA03hR2xRicvLw/640?wx_fmt=png" data-type="png" data-w="389" title="å›¾8 æ–‡çŒ®é™„å›¾S3 A" src="https://mmbiz.qpic.cn/mmbiz_png/iaRJcrq2LosicLWxG3A6KyWDUKMATWIIicLkg7YwicVseMtviaqWYOLdXIhh546x9LRak63oGrTJzJtA03hR2xRicvLw/640?wx_fmt=png"><figcaption><span>å›¾8 æ–‡çŒ®é™„å›¾S3 A</span><br></figcaption></figure><blockquote data-tool="mdniceç¼–è¾‘å™¨"><p><strong>S3 A Legendï¼š</strong> Selection of the number of cancer cell clusters based on Bayesian information criterion (BIC). The boxplot showing the BIC of models fitted for ğ‘˜ğ‘˜ = 1, 2, . . . , 25 with 10 different random initial parameter sets for each ğ‘˜. The red line indicates the selected number of clusters (ğ‘˜ = 12).</p></blockquote><span>- END -</span></section><p><br></p><p><mp-style-type data-value="10000"></mp-style-type></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/in2ju7wIAKVrdSk5jGxjPg",target="_blank" rel="noopener noreferrer">åŸæ–‡é“¾æ¥</a>
