---
title: "è½¬å½•ç»„+ç”²åŸºåŒ–æ¥ç‚¹æ–°èŠ±æ ·ï¼šELMERåŒ…"
date: 2023-01-01T05:30:34Z
draft: ["false"]
tags: [
  "fetched",
  "YuYuFiSH"
]
categories: ["Acdemic"]
---
è½¬å½•ç»„+ç”²åŸºåŒ–æ¥ç‚¹æ–°èŠ±æ ·ï¼šELMERåŒ… by YuYuFiSH
------
<div><section data-tool="mdniceç¼–è¾‘å™¨" data-website="https://www.mdnice.com"><h1 data-tool="mdniceç¼–è¾‘å™¨"><span></span>ELMER å¤šç»„å­¦RåŒ…</h1><blockquote data-type="2" data-url="" data-author-name="" data-content-utf8-length="40" data-source-title=""><section><section>å¦‚æœä½ æ‰‹ä¸Šæœ‰ç”²åŸºåŒ–å’Œè½¬å½•ç»„ä¸¤ç§æ•°æ®ï¼Œæœ‰å¤ªå¤šåˆ†æçš„è§’åº¦èƒ½è¾…åŠ©ä½ å®Œæˆä¸€ä¸ªæ¼‚äº®çš„æ•…äº‹äº†</section></section></blockquote><figure data-tool="mdniceç¼–è¾‘å™¨"><img data-ratio="0.41450313339301703" data-src="https://mmbiz.qpic.cn/mmbiz_png/PsnauFhHibc5IgrnANu8GpqP6TXajaybAHxPaj3Ko5OP0iajxzupfyeBxicdXicEOVgOKoabSlicVDs56QwicJzbwjJQ/640?wx_fmt=png" data-type="png" data-w="1117" src="https://mmbiz.qpic.cn/mmbiz_png/PsnauFhHibc5IgrnANu8GpqP6TXajaybAHxPaj3Ko5OP0iajxzupfyeBxicdXicEOVgOKoabSlicVDs56QwicJzbwjJQ/640?wx_fmt=png"></figure><h2 data-tool="mdniceç¼–è¾‘å™¨"><span></span>è‡ªæˆ‘ä»‹ç»</h2><p data-tool="mdniceç¼–è¾‘å™¨"><span>ELMERï¼ˆEnhancer Linking by Methylation/Expression Relationshipsï¼‰</span>ï¼Œå¢å¼ºå­ç›¸å…³ç”²åŸºåŒ–å’Œè¡¨è¾¾ç›¸å…³ç½‘ç»œ</p><p data-tool="mdniceç¼–è¾‘å™¨">å…³é”®æ˜¯ä»¥ç”²åŸºåŒ–ä½ç‚¹ä½œä¸ºç½‘ç»œçš„ä¸­å¿ƒæ¢çº½ï¼Œåœ¨ä¸‹æ¸¸ç­›é€‰å‡ºå·®å¼‚ç”²åŸºåŒ–ä½ç‚¹å¯¹åº”çš„ç¢±åŸºåºåˆ—ä¸å¯¹åº”çš„é¶åŸºå› ï¼Œè€Œä¸Šæ¸¸æ‰¾åˆ°å¯Œé›†çš„motifï¼ˆè½¬å½•å› å­ç»“åˆåŸºåºå³æœ‰è§„å¾‹æœ‰ç‰¹æ®ŠåŠŸèƒ½çš„ä¸€ä¸²ç¢±åŸºåºåˆ—ï¼‰ï¼Œå†åæ¨å‡ºä¸è¿™äº›motifç›¸äº’ä½œç”¨çš„è½¬å½•å› å­ï¼Œä»è€Œæ„å»ºå‡º<span><strong>TFâ€”ç”²åŸºåŒ–ä½ç‚¹â€”â€”é¶åŸºå› ç½‘ç»œ</strong></span></p><h2 data-tool="mdniceç¼–è¾‘å™¨"><span></span>ä¸»è¦æµç¨‹</h2><p data-tool="mdniceç¼–è¾‘å™¨">1.è¯†åˆ«è¿œç«¯æ¢é’ˆï¼ˆdistal probeï¼‰ï¼šHM450kæˆ–è€…EPICç”²åŸºåŒ–èŠ¯ç‰‡</p><p data-tool="mdniceç¼–è¾‘å™¨">2.è¯†åˆ«ç»„é—´ç”²åŸºåŒ–æ°´å¹³å·®å¼‚</p><p data-tool="mdniceç¼–è¾‘å™¨">3.è¯†åˆ«å·®å¼‚ç”²åŸºåŒ–æ¢é’ˆçš„é¶åŸºå› </p><p data-tool="mdniceç¼–è¾‘å™¨">4.ç¡®å®šå·®å¼‚ç”²åŸºåŒ–å·®å¼‚æ¢é’ˆä»¥åŠé¶åŸºå› ç›¸å…³æ¢é’ˆå¯Œé›†çš„motif</p><p data-tool="mdniceç¼–è¾‘å™¨">5.ç¡®è®¤ä¸motifä¸Šç”²åŸºåŒ–ç›¸å…³çš„è°ƒæ§è½¬å½•å› å­</p><blockquote data-type="2" data-url="" data-author-name="" data-content-utf8-length="173" data-source-title=""><section><section><p data-tool="mdniceç¼–è¾‘å™¨">RåŒ…ç¬¬ä¸€æ­¥å¯ä»¥ç†è§£ä¸ºè¯†åˆ«å…³é”®ä½ç½®çš„ç”²åŸºåŒ–æ¢é’ˆï¼Œè¿™é‡ŒæŒ‡å¢å¼ºå­ï¼ˆenhancerï¼‰æˆ–è€…å¯åŠ¨å­åŒºåŸŸï¼ˆpromoterï¼‰çš„æ¢é’ˆã€‚æ™®éè®¤ä¸ºåœ¨é¡ºå¼è°ƒæ§å…ƒä»¶éƒ¨ä½ï¼ˆåŒ…æ‹¬å¯åŠ¨å­ã€å¢å¼ºå­ç­‰éƒ¨ä½ï¼‰é«˜åº¦ç”²åŸºåŒ–ï¼Œå°†å½±å“DNAçš„ç»“æ„ï¼Œä»è€Œé˜»éè¯¥éƒ¨ä½åŸºå› çš„è½¬å½•</p><p data-tool="mdniceç¼–è¾‘å™¨">distal probeï¼šè·ç¦»è½¬å½•èµ·å§‹ä½ç‚¹ä½ç‚¹ä¸Šæ¸¸å¤§äº2kbçš„ç”²åŸºåŒ–æ¢é’ˆï¼Œé»˜è®¤è¿™ä¸€åŒºåŸŸçš„ç”²åŸºåŒ–æ¢é’ˆåŒºåŸŸä¸ºå¢å¼ºå­åŒºåŸŸã€‚</p></section></section></blockquote><h2 data-tool="mdniceç¼–è¾‘å™¨"><span></span>æœ€æ–°2019å¹´ç‰ˆæœ¬</h2><p data-tool="mdniceç¼–è¾‘å™¨">æ–°ç‰ˆæœ¬å†æ¬¡å‡çº§</p><figure data-tool="mdniceç¼–è¾‘å™¨"><img data-ratio="0.6323076923076923" data-src="https://mmbiz.qpic.cn/mmbiz_png/PsnauFhHibc5IgrnANu8GpqP6TXajaybAbkz8q6BxB80t35zHZn4Svz0bo8jLiaWQHaTzVd46wgKO1EYhv5VxsTQ/640?wx_fmt=png" data-type="png" data-w="1300" src="https://mmbiz.qpic.cn/mmbiz_png/PsnauFhHibc5IgrnANu8GpqP6TXajaybAbkz8q6BxB80t35zHZn4Svz0bo8jLiaWQHaTzVd46wgKO1EYhv5VxsTQ/640?wx_fmt=png"></figure><p data-tool="mdniceç¼–è¾‘å™¨">ğŸ¥ å¸®åŠ©æ–‡æ¡£ï¼šhttps://academic.oup.com/bioinformatics/article/35/11/1974/5145133?login=false</p><p data-tool="mdniceç¼–è¾‘å™¨">http://bioconductor.org/packages/release/bioc/vignettes/ELMER/inst/doc/analysis_data_input.html</p><p data-tool="mdniceç¼–è¾‘å™¨">http://bioconductor.org/packages/release/bioc/manuals/ELMER/man/ELMER.pdf</p><p data-tool="mdniceç¼–è¾‘å™¨"><span></span><span>ğŸ¥Â </span>å·¥ä½œæµç¨‹workshopï¼š</p><p data-tool="mdniceç¼–è¾‘å™¨">https://benbermanlab.com/assets/code/ELMER%20workshop.html</p><p data-tool="mdniceç¼–è¾‘å™¨">https://rpubs.com/tiagochst/elmer-data-workshop-2019</p><p data-tool="mdniceç¼–è¾‘å™¨"><span></span><span>ğŸ¥Â </span>æ–‡ç« TCGA-BRCAå®Œæ•´æŠ¥å‘Šï¼š</p><p data-tool="mdniceç¼–è¾‘å™¨">https://media.githubusercontent.com/media/tiagochst/ELMER_supplemental/master/supplemental_files.zip</p><h2 data-tool="mdniceç¼–è¾‘å™¨"><span>ç›®å½•</span></h2><ul data-tool="mdniceç¼–è¾‘å™¨"><li><section>è·å¾—è¿œç«¯æ¢é’ˆï¼š<code>get.feature.probe</code></section></li><li><section>è½½å…¥ç”²åŸºåŒ–ï¼Œè½¬å½•ç»„ï¼Œä¸´åºŠæ•°æ®</section></li><li><section>åˆ›å»ºMAEå¯¹è±¡:<code>createMAE</code></section></li><li><section>å·®å¼‚åˆ†æï¼š<code>get.diff.meth</code></section></li><li><section>å¯»æ‰¾æ¢é’ˆä¸Šä¸‹æ¸¸åŸºå› ï¼š<code>GetNearGenes</code></section></li><li><section>è¯†åˆ«probe-geneå¯¹ï¼š<code>get.pair</code></section></li><li><section>ç­›é€‰motifï¼š<code>get.enriched.motif</code></section></li><li><section>ç¡®å®šè½¬å½•å› å­TFï¼š<code>get.TFs</code></section></li><li><section>å„ç§å¯è§†åŒ–ï¼š</section></li><ul><li><section>æ•£ç‚¹å›¾ï¼š<code>scatter.plot</code></section></li><li><section>æŸ“è‰²ä½“ç¤ºæ„å›¾ï¼š<code>schematic.plot</code></section></li><li><section>motifå¯Œé›†æ£®æ—å›¾ï¼š<code>motif.enrichment.plot</code></section></li><li><section>è½¬å½•å› å­æ’åºå›¾ï¼š<code>TF.rank.plot</code></section></li><li><section>çƒ­å›¾ï¼š<code>heatmapPairs</code></section></li></ul><li><section>ç›´æ¥ä¸‹è½½TCGAæ•°æ®ï¼š<code>TCGA.pipe</code></section></li></ul><h2 data-tool="mdniceç¼–è¾‘å™¨"><span></span>é¢†åŸŸå±•å¼€----</h2><h3 data-tool="mdniceç¼–è¾‘å™¨"><span>1.RåŒ…å®‰è£…</span></h3><p data-tool="mdniceç¼–è¾‘å™¨">ä¸¤ç§æ–¹æ³•ï¼šä»githubä¸­å®‰è£… or Bioconductorå®‰è£…</p><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code>devtools::install_github(repoÂ =Â <span>"tiagochst/ELMER.data"</span>)<br>devtools::install_github(repoÂ =Â <span>"tiagochst/ELMER"</span>)<br><br><span>if</span>Â (!requireNamespace(<span>"BiocManager"</span>,Â quietly=TRUE))<br>Â Â Â Â install.packages(<span>"BiocManager"</span>)<br>BiocManager::install(<span>"ELMER"</span>)<br>BiocManager::install(<span>"sesameData"</span>)<br><br>library(ELMER)<br>library(sesameData)<br>library(DT)<br></code></pre><h3 data-tool="mdniceç¼–è¾‘å™¨"><span>2. è·å¾—è¿œç«¯æ¢é’ˆ</span></h3><p data-tool="mdniceç¼–è¾‘å™¨">è¿™é‡Œä»¥TCGAé¡¹ç›®ä¸­è‚ºé³ç™Œï¼ˆLUSCï¼‰é¡¹ç›®ä¸ºä¾‹ï¼Œåªé€‰æ‹©1å·æŸ“è‰²ä½“çš„ç”²åŸºåŒ–æ¢é’ˆå’ŒåŸºå› ä½œä¸ºä¾‹å­è¿›è¡Œåˆ†æã€‚</p><p data-tool="mdniceç¼–è¾‘å™¨">é¦–å…ˆï¼Œæ¢é’ˆè¦æå‰åšå‡ºç­›é€‰ï¼Œæ€ä¹ˆè·å¾—å¢å¼ºå­å’Œå¯åŠ¨å­ä½ç½®ä¸Šçš„å‘¢ï¼ŸråŒ…è¿™é‡Œå·²ç»æ ¹æ®æ•°æ®åº“GENCODEå’ŒUCSCåˆ†å¥½ç±»ï¼Œä¸è¿‡å¦‚æœéœ€è¦å¯åŠ¨å­ä¸Šçš„æ¢é’ˆï¼Œéœ€è¦ç”¨å‚æ•°<code>TSS.range</code>æå‰å®šä¹‰å¥½å¯åŠ¨å­åŒºåŸŸï¼Œé»˜è®¤æ˜¯TSSä¸Š2000bpå†…ï¼Œä¸è¿‡ä¹Ÿå¯ä»¥è‡ªå®šä¹‰ï¼Œæ¯•ç«Ÿæ²¡æœ‰ç¡¬æ€§æ ‡å‡†</p><ul data-tool="mdniceç¼–è¾‘å™¨"><li><section>è·å¾—è¿œç«¯æ¢é’ˆï¼š<code>get.feature.probe</code></section></li></ul><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code><span>#Â SelectionÂ ofÂ probesÂ withinÂ biofeatures</span><br><span>#Â getÂ distalÂ probesÂ thatÂ areÂ 2kbÂ awayÂ fromÂ TSSÂ onÂ chromosomeÂ 1</span><br>distal.probesÂ &lt;-Â get.feature.probe(<br>Â Â genomeÂ =Â <span>"hg19"</span>,Â <span>#Â hg38Â (default)Â orÂ hg19</span><br>Â Â met.platformÂ =Â <span>"450K"</span>,Â <span>#Â Â EPICÂ orÂ 450KÂ (default)</span><br>Â Â TSS.rangeÂ =Â list(upstreamÂ =Â 2000,Â downstreamÂ =Â 2000),Â <span>#Â defineÂ promoterÂ regionså®šä¹‰å¯åŠ¨å­åŒºåŸŸ</span><br>Â Â promoterÂ =Â FALSE,Â <span>#Â æ˜¯å¦é€‰æ‹©å¯åŠ¨å­</span><br>Â Â rm.chrÂ =Â paste0(<span>"chr"</span>,c(2:22,<span>"X"</span>,<span>"Y"</span>))Â <span>#Â ç§»é™¤çš„æŸ“è‰²ä½“</span><br><br>length(distal.probes)<span>#Â [1]Â 15561</span><br>head(distal.probes)<br></code></pre><figure data-tool="mdniceç¼–è¾‘å™¨"><img data-ratio="0.47194719471947194" data-src="https://mmbiz.qpic.cn/mmbiz_png/PsnauFhHibc5IgrnANu8GpqP6TXajaybAcfCrbDOLzrjAOYic3drIIYp08JFWEgXFIbOMa0yXgJ6A7pZTmSUbRXg/640?wx_fmt=png" data-type="png" data-w="1212" src="https://mmbiz.qpic.cn/mmbiz_png/PsnauFhHibc5IgrnANu8GpqP6TXajaybAcfCrbDOLzrjAOYic3drIIYp08JFWEgXFIbOMa0yXgJ6A7pZTmSUbRXg/640?wx_fmt=png"></figure><h3 data-tool="mdniceç¼–è¾‘å™¨"><span>3. è¾“å…¥æ•°æ®(TCGAæ ¼å¼)</span></h3><p data-tool="mdniceç¼–è¾‘å™¨">å¾—åˆ°1å·æŸ“è‰²ä½“çš„distal probeåï¼Œæˆ‘ä»¬å°±éœ€è¦</p><ul data-tool="mdniceç¼–è¾‘å™¨"><li><section>è½½å…¥ç”²åŸºåŒ–ï¼Œè½¬å½•ç»„æ•°æ®</section></li><li><section>åˆ›å»ºMAEå¯¹è±¡: <code>createMAE</code></section></li></ul><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code><span>#Â SelectionÂ ofÂ probesÂ withinÂ biofeatures</span><br>library(MultiAssayExperiment)<br>library(ELMER.data)<br><span>#Â è½¬å½•ç»„æ•°æ®</span><br>data(LUSC_RNA_refined,packageÂ =Â <span>"ELMER.data"</span>)<br><span>#Â ç”²åŸºåŒ–æ•°æ®</span><br>data(LUSC_meth_refined,packageÂ =Â <span>"ELMER.data"</span>)<br><br>dim(GeneExp)Â <span>#3842Â Â 234</span><br>dim(Meth)Â <span>#1728Â Â 268</span><br></code></pre><p data-tool="mdniceç¼–è¾‘å™¨">å¯ä»¥çœ‹åˆ°ï¼Œä¸¤ç§ç±»å‹çš„æ•°æ®æ ·æœ¬æ•°é‡ä¸å®Œå…¨ä¸€è‡´</p><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code>&gt;Â GeneExp[1:3,1:3]<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â TCGA-22-5472-01A-01R-1635-07Â TCGA-22-5489-01A-01R-1635-07<br>ENSG00000188984Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.0000000Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.00000<br>ENSG00000204518Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.4303923Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.00000<br>ENSG00000108270Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 10.0817831Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 10.71767<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â TCGA-22-5491-11A-01R-1858-07<br>ENSG00000188984Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.00000<br>ENSG00000204518Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.00000<br>ENSG00000108270Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 10.18086<br>&gt;Â Meth[1:3,1:3]<br>Â Â Â Â Â Â Â Â Â Â Â TCGA-43-3394-11A-01D-1551-05Â TCGA-43-3920-11B-01D-1551-05<br>cg00045114Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.8190894Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.8073763<br>cg00050294Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.8423084Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.8241138<br>cg00066722Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.9101127Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.9162212<br>Â Â Â Â Â Â Â Â Â Â Â TCGA-56-8305-01A-11D-2294-05<br>cg00045114Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.8907009<br>cg00050294Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.5597787<br>cg00066722Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 0.7228874<br></code></pre><p><br></p><ul data-tool="mdniceç¼–è¾‘å™¨"><li><section>åˆ›å»ºMAEå¯¹è±¡:<code>createMAE</code></section></li></ul><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code>maeÂ &lt;-Â createMAE(<br>Â Â expÂ =Â GeneExp,Â <br>Â Â metÂ =Â Meth,<br>Â Â saveÂ =Â TRUE,<br>Â Â linearize.expÂ =Â TRUE,Â <span>#Â Â log2(expÂ +Â 1)Â </span><br>Â Â save.filenameÂ =Â <span>"mae.rda"</span>,Â <span>#Â ä¿å­˜çš„æ–‡ä»¶</span><br>Â Â filter.probesÂ =Â distal.probes,Â <span>#Â promoterÂ regionsÂ orÂ distalÂ featureÂ regions</span><br>Â Â met.platformÂ =Â <span>"450K"</span>,<br>Â Â genomeÂ =Â <span>"hg19"</span>,<br>Â Â TCGAÂ =Â TRUE<br>)<br><span>#Â =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-</span><br><span>#Â Â Â CreatingÂ aÂ SummarizedExperimentÂ fromÂ geneÂ expressionÂ input</span><br><span>#Â =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-</span><br><span>#Â Â Â CreatingÂ aÂ SummarizedExperimentÂ fromÂ DNAÂ methylationÂ input</span><br><span>#Â CheckingÂ ifÂ samplesÂ haveÂ bothÂ DNAÂ methylationÂ andÂ GeneÂ expressionÂ andÂ ifÂ theyÂ areÂ inÂ theÂ sameÂ order...</span><br><span>#Â StartingÂ toÂ addÂ informationÂ toÂ samples</span><br><span>#Â =&gt;Â AddÂ clinicalÂ informationÂ toÂ samples</span><br><span>#Â =&gt;Â AddingÂ TCGAÂ molecularÂ informationÂ fromÂ markerÂ papers</span><br><span>#Â =&gt;Â InformationÂ willÂ haveÂ prefixÂ 'paper_'Â </span><br><span>#Â luscÂ subtypeÂ informationÂ from:doi:10.1038/nature11404<svg version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" x="0px" y="0px" width="32px" height="32px" viewbox="0 0 32 32" enable-background="new 0 0 32 32"><image width="32" height="32" x="0" y="0"></image></svg></span><br><span>#Â CreatingÂ MultiAssayExperiment</span><br><span>#Â MAEÂ savedÂ as:Â mae.rda</span><br></code></pre><p data-tool="mdniceç¼–è¾‘å™¨">MAEå¯¹è±¡å°†æ•°æ®è¿›è¡Œæ•´åˆï¼Œæå–åŒæ—¶æ‹¥æœ‰ä¸¤ç§æ•°æ®çš„234ä¸ªæ ·æœ¬ã€‚å¯ä»¥æ ¹æ®ä¸‹é¢æç¤ºçš„å‡½æ•°æ¢ç´¢ä¸‹MAEå¯¹è±¡</p><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code>&gt;Â mae<br>AÂ MultiAssayExperimentÂ objectÂ ofÂ 2Â listed<br>Â experimentsÂ withÂ user-definedÂ namesÂ andÂ respectiveÂ classes.<br>Â ContainingÂ anÂ ExperimentListÂ classÂ objectÂ ofÂ lengthÂ 2:<br>Â [1]Â DNAÂ methylation:Â RangedSummarizedExperimentÂ withÂ 1642Â rowsÂ andÂ 234Â columns<br>Â [2]Â GeneÂ expression:Â RangedSummarizedExperimentÂ withÂ 3808Â rowsÂ andÂ 234Â columns<br>Functionality:<br>Â experiments()Â -Â obtainÂ theÂ ExperimentListÂ instance<br>Â colData()Â -Â theÂ primary/phenotypeÂ DataFrame<br>Â sampleMap()Â -Â theÂ sampleÂ coordinationÂ DataFrame<br>Â `$`,Â `[`,Â `[[`Â -Â extractÂ colDataÂ columns,Â subset,Â orÂ experiment<br>Â *Format()Â -Â convertÂ intoÂ aÂ longÂ orÂ wideÂ DataFrame<br>Â assays()Â -Â convertÂ ExperimentListÂ toÂ aÂ SimpleListÂ ofÂ matrices<br>Â exportClass()Â -Â saveÂ dataÂ toÂ flatÂ files<br></code></pre><figure data-tool="mdniceç¼–è¾‘å™¨"><img data-ratio="0.5152603231597845" data-src="https://mmbiz.qpic.cn/mmbiz_png/PsnauFhHibc5IgrnANu8GpqP6TXajaybAy2PluJWwkeQJPsnXejAK2JkoxwuRvy16EmGH0eqJ4XucdhOWWLD3aw/640?wx_fmt=png" data-type="png" data-w="1114" src="https://mmbiz.qpic.cn/mmbiz_png/PsnauFhHibc5IgrnANu8GpqP6TXajaybAy2PluJWwkeQJPsnXejAK2JkoxwuRvy16EmGH0eqJ4XucdhOWWLD3aw/640?wx_fmt=png"></figure><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code><span>#Â è§£æmaeå¯¹è±¡</span><br><span>#è·å–ç”²åŸºåŒ–è¡¨è¾¾è°±æ•°æ®</span><br>getMet(mae)Â %&gt;%Â assay()<br><span>#è·å–MEAå¯¹è±¡è½¬å½•ç»„æ•°æ®</span><br>getExp(mae)Â %&gt;%Â assay()<br><span>#è·å–MEAå¯¹è±¡çš„metadataä¿¡æ¯</span><br>colData(mae)Â %&gt;%Â as.data.frame()<br></code></pre><h3 data-tool="mdniceç¼–è¾‘å™¨"><span>3.2 è¾“å…¥æ•°æ®ï¼ˆéTCGAæ ¼å¼ï¼‰</span></h3><p data-tool="mdniceç¼–è¾‘å™¨">å½“ç„¶ï¼Œå¦‚æœä½ æœ‰è‡ªå·±çš„æµ‹åºæ•°æ®ï¼Œä¹Ÿå¯ä»¥æ„å»ºå±äºè‡ªå·±çš„MAEå¯¹è±¡å’¯</p><p data-tool="mdniceç¼–è¾‘å™¨">éœ€è¦å‡†å¤‡ï¼š</p><p data-tool="mdniceç¼–è¾‘å™¨">1.ä»¥æ¢é’ˆä¸ºè¡Œï¼Œä»¥æ ·æœ¬ä¸ºåˆ—çš„Î²å€¼ç”²åŸºåŒ–çŸ©é˜µ</p><p data-tool="mdniceç¼–è¾‘å™¨">2.ä»¥åŸºå› ä¸ºè¡Œï¼Œä»¥æ ·æœ¬ä¸ºåˆ—å¹¶ä¸”æ ‡å‡†åŒ–å¤„ç†åçš„è½¬å½•ç»„çŸ©é˜µï¼ˆFPKMç­‰ï¼‰</p><p data-tool="mdniceç¼–è¾‘å™¨">3.ç”²åŸºåŒ–æ ·æœ¬åå’Œè½¬å½•ç»„æ ·æœ¬åå¯¹åº”çš„è½¬åŒ–çŸ©é˜µ</p><p data-tool="mdniceç¼–è¾‘å™¨">4.æ ·æœ¬æœ¬èº«çš„ä¸´åºŠä¿¡æ¯ï¼ˆmetadataï¼‰</p><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code>&gt;Â head(exp)<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Sample1Â Sample2Â Sample3Â Sample4Â Sample5<br>ENSG00000073282Â Â Â Â Â Â Â 0Â Â Â Â Â Â Â 0Â Â Â Â Â Â Â 0Â Â Â Â Â Â Â 0Â Â Â Â Â Â Â 0<br>ENSG00000078900Â Â Â Â Â Â Â 0Â Â Â Â Â Â Â 0Â Â Â Â Â Â Â 0Â Â Â Â Â Â Â 0Â Â Â Â Â Â Â 0<br>ENSG00000141510Â Â Â Â Â Â Â 0Â Â Â Â Â Â Â 0Â Â Â Â Â Â Â 0Â Â Â Â Â Â Â 0Â Â Â Â Â Â Â 0<br>&gt;Â head(met)<br>Â Â Â Â Â Â Â Â Â Â Â Sample1Â Sample2Â Sample3Â Sample4Â Sample5<br>cg26928153Â Â Â Â Â Â Â 0Â Â Â Â Â Â Â 0Â Â Â Â Â Â Â 0Â Â Â Â Â Â Â 0Â Â Â Â Â Â Â 0<br>cg16269199Â Â Â Â Â Â Â 0Â Â Â Â Â Â Â 0Â Â Â Â Â Â Â 0Â Â Â Â Â Â Â 0Â Â Â Â Â Â Â 0<br>cg13869341Â Â Â Â Â Â Â 0Â Â Â Â Â Â Â 0Â Â Â Â Â Â Â 0Â Â Â Â Â Â Â 0Â Â Â Â Â Â Â 0<br>&gt;Â head(colData)Â <span>#Â åç»­å¯ä»¥å¢åŠ å¤šå‡ åˆ—ä¸´åºŠä¿¡æ¯</span><br>Â Â Â Â Â Â Â Â Â sample<br>Sample1Â Sample1<br>Sample2Â Sample2<br>Sample3Â Sample3<br>Sample4Â Sample4<br>Sample5Â Sample5<br>&gt;Â sampleMapÂ <span>#Â æŒ‡å®šæ ·æœ¬ç±»å‹ï¼Œé˜²æ­¢æœ‰äº›æ ·æœ¬åä¸ä¸€è‡´ï¼Œæ­¤æ•°æ®é›†ä¸‰ä¸ªåˆ—åå›ºå®šä¸å˜</span><br>Â Â Â Â Â Â Â Â Â Â Â Â Â assayÂ primaryÂ colname<br>1Â Â DNAÂ methylationÂ Sample1Â Sample1<br>2Â Â DNAÂ methylationÂ Sample2Â Sample2<br>3Â Â DNAÂ methylationÂ Sample3Â Sample3<br>4Â Â DNAÂ methylationÂ Sample4Â Sample4<br>5Â Â DNAÂ methylationÂ Sample5Â Sample5<br>6Â Â GeneÂ expressionÂ Sample1Â Sample1<br>7Â Â GeneÂ expressionÂ Sample2Â Sample2<br>8Â Â GeneÂ expressionÂ Sample3Â Sample3<br>9Â Â GeneÂ expressionÂ Sample4Â Sample4<br>10Â GeneÂ expressionÂ Sample5Â Sample5<br></code></pre><p data-tool="mdniceç¼–è¾‘å™¨">è¿™ç§æƒ…å†µä¸‹ï¼Œä½ å°±ç…§ç€è‘«èŠ¦ç”»ç“¢ï¼Œæ¨¡ä»¿æ–‡æ¡£ç»™å‡ºçš„æ•°æ®æ ¼å¼æ„é€ å³å¯</p><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code>distal.probesÂ &lt;-Â get.feature.probe(<br>Â Â genomeÂ =Â <span>"hg19"</span>,Â <br>Â Â met.platformÂ =Â <span>"EPIC"</span><br>)<br><br>maeÂ &lt;-Â createMAE(<br>Â Â expÂ =Â exp,Â <br>Â Â metÂ =Â met,<br>Â Â saveÂ =Â TRUE,<br>Â Â filter.probesÂ =Â distal.probes,<br>Â Â colDataÂ =Â colData,<br>Â Â sampleMapÂ =Â sampleMap,<br>Â Â linearize.expÂ =Â TRUE,<br>Â Â save.filenameÂ =Â <span>"mae.rda"</span>,<br>Â Â met.platformÂ =Â <span>"EPIC"</span>,<br>Â Â genomeÂ =Â <span>"hg19"</span>,<br>Â Â TCGAÂ =Â FALSE<br>)<br></code></pre><h3 data-tool="mdniceç¼–è¾‘å™¨"><span>4. ç­›é€‰å·®å¼‚ç”²åŸºåŒ–æ¢é’ˆ</span></h3><ul data-tool="mdniceç¼–è¾‘å™¨"><li><section>å·®å¼‚åˆ†æï¼š<code>get.diff.meth</code></section></li></ul><p data-tool="mdniceç¼–è¾‘å™¨">ELMERæä¾›äº†ä¸¤ç§æ–¹å¼è¿›è¡Œç­›é€‰å·®å¼‚æ¢é’ˆ</p><p data-tool="mdniceç¼–è¾‘å™¨">ï¼ˆ1ï¼‰supervisedï¼Œå³ä½¿ç”¨æœªé…å¯¹çš„å•ä¾§tæ£€éªŒï¼Œå°†ç»„1ä¸­æ‰€æœ‰æ ·æœ¬çš„çš„DNAç”²åŸºåŒ–æ°´å¹³ä¸ç»„2ä¸­çš„æ‰€æœ‰æ ·æœ¬è¿›è¡Œæ¯”è¾ƒ</p><p data-tool="mdniceç¼–è¾‘å™¨">ï¼ˆ2ï¼‰unsupervisedï¼Œå³æ ¹æ®æ¯ä¸ªæ¢é’ˆçš„ç”²åŸºåŒ–æ°´å¹³åˆ†åˆ«å¯¹ä¸¤ä¸ªç»„çš„æ ·æœ¬å…ˆè¿›è¡Œæ’åºï¼Œæ¯ç»„ä¸­é€‰æ‹©æœ€ä½20%ç”²åŸºåŒ–æ°´å¹³çš„æ ·æœ¬ç”¨äºè¯†åˆ«æ­¤æ¢é’ˆæ˜¯å¦ä½ç”²åŸºåŒ–ï¼Œé«˜ç”²åŸºåŒ–æ¢é’ˆçš„é‰´å®šæ­£å¥½ç›¸åã€‚(åŸå› æ˜¯ï¼šç–¾ç—…ï¼ˆå°¤å…¶æ˜¯è‚¿ç˜¤ï¼‰å…·æœ‰æ˜ç¡®çš„å¼‚è´¨æ€§ï¼Œé€šè¿‡æå–20%çš„å­é›†èƒ½å¤Ÿè¿›ä¸€æ­¥æ‰¾åˆ°ç–¾ç—…ä¸­ç‰¹å®šçš„åˆ†å­äºšå‹)</p><p data-tool="mdniceç¼–è¾‘å™¨">å› æ­¤ï¼Œåœ¨æ— ç›‘ç£æ¨¡å¼ä¸‹ï¼Œåˆ†ææ¯ä¸ªæ¢é’ˆçš„æ ·æœ¬å¯èƒ½åŸºäºæ€»æ ·æœ¬çš„ä¸åŒå­é›†ï¼Œå¯ä»¥è¡¨ç¤ºæ¥è‡ªå¤šä¸ªåˆ†å­äºšå‹çš„æ¢é’ˆé›†çš„ç»“æœã€‚è€Œåœ¨ç›‘ç£æ¨¡å¼ä¸‹ï¼Œæ‰€æœ‰æµ‹è¯•éƒ½åŸºäºæ‰€æœ‰æ ·æœ¬ã€‚</p><figure data-tool="mdniceç¼–è¾‘å™¨"><img data-ratio="1.1711864406779662" data-src="https://mmbiz.qpic.cn/mmbiz_png/PsnauFhHibc5IgrnANu8GpqP6TXajaybAOykrvoNcYZIBOmmR74ia8wIskoBtsk6AN9GDyVT0y4ZEj7AFkmMmh4g/640?wx_fmt=png" data-type="png" data-w="1180" src="https://mmbiz.qpic.cn/mmbiz_png/PsnauFhHibc5IgrnANu8GpqP6TXajaybAOykrvoNcYZIBOmmR74ia8wIskoBtsk6AN9GDyVT0y4ZEj7AFkmMmh4g/640?wx_fmt=png"></figure><p data-tool="mdniceç¼–è¾‘å™¨">è¿™é‡Œä»¥unsupervisedæ–¹å¼ä¸ºä¾‹ï¼Œé€šè¿‡æ¯”è¾ƒä¸åŒç»„é—´æ ·æœ¬ä¸­æ‰€æœ‰è¿œç«¯æ¢é’ˆï¼ˆdistal probeï¼‰çš„ç”²åŸºåŒ–æ°´å¹³ï¼Œç­›é€‰å‡ºadj.P value&lt;0.01ï¼ŒÎ”Î²ï¼0.3çš„ç”²åŸºåŒ–ä½ç‚¹ï¼Œå¹¶ä¸”æ˜¯åœ¨è‚¿ç˜¤ç»„ç»‡ä½ç”²åŸºåŒ–æ°´å¹³çš„ä½ç‚¹</p><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code><span>###Â å·®å¼‚åˆ†æ----</span><br>table(colData(mae)[,<span>"definition"</span>])<br><span>#Â PrimaryÂ solidÂ TumorÂ SolidÂ TissueÂ NormalÂ </span><br><span>#Â 226Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â 8Â </span><br>sig.diffÂ &lt;-Â get.diff.meth(dataÂ =Â mae,Â <br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â group.colÂ =Â <span>"definition"</span>,Â <span>#Â ç»„åˆ«ç±»å‹å¯¹åº”çš„åˆ—å</span><br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â group1Â =Â Â <span>"PrimaryÂ solidÂ Tumor"</span>,Â <span>#Â Â ifÂ directionÂ isÂ hyper,Â getÂ probesÂ hypermethylatedÂ inÂ groupÂ 1Â comparedÂ toÂ groupÂ 2</span><br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â group2Â =Â <span>"SolidÂ TissueÂ Normal"</span>,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â minSubgroupFracÂ =Â 0.2,Â Â <span>#Â é»˜è®¤æŒ‡å®šæ’åºåçš„20%ç”²åŸºåŒ–æ°´å¹³çš„æ ·æœ¬è¿›è¡Œæ¯”è¾ƒ</span><br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â sig.difÂ =Â 0.3,Â <span>#Â ç»„é—´æœ€å°betaå€¼å·®å¼‚</span><br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â diff.dirÂ =Â <span>"hypo"</span>,Â <span>#Â "hypo",Â "hyper"Â orÂ "both"</span><br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â coresÂ =Â 1,Â <span>#Â å¹¶è¡Œæ ¸æ•°</span><br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â dir.outÂ =<span>"result"</span>,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â modeÂ =Â <span>"unsupervised"</span>,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â pvalueÂ =Â 0.01)<br><br>head(sig.diff)<br>dir(pathÂ =Â <span>"result"</span>,Â patternÂ =Â <span>"getMethdiff"</span>)Â <br></code></pre><p data-tool="mdniceç¼–è¾‘å™¨">è¿™é‡Œè¦æ³¨æ„å‚æ•°<code>diff.dir = "hypo"</code>ï¼Œæ„å‘³ç€åœ¨group1ä¸ºä½ç”²åŸºåŒ–æ°´å¹³çš„æ¢é’ˆ</p><p data-tool="mdniceç¼–è¾‘å™¨">"hypo" which is only selecting hypomethylated probes (one tailed test); "hyper" which is only selecting hypermethylated probes (one tailed test); or "both" which are probes differenly methylated (two tailed test).</p><p data-tool="mdniceç¼–è¾‘å™¨">è¾“å‡ºï¼šæ‰€æœ‰å’Œæ˜¾è‘—å·®å¼‚ä½ç‚¹çš„csvç»“æœï¼Œä»¥åŠä¸€å¼ ç«å±±å›¾</p><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code>&gt;Â dim(sig.diff)Â <span>#710Â Â Â 4</span><br>&gt;Â head(sig.diff)<br>Â Â Â Â Â Â Â probeÂ Â Â Â Â Â Â pvalueÂ Primary.solid.Tumor_Minus_Solid.Tissue.Normal<br>1Â cg25340711Â 3.898829e-05Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â -0.3193081<br>2Â cg24852892Â 5.418625e-24Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â -0.4664680<br>3Â cg06358191Â 2.617048e-09Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â -0.3832813<br>4Â cg06646682Â 8.147636e-18Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â -0.4730666<br>5Â cg15844005Â 2.104992e-30Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â -0.3818866<br>6Â cg06205922Â 2.302866e-22Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â -0.3646617<br>Â Â Â Â Â Â adjust.p<br>1Â 5.509361e-05<br>2Â 4.835534e-23<br>3Â 6.310122e-09<br>4Â 4.041818e-17<br>5Â 4.019066e-29<br>6Â 1.742537e-21<br><br>&gt;Â dir(pathÂ =Â <span>"result"</span>,Â patternÂ =Â <span>"getMethdiff"</span>)Â Â <br>[1]Â <span>"getMethdiff.hypo.probes.csv"</span>Â Â Â Â Â Â Â Â Â Â Â Â <br>[2]Â <span>"getMethdiff.hypo.probes.significant.csv"</span><br></code></pre><figure data-tool="mdniceç¼–è¾‘å™¨"><img data-ratio="0.48698884758364314" data-src="https://mmbiz.qpic.cn/mmbiz_png/PsnauFhHibc5IgrnANu8GpqP6TXajaybAbZGFUSZfoK1xOjOUHx9Ww1ZBPTEYyO8Tv7QC1Ax1rRY85NRMrVWjRg/640?wx_fmt=png" data-type="png" data-w="1345" src="https://mmbiz.qpic.cn/mmbiz_png/PsnauFhHibc5IgrnANu8GpqP6TXajaybAbZGFUSZfoK1xOjOUHx9Ww1ZBPTEYyO8Tv7QC1Ax1rRY85NRMrVWjRg/640?wx_fmt=png"></figure><h3 data-tool="mdniceç¼–è¾‘å™¨"><span>5. è¯†åˆ«ç”²åŸºåŒ–-åŸºå› å¯¹</span></h3><p data-tool="mdniceç¼–è¾‘å™¨">è¿™ä¸€æ­¥ï¼Œæˆ‘ä»¬å°†æŠŠè¿œç«¯æ¢é’ˆçš„ç”²åŸºåŒ–æ°´å¹³å’Œé¶åŸºå› çš„è¡¨è¾¾æ°´å¹³è¿›è¡Œç›¸å…³æ€§åˆ†æï¼Œä»è€Œæ„å»ºå‡ºç”²åŸºåŒ–-åŸºå› å¯¹ï¼ˆprobe-geneï¼‰ã€‚è¿™ä¸ªåŒ…å…·ä½“çš„åšæ³•æ˜¯åˆ†åˆ«æ‰¾åˆ°å·®å¼‚ç”²åŸºåŒ–è¿œç«¯æ¢é’ˆçš„ä¸Šæ¸¸å’Œä¸‹æ¸¸æœ€è¿‘çš„10ä¸ªåŸºå› ï¼Œåˆ†æå…¶ä½™å¯¹åº”æ¢é’ˆç”²åŸºåŒ–æ°´å¹³æ˜¯å¦å­˜åœ¨è´Ÿç›¸å…³ï¼Œä»è€Œç­›é€‰å‡ºæ¥çš„æ½œåœ¨probe-geneå¯¹ã€‚</p><p data-tool="mdniceç¼–è¾‘å™¨">åŒæ ·æä¾›äº†ä¸¤ç±»æ–¹å¼ï¼š</p><p data-tool="mdniceç¼–è¾‘å™¨">ï¼ˆ1ï¼‰supervisedï¼Œå³é€šè¿‡ä¸Šæ¸¸åˆ†æç–¾ç—…ç»„vs.æ­£å¸¸ç»„å·®å¼‚è¿œç«¯æ¢é’ˆï¼Œåœ¨æ‰€æœ‰æ ·æœ¬ä¸­è¿›è¡Œéå‚æ•°æ£€ï¼Œæ‰¾å‡ºåŒæ—¶åœ¨ç–¾ç—…ç»„é«˜è¡¨è¾¾çš„åŸºå› </p><p data-tool="mdniceç¼–è¾‘å™¨">ï¼ˆ2ï¼‰unsupervisedï¼Œæ ¹æ®åŸºå› çš„distal probeç”²åŸºåŒ–æ°´å¹³è¿›è¡Œæ’åºï¼Œæå–å‡ºç”²åŸºåŒ–æ°´å¹³åˆ†åˆ«åœ¨å‰20%å’Œå20%çš„æ ·æœ¬ï¼Œåˆ†åˆ«ä¸ºMå’ŒUç»„ï¼Œé€šè¿‡Uæ£€éªŒæ–¹å¼æ¯”è¾ƒä¸¤ç»„é—´åŸºå› è¡¨è¾¾æ°´å¹³,å¦‚æœåœ¨Mç»„çš„è¡¨è¾¾æ°´å¹³æ˜¾è‘—ä½äºUç»„ï¼Œå¹¶é€šè¿‡è¿­ä»£çš„æ–¹å¼è®¡ç®—å‡ºæ‰€æœ‰è¿œç«¯æ¢é’ˆå’Œé¶åŸºå› çš„ç›¸å…³æ€§P value,ä»è€Œç­›é€‰å‡ºå…·æœ‰æ„ä¹‰çš„å…³ç³»å¯¹ã€‚</p><figure data-tool="mdniceç¼–è¾‘å™¨"><img data-ratio="0.5716993051168667" data-src="https://mmbiz.qpic.cn/mmbiz_png/PsnauFhHibc5IgrnANu8GpqP6TXajaybADAN8ElYQ7sSz911LMuEob6dsqPfgwtIFEffQzobHJyGSR0qes5iakMg/640?wx_fmt=png" data-type="png" data-w="3166" src="https://mmbiz.qpic.cn/mmbiz_png/PsnauFhHibc5IgrnANu8GpqP6TXajaybADAN8ElYQ7sSz911LMuEob6dsqPfgwtIFEffQzobHJyGSR0qes5iakMg/640?wx_fmt=png"></figure><p><br></p><ul data-tool="mdniceç¼–è¾‘å™¨"><li><section>å¯»æ‰¾æ¢é’ˆä¸Šä¸‹æ¸¸åŸºå› ï¼š<code>GetNearGenes</code></section></li></ul><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code><span>###Â è¯†åˆ«probe-geneå¯¹-----</span><br><span>#Â åŠ è½½æˆ‘ä»¬ä¹‹å‰çš„æ•°æ®</span><br>maeÂ &lt;-Â get(load(<span>"mae.rda"</span>))<br><span>#åŠ è½½æˆ‘ä»¬ç­›é€‰å¾—åˆ°çš„å·®å¼‚ç”²åŸºåŒ–è¿œç«¯æ¢é’ˆï¼ˆdistalÂ probeï¼‰</span><br>sig.diffÂ &lt;-Â read.csv(<span>"result/getMethdiff.hypo.probes.significant.csv"</span>)<br><span>#æ‰¾åˆ°å·®å¼‚ç”²åŸºåŒ–è¿œç«¯æ¢é’ˆï¼ˆdistalÂ probeï¼‰ä¸Šä¸‹æ¸¸10ä¸ªåŸºå› </span><br>nearGenesÂ &lt;-Â GetNearGenes(dataÂ =Â mae,Â <br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â probesÂ =Â sig.diff<span>$probe</span>,Â <br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â numFlankingGenesÂ =Â 20)Â <span>#Â åˆ†åˆ«æ˜¯ä¸Šæ¸¸10ä¸ªå’Œä¸‹æ¸¸10åŸºå› </span><br><span>##Â SearchingÂ forÂ theÂ 20Â nearÂ genes</span><br><span>##Â IdentifyingÂ geneÂ positionÂ forÂ eachÂ probe</span><br>head(nearGenes)<br></code></pre><p><br></p><ul data-tool="mdniceç¼–è¾‘å™¨"><li><section>è¯†åˆ«probe-geneå¯¹ï¼š<code>get.pair</code></section></li></ul><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code><span>#å› ä¸ºå‰é¢ä½¿ç”¨unsupervisedï¼Œæˆ‘ä»¬è¿™è€Œä¹Ÿç”¨unsupervised</span><br>hypo.pairÂ &lt;-Â get.pair(dataÂ =Â mae,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â group.colÂ =Â <span>"definition"</span>,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â group1Â =Â Â <span>"PrimaryÂ solidÂ Tumor"</span>,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â group2Â =Â <span>"SolidÂ TissueÂ Normal"</span>,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â nearGenesÂ =Â nearGenes,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â minSubgroupFracÂ =Â 0.4,<span>#Â é»˜è®¤åˆ›å»ºMå’ŒUç»„æ ·æœ¬çš„æ¯”ä¾‹</span><br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â modeÂ =Â <span>"unsupervised"</span>,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â permu.dirÂ =Â <span>"result/permu"</span>,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â permu.sizeÂ =Â 100,Â <span>#Â ä¸€èˆ¬è¦è®¾ç½®è¿­ä»£æ¬¡æ•°ä¸º10000ï¼Œå¤„äºæ¼”ç¤ºç›®çš„å‡å°‘æ¬¡æ•°</span><br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â raw.pvalueÂ =Â 0.05,Â Â Â <br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â PeÂ =Â 0.01,Â <span>#Â På€¼ï¼Œé»˜è®¤0.001</span><br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â filter.probesÂ =Â TRUE,Â <br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â filter.percentageÂ =Â 0.05,Â <span>#Â æœ€å°è€ƒè™‘çš„æ ·æœ¬æ¯”ä¾‹</span><br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â filter.portionÂ =Â 0.3,Â <span>#Â ç”²åŸºåŒ–æ°´å¹³æˆªæ–­ç‚¹</span><br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â dir.outÂ =Â <span>"result"</span>,<span>#ç»“æœåŒæ ·è¾“å‡ºåˆ°resultç›®å½•</span><br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â coresÂ =Â 1,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â labelÂ =Â <span>"hypo"</span>)<br><span>#Â SelectingÂ UÂ (unmethylated)Â andÂ MÂ (methylated)Â groups.Â EachÂ groupsÂ hasÂ 20%Â ofÂ samples</span><br><span>#Â -------------------</span><br><span>#Â Â Â *Â FilteringÂ probes</span><br><span>#Â -------------------</span><br><span>#Â Â Â ForÂ moreÂ informationÂ seeÂ functionÂ preAssociationProbeFiltering</span><br><span>#Â MakingÂ sureÂ weÂ haveÂ atÂ leastÂ 5%Â ofÂ betaÂ valuesÂ lesserÂ thanÂ 0.3Â andÂ 5%Â ofÂ betaÂ valuesÂ greaterÂ 0.3.</span><br><span>#Â RemovingÂ 592Â probesÂ outÂ ofÂ 1642</span><br><span>#Â CalculatingÂ PpÂ (probeÂ -Â gene)Â forÂ allÂ nearbyÂ genes</span><br><span>#Â |============================================|100%Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â CompletedÂ afterÂ 5Â sÂ </span><br><span>#Â FileÂ created:Â result/getPair.hypo.all.pairs.statistic.csvÂ Â Â Â Â Â Â Â Â Â Â Â Â 474.73MB/s</span><br><span>#Â CalculatingÂ PrÂ (randomÂ probeÂ -Â gene).Â PermutatingÂ Â 100Â probesÂ forÂ 747Â genes</span><br><span>#Â CalculateÂ empiricalÂ PÂ value.Â Â Â Â Â Â Â Â Â Â Â Â Â Â </span><br>head(hypo.pair)<br>dim(hypo.pair)Â <span>#Â [1]Â 167Â Â 7</span><br></code></pre><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code>&gt;Â head(nearGenes)<br><span>#Â AÂ tibble:Â 6Â Ã—Â 5</span><br><span>#Â Groups:Â Â Â IDÂ [1]</span><br>Â Â IDÂ Â Â Â Â Â Â Â Â GeneIDÂ Â Â Â Â Â Â Â Â Â SymbolÂ Â DistanceÂ SideÂ <br>Â Â &lt;chr&gt;Â Â Â Â Â Â &lt;chr&gt;Â Â Â Â Â Â Â Â Â Â Â &lt;chr&gt;Â Â Â Â Â Â &lt;dbl&gt;Â &lt;chr&gt;<br>1Â cg00045114Â ENSG00000076258Â FMO4Â Â Â Â -1362935Â L10Â Â <br>2Â cg00045114Â ENSG00000117523Â PRRC2CÂ Â -1111508Â L9Â Â Â <br>3Â cg00045114Â ENSG00000034971Â MYOCÂ Â Â Â -1052335Â L8Â Â Â <br>4Â cg00045114Â ENSG00000117533Â VAMP4Â Â Â Â -962771Â L7Â Â Â <br>5Â cg00045114Â ENSG00000010165Â METTL13Â Â -890995Â L6Â Â Â <br>6Â cg00045114Â ENSG00000197959Â DNM3Â Â Â Â Â -286552Â L5Â <br><br>&gt;Â head(hypo.pair)<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â ProbeÂ Â Â Â Â Â Â Â Â Â GeneIDÂ SymbolÂ Distance<br>cg20701183.ENSG00000143013Â cg20701183Â ENSG00000143013Â Â Â LMO4Â Â Â Â Â Â Â Â 0<br>cg19403323.ENSG00000143469Â cg19403323Â ENSG00000143469Â Â SYT14Â Â Â Â 87476<br>cg12213388.ENSG00000143674Â cg12213388Â ENSG00000143674Â Â Â MLK4Â Â -984182<br>cg26607897.ENSG00000143199Â cg26607897Â ENSG00000143199Â ADCY10Â Â Â 267784<br>cg10574861.ENSG00000143013Â cg10574861Â ENSG00000143013Â Â Â LMO4Â Â Â Â Â Â Â Â 0<br>cg26607897.ENSG00000143147Â cg26607897Â ENSG00000143147Â GPR161Â Â Â 543156<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â SidesÂ Â Â Â Â Â Â Â Raw.pÂ Â Â Â Â Â Â Â Â Pe<br>cg20701183.ENSG00000143013Â Â Â Â L1Â 7.453984e-14Â 0.00990099<br>cg19403323.ENSG00000143469Â Â Â Â R1Â 1.671937e-12Â 0.00990099<br>cg12213388.ENSG00000143674Â Â Â Â L4Â 2.527644e-12Â 0.00990099<br>cg26607897.ENSG00000143199Â Â Â Â R3Â 4.593610e-12Â 0.00990099<br>cg10574861.ENSG00000143013Â Â Â Â L1Â 4.770162e-12Â 0.00990099<br>cg26607897.ENSG00000143147Â Â Â Â R6Â 8.048248e-12Â 0.00990099<br><br>&gt;Â dir(pathÂ =Â <span>"result"</span>,Â patternÂ =Â <span>"getPair"</span>)Â <br>[1]Â <span>"getPair.hypo.all.pairs.statistic.csv"</span>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â <br>[2]Â <span>"getPair.hypo.pairs.significant.csv"</span>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â <br>[3]Â <span>"getPair.hypo.pairs.statistic.with.empirical.pvalue.csv"</span><br></code></pre><h3 data-tool="mdniceç¼–è¾‘å™¨"><span>6. ç­›é€‰motif</span></h3><p data-tool="mdniceç¼–è¾‘å™¨">æå–æ¢é’ˆä¸Šä¸‹æ¸¸250bpåŒºåŸŸçš„ç¢±åŸºåºåˆ—ï¼Œä»è€Œæ‰¾åˆ°å¯Œé›†çš„motifã€‚æ¥ä¸‹æ¥å†é€šè¿‡è½¬å½•å› å­ï¼ˆTFï¼‰ç»“åˆmotifæ•°æ®åº“ï¼ˆHOMERæ•°æ®åº“ï¼‰ï¼Œä»è€Œé¢„æµ‹å‡ºmotifç»“åˆçš„è½¬å½•å› å­</p><ul data-tool="mdniceç¼–è¾‘å™¨"><li><section>ç­›é€‰motifï¼š<code>get.enriched.motif</code></section></li></ul><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code><span>###Â è¯†åˆ«motif-----</span><br><br>enriched.motifÂ &lt;-Â get.enriched.motif(dataÂ =Â mae,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â probesÂ =Â hypo.pair<span>$Probe</span>,Â <br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â dir.outÂ =Â <span>"result"</span>,Â <br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â labelÂ =Â <span>"hypo"</span>,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â min.incidenceÂ =Â 1,Â <span>#Â Â theÂ minimumÂ incidenceÂ ofÂ motif</span><br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â lower.ORÂ =Â 1.1)Â <span>#Â motifæœ€ä½95%ç½®ä¿¡åŒºé—´ORé˜ˆå€¼</span><br><span>##Â LoadingÂ object:Â Probes.motif.hg19.450K</span><br><br>names(enriched.motif)<br>enriched.motif[[1]]<br></code></pre><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code><span>#Â è¯†åˆ«å‡ºå…­ç§motif</span><br>&gt;Â names(enriched.motif)<br>[1]Â <span>"HXA9_HUMAN.H11MO.0.B"</span>Â Â <span>"FOXH1_HUMAN.H11MO.0.A"</span>Â <span>"PO5F1_HUMAN.H11MO.1.A"</span><br>[4]Â <span>"HXB13_HUMAN.H11MO.0.A"</span>Â <span>"IRF9_HUMAN.H11MO.0.C"</span>Â Â <span>"SOX10_HUMAN.H11MO.1.A"</span><br><br><span>#Â å±•ç¤ºç¬¬ä¸€ä¸ªmotifåŒ…å«çš„ç”²åŸºåŒ–æ¢é’ˆ</span><br>&gt;Â enriched.motif[[1]]<br>Â [1]Â <span>"cg13067635"</span>Â <span>"cg24873093"</span>Â <span>"cg25210796"</span>Â <span>"cg18345456"</span>Â <span>"cg14094320"</span><br>Â [6]Â <span>"cg17901924"</span>Â <span>"cg22736624"</span>Â <span>"cg12213388"</span>Â <span>"cg13305336"</span>Â <span>"cg26607897"</span><br>[11]Â <span>"cg11718886"</span>Â <span>"cg24593832"</span>Â <span>"cg25729466"</span>Â <span>"cg24446417"</span><br><br><span>#Â è¾“å‡ºçš„ç»“æœæ–‡ä»¶</span><br>&gt;Â dir(pathÂ =Â <span>"result"</span>,Â patternÂ =Â <span>"getMotif"</span>)Â <br>[1]Â <span>"getMotif.hypo.enriched.motifs.rda"</span>Â Â <span>"getMotif.hypo.motif.enrichment.csv"</span><br><br><span>#Â è¾“å‡ºçš„å›¾ç‰‡</span><br>&gt;Â dir(pathÂ =Â <span>"result"</span>,Â patternÂ =Â <span>"motif.enrichment.pdf"</span>)Â <br>[1]Â <span>"hypo.quality.A-DS_with_summary.motif.enrichment.pdf"</span><br>[2]Â <span>"hypo.quality.A-DS.motif.enrichment.pdf"</span><br></code></pre><figure data-tool="mdniceç¼–è¾‘å™¨"><img data-ratio="1.190800681431005" data-src="https://mmbiz.qpic.cn/mmbiz_png/PsnauFhHibc5IgrnANu8GpqP6TXajaybA6tp1m6iauC7c8eicaThPnxgSLfxz5vgN6zBvnT7n8CDJ1admKy5aib9gQ/640?wx_fmt=png" data-type="png" data-w="587" src="https://mmbiz.qpic.cn/mmbiz_png/PsnauFhHibc5IgrnANu8GpqP6TXajaybA6tp1m6iauC7c8eicaThPnxgSLfxz5vgN6zBvnT7n8CDJ1admKy5aib9gQ/640?wx_fmt=png"></figure><h3 data-tool="mdniceç¼–è¾‘å™¨"><span>7. ç¡®å®šè½¬å½•å› å­TF</span></h3><p data-tool="mdniceç¼–è¾‘å™¨">è¿™ä¸€æ­¥ï¼ŒELMERåŒ…å°†ä¸Šä¸€æ­¥æˆ‘ä»¬é€šè¿‡å¯Œé›†åˆ†ææ‰¾åˆ°å¯Œé›†çš„motifä»¥åŠå¯¹åº”motifçš„è½¬å½•å› å­è¿›è¡Œåˆ†æï¼Œç­›é€‰å‡ºå¯¹åº”è½¬å½•æ°´å¹³å‘ç”Ÿæ”¹å˜çš„è½¬å½•å› å­ã€‚</p><p data-tool="mdniceç¼–è¾‘å™¨">åŒæ ·åœ°ï¼Œæä¾›äº†ä¸¤ç§æ–¹å¼ï¼š</p><p data-tool="mdniceç¼–è¾‘å™¨">ï¼ˆ1ï¼‰unsupervisedï¼Œå³å°†å­˜åœ¨ç›¸åŒmotifçš„è¿œç«¯æ¢é’ˆï¼ˆdistal probeï¼‰æ ¹æ®ç”²åŸºåŒ–æ°´å¹³åˆ†ä¸ºé«˜ç”²åŸºåŒ–ç»„ï¼ˆä¸€èˆ¬å–å‰20%ï¼‰å’Œä½ç”²åŸºåŒ–ç»„ï¼ˆå20%ï¼‰ï¼Œæ¯”è¾ƒå…¶ä¸¤ç»„é—´å¯¹åº”TFçš„è¡¨è¾¾å€¼æ˜¯å¦å­˜åœ¨å·®å¼‚ã€‚</p><p data-tool="mdniceç¼–è¾‘å™¨">ï¼ˆ2ï¼‰Supervisedï¼Œå³å¯¹ç–¾ç—…ç»„vs.å¯¹ç…§ç»„ç›¸åŒmotifå¯¹åº”TFè¡¨è¾¾å€¼è¿›è¡Œåˆ†æã€‚è¿›è€Œç­›é€‰å‡ºå‰5% P valueæœ€å°çš„TFsï¼Œå¹¶è§†ä¸ºæ½œåœ¨çš„ä¸Šæ¸¸è°ƒæ§è½¬å½•å› å­ã€‚</p><ul data-tool="mdniceç¼–è¾‘å™¨"><li><section>ç¡®å®šè½¬å½•å› å­TFï¼š<code>get.TFs</code></section></li></ul><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code><span>##Â ç¡®å®šè½¬å½•å› å­TF-----</span><br>TFÂ &lt;-Â get.TFs(dataÂ =Â mae,Â <br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â group.colÂ =Â <span>"definition"</span>,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â group1Â =Â Â <span>"PrimaryÂ solidÂ Tumor"</span>,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â group2Â =Â <span>"SolidÂ TissueÂ Normal"</span>,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â modeÂ =Â <span>"unsupervised"</span>,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â enriched.motifÂ =Â enriched.motif,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â dir.outÂ =Â <span>"result"</span>,Â <br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â coresÂ =Â 1,Â <br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â labelÂ =Â <span>"hypo"</span>)<br>head(TF)<br>dir(pathÂ =Â <span>"result"</span>,Â patternÂ =Â <span>"getTF"</span>)Â <br></code></pre><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code>&gt;Â dir(pathÂ =Â <span>"result"</span>,Â patternÂ =Â <span>"getTF"</span>)Â <br>[1]Â <span>"getTF.hypo.significant.TFs.with.motif.summary.csv"</span><br>[2]Â <span>"getTF.hypo.TFs.with.motif.pvalue.rda"</span>Â <br></code></pre><figure data-tool="mdniceç¼–è¾‘å™¨"><img data-ratio="0.19132029339853301" data-src="https://mmbiz.qpic.cn/mmbiz_png/PsnauFhHibc5IgrnANu8GpqP6TXajaybAWzVbcegRL5ibrdIBcRQuUPlRbJUx7Z345u5icNKKFG3WNibNrslKXvRXg/640?wx_fmt=png" data-type="png" data-w="1636" src="https://mmbiz.qpic.cn/mmbiz_png/PsnauFhHibc5IgrnANu8GpqP6TXajaybAWzVbcegRL5ibrdIBcRQuUPlRbJUx7Z345u5icNKKFG3WNibNrslKXvRXg/640?wx_fmt=png"></figure><p data-tool="mdniceç¼–è¾‘å™¨"><br></p><p data-tool="mdniceç¼–è¾‘å™¨">æ€»ç»“ä¸€ä¸‹ï¼Œæˆ‘ä»¬é€šè¿‡åŒä¸€æ‰¹æ ·æœ¬çš„ç”²åŸºåŒ–æ•°æ®å’Œè½¬å½•ç»„æ•°æ®ï¼Œé¦–å…ˆæ‰¾åˆ°å…·æœ‰å·®å¼‚çš„è¿œç«¯æ¢é’ˆï¼ˆdistal probeï¼‰ï¼Œå¹¶æ ¹æ®è¿œç«¯æ¢é’ˆå’Œå¯¹åº”åŸºå› çš„è¡¨è¾¾æ°´å¹³æ„å»ºå‡ºprobe-geneå¯¹ã€‚å¹¶å°†probe-geneå¯¹ä¸­çš„probeè¿›ä¸€æ­¥è¿›è¡Œmotifå¯Œé›†åˆ†æï¼Œæ‰¾å‡ºæ½œåœ¨å¯Œé›†motifï¼Œæœ€åæ ¹æ®å¯Œé›†çš„motifæ¨æµ‹å‡ºä¸Šæ¸¸çš„è°ƒæ§TFs</p><h3 data-tool="mdniceç¼–è¾‘å™¨"><span>8. å„ç§å¯è§†åŒ–</span></h3><ul data-tool="mdniceç¼–è¾‘å™¨"><li><section>æ•£ç‚¹å›¾ï¼š<code>scatter.plot</code></section></li><li><section>æŸ“è‰²ä½“ç¤ºæ„å›¾ï¼š<code>schematic.plot</code></section></li><li><section>motifå¯Œé›†æ£®æ—å›¾ï¼š<code>motif.enrichment.plot</code></section></li><li><section>è½¬å½•å› å­æ’åºå›¾ï¼š<code>TF.rank.plot</code></section></li><li><section>çƒ­å›¾ï¼š<code>heatmapPairs</code></section></li></ul><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code><span>###Â å„ç§å¯è§†åŒ–----</span><br><span>#####Â 1.æ•£ç‚¹å›¾----</span><br><span>#åŠ è½½æ•°æ®</span><br>maeÂ &lt;-Â get(load(<span>"mae.rda"</span>))<br>load(<span>"result/getMotif.hypo.enriched.motifs.rda"</span>)<br><span>#æ¢é’ˆå’Œä¸Šä¸‹æ¸¸åˆ†åˆ«10ä¸ªåŸºå› çš„æ•£ç‚¹å›¾</span><br>scatter.plot(dataÂ =Â mae,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â byProbeÂ =Â list(probeÂ =Â c(<span>"cg19403323"</span>),Â <br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â numFlankingGenesÂ =Â 20),Â <br>Â Â Â Â Â Â Â Â Â Â Â Â Â categoryÂ =Â <span>"definition"</span>,Â <br>Â Â Â Â Â Â Â Â Â Â Â Â Â lmÂ =Â TRUE,Â <span>#Â ç”»å‡ºçº¿æ¡</span><br>Â Â Â Â Â Â Â Â Â Â Â Â Â saveÂ =Â FALSE)<br><span>##Â SearchingÂ forÂ theÂ 20Â nearÂ genes</span><br><span>##Â IdentifyingÂ geneÂ positionÂ forÂ eachÂ probe</span><br><br><span>#åªå±•ç°ä¸€å¯¹probe-geneæ•£ç‚¹å›¾</span><br>scatter.plot(dataÂ =Â mae,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â byPairÂ =Â list(probeÂ =Â c(<span>"cg19403323"</span>),Â geneÂ =Â c(<span>"ENSG00000143469"</span>)),Â <br>Â Â Â Â Â Â Â Â Â Â Â Â Â categoryÂ =Â <span>"definition"</span>,Â saveÂ =Â TRUE,Â lm_lineÂ =Â TRUE)<br><br><br><span>#ç‰¹å®šè½¬å½•å› å­è¡¨è¾¾æ°´å¹³ä»¥åŠç‰¹å®šmotifå¯¹åº”æ¢é’ˆç”²åŸºåŒ–æ°´å¹³æ•£ç‚¹å›¾</span><br>scatter.plot(dataÂ =Â mae,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â byTFÂ =Â list(TFÂ =Â c(<span>"SOX2"</span>),<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â probeÂ =Â enriched.motif[[names(enriched.motif)[1]]]),Â <br>Â Â Â Â Â Â Â Â Â Â Â Â Â categoryÂ =Â <span>"definition"</span>,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â saveÂ =Â F,Â <span>#ä¿å­˜ä¸ºPDFç‰ˆæœ¬</span><br>Â Â Â Â Â Â Â Â Â Â Â Â Â lm_lineÂ =Â TRUE)<br></code></pre><figure data-tool="mdniceç¼–è¾‘å™¨"><img data-ratio="0.6507936507936508" data-src="https://mmbiz.qpic.cn/mmbiz_png/PsnauFhHibc5IgrnANu8GpqP6TXajaybAbK5C64CgiauH2C6CtxGCS5Ceom0qun5c8wVQEYvRAuN2tJk19NK8k5Q/640?wx_fmt=png" data-type="png" data-w="1008" src="https://mmbiz.qpic.cn/mmbiz_png/PsnauFhHibc5IgrnANu8GpqP6TXajaybAbK5C64CgiauH2C6CtxGCS5Ceom0qun5c8wVQEYvRAuN2tJk19NK8k5Q/640?wx_fmt=png"></figure><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code><span>#####Â 2.æŸ“è‰²ä½“ç¤ºæ„å›¾----</span><br><span>#Â ä½¿ç”¨genome browserçš„æ–¹å¼å±•ç¤ºè¿œç«¯æ¢é’ˆï¼ˆdistal probeï¼‰ç”²åŸºåŒ–æ°´å¹³å’Œé™„è¿‘20ä¸ªåŸºå› åœ¨æŸ“è‰²ä½“ä¸Šçš„ä½ç½®ã€‚</span><br>library(GenomicInteractions)<br><span>#åŠ è½½probe-geneå¯¹</span><br>pairÂ &lt;-Â read.csv(<span>"result/getPair.hypo.pairs.significant.csv"</span>)<br><span>#å±•ç°æ¢é’ˆé™„è¿‘çš„gene</span><br>schematic.plot(pairÂ =Â pair,Â <br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â dataÂ =Â mae,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â group.colÂ =Â <span>"definition"</span>,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â byProbeÂ =Â pair<span>$Probe</span>[1],<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â saveÂ =Â FALSE)<br><span>##Â SearchingÂ forÂ theÂ 20Â nearÂ genes</span><br><span>##Â IdentifyingÂ geneÂ positionÂ forÂ eachÂ probe</span><br><br><span>#å±•ç¤ºåŸºå› é™„è¿‘çš„probe</span><br>schematic.plot(pairÂ =Â pair,Â <br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â dataÂ =Â mae,Â Â Â <br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â group.colÂ =Â <span>"definition"</span>,Â <br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â byGeneÂ =Â pair<span>$GeneID</span>[1],<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â saveÂ =Â FALSE)<br><br></code></pre><figure data-tool="mdniceç¼–è¾‘å™¨"><img data-ratio="0.547395388556789" data-src="https://mmbiz.qpic.cn/mmbiz_png/PsnauFhHibc5IgrnANu8GpqP6TXajaybAJKqicMSTiaSzdPnOrmlLpTAQJib9C0sHjrrjmZeP9Vuq2JNMXicuVbfw0A/640?wx_fmt=png" data-type="png" data-w="1171" src="https://mmbiz.qpic.cn/mmbiz_png/PsnauFhHibc5IgrnANu8GpqP6TXajaybAJKqicMSTiaSzdPnOrmlLpTAQJib9C0sHjrrjmZeP9Vuq2JNMXicuVbfw0A/640?wx_fmt=png"></figure><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code><span>#####Â 3.motifå¯Œé›†æ£®æ—å›¾-----</span><br><span>#åˆæ­¥å±•ç¤º</span><br>motif.enrichment.plot(motif.enrichmentÂ =Â <span>"result/getMotif.hypo.motif.enrichment.csv"</span>,Â <br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â significantÂ =Â list(ORÂ =Â 1.5,lowerORÂ =Â 1.3),Â <br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â labelÂ =Â <span>"hypo"</span>,Â <br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â saveÂ =Â FALSE)<br><br><span>#å±•ç°ORå€¼å’Œæ¢é’ˆæ•°</span><br>motif.enrichment.plot(motif.enrichmentÂ =Â <span>"result/getMotif.hypo.motif.enrichment.csv"</span>,Â <br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â significantÂ =Â list(ORÂ =Â 1.5,lowerORÂ =Â 1.3),Â <br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â labelÂ =Â <span>"hypo"</span>,Â <br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â summaryÂ =Â TRUE,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â saveÂ =Â FALSE)<br></code></pre><figure data-tool="mdniceç¼–è¾‘å™¨"><img data-ratio="0.5552742616033756" data-src="https://mmbiz.qpic.cn/mmbiz_png/PsnauFhHibc5IgrnANu8GpqP6TXajaybAiaa2kId4WwEs1L2zcrkrxAlCZicbdicYVVC2ZiaVQeCyric917sfIIHFZ3g/640?wx_fmt=png" data-type="png" data-w="1185" src="https://mmbiz.qpic.cn/mmbiz_png/PsnauFhHibc5IgrnANu8GpqP6TXajaybAiaa2kId4WwEs1L2zcrkrxAlCZicbdicYVVC2ZiaVQeCyric917sfIIHFZ3g/640?wx_fmt=png"></figure><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code><span>#####Â 4.è½¬å½•å› å­æ’åºå›¾-----</span><br>load(<span>"result/getTF.hypo.TFs.with.motif.pvalue.rda"</span>)<br>motifÂ &lt;-Â colnames(TF.meth.cor)[1]<br>TF.rank.plot(motif.pvalueÂ =Â TF.meth.cor,Â <br>Â Â Â Â Â Â Â Â Â Â Â Â Â motifÂ =Â motif,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â saveÂ =Â FALSE)<br><span>##Â RetrievingÂ TFClassÂ familyÂ classificationÂ fromÂ ELMER.data.</span><br><span>##Â RetrievingÂ TFClassÂ subfamilyÂ classificationÂ fromÂ ELMER.data.</span><br><span>##Â |Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â |Â Â 0%Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â |====================================================|100%Â ~0Â sÂ remainingÂ Â Â Â Â Â Â |====================================================|100%Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â Â CompletedÂ afterÂ 0Â s</span><br><span>##Â $HXA9_HUMAN.H11MO.0.B</span><br><span>##Â Warning:Â ggrepel:Â 69Â unlabeledÂ dataÂ pointsÂ (tooÂ manyÂ overlaps).Â Consider</span><br><span>##Â increasingÂ max.overlaps</span><br></code></pre><figure data-tool="mdniceç¼–è¾‘å™¨"><img data-ratio="0.7960264900662252" data-src="https://mmbiz.qpic.cn/mmbiz_png/PsnauFhHibc5IgrnANu8GpqP6TXajaybAq63yMoc8N3ncD41Q0hqPicV4w2DZnST0zpCiaIeKvhh3J74eSXiaCicABQ/640?wx_fmt=png" data-type="png" data-w="755" src="https://mmbiz.qpic.cn/mmbiz_png/PsnauFhHibc5IgrnANu8GpqP6TXajaybAq63yMoc8N3ncD41Q0hqPicV4w2DZnST0zpCiaIeKvhh3J74eSXiaCicABQ/640?wx_fmt=png"></figure><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code><span>#####Â 5.Â çƒ­å›¾-----</span><br>heatmapPairs(dataÂ =Â mae,Â <br>Â Â Â Â Â Â Â Â Â Â Â Â Â group.colÂ =Â <span>"definition"</span>,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â group1Â =Â <span>"PrimaryÂ solidÂ Tumor"</span>,Â <br>Â Â Â Â Â Â Â Â Â Â Â Â Â annotation.colÂ =Â c(<span>"years_smoked"</span>,<span>"gender"</span>),<br>Â Â Â Â Â Â Â Â Â Â Â Â Â group2Â =Â <span>"SolidÂ TissueÂ Normal"</span>,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â pairsÂ =Â pair,<br>Â Â Â Â Â Â Â Â Â Â Â Â Â filenameÂ =Â Â NULL)<br></code></pre><figure data-tool="mdniceç¼–è¾‘å™¨"><img data-ratio="0.8526077097505669" data-src="https://mmbiz.qpic.cn/mmbiz_png/PsnauFhHibc5IgrnANu8GpqP6TXajaybANrvHSID6yEMUvZAWVpUIShZgickIy53t6ibmSN9jMWcdUOmKVG88acAA/640?wx_fmt=png" data-type="png" data-w="882" src="https://mmbiz.qpic.cn/mmbiz_png/PsnauFhHibc5IgrnANu8GpqP6TXajaybANrvHSID6yEMUvZAWVpUIShZgickIy53t6ibmSN9jMWcdUOmKVG88acAA/640?wx_fmt=png"></figure><h3 data-tool="mdniceç¼–è¾‘å™¨"><span></span>9. ç›´æ¥ä¸‹è½½TCGAæ•°æ®<span></span></h3><p data-tool="mdniceç¼–è¾‘å™¨">å½“ç„¶ï¼Œå¯ä»¥ç›´æ¥ä½¿ç”¨<code>TCGA.pipe</code>ï¼Œè”åˆråŒ…TCGAbiolinksç›´æ¥ä¸‹è½½TCGAçš„æ•°æ®åˆ†æ</p><p data-tool="mdniceç¼–è¾‘å™¨">The following command will do distal DNA methylation analysis and predict putative target genes, motif analysis and identify regulatory transcription factors.</p><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code>TCGA.pipe(<span>"LUSC"</span>,<br>Â Â Â Â Â Â Â Â Â Â wdÂ =Â <span>"./ELMER.example"</span>,<br>Â Â Â Â Â Â Â Â Â Â coresÂ =Â parallel::detectCores()/2,<br>Â Â Â Â Â Â Â Â Â Â modeÂ =Â <span>"unsupervised"</span>,<br>Â Â Â Â Â Â Â Â Â Â permu.sizeÂ =Â 300,<br>Â Â Â Â Â Â Â Â Â Â PeÂ =Â 0.01,<br>Â Â Â Â Â Â Â Â Â Â analysisÂ =Â c(<span>"distal.probes"</span>,<span>"diffMeth"</span>,<span>"pair"</span>,<span>"motif"</span>,<span>"TF.search"</span>),<br>Â Â Â Â Â Â Â Â Â Â diff.dirÂ =Â <span>"hypo"</span>,<br>Â Â Â Â Â Â Â Â Â Â rm.chrÂ =Â paste0(<span>"chr"</span>,c(<span>"X"</span>,<span>"Y"</span>)))<br></code></pre><ul data-tool="mdniceç¼–è¾‘å™¨"><li><section>è¿˜å¯ä»¥ç›´æ¥åœ¨çº¿åˆ†æ</section></li></ul><pre data-tool="mdniceç¼–è¾‘å™¨"><span></span><code><span>###Â åœ¨çº¿ç‰ˆæœ¬----</span><br>BiocManager::install(<span>"TCGAbiolinksGUI"</span>)<br>library(TCGAbiolinksGUI)<br>TCGAbiolinksGUI()<br></code></pre><figure data-tool="mdniceç¼–è¾‘å™¨"><img data-ratio="0.5686433793663688" data-src="https://mmbiz.qpic.cn/mmbiz_png/PsnauFhHibc5IgrnANu8GpqP6TXajaybAAVLYyADUfZa8If58ep4iaW39LeTIHdUvcDibpjb9XlgibPOMyL58hicLrw/640?wx_fmt=png" data-type="png" data-w="1231" src="https://mmbiz.qpic.cn/mmbiz_png/PsnauFhHibc5IgrnANu8GpqP6TXajaybAAVLYyADUfZa8If58ep4iaW39LeTIHdUvcDibpjb9XlgibPOMyL58hicLrw/640?wx_fmt=png"></figure><h3 data-tool="mdniceç¼–è¾‘å™¨"><span>10. Nä¸ªå¥½å¥‡å¿ƒ</span></h3><h4 data-tool="mdniceç¼–è¾‘å™¨"><span></span>10.1 Supervised versus Unsupervised modeï¼Ÿ<span></span></h4><p data-tool="mdniceç¼–è¾‘å™¨">å¯ä»¥çœ‹åˆ°å‰é¢å¾ˆå¤šåŠŸèƒ½ï¼Œä½œè€…éƒ½æä¾›ä¸¤ç§æ¨¡å¼ï¼Œç›‘ç£ or éç›‘ç£æ¨¡å¼ï¼Œé‚£åœ¨å®é™…å·¥ä½œä¸­é‡‡ç”¨å“ªä¸ªæ¯”è¾ƒå¤šå‘¢ï¼Ÿ</p><p data-tool="mdniceç¼–è¾‘å™¨">åœ¨ç¬¬ä¸€ç‰ˆELMER v. 1æ—¶ï¼Œä½œè€…æ¨èå½“ä¸¤ç»„æ ·æœ¬å‡ä¸ºåŒè´¨æ€§(å³å¤„ç†è¿‡çš„vsæœªå¤„ç†çš„ï¼ŒAåˆ†å­äºšå‹vs Båˆ†å­äºšå‹)æ—¶ï¼Œåº”é‡‡ç”¨ç›‘ç£æ¨¡å¼;å½“è‡³å°‘æœ‰ä¸€ç»„æ ·æœ¬ä¸ºå¼‚è´¨æ€§(å³è‚¿ç˜¤æ ·æœ¬)æ—¶ï¼Œåº”é‡‡ç”¨æ— ç›‘ç£æ¨¡å¼</p><p data-tool="mdniceç¼–è¾‘å™¨">è€Œæœ€æ–°ç‰ˆï¼Œä½œè€…ä½¿ç”¨ELMER v. 2.4.3åˆ†æTCGA-BRCAæ•°æ®ï¼Œæ¢ç´¢å¢å¼ºå­-åŸºå› å¯¹ï¼ˆenhancer-geneï¼‰æ—¶ï¼Œå‘ç°ç›‘ç£æ¨¡å¼æœ‰æ›´å¼ºçš„ç»Ÿè®¡æ•ˆèƒ½ï¼Œå¹¶ä¸”èƒ½è¯†åˆ«æ›´å¤šçš„å¢å¼ºå­-åŸºå› å¯¹</p><p data-tool="mdniceç¼–è¾‘å™¨">This comparison suggests that while Unsupervised mode can serve as a useful exploratory tool when sample subtype is unknown a priori, Supervised mode offers greater statistical power when sample subtype is pre-defined.</p><figure data-tool="mdniceç¼–è¾‘å™¨"><img data-ratio="0.6625824693685203" data-src="https://mmbiz.qpic.cn/mmbiz_png/PsnauFhHibc5IgrnANu8GpqP6TXajaybAZpwfCYknCTkP6CZOrNHjwKHmenFv6AGeyKHkecp0yV0E7l6TQHqicLQ/640?wx_fmt=png" data-type="png" data-w="1061" src="https://mmbiz.qpic.cn/mmbiz_png/PsnauFhHibc5IgrnANu8GpqP6TXajaybAZpwfCYknCTkP6CZOrNHjwKHmenFv6AGeyKHkecp0yV0E7l6TQHqicLQ/640?wx_fmt=png"></figure><p data-tool="mdniceç¼–è¾‘å™¨">è€Œå¯¹äºäº§ç”Ÿçš„å‡é˜³æ€§ç‡æƒ…å†µï¼Œä½œè€…å»ºè®®ä¸¤ç§æ–¹æ³•éƒ½å¯ä»¥å°è¯•ä¸‹</p><p data-tool="mdniceç¼–è¾‘å™¨">For heterogeneous patient samples composed of multiple subtypes, it thus appears that Unsupervised and Supervised analyses can offer complementary merits, with Unsupervised analysis displaying a higher false negative rate, but a lower false positive rate. It is recommended to run both Supervised and Unsupervised analyses, as we demonstrated here, to gain maximum insight.</p><h4 data-tool="mdniceç¼–è¾‘å™¨"><span></span>10.2 èƒ½åˆ†æéè‚¿ç˜¤æ•°æ®å—ï¼Ÿ<span></span></h4><p data-tool="mdniceç¼–è¾‘å™¨">è™½ç„¶æ–‡ç« ä¸­æµ‹è¯„ç”¨çš„éƒ½æ˜¯è‚¿ç˜¤æ•°æ®ï¼Œå¹¶ä¸”èƒ½æ­é…TCGABiolinksä¸€èµ·ä½¿ç”¨ã€‚æˆ‘è§‰å¾—è¿˜æ˜¯å¯ä»¥ç”¨çš„ï¼Œå¾ˆå¤šæ•°æ®æŒ–æ˜ç±»å‹RåŒ…éƒ½å¯ä»¥é€šç”¨ï¼Œå¯èƒ½ä¼šæœ‰å±€é™ï¼Œå¯ä»¥æœç´¢ä¸‹æœ‰æ²¡æœ‰ç±»ä¼¼æ–‡ç« </p><p data-tool="mdniceç¼–è¾‘å™¨">we previously developed an R/Bioconductor package ELMER (Enhancer Linking by Methylation/Expression Relationships) to infer regulatory element landscapes and GRNs from cancer methylomes</p></section><p><mp-style-type data-value="3"></mp-style-type></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/COMXQ1vlW_CjK5mCWSgJ3A",target="_blank" rel="noopener noreferrer">åŸæ–‡é“¾æ¥</a>
