---
title: "勘误| infercnv亚群注释"
date: 2025-05-18T05:50:09Z
draft: ["false"]
tags: [
  "fetched",
  "Biomamba 生信基地"
]
categories: ["Acdemic"]
---
勘误| infercnv亚群注释 by Biomamba 生信基地
------
<div><section data-support="96编辑器" data-style-id="40226" data-mpa-powered-by="yiban.io" data-pm-slice="0 0 []"><section><section><p><strong><span leaf="">一、发现问题</span></strong></p></section><section data-width="100%"><section><span leaf=""><span textstyle="">有同学提问我们的</span><a href="https://mp.weixin.qq.com/s?__biz=MzAwMzIzOTk5OQ==&amp;mid=2247522733&amp;idx=1&amp;sn=4e3a0707fd8f83b5ad9377bc91ee8e53&amp;scene=21#wechat_redirect" textvalue="单细胞拷贝数变异学习手册" data-itemshowtype="0" target="_blank" linktype="text" data-linktype="2"><span textstyle="">单细胞拷贝数变异学习手册</span></a><span textstyle="">中根据infercnv score的聚类分群部分描述有问题，这是由于给大家编译教程的环境和我们之前测试环境不一致造成的。图片显示cluster1、3、5、6亚群的c明显高于2、4两个亚群，cluster7的界限就比较模糊了。</span></span></section><section nodeleaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/ImlBFVOwwpw0NDcqDcKsnLh9uoDUT3TwUHtyDG8kmTBjgA0aubSiaQ1krScvYpZEDchDr5nnKKWtQb32vlpeWAg/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.45185185185185184" data-s="300,640" data-type="png" data-w="1080" type="block" data-imgfileid="100045193" src="https://mmbiz.qpic.cn/mmbiz_png/ImlBFVOwwpw0NDcqDcKsnLh9uoDUT3TwUHtyDG8kmTBjgA0aubSiaQ1krScvYpZEDchDr5nnKKWtQb32vlpeWAg/640?wx_fmt=png&amp;from=appmsg"></section><section><span leaf=""><br></span></section><section><span leaf="">不过在先前注释的细胞比例中，cluster7与Malignant cells的交叉比例还是非常高的，考虑到infercnv score，那我们就姑且叫它"</span><span leaf="">Malignant-transition cells"。</span></section><section nodeleaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/ImlBFVOwwpw0NDcqDcKsnLh9uoDUT3TwhEKABCEfzIptMq6u58JopL6yo0iaJR9fAggaTib5zxBHDiaDV5RsLgwuQ/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.6921119592875318" data-s="300,640" data-type="png" data-w="1179" type="block" data-imgfileid="100045198" src="https://mmbiz.qpic.cn/mmbiz_png/ImlBFVOwwpw0NDcqDcKsnLh9uoDUT3TwhEKABCEfzIptMq6u58JopL6yo0iaJR9fAggaTib5zxBHDiaDV5RsLgwuQ/640?wx_fmt=png&amp;from=appmsg"></section></section></section></section><p><span leaf=""><br></span></p><section data-support="96编辑器" data-style-id="40226" data-mpa-powered-by="yiban.io" data-pm-slice="0 0 []"><section><section><p><strong><span leaf="">二、解决问题</span></strong></p></section><section data-width="100%"><span leaf="" data-pm-slice='1 1 ["para",{"tagName":"section","attributes":{},"namespaceURI":"http://www.w3.org/1999/xhtml"}]'>这部分修正后的流程如下，完整拷贝数变异分析教程与测试文件可见：<a href="https://mp.weixin.qq.com/s?__biz=MzAwMzIzOTk5OQ==&amp;mid=2247522733&amp;idx=1&amp;sn=4e3a0707fd8f83b5ad9377bc91ee8e53&amp;scene=21#wechat_redirect" textvalue="单细胞拷贝数变异学习手册" data-itemshowtype="0" target="_blank" linktype="text" data-linktype="2"><span textstyle="">单细胞拷贝数变异学习手册</span></a></span><h3 data-pm-slice="0 0 []"><span leaf=""><span textstyle="">5.2 infercnv流程</span></span></h3><p><span leaf="">这里我们再走一遍infercnv的基础流程</span></p><section><ul><li><li><li></ul><pre data-lang="apache"><code><span leaf=""><span># 加载包：</span></span></code><code><span leaf=""><span>library</span>(infercnv)</span></code><code><span leaf=""><span>library</span>(tidyverse)</span></code></pre></section><pre data-pm-slice="0 0 []"><code><span leaf="">## ── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──</span><span leaf=""><br></span><span leaf="">## ✔ dplyr     1.1.4     ✔ readr     2.1.5</span><span leaf=""><br></span><span leaf="">## ✔ forcats   1.0.0     ✔ stringr   1.5.1</span><span leaf=""><br></span><span leaf="">## ✔ ggplot2   3.5.1     ✔ tibble    3.2.1</span><span leaf=""><br></span><span leaf="">## ✔ lubridate 1.9.4     ✔ tidyr     1.3.1</span><span leaf=""><br></span><span leaf="">## ✔ purrr     1.0.2     </span><span leaf=""><br></span><span leaf="">## ── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──</span><span leaf=""><br></span><span leaf="">## ✖ dplyr::filter() masks stats::filter()</span><span leaf=""><br></span><span leaf="">## ✖ dplyr::lag()    masks stats::lag()</span><span leaf=""><br></span><span leaf="">## ℹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors</span></code></pre><section><ul><li><li></ul><pre data-lang=""><code><span leaf="">library(ComplexHeatmap)</span></code><code><span leaf="">library(circlize)</span></code></pre></section><section><ul><li><li><li><li></ul><pre data-lang="cs"><code><span leaf="">library(RColorBrewer)</span></code><code><span leaf="">library(infercnv)</span></code><code><span leaf=""><span>### 整理数据 ###</span></span></code><code><span leaf="">exprMatrix &lt;- <span>as</span>.matrix(GetAssayData(Epi, slot=<span>'counts'</span>))<span># 获取count矩阵</span></span></code></pre></section><pre data-pm-slice="0 0 []"><code><span leaf="">## Warning: The `slot` argument of `GetAssayData()` is deprecated as of SeuratObject 5.0.0.</span><span leaf=""><br></span><span leaf="">## ℹ Please use the `layer` argument instead.</span><span leaf=""><br></span><span leaf="">## This warning is displayed once every 8 hours.</span><span leaf=""><br></span><span leaf="">## Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span><span leaf=""><br></span><span leaf="">## generated.</span></code></pre><section><ul><li><li></ul><pre data-lang="nginx"><code><span leaf=""><span>cellAnnota</span> &lt;- subset(Epi<span>@meta</span>.data,<span>select</span>=<span>'Cell_subtype'</span>)<span># 取出分组注释信息</span></span></code><code><span leaf="">dir.create(<span>'Epi_infercnv/'</span>)<span># 创建文件夹</span></span></code></pre></section><pre data-pm-slice="0 0 []"><code><span leaf="">## Warning in dir.create("Epi_infercnv/"): 'Epi_infercnv' already exists</span></code></pre><section><ul><li><li><li></ul><pre data-lang="php"><code><span leaf="">write.<span>table</span>(cellAnnota,file =<span>"Epi_infercnv/groupFiles.txt"</span>,sep = <span>'\t'</span>,col.names = F)<span># 把分组注释信息写到本地</span></span></code><code><span leaf=""><span>### 制作基因与染色体位置的文件 ###</span></span></code><code><span leaf=""><span>if</span>(!<span>require</span>(AnnoProbe))install.<span>packages</span>(<span>'AnnoProbe'</span>)<span># 加载注释基因信息的包</span></span></code></pre></section><pre data-pm-slice="0 0 []"><code><span leaf="">## Loading required package: AnnoProbe</span><span leaf=""><br></span><span leaf="">## AnnoProbe v 0.1.7  welcome to use AnnoProbe!</span><span leaf=""><br></span><span leaf="">## If you use AnnoProbe in published research, please acknowledgements:</span><span leaf=""><br></span><span leaf="">## We thank Dr.Jianming Zeng(University of Macau), and all the members of his bioinformatics team, biotrainee, for generously sharing their experience and codes.</span></code></pre><section><ul><li></ul><pre data-lang="ini"><code><span leaf=""><span>geneInfor</span>=annoGene(rownames(Epi),<span>"SYMBOL"</span>,<span>'human'</span>)<span>#  获取基因注释信息，这里物种是人，如果是小鼠，把human换成mouse</span></span></code></pre></section><pre data-pm-slice="0 0 []"><code><span leaf="">## Warning in annoGene(rownames(Epi), "SYMBOL", "human"): 31.29% of input IDs are</span><span leaf=""><br></span><span leaf="">## fail to annotate...</span></code></pre><section><ul><li></ul><pre data-lang="bash"><code><span leaf="">colnames(geneInfor)<span># 包含以下信息</span></span></code></pre></section><pre data-pm-slice="0 0 []"><code><span leaf="">## [1] "SYMBOL"   "biotypes" "ENSEMBL"  "chr"      "start"    "end"</span></code></pre><section><ul><li><li><li></ul><pre data-lang="apache"><code><span leaf=""><span>geneInfor</span>=geneInfor[with(geneInfor, order(chr, start)),c(<span>1</span>,<span>4</span>:<span>6</span>)]# 选取基因的</span></code><code><span leaf=""><span>geneInfor</span>=geneInfor[!duplicated(geneInfor[,<span>1</span>]),]# 去重</span></code><code><span leaf=""><span>length</span>(unique(geneInfor[,<span>1</span>]))# 查看一共有多少基因</span></code></pre></section><pre data-pm-slice="0 0 []"><code><span leaf="">## [1] 20363</span></code></pre><section><ul><li></ul><pre data-lang="bash"><code><span leaf=""><span>head</span>(geneInfor)<span># 查看最终的注释信息</span></span></code></pre></section><pre data-pm-slice="0 0 []"><code><span leaf="">##        SYMBOL  chr  start    end</span><span leaf=""><br></span><span leaf="">## 23 AP006222.2 chr1 266855 268655</span><span leaf=""><br></span><span leaf="">## 51     FAM87B chr1 817371 819837</span><span leaf=""><br></span><span leaf="">## 53  LINC00115 chr1 826206 827522</span><span leaf=""><br></span><span leaf="">## 55     FAM41C chr1 868071 876903</span><span leaf=""><br></span><span leaf="">## 62     SAMD11 chr1 923928 944581</span><span leaf=""><br></span><span leaf="">## 63      NOC2L chr1 944203 959309</span></code></pre><section><ul><li><li><li><li><li><li><li></ul><pre data-lang="bash"><code><span leaf="">write.table(geneInfor, <span>'Epi_infercnv/geneLocate.txt'</span>, row.names=F, col.names=F, sep=<span>'\t'</span>)<span># 注释信息需要写到本地供infercnv读取</span></span></code><code><span leaf=""><span># 创建infercnv对象：</span></span></code><code><span leaf="">infercnv_obj = CreateInfercnvObject(raw_counts_matrix=exprMatrix,</span></code><code><span leaf="">                                    annotations_file=<span>"Epi_infercnv/groupFiles.txt"</span>,</span></code><code><span leaf="">                                    delim=<span>"\t"</span>,</span></code><code><span leaf="">                                    gene_order_file= <span>"Epi_infercnv/geneLocate.txt"</span>,</span></code><code><span leaf="">                                    ref_group_names=<span>"AT1"</span>)</span></code></pre></section><pre data-pm-slice="0 0 []"><code><span leaf="">## INFO [2025-05-16 04:45:04] Parsing gene order file: Epi_infercnv/geneLocate.txt</span><span leaf=""><br></span><span leaf="">## INFO [2025-05-16 04:45:04] Parsing cell annotations file: Epi_infercnv/groupFiles.txt</span><span leaf=""><br></span><span leaf="">## INFO [2025-05-16 04:45:04] ::order_reduce:Start.</span><span leaf=""><br></span><span leaf="">## INFO [2025-05-16 04:45:04] .order_reduce(): expr and order match.</span><span leaf=""><br></span><span leaf="">## INFO [2025-05-16 04:45:04] ::process_data:order_reduce:Reduction from positional data, new dimensions (r,c) = 29634,1500 Total=13628902 Min=0 Max=5783.</span><span leaf=""><br></span><span leaf="">## INFO [2025-05-16 04:45:05] num genes removed taking into account provided gene ordering list: 10109 = 34.1128433556051% removed.</span><span leaf=""><br></span><span leaf="">## INFO [2025-05-16 04:45:05] -filtering out cells &lt; 100 or &gt; Inf, removing 0 % of cells</span><span leaf=""><br></span><span leaf="">## WARN [2025-05-16 04:45:05] Please use "options(scipen = 100)" before running infercnv if you are using the analysis_mode="subclusters" option or you may encounter an error while the hclust is being generated.</span><span leaf=""><br></span><span leaf="">## INFO [2025-05-16 04:45:06] validating infercnv_obj</span></code></pre><section><ul><li><li><li><li><li><li><li><li></ul><pre data-lang="nginx"><code><span leaf=""><span># 运行infercnv：</span></span></code><code><span leaf=""><span>infercnv_obj</span> = infercnv::run(infercnv_obj,</span></code><code><span leaf="">                              cutoff=<span>0</span>.<span>1</span>, <span># use 1 for smart-seq, 0.1 for 10x-genomics </span></span></code><code><span leaf="">                              out_dir=  <span>'Epi_infercnv/output'</span> ,  <span># dir is auto-created for storing outputs</span></span></code><code><span leaf="">                              hclust_method=<span>"ward.D2"</span>, plot_steps=F,denoise=TRUE,</span></code><code><span leaf="">                              HMM=F,  <span>##特别耗时间</span></span></code><code><span leaf="">                              num_threads=<span>10</span></span></code><code><span leaf="">)</span></code></pre></section><pre data-pm-slice="0 0 []"><code><span leaf="">## INFO [2025-05-16 04:45:06] ::process_data:Start</span><span leaf=""><br></span><span leaf="">## INFO [2025-05-16 04:45:06] Checking for saved results.</span><span leaf=""><br></span><span leaf="">## INFO [2025-05-16 04:45:06] Trying to reload from step 22</span><span leaf=""><br></span><span leaf="">## INFO [2025-05-16 04:45:07] Using backup from step 22</span><span leaf=""><br></span><span leaf="">## INFO [2025-05-16 04:45:07] Trying to reload from step 15</span><span leaf=""><br></span><span leaf="">## INFO [2025-05-16 04:45:07] </span><span leaf=""><br></span><span leaf="">## </span><span leaf=""><br></span><span leaf="">##  STEP 1: incoming data</span><span leaf=""><br></span><span leaf="">## </span><span leaf=""><br></span><span leaf="">## INFO [2025-05-16 04:45:07] </span><span leaf=""><br></span><span leaf="">## </span><span leaf=""><br></span><span leaf="">##  STEP 02: Removing lowly expressed genes</span><span leaf=""><br></span><span leaf="">## </span><span leaf=""><br></span><span leaf="">## INFO [2025-05-16 04:45:07] </span><span leaf=""><br></span><span leaf="">## </span><span leaf=""><br></span><span leaf="">##  STEP 03: normalization by sequencing depth</span><span leaf=""><br></span><span leaf="">## </span><span leaf=""><br></span><span leaf="">## INFO [2025-05-16 04:45:07] </span><span leaf=""><br></span><span leaf="">## </span><span leaf=""><br></span><span leaf="">##  STEP 04: log transformation of data</span><span leaf=""><br></span><span leaf="">## </span><span leaf=""><br></span><span leaf="">## INFO [2025-05-16 04:45:07] </span><span leaf=""><br></span><span leaf="">## </span><span leaf=""><br></span><span leaf="">##  STEP 08: removing average of reference data (before smoothing)</span><span leaf=""><br></span><span leaf="">## </span><span leaf=""><br></span><span leaf="">## INFO [2025-05-16 04:45:07] </span><span leaf=""><br></span><span leaf="">## </span><span leaf=""><br></span><span leaf="">##  STEP 09: apply max centered expression threshold: 3</span><span leaf=""><br></span><span leaf="">## </span><span leaf=""><br></span><span leaf="">## INFO [2025-05-16 04:45:07] </span><span leaf=""><br></span><span leaf="">## </span><span leaf=""><br></span><span leaf="">##  STEP 10: Smoothing data per cell by chromosome</span><span leaf=""><br></span><span leaf="">## </span><span leaf=""><br></span><span leaf="">## INFO [2025-05-16 04:45:07] </span><span leaf=""><br></span><span leaf="">## </span><span leaf=""><br></span><span leaf="">##  STEP 11: re-centering data across chromosome after smoothing</span><span leaf=""><br></span><span leaf="">## </span><span leaf=""><br></span><span leaf="">## INFO [2025-05-16 04:45:07] </span><span leaf=""><br></span><span leaf="">## </span><span leaf=""><br></span><span leaf="">##  STEP 12: removing average of reference data (after smoothing)</span><span leaf=""><br></span><span leaf="">## </span><span leaf=""><br></span><span leaf="">## INFO [2025-05-16 04:45:07] </span><span leaf=""><br></span><span leaf="">## </span><span leaf=""><br></span><span leaf="">##  STEP 14: invert log2(FC) to FC</span><span leaf=""><br></span><span leaf="">## </span><span leaf=""><br></span><span leaf="">## INFO [2025-05-16 04:45:07] </span><span leaf=""><br></span><span leaf="">## </span><span leaf=""><br></span><span leaf="">##  STEP 15: computing tumor subclusters via leiden</span><span leaf=""><br></span><span leaf="">## </span><span leaf=""><br></span><span leaf="">## INFO [2025-05-16 04:45:07] </span><span leaf=""><br></span><span leaf="">## </span><span leaf=""><br></span><span leaf="">##  STEP 22: Denoising</span><span leaf=""><br></span><span leaf="">## </span><span leaf=""><br></span><span leaf="">## INFO [2025-05-16 04:45:10] </span><span leaf=""><br></span><span leaf="">## </span><span leaf=""><br></span><span leaf="">## ## Making the final infercnv heatmap ##</span><span leaf=""><br></span><span leaf="">## INFO [2025-05-16 04:45:11] ::plot_cnv:Start</span><span leaf=""><br></span><span leaf="">## INFO [2025-05-16 04:45:11] ::plot_cnv:Current data dimensions (r,c)=7292,1500 Total=10988883.0171584 Min=0.762388537428658 Max=1.57869976486215.</span><span leaf=""><br></span><span leaf="">## INFO [2025-05-16 04:45:11] ::plot_cnv:Depending on the size of the matrix this may take a moment.</span><span leaf=""><br></span><span leaf="">## INFO [2025-05-16 04:45:12] plot_cnv(): auto thresholding at: (0.829906 , 1.170094)</span><span leaf=""><br></span><span leaf="">## INFO [2025-05-16 04:45:12] plot_cnv_observation:Start</span><span leaf=""><br></span><span leaf="">## INFO [2025-05-16 04:45:12] Observation data size: Cells= 1000 Genes= 7292</span><span leaf=""><br></span><span leaf="">## INFO [2025-05-16 04:45:12] plot_cnv_observation:Writing observation groupings/color.</span><span leaf=""><br></span><span leaf="">## INFO [2025-05-16 04:45:12] plot_cnv_observation:Done writing observation groupings/color.</span><span leaf=""><br></span><span leaf="">## INFO [2025-05-16 04:45:12] plot_cnv_observation:Writing observation heatmap thresholds.</span><span leaf=""><br></span><span leaf="">## INFO [2025-05-16 04:45:12] plot_cnv_observation:Done writing observation heatmap thresholds.</span></code></pre><pre><code><span leaf="">## INFO [2025-05-16 04:45:13] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000</span><span leaf=""><br></span><span leaf="">## INFO [2025-05-16 04:45:13] Quantiles of plotted data range: 0.829906399244205,1.00117985183862,1.00117985183862,1.00117985183862,1.1700936007558</span></code></pre><pre><code><span leaf="">## INFO [2025-05-16 04:45:14] plot_cnv_references:Start</span><span leaf=""><br></span><span leaf="">## INFO [2025-05-16 04:45:14] Reference data size: Cells= 500 Genes= 7292</span><span leaf=""><br></span><span leaf="">## INFO [2025-05-16 04:45:16] plot_cnv_references:Number reference groups= 1</span><span leaf=""><br></span><span leaf="">## INFO [2025-05-16 04:45:16] plot_cnv_references:Plotting heatmap.</span></code></pre><pre><code><span leaf="">## INFO [2025-05-16 04:45:16] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000</span><span leaf=""><br></span><span leaf="">## INFO [2025-05-16 04:45:16] Quantiles of plotted data range: 0.829906399244205,1.00117985183862,1.00117985183862,1.00117985183862,1.1700936007558</span></code></pre><p><span leaf="">看一下”那一张热图“：</span><span leaf=""><br></span><span leaf=""><img data-imgfileid="100045200" data-ratio="0.9444444444444444" data-src="https://mmbiz.qpic.cn/mmbiz_png/ImlBFVOwwpw0NDcqDcKsnLh9uoDUT3TwvOoZEP7rWZl9ZoxbmUuic6Qto1Y4WrFmelib5dglMibnSgY81AfbAhFiaQ/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/ImlBFVOwwpw0NDcqDcKsnLh9uoDUT3TwvOoZEP7rWZl9ZoxbmUuic6Qto1Y4WrFmelib5dglMibnSgY81AfbAhFiaQ/640?wx_fmt=png&amp;from=appmsg"></span><span leaf=""><br></span><span leaf="">可以看到作为参考的</span><code><span leaf="">AT1</span></code><span leaf="">(上方热图)发生的拷贝数变异事件比较少。而</span><code><span leaf="">Malignant cells</span></code><span leaf="">出现了大量且密集的拷贝数变异。</span><code><span leaf="">tS1</span></code><span leaf="">似乎也存在着少量的拷贝数变异，所以单凭热图还不行，还得继续分析。</span></p><h3 data-pm-slice="0 0 []"><span leaf=""><span textstyle="">5.4 小提琴图</span></span></h3><h4><span leaf="">(1)从细胞类群的角度看</span></h4><section><ul><li><li></ul><pre data-lang="bash"><code><span leaf=""><span>### 转化cnvscore数据 ###</span></span></code><code><span leaf="">fivenum(<span>expr</span>)<span># 原始的cov_score是以1为中心分布的数据</span></span></code></pre></section><pre data-pm-slice="0 0 []"><code><span leaf="">## [1] 0.7623885 1.0011799 1.0011799 1.0011799 1.5786998</span></code></pre><section><ul><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li></ul><pre data-lang="apache"><code><span leaf=""><span>CNV_score</span>=expr-<span>1</span># 转换为以<span>0</span>为中心分布的数据</span></code><code><span leaf=""><span>CNV_score</span>=CNV_score ^ <span>2</span># 平方，防止累加的过程中cnv lose和cnv gain相互抵消，造成没有cnv事件发生的假象</span></code><code><span leaf=""><span>CNV_score</span>=as.data.frame(colMeans(CNV_score))# 以细胞为单位，获得cnv score</span></code><code><span leaf=""><span>colnames</span>(CNV_score)=<span>"CNV_score"</span># 更改列名</span></code><code><span leaf=""><span>CNV_score</span>$CB=rownames(CNV_score)# 加入cell barcode列</span></code><code><span leaf=""><span>kmeans_df_s</span>$CB=rownames(kmeans_df_s)# 把kmeans_df_s中的cell barcode也加上</span></code><code><span leaf=""><span>CNV_score</span>=CNV_score%&gt;%inner_join(kmeans_df_s,by=<span>"CB"</span>)# 合并cnvscore和聚类结果</span></code><code><span leaf=""><span>library</span>(ggplot2)</span></code><code><span leaf=""><span>library</span>(ggsignif)</span></code><code><span leaf=""> <span>ggplot</span>(CNV_score,aes(Sub_type,CNV_score,fill=Sub_type))+</span></code><code><span leaf="">      <span>geom_violin</span>(alpha=<span>0</span>.<span>4</span>)+ #小提琴图需要一定的透明度</span></code><code><span leaf="">      <span>stat_boxplot</span>(geom=<span>"errorbar"</span>,position = position_dodge(width = <span>0</span>.<span>1</span>),width=<span>0</span>.<span>1</span>)+ # 加上误差棒</span></code><code><span leaf="">      <span>geom_boxplot</span>(alpha=<span>0</span>.<span>5</span>,outlier.size=<span>0</span>, size=<span>0</span>.<span>3</span>, width=<span>0</span>.<span>3</span>)+ # 加上箱线图</span></code><code><span leaf="">      <span>scale_fill_manual</span>(values = ggsci::pal_aaas()(<span>3</span>))+ # 填充颜色</span></code><code><span leaf="">      <span>theme_bw</span>()+# 主题为空白</span></code><code><span leaf="">      <span>geom_signif</span>(comparisons = list(c(<span>"AT1"</span>,<span>"tS1"</span>),</span></code><code><span leaf="">                                     <span>c</span>(<span>"tS1"</span>, <span>"Malignant cells"</span>),</span></code><code><span leaf="">                                     <span>c</span>(<span>"AT1"</span> ,<span>"Malignant cells"</span>)),</span></code><code><span leaf="">                  <span>step_increase</span> = <span>0</span>.<span>1</span>,map_signif_level = F)</span></code></pre></section><section><span><span leaf=""><img data-imgfileid="100045201" data-ratio="0.7138888888888889" data-src="https://mmbiz.qpic.cn/mmbiz_png/ImlBFVOwwpw0NDcqDcKsnLh9uoDUT3TwLSEXciahub2cmjRxtltPqDH1YtAZXIFiarkgV82bI4iaDA3pBBrEmkiapA/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/ImlBFVOwwpw0NDcqDcKsnLh9uoDUT3TwLSEXciahub2cmjRxtltPqDH1YtAZXIFiarkgV82bI4iaDA3pBBrEmkiapA/640?wx_fmt=png&amp;from=appmsg"></span></span></section><p data-pm-slice="0 0 []"><span leaf="">作为正常细胞的</span><code><span leaf="">AT1</span></code><span leaf="">与</span><code><span leaf="">tS1</span></code><span leaf="">与</span><code><span leaf="">Malignant</span></code><span leaf=""> cells之间的统计结果呈现出极显著的关系，而AT1与tS1之间则没有显著性差异(0.086)</span></p><h4><span leaf="">(2) 从聚类的角度来看</span></h4><section><ul><li><li><li><li><li><li></ul><pre data-lang="apache"><code><span leaf=""><span>ggplot</span>(CNV_score,aes(kmeans_class,CNV_score,fill=kmeans_class))+</span></code><code><span leaf="">    <span>geom_violin</span>(alpha=<span>0</span>.<span>4</span>)+ #小提琴图需要一定的透明度</span></code><code><span leaf="">    <span>stat_boxplot</span>(geom=<span>"errorbar"</span>,position = position_dodge(width = <span>0</span>.<span>1</span>),width=<span>0</span>.<span>1</span>)+ # 加上误差棒</span></code><code><span leaf="">    <span>geom_boxplot</span>(alpha=<span>0</span>.<span>5</span>,outlier.size=<span>0</span>, size=<span>0</span>.<span>3</span>, width=<span>0</span>.<span>3</span>)+ # 加上箱线图</span></code><code><span leaf="">    <span>scale_fill_manual</span>(values = ggsci::pal_nejm()(<span>8</span>))+ # 填充颜色</span></code><code><span leaf="">    <span>theme_bw</span>()# 主题为空白</span></code></pre></section><section><span leaf=""><img data-imgfileid="100045202" data-ratio="0.7138888888888889" data-src="https://mmbiz.qpic.cn/mmbiz_png/ImlBFVOwwpw0NDcqDcKsnLh9uoDUT3TwOJH4VtycyibQWC3MBBAjK90ibQiaMPQqM8wjPI3WRkYY0UMRJTXXQggyQ/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/ImlBFVOwwpw0NDcqDcKsnLh9uoDUT3TwOJH4VtycyibQWC3MBBAjK90ibQiaMPQqM8wjPI3WRkYY0UMRJTXXQggyQ/640?wx_fmt=png&amp;from=appmsg"></span></section><section><span data-pm-slice="0 0 []"><span leaf="">实际上cluster1、3、5、6亚群的c明显高于2、4两个亚群，cluster7的界限就比较模糊了。我们可以通过cluster在Sub_type中的亚群来辅助判断它们的身份：</span></span></section><section><ul><li><li><li><li><li><li><li><li><li><li></ul><pre data-lang="apache"><code><span leaf=""><span>cell</span>.prop&lt;-as.data.frame(prop.table(table(CNV_score$kmeans_class, CNV_score$Sub_type)))</span></code><code><span leaf=""><span>colnames</span>(cell.prop)&lt;-c(<span>"kmeans_class"</span>,<span>"Sub_type"</span>,<span>"proportion"</span>)</span></code><code><span leaf=""><span>ggplot</span>(cell.prop,aes(kmeans_class,proportion,fill=Sub_type))+</span></code><code><span leaf=""><span>geom_bar</span>(stat=<span>"identity"</span>,position=<span>"fill"</span>)+</span></code><code><span leaf=""><span>ggtitle</span>(<span>""</span>)+</span></code><code><span leaf=""><span>theme_bw</span>()+</span></code><code><span leaf=""><span>theme</span>(axis.ticks.length=unit(<span>0</span>.<span>5</span>,'cm'))+</span></code><code><span leaf=""><span>guides</span>(fill=guide_legend(title=NULL))+</span></code><code><span leaf="">  <span>theme</span>( axis.text.x.bottom = element_text(angle = <span>45</span>,hjust = <span>1</span>,vjust = <span>1</span>))+</span></code><code><span leaf="">  <span>scale_fill_manual</span>(values = ggsci::pal_nejm()(<span>8</span>)) # 填充颜色</span></code></pre></section><section><span leaf=""><img data-imgfileid="100045203" data-ratio="0.7138888888888889" data-src="https://mmbiz.qpic.cn/mmbiz_png/ImlBFVOwwpw0NDcqDcKsnLh9uoDUT3TwxOTuKA7pTtHdMg4ktI7p07q3IkaA5usC8hMx0DmYAJJ20ovTnBnHbA/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/ImlBFVOwwpw0NDcqDcKsnLh9uoDUT3TwxOTuKA7pTtHdMg4ktI7p07q3IkaA5usC8hMx0DmYAJJ20ovTnBnHbA/640?wx_fmt=png&amp;from=appmsg"></span></section><section><span data-pm-slice="0 0 []"><span leaf="">果然不出所料，1、3、5、6中均存在很高比例的Malignant cells，而cluster7与Malignant cell的交叉比例较高，假设这是一个未经注释的数据集，我一定会这么命名：</span></span></section><section><ul><li><li><li><li><li><li><li><li><li></ul><pre data-lang="bash"><code><span leaf="">CNV_score<span>$Type</span> &lt;- recode(CNV_score<span>$kmeans_class</span>,</span></code><code><span leaf="">                                  <span>'1'</span>=<span>'Malignant cells'</span>,</span></code><code><span leaf="">                                  <span>'2'</span>=<span>'Normal Epi'</span>,</span></code><code><span leaf="">                                  <span>'3'</span>=<span>'Malignant cells'</span>,</span></code><code><span leaf="">                                  <span>'4'</span>=<span>'Normal Epi'</span>,</span></code><code><span leaf="">                                  <span>'5'</span>=<span>'Malignant cells'</span>,</span></code><code><span leaf="">                                  <span>'6'</span>=<span>'Malignant cells'</span>,</span></code><code><span leaf="">                                  <span>'7'</span>=<span>'Malignant-transition cells'</span>)</span></code><code><span leaf="">unique(CNV_score<span>$Type</span>)</span></code></pre></section><pre data-pm-slice="0 0 []"><code><span leaf="">## [1] Normal Epi                 Malignant cells           </span><span leaf=""><br></span><span leaf="">## [3] Malignant-transition cells</span><span leaf=""><br></span><span leaf="">## Levels: Malignant cells Normal Epi Malignant-transition cells</span></code></pre><section><ul><li><li><li><li><li><li><li><li><li><li><li></ul><pre data-lang="apache"><code><span leaf=""><span># 看一下重命名过的cnv score怎么分布：</span></span></code><code><span leaf=""><span>ggplot</span>(CNV_score,aes(Type,CNV_score,fill=Type))+</span></code><code><span leaf="">    <span>geom_violin</span>(alpha=<span>0</span>.<span>4</span>)+ #小提琴图需要一定的透明度</span></code><code><span leaf="">    <span>stat_boxplot</span>(geom=<span>"errorbar"</span>,position = position_dodge(width = <span>0</span>.<span>1</span>),width=<span>0</span>.<span>1</span>)+ # 加上误差棒</span></code><code><span leaf="">    <span>geom_boxplot</span>(alpha=<span>0</span>.<span>5</span>,outlier.size=<span>0</span>, size=<span>0</span>.<span>3</span>, width=<span>0</span>.<span>3</span>)+ # 加上箱线图</span></code><code><span leaf="">    <span>scale_fill_manual</span>(values = ggsci::pal_nejm()(<span>8</span>))+ # 填充颜色</span></code><code><span leaf="">    <span>theme_bw</span>()+# 主题为空白</span></code><code><span leaf="">    <span>geom_signif</span>(comparisons = list(c( 'Malignant cells', 'Normal Epi'),</span></code><code><span leaf="">                                   <span>c</span>('Normal Epi', 'Malignant-transition cells'),</span></code><code><span leaf="">                                   <span>c</span>('Malignant cells', 'Malignant-transition cells')),</span></code><code><span leaf="">                <span>step_increase</span> = <span>0</span>.<span>1</span>,map_signif_level = F)</span></code></pre></section><section><span leaf=""><img data-imgfileid="100045204" data-ratio="0.7138888888888889" data-src="https://mmbiz.qpic.cn/mmbiz_png/ImlBFVOwwpw0NDcqDcKsnLh9uoDUT3TwZQBncrCr2eyotJajOShhHNFeqh2kiaGP5YY1PibuunQAZNic7IjHyAVCA/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/ImlBFVOwwpw0NDcqDcKsnLh9uoDUT3TwZQBncrCr2eyotJajOShhHNFeqh2kiaGP5YY1PibuunQAZNic7IjHyAVCA/640?wx_fmt=png&amp;from=appmsg"></span></section><section><span data-pm-slice="0 0 []"><span leaf="">猜的没错，三个类型的cnv score两两间均有显著性差异，我们可以将这个注释导回原来的上皮细胞Seurat对象中：</span></span></section><pre data-pm-slice="0 0 []"><code><span><span><span leaf="">rownames</span></span><span leaf="">(CNV_score) </span><span><span leaf="">&lt;-</span></span><span leaf=""> CNV_score</span><span><span leaf="">$</span></span><span leaf="">CB</span><span><span leaf=""># 设置Cell barcode行名，AddMeta会按照行名匹配</span></span></span><span leaf=""><br></span><span></span><span leaf=""><br></span><span><span leaf="">Epi </span><span><span leaf="">&lt;-</span></span><span><span leaf="">AddMetaData</span></span><span leaf="">(Epi,</span><span><span leaf="">metadata =</span></span><span leaf=""> CNV_score)</span><span><span leaf=""># 添加CNV score注释信息</span></span></span><span leaf=""><br></span><span></span><span leaf=""><br></span><span><span><span leaf="">Idents</span></span><span leaf="">(Epi) </span><span><span leaf="">&lt;-</span></span><span><span leaf="">'Type'</span></span><span><span leaf=""># 改变Seurat对象默认标签</span></span></span><span leaf=""><br></span><span></span><span leaf=""><br></span><span><span><span leaf="">DimPlot</span></span><span leaf="">(Epi,</span><span><span leaf="">cols =</span></span><span leaf=""> ggsci</span><span><span leaf="">::</span></span><span><span leaf="">pal_nejm</span></span><span leaf="">()(</span><span><span leaf="">8</span></span><span leaf="">))</span><span><span leaf=""># 查看不同type的降维情况</span></span></span></code></pre><section><span leaf=""><img data-imgfileid="100045205" data-ratio="0.7138888888888889" data-src="https://mmbiz.qpic.cn/mmbiz_png/ImlBFVOwwpw0NDcqDcKsnLh9uoDUT3Twgq9qmTtnkSN2UyNaQuTaLNr0NXrBJEt6Pn8k4NNSUvhyMtqEmaMs9A/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/ImlBFVOwwpw0NDcqDcKsnLh9uoDUT3Twgq9qmTtnkSN2UyNaQuTaLNr0NXrBJEt6Pn8k4NNSUvhyMtqEmaMs9A/640?wx_fmt=png&amp;from=appmsg"></span></section><section><ul><li><li><li></ul><pre data-lang="bash"><code><span leaf=""><span># 下面的工作就是生物学方面的意义啦，比如做个差异分析：  </span></span></code><code><span leaf="">DEG &lt;- FindMarkers(Epi, ident.1 = <span>'Malignant cells'</span>,ident.2 = <span>'Normal Epi'</span>)</span></code><code><span leaf=""><span>head</span>(DEG)</span></code></pre></section><pre data-pm-slice="0 0 []"><code><span leaf="">##                  p_val avg_log2FC pct.1 pct.2     p_val_adj</span><span leaf=""><br></span><span leaf="">## IGFBP3   2.298313e-120   6.362855 0.526 0.021 6.810821e-116</span><span leaf=""><br></span><span leaf="">## TMEM176B  5.718259e-98   4.859484 0.558 0.060  1.694549e-93</span><span leaf=""><br></span><span leaf="">## HMGB3     2.088671e-95   4.098322 0.761 0.183  6.189567e-91</span><span leaf=""><br></span><span leaf="">## VSTM2L    1.889219e-94   4.029765 0.544 0.053  5.598510e-90</span><span leaf=""><br></span><span leaf="">## TMEM176A  1.495375e-91   4.775832 0.516 0.051  4.431395e-87</span><span leaf=""><br></span><span leaf="">## GAPDH     1.454763e-88   1.793153 0.986 0.830  4.311043e-84</span></code></pre><section><ul><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li><li></ul><pre data-lang="apache"><code><span leaf=""><span>DEG</span>$Gene &lt;- row.names(DEG)</span></code><code><span leaf=""><span># 画个火山图看一下：  </span></span></code><code><span leaf=""> <span>ggplot</span>(DEG,aes(avg_log2FC , -log10(p_val)))+</span></code><code><span leaf="">  <span># 横向水平参考线：</span></span></code><code><span leaf="">  <span>geom_hline</span>(yintercept = -log10(<span>0</span>.<span>05</span>), linetype = <span>"dashed"</span>, color = <span>"#999999"</span>)+</span></code><code><span leaf="">  <span># 纵向垂直参考线：</span></span></code><code><span leaf="">  <span>geom_vline</span>(xintercept = c(-<span>1</span>.<span>2</span>,<span>1</span>.<span>2</span>), linetype = <span>"dashed"</span>, color = <span>"#999999"</span>)+</span></code><code><span leaf="">  <span># 散点图:</span></span></code><code><span leaf="">  <span>geom_point</span>(aes(size=-log10(p_val), color= -log10(p_val)))+</span></code><code><span leaf="">  <span># 指定颜色渐变模式：</span></span></code><code><span leaf="">  <span>scale_color_gradientn</span>(values = seq(<span>0</span>,<span>1</span>,<span>0</span>.<span>2</span>),</span></code><code><span leaf="">                        <span>colors</span> = c(<span>"#39489f"</span>,<span>"#39bbec"</span>,'#FFA500',<span>"#f38466"</span>,<span>"#b81f25"</span>))+</span></code><code><span leaf="">  <span># 指定散点大小渐变模式：</span></span></code><code><span leaf="">  <span>scale_size_continuous</span>(range = c(<span>0</span>.<span>5</span>,<span>4</span>))+</span></code><code><span leaf="">  <span># 主题调整：</span></span></code><code><span leaf="">  <span>theme_bw</span>()+</span></code><code><span leaf="">  <span># 调整主题和图例位置：</span></span></code><code><span leaf="">  <span># theme(panel.grid = element_blank(),</span></span></code><code><span leaf="">  <span>#       legend.position = c(0.01,0.7),</span></span></code><code><span leaf="">  <span>#       legend.justification = c(0,1)</span></span></code><code><span leaf="">  <span># )+</span></span></code><code><span leaf="">  <span># 设置部分图例不显示：</span></span></code><code><span leaf="">  <span>guides</span>(col = guide_colourbar(title = <span>"-Log10_q-value"</span>),</span></code><code><span leaf="">         <span>size</span> = <span>"none"</span>)+</span></code><code><span leaf="">  <span># 添加标签：</span></span></code><code><span leaf="">  <span>geom_text</span>(aes(label=Gene, color = -log10(p_val)), size = <span>3</span>, vjust = <span>1</span>.<span>5</span>, hjust=<span>1</span>)+</span></code><code><span leaf="">  <span># 修改坐标轴：</span></span></code><code><span leaf="">  <span>xlab</span>(<span>"Log2FC"</span>)+</span></code><code><span leaf="">  <span>ylab</span>(<span>"-Log10(FDR q-value)"</span>)</span></code></pre></section><p><span leaf=""><img data-imgfileid="100045206" data-ratio="0.7138888888888889" data-src="https://mmbiz.qpic.cn/mmbiz_png/ImlBFVOwwpw0NDcqDcKsnLh9uoDUT3TwJN3Cu7ibibAznhticqWktbo61CL9GeUCGKzQSibNkjcXKm26EnHVTVyNwg/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/ImlBFVOwwpw0NDcqDcKsnLh9uoDUT3TwJN3Cu7ibibAznhticqWktbo61CL9GeUCGKzQSibNkjcXKm26EnHVTVyNwg/640?wx_fmt=png&amp;from=appmsg"></span></p></section></section></section><p><span leaf=""><br></span></p><section data-support="96编辑器" data-style-id="40226" data-mpa-powered-by="yiban.io" data-pm-slice="0 0 []"><section><section><p><strong><span leaf="">三、欢迎勘误</span></strong></p></section><section data-width="100%"><section><span><span leaf=""><span textstyle="">人要勇于承认自己的错误并改正<img data-src="https://res.wx.qq.com/t/wx_fed/we-emoji/res/assets/Expression/Expression_14@2x.png" data-ratio="1" data-w="20" src="https://res.wx.qq.com/t/wx_fed/we-emoji/res/assets/Expression/Expression_14@2x.png">。这里也允许我找个小借口，我们的很多教程都是数万字代码+注释+数十GB的文件，很难保证一桩桩一件件都没有错误。欢迎大家为我们勘误，不仅能够帮助我们做的更好，更能帮助其它粉丝朋友们少走弯路~</span></span></span><span leaf=""><br></span></section><section><span leaf=""><span textstyle="">Biomamba生信基地还给大家准备了大量的周边，可以作为大家为我们勘误的小礼品任选一件~</span></span></section><section><span leaf=""><br></span></section><section data-pm-slice="0 0 []"><span leaf=""><span textstyle="">Linux速查超大鼠标垫：</span></span></section><section nodeleaf=""><img alt="图片" data-imgfileid="100038919" data-ratio="0.3824074074074074" data-src="https://mmbiz.qpic.cn/mmbiz_jpg/ImlBFVOwwpwxvBkwhVBCGFWFIYwS7u9RO1DA22lRWyU5ONLdgL6VoUBXNTtpDib1VSS9eUOufbhH7LOr3gKtgyg/640?wx_fmt=other&amp;wxfrom=5&amp;wx_lazy=1&amp;wx_co=1&amp;tp=webp" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_jpg/ImlBFVOwwpwxvBkwhVBCGFWFIYwS7u9RO1DA22lRWyU5ONLdgL6VoUBXNTtpDib1VSS9eUOufbhH7LOr3gKtgyg/640?wx_fmt=other&amp;wxfrom=5&amp;wx_lazy=1&amp;wx_co=1&amp;tp=webp"></section><section><span leaf=""><br></span></section><section><span leaf=""><span textstyle="">Biomamba同款</span></span><span leaf=""><span textstyle="">背包与词云展示：</span></span></section><section nodeleaf=""><img alt="图片" data-imgfileid="100038920" data-ratio="0.7701149425287356" data-src="https://mmbiz.qpic.cn/mmbiz_jpg/ImlBFVOwwpzLmznOnvnQ1c5NalBlCLKxRK7BBFFVY92BicN8MMDdadYIxtuA17IJN8jvTtyJxoGoxxHoKEFicbKA/640?wx_fmt=other&amp;wxfrom=5&amp;wx_lazy=1&amp;wx_co=1&amp;tp=webp" data-w="783" src="https://mmbiz.qpic.cn/mmbiz_jpg/ImlBFVOwwpzLmznOnvnQ1c5NalBlCLKxRK7BBFFVY92BicN8MMDdadYIxtuA17IJN8jvTtyJxoGoxxHoKEFicbKA/640?wx_fmt=other&amp;wxfrom=5&amp;wx_lazy=1&amp;wx_co=1&amp;tp=webp"></section><section nodeleaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/ImlBFVOwwpw1iaU2Vvcj9zsGVYWvqIdzpvm0VWOKYTsNLpfwY2Ea0ic3YL4RxhibMgkAYrZLpOnBZCk7Xe9SiawaSw/640?wx_fmt=other&amp;wxfrom=5&amp;wx_lazy=1&amp;wx_co=1&amp;tp=webp" alt="图片" data-ratio="0.7357512953367875" data-w="965" data-imgfileid="100038921" src="https://mmbiz.qpic.cn/mmbiz_png/ImlBFVOwwpw1iaU2Vvcj9zsGVYWvqIdzpvm0VWOKYTsNLpfwY2Ea0ic3YL4RxhibMgkAYrZLpOnBZCk7Xe9SiawaSw/640?wx_fmt=other&amp;wxfrom=5&amp;wx_lazy=1&amp;wx_co=1&amp;tp=webp"></section><section><span leaf=""><br></span></section><p><span leaf=""><span textstyle="">金士顿64GB、3.0存储高速U盘：</span></span></p><section nodeleaf=""><img alt="图片" data-imgfileid="100038922" data-ratio="0.3726027397260274" data-s="300,640" data-src="https://mmbiz.qpic.cn/mmbiz_jpg/ImlBFVOwwpyWaOKGagbYpZrLeBC5pxjbV6G3Ksz4WGkql0hGTxtZZTX8j9q5ZIkaR8w9hic1QrAwkjKVF7oBAHw/640?wx_fmt=other&amp;wxfrom=5&amp;wx_lazy=1&amp;wx_co=1&amp;tp=webp" data-type="jpeg" data-w="365" src="https://mmbiz.qpic.cn/mmbiz_jpg/ImlBFVOwwpyWaOKGagbYpZrLeBC5pxjbV6G3Ksz4WGkql0hGTxtZZTX8j9q5ZIkaR8w9hic1QrAwkjKVF7oBAHw/640?wx_fmt=other&amp;wxfrom=5&amp;wx_lazy=1&amp;wx_co=1&amp;tp=webp"></section><section><span leaf=""><br></span></section><p data-pm-slice="0 0 []"><strong><span leaf=""><span textstyle="">Biomamba生信基地定制咖啡杯吧！Biomamba亲测好用，质感高级</span><img data-src="https://res.wx.qq.com/t/wx_fed/we-emoji/res/assets/newemoji/Yellowdog.png?tp=webp&amp;wxfrom=5&amp;wx_lazy=1" alt="图片" data-ratio="1" data-w="128" src="https://res.wx.qq.com/t/wx_fed/we-emoji/res/assets/newemoji/Yellowdog.png?tp=webp&amp;wxfrom=5&amp;wx_lazy=1"></span></strong></p><section nodeleaf=""><img alt="图片" data-croporisrc="https://mmbiz.qpic.cn/mmbiz_png/ImlBFVOwwpwDwviaZibEEU8S7ffoRDPY7GdymibxCrXiaibIG5HzxAauBNtRG5a8GB6nMG52ricZicV3Y0QVF8xr33Bkg/0?wx_fmt=png&amp;from=appmsg" data-cropx1="77.171875" data-cropx2="2188.875" data-cropy1="245.546875" data-cropy2="1585.53125" data-imgfileid="100044993" data-ratio="0.6351851851851852" data-s="300,640" data-src="https://mmbiz.qpic.cn/mmbiz_jpg/ImlBFVOwwpwDwviaZibEEU8S7ffoRDPY7GPxT1AhGe3OMIbO3BqlVVlaw0AZkXfOAs5U4IMmeCTJiaUgsK4NZjgIA/640?wx_fmt=jpeg&amp;tp=webp&amp;wxfrom=5&amp;wx_lazy=1" data-type="png" data-w="1080" type="block" src="https://mmbiz.qpic.cn/mmbiz_jpg/ImlBFVOwwpwDwviaZibEEU8S7ffoRDPY7GPxT1AhGe3OMIbO3BqlVVlaw0AZkXfOAs5U4IMmeCTJiaUgsK4NZjgIA/640?wx_fmt=jpeg&amp;tp=webp&amp;wxfrom=5&amp;wx_lazy=1"></section><section><span leaf=""><br></span></section></section></section></section><p><span leaf=""><br></span></p><section data-pm-slice="0 0 []"><section><section data-pm-slice="0 0 []"><section><section data-support="96编辑器" data-style-id="36531"><section data-align="title"><section><section nodeleaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/Ljib4So7yuWiaFyWDMZJ2713X3DwiaQicG5O8mjlC4kPOluxibs4oDot7AXe0RgoXgMeiaB3KYViadIQqXRoRicyian7KUg/640?wx_fmt=other&amp;wxfrom=5&amp;wx_lazy=1&amp;wx_co=1&amp;tp=webp" alt="图片" data-ratio="0.5106382978723404" data-w="47" data-width="100%" data-imgfileid="100042173" src="https://mmbiz.qpic.cn/mmbiz_png/Ljib4So7yuWiaFyWDMZJ2713X3DwiaQicG5O8mjlC4kPOluxibs4oDot7AXe0RgoXgMeiaB3KYViadIQqXRoRicyian7KUg/640?wx_fmt=other&amp;wxfrom=5&amp;wx_lazy=1&amp;wx_co=1&amp;tp=webp"></section><section><section data-lazy-bgimg="https://mmbiz.qpic.cn/mmbiz_png/Ljib4So7yuWiaFyWDMZJ2713X3DwiaQicG5Ohviaxjsp3ZhRgjJ0mLNVEwpbT2fzdQMNt832icGGUt1ib9G2DMiazWDahA/640?wx_fmt=png" data-fail="0"><section><p><strong><span leaf="">如何联系我们</span></strong><span leaf=""><br></span></p></section></section><section nodeleaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/Ljib4So7yuWiaFyWDMZJ2713X3DwiaQicG5OSxqKNRib39TuLic5NT2iaJaI0mIicuvjMZlRPvlqPfgjzg001qlRJTBDRA/640?wx_fmt=other&amp;wxfrom=5&amp;wx_lazy=1&amp;wx_co=1&amp;tp=webp" alt="图片" data-ratio="1.163265306122449" data-w="49" data-width="100%" data-imgfileid="100042169" src="https://mmbiz.qpic.cn/mmbiz_png/Ljib4So7yuWiaFyWDMZJ2713X3DwiaQicG5OSxqKNRib39TuLic5NT2iaJaI0mIicuvjMZlRPvlqPfgjzg001qlRJTBDRA/640?wx_fmt=other&amp;wxfrom=5&amp;wx_lazy=1&amp;wx_co=1&amp;tp=webp"></section></section></section></section></section><section data-support="96编辑器" data-style-id="34624"><section><section><section nodeleaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_gif/Ljib4So7yuWiaC0oer2QpXIolHVU3Xibtm9BNWiav660mHaB5qEd1dibMibUicNYmAzBBiadQP2s7HArdUxoWL5cDkTEbQ/640?wx_fmt=gif&amp;wxfrom=5&amp;wx_lazy=1&amp;tp=webp" alt="图片" data-ratio="0.2463768115942029" data-w="138" data-imgfileid="100042172" src="https://mmbiz.qpic.cn/mmbiz_gif/Ljib4So7yuWiaC0oer2QpXIolHVU3Xibtm9BNWiav660mHaB5qEd1dibMibUicNYmAzBBiadQP2s7HArdUxoWL5cDkTEbQ/640?wx_fmt=gif&amp;wxfrom=5&amp;wx_lazy=1&amp;tp=webp"></section><section><section data-pm-slice="0 0 []"><span><span leaf="">留一个领取资料、</span></span><span><span leaf=""><a href="https://mp.weixin.qq.com/s?__biz=MzAwMzIzOTk5OQ==&amp;mid=2247527585&amp;idx=1&amp;sn=fa32112c11b215b33096cab43baf08ea&amp;scene=21#wechat_redirect" textvalue="久违的免费两周学习版上架啦~" data-itemshowtype="0" target="_blank" linktype="text" data-linktype="2"><span textstyle="">免费两周学习版</span></a>服务器的</span></span><span><strong><span leaf="">微信号<span textstyle="">[Biomamba_yunying]</span></span></strong></span><span leaf="">，方便各位随时交流。同时我们也构建了<a href="https://mp.weixin.qq.com/s?__biz=MzAwMzIzOTk5OQ==&amp;mid=2247528057&amp;idx=1&amp;sn=bb348c2a88c977c7cce67ee6731a3591&amp;scene=21#wechat_redirect" textvalue="Biomamba生信基地交流群矩阵" data-itemshowtype="0" target="_blank" linktype="text" data-linktype="2"><span textstyle="">交流群矩阵</span></a>，欢迎大家入群讨论。</span></section><section data-pm-slice="0 0 []"><span leaf=""><br></span></section><section><span><span leaf="">大家可以阅读完这几篇之后添加</span></span></section><section><span><span leaf=""><a href="http://mp.weixin.qq.com/s?__biz=MzAwMzIzOTk5OQ==&amp;mid=2247493329&amp;idx=1&amp;sn=597870b7f7ce10346da3561e7c281627&amp;chksm=9b3c9f81ac4b169794088ce9abbcacb5f18e9fd174fc0d7e28da62d78facadf3100ef38aaf22&amp;scene=21#wechat_redirect" textvalue="给生信入门初学者的小贴士" data-itemshowtype="0" target="_blank" linktype="text" data-linktype="2"><span textstyle="">给生信入门初学者的小贴士</span></a></span><span leaf=""><br></span></span></section><section><span><span leaf=""><a href="http://mp.weixin.qq.com/s?__biz=MzAwMzIzOTk5OQ==&amp;mid=2247506266&amp;idx=2&amp;sn=22b1164bc9602c25620d058847785422&amp;chksm=9b3ca00aac4b291c16d738271746ffb90149282a728f484e8a1952463c46a8e7dc4f883ab928&amp;scene=21#wechat_redirect" textvalue="如何搜索公众号过往发布内容(内含集锦)" data-itemshowtype="0" target="_blank" linktype="text" data-linktype="2">如何搜索公众号过往发布内容</a></span></span><span leaf=""><br></span></section><p><span><strong><span leaf="">已有生信基地联系方式的同学</span></strong></span><span><strong><span leaf="">无需重复添加</span></strong></span></p><section nodeleaf=""><img alt="图片" data-imgfileid="100042171" data-ratio="0.46111111111111114" data-src="https://mmbiz.qpic.cn/mmbiz_png/ImlBFVOwwpyZBGOXDqRYHpgiczxr9PX8gliblp8giaPGsr7w6IeoU6lExTAeEjx48Ke4LEJQfbdiarzXwUGvoRFvyA/640?wx_fmt=other&amp;from=appmsg&amp;wxfrom=5&amp;wx_lazy=1&amp;wx_co=1&amp;tp=webp" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/ImlBFVOwwpyZBGOXDqRYHpgiczxr9PX8gliblp8giaPGsr7w6IeoU6lExTAeEjx48Ke4LEJQfbdiarzXwUGvoRFvyA/640?wx_fmt=other&amp;from=appmsg&amp;wxfrom=5&amp;wx_lazy=1&amp;wx_co=1&amp;tp=webp"></section></section></section></section></section></section></section><section><section><section nodeleaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_gif/Ljib4So7yuWhFTIP7DCMLPwWh8Yhjs079f1vMqlYdExqbWaZVJyiawKkr21EDSNW740b4yMzkeHxFJULTpZGzYwQ/640?wx_fmt=gif&amp;wxfrom=5&amp;wx_lazy=1&amp;tp=webp" alt="图片" data-ratio="0.7619047619047619" data-w="84" data-width="80%" data-imgfileid="100042174" src="https://mmbiz.qpic.cn/mmbiz_gif/Ljib4So7yuWhFTIP7DCMLPwWh8Yhjs079f1vMqlYdExqbWaZVJyiawKkr21EDSNW740b4yMzkeHxFJULTpZGzYwQ/640?wx_fmt=gif&amp;wxfrom=5&amp;wx_lazy=1&amp;tp=webp"></section><section><span><span leaf="">您点的每个赞和在看，我都认真当成了喜欢</span></span></section></section></section></section><section><span leaf=""><br></span></section></section><p><span leaf=""><br></span></p><p><mp-style-type data-value="3"></mp-style-type></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/wV6Gyh8DN2NPLjt26TjY2w",target="_blank" rel="noopener noreferrer">原文链接</a>
