---
title: "一个PBMC的scATAC-seq基础分析：Signac"
date: 2025-04-03T03:40:11Z
draft: ["false"]
tags: [
  "fetched",
  "生信技能树"
]
categories: ["Acdemic"]
---
一个PBMC的scATAC-seq基础分析：Signac by 生信技能树
------
<div><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com" data-pm-slice="0 0 []"><blockquote><span></span><p><span leaf="">前面在写ATACseq学习笔记的时候，后台有很多粉丝询问scATAC-seq有没有相关资料，现在就来学习，<span textstyle="">开个新专辑《scATAC-seq分析》~</span></span></p></blockquote><p data-tool="mdnice编辑器"><span leaf="">scATAC-seq用的最多的包就是</span><code><span leaf="">Signac</span></code><span leaf="">包了，这个包来自Seurat包的扩展，分析步骤等跟Seurat非常像，下面来看看。学习官网为：</span></p><blockquote><span></span><p><span leaf="">https://stuartlab.org/signac/</span></p></blockquote><p data-tool="mdnice编辑器"><span leaf=""><span textstyle="">单细胞的ATAC-seq学习还是需要了解一些背景知识的</span>，不然就是茫然的纯跑代码：</span><a href="https://mp.weixin.qq.com/mp/appmsgalbum?__biz=MzAxMDkxODM1Ng==&amp;action=getalbum&amp;album_id=3825619502127398912&amp;scene=173&amp;subscene=&amp;sessionid=1743257525&amp;enterid=0&amp;from_msgid=2247540277&amp;from_itemidx=1&amp;count=3&amp;nolastread=1#wechat_redirect"><span leaf="">ATAC-Seq 数据分析2025</span></a></p><h4 data-tool="mdnice编辑器"><span></span><span><span leaf="">这个包可以分析的测序数据类型包括</span><code><span leaf="">scATAC-seq</span></code><span leaf="">，</span><code><span leaf="">scCUT&amp;Tag</span></code><span leaf="">，</span><code><span leaf="">scNTT-seq</span></code><span leaf="">，功能包括：</span></span><span></span></h4><ul><li><section><span leaf="">peaks 识别</span></section></li><li><section><span leaf="">每个细胞中基因组区域的定量</span></section></li><li><section><span leaf="">单细胞QC</span></section></li><li><section><span leaf="">降维，可视化，聚类</span></section></li><li><section><span leaf="">鉴定 细胞类型特异性 peaks</span></section></li><li><section><span leaf="">可视化 ‘pseudo-bulk’ 覆盖区域</span></section></li><li><section><span leaf="">整合不同的数据集</span></section></li><li><section><span leaf="">与单细胞转录组数据进行整合</span></section></li><li><section><span leaf="">序列 motif 富集分析</span></section></li><li><section><span leaf="">转录组因子足迹分析</span></section></li><li><section><span leaf="">peaks 注释</span></section></li><li><section><span leaf="">加速并行分析</span></section></li><li><section><span leaf="">与</span><span leaf="">Seurat</span><span leaf="">、</span><span leaf="">SeuratWrappers</span><span leaf="">、</span><span leaf="">SeuratDisk</span><span leaf="">以及</span><span leaf="">SeuratData</span><span leaf="">（Seurat数据集分发）的功能实现无缝对接。</span></section></li></ul><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">安装</span></span><span></span><span><span leaf=""> </span></span></h2><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">## 使用西湖大学的 Bioconductor镜像</span></span><span leaf=""><br></span><span leaf="">options(BioC_mirror=</span><span><span leaf="">"https://mirrors.westlake.edu.cn/bioconductor"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">options(</span><span><span leaf="">"repos"</span></span><span leaf="">=c(CRAN=</span><span><span leaf="">"https://mirrors.westlake.edu.cn/CRAN/"</span></span><span leaf="">))</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">setRepositories(ind=1:3)</span><span leaf=""><br></span><span leaf="">install.packages(</span><span><span leaf="">"Signac"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">rm(list=ls())</span><span leaf=""><br></span><span><span leaf="">## 加载包</span></span><span leaf=""><br></span><span leaf="">library(Signac)</span><span leaf=""><br></span><span leaf="">library(Seurat)</span><span leaf=""><br></span><span leaf="">library(GenomicRanges)</span><span leaf=""><br></span><span leaf="">library(ggplot2)</span><span leaf=""><br></span><span leaf="">library(patchwork)</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">先来学习一波官网的案例，与Seurat软件一样，<span textstyle="">官方给了很多示例，今天学习：https://stuartlab.org/signac/articles/pbmc_vignette</span></span></p><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">0、示例数据</span></span><span></span><span><span leaf=""> </span></span></h2><p data-tool="mdnice编辑器"><span leaf="">这个数据来自10x Genomics，为组织类型为：</span><code><span leaf=""> human peripheral blood mononuclear cells (PBMCs)</span></code><span leaf="">。</span></p><p data-tool="mdnice编辑器"><span leaf="">下载地址为：</span></p><ul><li><section><span leaf=""><span textstyle="">rawdata</span>：https://cf.10xgenomics.com/samples/cell-atac/2.1.0/10k_pbmc_ATACv2_nextgem_Chromium_Controller/10k_pbmc_ATACv2_nextgem_Chromium_Controller_filtered_peak_bc_matrix.h5</span></section></li><li><section><span leaf=""><span textstyle="">metadata</span>：https://cf.10xgenomics.com/samples/cell-atac/2.1.0/10k_pbmc_ATACv2_nextgem_Chromium_Controller/10k_pbmc_ATACv2_nextgem_Chromium_Controller_singlecell.csv</span></section></li><li><section><span leaf=""><span textstyle="">fragment文件</span>：https://cf.10xgenomics.com/samples/cell-atac/2.1.0/10k_pbmc_ATACv2_nextgem_Chromium_Controller/10k_pbmc_ATACv2_nextgem_Chromium_Controller_fragments.tsv.gz</span></section></li><li><section><span leaf=""><span textstyle="">fragement index文件</span>：https://cf.10xgenomics.com/samples/cell-atac/2.1.0/10k_pbmc_ATACv2_nextgem_Chromium_Controller/10k_pbmc_ATACv2_nextgem_Chromium_Controller_fragments.tsv.gz.tbi</span></section></li></ul><p data-tool="mdnice编辑器"><span leaf="">下载：</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># bash命令，或者自己选择用浏览器，迅雷都可以</span></span><span leaf=""><br></span><span leaf="">wget https://cf.10xgenomics.com/samples/cell-atac/2.1.0/10k_pbmc_ATACv2_nextgem_Chromium_Controller/10k_pbmc_ATACv2_nextgem_Chromium_Controller_filtered_peak_bc_matrix.h5</span><span leaf=""><br></span><span leaf="">wget https://cf.10xgenomics.com/samples/cell-atac/2.1.0/10k_pbmc_ATACv2_nextgem_Chromium_Controller/10k_pbmc_ATACv2_nextgem_Chromium_Controller_singlecell.csv</span><span leaf=""><br></span><span leaf="">wget https://cf.10xgenomics.com/samples/cell-atac/2.1.0/10k_pbmc_ATACv2_nextgem_Chromium_Controller/10k_pbmc_ATACv2_nextgem_Chromium_Controller_fragments.tsv.gz</span><span leaf=""><br></span><span leaf="">wget https://cf.10xgenomics.com/samples/cell-atac/2.1.0/10k_pbmc_ATACv2_nextgem_Chromium_Controller/10k_pbmc_ATACv2_nextgem_Chromium_Controller_fragments.tsv.gz.tbi</span><span leaf=""><br></span></code></pre><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">1、数据预处理</span></span><span></span><span><span leaf=""> </span></span></h2><p data-tool="mdnice编辑器"><span leaf="">上游 CellRanger的输出数据为：</span></p><p data-tool="mdnice编辑器"><strong><span leaf="">Peak/Cell matrix</span></strong><span leaf="">：<span textstyle="">矩阵的每一行代表基因组的一个区域（一个peak），该区域被预测为开放染色质的区域。矩阵中的每个值表示每个barcode（即一个细胞）映射到每个峰值内的Tn5整合位点的数量。</span></span></p><p data-tool="mdnice编辑器"><span leaf="">更详细的说明：https://support.10xgenomics.com/single-cell-atac/software/pipelines/latest/output/matrices</span></p><p data-tool="mdnice编辑器"><span leaf="">这里下载的是h5文件，如果是文件夹，则是标准的三个文件：</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">filtered_peak_bc_matrix</span><span leaf=""><br></span><span leaf="">├── barcodes.tsv</span><span leaf=""><br></span><span leaf="">├── peaks.bed</span><span leaf=""><br></span><span leaf="">└── matrix.mtx</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">peaks.bed的内容：</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">head filtered_peak_bc_matrix/peaks.bed</span><span leaf=""><br></span><span leaf="">chr1    237588  237917</span><span leaf=""><br></span><span leaf="">chr1    564444  565537</span><span leaf=""><br></span><span leaf="">chr1    567478  568248</span><span leaf=""><br></span><span leaf="">chr1    569021  569641</span><span leaf=""><br></span><span leaf="">chr1    713461  715293</span><span leaf=""><br></span><span leaf="">chr1    752379  753032</span><span leaf=""><br></span><span leaf="">chr1    762073  763379</span><span leaf=""><br></span><span leaf="">chr1    773651  774064</span><span leaf=""><br></span><span leaf="">chr1    779547  780286</span><span leaf=""><br></span><span leaf="">chr1    793345  794375</span><span leaf=""><br></span></code></pre><h3 data-tool="mdnice编辑器"><span></span><span><span leaf="">读取数据并创建seurat对象：</span></span><span></span></h3><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># 1.读取数据：</span></span><span leaf=""><br></span><span leaf="">counts &lt;- Read10X_h5(filename = </span><span><span leaf="">"PBMCs/10k_pbmc_ATACv2_nextgem_Chromium_Controller_filtered_peak_bc_matrix.h5"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">metadata &lt;- read.csv(</span><span leaf=""><br></span><span leaf="">  file = </span><span><span leaf="">"PBMCs/10k_pbmc_ATACv2_nextgem_Chromium_Controller_singlecell.csv"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  header = </span><span><span leaf="">TRUE</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  row.names = </span><span><span leaf="">1</span></span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span><span leaf=""># 简单的探索一下数据</span></span><span leaf=""><br></span><span leaf="">head(metadata)</span><span leaf=""><br></span><span leaf="">dim(counts)</span><span leaf=""><br></span><span><span leaf=""># 每一行为一个peak区间，每一列为一个细胞</span></span><span leaf=""><br></span><span leaf="">counts[</span><span><span leaf="">1</span></span><span leaf="">:</span><span><span leaf="">5</span></span><span leaf="">,</span><span><span leaf="">1</span></span><span leaf="">:</span><span><span leaf="">5</span></span><span leaf="">]</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 创建对象</span></span><span leaf=""><br></span><span leaf="">chrom_assay &lt;- CreateChromatinAssay(</span><span leaf=""><br></span><span leaf="">  counts = counts,</span><span leaf=""><br></span><span leaf="">  sep = c(</span><span><span leaf="">":"</span></span><span leaf="">, </span><span><span leaf="">"-"</span></span><span leaf="">),</span><span leaf=""><br></span><span leaf="">  fragments = </span><span><span leaf="">"PBMCs/10k_pbmc_ATACv2_nextgem_Chromium_Controller_fragments.tsv.gz"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  min.cells = </span><span><span leaf="">10</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  min.features = </span><span><span leaf="">200</span></span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 生成seurat对象</span></span><span leaf=""><br></span><span leaf="">pbmc &lt;- CreateSeuratObject(</span><span leaf=""><br></span><span leaf="">  counts = chrom_assay,</span><span leaf=""><br></span><span leaf="">  assay = </span><span><span leaf="">"peaks"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  meta.data = metadata</span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">pbmc</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wz9EeI44eLtic8PibMsOMaHJIwyRDLunXCm6pyLrre4iblDTQASTmwcljs7kGyErBibWDRLW5CdicOMFXg/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.13769751693002258" data-type="png" data-w="886" data-imgfileid="100056630" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wz9EeI44eLtic8PibMsOMaHJIwyRDLunXCm6pyLrre4iblDTQASTmwcljs7kGyErBibWDRLW5CdicOMFXg/640?wx_fmt=png&amp;from=appmsg"></span></figure><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">## An object of class Seurat </span></span><span leaf=""><br></span><span><span leaf="">## 165434 features across 10246 samples within 1 assay </span></span><span leaf=""><br></span><span><span leaf="">## Active assay: peaks (165434 features, 0 variable features)</span></span><span leaf=""><br></span><span><span leaf="">##  2 layers present: counts, data</span></span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">ATACseq数据存储在</span><code><span leaf="">ChromatinAssay</span></code><span leaf="">里面：</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">pbmc[[</span><span><span leaf="">'peaks'</span></span><span leaf="">]]</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># ChromatinAssay data with 165434 features for 10246 cells</span></span><span leaf=""><br></span><span><span leaf=""># Variable features: 0 </span></span><span leaf=""><br></span><span><span leaf=""># Genome: </span></span><span leaf=""><br></span><span><span leaf="">#   Annotation present: FALSE </span></span><span leaf=""><br></span><span><span leaf=""># Motifs present: FALSE </span></span><span leaf=""><br></span><span><span leaf=""># Fragment files: 1 </span></span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 查看基因组范围</span></span><span leaf=""><br></span><span leaf="">granges(pbmc)</span><span leaf=""><br></span><span><span leaf=""># GRanges object with 165434 ranges and 0 metadata columns:</span></span><span leaf=""><br></span><span><span leaf="">#   seqnames        ranges strand</span></span><span leaf=""><br></span><span><span leaf=""># &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt;</span></span><span leaf=""><br></span><span><span leaf="">#   [1]       chr1    9772-10660      *</span></span><span leaf=""><br></span><span><span leaf="">#   [2]       chr1 180712-181178      *</span></span><span leaf=""><br></span><span><span leaf="">#   [3]       chr1 181200-181607      *</span></span><span leaf=""><br></span><span><span leaf="">#   [4]       chr1 191183-192084      *</span></span><span leaf=""><br></span><span><span leaf="">#   [5]       chr1 267576-268461      *</span></span><span leaf=""><br></span><span><span leaf="">#   ...        ...           ...    ...</span></span><span leaf=""><br></span><span><span leaf=""># [165430] KI270713.1   13054-13909      *</span></span><span leaf=""><br></span><span><span leaf="">#   [165431] KI270713.1   15212-15933      *</span></span><span leaf=""><br></span><span><span leaf="">#   [165432] KI270713.1   21459-22358      *</span></span><span leaf=""><br></span><span><span leaf="">#   [165433] KI270713.1   29676-30535      *</span></span><span leaf=""><br></span><span><span leaf="">#   [165434] KI270713.1   36913-37813      *</span></span><span leaf=""><br></span><span><span leaf="">#   -------</span></span><span leaf=""><br></span><span><span leaf="">#   seqinfo: 35 sequences from an unspecified genome; no seqlengths</span></span><span leaf=""><br></span></code></pre><h3 data-tool="mdnice编辑器"><span></span><span><span leaf="">去除非染色体序列的部分：</span></span><span></span></h3><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># 去除非染色体的部分</span></span><span leaf=""><br></span><span leaf="">standardChromosomes(granges(pbmc))</span><span leaf=""><br></span><span leaf="">seqnames(granges(pbmc)) </span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">peaks.keep &lt;- seqnames(granges(pbmc)) %</span><span><span leaf="">in</span></span><span leaf="">% standardChromosomes(granges(pbmc))</span><span leaf=""><br></span><span leaf="">table(peaks.keep)</span><span leaf=""><br></span><span leaf="">pbmc &lt;- pbmc[as.vector(peaks.keep), ]</span><span leaf=""><br></span><span leaf="">pbmc </span><span leaf=""><br></span></code></pre><h3 data-tool="mdnice编辑器"><span></span><span><span leaf="">添加基因信息</span></span><span></span></h3><p data-tool="mdnice编辑器"><span leaf="">这里<span textstyle="">需要注意与上游分析使用的参考基因需要是同一个版本</span>，这个数据的10x Genomics用的版本为：</span><code><span leaf="">GRCh38-2020-A</span></code><span leaf="">，对应</span><code><span leaf="">Ensembl v98 patch release</span></code><span leaf="">，他们的对应关系可以从这里获取：</span></p><p data-tool="mdnice编辑器"><span leaf="">https://www.ensembl.org/info/website/archives/assembly.html</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># 添加基因注释</span></span><span leaf=""><br></span><span leaf="">library(AnnotationHub)</span><span leaf=""><br></span><span leaf="">ah &lt;- AnnotationHub()</span><span leaf=""><br></span><span leaf="">save(ah, file = </span><span><span leaf="">"ah.RData"</span></span><span leaf="">) </span><span><span leaf=""># 保存以便下次使用</span></span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># Search for the Ensembl 98 EnsDb for Homo sapiens on AnnotationHub</span></span><span leaf=""><br></span><span leaf="">query(ah, </span><span><span leaf="">"EnsDb.Hsapiens.v98"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">ensdb_v98 &lt;- ah[[</span><span><span leaf="">"AH75011"</span></span><span leaf="">]]</span><span leaf=""><br></span><span><span leaf=""># extract gene annotations from EnsDb</span></span><span leaf=""><br></span><span leaf="">annotations &lt;- GetGRangesFromEnsDb(ensdb = ensdb_v98)</span><span leaf=""><br></span><span><span leaf=""># change to UCSC style since the data was mapped to hg38</span></span><span leaf=""><br></span><span leaf="">seqlevels(annotations) &lt;- paste0(</span><span><span leaf="">'chr'</span></span><span leaf="">, seqlevels(annotations))</span><span leaf=""><br></span><span leaf="">genome(annotations) &lt;- </span><span><span leaf="">"hg38"</span></span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 将基因信息添加到对象中</span></span><span leaf=""><br></span><span leaf="">Annotation(pbmc) &lt;- annotations</span><span leaf=""><br></span></code></pre><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">2、计算QC指标</span></span><span></span><span><span leaf=""> </span></span></h2><p data-tool="mdnice编辑器"><span leaf="">scATAC-seq的数据指标与bulk ATAC-seq的原理一致，有下面这些指标：</span></p><ul><li><section><strong><span leaf="">Nucleosome banding pattern</span></strong><span leaf="">：DNA片段大小的直方图应该显示出一个强烈的核小体条带模式，这对应于缠绕在单个核小体周围的DNA长度。我们针对每个单细胞计算这一模式，并定量估算单核小体片段与无核小体片段的近似比例（存储为</span><code><span leaf="">nucleosome_signal</span></code><span leaf="">）；</span></section></li><li><section><strong><span leaf="">TSS富集打分</span></strong><span leaf="">：ENCODE项目定义了一个基于转录起始位点（TSS）中心的片段与TSS侧翼区域片段比例的ATAC-seq靶向评分（详见</span><span leaf="">https://www.encodeproject.org/data-standards/terms/</span><span leaf="">）。通常，<span textstyle="">质量不佳的ATAC-seq实验会有较低的TSS富集评分。</span>我们可以使用</span><code><span leaf="">TSSEnrichment()</span></code><span leaf="">函数为每个细胞计算这一指标，结果存储在元数据的</span><code><span leaf="">TSS.enrichment</span></code><span leaf="">列中；</span></section></li><li><section><strong><span leaf="">peaks区域内的片段总数</span></strong><span leaf="">：用于衡量细胞测序深度/复杂度的指标。<span textstyle="">测序深度较低的细胞可能由于read数过少而需要被排除。而具有极高片段数的细胞可能代表双细胞、细胞核聚集或其他异常情况</span>；</span></section></li><li><section><strong><span leaf="">peaks区域内的片段比例</span></strong><span leaf="">：表示所有片段中落入ATAC-seq的peaks区域的比例。<span textstyle="">具有低值（例如&lt;15%-20%）的细胞通常代表低质量细胞或技术假象，应该被去除。</span>需要注意的是，这一比例可能对所使用的peaks集合较为敏感；</span></section></li><li><section><strong><span leaf="">基因组黑名单区域的reads比例</span></strong><span leaf="">：ENCODE项目提供了一个<span textstyle="">黑名单区域列表，这些区域中的reads通常与假象信号相关联。</span>如果细胞中映射到这些区域的reads比例较高（与映射到peaks区域的reads相比），则这些细胞通常代表技术假象，应该被去除。Signac软件包中包含了人类（hg19和hg38）、小鼠（mm9和mm10）、果蝇（dm3和dm6）以及秀丽隐杆线虫（ce10和ce11）的ENCODE黑名单区域。可以使用</span><code><span leaf="">FractionCountsInRegion()</span></code><span leaf="">函数来计算每个细胞在给定区域集合内的所有计数的比例。我们可以利用这个函数和黑名单区域来计算每个细胞的黑名单计数比例。</span></section></li></ul><p data-tool="mdnice编辑器"><span leaf="">这三个指标：</span><code><span leaf="">peaks区域内的片段总数</span></code><span leaf="">，</span><code><span leaf="">peaks区域内的片段比例</span></code><span leaf="">，</span><code><span leaf="">基因组黑名单区域的reads比例</span></code><span leaf="">在CellRanger 的标准输出结果中：</span></p><p data-tool="mdnice编辑器"><span leaf="">设置 </span><code><span leaf="">quantiles=TRUE</span></code><span leaf="">，可以快速的看出合适的过滤cutoff。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">## 2.计算QC指标</span></span><span leaf=""><br></span><span><span leaf=""># compute nucleosome signal score per cell</span></span><span leaf=""><br></span><span leaf="">pbmc &lt;- NucleosomeSignal(object = pbmc)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># compute TSS enrichment score per cell</span></span><span leaf=""><br></span><span leaf="">pbmc &lt;- TSSEnrichment(object = pbmc)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># add fraction of reads in peaks</span></span><span leaf=""><br></span><span leaf="">pbmc</span><span><span leaf="">$pct_reads_in_peaks</span></span><span leaf=""> &lt;- pbmc</span><span><span leaf="">$peak_region_fragments</span></span><span leaf=""> / pbmc</span><span><span leaf="">$passed_filters</span></span><span leaf=""> * 100</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># add blacklist ratio</span></span><span leaf=""><br></span><span leaf="">pbmc</span><span><span leaf="">$blacklist_ratio</span></span><span leaf=""> &lt;- FractionCountsInRegion(</span><span leaf=""><br></span><span leaf="">  object = pbmc, </span><span leaf=""><br></span><span leaf="">  assay = </span><span><span leaf="">'peaks'</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  regions = blacklist_hg38_unified</span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">head(pbmc@meta.data)</span><span leaf=""><br></span><span leaf="">colnames(pbmc@meta.data)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 可视化</span></span><span leaf=""><br></span><span leaf="">DensityScatter(pbmc, x = </span><span><span leaf="">'nCount_peaks'</span></span><span leaf="">, y = </span><span><span leaf="">'TSS.enrichment'</span></span><span leaf="">, log_x = TRUE, quantiles = TRUE)</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wz9EeI44eLtic8PibMsOMaHJIP0UPL1DRiatyG6icjhXtQfI4ic6FtDDvO3VpholZno0GSaPNJR8O51YUg/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.8163265306122449" data-type="png" data-w="784" data-imgfileid="100056634" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wz9EeI44eLtic8PibMsOMaHJIP0UPL1DRiatyG6icjhXtQfI4ic6FtDDvO3VpholZno0GSaPNJR8O51YUg/640?wx_fmt=png&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">我们还可以观察所有细胞的片段长度周期性，并根据核小体信号强度的高低对细胞进行分组。从前面的图表可以看出，单核小体/无核小体比例异常的细胞具有不同的核小体条带模式。其余细胞表现出典型的成功ATAC-seq实验的模式。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">pbmc</span><span><span leaf="">$nucleosome_group</span></span><span leaf=""> &lt;- ifelse(pbmc</span><span><span leaf="">$nucleosome_signal</span></span><span leaf=""> &gt; 4, </span><span><span leaf="">'NS &gt; 4'</span></span><span leaf="">, </span><span><span leaf="">'NS &lt; 4'</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">FragmentHistogram(object = pbmc, group.by = </span><span><span leaf="">'nucleosome_group'</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wz9EeI44eLtic8PibMsOMaHJImmR5DeaDBmZnyyDA6XObjY0qfia6byWMdib69AL8785diaNMcOfIGVYXQ/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.5294117647058824" data-type="png" data-w="867" data-imgfileid="100056632" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wz9EeI44eLtic8PibMsOMaHJImmR5DeaDBmZnyyDA6XObjY0qfia6byWMdib69AL8785diaNMcOfIGVYXQ/640?wx_fmt=png&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">还可以分别使用小提琴图来绘制每个质量控制指标的分布：</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">VlnPlot(</span><span leaf=""><br></span><span leaf="">  object = pbmc,</span><span leaf=""><br></span><span leaf="">  features = c(</span><span><span leaf="">'nCount_peaks'</span></span><span leaf="">, </span><span><span leaf="">'TSS.enrichment'</span></span><span leaf="">, </span><span><span leaf="">'blacklist_ratio'</span></span><span leaf="">, </span><span><span leaf="">'nucleosome_signal'</span></span><span leaf="">, </span><span><span leaf="">'pct_reads_in_peaks'</span></span><span leaf="">),</span><span leaf=""><br></span><span leaf="">  pt.size = 0.1,</span><span leaf=""><br></span><span leaf="">  ncol = 5</span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wz9EeI44eLtic8PibMsOMaHJIlRibW5vLMEdbI5MLibnLVYcdxgLCpDicmvVpPZtNxxFrBLXMWAX6Ficlhw/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.49444444444444446" data-type="png" data-w="1080" data-imgfileid="100056633" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wz9EeI44eLtic8PibMsOMaHJIlRibW5vLMEdbI5MLibnLVYcdxgLCpDicmvVpPZtNxxFrBLXMWAX6Ficlhw/640?wx_fmt=png&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">根据上面的数据分布，过滤条件如下，<span textstyle="">如果是自己的数据需要根据项目背景进行过滤cutoff调整：</span></span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">pbmc &lt;- subset(</span><span leaf=""><br></span><span leaf="">  x = pbmc,</span><span leaf=""><br></span><span leaf="">  subset = nCount_peaks &gt; 9000 &amp;</span><span leaf=""><br></span><span leaf="">    nCount_peaks &lt; 100000 &amp;</span><span leaf=""><br></span><span leaf="">    pct_reads_in_peaks &gt; 40 &amp;</span><span leaf=""><br></span><span leaf="">    blacklist_ratio &lt; 0.01 &amp;</span><span leaf=""><br></span><span leaf="">    nucleosome_signal &lt; 4 &amp;</span><span leaf=""><br></span><span leaf="">    TSS.enrichment &gt; 4</span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">pbmc</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># An object of class Seurat </span></span><span leaf=""><br></span><span><span leaf=""># 165376 features across 9651 samples within 1 assay </span></span><span leaf=""><br></span><span><span leaf=""># Active assay: peaks (165376 features, 0 variable features)</span></span><span leaf=""><br></span><span><span leaf=""># 2 layers present: counts, data</span></span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">还剩下9k多个细胞。</span></p><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">3、标准化和线性降维</span></span><span></span><span><span leaf=""> </span></span></h2><p data-tool="mdnice编辑器"><span leaf="">分为三个小步骤：</span></p><ul><li><section><strong><span leaf="">标准化</span></strong><span leaf="">：Signac进行</span><code><span leaf=""><span textstyle="">TF-IDF</span></span></code><span leaf=""><span textstyle="">标准化</span>，额还不知道是个什么标准化，mark一下后面再来学习这个细节</span></section></li><li><section><strong><span leaf="">特征选择</span></strong><span leaf="">：scATAC-seq数据的低动态范围使得我们难以像处理scRNA-seq数据那样进行可变特征选择。但可以选择仅使用排名top n%的特征（peaks）来进行降维，或者使用</span><code><span leaf="">FindTopFeatures()</span></code><span leaf="">函数移除出现在少于n个细胞中的特征</span></section></li><li><section><strong><span leaf="">降维</span></strong><span leaf="">：在上述选择的特征（峰值）所构成的TF-IDF矩阵上运行奇异值分解（SVD）</span></section></li></ul><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">## 3.标准化降维</span></span><span leaf=""><br></span><span leaf="">pbmc &lt;- RunTFIDF(pbmc)</span><span leaf=""><br></span><span leaf="">pbmc &lt;- FindTopFeatures(pbmc, min.cutoff = </span><span><span leaf="">'q0'</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">pbmc &lt;- RunSVD(pbmc)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">DepthCor(pbmc)</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf=""><span textstyle="">第一个 LSI 成分通常捕获测序深度（技术变异）而非生物学变异</span>。如果情况如此，那么该成分应从下游分析中移除。</span><strong><span leaf="">可以使用</span><code><span leaf="">DepthCor()</span></code><span leaf="">函数来评估每个LSI成分与测序深度之间的相关性：</span></strong></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wz9EeI44eLtic8PibMsOMaHJIiaXJTPicJ6pSRI5G2PSCsSJnErXBp6AKN3fUUf9eRj2ujichPZLVDzthw/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.7302452316076294" data-type="png" data-w="734" data-imgfileid="100056631" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wz9EeI44eLtic8PibMsOMaHJIiaXJTPicJ6pSRI5G2PSCsSJnErXBp6AKN3fUUf9eRj2ujichPZLVDzthw/640?wx_fmt=png&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">在这里，我们看到第一个 LSI 成分与细胞的总计数数量之间存在非常强的相关性（图中看到是就是前10个LSI，除了第一个基本上都在0.3以内）。我们将在后续步骤中不使用这个成分，因为我们不想根据细胞的总测序深度将它们分组，而是根据细胞类型特异性峰值的可及性模式对细胞进行分组。</span></p><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">4、非线性标准化和降维</span></span><span></span><span><span leaf=""> </span></span></h2><p data-tool="mdnice编辑器"><span leaf="">现在细胞已经被嵌入到低维空间中，我们可以使用通常用于scRNA-seq数据分析的方法来进行基于图的聚类和非线性降维以用于可视化。</span><code><span leaf="">RunUMAP()</span></code><span leaf="">、</span><code><span leaf="">FindNeighbors()</span></code><span leaf="">和</span><code><span leaf="">FindClusters()</span></code><span leaf="">这些函数都来自Seurat包。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">## 4.非线性标准化聚类</span></span><span leaf=""><br></span><span leaf="">pbmc &lt;- RunUMAP(object = pbmc, reduction = </span><span><span leaf="">'lsi'</span></span><span leaf="">, dims = <span textstyle="">2:30</span>)</span><span leaf=""><br></span><span leaf="">pbmc &lt;- FindNeighbors(object = pbmc, reduction = </span><span><span leaf="">'lsi'</span></span><span leaf="">, dims = <span textstyle="">2:30</span>)</span><span leaf=""><br></span><span leaf="">pbmc &lt;- FindClusters(object = pbmc, verbose = FALSE, algorithm = 3)</span><span leaf=""><br></span><span leaf="">DimPlot(object = pbmc, label = TRUE) + NoLegend()</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">聚类结果如下：</span></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wz9EeI44eLtic8PibMsOMaHJIkYnl3WUMJ8AjGYVasLf6AnrkC2FLPoMiaznIInRc08xv8EnKFZwsLyQ/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.7829181494661922" data-type="png" data-w="562" data-imgfileid="100056635" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wz9EeI44eLtic8PibMsOMaHJIkYnl3WUMJ8AjGYVasLf6AnrkC2FLPoMiaznIInRc08xv8EnKFZwsLyQ/640?wx_fmt=png&amp;from=appmsg"></span></figure><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">5、构建基因活性矩阵</span></span><span></span><span><span leaf=""> </span></span></h2><p data-tool="mdnice编辑器"><span leaf="">UMAP图展示了人类血液中存在多个细胞群体。如果你熟悉PBMC的scRNA-seq分析，你可能会在scATAC-seq数据中识别出某些髓系和淋巴系细胞群体的存在。</span></p><p data-tool="mdnice编辑器"><span leaf="">然而，<span textstyle="">在scATAC-seq数据中注释和解释细胞簇更具挑战性，因为人们对非编码基因组区域的功能角色的了解远少于对编码蛋白的基因区域的了解</span>。</span></p><p data-tool="mdnice编辑器"><span leaf="">我们可以通过评估与基因相关的染色质可及性来量化基因组中每个基因的活性，并从scATAC-seq数据中创建一个新的基因活性检测方法。在这里，我们将使用一种简单的方法，即汇总与基因主体和启动子区域相交的片段（我们还建议探索Cicero工具，它可以实现类似的目标，我们在这里提供了一个展示如何在Signac工作流中运行Cicero的教程：</span></p><p data-tool="mdnice编辑器"><span leaf="">https://stuartlab.org/signac/articles/cicero）。</span></p><p data-tool="mdnice编辑器"><span leaf="">为了创建基因活性矩阵，我们提取基因坐标，并将其扩展以包含上游2kb的区域（因为启动子的可及性通常与基因表达相关）。然后，我们使用</span><code><span leaf="">FeatureMatrix()</span></code><span leaf="">函数统计每个细胞映射到这些区域的片段数量。这些步骤可以通过</span><code><span leaf="">GeneActivity()</span></code><span leaf="">函数自动完成：</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">## 5.构建基因活性矩阵</span></span><span leaf=""><br></span><span leaf="">gene.activities &lt;- GeneActivity(pbmc)</span><span leaf=""><br></span><span leaf="">dim(gene.activities)</span><span leaf=""><br></span><span leaf="">gene.activities[1:5,1:5]</span><span leaf=""><br></span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># add the gene activity matrix to the Seurat object as a new assay and normalize it</span></span><span leaf=""><br></span><span leaf="">pbmc[[</span><span><span leaf="">'RNA'</span></span><span leaf="">]] &lt;- CreateAssayObject(counts = gene.activities)</span><span leaf=""><br></span><span leaf="">pbmc &lt;- NormalizeData(</span><span leaf=""><br></span><span leaf="">  object = pbmc,</span><span leaf=""><br></span><span leaf="">  assay = </span><span><span leaf="">'RNA'</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  normalization.method = </span><span><span leaf="">'LogNormalize'</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  scale.factor = median(pbmc</span><span><span leaf="">$nCount_RNA</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">现在可以可视化典型标记基因的活性，以帮助解释ATAC-seq聚类结果。以下是一些基因所对应的细胞类型：</span></p><p data-tool="mdnice编辑器"><strong><span leaf="">B细胞</span></strong></p><ul><li><section><strong><span leaf="">MS4A1（CD20）</span></strong><span leaf="">：是B细胞的一个经典标记基因，主要在成熟的B细胞上表达。</span></section></li></ul><p data-tool="mdnice编辑器"><strong><span leaf="">T细胞</span></strong></p><ul><li><section><strong><span leaf="">CD3D</span></strong><span leaf="">：是T细胞表面的CD3复合物的一部分，主要在T细胞上表达。</span></section></li><li><section><strong><span leaf="">LEF1</span></strong><span leaf="">：虽然LEF1在多种细胞类型中都有表达，但在T细胞中也有较高表达，且与T细胞的某些功能相关。</span></section></li></ul><p data-tool="mdnice编辑器"><strong><span leaf="">NK细胞</span></strong></p><ul><li><section><strong><span leaf="">NKG7</span></strong><span leaf="">：是NK细胞的一个标记基因，主要在NK细胞中表达。</span></section></li></ul><p data-tool="mdnice编辑器"><strong><span leaf="">单核细胞/巨噬细胞</span></strong></p><ul><li><section><strong><span leaf="">TREM1</span></strong><span leaf="">：主要在单核细胞/巨噬细胞等髓系细胞上表达。</span></section></li><li><section><strong><span leaf="">LYZ</span></strong><span leaf="">：编码溶菌酶，是单核细胞/巨噬细胞等髓系细胞的一个标记基因</span></section></li></ul><section data-tool="mdnice编辑器"><table><thead><tr><th><section><span leaf="">细胞类型</span></section></th><th><section><span leaf="">标记基因</span></section></th></tr></thead><tbody><tr><td><section><span leaf="">B细胞</span></section></td><td><section><span leaf="">MS4A1</span></section></td></tr><tr><td><section><span leaf="">T细胞</span></section></td><td><section><span leaf="">CD3D、LEF1</span></section></td></tr><tr><td><section><span leaf="">NK细胞</span></section></td><td><section><span leaf="">NKG7</span></section></td></tr><tr><td><section><span leaf="">单核细胞/巨噬细胞</span></section></td><td><section><span leaf="">TREM1、LYZ</span></section></td></tr></tbody></table></section><blockquote><span></span><p><span leaf=""><span textstyle="">需要注意的是，这些活性的测量值会比单细胞RNA-seq的测量值更加嘈杂。</span>这是因为它们代表了稀疏染色质数据的测量，并且它们假定基因主体/启动子的可及性与基因表达之间存在一般对应关系，而这种关系并不总是成立。尽管如此，我们仍然可以根据这些基因活性特征初步区分单核细胞、B细胞、T细胞和NK细胞的群体。然而，仅依靠监督分析来进一步细分这些细胞类型是具有挑战性的。</span></p></blockquote><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># 可视化</span></span><span leaf=""><br></span><span leaf="">DefaultAssay(pbmc) &lt;- </span><span><span leaf="">'RNA'</span></span><span leaf=""><br></span><span leaf="">FeaturePlot(</span><span leaf=""><br></span><span leaf="">  object = pbmc,</span><span leaf=""><br></span><span leaf="">  features = c(</span><span><span leaf="">'MS4A1'</span></span><span leaf="">, </span><span><span leaf="">'CD3D'</span></span><span leaf="">, </span><span><span leaf="">'LEF1'</span></span><span leaf="">, </span><span><span leaf="">'NKG7'</span></span><span leaf="">, </span><span><span leaf="">'TREM1'</span></span><span leaf="">, </span><span><span leaf="">'LYZ'</span></span><span leaf="">),</span><span leaf=""><br></span><span leaf="">  pt.size = 0.1,</span><span leaf=""><br></span><span leaf="">  max.cutoff = </span><span><span leaf="">'q95'</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  ncol = 3</span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wz9EeI44eLtic8PibMsOMaHJIUgdaI2qU2kLmWMrpfngllLiaF8RFn907VGvne46icMoCEysicxf3TUu7w/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.5453703703703704" data-type="png" data-w="1080" data-imgfileid="100056639" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wz9EeI44eLtic8PibMsOMaHJIUgdaI2qU2kLmWMrpfngllLiaF8RFn907VGvne46icMoCEysicxf3TUu7w/640?wx_fmt=png&amp;from=appmsg"></span></figure><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">6、与单细胞数据进行整合</span></span><span></span><span><span leaf=""> </span></span></h2><p data-tool="mdnice编辑器"><span leaf="">为了帮助解释scATAC-seq数据，我们可以基于同一组织类型（人类PBMC）的scRNA-seq实验来对细胞进行分类。</span></p><p data-tool="mdnice编辑器"><span leaf="">我们利用跨模态整合和标签转移的方法：https://doi.org/10.1016/j.cell.2019.05.031，旨在识别基因活性矩阵和scRNA-seq数据集中共享的相关性模式，以识别两种模态之间的匹配生物学状态。这一过程将为每个细胞返回每个基于单细胞RNA-seq定义的聚类标签的分类得分。</span></p><p data-tool="mdnice编辑器"><span leaf="">单细胞数据为10x Genomics提供的预处理过的人类PBM scRNA-seq数据集，下载：https://signac-objects.s3.amazonaws.com/pbmc_10k_v3.rds</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">## 6.与单细胞数据整合</span></span><span leaf=""><br></span><span><span leaf=""># Load the pre-processed scRNA-seq data for PBMCs</span></span><span leaf=""><br></span><span leaf="">pbmc_rna &lt;- readRDS(</span><span><span leaf="">"PBMCs/pbmc_10k_v3.rds"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">pbmc_rna &lt;- UpdateSeuratObject(pbmc_rna)</span><span leaf=""><br></span><span leaf="">DimPlot(object = pbmc_rna, label = TRUE) + NoLegend()</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">已经注释好的单细胞：</span></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wz9EeI44eLtic8PibMsOMaHJIzXEqvE7Xr4IGygWe8qHnANxke7T9n6mWLsEGcdJNFOhmoVUXTUICnw/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.8937784522003035" data-type="png" data-w="659" data-imgfileid="100056636" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wz9EeI44eLtic8PibMsOMaHJIzXEqvE7Xr4IGygWe8qHnANxke7T9n6mWLsEGcdJNFOhmoVUXTUICnw/640?wx_fmt=png&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">找到 anchors：</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">transfer.anchors &lt;- FindTransferAnchors(</span><span leaf=""><br></span><span leaf="">  reference = pbmc_rna,</span><span leaf=""><br></span><span leaf="">  query = pbmc,</span><span leaf=""><br></span><span leaf="">  reduction = </span><span><span leaf="">'cca'</span></span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">transfer.anchors</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">predicted.labels &lt;- TransferData(</span><span leaf=""><br></span><span leaf="">  anchorset = transfer.anchors,</span><span leaf=""><br></span><span leaf="">  refdata = pbmc_rna</span><span><span leaf="">$celltype</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  weight.reduction = pbmc[[</span><span><span leaf="">'lsi'</span></span><span leaf="">]],</span><span leaf=""><br></span><span leaf="">  dims = 2:30</span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">predicted.labels 的内容如下：</span></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wz9EeI44eLtic8PibMsOMaHJIjayBKB6C8Rt0hMOvvQQraSnCFk7rctaxHficzOqW7q4Y080iaZmctDdA/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.3194444444444444" data-type="png" data-w="1080" data-imgfileid="100056638" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wz9EeI44eLtic8PibMsOMaHJIjayBKB6C8Rt0hMOvvQQraSnCFk7rctaxHficzOqW7q4Y080iaZmctDdA/640?wx_fmt=png&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">可视化预测结果：</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># 添加到对象中</span></span><span leaf=""><br></span><span leaf="">pbmc &lt;- AddMetaData(object = pbmc, metadata = predicted.labels)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">plot1 &lt;- DimPlot(</span><span leaf=""><br></span><span leaf="">  object = pbmc_rna,</span><span leaf=""><br></span><span leaf="">  group.by = </span><span><span leaf="">'celltype'</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  label = TRUE,</span><span leaf=""><br></span><span leaf="">  repel = TRUE) + NoLegend() + ggtitle(</span><span><span leaf="">'scRNA-seq'</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">plot2 &lt;- DimPlot(</span><span leaf=""><br></span><span leaf="">  object = pbmc,</span><span leaf=""><br></span><span leaf="">  group.by = </span><span><span leaf="">'predicted.id'</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  label = TRUE,</span><span leaf=""><br></span><span leaf="">  repel = TRUE) + NoLegend() + ggtitle(</span><span><span leaf="">'scATAC-seq'</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">plot1 + plot2</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">现在这个图就是很多文献中scRNA-seq+scATAC-seq联合分析的结果啦：</span></p><p data-tool="mdnice编辑器"><span leaf="">可以看到，基于scRNA-seq的分类结果与使用scATAC-seq数据计算得到的UMAP可视化结果是一致的。然而，<span textstyle="">需要注意的是，在scATAC-seq数据集中，有一小部分细胞被预测为血小板。这是出乎意料的，因为血小板是无核的，本不应被scATAC-seq检测到。</span>有可能被预测为血小板的细胞实际上是血小板的前体细胞——巨核细胞，它们主要存在于骨髓中，但在健康患者的外周血液中很少见，例如这些PBMC所来自的个体。鉴于巨核细胞在正常骨髓中已经极为罕见（&lt; 0.1%），这种可能性似乎不大。</span></p><p data-tool="mdnice编辑器"><span leaf=""><span textstyle="">那是为什么？</span>我猜测是因为scRNA-seq中存在小部分的血小板，而scATAC-seq中的细胞类型预测标签都来自于scRNA-seq。</span></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wz9EeI44eLtic8PibMsOMaHJIhicy4lN0f8N1HvToclA81rSpx0OrvgzL6jicibEM6Ppdgfg2oKVpOEmBg/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.5138888888888888" data-type="png" data-w="1080" data-imgfileid="100056637" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wz9EeI44eLtic8PibMsOMaHJIhicy4lN0f8N1HvToclA81rSpx0OrvgzL6jicibEM6Ppdgfg2oKVpOEmBg/640?wx_fmt=png&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">比如我们前面学习的github文献：</span><a href="https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247538090&amp;idx=1&amp;sn=b3311f05b3a7364d7a61c726bf4b231c&amp;scene=21#wechat_redirect"><span leaf="">Github带有全套代码分享的文献复现2025</span></a></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wz9EeI44eLtic8PibMsOMaHJIUdJ47h8pSD0iaibMwOYyXWRuPicRrH57RQSTXsMrffYGtsGdjIiawrX3Qw/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.5850860420650096" data-type="png" data-w="1046" data-imgfileid="100056643" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wz9EeI44eLtic8PibMsOMaHJIUdJ47h8pSD0iaibMwOYyXWRuPicRrH57RQSTXsMrffYGtsGdjIiawrX3Qw/640?wx_fmt=png&amp;from=appmsg"></span></figure><h3 data-tool="mdnice编辑器"><span></span><span><span leaf="">探索一下：platelets</span></span><span></span></h3><p data-tool="mdnice编辑器"><span leaf="">绘制被分配到每个标签的细胞的预测得分，结果显示“血小板”细胞的得分相对较低（&lt;0.8），这表明对其分配的细胞身份的可信度较低。在大多数情况下，这些细胞最可能的细胞身份是“CD4 naive”。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">VlnPlot(pbmc, </span><span><span leaf="">'prediction.score.max'</span></span><span leaf="">, group.by = </span><span><span leaf="">'predicted.id'</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wz9EeI44eLtic8PibMsOMaHJIYkzqbz6FZiciaI4Z9T2FwicAnkl4iaOHRvJwyLvsRqAqh2jt4icKDGB3WXQ/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.5942796610169492" data-type="png" data-w="944" data-imgfileid="100056644" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wz9EeI44eLtic8PibMsOMaHJIYkzqbz6FZiciaI4Z9T2FwicAnkl4iaOHRvJwyLvsRqAqh2jt4icKDGB3WXQ/640?wx_fmt=png&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">提取每种细胞的预测打分：</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># Identify the metadata columns that start with "prediction.score."</span></span><span leaf=""><br></span><span leaf="">metadata_attributes &lt;- colnames(pbmc[[]])</span><span leaf=""><br></span><span leaf="">prediction_score_attributes &lt;- grep(</span><span><span leaf="">"^prediction.score."</span></span><span leaf="">, metadata_attributes, value = TRUE)</span><span leaf=""><br></span><span leaf="">prediction_score_attributes &lt;- setdiff(prediction_score_attributes, </span><span><span leaf="">"prediction.score.max"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">prediction_score_attributes</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># [1] "prediction.score.CD4.Memory"             "prediction.score.CD14..Monocytes"        "prediction.score.NK.dim"             </span></span><span leaf=""><br></span><span><span leaf=""># [4] "prediction.score.pre.B.cell"             "prediction.score.NK.bright"              "prediction.score.CD4.Naive"           </span></span><span leaf=""><br></span><span><span leaf=""># [7] "prediction.score.CD8.Naive"              "prediction.score.pDC"                    "prediction.score.Double.negative.T.cell"</span></span><span leaf=""><br></span><span><span leaf=""># [10] "prediction.score.CD16..Monocytes"        "prediction.score.Platelet"               "prediction.score.CD8.effector"       </span></span><span leaf=""><br></span><span><span leaf=""># [13] "prediction.score.B.cell.progenitor"      "prediction.score.Dendritic.cell"        </span></span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># Extract the prediction score attributes for these cells</span></span><span leaf=""><br></span><span leaf="">predicted_platelets &lt;- </span><span><span leaf="">which</span></span><span leaf="">(pbmc</span><span><span leaf="">$predicted</span></span><span leaf="">.id == </span><span><span leaf="">"Platelet"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">predicted_platelets</span><span leaf=""><br></span><span leaf="">platelet_scores &lt;- pbmc[[]][predicted_platelets, prediction_score_attributes]</span><span leaf=""><br></span><span leaf="">platelet_scores</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># Order the columns by their average values in descending order</span></span><span leaf=""><br></span><span leaf="">ordered_columns &lt;- names(sort(colMeans(platelet_scores, na.rm = TRUE), decreasing = TRUE))</span><span leaf=""><br></span><span leaf="">ordered_platelet_scores_df &lt;- platelet_scores[, ordered_columns]</span><span leaf=""><br></span><span><span leaf="">print</span></span><span leaf="">(ordered_platelet_scores_df)</span><span leaf=""><br></span></code></pre><blockquote><span></span><p><span leaf="">由于被分类为“血小板”的细胞数量极少（少于20个），很难确定它们的确切细胞身份。需要更大的数据集才能有信心识别这一细胞群体的特定峰值，并进行进一步分析以正确注释它们。因此，在下游分析中，我们将移除这些被预测的极为罕见的细胞状态，仅保留总细胞数超过20个的细胞注释。</span></p></blockquote><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># 移除血小板</span></span><span leaf=""><br></span><span leaf="">predicted_id_counts &lt;- table(pbmc</span><span><span leaf="">$predicted</span></span><span leaf="">.id)</span><span leaf=""><br></span><span leaf="">predicted_id_counts</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># Identify the predicted.id values that have more than 20 cells</span></span><span leaf=""><br></span><span leaf="">major_predicted_ids &lt;- names(predicted_id_counts[predicted_id_counts &gt; 20])</span><span leaf=""><br></span><span leaf="">pbmc &lt;- pbmc[, pbmc</span><span><span leaf="">$predicted</span></span><span leaf="">.id %</span><span><span leaf="">in</span></span><span leaf="">% major_predicted_ids]</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># change cell identities to the per-cell predicted labels</span></span><span leaf=""><br></span><span leaf="">Idents(pbmc) &lt;- pbmc</span><span><span leaf="">$predicted</span></span><span leaf="">.id</span><span leaf=""><br></span></code></pre><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">7、细胞类型差异peaks分析</span></span><span></span><span><span leaf=""> </span></span></h2><p data-tool="mdnice编辑器"><span leaf="">为了找到细胞亚群之间的差异可及性区域，我们可以进行差异可及性（DA）检验。一种简单的方法是执行Wilcoxon秩和检验，而</span><code><span leaf="">presto</span></code><span leaf="">包已经实现了一种极其快速的Wilcoxon检验，可以直接在Seurat对象上运行。</span></p><p data-tool="mdnice编辑器"><span leaf="">筛选 Naive CD4 cells and CD14 monocytes 之间的差异peaks：</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">## 7.差异peaks分析</span></span><span leaf=""><br></span><span><span leaf=""># remotes::install_github('immunogenomics/presto')</span></span><span leaf=""><br></span><span><span leaf=""># change back to working with peaks instead of gene activities</span></span><span leaf=""><br></span><span leaf="">DefaultAssay(pbmc) &lt;- </span><span><span leaf="">'peaks'</span></span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># wilcox is the default option for test.use</span></span><span leaf=""><br></span><span leaf="">da_peaks &lt;- FindMarkers(</span><span leaf=""><br></span><span leaf="">  object = pbmc,</span><span leaf=""><br></span><span leaf="">  ident.1 = </span><span><span leaf="">"CD4 Naive"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  ident.2 = </span><span><span leaf="">"CD14+ Monocytes"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  test.use = </span><span><span leaf="">'wilcox'</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  min.pct = 0.1</span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">head(da_peaks)</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wz9EeI44eLtic8PibMsOMaHJIOHDUJfMRtYk03DiaFhcanCdgImrmz7iaepwmneludJQlkTkibibG8ibHcNQ/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.21815286624203822" data-type="png" data-w="628" data-imgfileid="100056640" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wz9EeI44eLtic8PibMsOMaHJIOHDUJfMRtYk03DiaFhcanCdgImrmz7iaepwmneludJQlkTkibibG8ibHcNQ/640?wx_fmt=png&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">选一个peaks可视化：</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># 可视化差异peaks</span></span><span leaf=""><br></span><span leaf="">rownames(da_peaks)[1]</span><span leaf=""><br></span><span leaf="">plot1 &lt;- VlnPlot(</span><span leaf=""><br></span><span leaf="">  object = pbmc,</span><span leaf=""><br></span><span leaf="">  features = rownames(da_peaks)[1],</span><span leaf=""><br></span><span leaf="">  pt.size = 0.1,</span><span leaf=""><br></span><span leaf="">  idents = c(</span><span><span leaf="">"CD4 Naive"</span></span><span leaf="">,</span><span><span leaf="">"CD14+ Monocytes"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">plot2 &lt;- FeaturePlot(</span><span leaf=""><br></span><span leaf="">  object = pbmc,</span><span leaf=""><br></span><span leaf="">  features = rownames(da_peaks)[1],</span><span leaf=""><br></span><span leaf="">  pt.size = 0.1</span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">plot1 | plot2</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wz9EeI44eLtic8PibMsOMaHJIHDklcsBJ4sRbY6180sHtfibiaHy2WbzCNhriazdwYDe7GCnJfHSBxfxhQ/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.4775967413441955" data-type="png" data-w="982" data-imgfileid="100056642" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wz9EeI44eLtic8PibMsOMaHJIHDklcsBJ4sRbY6180sHtfibiaHy2WbzCNhriazdwYDe7GCnJfHSBxfxhQ/640?wx_fmt=png&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">对peaks进行注释，与基因关联：</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># 注释</span></span><span leaf=""><br></span><span leaf="">open_cd4naive &lt;- rownames(da_peaks[da_peaks</span><span><span leaf="">$avg_log2FC</span></span><span leaf=""> &gt; 3, ])</span><span leaf=""><br></span><span leaf="">open_cd14mono &lt;- rownames(da_peaks[da_peaks</span><span><span leaf="">$avg_log2FC</span></span><span leaf=""> &lt; -3, ])</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">closest_genes_cd4naive &lt;- ClosestFeature(pbmc, regions = open_cd4naive)</span><span leaf=""><br></span><span leaf="">closest_genes_cd14mono &lt;- ClosestFeature(pbmc, regions = open_cd14mono)</span><span leaf=""><br></span><span leaf="">head(closest_genes_cd4naive)</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wz9EeI44eLtic8PibMsOMaHJIj3aB7BiaDia69ShmIsQdrOxsOFqbyqApibVs2DWszuPia0A3QWJQ5GvlcQ/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.11049210770659239" data-type="png" data-w="1077" data-imgfileid="100056641" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wz9EeI44eLtic8PibMsOMaHJIj3aB7BiaDia69ShmIsQdrOxsOFqbyqApibVs2DWszuPia0A3QWJQ5GvlcQ/640?wx_fmt=png&amp;from=appmsg"></span></figure><h2 data-tool="mdnice编辑器"><span></span><span><span leaf="">8、绘制 genomic regions</span></span><span></span><span><span leaf=""> </span></span></h2><p data-tool="mdnice编辑器"><span leaf="">我们可以使用</span><code><span leaf="">CoveragePlot()</span></code><span leaf="">函数，针对pbmc中存储的任何基因组区域，按cluster、细胞类型或其他 metadata 对细胞进行分组，绘制Tn5整合在整个基因组区域的频率。这些表示 pseudo-bulk可及性轨迹，其中来自一个组内所有细胞的信号被平均在一起，以可视化一个区域内的DNA可及性（感谢Andrew Hill在他的优秀博客文章中为这个函数提供了灵感）。除了这些可及性轨迹外，我们还可以可视化其他重要信息，包括基因注释、峰值坐标和基因组链接（如果它们存储在对象中）。更多信息请参阅可视化教程：https://stuartlab.org/signac/articles/visualization。</span></p><p data-tool="mdnice编辑器"><span leaf="">为了绘图的需要，将相关细胞类型放在一起会更好。我们可以通过运行</span><code><span leaf="">SortIdents()</span></code><span leaf="">函数，根据注释细胞类型之间的相似性自动对绘图顺序进行排序：</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># 可视化</span></span><span leaf=""><br></span><span leaf="">pbmc &lt;- SortIdents(pbmc)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># find DA peaks overlapping gene of interest</span></span><span leaf=""><br></span><span leaf="">regions_highlight &lt;- subsetByOverlaps(StringToGRanges(open_cd4naive), LookupGeneCoords(pbmc, </span><span><span leaf="">"CD4"</span></span><span leaf="">))</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">CoveragePlot(</span><span leaf=""><br></span><span leaf="">  object = pbmc,</span><span leaf=""><br></span><span leaf="">  region = </span><span><span leaf="">"CD4"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">  region.highlight = regions_highlight,</span><span leaf=""><br></span><span leaf="">  extend.upstream = 1000,</span><span leaf=""><br></span><span leaf="">  extend.downstream = 1000</span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wz9EeI44eLtic8PibMsOMaHJIwZnBzDefzVfyyJe3OgFKWV9rFLnEOfYOFImrjMXTEDQxQqdDd6eLug/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.5583437892095358" data-type="png" data-w="797" data-imgfileid="100056650" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wz9EeI44eLtic8PibMsOMaHJIwZnBzDefzVfyyJe3OgFKWV9rFLnEOfYOFImrjMXTEDQxQqdDd6eLug/640?wx_fmt=png&amp;from=appmsg"></span></figure><h3 data-tool="mdnice编辑器"><span></span><span><span leaf="">文末友情宣传</span></span><span></span></h3><ul><li><section><a href="https://mp.weixin.qq.com/s?__biz=MzUzMTEwODk0Ng==&amp;mid=2247533159&amp;idx=2&amp;sn=a4e9d31ff11df0942ff2b6e7e468704e&amp;scene=21#wechat_redirect"><span leaf="">生信入门&amp;数据挖掘线上直播课4月班</span></a></section></li><li><section><a href="http://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247524148&amp;idx=1&amp;sn=7806da6feb41a36493c519c1cfc1d3ac&amp;chksm=9b4bdf8fac3c569960369602f1ef26639cb366b250f233b2297d1f059471c0458335bfc0b829&amp;scene=21#wechat_redirect"><span leaf="">时隔5年，我们的生信技能树VIP学徒继续招生啦</span></a></section></li><li><section><a href="https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247535760&amp;idx=2&amp;sn=1e02a2e982a046ecf6389231e6768d5b&amp;scene=21#wechat_redirect"><span leaf="">满足你生信分析计算需求的低价解决方案</span></a></section></li></ul></section><section><span leaf=""><br></span></section><p><mp-style-type data-value="3"></mp-style-type></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/W3lMsLN6x-iwwc_3c-pYnA",target="_blank" rel="noopener noreferrer">原文链接</a>
