---
title: "单细胞差异分析之pseudobulk的3种实现方法"
date: 2025-04-07T02:46:18Z
draft: ["false"]
tags: [
  "fetched",
  "生信技能树"
]
categories: ["Acdemic"]
---
单细胞差异分析之pseudobulk的3种实现方法 by 生信技能树
------
<div><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com"><blockquote data-tool="mdnice编辑器"><p>之前分享了：<a href="https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247517945&amp;idx=1&amp;sn=babf5948237875e5b8e8cf005d480b03&amp;scene=21#wechat_redirect" data-linktype="2">单细胞层面的表达量差异分析到底如何做</a>，提到了pseudobulks方法，因为找各个单细胞亚群特异性高表达量基因（FindAllMarkers函数）以及两个亚群针对性差异分析（FindMarkers函数）都不符合需求，所以才有pseudobulks的流行。之前我们在《单细胞天地》公众号分享过一个文献 ，解读在：https://cloud.tencent.com/developer/article/1901064</p></blockquote><ul data-tool="mdnice编辑器"><li><section>文章题目：Confronting false discoveries in single-cell differential expression</section></li><li><section>日期：2021-09-28</section></li><li><section>期刊：Nature Communications</section></li><li><section>链接：https://www.nature.com/articles/s41467-021-25960-2</section></li></ul><p data-tool="mdnice编辑器">里面提到的目前主流的单细胞差异分析方法都是Wilcoxon rank−sum test，但是它其实表现还不如pseudobulks 的方法。。。</p><p data-tool="mdnice编辑器">所以有必要从代码角度看看单细胞差异分析之pseudobulk的3种实现方法。</p><h3 data-tool="mdnice编辑器"><span></span>首先是rowSums方法<span></span></h3><p data-tool="mdnice编辑器">这个是非常容易理解的，我在之前分享了：<a href="https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247517945&amp;idx=1&amp;sn=babf5948237875e5b8e8cf005d480b03&amp;scene=21#wechat_redirect" data-linktype="2">单细胞层面的表达量差异分析到底如何做</a>，也是这样举例：</p><p data-tool="mdnice编辑器">前面的 compSce是一个seurat对象 ，它里面的comp是表型是两个分组，然后mice是表型是十几个小鼠。也就是说十几个小鼠各自的单细胞转录组样品是两分组，需要做差异分析。我实际上是创造了一个do.call(   cbind,lapply 的复杂语法，熟悉这些函数的小伙伴就容易理解。</p><pre data-tool="mdnice编辑器"><span></span><code>bs = split(colnames(compSce),compSce$mice)<br>names(bs)<br>ct = do.call(<br>  cbind,lapply(names(bs), <span>function</span>(x){ <br>    <span># x=names(bs)[[1]]</span><br>    kp =colnames(compSce) %<span>in</span>% bs[[x]]<br>    rowSums( as.matrix( compSce@assays$RNA@counts[, kp]  ))<br>  })<br>)<br><br>phe = unique(compSce@meta.data[,c(<span>'mice'</span>,<span>'comp'</span>)])<br>phe<br><br><span># 每次都要检测数据  </span><br>group_list = phe[match(names(bs),phe$mice),<span>'comp'</span>]<br>table(group_list)    <br>exprSet = ct<br>exprSet[<span>1</span>:<span>4</span>,<span>1</span>:<span>4</span>] <br>dim(exprSet) <br>exprSet=exprSet[apply(exprSet,<span>1</span>, <span>function</span>(x) sum(x&gt;<span>1</span>) &gt; <span>1</span>),]<br>dim(exprSet)  <br>table(group_list)<br>save(exprSet,group_list,file = <span>'input_for_deg.Rdata'</span>)<br><br>load(file = <span>'input_for_deg.Rdata'</span>)<br><br></code></pre><p data-tool="mdnice编辑器">有了表达量矩阵以及分组信息，后面就是常规的转录组差异分析啦。</p><p data-tool="mdnice编辑器">其实 https://jef.works/blog/2020/04/06/quickly-creating-pseudobulks/ 也是提出来了类似的代码实现，居然跟我说一模一样的！！！</p><pre data-tool="mdnice编辑器"><span></span><code><span>## pseudobulk gene expression per cell-type</span><br>getPseudobulk &lt;- <span>function</span>(mat, celltype) {<br>   mat.summary &lt;- do.call(cbind, lapply(levels(celltype), <span>function</span>(ct) {<br>     cells &lt;- names(celltype)[celltype==ct]<br>     pseudobulk &lt;- rowSums(mat[, cells])<br>     <span>return</span>(pseudobulk)<br>   }))<br>   colnames(mat.summary) &lt;- levels(celltype)<br>   <span>return</span>(mat.summary)<br>}<br><br><span>## test runtime</span><br>start_time1 &lt;- Sys.time()<br><span>## call function</span><br>mat.summary &lt;- getPseudobulk(mat.sparse, celltype)<br>end_time1 &lt;- Sys.time()<br><br><span>## take a look</span><br>dim(mat.summary)<br></code></pre><p data-tool="mdnice编辑器">然后 https://github.com/neurorestore/DE-analysis/blob/master/R/functions/run_DE.R 就是使用了另外一套语法体系：</p><pre data-tool="mdnice编辑器"><span></span><code><span># process data into gene X replicate X cell_type matrices  </span><br>      mm = model.matrix(~ <span>0</span> + replicate:label, data = meta_sub)<br>      mat_mm = GetAssayData(sc_sub, slot = <span>'counts'</span>) %*% mm<br>      keep_genes = rowSums(mat_mm &gt; <span>0</span>) &gt; <span>0</span><br>      mat_mm = mat_mm[keep_genes, ] %&gt;% as.data.frame()<br>      mat_mm %&lt;&gt;% as.data.frame()<br>      colnames(mat_mm) = gsub(<span>"replicate|label"</span>, <span>""</span>, colnames(mat_mm))<br>      <span># drop empty columns</span><br>      keep_samples = colSums(mat_mm) &gt; <span>0</span><br>      mat_mm %&lt;&gt;% extract(, keep_samples)<br>      <span>return</span>(mat_mm)<br>    }) %&gt;%<br>    setNames(keep)<br>  <span># drop NAs</span><br>  pseudobulks %&lt;&gt;% extract(!is.na(.))<br></code></pre><p data-tool="mdnice编辑器">这个代码实在是太复杂了，我仅仅是节选部分给大家看看，因为它考虑到的各种因素非常多，但是本质上还是表达量矩阵的提取和加和，是rowSums方法。。。</p><h3 data-tool="mdnice编辑器"><span></span>scater包的aggregateAcrossCells函数<span></span></h3><p data-tool="mdnice编辑器">我在文章看到了scater包的aggregateAcrossCells函数也可以做类似的：《Transcriptional changes in the mammary gland during lactation revealed by single cell sequencing of cells from human milk》，链接 https://www.nature.com/articles/s41467-021-27895-0</p><ul data-tool="mdnice编辑器"><li><section>DEGs were identified between subsetted groups by first generating pseudobulk samples using “aggregateAcrossCells” function in the <em>scater</em> package.</section></li><li><section><em>edgeR</em> version 3.34.3 was used to compute DEGs between groups by filtering and scaling sample library size using the “filterByExpr” and “calcNormFactors” functions.</section></li><li><section>Next the common, trended and tagwise negative binomial dispersions of the genes were calculated using “estimateDisp”.</section></li><li><section>Quasi-likelihood negative binomial generalized log-linear models were fitted using “glmQLFit” and “glmQLFTest”.</section></li><li><section>FDR corrections were applied to the resulting <em>p</em> values using the Benjamini–Hochberg method.</section></li><li><section>To visualize the DEGs, volcano plots were generated using the <em>EnhancedVolcano</em> package the FDR corrected <em>p</em> value cut off FDR &lt; 1 × 10−8.</section></li></ul><p data-tool="mdnice编辑器">我还没有测试scater包的aggregateAcrossCells函数，因为我比较喜欢自己的代码。</p><h2 data-tool="mdnice编辑器"><span></span>Matrix.utils包的aggregate.Matrix函数</h2><p data-tool="mdnice编辑器">这个是出自于  <span><strong> https://hbctraining.github.io/scRNA-seq/lessons/pseudobulk_DESeq2_scrnaseq.html</strong></span></p><p data-tool="mdnice编辑器">如下所示：</p><p><img data-galleryid="" data-ratio="0.7398148148148148" data-s="300,640" data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wyOxsrX37pGXZS1UcSAbVnnmQ7tbvDfKf6c427AL7GS2Aw3burG07TicpxjBvAR6YU21tp2W4oDdVA/640?wx_fmt=png" data-type="png" data-w="1080" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wyOxsrX37pGXZS1UcSAbVnnmQ7tbvDfKf6c427AL7GS2Aw3burG07TicpxjBvAR6YU21tp2W4oDdVA/640?wx_fmt=png"></p><figure data-tool="mdnice编辑器"><figcaption> </figcaption></figure></section><h4 data-tool="mdnice编辑器">文末友情宣传</h4><p data-tool="mdnice编辑器">强烈建议你推荐给身边的<strong>博士后以及年轻生物学PI</strong>，多一点数据认知，让他们的科研上一个台阶：</p><ul data-tool="mdnice编辑器"><li><section><a target="_blank" href="http://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247526014&amp;idx=1&amp;sn=44afb387fc49b89276386e5182db7bc9&amp;chksm=9b4b26c5ac3cafd35616b2fe9df7fea664e59d75e9970feb322a477beb222ac023f7daebb3dc&amp;scene=21#wechat_redirect" textvalue="生物信息学马拉松授课（买一得‍五）" linktype="text" imgurl="" imgdata="null" data-itemshowtype="0" tab="innerlink" data-linktype="2" hasload="1">生物信息学马拉松授课（买一得五）</a> ，你的生物信息学入门课</section></li><li><section><a target="_blank" href="http://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247524148&amp;idx=1&amp;sn=7806da6feb41a36493c519c1cfc1d3ac&amp;chksm=9b4bdf8fac3c569960369602f1ef26639cb366b250f233b2297d1f059471c0458335bfc0b829&amp;scene=21#wechat_redirect" textvalue="时隔5年，我们的生信技能树VIP学徒继续招生啦" linktype="text" imgurl="" imgdata="null" data-itemshowtype="0" tab="innerlink" data-linktype="2" hasload="1">时隔5年，我们的生信技能树VIP学徒继续招生啦</a><br></section></li><li><section><a target="_blank" href="http://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247522831&amp;idx=2&amp;sn=1744efdf428465425a145ff3a982198b&amp;chksm=9b4bdab4ac3c53a28fbecbbff4f254f470b54a7a20468bb753b295b930315e1ec45bcbabc10b&amp;scene=21#wechat_redirect" textvalue="144线程640Gb内存服务器共享一年‍仍然是仅需800" linktype="text" imgurl="" imgdata="null" data-itemshowtype="0" tab="innerlink" data-linktype="2" hasload="1">144线程640Gb内存服务器共享一年仍然是仅需800</a></section></li><li><section><a target="_blank" href="http://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247519765&amp;idx=1&amp;sn=ce5a8c8182f854c88043059f8c2cb9ff&amp;chksm=9b4bceaeac3c47b88c19941d43dbb1401f3a92206481a0afc41159927868199643f795d62a7e&amp;scene=21#wechat_redirect" textvalue="千呼万唤始出来的独享生物信息学云服务器" linktype="text" imgurl="" imgdata="null" data-itemshowtype="0" tab="innerlink" data-linktype="2" hasload="1">千呼万唤始出来的独享生物信息学云服务器</a></section></li><li><section><a target="_blank" href="http://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247519765&amp;idx=1&amp;sn=ce5a8c8182f854c88043059f8c2cb9ff&amp;chksm=9b4bceaeac3c47b88c19941d43dbb1401f3a92206481a0afc41159927868199643f795d62a7e&amp;scene=21#wechat_redirect" textvalue="千呼万唤始出来的独享生物信息学云服务器" linktype="text" imgurl="" imgdata="null" data-itemshowtype="0" tab="innerlink" data-linktype="2" hasload="1"></a><a target="_blank" href="http://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247524275&amp;idx=1&amp;sn=fa592ee29f636f34387491d0fceadd8e&amp;chksm=9b4bdf08ac3c561e0881974b3817beb0a0e514dc1a8df4c34c2b6653da6fa78e09acb03c70c2&amp;scene=21#wechat_redirect" textvalue="生信技能树知识整理实习生又又又开放申请啦" linktype="text" imgurl="" imgdata="null" data-itemshowtype="0" tab="innerlink" data-linktype="2" hasload="1">生信技能树知识整理实习生又又又开放申请啦</a><span><strong>（不招了，谢谢）</strong></span></section></li><li><p><a target="_blank" href="http://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247524432&amp;idx=1&amp;sn=5b33b0c6807a9e6939c332c58fabff89&amp;chksm=9b4b20ebac3ca9fdb3d8bfaf2bef5552f64eb70e7fae557cc7197fb1a23b3e8bc31b585bf829&amp;scene=21#wechat_redirect" textvalue="生信共享办公室出租" linktype="text" imgurl="" imgdata="null" data-itemshowtype="0" tab="innerlink" data-linktype="2" hasload="1">生信共享办公室出租</a></p></li></ul><p><br></p><p><mp-style-type data-value="3"></mp-style-type></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/jR2OdJQPfBAfxSLSYPyqUw",target="_blank" rel="noopener noreferrer">原文链接</a>
