---
title: "单细胞实战之inferCNV——入门到进阶(中级篇3）"
date: 2025-04-07T07:17:53Z
draft: ["false"]
tags: [
  "fetched",
  "单细胞天地"
]
categories: ["Acdemic"]
---
单细胞实战之inferCNV——入门到进阶(中级篇3） by 单细胞天地
------
<div><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com" data-pm-slice="0 0 []"><p data-tool="mdnice编辑器"><span leaf="">在前两讲中，我们已经完整介绍了CD4+ T细胞的提取及其亚群细分的流程。本讲将聚焦于单细胞分析领域中的重要工具——inferCNV，用于判断不同细胞亚群的CNV变化。尽管本讲内容在课程的连贯性上可能略显跳跃，但考虑到该工具在肿瘤研究中的重要性和以及其技术难度相对较低，我们仍在中级篇阶段对其进行回顾和学习。</span></p><p data-tool="mdnice编辑器"><span leaf="">本次内容涉及到的工程文件可通过网盘获得：初级篇2 链接: https://pan.baidu.com/s/1ETVEkJNCJSe9NU7_aLEPVA 提取码: znmb 。</span></p><p data-tool="mdnice编辑器"><span leaf="">上一讲CD4+T细胞细分后结果：中级篇2，链接: https://pan.baidu.com/s/1y-HHLXoXsJbgWKCdz26-gQ 提取码: yx93 。</span></p><p data-tool="mdnice编辑器"><span leaf="">此外，可以向“生信技能树”公众号发送关键词‘单细胞’，直接获取Seurat V5版本的完整代码。</span></p><p data-tool="mdnice编辑器"><strong><span leaf="">既往推文</span></strong><span leaf="">：</span></p><ol><li><section><span leaf=""><a href="https://mp.weixin.qq.com/s?__biz=MzI1Njk4ODE0MQ==&amp;mid=2247528271&amp;idx=1&amp;sn=333686533ec8b7cadf913a3868c79dd2&amp;scene=21#wechat_redirect" textvalue="单细胞实战之亚细胞分群从T/NK至CD4+T细胞 (中级篇1）" data-itemshowtype="0" target="_blank" linktype="text" data-linktype="2">单细胞实战之亚细胞分群从T/NK至CD4+T细胞 (中级篇1）</a></span></section></li><li><section><span leaf=""><a href="https://mp.weixin.qq.com/s?__biz=MzI1Njk4ODE0MQ==&amp;mid=2247528893&amp;idx=1&amp;sn=7483ee64d0c07fe684eccb86d2fef9a9&amp;scene=21#wechat_redirect" textvalue="单细胞实战之亚细胞分群之CD4+T细胞分群——从入门到进阶(中级篇2）" data-itemshowtype="0" target="_blank" linktype="text" data-linktype="2">单细胞实战之亚细胞分群之CD4+T细胞分群——从入门到进阶(中级篇2）</a></span></section></li></ol><p data-tool="mdnice编辑器"><strong><span leaf="">inferCNV分析简介</span></strong></p><p data-tool="mdnice编辑器"><span leaf="">InferCNV用于分析肿瘤单细胞RNA测序数据，通过比较肿瘤细胞与参考"正常"细胞的基因表达强度，识别体细胞大规模染色体拷贝数变异的证据（如整条染色体或大片段染色体的扩增或缺失）。该方法可生成展示肿瘤基因组各染色体区域相对表达强度的热图，相较于正常细胞，肿瘤基因组中过度富集或缺失的区域通常会直观显现。</span></p><p data-tool="mdnice编辑器"><strong><span leaf="">inferCNV分析的意义</span></strong></p><p data-tool="mdnice编辑器"><span leaf="">对单细胞分析的时候会根据关键基因/标记进行细胞分群，基本上可以把大部分的细胞亚群给区分开来，但如果某些基因/标记广泛地存在于多种细胞亚群时，此时应进一步寻找其他的关键基因/标记进行区分，这种分析策略是“常规流程”。对于肿瘤细胞和正常细胞而言，我们常会遇到某些关键基因/标记在两者中均表达的情况(转录组水平)，此时通过“常规流程”仍难以区分细胞的良恶性，因此换个角度来辅助区分良恶性细胞就显得尤为重要。</span></p><p data-tool="mdnice编辑器"><span leaf="">肿瘤细胞通常具有更多的CNV，而正常细胞则较为稳定，因此如果有一种工具能够可视化拷贝数变异的情况那么是不是就可以辅助鉴别良恶性细胞呢？基于这种情况inferCNV就应运而生了。</span></p><h4 data-tool="mdnice编辑器"><span></span><span leaf="">inferCNV分析结果</span><span></span></h4><h5 data-tool="mdnice编辑器"><span></span><span leaf="">1、官方示例展示</span><span></span></h5><p data-tool="mdnice编辑器"><span leaf="">图中的红色部分是References(Cells)，也就是我们自行定义的对照(正常)细胞。这些对照细胞可以是相对于肿瘤细胞的正常细胞(癌与非癌)，也可选用免疫细胞进行对照。蓝色部分是Observations(Cells),也就是我们想要重点分析的细胞。灰色部分是细胞的图注，上边的色块代表了对照细胞的情况，该图研究者纳入了Microglia/Macrophage 和 Oligodendrocytes (non-malignant) 作为对照细胞，恶性细胞作为观察细胞。<img data-imgfileid="100045316" data-ratio="0.7786096256684492" data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTiatwSmJZ1UmXCMjPgtOeTTC6FqajeISpicbyZbpPHVtV3n6k3SEPOteH35nzOM8yf5QeMsyYYCiaUw/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="935" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTiatwSmJZ1UmXCMjPgtOeTTC6FqajeISpicbyZbpPHVtV3n6k3SEPOteH35nzOM8yf5QeMsyYYCiaUw/640?wx_fmt=png&amp;from=appmsg">红色方框部分代表的是分层聚类树，其中第一列色柱代表的是所有观察组细胞的分层聚类，第二列色柱则是与所提供输入的观察组细胞分类相对应。<img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTiatwSmJZ1UmXCMjPgtOeTTbTyxR88jK1ibzS4IV03TMO1tVbY5nKGW2BP6ic2Wx1e8RsxLnpF3pJbw/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.6961748633879782" data-type="png" data-w="915" data-imgfileid="100045317" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTiatwSmJZ1UmXCMjPgtOeTTbTyxR88jK1ibzS4IV03TMO1tVbY5nKGW2BP6ic2Wx1e8RsxLnpF3pJbw/640?wx_fmt=png&amp;from=appmsg">按照箭头看向对照和观察组细胞，在红色箭头处代表了malignant_MGH36细胞群，这群细胞相比于观察组(蓝色箭头)在chr1, 4, 19中存在更显著的染色体缺失，在chr11, 21中存在更显著的染色体扩增。同样的我们可以看其他的三群恶性细胞，也均存在不同区域的染色体拷贝数异常。<img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTiatwSmJZ1UmXCMjPgtOeTTmwPh503Z8zk3giceDHicvqEsbDhRkjrnIyyxPYJXVS4YfHh9TIbeGp8g/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.7935828877005348" data-type="png" data-w="935" data-imgfileid="100045315" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTiatwSmJZ1UmXCMjPgtOeTTmwPh503Z8zk3giceDHicvqEsbDhRkjrnIyyxPYJXVS4YfHh9TIbeGp8g/640?wx_fmt=png&amp;from=appmsg"></span></p><h5 data-tool="mdnice编辑器"><span></span><span leaf="">2、其他数据展示—正常/肿瘤样本</span><span></span></h5><p data-tool="mdnice编辑器"><span leaf="">该结果显示1个正常样本作为对照组，10个肿瘤样本作为观察组。整体情况观察下来可以发现肿瘤组样本相比于对照组来说还是存在了很明显的染色体拷贝数异常。红色方框中的第二列色柱是代表了P10肿瘤样本的信息，中间还混了一小部分的P05患者的信息，从聚类树分层后的色柱来看，P10肿瘤样本还可以进一步的细分为多群。蓝色方框中的第二列色柱中存在很多肿瘤样本的信息，但从聚类树分层后的色柱来看，这些肿瘤样本信息应当分成同一群。<img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTiatwSmJZ1UmXCMjPgtOeTTHWZyuqejyDkyo79FcVbqwTp7m2vtfpgDCIN8cxvyHbbtB7e4TEicpNg/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.9138888888888889" data-type="png" data-w="1080" data-imgfileid="100045318" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTiatwSmJZ1UmXCMjPgtOeTTHWZyuqejyDkyo79FcVbqwTp7m2vtfpgDCIN8cxvyHbbtB7e4TEicpNg/640?wx_fmt=png&amp;from=appmsg">同样的数据，修改了cluster_by_groups的值之后聚类树的图形出现了变化。在不按照k_obs_groups分组之后，主要根据每个样本内部的情况进行分组，可以发现不同样本内部也存在着很大的异质性(红色方框)。<img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTiatwSmJZ1UmXCMjPgtOeTTCTYRyxLlBmPnvCaicupNNC8SXmspqJRX8t2Csqiax3Y5jF1j4MROe5SQ/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.9101851851851852" data-type="png" data-w="1080" data-imgfileid="100045319" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTiatwSmJZ1UmXCMjPgtOeTTCTYRyxLlBmPnvCaicupNNC8SXmspqJRX8t2Csqiax3Y5jF1j4MROe5SQ/640?wx_fmt=png&amp;from=appmsg"></span></p><h5 data-tool="mdnice编辑器"><span></span><span leaf="">3、其他数据展示—细胞亚群</span><span></span></h5><p data-tool="mdnice编辑器"><span leaf="">该结果显示前期分析得到了17个细胞亚群，其中第15个亚群根据关键基因/标记推测认为可能是正常细胞群，因此通过inferCNV进行辅助分析判断第15个亚群是否是正常细胞群，从结果来看确实也符合正常细胞亚群的猜测。而根据聚类树结果来看，k_obs_groups分组和细胞亚群分组一致。<img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTiatwSmJZ1UmXCMjPgtOeTTDXVKgQlHqgcqGC9kt4qCsOrxluBDibndukmfPkWibkUKua0ZXDDdQbow/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.962" data-type="png" data-w="1000" data-imgfileid="100045324" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTiatwSmJZ1UmXCMjPgtOeTTDXVKgQlHqgcqGC9kt4qCsOrxluBDibndukmfPkWibkUKua0ZXDDdQbow/640?wx_fmt=png&amp;from=appmsg">不按照k_obs_groups分组之后，我们可以明显发现很多细胞亚群之间没有十分明显的界限，也就是说这个结果提示我们在后续聚类的时候可以把很多细胞亚群合并成同一群。<img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTiatwSmJZ1UmXCMjPgtOeTTFicOhwlLehZ8HKDibnzDu6kh8vJFVEeQQFRiaAJTt9KpXThjYsrCAB6PQ/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.9617537313432836" data-type="png" data-w="1072" data-imgfileid="100045323" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTiatwSmJZ1UmXCMjPgtOeTTFicOhwlLehZ8HKDibnzDu6kh8vJFVEeQQFRiaAJTt9KpXThjYsrCAB6PQ/640?wx_fmt=png&amp;from=appmsg"></span></p><h4 data-tool="mdnice编辑器"><span></span><span leaf="">inferCNV分析流程</span><span></span></h4><h5 data-tool="mdnice编辑器"><span></span><span leaf="">1.加载对象和R包</span><span></span></h5><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">rm(list = ls())</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(infercnv)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(Seurat)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(gplots)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(ggplot2)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(qs)</span><span leaf=""><br></span><span leaf="">sce &lt;- qread(</span><span><span leaf="">"./4-Corrective-data/sce.all.qs"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf="">#检查一下自己导入进来的数据</span></span><span leaf=""><br></span><span leaf="">DimPlot(sce,reduction = </span><span><span leaf="">'umap'</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">        label = </span><span><span leaf="">TRUE</span></span><span leaf="">,pt.size = </span><span><span leaf="">0.5</span></span><span leaf="">) +NoLegend()</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">dir.create(</span><span><span leaf="">"11-infercnv"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">setwd(</span><span><span leaf="">"./11-infercnv"</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><h5 data-tool="mdnice编辑器"><span></span><span leaf="">2.数据预处理</span><span></span></h5><p data-tool="mdnice编辑器"><span leaf="">创建InferCNV对象前需要准备好三个文件：conut matrix；cell type annotations；gene ordering file</span></p><ol><li><section><span leaf="">conut matrix文件是行为基因，列为细胞的文件（图片来自示例数据）<img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTiatwSmJZ1UmXCMjPgtOeTTJM94Ng1IicxGdjjDr4LbQxlGnnFOscK7gxOkbeu05R0ag7uoX97ZC2g/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.32685185185185184" data-type="png" data-w="1080" data-imgfileid="100045321" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTiatwSmJZ1UmXCMjPgtOeTTJM94Ng1IicxGdjjDr4LbQxlGnnFOscK7gxOkbeu05R0ag7uoX97ZC2g/640?wx_fmt=png&amp;from=appmsg"></span></section></li><li><section><span leaf="">cell type annotations文件分为两列，左边为细胞名称，右边为已知的细胞类型，比如：细胞的生物学定义/数字编号的细胞簇名/等等（图片来自示例数据）<img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTiatwSmJZ1UmXCMjPgtOeTTmic6xibDC1YlJ9LNWA4JzPPaGJTx4h5aaA7WyIf3T3k3smWNCfnUadew/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.19722222222222222" data-type="png" data-w="1080" data-imgfileid="100045320" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTiatwSmJZ1UmXCMjPgtOeTTmic6xibDC1YlJ9LNWA4JzPPaGJTx4h5aaA7WyIf3T3k3smWNCfnUadew/640?wx_fmt=png&amp;from=appmsg"></span></section></li><li><section><span leaf="">gene ordering file文件提供了每个基因的染色体位置，格式为制表符分隔，没有列标题，提供基因名称、染色体和基因跨度（图片来自示例数据）<img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTiatwSmJZ1UmXCMjPgtOeTTHWb2GrZG9QpZNxS2f5uPG5AkticdJK8Rk88oM1licPSkPVUVQ2j4uBJw/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.17962962962962964" data-type="png" data-w="1080" data-imgfileid="100045322" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTiatwSmJZ1UmXCMjPgtOeTTHWb2GrZG9QpZNxS2f5uPG5AkticdJK8Rk88oM1licPSkPVUVQ2j4uBJw/640?wx_fmt=png&amp;from=appmsg">官方提供了基因排序文件：https://data.broadinstitute.org/Trinity/CTAT/cnv/，当然自行制作也十分迅速。<img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTiatwSmJZ1UmXCMjPgtOeTTumqUJMSD4ibf78NfYdWvTia9ia0fgby2e3XsMnRQ7jym2sLG2krg85l4w/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.40185185185185185" data-type="png" data-w="1080" data-imgfileid="100045327" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTiatwSmJZ1UmXCMjPgtOeTTumqUJMSD4ibf78NfYdWvTia9ia0fgby2e3XsMnRQ7jym2sLG2krg85l4w/640?wx_fmt=png&amp;from=appmsg"></span></section></li></ol><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">table(Idents(sce))</span><span leaf=""><br></span><span leaf="">table(sce@meta.data$seurat_clusters)</span><span leaf=""><br></span><span leaf="">table(sce@meta.data$orig.ident)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 提取部分细胞</span></span><span leaf=""><br></span><span leaf="">sce@meta.data</span><span leaf=""><br></span><span leaf="">sce &lt;- sce[,sce$celltype %</span><span><span leaf="">in</span></span><span leaf="">% c(</span><span><span leaf="">"plasma"</span></span><span leaf="">,</span><span><span leaf="">"fibroblasts"</span></span><span leaf="">,</span><span><span leaf="">"epithelial/cancer cells"</span></span><span leaf="">)]</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 文件制作1：表达量矩阵</span></span><span leaf=""><br></span><span><span leaf=""># 除了用GetAssayData函数其实也可以直接sce@assays$RNA@count即可</span></span><span leaf=""><br></span><span leaf="">dat &lt;- GetAssayData(sce,layer = </span><span><span leaf="">'counts'</span></span><span leaf="">,assay = </span><span><span leaf="">'RNA'</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">dat[</span><span><span leaf="">1</span></span><span leaf="">:</span><span><span leaf="">4</span></span><span leaf="">,</span><span><span leaf="">1</span></span><span leaf="">:</span><span><span leaf="">4</span></span><span leaf="">]</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 文件制作2：样本的描述</span></span><span leaf=""><br></span><span leaf="">groupinfo &lt;- data.frame(v1 = colnames(dat),v2 = sce@meta.data$celltype)</span><span leaf=""><br></span><span leaf="">head(groupinfo)</span><span leaf=""><br></span><span><span leaf="">#                                  v1                      v2</span></span><span leaf=""><br></span><span><span leaf=""># 1 GSM5688706_WGC_AACACACGTGCTTCAA-1 epithelial/cancer cells</span></span><span leaf=""><br></span><span><span leaf=""># 2 GSM5688706_WGC_AACACACTCCCGTTCA-1 epithelial/cancer cells</span></span><span leaf=""><br></span><span><span leaf=""># 3 GSM5688706_WGC_AACACACTCTGCTAGA-1                  plasma</span></span><span leaf=""><br></span><span><span leaf=""># 4 GSM5688706_WGC_AACCATGAGGCACAAC-1                  plasma</span></span><span leaf=""><br></span><span><span leaf=""># 5 GSM5688706_WGC_AACCATGCAAAGCTCT-1 epithelial/cancer cells</span></span><span leaf=""><br></span><span><span leaf=""># 6 GSM5688706_WGC_AACCCAAGTGCCGTAC-1                  plasma</span></span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 文件制作3：基因在染色体中的坐标</span></span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(AnnoProbe)</span><span leaf=""><br></span><span><span leaf="">#annoGene 函数返回一个数据框，包含输入基因的详细注释信息。</span></span><span leaf=""><br></span><span><span leaf="">#注释信息可能包括基因名称、染色体位置、基因描述等。</span></span><span leaf=""><br></span><span leaf="">geneInfor &lt;- annoGene(rownames(dat),</span><span><span leaf="">"SYMBOL"</span></span><span leaf="">,</span><span><span leaf="">"human"</span></span><span leaf="">) </span><span><span leaf="">#物种需要小心哦</span></span><span leaf=""><br></span><span><span leaf=""># 使用逻辑条件来移除包含 chrM, chrX, 和 chrY 的行</span></span><span leaf=""><br></span><span leaf="">geneInfor &lt;- geneInfor[!geneInfor$chr %</span><span><span leaf="">in</span></span><span leaf="">% c(</span><span><span leaf="">"chrM"</span></span><span leaf="">, </span><span><span leaf="">"chrX"</span></span><span leaf="">, </span><span><span leaf="">"chrY"</span></span><span leaf="">), ]</span><span leaf=""><br></span><span><span leaf="">#提取chr后后面的数字并转化为num，从而按这个num排序</span></span><span leaf=""><br></span><span><span leaf="">#sub函数用于把chr替换为空</span></span><span leaf=""><br></span><span leaf="">geneInfor$chr_num &lt;- as.numeric(sub(</span><span><span leaf="">"chr"</span></span><span leaf="">, </span><span><span leaf="">""</span></span><span leaf="">, geneInfor$chr))</span><span leaf=""><br></span><span leaf="">colnames(geneInfor)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf="">#with函数的作用是简化写法，可问gpt</span></span><span leaf=""><br></span><span leaf="">geneInfor &lt;- geneInfor[with(geneInfor,order(chr_num,start)),c(</span><span><span leaf="">1</span></span><span leaf="">,</span><span><span leaf="">4</span></span><span leaf="">:</span><span><span leaf="">6</span></span><span leaf="">)]</span><span leaf=""><br></span><span leaf="">geneInfor &lt;- geneInfor[!duplicated(geneInfor[,</span><span><span leaf="">1</span></span><span leaf="">]),]</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">length(unique(geneInfor[,</span><span><span leaf="">1</span></span><span leaf="">]))</span><span leaf=""><br></span><span leaf="">head(geneInfor)</span><span leaf=""><br></span><span><span leaf="">#        SYMBOL  chr  start    end</span></span><span leaf=""><br></span><span><span leaf=""># 23 AP006222.2 chr1 266855 268655</span></span><span leaf=""><br></span><span><span leaf=""># 51     FAM87B chr1 817371 819837</span></span><span leaf=""><br></span><span><span leaf=""># 53  LINC00115 chr1 826206 827522</span></span><span leaf=""><br></span><span><span leaf=""># 55     FAM41C chr1 868071 876903</span></span><span leaf=""><br></span><span><span leaf=""># 62     SAMD11 chr1 923928 944581</span></span><span leaf=""><br></span><span><span leaf=""># 63      NOC2L chr1 944203 959309</span></span><span leaf=""><br></span></code></pre><h5 data-tool="mdnice编辑器"><span></span><span leaf="">3.排序一下</span><span></span></h5><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">#保留行名在geneInfor第一列中存在的行。</span></span><span leaf=""><br></span><span leaf="">dat &lt;- dat[rownames(dat) %</span><span><span leaf="">in</span></span><span leaf="">% geneInfor[,</span><span><span leaf="">1</span></span><span leaf="">],] </span><span leaf=""><br></span><span><span leaf="">#match(x, table)：match函数返回x中每个元素在table中的位置索引。</span></span><span leaf=""><br></span><span><span leaf="">#获得位置后对dat进行重新排序使其跟geneInfor中的顺序一致</span></span><span leaf=""><br></span><span leaf="">dat &lt;- dat[match(geneInfor[,</span><span><span leaf="">1</span></span><span leaf="">],rownames(dat)),]</span><span leaf=""><br></span><span leaf="">dim(dat)</span><span leaf=""><br></span><span leaf="">head(groupinfo)</span><span leaf=""><br></span><span leaf="">dat[</span><span><span leaf="">1</span></span><span leaf="">:</span><span><span leaf="">4</span></span><span leaf="">,</span><span><span leaf="">1</span></span><span leaf="">:</span><span><span leaf="">4</span></span><span leaf="">]</span><span leaf=""><br></span><span leaf="">table(groupinfo$v2)</span><span leaf=""><br></span><span leaf="">dim(groupinfo)</span><span leaf=""><br></span></code></pre><h5 data-tool="mdnice编辑器"><span></span><span leaf="">4.保存/输出文件</span><span></span></h5><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">#为了节约计算机资源，直接抽样</span></span><span leaf=""><br></span><span><span leaf="">#如果真实项目请不要用抽样</span></span><span leaf=""><br></span><span><span leaf="">#kp &lt;- sample(1:nrow(groupinfo),500)</span></span><span leaf=""><br></span><span><span leaf="">#groupinfo &lt;- groupinfo[kp,]</span></span><span leaf=""><br></span><span><span leaf="">#dat &lt;- dat[,kp]</span></span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 保存矩阵文件</span></span><span leaf=""><br></span><span><span leaf="">#expFile:是一个变量，存储写入的文件的文件名或路径。在这里文件名是expFile.txt。</span></span><span leaf=""><br></span><span leaf="">expFile &lt;- </span><span><span leaf="">'expFile.txt'</span></span><span><span leaf="">#定义输出文件名</span></span><span leaf=""><br></span><span leaf="">colnames(dat) &lt;- gsub(</span><span><span leaf="">"-"</span></span><span leaf="">, </span><span><span leaf="">"_"</span></span><span leaf="">, colnames(dat))</span><span leaf=""><br></span><span leaf="">write.table(dat, file = expFile, sep = </span><span><span leaf="">'\t'</span></span><span leaf="">, quote = </span><span><span leaf="">F</span></span><span leaf="">)</span><span leaf=""><br></span><span><span leaf=""># 保存细胞注释文件</span></span><span leaf=""><br></span><span leaf="">groupFiles &lt;- </span><span><span leaf="">'groupFiles.txt'</span></span><span leaf=""><br></span><span leaf="">groupinfo$v1 &lt;- gsub(</span><span><span leaf="">"-"</span></span><span leaf="">, </span><span><span leaf="">"_"</span></span><span leaf="">, groupinfo$v1)</span><span leaf=""><br></span><span leaf="">write.table(groupinfo,file = groupFiles, sep = </span><span><span leaf="">'\t'</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">            quote = </span><span><span leaf="">F</span></span><span leaf="">, col.names = </span><span><span leaf="">F</span></span><span leaf="">, row.names = </span><span><span leaf="">F</span></span><span leaf="">)</span><span leaf=""><br></span><span><span leaf=""># 保存基因排序文件</span></span><span leaf=""><br></span><span leaf="">head(geneInfor)</span><span leaf=""><br></span><span leaf="">geneFile &lt;- </span><span><span leaf="">'geneFile.txt'</span></span><span leaf=""><br></span><span leaf="">write.table(geneInfor, file = geneFile, sep = </span><span><span leaf="">'\t'</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">            quote = </span><span><span leaf="">F</span></span><span leaf="">, col.names = </span><span><span leaf="">F</span></span><span leaf="">, row.names = </span><span><span leaf="">F</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><h5 data-tool="mdnice编辑器"><span></span><span leaf="">5.正式跑inferCNV</span><span></span></h5><p data-tool="mdnice编辑器"><span leaf="">创建infercnv_obj对象：</span></p><ol><li><section><span leaf="">ref_group_names选择已知的非肿瘤细胞即可，我们这里选择了plasma细胞；</span></section></li><li><section><span leaf="">如果没有参考细胞，可以将 ref_group_names 设置为 NULL，在这种情况下，将使用所有细胞的平均信号来定义基线。</span></section></li></ol><p data-tool="mdnice编辑器"><span leaf="">正式运行infercnv::run：</span></p><ol><li><section><span leaf="">cutoff截断值：截断值决定了决定了哪些基因将被纳入inferCNV分析，低于该阈值的基因（即细胞间平均表达计数不足的基因）将被排除。针对全长转录本测序技术（如基于细胞板实验的smart-seq平台），通常采用阈值1以适配其高检测灵敏度；而对于10x Genomics（及其他基于3’端测序的液滴检测技术，其基因计数矩阵数据稀疏性较高），经验表明阈值0.1普遍适用。</span></section></li><li><section><span leaf="">out_dir：指定输出目录的名称。</span></section></li><li><section><span leaf="">cluster_by_groups：参数设置为根据细胞注释文件中定义的患者类型对肿瘤细胞进行独立分群分析。</span></section></li><li><section><span leaf="">denoise：需要设定为True。同时还可以增加使用noise_filter参数设定具体噪音阈值（比如0.1）；或者增加使用sd_amplifier设定具体噪音阈值的标准差（比如1.5）；或者在固定sd_amplifier值的前提下，增加参数noise_logistic=T通过sigmoidal函数调整去噪强度。（但其实这些都是花里胡哨的，大部分使用者的目的只要能有判断CNV是否存在差别即可）</span></section></li><li><section><span leaf="">第一列色柱根据分析时cluster_by_groups参数设置的不同(True/False)会对结果产生差异，可以自行调试</span></section></li></ol><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">#创建对象，请注意文件需要一一对应哦！</span></span><span leaf=""><br></span><span><span leaf="">#ref_group_names 这里的细胞是正常对照，然后跟其他的细胞比较</span></span><span leaf=""><br></span><span leaf="">infercnv_obj &lt;- CreateInfercnvObject(raw_counts_matrix = expFile,</span><span leaf=""><br></span><span leaf="">                                     annotations_file = groupFiles,</span><span leaf=""><br></span><span leaf="">                                     delim = </span><span><span leaf="">"\t"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">                                     gene_order_file = geneFile,</span><span leaf=""><br></span><span leaf="">                                     ref_group_names = c(</span><span><span leaf="">"plasma"</span></span><span leaf="">,</span><span><span leaf="">"fibroblasts"</span></span><span leaf="">) </span><span><span leaf="">#这里用plasma和fibroblasts做参照</span></span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">infercnv_obj &lt;- infercnv::run(infercnv_obj,</span><span leaf=""><br></span><span leaf="">                               cutoff =  </span><span><span leaf="">0.1</span></span><span leaf="">, </span><span><span leaf="">#smart-seq选择1,10X选择0.1</span></span><span leaf=""><br></span><span leaf="">                               out_dir = </span><span><span leaf="">"infercnv_output"</span></span><span leaf="">, </span><span><span leaf=""># dir is auto</span></span><span leaf=""><br></span><span leaf="">                               </span><span><span leaf=""># 是否根据细胞注释文件的分组</span></span><span leaf=""><br></span><span leaf="">                               </span><span><span leaf=""># 对肿瘤细胞进行分组</span></span><span leaf=""><br></span><span leaf="">                               </span><span><span leaf=""># 影响read.dendrogram, 如果有多个细胞类型，且设置为TRUE，</span></span><span leaf=""><br></span><span leaf="">                               </span><span><span leaf=""># 后续的read.dendrogram无法执行</span></span><span leaf=""><br></span><span leaf="">                               cluster_by_groups =  </span><span><span leaf="">TRUE</span></span><span leaf="">, </span><span><span leaf="">#是否根据细胞类型(由细胞注释文件中定义)</span></span><span leaf=""><br></span><span leaf="">                               hclust_method = </span><span><span leaf="">"ward.D2"</span></span><span leaf="">,</span><span><span leaf=""># ward.D2 方法进行层次聚类</span></span><span leaf=""><br></span><span leaf="">                               analysis_mode = </span><span><span leaf="">"subclusters"</span></span><span leaf="">, </span><span><span leaf=""># 默认是samples，推荐是subclusters</span></span><span leaf=""><br></span><span leaf="">                               denoise = </span><span><span leaf="">TRUE</span></span><span leaf="">, </span><span><span leaf=""># 去噪音</span></span><span leaf=""><br></span><span leaf="">                               HMM = </span><span><span leaf="">F</span></span><span leaf="">,  </span><span><span leaf="">##特别耗时间,是否要去背景噪音</span></span><span leaf=""><br></span><span leaf="">                               plot_steps = </span><span><span leaf="">F</span></span><span leaf="">, </span><span><span leaf="">#不在每个步骤后生成图形。</span></span><span leaf=""><br></span><span leaf="">                               leiden_resolution = </span><span><span leaf="">"auto"</span></span><span leaf="">, </span><span><span leaf="">#可以手动调参数</span></span><span leaf=""><br></span><span leaf="">                               num_threads = </span><span><span leaf="">16</span></span><span><span leaf=""># 16线程工作，加快速度，一般电脑不建议哈</span></span><span leaf=""><br></span><span leaf="">                               )</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">去噪音前后图。去之前也看得出肿瘤细胞相比于其他两种细胞来说有更多CNV改变<img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTiatwSmJZ1UmXCMjPgtOeTTHzXzGnpQwUibSzSSelhaOxOcRZSGJ6y0jNM761iciaaEPRThXz8yqicnNA/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.45092592592592595" data-type="png" data-w="1080" data-imgfileid="100045328" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTiatwSmJZ1UmXCMjPgtOeTTHzXzGnpQwUibSzSSelhaOxOcRZSGJ6y0jNM761iciaaEPRThXz8yqicnNA/640?wx_fmt=png&amp;from=appmsg">增加中值滤波，让视觉效果更明显</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">infercnv_obj_medianfiltered = infercnv::apply_median_filtering(infercnv_obj)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">infercnv::plot_cnv(infercnv_obj_medianfiltered, </span><span leaf=""><br></span><span leaf="">                   out_dir=</span><span><span leaf="">'./example_output/'</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">                   output_filename=</span><span><span leaf="">'infercnv.median_filtered'</span></span><span leaf="">, </span><span leaf=""><br></span><span leaf="">                   x.range=</span><span><span leaf="">"auto"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">                   x.center=</span><span><span leaf="">1</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">                   title = </span><span><span leaf="">"infercnv"</span></span><span leaf="">, </span><span leaf=""><br></span><span leaf="">                   color_safe_pal = </span><span><span leaf="">FALSE</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">                   </span><span leaf=""><br></span><span leaf="">setwd(</span><span><span leaf="">".."</span></span><span leaf="">)                 </span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTiatwSmJZ1UmXCMjPgtOeTTLzOAIfpApib32qzmtSCXzKLlWgwmh3klXwKqPyT94BCkW7xCQtDaZag/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.9444444444444444" data-type="png" data-w="1080" data-imgfileid="100045329" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTiatwSmJZ1UmXCMjPgtOeTTLzOAIfpApib32qzmtSCXzKLlWgwmh3klXwKqPyT94BCkW7xCQtDaZag/640?wx_fmt=png&amp;from=appmsg"></span></figure><h5 data-tool="mdnice编辑器"><span></span><span leaf="">6、读取数据加载R包</span><span></span></h5><p data-tool="mdnice编辑器"><span leaf="">接下来使用曾老师的方法去检查拷贝数变异的情况</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">rm(list = ls())</span><span leaf=""><br></span><span leaf="">options(stringsAsFactiors = </span><span><span leaf="">F</span></span><span leaf="">)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(phylogram)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(gridExtra)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(grid)</span><span leaf=""><br></span><span><span leaf="">require</span></span><span leaf="">(dendextend)</span><span leaf=""><br></span><span><span leaf="">require</span></span><span leaf="">(ggthemes)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(tidyverse)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(Seurat)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(infercnv)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(miscTools)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">sce &lt;- qread(</span><span><span leaf="">"./4-Corrective-data/sce.all.qs"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">dir.create(</span><span><span leaf="">"11-infercnv"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">setwd(</span><span><span leaf="">"./11-infercnv"</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><h6 data-tool="mdnice编辑器"><span></span><span leaf="">7、把inferCNV中的run.final.infercnv_obje文件读取进来</span><span></span></h6><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># 提取想要的细胞</span></span><span leaf=""><br></span><span leaf="">sce &lt;- sce[,sce$celltype %</span><span><span leaf="">in</span></span><span leaf="">% c(</span><span><span leaf="">"plasma"</span></span><span leaf="">,</span><span><span leaf="">"fibroblasts"</span></span><span leaf="">,</span><span><span leaf="">"epithelial/cancer cells"</span></span><span leaf="">)]</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 提取infercnv_obj</span></span><span leaf=""><br></span><span leaf="">infer_CNV_obj&lt;-readRDS(</span><span><span leaf="">'./infercnv_output/run.final.infercnv_obj'</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">expr &lt;- infer_CNV_obj@expr.data</span><span leaf=""><br></span><span leaf="">expr[</span><span><span leaf="">1</span></span><span leaf="">:</span><span><span leaf="">4</span></span><span leaf="">,</span><span><span leaf="">1</span></span><span leaf="">:</span><span><span leaf="">4</span></span><span leaf="">]</span><span leaf=""><br></span><span leaf="">data_cnv &lt;- as.data.frame(expr)</span><span leaf=""><br></span><span leaf="">dim(expr)</span><span leaf=""><br></span><span leaf="">colnames(data_cnv)</span><span leaf=""><br></span><span leaf="">rownames(data_cnv)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">table(sce$orig.ident)</span><span leaf=""><br></span><span leaf="">table(sce$seurat_clusters)</span><span leaf=""><br></span><span><span leaf="">#记得要改一下sce中的列名</span></span><span leaf=""><br></span><span leaf="">colnames(sce) &lt;- gsub(</span><span><span leaf="">"-"</span></span><span leaf="">,</span><span><span leaf="">"_"</span></span><span leaf="">,colnames(sce))</span><span leaf=""><br></span><span leaf="">meta &lt;- sce@meta.data</span><span leaf=""><br></span></code></pre><h5 data-tool="mdnice编辑器"><span></span><span leaf="">8、计算CNV并绘图</span><span></span></h5><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">#要根据参数修改哦，这里的对照组有plasma和fibroblast</span></span><span leaf=""><br></span><span><span leaf="">if</span></span><span leaf="">(</span><span><span leaf="">T</span></span><span leaf="">){</span><span leaf=""><br></span><span leaf="">  tmp1 = expr[,infer_CNV_obj@reference_grouped_cell_indices$plasma]</span><span leaf=""><br></span><span leaf="">  tmp = tmp1</span><span leaf=""><br></span><span leaf="">  tmp2 = expr[,infer_CNV_obj@reference_grouped_cell_indices$fibroblast]</span><span leaf=""><br></span><span leaf="">  tmp= cbind(tmp1,tmp2)</span><span leaf=""><br></span><span leaf="">  down=mean(rowMeans(tmp)) - </span><span><span leaf="">2</span></span><span leaf=""> * mean( apply(tmp, </span><span><span leaf="">1</span></span><span leaf="">, sd))</span><span leaf=""><br></span><span leaf="">  up=mean(rowMeans(tmp)) + </span><span><span leaf="">2</span></span><span leaf=""> * mean( apply(tmp, </span><span><span leaf="">1</span></span><span leaf="">, sd))</span><span leaf=""><br></span><span leaf="">  oneCopy=up-down</span><span leaf=""><br></span><span leaf="">  oneCopy</span><span leaf=""><br></span><span leaf="">  a1= down- </span><span><span leaf="">2</span></span><span leaf="">*oneCopy</span><span leaf=""><br></span><span leaf="">  a2= down- </span><span><span leaf="">1</span></span><span leaf="">*oneCopy</span><span leaf=""><br></span><span leaf="">  down;up</span><span leaf=""><br></span><span leaf="">  a3= up +  </span><span><span leaf="">1</span></span><span leaf="">*oneCopy</span><span leaf=""><br></span><span leaf="">  a4= up + </span><span><span leaf="">2</span></span><span leaf="">*oneCopy </span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">  cnv_score_table&lt;-infer_CNV_obj@expr.data</span><span leaf=""><br></span><span leaf="">  cnv_score_table[</span><span><span leaf="">1</span></span><span leaf="">:</span><span><span leaf="">4</span></span><span leaf="">,</span><span><span leaf="">1</span></span><span leaf="">:</span><span><span leaf="">4</span></span><span leaf="">]</span><span leaf=""><br></span><span leaf="">  cnv_score_mat &lt;- as.matrix(cnv_score_table)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># Scoring</span></span><span leaf=""><br></span><span leaf="">  cnv_score_table[cnv_score_mat &gt; </span><span><span leaf="">0</span></span><span leaf=""> &amp; cnv_score_mat &lt; a2] &lt;- </span><span><span leaf="">"A"</span></span><span><span leaf="">#complete loss. 2pts</span></span><span leaf=""><br></span><span leaf="">  cnv_score_table[cnv_score_mat &gt;= a2 &amp; cnv_score_mat &lt; down] &lt;- </span><span><span leaf="">"B"</span></span><span><span leaf="">#loss of one copy. 1pts</span></span><span leaf=""><br></span><span leaf="">  cnv_score_table[cnv_score_mat &gt;= down &amp; cnv_score_mat &lt;  up ] &lt;- </span><span><span leaf="">"C"</span></span><span><span leaf="">#Neutral. 0pts</span></span><span leaf=""><br></span><span leaf="">  cnv_score_table[cnv_score_mat &gt;= up  &amp; cnv_score_mat &lt;= a3] &lt;- </span><span><span leaf="">"D"</span></span><span><span leaf="">#addition of one copy. 1pts</span></span><span leaf=""><br></span><span leaf="">  cnv_score_table[cnv_score_mat &gt; a3  &amp; cnv_score_mat &lt;= a4 ] &lt;- </span><span><span leaf="">"E"</span></span><span><span leaf="">#addition of two copies. 2pts</span></span><span leaf=""><br></span><span leaf="">  cnv_score_table[cnv_score_mat &gt; a4] &lt;- </span><span><span leaf="">"F"</span></span><span><span leaf="">#addition of more than two copies. 2pts</span></span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># Check</span></span><span leaf=""><br></span><span leaf="">  table(cnv_score_table[,</span><span><span leaf="">1</span></span><span leaf="">])</span><span leaf=""><br></span><span><span leaf=""># Replace with score </span></span><span leaf=""><br></span><span leaf="">  cnv_score_table_pts &lt;- cnv_score_mat</span><span leaf=""><br></span><span leaf="">  rm(cnv_score_mat)</span><span leaf=""><br></span><span><span leaf=""># </span></span><span leaf=""><br></span><span leaf="">  cnv_score_table_pts[cnv_score_table == </span><span><span leaf="">"A"</span></span><span leaf="">] &lt;- </span><span><span leaf="">2</span></span><span leaf=""><br></span><span leaf="">  cnv_score_table_pts[cnv_score_table == </span><span><span leaf="">"B"</span></span><span leaf="">] &lt;- </span><span><span leaf="">1</span></span><span leaf=""><br></span><span leaf="">  cnv_score_table_pts[cnv_score_table == </span><span><span leaf="">"C"</span></span><span leaf="">] &lt;- </span><span><span leaf="">0</span></span><span leaf=""><br></span><span leaf="">  cnv_score_table_pts[cnv_score_table == </span><span><span leaf="">"D"</span></span><span leaf="">] &lt;- </span><span><span leaf="">1</span></span><span leaf=""><br></span><span leaf="">  cnv_score_table_pts[cnv_score_table == </span><span><span leaf="">"E"</span></span><span leaf="">] &lt;- </span><span><span leaf="">2</span></span><span leaf=""><br></span><span leaf="">  cnv_score_table_pts[cnv_score_table == </span><span><span leaf="">"F"</span></span><span leaf="">] &lt;- </span><span><span leaf="">2</span></span><span leaf=""><br></span><span leaf="">   </span><span leaf=""><br></span><span leaf="">  cnv_score_table_pts[</span><span><span leaf="">1</span></span><span leaf="">:</span><span><span leaf="">4</span></span><span leaf="">,</span><span><span leaf="">1</span></span><span leaf="">:</span><span><span leaf="">4</span></span><span leaf="">]</span><span leaf=""><br></span><span leaf="">  str(as.data.frame(cnv_score_table_pts[</span><span><span leaf="">1</span></span><span leaf="">:</span><span><span leaf="">4</span></span><span leaf="">,</span><span><span leaf="">1</span></span><span leaf="">:</span><span><span leaf="">4</span></span><span leaf="">])) </span><span leaf=""><br></span><span leaf="">  cell_scores_CNV &lt;- as.data.frame(colSums(cnv_score_table_pts))</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">  colnames(cell_scores_CNV) &lt;- </span><span><span leaf="">"cnv_score"</span></span><span leaf=""><br></span><span leaf="">}</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf="">#可视化</span></span><span leaf=""><br></span><span leaf="">head(cell_scores_CNV) </span><span leaf=""><br></span><span leaf="">score=cell_scores_CNV</span><span leaf=""><br></span><span leaf="">head(score)</span><span leaf=""><br></span><span leaf="">meta$totalCNV = score[match(colnames(sce),</span><span leaf=""><br></span><span leaf="">                            rownames(score)),</span><span><span leaf="">1</span></span><span leaf="">] </span><span leaf=""><br></span><span leaf="">p &lt;- ggplot(meta, aes(x= celltype, </span><span leaf=""><br></span><span leaf="">                 y=totalCNV, </span><span leaf=""><br></span><span leaf="">                 fill=celltype)) +</span><span leaf=""><br></span><span leaf="">            geom_boxplot();print(p) </span><span leaf=""><br></span><span leaf="">ggsave(</span><span><span leaf="">"totalCNV.pdf"</span></span><span leaf="">,plot = p, width = </span><span><span leaf="">9</span></span><span leaf="">,height = </span><span><span leaf="">7</span></span><span leaf="">,dpi = </span><span><span leaf="">300</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">从CNV的结果确实能够看到epithelial/cancer cells与fibroblast和plasma细胞的差异。<img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTiatwSmJZ1UmXCMjPgtOeTTvrQ0zEXjrmwDqIL1yvgGLkShPXyvyxEb0zYH4jrFz2czZxUufkmNzw/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.7805555555555556" data-type="png" data-w="1080" data-imgfileid="100045326" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTiatwSmJZ1UmXCMjPgtOeTTvrQ0zEXjrmwDqIL1yvgGLkShPXyvyxEb0zYH4jrFz2czZxUufkmNzw/640?wx_fmt=png&amp;from=appmsg">本次分析完成了inferCNV工具的实践流程。inferCNV可以帮助我们从RNA水平推测细胞的CNV变化，并辅助开展多种分析工作。例如，它可以用于区分肿瘤细胞和正常细胞，识别基因组CNV，以及探索进化关系和克隆异质性。因此，inferCNV是单细胞分析中必须掌握的重要工具。</span></p><h4 data-tool="mdnice编辑器"><span></span><span leaf="">参考资料:</span><span></span></h4><ol><li><section><span leaf="">Single-cell RNA-seq highlights intratumoral heterogeneity in primary glioblastoma. Science. 2014 Jun 20;344(6190):1396-401.</span></section></li><li><section><span leaf="">inferCNV github：https://github.com/broadinstitute/infercnv https://github.com/broadinstitute/inferCNV/wiki</span></section></li><li><section><span leaf="">生信技能树：https://mp.weixin.qq.com/s/dcPYo9eR2CaPCr1-MbyC6w https://mp.weixin.qq.com/s/suH6jtGFfpaBluTKVCo_qA</span></section></li><li><section><span leaf="">单细胞天地：https://mp.weixin.qq.com/s/ByZFWLipJmyIu71W8TjLbA</span></section></li></ol><p data-tool="mdnice编辑器"><strong><span leaf="">致谢</span></strong><span leaf="">：感谢曾老师以及生信技能树团队全体成员。更多精彩内容可关注公众号：</span><strong><span leaf="">生信技能树，单细胞天地，生信菜鸟团</span></strong><span leaf="">等公众号。</span></p><p data-tool="mdnice编辑器"><strong><span leaf="">注</span></strong><span leaf="">：若对内容有疑惑或者有发现明确错误的朋友，请联系后台(欢迎交流)。</span></p><span><span leaf="">- END -</span></span></section><section><span leaf=""><br></span></section><p><mp-style-type data-value="3"></mp-style-type></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/y1Q_GGvVD-eHtrqg_rkKTg",target="_blank" rel="noopener noreferrer">原文链接</a>
