---
title: "单细胞实战之Ro/e、Augur、miloR——入门到进阶(中级篇4）"
date: 2025-04-06T13:05:16Z
draft: ["false"]
tags: [
  "fetched",
  "单细胞天地"
]
categories: ["Acdemic"]
---
单细胞实战之Ro/e、Augur、miloR——入门到进阶(中级篇4） by 单细胞天地
------
<div><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com" data-pm-slice="0 0 []"><p data-tool="mdnice编辑器"><span leaf="">在上一讲中完成了inferCNV分析流程，<a href="https://mp.weixin.qq.com/s?__biz=MzI1Njk4ODE0MQ==&amp;mid=2247528986&amp;idx=1&amp;sn=41d3a7cd4dea69bb5e03335dfe803060&amp;scene=21#wechat_redirect" textvalue="单细胞实战之inferCNV——入门到进阶(中级篇3）" data-itemshowtype="0" target="_blank" linktype="text" data-linktype="2">单细胞实战之inferCNV——入门到进阶(中级篇3）</a></span></p><p data-tool="mdnice编辑器"><span leaf="">本讲将回顾学习单细胞分析中的三个工具：Ro/e、Augur、miloR，这三个工具分别能够判断细胞簇在组织分布倾向、扰动影响和细胞差异丰度。</span></p><p data-tool="mdnice编辑器"><span leaf="">本次内容涉及到的工程文件可通过网盘获得：中级篇2，链接: https://pan.baidu.com/s/1y-HHLXoXsJbgWKCdz26-gQ 提取码: yx93 。</span></p><p data-tool="mdnice编辑器"><span leaf="">此外，可以向“生信技能树”公众号发送关键词‘单细胞’，直接获取Seurat V5版本的完整代码。</span></p><h5 data-tool="mdnice编辑器"><span></span><span leaf="">Ro/e——判断细胞簇在组织分布倾向的分析工具</span><span></span></h5><p data-tool="mdnice编辑器"><span leaf="">Ro/e分析又称为STARTRAC-distribution(张泽民院士团队开发的STARTRAC框架中的其中一个工具)的主要目的是通过计算组织中观测细胞数与期望细胞数（期望细胞数通过卡方检验计算得出）的比率。<img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTdjPHAyiaICzZXF6cjgvUqqZI6pL4qeibWicU56Bdt9vDTTWWWy4B42icbx2ibezwhMv65Ywmphcfich5g/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.9283387622149837" data-type="png" data-w="614" data-imgfileid="100045497" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTdjPHAyiaICzZXF6cjgvUqqZI6pL4qeibWicU56Bdt9vDTTWWWy4B42icbx2ibezwhMv65Ywmphcfich5g/640?wx_fmt=png&amp;from=appmsg">与卡方检验中定义的 (观测值-期望细胞数)²/期望细胞数仅能反映观测值与期望细胞数的偏离程度不同，STARTRAC通过Ro/e 可进一步指示特定细胞簇在特定组织中是否富集或耗竭。如下图中是分析不同的T细胞簇/巨噬细胞簇在不同组织中的Ro/e富集情况。<img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTdjPHAyiaICzZXF6cjgvUqqGJ01vMRkUxwIZH4rG3IGoBE8rNwaMf61p6iaHWUT6Au17qpCQajKANA/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.5648148148148148" data-type="png" data-w="1080" data-imgfileid="100045498" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTdjPHAyiaICzZXF6cjgvUqqGJ01vMRkUxwIZH4rG3IGoBE8rNwaMf61p6iaHWUT6Au17qpCQajKANA/640?wx_fmt=png&amp;from=appmsg"></span><strong><span leaf="">Ro/e 指数的评分规则 ：</span></strong></p><p data-tool="mdnice编辑器"><span leaf="">如果 Ro/e&gt;1，则表明在该组织中，某细胞簇的细胞数量高于期望细胞数，即表现为富集；如果 Ro/e&lt;1，则表明某细胞簇在该组织中的细胞数量低于期望细胞数，即表现为消耗。 +++, Ro/e &gt; 1; ++, 0.8 &lt; Ro/e ≤ 1; +, 0.2 ≤ Ro/e ≤ 0.8; +/−, 0 &lt; Ro/e &lt; 0.2; −, Ro/e = 0<img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTdjPHAyiaICzZXF6cjgvUqq3VqibZSCeBC0IXbg6LLkIo6xAEY5ibUekNJoCks606FheZY31AYbTDlQ/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.26785714285714285" data-type="png" data-w="616" data-imgfileid="100045494" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTdjPHAyiaICzZXF6cjgvUqq3VqibZSCeBC0IXbg6LLkIo6xAEY5ibUekNJoCks606FheZY31AYbTDlQ/640?wx_fmt=png&amp;from=appmsg">个人认为期望细胞数也是基于现有数据进行测算的并且最后生成的值是比率，所以其实最终结果跟细胞数本身很有关系，同时细胞数量也不应太少，容易出现很大的偏倚。</span></p><h5 data-tool="mdnice编辑器"><span></span><span leaf="">Ro/e 指数分析步骤</span><span></span></h5><h6 data-tool="mdnice编辑器"><span></span><span leaf="">1.导入</span><span></span></h6><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">rm(list = ls())</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(Startrac)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(ggplot2)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(tictoc)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(ggpubr)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(ComplexHeatmap)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(RColorBrewer)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(circlize)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(tidyverse)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(sscVis)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(Seurat)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(tidyverse)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(readr)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(qs)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(BiocParallel)</span><span leaf=""><br></span><span leaf="">register(MulticoreParam(workers = </span><span><span leaf="">8</span></span><span leaf="">, progressbar = </span><span><span leaf="">TRUE</span></span><span leaf="">)) </span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">scRNA &lt;- qread(</span><span><span leaf="">"./9-CD4+T/CD4+t_final.qs"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">Idents(scRNA) &lt;- </span><span><span leaf="">"celltype"</span></span><span leaf=""><br></span><span leaf="">DimPlot(scRNA)</span><span leaf=""><br></span></code></pre><h6 data-tool="mdnice编辑器"><span></span><span leaf="">2.数据预处理</span><span></span></h6><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">data  &lt;- scRNA@meta.data</span><span leaf=""><br></span><span leaf="">colnames(data)</span><span leaf=""><br></span><span><span leaf="">#  [1] "orig.ident"         "nCount_RNA"         "nFeature_RNA"       "percent_mito"      </span></span><span leaf=""><br></span><span><span leaf="">#  [5] "percent_ribo"       "percent_hb"         "RNA_snn_res.0.01"   "seurat_clusters"   </span></span><span leaf=""><br></span><span><span leaf="">#  [9] "RNA_snn_res.0.05"   "RNA_snn_res.0.1"    "RNA_snn_res.0.2"    "RNA_snn_res.0.3"   </span></span><span leaf=""><br></span><span><span leaf=""># [13] "RNA_snn_res.0.5"    "RNA_snn_res.0.8"    "RNA_snn_res.1"      "celltype"          </span></span><span leaf=""><br></span><span><span leaf=""># [17] "location"           "pANN_0.25"          "DF.classifications" "contamination"     </span></span><span leaf=""><br></span><span><span leaf=""># [21] "RNA_snn_res.0.4"    "RNA_snn_res.0.6"    "RNA_snn_res.0.7"    "RNA_snn_res.0.9"   </span></span><span leaf=""><br></span><span><span leaf=""># [25] "RNA_snn_res.1.1"    "RNA_snn_res.1.2"    "RNA_snn_res.1.3"    "RNA_snn_res.1.4"   </span></span><span leaf=""><br></span><span><span leaf=""># [29] "RNA_snn_res.1.5"    "RNA_snn_res.1.6"    "RNA_snn_res.1.7"    "RNA_snn_res.1.8"   </span></span><span leaf=""><br></span><span><span leaf=""># [33] "RNA_snn_res.1.9"    "RNA_snn_res.2" </span></span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 需要得到clone.id(ID标识),patient(样本),majorCluster(细胞亚群),loc(不同组织/位置)列</span></span><span leaf=""><br></span><span><span leaf=""># 并且需要重新命名</span></span><span leaf=""><br></span><span leaf="">data &lt;- data[,c(</span><span><span leaf="">34</span></span><span leaf="">,</span><span><span leaf="">1</span></span><span leaf="">,</span><span><span leaf="">16</span></span><span leaf="">,</span><span><span leaf="">17</span></span><span leaf="">)]</span><span leaf=""><br></span><span leaf="">colnames(data) &lt;- c(</span><span><span leaf="">"clone.id"</span></span><span leaf="">,</span><span><span leaf="">"patient"</span></span><span leaf="">,</span><span><span leaf="">"majorCluster"</span></span><span leaf="">,</span><span><span leaf="">"loc"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">data$Cell_Name &lt;- c(</span><span><span leaf="">"T cell"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">data$patient &lt;- as.character(data$patient)</span><span leaf=""><br></span><span leaf="">data$clone.id &lt;- as.character(data$clone.id)</span><span leaf=""><br></span><span leaf="">data$majorCluster &lt;- as.character(data$majorCluster)</span><span leaf=""><br></span><span leaf="">data$loc &lt;- as.character(data$loc)</span><span leaf=""><br></span></code></pre><h6 data-tool="mdnice编辑器"><span></span><span leaf="">3.Ro/e分析</span><span></span></h6><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">Roe &lt;- calTissueDist(data,</span><span leaf=""><br></span><span leaf="">         byPatient = </span><span><span leaf="">F</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">         colname.cluster = </span><span><span leaf="">"majorCluster"</span></span><span leaf="">, </span><span><span leaf=""># 不同细胞亚群</span></span><span leaf=""><br></span><span leaf="">         colname.patient = </span><span><span leaf="">"patient"</span></span><span leaf="">, </span><span><span leaf=""># 不同样本</span></span><span leaf=""><br></span><span leaf="">         colname.tissue = </span><span><span leaf="">"loc"</span></span><span leaf="">, </span><span><span leaf=""># 不同组织</span></span><span leaf=""><br></span><span leaf="">         method = </span><span><span leaf="">"chisq"</span></span><span leaf="">, </span><span><span leaf=""># "chisq", "fisher", and "freq" </span></span><span leaf=""><br></span><span leaf="">         min.rowSum = </span><span><span leaf="">0</span></span><span leaf="">) </span><span leaf=""><br></span><span leaf="">Roe</span><span leaf=""><br></span><span><span leaf="">#               left     right</span></span><span leaf=""><br></span><span><span leaf=""># ELK4+T    1.3286764 0.7309727</span></span><span leaf=""><br></span><span><span leaf=""># Naive T   0.6669732 1.2725882</span></span><span leaf=""><br></span><span><span leaf=""># Th1       0.8664169 1.1093401</span></span><span leaf=""><br></span><span><span leaf=""># Th17      1.2137669 0.8250281</span></span><span leaf=""><br></span><span><span leaf=""># Tm        0.9822346 1.0145413</span></span><span leaf=""><br></span><span><span leaf=""># Treg      1.3638576 0.7021762</span></span><span leaf=""><br></span><span><span leaf=""># ZNF793+T  1.0462073 0.9621785</span></span><span leaf=""><br></span><span><span leaf=""># ZSCAN12+T 0.5643888 1.3565553</span></span><span leaf=""><br></span><span><span leaf=""># 这是最关键的值，然后进行可视化即可</span></span><span leaf=""><br></span></code></pre><h6 data-tool="mdnice编辑器"><span></span><span leaf="">4.可视化</span><span></span></h6><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">table(data$majorCluster)</span><span leaf=""><br></span><span><span leaf=""># ELK4+T   Naive T       Th1      Th17        Tm      Treg  ZNF793+T ZSCAN12+T </span></span><span leaf=""><br></span><span><span leaf="">#    204      1459       818       421      1045      1528       189       248 </span></span><span leaf=""><br></span><span leaf="">col_fun &lt;- colorRamp2(c(min(Roe, na.rm = </span><span><span leaf="">TRUE</span></span><span leaf="">), </span><span><span leaf="">1</span></span><span leaf="">, max(Roe, na.rm = </span><span><span leaf="">TRUE</span></span><span leaf="">)), </span><span leaf=""><br></span><span leaf="">                      c(</span><span><span leaf="">"blue"</span></span><span leaf="">, </span><span><span leaf="">"white"</span></span><span leaf="">, </span><span><span leaf="">"red"</span></span><span leaf="">))</span><span leaf=""><br></span><span leaf="">Heatmap(as.matrix(Roe),</span><span leaf=""><br></span><span leaf="">        show_heatmap_legend = </span><span><span leaf="">TRUE</span></span><span leaf="">, </span><span leaf=""><br></span><span leaf="">        cluster_rows = </span><span><span leaf="">TRUE</span></span><span leaf="">, </span><span leaf=""><br></span><span leaf="">        cluster_columns = </span><span><span leaf="">TRUE</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">        row_names_side = </span><span><span leaf="">'right'</span></span><span leaf="">, </span><span leaf=""><br></span><span leaf="">        show_column_names = </span><span><span leaf="">TRUE</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">        show_row_names = </span><span><span leaf="">TRUE</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">        col = col_fun,</span><span leaf=""><br></span><span leaf="">        row_names_gp = gpar(fontsize = </span><span><span leaf="">10</span></span><span leaf="">),</span><span leaf=""><br></span><span leaf="">        column_names_gp = gpar(fontsize = </span><span><span leaf="">10</span></span><span leaf="">),</span><span leaf=""><br></span><span leaf="">        heatmap_legend_param = list(</span><span leaf=""><br></span><span leaf="">          title = </span><span><span leaf="">"Ro/e Index"</span></span><span leaf="">,  </span><span><span leaf=""># 自定义图注名称</span></span><span leaf=""><br></span><span leaf="">          at = seq(</span><span><span leaf="">0.5</span></span><span leaf="">, </span><span><span leaf="">1.5</span></span><span leaf="">, by = </span><span><span leaf="">0.5</span></span><span leaf="">), </span><span><span leaf=""># 例刻度的位置/自己的数据必须修改一下！</span></span><span leaf=""><br></span><span leaf="">          labels = seq(</span><span><span leaf="">0.5</span></span><span leaf="">, </span><span><span leaf="">1.5</span></span><span leaf="">, by = </span><span><span leaf="">0.5</span></span><span leaf="">) </span><span><span leaf=""># 每个刻度的标签/自己的数据必须修改一下！</span></span><span leaf=""><br></span><span leaf="">        ),</span><span leaf=""><br></span><span leaf="">        cell_fun = </span><span><span leaf="">function</span></span><span leaf="">(j, i, x, y, width, height, fill) {</span><span leaf=""><br></span><span leaf="">          grid.text(sprintf(</span><span><span leaf="">"%.2f"</span></span><span leaf="">, Roe[i, j]), x, y, gp = gpar(fontsize = </span><span><span leaf="">8</span></span><span leaf="">, col = </span><span><span leaf="">"black"</span></span><span leaf="">))</span><span leaf=""><br></span><span leaf="">        }</span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span><span leaf=""># cell_fun = function(j, i, x, y, width, height, fill)</span></span><span leaf=""><br></span><span><span leaf=""># j 和 i：分别表示单元格的列和行索引。</span></span><span leaf=""><br></span><span><span leaf=""># x 和 y：是单元格的中心位置坐标，用于确定文本的位置。</span></span><span leaf=""><br></span><span><span leaf=""># width 和 height：表示单元格的宽度和高度，可以用来调整文本位置或大小。</span></span><span leaf=""><br></span><span><span leaf=""># fill：表示单元格的背景颜色。</span></span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">setwd(</span><span><span leaf="">".."</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTdjPHAyiaICzZXF6cjgvUqq1bpUOl5AialmUSAFxiaNSibFL6vvra1iaeJmgExXjAlHHAgU5UjON1EMEw/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.4083333333333333" data-type="png" data-w="1080" data-imgfileid="100045495" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTdjPHAyiaICzZXF6cjgvUqq1bpUOl5AialmUSAFxiaNSibFL6vvra1iaeJmgExXjAlHHAgU5UjON1EMEw/640?wx_fmt=png&amp;from=appmsg">从细胞数和Ro/e值可以看到左半结肠中存在较多数量的Treg细胞以及较高的Ro/e值，Naive T细胞则在右半结肠中更多和显著的Ro/e值。因此，这也符合现有的生物学知识，右半结肠癌通常具有更高的免疫原性而左半结肠癌则相反。<img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTdjPHAyiaICzZXF6cjgvUqqTqLA3kIoaXLPenqJtZOjxGRiaxWvuKPFe2KVG6bg0icG5915Ku7vUQvw/640?wx_fmt=png&amp;from=appmsg" data-ratio="1.1453634085213034" data-type="png" data-w="798" data-imgfileid="100045496" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTdjPHAyiaICzZXF6cjgvUqqTqLA3kIoaXLPenqJtZOjxGRiaxWvuKPFe2KVG6bg0icG5915Ku7vUQvw/640?wx_fmt=png&amp;from=appmsg"></span></p><h4 data-tool="mdnice编辑器"><span></span><span leaf="">Augur——判断扰动影响的分析工具</span><span></span></h4><p data-tool="mdnice编辑器"><span leaf="">Augur能够在单细胞数据中优先考虑对生物扰动最敏感的细胞类型，其采用机器学习框架来量化扰动和未扰动细胞在高维空间中的可分性(比如不同的疾病状态，药物刺激等)。Augur 使用的机器学习框架依赖于分类算法，如随机森林（Random Forests）或逻辑回归（Logistic Regression），这些算法非常适合处理具有高维特征空间的单细胞数据。</span></p><p data-tool="mdnice编辑器"><span leaf="">Augur的具体做法是：通过判断是否能够仅凭分子特征准确预测出每个细胞的实验处理标签（如处理组 vs 对照组），来量化这种可分离性。在实际操作中，它会为每种细胞类型单独训练一个机器学习模型，用于预测每个细胞来自哪个实验条件。接着，通过交叉验证评估每个细胞类型特定分类器的准确性，从而为细胞类型的优先级排序提供一个定量的依据。</span></p><h5 data-tool="mdnice编辑器"><span></span><span leaf="">Augur分析流程</span><span></span></h5><h6 data-tool="mdnice编辑器"><span></span><span leaf="">1.导入</span><span></span></h6><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">rm(list = ls())</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(stringr)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(Seurat)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(Augur)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(dplyr)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(patchwork)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(viridis)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(qs)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(BiocParallel)</span><span leaf=""><br></span><span leaf="">register(MulticoreParam(workers = </span><span><span leaf="">8</span></span><span leaf="">, progressbar = </span><span><span leaf="">TRUE</span></span><span leaf="">)) </span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">scRNA &lt;- qread(</span><span><span leaf="">"./9-CD4+T/CD4+t_final.qs"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">Idents(scRNA) &lt;- </span><span><span leaf="">"celltype"</span></span><span leaf=""><br></span><span leaf="">DimPlot(scRNA)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">dir.create(</span><span><span leaf="">"./12-STARTRAC-miloR-Augur"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">setwd(</span><span><span leaf="">"./12-STARTRAC-miloR-Augur"</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><h6 data-tool="mdnice编辑器"><span></span><span leaf="">2.单一组别不同条件下augur分析</span><span></span></h6><p data-tool="mdnice编辑器"><span leaf="">这里的实验分组可以使用比如治疗前后等数据,但是本次分析就把左右半结肠位置数据当做实验分组了。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># 此函数的主要目的是评估并计算每种细胞类型在特定条件（如实验组别或时间点）下的表现，通常通过计算区域下曲线（AUC）来量化。它是单一条件下的细胞类型表现的度量，可以帮助识别在特定实验设置下哪些细胞类型表现最为显著或重要。</span></span><span leaf=""><br></span><span leaf="">augur &lt;- calculate_auc(scRNA,</span><span leaf=""><br></span><span leaf="">                       cell_type_col = </span><span><span leaf="">"celltype"</span></span><span leaf="">, </span><span><span leaf=""># 细胞亚群数据</span></span><span leaf=""><br></span><span leaf="">                       label_col = </span><span><span leaf="">"location"</span></span><span leaf="">, </span><span><span leaf=""># 实验分组数据</span></span><span leaf=""><br></span><span leaf="">                       n_threads = </span><span><span leaf="">8</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">qsave(augur,</span><span><span leaf="">"augur.qs"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">head(augur$AUC, </span><span><span leaf="">5</span></span><span leaf="">)</span><span leaf=""><br></span><span><span leaf=""># # A tibble: 5 × 2</span></span><span leaf=""><br></span><span><span leaf="">#   cell_type   auc</span></span><span leaf=""><br></span><span><span leaf="">#   &lt;chr&gt;     &lt;dbl&gt;</span></span><span leaf=""><br></span><span><span leaf=""># 1 Naive T   0.628</span></span><span leaf=""><br></span><span><span leaf=""># 2 Th17      0.621</span></span><span leaf=""><br></span><span><span leaf=""># 3 Th1       0.604</span></span><span leaf=""><br></span><span><span leaf=""># 4 ZSCAN12+T 0.588</span></span><span leaf=""><br></span><span><span leaf=""># 5 ELK4+T    0.586</span></span><span leaf=""><br></span></code></pre><h6 data-tool="mdnice编辑器"><span></span><span leaf="">3.可视化</span><span></span></h6><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># 明确哪个细胞群体的RNA水平变化最大</span></span><span leaf=""><br></span><span><span leaf=""># plot_lollipop函数是基于ggplot框架的,所以很容易就可以美化图片</span></span><span leaf=""><br></span><span><span leaf=""># 还可以提取原始函数进行修改</span></span><span leaf=""><br></span><span leaf="">augur &lt;- qread(</span><span><span leaf="">"augur.qs"</span></span><span leaf="">)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(ggplot2)</span><span leaf=""><br></span><span leaf="">plot_lollipop(augur) +</span><span leaf=""><br></span><span leaf="">  geom_segment(aes(xend = cell_type, yend = </span><span><span leaf="">0.5</span></span><span leaf="">), size = </span><span><span leaf="">1</span></span><span leaf="">) +</span><span leaf=""><br></span><span leaf="">  geom_point(size = </span><span><span leaf="">3</span></span><span leaf="">, aes(color = cell_type)) +  </span><span><span leaf=""># 增加点的大小和颜色映射</span></span><span leaf=""><br></span><span leaf="">  scale_color_manual(values = c(</span><span leaf=""><br></span><span leaf="">    </span><span><span leaf="">"Naive T"</span></span><span leaf=""> = </span><span><span leaf="">"#1f77b4"</span></span><span leaf="">,       </span><span><span leaf=""># 蓝色</span></span><span leaf=""><br></span><span leaf="">    </span><span><span leaf="">"Th17"</span></span><span leaf=""> = </span><span><span leaf="">"#ff7f0e"</span></span><span leaf="">,    </span><span><span leaf=""># 橙色</span></span><span leaf=""><br></span><span leaf="">    </span><span><span leaf="">"Th1"</span></span><span leaf=""> = </span><span><span leaf="">"#2ca02c"</span></span><span leaf="">,    </span><span><span leaf=""># 绿色</span></span><span leaf=""><br></span><span leaf="">    </span><span><span leaf="">"ZSCAN12+T"</span></span><span leaf=""> = </span><span><span leaf="">"#d62728"</span></span><span leaf="">,        </span><span><span leaf=""># 红色</span></span><span leaf=""><br></span><span leaf="">    </span><span><span leaf="">"ELK4+T"</span></span><span leaf=""> = </span><span><span leaf="">"#9467bd"</span></span><span leaf="">,   </span><span><span leaf=""># 紫色</span></span><span leaf=""><br></span><span leaf="">    </span><span><span leaf="">"Treg"</span></span><span leaf=""> = </span><span><span leaf="">"#8c564b"</span></span><span leaf="">,      </span><span><span leaf=""># 棕色</span></span><span leaf=""><br></span><span leaf="">    </span><span><span leaf="">"Tm"</span></span><span leaf=""> = </span><span><span leaf="">"#e377c2"</span></span><span leaf="">,     </span><span><span leaf=""># 粉色</span></span><span leaf=""><br></span><span leaf="">    </span><span><span leaf="">"ZNF793+T"</span></span><span leaf=""> = </span><span><span leaf="">"#7f7f7f"</span></span><span leaf="">    </span><span><span leaf=""># 灰色</span></span><span leaf=""><br></span><span leaf="">  )) </span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 结合umap</span></span><span leaf=""><br></span><span><span leaf=""># rank模式</span></span><span leaf=""><br></span><span leaf="">plot_umap(augur,scRNA,</span><span leaf=""><br></span><span leaf="">          mode = </span><span><span leaf="">"rank"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">          reduction = </span><span><span leaf="">"umap"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">          palette = </span><span><span leaf="">"cividis"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">          cell_type_col = </span><span><span leaf="">"celltype"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># default模式</span></span><span leaf=""><br></span><span leaf="">plot_umap(augur,scRNA,</span><span leaf=""><br></span><span leaf="">          mode = </span><span><span leaf="">"default"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">          reduction = </span><span><span leaf="">"umap"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">          palette = </span><span><span leaf="">"YlGnBu"</span></span><span leaf="">, </span><span><span leaf="">#  "viridis", "plasma", "magma", "inferno"</span></span><span leaf=""><br></span><span leaf="">          cell_type_col = </span><span><span leaf="">"celltype"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">setwd(</span><span><span leaf="">".."</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">plot_lollipop展示了在不同处理情况下(group)中变化最显著的细胞(AUC值)，从结果来看Naive T细胞受到的扰动程度是最大的，ZNF793+T收到的扰动程度则是最小的。<img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTdjPHAyiaICzZXF6cjgvUqqvvsA4aDqcy9icnoSclFz3smMeROH6eGxWQpS4axhsnwsnHYkj8ClH5w/640?wx_fmt=png&amp;from=appmsg" data-ratio="1.0855745721271393" data-type="png" data-w="409" data-imgfileid="100045501" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTdjPHAyiaICzZXF6cjgvUqqvvsA4aDqcy9icnoSclFz3smMeROH6eGxWQpS4axhsnwsnHYkj8ClH5w/640?wx_fmt=png&amp;from=appmsg">mode中的rank和default模式的区别在于是否用AUC值进行颜色映射。<img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTdjPHAyiaICzZXF6cjgvUqqe2lRpneaYrsnhmZrT36qMomV2hDWDQJnmCwJbgyQicdFfK8y2cfjFag/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.43004115226337447" data-type="png" data-w="486" data-imgfileid="100045500" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTdjPHAyiaICzZXF6cjgvUqqe2lRpneaYrsnhmZrT36qMomV2hDWDQJnmCwJbgyQicdFfK8y2cfjFag/640?wx_fmt=png&amp;from=appmsg"></span></p><h4 data-tool="mdnice编辑器"><span></span><span leaf="">miloR——判断细胞差异丰度的分析工具</span><span></span></h4><p data-tool="mdnice编辑器"><span leaf="">miloR能够随机定义细胞中的细胞节点，然后通过K最近邻法去识别所定义节点与周围的其他细胞节点之间的邻近关系，找到跟这些节点最近的(比如欧几里得法)其他细胞节点，将这些具有邻近关系的细胞小群看做不同的群组，此时可以想象一下，每个群组中都会存在具有疾病和正常信息的许多细胞，接着再将这些群组中疾病与否的细胞数分别进行比较和汇总，最后可以得到基于某一大类细胞中内部细胞变化趋势的结果(比如是否倾向与疾病的方向)。</span></p><p data-tool="mdnice编辑器"><span leaf="">有了这个结果之后，结合预先定义的细胞分群结果有助于使用者更好的去了解疾病与否、疾病状态、药物刺激等情况下的不同细胞亚群的变化状态，因此使用miloR可以有效的作为单细胞水平上细胞差异分析的补充工具。<img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTdjPHAyiaICzZXF6cjgvUqqJoByWGdMUfBzvnfBLIwqsAwVjWBKv9Xwp0STXl52t4ibkQib7eHDzuYQ/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.8672566371681416" data-type="png" data-w="904" data-imgfileid="100045503" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTdjPHAyiaICzZXF6cjgvUqqJoByWGdMUfBzvnfBLIwqsAwVjWBKv9Xwp0STXl52t4ibkQib7eHDzuYQ/640?wx_fmt=png&amp;from=appmsg"></span></p><h5 data-tool="mdnice编辑器"><span></span><span leaf="">miloR分析流程</span><span></span></h5><h6 data-tool="mdnice编辑器"><span></span><span leaf="">1.导入</span><span></span></h6><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">rm(list = ls())</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(stringr)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(Seurat)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(miloR)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(scater)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(scran)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(dplyr)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(patchwork)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(qs)</span><span leaf=""><br></span><span><span leaf="">library</span></span><span leaf="">(BiocParallel)</span><span leaf=""><br></span><span leaf="">register(MulticoreParam(workers = </span><span><span leaf="">8</span></span><span leaf="">, progressbar = </span><span><span leaf="">TRUE</span></span><span leaf="">)) </span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">scRNA &lt;- qread(</span><span><span leaf="">"./9-CD4+T/CD4+t_final.qs"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">Idents(scRNA) &lt;- </span><span><span leaf="">"celltype"</span></span><span leaf=""><br></span><span leaf="">DimPlot(scRNA)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">dir.create(</span><span><span leaf="">"./12-STARTRAC-miloR-Augur"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">setwd(</span><span><span leaf="">"./12-STARTRAC-miloR-Augur"</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><h6 data-tool="mdnice编辑器"><span></span><span leaf="">2.数据预处理</span><span></span></h6><p data-tool="mdnice编辑器"><span leaf="">需要一个seurat对象，建议提前预处理</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># check</span></span><span leaf=""><br></span><span leaf="">table(scRNA$orig.ident)</span><span leaf=""><br></span><span leaf=""> </span><span><span leaf="">#    GSM5688706_WGC     GSM5688707_JCA GSM5688708_LS-CRC3 GSM5688709_RS-CRC1  GSM5688710_R_CRC3 </span></span><span leaf=""><br></span><span leaf=""> </span><span><span leaf="">#               738                542               1381               1879                807 </span></span><span leaf=""><br></span><span leaf=""> </span><span><span leaf=""># GSM5688711_R_CRC4 </span></span><span leaf=""><br></span><span leaf=""> </span><span><span leaf="">#               565</span></span><span leaf=""><br></span><span leaf="">table(scRNA$celltype)</span><span leaf=""><br></span><span leaf="">   </span><span><span leaf=""># ELK4+T   Naive T       Th1      Th17        Tm      Treg  ZNF793+T ZSCAN12+T </span></span><span leaf=""><br></span><span leaf="">   </span><span><span leaf="">#    204      1459       818       421      1045      1528       189       248 </span></span><span leaf=""><br></span></code></pre><h6 data-tool="mdnice编辑器"><span></span><span leaf="">3.构建miloR对象</span><span></span></h6><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># 由于需要输入SingleCellExperiment对象</span></span><span leaf=""><br></span><span><span leaf=""># 因此先进行转化一下</span></span><span leaf=""><br></span><span leaf="">scRNA_pre &lt;- as.SingleCellExperiment(scRNA)</span><span leaf=""><br></span><span><span leaf=""># 构建miloR对象</span></span><span leaf=""><br></span><span leaf="">scRNA_milo &lt;- Milo(scRNA_pre)</span><span leaf=""><br></span><span leaf="">reducedDim(scRNA_milo,</span><span><span leaf="">"UMAP"</span></span><span leaf="">) &lt;- reducedDim(scRNA_pre,</span><span><span leaf="">"UMAP"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">scRNA_milo</span><span leaf=""><br></span><span><span leaf=""># class: Milo </span></span><span leaf=""><br></span><span><span leaf=""># dim: 21308 5912 </span></span><span leaf=""><br></span><span><span leaf=""># metadata(0):</span></span><span leaf=""><br></span><span><span leaf=""># assays(3): counts logcounts scaledata</span></span><span leaf=""><br></span><span><span leaf=""># rownames(21308): RP11-34P13.7 FO538757.2 ... GRIK1-AS1 AP001626.2</span></span><span leaf=""><br></span><span><span leaf=""># rowData names(0):</span></span><span leaf=""><br></span><span><span leaf=""># colnames(5912): GSM5688706_WGC_AACACACGTATCACGT-1 GSM5688706_WGC_AACACACTCAAGGTGG-1 ...</span></span><span leaf=""><br></span><span><span leaf="">#   GSM5688711_R_CRC4_TTTGGTTTCTGGGATT-1 GSM5688711_R_CRC4_TTTGTTGGTAATGCTC-1</span></span><span leaf=""><br></span><span><span leaf=""># colData names(35): orig.ident nCount_RNA ... RNA_snn_res.2 ident</span></span><span leaf=""><br></span><span><span leaf=""># reducedDimNames(3): PCA HARMONY UMAP</span></span><span leaf=""><br></span><span><span leaf=""># mainExpName: RNA</span></span><span leaf=""><br></span><span><span leaf=""># altExpNames(0):</span></span><span leaf=""><br></span><span><span leaf=""># nhoods dimensions(2): 1 1</span></span><span leaf=""><br></span><span><span leaf=""># nhoodCounts dimensions(2): 1 1</span></span><span leaf=""><br></span><span><span leaf=""># nhoodDistances dimension(1): 0</span></span><span leaf=""><br></span><span><span leaf=""># graph names(0):</span></span><span leaf=""><br></span><span><span leaf=""># nhoodIndex names(1): 0</span></span><span leaf=""><br></span><span><span leaf=""># nhoodExpression dimension(2): 1 1</span></span><span leaf=""><br></span><span><span leaf=""># nhoodReducedDim names(0):</span></span><span leaf=""><br></span><span><span leaf=""># nhoodGraph names(0):</span></span><span leaf=""><br></span><span><span leaf=""># nhoodAdjacency dimension(2): 1 1</span></span><span leaf=""><br></span></code></pre><h6 data-tool="mdnice编辑器"><span></span><span leaf="">4.构建K最邻图谱</span><span></span></h6><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># d值与降维时选定的pca值一致</span></span><span leaf=""><br></span><span><span leaf=""># k值与FindNeighbors时选定的值一致</span></span><span leaf=""><br></span><span leaf="">scRNA_milo &lt;- buildGraph(scRNA_milo, k = </span><span><span leaf="">30</span></span><span leaf="">, d = </span><span><span leaf="">15</span></span><span leaf="">,reduced.dim = </span><span><span leaf="">"PCA"</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><h6 data-tool="mdnice编辑器"><span></span><span leaf="">5.在 KNN 图上定义具有代表性的邻域</span><span></span></h6><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># 官网建议prop在0.1-0.2</span></span><span leaf=""><br></span><span><span leaf=""># K和d和KNN图参数一致</span></span><span leaf=""><br></span><span leaf="">scRNA_milo &lt;- makeNhoods(scRNA_milo, prop = </span><span><span leaf="">0.2</span></span><span leaf="">, k = </span><span><span leaf="">30</span></span><span leaf="">, d = </span><span><span leaf="">15</span></span><span leaf="">, </span><span leaf=""><br></span><span leaf="">                    refined = </span><span><span leaf="">TRUE</span></span><span leaf="">, reduced_dims = </span><span><span leaf="">"PCA"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">plotNhoodSizeHist(scRNA_milo)</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTdjPHAyiaICzZXF6cjgvUqq3Wck29ey232kLohpxarbPZ2dibQAAHSziaTibQRfr7tJgd1KQH8IGzMMA/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.20103986135181975" data-type="png" data-w="577" data-imgfileid="100045499" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTdjPHAyiaICzZXF6cjgvUqq3Wck29ey232kLohpxarbPZ2dibQAAHSziaTibQRfr7tJgd1KQH8IGzMMA/640?wx_fmt=png&amp;from=appmsg">开发者建议在50到100达到峰值，否则需要增加k/prop的值。<img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTdjPHAyiaICzZXF6cjgvUqqlVqQPSD4IichmSo2xeiadvxtr51twsiaQyHKO1D7q0TvSjO2HIQT2Llhw/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.6787037037037037" data-type="png" data-w="1080" data-imgfileid="100045502" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTdjPHAyiaICzZXF6cjgvUqqlVqQPSD4IichmSo2xeiadvxtr51twsiaQyHKO1D7q0TvSjO2HIQT2Llhw/640?wx_fmt=png&amp;from=appmsg"></span></p><h6 data-tool="mdnice编辑器"><span></span><span leaf="">6.计算neighbourhoods中的细胞</span><span></span></h6><p data-tool="mdnice编辑器"><span leaf="">计算每个样本中有多少细胞位于每个邻域中。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">scRNA_milo &lt;- countCells(scRNA_milo, </span><span leaf=""><br></span><span leaf="">                    meta.data = as.data.frame(colData(scRNA_milo)),</span><span leaf=""><br></span><span leaf="">                    sample=</span><span><span leaf="">"orig.ident"</span></span><span leaf="">) </span><span><span leaf=""># sample代表样本</span></span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># colData(scRNA) 相当于代表了metadata的信息</span></span><span leaf=""><br></span><span><span leaf=""># sample代表了样本数</span></span><span leaf=""><br></span><span leaf="">head(nhoodCounts(scRNA_milo))</span><span leaf=""><br></span><span><span leaf=""># 这个步骤能够给scRNA数据中增加一个n*m的matrix</span></span><span leaf=""><br></span><span><span leaf=""># n代表不同的neighborhood数量,m代表实验样本</span></span><span leaf=""><br></span><span><span leaf=""># 6 x 6 sparse Matrix of class "dgCMatrix"</span></span><span leaf=""><br></span><span><span leaf="">#   GSM5688706_WGC GSM5688707_JCA GSM5688708_LS-CRC3 GSM5688709_RS-CRC1 GSM5688710_R_CRC3 GSM5688711_R_CRC4</span></span><span leaf=""><br></span><span><span leaf=""># 1             19              7                 41                 18                 .                 .</span></span><span leaf=""><br></span><span><span leaf=""># 2              1              .                  5                  3                 2                61</span></span><span leaf=""><br></span><span><span leaf=""># 3              5              2                 84                  3                 .                 .</span></span><span leaf=""><br></span><span><span leaf=""># 4             10             21                  4                 30                 .                 .</span></span><span leaf=""><br></span><span><span leaf=""># 5             21             14                 59                 15                 .                 .</span></span><span leaf=""><br></span><span><span leaf=""># 6             71              .                 19                  4                 .                 3</span></span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">这为对象添加了一个矩阵，其中n是邻域的数量，𝑚是实验样本的数量。值表示每个样本在邻域中计数的细胞数量。这个计数矩阵将用于差异丰度测试。<img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTdjPHAyiaICzZXF6cjgvUqqp5Bw1Inic5iaicKqorsjq5r0mFmRUiafn4lGsem2LFYUumJibfYmYlVjCsw/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.14756944444444445" data-type="png" data-w="576" data-imgfileid="100045504" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTdjPHAyiaICzZXF6cjgvUqqp5Bw1Inic5iaicKqorsjq5r0mFmRUiafn4lGsem2LFYUumJibfYmYlVjCsw/640?wx_fmt=png&amp;from=appmsg"></span></p><h6 data-tool="mdnice编辑器"><span></span><span leaf="">7.实验设计定义</span><span></span></h6><p data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTdjPHAyiaICzZXF6cjgvUqqE2RSKRnHKdoxRZfWcyk3TaYTEJOBL2prgZ8zEflWlxOFjyqvRz8oJQ/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.2306368330464716" data-type="png" data-w="581" data-imgfileid="100045506" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTdjPHAyiaICzZXF6cjgvUqqE2RSKRnHKdoxRZfWcyk3TaYTEJOBL2prgZ8zEflWlxOFjyqvRz8oJQ/640?wx_fmt=png&amp;from=appmsg">研究者把在使用文件中把condition定义为covariate（协变量），然而在正式分组的时候又把condition定义为实验分组，个人觉得协变量是定义错了，condition就是实验分组。本次数据中location对标condition。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">traj_design &lt;- data.frame(colData(scRNA_milo))[,c(</span><span><span leaf="">"orig.ident"</span></span><span leaf="">, </span><span><span leaf="">"location"</span></span><span leaf="">)]</span><span leaf=""><br></span><span leaf="">traj_design &lt;- distinct(traj_design)</span><span leaf=""><br></span><span leaf="">rownames(traj_design) &lt;- traj_design$orig.ident</span><span leaf=""><br></span><span><span leaf="">#重新排序行名以匹配nhoodCounts（milo）的列</span></span><span leaf=""><br></span><span leaf="">traj_design &lt;- traj_design[colnames(nhoodCounts(scRNA_milo)), , drop=</span><span><span leaf="">FALSE</span></span><span leaf="">]</span><span leaf=""><br></span><span leaf="">traj_design$location &lt;- factor(traj_design$location, level = c(</span><span><span leaf="">"left"</span></span><span leaf="">, </span><span><span leaf="">"right"</span></span><span leaf="">))</span><span leaf=""><br></span><span leaf="">traj_design</span><span leaf=""><br></span><span><span leaf="">#                            orig.ident location</span></span><span leaf=""><br></span><span><span leaf=""># GSM5688706_WGC         GSM5688706_WGC     left</span></span><span leaf=""><br></span><span><span leaf=""># GSM5688707_JCA         GSM5688707_JCA     left</span></span><span leaf=""><br></span><span><span leaf=""># GSM5688708_LS-CRC3 GSM5688708_LS-CRC3     left</span></span><span leaf=""><br></span><span><span leaf=""># GSM5688709_RS-CRC1 GSM5688709_RS-CRC1    right</span></span><span leaf=""><br></span><span><span leaf=""># GSM5688710_R_CRC3   GSM5688710_R_CRC3    right</span></span><span leaf=""><br></span><span><span leaf=""># GSM5688711_R_CRC4   GSM5688711_R_CRC4    right</span></span><span leaf=""><br></span></code></pre><h6 data-tool="mdnice编辑器"><span></span><span leaf="">8.计算邻里连通性</span><span></span></h6><p data-tool="mdnice编辑器"><span leaf="">Milo使用cydar引入的Spatial FDR校正的变体，该校正考虑了邻域之间的重叠。具体来说，每个假设检验的P值都通过k个最近邻距离的倒数进行加权。要使用这个统计量，首先需要在Milo对象中存储最近邻之间的距离。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># Milo使用cydar引入的Spatial FDR校正的变体，该校正考虑了邻域之间的重叠。具体来说，每个假设检验的P值都通过k个最近邻距离的倒数进行加权。要使用这个统计量，首先需要在Milo对象中存储最近邻之间的距离。</span></span><span leaf=""><br></span><span leaf="">scRNA_milo &lt;- calcNhoodDistance(scRNA_milo, d=</span><span><span leaf="">15</span></span><span leaf="">, reduced.dim = </span><span><span leaf="">"PCA"</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><h6 data-tool="mdnice编辑器"><span></span><span leaf="">9.检测结果</span><span></span></h6><p data-tool="mdnice编辑器"><span leaf="">正式测试，明确定义实验设计，这里使用了location</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># 把批次和分组信息加到design中去</span></span><span leaf=""><br></span><span leaf="">da_results &lt;- testNhoods(scRNA_milo,design = ~ location, design.df = traj_design)</span><span leaf=""><br></span><span leaf="">da_results %&gt;%</span><span leaf=""><br></span><span leaf="">  arrange(SpatialFDR) %&gt;%</span><span leaf=""><br></span><span leaf="">  head() </span><span leaf=""><br></span><span><span leaf="">#         logFC   logCPM        F      PValue       FDR Nhood SpatialFDR</span></span><span leaf=""><br></span><span><span leaf=""># 15  -8.035409 10.53289 7.723075 0.005486977 0.1453135    15  0.1009631</span></span><span leaf=""><br></span><span><span leaf=""># 25   7.663408 10.18627 8.654687 0.003287987 0.1453135    25  0.1009631</span></span><span leaf=""><br></span><span><span leaf=""># 31   8.492067 10.93321 8.159168 0.004314811 0.1453135    31  0.1009631</span></span><span leaf=""><br></span><span><span leaf=""># 131 -8.311036 10.77975 7.523350 0.006127759 0.1453135   131  0.1009631</span></span><span leaf=""><br></span><span><span leaf=""># 183  8.511668 10.94935 8.097942 0.004462580 0.1453135   183  0.1009631</span></span><span leaf=""><br></span><span><span leaf=""># 231 -8.242297 10.71764 7.567548 0.005979667 0.1453135   231  0.1009631</span></span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 可以看一下柱状图和点图</span></span><span leaf=""><br></span><span leaf="">ggplot(da_results, aes(PValue)) + geom_histogram(bins=</span><span><span leaf="">50</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">ggplot(da_results, aes(logFC, -log10(SpatialFDR))) + </span><span leaf=""><br></span><span leaf="">  geom_point() +</span><span leaf=""><br></span><span leaf="">  geom_hline(yintercept = </span><span><span leaf="">1</span></span><span leaf="">) </span><span><span leaf="">## Mark significance threshold (10% FDR)</span></span><span leaf=""><br></span></code></pre><h6 data-tool="mdnice编辑器"><span></span><span leaf="">10.可视化</span><span></span></h6><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">scRNA_milo &lt;- buildNhoodGraph(scRNA_milo)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf="">## Plot single-cell UMAP</span></span><span leaf=""><br></span><span leaf="">umap_pl &lt;- plotReducedDim(scRNA_milo, dimred = </span><span><span leaf="">"UMAP"</span></span><span leaf="">, </span><span leaf=""><br></span><span leaf="">                          colour_by=</span><span><span leaf="">"location"</span></span><span leaf="">, text_by = </span><span><span leaf="">"celltype"</span></span><span leaf="">, </span><span leaf=""><br></span><span leaf="">                          text_size = </span><span><span leaf="">3</span></span><span leaf="">, point_size=</span><span><span leaf="">0.5</span></span><span leaf="">) + guides(fill=</span><span><span leaf="">"none"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">umap_pl</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf="">## Plot neighbourhood graph</span></span><span leaf=""><br></span><span leaf="">nh_graph_pl &lt;- plotNhoodGraphDA(scRNA_milo, da_results,</span><span leaf=""><br></span><span leaf="">                                layout=</span><span><span leaf="">"UMAP"</span></span><span leaf="">,alpha=</span><span><span leaf="">0.9</span></span><span leaf="">)  </span><span><span leaf="">#alpha默认0.1</span></span><span leaf=""><br></span><span leaf="">  </span><span leaf=""><br></span><span leaf="">umap_pl + nh_graph_pl +</span><span leaf=""><br></span><span leaf="">plot_layout(guides=</span><span><span leaf="">"collect"</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">UMAP 图（左图）：显示了不同细胞类型在两个实验组之间的分布。图中使用了两种颜色，蓝色（left） 和 橙色（right），分别代表left组和right组。不同的细胞类型被明确标注，并且可以看到left组和right组之间的细胞分布。</span></p><p data-tool="mdnice编辑器"><span leaf="">邻域网络图（右图）：右边的图是基于MiloR分析的邻域图。每个圈代表一个邻域，邻域的大小与其中细胞的数量成比例。颜色梯度表示了logFC值：正的 logFC（蓝色）意味着该邻域中的细胞在right组中富集，负的 logFC（红色）则表明该邻域中的细胞在left组中富集。邻域之间的连线表示邻域之间的相似性或重叠，线条的厚度表示邻域之间的重叠度（overlap size），越厚的线表示两个邻域之间的细胞更为相似。<img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTdjPHAyiaICzZXF6cjgvUqqFU2XWwScXiao128Wp5yOxjlLUbay2n3houGZciaHZEO9Rr0T4YnGoL0w/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.7583333333333333" data-type="png" data-w="1080" data-imgfileid="100045508" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTdjPHAyiaICzZXF6cjgvUqqFU2XWwScXiao128Wp5yOxjlLUbay2n3houGZciaHZEO9Rr0T4YnGoL0w/640?wx_fmt=png&amp;from=appmsg"></span></p><h6 data-tool="mdnice编辑器"><span></span><span leaf="">11.数据注释</span><span></span></h6><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">da_results &lt;- annotateNhoods(scRNA_milo, </span><span leaf=""><br></span><span leaf="">                             da_results, </span><span leaf=""><br></span><span leaf="">                             coldata_col = </span><span><span leaf="">"celltype"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">head(da_results)</span><span leaf=""><br></span><span><span leaf="">#       logFC    logCPM         F     PValue       FDR Nhood SpatialFDR celltype celltype_fraction</span></span><span leaf=""><br></span><span><span leaf=""># 1 -1.574239  9.515303 0.4603793 0.49750211 0.6649894     1  0.6489640  Naive T         0.8596491</span></span><span leaf=""><br></span><span><span leaf=""># 2  2.920511 10.169145 2.1800639 0.13991968 0.3845433     2  0.3461536  Naive T         0.8021978</span></span><span leaf=""><br></span><span><span leaf=""># 3  4.925942 10.883476 4.0560689 0.04410711 0.2888600     3  0.2281237  Naive T         0.8735632</span></span><span leaf=""><br></span><span><span leaf=""># 4 -2.618949 11.056240 1.2512551 0.26340764 0.4888901     4  0.4599119       Tm         0.8437500</span></span><span leaf=""><br></span><span><span leaf=""># 5 -2.062491 10.447327 1.0799157 0.29880501 0.5154568     5  0.4895583       Tm         0.4098361</span></span><span leaf=""><br></span><span><span leaf=""># 6  1.986811 10.452844 1.0385032 0.30825677 0.5173425     6  0.4939665      Th1         0.5952381</span></span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">ggplot(da_results, aes(celltype_fraction)) + geom_histogram(bins=</span><span><span leaf="">50</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">str(da_results)</span><span leaf=""><br></span><span leaf="">table(da_results$celltype)</span><span leaf=""><br></span><span leaf="">range(da_results$SpatialFDR)</span><span leaf=""><br></span><span leaf="">plotDAbeeswarm(da_results, group.by = </span><span><span leaf="">"celltype"</span></span><span leaf="">,alpha = </span><span><span leaf="">0.9</span></span><span leaf="">) </span><span><span leaf=""># alpha默认0.1</span></span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">celltype_fraction 直方图：展示了邻域中某种特定细胞类型的比例分布（celltype_fraction）。Y轴代表邻域的数量，X轴表示邻域中细胞类型所占的比例（范围从0到1）。如果邻域的细胞类型比例接近1，说明大多数邻域是单一细胞类型主导的，这意味着邻域中的细胞几乎都是同一种类型，没有明显的混合细胞类型。这种结果通常出现在较为纯净的细胞群或具有显著分离特征的细胞亚群。<img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTdjPHAyiaICzZXF6cjgvUqqysOjwjR8zNJsQGmQBuSoFCJ60Z3DeQ23kBUwhTic6Sq1FsDQzPv5hOg/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.8788870703764321" data-type="png" data-w="611" data-imgfileid="100045507" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTdjPHAyiaICzZXF6cjgvUqqysOjwjR8zNJsQGmQBuSoFCJ60Z3DeQ23kBUwhTic6Sq1FsDQzPv5hOg/640?wx_fmt=png&amp;from=appmsg">logFC 与细胞类型关联的 beeswarm 图：展示了不同细胞类型的logFC在不同邻域中的分布。Y轴表示细胞类型，X轴表示logFC（即不同组之间在邻域中的差异丰度变化），每个点代表一个邻域。比如Naive T细胞似乎在right组中的邻域具有更大的logFC值，Treg细胞在两组中的logFC值似乎差不多，但在left组中邻域数量更多且集中在-2.5左右。<img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTdjPHAyiaICzZXF6cjgvUqqxqBnzCcPfx4icUVRusmvF090RmuVAwe1Waic73DbXd8OlhzUOkmRfibvw/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.8788870703764321" data-type="png" data-w="611" data-imgfileid="100045505" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTdjPHAyiaICzZXF6cjgvUqqxqBnzCcPfx4icUVRusmvF090RmuVAwe1Waic73DbXd8OlhzUOkmRfibvw/640?wx_fmt=png&amp;from=appmsg"></span></p><h6 data-tool="mdnice编辑器"><span></span><span leaf="">12.找到DA群体的marker</span><span></span></h6><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">## Add log normalized count to Milo object</span></span><span leaf=""><br></span><span leaf="">scRNA_milo &lt;- logNormCounts(scRNA_milo)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># alpha默认0.1,所以SpatialFDR &lt; 0.9是修改过的</span></span><span leaf=""><br></span><span leaf="">da_results$NhoodGroup &lt;- as.numeric(da_results$SpatialFDR &lt; </span><span><span leaf="">0.9</span></span><span leaf=""> &amp; da_results$logFC &lt; </span><span><span leaf="">0</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">da_nhood_markers &lt;- findNhoodGroupMarkers(scRNA_milo, da_results, </span><span leaf=""><br></span><span leaf="">                                          subset.row = rownames(scRNA_milo), </span><span leaf=""><br></span><span leaf="">                                          aggregate.samples = </span><span><span leaf="">TRUE</span></span><span leaf="">, </span><span><span leaf=""># 建议设置为TRUE</span></span><span leaf=""><br></span><span leaf="">                                          sample_col = </span><span><span leaf="">"orig.ident"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">head(da_nhood_markers)</span><span leaf=""><br></span><span><span leaf="">#     GeneID      logFC_1 adj.P.Val_1      logFC_0 adj.P.Val_0</span></span><span leaf=""><br></span><span><span leaf=""># 1     A1BG  0.069556844   0.8845036 -0.069556844   0.8845036</span></span><span leaf=""><br></span><span><span leaf=""># 2 A1BG-AS1 -0.034653314   0.8845036  0.034653314   0.8845036</span></span><span leaf=""><br></span><span><span leaf=""># 3     A1CF  0.000000000   1.0000000  0.000000000   1.0000000</span></span><span leaf=""><br></span><span><span leaf=""># 4      A2M  0.002280312   0.9963105 -0.002280312   0.9963105</span></span><span leaf=""><br></span><span><span leaf=""># 5  A2M-AS1 -0.004230877   1.0000000  0.004230877   1.0000000</span></span><span leaf=""><br></span><span><span leaf=""># 6    A2ML1  0.000000000   1.0000000  0.000000000   1.0000000</span></span><span leaf=""><br></span></code></pre><h6 data-tool="mdnice编辑器"><span></span><span leaf="">12.1.自动对邻域进行分组</span><span></span></h6><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">#在许多情况下，DA邻域位于KNN图的不同区域，将所有重要的 DA 群体分组在一起可能并不理想,因为它们可能包括非常不同细胞类型的细胞。</span></span><span leaf=""><br></span><span><span leaf=""># 对于这种场景,开发者用了一个邻里函数,它使用社区检测根据(1)两个邻域之间共享的小区数量将邻域划分为组;(2)DA 社区的倍数变化方向;(3)折叠变化的差异。</span></span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf="">## Run buildNhoodGraph to store nhood adjacency matrix</span></span><span leaf=""><br></span><span leaf="">scRNA_milo &lt;- buildNhoodGraph(scRNA_milo)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf="">## Find groups</span></span><span leaf=""><br></span><span leaf="">da_results &lt;- groupNhoods(scRNA_milo, da_results, </span><span leaf=""><br></span><span leaf="">                          max.lfc.delta = </span><span><span leaf="">2</span></span><span leaf="">,da.fdr = </span><span><span leaf="">0.9</span></span><span leaf="">)</span><span><span leaf="">#da.fdr默认0，1</span></span><span leaf=""><br></span><span leaf="">head(da_results)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">plotNhoodGroups(scRNA_milo, da_results, layout=</span><span><span leaf="">"UMAP"</span></span><span leaf="">) </span><span leaf=""><br></span><span leaf="">plotDAbeeswarm(da_results, </span><span><span leaf="">"NhoodGroup"</span></span><span leaf="">,alpha = </span><span><span leaf="">0.9</span></span><span leaf="">) </span><span><span leaf="">#alpha默认是0.1</span></span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">这个步骤是按照DA的结果进行自动划分簇，后续步骤还可以找到差异基因。<img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTdjPHAyiaICzZXF6cjgvUqqnBnpfTkgstWHmdABib1N93VkKCjtPic3buEHY6QJu2Z4DYPoM6gVUBgg/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.5287037037037037" data-type="png" data-w="1080" data-imgfileid="100045513" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTdjPHAyiaICzZXF6cjgvUqqnBnpfTkgstWHmdABib1N93VkKCjtPic3buEHY6QJu2Z4DYPoM6gVUBgg/640?wx_fmt=png&amp;from=appmsg"></span></p><h6 data-tool="mdnice编辑器"><span></span><span leaf="">12.2 找到特征基因</span><span></span></h6><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">## Exclude zero counts genes</span></span><span leaf=""><br></span><span leaf="">keep.rows &lt;- rowSums(logcounts(scRNA_milo)) != </span><span><span leaf="">0</span></span><span leaf=""><br></span><span leaf="">scRNA_milo &lt;- scRNA_milo[keep.rows, ]</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf="">## Find HVGs</span></span><span leaf=""><br></span><span leaf="">dec &lt;- modelGeneVar(scRNA_milo)</span><span leaf=""><br></span><span leaf="">hvgs &lt;- getTopHVGs(dec, n=</span><span><span leaf="">2000</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">head(hvgs)</span><span leaf=""><br></span><span><span leaf=""># [1] "HSPA1A" "HSPA1B" "ANXA1"  "S100A4" "DNAJB1" "TXNIP" </span></span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 找到每个邻里群的one-vs-all差异基因表达</span></span><span leaf=""><br></span><span leaf="">nhood_markers &lt;- findNhoodGroupMarkers(scRNA_milo, da_results, subset.row = hvgs, </span><span leaf=""><br></span><span leaf="">                                       aggregate.samples = </span><span><span leaf="">TRUE</span></span><span leaf="">, sample_col = </span><span><span leaf="">"orig.ident"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">head(nhood_markers)[</span><span><span leaf="">1</span></span><span leaf="">:</span><span><span leaf="">4</span></span><span leaf="">,</span><span><span leaf="">1</span></span><span leaf="">:</span><span><span leaf="">4</span></span><span leaf="">]</span><span leaf=""><br></span><span><span leaf="">#   GeneID     logFC_1 adj.P.Val_1      logFC_2</span></span><span leaf=""><br></span><span><span leaf=""># 1  ABCA1  0.09845269   0.4335701 -0.034954072</span></span><span leaf=""><br></span><span><span leaf=""># 2  ABCA2  0.13412658   0.1315646 -0.047259161</span></span><span leaf=""><br></span><span><span leaf=""># 3  ABCA5  0.01590351   0.7922435  0.010316026</span></span><span leaf=""><br></span><span><span leaf=""># 4  ABCA7 -0.01051437   0.8831909  0.001969199</span></span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 找到group2的差异基因</span></span><span leaf=""><br></span><span leaf="">gr2_markers &lt;- nhood_markers[c(</span><span><span leaf="">"logFC_2"</span></span><span leaf="">, </span><span><span leaf="">"adj.P.Val_2"</span></span><span leaf="">)] </span><span leaf=""><br></span><span leaf="">colnames(gr2_markers) &lt;- c(</span><span><span leaf="">"logFC"</span></span><span leaf="">, </span><span><span leaf="">"adj.P.Val"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">head(gr2_markers[order(gr2_markers$adj.P.Val), ])</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 如果明确了对第二组细胞感兴趣</span></span><span leaf=""><br></span><span leaf="">nhood_markers &lt;- findNhoodGroupMarkers(scRNA_milo, da_results, subset.row = hvgs, </span><span leaf=""><br></span><span leaf="">                                       aggregate.samples = </span><span><span leaf="">TRUE</span></span><span leaf="">, sample_col = </span><span><span leaf="">"orig.ident"</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">                                       subset.groups = c(</span><span><span leaf="">"2"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">                                       )</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">head(nhood_markers)</span><span leaf=""><br></span><span><span leaf="">#          logFC_2  adj.P.Val_2 GeneID</span></span><span leaf=""><br></span><span><span leaf=""># LARP4  0.2920211 0.0000272445  LARP4</span></span><span leaf=""><br></span><span><span leaf=""># OAS3   0.2884843 0.0002802494   OAS3</span></span><span leaf=""><br></span><span><span leaf=""># NFKBID 0.3886595 0.0003353456 NFKBID</span></span><span leaf=""><br></span><span><span leaf=""># SSBP2  0.3669311 0.0007693966  SSBP2</span></span><span leaf=""><br></span><span><span leaf=""># CD8B   0.3266638 0.0010654284   CD8B</span></span><span leaf=""><br></span><span><span leaf=""># MRPL18 0.4196821 0.0067862948 MRPL18</span></span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 如果明确了相比较的邻域</span></span><span leaf=""><br></span><span leaf="">nhood_markers &lt;- findNhoodGroupMarkers(scRNA_milo, da_results, subset.row = hvgs,</span><span leaf=""><br></span><span leaf="">                                       subset.nhoods = da_results$NhoodGroup %</span><span><span leaf="">in</span></span><span leaf="">% c(</span><span><span leaf="">'2'</span></span><span leaf="">,</span><span><span leaf="">'5'</span></span><span leaf="">),</span><span leaf=""><br></span><span leaf="">                                       aggregate.samples = </span><span><span leaf="">TRUE</span></span><span leaf="">, sample_col = </span><span><span leaf="">"orig.ident"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">head(nhood_markers)</span><span leaf=""><br></span><span><span leaf="">#   GeneID      logFC_2 adj.P.Val_2      logFC_5 adj.P.Val_5</span></span><span leaf=""><br></span><span><span leaf=""># 1  ABCA1  0.007280498  0.63548063 -0.007280498  0.63548063</span></span><span leaf=""><br></span><span><span leaf=""># 2  ABCA2 -0.024945154  0.40911416  0.024945154  0.40911416</span></span><span leaf=""><br></span><span><span leaf=""># 3  ABCA5  0.021147177  0.25841475 -0.021147177  0.25841475</span></span><span leaf=""><br></span><span><span leaf=""># 4  ABCA7 -0.015377090  0.66034055  0.015377090  0.66034055</span></span><span leaf=""><br></span><span><span leaf=""># 5  ABCD2 -0.006936849  0.65999062  0.006936849  0.65999062</span></span><span leaf=""><br></span><span><span leaf=""># 6 ABHD18 -0.053001633  0.04725136  0.053001633  0.04725136</span></span><span leaf=""><br></span></code></pre><h6 data-tool="mdnice编辑器"><span></span><span leaf="">12.3 差异基因可视化</span><span></span></h6><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># 参数自行修改</span></span><span leaf=""><br></span><span leaf="">ggplot(nhood_markers, aes(logFC_2,-log10(adj.P.Val_2 ))) + </span><span leaf=""><br></span><span leaf="">  geom_point(alpha=</span><span><span leaf="">0.5</span></span><span leaf="">, size=</span><span><span leaf="">0.5</span></span><span leaf="">) +</span><span leaf=""><br></span><span leaf="">  geom_hline(yintercept = </span><span><span leaf="">0.05</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 筛选标准可以自行修改</span></span><span leaf=""><br></span><span leaf="">markers &lt;- nhood_markers$GeneID[nhood_markers$adj.P.Val_2 &lt; </span><span><span leaf="">0.05</span></span><span leaf=""> &amp; nhood_markers$logFC_2 &gt; </span><span><span leaf="">1</span></span><span leaf="">]</span><span leaf=""><br></span><span leaf="">plotNhoodExpressionGroups(scRNA_milo, da_results, features=markers,</span><span leaf=""><br></span><span leaf="">                          subset.nhoods = da_results$NhoodGroup %</span><span><span leaf="">in</span></span><span leaf="">% c(</span><span><span leaf="">'2'</span></span><span leaf="">,</span><span><span leaf="">'5'</span></span><span leaf="">), </span><span leaf=""><br></span><span leaf="">                          scale=</span><span><span leaf="">TRUE</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">                          grid.space = </span><span><span leaf="">"fixed"</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTdjPHAyiaICzZXF6cjgvUqqWJ81rmuYgzPM67BKaGwuP4xjgYVMCol25rQOm8YyV5yoACc52GNwow/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.5452898550724637" data-type="png" data-w="552" data-imgfileid="100045512" src="https://mmbiz.qpic.cn/mmbiz_png/siaia0BDGJdjTdjPHAyiaICzZXF6cjgvUqqWJ81rmuYgzPM67BKaGwuP4xjgYVMCol25rQOm8YyV5yoACc52GNwow/640?wx_fmt=png&amp;from=appmsg"></span></figure><h6 data-tool="mdnice编辑器"><span></span><span leaf="">12.4 同一邻域但是不同干预的细胞比较</span><span></span></h6><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">dge_5 &lt;- testDiffExp(scRNA_milo, da_results, design = ~ location, </span><span leaf=""><br></span><span leaf="">                     meta.data = data.frame(colData(scRNA_milo)),</span><span leaf=""><br></span><span leaf="">                     subset.row = rownames(scRNA_milo),</span><span leaf=""><br></span><span leaf="">                     subset.nhoods=da_results$NhoodGroup==</span><span><span leaf="">"5"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">dge_5</span><span leaf=""><br></span><span><span leaf=""># $`5`</span></span><span leaf=""><br></span><span><span leaf="">#                    logFC    AveExpr          t       P.Value     adj.P.Val          B Nhood.Group</span></span><span leaf=""><br></span><span><span leaf=""># RPS26         2.05155879 2.16078022  40.913548 2.133985e-259 3.832637e-255 583.478172           5</span></span><span leaf=""><br></span><span><span leaf=""># TSC22D3       1.18443050 1.71490401  18.676109  2.678121e-71  2.404953e-67 151.256496           5</span></span><span leaf=""><br></span><span><span leaf=""># TXNIP         0.92242005 0.78962782  17.282387  4.814448e-62  2.882250e-58 130.022537           5</span></span><span leaf=""><br></span><span><span leaf=""># IL7R          1.06746697 1.32047401  16.570152  1.643201e-57  7.377973e-54 119.625571           5</span></span><span leaf=""><br></span><span><span leaf=""># UGP2          0.80880323 0.85367836  16.342942  4.291897e-56  1.541649e-52 116.376331           5</span></span><span leaf=""><br></span><span><span leaf=""># FAM118A      -0.63992629 0.59464042 -15.560710  2.510389e-51  7.514431e-48 105.447340           5</span></span><span leaf=""><br></span><span><span leaf=""># ETS1          0.69424488 1.08136205  14.602241  9.954116e-46  2.553942e-42  92.618518           5</span></span><span leaf=""><br></span><span><span leaf=""># AIM1          0.36544366 0.22221315  13.380244  5.355135e-39  1.202228e-35  77.204983           5</span></span><span leaf=""><br></span><span><span leaf=""># MT-CO1       -0.58424851 5.05175258 -12.970636  7.565387e-37  1.509715e-33  72.284370           5</span></span><span leaf=""><br></span><span><span leaf=""># SMAP2         0.53193670 0.77612312  12.062342  2.816038e-32  5.057605e-29  61.829911           5</span></span><span leaf=""><br></span><span><span leaf=""># RPS9         -0.58244288 2.86336400 -11.929160  1.249072e-31  2.039394e-28  60.350991           5</span></span><span leaf=""><br></span><span><span leaf=""># ZNF765        0.17038370 0.06098921  11.890038  1.929662e-31  2.888060e-28  59.919222           5</span></span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">需要提醒一下，其中很多参数笔者做了修改也做了注释，使用的时候需要先从默认值开始分析！</span></p><p data-tool="mdnice编辑器"><span leaf="">本次内容完成了这三个工具的完整实践流程。虽然它们各自侧重的分析方向不同，但在实际应用中，它们的使用场景是存在交叉的。也就是说，我们可以针对同一批数据，分别使用这三个工具从不同的视角和分析维度进行解读，最终获得更加全面、立体的生物学信息。</span></p><h4 data-tool="mdnice编辑器"><span></span><span leaf="">参考资料：</span><span></span></h4><ol><li><section><span leaf="">Lineage tracking reveals dynamic relationships of T cells in colorectal cancer. Nature. 2018 Dec;564(7735):268-272.</span></section></li><li><section><span leaf="">A pan-cancer single-cell transcriptional atlas of tumor infiltrating myeloid cells. Cell. 2021 Feb 4;184(3):792-809.e23.</span></section></li><li><section><span leaf="">Global characterization of T cells in non-small-cell lung cancer by single-cell sequencing. Nat Med. 2018 Jul;24(7):978-985.</span></section></li><li><section><span leaf="">Cell type prioritization in single-cell data. Nat Biotechnol. 2021 Jan;39(1):30-34.</span></section></li><li><section><span leaf="">Differential abundance testing on single-cell data using k-nearest neighbor graphs. Nat Biotechnol. 2022 Feb;40(2):245-253.</span></section></li><li><section><span leaf="">STARTRAC github：https://github.com/Japrin/STARTRAC</span></section></li><li><section><span leaf="">Augur github：https://github.com/neurorestore/Augur</span></section></li><li><section><span leaf="">miloR github：https://rawcdn.githack.com/MarioniLab/miloR/7c7f906b94a73e62e36e095ddb3e3567b414144e/vignettes/milo_gastrulation.html#5_Finding_markers_of_DA_populations</span></section></li><li><section><span leaf="">单细胞天地：https://mp.weixin.qq.com/s/FlO4fN5AlcyZ81llzHE5A https://mp.weixin.qq.com/s/yYTxb_kOjeRSkzzuUoI4mA</span></section></li><li><section><span leaf="">生信菜鸟团：https://mp.weixin.qq.com/s/D0nGq3XGZRIlso7dKo0qkQ</span></section></li></ol><p data-tool="mdnice编辑器"><strong><span leaf="">致谢</span></strong><span leaf="">：感谢曾老师以及生信技能树团队全体成员。更多精彩内容可关注公众号：</span><strong><span leaf="">生信技能树，单细胞天地，生信菜鸟团</span></strong><span leaf="">等公众号。</span></p><p data-tool="mdnice编辑器"><strong><span leaf="">注</span></strong><span leaf="">：若对内容有疑惑或者有发现明确错误的朋友，请联系后台(欢迎交流)。</span></p><span><span leaf="">- END -</span></span></section><section><span leaf=""><br></span></section><p><mp-style-type data-value="3"></mp-style-type></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/AVQJEYiF7TApDnRzyYXFWA",target="_blank" rel="noopener noreferrer">原文链接</a>
