---
title: "空间转录组数据注释分析：Cell2location反卷积(Nature Biotechnology IF: 33.1)"
date: 2025-04-03T00:34:09Z
draft: ["false"]
tags: [
  "fetched",
  "生信技能树"
]
categories: ["Acdemic"]
---
空间转录组数据注释分析：Cell2location反卷积(Nature Biotechnology IF: 33.1) by 生信技能树
------
<div><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com" data-pm-slice="0 0 []"><blockquote><span></span><p><span leaf="">cell2location 于2022年发表于Nature Biotechnology [IF: 33.1]，<span textstyle="">是一种基于分层贝叶斯统计模型的空间转录组反卷积分析工具，旨在用于精确解析组织中不同细胞类型的空间分布及其丰度。</span>它需要使用单细胞数据作为参考，将单细胞水平的细胞类型注释信息“映射”到空间位点中，揭示复杂组织中细胞类型、状态与空间位置的关联，为理解组织微环境提供重要线索。</span></p></blockquote><p data-tool="mdnice编辑器"><span leaf=""><span textstyle="">这个方法需要性能比较高的计算资源配置，</span><span textstyle="">如果有GPU显卡会大大的缩短运行建模的时间</span>，如果你没有，可以考虑生信技能树的共享服务器，<span textstyle="">只要800一年，内存2T，配置20G显卡</span>，购买详情可看：</span><a href="https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247535760&amp;idx=2&amp;sn=1e02a2e982a046ecf6389231e6768d5b&amp;scene=21#wechat_redirect"><span leaf="">满足你生信分析计算需求的低价解决方案</span></a></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wx8HH4R5qMk8mu6icA8Mzb9kz32wJxyjARIS0vQdXBN8QiapyWu697qEN0DASicOL0msP28CxK8Z8nPw/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.47261009667024706" data-type="png" data-w="931" data-imgfileid="100056714" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wx8HH4R5qMk8mu6icA8Mzb9kz32wJxyjARIS0vQdXBN8QiapyWu697qEN0DASicOL0msP28CxK8Z8nPw/640?wx_fmt=png&amp;from=appmsg"></span></figure><h2 data-cacheurl="" data-remoteid="" data-tool="mdnice编辑器"><span></span><span><span leaf="">软件原理图如下</span></span><span></span></h2><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wx8HH4R5qMk8mu6icA8Mzb9k0UR6GMwiaMraQARfibwWvKCAt4fOxgMQFDVe4mA0uqHe7mcib25hvxlAw/640?wx_fmt=png&amp;from=appmsg" data-ratio="1.2010582010582012" data-type="png" data-w="567" data-imgfileid="100056717" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wx8HH4R5qMk8mu6icA8Mzb9k0UR6GMwiaMraQARfibwWvKCAt4fOxgMQFDVe4mA0uqHe7mcib25hvxlAw/640?wx_fmt=png&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">本次学习来自官网的教程：</span></p><blockquote><span></span><p><span leaf="">https://cell2location.readthedocs.io/en/latest/notebooks/cell2location_tutorial.html</span></p></blockquote><h2 data-cacheurl="" data-remoteid="" data-tool="mdnice编辑器"><span></span><span><span leaf="">0.环境配置</span></span><span></span></h2><p data-tool="mdnice编辑器"><span leaf="">conda 安装：</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">conda create -n cell2location -y python=3.9</span><span leaf=""><br></span><span leaf="">conda activate cell2location</span><span leaf=""><br></span><span leaf="">pip install -i https://mirrors.tuna.tsinghua.edu.cn/pypi/web/simple jupyterlab </span><span leaf=""><br></span><span><span leaf=""># 永久设置镜像</span></span><span leaf=""><br></span><span leaf="">pip config </span><span><span leaf="">set</span></span><span leaf=""> global.index-url https://pypi.mirrors.ustc.edu.cn/simple/</span><span leaf=""><br></span><span><span leaf=""># /home/zhangj/.config/pip/pip.conf</span></span><span leaf=""><br></span><span leaf="">pip install scanpy</span><span leaf=""><br></span><span leaf="">pip install cell2location[tutorials]</span><span leaf=""><br></span><span leaf="">pip install pandas==2.1.1</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">加载：</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">import</span></span><span leaf=""> os</span><span leaf=""><br></span><span><span leaf=""># this line should go before importing cell2location</span></span><span leaf=""><br></span><span leaf="">os.environ[</span><span><span leaf="">"THEANO_FLAGS"</span></span><span leaf="">] = </span><span><span leaf="">'device=cuda,floatX=float32,force_device=True'</span></span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># Loading packages</span></span><span leaf=""><br></span><span><span leaf="">import</span></span><span leaf=""> sys</span><span leaf=""><br></span><span><span leaf="">import</span></span><span leaf=""> scanpy </span><span><span leaf="">as</span></span><span leaf=""> sc</span><span leaf=""><br></span><span><span leaf="">import</span></span><span leaf=""> numpy </span><span><span leaf="">as</span></span><span leaf=""> np</span><span leaf=""><br></span><span><span leaf="">import</span></span><span leaf=""> matplotlib.pyplot </span><span><span leaf="">as</span></span><span leaf=""> plt</span><span leaf=""><br></span><span><span leaf="">import</span></span><span leaf=""> matplotlib </span><span><span leaf="">as</span></span><span leaf=""> mpl</span><span leaf=""><br></span><span><span leaf="">import</span></span><span leaf=""> cell2location</span><span leaf=""><br></span><span><span leaf="">from</span></span><span leaf=""> matplotlib </span><span><span leaf="">import</span></span><span leaf=""> rcParams</span><span leaf=""><br></span><span leaf="">rcParams[</span><span><span leaf="">'pdf.fonttype'</span></span><span leaf="">] = </span><span><span leaf="">42</span></span><span leaf=""> </span><span><span leaf=""># enables correct plotting of text for PDFs</span></span><span leaf=""><br></span></code></pre><h2 data-cacheurl="" data-remoteid="" data-tool="mdnice编辑器"><span></span><span><span leaf="">1.加载Visium空间数据</span></span><span></span></h2><h3 data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><span></span><span><span leaf="">1.1 数据介绍</span></span><span></span></h3><p data-tool="mdnice编辑器"><span leaf="">参考的单细胞数据来自人的次级淋巴器官包括34种细胞类型，空间数据来自10X Genomics的人类淋巴结Visium dataset，为10X Space Ranger软件的标准输出文件，来自 </span><span leaf="">Kleshchevnikov et al (section 4, Fig 4)</span><span leaf="">。</span></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wx8HH4R5qMk8mu6icA8Mzb9kGnXQIhnmNEENAyWW6g0OyOvHqAv9lznOpsicfKsD5Imva0eadUzMk7A/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.30537459283387625" data-type="png" data-w="1228" data-imgfileid="100056716" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wx8HH4R5qMk8mu6icA8Mzb9kGnXQIhnmNEENAyWW6g0OyOvHqAv9lznOpsicfKsD5Imva0eadUzMk7A/640?wx_fmt=png&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">Visium空间数据下载地址为10X Genomics 官网: <span textstyle="">https://support.10xgenomics.com/spatial-gene-expression/datasets/1.0.0/V1_Human_Lymph_Node</span>。格式为10X Space Ranger软件的输出，去网站下载到本地后导入：</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">adata_vis = sc.read_visium(</span><span><span leaf="">"../data/V1_Human_Lymph_Node/1.0.0/"</span></span><span leaf="">, count_file=</span><span><span leaf="">'filtered_feature_bc_matrix.h5'</span></span><span leaf="">, load_images=</span><span><span leaf="">True</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">adata_vis</span><span leaf=""><br></span><span leaf="">adata_vis.var_names</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># rename genes to ENSEMBL ID for correct matching between single cell and spatial data </span></span><span leaf=""><br></span><span leaf="">adata_vis.obs[</span><span><span leaf="">'sample'</span></span><span leaf="">] = list(adata_vis.uns[</span><span><span leaf="">'spatial'</span></span><span leaf="">].keys())[</span><span><span leaf="">0</span></span><span leaf="">]</span><span leaf=""><br></span><span leaf="">adata_vis.var[</span><span><span leaf="">'SYMBOL'</span></span><span leaf="">] = adata_vis.var_names</span><span leaf=""><br></span><span leaf="">adata_vis.var.set_index(</span><span><span leaf="">'gene_ids'</span></span><span leaf="">, drop=</span><span><span leaf="">True</span></span><span leaf="">, inplace=</span><span><span leaf="">True</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">adata_vis</span><span leaf=""><br></span><span leaf="">adata_vis.var_names</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">注意，数据 data/V1_Human_Lymph_Node/1.0.0/ 目录结构为：</span></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wx8HH4R5qMk8mu6icA8Mzb9kX2I9y4lEg04tVlP7dGCsOo99R9yZTlXmXEdmP4TwGCcsiasjI60hAiaA/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.537467700258398" data-type="png" data-w="387" data-imgfileid="100056713" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wx8HH4R5qMk8mu6icA8Mzb9kX2I9y4lEg04tVlP7dGCsOo99R9yZTlXmXEdmP4TwGCcsiasjI60hAiaA/640?wx_fmt=png&amp;from=appmsg"></span></figure><h3 data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><span></span><span><span leaf="">1.2 去除线粒体基因</span></span><span></span></h3><p data-tool="mdnice编辑器"><span leaf="">线粒体编码的基因(基因名称以前缀 MT-或 MT-开头)与空间mapping无关，因为它们的表达代表单细胞和细胞核数据中的技术artifacts，而不是线粒体的生物丰度。然而，这些基因在每个spot组成15-40% 的 mRNA 。因此，为了避免绘制artifacts，<span textstyle="">强烈建议去除线粒体基因</span>。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># find mitochondria-encoded (MT) genes</span></span><span leaf=""><br></span><span leaf="">adata_vis.var[</span><span><span leaf="">'MT_gene'</span></span><span leaf="">] = [gene.startswith(</span><span><span leaf="">'MT-'</span></span><span leaf="">) </span><span><span leaf="">for</span></span><span leaf=""> gene </span><span><span leaf="">in</span></span><span leaf=""> adata_vis.var[</span><span><span leaf="">'SYMBOL'</span></span><span leaf="">]]</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># remove MT genes for spatial mapping (keeping their counts in the object)</span></span><span leaf=""><br></span><span leaf="">adata_vis.obsm[</span><span><span leaf="">'MT'</span></span><span leaf="">] = adata_vis[:, adata_vis.var[</span><span><span leaf="">'MT_gene'</span></span><span leaf="">].values].X.toarray()</span><span leaf=""><br></span><span leaf="">adata_vis = adata_vis[:, ~adata_vis.var[</span><span><span leaf="">'MT_gene'</span></span><span leaf="">].values]</span><span leaf=""><br></span></code></pre><h2 data-cacheurl="" data-remoteid="" data-tool="mdnice编辑器"><span></span><span><span leaf="">2.加载单细胞参考集数据</span></span><span></span></h2><p data-tool="mdnice编辑器"><span leaf="">由于患者供体年龄的原因，已发表的淋巴结scRNA-seq数据集通常缺乏生发中心相关免疫细胞群的充分代表。因此，<span textstyle="">作者将综合淋巴结、脾脏和扁桃体的scRNA-seq数据集纳入单细胞参考，以确保捕获了空间转录组数据集中可能存在的免疫细胞状态的全部多样性</span>。</span></p><h3 data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><span></span><span><span leaf="">2.1 导入单细胞数据</span></span><span></span></h3><p data-tool="mdnice编辑器"><span leaf="">下载地址：https://cell2location.cog.sanger.ac.uk/paper/integrated_lymphoid_organ_scrna/RegressionNBV4Torch_57covariates_73260cells_10237genes/sc.h5ad</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># Read scrna data</span></span><span leaf=""><br></span><span leaf="">adata_ref = sc.read(</span><span><span leaf="">'../data/V1_Human_Lymph_Node/1.0.0/sc.h5ad'</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 将数据的genes转为ENSEMBL ID，与空间数据对应同样的id</span></span><span leaf=""><br></span><span leaf="">adata_ref.var[</span><span><span leaf="">'SYMBOL'</span></span><span leaf="">] = adata_ref.var.index</span><span leaf=""><br></span><span><span leaf=""># rename 'GeneID-2' as necessary for your data</span></span><span leaf=""><br></span><span leaf="">adata_ref.var.set_index(</span><span><span leaf="">'GeneID-2'</span></span><span leaf="">, drop=</span><span><span leaf="">True</span></span><span leaf="">, inplace=</span><span><span leaf="">True</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">adata_ref.var</span><span leaf=""><br></span></code></pre><h3 data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><span></span><span><span leaf="">2.2 very permissive genes选择</span></span><span></span></h3><p data-tool="mdnice编辑器"><span leaf="">在估计参考细胞类型特征之前，建议进行very permissive genes选择。<span textstyle="">与标准的高可变基因选择相比，这种方法保留了罕见标记基因，同时删除了大多数无信息的基因。</span>默认参数为：</span><code><span leaf="">cell_count_cutoff=5, cell_percentage_cutoff2=0.03, nonz_mean_cutoff=1.12</span></code></p><p data-tool="mdnice编辑器"><span leaf="">在实战项目过程中，可以适当增加这些阈值以便过滤更多基因，推荐 </span><code><span leaf="">cell_count_cutoff=5</span></code><span leaf="">，</span><code><span leaf="">cell_percentage_cutoff2</span></code><span leaf="">和</span><code><span leaf="">nonz_mean_cutoff</span></code><span leaf="">可以适当增加到基因个数为 8k-16k 个基因。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">from cell2location.utils.filtering import filter_genes</span><span leaf=""><br></span><span leaf="">selected = filter_genes(adata_ref, cell_count_cutoff=5, cell_percentage_cutoff2=0.03, nonz_mean_cutoff=1.12)</span><span leaf=""><br></span><span><span leaf=""># filter the object</span></span><span leaf=""><br></span><span leaf="">adata_ref = adata_ref[:, selected].copy()</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">过滤掉的基因示意图：</span></p><ul><li><section><p><span leaf="">每一个点表示一个基因</span></p></section></li><li><section><p><span leaf="">x轴：检测到该基因的细胞中的平均 RNA count值</span></p></section></li><li><section><p><span leaf="">y轴：表达该基因的细胞数量</span></p></section></li></ul><p data-tool="mdnice编辑器"><span leaf="">橙色矩形突出显示了基于x轴和y轴的组合而排除的基因。</span></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wx8HH4R5qMk8mu6icA8Mzb9kSTnDSHUDsOM7k4yZJgL73YiceauRibeqZYjjjyjphibyn06PwDI020nHg/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.78719723183391" data-type="png" data-w="578" data-imgfileid="100056715" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wx8HH4R5qMk8mu6icA8Mzb9kSTnDSHUDsOM7k4yZJgL73YiceauRibeqZYjjjyjphibyn06PwDI020nHg/640?wx_fmt=png&amp;from=appmsg"></span></figure><h2 data-cacheurl="" data-remoteid="" data-tool="mdnice编辑器"><span></span><span><span leaf="">3.参考细胞类型signatures评估</span></span><span></span></h2><p data-tool="mdnice编辑器"><span leaf="">根据scRNA-seq数据估计特征，考虑到批此效应，使用负二项回归模型进行分析。</span></p><h3 data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><span></span><span><span leaf="">3.1 首先，为回归模型准备一个数据对象:</span></span><span></span></h3><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># prepare anndata for the regression model</span></span><span leaf=""><br></span><span leaf="">cell2location.models.RegressionModel.setup_anndata(adata=adata_ref,</span><span leaf=""><br></span><span leaf="">                        </span><span><span leaf=""># 10X reaction / sample / batch</span></span><span leaf=""><br></span><span leaf="">                        batch_key=</span><span><span leaf="">'Sample'</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">                        </span><span><span leaf=""># cell type, covariate used for constructing signatures</span></span><span leaf=""><br></span><span leaf="">                        labels_key=</span><span><span leaf="">'Subset'</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">                        </span><span><span leaf=""># multiplicative technical effects (platform, 3' vs 5', donor effect)</span></span><span leaf=""><br></span><span leaf="">                        categorical_covariate_keys=[</span><span><span leaf="">'Method'</span></span><span leaf="">]</span><span leaf=""><br></span><span leaf="">                       )</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># create the regression model</span></span><span leaf=""><br></span><span leaf="">from cell2location.models import RegressionModel</span><span leaf=""><br></span><span leaf="">mod = RegressionModel(adata_ref)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># view anndata_setup as a sanity check</span></span><span leaf=""><br></span><span leaf="">mod.view_anndata_setup()</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">显示的内容如下：</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">Anndata setup with scvi-tools version 1.1.6.post2.</span><span leaf=""><br></span><span leaf="">Setup via `RegressionModel.setup_anndata` with arguments:</span><span leaf=""><br></span><span leaf="">{</span><span leaf=""><br></span><span leaf="">│   </span><span><span leaf="">'layer'</span></span><span leaf="">: None,</span><span leaf=""><br></span><span leaf="">│   </span><span><span leaf="">'batch_key'</span></span><span leaf="">: </span><span><span leaf="">'Sample'</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">│   </span><span><span leaf="">'labels_key'</span></span><span leaf="">: </span><span><span leaf="">'Subset'</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">│   </span><span><span leaf="">'categorical_covariate_keys'</span></span><span leaf="">: [</span><span><span leaf="">'Method'</span></span><span leaf="">],</span><span leaf=""><br></span><span leaf="">│   </span><span><span leaf="">'continuous_covariate_keys'</span></span><span leaf="">: None</span><span leaf=""><br></span><span leaf="">}</span><span leaf=""><br></span><span leaf="">         Summary Statistics         </span><span leaf=""><br></span><span leaf="">┏━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━┓</span><span leaf=""><br></span><span leaf="">┃     Summary Stat Key     ┃ Value ┃</span><span leaf=""><br></span><span leaf="">┡━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━┩</span><span leaf=""><br></span><span leaf="">│         n_batch          │  23   │</span><span leaf=""><br></span><span leaf="">│         n_cells          │ 73260 │</span><span leaf=""><br></span><span leaf="">│ n_extra_categorical_covs │   1   │</span><span leaf=""><br></span><span leaf="">│ n_extra_continuous_covs  │   0   │</span><span leaf=""><br></span><span leaf="">│         n_labels         │  34   │</span><span leaf=""><br></span><span leaf="">│          n_vars          │ 10237 │</span><span leaf=""><br></span><span leaf="">└──────────────────────────┴───────┘</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">                           Data Registry                             </span><span leaf=""><br></span><span leaf="">┏━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓</span><span leaf=""><br></span><span leaf="">┃      Registry Key      ┃            scvi-tools Location             ┃</span><span leaf=""><br></span><span leaf="">┡━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩</span><span leaf=""><br></span><span leaf="">│           X            │                  adata.X                   │</span><span leaf=""><br></span><span leaf="">│         batch          │          adata.obs[</span><span><span leaf="">'_scvi_batch'</span></span><span leaf="">]          │</span><span leaf=""><br></span><span leaf="">│ extra_categorical_covs │ adata.obsm[</span><span><span leaf="">'_scvi_extra_categorical_covs'</span></span><span leaf="">] │</span><span leaf=""><br></span><span leaf="">│         ind_x          │           adata.obs[</span><span><span leaf="">'_indices'</span></span><span leaf="">]            │</span><span leaf=""><br></span><span leaf="">│         labels         │         adata.obs[</span><span><span leaf="">'_scvi_labels'</span></span><span leaf="">]          │</span><span leaf=""><br></span><span leaf="">└────────────────────────┴────────────────────────────────────────────┘</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">                     batch State Registry                         </span><span leaf=""><br></span><span leaf="">┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━┓</span><span leaf=""><br></span><span leaf="">┃   Source Location   ┃       Categories       ┃ scvi-tools Encoding ┃</span><span leaf=""><br></span><span leaf="">┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━┩</span><span leaf=""><br></span><span leaf="">│ adata.obs[</span><span><span leaf="">'Sample'</span></span><span leaf="">] │    4861STDY7135913     │          0          │</span><span leaf=""><br></span><span leaf="">│                     │    4861STDY7135914     │          1          │</span><span leaf=""><br></span><span leaf="">│                     │    4861STDY7208412     │          2          │</span><span leaf=""><br></span><span leaf="">│                     │    4861STDY7208413     │          3          │</span><span leaf=""><br></span><span leaf="">│                     │    4861STDY7462253     │          4          │</span><span leaf=""><br></span><span leaf="">│                     │    4861STDY7462254     │          5          │</span><span leaf=""><br></span><span leaf="">│                     │    4861STDY7462255     │          6          │</span><span leaf=""><br></span><span leaf="">│                     │    4861STDY7462256     │          7          │</span><span leaf=""><br></span><span leaf="">│                     │    4861STDY7528597     │          8          │</span><span leaf=""><br></span><span leaf="">│                     │    4861STDY7528598     │          9          │</span><span leaf=""><br></span><span leaf="">│                     │    4861STDY7528599     │         10          │</span><span leaf=""><br></span><span leaf="">│                     │    4861STDY7528600     │         11          │</span><span leaf=""><br></span><span leaf="">│                     │      BCP002_Total      │         12          │</span><span leaf=""><br></span><span leaf="">│                     │      BCP003_Total      │         13          │</span><span leaf=""><br></span><span leaf="">│                     │      BCP004_Total      │         14          │</span><span leaf=""><br></span><span leaf="">│                     │      BCP005_Total      │         15          │</span><span leaf=""><br></span><span leaf="">│                     │      BCP006_Total      │         16          │</span><span leaf=""><br></span><span leaf="">│                     │      BCP008_Total      │         17          │</span><span leaf=""><br></span><span leaf="">│                     │      BCP009_Total      │         18          │</span><span leaf=""><br></span><span leaf="">│                     │ Human_colon_16S7255677 │         19          │</span><span leaf=""><br></span><span leaf="">│                     │ Human_colon_16S7255678 │         20          │</span><span leaf=""><br></span><span leaf="">│                     │ Human_colon_16S8000484 │         21          │</span><span leaf=""><br></span><span leaf="">│                     │      Pan_T7935494      │         22          │</span><span leaf=""><br></span><span leaf="">└─────────────────────┴────────────────────────┴─────────────────────┘</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">                  labels State Registry                      </span><span leaf=""><br></span><span leaf="">┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━┓</span><span leaf=""><br></span><span leaf="">┃   Source Location   ┃    Categories    ┃ scvi-tools Encoding ┃</span><span leaf=""><br></span><span leaf="">┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━┩</span><span leaf=""><br></span><span leaf="">│ adata.obs[</span><span><span leaf="">'Subset'</span></span><span leaf="">] │    B_Cycling     │          0          │</span><span leaf=""><br></span><span leaf="">│                     │     B_GC_DZ      │          1          │</span><span leaf=""><br></span><span leaf="">│                     │     B_GC_LZ      │          2          │</span><span leaf=""><br></span><span leaf="">│                     │    B_GC_prePB    │          3          │</span><span leaf=""><br></span><span leaf="">│                     │      B_IFN       │          4          │</span><span leaf=""><br></span><span leaf="">│                     │   B_activated    │          5          │</span><span leaf=""><br></span><span leaf="">│                     │      B_mem       │          6          │</span><span leaf=""><br></span><span leaf="">│                     │     B_naive      │          7          │</span><span leaf=""><br></span><span leaf="">│                     │     B_plasma     │          8          │</span><span leaf=""><br></span><span leaf="">│                     │     B_preGC      │          9          │</span><span leaf=""><br></span><span leaf="">│                     │     DC_CCR7+     │         10          │</span><span leaf=""><br></span><span leaf="">│                     │     DC_cDC1      │         11          │</span><span leaf=""><br></span><span leaf="">│                     │     DC_cDC2      │         12          │</span><span leaf=""><br></span><span leaf="">│                     │      DC_pDC      │         13          │</span><span leaf=""><br></span><span leaf="">│                     │       Endo       │         14          │</span><span leaf=""><br></span><span leaf="">│                     │       FDC        │         15          │</span><span leaf=""><br></span><span leaf="">│                     │       ILC        │         16          │</span><span leaf=""><br></span><span leaf="">│                     │  Macrophages_M1  │         17          │</span><span leaf=""><br></span><span leaf="">│                     │  Macrophages_M2  │         18          │</span><span leaf=""><br></span><span leaf="">│                     │       Mast       │         19          │</span><span leaf=""><br></span><span leaf="">│                     │    Monocytes     │         20          │</span><span leaf=""><br></span><span leaf="">│                     │        NK        │         21          │</span><span leaf=""><br></span><span leaf="">│                     │       NKT        │         22          │</span><span leaf=""><br></span><span leaf="">│                     │      T_CD4+      │         23          │</span><span leaf=""><br></span><span leaf="">│                     │    T_CD4+_TfH    │         24          │</span><span leaf=""><br></span><span leaf="">│                     │  T_CD4+_TfH_GC   │         25          │</span><span leaf=""><br></span><span leaf="">│                     │   T_CD4+_naive   │         26          │</span><span leaf=""><br></span><span leaf="">│                     │  T_CD8+_CD161+   │         27          │</span><span leaf=""><br></span><span leaf="">│                     │ T_CD8+_cytotoxic │         28          │</span><span leaf=""><br></span><span leaf="">│                     │   T_CD8+_naive   │         29          │</span><span leaf=""><br></span><span leaf="">│                     │     T_TIM3+      │         30          │</span><span leaf=""><br></span><span leaf="">│                     │      T_TfR       │         31          │</span><span leaf=""><br></span><span leaf="">│                     │      T_Treg      │         32          │</span><span leaf=""><br></span><span leaf="">│                     │       VSMC       │         33          │</span><span leaf=""><br></span><span leaf="">└─────────────────────┴──────────────────┴─────────────────────┘</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">extra_categorical_covs State Registry           </span><span leaf=""><br></span><span leaf="">┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━┓</span><span leaf=""><br></span><span leaf="">┃   Source Location   ┃ Categories ┃ scvi-tools Encoding ┃</span><span leaf=""><br></span><span leaf="">┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━┩</span><span leaf=""><br></span><span leaf="">│ adata.obs[</span><span><span leaf="">'Method'</span></span><span leaf="">] │    3GEX    │          0          │</span><span leaf=""><br></span><span leaf="">│                     │    5GEX    │          1          │</span><span leaf=""><br></span><span leaf="">│                     │            │                     │</span><span leaf=""><br></span><span leaf="">└─────────────────────┴────────────┴─────────────────────┘</span><span leaf=""><br></span></code></pre><h3 data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><span></span><span><span leaf="">3.2 训练模型</span></span><span></span></h3><p data-tool="mdnice编辑器"><span leaf="">estimate the reference cell type signatures，<span textstyle="">这一步骤没有GPU将非常耗时</span></span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf="">## 如果你有GPU超算，需要在前面导入cell2location模块的时候设置</span></span><span leaf=""><br></span><span leaf="">import os</span><span leaf=""><br></span><span><span leaf=""># this line should go before importing cell2location</span></span><span leaf=""><br></span><span leaf="">os.environ[</span><span><span leaf="">"THEANO_FLAGS"</span></span><span leaf="">] = </span><span><span leaf="">'device=cuda,floatX=float32,force_device=True'</span></span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 训练模型</span></span><span leaf=""><br></span><span leaf="">mod.train(max_epochs=250)</span><span leaf=""><br></span><span><span leaf=""><span textstyle=""># 我这里使用GPU，运行了18m24s</span></span></span><span leaf=""><br></span></code></pre><h3 data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><span></span><span><span leaf="">3.3 检查模型是否需要更多的训练</span></span><span></span></h3><p data-tool="mdnice编辑器"><span leaf="">可以绘制训练期间的 ELBO 损失历史，删除前20个 epoch。这个图在训练结束时应该有一个下降的趋势并趋于平稳。如果 max_epochs 仍在减小，则需要增加max_epochs：</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">mod.plot_history(20)</span><span leaf=""><br></span><span leaf="">plt.savefig(results_folder+</span><span><span leaf="">'01-mod.plot_history.png'</span></span><span leaf="">)  </span><span><span leaf=""># 保存为PNG格式</span></span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">训练结果图：</span></p><p data-tool="mdnice编辑器"><span leaf="">横轴为epochs，纵轴为ELBO loss，<span textstyle="">这个图在训练结束时应该有一个下降的趋势并趋于平稳</span>。</span></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wx8HH4R5qMk8mu6icA8Mzb9kbJLQja5SAANms1FtibkibRWPBiaSvcNtBtIz9OuAOkr5Y3YBCkSZt4xWg/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.746031746031746" data-type="png" data-w="630" data-imgfileid="100056718" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wx8HH4R5qMk8mu6icA8Mzb9kbJLQja5SAANms1FtibkibRWPBiaSvcNtBtIz9OuAOkr5Y3YBCkSZt4xWg/640?wx_fmt=png&amp;from=appmsg"></span></figure><h3 data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><span></span><span><span leaf="">3.4 保存训练后估计的细胞丰度</span></span><span></span></h3><p data-tool="mdnice编辑器"><span leaf="">输出训练的结果：</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># In this section, we export the estimated cell abundance (summary of the posterior distribution).</span></span><span leaf=""><br></span><span leaf="">adata_ref = mod.export_posterior(</span><span leaf=""><br></span><span leaf="">    adata_ref, sample_kwargs={</span><span><span leaf="">'num_samples'</span></span><span leaf="">: 1000, </span><span><span leaf="">'batch_size'</span></span><span leaf="">: 2500}</span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># Save model</span></span><span leaf=""><br></span><span leaf="">mod.save(</span><span><span leaf="">'./reference_signatures/'</span></span><span leaf="">, overwrite=True)</span><span leaf=""><br></span><span leaf="">adata_file = </span><span><span leaf="">"./reference_signatures/sc.h5ad"</span></span><span leaf=""><br></span><span leaf="">adata_ref.write(adata_file)</span><span leaf=""><br></span><span leaf="">adata_file</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">保存的数据后续还可以重新加载回来：</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">adata_file = </span><span><span leaf="">"./reference_signatures/sc.h5ad"</span></span><span leaf=""><br></span><span leaf="">adata_ref = sc.read_h5ad(adata_file)</span><span leaf=""><br></span><span leaf="">mod = cell2location.models.RegressionModel.load(</span><span><span leaf="">'./reference_signatures/'</span></span><span leaf="">, adata_ref)</span><span leaf=""><br></span></code></pre><h3 data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><span></span><span><span leaf="">3.4 检查评估结果</span></span><span></span></h3><p data-tool="mdnice编辑器"><span leaf="">评估标准：</span></p><p data-tool="mdnice编辑器"><span leaf="">1.Reconstruction accuracy图：这个二维直方图的大部分观测值应该沿着有噪声的对角线</span></p><p data-tool="mdnice编辑器"><span leaf="">2.图2：由于批处理效应，估计的表达特征与每个聚类的平均表达特征不同。对于不受批处理影响的scRNA-seq数据集(本数据集)，可以使用聚类平均表达来代替使用模型估计signature特征。当这个图与对角线图非常不同时(例如，y轴上的值非常低，到处都是密度)，这表明特征估计存在问题。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">mod.plot_QC()</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">评估图：</span></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wx8HH4R5qMk8mu6icA8Mzb9koatvG3LiaGzEs74rov8LWLqfC4HeCiaDLESWn1IwoicjLjLvUVgXq6M8Q/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.9233791748526523" data-type="png" data-w="509" data-imgfileid="100056719" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wx8HH4R5qMk8mu6icA8Mzb9koatvG3LiaGzEs74rov8LWLqfC4HeCiaDLESWn1IwoicjLjLvUVgXq6M8Q/640?wx_fmt=png&amp;from=appmsg"></span></figure><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wx8HH4R5qMk8mu6icA8Mzb9kQbpK7TO2IkibarQLlibEyymRvJmdydsZsBUx52fGjQLju9Acia8UKiaDfA/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.7619047619047619" data-type="png" data-w="567" data-imgfileid="100056720" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wx8HH4R5qMk8mu6icA8Mzb9kQbpK7TO2IkibarQLlibEyymRvJmdydsZsBUx52fGjQLju9Acia8UKiaDfA/640?wx_fmt=png&amp;from=appmsg"></span></figure><h3 data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><span></span><span><span leaf="">3.5 提取参考细胞类型特征</span></span><span></span></h3><p data-tool="mdnice编辑器"><span leaf="">提取参考细胞类型特征为pd.DataFrame：空间转录组数据的细胞类型映射只需要估计得到的signatures</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># export estimated expression in each cluster</span></span><span leaf=""><br></span><span><span leaf="">if</span></span><span leaf=""> </span><span><span leaf="">'means_per_cluster_mu_fg'</span></span><span leaf=""> </span><span><span leaf="">in</span></span><span leaf=""> adata_ref.varm.keys():</span><span leaf=""><br></span><span leaf="">    inf_aver = adata_ref.varm[</span><span><span leaf="">'means_per_cluster_mu_fg'</span></span><span leaf="">][[f</span><span><span leaf="">'means_per_cluster_mu_fg_{i}'</span></span><span leaf=""><br></span><span leaf="">                            </span><span><span leaf="">for</span></span><span leaf=""> i </span><span><span leaf="">in</span></span><span leaf=""> adata_ref.uns[</span><span><span leaf="">'mod'</span></span><span leaf="">][</span><span><span leaf="">'factor_names'</span></span><span leaf="">]]].copy()</span><span leaf=""><br></span><span><span leaf="">else</span></span><span leaf="">:</span><span leaf=""><br></span><span leaf="">    inf_aver = adata_ref.var[[f</span><span><span leaf="">'means_per_cluster_mu_fg_{i}'</span></span><span leaf=""><br></span><span leaf="">                            </span><span><span leaf="">for</span></span><span leaf=""> i </span><span><span leaf="">in</span></span><span leaf=""> adata_ref.uns[</span><span><span leaf="">'mod'</span></span><span leaf="">][</span><span><span leaf="">'factor_names'</span></span><span leaf="">]]].copy()</span><span leaf=""><br></span><span leaf="">                            </span><span leaf=""><br></span><span leaf="">inf_aver.columns = adata_ref.uns[</span><span><span leaf="">'mod'</span></span><span leaf="">][</span><span><span leaf="">'factor_names'</span></span><span leaf="">]</span><span leaf=""><br></span><span leaf="">inf_aver.iloc[0:12, 0:12]</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">这个就是用单细胞数据得到的细胞类型signature了，用于空转数据的解卷积。</span></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wx8HH4R5qMk8mu6icA8Mzb9kFe2E9gPk0bBQDbHWibVj5gSSQdezF9aJ32scxiaJDAiaE2Pj4MQRZic4yg/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.35359116022099446" data-type="png" data-w="1086" data-imgfileid="100056721" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wx8HH4R5qMk8mu6icA8Mzb9kFe2E9gPk0bBQDbHWibVj5gSSQdezF9aJ32scxiaJDAiaE2Pj4MQRZic4yg/640?wx_fmt=png&amp;from=appmsg"></span></figure><h2 data-cacheurl="" data-remoteid="" data-tool="mdnice编辑器"><span></span><span><span leaf="">4.运行Cell2location</span></span><span></span></h2><p data-tool="mdnice编辑器"><span leaf="">利用单细胞signature矩阵，对空转数据进行注释，首先提取单细胞与空转数据共同基因：</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># find shared genes and subset both anndata and reference signatures</span></span><span leaf=""><br></span><span leaf="">intersect = np.intersect1d(adata_vis.var_names, inf_aver.index)</span><span leaf=""><br></span><span leaf="">adata_vis = adata_vis[:, intersect].copy()</span><span leaf=""><br></span><span leaf="">inf_aver = inf_aver.loc[intersect, :].copy()</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># prepare anndata for cell2location model</span></span><span leaf=""><br></span><span leaf="">cell2location.models.Cell2location.setup_anndata(adata=adata_vis, batch_key=</span><span><span leaf="">"sample"</span></span><span leaf="">)</span><span leaf=""><br></span></code></pre><h3 data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><span></span><span><span leaf="">4.1 定义参数值</span></span><span></span></h3><p data-tool="mdnice编辑器"><span leaf="">在进行空转数据注释之前，<span textstyle="">需要定义两个比较重要的参数</span></span></p><p data-tool="mdnice编辑器"><strong><span leaf="">N_cells_per_location</span></strong><span leaf="">：每个spot 的细胞数，这个可以通过配对的 histology images进行评估，或者 通过squdipy 进行细胞分割，然后估计细胞数、或者通过 10X loupe browser 手动数</span></p><p data-tool="mdnice编辑器"><strong><span leaf="">detection_alpha</span></strong><span leaf="">：建议用 20，因为从已发表的研究来看，人体组织在实验中受到的技术影响较大。如果是使用的小鼠等受技术影响低的数据集，那可以用 200。当然了，最好是两个都试试。</span></p><p data-tool="mdnice编辑器"><span leaf="">更多详细的说明参考：</span><span leaf="">the flow diagram and the note，</span><span leaf="">https://github.com/BayraktarLab/cell2location/blob/master/docs/images/Note_on_selecting_hyperparameters.pdf</span></p><p data-tool="mdnice编辑器"><span leaf="">本教程：</span><code><span leaf="">N_cells_per_location=30，detection_alpha=20</span></code></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># create and train the model</span></span><span leaf=""><br></span><span leaf="">mod = cell2location.models.Cell2location(</span><span leaf=""><br></span><span leaf="">    adata_vis, cell_state_df=inf_aver,</span><span leaf=""><br></span><span leaf="">    </span><span><span leaf=""># the expected average cell abundance: tissue-dependent</span></span><span leaf=""><br></span><span leaf="">    </span><span><span leaf=""># hyper-prior which can be estimated from paired histology:</span></span><span leaf=""><br></span><span leaf="">    N_cells_per_location=30,</span><span leaf=""><br></span><span leaf="">    </span><span><span leaf=""># hyperparameter controlling normalisation of</span></span><span leaf=""><br></span><span leaf="">    </span><span><span leaf=""># within-experiment variation in RNA detection:</span></span><span leaf=""><br></span><span leaf="">    detection_alpha=20</span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">mod.view_anndata_setup()</span><span leaf=""><br></span></code></pre><h3 data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><span></span><span><span leaf="">4.2 Training cell2location</span></span><span></span></h3><p data-tool="mdnice编辑器"><span leaf=""><span textstyle="">此步骤也非常耗时：在使用GPU的情况下，总共运行了42m 34.3s</span></span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">mod.train(max_epochs=30000,</span><span leaf=""><br></span><span leaf="">          </span><span><span leaf=""># train using full data (batch_size=None)</span></span><span leaf=""><br></span><span leaf="">          batch_size=None,</span><span leaf=""><br></span><span leaf="">          </span><span><span leaf=""># use all data points in training because</span></span><span leaf=""><br></span><span leaf="">          </span><span><span leaf=""># we need to estimate cell abundance at all locations</span></span><span leaf=""><br></span><span leaf="">          train_size=1</span><span leaf=""><br></span><span leaf="">         )</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># plot ELBO loss history during training, removing first 100 epochs from the plot</span></span><span leaf=""><br></span><span leaf="">mod.plot_history(1000)</span><span leaf=""><br></span><span leaf="">plt.legend(labels=[</span><span><span leaf="">'full data training'</span></span><span leaf="">]);</span><span leaf=""><br></span><span leaf="">plt.savefig(</span><span><span leaf="">'03-mod.plot_history.png'</span></span><span leaf="">)  </span><span><span leaf=""># 保存为PNG格式</span></span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">横轴为epochs，纵轴为ELBO loss，这个图在训练结束时应该有一个下降的趋势并趋于平稳。</span></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wx8HH4R5qMk8mu6icA8Mzb9k4vpz9LicRjs4L4O3rNWgNa8xlAgeMA6J4ia9iaHS4hGWHtv4Kgg7ibcv7A/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.746031746031746" data-type="png" data-w="630" data-imgfileid="100056722" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wx8HH4R5qMk8mu6icA8Mzb9k4vpz9LicRjs4L4O3rNWgNa8xlAgeMA6J4ia9iaHS4hGWHtv4Kgg7ibcv7A/640?wx_fmt=png&amp;from=appmsg"></span></figure><h3 data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><span></span><span><span leaf="">4.3 保存结果</span></span><span></span></h3><p data-tool="mdnice编辑器"><span leaf="">训练完之后，将每个 spot 估计的细胞丰度，保存到空转数据中。并且保存模型：</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># In this section, we export the estimated cell abundance (summary of the posterior distribution).</span></span><span leaf=""><br></span><span leaf="">adata_vis = mod.export_posterior(</span><span leaf=""><br></span><span leaf="">    adata_vis, sample_kwargs={</span><span><span leaf="">'num_samples'</span></span><span leaf="">: 1000, </span><span><span leaf="">'batch_size'</span></span><span leaf="">: mod.adata.n_obs}</span><span leaf=""><br></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">mod.save(</span><span><span leaf="">"./"</span></span><span leaf="">, overwrite=True)</span><span leaf=""><br></span><span><span leaf=""># 生成 model.pt</span></span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># Save anndata object with results</span></span><span leaf=""><br></span><span leaf="">adata_file = </span><span><span leaf="">"./sp.h5ad"</span></span><span leaf=""><br></span><span leaf="">adata_vis.write(adata_file)</span><span leaf=""><br></span><span leaf="">adata_file</span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">数据后续又可以被加载回来：</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">adata_file = </span><span><span leaf="">"./sp.h5ad"</span></span><span leaf=""><br></span><span leaf="">adata_vis = sc.read_h5ad(adata_file)</span><span leaf=""><br></span><span leaf="">mod = cell2location.models.Cell2location.load(</span><span><span leaf="">"./"</span></span><span leaf="">, adata_vis)</span><span leaf=""><br></span></code></pre><h3 data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><span></span><span><span leaf="">4.4 mapping质量评估</span></span><span></span></h3><p data-tool="mdnice编辑器"><span leaf="">检查重建精度以评估映射是否存在任何问题，图应该大致是对角线的。</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">mod.plot_QC()</span><span leaf=""><br></span><span leaf="">plt.savefig(</span><span><span leaf="">'04-mod.plot_QC.png'</span></span><span leaf="">)  </span><span><span leaf=""># 保存为PNG格式</span></span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">结果图：</span></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wx8HH4R5qMk8mu6icA8Mzb9kUGWQU8xaUP25GgNrOuYQsbSfAMk2R8yuic0m1sYJ5DLKSSXea2sASFg/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.9710743801652892" data-type="png" data-w="484" data-imgfileid="100056724" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wx8HH4R5qMk8mu6icA8Mzb9kUGWQU8xaUP25GgNrOuYQsbSfAMk2R8yuic0m1sYJ5DLKSSXea2sASFg/640?wx_fmt=png&amp;from=appmsg"></span></figure><h2 data-cacheurl="" data-remoteid="" data-tool="mdnice编辑器"><span></span><span><span leaf="">5.在空间坐标上可视化评估的细胞丰度</span></span><span></span></h2><p data-tool="mdnice编辑器"><span leaf="">现在终于到最后一步啦，可以看看空间切片上不同spot中的细胞类型。</span></p><h3 data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><span></span><span><span leaf="">5.1 查看具体的每个spot上的细胞丰度：</span></span><span></span></h3><p data-tool="mdnice编辑器"><span leaf="">预测的结果中有这几种：</span><code><span leaf="">means_cell_abundance_w_sf</span></code><span leaf="">, </span><code><span leaf="">stds_cell_abundance_w_sf</span></code><span leaf="">, </span><code><span leaf="">q05_cell_abundance_w_sf</span></code><span leaf="">, </span><code><span leaf="">q95_cell_abundance_w_sf</span></code></p><ul><li><section><code><span leaf="">q05_cell_abundance_w_sf</span></code><span leaf=""> ：表示每个位置上每种细胞类型的预期细胞数量，包括可能的小数（fractional values），即细胞类型的密度</span></section></li><li><section><strong><span leaf="">5% 分位数</span></strong><span leaf="">：这个值是后验分布的 5% 分位数，意味着模型对这个值有较高的置信度，认为至少有这么多细胞存在</span></section></li></ul><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">adata_vis.obsm</span><span leaf=""><br></span><span leaf="">adata_vis.obsm[</span><span><span leaf="">'q05_cell_abundance_w_sf'</span></span><span leaf="">]</span><span leaf=""><br></span><span leaf="">adata_vis.obsm[</span><span><span leaf="">'q95_cell_abundance_w_sf'</span></span><span leaf="">]</span><span leaf=""><br></span></code></pre><figure data-tool="mdnice编辑器"><span leaf=""><img data-imgfileid="100056726" data-ratio="0.34649421375085093" data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wx8HH4R5qMk8mu6icA8Mzb9kCKrLqp8BlLRXhRfZap3wSibHWf0xddBseEjaxibTTpnMJytk62cBbuBA/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1469" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wx8HH4R5qMk8mu6icA8Mzb9kCKrLqp8BlLRXhRfZap3wSibHWf0xddBseEjaxibTTpnMJytk62cBbuBA/640?wx_fmt=png&amp;from=appmsg"></span></figure><h3 data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><span></span><span><span leaf="">5.2 在空间上可视化</span></span><span></span></h3><p data-tool="mdnice编辑器"><span leaf="">选择 </span><code><span leaf="">q05_cell_abundance_w_sf</span></code><span leaf="">的结果：</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># add 5% quantile, representing confident cell abundance, 'at least this amount is present',</span></span><span leaf=""><br></span><span><span leaf=""># to adata.obs with nice names for plotting</span></span><span leaf=""><br></span><span leaf="">adata_vis.obs[adata_vis.uns[</span><span><span leaf="">'mod'</span></span><span leaf="">][</span><span><span leaf="">'factor_names'</span></span><span leaf="">]] = adata_vis.obsm[</span><span><span leaf="">'q05_cell_abundance_w_sf'</span></span><span leaf="">]</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># select one slide</span></span><span leaf=""><br></span><span leaf="">from cell2location.utils import select_slide</span><span leaf=""><br></span><span leaf="">slide = select_slide(adata_vis, </span><span><span leaf="">'V1_Human_Lymph_Node'</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># plot in spatial coordinates</span></span><span leaf=""><br></span><span leaf="">with mpl.rc_context({</span><span><span leaf="">'axes.facecolor'</span></span><span leaf="">:  </span><span><span leaf="">'black'</span></span><span leaf="">,</span><span><span leaf="">'figure.figsize'</span></span><span leaf="">: [4.5, 5]}):</span><span leaf=""><br></span><span leaf="">    sc.pl.spatial(slide, cmap=</span><span><span leaf="">'magma'</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">                  </span><span><span leaf=""># show first 8 cell types</span></span><span leaf=""><br></span><span leaf="">                  color=[</span><span><span leaf="">'B_Cycling'</span></span><span leaf="">, </span><span><span leaf="">'B_GC_LZ'</span></span><span leaf="">, </span><span><span leaf="">'T_CD4+_TfH_GC'</span></span><span leaf="">, </span><span><span leaf="">'FDC'</span></span><span leaf="">,</span><span><span leaf="">'B_naive'</span></span><span leaf="">, </span><span><span leaf="">'T_CD4+_naive'</span></span><span leaf="">, </span><span><span leaf="">'B_plasma'</span></span><span leaf="">, </span><span><span leaf="">'Endo'</span></span><span leaf="">],</span><span leaf=""><br></span><span leaf="">                  ncols=4, size=1.3,img_key=</span><span><span leaf="">'hires'</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">                  </span><span><span leaf=""># limit color scale at 99.2% quantile of cell abundance</span></span><span leaf=""><br></span><span leaf="">                  vmin=0, vmax=</span><span><span leaf="">'p99.2'</span></span><span leaf=""><br></span><span leaf="">                 )</span><span leaf=""><br></span><span leaf="">plt.savefig(</span><span><span leaf="">'05-sc.pl.spatial.png'</span></span><span leaf="">)  </span><span><span leaf=""># 保存为PNG格式</span></span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">这里选择了8种细胞进行可视化：</span></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-imgfileid="100056728" data-ratio="0.5063218390804598" data-src="https://mmbiz.qpic.cn/mmbiz_jpg/cZNhZQ6j4wx8HH4R5qMk8mu6icA8Mzb9kvFL4asbCwUIsLyIiaKjbSGkWwziaDkciaOicMjiblXgUCtPh99Eha60Cagg/640?wx_fmt=jpeg&amp;from=appmsg" data-type="jpeg" data-w="1740" src="https://mmbiz.qpic.cn/mmbiz_jpg/cZNhZQ6j4wx8HH4R5qMk8mu6icA8Mzb9kvFL4asbCwUIsLyIiaKjbSGkWwziaDkciaOicMjiblXgUCtPh99Eha60Cagg/640?wx_fmt=jpeg&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">也可以选择多个细胞类型展示在一张切片上：这里选择了三种细胞 T_CD4+_naive，B_naive，FDC 进行共定位</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span><span leaf=""># Now we use cell2location plotter that allows showing multiple cell types in one panel</span></span><span leaf=""><br></span><span leaf="">from cell2location.plt import plot_spatial</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># select up to 6 clusters</span></span><span leaf=""><br></span><span leaf="">clust_labels = [</span><span><span leaf="">'T_CD4+_naive'</span></span><span leaf="">, </span><span><span leaf="">'B_naive'</span></span><span leaf="">, </span><span><span leaf="">'FDC'</span></span><span leaf="">]</span><span leaf=""><br></span><span leaf="">clust_col = [</span><span><span leaf="">''</span></span><span leaf=""> + str(i) </span><span><span leaf="">for</span></span><span leaf=""> i </span><span><span leaf="">in</span></span><span leaf=""> clust_labels] </span><span><span leaf=""># in case column names differ from labels</span></span><span leaf=""><br></span><span leaf="">clust_col</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">slide = select_slide(adata_vis, </span><span><span leaf="">'V1_Human_Lymph_Node'</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">with mpl.rc_context({</span><span><span leaf="">'figure.figsize'</span></span><span leaf="">: (15, 15)}):</span><span leaf=""><br></span><span leaf="">    fig = plot_spatial(</span><span leaf=""><br></span><span leaf="">        adata=slide,</span><span leaf=""><br></span><span leaf="">        </span><span><span leaf=""># labels to show on a plot</span></span><span leaf=""><br></span><span leaf="">        color=clust_col, labels=clust_labels,</span><span leaf=""><br></span><span leaf="">        show_img=True,</span><span leaf=""><br></span><span leaf="">        </span><span><span leaf=""># 'fast' (white background) or 'dark_background'</span></span><span leaf=""><br></span><span leaf="">        style=</span><span><span leaf="">'fast'</span></span><span leaf="">,</span><span leaf=""><br></span><span leaf="">        </span><span><span leaf=""># limit color scale at 99.2% quantile of cell abundance</span></span><span leaf=""><br></span><span leaf="">        max_color_quantile=0.992,</span><span leaf=""><br></span><span leaf="">        </span><span><span leaf=""># size of locations (adjust depending on figure size)</span></span><span leaf=""><br></span><span leaf="">        circle_diameter=6,</span><span leaf=""><br></span><span leaf="">        colorbar_position=</span><span><span leaf="">'right'</span></span><span leaf=""><br></span><span leaf="">    )</span><span leaf=""><br></span><span leaf="">plt.savefig(</span><span><span leaf="">'06-sc.pl.spatial.png'</span></span><span leaf="">)  </span><span><span leaf=""># 保存为PNG格式    </span></span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">结果：</span></p><p data-tool="mdnice编辑器"><span leaf="">这个跟其他空间工具解卷积结果来看，对于同一个spot如果有多种细胞类型，这里展示的是一个 颜色的叠加而不是饼图，嗯，我感觉还是饼图可能会更直观一点，这里如果颜色选取有强弱很容易造成视觉上的错觉吧。</span></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-imgfileid="100056725" data-ratio="0.8881524440762221" data-src="https://mmbiz.qpic.cn/mmbiz_jpg/cZNhZQ6j4wx8HH4R5qMk8mu6icA8Mzb9kssfOgiarwI75SZp0sBcSn8n3NNwUyfpfBghct0xxiboibtSlKJ8ZrWFiaw/640?wx_fmt=jpeg&amp;from=appmsg" data-type="jpeg" data-w="1207" src="https://mmbiz.qpic.cn/mmbiz_jpg/cZNhZQ6j4wx8HH4R5qMk8mu6icA8Mzb9kssfOgiarwI75SZp0sBcSn8n3NNwUyfpfBghct0xxiboibtSlKJ8ZrWFiaw/640?wx_fmt=jpeg&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">软件版本</span></p><pre data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><code><span leaf="">cell2location.__version__</span><span leaf=""><br></span><span><span leaf=""># '0.1.4'</span></span><span leaf=""><br></span></code></pre><h4 data-tool="mdnice编辑器"><span></span><span><span leaf="">下次见~</span></span><span></span></h4><h3 data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><span></span><span><strong><span leaf="">文末友情宣传</span></strong></span><span></span></h3><ul><li><section><a href="https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247539788&amp;idx=1&amp;sn=62a09c7af6373658bf81c149eb0b4026&amp;scene=21#wechat_redirect"><span leaf="">生信入门&amp;数据挖掘线上直播课4月班</span></a></section></li><li><section><a href="http://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247524148&amp;idx=1&amp;sn=7806da6feb41a36493c519c1cfc1d3ac&amp;chksm=9b4bdf8fac3c569960369602f1ef26639cb366b250f233b2297d1f059471c0458335bfc0b829&amp;scene=21#wechat_redirect"><span leaf="">时隔5年，我们的生信技能树VIP学徒继续招生啦</span></a></section></li><li><section><a href="https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247535760&amp;idx=2&amp;sn=1e02a2e982a046ecf6389231e6768d5b&amp;scene=21#wechat_redirect"><span leaf="">满足你生信分析计算需求的低价解决方案</span></a></section></li></ul></section><section><span leaf=""><br></span></section><p><mp-style-type data-value="3"></mp-style-type></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/8OnjHCfGvPgHdiKiXUeSgA",target="_blank" rel="noopener noreferrer">原文链接</a>
