---
title: "TCGA+GTEX 数据联合分析"
date: 2024-11-09T09:44:54Z
draft: ["false"]
tags: [
  "fetched",
  "老俊俊的生信笔记"
]
categories: ["Acdemic"]
---
TCGA+GTEX 数据联合分析 by 老俊俊的生信笔记
------
<div><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com" data-mpa-powered-by="yiban.io"><section><mp-common-profile data-pluginname="mpprofile" data-id="MzkyMTI1MTYxNA==" data-headimg="http://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4usezgsqIGqjITSMggCTSoViaYeoKe2xoZr1IIvNJoztibQxibYHLDDoiabwAc6Ggws3Tvdo8EPss2nLgaVQ/0?wx_fmt=png" data-nickname="老俊俊的生信笔记" data-alias="JunJunLab" data-signature="老俊俊的生信技能和知识分享,我不是巨人,但你可以站在我的肩膀上更进一步!" data-from="0" data-is_biz_ban="0"></mp-common-profile></section></section><section><mp-common-clmusic data-pluginname="insertaudio" type="1" music_name="%E8%A3%82%E7%BC%9D%E4%B8%AD%E7%9A%84%E9%98%B3%E5%85%89" albumurl="http://wx.y.gtimg.cn/music/photo_new/T002R500x500M000001gqOnU3DTg2S_2.jpg" singer="林俊杰" duration="228000" username="" music_source="1" is_vip="1" listenid="78221626689919136"></mp-common-clmusic></section><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com"><h2 data-tool="mdnice编辑器"><span></span><span>引言</span><span></span></h2><blockquote data-tool="mdnice编辑器"><span>❝</span><p><strong>TCGA</strong> 数据库里的很多癌症对应的<strong>正常组织的样本数量相对于癌症样本非常少</strong>，这对于后续的下游分析比如差异分析可能会带来一定的偏差。<strong>GTEX 数据库则收集来自健康捐献者的多个正常组织，通过对成千上万名捐献者的 RNA 测序 (RNA-seq) 、基因型测序 (Genotyping) 和 表型数据 ，建立了一个跨多个组织的基因表达数据库。</strong> 我们可以去里面下载对应组织的数据来作为正常样本，解决 TCGA 癌症正常样本数量过少的问题。本教程参考了部分 <strong>被炸熟的虾推文：</strong><a href="https://mp.weixin.qq.com/s?__biz=MzkwNDQwMDI5NQ==&amp;mid=2247484859&amp;idx=3&amp;sn=1865bf7e6118c70c8dc58dbd47bb2c3e&amp;chksm=c086d973f7f15065ca39ed19250367f9991dc1c66a9c2daac5e304fc45f8cafd45084a447e0f&amp;token=1657843347&amp;lang=zh_CN&amp;scene=21#wechat_redirect" data-linktype="2">矫正批次效应</a>。致以感谢。</p></blockquote><p data-tool="mdnice编辑器"><strong>https://gtexportal.org/home/</strong></p><figure data-tool="mdnice编辑器"><img data-imgfileid="100031897" data-ratio="0.3876906891109942" data-src="https://mmbiz.qpic.cn/sz_mmbiz_jpg/G5jjcE4useyr7icrw5wPEE8KibcnX88qiayQYE33XA7vKTcPBhTurmpicZlvLKJOlxPL2sQGbQunFwibd0423Hubn8g/640?wx_fmt=jpeg&amp;from=appmsg" data-type="jpeg" data-w="1901" src="https://mmbiz.qpic.cn/sz_mmbiz_jpg/G5jjcE4useyr7icrw5wPEE8KibcnX88qiayQYE33XA7vKTcPBhTurmpicZlvLKJOlxPL2sQGbQunFwibd0423Hubn8g/640?wx_fmt=jpeg&amp;from=appmsg"></figure><p data-tool="mdnice编辑器"><strong>GTEX 已经更新到 V10 版本，引入了 small RNAseq 的数据：</strong></p><figure data-tool="mdnice编辑器"><img data-imgfileid="100031898" data-ratio="0.41830761043108033" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useyr7icrw5wPEE8KibcnX88qiayltjzsUjSSibUZJ2Ynat7mvvukXCvhooAgvtqgoxDowWFEpZGMJcPS3A/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1879" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useyr7icrw5wPEE8KibcnX88qiayltjzsUjSSibUZJ2Ynat7mvvukXCvhooAgvtqgoxDowWFEpZGMJcPS3A/640?wx_fmt=png&amp;from=appmsg"></figure><p data-tool="mdnice编辑器"><strong>如果我们只想分析某个癌症，我们可以根据具体的组织名称去下载数据，而不用下载所有的数据：</strong></p><figure data-tool="mdnice编辑器"><img data-imgfileid="100031900" data-ratio="0.42258064516129035" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useyr7icrw5wPEE8KibcnX88qiayX23rvSE6KNhtlA5KJxmW3lNQa6ia5kKVcmUic2LoGR40N27gVEpD53bA/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1860" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useyr7icrw5wPEE8KibcnX88qiayX23rvSE6KNhtlA5KJxmW3lNQa6ia5kKVcmUic2LoGR40N27gVEpD53bA/640?wx_fmt=png&amp;from=appmsg"></figure><p data-tool="mdnice编辑器"><strong>我们这里下载了肺的数据：https://storage.googleapis.com/adult-gtex/bulk-gex/v10/rna-seq/counts-by-tissue/gene_reads_v10_lung.gct.gz</strong></p><p data-tool="mdnice编辑器"><strong>我们然后去 ucscxena 数据库下载 TCGA 的 lung 的表达数据</strong></p><figure data-tool="mdnice编辑器"><img data-imgfileid="100031899" data-ratio="0.5099935525467441" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useyr7icrw5wPEE8KibcnX88qiayOicKG6oia3Ond33hRuCnXnNc6RjrN9HPoNuUrqvsrDyYWF4d5ePkRPicw/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1551" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useyr7icrw5wPEE8KibcnX88qiayOicKG6oia3Ond33hRuCnXnNc6RjrN9HPoNuUrqvsrDyYWF4d5ePkRPicw/640?wx_fmt=png&amp;from=appmsg"></figure><blockquote data-tool="mdnice编辑器"><span>❝</span><ul><li><section><strong>表达矩阵：https://gdc-hub.s3.us-east-1.amazonaws.com/download/TCGA-LUAD.star_counts.tsv.gz</strong></section></li><li><section><strong>注释文件：https://gdc-hub.s3.us-east-1.amazonaws.com/download/gencode.v36.annotation.gtf.gene.probemap</strong></section></li><li><section><strong>phenotype：https://gdc-hub.s3.us-east-1.amazonaws.com/download/TCGA-LUAD.clinical.tsv.gz</strong></section></li></ul></blockquote><h2 data-tool="mdnice编辑器"><span></span><span>数据处理</span><span></span></h2><blockquote data-tool="mdnice编辑器"><span>❝</span><p>下载好相应的数据后，我们需要对数据进行预处理。把 <strong>GTEX</strong> 的矩阵和 <strong>TCGA</strong> 的进行合并。</p></blockquote><pre data-tool="mdnice编辑器"><span></span><code><span># BiocManager::install("sva")</span><br><span># BiocManager::install("apeglm")</span><br><span>library</span>(tidyverse)<br><span>library</span>(data.table)<br><span>library</span>(edgeR)<br><span>library</span>(sva)<br><span>library</span>(<span>"FactoMineR"</span>)<br><span>library</span>(<span>"factoextra"</span>)<br><br><span># ==============================================================================</span><br><span># pca plot</span><br>PCA_plot = <span>function</span>(data,col){<br>  df.pca &lt;- PCA(t(data), graph = <span>FALSE</span>)<br>  fviz_pca_ind(df.pca,<br>               geom.ind = <span>"point"</span>,<br>               col.ind = col ,<br>               addEllipses = <span>TRUE</span>,<br>               legend.title = <span>"Groups"</span>) +<br>    ggplot2::theme_bw() +<br>    ggplot2::theme(panel.grid = element_blank())<br>}<br></code></pre><p data-tool="mdnice编辑器">首先读入 TCGA 的表达数据和表型信息进行对应筛选，可以看到只有 <strong>76</strong> 的正常样本，<strong>我们使用 edgeR::filterByExpr 来过滤低表达基因</strong>：</p><pre data-tool="mdnice编辑器"><span></span><code><span># ==============================================================================</span><br><span># load raw counts</span><br>count &lt;- fread(<span>"TCGA-LUAD.star_counts.tsv.gz"</span>) %&gt;%<br>  column_to_rownames(var = <span>"Ensembl_ID"</span>)<br>count &lt;- <span>2</span>^count - <span>1</span><br>count &lt;- apply(count, c(<span>1</span>,<span>2</span>), as.integer)<br><br><span># filter low counts</span><br>count_ft &lt;- count[edgeR::filterByExpr(count,min.count = <span>1</span>),]<br><br><span># load phenotype</span><br>pheno &lt;- fread(<span>"TCGA-LUAD.clinical.tsv.gz"</span>) %&gt;%<br>  filter(sample %<span>in</span>% colnames(count)) %&gt;%<br>  mutate(type = ifelse(endsWith(sample,<span>"01A"</span>),<span>"tumor"</span>,<span>"normal"</span>),<br>         .after = sample) %&gt;%<br>  arrange(type)<br><br>table(pheno$type)<br><span># normal  tumor</span><br><span>#     76    513</span><br><br><span># order count matrix</span><br>count_ft &lt;- count_ft[,pheno$sample]<br><br><span># check</span><br>identical(colnames(count_ft),pheno$sample)<br><span># [1] TRUE</span><br></code></pre><p data-tool="mdnice编辑器">接着读取 GTEX 的矩阵，进行合并，<strong>使用 cpm 函数获取标准化的矩阵来绘制 PCA, PCA 图显示 GTEX 的样本和 TCGA 的正常样本具有明显的批次效应</strong>：</p><pre data-tool="mdnice编辑器"><span></span><code><span># ==============================================================================</span><br><span># load gtex data</span><br><span># ==============================================================================</span><br>gtex &lt;- data.table::fread(<span>"gene_reads_v10_lung.gct.gz"</span>,skip = <span>2</span>,header = <span>T</span>)<br>gtex_ct &lt;- gtex[,-<span>2</span>] %&gt;%<br>  column_to_rownames(var = <span>"Name"</span>)<br><br><span># filter common genes</span><br>co_gene &lt;- intersect(rownames(count_ft),rownames(gtex_ct))<br><br>count_ft &lt;- count_ft[co_gene,]<br>gtex_ct &lt;- gtex_ct[co_gene,]<br><br><span># check</span><br>identical(rownames(count_ft),rownames(gtex_ct))<br><span># [1] TRUE</span><br><br><span># combine matrix</span><br>cmb &lt;- cbind(count_ft,gtex_ct)<br><br>pheno_new &lt;- data.frame(sample = c(pheno$sample,colnames(gtex_ct)),<br>                        group = c(pheno$type,rep(<span>"normal"</span>,ncol(gtex_ct))),<br>                        batch = c(rep(<span>"TCGA"</span>,nrow(pheno)),rep(<span>"GTEX"</span>,ncol(gtex_ct)))) %&gt;%<br>  mutate(var = paste(batch,group,sep = <span>"-"</span>))<br><br><span># get cpm</span><br>cpm_mt &lt;- cpm(y = cmb,log = <span>T</span>,normalized.lib.sizes = <span>T</span>)<br><br><span># plot pca</span><br>PCA_plot(data = cpm_mt,col = pheno_new$var)<br></code></pre><figure data-tool="mdnice编辑器"><img data-imgfileid="100031896" data-ratio="0.6638655462184874" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useyr7icrw5wPEE8KibcnX88qiaybHegH3OL9a8Vg1PlEeItMXkc5skPaxyq6NEV43icc3TdwdEz8kVYQdQ/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="714" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useyr7icrw5wPEE8KibcnX88qiaybHegH3OL9a8Vg1PlEeItMXkc5skPaxyq6NEV43icc3TdwdEz8kVYQdQ/640?wx_fmt=png&amp;from=appmsg"></figure><blockquote data-tool="mdnice编辑器"><span>❝</span><p>既然有批次效应，我们就不能简单的进行合并了，我们使用 <strong>sva::ComBat_seq</strong> 函数来去除批次效应，这个是专门争对 <strong>RNASEQ</strong> 数据的。可以看到，<strong>去除批次效应后，GTEX 的正常样本可以和 TCGA 的正常样本比较好的混合了</strong>：</p></blockquote><pre data-tool="mdnice编辑器"><span></span><code><span># ==============================================================================</span><br><span># remove batch effect</span><br><span># ==============================================================================</span><br>rmb_cmb &lt;- sva::ComBat_seq(counts = as.matrix(cmb),<br>                           batch = pheno_new$batch,<br>                           group = pheno_new$group)<br><br><span># check</span><br>head(rmb_cmb[<span>1</span>:<span>4</span>,<span>1</span>:<span>4</span>])<br><span>#                    TCGA-44-6777-11A TCGA-55-6982-11A TCGA-50-6595-11A TCGA-55-6968-11A</span><br><span># ENSG00000000003.15              495              719             1132              385</span><br><span># ENSG00000000457.14              171              399              811              309</span><br><span># ENSG00000000460.17               41               64              106               67</span><br><span># ENSG00000000938.13             4677             3568             6744             5836</span><br><br><span># get cpm</span><br>rmb_cpm_mt &lt;- cpm(y = rmb_cmb,log = <span>T</span>,normalized.lib.sizes = <span>T</span>)<br><br><span># plot pca</span><br>PCA_plot(data = rmb_cpm_mt,col = pheno_new$var)<br></code></pre><figure data-tool="mdnice编辑器"><img data-imgfileid="100031902" data-ratio="0.6624825662482566" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useyr7icrw5wPEE8KibcnX88qiayFYwA5CPCdTgCjLu2TialXCQKRAiaF4ImmAurn7fdY6u7NicTKLfXGicjPA/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="717" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useyr7icrw5wPEE8KibcnX88qiayFYwA5CPCdTgCjLu2TialXCQKRAiaF4ImmAurn7fdY6u7NicTKLfXGicjPA/640?wx_fmt=png&amp;from=appmsg"></figure><hr data-tool="mdnice编辑器"><blockquote data-tool="mdnice编辑器"><span>❝</span><p><strong>sva::ComBat_seq</strong> 对于大点的数据，运行速度可能会比较慢，github 上面也有人提到这个问题。</p></blockquote><figure data-tool="mdnice编辑器"><img data-imgfileid="100031903" data-ratio="0.7123404255319149" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useyr7icrw5wPEE8KibcnX88qiaycR5JmYAibLr2rQOZsmszj1IicfnRasHnvwYQwPZLlIfTicla1X0E8N4sw/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1175" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useyr7icrw5wPEE8KibcnX88qiaycR5JmYAibLr2rQOZsmszj1IicfnRasHnvwYQwPZLlIfTicla1X0E8N4sw/640?wx_fmt=png&amp;from=appmsg"></figure><blockquote data-tool="mdnice编辑器"><span>❝</span><p><strong>于是我找到了这个，去年发表的 python 版的：</strong></p><ul><li><section><strong>github: https://github.com/epigenelabs/inmoose</strong></section></li><li><section><strong>说明文档：https://inmoose.readthedocs.io/en/latest/index.html</strong>感兴趣的可以去试试。</section></li></ul></blockquote><figure data-tool="mdnice编辑器"><img data-imgfileid="100031904" data-ratio="0.6594911937377691" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useyr7icrw5wPEE8KibcnX88qiayJD34V18YNRSzIr6icgZVxl8akK8aGCnyUkHcUKdwxKF5PzCPYHRqhMw/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1022" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useyr7icrw5wPEE8KibcnX88qiayJD34V18YNRSzIr6icgZVxl8akK8aGCnyUkHcUKdwxKF5PzCPYHRqhMw/640?wx_fmt=png&amp;from=appmsg"></figure><figure data-tool="mdnice编辑器"><img data-imgfileid="100031901" data-ratio="0.9036496350364963" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useyr7icrw5wPEE8KibcnX88qiayQ14Zn7gnTXcBvaEGnDsyMSNzD0l9qiaT7duVnV7klvERxzOUghlJictg/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="685" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useyr7icrw5wPEE8KibcnX88qiayQ14Zn7gnTXcBvaEGnDsyMSNzD0l9qiaT7duVnV7klvERxzOUghlJictg/640?wx_fmt=png&amp;from=appmsg"></figure><blockquote data-tool="mdnice编辑器"><span>❝</span><p>此外，<strong>ComBat</strong> 还提出了一种新的方法 <strong>ComBat-ref</strong>, <strong>采用负二项分布模型来调整计数数据，可以提高 RNA-seq 数据中差异表达分析的统计效力和可靠性。</strong></p></blockquote><p data-tool="mdnice编辑器"><strong>github：https://github.com/xiaoyu12/Combat-ref/tree/main</strong></p><figure data-tool="mdnice编辑器"><img data-imgfileid="100031905" data-ratio="0.5324341682723186" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useyr7icrw5wPEE8KibcnX88qiayoJcLVV3EhYCtnoH99oLI8NPicicOicyLib7XQ8uys4wO4hN6qiaDRYHItMA/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1557" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useyr7icrw5wPEE8KibcnX88qiayoJcLVV3EhYCtnoH99oLI8NPicicOicyLib7XQ8uys4wO4hN6qiaDRYHItMA/640?wx_fmt=png&amp;from=appmsg"></figure><figure data-tool="mdnice编辑器"><img data-imgfileid="100031910" data-ratio="1" data-src="https://mmbiz.qpic.cn/sz_mmbiz_jpg/G5jjcE4useyr7icrw5wPEE8KibcnX88qiayY8CeicPaj4rQVSt1EhePJeDm4tmpmVN1xicnVhyqVIQ0RC0ibFwTjlkJA/640?wx_fmt=jpeg&amp;from=appmsg" data-type="jpeg" data-w="2000" src="https://mmbiz.qpic.cn/sz_mmbiz_jpg/G5jjcE4useyr7icrw5wPEE8KibcnX88qiayY8CeicPaj4rQVSt1EhePJeDm4tmpmVN1xicnVhyqVIQ0RC0ibFwTjlkJA/640?wx_fmt=jpeg&amp;from=appmsg"></figure><h2 data-tool="mdnice编辑器"><span></span><span>差异分析</span><span></span></h2><blockquote data-tool="mdnice编辑器"><span>❝</span><p>处理好上面的数据，我们就可以进行简单的差异分析了：</p></blockquote><pre data-tool="mdnice编辑器"><span></span><code><span># ==============================================================================</span><br><span># diff analysis</span><br><span># ==============================================================================</span><br><span>library</span>(rtracklayer)<br><span>library</span>(DESeq2)<br><br><span># load gene annotation</span><br>g_anno &lt;- fread(<span>"gencode.v36.annotation.gtf.gene.probemap"</span>) %&gt;%<br>  select(id,gene) %&gt;%<br>  rowwise() %&gt;%<br>  mutate(gene_id = sapply(strsplit(id,split = <span>"\\."</span>),<span>"["</span>,<span>1</span>))<br><br><span># load gtf annotation</span><br>gtf &lt;- import.gff(<span>"Homo_sapiens.GRCh38.113.gtf.gz"</span>,format = <span>"gtf"</span>) %&gt;%<br>  data.frame() %&gt;%<br>  select(gene_id,gene_name,gene_biotype) %&gt;% unique()<br><br><span># merge annotation</span><br>g_anno &lt;- g_anno %&gt;%<br>  left_join(y = gtf,by = <span>"gene_id"</span>)<br><br><span># diff analysis</span><br>coldata &lt;- data.frame(row.names = colnames(rmb_cmb),<br>                      condition = pheno_new$group)<br><br>dds &lt;- DESeqDataSetFromMatrix(countData = rmb_cmb,<br>                              colData = coldata, design = ~condition)<br><br><br>dds &lt;- DESeq(dds)<br><br>resultsNames(dds)<br><span># [1] "Intercept"                 "condition_tumor_vs_normal"</span><br><br><span># res_lfc &lt;- lfcShrink(dds = dds,coef = 2,type = "apeglm")</span><br><span>#</span><br><span># plotMA(res_lfc)</span><br><span># plotMA(dds)</span><br><br>res &lt;- results(dds,contrast = c(<span>"condition"</span>,<span>"tumor"</span>,<span>"normal"</span>)) %&gt;%<br>  data.frame() %&gt;%<br>  rownames_to_column(var = <span>"gene_id"</span>) %&gt;%<br>  filter(log2FoldChange != <span>"NA"</span>) %&gt;%<br>  mutate(type = case_when(log2FoldChange &gt;= <span>1</span> &amp; pvalue &lt; <span>0.05</span> ~ <span>"sigUp"</span>,<br>                          log2FoldChange &lt;= -<span>1</span> &amp; pvalue &lt; <span>0.05</span> ~ <span>"sigDown"</span>,<br>                          .default = <span>"nonSig"</span>))<br><br><span># load gene anno</span><br>res_anno &lt;- res %&gt;% left_join(y = g_anno,by = c(<span>"gene_id"</span>=<span>"id"</span>))<br><br><span># output</span><br>write.csv(res_anno,file = <span>"all_deseq2_diff.csv"</span>,row.names = <span>F</span>)<br></code></pre><blockquote data-tool="mdnice编辑器"><span>❝</span><p><strong>讨论：在进行不同数据集之前的合并时，是需要考虑批次效应的问题，防止分析产生错误的结论。</strong></p></blockquote><h2 data-tool="mdnice编辑器"><span></span><span>结尾</span><span></span></h2><blockquote data-tool="mdnice编辑器"><span>❝</span><p><strong>路漫漫其修远兮,吾将上下而求索。</strong></p></blockquote><hr data-tool="mdnice编辑器"><p data-tool="mdnice编辑器">欢迎加入生信交流群。加我微信我也拉你进 <strong>微信群聊</strong> <strong>老俊俊生信交流群</strong> <strong>(微信交流群需收取 20 元入群费用,一旦交费,拒不退还!(防止骗子和便于管理))</strong> 。</p><figure data-tool="mdnice编辑器"><img data-imgfileid="100031909" data-ratio="0.6083707025411061" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useyr7icrw5wPEE8KibcnX88qiayqKJHic7yicrQgOe3pLdnhQzWt0Oich3MdnibgibYQKbOadBicRdbu9cooDJQ/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="669" src="https://mmbiz.qpic.cn/sz_mmbiz_png/G5jjcE4useyr7icrw5wPEE8KibcnX88qiayqKJHic7yicrQgOe3pLdnhQzWt0Oich3MdnibgibYQKbOadBicRdbu9cooDJQ/640?wx_fmt=png&amp;from=appmsg"></figure><hr data-tool="mdnice编辑器"><p data-tool="mdnice编辑器"><strong></strong></p><center data-tool="mdnice编辑器"><strong> 往期回顾目录 </strong></center><p data-tool="mdnice编辑器"><br></p><blockquote data-tool="mdnice编辑器"><span>❝</span><p><strong></strong></p><center><strong><a href="https://mp.weixin.qq.com/s?__biz=MzkyMTI1MTYxNA==&amp;mid=2247515543&amp;idx=1&amp;sn=49d6991c5d610fbe3fda761b7421934d&amp;chksm=c18487e6f6f30ef002f9e7d88bb19c2fed2de330786d1ad69d36c713b444f19b64aca79e27c2&amp;token=837342039&amp;lang=zh_CN&amp;scene=21#wechat_redirect" data-linktype="2">SNAF 新型计算工具鉴定广谱癌症新抗原</a></strong></center><strong><center><a href="https://mp.weixin.qq.com/s?__biz=MzkyMTI1MTYxNA==&amp;mid=2247515529&amp;idx=1&amp;sn=b047eb05ccbce0a9ad6ee734b4188416&amp;chksm=c18487f8f6f30eee3c027ac7c5f9b0babcc2fe737dfc2a6c5236e906ec6853b1f508e184af5f&amp;token=2137498509&amp;lang=zh_CN&amp;scene=21#wechat_redirect" data-linktype="2">听说你想给 visCluster 添加行注释？</a></center></strong><strong><center><a href="https://mp.weixin.qq.com/s?__biz=MzkyMTI1MTYxNA==&amp;mid=2247515507&amp;idx=1&amp;sn=bfd0f92a6dbad64dd41d8242e4868132&amp;chksm=c1848702f6f30e140bbe092108d9eeadf4e9e7b9ca13fa33a5575bfc778a7d2af01c982a622f&amp;token=2137498509&amp;lang=zh_CN&amp;scene=21#wechat_redirect" data-linktype="2">Julia 中也可以使用 ggplot？</a></center></strong><strong><center><a href="https://mp.weixin.qq.com/s?__biz=MzkyMTI1MTYxNA==&amp;mid=2247515491&amp;idx=1&amp;sn=d2cfe70986882a8cbc4aecb23fbc1dad&amp;chksm=c1848712f6f30e0413b5808b4cb284b7a98b4e374812d71b8e7e1f3278a3308d6ea1e8a7600d&amp;token=697561698&amp;lang=zh_CN&amp;scene=21#wechat_redirect" data-linktype="2">trackc：HiC 数据可视化神器！</a></center></strong><strong><center><a href="https://mp.weixin.qq.com/s?__biz=MzkyMTI1MTYxNA==&amp;mid=2247515447&amp;idx=1&amp;sn=d8e01ea8fe0b37ee430142ea058a0f9d&amp;chksm=c1848746f6f30e50d75249f9a198946d211e08317960ce7388f2dde95db27d3f9351195aa211&amp;token=697561698&amp;lang=zh_CN&amp;scene=21#wechat_redirect" data-linktype="2">scplotter：单细胞可视化神包！</a></center></strong><strong><center><a href="https://mp.weixin.qq.com/s?__biz=MzkyMTI1MTYxNA==&amp;mid=2247515426&amp;idx=1&amp;sn=a4911c55b9f53ea046cf7aebc5fc0855&amp;chksm=c1848753f6f30e4561303f7f5da0edcb9008ca7925b111e25cb365f9df7962bd57c937c341cb&amp;token=96515316&amp;lang=zh_CN&amp;scene=21#wechat_redirect" data-linktype="2">GseaVis 绘制山脊图</a></center></strong><strong><center><a href="https://mp.weixin.qq.com/s?__biz=MzkyMTI1MTYxNA==&amp;mid=2247515405&amp;idx=1&amp;sn=d3b3ba1ffe1f533ea0fe2f0daf1823a6&amp;chksm=c184877cf6f30e6a36e6e4a32c2677a37dd3ebd24f4e9da823d435f6af9e28678b99676e860b&amp;token=96515316&amp;lang=zh_CN&amp;scene=21#wechat_redirect" data-linktype="2">ClusterGvis 综合使用手册</a></center></strong><strong><center><a href="https://mp.weixin.qq.com/s?__biz=MzkyMTI1MTYxNA==&amp;mid=2247515358&amp;idx=1&amp;sn=98fe527ac548ee46720c11a26b723d39&amp;chksm=c18486aff6f30fb9c0953f5a8d7a1aa39b4dcb8202bb5c34cf61cadae042f8eb2e8631dc36c1&amp;token=2130598365&amp;lang=zh_CN&amp;scene=21#wechat_redirect" data-linktype="2">ClusterGvis 进行 TCseq 聚类分析</a></center></strong><strong><center><a href="https://mp.weixin.qq.com/s?__biz=MzkyMTI1MTYxNA==&amp;mid=2247515345&amp;idx=1&amp;sn=a57b858a3b02af88baf27bc990c8540d&amp;chksm=c18486a0f6f30fb627fa948b07e5b59348fa16eb9e9f9bff3f0895a61ab7c2fdbaa8f9a23d76&amp;token=2130598365&amp;lang=zh_CN&amp;scene=21#wechat_redirect" data-linktype="2">GseaVis 使用文档</a></center></strong><strong><center><a href="https://mp.weixin.qq.com/s?__biz=MzkyMTI1MTYxNA==&amp;mid=2247515329&amp;idx=1&amp;sn=48aef9779d113c6c0bbdb5848de285f7&amp;chksm=c18486b0f6f30fa6fd5437d175f8fa42e4c811f23c9f982572237305d57c6d490088381ec4a5&amp;token=1520616408&amp;lang=zh_CN&amp;scene=21#wechat_redirect" data-linktype="2">R 语言里的非标准求值</a></center></strong><p><br></p></blockquote></section><p><br></p><p><mp-style-type data-value="3"></mp-style-type></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/ahgmT2cW3m5TXYtndClwhQ",target="_blank" rel="noopener noreferrer">原文链接</a>
