---
title: "文章复现学习 | ROS（8）突变频谱、免疫浸润、gsva"
date: 2024-11-13T02:29:09Z
draft: ["false"]
tags: [
  "fetched",
  "生信菜鸟团"
]
categories: ["Acdemic"]
---
文章复现学习 | ROS（8）突变频谱、免疫浸润、gsva by 生信菜鸟团
------
<div><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com"><blockquote data-type="2" data-url="" data-author-name="" data-content-utf8-length="19" data-source-title=""><section><p><span>学习笔记总结于『生信技能树』马拉松课程</span></p></section></blockquote><section data-tool="mdnice编辑器"><section><p>本文学习复现《Oxidative Stress Response Biomarkers of Ovarian Cancer Based on Single-Cell and Bulk RNA Sequencing》（基于单细胞和Bulk转录组的卵巢癌氧化应激反应生物标志物）一文中的图，其中Oxidative Stress Response缩写为ROS。文章包含了芯片、转录组、单细胞等技术，适合用来复现及学习</p></section></section><h1 data-tool="mdnice编辑器"><span></span><span>八、突变频谱、免疫浸润、gsva</span><span></span></h1><h2 data-tool="mdnice编辑器"><span></span><span>1.maftools</span><span></span></h2><h3 data-tool="mdnice编辑器"><span></span><span>1.1突变频谱图</span><span></span></h3><pre data-tool="mdnice编辑器"><span></span><code>rm(list = ls())<br>load(<span>"../6.model/lasso_model.Rdata"</span>)<br>load(<span>"../6.model/rsurv.Rdata"</span>)<br><br><span># TCGAmutations包整合了TCGA中全部样本的maf文件，所以数据就来源这个R包，tcga_load()会给出一个maf文件</span><br><span># devtools::install_github(repo = "PoisonAlien/TCGAmutations")</span><br>library(TCGAmutations)<br>library(tidyverse)<br>maf = tcga_load(<span>"OV"</span>)<br></code></pre><p data-tool="mdnice编辑器">简单查看maf，其中data为主要，clinical.data是临床信息，其余为补充信息</p><figure data-tool="mdnice编辑器"><img data-imgfileid="100002248" data-ratio="0.39892665474060823" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/A8HZFc8vBSXVLvv8JHoBbuO3ib07zbVzQfWtiaA3GrCpcXbibJFGUqPUKIia2zpbbhfnicgHe7uDia02vvF05kqRuQtw/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="559" src="https://mmbiz.qpic.cn/sz_mmbiz_png/A8HZFc8vBSXVLvv8JHoBbuO3ib07zbVzQfWtiaA3GrCpcXbibJFGUqPUKIia2zpbbhfnicgHe7uDia02vvF05kqRuQtw/640?wx_fmt=png&amp;from=appmsg"><figcaption>图83</figcaption></figure><p data-tool="mdnice编辑器">再来看一下主体data的内容，如图84所示</p><figure data-tool="mdnice编辑器"><img data-imgfileid="100002251" data-ratio="0.391" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/A8HZFc8vBSXVLvv8JHoBbuO3ib07zbVzQRcUGT0qAL7Q1AYdkpIFJrYuS39nqE5cxEvVJicN1AA7026frRunCs9w/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="2000" src="https://mmbiz.qpic.cn/sz_mmbiz_png/A8HZFc8vBSXVLvv8JHoBbuO3ib07zbVzQRcUGT0qAL7Q1AYdkpIFJrYuS39nqE5cxEvVJicN1AA7026frRunCs9w/640?wx_fmt=png&amp;from=appmsg"><figcaption>图84</figcaption></figure><pre data-tool="mdnice编辑器"><span></span><code>dim(maf@data)<br><span># [1] 35985    21</span><br><span># 35985行代表着35985条突变，至于是多少人、多少基因，需要另做统计</span><br>length(unique(maf@data<span>$Tumor_Sample_Barcode</span>))<br><span># [1] 411  # 病人数</span><br>length(unique(maf@data<span>$Hugo_Symbol</span>))<br><span># [1] 12861 # 基因数</span><br><br><span># lasso回归模型里面有367个病人，这些病人的ID是16位的，而图84中的突变病人ID是完整的</span><br><span># 为了能将两个数据连接联动起来，需要把后者也截成16位</span><br>samp = data.frame(ID = str_sub(unique(maf@data<span>$Tumor_Sample_Barcode</span>),1,16),<br>                  long = unique(maf@data<span>$Tumor_Sample_Barcode</span>))<br>samp= merge(rsurv,samp,by = <span>"ID"</span>) <span># 一一对应后只剩249人</span><br></code></pre><figure data-tool="mdnice编辑器"><img data-imgfileid="100002250" data-ratio="0.4259776536312849" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/A8HZFc8vBSXVLvv8JHoBbuO3ib07zbVzQZS441paxasiaO5hHJ8B6qo6qGN5WcyvTKVPGmFhm3FaqV6ZOu5jIyBw/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1432" src="https://mmbiz.qpic.cn/sz_mmbiz_png/A8HZFc8vBSXVLvv8JHoBbuO3ib07zbVzQZS441paxasiaO5hHJ8B6qo6qGN5WcyvTKVPGmFhm3FaqV6ZOu5jIyBw/640?wx_fmt=png&amp;from=appmsg"><figcaption>图85 突变频谱图</figcaption></figure><h3 data-tool="mdnice编辑器"><span></span><span>突变频谱图（瀑布图）的含义</span><span></span></h3><figure data-tool="mdnice编辑器"><img data-imgfileid="100002252" data-ratio="0.6497041420118344" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/A8HZFc8vBSXVLvv8JHoBbuO3ib07zbVzQ1hxNTPacxibOhxRaTGsicKibIX6SLCichVqicyEOB0j1FYMcKXhGoTJeu2w/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1690" src="https://mmbiz.qpic.cn/sz_mmbiz_png/A8HZFc8vBSXVLvv8JHoBbuO3ib07zbVzQ1hxNTPacxibOhxRaTGsicKibIX6SLCichVqicyEOB0j1FYMcKXhGoTJeu2w/640?wx_fmt=png&amp;from=appmsg"><figcaption>图86</figcaption></figure><h3 data-tool="mdnice编辑器"><span></span><span>1.2 TMB肿瘤突变负荷</span><span></span></h3><p data-tool="mdnice编辑器">TMB：衡量每个病人的突变情况是否严重，数字越大，突变越多</p><figure data-tool="mdnice编辑器"><img data-imgfileid="100002249" data-ratio="0.6051800379027164" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/A8HZFc8vBSXVLvv8JHoBbuO3ib07zbVzQtO571PSFqCiaD9n83jKkqbib43fvicia1DrCbnAlT0sYOa3dicgV2T69wew/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1583" src="https://mmbiz.qpic.cn/sz_mmbiz_png/A8HZFc8vBSXVLvv8JHoBbuO3ib07zbVzQtO571PSFqCiaD9n83jKkqbib43fvicia1DrCbnAlT0sYOa3dicgV2T69wew/640?wx_fmt=png&amp;from=appmsg"><figcaption>图87</figcaption></figure><pre data-tool="mdnice编辑器"><span></span><code>identical(tmb<span>$Tumor_Sample_Barcode</span>,samp<span>$long</span>)<br><span># [1] FALSE  </span><br><span># 发现顺序不同</span><br>samp = merge(samp,tmb,by.x = <span>"long"</span>,by.y = <span>"Tumor_Sample_Barcode"</span>)<br>samp<span>$tmb_group</span> = ifelse(samp<span>$total_perMB_log</span>&gt;median(samp<span>$total_perMB_log</span>),<span>"high"</span>,<span>"low"</span>)<br>library(ggplot2)<br>library(ggpubr)<br>ggplot(samp,aes(x = total_perMB_log,y = RS))+<br>  geom_point(aes(color = group))+<br>  geom_smooth(method = <span>"lm"</span>)+<br>  theme_bw()+<br>  stat_cor(method = <span>"pearson"</span>)<br></code></pre><p data-tool="mdnice编辑器">相关系数接近于0，差不多不相关了</p><figure data-tool="mdnice编辑器"><img data-imgfileid="100002253" data-ratio="0.8345864661654135" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/A8HZFc8vBSXVLvv8JHoBbuO3ib07zbVzQ5EMica85bWC4lvJNlfjsibA9ZibppVfy9iattCiabfhA1JEu985GicYnyZJA/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="665" src="https://mmbiz.qpic.cn/sz_mmbiz_png/A8HZFc8vBSXVLvv8JHoBbuO3ib07zbVzQ5EMica85bWC4lvJNlfjsibA9ZibppVfy9iattCiabfhA1JEu985GicYnyZJA/640?wx_fmt=png&amp;from=appmsg"><figcaption>图88 相关性</figcaption></figure><pre data-tool="mdnice编辑器"><span></span><code>load(<span>"../6.model/TCGA-OV_sur_model.Rdata"</span>)<br>library(survival)<br>library(survminer)<br><br>sfit &lt;- survfit(Surv(time, event)~tmb_group, data=samp)<br>ggsurvplot(sfit,<br>           palette = <span>"jco"</span>,<br>           risk.table =TRUE,<br>           pval =TRUE,<br>           conf.int =TRUE)<br></code></pre><figure data-tool="mdnice编辑器"><img data-imgfileid="100002254" data-ratio="0.8345864661654135" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/A8HZFc8vBSXVLvv8JHoBbuO3ib07zbVzQwvRU9NYOG4jvbRxiaynjbDervqVoicLFrEfnJIrkz3CQV5DRRpqiaj8gg/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="665" src="https://mmbiz.qpic.cn/sz_mmbiz_png/A8HZFc8vBSXVLvv8JHoBbuO3ib07zbVzQwvRU9NYOG4jvbRxiaynjbDervqVoicLFrEfnJIrkz3CQV5DRRpqiaj8gg/640?wx_fmt=png&amp;from=appmsg"><figcaption>图89</figcaption></figure><h2 data-tool="mdnice编辑器"><mpcpc js_editor_cpcad="" src="/cgi-bin/readtemplate?t=tmpl/cpc_tmpl#1713362815691" data-category_id_list="1|16|17|2|21|24|28|29|31|35|36|37|39|41|42|43|46|47|48|5|50|51|55|56|57|58|59|6|60|61|62|63|64|65|66|67|68|7|8" data-id="1713362815691"></mpcpc><span>2.免疫浸润</span><span></span></h2><p data-tool="mdnice编辑器">免疫浸润：根据基因表达量推断每一种细胞的组成比例和相对丰度是多少</p><p data-tool="mdnice编辑器">目前主流的免疫浸润计算方法是CIBERSORT和ssgsea，这里介绍CIBERSORT</p><h3 data-tool="mdnice编辑器"><span></span><span>2.1对输入数据的要求</span><span></span></h3><pre data-tool="mdnice编辑器"><span></span><code>下面这段话摘自CIBERSORT的介绍<br>&gt; Importantly, all expression data should be non-negative, devoid of missing values, and represented <span>in</span> non-log linear space. <br>&gt; For Affymetrix microarrays, a custom chip definition file (CDF) is recommended (see Subheading 3.2.2) and should be normalized with MAS5 or RMA. <br>&gt; Illumina Beadchip and single color Agilent arrays should be processed as described <span>in</span> the limma package. <br>&gt; Standard RNA-Seq expression quantification metrics, such as frag- ments per kilobase per million (FPKM) and transcripts per kilobase million (TPM), are suitable <span>for</span> use with CIBERSORT.<br>--《Profiling Tumor Infiltrating Immune Cells with CIBERSORT》<br><br>非常清楚的写出了输入数据的要求：<br>1.不可以有负值和缺失值<br>2.不要取<span>log</span><br>3.如果是芯片数据，昂飞芯片使用RMA标准化，Illumina 的Beadchip 和Agilent的单色芯片，用limma处理。 <br>4.如果是RNA-Seq表达量，使用FPKM和TPM都很合适。<br>GEO常规的表达矩阵都是这样得到的，直接下载使用即可。注意有的表达矩阵下载下来就已经取过<span>log</span>，需要逆转回去。有的经过了标准化或者有负值，需要处理原始数据，前面写过介绍文了<br></code></pre><h3 data-tool="mdnice编辑器"><span></span><span>2.2做成cibersort要求的输入文件</span><span></span></h3><p data-tool="mdnice编辑器">当年这个算法并没有被写成R包，而是只有一个放着函数的脚本<code>CIBERSORT.R</code>，把它下载下来放在工作目录即可。现在这个R包写出来了，但还是要求提供一个txt文件，详见带代码</p><pre data-tool="mdnice编辑器"><span></span><code>rm(list = ls())<br>library(tidyverse)<br>load(<span>"../6.model/TCGA-OV_sur_model.Rdata"</span>)<span>#生存分析的数据</span><br>load(<span>"../6.model/lasso_model.Rdata"</span>)<span>#模型</span><br>load(<span>"../6.model/rsurv.Rdata"</span>)<br><br><span># 需要两个输入文件:</span><br><span># 一个是表达矩阵文件</span><br><span># 一个是官网提供的LM22.txt，记录了22种免疫细胞的基因表达特征数据，现在这个文件已经内置在R包里面了</span><br><br><span># 要求提供一个txt文件</span><br>library(tidyverse)<br>exp = exprSet<span>#完整的表达矩阵，而不是lasso回归后的基因</span><br>exp2 = as.data.frame(exp)<br>exp2 = rownames_to_column(exp2)<span>#由于CIBERSORT.R读取文件的代码比较粗暴，为了适应它，导出文件之前需要把行名变成一列。不然后面就会有报错。</span><br>write.table(exp2,file = <span>"exp.txt"</span>,row.names = F,quote = F,sep = <span>"\t"</span>)<br></code></pre><h3 data-tool="mdnice编辑器"><span></span><span>2.3运行CIBERSORT</span><span></span></h3><pre data-tool="mdnice编辑器"><span></span><code><span># 这段代码大致需要半小时才能运行完毕，不过已经设置了限速步骤了</span><br>f = <span>"ciber_OV.Rdata"</span><br><span>if</span>(!file.exists(f)){<br>  <span>#devtools:: install_github ("Moonerss/CIBERSORT")</span><br>  library(CIBERSORT)<br>  lm22f = system.file(<span>"extdata"</span>, <span>"LM22.txt"</span>, package = <span>"CIBERSORT"</span>) <span>#调取R包的内置数据LM22.txt</span><br>  TME.results = cibersort(lm22f, <br>                          <span>"exp.txt"</span> , <br>                          perm = 1000, <br>                          QN = T)<br>  save(TME.results,file = f)<br>}<br><br>load(f)<br>TME.results[1:4,1:4]<br><span># TME.results即运行的结果，一行是一个样本，一列是一种细胞，这些数字即这个细胞在该样本中的相对丰度</span><br><span># 计算方法：每一种细胞是有自己的marker gene的，根据这些基因的表达量推断，所以才要求我们的输入数据中有表达矩阵</span><br>re &lt;- TME.results[,-(23:25)]<span># 前22列是计算结果，后三列是总结，这三列一般用不上</span><br></code></pre><h3 data-tool="mdnice编辑器"><span></span><span>2.4经典的免疫细胞丰度热图</span><span></span></h3><pre data-tool="mdnice编辑器"><span></span><code>library(pheatmap)<br>k &lt;- apply(re,2,<span>function</span>(x) {sum(x == 0) &lt; nrow(TME.results)/2})<span>#在一半以上样本里丰度为0的免疫细胞，就不展示在热图里了</span><br>table(k)<br><span># k</span><br><span># FALSE  TRUE </span><br><span>#    10    12 </span><br>re2 &lt;- as.data.frame(t(re[,k]))<br>Group = rsurv<span>$group</span><br>table(Group)<br><span># Group</span><br><span># low high </span><br><span># 184  183 </span><br>an = data.frame(group = Group,<br>                row.names = colnames(exp))<br>pheatmap(re2,scale = <span>"row"</span>,<br>         show_colnames = F,<br>         cluster_cols = F,<br>         annotation_col = an,<br>         color = colorRampPalette(c(<span>"navy"</span>, <span>"white"</span>, <span>"firebrick3"</span>))(50))<br></code></pre><p data-tool="mdnice编辑器">得到如图90所示，不过这个就比较凌乱了，聚类与否都是乱的。可以尝试用分组聚类的方法把图变好看点，不过没必要，既然这么乱就不用放进文章了</p><figure data-tool="mdnice编辑器"><img data-imgfileid="100002256" data-ratio="0.8345864661654135" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/A8HZFc8vBSXVLvv8JHoBbuO3ib07zbVzQfvEKJibNgbnMaRCwPgZEU08b6Ce7QTdapX8cHUsxGkd5HrzLWTTeLow/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="665" src="https://mmbiz.qpic.cn/sz_mmbiz_png/A8HZFc8vBSXVLvv8JHoBbuO3ib07zbVzQfvEKJibNgbnMaRCwPgZEU08b6Ce7QTdapX8cHUsxGkd5HrzLWTTeLow/640?wx_fmt=png&amp;from=appmsg"><figcaption>图90 免疫细胞丰度热图</figcaption></figure><h3 data-tool="mdnice编辑器"><span></span><span>2.5直方图</span><span></span></h3><pre data-tool="mdnice编辑器"><span></span><code><span># 可以展示出每个样本的免疫细胞比例</span><br>library(RColorBrewer)<br><span># mypalette &lt;- colorRampPalette(brewer.pal(8,"Set1"))#设置配色，不过颜色不好看，换一个</span><br>?draw_boxplot()<br>grcolor = c(<span>"#2874C5"</span>, <span>"#f87669"</span>, <span>"#e6b707"</span>, <span>"#868686"</span>, <span>"#66C2A5"</span>, <span>"#FC8D62"</span>, <span>"#8DA0CB"</span>,<br>          <span>"#E78AC3"</span>, <span>"#A6D854"</span>, <span>"#FFD92F"</span>, <span>"#E5C494"</span>, <span>"#B3B3B3"</span>)<br><br><span># 调整格式：宽变长</span><br>dat &lt;- re %&gt;% <br>  as.data.frame() %&gt;%<br>  rownames_to_column(<span>"Sample"</span>) %&gt;% <br>  mutate(group = Group) %&gt;% <br>  gather(key = Cell_type,value = Proportion,-Sample,-group) %&gt;% <br>  arrange(group)<br><br>dat<span>$Sample</span> = factor(dat<span>$Sample</span>,ordered = T,levels = unique(dat<span>$Sample</span>)) <span>#为了固定横坐标的顺序，以因子的时候把细胞分出顺序，这样在画图的时候能一一对应</span><br><span># 先把group排序，然后将sample设为了因子，确定排序后的顺序为水平，所以两图的顺序是对应的</span><br><br>dat2 = data.frame(a = 1:ncol(exp),<br>                  b = 1,<br>                  group = sort(Group)) <span>#这个dat2是为了画注释条</span><br><br><span># p2才是核心</span><br>p1 = ggplot(dat2,aes(x = a, y = b)) + <br>  geom_tile(aes(fill = group)) + <br>  scale_fill_manual(values = grcolor) + <span># mypalette(22)[1:length(unique(Group))]</span><br>  theme(panel.grid = element_blank(), <br>        panel.background = element_blank(), <br>        axis.line = element_blank(), <br>        axis.ticks = element_blank(), <br>        axis.text = element_blank(), <br>        axis.title = element_blank()) + <br>  scale_x_continuous(expand = c(0, 0)) +<br>  labs(fill = <span>"Group"</span>)<br><br>p2 = ggplot(dat,aes(Sample, Proportion,fill = Cell_type)) + <br>  geom_bar(<span>stat</span> = <span>"identity"</span>) +<br>  labs(fill = <span>"Cell Type"</span>,x = <span>""</span>,y = <span>"Estiamted Proportion"</span>) + <br>  theme_bw() +<br>  theme(<span>#axis.text.x = element_blank(),</span><br>    axis.ticks.x = element_blank()<br>  ) + <br>  scale_y_continuous(expand = c(0.01,0)) +<br>  scale_fill_manual(values = mypalette(22))<br><br>library(patchwork)<br>p1 / p2 + plot_layout(heights = c(1,10),guides = <span>"collect"</span> ) &amp;<br>  theme(legend.position = <span>"bottom"</span>)<br></code></pre><p data-tool="mdnice编辑器">得到图91，为了看高风险组和低风险组的病人的细胞组成是否存在差别，但肉眼感觉看不出来</p><figure data-tool="mdnice编辑器"><img data-imgfileid="100002257" data-ratio="0.8345864661654135" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/A8HZFc8vBSXVLvv8JHoBbuO3ib07zbVzQJN8bCeLdcd6SgqW1YpH1b9P9p4IOHTUCjdGJGVwPs0gKGdibXVFulicA/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="665" src="https://mmbiz.qpic.cn/sz_mmbiz_png/A8HZFc8vBSXVLvv8JHoBbuO3ib07zbVzQJN8bCeLdcd6SgqW1YpH1b9P9p4IOHTUCjdGJGVwPs0gKGdibXVFulicA/640?wx_fmt=png&amp;from=appmsg"><figcaption>图91 直方图</figcaption></figure><h3 data-tool="mdnice编辑器"><span></span><span>2.6箱线图</span><span></span></h3><pre data-tool="mdnice编辑器"><span></span><code><span># 展示免疫细胞之间的比较</span><br>ggplot(dat,aes(Cell_type,Proportion,fill = Cell_type)) + <br>  geom_boxplot() + <br>  theme_bw() + <br>  labs(x = <span>"Cell Type"</span>, y = <span>"Estimated Proportion"</span>) +<br>  theme(axis.text.x = element_blank(),<br>        axis.ticks.x = element_blank(),<br>        legend.position = <span>"bottom"</span>) + <br>  scale_fill_manual(values = mypalette(22))<br></code></pre><p data-tool="mdnice编辑器">得到图92，不同的免疫细胞的丰度比较，有些完全是0</p><figure data-tool="mdnice编辑器"><img data-imgfileid="100002255" data-ratio="0.8345864661654135" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/A8HZFc8vBSXVLvv8JHoBbuO3ib07zbVzQicF9HWbYQNTBnQicVWDhRbZH5SD9duBUQIomuyCv204CDCPdWHELpjpg/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="665" src="https://mmbiz.qpic.cn/sz_mmbiz_png/A8HZFc8vBSXVLvv8JHoBbuO3ib07zbVzQicF9HWbYQNTBnQicVWDhRbZH5SD9duBUQIomuyCv204CDCPdWHELpjpg/640?wx_fmt=png&amp;from=appmsg"><figcaption>图92 箱线图</figcaption></figure><p data-tool="mdnice编辑器">下面把每种免疫细胞不同分组的结果做个比较</p><pre data-tool="mdnice编辑器"><span></span><code><span># 画出全部在组间差异显著的细胞</span><br><span># 全是0的行去掉</span><br>k = colSums(re)&gt;0;table(k)<br>re = re[,k]<br>library(tinyarray)<br>draw_boxplot(t(re),factor(Group),<br>             drop = F,<span>#color = mypalette(length(unique(Group)))</span><br>             )+<br>  labs(x = <span>"Cell Type"</span>, y = <span>"Estimated Proportion"</span>) <br></code></pre><p data-tool="mdnice编辑器">得到图93</p><figure data-tool="mdnice编辑器"><img data-imgfileid="100002260" data-ratio="0.8345864661654135" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/A8HZFc8vBSXVLvv8JHoBbuO3ib07zbVzQIvbHcMH7ZaKzPSuxQvn3M8Bz9hdGaWia8NYLsUzuIw1PWichialibT7ahQ/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="665" src="https://mmbiz.qpic.cn/sz_mmbiz_png/A8HZFc8vBSXVLvv8JHoBbuO3ib07zbVzQIvbHcMH7ZaKzPSuxQvn3M8Bz9hdGaWia8NYLsUzuIw1PWichialibT7ahQ/640?wx_fmt=png&amp;from=appmsg"><figcaption>图93 箱线图</figcaption></figure><h3 data-tool="mdnice编辑器"><span></span><span>单画某一个感兴趣的免疫细胞</span><span></span></h3><pre data-tool="mdnice编辑器"><span></span><code>dat2 = dat[dat<span>$Cell_type</span>==<span>"Dendritic cells activated"</span>,]<br>library(ggpubr)<br>ggplot(dat2,aes(group,Proportion,fill = group)) + <br>  geom_boxplot() + <br>  theme_bw() + <br>  labs(x = <span>"Group"</span>, y = <span>"Estimated Proportion"</span>) +<br>  theme(legend.position = <span>"top"</span>) + <br>  scale_fill_manual(values = grcolor)+ <span>#mypalette(length(unique(Group)))</span><br>  stat_compare_means(aes(group = group,label = ..p.signif..),method = <span>"kruskal.test"</span>)<br></code></pre><figure data-tool="mdnice编辑器"><img data-imgfileid="100002259" data-ratio="0.8345864661654135" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/A8HZFc8vBSXVLvv8JHoBbuO3ib07zbVzQIPed6ic5UKFJSrnbXViasTjBEB43v0rFJeazW5mqB3sRvIXshnnjEWGQ/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="665" src="https://mmbiz.qpic.cn/sz_mmbiz_png/A8HZFc8vBSXVLvv8JHoBbuO3ib07zbVzQIPed6ic5UKFJSrnbXViasTjBEB43v0rFJeazW5mqB3sRvIXshnnjEWGQ/640?wx_fmt=png&amp;from=appmsg"><figcaption>图94</figcaption></figure><h2 data-tool="mdnice编辑器"><mpcpc js_editor_cpcad="" src="/cgi-bin/readtemplate?t=tmpl/cpc_tmpl#1713362802954" data-category_id_list="1|16|17|2|21|24|28|29|31|35|36|37|39|41|42|43|46|47|48|5|50|51|55|56|57|58|59|6|60|61|62|63|64|65|66|67|68|7|8" data-id="1713362802954"></mpcpc><span>3.gsva</span><span></span></h2><p data-tool="mdnice编辑器">文章中的图如图95，是为了看免疫检查点基因和构建模型使用的基因的相关性</p><figure data-tool="mdnice编辑器"><img data-imgfileid="100002261" data-ratio="2.533582089552239" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/A8HZFc8vBSXVLvv8JHoBbuO3ib07zbVzQjoJbdefRs1jaFgRyBsz8RgCKbZWHExVnjQoNxb7L0pPIicl0hwc5hMA/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="268" src="https://mmbiz.qpic.cn/sz_mmbiz_png/A8HZFc8vBSXVLvv8JHoBbuO3ib07zbVzQjoJbdefRs1jaFgRyBsz8RgCKbZWHExVnjQoNxb7L0pPIicl0hwc5hMA/640?wx_fmt=png&amp;from=appmsg"><figcaption>图95</figcaption></figure><h3 data-tool="mdnice编辑器"><span></span><span>3.1免疫检查点基因</span><span></span></h3><pre data-tool="mdnice编辑器"><span></span><code><span># 免疫检查点基因参考自文献 https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7786136<span es-id-type="PMCID" es-id="7786136" es-collect-button-id="0" cslsource="pubmedPmid"><svg version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" x="0px" y="0px" viewbox="0 0 1000 1000" enable-background="new 0 0 1000 1000" height="1em"></svg><br>        <g><path d="M852.3,135.6l-47.7,48.3c157.6,178.7,157.4,451,0.1,631l48.8,49.5C1035.9,672.1,1035.5,327.2,852.3,135.6z M740.9,248.6l-49.3,49.9c96.1,115.5,96.1,285.8,0.5,402.1l48.8,49.6C863.4,606.1,863.6,391.5,740.9,248.6z M195.4,183.9l-47.7-48.3c-183.2,191.6-183.6,536.5-1.2,728.8l48.8-49.5C37.9,635,37.8,362.6,195.4,183.9z M259.1,248.6C136.4,391.5,136.6,606.1,259,750.3l48.8-49.6c-95.6-116.4-95.6-286.6,0.4-402.1L259.1,248.6z M499.8,340.4c-87.3,0-158.1,71.8-158.1,160.4c0,88.5,70.8,160.4,158.1,160.4c87.3,0,158.1-71.8,158.1-160.4C657.8,412.2,587,340.4,499.8,340.4z"></path></g><br>        <span><span value="4.7">IF: 4.7 </span><span value="Q2">Q2 </span><span value="B3">B3</span></span></span>/</span><br><br>rm(list = ls())<br>library(tidyverse)<br>tmp = readLines(<span>"immune_checkpoint.txt"</span>)<br>genes = str_split(tmp,<span>", "</span>)[[1]]<span>#将tmp的内容拆分，这些基因如果之间从表达矩阵里面找，有7个是找不到的。有可能是这7个有别名，只能手动找了，别名也是可以用的</span><br>load(<span>"../6.model/TCGA-OV_sur_model.Rdata"</span>)<br>k = genes %<span>in</span>% rownames(exprSet) ;table(k)<span>#不过也有41个了，偷点懒不找了</span><br>genes = genes[k]<br>genes<br><br>load(<span>"../6.model/lassogene.Rdata"</span>)<br>library(Hmisc)<br>exp = exprSet<br>da = t(exp[c(genes,lassoGene),])<span>#把目标基因和lasso回归基因拼到一起，并转置。计算基因之间的相关性要把基因放到列名上。如果不放，列名是样本，那就变成样本的相关性了</span><br><span># Hmisc包的rcorr函数能算R值和P值</span><br>m = rcorr(da)<span>$r</span>[1:length(genes),(ncol(da)-length(lassoGene)+1):ncol(da)]<span>#前41行，后26列，右上角</span><br>p = rcorr(da)<span>$P</span>[1:length(genes),(ncol(da)-length(lassoGene)+1):ncol(da)]<br>p[1:3,1:4]<br><span># 倒数第26列：ncol(da)基因的总数67个 - length(lassoGene)26个 + 1</span><br><br>library(dplyr)<br>tmp = matrix(case_when(as.logical(p&lt;0.01)~<span>"**"</span>,<br>                       as.logical(p&lt;0.05)~<span>"*"</span>,<br>                       T~<span>""</span>),nrow = nrow(p))<br>library(pheatmap)<br><span>if</span> (as.numeric(grDevices::dev.cur()) != 1) grDevices::graphics.off() <span>#为了关闭画板</span><br>png(<span>"ff.png"</span>,width = 1200,height = 900)<br>pheatmap(t(m),<br>         display_numbers =t(tmp),<br>         angle_col =45,<br>         color = colorRampPalette(c(<span>"#2fa1dd"</span>, <span>"white"</span>, <span>"#f87669"</span>))(100),<br>         border_color = <span>"white"</span>,<br>         width = 7, <br>         height=9.1,<br>         treeheight_col = 0,<br>         treeheight_row = 0)<br>dev.off()<br></code></pre><p data-tool="mdnice编辑器">得到如图96，文章是竖着画的，我们是横着画的</p><p data-tool="mdnice编辑器">不过注意色带，虽然很蓝，但其实数值也只到0.2左右。可以试着对<code>pheatmap()</code>函数调整颜色范围</p><figure data-tool="mdnice编辑器"><img data-imgfileid="100002262" data-ratio="0.75" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/A8HZFc8vBSXVLvv8JHoBbuO3ib07zbVzQCQj8y22nvpf4CdhaG5A9TNM3wZRUpxNicDZhlr18KMRLuWAOwqEFs0w/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1200" src="https://mmbiz.qpic.cn/sz_mmbiz_png/A8HZFc8vBSXVLvv8JHoBbuO3ib07zbVzQCQj8y22nvpf4CdhaG5A9TNM3wZRUpxNicDZhlr18KMRLuWAOwqEFs0w/640?wx_fmt=png&amp;from=appmsg"><figcaption>图96</figcaption></figure><h3 data-tool="mdnice编辑器"><span></span><span>3.2 gsva</span><span></span></h3><p data-tool="mdnice编辑器">gsva是为了做样本（或分组）之间的富集程度比较</p><pre data-tool="mdnice编辑器"><span></span><code>rm(list = ls())<br>load(<span>"../6.model/TCGA-OV_sur_model.Rdata"</span>)<br>load(<span>"../6.model/lasso_model.Rdata"</span>)<br><br><br><br><br>library(GSVA)<span># 和gsea一样，需要基因和通路之间的对应关系</span><br>library(msigdbr)<span># 这个对应关系可以从msigdbr里面找，也可以从网站上下载</span><br>KEGG_df = msigdbr(species = <span>"Homo sapiens"</span>,category = <span>"C2"</span>,subcategory = <span>"CP:KEGG"</span>) %&gt;% <br>  dplyr::select(gs_exact_source,gs_name,gene_symbol)<br>head(KEGG_df)<br><br>H_df = msigdbr(species = <span>"Homo sapiens"</span>,category = <span>"H"</span>) %&gt;% <br>  dplyr::select(gene_symbol,gs_exact_source,gs_subcat,gs_name)<br>dim(H_df)<br><br><span>#把数据框拆分成列表</span><br>kegg_list = split(KEGG_df<span>$gene_symbol</span>,KEGG_df<span>$gs_exact_source</span>)<br>lapply(kegg_list[1:3], head)<br><br>H_list = split(H_df<span>$gene_symbol</span>,H_df<span>$gs_name</span>)<br>lapply(H_list[1:3], head)<br><br><br>f = <span>"gsva.Rdata"</span><br><span>if</span>(!file.exists(f)){<br>  KEGG_ES &lt;- gsva(expr=exprSet, <br>                  gset.idx.list=kegg_list, <br>                  parallel.sz=10)<span># 注意设置10线程对内存有要求，实在不够，设置为1，让它慢慢跑吧//gsva只需要表达矩阵和基因通路对应关系列表就能完成分析</span><br>  H_ES &lt;- gsva(expr=exprSet, <br>               gset.idx.list=H_list, <br>               parallel.sz=10)<br>  save(KEGG_ES,H_ES,file = f)<br>}<br>load(f)<span>#载入得到的结果H_ES，行名是通路，列名是样本，数字是富集分数</span><br><span># 有了这个结果，可以用limma做差异分析</span><br><br><br><br>load(<span>"../6.model/rsurv.Rdata"</span>)<br>library(limma)<br>design = model.matrix(~rsurv<span>$group</span>)<br>fit = lmFit(KEGG_ES, design)<br>fit = eBayes(fit)<br>DEG_KEGG = topTable(fit, coef = 2, number = Inf)<br>head(DEG_KEGG)<br><br>fit2 = lmFit(H_ES, design)<br>fit2 = eBayes(fit2)<br>DEG_H = topTable(fit2, coef = 2, number = Inf)<br>head(DEG_H)<br><br><span># 做热图，那图个方便，就拿前20个来做</span><br>top20_kegg = head(rownames(DEG_KEGG)[order(DEG_KEGG<span>$logFC</span>,decreasing = T)],20)<br>top20_go = head(rownames(DEG_H)[order(DEG_H<span>$logFC</span>,decreasing = T)],20)<br>n = KEGG_ES[top20_kegg,]<br>n = n[,order(rsurv<span>$RS</span>)]<br>rownames(n) = KEGG_df<span>$gs_name</span>[match(rownames(n),KEGG_df<span>$gs_exact_source</span>)]<br>g = sort(rsurv<span>$group</span>)<br>library(tinyarray)<br>draw_heatmap(n,g,show_rownames = T,cluster_cols = F,scale = F,legend = T,annotation_legend = T)<br></code></pre><figure data-tool="mdnice编辑器"><img data-imgfileid="100002263" data-ratio="0.5299635226680562" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/A8HZFc8vBSXVLvv8JHoBbuO3ib07zbVzQz5X3pyTsVb1ibcYzUCTAE8pjWgNf64w0YZePa0w6kriaFTXop1qkdZ9w/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1919" src="https://mmbiz.qpic.cn/sz_mmbiz_png/A8HZFc8vBSXVLvv8JHoBbuO3ib07zbVzQz5X3pyTsVb1ibcYzUCTAE8pjWgNf64w0YZePa0w6kriaFTXop1qkdZ9w/640?wx_fmt=png&amp;from=appmsg"><figcaption>图97 KEGG_ES</figcaption></figure><pre data-tool="mdnice编辑器"><span></span><code>n2 = H_ES[top20_go,]<br>n2 = n2[,order(rsurv<span>$RS</span>)]<br>rownames(n2) = H_df<span>$gs_name</span>[match(rownames(n2),H_df<span>$gs_name</span>)]<br><br>draw_heatmap(n2,g,show_rownames = T,cluster_cols = F,scale = F,legend = T,annotation_legend = T)<br></code></pre><figure data-tool="mdnice编辑器"><img data-imgfileid="100002268" data-ratio="0.5299635226680562" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/A8HZFc8vBSXVLvv8JHoBbuO3ib07zbVzQcDFqNkF2RUfoPWBWHyG3DaByiaaDNrgia8TZMhcibLOAHs1s2hIMxdJjQ/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1919" src="https://mmbiz.qpic.cn/sz_mmbiz_png/A8HZFc8vBSXVLvv8JHoBbuO3ib07zbVzQcDFqNkF2RUfoPWBWHyG3DaByiaaDNrgia8TZMhcibLOAHs1s2hIMxdJjQ/640?wx_fmt=png&amp;from=appmsg"><figcaption>图98 H_ES</figcaption></figure><p data-tool="mdnice编辑器">文章还有一张热图，为了显示差异富集的 HALLMARK 通路的富集评分与 RS 之间的相关性，我们也来做一下</p><pre data-tool="mdnice编辑器"><span></span><code>library(corrplot)<br>M = cor(cbind(t(n),riskscore = rsurv<span>$RS</span>))<br>p.mat &lt;- cor.mtest(cbind(t(n),riskscore = rsurv<span>$RS</span>))<span>$p</span><br>library(paletteer)<br>my_color = rev(paletteer_d(<span>"RColorBrewer::RdYlBu"</span>)[-1])<br>my_color = colorRampPalette(my_color)(10)<br>png(<span>"corrplot.png"</span>,height = 1000,width = 1000)<br>corrplot(M, <span>type</span>=<span>"lower"</span>, <br>         order=<span>"hclust"</span>, <br>         col = my_color,<br>         p.mat = p.mat, <br>         sig.level = 0.01, <br>         insig = <span>"blank"</span>,<br>         tl.col = <span>"black"</span>,<br>         tl.pos = <span>'l'</span>,<br>         cl.pos = <span>'r'</span>)<br>dev.off()<br></code></pre><p data-tool="mdnice编辑器">得到如图99，不过看起来都不怎么相关</p><figure data-tool="mdnice编辑器"><img data-imgfileid="100002267" data-ratio="0.521" data-src="https://mmbiz.qpic.cn/sz_mmbiz_png/A8HZFc8vBSXVLvv8JHoBbuO3ib07zbVzQ9jibDEDNYmibfN08nWBEfNcxJzCsmpLy5Tzud9fbedBCnHYoqNkWVrKw/640?wx_fmt=png&amp;from=appmsg" data-type="png" data-w="1000" src="https://mmbiz.qpic.cn/sz_mmbiz_png/A8HZFc8vBSXVLvv8JHoBbuO3ib07zbVzQ9jibDEDNYmibfN08nWBEfNcxJzCsmpLy5Tzud9fbedBCnHYoqNkWVrKw/640?wx_fmt=png&amp;from=appmsg"><figcaption>图99</figcaption></figure></section><p><br></p><p>谢谢观看</p><p><br></p><p><img data-galleryid="" data-imgfileid="100002282" data-ratio="0.675" data-s="300,640" data-src="https://mmbiz.qpic.cn/sz_mmbiz_jpg/A8HZFc8vBSXVLvv8JHoBbuO3ib07zbVzQF60XnZjRCTOHBPIhBHaRbLmqCAQXs4TVR2dMkRj2Zia12KgTI4fGoSw/640?wx_fmt=jpeg&amp;from=appmsg" data-type="jpeg" data-w="1600" src="https://mmbiz.qpic.cn/sz_mmbiz_jpg/A8HZFc8vBSXVLvv8JHoBbuO3ib07zbVzQF60XnZjRCTOHBPIhBHaRbLmqCAQXs4TVR2dMkRj2Zia12KgTI4fGoSw/640?wx_fmt=jpeg&amp;from=appmsg"></p><p><mp-style-type data-value="3"></mp-style-type></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/GmYkdRJKnkneHJzNL-NHkA",target="_blank" rel="noopener noreferrer">原文链接</a>
