---
title: "单细胞数据不同亚群间的拉丝到底是不是双包体？"
date: 2026-01-04T14:04:59Z
draft: ["false"]
tags: [
  "fetched",
  "生信技能树"
]
categories: ["Acdemic"]
---
单细胞数据不同亚群间的拉丝到底是不是双包体？ by 生信技能树
------
<div><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com" data-pm-slice="0 0 []"><blockquote><p><span leaf="">前面在做一个单细胞的数据分析时，发现我出来的umap结果上面有很多的拉丝情况，见稿子：</span></p><p><span leaf=""><a target="_blank" href="https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247547485&amp;idx=1&amp;sn=9401d6e38a917b4b3451cc72857ff009&amp;scene=21#wechat_redirect" textvalue="" linktype="text" data-linktype="2">破防了，你的单细胞UMAP图也是这样吗：整体“毛茸茸”、点群边界不清？</a></span></p><p><span leaf=""><a target="_blank" href="https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247547515&amp;idx=1&amp;sn=67bc091927bac92f20e35a18c5611a01&amp;scene=21#wechat_redirect" textvalue="" linktype="text" data-linktype="2">单细胞数据异质性太强？对每个样本单独设置过滤指标看看</a></span></p></blockquote><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wymGtr3yQ4WpjnEzqWZpDRfmXRXibgIJRenMJwdsy8TlDnHFjtvTFaXiaHpVWfbypnDy8JtaNFoEmAw/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.500965250965251" data-type="png" data-w="1036" data-imgfileid="100063963" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wymGtr3yQ4WpjnEzqWZpDRfmXRXibgIJRenMJwdsy8TlDnHFjtvTFaXiaHpVWfbypnDy8JtaNFoEmAw/640?wx_fmt=png&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">今天看看里面很明显的拉丝会不会是双细胞呢？</span></p><p data-tool="mdnice编辑器"><span leaf="">这里我使用 DoubletFinder 来看看。</span></p><h2 data-cacheurl="" data-remoteid="" data-tool="mdnice编辑器"><span></span><span><span leaf="">DoubletFinder算法</span></span></h2><p data-tool="mdnice编辑器"><span leaf="">单细胞测序期望每个GEM里包含一个细胞或者细胞核，即每个barcode对应一个真实的细胞，但是实际数据中会有两个或多个细胞共用一个barcode的情况，我们称之为doublets。doublets可分为两类：</span></p><ul><li><section><span leaf="">Homotypic doublets 同源双包体: doublets derived from transcriptionally similar cellls</span></section></li><li><section><span leaf="">Heterotypic doublets 异源双包体: doublets derived from transcriptionally distinct cellls</span></section></li></ul><h3 data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><span></span><span><span leaf="">特点：</span></span><span></span></h3><ul><li><section><span leaf="">DoubletFinder 使用基因表达特征预测单细胞数据中的双包体doublets</span></section></li><li><section><span leaf="">DoubletFinder 可以识别转录组不同细胞的双包体doublets</span></section></li><li><section><span leaf="">Doublet 去除可以提高差异表达分析性能</span></section></li><li><section><span leaf="">DoubletFinder 对 bona fide cells with “hybrid”expression profiles不敏感（同源双包体）</span></section></li></ul><h3 data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><span></span><span><span leaf="">输入数据要求：</span></span><span></span></h3><ul><li><section><span leaf="">1.<span textstyle="">数据为单样本，不能为多样本</span>：算法第一步是对单细胞数据生成模拟双包体，</span><strong><span leaf="">如果为多样本，则可能模拟出来自两个样本的双包体，这种情况在实际中是不存在的</span></strong></section></li><li><section><span leaf="">2.数据经过了质控，即去除低质量细胞</span></section></li><li><section><span leaf="">3.数据经过了预处理：i.e., After NormalizeData, FindVariableGenes, ScaleData, RunPCA, and RunTSNE/UMAP have all been performed</span></section></li><li><section><span leaf="">4.数据为Seurat对象</span></section></li></ul><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wymGtr3yQ4WpjnEzqWZpDRfP9icmbgoOT03eRhIuTpPla4O1aInWTLjjHSicaWXmbMD9qEemfibGRaiaA/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.44456521739130433" data-type="png" data-w="920" data-imgfileid="100063961" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wymGtr3yQ4WpjnEzqWZpDRfP9icmbgoOT03eRhIuTpPla4O1aInWTLjjHSicaWXmbMD9qEemfibGRaiaA/640?wx_fmt=png&amp;from=appmsg"></span></figure><h3 data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><span></span><span><span leaf="">双细胞率参考：</span></span><span></span></h3><p data-tool="mdnice编辑器"><span leaf="">随着细胞数增多，双细胞比例会增加。可以参照下表计算自己的数据中双细胞的比例。multiplet_rate &lt;- c('500' = 0.004, '1000' = 0.008, '2000' = 0.016, '3000' = 0.023, '4000' = 0.031, '5000' = 0.039, '6000' = 0.046, '7000' = 0.054, '8000' = 0.061, '9000' = 0.069, '10000' = 0.076)</span></p><p data-tool="mdnice编辑器"><span leaf="">10X genomics单细胞平台的dulblets比率ref：</span><strong><span leaf="">Chromium Next GEM Single Cell 3ʹ Reagent Kits v3.1 User Guide • Rev D</span></strong></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wymGtr3yQ4WpjnEzqWZpDRfcgdVbxCjtiac7Trw4cicQTq3jMreRt33TZbZxHte56EpibibfmoVFFsyiaQ/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.9418132611637348" data-type="png" data-w="739" data-imgfileid="100063962" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wymGtr3yQ4WpjnEzqWZpDRfcgdVbxCjtiac7Trw4cicQTq3jMreRt33TZbZxHte56EpibibfmoVFFsyiaQ/640?wx_fmt=png&amp;from=appmsg"></span></figure><h2 data-cacheurl="" data-remoteid="" data-tool="mdnice编辑器"><span></span><span><span leaf="">代码实现</span></span></h2><p data-tool="mdnice编辑器"><span leaf="">先定义一个函数：</span></p><pre data-tool="mdnice编辑器"><code><span><span leaf=""># 定义主要参数和函数</span></span><span leaf=""><br></span><span leaf="">run_doublet_finder &lt;- </span><span><span leaf="">function</span></span><span leaf="">(sce_object, doublet_rate, pcs = 1:20, sct = FALSE){</span><span leaf=""><br></span><span><span leaf=""><a topic-id="mj1l9m7v-i4f8h3" data-topic="1">#参数扫描</a>，选择最优pK值</span></span><span leaf=""><br></span><span><span leaf=""># sce_object &lt;- sce_list[[1]]</span></span><span leaf=""><br></span><span leaf="">  sce_object &lt;- NormalizeData(sce_object) </span><span leaf=""><br></span><span leaf="">  sce_object &lt;- FindVariableFeatures(sce_object, nfeatures = 2000)</span><span leaf=""><br></span><span leaf="">  sce_object &lt;- ScaleData(sce_object)</span><span leaf=""><br></span><span leaf="">  sce_object &lt;- RunPCA(sce_object)</span><span leaf=""><br></span><span leaf="">  sce_object &lt;- FindNeighbors(sce_object, dims = 1:20)</span><span leaf=""><br></span><span leaf="">  sce_object &lt;- FindClusters(sce_object, resolution = 0.5)</span><span leaf=""><br></span><span leaf="">  sce_object &lt;- RunUMAP(sce,  dims = 1:20, reduction = </span><span><span leaf="">"pca"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf="">## pK Identification (no ground-truth)</span></span><span leaf=""><br></span><span leaf="">  sweep.res &lt;- paramSweep(sce_object, PCs = 1:20, sct = F, num.cores = 20)</span><span leaf=""><br></span><span leaf="">  sweep.stats &lt;- summarizeSweep(sweep.res, GT = FALSE)</span><span leaf=""><br></span><span leaf="">  bcmvn &lt;- find.pK(sweep.stats)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf="">## pK_select based bcmvn</span></span><span leaf=""><br></span><span><span leaf=""># p &lt;- ggplot(bcmvn,aes(pK, BCmetric, group=1)) + geom_point() + geom_line() + </span></span><span leaf=""><br></span><span><span leaf="">#   geom_vline(xintercept = which.max(bcmvn$BCmetric), linewidth=1, col="red") + </span></span><span leaf=""><br></span><span><span leaf="">#   theme_bw() +</span></span><span leaf=""><br></span><span><span leaf="">#   mytheme</span></span><span leaf=""><br></span><span><span leaf=""># p</span></span><span leaf=""><br></span><span><span leaf=""># This defines the PC neighborhood size used to compute pANN, expressed as a proportion of the merged real-artificial data. </span></span><span leaf=""><br></span><span><span leaf=""># No default is set, as pK should be adjusted for each scRNA-seq dataset. </span></span><span leaf=""><br></span><span><span leaf=""># Optimal pK values should be estimated using the strategy described below.</span></span><span leaf=""><br></span><span leaf="">  pK &lt;- bcmvn</span><span><span leaf="">$pK</span></span><span leaf="">[which.max(bcmvn</span><span><span leaf="">$BCmetric</span></span><span leaf="">)] %&gt;% as.character() %&gt;% as.numeric()</span><span leaf=""><br></span><span leaf="">  pK</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># This defines the number of generated artificial doublets, expressed as a proportion of the merged real-artificial data. </span></span><span leaf=""><br></span><span><span leaf=""># Default is set to 25%, based on observation that DoubletFinder performance is largely pN-invariant </span></span><span leaf=""><br></span><span><span leaf=""># (see McGinnis, Murrow and Gartner 2019, Cell Systems)</span></span><span leaf=""><br></span><span><span leaf=""># pN &lt;- opt$pN</span></span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 计算 homotypicdoublets 和调整的 doublet 数目</span></span><span leaf=""><br></span><span><span leaf=""># This defines the pANN threshold used to make final doublet/singlet predictions. </span></span><span leaf=""><br></span><span><span leaf=""># This value can best be estimated from cell loading densities into the 10X/Drop-Seq device, </span></span><span leaf=""><br></span><span><span leaf=""># and adjusted according to the estimated proportion of homotypic doublets.</span></span><span leaf=""><br></span><span leaf="">  homotypic.prop &lt;- modelHomotypic(sce_object</span><span><span leaf="">$seurat_clusters</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">  nExp_poi &lt;- round(doublet_rate * ncol(sce_object))</span><span leaf=""><br></span><span leaf="">  nExp_poi.adj &lt;- round(nExp_poi * (1 - homotypic.prop))</span><span leaf=""><br></span><span leaf="">  nExp_poi.adj   </span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># print the paratmer</span></span><span leaf=""><br></span><span><span leaf="">print</span></span><span leaf="">(paste0(</span><span><span leaf="">"pN_pK_nExp_poi.adj: "</span></span><span leaf="">, paste0(</span><span><span leaf="">"0.25_"</span></span><span leaf="">,pK,</span><span><span leaf="">"_"</span></span><span leaf="">,nExp_poi.adj )))</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf="">## 鉴定doublets</span></span><span leaf=""><br></span><span leaf="">  sce_object &lt;- doubletFinder(sce_object, PCs = 1:20, pN = 0.25, pK = pK, nExp = nExp_poi.adj, reuse.pANN = FALSE, sct = F)</span><span leaf=""><br></span><span leaf="">  head(sce_object)</span><span leaf=""><br></span><span leaf="">  select &lt;- paste0(</span><span><span leaf="">'DF.classifications_'</span></span><span leaf="">, 0.25, </span><span><span leaf="">'_'</span></span><span leaf="">, pK, </span><span><span leaf="">'_'</span></span><span leaf="">, nExp_poi.adj)</span><span leaf=""><br></span><span><span leaf="">print</span></span><span leaf="">(select)</span><span leaf=""><br></span><span leaf="">  sce_object</span><span><span leaf="">$ls_doublet</span></span><span leaf=""> &lt;- sce_object[[select]]</span><span leaf=""><br></span><span leaf="">  DimPlot(sce_object,reduction = </span><span><span leaf="">"umap"</span></span><span leaf="">, group.by = select)</span><span leaf=""><br></span><span><span leaf=""># DimPlot(sce_object,reduction = "umap", split.by  = select)</span></span><span leaf=""><br></span><span><span leaf="">return</span></span><span leaf="">(sce_object)</span><span leaf=""><br></span><span leaf="">}</span><span leaf=""><br></span></code></pre><h3 data-tool="mdnice编辑器"><span data-cacheurl="" data-remoteid=""></span><span></span><span><span leaf="">运行：</span></span><span></span></h3><pre data-tool="mdnice编辑器"><code><span leaf="">library(tidyverse)</span><span leaf=""><br></span><span leaf="">library(Seurat)</span><span leaf=""><br></span><span leaf="">library(patchwork)</span><span leaf=""><br></span><span leaf="">library(DoubletFinder)</span><span leaf=""><br></span><span leaf="">packageVersion(</span><span><span leaf="">"DoubletFinder"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf="">## 运行 DoubletFinder分析</span></span><span leaf=""><br></span><span><span leaf="">## load Seurat object</span></span><span leaf=""><br></span><span leaf="">seurat_obj &lt;- load(</span><span><span leaf="">"2-harmony/sce.all_int_merge.RData"</span></span><span leaf="">) </span><span><span leaf=""><a topic-id="mj1l9m7v-1wgaxd" data-topic="1">#这里为qc后的数据</a></span></span><span leaf=""><br></span><span leaf="">seurat_obj &lt;- sce.all.filt</span><span leaf=""><br></span><span><span leaf="">## 将seurat拆分成单个样本数据----</span></span><span leaf=""><br></span><span leaf="">head(seurat_obj@meta.data)</span><span leaf=""><br></span><span leaf="">sce_list &lt;- SplitObject(seurat_obj , split.by = </span><span><span leaf="">"orig.ident"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf="">sce_list</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># check每个样本的细胞数</span></span><span leaf=""><br></span><span leaf="">sce_list</span><span leaf=""><br></span><span leaf="">unlist(lapply(sce_list, ncol))</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># 单独运行一个看看那</span></span><span leaf=""><br></span><span leaf="">sce_list[[</span><span><span leaf="">"GSM7937475_CNCL-25-1-clean"</span></span><span leaf="">]] &lt;- run_doublet_finder(sce_list[[</span><span><span leaf="">"GSM7937475_CNCL-25-1-clean"</span></span><span leaf="">]], doublet_rate = 0.15) </span><span leaf=""><br></span><span leaf="">DimPlot(sce_list[[</span><span><span leaf="">"GSM7937475_CNCL-25-1-clean"</span></span><span leaf="">]],reduction = </span><span><span leaf="">"umap"</span></span><span leaf="">, group.by = </span><span><span leaf="">"ls_doublet"</span></span><span leaf="">)</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf=""># doublet_rate中的值需要根据 由于每个样本的细胞数都不一样，所以每次都需要重新选择样本并且设定doublet比例</span></span><span leaf=""><br></span><span leaf="">multiplet_rate &lt;- c(</span><span><span leaf="">'500'</span></span><span leaf=""> = 0.004, </span><span leaf=""><br></span><span leaf="">                    </span><span><span leaf="">'1000'</span></span><span leaf=""> = 0.008, </span><span leaf=""><br></span><span leaf="">                    </span><span><span leaf="">'2000'</span></span><span leaf=""> = 0.016, </span><span leaf=""><br></span><span leaf="">                    </span><span><span leaf="">'3000'</span></span><span leaf=""> = 0.023, </span><span leaf=""><br></span><span leaf="">                    </span><span><span leaf="">'4000'</span></span><span leaf=""> = 0.031, </span><span leaf=""><br></span><span leaf="">                    </span><span><span leaf="">'5000'</span></span><span leaf=""> = 0.039, </span><span leaf=""><br></span><span leaf="">                    </span><span><span leaf="">'6000'</span></span><span leaf=""> = 0.046, </span><span leaf=""><br></span><span leaf="">                    </span><span><span leaf="">'7000'</span></span><span leaf=""> = 0.054, </span><span leaf=""><br></span><span leaf="">                    </span><span><span leaf="">'8000'</span></span><span leaf=""> = 0.061, </span><span leaf=""><br></span><span leaf="">                    </span><span><span leaf="">'9000'</span></span><span leaf=""> = 0.069, </span><span leaf=""><br></span><span leaf="">                    </span><span><span leaf="">'10000'</span></span><span leaf=""> = 0.076</span><span leaf=""><br></span><span leaf="">                    )</span><span leaf=""><br></span><span leaf=""><br></span><span leaf="">rate &lt;- c(0.15,   </span><span><span leaf=""># GSM7937475_CNCL-25-1-clean  19302</span></span><span leaf=""><br></span><span leaf="">          0.021,  </span><span><span leaf=""># GSM7937476_CNCL-25-2-clean  2810</span></span><span leaf=""><br></span><span leaf="">          0.061,  </span><span><span leaf=""># GSM7937477_CNCL-34-1-clean  8338</span></span><span leaf=""><br></span><span leaf="">          0.012,  </span><span><span leaf=""># GSM7937478_CNCL-34-2-clean  1555</span></span><span leaf=""><br></span><span leaf="">          0.077,  </span><span><span leaf=""># GSM7937479_CNCL-36-1-clean  11527</span></span><span leaf=""><br></span><span leaf="">          0.023,  </span><span><span leaf=""># GSM7937480_CNCL-36-2-clean  2478</span></span><span leaf=""><br></span><span leaf="">          0.050,  </span><span><span leaf=""># GSM7937481_CNCL-37-1-clean  6489</span></span><span leaf=""><br></span><span leaf="">          0.013,  </span><span><span leaf=""># GSM7937482_CNCL-37-2-clean  1615</span></span><span leaf=""><br></span><span leaf="">          0.046,  </span><span><span leaf=""># GSM7937483_CNCL-43-1        6080</span></span><span leaf=""><br></span><span leaf="">          0.022,  </span><span><span leaf=""># GSM7937484_CNCL-43-2        2835</span></span><span leaf=""><br></span><span leaf="">          0.031,  </span><span><span leaf=""># GSM7937485_CNCL-48-1        4063</span></span><span leaf=""><br></span><span leaf="">          0.039,  </span><span><span leaf=""># GSM7937486_CNCL-48-2        5114</span></span><span leaf=""><br></span><span leaf="">          0.051,  </span><span><span leaf=""># GSM7937487_CNCL-55-1-clean  6569</span></span><span leaf=""><br></span><span leaf="">          0.025,  </span><span><span leaf=""># GSM7937488_CNCL-55-2-clean  3509</span></span><span leaf=""><br></span><span leaf="">          0.065,  </span><span><span leaf=""># GSM7937489_CNCL-56-1-clean  8680</span></span><span leaf=""><br></span><span leaf="">          0.042,  </span><span><span leaf=""># GSM7937490_CNCL-56-2-clean  5653</span></span><span leaf=""><br></span><span leaf="">          0.047,  </span><span><span leaf=""># GSM7937491_CNCL-59-1        6254</span></span><span leaf=""><br></span><span leaf="">          0.023,  </span><span><span leaf=""># GSM7937492_CNCL-59-2        3181</span></span><span leaf=""><br></span><span leaf="">          0.075,  </span><span><span leaf=""># GSM7937493_CNCL-60-1-clean  9873</span></span><span leaf=""><br></span><span leaf="">          0.031,  </span><span><span leaf=""># GSM7937494_CNCL-60-2-clean  4025</span></span><span leaf=""><br></span><span leaf="">          0.004,  </span><span><span leaf=""># GSM7937495_CNCL-61-1        475</span></span><span leaf=""><br></span><span leaf="">          0.038)  </span><span><span leaf=""># GSM7937496_CNCL-61-2       4777</span></span><span leaf=""><br></span><span leaf="">names(rate) &lt;- names(sce_list)</span><span leaf=""><br></span><span leaf="">rate</span><span leaf=""><br></span><span leaf=""><br></span><span><span leaf="">for</span></span><span leaf="">(i </span><span><span leaf="">in</span></span><span leaf=""> 1:length(sce_list)) {</span><span leaf=""><br></span><span><span leaf="">print</span></span><span leaf="">(i)</span><span leaf=""><br></span><span leaf="">  sce_list[[i]] &lt;- run_doublet_finder(sce_list[[i]], doublet_rate = rate[i]) </span><span leaf=""><br></span><span leaf="">}</span><span leaf=""><br></span><span leaf=""><br></span></code></pre><p data-tool="mdnice编辑器"><span leaf="">我这里只跑完了一个样本：GSM7937475_CNCL-25-1-clean，看看如下：</span></p><figure data-tool="mdnice编辑器"><span leaf=""><img data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wymGtr3yQ4WpjnEzqWZpDRf2AQkUkicjjJ2YsMV9xlSzUicMdrZVX8CfR6iaAyZ9Ribn4KiaMD3vR8EBPg/640?wx_fmt=png&amp;from=appmsg" data-ratio="0.831275720164609" data-type="png" data-w="972" data-imgfileid="100063964" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wymGtr3yQ4WpjnEzqWZpDRf2AQkUkicjjJ2YsMV9xlSzUicMdrZVX8CfR6iaAyZ9Ribn4KiaMD3vR8EBPg/640?wx_fmt=png&amp;from=appmsg"></span></figure><p data-tool="mdnice编辑器"><span leaf="">最后的结果：将上面的list对象merge到一起，然后看降维聚类分群的结果，</span><strong><span leaf="">拉丝的部分有很大的比例能判断为双细胞，快去做做看着~</span></strong></p><p data-tool="mdnice编辑器"><span leaf="">今天分享到这~</span></p></section><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com" data-pm-slice="0 0 []"><p data-tool="mdnice编辑器"><span leaf=""><span textstyle="">转发：</span></span></p></section><ul><li><section><span leaf=""><a target="_blank" href="https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247547012&amp;idx=1&amp;sn=f55923d9a6d9e04c3e923c2a3cae6e56&amp;scene=21#wechat_redirect" textvalue="生信入门&amp;数据挖掘线上直播课12月班" data-itemshowtype="0" linktype="text" data-linktype="2">生信入门&amp;数据挖掘线上直播课12月班</a><span textstyle="">，你的生物信息学入门课</span></span></section></li><li><section><span leaf=""><a target="_blank" href="https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247525079&amp;idx=1&amp;sn=0b997af16a58195b4192691373048fd5&amp;scene=21#wechat_redirect" textvalue="时隔5年，我们的生信技能树VIP学徒继续招生啦" data-itemshowtype="0" linktype="text" data-linktype="2"><span textstyle="">时隔5年，我们的生信技能树VIP学徒继续招生啦</span></a></span></section></li><li><section><span leaf=""><a target="_blank" href="https://mp.weixin.qq.com/s?__biz=MzUzMTEwODk0Ng==&amp;mid=2247530048&amp;idx=1&amp;sn=28aa7bbd5e00521f79e074496a5f5d66&amp;scene=21#wechat_redirect" textvalue="满足你生信分析计算需求的低价解决方案" data-itemshowtype="0" linktype="text" data-linktype="2">满足你生信分析计算需求的低价解决方案</a></span></section></li><li><section><span leaf=""><a target="_blank" href="https://mp.weixin.qq.com/mp/appmsgalbum?__biz=MzAxMDkxODM1Ng==&amp;action=getalbum&amp;album_id=1679199708449144836&amp;scene=173&amp;subscene=207&amp;sessionid=1745492310&amp;enterid=1745492314&amp;from_msgid=2247541298&amp;from_itemidx=1&amp;count=3&amp;nolastread=1#wechat_redirect" textvalue="生信故事会" data-itemshowtype="0" linktype="text" data-linktype="2">生信故事会</a>，来看看他们的生信入门故事</span></section></li><li><section><span leaf=""><a target="_blank" href="https://mp.weixin.qq.com/mp/appmsgalbum?__biz=MzAxMDkxODM1Ng==&amp;action=getalbum&amp;album_id=3690970204957147140&amp;scene=173&amp;subscene=207&amp;sessionid=1745066271&amp;enterid=1745066274&amp;from_msgid=2247540702&amp;from_itemidx=1&amp;count=3&amp;nolastread=1#wechat_redirect" textvalue="生信马拉松答疑专辑" data-itemshowtype="0" linktype="text" data-linktype="2">生信马拉松答疑专辑</a>，获取你的生信专属答疑</span></section></li></ul><section><span leaf=""><br></span></section><p><mp-style-type data-value="3"></mp-style-type></p></div>  
<hr>
<a href="https://mp.weixin.qq.com/s/nIeCkDHFMt8qifrooGlkZQ",target="_blank" rel="noopener noreferrer">原文链接</a>
